// echarts@6.0.0/lib/chart/lines/LinesView.js
import { __extends as c } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import u from "../helper/LineDraw.js";
import v from "../helper/EffectLine.js";
import g from "../helper/Line.js";
import m from "../helper/Polyline.js";
import _ from "../helper/EffectPolyline.js";
import y from "../helper/LargeLineDraw.js";
import w from "./linesLayout.js";
import { createClipPath as d } from "../helper/createClipPathFromCoordSys.js";
import D from "../../view/Chart.js";
var S = (
  /** @class */
  (function(s) {
    c(a, s);
    function a() {
      var e = s !== null && s.apply(this, arguments) || this;
      return e.type = a.type, e;
    }
    return a.prototype.render = function(e, t, r) {
      var i = e.getData(), n = this._updateLineDraw(i, e), o = e.get("zlevel"), l = e.get(["effect", "trailLength"]), p = r.getZr(), f = p.painter.getType() === "svg";
      f || p.painter.getLayer(o).clear(!0), this._lastZlevel != null && !f && p.configLayer(this._lastZlevel, {
        motionBlur: !1
      }), this._showEffect(e) && l > 0 && (f ? process.env.NODE_ENV !== "production" && console.warn("SVG render mode doesn't support lines with trail effect") : p.configLayer(o, {
        motionBlur: !0,
        lastFrameAlpha: Math.max(Math.min(l / 10 + 0.9, 1), 0)
      })), n.updateData(i);
      var h = e.get("clip", !0) && d(e.coordinateSystem, !1, e);
      h ? this.group.setClipPath(h) : this.group.removeClipPath(), this._lastZlevel = o, this._finished = !0;
    }, a.prototype.incrementalPrepareRender = function(e, t, r) {
      var i = e.getData(), n = this._updateLineDraw(i, e);
      n.incrementalPrepareUpdate(i), this._clearLayer(r), this._finished = !1;
    }, a.prototype.incrementalRender = function(e, t, r) {
      this._lineDraw.incrementalUpdate(e, t.getData()), this._finished = e.end === t.getData().count();
    }, a.prototype.eachRendered = function(e) {
      this._lineDraw && this._lineDraw.eachRendered(e);
    }, a.prototype.updateTransform = function(e, t, r) {
      var i = e.getData(), n = e.pipelineContext;
      if (!this._finished || n.large || n.progressiveRender)
        return {
          update: !0
        };
      var o = w.reset(e, t, r);
      o.progress && o.progress({
        start: 0,
        end: i.count(),
        count: i.count()
      }, i), this._lineDraw.updateLayout(), this._clearLayer(r);
    }, a.prototype._updateLineDraw = function(e, t) {
      var r = this._lineDraw, i = this._showEffect(t), n = !!t.get("polyline"), o = t.pipelineContext, l = o.large;
      return process.env.NODE_ENV !== "production" && i && l && console.warn("Large lines not support effect"), (!r || i !== this._hasEffet || n !== this._isPolyline || l !== this._isLargeDraw) && (r && r.remove(), r = this._lineDraw = l ? new y() : new u(n ? i ? _ : m : i ? v : g), this._hasEffet = i, this._isPolyline = n, this._isLargeDraw = l), this.group.add(r.group), r;
    }, a.prototype._showEffect = function(e) {
      return !!e.get(["effect", "show"]);
    }, a.prototype._clearLayer = function(e) {
      var t = e.getZr(), r = t.painter.getType() === "svg";
      !r && this._lastZlevel != null && t.painter.getLayer(this._lastZlevel).clear(!0);
    }, a.prototype.remove = function(e, t) {
      this._lineDraw && this._lineDraw.remove(), this._lineDraw = null, this._clearLayer(t);
    }, a.prototype.dispose = function(e, t) {
      this.remove(e, t);
    }, a.type = "lines", a;
  })(D)
);
export {
  S as default
};
