// echarts@6.0.0/lib/chart/lines/LinesSeries.js
import { __extends as u } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import m from "../../model/Series.js";
import C from "../../data/SeriesData.js";
import { concatArray as c, isNumber as _, map as A, mergeAll as w } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import N from "../../core/CoordinateSystem.js";
import { createTooltipMarkup as O } from "../../component/tooltip/tooltipMarkup.js";
var D = typeof Uint32Array > "u" ? Array : Uint32Array, M = typeof Float64Array > "u" ? Array : Float64Array;
function y(f) {
  var o = f.data;
  o && o[0] && o[0][0] && o[0][0].coord && (process.env.NODE_ENV !== "production" && console.warn("Lines data configuration has been changed to { coords:[[1,2],[2,3]] }"), f.data = A(o, function(t) {
    var r = [t[0].coord, t[1].coord], a = {
      coords: r
    };
    return t[0].name && (a.fromName = t[0].name), t[1].name && (a.toName = t[1].name), w([a, t[0], t[1]]);
  }));
}
var b = (
  /** @class */
  (function(f) {
    u(o, f);
    function o() {
      var t = f !== null && f.apply(this, arguments) || this;
      return t.type = o.type, t.visualStyleAccessPath = "lineStyle", t.visualDrawType = "stroke", t;
    }
    return o.prototype.init = function(t) {
      t.data = t.data || [], y(t);
      var r = this._processFlatCoordsArray(t.data);
      this._flatCoords = r.flatCoords, this._flatCoordsOffset = r.flatCoordsOffset, r.flatCoords && (t.data = new Float32Array(r.count)), f.prototype.init.apply(this, arguments);
    }, o.prototype.mergeOption = function(t) {
      if (y(t), t.data) {
        var r = this._processFlatCoordsArray(t.data);
        this._flatCoords = r.flatCoords, this._flatCoordsOffset = r.flatCoordsOffset, r.flatCoords && (t.data = new Float32Array(r.count));
      }
      f.prototype.mergeOption.apply(this, arguments);
    }, o.prototype.appendData = function(t) {
      var r = this._processFlatCoordsArray(t.data);
      r.flatCoords && (this._flatCoords ? (this._flatCoords = c(this._flatCoords, r.flatCoords), this._flatCoordsOffset = c(this._flatCoordsOffset, r.flatCoordsOffset)) : (this._flatCoords = r.flatCoords, this._flatCoordsOffset = r.flatCoordsOffset), t.data = new Float32Array(r.count)), this.getRawData().appendData(t.data);
    }, o.prototype._getCoordsFromItemModel = function(t) {
      var r = this.getData().getItemModel(t), a = r.option instanceof Array ? r.option : r.getShallow("coords");
      if (process.env.NODE_ENV !== "production" && !(a instanceof Array && a.length > 0 && a[0] instanceof Array))
        throw new Error("Invalid coords " + JSON.stringify(a) + ". Lines must have 2d coords array in data item.");
      return a;
    }, o.prototype.getLineCoordsCount = function(t) {
      return this._flatCoordsOffset ? this._flatCoordsOffset[t * 2 + 1] : this._getCoordsFromItemModel(t).length;
    }, o.prototype.getLineCoords = function(t, r) {
      if (this._flatCoordsOffset) {
        for (var a = this._flatCoordsOffset[t * 2], n = this._flatCoordsOffset[t * 2 + 1], e = 0; e < n; e++)
          r[e] = r[e] || [], r[e][0] = this._flatCoords[a + e * 2], r[e][1] = this._flatCoords[a + e * 2 + 1];
        return n;
      } else {
        for (var i = this._getCoordsFromItemModel(t), e = 0; e < i.length; e++)
          r[e] = r[e] || [], r[e][0] = i[e][0], r[e][1] = i[e][1];
        return i.length;
      }
    }, o.prototype._processFlatCoordsArray = function(t) {
      var r = 0;
      if (this._flatCoords && (r = this._flatCoords.length), _(t[0])) {
        for (var a = t.length, n = new D(a), e = new M(a), i = 0, l = 0, d = 0, s = 0; s < a; ) {
          d++;
          var h = t[s++];
          n[l++] = i + r, n[l++] = h;
          for (var p = 0; p < h; p++) {
            var g = t[s++], v = t[s++];
            if (e[i++] = g, e[i++] = v, s > a && process.env.NODE_ENV !== "production")
              throw new Error("Invalid data format.");
          }
        }
        return {
          flatCoordsOffset: new Uint32Array(n.buffer, 0, l),
          flatCoords: e,
          count: d
        };
      }
      return {
        flatCoordsOffset: null,
        flatCoords: null,
        count: t.length
      };
    }, o.prototype.getInitialData = function(t, r) {
      if (process.env.NODE_ENV !== "production") {
        var a = N.get(t.coordinateSystem);
        if (!a)
          throw new Error("Unknown coordinate system " + t.coordinateSystem);
      }
      var n = new C(["value"], this);
      return n.hasItemOption = !1, n.initData(t.data, [], function(e, i, l, d) {
        if (e instanceof Array)
          return NaN;
        n.hasItemOption = !0;
        var s = e.value;
        if (s != null)
          return s instanceof Array ? s[d] : s;
      }), n;
    }, o.prototype.formatTooltip = function(t, r, a) {
      var n = this.getData(), e = n.getItemModel(t), i = e.get("name");
      if (i)
        return i;
      var l = e.get("fromName"), d = e.get("toName"), s = [];
      return l != null && s.push(l), d != null && s.push(d), O("nameValue", {
        name: s.join(" > ")
      });
    }, o.prototype.preventIncremental = function() {
      return !!this.get(["effect", "show"]);
    }, o.prototype.getProgressive = function() {
      var t = this.option.progressive;
      return t ?? (this.option.large ? 1e4 : this.get("progressive"));
    }, o.prototype.getProgressiveThreshold = function() {
      var t = this.option.progressiveThreshold;
      return t ?? (this.option.large ? 2e4 : this.get("progressiveThreshold"));
    }, o.prototype.getZLevelKey = function() {
      var t = this.getModel("effect"), r = t.get("trailLength");
      return this.getData().count() > this.getProgressiveThreshold() ? this.id : t.get("show") && r > 0 ? r + "" : "";
    }, o.type = "series.lines", o.dependencies = ["grid", "polar", "geo", "calendar"], o.defaultOption = {
      coordinateSystem: "geo",
      // zlevel: 0,
      z: 2,
      legendHoverLink: !0,
      // Cartesian coordinate system
      xAxisIndex: 0,
      yAxisIndex: 0,
      symbol: ["none", "none"],
      symbolSize: [10, 10],
      // Geo coordinate system
      geoIndex: 0,
      effect: {
        show: !1,
        period: 4,
        constantSpeed: 0,
        symbol: "circle",
        symbolSize: 3,
        loop: !0,
        trailLength: 0.2
      },
      large: !1,
      // Available when large is true
      largeThreshold: 2e3,
      polyline: !1,
      clip: !0,
      label: {
        show: !1,
        position: "end"
        // distance: 5,
        // formatter: 标签文本格式器，同Tooltip.formatter，不支持异步回调
      },
      lineStyle: {
        opacity: 0.5
      }
    }, o;
  })(m)
);
export {
  b as default
};
