// @radix-ui/react-popover@1.1.5/dist/index.js
import * as c from "react";
import { composeEventHandlers as P } from "../../../../../@radix-ui_primitive@1.1.1/node_modules/@radix-ui/primitive/dist/index.js";
import { useComposedRefs as O } from "../../../../../@radix-ui_react-compose-refs@1.1.1_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-compose-refs/dist/index.js";
import { createContextScope as G } from "../../../../../@radix-ui_react-context@1.1.1_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-context/dist/index.js";
import { DismissableLayer as H } from "../../../../../@radix-ui_react-dismissable-layer@1.1.4_@types_react-dom@18.3.1_@types_react@18.3.18_react-do_4tz2ji7xudyyxfit74kdiyd35a/node_modules/@radix-ui/react-dismissable-layer/dist/index.js";
import { useFocusGuards as K } from "../../../../../@radix-ui_react-focus-guards@1.1.1_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-focus-guards/dist/index.js";
import { FocusScope as $ } from "../../../../../@radix-ui_react-focus-scope@1.1.1_@types_react-dom@18.3.1_@types_react@18.3.18_react-dom@18.3_5sl5w2z6hfwxfunfgzrt77oufa/node_modules/@radix-ui/react-focus-scope/dist/index.js";
import { useId as j } from "../../../../../@radix-ui_react-id@1.1.0_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-id/dist/index.js";
import { createPopperScope as _, Root as U, Anchor as x, Content as V, Arrow as W } from "../../../../../@radix-ui_react-popper@1.2.1_@types_react-dom@18.3.1_@types_react@18.3.18_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/@radix-ui/react-popper/dist/index.js";
import { Portal as Z } from "../../../../../@radix-ui_react-portal@1.1.3_@types_react-dom@18.3.1_@types_react@18.3.18_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/@radix-ui/react-portal/dist/index.js";
import { Presence as w } from "../../../../../@radix-ui_react-presence@1.1.2_@types_react-dom@18.3.1_@types_react@18.3.18_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/@radix-ui/react-presence/dist/index.js";
import { Primitive as E } from "../../../../../@radix-ui_react-primitive@2.0.1_@types_react-dom@18.3.1_@types_react@18.3.18_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/@radix-ui/react-primitive/dist/index.js";
import { Slot as q } from "../../../../../@radix-ui_react-slot@1.1.1_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-slot/dist/index.js";
import { useControllableState as z } from "../../../../../@radix-ui_react-use-controllable-state@1.1.0_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-use-controllable-state/dist/index.js";
import { hideOthers as B } from "../../../../../aria-hidden@1.2.6/node_modules/aria-hidden/dist/es2015/index.js";
import J from "../../../../../react-remove-scroll@2.7.1_@types_react@18.3.18_react@18.3.1/node_modules/react-remove-scroll/dist/es2015/Combination.js";
import { jsx as p } from "react/jsx-runtime";
var R = "Popover", [b] = G(R, [
  _
]), m = _(), [Q, u] = b(R), F = (o) => {
  const {
    __scopePopover: n,
    children: t,
    open: a,
    defaultOpen: e,
    onOpenChange: r,
    modal: s = !1
  } = o, i = m(n), d = c.useRef(null), [v, g] = c.useState(!1), [h = !1, l] = z({
    prop: a,
    defaultProp: e,
    onChange: r
  });
  return /* @__PURE__ */ p(U, { ...i, children: /* @__PURE__ */ p(
    Q,
    {
      scope: n,
      contentId: j(),
      triggerRef: d,
      open: h,
      onOpenChange: l,
      onOpenToggle: c.useCallback(() => l((C) => !C), [l]),
      hasCustomAnchor: v,
      onCustomAnchorAdd: c.useCallback(() => g(!0), []),
      onCustomAnchorRemove: c.useCallback(() => g(!1), []),
      modal: s,
      children: t
    }
  ) });
};
F.displayName = R;
var N = "PopoverAnchor", y = c.forwardRef(
  (o, n) => {
    const { __scopePopover: t, ...a } = o, e = u(N, t), r = m(t), { onCustomAnchorAdd: s, onCustomAnchorRemove: i } = e;
    return c.useEffect(() => (s(), () => i()), [s, i]), /* @__PURE__ */ p(x, { ...r, ...a, ref: n });
  }
);
y.displayName = N;
var S = "PopoverTrigger", D = c.forwardRef(
  (o, n) => {
    const { __scopePopover: t, ...a } = o, e = u(S, t), r = m(t), s = O(n, e.triggerRef), i = /* @__PURE__ */ p(
      E.button,
      {
        type: "button",
        "aria-haspopup": "dialog",
        "aria-expanded": e.open,
        "aria-controls": e.contentId,
        "data-state": L(e.open),
        ...a,
        ref: s,
        onClick: P(o.onClick, e.onOpenToggle)
      }
    );
    return e.hasCustomAnchor ? i : /* @__PURE__ */ p(x, { asChild: !0, ...r, children: i });
  }
);
D.displayName = S;
var A = "PopoverPortal", [X, Y] = b(A, {
  forceMount: void 0
}), M = (o) => {
  const { __scopePopover: n, forceMount: t, children: a, container: e } = o, r = u(A, n);
  return /* @__PURE__ */ p(X, { scope: n, forceMount: t, children: /* @__PURE__ */ p(w, { present: t || r.open, children: /* @__PURE__ */ p(Z, { asChild: !0, container: e, children: a }) }) });
};
M.displayName = A;
var f = "PopoverContent", k = c.forwardRef(
  (o, n) => {
    const t = Y(f, o.__scopePopover), { forceMount: a = t.forceMount, ...e } = o, r = u(f, o.__scopePopover);
    return /* @__PURE__ */ p(w, { present: a || r.open, children: r.modal ? /* @__PURE__ */ p(oo, { ...e, ref: n }) : /* @__PURE__ */ p(eo, { ...e, ref: n }) });
  }
);
k.displayName = f;
var oo = c.forwardRef(
  (o, n) => {
    const t = u(f, o.__scopePopover), a = c.useRef(null), e = O(n, a), r = c.useRef(!1);
    return c.useEffect(() => {
      const s = a.current;
      if (s) return B(s);
    }, []), /* @__PURE__ */ p(J, { as: q, allowPinchZoom: !0, children: /* @__PURE__ */ p(
      I,
      {
        ...o,
        ref: e,
        trapFocus: t.open,
        disableOutsidePointerEvents: !0,
        onCloseAutoFocus: P(o.onCloseAutoFocus, (s) => {
          s.preventDefault(), r.current || t.triggerRef.current?.focus();
        }),
        onPointerDownOutside: P(
          o.onPointerDownOutside,
          (s) => {
            const i = s.detail.originalEvent, d = i.button === 0 && i.ctrlKey === !0, v = i.button === 2 || d;
            r.current = v;
          },
          { checkForDefaultPrevented: !1 }
        ),
        onFocusOutside: P(
          o.onFocusOutside,
          (s) => s.preventDefault(),
          { checkForDefaultPrevented: !1 }
        )
      }
    ) });
  }
), eo = c.forwardRef(
  (o, n) => {
    const t = u(f, o.__scopePopover), a = c.useRef(!1), e = c.useRef(!1);
    return /* @__PURE__ */ p(
      I,
      {
        ...o,
        ref: n,
        trapFocus: !1,
        disableOutsidePointerEvents: !1,
        onCloseAutoFocus: (r) => {
          o.onCloseAutoFocus?.(r), r.defaultPrevented || (a.current || t.triggerRef.current?.focus(), r.preventDefault()), a.current = !1, e.current = !1;
        },
        onInteractOutside: (r) => {
          o.onInteractOutside?.(r), r.defaultPrevented || (a.current = !0, r.detail.originalEvent.type === "pointerdown" && (e.current = !0));
          const s = r.target;
          t.triggerRef.current?.contains(s) && r.preventDefault(), r.detail.originalEvent.type === "focusin" && e.current && r.preventDefault();
        }
      }
    );
  }
), I = c.forwardRef(
  (o, n) => {
    const {
      __scopePopover: t,
      trapFocus: a,
      onOpenAutoFocus: e,
      onCloseAutoFocus: r,
      disableOutsidePointerEvents: s,
      onEscapeKeyDown: i,
      onPointerDownOutside: d,
      onFocusOutside: v,
      onInteractOutside: g,
      ...h
    } = o, l = u(f, t), C = m(t);
    return K(), /* @__PURE__ */ p(
      $,
      {
        asChild: !0,
        loop: !0,
        trapped: a,
        onMountAutoFocus: e,
        onUnmountAutoFocus: r,
        children: /* @__PURE__ */ p(
          H,
          {
            asChild: !0,
            disableOutsidePointerEvents: s,
            onInteractOutside: g,
            onEscapeKeyDown: i,
            onPointerDownOutside: d,
            onFocusOutside: v,
            onDismiss: () => l.onOpenChange(!1),
            children: /* @__PURE__ */ p(
              V,
              {
                "data-state": L(l.open),
                role: "dialog",
                id: l.contentId,
                ...C,
                ...h,
                ref: n,
                style: {
                  ...h.style,
                  "--radix-popover-content-transform-origin": "var(--radix-popper-transform-origin)",
                  "--radix-popover-content-available-width": "var(--radix-popper-available-width)",
                  "--radix-popover-content-available-height": "var(--radix-popper-available-height)",
                  "--radix-popover-trigger-width": "var(--radix-popper-anchor-width)",
                  "--radix-popover-trigger-height": "var(--radix-popper-anchor-height)"
                }
              }
            )
          }
        )
      }
    );
  }
), T = "PopoverClose", ro = c.forwardRef(
  (o, n) => {
    const { __scopePopover: t, ...a } = o, e = u(T, t);
    return /* @__PURE__ */ p(
      E.button,
      {
        type: "button",
        ...a,
        ref: n,
        onClick: P(o.onClick, () => e.onOpenChange(!1))
      }
    );
  }
);
ro.displayName = T;
var to = "PopoverArrow", no = c.forwardRef(
  (o, n) => {
    const { __scopePopover: t, ...a } = o, e = m(t);
    return /* @__PURE__ */ p(W, { ...e, ...a, ref: n });
  }
);
no.displayName = to;
function L(o) {
  return o ? "open" : "closed";
}
var Oo = F, _o = y, xo = D, wo = M, Eo = k;
export {
  _o as Anchor,
  Eo as Content,
  F as Popover,
  y as PopoverAnchor,
  no as PopoverArrow,
  ro as PopoverClose,
  k as PopoverContent,
  M as PopoverPortal,
  D as PopoverTrigger,
  wo as Portal,
  Oo as Root,
  xo as Trigger
};
