// embla-carousel-wheel-gestures@8.0.1/dist/embla-carousel-wheel-gestures.esm.js
import P from "../../../../wheel-gestures@2.2.48/node_modules/wheel-gestures/dist/wheel-gestures.esm.js";
var T = {
  active: !0,
  breakpoints: {},
  wheelDraggingClass: "is-wheel-dragging",
  forceWheelAxis: void 0,
  target: void 0
};
N.globalOptions = void 0;
var V = process.env.NODE_ENV !== "production";
function N(v) {
  v === void 0 && (v = {});
  var r, f = function() {
  };
  function y(u, w) {
    var M, x, b = w.mergeOptions, G = w.optionsAtMedia, X = b(T, N.globalOptions), Y = b(X, v);
    r = G(Y);
    var o = u.internalEngine(), h = (M = r.target) != null ? M : u.containerNode().parentNode, c = (x = r.forceWheelAxis) != null ? x : o.options.axis, D = P({
      preventWheelAction: c,
      reverseSign: [!0, !0, !1]
    }), _ = D.observe(h), C = D.on("wheel", k), l = !1, i;
    function O(e) {
      try {
        i = new MouseEvent("mousedown", e.event), E(i);
      } catch {
        return V && console.warn("Legacy browser requires events-polyfill (https://github.com/xiel/embla-carousel-wheel-gestures#legacy-browsers)"), f();
      }
      l = !0, S(), r.wheelDraggingClass && h.classList.add(r.wheelDraggingClass);
    }
    function R(e) {
      l = !1, E(A("mouseup", e)), L(), r.wheelDraggingClass && h.classList.remove(r.wheelDraggingClass);
    }
    function S() {
      document.documentElement.addEventListener("mousemove", s, !0), document.documentElement.addEventListener("mouseup", s, !0), document.documentElement.addEventListener("mousedown", s, !0);
    }
    function L() {
      document.documentElement.removeEventListener("mousemove", s, !0), document.documentElement.removeEventListener("mouseup", s, !0), document.documentElement.removeEventListener("mousedown", s, !0);
    }
    function s(e) {
      l && e.isTrusted && e.stopImmediatePropagation();
    }
    function A(e, a) {
      var n, t;
      if (c === o.options.axis) {
        var m = a.axisMovement;
        n = m[0], t = m[1];
      } else {
        var d = a.axisMovement;
        t = d[0], n = d[1];
      }
      if (!o.options.skipSnaps && !o.options.dragFree) {
        var g = o.containerRect.width, p = o.containerRect.height;
        n = n < 0 ? Math.max(n, -g) : Math.min(n, g), t = t < 0 ? Math.max(t, -p) : Math.min(t, p);
      }
      return new MouseEvent(e, {
        clientX: i.clientX + n,
        clientY: i.clientY + t,
        screenX: i.screenX + n,
        screenY: i.screenY + t,
        movementX: n,
        movementY: t,
        button: 0,
        bubbles: !0,
        cancelable: !0,
        composed: !0
      });
    }
    function E(e) {
      u.containerNode().dispatchEvent(e);
    }
    function k(e) {
      var a = e.axisDelta, n = a[0], t = a[1], m = c === "x" ? n : t, d = c === "x" ? t : n, g = e.isMomentum && e.previous && !e.previous.isMomentum, p = e.isEnding && !e.isMomentum || g, I = Math.abs(m) > Math.abs(d);
      I && !l && !e.isMomentum && O(e), l && (p ? R(e) : E(A("mousemove", e)));
    }
    f = function() {
      _(), C(), L();
    };
  }
  var W = {
    name: "wheelGestures",
    options: v,
    init: y,
    destroy: function() {
      return f();
    }
  };
  return W;
}
export {
  N as WheelGesturesPlugin
};
