// echarts@6.0.0/lib/chart/scatter/ScatterView.js
import { __extends as p } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import s from "../helper/SymbolDraw.js";
import u from "../helper/LargeSymbolDraw.js";
import h from "../../layout/points.js";
import l from "../../view/Chart.js";
var D = (
  /** @class */
  (function(n) {
    p(r, n);
    function r() {
      var t = n !== null && n.apply(this, arguments) || this;
      return t.type = r.type, t;
    }
    return r.prototype.render = function(t, e, o) {
      var a = t.getData(), i = this._updateSymbolDraw(a, t);
      i.updateData(a, {
        // TODO
        // If this parameter should be a shape or a bounding volume
        // shape will be more general.
        // But bounding volume like bounding rect will be much faster in the contain calculation
        clipShape: this._getClipShape(t)
      }), this._finished = !0;
    }, r.prototype.incrementalPrepareRender = function(t, e, o) {
      var a = t.getData(), i = this._updateSymbolDraw(a, t);
      i.incrementalPrepareUpdate(a), this._finished = !1;
    }, r.prototype.incrementalRender = function(t, e, o) {
      this._symbolDraw.incrementalUpdate(t, e.getData(), {
        clipShape: this._getClipShape(e)
      }), this._finished = t.end === e.getData().count();
    }, r.prototype.updateTransform = function(t, e, o) {
      var a = t.getData();
      if (this.group.dirty(), !this._finished || a.count() > 1e4)
        return {
          update: !0
        };
      var i = h("").reset(t, e, o);
      i.progress && i.progress({
        start: 0,
        end: a.count(),
        count: a.count()
      }, a), this._symbolDraw.updateLayout(a);
    }, r.prototype.eachRendered = function(t) {
      this._symbolDraw && this._symbolDraw.eachRendered(t);
    }, r.prototype._getClipShape = function(t) {
      if (t.get("clip", !0)) {
        var e = t.coordinateSystem;
        return e && e.getArea && e.getArea(0.1);
      }
    }, r.prototype._updateSymbolDraw = function(t, e) {
      var o = this._symbolDraw, a = e.pipelineContext, i = a.large;
      return (!o || i !== this._isLargeDraw) && (o && o.remove(), o = this._symbolDraw = i ? new u() : new s(), this._isLargeDraw = i, this.group.removeAll()), this.group.add(o.group), o;
    }, r.prototype.remove = function(t, e) {
      this._symbolDraw && this._symbolDraw.remove(!0), this._symbolDraw = null;
    }, r.prototype.dispose = function() {
    }, r.type = "scatter", r;
  })(l)
);
export {
  D as default
};
