// zrender@6.0.0/lib/contain/path.js
import K from "../core/PathProxy.js";
import { containStroke as Q } from "./line.js";
import { containStroke as U } from "./cubic.js";
import { containStroke as V } from "./quadratic.js";
import { containStroke as X } from "./arc.js";
import { quadraticRootAt as Y, quadraticAt as Z, cubicRootAt as i, cubicExtrema as w, quadraticExtremum as F, cubicAt as j } from "../core/curve.js";
import L from "./windingLine.js";
var $ = K.CMD, S = Math.PI * 2, T = 1e-4;
function d(b, e) {
  return Math.abs(b - e) < T;
}
var k = [-1, -1, -1], q = [-1, -1];
function h() {
  var b = q[0];
  q[0] = q[1], q[1] = b;
}
function A(b, e, M, f, u, r, I, o, n, v) {
  if (v > e && v > f && v > r && v > o || v < e && v < f && v < r && v < o)
    return 0;
  var t = i(e, f, r, o, v, k);
  if (t === 0)
    return 0;
  for (var c = 0, m = -1, s = void 0, a = void 0, P = 0; P < t; P++) {
    var R = k[P], p = R === 0 || R === 1 ? 0.5 : 1, D = j(b, M, u, I, R);
    D < n || (m < 0 && (m = w(e, f, r, o, q), q[1] < q[0] && m > 1 && h(), s = j(e, f, r, o, q[0]), m > 1 && (a = j(e, f, r, o, q[1]))), m === 2 ? R < q[0] ? c += s < e ? p : -p : R < q[1] ? c += a < s ? p : -p : c += o < a ? p : -p : R < q[0] ? c += s < e ? p : -p : c += o < s ? p : -p);
  }
  return c;
}
function W(b, e, M, f, u, r, I, o) {
  if (o > e && o > f && o > r || o < e && o < f && o < r)
    return 0;
  var n = Y(e, f, r, o, k);
  if (n === 0)
    return 0;
  var v = F(e, f, r);
  if (v >= 0 && v <= 1) {
    for (var t = 0, c = Z(e, f, r, v), m = 0; m < n; m++) {
      var s = k[m] === 0 || k[m] === 1 ? 0.5 : 1, a = Z(b, M, u, k[m]);
      a < I || (k[m] < v ? t += c < e ? s : -s : t += r < c ? s : -s);
    }
    return t;
  } else {
    var s = k[0] === 0 || k[0] === 1 ? 0.5 : 1, a = Z(b, M, u, k[0]);
    return a < I ? 0 : r < e ? s : -s;
  }
}
function _(b, e, M, f, u, r, I, o) {
  if (o -= e, o > M || o < -M)
    return 0;
  var n = Math.sqrt(M * M - o * o);
  k[0] = -n, k[1] = n;
  var v = Math.abs(f - u);
  if (v < 1e-4)
    return 0;
  if (v >= S - 1e-4) {
    f = 0, u = S;
    var t = r ? 1 : -1;
    return I >= k[0] + b && I <= k[1] + b ? t : 0;
  }
  if (f > u) {
    var c = f;
    f = u, u = c;
  }
  f < 0 && (f += S, u += S);
  for (var m = 0, s = 0; s < 2; s++) {
    var a = k[s];
    if (a + b > I) {
      var P = Math.atan2(o, a), t = r ? 1 : -1;
      P < 0 && (P = S + P), (P >= f && P <= u || P + S >= f && P + S <= u) && (P > Math.PI / 2 && P < Math.PI * 1.5 && (t = -t), m += t);
    }
  }
  return m;
}
function G(b, e, M, f, u) {
  for (var r = b.data, I = b.len(), o = 0, n = 0, v = 0, t = 0, c = 0, m, s, a = 0; a < I; ) {
    var P = r[a++], R = a === 1;
    switch (P === $.M && a > 1 && (M || (o += L(n, v, t, c, f, u))), R && (n = r[a], v = r[a + 1], t = n, c = v), P) {
      case $.M:
        t = r[a++], c = r[a++], n = t, v = c;
        break;
      case $.L:
        if (M) {
          if (Q(n, v, r[a], r[a + 1], e, f, u))
            return !0;
        } else
          o += L(n, v, r[a], r[a + 1], f, u) || 0;
        n = r[a++], v = r[a++];
        break;
      case $.C:
        if (M) {
          if (U(n, v, r[a++], r[a++], r[a++], r[a++], r[a], r[a + 1], e, f, u))
            return !0;
        } else
          o += A(n, v, r[a++], r[a++], r[a++], r[a++], r[a], r[a + 1], f, u) || 0;
        n = r[a++], v = r[a++];
        break;
      case $.Q:
        if (M) {
          if (V(n, v, r[a++], r[a++], r[a], r[a + 1], e, f, u))
            return !0;
        } else
          o += W(n, v, r[a++], r[a++], r[a], r[a + 1], f, u) || 0;
        n = r[a++], v = r[a++];
        break;
      case $.A:
        var p = r[a++], D = r[a++], O = r[a++], E = r[a++], C = r[a++], N = r[a++];
        a += 1;
        var z = !!(1 - r[a++]);
        m = Math.cos(C) * O + p, s = Math.sin(C) * E + D, R ? (t = m, c = s) : o += L(n, v, m, s, f, u);
        var B = (f - p) * E / O + p;
        if (M) {
          if (X(p, D, E, C, C + N, z, e, B, u))
            return !0;
        } else
          o += _(p, D, E, C, C + N, z, B, u);
        n = Math.cos(C + N) * O + p, v = Math.sin(C + N) * E + D;
        break;
      case $.R:
        t = n = r[a++], c = v = r[a++];
        var H = r[a++], J = r[a++];
        if (m = t + H, s = c + J, M) {
          if (Q(t, c, m, c, e, f, u) || Q(m, c, m, s, e, f, u) || Q(m, s, t, s, e, f, u) || Q(t, s, t, c, e, f, u))
            return !0;
        } else
          o += L(m, c, m, s, f, u), o += L(t, s, t, c, f, u);
        break;
      case $.Z:
        if (M) {
          if (Q(n, v, t, c, e, f, u))
            return !0;
        } else
          o += L(n, v, t, c, f, u);
        n = t, v = c;
        break;
    }
  }
  return !M && !d(v, c) && (o += L(n, v, t, c, f, u) || 0), o !== 0;
}
function vr(b, e, M) {
  return G(b, 0, !1, e, M);
}
function or(b, e, M, f) {
  return G(b, e, !0, M, f);
}
export {
  vr as contain,
  or as containStroke
};
