// zrender@6.0.0/lib/contain/text.js
import b from "../core/BoundingRect.js";
import g from "../core/LRU.js";
import { platformApi as m, DEFAULT_FONT as T } from "../core/platform.js";
function C(i) {
  o || (o = new g(100)), i = i || T;
  var e = o.get(i);
  return e || (e = {
    font: i,
    strWidthCache: new g(500),
    asciiWidthMap: null,
    asciiWidthMapTried: !1,
    stWideCharWidth: m.measureText("国", i).width,
    asciiCharWidth: m.measureText("a", i).width
  }, o.put(i, e)), e;
}
var o;
function k(i) {
  if (!(p >= f)) {
    i = i || T;
    for (var e = [], a = +/* @__PURE__ */ new Date(), r = 0; r <= 127; r++)
      e[r] = m.measureText(String.fromCharCode(r), i).width;
    var t = +/* @__PURE__ */ new Date() - a;
    return t > 16 ? p = f : t > 2 && p++, e;
  }
}
var p = 0, f = 5;
function M(i, e) {
  return i.asciiWidthMapTried || (i.asciiWidthMap = k(i.font), i.asciiWidthMapTried = !0), 0 <= e && e <= 127 ? i.asciiWidthMap != null ? i.asciiWidthMap[e] : i.asciiCharWidth : i.stWideCharWidth;
}
function w(i, e) {
  var a = i.strWidthCache, r = a.get(e);
  return r == null && (r = m.measureText(e, i.font).width, a.put(e, r)), r;
}
function W(i, e, a, r) {
  var t = w(C(e), i), u = _(e), s = L(0, t, a), h = x(0, u, r), n = new b(s, h, t, u);
  return n;
}
function R(i, e, a, r) {
  var t = ((i || "") + "").split(`
`), u = t.length;
  if (u === 1)
    return W(t[0], e, a, r);
  for (var s = new b(0, 0, 0, 0), h = 0; h < t.length; h++) {
    var n = W(t[h], e, a, r);
    h === 0 ? s.copy(n) : s.union(n);
  }
  return s;
}
function L(i, e, a, r) {
  return a === "right" ? r ? i += e : i -= e : a === "center" && (r ? i += e / 2 : i -= e / 2), i;
}
function x(i, e, a, r) {
  return a === "middle" ? r ? i += e / 2 : i -= e / 2 : a === "bottom" && (r ? i += e : i -= e), i;
}
function _(i) {
  return C(i).stWideCharWidth;
}
function v(i, e) {
  return typeof i == "string" ? i.lastIndexOf("%") >= 0 ? parseFloat(i) / 100 * e : parseFloat(i) : i;
}
function y(i, e, a) {
  var r = e.position || "inside", t = e.distance != null ? e.distance : 5, u = a.height, s = a.width, h = u / 2, n = a.x, d = a.y, c = "left", l = "top";
  if (r instanceof Array)
    n += v(r[0], a.width), d += v(r[1], a.height), c = null, l = null;
  else
    switch (r) {
      case "left":
        n -= t, d += h, c = "right", l = "middle";
        break;
      case "right":
        n += t + s, d += h, l = "middle";
        break;
      case "top":
        n += s / 2, d -= t, c = "center", l = "bottom";
        break;
      case "bottom":
        n += s / 2, d += u + t, c = "center";
        break;
      case "inside":
        n += s / 2, d += h, c = "center", l = "middle";
        break;
      case "insideLeft":
        n += t, d += h, l = "middle";
        break;
      case "insideRight":
        n += s - t, d += h, c = "right", l = "middle";
        break;
      case "insideTop":
        n += s / 2, d += t, c = "center";
        break;
      case "insideBottom":
        n += s / 2, d += u - t, c = "center", l = "bottom";
        break;
      case "insideTopLeft":
        n += t, d += t;
        break;
      case "insideTopRight":
        n += s - t, d += t, c = "right";
        break;
      case "insideBottomLeft":
        n += t, d += u - t, l = "bottom";
        break;
      case "insideBottomRight":
        n += s - t, d += u - t, c = "right", l = "bottom";
        break;
    }
  return i = i || {}, i.x = n, i.y = d, i.align = c, i.verticalAlign = l, i;
}
export {
  L as adjustTextX,
  x as adjustTextY,
  y as calculateTextPosition,
  C as ensureFontMeasureInfo,
  R as getBoundingRect,
  _ as getLineHeight,
  W as innerGetBoundingRect,
  M as measureCharWidth,
  w as measureWidth,
  v as parsePercent
};
