// @atlaskit/pragmatic-drag-and-drop@1.7.4/dist/esm/util/changing-window/count-events-for-safari.js
import { d } from "../../../../../../../../../../_virtual/index4.js";
import { isSafari as a } from "../is-safari.js";
var i = {
  isLeavingWindow: /* @__PURE__ */ Symbol("leaving"),
  isEnteringWindow: /* @__PURE__ */ Symbol("entering")
};
function l(r) {
  var n = r.dragLeave;
  return a() ? n.hasOwnProperty(i.isLeavingWindow) : !1;
}
(function() {
  if (typeof window > "u" || process.env.NODE_ENV === "test" || !a())
    return;
  function n() {
    return {
      enterCount: 0,
      isOverWindow: !1
    };
  }
  var e = n();
  function o() {
    e = n();
  }
  d.bindAll(
    window,
    [{
      type: "dragstart",
      listener: function() {
        e.enterCount = 0, e.isOverWindow = !0;
      }
    }, {
      type: "drop",
      listener: o
    }, {
      type: "dragend",
      listener: o
    }, {
      type: "dragenter",
      listener: function(t) {
        !e.isOverWindow && e.enterCount === 0 && (t[i.isEnteringWindow] = !0), e.isOverWindow = !0, e.enterCount++;
      }
    }, {
      type: "dragleave",
      listener: function(t) {
        e.enterCount--, e.isOverWindow && e.enterCount === 0 && (t[i.isLeavingWindow] = !0, e.isOverWindow = !1);
      }
    }],
    // using `capture: true` so that adding event listeners
    // in bubble phase will have the correct symbols
    {
      capture: !0
    }
  );
})();
export {
  l as isLeavingWindowInSafari
};
