// date-fns@3.6.0/parse.js
import { constructFrom as s } from "./constructFrom.js";
import { getDefaultOptions as x } from "./getDefaultOptions.js";
import { toDate as w } from "./toDate.js";
import { longFormatters as N } from "./_lib/format/longFormatters.js";
import { isProtectedWeekYearToken as P, warnOrThrowProtectedError as y, isProtectedDayOfYearToken as b } from "./_lib/protectedTokens.js";
import { parsers as C } from "./parse/_lib/parsers.js";
import { DateToSystemTimezoneSetter as F } from "./parse/_lib/Setter.js";
import { enUS as $ } from "./locale/en-US.js";
const A = /[yYQqMLwIdDecihHKkms]o|(\w)\1*|''|'(''|[^'])+('|$)|./g, W = /P+p+|P+|p+|''|'(''|[^'])+('|$)|./g, L = /^'([^]*?)'?$/, Y = /''/g, j = /\S/, q = /[a-zA-Z]/;
function D(n, c, r, Q) {
  const a = x(), l = a.locale ?? $, E = a.firstWeekContainsDate ?? a.locale?.options?.firstWeekContainsDate ?? 1, O = a.weekStartsOn ?? a.locale?.options?.weekStartsOn ?? 0;
  if (c === "")
    return n === "" ? w(r) : s(r, NaN);
  const p = {
    firstWeekContainsDate: E,
    weekStartsOn: O,
    locale: l
  }, m = [new F()], d = c.match(W).map((t) => {
    const o = t[0];
    if (o in N) {
      const e = N[o];
      return e(t, l.formatLong);
    }
    return t;
  }).join("").match(A), f = [];
  for (let t of d) {
    P(t) && y(t, c, n), b(t) && y(t, c, n);
    const o = t[0], e = C[o];
    if (e) {
      const { incompatibleTokens: h } = e;
      if (Array.isArray(h)) {
        const k = f.find(
          (T) => h.includes(T.token) || T.token === o
        );
        if (k)
          throw new RangeError(
            `The format string mustn't contain \`${k.fullToken}\` and \`${t}\` at the same time`
          );
      } else if (e.incompatibleTokens === "*" && f.length > 0)
        throw new RangeError(
          `The format string mustn't contain \`${t}\` and any other token at the same time`
        );
      f.push({ token: o, fullToken: t });
      const u = e.run(
        n,
        t,
        l.match,
        p
      );
      if (!u)
        return s(r, NaN);
      m.push(u.setter), n = u.rest;
    } else {
      if (o.match(q))
        throw new RangeError(
          "Format string contains an unescaped latin alphabet character `" + o + "`"
        );
      if (t === "''" ? t = "'" : o === "'" && (t = z(t)), n.indexOf(t) === 0)
        n = n.slice(t.length);
      else
        return s(r, NaN);
    }
  }
  if (n.length > 0 && j.test(n))
    return s(r, NaN);
  const R = m.map((t) => t.priority).sort((t, o) => o - t).filter((t, o, e) => e.indexOf(t) === o).map(
    (t) => m.filter((o) => o.priority === t).sort((o, e) => e.subPriority - o.subPriority)
  ).map((t) => t[0]);
  let i = w(r);
  if (isNaN(i.getTime()))
    return s(r, NaN);
  const g = {};
  for (const t of R) {
    if (!t.validate(i, p))
      return s(r, NaN);
    const o = t.set(i, g, p);
    Array.isArray(o) ? (i = o[0], Object.assign(g, o[1])) : i = o;
  }
  return s(r, i);
}
function z(n) {
  return n.match(L)[1].replace(Y, "'");
}
export {
  D as default,
  N as longFormatters,
  D as parse,
  C as parsers
};
