// @tiptap/extension-drag-handle@2.24.0/dist/index.js
import { Extension as A } from "../../../../../@tiptap_core@2.24.0_@tiptap_pm@2.24.0/node_modules/@tiptap/core/dist/index.js";
import { isChangeOrigin as H } from "../../../../../@tiptap_extension-collaboration@2.24.0_@tiptap_core@2.24.0_@tiptap_pm@2.24.0__@tiptap_pm@2.24_pf3qwaorsbxk47ohq5ljq6pmoi/node_modules/@tiptap/extension-collaboration/dist/index.js";
import { PluginKey as b, Plugin as T } from "../../../../../prosemirror-state@1.4.3/node_modules/prosemirror-state/dist/index.js";
import $ from "../../../../../tippy.js@6.3.7/node_modules/tippy.js/dist/tippy.esm.js";
import { getSelectionRanges as x, NodeRangeSelection as C } from "../../../../../@tiptap_extension-node-range@2.24.0_@tiptap_core@2.24.0_@tiptap_pm@2.24.0__@tiptap_pm@2.24.0/node_modules/@tiptap/extension-node-range/dist/index.js";
import { absolutePositionToRelativePosition as S, relativePositionToAbsolutePosition as w } from "../../../../../y-prosemirror@1.3.6_prosemirror-model@1.25.0_prosemirror-state@1.4.3_prosemirror-view@1.38.1__5io44ux6u7ky2cicdaagpya6xa/node_modules/y-prosemirror/src/lib.js";
import { ySyncPluginKey as M } from "../../../../../y-prosemirror@1.3.6_prosemirror-model@1.25.0_prosemirror-state@1.4.3_prosemirror-view@1.38.1__5io44ux6u7ky2cicdaagpya6xa/node_modules/y-prosemirror/src/plugins/keys.js";
function L(e) {
  let o = "";
  const t = getComputedStyle(e);
  for (let i = 0; i < t.length; i += 1)
    o += `${t[i]}:${t.getPropertyValue(t[i])};`;
  return o;
}
function I(e) {
  const o = e.cloneNode(!0), t = [e, ...Array.from(e.getElementsByTagName("*"))], i = [o, ...Array.from(o.getElementsByTagName("*"))];
  return t.forEach((s, d) => {
    i[d].style.cssText = L(s);
  }), o;
}
const R = (e) => {
  const { x: o, y: t, direction: i, editor: s } = e;
  let d = null, n = null, a = null, u = o;
  for (; n === null && u < window.innerWidth && u > 0; ) {
    const l = document.elementsFromPoint(u, t), y = l.findIndex((c) => c.classList.contains("ProseMirror")), r = l.slice(0, y);
    if (r.length > 0) {
      const c = r[0];
      if (d = c, a = s.view.posAtDOM(c, 0), a >= 0) {
        n = s.state.doc.nodeAt(Math.max(a - 1, 0)), n?.isText && (n = s.state.doc.nodeAt(Math.max(a - 1, 0))), n || (n = s.state.doc.nodeAt(Math.max(a, 0)));
        break;
      }
    }
    i === "left" ? u -= 1 : u += 1;
  }
  return { resultElement: d, resultNode: n, pos: a ?? null };
};
function h(e, o) {
  return window.getComputedStyle(e)[o];
}
function B(e = 0, o = 0, t = 0) {
  return Math.min(Math.max(e, o), t);
}
function W(e, o, t) {
  const i = parseInt(h(e.dom, "paddingLeft"), 10), s = parseInt(h(e.dom, "paddingRight"), 10), d = parseInt(h(e.dom, "borderLeftWidth"), 10), n = parseInt(h(e.dom, "borderLeftWidth"), 10), a = e.dom.getBoundingClientRect();
  return {
    left: B(o, a.left + i + d, a.right - s - n),
    top: t
  };
}
function O(e) {
  var o;
  (o = e.parentNode) === null || o === void 0 || o.removeChild(e);
}
function X(e, o) {
  const { doc: t } = o.view.state, i = R({
    editor: o,
    x: e.clientX,
    y: e.clientY,
    direction: "right"
  });
  if (!i.resultNode || i.pos === null)
    return [];
  const s = e.clientX, d = W(o.view, s, e.clientY), n = o.view.posAtCoords(d);
  if (!n)
    return [];
  const { pos: a } = n;
  if (!t.resolve(a).parent)
    return [];
  const l = t.resolve(i.pos), y = t.resolve(i.pos + 1);
  return x(l, y, 0);
}
function _(e, o) {
  const { view: t } = o;
  if (!e.dataTransfer)
    return;
  const { empty: i, $from: s, $to: d } = t.state.selection, n = X(e, o), a = x(s, d, 0), u = a.some((g) => n.find((m) => m.$from === g.$from && m.$to === g.$to)), l = i || !u ? n : a;
  if (!l.length)
    return;
  const { tr: y } = t.state, r = document.createElement("div"), c = l[0].$from.pos, v = l[l.length - 1].$to.pos, p = C.create(t.state.doc, c, v), f = p.content();
  l.forEach((g) => {
    const m = t.nodeDOM(g.$from.pos), P = I(m);
    r.append(P);
  }), r.style.position = "absolute", r.style.top = "-10000px", document.body.append(r), e.dataTransfer.clearData(), e.dataTransfer.setDragImage(r, 0, 0), t.dragging = { slice: f, move: !0 }, y.setSelection(p), t.dispatch(y), document.addEventListener("drop", () => O(r), { once: !0 });
}
const D = (e, o) => {
  const t = e.resolve(o), { depth: i } = t;
  return i === 0 ? o : t.pos - t.parentOffset - 1;
}, k = (e, o) => {
  const t = e.nodeAt(o), i = e.resolve(o);
  let { depth: s } = i, d = t;
  for (; s > 0; ) {
    const n = i.node(s);
    s -= 1, s === 0 && (d = n);
  }
  return d;
}, E = (e, o) => {
  const t = M.getState(e);
  return t ? S(o, t.type, t.binding.mapping) : null;
}, Y = (e, o) => {
  const t = M.getState(e);
  return t ? w(t.doc, t.type, o, t.binding.mapping) || 0 : -1;
}, N = (e, o) => {
  let t = o;
  for (; t && t.parentNode && t.parentNode !== e.dom; )
    t = t.parentNode;
  return t;
}, q = new b("dragHandle"), F = ({ pluginKey: e = q, element: o, editor: t, tippyOptions: i, onNodeChange: s }) => {
  const d = document.createElement("div");
  let n = null, a = !1, u = null, l = -1, y;
  return o.addEventListener("dragstart", (r) => {
    _(r, t), setTimeout(() => {
      o && (o.style.pointerEvents = "none");
    }, 0);
  }), o.addEventListener("dragend", () => {
    o && (o.style.pointerEvents = "auto");
  }), new T({
    key: typeof e == "string" ? new b(e) : e,
    state: {
      init() {
        return { locked: !1 };
      },
      apply(r, c, v, p) {
        const f = r.getMeta("lockDragHandle"), g = r.getMeta("hideDragHandle");
        if (f !== void 0 && (a = f), g && n)
          return n.hide(), a = !1, u = null, l = -1, s?.({ editor: t, node: null, pos: -1 }), c;
        if (r.docChanged && l !== -1 && o && n)
          if (H(r)) {
            const m = Y(p, y);
            m !== l && (l = m);
          } else {
            const m = r.mapping.map(l);
            m !== l && (l = m, y = E(p, l));
          }
        return c;
      }
    },
    view: (r) => {
      var c;
      return o.draggable = !0, o.style.pointerEvents = "auto", (c = t.view.dom.parentElement) === null || c === void 0 || c.appendChild(d), d.appendChild(o), d.style.pointerEvents = "none", d.style.position = "absolute", d.style.top = "0", d.style.left = "0", {
        update(v, p) {
          if (!o)
            return;
          if (!t.isEditable) {
            n?.destroy(), n = null;
            return;
          }
          if (n || (n = $(r.dom, {
            getReferenceClientRect: null,
            interactive: !0,
            trigger: "manual",
            placement: "left-start",
            hideOnClick: !1,
            duration: 100,
            popperOptions: {
              modifiers: [
                { name: "flip", enabled: !1 },
                {
                  name: "preventOverflow",
                  options: {
                    rootBoundary: "document",
                    mainAxis: !1
                  }
                }
              ]
            },
            ...i,
            appendTo: d,
            content: o
          })), a ? o.draggable = !1 : o.draggable = !0, r.state.doc.eq(p.doc) || l === -1)
            return;
          let f = r.nodeDOM(l);
          if (f = N(r, f), f === r.dom || f?.nodeType !== 1)
            return;
          const g = r.posAtDOM(f, 0), m = k(t.state.doc, g), P = D(t.state.doc, g);
          u = m, l = P, y = E(r.state, l), s?.({ editor: t, node: u, pos: l }), n.setProps({
            getReferenceClientRect: () => f.getBoundingClientRect()
          });
        },
        // TODO: Kills even on hot reload
        destroy() {
          n?.destroy(), o && O(d);
        }
      };
    },
    props: {
      handleDOMEvents: {
        mouseleave(r, c) {
          return a || c.target && !d.contains(c.relatedTarget) && (n?.hide(), u = null, l = -1, s?.({ editor: t, node: null, pos: -1 })), !1;
        },
        mousemove(r, c) {
          if (!o || !n || a)
            return !1;
          const v = R({
            x: c.clientX,
            y: c.clientY,
            direction: "right",
            editor: t
          });
          if (!v.resultElement)
            return !1;
          let p = v.resultElement;
          if (p = N(r, p), p === r.dom || p?.nodeType !== 1)
            return !1;
          const f = r.posAtDOM(p, 0), g = k(t.state.doc, f);
          if (g !== u) {
            const m = D(t.state.doc, f);
            u = g, l = m, y = E(r.state, l), s?.({ editor: t, node: u, pos: l }), n.setProps({
              getReferenceClientRect: () => p.getBoundingClientRect()
            }), n.show();
          }
          return !1;
        }
      }
    }
  });
};
A.create({
  name: "dragHandle",
  addOptions() {
    return {
      render() {
        const e = document.createElement("div");
        return e.classList.add("drag-handle"), e;
      },
      tippyOptions: {},
      locked: !1,
      onNodeChange: () => null
    };
  },
  addCommands() {
    return {
      lockDragHandle: () => ({ editor: e }) => (this.options.locked = !0, e.commands.setMeta("lockDragHandle", this.options.locked)),
      unlockDragHandle: () => ({ editor: e }) => (this.options.locked = !1, e.commands.setMeta("lockDragHandle", this.options.locked)),
      toggleDragHandle: () => ({ editor: e }) => (this.options.locked = !this.options.locked, e.commands.setMeta("lockDragHandle", this.options.locked))
    };
  },
  addProseMirrorPlugins() {
    const e = this.options.render();
    return [
      F({
        tippyOptions: this.options.tippyOptions,
        element: e,
        editor: this.editor,
        onNodeChange: this.options.onNodeChange
      })
    ];
  }
});
export {
  F as DragHandlePlugin,
  q as dragHandlePluginDefaultKey
};
