// recharts@2.15.0/es6/polar/PolarRadiusAxis.js
import f, { PureComponent as D } from "react";
import V from "../../../../../../../_virtual/maxBy.js";
import C from "../../../../../../../_virtual/minBy.js";
import $ from "../../../../../../../_virtual/isFunction.js";
import { clsx as E } from "../../../../../clsx@2.1.1/node_modules/clsx/dist/clsx.js";
import { Text as F } from "../component/Text.js";
import { Label as M } from "../component/Label.js";
import { Layer as w } from "../container/Layer.js";
import { polarToCartesian as j, getTickClassName as W } from "../util/PolarUtils.js";
import { adaptEventsOfChild as z } from "../util/types.js";
import { filterProps as g } from "../util/ReactUtils.js";
var K = ["cx", "cy", "angle", "ticks", "axisLine"], q = ["ticks", "tick", "angle", "tickFormatter", "stroke"];
function y(t) {
  "@babel/helpers - typeof";
  return y = typeof Symbol == "function" && typeof Symbol.iterator == "symbol" ? function(e) {
    return typeof e;
  } : function(e) {
    return e && typeof Symbol == "function" && e.constructor === Symbol && e !== Symbol.prototype ? "symbol" : typeof e;
  }, y(t);
}
function d() {
  return d = Object.assign ? Object.assign.bind() : function(t) {
    for (var e = 1; e < arguments.length; e++) {
      var n = arguments[e];
      for (var r in n)
        Object.prototype.hasOwnProperty.call(n, r) && (t[r] = n[r]);
    }
    return t;
  }, d.apply(this, arguments);
}
function R(t, e) {
  var n = Object.keys(t);
  if (Object.getOwnPropertySymbols) {
    var r = Object.getOwnPropertySymbols(t);
    e && (r = r.filter(function(i) {
      return Object.getOwnPropertyDescriptor(t, i).enumerable;
    })), n.push.apply(n, r);
  }
  return n;
}
function p(t) {
  for (var e = 1; e < arguments.length; e++) {
    var n = arguments[e] != null ? arguments[e] : {};
    e % 2 ? R(Object(n), !0).forEach(function(r) {
      k(t, r, n[r]);
    }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(t, Object.getOwnPropertyDescriptors(n)) : R(Object(n)).forEach(function(r) {
      Object.defineProperty(t, r, Object.getOwnPropertyDescriptor(n, r));
    });
  }
  return t;
}
function S(t, e) {
  if (t == null) return {};
  var n = G(t, e), r, i;
  if (Object.getOwnPropertySymbols) {
    var a = Object.getOwnPropertySymbols(t);
    for (i = 0; i < a.length; i++)
      r = a[i], !(e.indexOf(r) >= 0) && Object.prototype.propertyIsEnumerable.call(t, r) && (n[r] = t[r]);
  }
  return n;
}
function G(t, e) {
  if (t == null) return {};
  var n = {};
  for (var r in t)
    if (Object.prototype.hasOwnProperty.call(t, r)) {
      if (e.indexOf(r) >= 0) continue;
      n[r] = t[r];
    }
  return n;
}
function H(t, e) {
  if (!(t instanceof e))
    throw new TypeError("Cannot call a class as a function");
}
function L(t, e) {
  for (var n = 0; n < e.length; n++) {
    var r = e[n];
    r.enumerable = r.enumerable || !1, r.configurable = !0, "value" in r && (r.writable = !0), Object.defineProperty(t, I(r.key), r);
  }
}
function J(t, e, n) {
  return e && L(t.prototype, e), n && L(t, n), Object.defineProperty(t, "prototype", { writable: !1 }), t;
}
function Q(t, e, n) {
  return e = x(e), U(t, N() ? Reflect.construct(e, n || [], x(t).constructor) : e.apply(t, n));
}
function U(t, e) {
  if (e && (y(e) === "object" || typeof e == "function"))
    return e;
  if (e !== void 0)
    throw new TypeError("Derived constructors may only return object or undefined");
  return X(t);
}
function X(t) {
  if (t === void 0)
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  return t;
}
function N() {
  try {
    var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function() {
    }));
  } catch {
  }
  return (N = function() {
    return !!t;
  })();
}
function x(t) {
  return x = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function(n) {
    return n.__proto__ || Object.getPrototypeOf(n);
  }, x(t);
}
function Y(t, e) {
  if (typeof e != "function" && e !== null)
    throw new TypeError("Super expression must either be null or a function");
  t.prototype = Object.create(e && e.prototype, { constructor: { value: t, writable: !0, configurable: !0 } }), Object.defineProperty(t, "prototype", { writable: !1 }), e && T(t, e);
}
function T(t, e) {
  return T = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function(r, i) {
    return r.__proto__ = i, r;
  }, T(t, e);
}
function k(t, e, n) {
  return e = I(e), e in t ? Object.defineProperty(t, e, { value: n, enumerable: !0, configurable: !0, writable: !0 }) : t[e] = n, t;
}
function I(t) {
  var e = Z(t, "string");
  return y(e) == "symbol" ? e : e + "";
}
function Z(t, e) {
  if (y(t) != "object" || !t) return t;
  var n = t[Symbol.toPrimitive];
  if (n !== void 0) {
    var r = n.call(t, e);
    if (y(r) != "object") return r;
    throw new TypeError("@@toPrimitive must return a primitive value.");
  }
  return (e === "string" ? String : Number)(t);
}
var A = /* @__PURE__ */ (function(t) {
  function e() {
    return H(this, e), Q(this, e, arguments);
  }
  return Y(e, t), J(e, [{
    key: "getTickValueCoord",
    value: (
      /**
       * Calculate the coordinate of tick
       * @param  {Number} coordinate The radius of tick
       * @return {Object} (x, y)
       */
      function(r) {
        var i = r.coordinate, a = this.props, o = a.angle, s = a.cx, u = a.cy;
        return j(s, u, i, o);
      }
    )
  }, {
    key: "getTickTextAnchor",
    value: function() {
      var r = this.props.orientation, i;
      switch (r) {
        case "left":
          i = "end";
          break;
        case "right":
          i = "start";
          break;
        default:
          i = "middle";
          break;
      }
      return i;
    }
  }, {
    key: "getViewBox",
    value: function() {
      var r = this.props, i = r.cx, a = r.cy, o = r.angle, s = r.ticks, u = V(s, function(l) {
        return l.coordinate || 0;
      }), m = C(s, function(l) {
        return l.coordinate || 0;
      });
      return {
        cx: i,
        cy: a,
        startAngle: o,
        endAngle: o,
        innerRadius: m.coordinate || 0,
        outerRadius: u.coordinate || 0
      };
    }
  }, {
    key: "renderAxisLine",
    value: function() {
      var r = this.props, i = r.cx, a = r.cy, o = r.angle, s = r.ticks, u = r.axisLine, m = S(r, K), l = s.reduce(function(h, c) {
        return [Math.min(h[0], c.coordinate), Math.max(h[1], c.coordinate)];
      }, [1 / 0, -1 / 0]), b = j(i, a, l[0], o), v = j(i, a, l[1], o), O = p(p(p({}, g(m, !1)), {}, {
        fill: "none"
      }, g(u, !1)), {}, {
        x1: b.x,
        y1: b.y,
        x2: v.x,
        y2: v.y
      });
      return /* @__PURE__ */ f.createElement("line", d({
        className: "recharts-polar-radius-axis-line"
      }, O));
    }
  }, {
    key: "renderTicks",
    value: function() {
      var r = this, i = this.props, a = i.ticks, o = i.tick, s = i.angle, u = i.tickFormatter, m = i.stroke, l = S(i, q), b = this.getTickTextAnchor(), v = g(l, !1), O = g(o, !1), h = a.map(function(c, P) {
        var _ = r.getTickValueCoord(c), B = p(p(p(p({
          textAnchor: b,
          transform: "rotate(".concat(90 - s, ", ").concat(_.x, ", ").concat(_.y, ")")
        }, v), {}, {
          stroke: "none",
          fill: m
        }, O), {}, {
          index: P
        }, _), {}, {
          payload: c
        });
        return /* @__PURE__ */ f.createElement(w, d({
          className: E("recharts-polar-radius-axis-tick", W(o)),
          key: "tick-".concat(c.coordinate)
        }, z(r.props, c, P)), e.renderTickItem(o, B, u ? u(c.value, P) : c.value));
      });
      return /* @__PURE__ */ f.createElement(w, {
        className: "recharts-polar-radius-axis-ticks"
      }, h);
    }
  }, {
    key: "render",
    value: function() {
      var r = this.props, i = r.ticks, a = r.axisLine, o = r.tick;
      return !i || !i.length ? null : /* @__PURE__ */ f.createElement(w, {
        className: E("recharts-polar-radius-axis", this.props.className)
      }, a && this.renderAxisLine(), o && this.renderTicks(), M.renderCallByParent(this.props, this.getViewBox()));
    }
  }], [{
    key: "renderTickItem",
    value: function(r, i, a) {
      var o;
      return /* @__PURE__ */ f.isValidElement(r) ? o = /* @__PURE__ */ f.cloneElement(r, i) : $(r) ? o = r(i) : o = /* @__PURE__ */ f.createElement(F, d({}, i, {
        className: "recharts-polar-radius-axis-tick-value"
      }), a), o;
    }
  }]);
})(D);
k(A, "displayName", "PolarRadiusAxis");
k(A, "axisType", "radiusAxis");
k(A, "defaultProps", {
  type: "number",
  radiusAxisId: 0,
  cx: 0,
  cy: 0,
  angle: 0,
  orientation: "right",
  stroke: "#ccc",
  axisLine: !0,
  tick: !0,
  tickCount: 5,
  allowDataOverflow: !1,
  scale: "auto",
  allowDuplicatedCategory: !0
});
export {
  A as PolarRadiusAxis
};
