// recharts@2.15.0/es6/polar/PolarGrid.js
import o from "react";
import { clsx as h } from "../../../../../clsx@2.1.1/node_modules/clsx/dist/clsx.js";
import { polarToCartesian as p } from "../util/PolarUtils.js";
import { filterProps as v } from "../util/ReactUtils.js";
var j = ["cx", "cy", "innerRadius", "outerRadius", "gridType", "radialLines"];
function g(r) {
  "@babel/helpers - typeof";
  return g = typeof Symbol == "function" && typeof Symbol.iterator == "symbol" ? function(e) {
    return typeof e;
  } : function(e) {
    return e && typeof Symbol == "function" && e.constructor === Symbol && e !== Symbol.prototype ? "symbol" : typeof e;
  }, g(r);
}
function R(r, e) {
  if (r == null) return {};
  var t = E(r, e), n, i;
  if (Object.getOwnPropertySymbols) {
    var a = Object.getOwnPropertySymbols(r);
    for (i = 0; i < a.length; i++)
      n = a[i], !(e.indexOf(n) >= 0) && Object.prototype.propertyIsEnumerable.call(r, n) && (t[n] = r[n]);
  }
  return t;
}
function E(r, e) {
  if (r == null) return {};
  var t = {};
  for (var n in r)
    if (Object.prototype.hasOwnProperty.call(r, n)) {
      if (e.indexOf(n) >= 0) continue;
      t[n] = r[n];
    }
  return t;
}
function u() {
  return u = Object.assign ? Object.assign.bind() : function(r) {
    for (var e = 1; e < arguments.length; e++) {
      var t = arguments[e];
      for (var n in t)
        Object.prototype.hasOwnProperty.call(t, n) && (r[n] = t[n]);
    }
    return r;
  }, u.apply(this, arguments);
}
function x(r, e) {
  var t = Object.keys(r);
  if (Object.getOwnPropertySymbols) {
    var n = Object.getOwnPropertySymbols(r);
    e && (n = n.filter(function(i) {
      return Object.getOwnPropertyDescriptor(r, i).enumerable;
    })), t.push.apply(t, n);
  }
  return t;
}
function m(r) {
  for (var e = 1; e < arguments.length; e++) {
    var t = arguments[e] != null ? arguments[e] : {};
    e % 2 ? x(Object(t), !0).forEach(function(n) {
      w(r, n, t[n]);
    }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(r, Object.getOwnPropertyDescriptors(t)) : x(Object(t)).forEach(function(n) {
      Object.defineProperty(r, n, Object.getOwnPropertyDescriptor(t, n));
    });
  }
  return r;
}
function w(r, e, t) {
  return e = S(e), e in r ? Object.defineProperty(r, e, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : r[e] = t, r;
}
function S(r) {
  var e = _(r, "string");
  return g(e) == "symbol" ? e : e + "";
}
function _(r, e) {
  if (g(r) != "object" || !r) return r;
  var t = r[Symbol.toPrimitive];
  if (t !== void 0) {
    var n = t.call(r, e);
    if (g(n) != "object") return n;
    throw new TypeError("@@toPrimitive must return a primitive value.");
  }
  return (e === "string" ? String : Number)(r);
}
var C = function(e, t, n, i) {
  var a = "";
  return i.forEach(function(c, f) {
    var l = p(t, n, e, c);
    f ? a += "L ".concat(l.x, ",").concat(l.y) : a += "M ".concat(l.x, ",").concat(l.y);
  }), a += "Z", a;
}, L = function(e) {
  var t = e.cx, n = e.cy, i = e.innerRadius, a = e.outerRadius, c = e.polarAngles, f = e.radialLines;
  if (!c || !c.length || !f)
    return null;
  var l = m({
    stroke: "#ccc"
  }, v(e, !1));
  return /* @__PURE__ */ o.createElement("g", {
    className: "recharts-polar-grid-angle"
  }, c.map(function(s) {
    var d = p(t, n, i, s), y = p(t, n, a, s);
    return /* @__PURE__ */ o.createElement("line", u({}, l, {
      key: "line-".concat(s),
      x1: d.x,
      y1: d.y,
      x2: y.x,
      y2: y.y
    }));
  }));
}, T = function(e) {
  var t = e.cx, n = e.cy, i = e.radius, a = e.index, c = m(m({
    stroke: "#ccc"
  }, v(e, !1)), {}, {
    fill: "none"
  });
  return /* @__PURE__ */ o.createElement("circle", u({}, c, {
    className: h("recharts-polar-grid-concentric-circle", e.className),
    key: "circle-".concat(a),
    cx: t,
    cy: n,
    r: i
  }));
}, N = function(e) {
  var t = e.radius, n = e.index, i = m(m({
    stroke: "#ccc"
  }, v(e, !1)), {}, {
    fill: "none"
  });
  return /* @__PURE__ */ o.createElement("path", u({}, i, {
    className: h("recharts-polar-grid-concentric-polygon", e.className),
    key: "path-".concat(n),
    d: C(t, e.cx, e.cy, e.polarAngles)
  }));
}, A = function(e) {
  var t = e.polarRadius, n = e.gridType;
  return !t || !t.length ? null : /* @__PURE__ */ o.createElement("g", {
    className: "recharts-polar-grid-concentric"
  }, t.map(function(i, a) {
    var c = a;
    return n === "circle" ? /* @__PURE__ */ o.createElement(T, u({
      key: c
    }, e, {
      radius: i,
      index: a
    })) : /* @__PURE__ */ o.createElement(N, u({
      key: c
    }, e, {
      radius: i,
      index: a
    }));
  }));
}, $ = function(e) {
  var t = e.cx, n = t === void 0 ? 0 : t, i = e.cy, a = i === void 0 ? 0 : i, c = e.innerRadius, f = c === void 0 ? 0 : c, l = e.outerRadius, s = l === void 0 ? 0 : l, d = e.gridType, y = d === void 0 ? "polygon" : d, b = e.radialLines, P = b === void 0 ? !0 : b, O = R(e, j);
  return s <= 0 ? null : /* @__PURE__ */ o.createElement("g", {
    className: "recharts-polar-grid"
  }, /* @__PURE__ */ o.createElement(L, u({
    cx: n,
    cy: a,
    innerRadius: f,
    outerRadius: s,
    gridType: y,
    radialLines: P
  }, O)), /* @__PURE__ */ o.createElement(A, u({
    cx: n,
    cy: a,
    innerRadius: f,
    outerRadius: s,
    gridType: y,
    radialLines: P
  }, O)));
};
$.displayName = "PolarGrid";
export {
  $ as PolarGrid
};
