// recharts@2.15.0/es6/polar/Radar.js
import y, { PureComponent as V } from "react";
import q from "../../../../../react-smooth@4.0.4_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/react-smooth/es6/Animate.js";
import N from "../../../../../../../_virtual/isNil.js";
import C from "../../../../../../../_virtual/last.js";
import G from "../../../../../../../_virtual/first.js";
import X from "../../../../../../../_virtual/isEqual.js";
import E from "../../../../../../../_virtual/isFunction.js";
import { clsx as I } from "../../../../../clsx@2.1.1/node_modules/clsx/dist/clsx.js";
import { interpolateNumber as j } from "../util/DataUtils.js";
import { Global as Y } from "../util/Global.js";
import { polarToCartesian as L } from "../util/PolarUtils.js";
import { getValueByDataKey as M } from "../util/ChartUtils.js";
import { Polygon as H } from "../shape/Polygon.js";
import { Dot as J } from "../shape/Dot.js";
import { Layer as w } from "../container/Layer.js";
import { LabelList as Q } from "../component/LabelList.js";
import { filterProps as D } from "../util/ReactUtils.js";
var U = ["key"];
function O(t) {
  "@babel/helpers - typeof";
  return O = typeof Symbol == "function" && typeof Symbol.iterator == "symbol" ? function(e) {
    return typeof e;
  } : function(e) {
    return e && typeof Symbol == "function" && e.constructor === Symbol && e !== Symbol.prototype ? "symbol" : typeof e;
  }, O(t);
}
function Z(t, e) {
  if (t == null) return {};
  var r = tt(t, e), n, i;
  if (Object.getOwnPropertySymbols) {
    var o = Object.getOwnPropertySymbols(t);
    for (i = 0; i < o.length; i++)
      n = o[i], !(e.indexOf(n) >= 0) && Object.prototype.propertyIsEnumerable.call(t, n) && (r[n] = t[n]);
  }
  return r;
}
function tt(t, e) {
  if (t == null) return {};
  var r = {};
  for (var n in t)
    if (Object.prototype.hasOwnProperty.call(t, n)) {
      if (e.indexOf(n) >= 0) continue;
      r[n] = t[n];
    }
  return r;
}
function _() {
  return _ = Object.assign ? Object.assign.bind() : function(t) {
    for (var e = 1; e < arguments.length; e++) {
      var r = arguments[e];
      for (var n in r)
        Object.prototype.hasOwnProperty.call(r, n) && (t[n] = r[n]);
    }
    return t;
  }, _.apply(this, arguments);
}
function F(t, e) {
  var r = Object.keys(t);
  if (Object.getOwnPropertySymbols) {
    var n = Object.getOwnPropertySymbols(t);
    e && (n = n.filter(function(i) {
      return Object.getOwnPropertyDescriptor(t, i).enumerable;
    })), r.push.apply(r, n);
  }
  return r;
}
function l(t) {
  for (var e = 1; e < arguments.length; e++) {
    var r = arguments[e] != null ? arguments[e] : {};
    e % 2 ? F(Object(r), !0).forEach(function(n) {
      h(t, n, r[n]);
    }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(t, Object.getOwnPropertyDescriptors(r)) : F(Object(r)).forEach(function(n) {
      Object.defineProperty(t, n, Object.getOwnPropertyDescriptor(r, n));
    });
  }
  return t;
}
function et(t, e) {
  if (!(t instanceof e))
    throw new TypeError("Cannot call a class as a function");
}
function B(t, e) {
  for (var r = 0; r < e.length; r++) {
    var n = e[r];
    n.enumerable = n.enumerable || !1, n.configurable = !0, "value" in n && (n.writable = !0), Object.defineProperty(t, T(n.key), n);
  }
}
function nt(t, e, r) {
  return e && B(t.prototype, e), r && B(t, r), Object.defineProperty(t, "prototype", { writable: !1 }), t;
}
function rt(t, e, r) {
  return e = S(e), it(t, K() ? Reflect.construct(e, r || [], S(t).constructor) : e.apply(t, r));
}
function it(t, e) {
  if (e && (O(e) === "object" || typeof e == "function"))
    return e;
  if (e !== void 0)
    throw new TypeError("Derived constructors may only return object or undefined");
  return ot(t);
}
function ot(t) {
  if (t === void 0)
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  return t;
}
function K() {
  try {
    var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function() {
    }));
  } catch {
  }
  return (K = function() {
    return !!t;
  })();
}
function S(t) {
  return S = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function(r) {
    return r.__proto__ || Object.getPrototypeOf(r);
  }, S(t);
}
function at(t, e) {
  if (typeof e != "function" && e !== null)
    throw new TypeError("Super expression must either be null or a function");
  t.prototype = Object.create(e && e.prototype, { constructor: { value: t, writable: !0, configurable: !0 } }), Object.defineProperty(t, "prototype", { writable: !1 }), e && x(t, e);
}
function x(t, e) {
  return x = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function(n, i) {
    return n.__proto__ = i, n;
  }, x(t, e);
}
function h(t, e, r) {
  return e = T(e), e in t ? Object.defineProperty(t, e, { value: r, enumerable: !0, configurable: !0, writable: !0 }) : t[e] = r, t;
}
function T(t) {
  var e = st(t, "string");
  return O(e) == "symbol" ? e : e + "";
}
function st(t, e) {
  if (O(t) != "object" || !t) return t;
  var r = t[Symbol.toPrimitive];
  if (r !== void 0) {
    var n = r.call(t, e);
    if (O(n) != "object") return n;
    throw new TypeError("@@toPrimitive must return a primitive value.");
  }
  return String(t);
}
var R = /* @__PURE__ */ (function(t) {
  function e() {
    var r;
    et(this, e);
    for (var n = arguments.length, i = new Array(n), o = 0; o < n; o++)
      i[o] = arguments[o];
    return r = rt(this, e, [].concat(i)), h(r, "state", {
      isAnimationFinished: !1
    }), h(r, "handleAnimationEnd", function() {
      var a = r.props.onAnimationEnd;
      r.setState({
        isAnimationFinished: !0
      }), E(a) && a();
    }), h(r, "handleAnimationStart", function() {
      var a = r.props.onAnimationStart;
      r.setState({
        isAnimationFinished: !1
      }), E(a) && a();
    }), h(r, "handleMouseEnter", function(a) {
      var s = r.props.onMouseEnter;
      s && s(r.props, a);
    }), h(r, "handleMouseLeave", function(a) {
      var s = r.props.onMouseLeave;
      s && s(r.props, a);
    }), r;
  }
  return at(e, t), nt(e, [{
    key: "renderDots",
    value: function(n) {
      var i = this.props, o = i.dot, a = i.dataKey, s = D(this.props, !1), f = D(o, !0), p = n.map(function(c, m) {
        var u = l(l(l({
          key: "dot-".concat(m),
          r: 3
        }, s), f), {}, {
          dataKey: a,
          cx: c.x,
          cy: c.y,
          index: m,
          payload: c
        });
        return e.renderDotItem(o, u);
      });
      return /* @__PURE__ */ y.createElement(w, {
        className: "recharts-radar-dots"
      }, p);
    }
  }, {
    key: "renderPolygonStatically",
    value: function(n) {
      var i = this.props, o = i.shape, a = i.dot, s = i.isRange, f = i.baseLinePoints, p = i.connectNulls, c;
      return /* @__PURE__ */ y.isValidElement(o) ? c = /* @__PURE__ */ y.cloneElement(o, l(l({}, this.props), {}, {
        points: n
      })) : E(o) ? c = o(l(l({}, this.props), {}, {
        points: n
      })) : c = /* @__PURE__ */ y.createElement(H, _({}, D(this.props, !0), {
        onMouseEnter: this.handleMouseEnter,
        onMouseLeave: this.handleMouseLeave,
        points: n,
        baseLinePoints: s ? f : null,
        connectNulls: p
      })), /* @__PURE__ */ y.createElement(w, {
        className: "recharts-radar-polygon"
      }, c, a ? this.renderDots(n) : null);
    }
  }, {
    key: "renderPolygonWithAnimation",
    value: function() {
      var n = this, i = this.props, o = i.points, a = i.isAnimationActive, s = i.animationBegin, f = i.animationDuration, p = i.animationEasing, c = i.animationId, m = this.state.prevPoints;
      return /* @__PURE__ */ y.createElement(q, {
        begin: s,
        duration: f,
        isActive: a,
        easing: p,
        from: {
          t: 0
        },
        to: {
          t: 1
        },
        key: "radar-".concat(c),
        onAnimationEnd: this.handleAnimationEnd,
        onAnimationStart: this.handleAnimationStart
      }, function(u) {
        var v = u.t, g = m && m.length / o.length, b = o.map(function(d, A) {
          var P = m && m[Math.floor(A * g)];
          if (P) {
            var W = j(P.x, d.x), $ = j(P.y, d.y);
            return l(l({}, d), {}, {
              x: W(v),
              y: $(v)
            });
          }
          var k = j(d.cx, d.x), z = j(d.cy, d.y);
          return l(l({}, d), {}, {
            x: k(v),
            y: z(v)
          });
        });
        return n.renderPolygonStatically(b);
      });
    }
  }, {
    key: "renderPolygon",
    value: function() {
      var n = this.props, i = n.points, o = n.isAnimationActive, a = n.isRange, s = this.state.prevPoints;
      return o && i && i.length && !a && (!s || !X(s, i)) ? this.renderPolygonWithAnimation() : this.renderPolygonStatically(i);
    }
  }, {
    key: "render",
    value: function() {
      var n = this.props, i = n.hide, o = n.className, a = n.points, s = n.isAnimationActive;
      if (i || !a || !a.length)
        return null;
      var f = this.state.isAnimationFinished, p = I("recharts-radar", o);
      return /* @__PURE__ */ y.createElement(w, {
        className: p
      }, this.renderPolygon(), (!s || f) && Q.renderCallByParent(this.props, a));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function(n, i) {
      return n.animationId !== i.prevAnimationId ? {
        prevAnimationId: n.animationId,
        curPoints: n.points,
        prevPoints: i.curPoints
      } : n.points !== i.curPoints ? {
        curPoints: n.points
      } : null;
    }
  }, {
    key: "renderDotItem",
    value: function(n, i) {
      var o;
      if (/* @__PURE__ */ y.isValidElement(n))
        o = /* @__PURE__ */ y.cloneElement(n, i);
      else if (E(n))
        o = n(i);
      else {
        var a = i.key, s = Z(i, U);
        o = /* @__PURE__ */ y.createElement(J, _({}, s, {
          key: a,
          className: I("recharts-radar-dot", typeof n != "boolean" ? n.className : "")
        }));
      }
      return o;
    }
  }]);
})(V);
h(R, "displayName", "Radar");
h(R, "defaultProps", {
  angleAxisId: 0,
  radiusAxisId: 0,
  hide: !1,
  activeDot: !0,
  dot: !1,
  legendType: "rect",
  isAnimationActive: !Y.isSsr,
  animationBegin: 0,
  animationDuration: 1500,
  animationEasing: "ease"
});
h(R, "getComposedData", function(t) {
  var e = t.radiusAxis, r = t.angleAxis, n = t.displayedData, i = t.dataKey, o = t.bandSize, a = r.cx, s = r.cy, f = !1, p = [], c = r.type !== "number" ? o ?? 0 : 0;
  n.forEach(function(u, v) {
    var g = M(u, r.dataKey, v), b = M(u, i), d = r.scale(g) + c, A = Array.isArray(b) ? C(b) : b, P = N(A) ? void 0 : e.scale(A);
    Array.isArray(b) && b.length >= 2 && (f = !0), p.push(l(l({}, L(a, s, P, d)), {}, {
      name: g,
      value: b,
      cx: a,
      cy: s,
      radius: P,
      angle: d,
      payload: u
    }));
  });
  var m = [];
  return f && p.forEach(function(u) {
    if (Array.isArray(u.value)) {
      var v = G(u.value), g = N(v) ? void 0 : e.scale(v);
      m.push(l(l({}, u), {}, {
        radius: g
      }, L(a, s, g, u.angle)));
    } else
      m.push(u);
  }), {
    points: p,
    isRange: f,
    baseLinePoints: m
  };
});
export {
  R as Radar
};
