// echarts@6.0.0/lib/component/matrix/MatrixView.js
import { __extends as k } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import P from "../../view/Component.js";
import O from "../../model/Model.js";
import T from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/BoundingRect.js";
import { set as A } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/vector.js";
import { subPixelOptimize as D } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/helper/subPixelOptimize.js";
import { XY as Y, expandOrShrinkRect as F, setTooltipConfig as V } from "../../util/graphic.js";
import { ListIterator as W, clearTmpModel as E } from "../../util/model.js";
import { retrieve2 as U, clone as X } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { invert as G } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/matrix.js";
import { setLabelStyle as H } from "../../label/labelStyle.js";
import z from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Rect.js";
import K from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Line.js";
var R = Math.round, j = 0, q = 99, J = {
  normal: 25,
  special: 100
}, Q = {
  normal: 50,
  special: 125
}, yt = (
  /** @class */
  (function(t) {
    k(e, t);
    function e() {
      var r = t !== null && t.apply(this, arguments) || this;
      return r.type = e.type, r;
    }
    return e.prototype.render = function(r, o) {
      this.group.removeAll();
      var n = this.group, u = r.coordinateSystem, a = u.getRect(), S = r.getDimensionModel("x"), s = r.getDimensionModel("y"), y = S.dim, f = s.dim;
      $(n, r, o), C(n, r, y, f, o);
      var g = r.getShallow("borderZ2", !0), d = U(g, q), m = d - 1, x = r.getModel("backgroundStyle").getItemStyle(["borderWidth"]);
      x.lineWidth = 0;
      var i = r.getModel("backgroundStyle").getItemStyle(["color", "decal", "shadowColor", "shadowBlur", "shadowOffsetX", "shadowOffsetY"]);
      i.fill = "none";
      var p = Z(a.clone(), x, j), v = Z(a.clone(), i, d);
      p.silent = !0, v.silent = !0, n.add(p), n.add(v);
      var w = y.getUnitLayoutInfo(0, 0), l = f.getUnitLayoutInfo(1, 0);
      w && l && (y.shouldShow() && n.add(b({
        x1: a.x,
        y1: l.xy,
        x2: a.x + a.width,
        y2: l.xy
      }, S.getModel("dividerLineStyle").getLineStyle(), m)), f.shouldShow() && n.add(b({
        x1: w.xy,
        y1: a.y,
        x2: w.xy,
        y2: a.y + a.height
      }, s.getModel("dividerLineStyle").getLineStyle(), m)));
    }, e.type = "matrix", e;
  })(P)
);
function $(t, e, r) {
  o(0), o(1);
  function o(n) {
    var u = e.getDimensionModel(Y[n]), a = u.dim;
    if (a.shouldShow())
      for (var S = u.getModel("itemStyle"), s = u.getModel("label"), y = e.getShallow("tooltip", !0), f = [], g = a.resetCellIterator(); g.next(); ) {
        var d = g.item, m = {};
        T.copy(m, d.rect), A(f, d.id.x, d.id.y), N(f, e, t, r, d.option, S, s, u, m, d.option.value, Q, y);
      }
  }
}
function C(t, e, r, o, n) {
  u("body", e.getBody(), r, o), r.shouldShow() && o.shouldShow() && u("corner", e.getCorner(), o, r);
  function u(a, S, s, y) {
    var f = new O(e.getShallow(a, !0)), g = f.getModel("itemStyle"), d = f.getModel("label"), m = new W(), x = new W(), i = [], p = e.getShallow("tooltip", !0);
    for (y.resetLayoutIterator(x, 1); x.next(); )
      for (s.resetLayoutIterator(m, 0); m.next(); ) {
        var v = m.item, w = x.item;
        A(i, v.id.x, w.id.y);
        var l = S.getCell(i);
        if (!(l && l.inSpanOf && l.inSpanOf !== l)) {
          var c = {};
          l && l.span ? T.copy(c, l.spanRect) : (v.dim.getLayout(c, 0, i[0]), w.dim.getLayout(c, 1, i[1]));
          var L = l ? l.option : null;
          N(i, e, t, n, L, g, d, f, c, L ? L.value : null, J, p);
        }
      }
  }
}
function N(t, e, r, o, n, u, a, S, s, y, f, g) {
  var d;
  I.option = n ? n.itemStyle : null, I.parentModel = u, h.option = n, h.parentModel = S;
  var m = U(h.getShallow("z2"), n && n.itemStyle ? f.special : f.normal), x = g && g.show, i = Z(s, I.getItemStyle(), m);
  r.add(i);
  var p = h.get("cursor");
  p != null && i.attr("cursor", p);
  var v;
  if (y != null) {
    var w = y + "";
    if (M.option = n ? n.label : null, M.parentModel = a, M.ecModel = o, H(
      i,
      // Currently do not support other states (`emphasis`, `select`, `blur`)
      {
        normal: M
      },
      {
        defaultText: w,
        autoOverflowArea: !0,
        // By default based on boundingRect. But boundingRect contains borderWidth,
        // and borderWidth is half outside the cell. Thus specific `layoutRect` explicitly.
        layoutRect: X(i.shape)
      }
    ), v = i.getTextContent(), v) {
      v.z2 = m + 1;
      var l = v.style;
      if (l && l.overflow && l.overflow !== "none" && l.lineOverflow) {
        var c = {};
        T.copy(c, s), F(c, (((d = i.style) === null || d === void 0 ? void 0 : d.lineWidth) || 0) / 2, !0, !0), i.updateInnerText(), v.getLocalTransform(_), G(_, _), T.applyTransform(c, c, _), v.setClipPath(new z({
          shape: c
        }));
      }
    }
    V({
      el: i,
      componentModel: e,
      itemName: w,
      itemTooltipOption: g,
      formatterParamsExtra: {
        xyLocator: t.slice()
      }
    });
  }
  if (v) {
    var L = M.get("silent");
    L == null && (L = !x), v.silent = L, v.ignoreHostSilent = !0;
  }
  var B = h.get("silent");
  B == null && (B = // If no background color in cell, set `rect.silent: false` will cause that only
  // the border response to mouse hovering, which is probably weird.
  !i.style || i.style.fill === "none" || !i.style.fill), i.silent = B, E(h), E(I), E(M);
}
var h = new O(), I = new O(), M = new O(), _ = [];
function Z(t, e, r) {
  var o = e.lineWidth;
  if (o) {
    var n = t.x + t.width, u = t.y + t.height;
    t.x = D(t.x, o, !0), t.y = D(t.y, o, !0), t.width = D(n, o, !0) - t.x, t.height = D(u, o, !0) - t.y;
  }
  return new z({
    shape: t,
    style: e,
    z2: r
  });
}
function b(t, e, r) {
  var o = e.lineWidth;
  return o && (R(t.x1 * 2) === R(t.x2 * 2) && (t.x1 = t.x2 = D(t.x1, o, !0)), R(t.y1 * 2) === R(t.y2 * 2) && (t.y1 = t.y2 = D(t.y1, o, !0))), new K({
    shape: t,
    style: e,
    silent: !0,
    z2: r
  });
}
export {
  yt as default
};
