// echarts@6.0.0/lib/component/transform/filterTransform.js
import { parseConditionalExpression as c } from "../../util/conditionalExpression.js";
import { createHashMap as v, hasOwn as p } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { makePrintable as d, throwError as f } from "../../util/log.js";
var h = {
  type: "echarts:filter",
  // PENDING: enhance to filter by index rather than create new data
  transform: function(t) {
    for (var e = t.upstream, r, l = c(t.config, {
      valueGetterAttrMap: v({
        dimension: !0
      }),
      prepareGetValue: function(n) {
        var i = "", s = n.dimension;
        p(n, "dimension") || (process.env.NODE_ENV !== "production" && (i = d('Relation condition must has prop "dimension" specified.', "Illegal condition:", n)), f(i));
        var m = e.getDimensionInfo(s);
        return m || (process.env.NODE_ENV !== "production" && (i = d("Can not find dimension info via: " + s + `.
`, "Existing dimensions: ", e.cloneAllDimensionInfo(), `.
`, "Illegal condition:", n, `.
`)), f(i)), {
          dimIdx: m.index
        };
      },
      getValue: function(n) {
        return e.retrieveValueFromItem(r, n.dimIdx);
      }
    }), a = [], o = 0, u = e.count(); o < u; o++)
      r = e.getRawDataItem(o), l.evaluate() && a.push(r);
    return {
      data: a
    };
  }
};
export {
  h as filterTransform
};
