// echarts@6.0.0/lib/component/transform/sortTransform.js
import { SOURCE_FORMAT_ARRAY_ROWS as _, SOURCE_FORMAT_OBJECT_ROWS as I } from "../../util/types.js";
import { throwError as i, makePrintable as O } from "../../util/log.js";
import { each as R } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { normalizeToArray as D } from "../../util/model.js";
import { getRawValueParser as F, SortOrderComparator as S } from "../../data/helper/dataValueHelper.js";
var v = "";
process.env.NODE_ENV !== "production" && (v = ["Valid config is like:", '{ dimension: "age", order: "asc" }', 'or [{ dimension: "age", order: "asc"], { dimension: "date", order: "desc" }]'].join(" "));
var x = {
  type: "echarts:sort",
  transform: function(l) {
    var s = l.upstream, h = l.config, r = "", E = D(h);
    E.length || (process.env.NODE_ENV !== "production" && (r = "Empty `config` in sort transform."), i(r));
    var f = [];
    R(E, function(a) {
      var m = a.dimension, e = a.order, o = a.parser, n = a.incomparable;
      if (m == null && (process.env.NODE_ENV !== "production" && (r = 'Sort transform config must has "dimension" specified.' + v), i(r)), e !== "asc" && e !== "desc" && (process.env.NODE_ENV !== "production" && (r = 'Sort transform config must has "order" specified.' + v), i(r)), n && n !== "min" && n !== "max") {
        var t = "";
        process.env.NODE_ENV !== "production" && (t = 'incomparable must be "min" or "max" rather than "' + n + '".'), i(t);
      }
      if (e !== "asc" && e !== "desc") {
        var c = "";
        process.env.NODE_ENV !== "production" && (c = 'order must be "asc" or "desc" rather than "' + e + '".'), i(c);
      }
      var N = s.getDimensionInfo(m);
      N || (process.env.NODE_ENV !== "production" && (r = O("Can not find dimension info via: " + m + `.
`, "Existing dimensions: ", s.cloneAllDimensionInfo(), `.
`, "Illegal config:", a, `.
`)), i(r));
      var g = o ? F(o) : null;
      o && !g && (process.env.NODE_ENV !== "production" && (r = O("Invalid parser name " + o + `.
`, "Illegal config:", a, `.
`)), i(r)), f.push({
        dimIdx: N.index,
        parser: g,
        comparator: new S(e, n)
      });
    });
    var p = s.sourceFormat;
    p !== _ && p !== I && (process.env.NODE_ENV !== "production" && (r = 'sourceFormat "' + p + '" is not supported yet'), i(r));
    for (var d = [], u = 0, V = s.count(); u < V; u++)
      d.push(s.getRawDataItem(u));
    return d.sort(function(a, m) {
      for (var e = 0; e < f.length; e++) {
        var o = f[e], n = s.retrieveValueFromItem(a, o.dimIdx), t = s.retrieveValueFromItem(m, o.dimIdx);
        o.parser && (n = o.parser(n), t = o.parser(t));
        var c = o.comparator.evaluate(n, t);
        if (c !== 0)
          return c;
      }
      return 0;
    }), {
      data: d
    };
  }
};
export {
  x as sortTransform
};
