// y-prosemirror@1.3.6/src/plugins/sync-plugin.js
import { createMutex as Q } from "../../../../../lib0@0.2.109/node_modules/lib0/mutex.js";
import { Slice as _, Fragment as T } from "../../../../../prosemirror-model@1.25.0/node_modules/prosemirror-model/dist/index.js";
import { Plugin as $, TextSelection as B, AllSelection as ee, NodeSelection as te } from "../../../../../prosemirror-state@1.4.3/node_modules/prosemirror-state/dist/index.js";
import { min as V, max as Y } from "../../../../../lib0@0.2.109/node_modules/lib0/math.js";
import { every as ne, keys as ie } from "../../../../../lib0@0.2.109/node_modules/lib0/object.js";
import { create as se } from "../../../../../lib0@0.2.109/node_modules/lib0/set.js";
import { simpleDiff as re } from "../../../../../lib0@0.2.109/node_modules/lib0/diff.js";
import { unexpectedCase as oe, methodUnimplemented as ce } from "../../../../../lib0@0.2.109/node_modules/lib0/error.js";
import { ySyncPluginKey as p, yUndoPluginKey as le } from "./keys.js";
import { snapshot as I, Doc as ae, applyUpdateV2 as q, findIndexSS as de, iterateDeletedStructs as z, typeListToArraySnapshot as K, Snapshot as J, Item as fe, UndoManager as he, createSnapshot as ue, isDeleted as pe, XmlElement as m, XmlText as b, Text as me, ContentString as ge, ContentFormat as ye, createDeleteSet as we } from "../../../../../yjs@13.6.27/node_modules/yjs/dist/yjs.js";
import { absolutePositionToRelativePosition as H, relativePositionToAbsolutePosition as P } from "../lib.js";
import { oneOf as Ce } from "../../../../../lib0@0.2.109/node_modules/lib0/random.js";
import { isBrowser as _e } from "../../../../../lib0@0.2.109/node_modules/lib0/environment.js";
import { doc as Te } from "../../../../../lib0@0.2.109/node_modules/lib0/dom.js";
import { timeout as W } from "../../../../../lib0@0.2.109/node_modules/lib0/eventloop.js";
import { setIfUndefined as Ae } from "../../../../../lib0@0.2.109/node_modules/lib0/map.js";
import { hashOfJSON as Ve } from "../utils.js";
const S = (t, e) => e === void 0 ? !t.deleted : e.sv.has(t.id.client) && /** @type {number} */
e.sv.get(t.id.client) > t.id.clock && !pe(e.ds, t.id), be = [{ light: "#ecd44433", dark: "#ecd444" }], Se = (t, e, n) => {
  if (!t.has(n)) {
    if (t.size < e.length) {
      const i = se();
      t.forEach((s) => i.add(s)), e = e.filter((s) => !i.has(s));
    }
    t.set(n, Ce(e));
  }
  return (
    /** @type {ColorDef} */
    t.get(n)
  );
}, Qe = (t, {
  colors: e = be,
  colorMapping: n = /* @__PURE__ */ new Map(),
  permanentUserData: i = null,
  onFirstRender: s = () => {
  },
  mapping: l
} = {}) => {
  let a = !1;
  const c = new ke(t, l), f = new $({
    props: {
      editable: (r) => {
        const d = p.getState(r);
        return d.snapshot == null && d.prevSnapshot == null;
      }
    },
    key: p,
    state: {
      /**
       * @returns {any}
       */
      init: (r, d) => ({
        type: t,
        doc: t.doc,
        binding: c,
        snapshot: null,
        prevSnapshot: null,
        isChangeOrigin: !1,
        isUndoRedoOperation: !1,
        addToHistory: !0,
        colors: e,
        colorMapping: n,
        permanentUserData: i
      }),
      apply: (r, d) => {
        const o = r.getMeta(p);
        if (o !== void 0) {
          d = Object.assign({}, d);
          for (const h in o)
            d[h] = o[h];
        }
        return d.addToHistory = r.getMeta("addToHistory") !== !1, d.isChangeOrigin = o !== void 0 && !!o.isChangeOrigin, d.isUndoRedoOperation = o !== void 0 && !!o.isChangeOrigin && !!o.isUndoRedoOperation, c.prosemirrorView !== null && o !== void 0 && (o.snapshot != null || o.prevSnapshot != null) && W(0, () => {
          c.prosemirrorView != null && (o.restore == null ? c._renderSnapshot(
            o.snapshot,
            o.prevSnapshot,
            d
          ) : (c._renderSnapshot(
            o.snapshot,
            o.snapshot,
            d
          ), delete d.restore, delete d.snapshot, delete d.prevSnapshot, c.mux(() => {
            c._prosemirrorChanged(
              c.prosemirrorView.state.doc
            );
          })));
        }), d;
      }
    },
    view: (r) => (c.initView(r), l == null && c._forceRerender(), s(), {
      update: () => {
        const d = f.getState(r.state);
        if (d.snapshot == null && d.prevSnapshot == null && // If the content doesn't change initially, we don't render anything to Yjs
        // If the content was cleared by a user action, we want to catch the change and
        // represent it in Yjs
        (a || r.state.doc.content.findDiffStart(
          r.state.doc.type.createAndFill().content
        ) !== null)) {
          if (a = !0, d.addToHistory === !1 && !d.isChangeOrigin) {
            const o = le.getState(r.state), h = o && o.undoManager;
            h && h.stopCapturing();
          }
          c.mux(() => {
            d.doc.transact((o) => {
              o.meta.set("addToHistory", d.addToHistory), c._prosemirrorChanged(r.state.doc);
            }, p);
          });
        }
      },
      destroy: () => {
        c.destroy();
      }
    })
  });
  return f;
}, xe = (t, e, n) => {
  if (e !== null && e.anchor !== null && e.head !== null)
    if (e.type === "all")
      t.setSelection(new ee(t.doc));
    else if (e.type === "node") {
      const i = P(
        n.doc,
        n.type,
        e.anchor,
        n.mapping
      );
      t.setSelection(te.create(t.doc, i));
    } else {
      const i = P(
        n.doc,
        n.type,
        e.anchor,
        n.mapping
      ), s = P(
        n.doc,
        n.type,
        e.head,
        n.mapping
      );
      if (i !== null && s !== null) {
        const l = B.between(t.doc.resolve(i), t.doc.resolve(s));
        t.setSelection(l);
      }
    }
}, D = (t, e) => ({
  type: (
    /** @type {any} */
    e.selection.jsonID
  ),
  anchor: H(
    e.selection.anchor,
    t.type,
    t.mapping
  ),
  head: H(
    e.selection.head,
    t.type,
    t.mapping
  )
});
class ke {
  /**
   * @param {Y.XmlFragment} yXmlFragment The bind source
   * @param {ProsemirrorMapping} mapping
   */
  constructor(e, n = /* @__PURE__ */ new Map()) {
    this.type = e, this.prosemirrorView = null, this.mux = Q(), this.mapping = n, this.isOMark = /* @__PURE__ */ new Map(), this._observeFunction = this._typeChanged.bind(this), this.doc = e.doc, this.beforeTransactionSelection = null, this.beforeAllTransactions = () => {
      this.beforeTransactionSelection === null && this.prosemirrorView != null && (this.beforeTransactionSelection = D(
        this,
        this.prosemirrorView.state
      ));
    }, this.afterAllTransactions = () => {
      this.beforeTransactionSelection = null;
    }, this._domSelectionInView = null;
  }
  /**
   * Create a transaction for changing the prosemirror state.
   *
   * @returns
   */
  get _tr() {
    return this.prosemirrorView.state.tr.setMeta("addToHistory", !1);
  }
  _isLocalCursorInView() {
    return this.prosemirrorView.hasFocus() ? (_e && this._domSelectionInView === null && (W(0, () => {
      this._domSelectionInView = null;
    }), this._domSelectionInView = this._isDomSelectionInView()), this._domSelectionInView) : !1;
  }
  _isDomSelectionInView() {
    const e = this.prosemirrorView._root.getSelection();
    if (e == null || e.anchorNode == null) return !1;
    const n = this.prosemirrorView._root.createRange();
    n.setStart(e.anchorNode, e.anchorOffset), n.setEnd(e.focusNode, e.focusOffset), n.getClientRects().length === 0 && n.startContainer && n.collapsed && n.selectNodeContents(n.startContainer);
    const s = n.getBoundingClientRect(), l = Te.documentElement;
    return s.bottom >= 0 && s.right >= 0 && s.left <= (window.innerWidth || l.clientWidth || 0) && s.top <= (window.innerHeight || l.clientHeight || 0);
  }
  /**
   * @param {Y.Snapshot} snapshot
   * @param {Y.Snapshot} prevSnapshot
   */
  renderSnapshot(e, n) {
    n || (n = ue(we(), /* @__PURE__ */ new Map())), this.prosemirrorView.dispatch(
      this._tr.setMeta(p, { snapshot: e, prevSnapshot: n })
    );
  }
  unrenderSnapshot() {
    this.mapping.clear(), this.mux(() => {
      const e = this.type.toArray().map(
        (i) => A(
          /** @type {Y.XmlElement} */
          i,
          this.prosemirrorView.state.schema,
          this
        )
      ).filter((i) => i !== null), n = this._tr.replace(
        0,
        this.prosemirrorView.state.doc.content.size,
        new _(T.from(e), 0, 0)
      );
      n.setMeta(p, { snapshot: null, prevSnapshot: null }), this.prosemirrorView.dispatch(n);
    });
  }
  _forceRerender() {
    this.mapping.clear(), this.mux(() => {
      const e = this.beforeTransactionSelection !== null ? null : this.prosemirrorView.state.selection, n = this.type.toArray().map(
        (s) => A(
          /** @type {Y.XmlElement} */
          s,
          this.prosemirrorView.state.schema,
          this
        )
      ).filter((s) => s !== null), i = this._tr.replace(
        0,
        this.prosemirrorView.state.doc.content.size,
        new _(T.from(n), 0, 0)
      );
      if (e) {
        const s = V(Y(e.anchor, 0), i.doc.content.size), l = V(Y(e.head, 0), i.doc.content.size);
        i.setSelection(B.create(i.doc, s, l));
      }
      this.prosemirrorView.dispatch(
        i.setMeta(p, { isChangeOrigin: !0, binding: this })
      );
    });
  }
  /**
   * @param {Y.Snapshot|Uint8Array} snapshot
   * @param {Y.Snapshot|Uint8Array} prevSnapshot
   * @param {Object} pluginState
   */
  _renderSnapshot(e, n, i) {
    let s = this.doc, l = this.type;
    if (e || (e = I(this.doc)), e instanceof Uint8Array || n instanceof Uint8Array)
      if ((!(e instanceof Uint8Array) || !(n instanceof Uint8Array)) && oe(), s = new ae({ gc: !1 }), q(s, n), n = I(s), q(s, e), e = I(s), l._item === null) {
        const a = Array.from(this.doc.share.keys()).find(
          (c) => this.doc.share.get(c) === this.type
        );
        l = s.getXmlFragment(a);
      } else {
        const a = s.store.clients.get(l._item.id.client) ?? [], c = de(
          a,
          l._item.id.clock
        );
        l = /** @type {Y.XmlFragment} */
        /** @type {Y.ContentType} */
        /** @type {Y.Item} */
        a[c].content.type;
      }
    this.mapping.clear(), this.mux(() => {
      s.transact((a) => {
        const c = i.permanentUserData;
        c && c.dss.forEach((o) => {
          z(a, o, (h) => {
          });
        });
        const f = (o, h) => {
          const u = o === "added" ? c.getUserByClientId(h.client) : c.getUserByDeletedId(h);
          return {
            user: u,
            type: o,
            color: Se(
              i.colorMapping,
              i.colors,
              u
            )
          };
        }, r = K(
          l,
          new J(n.ds, e.sv)
        ).map((o) => !o._item.deleted || S(o._item, e) || S(o._item, n) ? A(
          o,
          this.prosemirrorView.state.schema,
          { mapping: /* @__PURE__ */ new Map(), isOMark: /* @__PURE__ */ new Map() },
          e,
          n,
          f
        ) : null).filter((o) => o !== null), d = this._tr.replace(
          0,
          this.prosemirrorView.state.doc.content.size,
          new _(T.from(r), 0, 0)
        );
        this.prosemirrorView.dispatch(
          d.setMeta(p, { isChangeOrigin: !0 })
        );
      }, p);
    });
  }
  /**
   * @param {Array<Y.YEvent<any>>} events
   * @param {Y.Transaction} transaction
   */
  _typeChanged(e, n) {
    if (this.prosemirrorView == null) return;
    const i = p.getState(this.prosemirrorView.state);
    if (e.length === 0 || i.snapshot != null || i.prevSnapshot != null) {
      this.renderSnapshot(i.snapshot, i.prevSnapshot);
      return;
    }
    this.mux(() => {
      const s = (c, f) => this.mapping.delete(f);
      z(
        n,
        n.deleteSet,
        (c) => {
          if (c.constructor === fe) {
            const f = (
              /** @type {Y.ContentType} */
              /** @type {Y.Item} */
              c.content.type
            );
            f && this.mapping.delete(f);
          }
        }
      ), n.changed.forEach(s), n.changedParentTypes.forEach(s);
      const l = this.type.toArray().map(
        (c) => X(
          /** @type {Y.XmlElement | Y.XmlHook} */
          c,
          this.prosemirrorView.state.schema,
          this
        )
      ).filter((c) => c !== null);
      let a = this._tr.replace(
        0,
        this.prosemirrorView.state.doc.content.size,
        new _(T.from(l), 0, 0)
      );
      xe(a, this.beforeTransactionSelection, this), a = a.setMeta(p, { isChangeOrigin: !0, isUndoRedoOperation: n.origin instanceof he }), this.beforeTransactionSelection !== null && this._isLocalCursorInView() && a.scrollIntoView(), this.prosemirrorView.dispatch(a);
    });
  }
  /**
   * @param {import('prosemirror-model').Node} doc
   */
  _prosemirrorChanged(e) {
    this.doc.transact(() => {
      v(this.doc, this.type, e, this), this.beforeTransactionSelection = D(
        this,
        this.prosemirrorView.state
      );
    }, p);
  }
  /**
   * View is ready to listen to changes. Register observers.
   * @param {any} prosemirrorView
   */
  initView(e) {
    this.prosemirrorView != null && this.destroy(), this.prosemirrorView = e, this.doc.on("beforeAllTransactions", this.beforeAllTransactions), this.doc.on("afterAllTransactions", this.afterAllTransactions), this.type.observeDeep(this._observeFunction);
  }
  destroy() {
    this.prosemirrorView != null && (this.prosemirrorView = null, this.type.unobserveDeep(this._observeFunction), this.doc.off("beforeAllTransactions", this.beforeAllTransactions), this.doc.off("afterAllTransactions", this.afterAllTransactions));
  }
}
const X = (t, e, n, i, s, l) => {
  const a = (
    /** @type {PModel.Node} */
    n.mapping.get(t)
  );
  if (a === void 0) {
    if (t instanceof m)
      return A(
        t,
        e,
        n,
        i,
        s,
        l
      );
    throw ce();
  }
  return a;
}, A = (t, e, n, i, s, l) => {
  const a = [], c = (f) => {
    if (f instanceof m) {
      const r = X(
        f,
        e,
        n,
        i,
        s,
        l
      );
      r !== null && a.push(r);
    } else {
      const r = (
        /** @type {Y.ContentType} */
        f._item.right?.content?.type
      );
      r instanceof me && !r._item.deleted && r._item.id.client === r.doc.clientID && (f.applyDelta([
        { retain: f.length },
        ...r.toDelta()
      ]), r.doc.transact((o) => {
        r._item.delete(o);
      }));
      const d = Oe(
        f,
        e,
        n,
        i,
        s,
        l
      );
      d !== null && d.forEach((o) => {
        o !== null && a.push(o);
      });
    }
  };
  i === void 0 || s === void 0 ? t.toArray().forEach(c) : K(t, new J(s.ds, i.sv)).forEach(c);
  try {
    const f = t.getAttributes(i);
    i !== void 0 && (S(
      /** @type {Y.Item} */
      t._item,
      i
    ) ? S(
      /** @type {Y.Item} */
      t._item,
      s
    ) || (f.ychange = l ? l(
      "added",
      /** @type {Y.Item} */
      t._item.id
    ) : { type: "added" }) : f.ychange = l ? l(
      "removed",
      /** @type {Y.Item} */
      t._item.id
    ) : { type: "removed" });
    const r = e.node(t.nodeName, f, a);
    return n.mapping.set(t, r), r;
  } catch {
    return t.doc.transact((r) => {
      t._item.delete(r);
    }, p), n.mapping.delete(t), null;
  }
}, Oe = (t, e, n, i, s, l) => {
  const a = [], c = t.toDelta(i, s, l);
  try {
    for (let f = 0; f < c.length; f++) {
      const r = c[f];
      a.push(e.text(r.insert, ve(r.attributes, e)));
    }
  } catch {
    return t.doc.transact((r) => {
      t._item.delete(r);
    }, p), null;
  }
  return a;
}, Me = (t, e) => {
  const n = new b(), i = t.map((s) => ({
    // @ts-ignore
    insert: s.text,
    attributes: G(s.marks, e)
  }));
  return n.applyDelta(i), e.mapping.set(n, t), n;
}, Ee = (t, e) => {
  const n = new m(t.type.name);
  for (const i in t.attrs) {
    const s = t.attrs[i];
    s !== null && i !== "ychange" && n.setAttribute(i, s);
  }
  return n.insert(
    0,
    k(t).map(
      (i) => N(i, e)
    )
  ), e.mapping.set(n, t), n;
}, N = (t, e) => t instanceof Array ? Me(t, e) : Ee(t, e), j = (t) => typeof t == "object" && t !== null, R = (t, e) => {
  const n = Object.keys(t).filter((s) => t[s] !== null);
  let i = n.length === (e == null ? 0 : Object.keys(e).filter((s) => e[s] !== null).length);
  for (let s = 0; s < n.length && i; s++) {
    const l = n[s], a = t[l], c = e[l];
    i = l === "ychange" || a === c || j(a) && j(c) && R(a, c);
  }
  return i;
}, k = (t) => {
  const e = t.content.content, n = [];
  for (let i = 0; i < e.length; i++) {
    const s = e[i];
    if (s.isText) {
      const l = [];
      for (let a = e[i]; i < e.length && a.isText; a = e[++i])
        l.push(a);
      i--, n.push(l);
    } else
      n.push(s);
  }
  return n;
}, F = (t, e) => {
  const n = t.toDelta();
  return n.length === e.length && n.every(
    /** @type {(d:any,i:number) => boolean} */
    (i, s) => i.insert === /** @type {any} */
    e[s].text && ie(i.attributes || {}).length === e[s].marks.length && ne(i.attributes, (l, a) => {
      const c = Z(a), f = e[s].marks;
      return R(l, f.find(
        /** @param {any} mark */
        (r) => r.type.name === c
      )?.attrs);
    })
  );
}, g = (t, e) => {
  if (t instanceof m && !(e instanceof Array) && U(t, e)) {
    const n = k(e);
    return t._length === n.length && R(t.getAttributes(), e.attrs) && t.toArray().every(
      (i, s) => g(i, n[s])
    );
  }
  return t instanceof b && e instanceof Array && F(t, e);
}, x = (t, e) => t === e || t instanceof Array && e instanceof Array && t.length === e.length && t.every(
  (n, i) => e[i] === n
), L = (t, e, n) => {
  const i = t.toArray(), s = k(e), l = s.length, a = i.length, c = V(a, l);
  let f = 0, r = 0, d = !1;
  for (; f < c; f++) {
    const o = i[f], h = s[f];
    if (x(n.mapping.get(o), h))
      d = !0;
    else if (!g(o, h))
      break;
  }
  for (; f + r < c; r++) {
    const o = i[a - r - 1], h = s[l - r - 1];
    if (x(n.mapping.get(o), h))
      d = !0;
    else if (!g(o, h))
      break;
  }
  return {
    equalityFactor: f + r,
    foundMappedChild: d
  };
}, Ie = (t) => {
  let e = "", n = t._start;
  const i = {};
  for (; n !== null; )
    n.deleted || (n.countable && n.content instanceof ge ? e += n.content.str : n.content instanceof ye && (i[n.content.key] = null)), n = n.right;
  return {
    str: e,
    nAttrs: i
  };
}, Pe = (t, e, n) => {
  n.mapping.set(t, e);
  const { nAttrs: i, str: s } = Ie(t), l = e.map((r) => ({
    insert: (
      /** @type {any} */
      r.text
    ),
    attributes: Object.assign({}, i, G(r.marks, n))
  })), { insert: a, remove: c, index: f } = re(
    s,
    l.map((r) => r.insert).join("")
  );
  t.delete(f, c), t.insert(f, a), t.applyDelta(
    l.map((r) => ({ retain: r.insert.length, attributes: r.attributes }))
  );
}, Ne = /(.*)(--[a-zA-Z0-9+/=]{8})$/, Z = (t) => Ne.exec(t)?.[1] ?? t, ve = (t, e) => {
  const n = [];
  for (const i in t)
    n.push(e.mark(Z(i), t[i]));
  return n;
}, G = (t, e) => {
  const n = {};
  return t.forEach((i) => {
    if (i.type.name !== "ychange") {
      const s = Ae(e.isOMark, i.type, () => !i.type.excludes(i.type));
      n[s ? `${i.type.name}--${Ve(i.toJSON())}` : i.type.name] = i.attrs;
    }
  }), n;
}, v = (t, e, n, i) => {
  if (e instanceof m && e.nodeName !== n.type.name)
    throw new Error("node name mismatch!");
  if (i.mapping.set(e, n), e instanceof m) {
    const o = e.getAttributes(), h = n.attrs;
    for (const u in h)
      h[u] !== null ? o[u] !== h[u] && u !== "ychange" && e.setAttribute(u, h[u]) : e.removeAttribute(u);
    for (const u in o)
      h[u] === void 0 && e.removeAttribute(u);
  }
  const s = k(n), l = s.length, a = e.toArray(), c = a.length, f = V(l, c);
  let r = 0, d = 0;
  for (; r < f; r++) {
    const o = a[r], h = s[r];
    if (!x(i.mapping.get(o), h))
      if (g(o, h))
        i.mapping.set(o, h);
      else
        break;
  }
  for (; d + r + 1 < f; d++) {
    const o = a[c - d - 1], h = s[l - d - 1];
    if (!x(i.mapping.get(o), h))
      if (g(o, h))
        i.mapping.set(o, h);
      else
        break;
  }
  t.transact(() => {
    for (; c - r - d > 0 && l - r - d > 0; ) {
      const h = a[r], u = s[r], y = a[c - d - 1], O = s[l - d - 1];
      if (h instanceof b && u instanceof Array)
        F(h, u) || Pe(h, u, i), r += 1;
      else {
        let w = h instanceof m && U(h, u), C = y instanceof m && U(y, O);
        if (w && C) {
          const M = L(
            /** @type {Y.XmlElement} */
            h,
            /** @type {PModel.Node} */
            u,
            i
          ), E = L(
            /** @type {Y.XmlElement} */
            y,
            /** @type {PModel.Node} */
            O,
            i
          );
          M.foundMappedChild && !E.foundMappedChild ? C = !1 : !M.foundMappedChild && E.foundMappedChild || M.equalityFactor < E.equalityFactor ? w = !1 : C = !1;
        }
        w ? (v(
          t,
          /** @type {Y.XmlFragment} */
          h,
          /** @type {PModel.Node} */
          u,
          i
        ), r += 1) : C ? (v(
          t,
          /** @type {Y.XmlFragment} */
          y,
          /** @type {PModel.Node} */
          O,
          i
        ), d += 1) : (i.mapping.delete(e.get(r)), e.delete(r, 1), e.insert(r, [
          N(u, i)
        ]), r += 1);
      }
    }
    const o = c - r - d;
    if (c === 1 && l === 0 && a[0] instanceof b ? (i.mapping.delete(a[0]), a[0].delete(0, a[0].length)) : o > 0 && (e.slice(r, r + o).forEach((h) => i.mapping.delete(h)), e.delete(r, o)), r + d < l) {
      const h = [];
      for (let u = r; u < l - d; u++)
        h.push(N(s[u], i));
      e.insert(r, h);
    }
  }, p);
}, U = (t, e) => !(e instanceof Array) && t.nodeName === e.type.name;
export {
  ke as ProsemirrorBinding,
  ve as attributesToMarks,
  A as createNodeFromYElement,
  D as getRelativeSelection,
  S as isVisible,
  v as updateYFragment,
  Qe as ySyncPlugin,
  Z as yattr2markname
};
