// y-prosemirror@1.3.6/src/plugins/undo-plugin.js
import { Plugin as u } from "../../../../../prosemirror-state@1.4.3/node_modules/prosemirror-state/dist/index.js";
import { getRelativeSelection as S } from "./sync-plugin.js";
import { UndoManager as y, Item as f, ContentType as h, Text as m, XmlElement as k } from "../../../../../yjs@13.6.27/node_modules/yjs/dist/yjs.js";
import { yUndoPluginKey as d, ySyncPluginKey as c } from "./keys.js";
const R = (t) => d.getState(t)?.undoManager?.undo() != null, w = (t) => d.getState(t)?.undoManager?.redo() != null, O = /* @__PURE__ */ new Set(["paragraph"]), b = (t, i) => !(t instanceof f) || !(t.content instanceof h) || !(t.content.type instanceof m || t.content.type instanceof k && i.has(t.content.type.nodeName)) || t.content.type._length === 0, x = ({ protectedNodes: t = O, trackedOrigins: i = [], undoManager: p = null } = {}) => new u({
  key: d,
  state: {
    init: (s, n) => {
      const a = c.getState(n), o = p || new y(a.type, {
        trackedOrigins: new Set([c].concat(i)),
        deleteFilter: (e) => b(e, t),
        captureTransaction: (e) => e.meta.get("addToHistory") !== !1
      });
      return {
        undoManager: o,
        prevSel: null,
        hasUndoOps: o.undoStack.length > 0,
        hasRedoOps: o.redoStack.length > 0
      };
    },
    apply: (s, n, a, o) => {
      const e = c.getState(o).binding, r = n.undoManager, g = r.undoStack.length > 0, l = r.redoStack.length > 0;
      return e ? {
        undoManager: r,
        prevSel: S(e, a),
        hasUndoOps: g,
        hasRedoOps: l
      } : g !== n.hasUndoOps || l !== n.hasRedoOps ? Object.assign({}, n, {
        hasUndoOps: r.undoStack.length > 0,
        hasRedoOps: r.redoStack.length > 0
      }) : n;
    }
  },
  view: (s) => {
    const n = c.getState(s.state), a = d.getState(s.state).undoManager;
    return a.on("stack-item-added", ({ stackItem: o }) => {
      const e = n.binding;
      e && o.meta.set(e, d.getState(s.state).prevSel);
    }), a.on("stack-item-popped", ({ stackItem: o }) => {
      const e = n.binding;
      e && (e.beforeTransactionSelection = o.meta.get(e) || e.beforeTransactionSelection);
    }), {
      destroy: () => {
        a.destroy();
      }
    };
  }
});
export {
  b as defaultDeleteFilter,
  O as defaultProtectedNodes,
  w as redo,
  R as undo,
  x as yUndoPlugin
};
