// screenfull@5.2.0/dist/screenfull.js
import { __module as o } from "../../../../../../_virtual/screenfull2.js";
var h;
function b() {
  return h ? o.exports : (h = 1, (function(s) {
    (function() {
      var t = typeof window < "u" && typeof window.document < "u" ? window.document : {}, f = s.exports, l = (function() {
        for (var e, r = [
          [
            "requestFullscreen",
            "exitFullscreen",
            "fullscreenElement",
            "fullscreenEnabled",
            "fullscreenchange",
            "fullscreenerror"
          ],
          // New WebKit
          [
            "webkitRequestFullscreen",
            "webkitExitFullscreen",
            "webkitFullscreenElement",
            "webkitFullscreenEnabled",
            "webkitfullscreenchange",
            "webkitfullscreenerror"
          ],
          // Old WebKit
          [
            "webkitRequestFullScreen",
            "webkitCancelFullScreen",
            "webkitCurrentFullScreenElement",
            "webkitCancelFullScreen",
            "webkitfullscreenchange",
            "webkitfullscreenerror"
          ],
          [
            "mozRequestFullScreen",
            "mozCancelFullScreen",
            "mozFullScreenElement",
            "mozFullScreenEnabled",
            "mozfullscreenchange",
            "mozfullscreenerror"
          ],
          [
            "msRequestFullscreen",
            "msExitFullscreen",
            "msFullscreenElement",
            "msFullscreenEnabled",
            "MSFullscreenChange",
            "MSFullscreenError"
          ]
        ], n = 0, u = r.length, i = {}; n < u; n++)
          if (e = r[n], e && e[1] in t) {
            for (n = 0; n < e.length; n++)
              i[r[0][n]] = e[n];
            return i;
          }
        return !1;
      })(), a = {
        change: l.fullscreenchange,
        error: l.fullscreenerror
      }, c = {
        request: function(e, r) {
          return new Promise((function(n, u) {
            var i = (function() {
              this.off("change", i), n();
            }).bind(this);
            this.on("change", i), e = e || t.documentElement;
            var m = e[l.requestFullscreen](r);
            m instanceof Promise && m.then(i).catch(u);
          }).bind(this));
        },
        exit: function() {
          return new Promise((function(e, r) {
            if (!this.isFullscreen) {
              e();
              return;
            }
            var n = (function() {
              this.off("change", n), e();
            }).bind(this);
            this.on("change", n);
            var u = t[l.exitFullscreen]();
            u instanceof Promise && u.then(n).catch(r);
          }).bind(this));
        },
        toggle: function(e, r) {
          return this.isFullscreen ? this.exit() : this.request(e, r);
        },
        onchange: function(e) {
          this.on("change", e);
        },
        onerror: function(e) {
          this.on("error", e);
        },
        on: function(e, r) {
          var n = a[e];
          n && t.addEventListener(n, r, !1);
        },
        off: function(e, r) {
          var n = a[e];
          n && t.removeEventListener(n, r, !1);
        },
        raw: l
      };
      if (!l) {
        f ? s.exports = { isEnabled: !1 } : window.screenfull = { isEnabled: !1 };
        return;
      }
      Object.defineProperties(c, {
        isFullscreen: {
          get: function() {
            return !!t[l.fullscreenElement];
          }
        },
        element: {
          enumerable: !0,
          get: function() {
            return t[l.fullscreenElement];
          }
        },
        isEnabled: {
          enumerable: !0,
          get: function() {
            return !!t[l.fullscreenEnabled];
          }
        }
      }), f ? s.exports = c : window.screenfull = c;
    })();
  })(o), o.exports);
}
export {
  b as __require
};
