// @tiptap/extension-details@2.23.0/dist/index.js
import { Node as M, isActive as N, findChildren as C, defaultBlockAt as T, findParentNode as w, mergeAttributes as S } from "../../../../../@tiptap_core@2.24.0_@tiptap_pm@2.24.0/node_modules/@tiptap/core/dist/index.js";
import { Plugin as L, PluginKey as O, TextSelection as D, Selection as E } from "../../../../../prosemirror-state@1.4.3/node_modules/prosemirror-state/dist/index.js";
import { GapCursor as b } from "../../../../../prosemirror-gapcursor@1.3.2/node_modules/prosemirror-gapcursor/dist/index.js";
const v = (t, i) => i.view.domAtPos(t).node.offsetParent !== null, V = (t, i, o) => {
  for (let s = t.depth; s > 0; s -= 1) {
    const e = t.node(s), r = i(e), l = v(t.start(s), o);
    if (r && l)
      return {
        pos: s > 0 ? t.before(s) : 0,
        start: t.start(s),
        depth: s,
        node: e
      };
  }
}, A = (t, i) => {
  const { state: o, view: s, extensionManager: e } = t, { schema: r, selection: l } = o, { empty: a, $anchor: u } = l, n = !!e.extensions.find((m) => m.name === "gapCursor");
  if (!a || u.parent.type !== r.nodes.detailsSummary || !n || i === "right" && u.parentOffset !== u.parent.nodeSize - 2)
    return !1;
  const c = w((m) => m.type === r.nodes.details)(l);
  if (!c)
    return !1;
  const d = C(c.node, (m) => m.type === r.nodes.detailsContent);
  if (!d.length || v(c.start + d[0].pos + 1, t))
    return !1;
  const p = o.doc.resolve(c.pos + c.node.nodeSize), h = b.findFrom(p, 1, !1);
  if (!h)
    return !1;
  const { tr: g } = o, y = new b(h);
  return g.setSelection(y), g.scrollIntoView(), s.dispatch(g), !0;
}, H = M.create({
  name: "details",
  content: "detailsSummary detailsContent",
  group: "block",
  defining: !0,
  isolating: !0,
  allowGapCursor: !1,
  addOptions() {
    return {
      persist: !1,
      openClassName: "is-open",
      HTMLAttributes: {}
    };
  },
  addAttributes() {
    return this.options.persist ? {
      open: {
        default: !1,
        parseHTML: (t) => t.hasAttribute("open"),
        renderHTML: ({ open: t }) => t ? { open: "" } : {}
      }
    } : [];
  },
  parseHTML() {
    return [
      {
        tag: "details"
      }
    ];
  },
  renderHTML({ HTMLAttributes: t }) {
    return [
      "details",
      S(this.options.HTMLAttributes, t),
      0
    ];
  },
  addNodeView() {
    return ({ editor: t, getPos: i, node: o, HTMLAttributes: s }) => {
      const e = document.createElement("div"), r = S(this.options.HTMLAttributes, s, {
        "data-type": this.name
      });
      Object.entries(r).forEach(([n, c]) => e.setAttribute(n, c));
      const l = document.createElement("button");
      l.type = "button", e.append(l);
      const a = document.createElement("div");
      e.append(a);
      const u = (n) => {
        if (n !== void 0)
          if (n) {
            if (e.classList.contains(this.options.openClassName))
              return;
            e.classList.add(this.options.openClassName);
          } else {
            if (!e.classList.contains(this.options.openClassName))
              return;
            e.classList.remove(this.options.openClassName);
          }
        else
          e.classList.toggle(this.options.openClassName);
        const c = new Event("toggleDetailsContent"), d = a.querySelector(':scope > div[data-type="detailsContent"]');
        d?.dispatchEvent(c);
      };
      return o.attrs.open && setTimeout(() => u()), l.addEventListener("click", () => {
        if (u(), !this.options.persist) {
          t.commands.focus(void 0, { scrollIntoView: !1 });
          return;
        }
        if (t.isEditable && typeof i == "function") {
          const { from: n, to: c } = t.state.selection;
          t.chain().command(({ tr: d }) => {
            const f = i(), p = d.doc.nodeAt(f);
            return p?.type !== this.type ? !1 : (d.setNodeMarkup(f, void 0, {
              open: !p.attrs.open
            }), !0);
          }).setTextSelection({
            from: n,
            to: c
          }).focus(void 0, { scrollIntoView: !1 }).run();
        }
      }), {
        dom: e,
        contentDOM: a,
        ignoreMutation(n) {
          return n.type === "selection" ? !1 : !e.contains(n.target) || e === n.target;
        },
        update: (n) => n.type !== this.type ? !1 : (n.attrs.open !== void 0 && u(n.attrs.open), !0)
      };
    };
  },
  addCommands() {
    return {
      setDetails: () => ({ state: t, chain: i }) => {
        var o;
        const { schema: s, selection: e } = t, { $from: r, $to: l } = e, a = r.blockRange(l);
        if (!a)
          return !1;
        const u = t.doc.slice(a.start, a.end);
        if (!s.nodes.detailsContent.contentMatch.matchFragment(u.content))
          return !1;
        const c = ((o = u.toJSON()) === null || o === void 0 ? void 0 : o.content) || [];
        return i().insertContentAt({ from: a.start, to: a.end }, {
          type: this.name,
          content: [
            {
              type: "detailsSummary"
            },
            {
              type: "detailsContent",
              content: c
            }
          ]
        }).setTextSelection(a.start + 2).run();
      },
      unsetDetails: () => ({ state: t, chain: i }) => {
        const { selection: o, schema: s } = t, e = w((m) => m.type === this.type)(o);
        if (!e)
          return !1;
        const r = C(e.node, (m) => m.type === s.nodes.detailsSummary), l = C(e.node, (m) => m.type === s.nodes.detailsContent);
        if (!r.length || !l.length)
          return !1;
        const a = r[0], u = l[0], n = e.pos, c = t.doc.resolve(n), d = n + e.node.nodeSize, f = { from: n, to: d }, p = u.node.content.toJSON() || [], h = c.parent.type.contentMatch.defaultType, y = [
          h?.create(null, a.node.content).toJSON(),
          ...p
        ];
        return i().insertContentAt(f, y).setTextSelection(n + 1).run();
      }
    };
  },
  addKeyboardShortcuts() {
    return {
      Backspace: () => {
        const { schema: t, selection: i } = this.editor.state, { empty: o, $anchor: s } = i;
        return !o || s.parent.type !== t.nodes.detailsSummary ? !1 : s.parentOffset !== 0 ? this.editor.commands.command(({ tr: e }) => {
          const r = s.pos - 1, l = s.pos;
          return e.delete(r, l), !0;
        }) : this.editor.commands.unsetDetails();
      },
      // Creates a new node below it if it is closed.
      // Otherwise inside `DetailsContent`.
      Enter: ({ editor: t }) => {
        const { state: i, view: o } = t, { schema: s, selection: e } = i, { $head: r } = e;
        if (r.parent.type !== s.nodes.detailsSummary)
          return !1;
        const l = v(r.after() + 1, t), a = l ? i.doc.nodeAt(r.after()) : r.node(-2);
        if (!a)
          return !1;
        const u = l ? 0 : r.indexAfter(-1), n = T(a.contentMatchAt(u));
        if (!n || !a.canReplaceWith(u, u, n))
          return !1;
        const c = n.createAndFill();
        if (!c)
          return !1;
        const d = l ? r.after() + 1 : r.after(-1), f = i.tr.replaceWith(d, d, c), p = f.doc.resolve(d), h = E.near(p, 1);
        return f.setSelection(h), f.scrollIntoView(), o.dispatch(f), !0;
      },
      // The default gapcursor implementation can’t handle hidden content, so we need to fix this.
      ArrowRight: ({ editor: t }) => A(t, "right"),
      // The default gapcursor implementation can’t handle hidden content, so we need to fix this.
      ArrowDown: ({ editor: t }) => A(t, "down")
    };
  },
  addProseMirrorPlugins() {
    return [
      // This plugin prevents text selections within the hidden content in `DetailsContent`.
      // The cursor is moved to the next visible position.
      new L({
        key: new O("detailsSelection"),
        appendTransaction: (t, i, o) => {
          const { editor: s, type: e } = this;
          if (!t.some((y) => y.selectionSet) || !i.selection.empty || !o.selection.empty || !N(o, e.name))
            return;
          const { $from: a } = o.selection;
          if (v(a.pos, s))
            return;
          const n = V(a, (y) => y.type === e, s);
          if (!n)
            return;
          const c = C(n.node, (y) => y.type === o.schema.nodes.detailsSummary);
          if (!c.length)
            return;
          const d = c[0], p = (i.selection.from < o.selection.from ? "forward" : "backward") === "forward" ? n.start + d.pos : n.pos + d.pos + d.node.nodeSize, h = D.create(o.doc, p);
          return o.tr.setSelection(h);
        }
      })
    ];
  }
});
export {
  H as Details,
  H as default
};
