// echarts@6.0.0/lib/chart/gauge/GaugeView.js
import { __extends as it } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import nt from "./PointerPath.js";
import { setStatesStylesFromModel as Q, toggleHoverEmphasis as U } from "../../util/states.js";
import { createTextStyle as Y, setLabelValueAnimation as lt, animateLabelValue as vt } from "../../label/labelStyle.js";
import gt from "../../view/Chart.js";
import { parsePercent as M, round as ht, linearMap as W } from "../../util/number.js";
import $ from "../../util/shape/sausage.js";
import { createSymbol as F } from "../../util/symbol.js";
import ut from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Image.js";
import { each as mt, extend as ot, isString as ct, isFunction as st, isNumber as dt } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { setCommonECData as j } from "../../util/innerStore.js";
import { normalizeArcAngles as ft } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/PathProxy.js";
import I from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Sector.js";
import tt from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Line.js";
import q from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Text.js";
import et from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import { initProps as at, updateProps as rt } from "../../animation/basicTransition.js";
function yt(O, c) {
  var t = O.get("center"), b = c.getWidth(), y = c.getHeight(), m = Math.min(b, y), a = M(t[0], c.getWidth()), A = M(t[1], c.getHeight()), s = M(O.get("radius"), m / 2);
  return {
    cx: a,
    cy: A,
    r: s
  };
}
function B(O, c) {
  var t = O == null ? "" : O + "";
  return c && (ct(c) ? t = c.replace("{value}", t) : st(c) && (t = c(O))), t;
}
var Tt = (
  /** @class */
  (function(O) {
    it(c, O);
    function c() {
      var t = O !== null && O.apply(this, arguments) || this;
      return t.type = c.type, t;
    }
    return c.prototype.render = function(t, b, y) {
      this.group.removeAll();
      var m = t.get(["axisLine", "lineStyle", "color"]), a = yt(t, y);
      this._renderMain(t, b, y, m, a), this._data = t.getData();
    }, c.prototype.dispose = function() {
    }, c.prototype._renderMain = function(t, b, y, m, a) {
      var A = this.group, s = t.get("clockwise"), k = -t.get("startAngle") / 180 * Math.PI, d = -t.get("endAngle") / 180 * Math.PI, _ = t.getModel("axisLine"), L = _.get("roundCap"), V = L ? $ : I, u = _.get("show"), z = _.getModel("lineStyle"), D = z.get("width"), g = [k, d];
      ft(g, !s), k = g[0], d = g[1];
      for (var r = d - k, p = k, S = [], N = 0; u && N < m.length; N++) {
        var x = Math.min(Math.max(m[N][0], 0), 1);
        d = k + r * x;
        var E = new V({
          shape: {
            startAngle: p,
            endAngle: d,
            cx: a.cx,
            cy: a.cy,
            clockwise: s,
            r0: a.r - D,
            r: a.r
          },
          silent: !0
        });
        E.setStyle({
          fill: m[N][1]
        }), E.setStyle(z.getLineStyle(
          // Because we use sector to simulate arc
          // so the properties for stroking are useless
          ["color", "width"]
        )), S.push(E), p = d;
      }
      S.reverse(), mt(S, function(C) {
        return A.add(C);
      });
      var P = function(C) {
        if (C <= 0)
          return m[0][1];
        var e;
        for (e = 0; e < m.length; e++)
          if (m[e][0] >= C && (e === 0 ? 0 : m[e - 1][0]) < C)
            return m[e][1];
        return m[e - 1][1];
      };
      this._renderTicks(t, b, y, P, a, k, d, s, D), this._renderTitleAndDetail(t, b, y, P, a), this._renderAnchor(t, a), this._renderPointer(t, b, y, P, a, k, d, s, D);
    }, c.prototype._renderTicks = function(t, b, y, m, a, A, s, k, d) {
      for (var _ = this.group, L = a.cx, V = a.cy, u = a.r, z = +t.get("min"), D = +t.get("max"), g = t.getModel("splitLine"), r = t.getModel("axisTick"), p = t.getModel("axisLabel"), S = t.get("splitNumber"), N = r.get("splitNumber"), x = M(g.get("length"), u), E = M(r.get("length"), u), P = A, C = (s - A) / S, e = C / N, h = g.getModel("lineStyle").getLineStyle(), f = r.getModel("lineStyle").getLineStyle(), o = g.get("distance"), n, l, i = 0; i <= S; i++) {
        if (n = Math.cos(P), l = Math.sin(P), g.get("show")) {
          var v = o ? o + d : d, w = new tt({
            shape: {
              x1: n * (u - v) + L,
              y1: l * (u - v) + V,
              x2: n * (u - x - v) + L,
              y2: l * (u - x - v) + V
            },
            style: h,
            silent: !0
          });
          h.stroke === "auto" && w.setStyle({
            stroke: m(i / S)
          }), _.add(w);
        }
        if (p.get("show")) {
          var v = p.get("distance") + o, G = B(ht(i / S * (D - z) + z), p.get("formatter")), T = m(i / S), H = n * (u - x - v) + L, X = l * (u - x - v) + V, K = p.get("rotate"), R = 0;
          K === "radial" ? (R = -P + 2 * Math.PI, R > Math.PI / 2 && (R += Math.PI)) : K === "tangential" ? R = -P - Math.PI / 2 : dt(K) && (R = K * Math.PI / 180), R === 0 ? _.add(new q({
            style: Y(p, {
              text: G,
              x: H,
              y: X,
              verticalAlign: l < -0.8 ? "top" : l > 0.8 ? "bottom" : "middle",
              align: n < -0.4 ? "left" : n > 0.4 ? "right" : "center"
            }, {
              inheritColor: T
            }),
            silent: !0
          })) : _.add(new q({
            style: Y(p, {
              text: G,
              x: H,
              y: X,
              verticalAlign: "middle",
              align: "center"
            }, {
              inheritColor: T
            }),
            silent: !0,
            originX: H,
            originY: X,
            rotation: R
          }));
        }
        if (r.get("show") && i !== S) {
          var v = r.get("distance");
          v = v ? v + d : d;
          for (var Z = 0; Z <= N; Z++) {
            n = Math.cos(P), l = Math.sin(P);
            var J = new tt({
              shape: {
                x1: n * (u - v) + L,
                y1: l * (u - v) + V,
                x2: n * (u - E - v) + L,
                y2: l * (u - E - v) + V
              },
              silent: !0,
              style: f
            });
            f.stroke === "auto" && J.setStyle({
              stroke: m((i + Z / N) / S)
            }), _.add(J), P += e;
          }
          P -= e;
        } else
          P += C;
      }
    }, c.prototype._renderPointer = function(t, b, y, m, a, A, s, k, d) {
      var _ = this.group, L = this._data, V = this._progressEls, u = [], z = t.get(["pointer", "show"]), D = t.getModel("progress"), g = D.get("show"), r = t.getData(), p = r.mapDimension("value"), S = +t.get("min"), N = +t.get("max"), x = [S, N], E = [A, s];
      function P(e, h) {
        var f = r.getItemModel(e), o = f.getModel("pointer"), n = M(o.get("width"), a.r), l = M(o.get("length"), a.r), i = t.get(["pointer", "icon"]), v = o.get("offsetCenter"), w = M(v[0], a.r), G = M(v[1], a.r), T = o.get("keepAspect"), H;
        return i ? H = F(i, w - n / 2, G - l, n, l, null, T) : H = new nt({
          shape: {
            angle: -Math.PI / 2,
            width: n,
            r: l,
            x: w,
            y: G
          }
        }), H.rotation = -(h + Math.PI / 2), H.x = a.cx, H.y = a.cy, H;
      }
      function C(e, h) {
        var f = D.get("roundCap"), o = f ? $ : I, n = D.get("overlap"), l = n ? D.get("width") : d / r.count(), i = n ? a.r - l : a.r - (e + 1) * l, v = n ? a.r : a.r - e * l, w = new o({
          shape: {
            startAngle: A,
            endAngle: h,
            cx: a.cx,
            cy: a.cy,
            clockwise: k,
            r0: i,
            r: v
          }
        });
        return n && (w.z2 = W(r.get(p, e), [S, N], [100, 0], !0)), w;
      }
      (g || z) && (r.diff(L).add(function(e) {
        var h = r.get(p, e);
        if (z) {
          var f = P(e, A);
          at(f, {
            rotation: -((isNaN(+h) ? E[0] : W(h, x, E, !0)) + Math.PI / 2)
          }, t), _.add(f), r.setItemGraphicEl(e, f);
        }
        if (g) {
          var o = C(e, A), n = D.get("clip");
          at(o, {
            shape: {
              endAngle: W(h, x, E, n)
            }
          }, t), _.add(o), j(t.seriesIndex, r.dataType, e, o), u[e] = o;
        }
      }).update(function(e, h) {
        var f = r.get(p, e);
        if (z) {
          var o = L.getItemGraphicEl(h), n = o ? o.rotation : A, l = P(e, n);
          l.rotation = n, rt(l, {
            rotation: -((isNaN(+f) ? E[0] : W(f, x, E, !0)) + Math.PI / 2)
          }, t), _.add(l), r.setItemGraphicEl(e, l);
        }
        if (g) {
          var i = V[h], v = i ? i.shape.endAngle : A, w = C(e, v), G = D.get("clip");
          rt(w, {
            shape: {
              endAngle: W(f, x, E, G)
            }
          }, t), _.add(w), j(t.seriesIndex, r.dataType, e, w), u[e] = w;
        }
      }).execute(), r.each(function(e) {
        var h = r.getItemModel(e), f = h.getModel("emphasis"), o = f.get("focus"), n = f.get("blurScope"), l = f.get("disabled");
        if (z) {
          var i = r.getItemGraphicEl(e), v = r.getItemVisual(e, "style"), w = v.fill;
          if (i instanceof ut) {
            var G = i.style;
            i.useStyle(ot({
              image: G.image,
              x: G.x,
              y: G.y,
              width: G.width,
              height: G.height
            }, v));
          } else
            i.useStyle(v), i.type !== "pointer" && i.setColor(w);
          i.setStyle(h.getModel(["pointer", "itemStyle"]).getItemStyle()), i.style.fill === "auto" && i.setStyle("fill", m(W(r.get(p, e), x, [0, 1], !0))), i.z2EmphasisLift = 0, Q(i, h), U(i, o, n, l);
        }
        if (g) {
          var T = u[e];
          T.useStyle(r.getItemVisual(e, "style")), T.setStyle(h.getModel(["progress", "itemStyle"]).getItemStyle()), T.z2EmphasisLift = 0, Q(T, h), U(T, o, n, l);
        }
      }), this._progressEls = u);
    }, c.prototype._renderAnchor = function(t, b) {
      var y = t.getModel("anchor"), m = y.get("show");
      if (m) {
        var a = y.get("size"), A = y.get("icon"), s = y.get("offsetCenter"), k = y.get("keepAspect"), d = F(A, b.cx - a / 2 + M(s[0], b.r), b.cy - a / 2 + M(s[1], b.r), a, a, null, k);
        d.z2 = y.get("showAbove") ? 1 : 0, d.setStyle(y.getModel("itemStyle").getItemStyle()), this.group.add(d);
      }
    }, c.prototype._renderTitleAndDetail = function(t, b, y, m, a) {
      var A = this, s = t.getData(), k = s.mapDimension("value"), d = +t.get("min"), _ = +t.get("max"), L = new et(), V = [], u = [], z = t.isAnimationEnabled(), D = t.get(["pointer", "showAbove"]);
      s.diff(this._data).add(function(g) {
        V[g] = new q({
          silent: !0
        }), u[g] = new q({
          silent: !0
        });
      }).update(function(g, r) {
        V[g] = A._titleEls[r], u[g] = A._detailEls[r];
      }).execute(), s.each(function(g) {
        var r = s.getItemModel(g), p = s.get(k, g), S = new et(), N = m(W(p, [d, _], [0, 1], !0)), x = r.getModel("title");
        if (x.get("show")) {
          var E = x.get("offsetCenter"), P = a.cx + M(E[0], a.r), C = a.cy + M(E[1], a.r), e = V[g];
          e.attr({
            z2: D ? 0 : 2,
            style: Y(x, {
              x: P,
              y: C,
              text: s.getName(g),
              align: "center",
              verticalAlign: "middle"
            }, {
              inheritColor: N
            })
          }), S.add(e);
        }
        var h = r.getModel("detail");
        if (h.get("show")) {
          var f = h.get("offsetCenter"), o = a.cx + M(f[0], a.r), n = a.cy + M(f[1], a.r), l = M(h.get("width"), a.r), i = M(h.get("height"), a.r), v = t.get(["progress", "show"]) ? s.getItemVisual(g, "style").fill : N, e = u[g], w = h.get("formatter");
          e.attr({
            z2: D ? 0 : 2,
            style: Y(h, {
              x: o,
              y: n,
              text: B(p, w),
              width: isNaN(l) ? null : l,
              height: isNaN(i) ? null : i,
              align: "center",
              verticalAlign: "middle"
            }, {
              inheritColor: v
            })
          }), lt(e, {
            normal: h
          }, p, function(T) {
            return B(T, w);
          }), z && vt(e, g, s, t, {
            getFormattedLabel: function(T, H, X, K, R, Z) {
              return B(Z ? Z.interpolatedValue : p, w);
            }
          }), S.add(e);
        }
        L.add(S);
      }), this.group.add(L), this._titleEls = V, this._detailEls = u;
    }, c.type = "gauge", c;
  })(gt)
);
export {
  Tt as default
};
