// echarts@6.0.0/lib/chart/gauge/PointerPath.js
import { __extends as h } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import y from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Path.js";
var P = (
  /** @class */
  /* @__PURE__ */ (function() {
    function o() {
      this.angle = 0, this.width = 10, this.r = 10, this.x = 0, this.y = 0;
    }
    return o;
  })()
), d = (
  /** @class */
  (function(o) {
    h(a, o);
    function a(r) {
      var t = o.call(this, r) || this;
      return t.type = "pointer", t;
    }
    return a.prototype.getDefaultShape = function() {
      return new P();
    }, a.prototype.buildPath = function(r, t) {
      var e = Math.cos, l = Math.sin, u = t.r, n = t.width, i = t.angle, v = t.x - e(i) * n * (n >= u / 3 ? 1 : 2), f = t.y - l(i) * n * (n >= u / 3 ? 1 : 2);
      i = t.angle - Math.PI / 2, r.moveTo(v, f), r.lineTo(t.x + e(i) * n, t.y + l(i) * n), r.lineTo(t.x + e(t.angle) * u, t.y + l(t.angle) * u), r.lineTo(t.x - e(i) * n, t.y - l(i) * n), r.lineTo(v, f);
    }, a;
  })(y)
);
export {
  d as default
};
