// echarts@6.0.0/lib/extension.js
import { PRIORITY as g, registerMap as a, registerLoading as m, registerTransform as p, registerVisual as f, registerLayout as u, registerCoordinateSystem as l, registerAction as c, registerUpdateLifecycle as C, registerPostUpdate as d, registerPostInit as P, registerProcessor as y, registerPreprocessor as S } from "./core/echarts.js";
import t from "./view/Component.js";
import s from "./view/Chart.js";
import i from "./model/Component.js";
import o from "./model/Series.js";
import { isArray as I, each as M, indexOf as V, isFunction as w } from "../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { registerImpl as L } from "./core/impl.js";
import { registerPainter as T } from "../../../../zrender@6.0.0/node_modules/zrender/lib/zrender.js";
import { registerCustomSeries as h } from "./chart/custom/customSeriesRegister.js";
var n = [], R = {
  registerPreprocessor: S,
  registerProcessor: y,
  registerPostInit: P,
  registerPostUpdate: d,
  registerUpdateLifecycle: C,
  registerAction: c,
  registerCoordinateSystem: l,
  registerLayout: u,
  registerVisual: f,
  registerTransform: p,
  registerLoading: m,
  registerMap: a,
  registerImpl: L,
  PRIORITY: g,
  ComponentModel: i,
  ComponentView: t,
  SeriesModel: o,
  ChartView: s,
  // TODO Use ComponentModel and SeriesModel instead of Constructor
  registerComponentModel: function(e) {
    i.registerClass(e);
  },
  registerComponentView: function(e) {
    t.registerClass(e);
  },
  registerSeriesModel: function(e) {
    o.registerClass(e);
  },
  registerChartView: function(e) {
    s.registerClass(e);
  },
  registerCustomSeries: function(e, r) {
    h(e, r);
  },
  registerSubTypeDefaulter: function(e, r) {
    i.registerSubTypeDefaulter(e, r);
  },
  registerPainter: function(e, r) {
    T(e, r);
  }
};
function U(e) {
  if (I(e)) {
    M(e, function(r) {
      U(r);
    });
    return;
  }
  V(n, e) >= 0 || (n.push(e), w(e) && (e = {
    install: e
  }), e.install(R));
}
export {
  U as use
};
