// recharts@2.15.0/es6/container/Layer.js
import l from "react";
import { clsx as f } from "../../../../../clsx@2.1.1/node_modules/clsx/dist/clsx.js";
import { filterProps as c } from "../util/ReactUtils.js";
var s = ["children", "className"];
function o() {
  return o = Object.assign ? Object.assign.bind() : function(t) {
    for (var n = 1; n < arguments.length; n++) {
      var e = arguments[n];
      for (var r in e)
        Object.prototype.hasOwnProperty.call(e, r) && (t[r] = e[r]);
    }
    return t;
  }, o.apply(this, arguments);
}
function p(t, n) {
  if (t == null) return {};
  var e = m(t, n), r, a;
  if (Object.getOwnPropertySymbols) {
    var i = Object.getOwnPropertySymbols(t);
    for (a = 0; a < i.length; a++)
      r = i[a], !(n.indexOf(r) >= 0) && Object.prototype.propertyIsEnumerable.call(t, r) && (e[r] = t[r]);
  }
  return e;
}
function m(t, n) {
  if (t == null) return {};
  var e = {};
  for (var r in t)
    if (Object.prototype.hasOwnProperty.call(t, r)) {
      if (n.indexOf(r) >= 0) continue;
      e[r] = t[r];
    }
  return e;
}
var h = /* @__PURE__ */ l.forwardRef(function(t, n) {
  var e = t.children, r = t.className, a = p(t, s), i = f("recharts-layer", r);
  return /* @__PURE__ */ l.createElement("g", o({
    className: i
  }, c(a, !0), {
    ref: n
  }), e);
});
export {
  h as Layer
};
