// echarts@6.0.0/lib/layout/barGrid.js
import { each as l, keys as Y, defaults as Z } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { parsePercent as A } from "../util/number.js";
import { isDimensionStacked as $ } from "../data/helper/dataStackHelper.js";
import N from "../chart/helper/createRenderPlanner.js";
import { createFloat32Array as R } from "../util/vendor.js";
var E = "__ec_stack_";
function X(e) {
  return e.get("stack") || E + e.seriesIndex;
}
function F(e) {
  return e.dim + e.index;
}
function oa(e) {
  var i = [], t = e.axis, a = "axis0";
  if (t.type === "category") {
    for (var u = t.getBandWidth(), r = 0; r < e.count; r++)
      i.push(Z({
        bandWidth: u,
        axisKey: a,
        stackId: E + r
      }, e));
    for (var f = q(i), n = [], r = 0; r < e.count; r++) {
      var d = f[a][E + r];
      d.offsetCenter = d.offset + d.width / 2, n.push(d);
    }
    return n;
  }
}
function j(e, i) {
  var t = [];
  return i.eachSeriesByType(e, function(a) {
    J(a) && t.push(a);
  }), t;
}
function aa(e) {
  var i = {};
  l(e, function(d) {
    var o = d.coordinateSystem, v = o.getBaseAxis();
    if (!(v.type !== "time" && v.type !== "value"))
      for (var s = d.getData(), h = v.dim + "_" + v.index, m = s.getDimensionIndex(s.mapDimension(v.dim)), p = s.getStore(), x = 0, W = p.count(); x < W; ++x) {
        var c = p.get(m, x);
        i[h] ? i[h].push(c) : i[h] = [c];
      }
  });
  var t = {};
  for (var a in i)
    if (i.hasOwnProperty(a)) {
      var u = i[a];
      if (u) {
        u.sort(function(d, o) {
          return d - o;
        });
        for (var r = null, f = 1; f < u.length; ++f) {
          var n = u[f] - u[f - 1];
          n > 0 && (r = r === null ? n : Math.min(r, n));
        }
        t[a] = r;
      }
    }
  return t;
}
function ta(e) {
  var i = aa(e), t = [];
  return l(e, function(a) {
    var u = a.coordinateSystem, r = u.getBaseAxis(), f = r.getExtent(), n;
    if (r.type === "category")
      n = r.getBandWidth();
    else if (r.type === "value" || r.type === "time") {
      var d = r.dim + "_" + r.index, o = i[d], v = Math.abs(f[1] - f[0]), s = r.scale.getExtent(), h = Math.abs(s[1] - s[0]);
      n = o ? v / h * o : v;
    } else {
      var m = a.getData();
      n = Math.abs(f[1] - f[0]) / m.count();
    }
    var p = A(a.get("barWidth"), n), x = A(a.get("barMaxWidth"), n), W = A(
      // barMinWidth by default is 0.5 / 1 in cartesian. Because in value axis,
      // the auto-calculated bar width might be less than 0.5 / 1.
      a.get("barMinWidth") || (Q(a) ? 0.5 : 1),
      n
    ), c = a.get("barGap"), y = a.get("barCategoryGap"), b = a.get("defaultBarGap");
    t.push({
      bandWidth: n,
      barWidth: p,
      barMaxWidth: x,
      barMinWidth: W,
      barGap: c,
      barCategoryGap: y,
      defaultBarGap: b,
      axisKey: F(r),
      stackId: X(a)
    });
  }), q(t);
}
function q(e) {
  var i = {};
  l(e, function(a, u) {
    var r = a.axisKey, f = a.bandWidth, n = i[r] || {
      bandWidth: f,
      remainedWidth: f,
      autoWidthCount: 0,
      categoryGap: null,
      gap: a.defaultBarGap || 0,
      stacks: {}
    }, d = n.stacks;
    i[r] = n;
    var o = a.stackId;
    d[o] || n.autoWidthCount++, d[o] = d[o] || {
      width: 0,
      maxWidth: 0
    };
    var v = a.barWidth;
    v && !d[o].width && (d[o].width = v, v = Math.min(n.remainedWidth, v), n.remainedWidth -= v);
    var s = a.barMaxWidth;
    s && (d[o].maxWidth = s);
    var h = a.barMinWidth;
    h && (d[o].minWidth = h);
    var m = a.barGap;
    m != null && (n.gap = m);
    var p = a.barCategoryGap;
    p != null && (n.categoryGap = p);
  });
  var t = {};
  return l(i, function(a, u) {
    t[u] = {};
    var r = a.stacks, f = a.bandWidth, n = a.categoryGap;
    if (n == null) {
      var d = Y(r).length;
      n = Math.max(35 - d * 4, 15) + "%";
    }
    var o = A(n, f), v = A(a.gap, 1), s = a.remainedWidth, h = a.autoWidthCount, m = (s - o) / (h + (h - 1) * v);
    m = Math.max(m, 0), l(r, function(c) {
      var y = c.maxWidth, b = c.minWidth;
      if (c.width) {
        var g = c.width;
        y && (g = Math.min(g, y)), b && (g = Math.max(g, b)), c.width = g, s -= g + v * g, h--;
      } else {
        var g = m;
        y && y < g && (g = Math.min(y, s)), b && b > g && (g = b), g !== m && (c.width = g, s -= g + v * g, h--);
      }
    }), m = (s - o) / (h + (h - 1) * v), m = Math.max(m, 0);
    var p = 0, x;
    l(r, function(c, y) {
      c.width || (c.width = m), x = c, p += c.width * (1 + v);
    }), x && (p -= x.width * v);
    var W = -p / 2;
    l(r, function(c, y) {
      t[u][y] = t[u][y] || {
        bandWidth: f,
        offset: W,
        width: c.width
      }, W += c.width * (1 + v);
    });
  }), t;
}
function ua(e, i, t) {
  if (e && i) {
    var a = e[F(i)];
    return a;
  }
}
function fa(e, i) {
  var t = j(e, i), a = ta(t);
  l(t, function(u) {
    var r = u.getData(), f = u.coordinateSystem, n = f.getBaseAxis(), d = X(u), o = a[F(n)][d], v = o.offset, s = o.width;
    r.setLayout({
      bandWidth: o.bandWidth,
      offset: v,
      size: s
    });
  });
}
function sa(e) {
  return {
    seriesType: e,
    plan: N(),
    reset: function(i) {
      if (J(i)) {
        var t = i.getData(), a = i.coordinateSystem, u = a.getBaseAxis(), r = a.getOtherAxis(u), f = t.getDimensionIndex(t.mapDimension(r.dim)), n = t.getDimensionIndex(t.mapDimension(u.dim)), d = i.get("showBackground", !0), o = t.mapDimension(r.dim), v = t.getCalculationInfo("stackResultDimension"), s = $(t, o) && !!t.getCalculationInfo("stackedOnSeries"), h = r.isHorizontal(), m = ra(u, r), p = Q(i), x = i.get("barMinHeight") || 0, W = v && t.getDimensionIndex(v), c = t.getLayout("size"), y = t.getLayout("offset");
        return {
          progress: function(b, g) {
            for (var K = b.count, I = p && R(K * 3), B = p && d && R(K * 3), V = p && R(K), M = a.master.getRect(), U = h ? M.width : M.height, G, _ = g.getStore(), k = 0; (G = b.next()) != null; ) {
              var O = _.get(s ? W : f, G), T = _.get(n, G), C = m, z = void 0;
              s && (z = +O - _.get(f, G));
              var L = void 0, P = void 0, D = void 0, w = void 0;
              if (h) {
                var S = a.dataToPoint([O, T]);
                if (s) {
                  var H = a.dataToPoint([z, T]);
                  C = H[0];
                }
                L = C, P = S[1] + y, D = S[0] - C, w = c, Math.abs(D) < x && (D = (D < 0 ? -1 : 1) * x);
              } else {
                var S = a.dataToPoint([T, O]);
                if (s) {
                  var H = a.dataToPoint([T, z]);
                  C = H[1];
                }
                L = S[0] + y, P = C, D = c, w = S[1] - C, Math.abs(w) < x && (w = (w <= 0 ? -1 : 1) * x);
              }
              p ? (I[k] = L, I[k + 1] = P, I[k + 2] = h ? D : w, B && (B[k] = h ? M.x : L, B[k + 1] = h ? P : M.y, B[k + 2] = U), V[G] = G) : g.setItemLayout(G, {
                x: L,
                y: P,
                width: D,
                height: w
              }), k += 3;
            }
            p && g.setLayout({
              largePoints: I,
              largeDataIndices: V,
              largeBackgroundPoints: B,
              valueAxisHorizontal: h
            });
          }
        };
      }
    }
  };
}
function J(e) {
  return e.coordinateSystem && e.coordinateSystem.type === "cartesian2d";
}
function Q(e) {
  return e.pipelineContext && e.pipelineContext.large;
}
function ra(e, i) {
  var t = i.model.get("startValue");
  return t || (t = 0), i.toGlobalCoord(i.dataToCoord(i.type === "log" ? t > 0 ? t : 1 : t));
}
export {
  sa as createProgressiveLayout,
  oa as getLayoutOnAxis,
  fa as layout,
  ta as makeColumnLayout,
  j as prepareLayoutBarSeries,
  ua as retrieveColumnLayout
};
