// echarts@6.0.0/lib/layout/barPolar.js
import { filter as U, each as s } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { parsePercent as T } from "../util/number.js";
import { isDimensionStacked as X } from "../data/helper/dataStackHelper.js";
function E(m) {
  return m.get("stack") || "__ec_stack_" + m.seriesIndex;
}
function H(m, u) {
  return u.dim + m.model.componentIndex;
}
function O(m, u, x) {
  var a = {}, g = Y(U(u.getSeriesByType(m), function(t) {
    return !u.isSeriesFiltered(t) && t.coordinateSystem && t.coordinateSystem.type === "polar";
  }));
  u.eachSeriesByType(m, function(t) {
    if (t.coordinateSystem.type === "polar") {
      var f = t.getData(), e = t.coordinateSystem, d = e.getBaseAxis(), p = H(e, d), r = E(t), i = g[p][r], n = i.offset, c = i.width, o = e.getOtherAxis(d), v = t.coordinateSystem.cx, y = t.coordinateSystem.cy, h = t.get("barMinHeight") || 0, K = t.get("barMinAngle") || 0;
      a[r] = a[r] || [];
      for (var A = f.mapDimension(o.dim), F = f.mapDimension(d.dim), P = X(
        f,
        A
        /* , baseDim */
      ), R = d.dim !== "radius" || !t.get("roundCap", !0), j = o.model, q = j.get("startValue"), w = o.dataToCoord(q || 0), b = 0, z = f.count(); b < z; b++) {
        var L = f.get(A, b), W = f.get(F, b), _ = L >= 0 ? "p" : "n", k = w;
        P && (a[r][W] || (a[r][W] = {
          p: w,
          n: w
          // Negative stack
        }), k = a[r][W][_]);
        var I = void 0, S = void 0, C = void 0, G = void 0;
        if (o.dim === "radius") {
          var D = o.dataToCoord(L) - w, J = d.dataToCoord(W);
          Math.abs(D) < h && (D = (D < 0 ? -1 : 1) * h), I = k, S = k + D, C = J - n, G = C - c, P && (a[r][W][_] = S);
        } else {
          var B = o.dataToCoord(L, R) - w, Q = d.dataToCoord(W);
          Math.abs(B) < K && (B = (B < 0 ? -1 : 1) * K), I = Q + n, S = I + c, C = k, G = k + B, P && (a[r][W][_] = G);
        }
        f.setItemLayout(b, {
          cx: v,
          cy: y,
          r0: I,
          r: S,
          // Consider that positive angle is anti-clockwise,
          // while positive radian of sector is clockwise
          startAngle: -C * Math.PI / 180,
          endAngle: -G * Math.PI / 180,
          /**
           * Keep the same logic with bar in catesion: use end value to
           * control direction. Notice that if clockwise is true (by
           * default), the sector will always draw clockwisely, no matter
           * whether endAngle is greater or less than startAngle.
           */
          clockwise: C >= G
        });
      }
    }
  });
}
function Y(m) {
  var u = {};
  s(m, function(a, g) {
    var t = a.getData(), f = a.coordinateSystem, e = f.getBaseAxis(), d = H(f, e), p = e.getExtent(), r = e.type === "category" ? e.getBandWidth() : Math.abs(p[1] - p[0]) / t.count(), i = u[d] || {
      bandWidth: r,
      remainedWidth: r,
      autoWidthCount: 0,
      categoryGap: "20%",
      gap: "30%",
      stacks: {}
    }, n = i.stacks;
    u[d] = i;
    var c = E(a);
    n[c] || i.autoWidthCount++, n[c] = n[c] || {
      width: 0,
      maxWidth: 0
    };
    var o = T(a.get("barWidth"), r), v = T(a.get("barMaxWidth"), r), y = a.get("barGap"), h = a.get("barCategoryGap");
    o && !n[c].width && (o = Math.min(i.remainedWidth, o), n[c].width = o, i.remainedWidth -= o), v && (n[c].maxWidth = v), y != null && (i.gap = y), h != null && (i.categoryGap = h);
  });
  var x = {};
  return s(u, function(a, g) {
    x[g] = {};
    var t = a.stacks, f = a.bandWidth, e = T(a.categoryGap, f), d = T(a.gap, 1), p = a.remainedWidth, r = a.autoWidthCount, i = (p - e) / (r + (r - 1) * d);
    i = Math.max(i, 0), s(t, function(v, y) {
      var h = v.maxWidth;
      h && h < i && (h = Math.min(h, p), v.width && (h = Math.min(h, v.width)), p -= h, v.width = h, r--);
    }), i = (p - e) / (r + (r - 1) * d), i = Math.max(i, 0);
    var n = 0, c;
    s(t, function(v, y) {
      v.width || (v.width = i), c = v, n += v.width * (1 + d);
    }), c && (n -= c.width * d);
    var o = -n / 2;
    s(t, function(v, y) {
      x[g][y] = x[g][y] || {
        offset: o,
        width: v.width
      }, o += v.width * (1 + d);
    });
  }), x;
}
export {
  O as default
};
