// echarts@6.0.0/lib/coord/matrix/MatrixBodyCorner.js
import { createHashMap as D, isArray as E, each as N, isObject as b, extend as S } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { error as C } from "../../util/log.js";
import g from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/Point.js";
import { resetXYLocatorRange as V, parseCoordRangeOption as X, MatrixClampOption as d, isXYLocatorRangeInvalidOnDim as _, fillIdSpanFromLocatorRange as k, resolveXYLocatorRangeByCellMerge as B, createNaNRectLike as Y, cloneXYLocatorRange as j } from "./matrixCoordHelper.js";
var q = (
  /** @class */
  (function() {
    function c(e, n, u) {
      this._model = n, this._dims = u, this._kind = e, this._cellMergeOwnerList = [];
    }
    return c.prototype._ensureCellMap = function() {
      var e = this, n = e._cellMap;
      return n || (n = e._cellMap = D(), u()), n;
      function u() {
        var l = [], a = e._model.getShallow("data");
        a && !E(a) && (process.env.NODE_ENV !== "production" && C("matrix." + a + ".data must be an array if specified."), a = null), N(a, function(r, i) {
          if (!b(r) || !E(r.coord)) {
            process.env.NODE_ENV !== "production" && C("Illegal matrix." + e._kind + ".data[" + i + "], must be a {coord: [...], ...}");
            return;
          }
          var o = V([]), f = null;
          if (process.env.NODE_ENV !== "production" && (f = []), X(o, f, r.coord, e._dims, r.coordClamp ? d[e._kind] : d.none), _(o, 0) || _(o, 1)) {
            process.env.NODE_ENV !== "production" && C("Can not determine cells by option matrix." + e._kind + ".data[" + i + "]: " + ("" + f.join(" ")));
            return;
          }
          var O = r && r.mergeCells, L = {
            id: new g(),
            span: new g(),
            locatorRange: o,
            option: r,
            cellMergeOwner: O
          };
          k(L, o), l.push(L);
        });
        for (var v = [], t = 0; t < l.length; t++) {
          var s = l[t];
          if (s.cellMergeOwner) {
            var M = s.locatorRange;
            B(M, v, l, t);
            for (var m = 0; m < t; m++)
              v[m] && (l[m].cellMergeOwner = !1);
            if (M[0][0] !== s.id.x || M[1][0] !== s.id.y) {
              s.cellMergeOwner = !1;
              var w = S({}, s.option);
              w.coord = null;
              var y = {
                id: new g(),
                span: new g(),
                locatorRange: M,
                option: w,
                cellMergeOwner: !0
              };
              k(y, M), l.push(y);
            }
          }
        }
        N(l, function(r) {
          var i = p(r.id.x, r.id.y);
          if (r.cellMergeOwner && (i.cellMergeOwner = !0, i.span = r.span, i.locatorRange = r.locatorRange, i.spanRect = Y(), e._cellMergeOwnerList.push(i)), !(!r.cellMergeOwner && !r.option))
            for (var o = 0; o < r.span.y; o++)
              for (var f = 0; f < r.span.x; f++) {
                var O = p(r.id.x + f, r.id.y + o);
                O.option = r.option, r.cellMergeOwner && (O.inSpanOf = i);
              }
        });
      }
      function p(l, a) {
        var v = R(l, a), t = n.get(v);
        return t || (t = n.set(v, {
          id: new g(l, a),
          option: null,
          inSpanOf: null,
          span: null,
          spanRect: null,
          locatorRange: null,
          cellMergeOwner: !1
        })), t;
      }
    }, c.prototype.getCell = function(e) {
      return this._ensureCellMap().get(R(e[0], e[1]));
    }, c.prototype.travelExistingCells = function(e) {
      this._ensureCellMap().each(e);
    }, c.prototype.expandRangeByCellMerge = function(e) {
      if (!_(e, 0) && !_(e, 1) && e[0][0] === e[0][1] && e[1][0] === e[1][1]) {
        h[0] = e[0][0], h[1] = e[1][0];
        var n = this.getCell(h), u = n && n.inSpanOf;
        if (u) {
          j(e, u.locatorRange);
          return;
        }
      }
      var p = this._cellMergeOwnerList;
      B(e, null, p, p.length);
    }, c;
  })()
), h = [];
function R(c, e) {
  return c + "|" + e;
}
export {
  q as MatrixBodyCorner
};
