// echarts@6.0.0/lib/coord/matrix/MatrixModel.js
import { __extends as p } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import c from "../../model/Component.js";
import n from "../../model/Model.js";
import { MatrixDim as a } from "./MatrixDim.js";
import { MatrixBodyCorner as u } from "./MatrixBodyCorner.js";
import r from "../../visual/tokens.js";
var i = {
  show: !0,
  color: r.color.secondary,
  // overflow: 'truncate',
  overflow: "break",
  lineOverflow: "truncate",
  padding: [2, 3, 2, 3],
  // Prefer to use `padding`, rather than distance.
  distance: 0
};
function l(o) {
  return {
    color: "none",
    borderWidth: 1,
    borderColor: o ? "none" : r.color.borderTint
  };
}
var y = {
  show: !0,
  label: i,
  itemStyle: l(!1),
  silent: void 0,
  dividerLineStyle: {
    width: 1,
    color: r.color.border
  }
}, h = {
  label: i,
  itemStyle: l(!1),
  silent: void 0
}, x = {
  label: i,
  itemStyle: l(!0),
  silent: void 0
}, M = {
  // As a most basic coord sys, `z` should be lower than
  // other series and coord sys, such as, grid.
  z: -50,
  left: "10%",
  top: "10%",
  right: "10%",
  bottom: "10%",
  x: y,
  y,
  body: h,
  corner: x,
  backgroundStyle: {
    color: "none",
    borderColor: r.color.axisLine,
    borderWidth: 1
  }
}, S = (
  /** @class */
  (function(o) {
    p(t, o);
    function t() {
      var e = o !== null && o.apply(this, arguments) || this;
      return e.type = t.type, e;
    }
    return t.prototype.optionUpdated = function() {
      var e = this._dimModels = {
        // Do not use matrixModel as the parent model, for preventing from cascade-fetching options to it.
        x: new m(this.get("x", !0) || {}),
        y: new m(this.get("y", !0) || {})
      };
      e.x.option.type = e.y.option.type = "category";
      var s = e.x.dim = new a("x", e.x), f = e.y.dim = new a("y", e.y), d = {
        x: s,
        y: f
      };
      this._body = new u("body", new n(this.getShallow("body")), d), this._corner = new u("corner", new n(this.getShallow("corner")), d);
    }, t.prototype.getDimensionModel = function(e) {
      return this._dimModels[e];
    }, t.prototype.getBody = function() {
      return this._body;
    }, t.prototype.getCorner = function() {
      return this._corner;
    }, t.type = "matrix", t.layoutMode = "box", t.defaultOption = M, t;
  })(c)
), m = (
  /** @class */
  (function(o) {
    p(t, o);
    function t() {
      return o !== null && o.apply(this, arguments) || this;
    }
    return t.prototype.getOrdinalMeta = function() {
      return this.dim.getOrdinalMeta();
    }, t;
  })(n)
);
export {
  m as MatrixDimensionModel,
  S as default
};
