// echarts@6.0.0/lib/coord/matrix/Matrix.js
import { getLayoutRect as d } from "../../util/layout.js";
import { ListIterator as m } from "../../util/model.js";
import { isArray as I, retrieve2 as J, eqNaN as X } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { XY as i, WH as l } from "../../util/graphic.js";
import Y from "../../model/Model.js";
import { mathMin as H, mathMax as A, parsePositionSizeOption as ee } from "../../util/number.js";
import { createNaNRectLike as te, resetXYLocatorRange as re, parseCoordRangeOption as oe, MatrixClampOption as T, xyLocatorRangeToRectOneDim as E, MatrixCellLayoutInfoType as K } from "./matrixCoordHelper.js";
import { error as ne } from "../../util/log.js";
import { injectCoordSysByOption as ae, simpleCoordSysInjectionProvider as ie } from "../../core/CoordinateSystem.js";
var Ce = (
  /** @class */
  (function() {
    function o(e, r, t) {
      this.dimensions = o.dimensions, this.type = "matrix", this._model = e;
      var n = this._dimModels = {
        x: e.getDimensionModel("x"),
        y: e.getDimensionModel("y")
      };
      this._dims = {
        x: n.x.dim,
        y: n.y.dim
      }, this._resize(e, t);
    }
    return o.getDimensionsInfo = function() {
      return [{
        name: "x",
        type: "ordinal"
      }, {
        name: "y",
        type: "ordinal"
      }, {
        name: "value"
      }];
    }, o.create = function(e, r) {
      var t = [];
      return e.eachComponent("matrix", function(n) {
        var a = new o(n, e, r);
        t.push(a), n.coordinateSystem = a;
      }), e.eachComponent(function(n, a) {
        ae({
          targetModel: a,
          coordSysType: "matrix",
          coordSysProvider: ie
        });
      }), t;
    }, o.prototype.getRect = function() {
      return this._rect;
    }, o.prototype._resize = function(e, r) {
      var t = this._dims, n = this._dimModels, a = this._rect = d(e.getBoxLayoutParams(), {
        width: r.getWidth(),
        height: r.getHeight()
      });
      F(n, t, a, 0), F(n, t, a, 1), W(0, t), W(1, t), j(this._model.getBody(), t), j(this._model.getCorner(), t);
    }, o.prototype.dataToPoint = function(e, r, t) {
      return t = t || [], this.dataToLayout(e, r, D), t[0] = D.rect.x + D.rect.width / 2, t[1] = D.rect.y + D.rect.height / 2, t;
    }, o.prototype.dataToLayout = function(e, r, t) {
      var n = this._dims;
      t = t || {};
      var a = t.rect = t.rect || {};
      a.x = a.y = a.width = a.height = NaN;
      var y = t.matrixXYLocatorRange = re(t.matrixXYLocatorRange);
      return I(e) ? (oe(y, null, e, n, J(r && r.clamp, T.none)), (!r || !r.ignoreMergeCells) && ((!r || r.clamp !== T.corner) && this._model.getBody().expandRangeByCellMerge(y), (!r || r.clamp !== T.body) && this._model.getCorner().expandRangeByCellMerge(y)), E(a, y, n, 0), E(a, y, n, 1), t) : (process.env.NODE_ENV !== "production" && ne("Input data must be an array in `convertToLayout`, `convertToPixel`"), t);
    }, o.prototype.pointToData = function(e, r, t) {
      var n = this._dims;
      return k(h, 0, n, e, r && r.clamp), k(h, 1, n, e, r && r.clamp), t = t || [], t[0] = t[1] = NaN, h.y === u.inCorner && h.x === u.inBody ? q(h, t, 0, n) : h.x === u.inCorner && h.y === u.inBody ? q(h, t, 1, n) : (V(h, t, 0, n), V(h, t, 1, n)), t;
    }, o.prototype.convertToPixel = function(e, r, t, n) {
      var a = U(r);
      return a === this ? a.dataToPoint(t, n) : void 0;
    }, o.prototype.convertToLayout = function(e, r, t, n) {
      var a = U(r);
      return a === this ? a.dataToLayout(t, n) : void 0;
    }, o.prototype.convertFromPixel = function(e, r, t, n) {
      var a = U(r);
      return a === this ? a.pointToData(t, n) : void 0;
    }, o.prototype.containPoint = function(e) {
      return this._rect.contain(e[0], e[1]);
    }, o.dimensions = ["x", "y", "value"], o;
  })()
), D = {
  rect: te()
}, O = new m(), B = new m();
function F(o, e, r, t) {
  for (var n = 1 - t, a = e[i[t]], y = e[i[n]], v = y.shouldShow(), p = a.resetCellIterator(); p.next(); )
    p.item.wh = p.item.xy = NaN;
  for (var L = y.resetLayoutIterator(null, t); L.next(); )
    L.item.wh = L.item.xy = NaN;
  for (var f = r[l[t]], s = a.getLocatorCount(t) + y.getLocatorCount(t), g = new Y(), C = y.resetLevelIterator(); C.next(); )
    g.option = C.item.option, g.parentModel = o[i[n]], _(C.item, v ? g.get("levelSize") : 0);
  for (var M = new Y(), S = a.resetCellIterator(); S.next(); )
    S.item.type === K.leaf && (M.option = S.item.option, M.parentModel = void 0, _(S.item, M.get("size")));
  function _(c, x) {
    var b = se(x, t, r);
    X(b) || (c.wh = P(b, f), f = P(f - c.wh), s--);
  }
  var Q = s ? f / s : 0, Z = !s && f >= 1, z = r[i[t]], $ = a.getLocatorCount(t) - 1, w = new m();
  for (y.resetLayoutIterator(w, t); w.next(); )
    R(w.item);
  for (a.resetLayoutIterator(w, t); w.next(); )
    R(w.item);
  function R(c) {
    X(c.wh) && (c.wh = Q), c.xy = z, c.id[i[t]] === $ && !Z && (c.wh = r[i[t]] + r[l[t]] - c.xy), z += c.wh;
  }
}
function W(o, e) {
  for (var r = e[i[o]].resetCellIterator(); r.next(); ) {
    var t = r.item;
    N(t.rect, o, t.id, t.span, e), N(t.rect, 1 - o, t.id, t.span, e), t.type === K.nonLeaf && (t.xy = t.rect[i[o]], t.wh = t.rect[l[o]]);
  }
}
function j(o, e) {
  o.travelExistingCells(function(r) {
    var t = r.span;
    if (t) {
      var n = r.spanRect, a = r.id;
      N(n, 0, a, t, e), N(n, 1, a, t, e);
    }
  });
}
function N(o, e, r, t, n) {
  o[l[e]] = 0;
  var a = r[i[e]], y = a < 0 ? n[i[1 - e]] : n[i[e]], v = y.getUnitLayoutInfo(e, r[i[e]]);
  if (o[i[e]] = v.xy, o[l[e]] = v.wh, t[i[e]] > 1) {
    var p = y.getUnitLayoutInfo(e, r[i[e]] + t[i[e]] - 1);
    o[l[e]] = p.xy + p.wh - v.xy;
  }
}
function se(o, e, r) {
  var t = ee(o, r[l[e]]);
  return P(t, r[l[e]]);
}
function P(o, e) {
  return Math.max(Math.min(o, J(e, 1 / 0)), 0);
}
function U(o) {
  var e = o.matrixModel, r = o.seriesModel, t = e ? e.coordinateSystem : r ? r.coordinateSystem : null;
  return t;
}
var u = {
  inBody: 1,
  inCorner: 2,
  outside: 3
}, h = {
  x: null,
  y: null,
  point: []
};
function k(o, e, r, t, n) {
  var a = r[i[e]], y = r[i[1 - e]], v = a.getUnitLayoutInfo(e, a.getLocatorCount(e) - 1), p = a.getUnitLayoutInfo(e, 0), L = y.getUnitLayoutInfo(e, -y.getLocatorCount(e)), f = y.shouldShow() ? y.getUnitLayoutInfo(e, -1) : null, s = o.point[e] = t[e];
  if (!p && !f) {
    o[i[e]] = u.outside;
    return;
  }
  if (n === T.body) {
    p ? (o[i[e]] = u.inBody, s = H(v.xy + v.wh, A(p.xy, s)), o.point[e] = s) : o[i[e]] = u.outside;
    return;
  } else if (n === T.corner) {
    f ? (o[i[e]] = u.inCorner, s = H(f.xy + f.wh, A(L.xy, s)), o.point[e] = s) : o[i[e]] = u.outside;
    return;
  }
  var g = p ? p.xy : f ? f.xy + f.wh : NaN, C = L ? L.xy : g, M = v ? v.xy + v.wh : g;
  if (s < C) {
    if (!n) {
      o[i[e]] = u.outside;
      return;
    }
    s = C;
  } else if (s > M) {
    if (!n) {
      o[i[e]] = u.outside;
      return;
    }
    s = M;
  }
  o.point[e] = s, o[i[e]] = g <= s && s <= M ? u.inBody : C <= s && s <= g ? u.inCorner : u.outside;
}
function q(o, e, r, t) {
  var n = 1 - r;
  if (o[i[r]] !== u.outside)
    for (t[i[r]].resetCellIterator(B); B.next(); ) {
      var a = B.item;
      if (G(o.point[r], a.rect, r) && G(o.point[n], a.rect, n)) {
        e[r] = a.ordinal, e[n] = a.id[i[n]];
        return;
      }
    }
}
function V(o, e, r, t) {
  if (o[i[r]] !== u.outside) {
    var n = o[i[r]] === u.inCorner ? t[i[1 - r]] : t[i[r]];
    for (n.resetLayoutIterator(O, r); O.next(); )
      if (ye(o.point[r], O.item)) {
        e[r] = O.item.id[i[r]];
        return;
      }
  }
}
function ye(o, e) {
  return e.xy <= o && o <= e.xy + e.wh;
}
function G(o, e, r) {
  return e[i[r]] <= o && o <= e[i[r]] + e[l[r]];
}
export {
  Ce as default
};
