// echarts@6.0.0/lib/coord/matrix/matrixCoordHelper.js
import { isArray as F, eqNaN as l, isNumber as H } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { XY as v, WH as E } from "../../util/graphic.js";
import { mathMax as Y, mathMin as b } from "../../util/number.js";
var M = {
  level: 1,
  leaf: 2,
  nonLeaf: 3
}, T = {
  // No clamp, be falsy, equals to null/undefined. It means if the input part is
  // null/undefined/NaN/outOfBoundary, the result part is NaN, rather than clamp to
  // the boundary of the matrix.
  none: 0,
  // Clamp, where null/undefined/NaN/outOfBoundary can be used to cover the entire row/column.
  all: 1,
  body: 2,
  corner: 3
};
function A(r, n, t) {
  var a = n[v[t]].getCell(r);
  return !a && H(r) && r < 0 && (a = n[v[1 - t]].getUnitLayoutInfo(t, Math.round(r))), a;
}
function G(r) {
  var n = r || [];
  return n[0] = n[0] || [], n[1] = n[1] || [], n[0][0] = n[0][1] = n[1][0] = n[1][1] = NaN, n;
}
function J(r, n, t, a, f) {
  _(r[0], n, f, t, a, 0), _(r[1], n, f, t, a, 1);
}
function _(r, n, t, a, f, e) {
  r[0] = 1 / 0, r[1] = -1 / 0;
  var y = a[e], C = F(y) ? y : [y], i = C.length, p = !!t;
  if (i >= 1 ? (S(r, n, C, p, f, e, 0), i > 1 && S(r, n, C, p, f, e, i - 1)) : (process.env.NODE_ENV !== "production" && n && n.push('Should be like [["x1", "x2"], ["y1", "y2"]], or ["x1", "y1"], rather than empty.'), r[0] = r[1] = NaN), p) {
    var N = -f[v[1 - e]].getLocatorCount(e), X = f[v[e]].getLocatorCount(e) - 1;
    t === T.body ? N = Y(0, N) : t === T.corner && (X = b(-1, X)), X < N && (N = X = NaN), l(r[0]) && (r[0] = N), l(r[1]) && (r[1] = X), r[0] = Y(b(r[0], X), N), r[1] = Y(b(r[1], X), N);
  }
}
function S(r, n, t, a, f, e, y) {
  var C = A(t[y], f, e);
  if (!C) {
    process.env.NODE_ENV !== "production" && !a && n && n.push("Can not find cell by coord[" + e + "][" + y + "]."), r[0] = r[1] = NaN;
    return;
  }
  var i = C.id[v[e]], p = i, N = W(C);
  N && (p += N.span[v[e]] - 1), r[0] = b(r[0], i, p), r[1] = Y(r[1], i, p);
}
function K(r, n) {
  return l(r[n][0]) || l(r[n][1]);
}
function Q(r, n, t, a) {
  n = n || P;
  for (var f = 0; f < a; f++)
    n[f] = !1;
  for (; ; ) {
    for (var e = !1, f = 0; f < a; f++) {
      var y = t[f];
      !n[f] && y.cellMergeOwner && V(r, y.locatorRange) && (n[f] = !0, e = !0);
    }
    if (!e)
      break;
  }
}
var P = [];
function V(r, n) {
  return !q(r[0], n[0]) || !q(r[1], n[1]) ? !1 : (r[0][0] = b(r[0][0], n[0][0]), r[0][1] = Y(r[0][1], n[0][1]), r[1][0] = b(r[1][0], n[1][0]), r[1][1] = Y(r[1][1], n[1][1]), !0);
}
function q(r, n) {
  return r[1] >= n[0] && r[0] <= n[1];
}
function Z(r, n) {
  r.id.set(n[0][0], n[1][0]), r.span.set(n[0][1] - r.id.x + 1, n[1][1] - r.id.y + 1);
}
function $(r, n) {
  r[0][0] = n[0][0], r[0][1] = n[0][1], r[1][0] = n[1][0], r[1][1] = n[1][1];
}
function h(r, n, t, a) {
  var f = A(n[a][0], t, a), e = A(n[a][1], t, a);
  r[v[a]] = r[E[a]] = NaN, f && e && (r[v[a]] = f.xy, r[E[a]] = e.xy + e.wh - f.xy);
}
function s(r, n, t, a) {
  return r[v[n]] = t, r[v[1 - n]] = a, r;
}
function W(r) {
  return r && (r.type === M.leaf || r.type === M.nonLeaf) ? r : null;
}
function u() {
  return {
    x: NaN,
    y: NaN,
    width: NaN,
    height: NaN
  };
}
export {
  M as MatrixCellLayoutInfoType,
  T as MatrixClampOption,
  $ as cloneXYLocatorRange,
  A as coordDataToAllCellLevelLayout,
  u as createNaNRectLike,
  Z as fillIdSpanFromLocatorRange,
  K as isXYLocatorRangeInvalidOnDim,
  J as parseCoordRangeOption,
  G as resetXYLocatorRange,
  Q as resolveXYLocatorRangeByCellMerge,
  s as setDimXYValue,
  h as xyLocatorRangeToRectOneDim
};
