// echarts@6.0.0/lib/coord/matrix/MatrixDim.js
import { isArray as M, createHashMap as U, defaults as B, each as g, isString as L, isObject as E, eqNaN as G } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import m from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/Point.js";
import D from "../../data/OrdinalMeta.js";
import O from "../../scale/Ordinal.js";
import { XY as w, WH as R } from "../../util/graphic.js";
import { ListIterator as x } from "../../util/model.js";
import { setDimXYValue as y, createNaNRectLike as V, MatrixCellLayoutInfoType as I } from "./matrixCoordHelper.js";
import { error as q } from "../../util/log.js";
import { mathMax as b } from "../../util/number.js";
var z = (
  /** @class */
  (function() {
    function a(e, t) {
      this._cells = [], this._levels = [], this.dim = e, this.dimIdx = e === "x" ? 0 : 1, this._model = t, this._uniqueValueGen = k(e);
      var i = t.get("data", !0);
      i != null && !M(i) && (process.env.NODE_ENV !== "production" && q("Illegal echarts option - matrix." + this.dim + ".data must be an array if specified."), i = []), i ? this._initByDimModelData(i) : this._initBySeriesData();
    }
    return a.prototype._initByDimModelData = function(e) {
      var t = this, i = t._cells, n = t._levels, l = [], o = 0;
      t._leavesCount = p(e, 0, 0), h();
      return;
      function p(u, r, s) {
        var f = 0;
        return u && g(u, function(v, c) {
          var d = !1, _;
          L(v) ? _ = {
            value: v
          } : E(v) ? (_ = v, v.value != null && !L(v.value) && (d = !0, _ = {
            value: null
          })) : (_ = {
            value: null
          }, v != null && (d = !0)), d && process.env.NODE_ENV !== "production" && q("Illegal echarts option - matrix." + t.dim + ".data[" + c + "] must be `string | {value: string}`.");
          var C = {
            type: I.nonLeaf,
            ordinal: NaN,
            level: s,
            firstLeafLocator: r,
            id: new m(),
            span: y(new m(), t.dimIdx, 1, 1),
            option: _,
            xy: NaN,
            wh: NaN,
            dim: t,
            rect: V()
          };
          o++, (l[r] || (l[r] = [])).push(C), n[s] || (n[s] = {
            type: I.level,
            xy: NaN,
            wh: NaN,
            option: null,
            id: new m(),
            dim: t
          });
          var S = p(_.children, r, s + 1), N = Math.max(1, S);
          C.span[w[t.dimIdx]] = N, f += N, r += N;
        }), f;
      }
      function h() {
        for (var u = []; i.length < o; )
          for (var r = 0; r < l.length; r++) {
            var s = l[r].pop();
            if (s) {
              s.ordinal = u.length;
              var f = s.option.value;
              u.push(f), i.push(s), t._uniqueValueGen.calcDupBase(f);
            }
          }
        t._uniqueValueGen.ensureValueUnique(u, i);
        var v = t._ordinalMeta = new D({
          categories: u,
          needCollect: !1,
          deduplication: !1
        });
        t._scale = new O({
          ordinalMeta: v
        });
        for (var c = 0; c < t._leavesCount; c++) {
          var d = t._cells[c];
          d.type = I.leaf, d.span[w[1 - t.dimIdx]] = t._levels.length - d.level;
        }
        t._initCellsId(), t._initLevelIdOptions();
      }
    }, a.prototype._initBySeriesData = function() {
      var e = this;
      e._leavesCount = 0, e._levels = [{
        type: I.level,
        xy: NaN,
        wh: NaN,
        option: null,
        id: new m(),
        dim: e
      }], e._initLevelIdOptions();
      var t = e._ordinalMeta = new D({
        needCollect: !0,
        deduplication: !0,
        onCollect: function(i, n) {
          var l = e._cells[n] = {
            type: I.leaf,
            ordinal: n,
            level: 0,
            firstLeafLocator: n,
            id: new m(),
            span: y(new m(), e.dimIdx, 1, 1),
            // Theoretically `value` is from `dataset` or `series.data`, so it may be any type.
            // Do not restrict this case for user's convenience, and here simply convert it to
            // string for display.
            option: {
              value: i + ""
            },
            xy: NaN,
            wh: NaN,
            dim: e,
            rect: V()
          };
          e._leavesCount++, e._setCellId(l);
        }
      });
      e._scale = new O({
        ordinalMeta: t
      });
    }, a.prototype._setCellId = function(e) {
      var t = this._levels.length, i = this.dimIdx;
      y(e.id, i, e.firstLeafLocator, e.level - t);
    }, a.prototype._initCellsId = function() {
      var e = this._levels.length, t = this.dimIdx;
      g(this._cells, function(i) {
        y(i.id, t, i.firstLeafLocator, i.level - e);
      });
    }, a.prototype._initLevelIdOptions = function() {
      var e = this._levels.length, t = this.dimIdx, i = this._model.get("levels", !0);
      i = M(i) ? i : [], g(this._levels, function(n, l) {
        y(n.id, t, 0, l - e), n.option = i[l];
      });
    }, a.prototype.shouldShow = function() {
      return !!this._model.getShallow("show", !0);
    }, a.prototype.resetLayoutIterator = function(e, t, i, n) {
      if (e = e || new x(), t === this.dimIdx) {
        var l = this._leavesCount, o = i != null ? Math.max(0, i) : 0;
        n = n != null ? Math.min(n, l) : l, e.reset(this._cells, o, o + n);
      } else {
        var l = this._levels.length, o = i != null ? Math.max(0, i + l) : 0;
        n = n != null ? Math.min(n, l) : l, e.reset(this._levels, o, o + n);
      }
      return e;
    }, a.prototype.resetCellIterator = function(e) {
      return (e || new x()).reset(this._cells, 0);
    }, a.prototype.resetLevelIterator = function(e) {
      return (e || new x()).reset(this._levels, 0);
    }, a.prototype.getLayout = function(e, t, i) {
      var n = this.getUnitLayoutInfo(t, i);
      e[w[t]] = n ? n.xy : NaN, e[R[t]] = n ? n.wh : NaN;
    }, a.prototype.getUnitLayoutInfo = function(e, t) {
      return e === this.dimIdx ? t < this._leavesCount ? this._cells[t] : void 0 : this._levels[t + this._levels.length];
    }, a.prototype.getCell = function(e) {
      var t = this._scale.parse(e);
      return G(t) ? void 0 : this._cells[t];
    }, a.prototype.getLocatorCount = function(e) {
      return e === this.dimIdx ? this._leavesCount : this._levels.length;
    }, a.prototype.getOrdinalMeta = function() {
      return this._ordinalMeta;
    }, a;
  })()
);
function k(a) {
  var e = a.toUpperCase(), t = new RegExp("^" + e + "([0-9]+)$"), i = 0;
  function n(p) {
    var h;
    p != null && (h = p.match(t)) && (i = b(i, +h[1] + 1));
  }
  function l() {
    return "" + e + i++;
  }
  function o(p, h) {
    for (var u = U(), r = 0; r < p.length; r++) {
      var s = p[r];
      (s == null || u.get(s) != null) && (p[r] = s = l(), h[r].option = B({
        value: s
      }, h[r].option)), u.set(s, !0);
    }
  }
  return {
    calcDupBase: n,
    ensureValueUnique: o
  };
}
export {
  z as MatrixDim
};
