// echarts@6.0.0/lib/component/visualMap/ContinuousModel.js
import { __extends as S } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { isArray as R, each as O, retrieve2 as I } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import g from "./VisualMapModel.js";
import { asc as M } from "../../util/number.js";
import { inheritDefaultOption as x } from "../../util/component.js";
import h from "../../visual/tokens.js";
var d = [20, 140], L = (
  /** @class */
  (function(l) {
    S(r, l);
    function r() {
      var t = l !== null && l.apply(this, arguments) || this;
      return t.type = r.type, t;
    }
    return r.prototype.optionUpdated = function(t, e) {
      l.prototype.optionUpdated.apply(this, arguments), this.resetExtent(), this.resetVisual(function(o) {
        o.mappingMethod = "linear", o.dataExtent = this.getExtent();
      }), this._resetRange();
    }, r.prototype.resetItemSize = function() {
      l.prototype.resetItemSize.apply(this, arguments);
      var t = this.itemSize;
      (t[0] == null || isNaN(t[0])) && (t[0] = d[0]), (t[1] == null || isNaN(t[1])) && (t[1] = d[1]);
    }, r.prototype._resetRange = function() {
      var t = this.getExtent(), e = this.option.range;
      !e || e.auto ? (t.auto = 1, this.option.range = t) : R(e) && (e[0] > e[1] && e.reverse(), e[0] = Math.max(e[0], t[0]), e[1] = Math.min(e[1], t[1]));
    }, r.prototype.completeVisualOption = function() {
      l.prototype.completeVisualOption.apply(this, arguments), O(this.stateList, function(t) {
        var e = this.option.controller[t].symbolSize;
        e && e[0] !== e[1] && (e[0] = e[1] / 3);
      }, this);
    }, r.prototype.setSelected = function(t) {
      this.option.range = t.slice(), this._resetRange();
    }, r.prototype.getSelected = function() {
      var t = this.getExtent(), e = M((this.get("range") || []).slice());
      return e[0] > t[1] && (e[0] = t[1]), e[1] > t[1] && (e[1] = t[1]), e[0] < t[0] && (e[0] = t[0]), e[1] < t[0] && (e[1] = t[0]), e;
    }, r.prototype.getValueState = function(t) {
      var e = this.option.range, o = this.getExtent(), n = I(this.option.unboundedRange, !0);
      return (n && e[0] <= o[0] || e[0] <= t) && (n && e[1] >= o[1] || t <= e[1]) ? "inRange" : "outOfRange";
    }, r.prototype.findTargetDataIndices = function(t) {
      var e = [];
      return this.eachTargetSeries(function(o) {
        var n = [], a = o.getData();
        a.each(this.getDataDimensionIndex(a), function(i, s) {
          t[0] <= i && i <= t[1] && n.push(s);
        }, this), e.push({
          seriesId: o.id,
          dataIndex: n
        });
      }, this), e;
    }, r.prototype.getVisualMeta = function(t) {
      var e = v(this, "outOfRange", this.getExtent()), o = v(this, "inRange", this.option.range.slice()), n = [];
      function a(c, y) {
        n.push({
          value: c,
          color: t(c, y)
        });
      }
      for (var i = 0, s = 0, m = o.length, f = e.length; s < f && (!o.length || e[s] <= o[0]); s++)
        e[s] < o[i] && a(e[s], "outOfRange");
      for (var u = 1; i < m; i++, u = 0)
        u && n.length && a(o[i], "outOfRange"), a(o[i], "inRange");
      for (var u = 1; s < f; s++)
        (!o.length || o[o.length - 1] < e[s]) && (u && (n.length && a(n[n.length - 1].value, "outOfRange"), u = 0), a(e[s], "outOfRange"));
      var p = n.length;
      return {
        stops: n,
        outerColors: [p ? n[0].color : "transparent", p ? n[p - 1].color : "transparent"]
      };
    }, r.type = "visualMap.continuous", r.defaultOption = x(g.defaultOption, {
      align: "auto",
      calculable: !1,
      hoverLink: !0,
      realtime: !0,
      handleIcon: "path://M-11.39,9.77h0a3.5,3.5,0,0,1-3.5,3.5h-22a3.5,3.5,0,0,1-3.5-3.5h0a3.5,3.5,0,0,1,3.5-3.5h22A3.5,3.5,0,0,1-11.39,9.77Z",
      handleSize: "120%",
      handleStyle: {
        borderColor: h.color.neutral00,
        borderWidth: 1
      },
      indicatorIcon: "circle",
      indicatorSize: "50%",
      indicatorStyle: {
        borderColor: h.color.neutral00,
        borderWidth: 2,
        shadowBlur: 2,
        shadowOffsetX: 1,
        shadowOffsetY: 1,
        shadowColor: h.color.shadow
      }
      // emphasis: {
      //     handleStyle: {
      //         shadowBlur: 3,
      //         shadowOffsetX: 1,
      //         shadowOffsetY: 1,
      //         shadowColor: tokens.color.shadow
      //     }
      // }
    }), r;
  })(g)
);
function v(l, r, t) {
  if (t[0] === t[1])
    return t.slice();
  for (var e = 200, o = (t[1] - t[0]) / e, n = t[0], a = [], i = 0; i <= e && n < t[1]; i++)
    a.push(n), n += o;
  return a.push(t[1]), a;
}
export {
  L as default
};
