// echarts@6.0.0/lib/component/visualMap/ContinuousView.js
import { __extends as Z } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { bind as y, extend as Y, each as K, isArray as $ } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import U from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/LinearGradient.js";
import { stop as w } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/event.js";
import W from "./VisualMapView.js";
import { applyTransform as x, getTransform as L, transformDirection as j } from "../../util/graphic.js";
import { linearMap as q } from "../../util/number.js";
import J from "../helper/sliderMove.js";
import { getItemAlign as Q, makeHighDownBatch as H } from "./helper.js";
import { compressBatches as tt } from "../../util/model.js";
import { parsePercent as C } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/contain/text.js";
import { setAsHighDownDispatcher as et } from "../../util/states.js";
import { createSymbol as P } from "../../util/symbol.js";
import it from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Image.js";
import { getECData as rt } from "../../util/innerStore.js";
import { createTextStyle as z } from "../../label/labelStyle.js";
import { findEventDispatcher as at } from "../../util/event.js";
import T from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Text.js";
import O from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import ot from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Rect.js";
import nt from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Polygon.js";
var f = q, st = K, B = Math.min, k = Math.max, lt = 12, ht = 6, Dt = (
  /** @class */
  (function(c) {
    Z(l, c);
    function l() {
      var t = c !== null && c.apply(this, arguments) || this;
      return t.type = l.type, t._shapes = {}, t._dataInterval = [], t._handleEnds = [], t._hoverLinkDataIndices = [], t;
    }
    return l.prototype.init = function(t, i) {
      c.prototype.init.call(this, t, i), this._hoverLinkFromSeriesMouseOver = y(this._hoverLinkFromSeriesMouseOver, this), this._hideIndicator = y(this._hideIndicator, this);
    }, l.prototype.doRender = function(t, i, e, r) {
      (!r || r.type !== "selectDataRange" || r.from !== this.uid) && this._buildView();
    }, l.prototype._buildView = function() {
      this.group.removeAll();
      var t = this.visualMapModel, i = this.group;
      this._orient = t.get("orient"), this._useHandle = t.get("calculable"), this._resetInterval(), this._renderBar(i);
      var e = t.get("text");
      this._renderEndsText(i, e, 0), this._renderEndsText(i, e, 1), this._updateView(!0), this.renderBackground(i), this._updateView(), this._enableHoverLinkToSeries(), this._enableHoverLinkFromSeries(), this.positionGroup(i);
    }, l.prototype._renderEndsText = function(t, i, e) {
      if (i) {
        var r = i[1 - e];
        r = r != null ? r + "" : "";
        var a = this.visualMapModel, s = a.get("textGap"), n = a.itemSize, v = this._shapes.mainGroup, h = this._applyTransform([n[0] / 2, e === 0 ? -s : n[1] + s], v), o = this._applyTransform(e === 0 ? "bottom" : "top", v), u = this._orient, p = this.visualMapModel.textStyleModel;
        this.group.add(new T({
          style: z(p, {
            x: h[0],
            y: h[1],
            verticalAlign: p.get("verticalAlign") || (u === "horizontal" ? "middle" : o),
            align: p.get("align") || (u === "horizontal" ? o : "center"),
            text: r
          })
        }));
      }
    }, l.prototype._renderBar = function(t) {
      var i = this.visualMapModel, e = this._shapes, r = i.itemSize, a = this._orient, s = this._useHandle, n = Q(i, this.api, r), v = e.mainGroup = this._createBarGroup(n), h = new O();
      v.add(h), h.add(e.outOfRange = A()), h.add(e.inRange = A(null, s ? N(this._orient) : null, y(this._dragHandle, this, "all", !1), y(this._dragHandle, this, "all", !0))), h.setClipPath(new ot({
        shape: {
          x: 0,
          y: 0,
          width: r[0],
          height: r[1],
          r: 3
        }
      }));
      var o = i.textStyleModel.getTextRect("国"), u = k(o.width, o.height);
      s && (e.handleThumbs = [], e.handleLabels = [], e.handleLabelPoints = [], this._createHandle(i, v, 0, r, u, a), this._createHandle(i, v, 1, r, u, a)), this._createIndicator(i, v, r, u, a), t.add(v);
    }, l.prototype._createHandle = function(t, i, e, r, a, s) {
      var n = y(this._dragHandle, this, e, !1), v = y(this._dragHandle, this, e, !0), h = C(t.get("handleSize"), r[0]), o = P(t.get("handleIcon"), -h / 2, -h / 2, h, h, null, !0), u = N(this._orient);
      o.attr({
        cursor: u,
        draggable: !0,
        drift: n,
        ondragend: v,
        onmousemove: function(_) {
          w(_.event);
        }
      }), o.x = r[0] / 2, o.useStyle(t.getModel("handleStyle").getItemStyle()), o.setStyle({
        strokeNoScale: !0,
        strokeFirst: !0
      }), o.style.lineWidth *= 2, o.ensureState("emphasis").style = t.getModel(["emphasis", "handleStyle"]).getItemStyle(), et(o, !0), i.add(o);
      var p = this.visualMapModel.textStyleModel, d = new T({
        cursor: u,
        draggable: !0,
        drift: n,
        onmousemove: function(_) {
          w(_.event);
        },
        ondragend: v,
        style: z(p, {
          x: 0,
          y: 0,
          text: ""
        })
      });
      d.ensureState("blur").style = {
        opacity: 0.1
      }, d.stateTransition = {
        duration: 200
      }, this.group.add(d);
      var m = [h, 0], g = this._shapes;
      g.handleThumbs[e] = o, g.handleLabelPoints[e] = m, g.handleLabels[e] = d;
    }, l.prototype._createIndicator = function(t, i, e, r, a) {
      var s = C(t.get("indicatorSize"), e[0]), n = P(t.get("indicatorIcon"), -s / 2, -s / 2, s, s, null, !0);
      n.attr({
        cursor: "move",
        invisible: !0,
        silent: !0,
        x: e[0] / 2
      });
      var v = t.getModel("indicatorStyle").getItemStyle();
      if (n instanceof it) {
        var h = n.style;
        n.useStyle(Y({
          // TODO other properties like x, y ?
          image: h.image,
          x: h.x,
          y: h.y,
          width: h.width,
          height: h.height
        }, v));
      } else
        n.useStyle(v);
      i.add(n);
      var o = this.visualMapModel.textStyleModel, u = new T({
        silent: !0,
        invisible: !0,
        style: z(o, {
          x: 0,
          y: 0,
          text: ""
        })
      });
      this.group.add(u);
      var p = [(a === "horizontal" ? r / 2 : ht) + e[0] / 2, 0], d = this._shapes;
      d.indicator = n, d.indicatorLabel = u, d.indicatorLabelPoint = p, this._firstShowIndicator = !0;
    }, l.prototype._dragHandle = function(t, i, e, r) {
      if (this._useHandle) {
        if (this._dragging = !i, !i) {
          var a = this._applyTransform([e, r], this._shapes.mainGroup, !0);
          this._updateInterval(t, a[1]), this._hideIndicator(), this._updateView();
        }
        i === !this.visualMapModel.get("realtime") && this.api.dispatchAction({
          type: "selectDataRange",
          from: this.uid,
          visualMapId: this.visualMapModel.id,
          selected: this._dataInterval.slice()
        }), i ? !this._hovering && this._clearHoverLinkToSeries() : F(this.visualMapModel) && this._doHoverLinkToSeries(this._handleEnds[t], !1);
      }
    }, l.prototype._resetInterval = function() {
      var t = this.visualMapModel, i = this._dataInterval = t.getSelected(), e = t.getExtent(), r = [0, t.itemSize[1]];
      this._handleEnds = [f(i[0], e, r, !0), f(i[1], e, r, !0)];
    }, l.prototype._updateInterval = function(t, i) {
      i = i || 0;
      var e = this.visualMapModel, r = this._handleEnds, a = [0, e.itemSize[1]];
      J(
        i,
        r,
        a,
        t,
        // cross is forbidden
        0
      );
      var s = e.getExtent();
      this._dataInterval = [f(r[0], a, s, !0), f(r[1], a, s, !0)];
    }, l.prototype._updateView = function(t) {
      var i = this.visualMapModel, e = i.getExtent(), r = this._shapes, a = [0, i.itemSize[1]], s = t ? a : this._handleEnds, n = this._createBarVisual(this._dataInterval, e, s, "inRange"), v = this._createBarVisual(e, e, a, "outOfRange");
      r.inRange.setStyle({
        fill: n.barColor
        // opacity: visualInRange.opacity
      }).setShape("points", n.barPoints), r.outOfRange.setStyle({
        fill: v.barColor
        // opacity: visualOutOfRange.opacity
      }).setShape("points", v.barPoints), this._updateHandle(s, n);
    }, l.prototype._createBarVisual = function(t, i, e, r) {
      var a = {
        forceState: r,
        convertOpacityToAlpha: !0
      }, s = this._makeColorGradient(t, a), n = [this.getControllerVisual(t[0], "symbolSize", a), this.getControllerVisual(t[1], "symbolSize", a)], v = this._createBarPoints(e, n);
      return {
        barColor: new U(0, 0, 0, 1, s),
        barPoints: v,
        handlesColor: [s[0].color, s[s.length - 1].color]
      };
    }, l.prototype._makeColorGradient = function(t, i) {
      var e = 100, r = [], a = (t[1] - t[0]) / e;
      r.push({
        color: this.getControllerVisual(t[0], "color", i),
        offset: 0
      });
      for (var s = 1; s < e; s++) {
        var n = t[0] + a * s;
        if (n > t[1])
          break;
        r.push({
          color: this.getControllerVisual(n, "color", i),
          offset: s / e
        });
      }
      return r.push({
        color: this.getControllerVisual(t[1], "color", i),
        offset: 1
      }), r;
    }, l.prototype._createBarPoints = function(t, i) {
      var e = this.visualMapModel.itemSize;
      return [[e[0] - i[0], t[0]], [e[0], t[0]], [e[0], t[1]], [e[0] - i[1], t[1]]];
    }, l.prototype._createBarGroup = function(t) {
      var i = this._orient, e = this.visualMapModel.get("inverse");
      return new O(i === "horizontal" && !e ? {
        scaleX: t === "bottom" ? 1 : -1,
        rotation: Math.PI / 2
      } : i === "horizontal" && e ? {
        scaleX: t === "bottom" ? -1 : 1,
        rotation: -Math.PI / 2
      } : i === "vertical" && !e ? {
        scaleX: t === "left" ? 1 : -1,
        scaleY: -1
      } : {
        scaleX: t === "left" ? 1 : -1
      });
    }, l.prototype._updateHandle = function(t, i) {
      if (this._useHandle) {
        var e = this._shapes, r = this.visualMapModel, a = e.handleThumbs, s = e.handleLabels, n = r.itemSize, v = r.getExtent(), h = this._applyTransform("left", e.mainGroup);
        st([0, 1], function(o) {
          var u = a[o];
          u.setStyle("fill", i.handlesColor[o]), u.y = t[o];
          var p = f(t[o], [0, n[1]], v, !0), d = this.getControllerVisual(p, "symbolSize");
          u.scaleX = u.scaleY = d / n[0], u.x = n[0] - d / 2;
          var m = x(e.handleLabelPoints[o], L(u, this.group));
          if (this._orient === "horizontal") {
            var g = h === "left" || h === "top" ? (n[0] - d) / 2 : (n[0] - d) / -2;
            m[1] += g;
          }
          s[o].setStyle({
            x: m[0],
            y: m[1],
            text: r.formatValueText(this._dataInterval[o]),
            verticalAlign: "middle",
            align: this._orient === "vertical" ? this._applyTransform("left", e.mainGroup) : "center"
          });
        }, this);
      }
    }, l.prototype._showIndicator = function(t, i, e, r) {
      var a = this.visualMapModel, s = a.getExtent(), n = a.itemSize, v = [0, n[1]], h = this._shapes, o = h.indicator;
      if (o) {
        o.attr("invisible", !1);
        var u = {
          convertOpacityToAlpha: !0
        }, p = this.getControllerVisual(t, "color", u), d = this.getControllerVisual(t, "symbolSize"), m = f(t, s, v, !0), g = n[0] - d / 2, _ = {
          x: o.x,
          y: o.y
        };
        o.y = m, o.x = g;
        var I = x(h.indicatorLabelPoint, L(o, this.group)), M = h.indicatorLabel;
        M.attr("invisible", !1);
        var E = this._applyTransform("left", h.mainGroup), X = this._orient, D = X === "horizontal";
        M.setStyle({
          text: (e || "") + a.formatValueText(i),
          verticalAlign: D ? E : "middle",
          align: D ? "center" : E
        });
        var V = {
          x: g,
          y: m,
          style: {
            fill: p
          }
        }, R = {
          style: {
            x: I[0],
            y: I[1]
          }
        };
        if (a.ecModel.isAnimationEnabled() && !this._firstShowIndicator) {
          var G = {
            duration: 100,
            easing: "cubicInOut",
            additive: !0
          };
          o.x = _.x, o.y = _.y, o.animateTo(V, G), M.animateTo(R, G);
        } else
          o.attr(V), M.attr(R);
        this._firstShowIndicator = !1;
        var S = this._shapes.handleLabels;
        if (S)
          for (var b = 0; b < S.length; b++)
            this.api.enterBlur(S[b]);
      }
    }, l.prototype._enableHoverLinkToSeries = function() {
      var t = this;
      this._shapes.mainGroup.on("mousemove", function(i) {
        if (t._hovering = !0, !t._dragging) {
          var e = t.visualMapModel.itemSize, r = t._applyTransform([i.offsetX, i.offsetY], t._shapes.mainGroup, !0, !0);
          r[1] = B(k(0, r[1]), e[1]), t._doHoverLinkToSeries(r[1], 0 <= r[0] && r[0] <= e[0]);
        }
      }).on("mouseout", function() {
        t._hovering = !1, !t._dragging && t._clearHoverLinkToSeries();
      });
    }, l.prototype._enableHoverLinkFromSeries = function() {
      var t = this.api.getZr();
      this.visualMapModel.option.hoverLink ? (t.on("mouseover", this._hoverLinkFromSeriesMouseOver, this), t.on("mouseout", this._hideIndicator, this)) : this._clearHoverLinkFromSeries();
    }, l.prototype._doHoverLinkToSeries = function(t, i) {
      var e = this.visualMapModel, r = e.itemSize;
      if (e.option.hoverLink) {
        var a = [0, r[1]], s = e.getExtent();
        t = B(k(a[0], t), a[1]);
        var n = vt(e, s, a), v = [t - n, t + n], h = f(t, a, s, !0), o = [f(v[0], a, s, !0), f(v[1], a, s, !0)];
        v[0] < a[0] && (o[0] = -1 / 0), v[1] > a[1] && (o[1] = 1 / 0), i && (o[0] === -1 / 0 ? this._showIndicator(h, o[1], "< ", n) : o[1] === 1 / 0 ? this._showIndicator(h, o[0], "> ", n) : this._showIndicator(h, h, "≈ ", n));
        var u = this._hoverLinkDataIndices, p = [];
        (i || F(e)) && (p = this._hoverLinkDataIndices = e.findTargetDataIndices(o));
        var d = tt(u, p);
        this._dispatchHighDown("downplay", H(d[0], e)), this._dispatchHighDown("highlight", H(d[1], e));
      }
    }, l.prototype._hoverLinkFromSeriesMouseOver = function(t) {
      var i;
      if (at(t.target, function(n) {
        var v = rt(n);
        if (v.dataIndex != null)
          return i = v, !0;
      }, !0), !!i) {
        var e = this.ecModel.getSeriesByIndex(i.seriesIndex), r = this.visualMapModel;
        if (r.isTargetSeries(e)) {
          var a = e.getData(i.dataType), s = a.getStore().get(r.getDataDimensionIndex(a), i.dataIndex);
          isNaN(s) || this._showIndicator(s, s);
        }
      }
    }, l.prototype._hideIndicator = function() {
      var t = this._shapes;
      t.indicator && t.indicator.attr("invisible", !0), t.indicatorLabel && t.indicatorLabel.attr("invisible", !0);
      var i = this._shapes.handleLabels;
      if (i)
        for (var e = 0; e < i.length; e++)
          this.api.leaveBlur(i[e]);
    }, l.prototype._clearHoverLinkToSeries = function() {
      this._hideIndicator();
      var t = this._hoverLinkDataIndices;
      this._dispatchHighDown("downplay", H(t, this.visualMapModel)), t.length = 0;
    }, l.prototype._clearHoverLinkFromSeries = function() {
      this._hideIndicator();
      var t = this.api.getZr();
      t.off("mouseover", this._hoverLinkFromSeriesMouseOver), t.off("mouseout", this._hideIndicator);
    }, l.prototype._applyTransform = function(t, i, e, r) {
      var a = L(i, r ? null : this.group);
      return $(t) ? x(t, a, e) : j(t, a, e);
    }, l.prototype._dispatchHighDown = function(t, i) {
      i && i.length && this.api.dispatchAction({
        type: t,
        batch: i
      });
    }, l.prototype.dispose = function() {
      this._clearHoverLinkFromSeries(), this._clearHoverLinkToSeries();
    }, l.type = "visualMap.continuous", l;
  })(W)
);
function A(c, l, t, i) {
  return new nt({
    shape: {
      points: c
    },
    draggable: !!t,
    cursor: l,
    drift: t,
    onmousemove: function(e) {
      w(e.event);
    },
    ondragend: i
  });
}
function vt(c, l, t) {
  var i = lt / 2, e = c.get("hoverLinkDataSize");
  return e && (i = f(e, l, t, !0) / 2), i;
}
function F(c) {
  var l = c.get("hoverLinkOnHandle");
  return !!(l ?? c.get("realtime"));
}
function N(c) {
  return c === "vertical" ? "ns-resize" : "ew-resize";
}
export {
  Dt as default
};
