// echarts@6.0.0/lib/component/visualMap/PiecewiseModel.js
import { __extends as I } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { clone as y, map as V, each as h, isObject as M } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import m from "./VisualMapModel.js";
import d from "../../visual/VisualMapping.js";
import O from "../../visual/visualDefault.js";
import { reformIntervals as g } from "../../util/number.js";
import { inheritDefaultOption as b } from "../../util/component.js";
var K = (
  /** @class */
  (function(l) {
    I(s, l);
    function s() {
      var t = l !== null && l.apply(this, arguments) || this;
      return t.type = s.type, t._pieceList = [], t;
    }
    return s.prototype.optionUpdated = function(t, n) {
      l.prototype.optionUpdated.apply(this, arguments), this.resetExtent();
      var e = this._mode = this._determineMode();
      this._pieceList = [], S[this._mode].call(this, this._pieceList), this._resetSelected(t, n);
      var u = this.option.categories;
      this.resetVisual(function(i, r) {
        e === "categories" ? (i.mappingMethod = "category", i.categories = y(u)) : (i.dataExtent = this.getExtent(), i.mappingMethod = "piecewise", i.pieceList = V(this._pieceList, function(o) {
          return o = y(o), r !== "inRange" && (o.visual = null), o;
        }));
      });
    }, s.prototype.completeVisualOption = function() {
      var t = this.option, n = {}, e = d.listVisualTypes(), u = this.isCategory();
      h(t.pieces, function(r) {
        h(e, function(o) {
          r.hasOwnProperty(o) && (n[o] = 1);
        });
      }), h(n, function(r, o) {
        var f = !1;
        h(this.stateList, function(a) {
          f = f || i(t, a, o) || i(t.target, a, o);
        }, this), !f && h(this.stateList, function(a) {
          (t[a] || (t[a] = {}))[o] = O.get(o, a === "inRange" ? "active" : "inactive", u);
        });
      }, this);
      function i(r, o, f) {
        return r && r[o] && r[o].hasOwnProperty(f);
      }
      l.prototype.completeVisualOption.apply(this, arguments);
    }, s.prototype._resetSelected = function(t, n) {
      var e = this.option, u = this._pieceList, i = (n ? e : t).selected || {};
      if (e.selected = i, h(u, function(o, f) {
        var a = this.getSelectedMapKey(o);
        i.hasOwnProperty(a) || (i[a] = !0);
      }, this), e.selectedMode === "single") {
        var r = !1;
        h(u, function(o, f) {
          var a = this.getSelectedMapKey(o);
          i[a] && (r ? i[a] = !1 : r = !0);
        }, this);
      }
    }, s.prototype.getItemSymbol = function() {
      return this.get("itemSymbol");
    }, s.prototype.getSelectedMapKey = function(t) {
      return this._mode === "categories" ? t.value + "" : t.index + "";
    }, s.prototype.getPieceList = function() {
      return this._pieceList;
    }, s.prototype._determineMode = function() {
      var t = this.option;
      return t.pieces && t.pieces.length > 0 ? "pieces" : this.option.categories ? "categories" : "splitNumber";
    }, s.prototype.setSelected = function(t) {
      this.option.selected = y(t);
    }, s.prototype.getValueState = function(t) {
      var n = d.findPieceIndex(t, this._pieceList);
      return n != null && this.option.selected[this.getSelectedMapKey(this._pieceList[n])] ? "inRange" : "outOfRange";
    }, s.prototype.findTargetDataIndices = function(t) {
      var n = [], e = this._pieceList;
      return this.eachTargetSeries(function(u) {
        var i = [], r = u.getData();
        r.each(this.getDataDimensionIndex(r), function(o, f) {
          var a = d.findPieceIndex(o, e);
          a === t && i.push(f);
        }, this), n.push({
          seriesId: u.id,
          dataIndex: i
        });
      }, this), n;
    }, s.prototype.getRepresentValue = function(t) {
      var n;
      if (this.isCategory())
        n = t.value;
      else if (t.value != null)
        n = t.value;
      else {
        var e = t.interval || [];
        n = e[0] === -1 / 0 && e[1] === 1 / 0 ? 0 : (e[0] + e[1]) / 2;
      }
      return n;
    }, s.prototype.getVisualMeta = function(t) {
      if (this.isCategory())
        return;
      var n = [], e = ["", ""], u = this;
      function i(a, p) {
        var v = u.getRepresentValue({
          interval: a
        });
        p || (p = u.getValueState(v));
        var c = t(v, p);
        a[0] === -1 / 0 ? e[0] = c : a[1] === 1 / 0 ? e[1] = c : n.push({
          value: a[0],
          color: c
        }, {
          value: a[1],
          color: c
        });
      }
      var r = this._pieceList.slice();
      if (!r.length)
        r.push({
          interval: [-1 / 0, 1 / 0]
        });
      else {
        var o = r[0].interval[0];
        o !== -1 / 0 && r.unshift({
          interval: [-1 / 0, o]
        }), o = r[r.length - 1].interval[1], o !== 1 / 0 && r.push({
          interval: [o, 1 / 0]
        });
      }
      var f = -1 / 0;
      return h(r, function(a) {
        var p = a.interval;
        p && (p[0] > f && i([f, p[0]], "outOfRange"), i(p.slice()), f = p[1]);
      }, this), {
        stops: n,
        outerColors: e
      };
    }, s.type = "visualMap.piecewise", s.defaultOption = b(m.defaultOption, {
      selected: null,
      minOpen: !1,
      maxOpen: !1,
      align: "auto",
      itemWidth: 20,
      itemHeight: 14,
      itemSymbol: "roundRect",
      pieces: null,
      categories: null,
      splitNumber: 5,
      selectedMode: "multiple",
      itemGap: 10,
      hoverLink: !0
      // Enable hover highlight.
    }), s;
  })(m)
), S = {
  splitNumber: function(l) {
    var s = this.option, t = Math.min(s.precision, 20), n = this.getExtent(), e = s.splitNumber;
    e = Math.max(parseInt(e, 10), 1), s.splitNumber = e;
    for (var u = (n[1] - n[0]) / e; +u.toFixed(t) !== u && t < 5; )
      t++;
    s.precision = t, u = +u.toFixed(t), s.minOpen && l.push({
      interval: [-1 / 0, n[0]],
      close: [0, 0]
    });
    for (var i = 0, r = n[0]; i < e; r += u, i++) {
      var o = i === e - 1 ? n[1] : r + u;
      l.push({
        interval: [r, o],
        close: [1, 1]
      });
    }
    s.maxOpen && l.push({
      interval: [n[1], 1 / 0],
      close: [0, 0]
    }), g(l), h(l, function(f, a) {
      f.index = a, f.text = this.formatValueText(f.interval);
    }, this);
  },
  categories: function(l) {
    var s = this.option;
    h(s.categories, function(t) {
      l.push({
        text: this.formatValueText(t, !0),
        value: t
      });
    }, this), x(s, l);
  },
  pieces: function(l) {
    var s = this.option;
    h(s.pieces, function(t, n) {
      M(t) || (t = {
        value: t
      });
      var e = {
        text: "",
        index: n
      };
      if (t.label != null && (e.text = t.label), t.hasOwnProperty("value")) {
        var u = e.value = t.value;
        e.interval = [u, u], e.close = [1, 1];
      } else {
        for (var i = e.interval = [], r = e.close = [0, 0], o = [1, 0, 1], f = [-1 / 0, 1 / 0], a = [], p = 0; p < 2; p++) {
          for (var v = [["gte", "gt", "min"], ["lte", "lt", "max"]][p], c = 0; c < 3 && i[p] == null; c++)
            i[p] = t[v[c]], r[p] = o[c], a[p] = c === 2;
          i[p] == null && (i[p] = f[p]);
        }
        a[0] && i[1] === 1 / 0 && (r[0] = 0), a[1] && i[0] === -1 / 0 && (r[1] = 0), process.env.NODE_ENV !== "production" && i[0] > i[1] && console.warn("Piece " + n + "is illegal: " + i + " lower bound should not greater then uppper bound."), i[0] === i[1] && r[0] && r[1] && (e.value = i[0]);
      }
      e.visual = d.retrieveVisuals(t), l.push(e);
    }, this), x(s, l), g(l), h(l, function(t) {
      var n = t.close, e = [["<", "≤"][n[1]], [">", "≥"][n[0]]];
      t.text = t.text || this.formatValueText(t.value != null ? t.value : t.interval, !1, e);
    }, this);
  }
};
function x(l, s) {
  var t = l.inverse;
  (l.orient === "vertical" ? !t : t) && s.reverse();
}
export {
  K as default
};
