// echarts@6.0.0/lib/component/visualMap/PiecewiseView.js
import { __extends as _ } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { retrieve as S, each as g, bind as V, retrieve2 as b, map as I, clone as L } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import T from "./VisualMapView.js";
import { createSymbol as A } from "../../util/symbol.js";
import { box as k } from "../../util/layout.js";
import { makeHighDownBatch as G, getItemAlign as D } from "./helper.js";
import { createTextStyle as f } from "../../label/labelStyle.js";
import y from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import x from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Text.js";
var Z = (
  /** @class */
  (function(d) {
    _(n, d);
    function n() {
      var e = d !== null && d.apply(this, arguments) || this;
      return e.type = n.type, e;
    }
    return n.prototype.doRender = function() {
      var e = this.group;
      e.removeAll();
      var i = this.visualMapModel, t = i.get("textGap"), a = i.textStyleModel, r = this._getItemAlign(), o = i.itemSize, s = this._getViewData(), l = s.endsText, v = S(i.get("showLabel", !0), !l), c = !i.get("selectedMode");
      l && this._renderEndsText(e, l[0], o, v, r), g(s.viewPieceList, function(h) {
        var u = h.piece, p = new y();
        p.onclick = V(this._onItemClick, this, u), this._enableHoverLink(p, h.indexInModelPieceList);
        var m = i.getRepresentValue(u);
        if (this._createItemSymbol(p, m, [0, 0, o[0], o[1]], c), v) {
          var w = this.visualMapModel.getValueState(m), M = a.get("align") || r;
          p.add(new x({
            style: f(a, {
              x: M === "right" ? -t : o[0] + t,
              y: o[1] / 2,
              text: u.text,
              verticalAlign: a.get("verticalAlign") || "middle",
              align: M,
              opacity: b(a.get("opacity"), w === "outOfRange" ? 0.5 : 1)
            }),
            silent: c
          }));
        }
        e.add(p);
      }, this), l && this._renderEndsText(e, l[1], o, v, r), k(i.get("orient"), e, i.get("itemGap")), this.renderBackground(e), this.positionGroup(e);
    }, n.prototype._enableHoverLink = function(e, i) {
      var t = this;
      e.on("mouseover", function() {
        return a("highlight");
      }).on("mouseout", function() {
        return a("downplay");
      });
      var a = function(r) {
        var o = t.visualMapModel;
        o.option.hoverLink && t.api.dispatchAction({
          type: r,
          batch: G(o.findTargetDataIndices(i), o)
        });
      };
    }, n.prototype._getItemAlign = function() {
      var e = this.visualMapModel, i = e.option;
      if (i.orient === "vertical")
        return D(e, this.api, e.itemSize);
      var t = i.align;
      return (!t || t === "auto") && (t = "left"), t;
    }, n.prototype._renderEndsText = function(e, i, t, a, r) {
      if (i) {
        var o = new y(), s = this.visualMapModel.textStyleModel;
        o.add(new x({
          style: f(s, {
            x: a ? r === "right" ? t[0] : 0 : t[0] / 2,
            y: t[1] / 2,
            verticalAlign: "middle",
            align: a ? r : "center",
            text: i
          })
        })), e.add(o);
      }
    }, n.prototype._getViewData = function() {
      var e = this.visualMapModel, i = I(e.getPieceList(), function(o, s) {
        return {
          piece: o,
          indexInModelPieceList: s
        };
      }), t = e.get("text"), a = e.get("orient"), r = e.get("inverse");
      return (a === "horizontal" ? r : !r) ? i.reverse() : t && (t = t.slice().reverse()), {
        viewPieceList: i,
        endsText: t
      };
    }, n.prototype._createItemSymbol = function(e, i, t, a) {
      var r = A(
        // symbol will be string
        this.getControllerVisual(i, "symbol"),
        t[0],
        t[1],
        t[2],
        t[3],
        // color will be string
        this.getControllerVisual(i, "color")
      );
      r.silent = a, e.add(r);
    }, n.prototype._onItemClick = function(e) {
      var i = this.visualMapModel, t = i.option, a = t.selectedMode;
      if (a) {
        var r = L(t.selected), o = i.getSelectedMapKey(e);
        a === "single" || a === !0 ? (r[o] = !0, g(r, function(s, l) {
          r[l] = l === o;
        })) : r[o] = !r[o], this.api.dispatchAction({
          type: "selectDataRange",
          from: this.uid,
          visualMapId: this.visualMapModel.id,
          selected: r
        });
      }
    }, n.type = "visualMap.piecewise", n;
  })(T)
);
export {
  Z as default
};
