// echarts@6.0.0/lib/component/visualMap/VisualMapView.js
import { __extends as v } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { each as g } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import "../../util/graphic.js";
import { normalizeCssArray as y } from "../../util/format.js";
import { createBoxLayoutReference as M, positionElement as V } from "../../util/layout.js";
import d from "../../visual/VisualMapping.js";
import w from "../../view/Component.js";
import x from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Rect.js";
var T = (
  /** @class */
  (function(l) {
    v(i, l);
    function i() {
      var t = l !== null && l.apply(this, arguments) || this;
      return t.type = i.type, t.autoPositionValues = {
        left: 1,
        right: 1,
        top: 1,
        bottom: 1
      }, t;
    }
    return i.prototype.init = function(t, o) {
      this.ecModel = t, this.api = o;
    }, i.prototype.render = function(t, o, e, r) {
      if (this.visualMapModel = t, t.get("show") === !1) {
        this.group.removeAll();
        return;
      }
      this.doRender(t, o, e, r);
    }, i.prototype.renderBackground = function(t) {
      var o = this.visualMapModel, e = y(o.get("padding") || 0), r = t.getBoundingRect();
      t.add(new x({
        z2: -1,
        silent: !0,
        shape: {
          x: r.x - e[3],
          y: r.y - e[0],
          width: r.width + e[3] + e[1],
          height: r.height + e[0] + e[2]
        },
        style: {
          fill: o.get("backgroundColor"),
          stroke: o.get("borderColor"),
          lineWidth: o.get("borderWidth")
        }
      }));
    }, i.prototype.getControllerVisual = function(t, o, e) {
      e = e || {};
      var r = e.forceState, s = this.visualMapModel, p = {};
      if (o === "color") {
        var c = s.get("contentColor");
        p.color = c;
      }
      function f(a) {
        return p[a];
      }
      function h(a, n) {
        p[a] = n;
      }
      var u = s.controllerVisuals[r || s.getValueState(t)], m = d.prepareVisualTypes(u);
      return g(m, function(a) {
        var n = u[a];
        e.convertOpacityToAlpha && a === "opacity" && (a = "colorAlpha", n = u.__alphaForOpacity), d.dependsOn(a, o) && n && n.applyVisual(t, f, h);
      }), p[o];
    }, i.prototype.positionGroup = function(t) {
      var o = this.visualMapModel, e = this.api, r = M(o, e).refContainer;
      V(t, o.getBoxLayoutParams(), r);
    }, i.prototype.doRender = function(t, o, e, r) {
    }, i.type = "visualMap", i;
  })(w)
);
export {
  T as default
};
