// echarts@6.0.0/lib/component/visualMap/VisualMapModel.js
import { __extends as T } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { bind as F, map as B, each as E, isArray as A, isString as L, isFunction as $, merge as V, clone as z } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import w from "../../visual/visualDefault.js";
import I from "../../visual/VisualMapping.js";
import { replaceVisualOption as W, createVisualMappings as C } from "../../visual/visualSolution.js";
import { queryReferringComponents as k } from "../../util/model.js";
import { asc as H, linearMap as K } from "../../util/number.js";
import N from "../../model/Component.js";
import y from "../../visual/tokens.js";
var O = I.mapVisual, P = I.eachVisual, q = A, D = E, G = H, U = K, it = (
  /** @class */
  (function(M) {
    T(i, M);
    function i() {
      var t = M !== null && M.apply(this, arguments) || this;
      return t.type = i.type, t.stateList = ["inRange", "outOfRange"], t.replacableOptionKeys = ["inRange", "outOfRange", "target", "controller", "color"], t.layoutMode = {
        type: "box",
        ignoreSize: !0
      }, t.dataBound = [-1 / 0, 1 / 0], t.targetVisuals = {}, t.controllerVisuals = {}, t;
    }
    return i.prototype.init = function(t, e, o) {
      this.mergeDefaultAndTheme(t, o);
    }, i.prototype.optionUpdated = function(t, e) {
      var o = this.option;
      !e && W(o, t, this.replacableOptionKeys), this.textStyleModel = this.getModel("textStyle"), this.resetItemSize(), this.completeVisualOption();
    }, i.prototype.resetVisual = function(t) {
      var e = this.stateList;
      t = F(t, this), this.controllerVisuals = C(this.option.controller, e, t), this.targetVisuals = C(this.option.target, e, t);
    }, i.prototype.getItemSymbol = function() {
      return null;
    }, i.prototype.getTargetSeriesIndices = function() {
      var t = this.option.seriesId, e = this.option.seriesIndex;
      e == null && t == null && (e = "all");
      var o = k(this.ecModel, "series", {
        index: e,
        id: t
      }, {
        useDefault: !1,
        enableAll: !0,
        enableNone: !1
      }).models;
      return B(o, function(n) {
        return n.componentIndex;
      });
    }, i.prototype.eachTargetSeries = function(t, e) {
      E(this.getTargetSeriesIndices(), function(o) {
        var n = this.ecModel.getSeriesByIndex(o);
        n && t.call(e, n);
      }, this);
    }, i.prototype.isTargetSeries = function(t) {
      var e = !1;
      return this.eachTargetSeries(function(o) {
        o === t && (e = !0);
      }), e;
    }, i.prototype.formatValueText = function(t, e, o) {
      var n = this.option, m = n.precision, a = this.dataBound, p = n.formatter, c;
      o = o || ["<", ">"], A(t) && (t = t.slice(), c = !0);
      var s = e ? t : c ? [r(t[0]), r(t[1])] : r(t);
      if (L(p))
        return p.replace("{value}", c ? s[0] : s).replace("{value2}", c ? s[1] : s);
      if ($(p))
        return c ? p(t[0], t[1]) : p(t);
      if (c)
        return t[0] === a[0] ? o[0] + " " + s[1] : t[1] === a[1] ? o[1] + " " + s[0] : s[0] + " - " + s[1];
      return s;
      function r(u) {
        return u === a[0] ? "min" : u === a[1] ? "max" : (+u).toFixed(Math.min(m, 20));
      }
    }, i.prototype.resetExtent = function() {
      var t = this.option, e = G([t.min, t.max]);
      this._dataExtent = e;
    }, i.prototype.getDataDimensionIndex = function(t) {
      var e = this.option.dimension;
      if (e != null)
        return t.getDimensionIndex(e);
      for (var o = t.dimensions, n = o.length - 1; n >= 0; n--) {
        var m = o[n], a = t.getDimensionInfo(m);
        if (!a.isCalculationCoord)
          return a.storeDimIndex;
      }
    }, i.prototype.getExtent = function() {
      return this._dataExtent.slice();
    }, i.prototype.completeVisualOption = function() {
      var t = this.ecModel, e = this.option, o = {
        inRange: e.inRange,
        outOfRange: e.outOfRange
      }, n = e.target || (e.target = {}), m = e.controller || (e.controller = {});
      V(n, o), V(m, o);
      var a = this.isCategory();
      p.call(this, n), p.call(this, m), c.call(this, n, "inRange", "outOfRange"), s.call(this, m);
      function p(r) {
        q(e.color) && !r.inRange && (r.inRange = {
          color: e.color.slice().reverse()
        }), r.inRange = r.inRange || {
          color: t.get("gradientColor")
        };
      }
      function c(r, u, v) {
        var S = r[u], f = r[v];
        S && !f && (f = r[v] = {}, D(S, function(x, h) {
          if (I.isValidType(h)) {
            var g = w.get(h, "inactive", a);
            g != null && (f[h] = g, h === "color" && !f.hasOwnProperty("opacity") && !f.hasOwnProperty("colorAlpha") && (f.opacity = [0, 0]));
          }
        }));
      }
      function s(r) {
        var u = (r.inRange || {}).symbol || (r.outOfRange || {}).symbol, v = (r.inRange || {}).symbolSize || (r.outOfRange || {}).symbolSize, S = this.get("inactiveColor"), f = this.getItemSymbol(), x = f || "roundRect";
        D(this.stateList, function(h) {
          var g = this.itemSize, l = r[h];
          l || (l = r[h] = {
            color: a ? S : [S]
          }), l.symbol == null && (l.symbol = u && z(u) || (a ? x : [x])), l.symbolSize == null && (l.symbolSize = v && z(v) || (a ? g[0] : [g[0], g[0]])), l.symbol = O(l.symbol, function(d) {
            return d === "none" ? x : d;
          });
          var b = l.symbolSize;
          if (b != null) {
            var R = -1 / 0;
            P(b, function(d) {
              d > R && (R = d);
            }), l.symbolSize = O(b, function(d) {
              return U(d, [0, R], [0, g[0]], !0);
            });
          }
        }, this);
      }
    }, i.prototype.resetItemSize = function() {
      this.itemSize = [parseFloat(this.get("itemWidth")), parseFloat(this.get("itemHeight"))];
    }, i.prototype.isCategory = function() {
      return !!this.option.categories;
    }, i.prototype.setSelected = function(t) {
    }, i.prototype.getSelected = function() {
      return null;
    }, i.prototype.getValueState = function(t) {
      return null;
    }, i.prototype.getVisualMeta = function(t) {
      return null;
    }, i.type = "visualMap", i.dependencies = ["series"], i.defaultOption = {
      show: !0,
      // zlevel: 0,
      z: 4,
      // seriesIndex: 'all',
      min: 0,
      max: 200,
      left: 0,
      right: null,
      top: null,
      bottom: 0,
      itemWidth: null,
      itemHeight: null,
      inverse: !1,
      orient: "vertical",
      backgroundColor: y.color.transparent,
      borderColor: y.color.borderTint,
      contentColor: y.color.theme[0],
      inactiveColor: y.color.disabled,
      borderWidth: 0,
      padding: y.size.m,
      // 接受数组分别设定上右下左边距，同css
      textGap: 10,
      precision: 0,
      textStyle: {
        color: y.color.secondary
        // 值域文字颜色
      }
    }, i;
  })(N)
);
export {
  it as default
};
