// echarts@6.0.0/lib/component/visualMap/visualEncoding.js
import { bind as l } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { incrementalApplyVisual as v } from "../../visual/visualSolution.js";
import D from "../../visual/VisualMapping.js";
import { getVisualFromData as h } from "../../visual/helper.js";
var A = [
  {
    createOnAllSeries: !0,
    reset: function(a, i) {
      var e = [];
      return i.eachComponent("visualMap", function(t) {
        var r = a.pipelineContext;
        !t.isTargetSeries(a) || r && r.large || e.push(v(t.stateList, t.targetVisuals, l(t.getValueState, t), t.getDataDimensionIndex(a.getData())));
      }), e;
    }
  },
  // Only support color.
  {
    createOnAllSeries: !0,
    reset: function(a, i) {
      var e = a.getData(), t = [];
      i.eachComponent("visualMap", function(r) {
        if (r.isTargetSeries(a)) {
          var s = r.getVisualMeta(l(C, null, a, r)) || {
            stops: [],
            outerColors: []
          }, n = r.getDataDimensionIndex(e);
          n >= 0 && (s.dimension = n, t.push(s));
        }
      }), a.getData().setVisual("visualMeta", t);
    }
  }
];
function C(a, i, e, t) {
  for (var r = i.targetVisuals[t], s = D.prepareVisualTypes(r), n = {
    color: h(a.getData(), "color")
    // default color.
  }, o = 0, c = s.length; o < c; o++) {
    var p = s[o], g = r[p === "opacity" ? "__alphaForOpacity" : p];
    g && g.applyVisual(e, f, m);
  }
  return n.color;
  function f(u) {
    return n[u];
  }
  function m(u, V) {
    n[u] = V;
  }
}
export {
  A as visualMapEncodingHandlers
};
