// echarts@6.0.0/lib/component/radar/RadarView.js
import { __extends as G } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { map as V, each as b, isArray as E, defaults as N } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import H from "../axis/AxisBuilder.js";
import { mergePath as P } from "../../util/graphic.js";
import J from "../../view/Component.js";
import K from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Circle.js";
import O from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Ring.js";
import Q from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Polyline.js";
import U from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Polygon.js";
var oe = (
  /** @class */
  (function(S) {
    G(a, S);
    function a() {
      var r = S !== null && S.apply(this, arguments) || this;
      return r.type = a.type, r;
    }
    return a.prototype.render = function(r, i, o) {
      var p = this.group;
      p.removeAll(), this._buildAxes(r, o), this._buildSplitLineAndArea(r);
    }, a.prototype._buildAxes = function(r, i) {
      var o = r.coordinateSystem, p = o.getIndicatorAxes(), f = V(p, function(t) {
        var m = t.model.get("showName") ? t.name : "", y = new H(t.model, i, {
          axisName: m,
          position: [o.cx, o.cy],
          rotation: t.angle,
          labelDirection: -1,
          tickDirection: -1,
          nameDirection: 1
        });
        return y;
      });
      b(f, function(t) {
        t.build(), this.group.add(t.group);
      }, this);
    }, a.prototype._buildSplitLineAndArea = function(r) {
      var i = r.coordinateSystem, o = i.getIndicatorAxes();
      if (!o.length)
        return;
      var p = r.get("shape"), f = r.getModel("splitLine"), t = r.getModel("splitArea"), m = f.getModel("lineStyle"), y = t.getModel("areaStyle"), T = f.get("show"), _ = t.get("show"), L = m.get("color"), I = y.get("color"), A = E(L) ? L : [L], x = E(I) ? I : [I], v = [], d = [];
      function w(n, c, u) {
        var g = u % c.length;
        return n[g] = n[g] || [], g;
      }
      if (p === "circle")
        for (var h = o[0].getTicksCoords(), k = i.cx, B = i.cy, e = 0; e < h.length; e++) {
          if (T) {
            var l = w(v, A, e);
            v[l].push(new K({
              shape: {
                cx: k,
                cy: B,
                r: h[e].coord
              }
            }));
          }
          if (_ && e < h.length - 1) {
            var l = w(d, x, e);
            d[l].push(new O({
              shape: {
                cx: k,
                cy: B,
                r0: h[e].coord,
                r: h[e + 1].coord
              }
            }));
          }
        }
      else
        for (var C, q = V(o, function(n, c) {
          var u = n.getTicksCoords();
          return C = C == null ? u.length - 1 : Math.min(u.length - 1, C), V(u, function(g) {
            return i.coordToPoint(g.coord, c);
          });
        }), M = [], e = 0; e <= C; e++) {
          for (var s = [], D = 0; D < o.length; D++)
            s.push(q[D][e]);
          if (s[0] ? s.push(s[0].slice()) : process.env.NODE_ENV !== "production" && console.error("Can't draw value axis " + e), T) {
            var l = w(v, A, e);
            v[l].push(new Q({
              shape: {
                points: s
              }
            }));
          }
          if (_ && M) {
            var l = w(d, x, e - 1);
            d[l].push(new U({
              shape: {
                points: s.concat(M)
              }
            }));
          }
          M = s.slice().reverse();
        }
      var z = m.getLineStyle(), F = y.getAreaStyle();
      b(d, function(n, c) {
        this.group.add(P(n, {
          style: N({
            stroke: "none",
            fill: x[c % x.length]
          }, F),
          silent: !0
        }));
      }, this), b(v, function(n, c) {
        this.group.add(P(n, {
          style: N({
            fill: "none",
            stroke: A[c % A.length]
          }, z),
          silent: !0
        }));
      }, this);
    }, a.type = "radar", a;
  })(J)
);
export {
  oe as default
};
