// usehooks-ts@3.1.1/dist/index.js
import { useState as m, useLayoutEffect as C, useEffect as l, useRef as a, useMemo as O, useCallback as y } from "react";
import z from "../../../../../../_virtual/index2.js";
var I = typeof window < "u" ? C : l;
function A(e, n, r, c) {
  const i = a(n);
  I(() => {
    i.current = n;
  }, [n]), l(() => {
    const o = window;
    if (!(o && o.addEventListener))
      return;
    const s = (t) => {
      i.current(t);
    };
    return o.addEventListener(e, s, c), () => {
      o.removeEventListener(e, s, c);
    };
  }, [e, r, c]);
}
var k = typeof window > "u";
function H(e, {
  defaultValue: n = !1,
  initializeWithValue: r = !0
} = {}) {
  const c = (t) => k ? n : window.matchMedia(t).matches, [i, o] = m(() => r ? c(e) : n);
  function s() {
    o(c(e));
  }
  return I(() => {
    const t = window.matchMedia(e);
    return s(), t.addListener ? t.addListener(s) : t.addEventListener("change", s), () => {
      t.removeListener ? t.removeListener(s) : t.removeEventListener("change", s);
    };
  }, [e]), i;
}
function B(e) {
  const n = a(e);
  n.current = e, l(
    () => () => {
      n.current();
    },
    []
  );
}
function D(e, n = 500, r) {
  const c = a();
  B(() => {
    c.current && c.current.cancel();
  });
  const i = O(() => {
    const o = z(e, n, r), s = (...t) => o(...t);
    return s.cancel = () => {
      o.cancel();
    }, s.isPending = () => !!c.current, s.flush = () => o.flush(), s;
  }, [e, n, r]);
  return l(() => {
    c.current = z(e, n, r);
  }, [e, n, r]), i;
}
function J(e, n, r) {
  const c = ((u, w) => u === w), i = e instanceof Function ? e() : e, [o, s] = m(i), t = a(i), f = D(
    s,
    n,
    r
  );
  return c(t.current, i) || (f(i), t.current = i), [o, f];
}
function Q({
  threshold: e = 0,
  root: n = null,
  rootMargin: r = "0%",
  freezeOnceVisible: c = !1,
  initialIsIntersecting: i = !1,
  onChange: o
} = {}) {
  var s;
  const [t, f] = m(null), [u, w] = m(() => ({
    isIntersecting: i,
    entry: void 0
  })), g = a();
  g.current = o;
  const h = ((s = u.entry) == null ? void 0 : s.isIntersecting) && c;
  l(() => {
    if (!t || !("IntersectionObserver" in window) || h)
      return;
    const d = new IntersectionObserver(
      (p) => {
        const x = Array.isArray(d.thresholds) ? d.thresholds : [d.thresholds];
        p.forEach((S) => {
          const R = S.isIntersecting && x.some((M) => S.intersectionRatio >= M);
          w({ isIntersecting: R, entry: S }), g.current && g.current(R, S);
        });
      },
      { threshold: e, root: n, rootMargin: r }
    );
    return d.observe(t), () => {
      d.disconnect();
    };
  }, [
    t,
    // eslint-disable-next-line react-hooks/exhaustive-deps
    JSON.stringify(e),
    n,
    r,
    h,
    c
  ]);
  const b = a(null);
  l(() => {
    var d;
    !t && ((d = u.entry) != null && d.target) && !c && !h && b.current !== u.entry.target && (b.current = u.entry.target, w({ isIntersecting: i, entry: void 0 }));
  }, [t, u.entry, c, h, i]);
  const v = [
    f,
    !!u.isIntersecting,
    u.entry
  ];
  return v.ref = v[0], v.isIntersecting = v[1], v.entry = v[2], v;
}
function F() {
  const e = a(!1);
  return l(() => (e.current = !0, () => {
    e.current = !1;
  }), []), y(() => e.current, []);
}
function U(e, n, r = "mousedown", c = {}) {
  A(
    r,
    (i) => {
      const o = i.target;
      if (!o || !o.isConnected)
        return;
      (Array.isArray(e) ? e.filter((t) => !!t.current).every((t) => t.current && !t.current.contains(o)) : e.current && !e.current.contains(o)) && n(i);
    },
    void 0,
    c
  );
}
var E = {
  width: void 0,
  height: void 0
};
function V(e) {
  const { ref: n, box: r = "content-box" } = e, [{ width: c, height: i }, o] = m(E), s = F(), t = a({ ...E }), f = a(void 0);
  return f.current = e.onResize, l(() => {
    if (!n.current || typeof window > "u" || !("ResizeObserver" in window))
      return;
    const u = new ResizeObserver(([w]) => {
      const g = r === "border-box" ? "borderBoxSize" : r === "device-pixel-content-box" ? "devicePixelContentBoxSize" : "contentBoxSize", h = L(w, g, "inlineSize"), b = L(w, g, "blockSize");
      if (t.current.width !== h || t.current.height !== b) {
        const d = { width: h, height: b };
        t.current.width = h, t.current.height = b, f.current ? f.current(d) : s() && o(d);
      }
    });
    return u.observe(n.current, { box: r }), () => {
      u.disconnect();
    };
  }, [r, n, s]), { width: c, height: i };
}
function L(e, n, r) {
  return e[n] ? Array.isArray(e[n]) ? e[n][0][r] : (
    // @ts-ignore Support Firefox's non-standard behavior
    e[n][r]
  ) : n === "contentBoxSize" ? e.contentRect[r === "inlineSize" ? "width" : "height"] : void 0;
}
export {
  D as useDebounceCallback,
  J as useDebounceValue,
  A as useEventListener,
  Q as useIntersectionObserver,
  F as useIsMounted,
  I as useIsomorphicLayoutEffect,
  H as useMediaQuery,
  U as useOnClickOutside,
  V as useResizeObserver,
  B as useUnmount
};
