// echarts@6.0.0/lib/view/Chart.js
import { each as f } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import c from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import { getUID as g } from "../util/component.js";
import { enableClassExtend as v, enableClassManagement as y } from "../util/clazz.js";
import { makeInner as M, queryDataIndex as w, normalizeToArray as E } from "../util/model.js";
import { getHighlightDigit as T, isHighDownDispatcher as D, enterEmphasis as k, leaveEmphasis as R } from "../util/states.js";
import { createTask as I } from "../core/task.js";
import N from "../chart/helper/createRenderPlanner.js";
import { traverseElements as V } from "../util/graphic.js";
import { error as s } from "../util/log.js";
var h = M(), C = N(), m = (
  /** @class */
  (function() {
    function r() {
      this.group = new c(), this.uid = g("viewChart"), this.renderTask = I({
        plan: P,
        reset: U
      }), this.renderTask.context = {
        view: this
      };
    }
    return r.prototype.init = function(e, i) {
    }, r.prototype.render = function(e, i, o, t) {
      if (process.env.NODE_ENV !== "production")
        throw new Error("render method must been implemented");
    }, r.prototype.highlight = function(e, i, o, t) {
      var n = e.getData(t && t.dataType);
      if (!n) {
        process.env.NODE_ENV !== "production" && s("Unknown dataType " + t.dataType);
        return;
      }
      u(n, t, "emphasis");
    }, r.prototype.downplay = function(e, i, o, t) {
      var n = e.getData(t && t.dataType);
      if (!n) {
        process.env.NODE_ENV !== "production" && s("Unknown dataType " + t.dataType);
        return;
      }
      u(n, t, "normal");
    }, r.prototype.remove = function(e, i) {
      this.group.removeAll();
    }, r.prototype.dispose = function(e, i) {
    }, r.prototype.updateView = function(e, i, o, t) {
      this.render(e, i, o, t);
    }, r.prototype.updateLayout = function(e, i, o, t) {
      this.render(e, i, o, t);
    }, r.prototype.updateVisual = function(e, i, o, t) {
      this.render(e, i, o, t);
    }, r.prototype.eachRendered = function(e) {
      V(this.group, e);
    }, r.markUpdateMethod = function(e, i) {
      h(e).updateMethod = i;
    }, r.protoInitialize = (function() {
      var e = r.prototype;
      e.type = "chart";
    })(), r;
  })()
);
function l(r, e, i) {
  r && D(r) && (e === "emphasis" ? k : R)(r, i);
}
function u(r, e, i) {
  var o = w(r, e), t = e && e.highlightKey != null ? T(e.highlightKey) : null;
  o != null ? f(E(o), function(n) {
    l(r.getItemGraphicEl(n), i, t);
  }) : r.eachItemGraphicEl(function(n) {
    l(n, i, t);
  });
}
v(m, ["dispose"]);
y(m);
function P(r) {
  return C(r.model);
}
function U(r) {
  var e = r.model, i = r.ecModel, o = r.api, t = r.payload, n = e.pipelineContext.progressiveRender, d = r.view, a = t && h(t).updateMethod, p = n ? "incrementalPrepareRender" : a && d[a] ? a : "render";
  return p !== "render" && d[p](e, i, o, t), b[p];
}
var b = {
  incrementalPrepareRender: {
    progress: function(r, e) {
      e.view.incrementalRender(r, e.model, e.ecModel, e.api, e.payload);
    }
  },
  render: {
    // Put view.render in `progress` to support appendData. But in this case
    // view.render should not be called in reset, otherwise it will be called
    // twise. Use `forceFirstProgress` to make sure that view.render is called
    // in any cases.
    forceFirstProgress: !0,
    progress: function(r, e) {
      e.view.render(e.model, e.ecModel, e.api, e.payload);
    }
  }
};
export {
  m as default
};
