// echarts@6.0.0/lib/chart/chord/chordLayout.js
import { normalizeArcAngles as X } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/PathProxy.js";
import { getCircleLayout as Y } from "../../util/layout.js";
var z = Math.PI / 180;
function ta(u, E) {
  u.eachSeriesByType("chord", function(y) {
    _(y, E);
  });
}
function _(u, E) {
  var y = u.getData(), g = y.graph, P = u.getEdgeData(), j = P.count();
  if (j) {
    var m = Y(u, E), c = m.cx, d = m.cy, q = m.r, r = m.r0, h = Math.max((u.get("padAngle") || 0) * z, 0), e = Math.max((u.get("minAngle") || 0) * z, 0), R = -u.get("startAngle") * z, F = R + Math.PI * 2, p = u.get("clockwise"), s = p ? 1 : -1, D = [R, F];
    X(D, !p);
    var B = D[0], H = D[1], f = H - B, I = y.getSum("value") === 0 && P.getSum("value") === 0, v = [], l = 0;
    g.eachEdge(function(a) {
      var t = I ? 1 : a.getValue("value");
      I && (t > 0 || e) && (l += 2);
      var n = a.node1.dataIndex, o = a.node2.dataIndex;
      v[n] = (v[n] || 0) + t, v[o] = (v[o] || 0) + t;
    });
    var M = 0;
    if (g.eachNode(function(a) {
      var t = a.getValue("value");
      isNaN(t) || (v[a.dataIndex] = Math.max(t, v[a.dataIndex] || 0)), !I && (v[a.dataIndex] > 0 || e) && l++, M += v[a.dataIndex] || 0;
    }), !(l === 0 || M === 0)) {
      h * l >= Math.abs(f) && (h = Math.max(0, (Math.abs(f) - e * l) / l)), (h + e) * l >= Math.abs(f) && (e = (Math.abs(f) - h * l) / l);
      var G = (f - h * l * s) / M, A = 0, V = 0, k = 0;
      g.eachNode(function(a) {
        var t = v[a.dataIndex] || 0, n = G * (M ? t : 1) * s;
        Math.abs(n) < e ? A += e - Math.abs(n) : (V += Math.abs(n) - e, k += Math.abs(n)), a.setLayout({
          angle: n,
          value: t
        });
      });
      var w = !1;
      if (A > V) {
        var T = A / V;
        g.eachNode(function(a) {
          var t = a.getLayout().angle;
          Math.abs(t) >= e ? a.setLayout({
            angle: t * T,
            ratio: T
          }, !0) : a.setLayout({
            angle: e,
            ratio: e === 0 ? 1 : t / e
          }, !0);
        });
      } else
        g.eachNode(function(a) {
          if (!w) {
            var t = a.getLayout().angle, n = Math.min(t / k, 1), o = n * A;
            t - o < e && (w = !0);
          }
        });
      var C = A;
      g.eachNode(function(a) {
        if (!(C <= 0)) {
          var t = a.getLayout().angle;
          if (t > e && e > 0) {
            var n = w ? 1 : Math.min(t / k, 1), o = t - e, i = Math.min(o, Math.min(C, A * n));
            C -= i, a.setLayout({
              angle: t - i,
              ratio: (t - i) / t
            }, !0);
          } else e > 0 && a.setLayout({
            angle: e,
            ratio: t === 0 ? 1 : e / t
          }, !0);
        }
      });
      var L = B, x = [];
      g.eachNode(function(a) {
        var t = Math.max(a.getLayout().angle, e);
        a.setLayout({
          cx: c,
          cy: d,
          r0: r,
          r: q,
          startAngle: L,
          endAngle: L + t * s,
          clockwise: p
        }, !0), x[a.dataIndex] = L, L += (t + h) * s;
      }), g.eachEdge(function(a) {
        var t = I ? 1 : a.getValue("value"), n = G * (M ? t : 1) * s, o = a.node1.dataIndex, i = x[o] || 0, J = Math.abs((a.node1.getLayout().ratio || 1) * n), S = i + J * s, K = [c + r * Math.cos(i), d + r * Math.sin(i)], O = [c + r * Math.cos(S), d + r * Math.sin(S)], Z = a.node2.dataIndex, b = x[Z] || 0, Q = Math.abs((a.node2.getLayout().ratio || 1) * n), N = b + Q * s, U = [c + r * Math.cos(b), d + r * Math.sin(b)], W = [c + r * Math.cos(N), d + r * Math.sin(N)];
        a.setLayout({
          s1: K,
          s2: O,
          sStartAngle: i,
          sEndAngle: S,
          t1: U,
          t2: W,
          tStartAngle: b,
          tEndAngle: N,
          cx: c,
          cy: d,
          r,
          value: t,
          clockwise: p
        }), x[o] = S, x[Z] = N;
      });
    }
  }
}
export {
  ta as default
};
