// echarts@6.0.0/lib/chart/chord/ChordPiece.js
import { __extends as I } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { extend as h, retrieve3 as P } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { getSectorCornerRadius as f } from "../helper/sectorHelper.js";
import { getLabelStatesModels as T, setLabelStyle as D } from "../../label/labelStyle.js";
import { setStatesStylesFromModel as y, toggleHoverEmphasis as F } from "../../util/states.js";
import { getECData as _ } from "../../util/innerStore.js";
import E from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Text.js";
import N from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Sector.js";
import { updateProps as j } from "../../animation/basicTransition.js";
var q = (
  /** @class */
  (function(v) {
    I(u, v);
    function u(o, t, n) {
      var a = v.call(this) || this;
      _(a).dataType = "node", a.z2 = 2;
      var e = new E();
      return a.setTextContent(e), a.updateData(o, t, n, !0), a;
    }
    return u.prototype.updateData = function(o, t, n, a) {
      var e = this, m = o.graph.getNodeByIndex(t), p = o.hostModel, l = m.getModel(), i = l.getModel("emphasis"), s = o.getItemLayout(t), d = h(f(l.getModel("itemStyle"), s, !0), s), r = this;
      if (isNaN(d.startAngle)) {
        r.setShape(d);
        return;
      }
      a ? r.setShape(d) : j(r, {
        shape: d
      }, p, t);
      var c = h(f(l.getModel("itemStyle"), s, !0), s);
      e.setShape(c), e.useStyle(o.getItemVisual(t, "style")), y(e, l), this._updateLabel(p, l, m), o.setItemGraphicEl(t, r), y(r, l, "itemStyle");
      var g = i.get("focus");
      F(this, g === "adjacency" ? m.getAdjacentDataIndices() : g, i.get("blurScope"), i.get("disabled"));
    }, u.prototype._updateLabel = function(o, t, n) {
      var a = this.getTextContent(), e = n.getLayout(), m = (e.startAngle + e.endAngle) / 2, p = Math.cos(m), l = Math.sin(m), i = t.getModel("label");
      a.ignore = !i.get("show");
      var s = T(t), d = n.getVisual("style");
      D(a, s, {
        labelFetcher: {
          getFormattedLabel: function(M, x, w, A, L, C) {
            return o.getFormattedLabel(
              M,
              x,
              "node",
              A,
              // ensure edgeLabel formatter is provided
              // to prevent the inheritance from `label.formatter` of the series
              P(L, s.normal && s.normal.get("formatter"), t.get("name")),
              C
            );
          }
        },
        labelDataIndex: n.dataIndex,
        defaultText: n.dataIndex + "",
        inheritColor: d.fill,
        defaultOpacity: d.opacity,
        defaultOutsidePosition: "startArc"
      });
      var r = i.get("position") || "outside", c = i.get("distance") || 0, g;
      r === "outside" ? g = e.r + c : g = (e.r + e.r0) / 2, this.textConfig = {
        inside: r !== "outside"
      };
      var b = r !== "outside" ? i.get("align") || "center" : p > 0 ? "left" : "right", S = r !== "outside" ? i.get("verticalAlign") || "middle" : l > 0 ? "top" : "bottom";
      a.attr({
        x: p * g + e.cx,
        y: l * g + e.cy,
        rotation: 0,
        style: {
          align: b,
          verticalAlign: S
        }
      });
    }, u;
  })(N)
);
export {
  q as default
};
