// echarts@6.0.0/lib/chart/chord/ChordEdge.js
import { __extends as x } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { extend as M, isString as S } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { getSectorCornerRadius as E } from "../helper/sectorHelper.js";
import { saveOldStyle as b, updateProps as A } from "../../animation/basicTransition.js";
import { toggleHoverEmphasis as V, setStatesStylesFromModel as k } from "../../util/states.js";
import { getECData as w } from "../../util/innerStore.js";
import C from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Path.js";
import N from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/LinearGradient.js";
var B = (
  /** @class */
  (function(a) {
    x(c, a);
    function c(e, t, r, d) {
      var l = a.call(this) || this;
      return w(l).dataType = "edge", l.updateData(e, t, r, d, !0), l;
    }
    return c.prototype.buildPath = function(e, t) {
      e.moveTo(t.s1[0], t.s1[1]);
      var r = 0.7, d = t.clockwise;
      e.arc(t.cx, t.cy, t.r, t.sStartAngle, t.sEndAngle, !d), e.bezierCurveTo((t.cx - t.s2[0]) * r + t.s2[0], (t.cy - t.s2[1]) * r + t.s2[1], (t.cx - t.t1[0]) * r + t.t1[0], (t.cy - t.t1[1]) * r + t.t1[1], t.t1[0], t.t1[1]), e.arc(t.cx, t.cy, t.r, t.tStartAngle, t.tEndAngle, !d), e.bezierCurveTo((t.cx - t.t2[0]) * r + t.t2[0], (t.cy - t.t2[1]) * r + t.t2[1], (t.cx - t.s1[0]) * r + t.s1[0], (t.cy - t.s1[1]) * r + t.s1[1], t.s1[0], t.s1[1]), e.closePath();
    }, c.prototype.updateData = function(e, t, r, d, l) {
      var u = e.hostModel, o = t.graph.getEdgeByIndex(r), g = o.getLayout(), i = o.node1.getModel(), s = t.getItemModel(o.dataIndex), m = s.getModel("lineStyle"), v = s.getModel("emphasis"), y = v.get("focus"), f = M(E(i.getModel("itemStyle"), g, !0), g), n = this;
      if (isNaN(f.sStartAngle) || isNaN(f.tStartAngle)) {
        n.setShape(f);
        return;
      }
      l ? (n.setShape(f), I(n, o, e, m)) : (b(n), I(n, o, e, m), A(n, {
        shape: f
      }, u, r)), V(this, y === "adjacency" ? o.getAdjacentDataIndices() : y, v.get("blurScope"), v.get("disabled")), k(n, s, "lineStyle"), t.setItemGraphicEl(o.dataIndex, n);
    }, c;
  })(C)
);
function I(a, c, e, t) {
  var r = c.node1, d = c.node2, l = a.style;
  a.setStyle(t.getLineStyle());
  var u = t.get("color");
  switch (u) {
    case "source":
      l.fill = e.getItemVisual(r.dataIndex, "style").fill, l.decal = r.getVisual("style").decal;
      break;
    case "target":
      l.fill = e.getItemVisual(d.dataIndex, "style").fill, l.decal = d.getVisual("style").decal;
      break;
    case "gradient":
      var o = e.getItemVisual(r.dataIndex, "style").fill, g = e.getItemVisual(d.dataIndex, "style").fill;
      if (S(o) && S(g)) {
        var i = a.shape, s = (i.s1[0] + i.s2[0]) / 2, m = (i.s1[1] + i.s2[1]) / 2, v = (i.t1[0] + i.t2[0]) / 2, y = (i.t1[1] + i.t2[1]) / 2;
        l.fill = new N(s, m, v, y, [{
          offset: 0,
          color: o
        }, {
          offset: 1,
          color: g
        }], !0);
      }
      break;
  }
}
export {
  B as ChordEdge
};
