// echarts@6.0.0/lib/chart/chord/ChordView.js
import { __extends as s } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import y from "../../view/Chart.js";
import m from "./ChordPiece.js";
import { ChordEdge as E } from "./ChordEdge.js";
import { parsePercent as v } from "../../util/number.js";
import { getECData as l } from "../../util/innerStore.js";
import { removeElementWithFadeOut as c, initProps as C } from "../../animation/basicTransition.js";
var _ = Math.PI / 180, L = (
  /** @class */
  (function(f) {
    s(n, f);
    function n() {
      var t = f !== null && f.apply(this, arguments) || this;
      return t.type = n.type, t;
    }
    return n.prototype.init = function(t, h) {
    }, n.prototype.render = function(t, h, g) {
      var e = t.getData(), p = this._data, d = this.group, r = -t.get("startAngle") * _;
      if (e.diff(p).add(function(a) {
        var u = e.getItemLayout(a);
        if (u) {
          var i = new m(e, a, r);
          l(i).dataIndex = a, d.add(i);
        }
      }).update(function(a, u) {
        var i = p.getItemGraphicEl(u), D = e.getItemLayout(a);
        if (!D) {
          i && c(i, t, u);
          return;
        }
        i ? i.updateData(e, a, r) : i = new m(e, a, r), d.add(i);
      }).remove(function(a) {
        var u = p.getItemGraphicEl(a);
        u && c(u, t, a);
      }).execute(), !p) {
        var o = t.get("center");
        this.group.scaleX = 0.01, this.group.scaleY = 0.01, this.group.originX = v(o[0], g.getWidth()), this.group.originY = v(o[1], g.getHeight()), C(this.group, {
          scaleX: 1,
          scaleY: 1
        }, t);
      }
      this._data = e, this.renderEdges(t, r);
    }, n.prototype.renderEdges = function(t, h) {
      var g = t.getData(), e = t.getEdgeData(), p = this._edgeData, d = this.group;
      e.diff(p).add(function(r) {
        var o = new E(g, e, r, h);
        l(o).dataIndex = r, d.add(o);
      }).update(function(r, o) {
        var a = p.getItemGraphicEl(o);
        a.updateData(g, e, r, h), d.add(a);
      }).remove(function(r) {
        var o = p.getItemGraphicEl(r);
        o && c(o, t, r);
      }).execute(), this._edgeData = e;
    }, n.prototype.dispose = function() {
    }, n.type = "chord", n;
  })(y)
);
export {
  L as default
};
