// echarts@6.0.0/lib/chart/chord/ChordSeries.js
import { __extends as y } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import D from "../../model/Model.js";
import M from "../../model/Series.js";
import b from "../helper/createGraphFromNodeEdge.js";
import { createTooltipMarkup as h } from "../../component/tooltip/tooltipMarkup.js";
import S from "../../visual/LegendVisualProvider.js";
import { bind as v } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
var B = (
  /** @class */
  (function(i) {
    y(a, i);
    function a() {
      var e = i !== null && i.apply(this, arguments) || this;
      return e.type = a.type, e;
    }
    return a.prototype.init = function(e) {
      i.prototype.init.apply(this, arguments), this.fillDataTextStyle(e.edges || e.links), this.legendVisualProvider = new S(v(this.getData, this), v(this.getRawData, this));
    }, a.prototype.mergeOption = function(e) {
      i.prototype.mergeOption.apply(this, arguments), this.fillDataTextStyle(e.edges || e.links);
    }, a.prototype.getInitialData = function(e, d) {
      var o = e.edges || e.links || [], l = e.data || e.nodes || [];
      if (l && o) {
        var r = b(l, o, this, !0, n);
        return r.data;
      }
      function n(p, g) {
        var u = D.prototype.getModel;
        function c(t, s) {
          var f = u.call(this, t, s);
          return f.resolveParentPath = m, f;
        }
        g.wrapMethod("getItemModel", function(t) {
          return t.resolveParentPath = m, t.getModel = c, t;
        });
        function m(t) {
          if (t && (t[0] === "label" || t[1] === "label")) {
            var s = t.slice();
            return t[0] === "label" ? s[0] = "edgeLabel" : t[1] === "label" && (s[1] = "edgeLabel"), s;
          }
          return t;
        }
      }
    }, a.prototype.getGraph = function() {
      return this.getData().graph;
    }, a.prototype.getEdgeData = function() {
      return this.getGraph().edgeData;
    }, a.prototype.formatTooltip = function(e, d, o) {
      var l = this.getDataParams(e, o);
      if (o === "edge") {
        var r = this.getData(), n = r.graph.getEdgeByIndex(e), p = r.getName(n.node1.dataIndex), g = r.getName(n.node2.dataIndex), u = [];
        return p != null && u.push(p), g != null && u.push(g), h("nameValue", {
          name: u.join(" > "),
          value: l.value,
          noValue: l.value == null
        });
      }
      return h("nameValue", {
        name: l.name,
        value: l.value,
        noValue: l.value == null
      });
    }, a.prototype.getDataParams = function(e, d) {
      var o = i.prototype.getDataParams.call(this, e, d);
      if (d === "node") {
        var l = this.getData(), r = this.getGraph().getNodeByIndex(e);
        if (o.name == null && (o.name = l.getName(e)), o.value == null) {
          var n = r.getLayout().value;
          o.value = n;
        }
      }
      return o;
    }, a.type = "series.chord", a.defaultOption = {
      // zlevel: 0,
      z: 2,
      coordinateSystem: "none",
      legendHoverLink: !0,
      colorBy: "data",
      left: 0,
      top: 0,
      right: 0,
      bottom: 0,
      width: null,
      height: null,
      center: ["50%", "50%"],
      radius: ["70%", "80%"],
      clockwise: !0,
      startAngle: 90,
      endAngle: "auto",
      minAngle: 0,
      padAngle: 3,
      itemStyle: {
        borderRadius: [0, 0, 5, 5]
      },
      lineStyle: {
        width: 0,
        color: "source",
        opacity: 0.2
      },
      label: {
        show: !0,
        position: "outside",
        distance: 5
      },
      emphasis: {
        focus: "adjacency",
        lineStyle: {
          opacity: 0.5
        }
      }
    }, a;
  })(M)
);
export {
  B as default
};
