// lib0@0.2.109/decoding.js
import { BIT8 as a, BITS7 as p, BIT7 as D, BITS6 as U } from "./binary.js";
import { floor as b, isNegativeZero as x } from "./math.js";
import { MAX_SAFE_INTEGER as g } from "./number.js";
import { utf8TextDecoder as m } from "./string.js";
import { create as w } from "./error.js";
const y = w("Unexpected end of array"), d = w("Integer out of Range");
class i {
  /**
   * @param {Uint8Array} uint8Array Binary data to decode
   */
  constructor(s) {
    this.arr = s, this.pos = 0;
  }
}
const k = (t) => new i(t), A = (t) => t.pos !== t.arr.length, V = (t, s) => {
  const r = new Uint8Array(t.arr.buffer, t.pos + t.arr.byteOffset, s);
  return t.pos += s, r;
}, I = (t) => V(t, o(t)), h = (t) => t.arr[t.pos++], o = (t) => {
  let s = 0, r = 1;
  const n = t.arr.length;
  for (; t.pos < n; ) {
    const e = t.arr[t.pos++];
    if (s = s + (e & p) * r, r *= 128, e < a)
      return s;
    if (s > g)
      throw d;
  }
  throw y;
}, u = (t) => {
  let s = t.arr[t.pos++], r = s & U, n = 64;
  const e = (s & D) > 0 ? -1 : 1;
  if ((s & a) === 0)
    return e * r;
  const S = t.arr.length;
  for (; t.pos < S; ) {
    if (s = t.arr[t.pos++], r = r + (s & p) * n, n *= 128, s < a)
      return e * r;
    if (r > g)
      throw d;
  }
  throw y;
}, O = (t) => {
  let s = o(t);
  if (s === 0)
    return "";
  {
    let r = String.fromCodePoint(h(t));
    if (--s < 100)
      for (; s--; )
        r += String.fromCodePoint(h(t));
    else
      for (; s > 0; ) {
        const n = s < 1e4 ? s : 1e4, e = t.arr.subarray(t.pos, t.pos + n);
        t.pos += n, r += String.fromCodePoint.apply(
          null,
          /** @type {any} */
          e
        ), s -= n;
      }
    return decodeURIComponent(escape(r));
  }
}, R = (t) => (
  /** @type any */
  m.decode(I(t))
), c = m ? R : O, f = (t, s) => {
  const r = new DataView(t.arr.buffer, t.arr.byteOffset + t.pos, s);
  return t.pos += s, r;
}, T = (t) => f(t, 4).getFloat32(0, !1), B = (t) => f(t, 8).getFloat64(0, !1), C = (t) => (
  /** @type {any} */
  f(t, 8).getBigInt64(0, !1)
), F = [
  (t) => {
  },
  // CASE 127: undefined
  (t) => null,
  // CASE 126: null
  u,
  // CASE 125: integer
  T,
  // CASE 124: float32
  B,
  // CASE 123: float64
  C,
  // CASE 122: bigint
  (t) => !1,
  // CASE 121: boolean (false)
  (t) => !0,
  // CASE 120: boolean (true)
  c,
  // CASE 119: string
  (t) => {
    const s = o(t), r = {};
    for (let n = 0; n < s; n++) {
      const e = c(t);
      r[e] = l(t);
    }
    return r;
  },
  (t) => {
    const s = o(t), r = [];
    for (let n = 0; n < s; n++)
      r.push(l(t));
    return r;
  },
  I
  // CASE 116: Uint8Array
], l = (t) => F[127 - h(t)](t);
class j extends i {
  /**
   * @param {Uint8Array} uint8Array
   * @param {function(Decoder):T} reader
   */
  constructor(s, r) {
    super(s), this.reader = r, this.s = null, this.count = 0;
  }
  read() {
    return this.count === 0 && (this.s = this.reader(this), A(this) ? this.count = o(this) + 1 : this.count = -1), this.count--, /** @type {T} */
    this.s;
  }
}
class v extends i {
  /**
   * @param {Uint8Array} uint8Array
   */
  constructor(s) {
    super(s), this.s = 0, this.count = 0;
  }
  read() {
    if (this.count === 0) {
      this.s = u(this);
      const s = x(this.s);
      this.count = 1, s && (this.s = -this.s, this.count = o(this) + 2);
    }
    return this.count--, /** @type {number} */
    this.s;
  }
}
class G extends i {
  /**
   * @param {Uint8Array} uint8Array
   */
  constructor(s) {
    super(s), this.s = 0, this.count = 0, this.diff = 0;
  }
  /**
   * @return {number}
   */
  read() {
    if (this.count === 0) {
      const s = u(this), r = s & 1;
      this.diff = b(s / 2), this.count = 1, r && (this.count = o(this) + 2);
    }
    return this.s += this.diff, this.count--, this.s;
  }
}
class M {
  /**
   * @param {Uint8Array} uint8Array
   */
  constructor(s) {
    this.decoder = new v(s), this.str = c(this.decoder), this.spos = 0;
  }
  /**
   * @return {string}
   */
  read() {
    const s = this.spos + this.decoder.read(), r = this.str.slice(this.spos, s);
    return this.spos = s, r;
  }
}
export {
  i as Decoder,
  G as IntDiffOptRleDecoder,
  j as RleDecoder,
  M as StringDecoder,
  v as UintOptRleDecoder,
  R as _readVarStringNative,
  O as _readVarStringPolyfill,
  k as createDecoder,
  A as hasContent,
  l as readAny,
  C as readBigInt64,
  T as readFloat32,
  B as readFloat64,
  f as readFromDataView,
  h as readUint8,
  V as readUint8Array,
  u as readVarInt,
  c as readVarString,
  o as readVarUint,
  I as readVarUint8Array
};
