// lib0@0.2.109/encoding.js
import { floor as u, abs as F, min as L, max as S, isNegativeZero as T } from "./math.js";
import { isInteger as D } from "./number.js";
import { BITS7 as h, BITS31 as O, BITS6 as U, BIT8 as w, BIT7 as R } from "./binary.js";
import { utf8TextEncoder as b, encodeUtf8 as x } from "./string.js";
import { isArray as _ } from "./array.js";
class e {
  constructor() {
    this.cpos = 0, this.cbuf = new Uint8Array(100), this.bufs = [];
  }
}
const j = () => new e(), $ = (t) => {
  const s = j();
  return t(s), c(s);
}, C = (t) => {
  let s = t.cpos;
  for (let i = 0; i < t.bufs.length; i++)
    s += t.bufs[i].length;
  return s;
}, c = (t) => {
  const s = new Uint8Array(C(t));
  let i = 0;
  for (let f = 0; f < t.bufs.length; f++) {
    const o = t.bufs[f];
    s.set(o, i), i += o.length;
  }
  return s.set(new Uint8Array(t.cbuf.buffer, 0, t.cpos), i), s;
}, N = (t, s) => {
  const i = t.cbuf.length;
  i - t.cpos < s && (t.bufs.push(new Uint8Array(t.cbuf.buffer, 0, t.cpos)), t.cbuf = new Uint8Array(S(i, s) * 2), t.cpos = 0);
}, n = (t, s) => {
  const i = t.cbuf.length;
  t.cpos === i && (t.bufs.push(t.cbuf), t.cbuf = new Uint8Array(i * 2), t.cpos = 0), t.cbuf[t.cpos++] = s;
}, m = n, r = (t, s) => {
  for (; s > h; )
    n(t, w | h & s), s = u(s / 128);
  n(t, h & s);
}, g = (t, s) => {
  const i = T(s);
  for (i && (s = -s), n(t, (s > U ? w : 0) | (i ? R : 0) | U & s), s = u(s / 64); s > 0; )
    n(t, (s > h ? w : 0) | h & s), s = u(s / 128);
}, a = new Uint8Array(3e4), P = a.length / 3, z = (t, s) => {
  if (s.length < P) {
    const i = b.encodeInto(s, a).written || 0;
    r(t, i);
    for (let f = 0; f < i; f++)
      n(t, a[f]);
  } else
    k(t, x(s));
}, Z = (t, s) => {
  const i = unescape(encodeURIComponent(s)), f = i.length;
  r(t, f);
  for (let o = 0; o < f; o++)
    n(
      t,
      /** @type {number} */
      i.codePointAt(o)
    );
}, p = b && /** @type {any} */
b.encodeInto ? z : Z, V = (t, s) => {
  const i = t.cbuf.length, f = t.cpos, o = L(i - f, s.length), l = s.length - o;
  t.cbuf.set(s.subarray(0, o), f), t.cpos += o, l > 0 && (t.bufs.push(t.cbuf), t.cbuf = new Uint8Array(S(i * 2, l)), t.cbuf.set(s.subarray(o)), t.cpos = l);
}, k = (t, s) => {
  r(t, s.byteLength), V(t, s);
}, y = (t, s) => {
  N(t, s);
  const i = new DataView(t.cbuf.buffer, t.cpos, s);
  return t.cpos += s, i;
}, q = (t, s) => y(t, 4).setFloat32(0, s, !1), G = (t, s) => y(t, 8).setFloat64(0, s, !1), H = (t, s) => (
  /** @type {any} */
  y(t, 8).setBigInt64(0, s, !1)
), A = new DataView(new ArrayBuffer(4)), J = (t) => (A.setFloat32(0, t), A.getFloat32(0) === t), I = (t, s) => {
  switch (typeof s) {
    case "string":
      n(t, 119), p(t, s);
      break;
    case "number":
      D(s) && F(s) <= O ? (n(t, 125), g(t, s)) : J(s) ? (n(t, 124), q(t, s)) : (n(t, 123), G(t, s));
      break;
    case "bigint":
      n(t, 122), H(t, s);
      break;
    case "object":
      if (s === null)
        n(t, 126);
      else if (_(s)) {
        n(t, 117), r(t, s.length);
        for (let i = 0; i < s.length; i++)
          I(t, s[i]);
      } else if (s instanceof Uint8Array)
        n(t, 116), k(t, s);
      else {
        n(t, 118);
        const i = Object.keys(s);
        r(t, i.length);
        for (let f = 0; f < i.length; f++) {
          const o = i[f];
          p(t, o), I(t, s[o]);
        }
      }
      break;
    case "boolean":
      n(t, s ? 120 : 121);
      break;
    default:
      n(t, 127);
  }
};
class v extends e {
  /**
   * @param {function(Encoder, T):void} writer
   */
  constructor(s) {
    super(), this.w = s, this.s = null, this.count = 0;
  }
  /**
   * @param {T} v
   */
  write(s) {
    this.s === s ? this.count++ : (this.count > 0 && r(this, this.count - 1), this.count = 1, this.w(this, s), this.s = s);
  }
}
const E = (t) => {
  t.count > 0 && (g(t.encoder, t.count === 1 ? t.s : -t.s), t.count > 1 && r(t.encoder, t.count - 2));
};
class K {
  constructor() {
    this.encoder = new e(), this.s = 0, this.count = 0;
  }
  /**
   * @param {number} v
   */
  write(s) {
    this.s === s ? this.count++ : (E(this), this.count = 1, this.s = s);
  }
  /**
   * Flush the encoded state and transform this to a Uint8Array.
   *
   * Note that this should only be called once.
   */
  toUint8Array() {
    return E(this), c(this.encoder);
  }
}
const B = (t) => {
  if (t.count > 0) {
    const s = t.diff * 2 + (t.count === 1 ? 0 : 1);
    g(t.encoder, s), t.count > 1 && r(t.encoder, t.count - 2);
  }
};
class d {
  constructor() {
    this.encoder = new e(), this.s = 0, this.count = 0, this.diff = 0;
  }
  /**
   * @param {number} v
   */
  write(s) {
    this.diff === s - this.s ? (this.s = s, this.count++) : (B(this), this.count = 1, this.diff = s - this.s, this.s = s);
  }
  /**
   * Flush the encoded state and transform this to a Uint8Array.
   *
   * Note that this should only be called once.
   */
  toUint8Array() {
    return B(this), c(this.encoder);
  }
}
class tt {
  constructor() {
    this.sarr = [], this.s = "", this.lensE = new K();
  }
  /**
   * @param {string} string
   */
  write(s) {
    this.s += s, this.s.length > 19 && (this.sarr.push(this.s), this.s = ""), this.lensE.write(s.length);
  }
  toUint8Array() {
    const s = new e();
    return this.sarr.push(this.s), this.s = "", p(s, this.sarr.join("")), V(s, this.lensE.toUint8Array()), c(s);
  }
}
export {
  e as Encoder,
  d as IntDiffOptRleEncoder,
  v as RleEncoder,
  tt as StringEncoder,
  K as UintOptRleEncoder,
  z as _writeVarStringNative,
  Z as _writeVarStringPolyfill,
  j as createEncoder,
  $ as encode,
  C as length,
  c as toUint8Array,
  N as verifyLen,
  n as write,
  I as writeAny,
  H as writeBigInt64,
  q as writeFloat32,
  G as writeFloat64,
  y as writeOnDataView,
  m as writeUint8,
  V as writeUint8Array,
  g as writeVarInt,
  p as writeVarString,
  r as writeVarUint,
  k as writeVarUint8Array
};
