// lib0@0.2.109/logging.js
import { supportsColor as m } from "./environment.js";
import { create as h } from "./set.js";
import { create as r } from "./pair.js";
import { mapToStyleString as a } from "./dom.js";
import { create as E } from "./map.js";
import { computeNoColorLoggingArgs as b, ORANGE as p, UNCOLOR as y, PURPLE as L, RED as d, GREY as w, GREEN as N, BLUE as R, UNBOLD as S, BOLD as A } from "./logging.common.js";
const B = {
  [A]: r("font-weight", "bold"),
  [S]: r("font-weight", "normal"),
  [R]: r("color", "blue"),
  [N]: r("color", "green"),
  [w]: r("color", "grey"),
  [d]: r("color", "red"),
  [L]: r("color", "purple"),
  [p]: r("color", "orange"),
  // not well supported in chrome when debugging node with inspector - TODO: deprecate
  [y]: r("color", "black")
}, O = (o) => {
  o.length === 1 && o[0]?.constructor === Function && (o = /** @type {Array<string|Symbol|Object|number>} */
  /** @type {[function]} */
  o[0]());
  const n = [], s = [], i = E();
  let c = [], e = 0;
  for (; e < o.length; e++) {
    const t = o[e], l = B[t];
    if (l !== void 0)
      i.set(l.left, l.right);
    else {
      if (t === void 0)
        break;
      if (t.constructor === String || t.constructor === Number) {
        const f = a(i);
        e > 0 || f.length > 0 ? (n.push("%c" + t), s.push(f)) : n.push(t);
      } else
        break;
    }
  }
  for (e > 0 && (c = s, c.unshift(n.join(""))); e < o.length; e++) {
    const t = o[e];
    t instanceof Symbol || c.push(t);
  }
  return c;
}, u = m ? O : b, $ = (...o) => {
  console.log(...u(o)), g.forEach((n) => n.print(o));
}, j = (...o) => {
  console.warn(...u(o)), o.unshift(p), g.forEach((n) => n.print(o));
}, g = h();
export {
  R as BLUE,
  A as BOLD,
  N as GREEN,
  w as GREY,
  p as ORANGE,
  L as PURPLE,
  d as RED,
  S as UNBOLD,
  y as UNCOLOR,
  $ as print,
  g as vconsoles,
  j as warn
};
