// lib0@0.2.109/observable.js
import { create as t, setIfUndefined as i } from "./map.js";
import { create as f } from "./set.js";
import { from as c } from "./array.js";
class v {
  constructor() {
    this._observers = t();
  }
  /**
   * @template {keyof EVENTS & string} NAME
   * @param {NAME} name
   * @param {EVENTS[NAME]} f
   */
  on(e, r) {
    return i(
      this._observers,
      /** @type {string} */
      e,
      f
    ).add(r), r;
  }
  /**
   * @template {keyof EVENTS & string} NAME
   * @param {NAME} name
   * @param {EVENTS[NAME]} f
   */
  once(e, r) {
    const s = (...o) => {
      this.off(
        e,
        /** @type {any} */
        s
      ), r(...o);
    };
    this.on(
      e,
      /** @type {any} */
      s
    );
  }
  /**
   * @template {keyof EVENTS & string} NAME
   * @param {NAME} name
   * @param {EVENTS[NAME]} f
   */
  off(e, r) {
    const s = this._observers.get(e);
    s !== void 0 && (s.delete(r), s.size === 0 && this._observers.delete(e));
  }
  /**
   * Emit a named event. All registered event listeners that listen to the
   * specified name will receive the event.
   *
   * @todo This should catch exceptions
   *
   * @template {keyof EVENTS & string} NAME
   * @param {NAME} name The event name.
   * @param {Parameters<EVENTS[NAME]>} args The arguments that are applied to the event listener.
   */
  emit(e, r) {
    return c((this._observers.get(e) || t()).values()).forEach((s) => s(...r));
  }
  destroy() {
    this._observers = t();
  }
}
export {
  v as ObservableV2
};
