// echarts@6.0.0/lib/chart/sankey/install.js
import m from "./SankeyView.js";
import u from "./SankeySeries.js";
import c from "./sankeyLayout.js";
import y from "./sankeyVisual.js";
import { updateCenterAndZoomInAction as p } from "../../component/helper/roamHelper.js";
function C(n) {
  n.registerChartView(m), n.registerSeriesModel(u), n.registerLayout(c), n.registerVisual(y), n.registerAction({
    type: "dragNode",
    event: "dragnode",
    // here can only use 'update' now, other value is not support in echarts.
    update: "update"
  }, function(e, o) {
    o.eachComponent({
      mainType: "series",
      subType: "sankey",
      query: e
    }, function(i) {
      i.setNodePosition(e.dataIndex, [e.localX, e.localY]);
    });
  }), n.registerAction({
    type: "sankeyRoam",
    event: "sankeyRoam",
    update: "none"
  }, function(e, o, i) {
    o.eachComponent({
      mainType: "series",
      subType: "sankey",
      query: e
    }, function(t) {
      var r = t.coordinateSystem, a = p(r, e, t.get("scaleLimit"));
      t.setCenter(a.center), t.setZoom(a.zoom);
    });
  });
}
export {
  C as install
};
