// echarts@6.0.0/lib/chart/sankey/SankeyView.js
import { __extends as O } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { setStatesStylesFromModel as P, toggleHoverEmphasis as B, enterEmphasis as et, leaveEmphasis as at } from "../../util/states.js";
import ot from "../../view/Chart.js";
import { getLabelStatesModels as N, setLabelStyle as Z } from "../../label/labelStyle.js";
import { getECData as q } from "../../util/innerStore.js";
import { isString as J, retrieve3 as rt } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import lt from "../../component/helper/RoamController.js";
import { updateController as it } from "../../component/helper/roamHelper.js";
import nt from "../../coord/View.js";
import ct from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import Q from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Rect.js";
import dt from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Path.js";
import yt from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/LinearGradient.js";
import { initProps as ut } from "../../animation/basicTransition.js";
var st = (
  /** @class */
  /* @__PURE__ */ (function() {
    function o() {
      this.x1 = 0, this.y1 = 0, this.x2 = 0, this.y2 = 0, this.cpx1 = 0, this.cpy1 = 0, this.cpx2 = 0, this.cpy2 = 0, this.extent = 0;
    }
    return o;
  })()
), gt = (
  /** @class */
  (function(o) {
    O(r, o);
    function r(t) {
      return o.call(this, t) || this;
    }
    return r.prototype.getDefaultShape = function() {
      return new st();
    }, r.prototype.buildPath = function(t, e) {
      var i = e.extent;
      t.moveTo(e.x1, e.y1), t.bezierCurveTo(e.cpx1, e.cpy1, e.cpx2, e.cpy2, e.x2, e.y2), e.orient === "vertical" ? (t.lineTo(e.x2 + i, e.y2), t.bezierCurveTo(e.cpx2 + i, e.cpy2, e.cpx1 + i, e.cpy1, e.x1 + i, e.y1)) : (t.lineTo(e.x2, e.y2 + i), t.bezierCurveTo(e.cpx2, e.cpy2 + i, e.cpx1, e.cpy1 + i, e.x1, e.y1 + i)), t.closePath();
    }, r.prototype.highlight = function() {
      et(this);
    }, r.prototype.downplay = function() {
      at(this);
    }, r;
  })(dt)
), jt = (
  /** @class */
  (function(o) {
    O(r, o);
    function r() {
      var t = o !== null && o.apply(this, arguments) || this;
      return t.type = r.type, t._mainGroup = new ct(), t._focusAdjacencyDisabled = !1, t;
    }
    return r.prototype.init = function(t, e) {
      this._controller = new lt(e.getZr()), this._controllerHost = {
        target: this.group
      }, this.group.add(this._mainGroup);
    }, r.prototype.render = function(t, e, i) {
      var L = this, _ = t.getGraph(), n = this._mainGroup, D = t.layoutInfo, p = D.width, f = D.height, T = t.getData(), U = t.getData("edge"), V = t.get("orient");
      this._model = t, n.removeAll(), n.x = D.x, n.y = D.y, this._updateViewCoordSys(t, i), it(t, i, n, this._controller, this._controllerHost, null), _.eachEdge(function(a) {
        var l = new gt(), c = q(l);
        c.dataIndex = a.dataIndex, c.seriesIndex = t.seriesIndex, c.dataType = "edge";
        var u = a.getModel(), h = u.getModel("lineStyle"), d = h.get("curveness"), s = a.node1.getLayout(), y = a.node1.getModel(), g = y.get("localX"), m = y.get("localY"), v = a.node2.getLayout(), M = a.node2.getModel(), C = M.get("localX"), j = M.get("localY"), I = a.getLayout(), x, S, b, w, G, k, A, X;
        l.shape.extent = Math.max(1, I.dy), l.shape.orient = V, V === "vertical" ? (x = (g != null ? g * p : s.x) + I.sy, S = (m != null ? m * f : s.y) + s.dy, b = (C != null ? C * p : v.x) + I.ty, w = j != null ? j * f : v.y, G = x, k = S * (1 - d) + w * d, A = b, X = S * d + w * (1 - d)) : (x = (g != null ? g * p : s.x) + s.dx, S = (m != null ? m * f : s.y) + I.sy, b = C != null ? C * p : v.x, w = (j != null ? j * f : v.y) + I.ty, G = x * (1 - d) + b * d, k = S, A = x * d + b * (1 - d), X = w), l.setShape({
          x1: x,
          y1: S,
          x2: b,
          y2: w,
          cpx1: G,
          cpy1: k,
          cpx2: A,
          cpy2: X
        }), l.useStyle(h.getItemStyle()), K(l.style, V, a);
        var H = "" + u.get("value"), Y = N(u, "edgeLabel");
        Z(l, Y, {
          labelFetcher: {
            getFormattedLabel: function(R, E, pt, W, $, tt) {
              return t.getFormattedLabel(
                R,
                E,
                "edge",
                W,
                // ensure edgeLabel formatter is provided
                // to prevent the inheritance from `label.formatter` of the series
                rt($, Y.normal && Y.normal.get("formatter"), H),
                tt
              );
            }
          },
          labelDataIndex: a.dataIndex,
          defaultText: H
        }), l.setTextConfig({
          position: "inside"
        });
        var z = u.getModel("emphasis");
        P(l, u, "lineStyle", function(R) {
          var E = R.getItemStyle();
          return K(E, V, a), E;
        }), n.add(l), U.setItemGraphicEl(a.dataIndex, l);
        var F = z.get("focus");
        B(l, F === "adjacency" ? a.getAdjacentDataIndices() : F === "trajectory" ? a.getTrajectoryDataIndices() : F, z.get("blurScope"), z.get("disabled"));
      }), _.eachNode(function(a) {
        var l = a.getLayout(), c = a.getModel(), u = c.get("localX"), h = c.get("localY"), d = c.getModel("emphasis"), s = c.get(["itemStyle", "borderRadius"]) || 0, y = new Q({
          shape: {
            x: u != null ? u * p : l.x,
            y: h != null ? h * f : l.y,
            width: l.dx,
            height: l.dy,
            r: s
          },
          style: c.getModel("itemStyle").getItemStyle(),
          z2: 10
        });
        Z(y, N(c), {
          labelFetcher: {
            getFormattedLabel: function(m, v) {
              return t.getFormattedLabel(m, v, "node");
            }
          },
          labelDataIndex: a.dataIndex,
          defaultText: a.id
        }), y.disableLabelAnimation = !0, y.setStyle("fill", a.getVisual("color")), y.setStyle("decal", a.getVisual("style").decal), P(y, c), n.add(y), T.setItemGraphicEl(a.dataIndex, y), q(y).dataType = "node";
        var g = d.get("focus");
        B(y, g === "adjacency" ? a.getAdjacentDataIndices() : g === "trajectory" ? a.getTrajectoryDataIndices() : g, d.get("blurScope"), d.get("disabled"));
      }), T.eachItemGraphicEl(function(a, l) {
        var c = T.getItemModel(l);
        c.get("draggable") && (a.drift = function(u, h) {
          L._focusAdjacencyDisabled = !0, this.shape.x += u, this.shape.y += h, this.dirty(), i.dispatchAction({
            type: "dragNode",
            seriesId: t.id,
            dataIndex: T.getRawIndex(l),
            localX: this.shape.x / p,
            localY: this.shape.y / f
          });
        }, a.ondragend = function() {
          L._focusAdjacencyDisabled = !1;
        }, a.draggable = !0, a.cursor = "move");
      }), !this._data && t.isAnimationEnabled() && n.setClipPath(ht(n.getBoundingRect(), t, function() {
        n.removeClipPath();
      })), this._data = t.getData();
    }, r.prototype.dispose = function() {
      this._controller && this._controller.dispose(), this._controllerHost = null;
    }, r.prototype._updateViewCoordSys = function(t, e) {
      var i = t.layoutInfo, L = i.width, _ = i.height, n = t.coordinateSystem = new nt(null, {
        api: e,
        ecModel: t.ecModel
      });
      n.zoomLimit = t.get("scaleLimit"), n.setBoundingRect(0, 0, L, _), n.setCenter(t.get("center")), n.setZoom(t.get("zoom")), this._controllerHost.target.attr({
        x: n.x,
        y: n.y,
        scaleX: n.scaleX,
        scaleY: n.scaleY
      });
    }, r.type = "sankey", r;
  })(ot)
);
function K(o, r, t) {
  switch (o.fill) {
    case "source":
      o.fill = t.node1.getVisual("color"), o.decal = t.node1.getVisual("style").decal;
      break;
    case "target":
      o.fill = t.node2.getVisual("color"), o.decal = t.node2.getVisual("style").decal;
      break;
    case "gradient":
      var e = t.node1.getVisual("color"), i = t.node2.getVisual("color");
      J(e) && J(i) && (o.fill = new yt(0, 0, +(r === "horizontal"), +(r === "vertical"), [{
        color: e,
        offset: 0
      }, {
        color: i,
        offset: 1
      }]));
  }
}
function ht(o, r, t) {
  var e = new Q({
    shape: {
      x: o.x - 10,
      y: o.y - 10,
      width: 0,
      height: o.height + 20
    }
  });
  return ut(e, {
    shape: {
      width: o.width + 20
    }
  }, r, t), e;
}
export {
  jt as default
};
