// echarts@6.0.0/lib/chart/sankey/SankeySeries.js
import { __extends as D } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import k from "../../model/Series.js";
import N from "../helper/createGraphFromNodeEdge.js";
import w from "../../model/Model.js";
import { createTooltipMarkup as b } from "../../component/tooltip/tooltipMarkup.js";
import S from "../../visual/tokens.js";
var x = (
  /** @class */
  (function(u) {
    D(t, u);
    function t() {
      var e = u !== null && u.apply(this, arguments) || this;
      return e.type = t.type, e;
    }
    return t.prototype.getInitialData = function(e, i) {
      var o = e.edges || e.links || [], n = e.data || e.nodes || [], r = e.levels || [];
      this.levelModels = [];
      for (var p = this.levelModels, l = 0; l < r.length; l++)
        if (r[l].depth != null && r[l].depth >= 0)
          p[r[l].depth] = new w(r[l], this, i);
        else if (process.env.NODE_ENV !== "production")
          throw new Error("levels[i].depth is mandatory and should be natural number");
      var m = N(n, o, this, !0, y);
      return m.data;
      function y(d, h) {
        d.wrapMethod("getItemModel", function(a, c) {
          var s = a.parentModel, g = s.getData().getItemLayout(c);
          if (g) {
            var v = g.depth, f = s.levelModels[v];
            f && (a.parentModel = f);
          }
          return a;
        }), h.wrapMethod("getItemModel", function(a, c) {
          var s = a.parentModel, g = s.getGraph().getEdgeByIndex(c), v = g.node1.getLayout();
          if (v) {
            var f = v.depth, M = s.levelModels[f];
            M && (a.parentModel = M);
          }
          return a;
        });
      }
    }, t.prototype.setNodePosition = function(e, i) {
      var o = this.option.data || this.option.nodes, n = o[e];
      n.localX = i[0], n.localY = i[1];
    }, t.prototype.setCenter = function(e) {
      this.option.center = e;
    }, t.prototype.setZoom = function(e) {
      this.option.zoom = e;
    }, t.prototype.getGraph = function() {
      return this.getData().graph;
    }, t.prototype.getEdgeData = function() {
      return this.getGraph().edgeData;
    }, t.prototype.formatTooltip = function(e, i, o) {
      function n(a) {
        return isNaN(a) || a == null;
      }
      if (o === "edge") {
        var r = this.getDataParams(e, o), p = r.data, l = r.value, m = p.source + " -- " + p.target;
        return b("nameValue", {
          name: m,
          value: l,
          noValue: n(l)
        });
      } else {
        var y = this.getGraph().getNodeByIndex(e), d = y.getLayout().value, h = this.getDataParams(e, o).data.name;
        return b("nameValue", {
          name: h != null ? h + "" : null,
          value: d,
          noValue: n(d)
        });
      }
    }, t.prototype.optionUpdated = function() {
    }, t.prototype.getDataParams = function(e, i) {
      var o = u.prototype.getDataParams.call(this, e, i);
      if (o.value == null && i === "node") {
        var n = this.getGraph().getNodeByIndex(e), r = n.getLayout().value;
        o.value = r;
      }
      return o;
    }, t.type = "series.sankey", t.layoutMode = "box", t.defaultOption = {
      // zlevel: 0,
      z: 2,
      // `coordinateSystem` can be declared as 'matrix', 'calendar',
      //  which provides box layout container.
      coordinateSystemUsage: "box",
      left: "5%",
      top: "5%",
      right: "20%",
      bottom: "5%",
      orient: "horizontal",
      nodeWidth: 20,
      nodeGap: 8,
      draggable: !0,
      layoutIterations: 32,
      // true | false | 'move' | 'scale', see module:component/helper/RoamController.
      roam: !1,
      roamTrigger: "global",
      center: null,
      zoom: 1,
      label: {
        show: !0,
        position: "right",
        fontSize: 12
      },
      edgeLabel: {
        show: !1,
        fontSize: 12
      },
      levels: [],
      nodeAlign: "justify",
      lineStyle: {
        color: S.color.neutral50,
        opacity: 0.2,
        curveness: 0.5
      },
      emphasis: {
        label: {
          show: !0
        },
        lineStyle: {
          opacity: 0.5
        }
      },
      select: {
        itemStyle: {
          borderColor: S.color.primary
        }
      },
      animationEasing: "linear",
      animationDuration: 1e3
    }, t;
  })(k)
);
export {
  x as default
};
