// echarts@6.0.0/lib/chart/sankey/sankeyVisual.js
import { each as r } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import p from "../../visual/VisualMapping.js";
function v(g) {
  g.eachSeriesByType("sankey", function(s) {
    var n = s.getGraph(), a = n.nodes, u = n.edges;
    if (a.length) {
      var l = 1 / 0, i = -1 / 0;
      r(a, function(e) {
        var t = e.getLayout().value;
        t < l && (l = t), t > i && (i = t);
      }), r(a, function(e) {
        var t = new p({
          type: "color",
          mappingMethod: "linear",
          dataExtent: [l, i],
          visual: s.get("color")
        }), f = t.mapValueToVisual(e.getLayout().value), o = e.getModel().get(["itemStyle", "color"]);
        o != null ? (e.setVisual("color", o), e.setVisual("style", {
          fill: o
        })) : (e.setVisual("color", f), e.setVisual("style", {
          fill: f
        }));
      });
    }
    u.length && r(u, function(e) {
      var t = e.getModel().get("lineStyle");
      e.setVisual("style", t);
    });
  });
}
export {
  v as default
};
