// echarts@6.0.0/lib/chart/sankey/sankeyLayout.js
import { filter as T, each as c } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { groupData as O } from "../../util/model.js";
import { createBoxLayoutReference as C, getLayoutRect as S } from "../../util/layout.js";
function tt(a, r) {
  a.eachSeriesByType("sankey", function(e) {
    var f = e.get("nodeWidth"), t = e.get("nodeGap"), u = C(e, r).refContainer, y = S(e.getBoxLayoutParams(), u);
    e.layoutInfo = y;
    var n = y.width, v = y.height, i = e.getGraph(), o = i.nodes, s = i.edges;
    b(o);
    var g = T(o, function(h) {
      return h.getLayout().value === 0;
    }), l = g.length !== 0 ? 0 : e.get("layoutIterations"), E = e.get("orient"), L = e.get("nodeAlign");
    z(o, s, f, t, n, v, l, E, L);
  });
}
function z(a, r, e, f, t, u, y, n, v) {
  H(a, r, e, t, u, n, v), X(a, r, u, t, f, y, n), Z(a, n);
}
function b(a) {
  c(a, function(r) {
    var e = p(r.outEdges, k), f = p(r.inEdges, k), t = r.getValue() || 0, u = Math.max(e, f, t);
    r.setLayout({
      value: u
    }, !0);
  });
}
function H(a, r, e, f, t, u, y) {
  for (var n = [], v = [], i = [], o = [], s = 0, g = 0; g < r.length; g++)
    n[g] = 1;
  for (var g = 0; g < a.length; g++)
    v[g] = a[g].inEdges.length, v[g] === 0 && i.push(a[g]);
  for (var l = -1; i.length; ) {
    for (var E = 0; E < i.length; E++) {
      var L = i[E], h = L.hostGraph.data.getRawDataItem(L.dataIndex), B = h.depth != null && h.depth >= 0;
      B && h.depth > l && (l = h.depth), L.setLayout({
        depth: B ? h.depth : s
      }, !0), u === "vertical" ? L.setLayout({
        dy: e
      }, !0) : L.setLayout({
        dx: e
      }, !0);
      for (var m = 0; m < L.outEdges.length; m++) {
        var R = L.outEdges[m], d = r.indexOf(R);
        n[d] = 0;
        var D = R.node2, A = a.indexOf(D);
        --v[A] === 0 && o.indexOf(D) < 0 && o.push(D);
      }
    }
    ++s, i = o, o = [];
  }
  for (var g = 0; g < n.length; g++)
    if (n[g] === 1)
      throw new Error("Sankey is a DAG, the original data has cycle!");
  var N = l > s - 1 ? l : s - 1;
  y && y !== "left" && K(a, y, u, N);
  var I = u === "vertical" ? (t - e) / N : (f - e) / N;
  W(a, I, u);
}
function V(a) {
  var r = a.hostGraph.data.getRawDataItem(a.dataIndex);
  return r.depth != null && r.depth >= 0;
}
function K(a, r, e, f) {
  if (r === "right") {
    for (var t = [], u = a, y = 0; u.length; ) {
      for (var n = 0; n < u.length; n++) {
        var v = u[n];
        v.setLayout({
          skNodeHeight: y
        }, !0);
        for (var i = 0; i < v.inEdges.length; i++) {
          var o = v.inEdges[i];
          t.indexOf(o.node1) < 0 && t.push(o.node1);
        }
      }
      u = t, t = [], ++y;
    }
    c(a, function(s) {
      V(s) || s.setLayout({
        depth: Math.max(0, f - s.getLayout().skNodeHeight)
      }, !0);
    });
  } else r === "justify" && P(a, f);
}
function P(a, r) {
  c(a, function(e) {
    !V(e) && !e.outEdges.length && e.setLayout({
      depth: r
    }, !0);
  });
}
function W(a, r, e) {
  c(a, function(f) {
    var t = f.getLayout().depth * r;
    e === "vertical" ? f.setLayout({
      y: t
    }, !0) : f.setLayout({
      x: t
    }, !0);
  });
}
function X(a, r, e, f, t, u, y) {
  var n = j(a, y);
  q(n, r, e, f, t, y), w(n, t, e, f, y);
  for (var v = 1; u > 0; u--)
    v *= 0.99, F(n, v, y), w(n, t, e, f, y), Y(n, v, y), w(n, t, e, f, y);
}
function j(a, r) {
  var e = [], f = r === "vertical" ? "y" : "x", t = O(a, function(u) {
    return u.getLayout()[f];
  });
  return t.keys.sort(function(u, y) {
    return u - y;
  }), c(t.keys, function(u) {
    e.push(t.buckets.get(u));
  }), e;
}
function q(a, r, e, f, t, u) {
  var y = 1 / 0;
  c(a, function(n) {
    var v = n.length, i = 0;
    c(n, function(s) {
      i += s.getLayout().value;
    });
    var o = u === "vertical" ? (f - (v - 1) * t) / i : (e - (v - 1) * t) / i;
    o < y && (y = o);
  }), c(a, function(n) {
    c(n, function(v, i) {
      var o = v.getLayout().value * y;
      u === "vertical" ? (v.setLayout({
        x: i
      }, !0), v.setLayout({
        dx: o
      }, !0)) : (v.setLayout({
        y: i
      }, !0), v.setLayout({
        dy: o
      }, !0));
    });
  }), c(r, function(n) {
    var v = +n.getValue() * y;
    n.setLayout({
      dy: v
    }, !0);
  });
}
function w(a, r, e, f, t) {
  var u = t === "vertical" ? "x" : "y";
  c(a, function(y) {
    y.sort(function(L, h) {
      return L.getLayout()[u] - h.getLayout()[u];
    });
    for (var n, v, i, o = 0, s = y.length, g = t === "vertical" ? "dx" : "dy", l = 0; l < s; l++)
      v = y[l], i = o - v.getLayout()[u], i > 0 && (n = v.getLayout()[u] + i, t === "vertical" ? v.setLayout({
        x: n
      }, !0) : v.setLayout({
        y: n
      }, !0)), o = v.getLayout()[u] + v.getLayout()[g] + r;
    var E = t === "vertical" ? f : e;
    if (i = o - r - E, i > 0) {
      n = v.getLayout()[u] - i, t === "vertical" ? v.setLayout({
        x: n
      }, !0) : v.setLayout({
        y: n
      }, !0), o = n;
      for (var l = s - 2; l >= 0; --l)
        v = y[l], i = v.getLayout()[u] + v.getLayout()[g] + r - o, i > 0 && (n = v.getLayout()[u] - i, t === "vertical" ? v.setLayout({
          x: n
        }, !0) : v.setLayout({
          y: n
        }, !0)), o = v.getLayout()[u];
    }
  });
}
function F(a, r, e) {
  c(a.slice().reverse(), function(f) {
    c(f, function(t) {
      if (t.outEdges.length) {
        var u = p(t.outEdges, G, e) / p(t.outEdges, k);
        if (isNaN(u)) {
          var y = t.outEdges.length;
          u = y ? p(t.outEdges, J, e) / y : 0;
        }
        if (e === "vertical") {
          var n = t.getLayout().x + (u - x(t, e)) * r;
          t.setLayout({
            x: n
          }, !0);
        } else {
          var v = t.getLayout().y + (u - x(t, e)) * r;
          t.setLayout({
            y: v
          }, !0);
        }
      }
    });
  });
}
function G(a, r) {
  return x(a.node2, r) * a.getValue();
}
function J(a, r) {
  return x(a.node2, r);
}
function Q(a, r) {
  return x(a.node1, r) * a.getValue();
}
function U(a, r) {
  return x(a.node1, r);
}
function x(a, r) {
  return r === "vertical" ? a.getLayout().x + a.getLayout().dx / 2 : a.getLayout().y + a.getLayout().dy / 2;
}
function k(a) {
  return a.getValue();
}
function p(a, r, e) {
  for (var f = 0, t = a.length, u = -1; ++u < t; ) {
    var y = +r(a[u], e);
    isNaN(y) || (f += y);
  }
  return f;
}
function Y(a, r, e) {
  c(a, function(f) {
    c(f, function(t) {
      if (t.inEdges.length) {
        var u = p(t.inEdges, Q, e) / p(t.inEdges, k);
        if (isNaN(u)) {
          var y = t.inEdges.length;
          u = y ? p(t.inEdges, U, e) / y : 0;
        }
        if (e === "vertical") {
          var n = t.getLayout().x + (u - x(t, e)) * r;
          t.setLayout({
            x: n
          }, !0);
        } else {
          var v = t.getLayout().y + (u - x(t, e)) * r;
          t.setLayout({
            y: v
          }, !0);
        }
      }
    });
  });
}
function Z(a, r) {
  var e = r === "vertical" ? "x" : "y";
  c(a, function(f) {
    f.outEdges.sort(function(t, u) {
      return t.node2.getLayout()[e] - u.node2.getLayout()[e];
    }), f.inEdges.sort(function(t, u) {
      return t.node1.getLayout()[e] - u.node1.getLayout()[e];
    });
  }), c(a, function(f) {
    var t = 0, u = 0;
    c(f.outEdges, function(y) {
      y.setLayout({
        sy: t
      }, !0), t += y.getLayout().dy;
    }), c(f.inEdges, function(y) {
      y.setLayout({
        ty: u
      }, !0), u += y.getLayout().dy;
    });
  });
}
export {
  tt as default
};
