// @dnd-kit/sortable@10.0.0/dist/sortable.esm.js
import B, { useMemo as E, useRef as P, useEffect as L, useContext as fe, useState as ge } from "react";
import { KeyboardCode as h, useDndContext as pe, closestCorners as be, getFirstCollision as ve, getScrollableAncestors as me, useDroppable as Ie, useDraggable as xe, getClientRect as he } from "../../../../../@dnd-kit_core@6.3.1_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/@dnd-kit/core/dist/core.esm.js";
import { useUniqueId as Ce, useIsomorphicLayoutEffect as H, subtract as we, useCombinedRefs as De, CSS as V, isKeyboardEvent as ye } from "../../../../../@dnd-kit_utilities@3.2.2_react@18.3.1/node_modules/@dnd-kit/utilities/dist/utilities.esm.js";
function Y(e, t, r) {
  const n = e.slice();
  return n.splice(r < 0 ? n.length + r : r, 0, n.splice(t, 1)[0]), n;
}
function Re(e, t) {
  return e.reduce((r, n, a) => {
    const d = t.get(n);
    return d && (r[a] = d), r;
  }, Array(e.length));
}
function N(e) {
  return e !== null && e >= 0;
}
function Se(e, t) {
  if (e === t)
    return !0;
  if (e.length !== t.length)
    return !1;
  for (let r = 0; r < e.length; r++)
    if (e[r] !== t[r])
      return !1;
  return !0;
}
function Ae(e) {
  return typeof e == "boolean" ? {
    draggable: e,
    droppable: e
  } : e;
}
const $ = (e) => {
  let {
    rects: t,
    activeIndex: r,
    overIndex: n,
    index: a
  } = e;
  const d = Y(t, n, r), o = t[a], u = d[a];
  return !u || !o ? null : {
    x: u.left - o.left,
    y: u.top - o.top,
    scaleX: u.width / o.width,
    scaleY: u.height / o.height
  };
}, G = "Sortable", J = /* @__PURE__ */ B.createContext({
  activeIndex: -1,
  containerId: G,
  disableTransforms: !1,
  items: [],
  overIndex: -1,
  useDragOverlay: !1,
  sortedRects: [],
  strategy: $,
  disabled: {
    draggable: !1,
    droppable: !1
  }
});
function Xe(e) {
  let {
    children: t,
    id: r,
    items: n,
    strategy: a = $,
    disabled: d = !1
  } = e;
  const {
    active: o,
    dragOverlay: u,
    droppableRects: f,
    over: g,
    measureDroppableContainers: s
  } = pe(), i = Ce(G, r), c = u.rect !== null, l = E(() => n.map((x) => typeof x == "object" && "id" in x ? x.id : x), [n]), C = o != null, w = o ? l.indexOf(o.id) : -1, v = g ? l.indexOf(g.id) : -1, D = P(l), I = !Se(l, D.current), m = v !== -1 && w === -1 || I, b = Ae(d);
  H(() => {
    I && C && s(l);
  }, [I, l, C, s]), L(() => {
    D.current = l;
  }, [l]);
  const y = E(
    () => ({
      activeIndex: w,
      containerId: i,
      disabled: b,
      disableTransforms: m,
      items: l,
      overIndex: v,
      useDragOverlay: c,
      sortedRects: Re(l, f),
      strategy: a
    }),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [w, i, b.draggable, b.droppable, m, l, v, f, c, a]
  );
  return B.createElement(J.Provider, {
    value: y
  }, t);
}
const Te = (e) => {
  let {
    id: t,
    items: r,
    activeIndex: n,
    overIndex: a
  } = e;
  return Y(r, n, a).indexOf(t);
}, Oe = (e) => {
  let {
    containerId: t,
    isSorting: r,
    wasDragging: n,
    index: a,
    items: d,
    newIndex: o,
    previousItems: u,
    previousContainerId: f,
    transition: g
  } = e;
  return !g || !n || u !== d && a === o ? !1 : r ? !0 : o !== a && t === f;
}, Ne = {
  duration: 200,
  easing: "ease"
}, Q = "transform", Ee = /* @__PURE__ */ V.Transition.toString({
  property: Q,
  duration: 0,
  easing: "linear"
}), Le = {
  roleDescription: "sortable"
};
function _e(e) {
  let {
    disabled: t,
    index: r,
    node: n,
    rect: a
  } = e;
  const [d, o] = ge(null), u = P(r);
  return H(() => {
    if (!t && r !== u.current && n.current) {
      const f = a.current;
      if (f) {
        const g = he(n.current, {
          ignoreTransform: !0
        }), s = {
          x: f.left - g.left,
          y: f.top - g.top,
          scaleX: f.width / g.width,
          scaleY: f.height / g.height
        };
        (s.x || s.y) && o(s);
      }
    }
    r !== u.current && (u.current = r);
  }, [t, r, n, a]), L(() => {
    d && o(null);
  }, [d]), d;
}
function je(e) {
  let {
    animateLayoutChanges: t = Oe,
    attributes: r,
    disabled: n,
    data: a,
    getNewIndex: d = Te,
    id: o,
    strategy: u,
    resizeObserverConfig: f,
    transition: g = Ne
  } = e;
  const {
    items: s,
    containerId: i,
    activeIndex: c,
    disabled: l,
    disableTransforms: C,
    sortedRects: w,
    overIndex: v,
    useDragOverlay: D,
    strategy: I
  } = fe(J), m = ke(n, l), b = s.indexOf(o), y = E(() => ({
    sortable: {
      containerId: i,
      index: b,
      items: s
    },
    ...a
  }), [i, a, b, s]), x = E(() => s.slice(s.indexOf(o)), [s, o]), {
    rect: O,
    node: F,
    isOver: Z,
    setNodeRef: K
  } = Ie({
    id: o,
    data: y,
    disabled: m.droppable,
    resizeObserverConfig: {
      updateMeasurementsFor: x,
      ...f
    }
  }), {
    active: A,
    activatorEvent: ee,
    activeNodeRect: te,
    attributes: re,
    setNodeRef: M,
    listeners: ne,
    isDragging: k,
    over: oe,
    setActivatorNodeRef: se,
    transform: ie
  } = xe({
    id: o,
    data: y,
    attributes: {
      ...Le,
      ...r
    },
    disabled: m.draggable
  }), ae = De(K, M), R = !!A, U = R && !C && N(c) && N(v), X = !D && k, j = X && U ? ie : null, ce = U ? j ?? (u ?? I)({
    rects: w,
    activeNodeRect: te,
    activeIndex: c,
    overIndex: v,
    index: b
  }) : null, T = N(c) && N(v) ? d({
    id: o,
    items: s,
    activeIndex: c,
    overIndex: v
  }) : b, S = A?.id, p = P({
    activeId: S,
    items: s,
    newIndex: T,
    containerId: i
  }), le = s !== p.current.items, q = t({
    active: A,
    containerId: i,
    isDragging: k,
    isSorting: R,
    id: o,
    index: b,
    items: s,
    newIndex: p.current.newIndex,
    previousItems: p.current.items,
    previousContainerId: p.current.containerId,
    transition: g,
    wasDragging: p.current.activeId != null
  }), z = _e({
    disabled: !q,
    index: b,
    node: F,
    rect: O
  });
  return L(() => {
    R && p.current.newIndex !== T && (p.current.newIndex = T), i !== p.current.containerId && (p.current.containerId = i), s !== p.current.items && (p.current.items = s);
  }, [R, T, i, s]), L(() => {
    if (S === p.current.activeId)
      return;
    if (S != null && p.current.activeId == null) {
      p.current.activeId = S;
      return;
    }
    const ue = setTimeout(() => {
      p.current.activeId = S;
    }, 50);
    return () => clearTimeout(ue);
  }, [S]), {
    active: A,
    activeIndex: c,
    attributes: re,
    data: y,
    rect: O,
    index: b,
    newIndex: T,
    items: s,
    isOver: Z,
    isSorting: R,
    isDragging: k,
    listeners: ne,
    node: F,
    overIndex: v,
    over: oe,
    setNodeRef: ae,
    setActivatorNodeRef: se,
    setDroppableNodeRef: K,
    setDraggableNodeRef: M,
    transform: z ?? ce,
    transition: de()
  };
  function de() {
    if (
      // Temporarily disable transitions for a single frame to set up derived transforms
      z || // Or to prevent items jumping to back to their "new" position when items change
      le && p.current.newIndex === b
    )
      return Ee;
    if (!(X && !ye(ee) || !g) && (R || q))
      return V.Transition.toString({
        ...g,
        property: Q
      });
  }
}
function ke(e, t) {
  var r, n;
  return typeof e == "boolean" ? {
    draggable: e,
    // Backwards compatibility
    droppable: !1
  } : {
    draggable: (r = e?.draggable) != null ? r : t.draggable,
    droppable: (n = e?.droppable) != null ? n : t.droppable
  };
}
function _(e) {
  if (!e)
    return !1;
  const t = e.data.current;
  return !!(t && "sortable" in t && typeof t.sortable == "object" && "containerId" in t.sortable && "items" in t.sortable && "index" in t.sortable);
}
const ze = [h.Down, h.Right, h.Up, h.Left], qe = (e, t) => {
  let {
    context: {
      active: r,
      collisionRect: n,
      droppableRects: a,
      droppableContainers: d,
      over: o,
      scrollableAncestors: u
    }
  } = t;
  if (ze.includes(e.code)) {
    if (e.preventDefault(), !r || !n)
      return;
    const f = [];
    d.getEnabled().forEach((i) => {
      if (!i || i != null && i.disabled)
        return;
      const c = a.get(i.id);
      if (c)
        switch (e.code) {
          case h.Down:
            n.top < c.top && f.push(i);
            break;
          case h.Up:
            n.top > c.top && f.push(i);
            break;
          case h.Left:
            n.left > c.left && f.push(i);
            break;
          case h.Right:
            n.left < c.left && f.push(i);
            break;
        }
    });
    const g = be({
      collisionRect: n,
      droppableRects: a,
      droppableContainers: f
    });
    let s = ve(g, "id");
    if (s === o?.id && g.length > 1 && (s = g[1].id), s != null) {
      const i = d.get(r.id), c = d.get(s), l = c ? a.get(c.id) : null, C = c?.node.current;
      if (C && l && i && c) {
        const v = me(C).some((x, O) => u[O] !== x), D = W(i, c), I = Pe(i, c), m = v || !D ? {
          x: 0,
          y: 0
        } : {
          x: I ? n.width - l.width : 0,
          y: I ? n.height - l.height : 0
        }, b = {
          x: l.left,
          y: l.top
        };
        return m.x && m.y ? b : we(b, m);
      }
    }
  }
};
function W(e, t) {
  return !_(e) || !_(t) ? !1 : e.data.current.sortable.containerId === t.data.current.sortable.containerId;
}
function Pe(e, t) {
  return !_(e) || !_(t) || !W(e, t) ? !1 : e.data.current.sortable.index < t.data.current.sortable.index;
}
export {
  Xe as SortableContext,
  Y as arrayMove,
  Oe as defaultAnimateLayoutChanges,
  Te as defaultNewIndexGetter,
  _ as hasSortableData,
  $ as rectSortingStrategy,
  qe as sortableKeyboardCoordinates,
  je as useSortable
};
