// zrender@6.0.0/lib/animation/Animator.js
import Z from "./Clip.js";
import { parse as R } from "../tool/color.js";
import { logError as $, isArrayLike as S, keys as q, map as M, isNumber as D, eqNaN as w, isString as H, isGradientObject as V, extend as j, isFunction as tt } from "../core/util.js";
import et from "./easing.js";
import { createCubicEasingFunc as rt } from "./cubicEasing.js";
import { isLinearGradient as it, isRadialGradient as st } from "../svg/helper.js";
var P = Array.prototype.slice;
function y(i, t, r) {
  return (t - i) * r + i;
}
function N(i, t, r, e) {
  for (var s = t.length, n = 0; n < s; n++)
    i[n] = y(t[n], r[n], e);
  return i;
}
function nt(i, t, r, e) {
  for (var s = t.length, n = s && t[0].length, o = 0; o < s; o++) {
    i[o] || (i[o] = []);
    for (var a = 0; a < n; a++)
      i[o][a] = y(t[o][a], r[o][a], e);
  }
  return i;
}
function C(i, t, r, e) {
  for (var s = t.length, n = 0; n < s; n++)
    i[n] = t[n] + r[n] * e;
  return i;
}
function B(i, t, r, e) {
  for (var s = t.length, n = s && t[0].length, o = 0; o < s; o++) {
    i[o] || (i[o] = []);
    for (var a = 0; a < n; a++)
      i[o][a] = t[o][a] + r[o][a] * e;
  }
  return i;
}
function at(i, t) {
  for (var r = i.length, e = t.length, s = r > e ? t : i, n = Math.min(r, e), o = s[n - 1] || { color: [0, 0, 0, 0], offset: 0 }, a = n; a < Math.max(r, e); a++)
    s.push({
      offset: o.offset,
      color: o.color.slice()
    });
}
function ot(i, t, r) {
  var e = i, s = t;
  if (!(!e.push || !s.push)) {
    var n = e.length, o = s.length;
    if (n !== o) {
      var a = n > o;
      if (a)
        e.length = o;
      else
        for (var h = n; h < o; h++)
          e.push(r === 1 ? s[h] : P.call(s[h]));
    }
    for (var l = e[0] && e[0].length, h = 0; h < e.length; h++)
      if (r === 1)
        isNaN(e[h]) && (e[h] = s[h]);
      else
        for (var f = 0; f < l; f++)
          isNaN(e[h][f]) && (e[h][f] = s[h][f]);
  }
}
function K(i) {
  if (S(i)) {
    var t = i.length;
    if (S(i[0])) {
      for (var r = [], e = 0; e < t; e++)
        r.push(P.call(i[e]));
      return r;
    }
    return P.call(i);
  }
  return i;
}
function L(i) {
  return i[0] = Math.floor(i[0]) || 0, i[1] = Math.floor(i[1]) || 0, i[2] = Math.floor(i[2]) || 0, i[3] = i[3] == null ? 1 : i[3], "rgba(" + i.join(",") + ")";
}
function ft(i) {
  return S(i && i[0]) ? 2 : 1;
}
var F = 0, x = 1, Q = 2, b = 3, U = 4, Y = 5, z = 6;
function J(i) {
  return i === U || i === Y;
}
function E(i) {
  return i === x || i === Q;
}
var k = [0, 0, 0, 0], ht = (function() {
  function i(t) {
    this.keyframes = [], this.discrete = !1, this._invalid = !1, this._needsSort = !1, this._lastFr = 0, this._lastFrP = 0, this.propName = t;
  }
  return i.prototype.isFinished = function() {
    return this._finished;
  }, i.prototype.setFinished = function() {
    this._finished = !0, this._additiveTrack && this._additiveTrack.setFinished();
  }, i.prototype.needsAnimate = function() {
    return this.keyframes.length >= 1;
  }, i.prototype.getAdditiveTrack = function() {
    return this._additiveTrack;
  }, i.prototype.addKeyframe = function(t, r, e) {
    this._needsSort = !0;
    var s = this.keyframes, n = s.length, o = !1, a = z, h = r;
    if (S(r)) {
      var l = ft(r);
      a = l, (l === 1 && !D(r[0]) || l === 2 && !D(r[0][0])) && (o = !0);
    } else if (D(r) && !w(r))
      a = F;
    else if (H(r))
      if (!isNaN(+r))
        a = F;
      else {
        var f = R(r);
        f && (h = f, a = b);
      }
    else if (V(r)) {
      var p = j({}, h);
      p.colorStops = M(r.colorStops, function(u) {
        return {
          offset: u.offset,
          color: R(u.color)
        };
      }), it(r) ? a = U : st(r) && (a = Y), h = p;
    }
    n === 0 ? this.valType = a : (a !== this.valType || a === z) && (o = !0), this.discrete = this.discrete || o;
    var d = {
      time: t,
      value: h,
      rawValue: r,
      percent: 0
    };
    return e && (d.easing = e, d.easingFunc = tt(e) ? e : et[e] || rt(e)), s.push(d), d;
  }, i.prototype.prepare = function(t, r) {
    var e = this.keyframes;
    this._needsSort && e.sort(function(g, c) {
      return g.time - c.time;
    });
    for (var s = this.valType, n = e.length, o = e[n - 1], a = this.discrete, h = E(s), l = J(s), f = 0; f < n; f++) {
      var p = e[f], d = p.value, u = o.value;
      p.percent = p.time / t, a || (h && f !== n - 1 ? ot(d, u, s) : l && at(d.colorStops, u.colorStops));
    }
    if (!a && s !== Y && r && this.needsAnimate() && r.needsAnimate() && s === r.valType && !r._finished) {
      this._additiveTrack = r;
      for (var v = e[0].value, f = 0; f < n; f++)
        s === F ? e[f].additiveValue = e[f].value - v : s === b ? e[f].additiveValue = C([], e[f].value, v, -1) : E(s) && (e[f].additiveValue = s === x ? C([], e[f].value, v, -1) : B([], e[f].value, v, -1));
    }
  }, i.prototype.step = function(t, r) {
    if (!this._finished) {
      this._additiveTrack && this._additiveTrack._finished && (this._additiveTrack = null);
      var e = this._additiveTrack != null, s = e ? "additiveValue" : "value", n = this.valType, o = this.keyframes, a = o.length, h = this.propName, l = n === b, f, p = this._lastFr, d = Math.min, u, v;
      if (a === 1)
        u = v = o[0];
      else {
        if (r < 0)
          f = 0;
        else if (r < this._lastFrP) {
          var g = d(p + 1, a - 1);
          for (f = g; f >= 0 && !(o[f].percent <= r); f--)
            ;
          f = d(f, a - 2);
        } else {
          for (f = p; f < a && !(o[f].percent > r); f++)
            ;
          f = d(f - 1, a - 2);
        }
        v = o[f + 1], u = o[f];
      }
      if (u && v) {
        this._lastFr = f, this._lastFrP = r;
        var c = v.percent - u.percent, _ = c === 0 ? 1 : d((r - u.percent) / c, 1);
        v.easingFunc && (_ = v.easingFunc(_));
        var T = e ? this._additiveValue : l ? k : t[h];
        if ((E(n) || l) && !T && (T = this._additiveValue = []), this.discrete)
          t[h] = _ < 1 ? u.rawValue : v.rawValue;
        else if (E(n))
          n === x ? N(T, u[s], v[s], _) : nt(T, u[s], v[s], _);
        else if (J(n)) {
          var A = u[s], m = v[s], G = n === U;
          t[h] = {
            type: G ? "linear" : "radial",
            x: y(A.x, m.x, _),
            y: y(A.y, m.y, _),
            colorStops: M(A.colorStops, function(O, X) {
              var W = m.colorStops[X];
              return {
                offset: y(O.offset, W.offset, _),
                color: L(N([], O.color, W.color, _))
              };
            }),
            global: m.global
          }, G ? (t[h].x2 = y(A.x2, m.x2, _), t[h].y2 = y(A.y2, m.y2, _)) : t[h].r = y(A.r, m.r, _);
        } else if (l)
          N(T, u[s], v[s], _), e || (t[h] = L(T));
        else {
          var I = y(u[s], v[s], _);
          e ? this._additiveValue = I : t[h] = I;
        }
        e && this._addToTarget(t);
      }
    }
  }, i.prototype._addToTarget = function(t) {
    var r = this.valType, e = this.propName, s = this._additiveValue;
    r === F ? t[e] = t[e] + s : r === b ? (R(t[e], k), C(k, k, s, 1), t[e] = L(k)) : r === x ? C(t[e], t[e], s, 1) : r === Q && B(t[e], t[e], s, 1);
  }, i;
})(), pt = (function() {
  function i(t, r, e, s) {
    if (this._tracks = {}, this._trackKeys = [], this._maxTime = 0, this._started = 0, this._clip = null, this._target = t, this._loop = r, r && s) {
      $("Can' use additive animation on looped animation.");
      return;
    }
    this._additiveAnimators = s, this._allowDiscrete = e;
  }
  return i.prototype.getMaxTime = function() {
    return this._maxTime;
  }, i.prototype.getDelay = function() {
    return this._delay;
  }, i.prototype.getLoop = function() {
    return this._loop;
  }, i.prototype.getTarget = function() {
    return this._target;
  }, i.prototype.changeTarget = function(t) {
    this._target = t;
  }, i.prototype.when = function(t, r, e) {
    return this.whenWithKeys(t, r, q(r), e);
  }, i.prototype.whenWithKeys = function(t, r, e, s) {
    for (var n = this._tracks, o = 0; o < e.length; o++) {
      var a = e[o], h = n[a];
      if (!h) {
        h = n[a] = new ht(a);
        var l = void 0, f = this._getAdditiveTrack(a);
        if (f) {
          var p = f.keyframes, d = p[p.length - 1];
          l = d && d.value, f.valType === b && l && (l = L(l));
        } else
          l = this._target[a];
        if (l == null)
          continue;
        t > 0 && h.addKeyframe(0, K(l), s), this._trackKeys.push(a);
      }
      h.addKeyframe(t, K(r[a]), s);
    }
    return this._maxTime = Math.max(this._maxTime, t), this;
  }, i.prototype.pause = function() {
    this._clip.pause(), this._paused = !0;
  }, i.prototype.resume = function() {
    this._clip.resume(), this._paused = !1;
  }, i.prototype.isPaused = function() {
    return !!this._paused;
  }, i.prototype.duration = function(t) {
    return this._maxTime = t, this._force = !0, this;
  }, i.prototype._doneCallback = function() {
    this._setTracksFinished(), this._clip = null;
    var t = this._doneCbs;
    if (t)
      for (var r = t.length, e = 0; e < r; e++)
        t[e].call(this);
  }, i.prototype._abortedCallback = function() {
    this._setTracksFinished();
    var t = this.animation, r = this._abortedCbs;
    if (t && t.removeClip(this._clip), this._clip = null, r)
      for (var e = 0; e < r.length; e++)
        r[e].call(this);
  }, i.prototype._setTracksFinished = function() {
    for (var t = this._tracks, r = this._trackKeys, e = 0; e < r.length; e++)
      t[r[e]].setFinished();
  }, i.prototype._getAdditiveTrack = function(t) {
    var r, e = this._additiveAnimators;
    if (e)
      for (var s = 0; s < e.length; s++) {
        var n = e[s].getTrack(t);
        n && (r = n);
      }
    return r;
  }, i.prototype.start = function(t) {
    if (!(this._started > 0)) {
      this._started = 1;
      for (var r = this, e = [], s = this._maxTime || 0, n = 0; n < this._trackKeys.length; n++) {
        var o = this._trackKeys[n], a = this._tracks[o], h = this._getAdditiveTrack(o), l = a.keyframes, f = l.length;
        if (a.prepare(s, h), a.needsAnimate())
          if (!this._allowDiscrete && a.discrete) {
            var p = l[f - 1];
            p && (r._target[a.propName] = p.rawValue), a.setFinished();
          } else
            e.push(a);
      }
      if (e.length || this._force) {
        var d = new Z({
          life: s,
          loop: this._loop,
          delay: this._delay || 0,
          onframe: function(u) {
            r._started = 2;
            var v = r._additiveAnimators;
            if (v) {
              for (var g = !1, c = 0; c < v.length; c++)
                if (v[c]._clip) {
                  g = !0;
                  break;
                }
              g || (r._additiveAnimators = null);
            }
            for (var c = 0; c < e.length; c++)
              e[c].step(r._target, u);
            var _ = r._onframeCbs;
            if (_)
              for (var c = 0; c < _.length; c++)
                _[c](r._target, u);
          },
          ondestroy: function() {
            r._doneCallback();
          }
        });
        this._clip = d, this.animation && this.animation.addClip(d), t && d.setEasing(t);
      } else
        this._doneCallback();
      return this;
    }
  }, i.prototype.stop = function(t) {
    if (this._clip) {
      var r = this._clip;
      t && r.onframe(1), this._abortedCallback();
    }
  }, i.prototype.delay = function(t) {
    return this._delay = t, this;
  }, i.prototype.during = function(t) {
    return t && (this._onframeCbs || (this._onframeCbs = []), this._onframeCbs.push(t)), this;
  }, i.prototype.done = function(t) {
    return t && (this._doneCbs || (this._doneCbs = []), this._doneCbs.push(t)), this;
  }, i.prototype.aborted = function(t) {
    return t && (this._abortedCbs || (this._abortedCbs = []), this._abortedCbs.push(t)), this;
  }, i.prototype.getClip = function() {
    return this._clip;
  }, i.prototype.getTrack = function(t) {
    return this._tracks[t];
  }, i.prototype.getTracks = function() {
    var t = this;
    return M(this._trackKeys, function(r) {
      return t._tracks[r];
    });
  }, i.prototype.stopTracks = function(t, r) {
    if (!t.length || !this._clip)
      return !0;
    for (var e = this._tracks, s = this._trackKeys, n = 0; n < t.length; n++) {
      var o = e[t[n]];
      o && !o.isFinished() && (r ? o.step(this._target, 1) : this._started === 1 && o.step(this._target, 0), o.setFinished());
    }
    for (var a = !0, n = 0; n < s.length; n++)
      if (!e[s[n]].isFinished()) {
        a = !1;
        break;
      }
    return a && this._abortedCallback(), a;
  }, i.prototype.saveTo = function(t, r, e) {
    if (t) {
      r = r || this._trackKeys;
      for (var s = 0; s < r.length; s++) {
        var n = r[s], o = this._tracks[n];
        if (!(!o || o.isFinished())) {
          var a = o.keyframes, h = a[e ? 0 : a.length - 1];
          h && (t[n] = K(h.rawValue));
        }
      }
    }
  }, i.prototype.__changeFinalValue = function(t, r) {
    r = r || q(t);
    for (var e = 0; e < r.length; e++) {
      var s = r[e], n = this._tracks[s];
      if (n) {
        var o = n.keyframes;
        if (o.length > 1) {
          var a = o.pop();
          n.addKeyframe(a.time, t[s]), n.prepare(this._maxTime, n.getAdditiveTrack());
        }
      }
    }
  }, i;
})();
export {
  K as cloneValue,
  pt as default
};
