// zrender@6.0.0/lib/animation/Animation.js
import { __extends as h } from "../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import f from "../core/Eventful.js";
import u from "./requestAnimationFrame.js";
import m from "./Animator.js";
function r() {
  return (/* @__PURE__ */ new Date()).getTime();
}
var g = (function(s) {
  h(e, s);
  function e(t) {
    var i = s.call(this) || this;
    return i._running = !1, i._time = 0, i._pausedTime = 0, i._pauseStart = 0, i._paused = !1, t = t || {}, i.stage = t.stage || {}, i;
  }
  return e.prototype.addClip = function(t) {
    t.animation && this.removeClip(t), this._head ? (this._tail.next = t, t.prev = this._tail, t.next = null, this._tail = t) : this._head = this._tail = t, t.animation = this;
  }, e.prototype.addAnimator = function(t) {
    t.animation = this;
    var i = t.getClip();
    i && this.addClip(i);
  }, e.prototype.removeClip = function(t) {
    if (t.animation) {
      var i = t.prev, n = t.next;
      i ? i.next = n : this._head = n, n ? n.prev = i : this._tail = i, t.next = t.prev = t.animation = null;
    }
  }, e.prototype.removeAnimator = function(t) {
    var i = t.getClip();
    i && this.removeClip(i), t.animation = null;
  }, e.prototype.update = function(t) {
    for (var i = r() - this._pausedTime, n = i - this._time, a = this._head; a; ) {
      var o = a.next, p = a.step(i, n);
      p && (a.ondestroy(), this.removeClip(a)), a = o;
    }
    this._time = i, t || (this.trigger("frame", n), this.stage.update && this.stage.update());
  }, e.prototype._startLoop = function() {
    var t = this;
    this._running = !0;
    function i() {
      t._running && (u(i), !t._paused && t.update());
    }
    u(i);
  }, e.prototype.start = function() {
    this._running || (this._time = r(), this._pausedTime = 0, this._startLoop());
  }, e.prototype.stop = function() {
    this._running = !1;
  }, e.prototype.pause = function() {
    this._paused || (this._pauseStart = r(), this._paused = !0);
  }, e.prototype.resume = function() {
    this._paused && (this._pausedTime += r() - this._pauseStart, this._paused = !1);
  }, e.prototype.clear = function() {
    for (var t = this._head; t; ) {
      var i = t.next;
      t.prev = t.next = t.animation = null, t = i;
    }
    this._head = this._tail = null;
  }, e.prototype.isFinished = function() {
    return this._head == null;
  }, e.prototype.animate = function(t, i) {
    i = i || {}, this.start();
    var n = new m(t, i.loop);
    return this.addAnimator(n), n;
  }, e;
})(f);
export {
  g as default,
  r as getTime
};
