// zrender@6.0.0/lib/animation/Clip.js
import f from "./easing.js";
import { noop as s, isFunction as p } from "../core/util.js";
import { createCubicEasingFunc as d } from "./cubicEasing.js";
var c = (function() {
  function e(t) {
    this._inited = !1, this._startTime = 0, this._pausedTime = 0, this._paused = !1, this._life = t.life || 1e3, this._delay = t.delay || 0, this.loop = t.loop || !1, this.onframe = t.onframe || s, this.ondestroy = t.ondestroy || s, this.onrestart = t.onrestart || s, t.easing && this.setEasing(t.easing);
  }
  return e.prototype.step = function(t, h) {
    if (this._inited || (this._startTime = t + this._delay, this._inited = !0), this._paused) {
      this._pausedTime += h;
      return;
    }
    var r = this._life, a = t - this._startTime - this._pausedTime, i = a / r;
    i < 0 && (i = 0), i = Math.min(i, 1);
    var n = this.easingFunc, u = n ? n(i) : i;
    if (this.onframe(u), i === 1)
      if (this.loop) {
        var o = a % r;
        this._startTime = t - o, this._pausedTime = 0, this.onrestart();
      } else
        return !0;
    return !1;
  }, e.prototype.pause = function() {
    this._paused = !0;
  }, e.prototype.resume = function() {
    this._paused = !1;
  }, e.prototype.setEasing = function(t) {
    this.easing = t, this.easingFunc = p(t) ? t : f[t] || d(t);
  }, e;
})();
export {
  c as default
};
