// zrender@6.0.0/lib/animation/easing.js
var i = {
  linear: function(n) {
    return n;
  },
  quadraticIn: function(n) {
    return n * n;
  },
  quadraticOut: function(n) {
    return n * (2 - n);
  },
  quadraticInOut: function(n) {
    return (n *= 2) < 1 ? 0.5 * n * n : -0.5 * (--n * (n - 2) - 1);
  },
  cubicIn: function(n) {
    return n * n * n;
  },
  cubicOut: function(n) {
    return --n * n * n + 1;
  },
  cubicInOut: function(n) {
    return (n *= 2) < 1 ? 0.5 * n * n * n : 0.5 * ((n -= 2) * n * n + 2);
  },
  quarticIn: function(n) {
    return n * n * n * n;
  },
  quarticOut: function(n) {
    return 1 - --n * n * n * n;
  },
  quarticInOut: function(n) {
    return (n *= 2) < 1 ? 0.5 * n * n * n * n : -0.5 * ((n -= 2) * n * n * n - 2);
  },
  quinticIn: function(n) {
    return n * n * n * n * n;
  },
  quinticOut: function(n) {
    return --n * n * n * n * n + 1;
  },
  quinticInOut: function(n) {
    return (n *= 2) < 1 ? 0.5 * n * n * n * n * n : 0.5 * ((n -= 2) * n * n * n * n + 2);
  },
  sinusoidalIn: function(n) {
    return 1 - Math.cos(n * Math.PI / 2);
  },
  sinusoidalOut: function(n) {
    return Math.sin(n * Math.PI / 2);
  },
  sinusoidalInOut: function(n) {
    return 0.5 * (1 - Math.cos(Math.PI * n));
  },
  exponentialIn: function(n) {
    return n === 0 ? 0 : Math.pow(1024, n - 1);
  },
  exponentialOut: function(n) {
    return n === 1 ? 1 : 1 - Math.pow(2, -10 * n);
  },
  exponentialInOut: function(n) {
    return n === 0 ? 0 : n === 1 ? 1 : (n *= 2) < 1 ? 0.5 * Math.pow(1024, n - 1) : 0.5 * (-Math.pow(2, -10 * (n - 1)) + 2);
  },
  circularIn: function(n) {
    return 1 - Math.sqrt(1 - n * n);
  },
  circularOut: function(n) {
    return Math.sqrt(1 - --n * n);
  },
  circularInOut: function(n) {
    return (n *= 2) < 1 ? -0.5 * (Math.sqrt(1 - n * n) - 1) : 0.5 * (Math.sqrt(1 - (n -= 2) * n) + 1);
  },
  elasticIn: function(n) {
    var t, r = 0.1, u = 0.4;
    return n === 0 ? 0 : n === 1 ? 1 : (!r || r < 1 ? (r = 1, t = u / 4) : t = u * Math.asin(1 / r) / (2 * Math.PI), -(r * Math.pow(2, 10 * (n -= 1)) * Math.sin((n - t) * (2 * Math.PI) / u)));
  },
  elasticOut: function(n) {
    var t, r = 0.1, u = 0.4;
    return n === 0 ? 0 : n === 1 ? 1 : (!r || r < 1 ? (r = 1, t = u / 4) : t = u * Math.asin(1 / r) / (2 * Math.PI), r * Math.pow(2, -10 * n) * Math.sin((n - t) * (2 * Math.PI) / u) + 1);
  },
  elasticInOut: function(n) {
    var t, r = 0.1, u = 0.4;
    return n === 0 ? 0 : n === 1 ? 1 : (!r || r < 1 ? (r = 1, t = u / 4) : t = u * Math.asin(1 / r) / (2 * Math.PI), (n *= 2) < 1 ? -0.5 * (r * Math.pow(2, 10 * (n -= 1)) * Math.sin((n - t) * (2 * Math.PI) / u)) : r * Math.pow(2, -10 * (n -= 1)) * Math.sin((n - t) * (2 * Math.PI) / u) * 0.5 + 1);
  },
  backIn: function(n) {
    var t = 1.70158;
    return n * n * ((t + 1) * n - t);
  },
  backOut: function(n) {
    var t = 1.70158;
    return --n * n * ((t + 1) * n + t) + 1;
  },
  backInOut: function(n) {
    var t = 2.5949095;
    return (n *= 2) < 1 ? 0.5 * (n * n * ((t + 1) * n - t)) : 0.5 * ((n -= 2) * n * ((t + 1) * n + t) + 2);
  },
  bounceIn: function(n) {
    return 1 - i.bounceOut(1 - n);
  },
  bounceOut: function(n) {
    return n < 0.36363636363636365 ? 7.5625 * n * n : n < 0.7272727272727273 ? 7.5625 * (n -= 0.5454545454545454) * n + 0.75 : n < 0.9090909090909091 ? 7.5625 * (n -= 0.8181818181818182) * n + 0.9375 : 7.5625 * (n -= 0.9545454545454546) * n + 0.984375;
  },
  bounceInOut: function(n) {
    return n < 0.5 ? i.bounceIn(n * 2) * 0.5 : i.bounceOut(n * 2 - 1) * 0.5 + 0.5;
  }
};
export {
  i as default
};
