// echarts@6.0.0/lib/chart/themeRiver/themeRiverLayout.js
import { map as x } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { parsePercent as g } from "../../util/number.js";
function R(a, c) {
  a.eachSeriesByType("themeRiver", function(o) {
    var y = o.getData(), n = o.coordinateSystem, e = {}, t = n.getRect();
    e.rect = t;
    var r = o.get("boundaryGap"), u = n.getAxis();
    if (e.boundaryGap = r, u.orient === "horizontal") {
      r[0] = g(r[0], t.height), r[1] = g(r[1], t.height);
      var m = t.height - r[0] - r[1];
      L(y, o, m);
    } else {
      r[0] = g(r[0], t.width), r[1] = g(r[1], t.width);
      var i = t.width - r[0] - r[1];
      L(y, o, i);
    }
    y.setLayout("layoutInfo", e);
  });
}
function L(a, c, o) {
  if (a.count())
    for (var y = c.coordinateSystem, n = c.getLayerSeries(), e = a.mapDimension("single"), t = a.mapDimension("value"), r = x(n, function(I) {
      return x(I.indices, function(p) {
        var l = y.dataToPoint(a.get(e, p));
        return l[1] = a.get(t, p), l;
      });
    }), u = S(r), m = u.y0, i = o / u.max, h = n.length, b = n[0].indices.length, s, v = 0; v < b; ++v) {
      s = m[v] * i, a.setItemLayout(n[0].indices[v], {
        layerIndex: 0,
        x: r[0][v][0],
        y0: s,
        y: r[0][v][1] * i
      });
      for (var f = 1; f < h; ++f)
        s += r[f - 1][v][1] * i, a.setItemLayout(n[f].indices[v], {
          layerIndex: f,
          x: r[f][v][0],
          y0: s,
          y: r[f][v][1] * i
        });
    }
}
function S(a) {
  for (var c = a.length, o = a[0].length, y = [], n = [], e = 0, t = 0; t < o; ++t) {
    for (var r = 0, u = 0; u < c; ++u)
      r += a[u][t][1];
    r > e && (e = r), y.push(r);
  }
  for (var m = 0; m < o; ++m)
    n[m] = (e - y[m]) / 2;
  e = 0;
  for (var i = 0; i < o; ++i) {
    var h = y[i] + n[i];
    h > e && (e = h);
  }
  return {
    y0: n,
    max: e
  };
}
export {
  R as default
};
