// echarts@6.0.0/lib/chart/themeRiver/ThemeRiverView.js
import { __extends as A } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { ECPolygon as j } from "../line/poly.js";
import { setStatesStylesFromModel as z, toggleHoverEmphasis as B } from "../../util/states.js";
import { setLabelStyle as F, getLabelStatesModels as H } from "../../label/labelStyle.js";
import { bind as x } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import N from "../../data/DataDiffer.js";
import q from "../../view/Chart.js";
import { updateProps as J, saveOldStyle as K, initProps as Q } from "../../animation/basicTransition.js";
import U from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import W from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Rect.js";
var se = (
  /** @class */
  (function(a) {
    A(r, a);
    function r() {
      var e = a !== null && a.apply(this, arguments) || this;
      return e.type = r.type, e._layers = [], e;
    }
    return r.prototype.render = function(e, h, Y) {
      var o = e.getData(), E = this, l = this.group, v = e.getLayerSeries(), C = o.getLayout("layoutInfo"), _ = C.rect, D = C.boundaryGap;
      l.x = 0, l.y = _.y + D[0];
      function L(m) {
        return m.name;
      }
      var R = new N(this._layersSeries || [], v, L, L), d = [];
      R.add(x(u, this, "add")).update(x(u, this, "update")).remove(x(u, this, "remove")).execute();
      function u(m, p, T) {
        var G = E._layers;
        if (m === "remove") {
          l.remove(G[p]);
          return;
        }
        for (var y = [], f = [], g, i = v[p].indices, n = 0; n < i.length; n++) {
          var c = o.getItemLayout(i[n]), I = c.x, P = c.y0, k = c.y;
          y.push(I, P), f.push(I, P + k), g = o.getItemVisual(i[n], "style");
        }
        var t, S = o.getItemLayout(i[0]), O = e.getModel("label"), V = O.get("margin"), b = e.getModel("emphasis");
        if (m === "add") {
          var s = d[p] = new U();
          t = new j({
            shape: {
              points: y,
              stackedOnPoints: f,
              smooth: 0.4,
              stackedOnSmooth: 0.4,
              smoothConstraint: !1
            },
            z2: 0
          }), s.add(t), l.add(s), e.isAnimationEnabled() && t.setClipPath(X(t.getBoundingRect(), e, function() {
            t.removeClipPath();
          }));
        } else {
          var s = G[T];
          t = s.childAt(0), l.add(s), d[p] = s, J(t, {
            shape: {
              points: y,
              stackedOnPoints: f
            }
          }, e), K(t);
        }
        F(t, H(e), {
          labelDataIndex: i[n - 1],
          defaultText: o.getName(i[n - 1]),
          inheritColor: g.fill
        }, {
          normal: {
            verticalAlign: "middle"
            // align: 'right'
          }
        }), t.setTextConfig({
          position: null,
          local: !0
        });
        var w = t.getTextContent();
        w && (w.x = S.x - V, w.y = S.y0 + S.y / 2), t.useStyle(g), o.setItemGraphicEl(p, t), z(t, e), B(t, b.get("focus"), b.get("blurScope"), b.get("disabled"));
      }
      this._layersSeries = v, this._layers = d;
    }, r.type = "themeRiver", r;
  })(q)
);
function X(a, r, e) {
  var h = new W({
    shape: {
      x: a.x - 10,
      y: a.y - 10,
      width: 0,
      height: a.height + 20
    }
  });
  return Q(h, {
    shape: {
      x: a.x - 50,
      width: a.width + 100,
      height: a.height + 20
    }
  }, r, e), h;
}
export {
  se as default
};
