// echarts@6.0.0/lib/chart/themeRiver/ThemeRiverSeries.js
import { __extends as A } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import S from "../../model/Series.js";
import x from "../../data/helper/createDimensions.js";
import { getDimensionTypeByAxis as M } from "../../data/helper/dimensionHelper.js";
import L from "../../data/SeriesData.js";
import { bind as y, filter as R, createHashMap as b, isArray as N } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { groupData as D, SINGLE_REFERRING as T } from "../../util/model.js";
import w from "../../visual/LegendVisualProvider.js";
import { createTooltipMarkup as E } from "../../component/tooltip/tooltipMarkup.js";
var h = 2, X = (
  /** @class */
  (function(g) {
    A(l, g);
    function l() {
      var e = g !== null && g.apply(this, arguments) || this;
      return e.type = l.type, e;
    }
    return l.prototype.init = function(e) {
      g.prototype.init.apply(this, arguments), this.legendVisualProvider = new w(y(this.getData, this), y(this.getRawData, this));
    }, l.prototype.fixData = function(e) {
      var m = e.length, n = {}, o = D(e, function(u) {
        return n.hasOwnProperty(u[0] + "") || (n[u[0] + ""] = -1), u[2];
      }), t = [];
      o.buckets.each(function(u, v) {
        t.push({
          name: v,
          dataList: u
        });
      });
      for (var r = t.length, i = 0; i < r; ++i) {
        for (var p = t[i].name, s = 0; s < t[i].dataList.length; ++s) {
          var a = t[i].dataList[s][0] + "";
          n[a] = i;
        }
        for (var a in n)
          n.hasOwnProperty(a) && n[a] !== i && (n[a] = i, e[m] = [a, 0, p], m++);
      }
      return e;
    }, l.prototype.getInitialData = function(e, m) {
      for (var n = this.getReferringComponents("singleAxis", T).models[0], o = n.get("type"), t = R(e.data, function(f) {
        return f[2] !== void 0;
      }), r = this.fixData(t || []), i = [], p = this.nameMap = b(), s = 0, a = 0; a < r.length; ++a)
        i.push(r[a][h]), p.get(r[a][h]) || (p.set(r[a][h], s), s++);
      var u = x(r, {
        coordDimensions: ["single"],
        dimensionsDefine: [{
          name: "time",
          type: M(o)
        }, {
          name: "value",
          type: "float"
        }, {
          name: "name",
          type: "ordinal"
        }],
        encodeDefine: {
          single: 0,
          value: 1,
          itemName: 2
        }
      }).dimensions, v = new L(u, this);
      return v.initData(r), v;
    }, l.prototype.getLayerSeries = function() {
      for (var e = this.getData(), m = e.count(), n = [], o = 0; o < m; ++o)
        n[o] = o;
      var t = e.mapDimension("single"), r = D(n, function(p) {
        return e.get("name", p);
      }), i = [];
      return r.buckets.each(function(p, s) {
        p.sort(function(a, u) {
          return e.get(t, a) - e.get(t, u);
        }), i.push({
          name: s,
          indices: p
        });
      }), i;
    }, l.prototype.getAxisTooltipData = function(e, m, n) {
      N(e) || (e = e ? [e] : []);
      for (var o = this.getData(), t = this.getLayerSeries(), r = [], i = t.length, p, s = 0; s < i; ++s) {
        for (var a = Number.MAX_VALUE, u = -1, v = t[s].indices.length, f = 0; f < v; ++f) {
          var c = o.get(e[0], t[s].indices[f]), d = Math.abs(c - m);
          d <= a && (p = c, a = d, u = t[s].indices[f]);
        }
        r.push(u);
      }
      return {
        dataIndices: r,
        nestestValue: p
      };
    }, l.prototype.formatTooltip = function(e, m, n) {
      var o = this.getData(), t = o.getName(e), r = o.get(o.mapDimension("value"), e);
      return E("nameValue", {
        name: t,
        value: r
      });
    }, l.type = "series.themeRiver", l.dependencies = ["singleAxis"], l.defaultOption = {
      // zlevel: 0,
      z: 2,
      colorBy: "data",
      coordinateSystem: "singleAxis",
      // gap in axis's orthogonal orientation
      boundaryGap: ["10%", "10%"],
      // legendHoverLink: true,
      singleAxisIndex: 0,
      animationEasing: "linear",
      label: {
        margin: 4,
        show: !0,
        position: "left",
        fontSize: 11
      },
      emphasis: {
        label: {
          show: !0
        }
      }
    }, l;
  })(S)
);
export {
  X as default
};
