// @radix-ui/react-tooltip@1.2.8/dist/index.js
import * as s from "react";
import { composeEventHandlers as x } from "../../../../../@radix-ui_primitive@1.1.3/node_modules/@radix-ui/primitive/dist/index.js";
import { useComposedRefs as N } from "../../../../../@radix-ui_react-compose-refs@1.1.2_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-compose-refs/dist/index.js";
import { createContextScope as K } from "../../../../../@radix-ui_react-context@1.1.2_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-context/dist/index.js";
import { DismissableLayer as W } from "../../../../../@radix-ui_react-dismissable-layer@1.1.11_@types_react-dom@18.3.1_@types_react@18.3.18_react-d_yogtd4jid4ravehalgybjmgjoa/node_modules/@radix-ui/react-dismissable-layer/dist/index.js";
import { useId as $ } from "../../../../../@radix-ui_react-id@1.1.1_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-id/dist/index.js";
import { createPopperScope as S, Root as z, Anchor as J, Content as Q, Arrow as Z } from "../../../../../@radix-ui_react-popper@1.2.8_@types_react-dom@18.3.1_@types_react@18.3.18_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/@radix-ui/react-popper/dist/index.js";
import { Portal as ee } from "../../../../../@radix-ui_react-portal@1.1.9_@types_react-dom@18.3.1_@types_react@18.3.18_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/@radix-ui/react-portal/dist/index.js";
import { Presence as G } from "../../../../../@radix-ui_react-presence@1.1.5_@types_react-dom@18.3.1_@types_react@18.3.18_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/@radix-ui/react-presence/dist/index.js";
import { Primitive as te } from "../../../../../@radix-ui_react-primitive@2.1.3_@types_react-dom@18.3.1_@types_react@18.3.18_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/@radix-ui/react-primitive/dist/index.js";
import { createSlottable as oe } from "../../../../../@radix-ui_react-slot@1.2.3_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-slot/dist/index.js";
import { useControllableState as re } from "../../../../../@radix-ui_react-use-controllable-state@1.2.2_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-use-controllable-state/dist/index.js";
import { Root as ne } from "../../../../../@radix-ui_react-visually-hidden@1.2.3_@types_react-dom@18.3.1_@types_react@18.3.18_react-dom@_jh4b6g5hyzelaffwdc575htfhy/node_modules/@radix-ui/react-visually-hidden/dist/index.js";
import { jsx as f, jsxs as se } from "react/jsx-runtime";
var [O] = K("Tooltip", [
  S
]), D = S(), j = "TooltipProvider", ie = 700, L = "tooltip.open", [ae, k] = O(j), F = (t) => {
  const {
    __scopeTooltip: o,
    delayDuration: e = ie,
    skipDelayDuration: r = 300,
    disableHoverableContent: n = !1,
    children: c
  } = t, a = s.useRef(!0), v = s.useRef(!1), i = s.useRef(0);
  return s.useEffect(() => {
    const p = i.current;
    return () => window.clearTimeout(p);
  }, []), /* @__PURE__ */ f(
    ae,
    {
      scope: o,
      isOpenDelayedRef: a,
      delayDuration: e,
      onOpen: s.useCallback(() => {
        window.clearTimeout(i.current), a.current = !1;
      }, []),
      onClose: s.useCallback(() => {
        window.clearTimeout(i.current), i.current = window.setTimeout(
          () => a.current = !0,
          r
        );
      }, [r]),
      isPointerInTransitRef: v,
      onPointerInTransitChange: s.useCallback((p) => {
        v.current = p;
      }, []),
      disableHoverableContent: n,
      children: c
    }
  );
};
F.displayName = j;
var R = "Tooltip", [ce, _] = O(R), B = (t) => {
  const {
    __scopeTooltip: o,
    children: e,
    open: r,
    defaultOpen: n,
    onOpenChange: c,
    disableHoverableContent: a,
    delayDuration: v
  } = t, i = k(R, t.__scopeTooltip), p = D(o), [l, d] = s.useState(null), h = $(), u = s.useRef(0), m = a ?? i.disableHoverableContent, C = v ?? i.delayDuration, T = s.useRef(!1), [g, y] = re({
    prop: r,
    defaultProp: n ?? !1,
    onChange: (H) => {
      H ? (i.onOpen(), document.dispatchEvent(new CustomEvent(L))) : i.onClose(), c?.(H);
    },
    caller: R
  }), w = s.useMemo(() => g ? T.current ? "delayed-open" : "instant-open" : "closed", [g]), P = s.useCallback(() => {
    window.clearTimeout(u.current), u.current = 0, T.current = !1, y(!0);
  }, [y]), E = s.useCallback(() => {
    window.clearTimeout(u.current), u.current = 0, y(!1);
  }, [y]), I = s.useCallback(() => {
    window.clearTimeout(u.current), u.current = window.setTimeout(() => {
      T.current = !0, y(!0), u.current = 0;
    }, C);
  }, [C, y]);
  return s.useEffect(() => () => {
    u.current && (window.clearTimeout(u.current), u.current = 0);
  }, []), /* @__PURE__ */ f(z, { ...p, children: /* @__PURE__ */ f(
    ce,
    {
      scope: o,
      contentId: h,
      open: g,
      stateAttribute: w,
      trigger: l,
      onTriggerChange: d,
      onTriggerEnter: s.useCallback(() => {
        i.isOpenDelayedRef.current ? I() : P();
      }, [i.isOpenDelayedRef, I, P]),
      onTriggerLeave: s.useCallback(() => {
        m ? E() : (window.clearTimeout(u.current), u.current = 0);
      }, [E, m]),
      onOpen: P,
      onClose: E,
      disableHoverableContent: m,
      children: e
    }
  ) });
};
B.displayName = R;
var A = "TooltipTrigger", U = s.forwardRef(
  (t, o) => {
    const { __scopeTooltip: e, ...r } = t, n = _(A, e), c = k(A, e), a = D(e), v = s.useRef(null), i = N(o, v, n.onTriggerChange), p = s.useRef(!1), l = s.useRef(!1), d = s.useCallback(() => p.current = !1, []);
    return s.useEffect(() => () => document.removeEventListener("pointerup", d), [d]), /* @__PURE__ */ f(J, { asChild: !0, ...a, children: /* @__PURE__ */ f(
      te.button,
      {
        "aria-describedby": n.open ? n.contentId : void 0,
        "data-state": n.stateAttribute,
        ...r,
        ref: i,
        onPointerMove: x(t.onPointerMove, (h) => {
          h.pointerType !== "touch" && !l.current && !c.isPointerInTransitRef.current && (n.onTriggerEnter(), l.current = !0);
        }),
        onPointerLeave: x(t.onPointerLeave, () => {
          n.onTriggerLeave(), l.current = !1;
        }),
        onPointerDown: x(t.onPointerDown, () => {
          n.open && n.onClose(), p.current = !0, document.addEventListener("pointerup", d, { once: !0 });
        }),
        onFocus: x(t.onFocus, () => {
          p.current || n.onOpen();
        }),
        onBlur: x(t.onBlur, n.onClose),
        onClick: x(t.onClick, n.onClose)
      }
    ) });
  }
);
U.displayName = A;
var M = "TooltipPortal", [le, ue] = O(M, {
  forceMount: void 0
}), V = (t) => {
  const { __scopeTooltip: o, forceMount: e, children: r, container: n } = t, c = _(M, o);
  return /* @__PURE__ */ f(le, { scope: o, forceMount: e, children: /* @__PURE__ */ f(G, { present: e || c.open, children: /* @__PURE__ */ f(ee, { asChild: !0, container: n, children: r }) }) });
};
V.displayName = M;
var b = "TooltipContent", Y = s.forwardRef(
  (t, o) => {
    const e = ue(b, t.__scopeTooltip), { forceMount: r = e.forceMount, side: n = "top", ...c } = t, a = _(b, t.__scopeTooltip);
    return /* @__PURE__ */ f(G, { present: r || a.open, children: a.disableHoverableContent ? /* @__PURE__ */ f(q, { side: n, ...c, ref: o }) : /* @__PURE__ */ f(pe, { side: n, ...c, ref: o }) });
  }
), pe = s.forwardRef((t, o) => {
  const e = _(b, t.__scopeTooltip), r = k(b, t.__scopeTooltip), n = s.useRef(null), c = N(o, n), [a, v] = s.useState(null), { trigger: i, onClose: p } = e, l = n.current, { onPointerInTransitChange: d } = r, h = s.useCallback(() => {
    v(null), d(!1);
  }, [d]), u = s.useCallback(
    (m, C) => {
      const T = m.currentTarget, g = { x: m.clientX, y: m.clientY }, y = me(g, T.getBoundingClientRect()), w = Te(g, y), P = Ce(C.getBoundingClientRect()), E = ge([...w, ...P]);
      v(E), d(!0);
    },
    [d]
  );
  return s.useEffect(() => () => h(), [h]), s.useEffect(() => {
    if (i && l) {
      const m = (T) => u(T, l), C = (T) => u(T, i);
      return i.addEventListener("pointerleave", m), l.addEventListener("pointerleave", C), () => {
        i.removeEventListener("pointerleave", m), l.removeEventListener("pointerleave", C);
      };
    }
  }, [i, l, u, h]), s.useEffect(() => {
    if (a) {
      const m = (C) => {
        const T = C.target, g = { x: C.clientX, y: C.clientY }, y = i?.contains(T) || l?.contains(T), w = !ye(g, a);
        y ? h() : w && (h(), p());
      };
      return document.addEventListener("pointermove", m), () => document.removeEventListener("pointermove", m);
    }
  }, [i, l, a, p, h]), /* @__PURE__ */ f(q, { ...t, ref: c });
}), [de, fe] = O(R, { isInside: !1 }), ve = oe("TooltipContent"), q = s.forwardRef(
  (t, o) => {
    const {
      __scopeTooltip: e,
      children: r,
      "aria-label": n,
      onEscapeKeyDown: c,
      onPointerDownOutside: a,
      ...v
    } = t, i = _(b, e), p = D(e), { onClose: l } = i;
    return s.useEffect(() => (document.addEventListener(L, l), () => document.removeEventListener(L, l)), [l]), s.useEffect(() => {
      if (i.trigger) {
        const d = (h) => {
          h.target?.contains(i.trigger) && l();
        };
        return window.addEventListener("scroll", d, { capture: !0 }), () => window.removeEventListener("scroll", d, { capture: !0 });
      }
    }, [i.trigger, l]), /* @__PURE__ */ f(
      W,
      {
        asChild: !0,
        disableOutsidePointerEvents: !1,
        onEscapeKeyDown: c,
        onPointerDownOutside: a,
        onFocusOutside: (d) => d.preventDefault(),
        onDismiss: l,
        children: /* @__PURE__ */ se(
          Q,
          {
            "data-state": i.stateAttribute,
            ...p,
            ...v,
            ref: o,
            style: {
              ...v.style,
              "--radix-tooltip-content-transform-origin": "var(--radix-popper-transform-origin)",
              "--radix-tooltip-content-available-width": "var(--radix-popper-available-width)",
              "--radix-tooltip-content-available-height": "var(--radix-popper-available-height)",
              "--radix-tooltip-trigger-width": "var(--radix-popper-anchor-width)",
              "--radix-tooltip-trigger-height": "var(--radix-popper-anchor-height)"
            },
            children: [
              /* @__PURE__ */ f(ve, { children: r }),
              /* @__PURE__ */ f(de, { scope: e, isInside: !0, children: /* @__PURE__ */ f(ne, { id: i.contentId, role: "tooltip", children: n || r }) })
            ]
          }
        )
      }
    );
  }
);
Y.displayName = b;
var X = "TooltipArrow", he = s.forwardRef(
  (t, o) => {
    const { __scopeTooltip: e, ...r } = t, n = D(e);
    return fe(
      X,
      e
    ).isInside ? null : /* @__PURE__ */ f(Z, { ...n, ...r, ref: o });
  }
);
he.displayName = X;
function me(t, o) {
  const e = Math.abs(o.top - t.y), r = Math.abs(o.bottom - t.y), n = Math.abs(o.right - t.x), c = Math.abs(o.left - t.x);
  switch (Math.min(e, r, n, c)) {
    case c:
      return "left";
    case n:
      return "right";
    case e:
      return "top";
    case r:
      return "bottom";
    default:
      throw new Error("unreachable");
  }
}
function Te(t, o, e = 5) {
  const r = [];
  switch (o) {
    case "top":
      r.push(
        { x: t.x - e, y: t.y + e },
        { x: t.x + e, y: t.y + e }
      );
      break;
    case "bottom":
      r.push(
        { x: t.x - e, y: t.y - e },
        { x: t.x + e, y: t.y - e }
      );
      break;
    case "left":
      r.push(
        { x: t.x + e, y: t.y - e },
        { x: t.x + e, y: t.y + e }
      );
      break;
    case "right":
      r.push(
        { x: t.x - e, y: t.y - e },
        { x: t.x - e, y: t.y + e }
      );
      break;
  }
  return r;
}
function Ce(t) {
  const { top: o, right: e, bottom: r, left: n } = t;
  return [
    { x: n, y: o },
    { x: e, y: o },
    { x: e, y: r },
    { x: n, y: r }
  ];
}
function ye(t, o) {
  const { x: e, y: r } = t;
  let n = !1;
  for (let c = 0, a = o.length - 1; c < o.length; a = c++) {
    const v = o[c], i = o[a], p = v.x, l = v.y, d = i.x, h = i.y;
    l > r != h > r && e < (d - p) * (r - l) / (h - l) + p && (n = !n);
  }
  return n;
}
function ge(t) {
  const o = t.slice();
  return o.sort((e, r) => e.x < r.x ? -1 : e.x > r.x ? 1 : e.y < r.y ? -1 : e.y > r.y ? 1 : 0), xe(o);
}
function xe(t) {
  if (t.length <= 1) return t.slice();
  const o = [];
  for (let r = 0; r < t.length; r++) {
    const n = t[r];
    for (; o.length >= 2; ) {
      const c = o[o.length - 1], a = o[o.length - 2];
      if ((c.x - a.x) * (n.y - a.y) >= (c.y - a.y) * (n.x - a.x)) o.pop();
      else break;
    }
    o.push(n);
  }
  o.pop();
  const e = [];
  for (let r = t.length - 1; r >= 0; r--) {
    const n = t[r];
    for (; e.length >= 2; ) {
      const c = e[e.length - 1], a = e[e.length - 2];
      if ((c.x - a.x) * (n.y - a.y) >= (c.y - a.y) * (n.x - a.x)) e.pop();
      else break;
    }
    e.push(n);
  }
  return e.pop(), o.length === 1 && e.length === 1 && o[0].x === e[0].x && o[0].y === e[0].y ? o : o.concat(e);
}
var He = F, Ne = B, Se = U, Ge = V, je = Y;
export {
  je as Content,
  Ge as Portal,
  He as Provider,
  Ne as Root,
  B as Tooltip,
  he as TooltipArrow,
  Y as TooltipContent,
  V as TooltipPortal,
  F as TooltipProvider,
  U as TooltipTrigger,
  Se as Trigger
};
