// echarts@6.0.0/lib/chart/candlestick/candlestickLayout.js
import { subPixelOptimize as M } from "../../util/graphic.js";
import U from "../helper/createRenderPlanner.js";
import { parsePercent as O } from "../../util/number.js";
import { map as X, retrieve2 as F } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { createFloat32Array as Y } from "../../util/vendor.js";
var rt = {
  seriesType: "candlestick",
  plan: U(),
  reset: function(v) {
    var h = v.coordinateSystem, e = v.getData(), a = Z(v, e), t = 0, u = 1, d = ["x", "y"], y = e.getDimensionIndex(e.mapDimension(d[t])), l = X(e.mapDimensionsAll(d[u]), e.getDimensionIndex, e), R = l[0], P = l[1], z = l[2], E = l[3];
    if (e.setLayout({
      candleWidth: a,
      // The value is experimented visually.
      isSimpleBox: a <= 1.3
    }), y < 0 || l.length < 4)
      return;
    return {
      progress: v.pipelineContext.large ? G : w
    };
    function w(S, b) {
      for (var r, i = b.getStore(); (r = S.next()) != null; ) {
        var n = i.get(y, r), g = i.get(R, r), f = i.get(P, r), m = i.get(z, r), s = i.get(E, r), k = Math.min(g, f), B = Math.max(g, f), W = I(k, n), D = I(B, n), C = I(m, n), A = I(s, n), T = [];
        L(T, D, 0), L(T, W, 1), T.push(j(A), j(D), j(C), j(W));
        var J = b.getItemModel(r), K = !!J.get(["itemStyle", "borderColorDoji"]);
        b.setItemLayout(r, {
          sign: q(i, r, g, f, P, K),
          initBaseline: g > f ? D[u] : W[u],
          ends: T,
          brushRect: Q(m, s, n)
        });
      }
      function I(o, N) {
        var p = [];
        return p[t] = N, p[u] = o, isNaN(N) || isNaN(o) ? [NaN, NaN] : h.dataToPoint(p);
      }
      function L(o, N, p) {
        var c = N.slice(), x = N.slice();
        c[t] = M(c[t] + a / 2, 1, !1), x[t] = M(x[t] - a / 2, 1, !0), p ? o.push(c, x) : o.push(x, c);
      }
      function Q(o, N, p) {
        var c = I(o, p), x = I(N, p);
        return c[t] -= a / 2, x[t] -= a / 2, {
          x: c[0],
          y: c[1],
          width: a,
          height: x[1] - c[1]
        };
      }
      function j(o) {
        return o[t] = M(o[t], 1), o;
      }
    }
    function G(S, b) {
      for (var r = Y(S.count * 4), i = 0, n, g = [], f = [], m, s = b.getStore(), k = !!v.get(["itemStyle", "borderColorDoji"]); (m = S.next()) != null; ) {
        var B = s.get(y, m), W = s.get(R, m), D = s.get(P, m), C = s.get(z, m), A = s.get(E, m);
        if (isNaN(B) || isNaN(C) || isNaN(A)) {
          r[i++] = NaN, i += 3;
          continue;
        }
        r[i++] = q(s, m, W, D, P, k), g[t] = B, g[u] = C, n = h.dataToPoint(g, null, f), r[i++] = n ? n[0] : NaN, r[i++] = n ? n[1] : NaN, g[u] = A, n = h.dataToPoint(g, null, f), r[i++] = n ? n[1] : NaN;
      }
      b.setLayout("largePoints", r);
    }
  }
};
function q(v, h, e, a, t, u) {
  var d;
  return e > a ? d = -1 : e < a ? d = 1 : d = u ? 0 : h > 0 ? v.get(t, h - 1) <= a ? 1 : -1 : 1, d;
}
function Z(v, h) {
  var e = v.getBaseAxis(), a, t = e.type === "category" ? e.getBandWidth() : (a = e.getExtent(), Math.abs(a[1] - a[0]) / h.count()), u = O(F(v.get("barMaxWidth"), t), t), d = O(F(v.get("barMinWidth"), 1), t), y = v.get("barWidth");
  return y != null ? O(y, t) : Math.max(Math.min(t / 2, u), d);
}
export {
  rt as default
};
