// echarts@6.0.0/lib/chart/candlestick/CandlestickView.js
import { __extends as _ } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { each as L, map as D } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import N from "../../view/Chart.js";
import { traverseElements as T } from "../../util/graphic.js";
import { setStatesStylesFromModel as w, toggleHoverEmphasis as E } from "../../util/states.js";
import S from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Path.js";
import { createClipPath as k } from "../helper/createClipPathFromCoordSys.js";
import { initProps as b, updateProps as R, saveOldStyle as V } from "../../animation/basicTransition.js";
import { getColor as B, getBorderColor as C } from "./candlestickVisual.js";
var A = ["color", "borderColor"], U = (
  /** @class */
  (function(o) {
    _(a, o);
    function a() {
      var e = o !== null && o.apply(this, arguments) || this;
      return e.type = a.type, e;
    }
    return a.prototype.render = function(e, t, r) {
      this.group.removeClipPath(), this._progressiveEls = null, this._updateDrawMode(e), this._isLargeDraw ? this._renderLarge(e) : this._renderNormal(e);
    }, a.prototype.incrementalPrepareRender = function(e, t, r) {
      this._clear(), this._updateDrawMode(e);
    }, a.prototype.incrementalRender = function(e, t, r, i) {
      this._progressiveEls = [], this._isLargeDraw ? this._incrementalRenderLarge(e, t) : this._incrementalRenderNormal(e, t);
    }, a.prototype.eachRendered = function(e) {
      T(this._progressiveEls || this.group, e);
    }, a.prototype._updateDrawMode = function(e) {
      var t = e.pipelineContext.large;
      (this._isLargeDraw == null || t !== this._isLargeDraw) && (this._isLargeDraw = t, this._clear());
    }, a.prototype._renderNormal = function(e) {
      var t = e.getData(), r = this._data, i = this.group, n = t.getLayout("isSimpleBox"), p = e.get("clip", !0), u = e.coordinateSystem, m = u.getArea && u.getArea();
      this._data || i.removeAll(), t.diff(r).add(function(l) {
        if (t.hasValue(l)) {
          var h = t.getItemLayout(l);
          if (p && y(m, h))
            return;
          var s = g(h, l, !0);
          b(s, {
            shape: {
              points: h.ends
            }
          }, e, l), f(s, t, l, n), i.add(s), t.setItemGraphicEl(l, s);
        }
      }).update(function(l, h) {
        var s = r.getItemGraphicEl(h);
        if (!t.hasValue(l)) {
          i.remove(s);
          return;
        }
        var c = t.getItemLayout(l);
        if (p && y(m, c)) {
          i.remove(s);
          return;
        }
        s ? (R(s, {
          shape: {
            points: c.ends
          }
        }, e, l), V(s)) : s = g(c), f(s, t, l, n), i.add(s), t.setItemGraphicEl(l, s);
      }).remove(function(l) {
        var h = r.getItemGraphicEl(l);
        h && i.remove(h);
      }).execute(), this._data = t;
    }, a.prototype._renderLarge = function(e) {
      this._clear(), P(e, this.group);
      var t = e.get("clip", !0) ? k(e.coordinateSystem, !1, e) : null;
      t ? this.group.setClipPath(t) : this.group.removeClipPath();
    }, a.prototype._incrementalRenderNormal = function(e, t) {
      for (var r = t.getData(), i = r.getLayout("isSimpleBox"), n; (n = e.next()) != null; ) {
        var p = r.getItemLayout(n), u = g(p);
        f(u, r, n, i), u.incremental = !0, this.group.add(u), this._progressiveEls.push(u);
      }
    }, a.prototype._incrementalRenderLarge = function(e, t) {
      P(t, this.group, this._progressiveEls, !0);
    }, a.prototype.remove = function(e) {
      this._clear();
    }, a.prototype._clear = function() {
      this.group.removeAll(), this._data = null;
    }, a.type = "candlestick", a;
  })(N)
), G = (
  /** @class */
  /* @__PURE__ */ (function() {
    function o() {
    }
    return o;
  })()
), I = (
  /** @class */
  (function(o) {
    _(a, o);
    function a(e) {
      var t = o.call(this, e) || this;
      return t.type = "normalCandlestickBox", t;
    }
    return a.prototype.getDefaultShape = function() {
      return new G();
    }, a.prototype.buildPath = function(e, t) {
      var r = t.points;
      this.__simpleBox ? (e.moveTo(r[4][0], r[4][1]), e.lineTo(r[6][0], r[6][1])) : (e.moveTo(r[0][0], r[0][1]), e.lineTo(r[1][0], r[1][1]), e.lineTo(r[2][0], r[2][1]), e.lineTo(r[3][0], r[3][1]), e.closePath(), e.moveTo(r[4][0], r[4][1]), e.lineTo(r[5][0], r[5][1]), e.moveTo(r[6][0], r[6][1]), e.lineTo(r[7][0], r[7][1]));
    }, a;
  })(S)
);
function g(o, a, e) {
  var t = o.ends;
  return new I({
    shape: {
      points: e ? x(t, o) : t
    },
    z2: 100
  });
}
function y(o, a) {
  for (var e = !0, t = 0; t < a.ends.length; t++)
    if (o.contain(a.ends[t][0], a.ends[t][1])) {
      e = !1;
      break;
    }
  return e;
}
function f(o, a, e, t) {
  var r = a.getItemModel(e);
  o.useStyle(a.getItemVisual(e, "style")), o.style.strokeNoScale = !0, o.__simpleBox = t, w(o, r);
  var i = a.getItemLayout(e).sign;
  L(o.states, function(p, u) {
    var m = r.getModel(u), l = B(i, m), h = C(i, m) || l, s = p.style || (p.style = {});
    l && (s.fill = l), h && (s.stroke = h);
  });
  var n = r.getModel("emphasis");
  E(o, n.get("focus"), n.get("blurScope"), n.get("disabled"));
}
function x(o, a) {
  return D(o, function(e) {
    return e = e.slice(), e[1] = a.initBaseline, e;
  });
}
var O = (
  /** @class */
  /* @__PURE__ */ (function() {
    function o() {
    }
    return o;
  })()
), v = (
  /** @class */
  (function(o) {
    _(a, o);
    function a(e) {
      var t = o.call(this, e) || this;
      return t.type = "largeCandlestickBox", t;
    }
    return a.prototype.getDefaultShape = function() {
      return new O();
    }, a.prototype.buildPath = function(e, t) {
      for (var r = t.points, i = 0; i < r.length; )
        if (this.__sign === r[i++]) {
          var n = r[i++];
          e.moveTo(n, r[i++]), e.lineTo(n, r[i++]);
        } else
          i += 3;
    }, a;
  })(S)
);
function P(o, a, e, t) {
  var r = o.getData(), i = r.getLayout("largePoints"), n = new v({
    shape: {
      points: i
    },
    __sign: 1,
    ignoreCoarsePointer: !0
  });
  a.add(n);
  var p = new v({
    shape: {
      points: i
    },
    __sign: -1,
    ignoreCoarsePointer: !0
  });
  a.add(p);
  var u = new v({
    shape: {
      points: i
    },
    __sign: 0,
    ignoreCoarsePointer: !0
  });
  a.add(u), d(1, n, o), d(-1, p, o), d(0, u, o), t && (n.incremental = !0, p.incremental = !0), e && e.push(n, p);
}
function d(o, a, e, t) {
  var r = C(o, e) || B(o, e), i = e.getModel("itemStyle").getItemStyle(A);
  a.useStyle(i), a.style.fill = null, a.style.stroke = r;
}
export {
  U as default
};
