// echarts@6.0.0/lib/chart/candlestick/candlestickVisual.js
import d from "../helper/createRenderPlanner.js";
import { extend as s } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
var m = ["itemStyle", "borderColor"], c = ["itemStyle", "borderColor0"], f = ["itemStyle", "borderColorDoji"], g = ["itemStyle", "color"], p = ["itemStyle", "color0"];
function y(e, r) {
  return r.get(e > 0 ? g : p);
}
function C(e, r) {
  return r.get(e === 0 ? f : e > 0 ? m : c);
}
var I = {
  seriesType: "candlestick",
  plan: d(),
  // For legend.
  performRawSeries: !0,
  reset: function(e, r) {
    if (!r.isSeriesFiltered(e)) {
      var a = e.pipelineContext.large;
      return !a && {
        progress: function(u, i) {
          for (var o; (o = u.next()) != null; ) {
            var l = i.getItemModel(o), n = i.getItemLayout(o).sign, t = l.getItemStyle();
            t.fill = y(n, l), t.stroke = C(n, l) || t.fill;
            var v = i.ensureUniqueItemVisual(o, "style");
            s(v, t);
          }
        }
      };
    }
  }
};
export {
  I as default,
  C as getBorderColor,
  y as getColor
};
