// echarts@6.0.0/lib/core/CoordinateSystem.js
import { each as S, assert as f, createHashMap as g } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { SINGLE_REFERRING as N } from "../util/model.js";
import { error as l } from "../util/log.js";
var u = {}, p = {}, D = (
  /** @class */
  (function() {
    function o() {
      this._normalMasterList = [], this._nonSeriesBoxMasterList = [];
    }
    return o.prototype.create = function(e, r) {
      this._nonSeriesBoxMasterList = t(u, !0), this._normalMasterList = t(p, !1);
      function t(i, a) {
        var n = [];
        return S(i, function(d, y) {
          var c = d.create(e, r);
          n = n.concat(c || []), process.env.NODE_ENV !== "production" && a && S(c, function(v) {
            return f(!v.update);
          });
        }), n;
      }
    }, o.prototype.update = function(e, r) {
      S(this._normalMasterList, function(t) {
        t.update && t.update(e, r);
      });
    }, o.prototype.getCoordinateSystems = function() {
      return this._normalMasterList.concat(this._nonSeriesBoxMasterList);
    }, o.register = function(e, r) {
      if (e === "matrix" || e === "calendar") {
        u[e] = r;
        return;
      }
      p[e] = r;
    }, o.get = function(e) {
      return p[e] || u[e];
    }, o;
  })()
);
function x(o) {
  return !!u[o];
}
var m = {
  // By default fetch coord from `model.get('coord')`.
  coord: 1,
  // Some model/series, such as pie, is allowed to also get coord from `model.get('center')`,
  // if cannot get from `model.get('coord')`. But historically pie use `center` option, but
  // geo use `layoutCenter` option to specify layout center; they are not able to be unified.
  // Therefor it is not recommended.
  coord2: 2
};
function O(o) {
  process.env.NODE_ENV !== "production" && f(!C.get(o.fullType)), C.set(o.fullType, {
    getCoord2: void 0
  }).getCoord2 = o.getCoord2;
}
var C = g();
function T(o) {
  var e = o.getShallow("coord", !0), r = m.coord;
  if (e == null) {
    var t = C.get(o.type);
    t && t.getCoord2 && (r = m.coord2, e = t.getCoord2(o));
  }
  return {
    coord: e,
    from: r
  };
}
var s = {
  none: 0,
  dataCoordSys: 1,
  boxCoordSys: 2
};
function E(o, e) {
  var r = o.getShallow("coordinateSystem"), t = o.getShallow("coordinateSystemUsage", !0), i = t != null, a = s.none;
  if (r) {
    var n = o.mainType === "series";
    t == null && (t = n ? "data" : "box"), t === "data" ? (a = s.dataCoordSys, n || (process.env.NODE_ENV !== "production" && i && e && l('coordinateSystemUsage "data" is not supported in non-series components.'), a = s.none)) : t === "box" && (a = s.boxCoordSys, !n && !x(r) && (process.env.NODE_ENV !== "production" && i && e && l('coordinateSystem "' + r + '" cannot be used' + (' as coordinateSystemUsage "box" for "' + o.type + '" yet.')), a = s.none));
  }
  return {
    coordSysType: r,
    kind: a
  };
}
function B(o) {
  var e = o.targetModel, r = o.coordSysType, t = o.coordSysProvider, i = o.isDefaultDataCoordSys, a = o.allowNotFound;
  process.env.NODE_ENV !== "production" && f(!!r);
  var n = E(e, !0), d = n.kind, y = n.coordSysType;
  if (i && d !== s.dataCoordSys && (d = s.dataCoordSys, y = r), d === s.none || y !== r)
    return !1;
  var c = t(r, e);
  return c ? (d === s.dataCoordSys ? (process.env.NODE_ENV !== "production" && f(e.mainType === "series"), e.coordinateSystem = c) : e.boxCoordinateSystem = c, !0) : (process.env.NODE_ENV !== "production" && (a || l(r + " cannot be found for" + (" " + e.type + " (index: " + e.componentIndex + ")."))), !1);
}
var L = function(o, e) {
  var r = e.getReferringComponents(o, N).models[0];
  return r && r.coordinateSystem;
};
export {
  m as BoxCoordinateSystemCoordFrom,
  s as CoordinateSystemUsageKind,
  E as decideCoordSysUsageKind,
  D as default,
  T as getCoordForBoxCoordSys,
  B as injectCoordSysByOption,
  O as registerLayOutOnCoordSysUsage,
  L as simpleCoordSysInjectionProvider
};
