// echarts@6.0.0/lib/core/echarts.js
import { __extends as ye } from "../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { init as be, registerSSRDataGetter as Vt } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/zrender.js";
import { retrieve2 as X, bind as ze, setAsPrimitive as It, isObject as B, isString as Nt, clone as Fe, each as C, extend as z, assert as G, createHashMap as de, map as Mt, defaults as He, indexOf as ut, isDom as Lt, isFunction as ft, noop as j } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import q from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/env.js";
import pe from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/timsort.js";
import Pe from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/Eventful.js";
import xt from "../model/Global.js";
import Ut from "./ExtensionAPI.js";
import dt from "./CoordinateSystem.js";
import kt from "../model/OptionManager.js";
import pt from "../preprocessor/backwardCompat.js";
import bt from "../processor/dataStack.js";
import Ye from "../model/Series.js";
import zt from "../view/Component.js";
import le from "../view/Chart.js";
import { retrieveZInfo as Ft, traverseUpdateZ as Ht } from "../util/graphic.js";
import { getECData as J } from "../util/innerStore.js";
import { isSelectChangePayload as Be, isHighDownPayload as Ge, allLeaveBlur as Yt, updateSeriesElementSelection as Ze, HOVER_STATE_EMPHASIS as Bt, HOVER_STATE_BLUR as Gt, savePathStates as Zt, enterEmphasis as We, leaveEmphasis as Xe, enterBlur as Wt, leaveBlur as Xt, enterSelect as qt, leaveSelect as Kt, isHighDownDispatcher as qe, handleGlobalMouseOverForHighDown as jt, handleGlobalMouseOutForHighDown as $t, HIGHLIGHT_ACTION_TYPE as Z, DOWNPLAY_ACTION_TYPE as ce, SELECT_ACTION_TYPE as Ke, SELECT_CHANGED_EVENT_TYPE as Ae, UNSELECT_ACTION_TYPE as je, TOGGLE_SELECT_ACTION_TYPE as $e, getAllSelectedIndices as Jt, blurSeriesFromHighlightPayload as Qt, findComponentHighDownDispatchers as er, blurComponent as tr, toggleSelectionFromPayload as rr } from "../util/states.js";
import { getAttribute as ir, parseFinder as he, setAttribute as lt, preParseFinder as nr, convertOptionIdName as ar, normalizeToArray as or } from "../util/model.js";
import { throttle as sr } from "../util/throttle.js";
import { seriesStyleTask as ur, dataStyleTask as fr, dataColorPaletteTask as dr } from "../visual/style.js";
import pr from "../loading/default.js";
import ct from "./Scheduler.js";
import lr from "../theme/dark.js";
import { parseClassType as Je } from "../util/clazz.js";
import { ECEventProcessor as cr } from "../util/ECEventProcessor.js";
import { seriesSymbolTask as hr, dataSymbolTask as vr } from "../visual/symbol.js";
import { getItemVisualFromData as _r, getVisualFromData as gr } from "../visual/helper.js";
import { deprecateLog as mr, error as ne, deprecateReplaceLog as Er, warn as x } from "../util/log.js";
import { handleLegacySelectEvents as Tr } from "../legacy/dataSelectAction.js";
import { registerExternalTransform as Sr } from "../data/helper/transform.js";
import { createLocaleObject as yr, SYSTEM_LANG as Cr } from "./locale.js";
import { registerLocale as Wi } from "./locale.js";
import { findEventDispatcher as Q } from "../util/event.js";
import Rr from "../visual/decal.js";
import N from "./lifecycle.js";
import { platformApi as Or } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/platform.js";
import { getImpl as wr } from "./impl.js";
import Pr from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Rect.js";
import Ar from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Path.js";
import Dr from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Image.js";
import { isElementRemoved as ve } from "../animation/basicTransition.js";
var Vr = 1, Ir = 800, Nr = 900, Mr = 1e3, Lr = 2e3, xr = 5e3, ht = 1e3, Ur = 1100, De = 2e3, vt = 3e3, kr = 4e3, ue = 4500, br = 4600, zr = 5e3, Fr = 6e3, _t = 7e3, Ui = {
  PROCESSOR: {
    FILTER: Mr,
    SERIES_FILTER: Ir,
    STATISTIC: xr
  },
  VISUAL: {
    LAYOUT: ht,
    PROGRESSIVE_LAYOUT: Ur,
    GLOBAL: De,
    CHART: vt,
    POST_CHART_LAYOUT: br,
    COMPONENT: kr,
    BRUSH: zr,
    CHART_ITEM: ue,
    ARIA: Fr,
    DECAL: _t
  }
}, R = "__flagInMainProcess", ee = "__mainProcessVersion", O = "__pendingUpdate", _e = "__needsUpdateStatus", Qe = /^[a-zA-Z0-9_]+$/, ge = "__connectUpdateStatus", et = 0, Hr = 1, Yr = 2;
function gt(c) {
  return function() {
    for (var u = [], r = 0; r < arguments.length; r++)
      u[r] = arguments[r];
    if (this.isDisposed()) {
      D(this.id);
      return;
    }
    return Et(this, c, u);
  };
}
function mt(c) {
  return function() {
    for (var u = [], r = 0; r < arguments.length; r++)
      u[r] = arguments[r];
    return Et(this, c, u);
  };
}
function Et(c, u, r) {
  return r[0] = r[0] && r[0].toLowerCase(), Pe.prototype[u].apply(c, r);
}
var Tt = (
  /** @class */
  (function(c) {
    ye(u, c);
    function u() {
      return c !== null && c.apply(this, arguments) || this;
    }
    return u;
  })(Pe)
), St = Tt.prototype;
St.on = mt("on");
St.off = mt("off");
var b, me, te, L, re, Ee, Te, F, H, tt, rt, Se, it, ie, nt, yt, I, at, Y, Ct = (
  /** @class */
  (function(c) {
    ye(u, c);
    function u(r, n, f) {
      var o = c.call(this, new cr()) || this;
      o._chartsViews = [], o._chartsMap = {}, o._componentsViews = [], o._componentsMap = {}, o._pendingActions = [], f = f || {}, o._dom = r;
      var g = "canvas", S = "auto", E = !1;
      if (o[ee] = 1, process.env.NODE_ENV !== "production") {
        var m = (
          /* eslint-disable-next-line */
          q.hasGlobalWindow ? window : global
        );
        m && (g = X(m.__ECHARTS__DEFAULT__RENDERER__, g), S = X(m.__ECHARTS__DEFAULT__COARSE_POINTER, S), E = X(m.__ECHARTS__DEFAULT__USE_DIRTY_RECT__, E));
      }
      f.ssr && Vt(function(t) {
        var a = J(t), s = a.dataIndex;
        if (s != null) {
          var p = de();
          return p.set("series_index", a.seriesIndex), p.set("data_index", s), a.ssrType && p.set("ssr_type", a.ssrType), p;
        }
      });
      var y = o._zr = be(r, {
        renderer: f.renderer || g,
        devicePixelRatio: f.devicePixelRatio,
        width: f.width,
        height: f.height,
        ssr: f.ssr,
        useDirtyRect: X(f.useDirtyRect, E),
        useCoarsePointer: X(f.useCoarsePointer, S),
        pointerSize: f.pointerSize
      });
      o._ssr = f.ssr, o._throttledZrFlush = sr(ze(y.flush, y), 17), o._updateTheme(n), o._locale = yr(f.locale || Cr), o._coordSysMgr = new dt();
      var e = o._api = nt(o);
      function i(t, a) {
        return t.__prio - a.__prio;
      }
      return pe(se, i), pe(Re, i), o._scheduler = new ct(o, e, Re, se), o._messageCenter = new Tt(), o._initEvents(), o.resize = ze(o.resize, o), y.animation.on("frame", o._onframe, o), tt(y, o), rt(y, o), It(o), o;
    }
    return u.prototype._onframe = function() {
      if (!this._disposed) {
        at(this);
        var r = this._scheduler;
        if (this[O]) {
          var n = this[O].silent;
          this[R] = !0, Y(this);
          try {
            b(this), L.update.call(this, null, this[O].updateParams);
          } catch (E) {
            throw this[R] = !1, this[O] = null, E;
          }
          this._zr.flush(), this[R] = !1, this[O] = null, F.call(this, n), H.call(this, n);
        } else if (r.unfinished) {
          var f = Vr, o = this._model, g = this._api;
          r.unfinished = !1;
          do {
            var S = +/* @__PURE__ */ new Date();
            r.performSeriesTasks(o), r.performDataProcessorTasks(o), Ee(this, o), r.performVisualTasks(o), ie(this, this._model, g, "remain", {}), f -= +/* @__PURE__ */ new Date() - S;
          } while (f > 0 && r.unfinished);
          r.unfinished || this._zr.flush();
        }
      }
    }, u.prototype.getDom = function() {
      return this._dom;
    }, u.prototype.getId = function() {
      return this.id;
    }, u.prototype.getZr = function() {
      return this._zr;
    }, u.prototype.isSSR = function() {
      return this._ssr;
    }, u.prototype.setOption = function(r, n, f) {
      if (this[R]) {
        process.env.NODE_ENV !== "production" && ne("`setOption` should not be called during main process.");
        return;
      }
      if (this._disposed) {
        D(this.id);
        return;
      }
      var o, g, S;
      if (B(n) && (f = n.lazyUpdate, o = n.silent, g = n.replaceMerge, S = n.transition, n = n.notMerge), this[R] = !0, Y(this), !this._model || n) {
        var E = new kt(this._api), m = this._theme, y = this._model = new xt();
        y.scheduler = this._scheduler, y.ssr = this._ssr, y.init(null, null, null, m, this._locale, E);
      }
      this._model.setOption(r, {
        replaceMerge: g
      }, Oe);
      var e = {
        seriesTransition: S,
        optionChanged: !0
      };
      if (f)
        this[O] = {
          silent: o,
          updateParams: e
        }, this[R] = !1, this.getZr().wakeUp();
      else {
        try {
          b(this), L.update.call(this, null, e);
        } catch (i) {
          throw this[O] = null, this[R] = !1, i;
        }
        this._ssr || this._zr.flush(), this[O] = null, this[R] = !1, F.call(this, o), H.call(this, o);
      }
    }, u.prototype.setTheme = function(r, n) {
      if (this[R]) {
        process.env.NODE_ENV !== "production" && ne("`setTheme` should not be called during main process.");
        return;
      }
      if (this._disposed) {
        D(this.id);
        return;
      }
      var f = this._model;
      if (f) {
        var o = n && n.silent, g = null;
        this[O] && (o == null && (o = this[O].silent), g = this[O].updateParams, this[O] = null), this[R] = !0, Y(this);
        try {
          this._updateTheme(r), f.setTheme(this._theme), b(this), L.update.call(this, {
            type: "setTheme"
          }, g);
        } catch (S) {
          throw this[R] = !1, S;
        }
        this[R] = !1, F.call(this, o), H.call(this, o);
      }
    }, u.prototype._updateTheme = function(r) {
      Nt(r) && (r = Rt[r]), r && (r = Fe(r), r && pt(r, !0), this._theme = r);
    }, u.prototype.getModel = function() {
      return this._model;
    }, u.prototype.getOption = function() {
      return this._model && this._model.getOption();
    }, u.prototype.getWidth = function() {
      return this._zr.getWidth();
    }, u.prototype.getHeight = function() {
      return this._zr.getHeight();
    }, u.prototype.getDevicePixelRatio = function() {
      return this._zr.painter.dpr || q.hasGlobalWindow && window.devicePixelRatio || 1;
    }, u.prototype.getRenderedCanvas = function(r) {
      return process.env.NODE_ENV !== "production" && Er("getRenderedCanvas", "renderToCanvas"), this.renderToCanvas(r);
    }, u.prototype.renderToCanvas = function(r) {
      r = r || {};
      var n = this._zr.painter;
      if (process.env.NODE_ENV !== "production" && n.type !== "canvas")
        throw new Error("renderToCanvas can only be used in the canvas renderer.");
      return n.getRenderedCanvas({
        backgroundColor: r.backgroundColor || this._model.get("backgroundColor"),
        pixelRatio: r.pixelRatio || this.getDevicePixelRatio()
      });
    }, u.prototype.renderToSVGString = function(r) {
      r = r || {};
      var n = this._zr.painter;
      if (process.env.NODE_ENV !== "production" && n.type !== "svg")
        throw new Error("renderToSVGString can only be used in the svg renderer.");
      return n.renderToString({
        useViewBox: r.useViewBox
      });
    }, u.prototype.getSvgDataURL = function() {
      var r = this._zr, n = r.storage.getDisplayList();
      return C(n, function(f) {
        f.stopAnimation(null, !0);
      }), r.painter.toDataURL();
    }, u.prototype.getDataURL = function(r) {
      if (this._disposed) {
        D(this.id);
        return;
      }
      r = r || {};
      var n = r.excludeComponents, f = this._model, o = [], g = this;
      C(n, function(E) {
        f.eachComponent({
          mainType: E
        }, function(m) {
          var y = g._componentsMap[m.__viewId];
          y.group.ignore || (o.push(y), y.group.ignore = !0);
        });
      });
      var S = this._zr.painter.getType() === "svg" ? this.getSvgDataURL() : this.renderToCanvas(r).toDataURL("image/" + (r && r.type || "png"));
      return C(o, function(E) {
        E.group.ignore = !1;
      }), S;
    }, u.prototype.getConnectedDataURL = function(r) {
      if (this._disposed) {
        D(this.id);
        return;
      }
      var n = r.type === "svg", f = this.group, o = Math.min, g = Math.max, S = 1 / 0;
      if (ot[f]) {
        var E = S, m = S, y = -S, e = -S, i = [], t = r && r.pixelRatio || this.getDevicePixelRatio();
        C(K, function(l, v) {
          if (l.group === f) {
            var T = n ? l.getZr().painter.getSvgDom().innerHTML : l.renderToCanvas(Fe(r)), _ = l.getDom().getBoundingClientRect();
            E = o(_.left, E), m = o(_.top, m), y = g(_.right, y), e = g(_.bottom, e), i.push({
              dom: T,
              left: _.left,
              top: _.top
            });
          }
        }), E *= t, m *= t, y *= t, e *= t;
        var a = y - E, s = e - m, p = Or.createCanvas(), d = be(p, {
          renderer: n ? "svg" : "canvas"
        });
        if (d.resize({
          width: a,
          height: s
        }), n) {
          var h = "";
          return C(i, function(l) {
            var v = l.left - E, T = l.top - m;
            h += '<g transform="translate(' + v + "," + T + ')">' + l.dom + "</g>";
          }), d.painter.getSvgRoot().innerHTML = h, r.connectedBackgroundColor && d.painter.setBackgroundColor(r.connectedBackgroundColor), d.refreshImmediately(), d.painter.toDataURL();
        } else
          return r.connectedBackgroundColor && d.add(new Pr({
            shape: {
              x: 0,
              y: 0,
              width: a,
              height: s
            },
            style: {
              fill: r.connectedBackgroundColor
            }
          })), C(i, function(l) {
            var v = new Dr({
              style: {
                x: l.left * t - E,
                y: l.top * t - m,
                image: l.dom
              }
            });
            d.add(v);
          }), d.refreshImmediately(), p.toDataURL("image/" + (r && r.type || "png"));
      } else
        return this.getDataURL(r);
    }, u.prototype.convertToPixel = function(r, n, f) {
      return re(this, "convertToPixel", r, n, f);
    }, u.prototype.convertToLayout = function(r, n, f) {
      return re(this, "convertToLayout", r, n, f);
    }, u.prototype.convertFromPixel = function(r, n, f) {
      return re(this, "convertFromPixel", r, n, f);
    }, u.prototype.containPixel = function(r, n) {
      if (this._disposed) {
        D(this.id);
        return;
      }
      var f = this._model, o, g = he(f, r);
      return C(g, function(S, E) {
        E.indexOf("Models") >= 0 && C(S, function(m) {
          var y = m.coordinateSystem;
          if (y && y.containPoint)
            o = o || !!y.containPoint(n);
          else if (E === "seriesModels") {
            var e = this._chartsMap[m.__viewId];
            e && e.containPoint ? o = o || e.containPoint(n, m) : process.env.NODE_ENV !== "production" && x(E + ": " + (e ? "The found component do not support containPoint." : "No view mapping to the found component."));
          } else
            process.env.NODE_ENV !== "production" && x(E + ": containPoint is not supported");
        }, this);
      }, this), !!o;
    }, u.prototype.getVisual = function(r, n) {
      var f = this._model, o = he(f, r, {
        defaultMainType: "series"
      }), g = o.seriesModel;
      process.env.NODE_ENV !== "production" && (g || x("There is no specified series model"));
      var S = g.getData(), E = o.hasOwnProperty("dataIndexInside") ? o.dataIndexInside : o.hasOwnProperty("dataIndex") ? S.indexOfRawIndex(o.dataIndex) : null;
      return E != null ? _r(S, E, n) : gr(S, n);
    }, u.prototype.getViewOfComponentModel = function(r) {
      return this._componentsMap[r.__viewId];
    }, u.prototype.getViewOfSeriesModel = function(r) {
      return this._chartsMap[r.__viewId];
    }, u.prototype._initEvents = function() {
      var r = this;
      C(Br, function(f) {
        var o = function(g) {
          var S = r.getModel(), E = g.target, m, y = f === "globalout";
          if (y ? m = {} : E && Q(E, function(s) {
            var p = J(s);
            if (p && p.dataIndex != null) {
              var d = p.dataModel || S.getSeriesByIndex(p.seriesIndex);
              return m = d && d.getDataParams(p.dataIndex, p.dataType, E) || {}, !0;
            } else if (p.eventData)
              return m = z({}, p.eventData), !0;
          }, !0), m) {
            var e = m.componentType, i = m.componentIndex;
            (e === "markLine" || e === "markPoint" || e === "markArea") && (e = "series", i = m.seriesIndex);
            var t = e && i != null && S.getComponent(e, i), a = t && r[t.mainType === "series" ? "_chartsMap" : "_componentsMap"][t.__viewId];
            process.env.NODE_ENV !== "production" && !y && !(t && a) && x("model or view can not be found by params"), m.event = g, m.type = f, r._$eventProcessor.eventInfo = {
              targetEl: E,
              packedEvent: m,
              model: t,
              view: a
            }, r.trigger(f, m);
          }
        };
        o.zrEventfulCallAtLast = !0, r._zr.on(f, o, r);
      });
      var n = this._messageCenter;
      C(Ce, function(f, o) {
        n.on(o, function(g) {
          r.trigger(o, g);
        });
      }), Tr(n, this, this._api);
    }, u.prototype.isDisposed = function() {
      return this._disposed;
    }, u.prototype.clear = function() {
      if (this._disposed) {
        D(this.id);
        return;
      }
      this.setOption({
        series: []
      }, !0);
    }, u.prototype.dispose = function() {
      if (this._disposed) {
        D(this.id);
        return;
      }
      this._disposed = !0;
      var r = this.getDom();
      r && lt(this.getDom(), Ie, "");
      var n = this, f = n._api, o = n._model;
      C(n._componentsViews, function(g) {
        g.dispose(o, f);
      }), C(n._chartsViews, function(g) {
        g.dispose(o, f);
      }), n._zr.dispose(), n._dom = n._model = n._chartsMap = n._componentsMap = n._chartsViews = n._componentsViews = n._scheduler = n._api = n._zr = n._throttledZrFlush = n._theme = n._coordSysMgr = n._messageCenter = null, delete K[n.id];
    }, u.prototype.resize = function(r) {
      if (this[R]) {
        process.env.NODE_ENV !== "production" && ne("`resize` should not be called during main process.");
        return;
      }
      if (this._disposed) {
        D(this.id);
        return;
      }
      this._zr.resize(r);
      var n = this._model;
      if (this._loadingFX && this._loadingFX.resize(), !!n) {
        var f = n.resetOption("media"), o = r && r.silent;
        this[O] && (o == null && (o = this[O].silent), f = !0, this[O] = null), this[R] = !0, Y(this);
        try {
          f && b(this), L.update.call(this, {
            type: "resize",
            animation: z({
              // Disable animation
              duration: 0
            }, r && r.animation)
          });
        } catch (g) {
          throw this[R] = !1, g;
        }
        this[R] = !1, F.call(this, o), H.call(this, o);
      }
    }, u.prototype.showLoading = function(r, n) {
      if (this._disposed) {
        D(this.id);
        return;
      }
      if (B(r) && (n = r, r = ""), r = r || "default", this.hideLoading(), !we[r]) {
        process.env.NODE_ENV !== "production" && x("Loading effects " + r + " not exists.");
        return;
      }
      var f = we[r](this._api, n), o = this._zr;
      this._loadingFX = f, o.add(f);
    }, u.prototype.hideLoading = function() {
      if (this._disposed) {
        D(this.id);
        return;
      }
      this._loadingFX && this._zr.remove(this._loadingFX), this._loadingFX = null;
    }, u.prototype.makeActionFromEvent = function(r) {
      var n = z({}, r);
      return n.type = oe[r.type], n;
    }, u.prototype.dispatchAction = function(r, n) {
      if (this._disposed) {
        D(this.id);
        return;
      }
      if (B(n) || (n = {
        silent: !!n
      }), !!ae[r.type] && this._model) {
        if (this[R]) {
          this._pendingActions.push(r);
          return;
        }
        var f = n.silent;
        Te.call(this, r, f);
        var o = n.flush;
        o ? this._zr.flush() : o !== !1 && q.browser.weChat && this._throttledZrFlush(), F.call(this, f), H.call(this, f);
      }
    }, u.prototype.updateLabelLayout = function() {
      N.trigger("series:layoutlabels", this._model, this._api, {
        // Not adding series labels.
        // TODO
        updatedSeries: []
      });
    }, u.prototype.appendData = function(r) {
      if (this._disposed) {
        D(this.id);
        return;
      }
      var n = r.seriesIndex, f = this.getModel(), o = f.getSeriesByIndex(n);
      process.env.NODE_ENV !== "production" && G(r.data && o), o.appendData(r), this._scheduler.unfinished = !0, this.getZr().wakeUp();
    }, u.internalField = (function() {
      b = function(e) {
        var i = e._scheduler;
        i.restorePipelines(e._model), i.prepareStageTasks(), me(e, !0), me(e, !1), i.plan();
      }, me = function(e, i) {
        for (var t = e._model, a = e._scheduler, s = i ? e._componentsViews : e._chartsViews, p = i ? e._componentsMap : e._chartsMap, d = e._zr, h = e._api, l = 0; l < s.length; l++)
          s[l].__alive = !1;
        i ? t.eachComponent(function(_, A) {
          _ !== "series" && v(A);
        }) : t.eachSeries(v);
        function v(_) {
          var A = _.__requireNewView;
          _.__requireNewView = !1;
          var w = "_ec_" + _.id + "_" + _.type, P = !A && p[w];
          if (!P) {
            var M = Je(_.type), U = i ? zt.getClass(M.main, M.sub) : (
              // FIXME:TS
              // (ChartView as ChartViewConstructor).getClass('series', classType.sub)
              // For backward compat, still support a chart type declared as only subType
              // like "liquidfill", but recommend "series.liquidfill"
              // But need a base class to make a type series.
              le.getClass(M.sub)
            );
            process.env.NODE_ENV !== "production" && G(U, M.sub + " does not exist."), P = new U(), P.init(t, h), p[w] = P, s.push(P), d.add(P.group);
          }
          _.__viewId = P.__id = w, P.__alive = !0, P.__model = _, P.group.__ecComponentInfo = {
            mainType: _.mainType,
            index: _.componentIndex
          }, !i && a.prepareView(P, _, t, h);
        }
        for (var l = 0; l < s.length; ) {
          var T = s[l];
          T.__alive ? l++ : (!i && T.renderTask.dispose(), d.remove(T.group), T.dispose(t, h), s.splice(l, 1), p[T.__id] === T && delete p[T.__id], T.__id = T.group.__ecComponentInfo = null);
        }
      }, te = function(e, i, t, a, s) {
        var p = e._model;
        if (p.setUpdatePayload(t), !a) {
          C([].concat(e._componentsViews).concat(e._chartsViews), T);
          return;
        }
        var d = {};
        d[a + "Id"] = t[a + "Id"], d[a + "Index"] = t[a + "Index"], d[a + "Name"] = t[a + "Name"];
        var h = {
          mainType: a,
          query: d
        };
        s && (h.subType = s);
        var l = t.excludeSeriesId, v;
        l != null && (v = de(), C(or(l), function(_) {
          var A = ar(_, null);
          A != null && v.set(A, !0);
        })), p && p.eachComponent(h, function(_) {
          var A = v && v.get(_.id) != null;
          if (!A)
            if (Ge(t))
              if (_ instanceof Ye)
                t.type === Z && !t.notBlur && !_.get(["emphasis", "disabled"]) && Qt(_, t, e._api);
              else {
                var w = er(_.mainType, _.componentIndex, t.name, e._api), P = w.focusSelf, M = w.dispatchers;
                t.type === Z && P && !t.notBlur && tr(_.mainType, _.componentIndex, e._api), M && C(M, function(U) {
                  t.type === Z ? We(U) : Xe(U);
                });
              }
            else Be(t) && _ instanceof Ye && (rr(_, t, e._api), Ze(_), I(e));
        }, e), p && p.eachComponent(h, function(_) {
          var A = v && v.get(_.id) != null;
          A || T(e[a === "series" ? "_chartsMap" : "_componentsMap"][_.__viewId]);
        }, e);
        function T(_) {
          _ && _.__alive && _[i] && _[i](_.__model, p, e._api, t);
        }
      }, L = {
        prepareAndUpdate: function(e) {
          b(this), L.update.call(this, e, e && {
            // Needs to mark option changed if newOption is given.
            // It's from MagicType.
            // TODO If use a separate flag optionChanged in payload?
            optionChanged: e.newOption != null
          });
        },
        update: function(e, i) {
          var t = this._model, a = this._api, s = this._zr, p = this._coordSysMgr, d = this._scheduler;
          if (t) {
            t.setUpdatePayload(e), d.restoreData(t, e), d.performSeriesTasks(t), p.create(t, a), d.performDataProcessorTasks(t, e), Ee(this, t), p.update(t, a), n(t), d.performVisualTasks(t, e);
            var h = t.get("backgroundColor") || "transparent";
            s.setBackgroundColor(h);
            var l = t.get("darkMode");
            l != null && l !== "auto" && s.setDarkMode(l), Se(this, t, a, e, i), N.trigger("afterupdate", t, a);
          }
        },
        updateTransform: function(e) {
          var i = this, t = this._model, a = this._api;
          if (t) {
            t.setUpdatePayload(e);
            var s = [];
            t.eachComponent(function(d, h) {
              if (d !== "series") {
                var l = i.getViewOfComponentModel(h);
                if (l && l.__alive)
                  if (l.updateTransform) {
                    var v = l.updateTransform(h, t, a, e);
                    v && v.update && s.push(l);
                  } else
                    s.push(l);
              }
            });
            var p = de();
            t.eachSeries(function(d) {
              var h = i._chartsMap[d.__viewId];
              if (h.updateTransform) {
                var l = h.updateTransform(d, t, a, e);
                l && l.update && p.set(d.uid, 1);
              } else
                p.set(d.uid, 1);
            }), n(t), this._scheduler.performVisualTasks(t, e, {
              setDirty: !0,
              dirtyMap: p
            }), ie(this, t, a, e, {}, p), N.trigger("afterupdate", t, a);
          }
        },
        updateView: function(e) {
          var i = this._model;
          i && (i.setUpdatePayload(e), le.markUpdateMethod(e, "updateView"), n(i), this._scheduler.performVisualTasks(i, e, {
            setDirty: !0
          }), Se(this, i, this._api, e, {}), N.trigger("afterupdate", i, this._api));
        },
        updateVisual: function(e) {
          var i = this, t = this._model;
          t && (t.setUpdatePayload(e), t.eachSeries(function(a) {
            a.getData().clearAllVisual();
          }), le.markUpdateMethod(e, "updateVisual"), n(t), this._scheduler.performVisualTasks(t, e, {
            visualType: "visual",
            setDirty: !0
          }), t.eachComponent(function(a, s) {
            if (a !== "series") {
              var p = i.getViewOfComponentModel(s);
              p && p.__alive && p.updateVisual(s, t, i._api, e);
            }
          }), t.eachSeries(function(a) {
            var s = i._chartsMap[a.__viewId];
            s.updateVisual(a, t, i._api, e);
          }), N.trigger("afterupdate", t, this._api));
        },
        updateLayout: function(e) {
          L.update.call(this, e);
        }
      };
      function r(e, i, t, a, s) {
        if (e._disposed) {
          D(e.id);
          return;
        }
        for (var p = e._model, d = e._coordSysMgr.getCoordinateSystems(), h, l = he(p, t), v = 0; v < d.length; v++) {
          var T = d[v];
          if (T[i] && (h = T[i](p, l, a, s)) != null)
            return h;
        }
        process.env.NODE_ENV !== "production" && x("No coordinate system that supports " + i + " found by the given finder.");
      }
      re = r, Ee = function(e, i) {
        var t = e._chartsMap, a = e._scheduler;
        i.eachSeries(function(s) {
          a.updateStreamModes(s, t[s.__viewId]);
        });
      }, Te = function(e, i) {
        var t = this, a = this.getModel(), s = e.type, p = e.escapeConnect, d = ae[s], h = (d.update || "update").split(":"), l = h.pop(), v = h[0] != null && Je(h[0]);
        this[R] = !0, Y(this);
        var T = [e], _ = !1;
        e.batch && (_ = !0, T = Mt(e.batch, function(V) {
          return V = He(z({}, V), e), V.batch = null, V;
        }));
        var A = [], w, P = [], M = d.nonRefinedEventType, U = Be(e), fe = Ge(e);
        if (fe && Yt(this._api), C(T, function(V) {
          var Ue = d.action(V, a, t._api);
          if (d.refineEvent ? P.push(Ue) : w = Ue, w = w || z({}, V), w.type = M, A.push(w), fe) {
            var ke = nr(e), Pt = ke.queryOptionMap, At = ke.mainTypeSpecified, Dt = At ? Pt.keys()[0] : "series";
            te(t, l, V, Dt), I(t);
          } else U ? (te(t, l, V, "series"), I(t)) : v && te(t, l, V, v.main, v.sub);
        }), l !== "none" && !fe && !U && !v)
          try {
            this[O] ? (b(this), L.update.call(this, e), this[O] = null) : L[l].call(this, e);
          } catch (V) {
            throw this[R] = !1, V;
          }
        if (_ ? w = {
          type: M,
          escapeConnect: p,
          batch: A
        } : w = A[0], this[R] = !1, !i) {
          var k = void 0;
          if (d.refineEvent) {
            var Le = d.refineEvent(P, e, a, this._api).eventContent;
            G(B(Le)), k = He({
              type: d.refinedEventType
            }, Le), k.fromAction = e.type, k.fromActionPayload = e, k.escapeConnect = !0;
          }
          var xe = this._messageCenter;
          xe.trigger(w.type, w), k && xe.trigger(k.type, k);
        }
      }, F = function(e) {
        for (var i = this._pendingActions; i.length; ) {
          var t = i.shift();
          Te.call(this, t, e);
        }
      }, H = function(e) {
        !e && this.trigger("updated");
      }, tt = function(e, i) {
        e.on("rendered", function(t) {
          i.trigger("rendered", t), // Although zr is dirty if initial animation is not finished
          // and this checking is called on frame, we also check
          // animation finished for robustness.
          e.animation.isFinished() && !i[O] && !i._scheduler.unfinished && !i._pendingActions.length && i.trigger("finished");
        });
      }, rt = function(e, i) {
        e.on("mouseover", function(t) {
          var a = t.target, s = Q(a, qe);
          s && (jt(s, t, i._api), I(i));
        }).on("mouseout", function(t) {
          var a = t.target, s = Q(a, qe);
          s && ($t(s, t, i._api), I(i));
        }).on("click", function(t) {
          var a = t.target, s = Q(a, function(h) {
            return J(h).dataIndex != null;
          }, !0);
          if (s) {
            var p = s.selected ? "unselect" : "select", d = J(s);
            i._api.dispatchAction({
              type: p,
              dataType: d.dataType,
              dataIndexInside: d.dataIndex,
              seriesIndex: d.seriesIndex,
              isFromClick: !0
            });
          }
        });
      };
      function n(e) {
        e.clearColorPalette(), e.eachSeries(function(i) {
          i.clearColorPalette();
        });
      }
      function f(e) {
        var i = [], t = [], a = !1;
        if (e.eachComponent(function(h, l) {
          var v = l.get("zlevel") || 0, T = l.get("z") || 0, _ = l.getZLevelKey();
          a = a || !!_, (h === "series" ? t : i).push({
            zlevel: v,
            z: T,
            idx: l.componentIndex,
            type: h,
            key: _
          });
        }), a) {
          var s = i.concat(t), p, d;
          pe(s, function(h, l) {
            return h.zlevel === l.zlevel ? h.z - l.z : h.zlevel - l.zlevel;
          }), C(s, function(h) {
            var l = e.getComponent(h.type, h.idx), v = h.zlevel, T = h.key;
            p != null && (v = Math.max(p, v)), T ? (v === p && T !== d && v++, d = T) : d && (v === p && v++, d = ""), p = v, l.setZLevel(v);
          });
        }
      }
      Se = function(e, i, t, a, s) {
        f(i), it(e, i, t, a, s), C(e._chartsViews, function(p) {
          p.__alive = !1;
        }), ie(e, i, t, a, s), C(e._chartsViews, function(p) {
          p.__alive || p.remove(i, t);
        });
      }, it = function(e, i, t, a, s, p) {
        C(p || e._componentsViews, function(d) {
          var h = d.__model;
          m(h, d), d.render(h, i, t, a), E(h, d), y(h, d);
        });
      }, ie = function(e, i, t, a, s, p) {
        var d = e._scheduler;
        s = z(s || {}, {
          updatedSeries: i.getSeries()
        }), N.trigger("series:beforeupdate", i, t, s);
        var h = !1;
        i.eachSeries(function(l) {
          var v = e._chartsMap[l.__viewId];
          v.__alive = !0;
          var T = v.renderTask;
          d.updatePayload(T, a), m(l, v), p && p.get(l.uid) && T.dirty(), T.perform(d.getPerformArgs(T)) && (h = !0), v.group.silent = !!l.get("silent"), S(l, v), Ze(l);
        }), d.unfinished = h || d.unfinished, N.trigger("series:layoutlabels", i, t, s), N.trigger("series:transition", i, t, s), i.eachSeries(function(l) {
          var v = e._chartsMap[l.__viewId];
          E(l, v), y(l, v);
        }), g(e, i), N.trigger("series:afterupdate", i, t, s);
      }, I = function(e) {
        e[_e] = !0, e.getZr().wakeUp();
      }, Y = function(e) {
        e[ee] = (e[ee] + 1) % 1e3;
      }, at = function(e) {
        e[_e] && (e.getZr().storage.traverse(function(i) {
          ve(i) || o(i);
        }), e[_e] = !1);
      };
      function o(e) {
        for (var i = [], t = e.currentStates, a = 0; a < t.length; a++) {
          var s = t[a];
          s === "emphasis" || s === "blur" || s === "select" || i.push(s);
        }
        e.selected && e.states.select && i.push("select"), e.hoverState === Bt && e.states.emphasis ? i.push("emphasis") : e.hoverState === Gt && e.states.blur && i.push("blur"), e.useStates(i);
      }
      function g(e, i) {
        var t = e._zr, a = t.storage, s = 0;
        a.traverse(function(p) {
          p.isGroup || s++;
        }), s > i.get("hoverLayerThreshold") && !q.node && !q.worker && i.eachSeries(function(p) {
          if (!p.preventUsingHoverLayer) {
            var d = e._chartsMap[p.__viewId];
            d.__alive && d.eachRendered(function(h) {
              h.states.emphasis && (h.states.emphasis.hoverLayer = !0);
            });
          }
        });
      }
      function S(e, i) {
        var t = e.get("blendMode") || null;
        i.eachRendered(function(a) {
          a.isGroup || (a.style.blend = t);
        });
      }
      function E(e, i) {
        if (!e.preventAutoZ) {
          var t = Ft(e);
          i.eachRendered(function(a) {
            return Ht(a, t.z, t.zlevel), !0;
          });
        }
      }
      function m(e, i) {
        i.eachRendered(function(t) {
          if (!ve(t)) {
            var a = t.getTextContent(), s = t.getTextGuideLine();
            t.stateTransition && (t.stateTransition = null), a && a.stateTransition && (a.stateTransition = null), s && s.stateTransition && (s.stateTransition = null), t.hasState() ? (t.prevStates = t.currentStates, t.clearStates()) : t.prevStates && (t.prevStates = null);
          }
        });
      }
      function y(e, i) {
        var t = e.getModel("stateAnimation"), a = e.isAnimationEnabled(), s = t.get("duration"), p = s > 0 ? {
          duration: s,
          delay: t.get("delay"),
          easing: t.get("easing")
          // additive: stateAnimationModel.get('additive')
        } : null;
        i.eachRendered(function(d) {
          if (d.states && d.states.emphasis) {
            if (ve(d))
              return;
            if (d instanceof Ar && Zt(d), d.__dirty) {
              var h = d.prevStates;
              h && d.useStates(h);
            }
            if (a) {
              d.stateTransition = p;
              var l = d.getTextContent(), v = d.getTextGuideLine();
              l && (l.stateTransition = p), v && (v.stateTransition = p);
            }
            d.__dirty && o(d);
          }
        });
      }
      nt = function(e) {
        return new /** @class */
        ((function(i) {
          ye(t, i);
          function t() {
            return i !== null && i.apply(this, arguments) || this;
          }
          return t.prototype.getCoordinateSystems = function() {
            return e._coordSysMgr.getCoordinateSystems();
          }, t.prototype.getComponentByElement = function(a) {
            for (; a; ) {
              var s = a.__ecComponentInfo;
              if (s != null)
                return e._model.getComponent(s.mainType, s.index);
              a = a.parent;
            }
          }, t.prototype.enterEmphasis = function(a, s) {
            We(a, s), I(e);
          }, t.prototype.leaveEmphasis = function(a, s) {
            Xe(a, s), I(e);
          }, t.prototype.enterBlur = function(a) {
            Wt(a), I(e);
          }, t.prototype.leaveBlur = function(a) {
            Xt(a), I(e);
          }, t.prototype.enterSelect = function(a) {
            qt(a), I(e);
          }, t.prototype.leaveSelect = function(a) {
            Kt(a), I(e);
          }, t.prototype.getModel = function() {
            return e.getModel();
          }, t.prototype.getViewOfComponentModel = function(a) {
            return e.getViewOfComponentModel(a);
          }, t.prototype.getViewOfSeriesModel = function(a) {
            return e.getViewOfSeriesModel(a);
          }, t.prototype.getMainProcessVersion = function() {
            return e[ee];
          }, t;
        })(Ut))(e);
      }, yt = function(e) {
        function i(t, a) {
          for (var s = 0; s < t.length; s++) {
            var p = t[s];
            p[ge] = a;
          }
        }
        C(oe, function(t, a) {
          e._messageCenter.on(a, function(s) {
            if (ot[e.group] && e[ge] !== et) {
              if (s && s.escapeConnect)
                return;
              var p = e.makeActionFromEvent(s), d = [];
              C(K, function(h) {
                h !== e && h.group === e.group && d.push(h);
              }), i(d, et), C(d, function(h) {
                h[ge] !== Hr && h.dispatchAction(p);
              }), i(d, Yr);
            }
          });
        });
      };
    })(), u;
  })(Pe)
), Ve = Ct.prototype;
Ve.on = gt("on");
Ve.off = gt("off");
Ve.one = function(c, u, r) {
  var n = this;
  mr("ECharts#one is deprecated.");
  function f() {
    for (var o = [], g = 0; g < arguments.length; g++)
      o[g] = arguments[g];
    u && u.apply && u.apply(this, o), n.off(c, f);
  }
  this.on.call(this, c, f, r);
};
var Br = ["click", "dblclick", "mouseover", "mouseout", "mousemove", "mousedown", "mouseup", "globalout", "contextmenu"];
function D(c) {
  process.env.NODE_ENV !== "production" && x("Instance " + c + " has been disposed");
}
var ae = {}, oe = {}, Ce = {}, Re = [], Oe = [], se = [], Rt = {}, we = {}, K = {}, ot = {}, Gr = +/* @__PURE__ */ new Date() - 0, Ie = "_echarts_instance_";
function ki(c, u, r) {
  {
    if (process.env.NODE_ENV !== "production" && !c)
      throw new Error("Initialize failed: invalid dom.");
    var n = Zr(c);
    if (n)
      return process.env.NODE_ENV !== "production" && x("There is a chart instance already initialized on the dom."), n;
    process.env.NODE_ENV !== "production" && Lt(c) && c.nodeName.toUpperCase() !== "CANVAS" && (!c.clientWidth && !r || !c.clientHeight && !r) && x("Can't get DOM width or height. Please check dom.clientWidth and dom.clientHeight. They should not be 0.For example, you may need to call this in the callback of window.onload.");
  }
  var f = new Ct(c, u, r);
  return f.id = "ec_" + Gr++, K[f.id] = f, lt(c, Ie, f.id), yt(f), N.trigger("afterinit", f), f;
}
function Zr(c) {
  return K[ir(c, Ie)];
}
function Ot(c, u) {
  Rt[c] = u;
}
function Wr(c) {
  ut(Oe, c) < 0 && Oe.push(c);
}
function Xr(c, u) {
  Ne(Re, c, u, Lr);
}
function bi(c) {
  wt("afterinit", c);
}
function zi(c) {
  wt("afterupdate", c);
}
function wt(c, u) {
  N.on(c, u);
}
function $(c, u, r) {
  var n, f, o, g, S;
  ft(u) && (r = u, u = ""), B(c) ? (n = c.type, f = c.event, g = c.update, S = c.publishNonRefinedEvent, r || (r = c.action), o = c.refineEvent) : (n = c, f = u);
  function E(y) {
    return y.toLowerCase();
  }
  f = E(f || n);
  var m = o ? E(n) : f;
  ae[n] || (G(Qe.test(n) && Qe.test(f)), o && G(f !== n), ae[n] = {
    actionType: n,
    refinedEventType: f,
    nonRefinedEventType: m,
    update: g,
    action: r,
    refineEvent: o
  }, Ce[f] = 1, o && S && (Ce[m] = 1), process.env.NODE_ENV !== "production" && oe[m] && ne(m + ' must not be shared; use "refineEvent" if you intend to share an event name.'), oe[m] = n);
}
function Fi(c, u) {
  dt.register(c, u);
}
function Hi(c, u) {
  Ne(se, c, u, ht, "layout");
}
function W(c, u) {
  Ne(se, c, u, vt, "visual");
}
var st = [];
function Ne(c, u, r, n, f) {
  if ((ft(u) || B(u)) && (r = u, u = n), process.env.NODE_ENV !== "production") {
    if (isNaN(u) || u == null)
      throw new Error("Illegal priority");
    C(c, function(g) {
      G(g.__raw !== r);
    });
  }
  if (!(ut(st, r) >= 0)) {
    st.push(r);
    var o = ct.wrapStageHandler(r, f);
    o.__prio = u, o.__raw = r, c.push(o);
  }
}
function qr(c, u) {
  we[c] = u;
}
function Yi(c, u, r) {
  var n = wr("registerMap");
  n && n(c, u, r);
}
var Bi = Sr;
W(De, ur);
W(ue, fr);
W(ue, dr);
W(De, hr);
W(ue, vr);
W(_t, Rr);
Wr(pt);
Xr(Nr, bt);
qr("default", pr);
$({
  type: Z,
  event: Z,
  update: Z
}, j);
$({
  type: ce,
  event: ce,
  update: ce
}, j);
$({
  type: Ke,
  event: Ae,
  update: Ke,
  action: j,
  refineEvent: Me,
  publishNonRefinedEvent: !0
});
$({
  type: je,
  event: Ae,
  update: je,
  action: j,
  refineEvent: Me,
  publishNonRefinedEvent: !0
});
$({
  type: $e,
  event: Ae,
  update: $e,
  action: j,
  refineEvent: Me,
  publishNonRefinedEvent: !0
});
function Me(c, u, r, n) {
  return {
    eventContent: {
      selected: Jt(r),
      isFromClick: u.isFromClick || !1
    }
  };
}
Ot("default", {});
Ot("dark", lr);
export {
  Ui as PRIORITY,
  Zr as getInstanceByDom,
  ki as init,
  $ as registerAction,
  Fi as registerCoordinateSystem,
  Hi as registerLayout,
  qr as registerLoading,
  Wi as registerLocale,
  Yi as registerMap,
  bi as registerPostInit,
  zi as registerPostUpdate,
  Wr as registerPreprocessor,
  Xr as registerProcessor,
  Ot as registerTheme,
  Bi as registerTransform,
  wt as registerUpdateLifecycle,
  W as registerVisual
};
