// echarts@6.0.0/lib/core/locale.js
import m from "../model/Model.js";
import p from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/env.js";
import L from "../i18n/langEN.js";
import s from "../i18n/langZH.js";
import { isString as v, clone as n, merge as f } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
var a = "ZH", i = "EN", e = i, o = {}, u = {}, S = p.domSupported ? (function() {
  var r = (
    /* eslint-disable-next-line */
    (document.documentElement.lang || navigator.language || navigator.browserLanguage || e).toUpperCase()
  );
  return r.indexOf(a) > -1 ? a : e;
})() : e;
function g(r, t) {
  r = r.toUpperCase(), u[r] = new m(t), o[r] = t;
}
function A(r) {
  if (v(r)) {
    var t = o[r.toUpperCase()] || {};
    return r === a || r === i ? n(t) : f(n(t), n(o[e]), !1);
  } else
    return f(n(r), n(o[e]), !1);
}
function M(r) {
  return u[r];
}
function N() {
  return u[e];
}
g(i, L);
g(a, s);
export {
  S as SYSTEM_LANG,
  A as createLocaleObject,
  N as getDefaultLocaleModel,
  M as getLocaleModel,
  g as registerLocale
};
