// echarts@6.0.0/lib/core/task.js
import { assert as p, isArray as D } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
function N(i) {
  return new g(i);
}
var g = (
  /** @class */
  (function() {
    function i(t) {
      t = t || {}, this._reset = t.reset, this._plan = t.plan, this._count = t.count, this._onDirty = t.onDirty, this._dirty = !0;
    }
    return i.prototype.perform = function(t) {
      var s = this._upstream, e = t && t.skip;
      if (this._dirty && s) {
        var n = this.context;
        n.data = n.outputData = s.context.outputData;
      }
      this.__pipeline && (this.__pipeline.currentTask = this);
      var o;
      this._plan && !e && (o = this._plan(this.context));
      var c = _(this._modBy), l = this._modDataCount || 0, r = _(t && t.modBy), u = t && t.modDataCount || 0;
      (c !== r || l !== u) && (o = "reset");
      function _(v) {
        return !(v >= 1) && (v = 1), v;
      }
      var d;
      (this._dirty || o === "reset") && (this._dirty = !1, d = this._doReset(e)), this._modBy = r, this._modDataCount = u;
      var m = t && t.step;
      if (s ? (process.env.NODE_ENV !== "production" && p(s._outputDueEnd != null), this._dueEnd = s._outputDueEnd) : (process.env.NODE_ENV !== "production" && p(!this._progress || this._count), this._dueEnd = this._count ? this._count(this.context) : 1 / 0), this._progress) {
        var f = this._dueIndex, h = Math.min(m != null ? this._dueIndex + m : 1 / 0, this._dueEnd);
        if (!e && (d || f < h)) {
          var a = this._progress;
          if (D(a))
            for (var y = 0; y < a.length; y++)
              this._doProgress(a[y], f, h, r, u);
          else
            this._doProgress(a, f, h, r, u);
        }
        this._dueIndex = h;
        var E = this._settedOutputEnd != null ? this._settedOutputEnd : h;
        process.env.NODE_ENV !== "production" && p(E >= this._outputDueEnd), this._outputDueEnd = E;
      } else
        this._dueIndex = this._outputDueEnd = this._settedOutputEnd != null ? this._settedOutputEnd : this._dueEnd;
      return this.unfinished();
    }, i.prototype.dirty = function() {
      this._dirty = !0, this._onDirty && this._onDirty(this.context);
    }, i.prototype._doProgress = function(t, s, e, n, o) {
      x.reset(s, e, n, o), this._callingProgress = t, this._callingProgress({
        start: s,
        end: e,
        count: e - s,
        next: x.next
      }, this.context);
    }, i.prototype._doReset = function(t) {
      this._dueIndex = this._outputDueEnd = this._dueEnd = 0, this._settedOutputEnd = null;
      var s, e;
      !t && this._reset && (s = this._reset(this.context), s && s.progress && (e = s.forceFirstProgress, s = s.progress), D(s) && !s.length && (s = null)), this._progress = s, this._modBy = this._modDataCount = null;
      var n = this._downstream;
      return n && n.dirty(), e;
    }, i.prototype.unfinished = function() {
      return this._progress && this._dueIndex < this._dueEnd;
    }, i.prototype.pipe = function(t) {
      process.env.NODE_ENV !== "production" && p(t && !t._disposed && t !== this), (this._downstream !== t || this._dirty) && (this._downstream = t, t._upstream = this, t.dirty());
    }, i.prototype.dispose = function() {
      this._disposed || (this._upstream && (this._upstream._downstream = null), this._downstream && (this._downstream._upstream = null), this._dirty = !1, this._disposed = !0);
    }, i.prototype.getUpstream = function() {
      return this._upstream;
    }, i.prototype.getDownstream = function() {
      return this._downstream;
    }, i.prototype.setOutputEnd = function(t) {
      this._outputDueEnd = this._settedOutputEnd = t;
    }, i;
  })()
), x = /* @__PURE__ */ (function() {
  var i, t, s, e, n, o = {
    reset: function(r, u, _, d) {
      t = r, i = u, s = _, e = d, n = Math.ceil(e / s), o.next = s > 1 && e > 0 ? l : c;
    }
  };
  return o;
  function c() {
    return t < i ? t++ : null;
  }
  function l() {
    var r = t % n * s + Math.ceil(t / n), u = t >= i ? null : r < e ? r : t;
    return t++, u;
  }
})();
export {
  g as Task,
  N as createTask
};
