// echarts@6.0.0/lib/processor/dataStack.js
import { createHashMap as f, each as y } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { addSafe as B } from "../util/number.js";
function S(u) {
  var i = f();
  u.eachSeries(function(e) {
    var n = e.get("stack");
    if (n) {
      var c = i.get(n) || i.set(n, []), s = e.getData(), a = {
        // Used for calculate axis extent automatically.
        // TODO: Type getCalculationInfo return more specific type?
        stackResultDimension: s.getCalculationInfo("stackResultDimension"),
        stackedOverDimension: s.getCalculationInfo("stackedOverDimension"),
        stackedDimension: s.getCalculationInfo("stackedDimension"),
        stackedByDimension: s.getCalculationInfo("stackedByDimension"),
        isStackedByIndex: s.getCalculationInfo("isStackedByIndex"),
        data: s,
        seriesModel: e
      };
      if (!a.stackedDimension || !(a.isStackedByIndex || a.stackedByDimension))
        return;
      c.push(a);
    }
  }), i.each(function(e) {
    if (e.length !== 0) {
      var n = e[0].seriesModel, c = n.get("stackOrder") || "seriesAsc";
      c === "seriesDesc" && e.reverse(), y(e, function(s, a) {
        s.data.setCalculationInfo("stackedOnSeries", a > 0 ? e[a - 1].seriesModel : null);
      }), N(e);
    }
  });
}
function N(u) {
  y(u, function(i, e) {
    var n = [], c = [NaN, NaN], s = [i.stackResultDimension, i.stackedOverDimension], a = i.data, D = i.isStackedByIndex, d = i.seriesModel.get("stackStrategy") || "samesign";
    a.modify(s, function(O, p, m) {
      var r = a.get(i.stackedDimension, m);
      if (isNaN(r))
        return c;
      var g, o;
      D ? o = a.getRawIndex(m) : g = a.get(i.stackedByDimension, m);
      for (var k = NaN, v = e - 1; v >= 0; v--) {
        var l = u[v];
        if (D || (o = l.data.rawIndexOf(l.stackedByDimension, g)), o >= 0) {
          var t = l.data.getByRawIndex(l.stackResultDimension, o);
          if (d === "all" || d === "positive" && t > 0 || d === "negative" && t < 0 || d === "samesign" && r >= 0 && t > 0 || d === "samesign" && r <= 0 && t < 0) {
            r = B(r, t), k = t;
            break;
          }
        }
      }
      return n[0] = r, n[1] = k, n;
    });
  });
}
export {
  S as default
};
