// echarts@6.0.0/lib/processor/dataSample.js
import { isString as d, isFunction as l } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
var D = {
  average: function(t) {
    for (var n = 0, a = 0, o = 0; o < t.length; o++)
      isNaN(t[o]) || (n += t[o], a++);
    return a === 0 ? NaN : n / a;
  },
  sum: function(t) {
    for (var n = 0, a = 0; a < t.length; a++)
      n += t[a] || 0;
    return n;
  },
  max: function(t) {
    for (var n = -1 / 0, a = 0; a < t.length; a++)
      t[a] > n && (n = t[a]);
    return isFinite(n) ? n : NaN;
  },
  min: function(t) {
    for (var n = 1 / 0, a = 0; a < t.length; a++)
      t[a] < n && (n = t[a]);
    return isFinite(n) ? n : NaN;
  },
  // TODO
  // Median
  nearest: function(t) {
    return t[0];
  }
}, h = function(t) {
  return Math.round(t.length / 2);
};
function S(t) {
  return {
    seriesType: t,
    // FIXME:TS never used, so comment it
    // modifyOutputEnd: true,
    reset: function(n, a, o) {
      var i = n.getData(), r = n.get("sampling"), v = n.coordinateSystem, c = i.count();
      if (c > 10 && v.type === "cartesian2d" && r) {
        var f = v.getBaseAxis(), m = v.getOtherAxis(f), s = f.getExtent(), g = o.getDevicePixelRatio(), p = Math.abs(s[1] - s[0]) * (g || 1), e = Math.round(c / p);
        if (isFinite(e) && e > 1) {
          r === "lttb" ? n.setData(i.lttbDownSample(i.mapDimension(m.dim), 1 / e)) : r === "minmax" && n.setData(i.minmaxDownSample(i.mapDimension(m.dim), 1 / e));
          var u = void 0;
          d(r) ? u = D[r] : l(r) && (u = r), u && n.setData(i.downSample(i.mapDimension(m.dim), 1 / e, u, h));
        }
      }
    }
  };
}
export {
  S as default
};
