// zod@3.25.76/v3/helpers/parseUtil.js
import { getErrorMap as p } from "../errors.js";
import d from "../locales/en.js";
const y = (t) => {
  const { data: e, path: r, errorMaps: a, issueData: s } = t, o = [...r, ...s.path || []], i = {
    ...s,
    path: o
  };
  if (s.message !== void 0)
    return {
      ...s,
      path: o,
      message: s.message
    };
  let u = "";
  const f = a.filter((n) => !!n).slice().reverse();
  for (const n of f)
    u = n(i, { data: e, defaultError: u }).message;
  return {
    ...s,
    path: o,
    message: u
  };
};
function h(t, e) {
  const r = p(), a = y({
    issueData: e,
    data: t.data,
    path: t.path,
    errorMaps: [
      t.common.contextualErrorMap,
      // contextual error map is first priority
      t.schemaErrorMap,
      // then schema-bound map if available
      r,
      // then global override map
      r === d ? void 0 : d
      // then global default map
    ].filter((s) => !!s)
  });
  t.common.issues.push(a);
}
class l {
  constructor() {
    this.value = "valid";
  }
  dirty() {
    this.value === "valid" && (this.value = "dirty");
  }
  abort() {
    this.value !== "aborted" && (this.value = "aborted");
  }
  static mergeArray(e, r) {
    const a = [];
    for (const s of r) {
      if (s.status === "aborted")
        return c;
      s.status === "dirty" && e.dirty(), a.push(s.value);
    }
    return { status: e.value, value: a };
  }
  static async mergeObjectAsync(e, r) {
    const a = [];
    for (const s of r) {
      const o = await s.key, i = await s.value;
      a.push({
        key: o,
        value: i
      });
    }
    return l.mergeObjectSync(e, a);
  }
  static mergeObjectSync(e, r) {
    const a = {};
    for (const s of r) {
      const { key: o, value: i } = s;
      if (o.status === "aborted" || i.status === "aborted")
        return c;
      o.status === "dirty" && e.dirty(), i.status === "dirty" && e.dirty(), o.value !== "__proto__" && (typeof i.value < "u" || s.alwaysSet) && (a[o.value] = i.value);
    }
    return { status: e.value, value: a };
  }
}
const c = Object.freeze({
  status: "aborted"
}), b = (t) => ({ status: "dirty", value: t }), g = (t) => ({ status: "valid", value: t }), M = (t) => t.status === "aborted", I = (t) => t.status === "dirty", O = (t) => t.status === "valid", j = (t) => typeof Promise < "u" && t instanceof Promise;
export {
  b as DIRTY,
  c as INVALID,
  g as OK,
  l as ParseStatus,
  h as addIssueToContext,
  M as isAborted,
  j as isAsync,
  I as isDirty,
  O as isValid,
  y as makeIssue
};
