// zod@3.25.76/v3/helpers/util.js
var u;
(function(n) {
  n.assertEqual = (e) => {
  };
  function s(e) {
  }
  n.assertIs = s;
  function f(e) {
    throw new Error();
  }
  n.assertNever = f, n.arrayToEnum = (e) => {
    const r = {};
    for (const o of e)
      r[o] = o;
    return r;
  }, n.getValidEnumValues = (e) => {
    const r = n.objectKeys(e).filter((i) => typeof e[e[i]] != "number"), o = {};
    for (const i of r)
      o[i] = e[i];
    return n.objectValues(o);
  }, n.objectValues = (e) => n.objectKeys(e).map(function(r) {
    return e[r];
  }), n.objectKeys = typeof Object.keys == "function" ? (e) => Object.keys(e) : (e) => {
    const r = [];
    for (const o in e)
      Object.prototype.hasOwnProperty.call(e, o) && r.push(o);
    return r;
  }, n.find = (e, r) => {
    for (const o of e)
      if (r(o))
        return o;
  }, n.isInteger = typeof Number.isInteger == "function" ? (e) => Number.isInteger(e) : (e) => typeof e == "number" && Number.isFinite(e) && Math.floor(e) === e;
  function a(e, r = " | ") {
    return e.map((o) => typeof o == "string" ? `'${o}'` : o).join(r);
  }
  n.joinValues = a, n.jsonStringifyReplacer = (e, r) => typeof r == "bigint" ? r.toString() : r;
})(u || (u = {}));
var c;
(function(n) {
  n.mergeShapes = (s, f) => ({
    ...s,
    ...f
    // second overwrites first
  });
})(c || (c = {}));
const t = u.arrayToEnum([
  "string",
  "nan",
  "number",
  "integer",
  "float",
  "boolean",
  "date",
  "bigint",
  "symbol",
  "function",
  "undefined",
  "null",
  "array",
  "object",
  "unknown",
  "promise",
  "void",
  "never",
  "map",
  "set"
]), y = (n) => {
  switch (typeof n) {
    case "undefined":
      return t.undefined;
    case "string":
      return t.string;
    case "number":
      return Number.isNaN(n) ? t.nan : t.number;
    case "boolean":
      return t.boolean;
    case "function":
      return t.function;
    case "bigint":
      return t.bigint;
    case "symbol":
      return t.symbol;
    case "object":
      return Array.isArray(n) ? t.array : n === null ? t.null : n.then && typeof n.then == "function" && n.catch && typeof n.catch == "function" ? t.promise : typeof Map < "u" && n instanceof Map ? t.map : typeof Set < "u" && n instanceof Set ? t.set : typeof Date < "u" && n instanceof Date ? t.date : t.object;
    default:
      return t.unknown;
  }
};
export {
  t as ZodParsedType,
  y as getParsedType,
  c as objectUtil,
  u as util
};
