// @tiptap/extension-file-handler@3.15.3/dist/index.js
import { Extension as u } from "../../../../../@tiptap_core@2.24.0_@tiptap_pm@2.24.0/node_modules/@tiptap/core/dist/index.js";
import { PluginKey as p, Plugin as h } from "../../../../../prosemirror-state@1.4.3/node_modules/prosemirror-state/dist/index.js";
var g = ({ key: d, editor: o, onPaste: n, onDrop: s, allowedMimeTypes: a }) => new h({
  key: d || new p("fileHandler"),
  props: {
    handleDrop(f, r) {
      var i;
      if (!s || !((i = r.dataTransfer) != null && i.files.length))
        return !1;
      const e = f.posAtCoords({
        left: r.clientX,
        top: r.clientY
      });
      let t = Array.from(r.dataTransfer.files);
      return a && (t = t.filter((l) => a.includes(l.type))), t.length === 0 ? !1 : (r.preventDefault(), r.stopPropagation(), s(o, t, e?.pos || 0), !0);
    },
    handlePaste(f, r) {
      var i;
      if (!n || !((i = r.clipboardData) != null && i.files.length))
        return !1;
      let e = Array.from(r.clipboardData.files);
      const t = r.clipboardData.getData("text/html");
      return a && (e = e.filter((l) => a.includes(l.type))), !(e.length === 0 || (r.preventDefault(), r.stopPropagation(), n(o, e, t), t.length > 0));
    }
  }
}), m = u.create({
  name: "fileHandler",
  addOptions() {
    return {
      onPaste: void 0,
      onDrop: void 0,
      allowedMimeTypes: void 0
    };
  },
  addProseMirrorPlugins() {
    return [
      g({
        key: new p(this.name),
        editor: this.editor,
        allowedMimeTypes: this.options.allowedMimeTypes,
        onDrop: this.options.onDrop,
        onPaste: this.options.onPaste
      })
    ];
  }
});
export {
  g as FileHandlePlugin,
  m as FileHandler
};
