// echarts@6.0.0/lib/component/axisPointer/axisTrigger.js
import { makeInner as H } from "../../util/model.js";
import { makeKey as b } from "./modelHelper.js";
import O from "./findPointFromSeries.js";
import { bind as V, curry as S, each as p, extend as C } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
var B = H();
function j(e, i, a) {
  var n = e.currTrigger, t = [e.x, e.y], d = e, r = e.dispatchAction || V(a.dispatchAction, a), s = i.getComponent("axisPointer").coordSysAxesInfo;
  if (s) {
    w(t) && (t = O({
      seriesIndex: d.seriesIndex,
      // Do not use dataIndexInside from other ec instance.
      // FIXME: auto detect it?
      dataIndex: d.dataIndex
    }, i).point);
    var o = w(t), u = d.axesInfo, l = s.axesInfo, v = n === "leave" || w(t), c = {}, x = {}, f = {
      list: [],
      map: {}
    }, y = {
      showPointer: S(L, x),
      showTooltip: S(E, f)
    };
    p(s.coordSysMap, function(h, T) {
      var g = o || h.containPoint(t);
      p(s.coordSysAxesInfo[T], function(I, M) {
        var A = I.axis, m = R(u, I);
        if (!v && g && (!u || m)) {
          var P = m && m.value;
          P == null && !o && (P = A.pointToData(t)), P != null && k(I, P, y, !1, c);
        }
      });
    });
    var D = {};
    return p(l, function(h, T) {
      var g = h.linkGroup;
      g && !x[T] && p(g.axesInfo, function(I, M) {
        var A = x[M];
        if (I !== h && A) {
          var m = A.value;
          g.mapper && (m = h.axis.scale.parse(g.mapper(m, N(I), N(h)))), D[h.key] = m;
        }
      });
    }), p(D, function(h, T) {
      k(l[T], h, y, !0, c);
    }), G(x, l, c), z(f, t, e, r), F(l, r, a), c;
  }
}
function k(e, i, a, n, t) {
  var d = e.axis;
  if (!(d.scale.isBlank() || !d.containData(i))) {
    if (!e.involveSeries) {
      a.showPointer(e, i);
      return;
    }
    var r = K(i, e), s = r.payloadBatch, o = r.snapToValue;
    s[0] && t.seriesIndex == null && C(t, s[0]), !n && e.snap && d.containData(o) && o != null && (i = o), a.showPointer(e, i, s), a.showTooltip(e, r, o);
  }
}
function K(e, i) {
  var a = i.axis, n = a.dim, t = e, d = [], r = Number.MAX_VALUE, s = -1;
  return p(i.seriesModels, function(o, u) {
    var l = o.getData().mapDimensionsAll(n), v, c;
    if (o.getAxisTooltipData) {
      var x = o.getAxisTooltipData(l, e, a);
      c = x.dataIndices, v = x.nestestValue;
    } else {
      if (c = o.indicesOfNearest(
        n,
        l[0],
        e,
        // Add a threshold to avoid find the wrong dataIndex
        // when data length is not same.
        // false,
        a.type === "category" ? 0.5 : null
      ), !c.length)
        return;
      v = o.getData().get(l[0], c[0]);
    }
    if (!(v == null || !isFinite(v))) {
      var f = e - v, y = Math.abs(f);
      y <= r && ((y < r || f >= 0 && s < 0) && (r = y, s = f, t = v, d.length = 0), p(c, function(D) {
        d.push({
          seriesIndex: o.seriesIndex,
          dataIndexInside: D,
          dataIndex: o.getData().getRawIndex(D)
        });
      }));
    }
  }), {
    payloadBatch: d,
    snapToValue: t
  };
}
function L(e, i, a, n) {
  e[i.key] = {
    value: a,
    payloadBatch: n
  };
}
function E(e, i, a, n) {
  var t = a.payloadBatch, d = i.axis, r = d.model, s = i.axisPointerModel;
  if (!(!i.triggerTooltip || !t.length)) {
    var o = i.coordSys.model, u = b(o), l = e.map[u];
    l || (l = e.map[u] = {
      coordSysId: o.id,
      coordSysIndex: o.componentIndex,
      coordSysType: o.type,
      coordSysMainType: o.mainType,
      dataByAxis: []
    }, e.list.push(l)), l.dataByAxis.push({
      axisDim: d.dim,
      axisIndex: r.componentIndex,
      axisType: r.type,
      axisId: r.id,
      value: n,
      // Caustion: viewHelper.getValueLabel is actually on "view stage", which
      // depends that all models have been updated. So it should not be performed
      // here. Considering axisPointerModel used here is volatile, which is hard
      // to be retrieve in TooltipView, we prepare parameters here.
      valueLabelOpt: {
        precision: s.get(["label", "precision"]),
        formatter: s.get(["label", "formatter"])
      },
      seriesDataIndices: t.slice()
    });
  }
}
function G(e, i, a) {
  var n = a.axesInfo = [];
  p(i, function(t, d) {
    var r = t.axisPointerModel.option, s = e[d];
    s ? (!t.useHandle && (r.status = "show"), r.value = s.value, r.seriesDataIndices = (s.payloadBatch || []).slice()) : !t.useHandle && (r.status = "hide"), r.status === "show" && n.push({
      axisDim: t.axis.dim,
      axisIndex: t.axis.model.componentIndex,
      value: r.value
    });
  });
}
function z(e, i, a, n) {
  if (w(i) || !e.list.length) {
    n({
      type: "hideTip"
    });
    return;
  }
  var t = ((e.list[0].dataByAxis[0] || {}).seriesDataIndices || [])[0] || {};
  n({
    type: "showTip",
    escapeConnect: !0,
    x: i[0],
    y: i[1],
    tooltipOption: a.tooltipOption,
    position: a.position,
    dataIndexInside: t.dataIndexInside,
    dataIndex: t.dataIndex,
    seriesIndex: t.seriesIndex,
    dataByCoordSys: e.list
  });
}
function F(e, i, a) {
  var n = a.getZr(), t = "axisPointerLastHighlights", d = B(n)[t] || {}, r = B(n)[t] = {};
  p(e, function(u, l) {
    var v = u.axisPointerModel.option;
    v.status === "show" && u.triggerEmphasis && p(v.seriesDataIndices, function(c) {
      var x = c.seriesIndex + " | " + c.dataIndex;
      r[x] = c;
    });
  });
  var s = [], o = [];
  p(d, function(u, l) {
    !r[l] && o.push(u);
  }), p(r, function(u, l) {
    !d[l] && s.push(u);
  }), o.length && a.dispatchAction({
    type: "downplay",
    escapeConnect: !0,
    // Not blur others when highlight in axisPointer.
    notBlur: !0,
    batch: o
  }), s.length && a.dispatchAction({
    type: "highlight",
    escapeConnect: !0,
    // Not blur others when highlight in axisPointer.
    notBlur: !0,
    batch: s
  });
}
function R(e, i) {
  for (var a = 0; a < (e || []).length; a++) {
    var n = e[a];
    if (i.axis.dim === n.axisDim && i.axis.model.componentIndex === n.axisIndex)
      return n;
  }
}
function N(e) {
  var i = e.axis.model, a = {}, n = a.axisDim = e.axis.dim;
  return a.axisIndex = a[n + "AxisIndex"] = i.componentIndex, a.axisName = a[n + "AxisName"] = i.name, a.axisId = a[n + "AxisId"] = i.id, a;
}
function w(e) {
  return !e || e[0] == null || isNaN(e[0]) || e[1] == null || isNaN(e[1]);
}
export {
  j as default
};
