// echarts@6.0.0/lib/component/axisPointer/AxisPointerModel.js
import { __extends as l } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import n from "../../model/Component.js";
import t from "../../visual/tokens.js";
var d = (
  /** @class */
  (function(r) {
    l(o, r);
    function o() {
      var e = r !== null && r.apply(this, arguments) || this;
      return e.type = o.type, e;
    }
    return o.type = "axisPointer", o.defaultOption = {
      // 'auto' means that show when triggered by tooltip or handle.
      show: "auto",
      // zlevel: 0,
      z: 50,
      type: "line",
      // axispointer triggered by tootip determine snap automatically,
      // see `modelHelper`.
      snap: !1,
      triggerTooltip: !0,
      triggerEmphasis: !0,
      value: null,
      status: null,
      link: [],
      // Do not set 'auto' here, otherwise global animation: false
      // will not effect at this axispointer.
      animation: null,
      animationDurationUpdate: 200,
      lineStyle: {
        color: t.color.border,
        width: 1,
        type: "dashed"
      },
      shadowStyle: {
        color: t.color.shadowTint
      },
      label: {
        show: !0,
        formatter: null,
        precision: "auto",
        margin: 3,
        color: t.color.neutral00,
        padding: [5, 7, 5, 7],
        backgroundColor: t.color.accent60,
        borderColor: null,
        borderWidth: 0,
        borderRadius: 3
      },
      handle: {
        show: !1,
        // eslint-disable-next-line
        icon: "M10.7,11.9v-1.3H9.3v1.3c-4.9,0.3-8.8,4.4-8.8,9.4c0,5,3.9,9.1,8.8,9.4h1.3c4.9-0.3,8.8-4.4,8.8-9.4C19.5,16.3,15.6,12.2,10.7,11.9z M13.3,24.4H6.7v-1.2h6.6z M13.3,22H6.7v-1.2h6.6z M13.3,19.6H6.7v-1.2h6.6z",
        size: 45,
        // handle margin is from symbol center to axis, which is stable when circular move.
        margin: 50,
        // color: '#1b8bbd'
        // color: '#2f4554'
        color: t.color.accent40,
        // For mobile performance
        throttle: 40
      }
    }, o;
  })(n)
);
export {
  d as default
};
