// echarts@6.0.0/lib/component/axisPointer/AxisPointerView.js
import { __extends as m } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { register as g, unregister as f } from "./globalListener.js";
import u from "../../view/Component.js";
var y = (
  /** @class */
  (function(n) {
    m(t, n);
    function t() {
      var e = n !== null && n.apply(this, arguments) || this;
      return e.type = t.type, e;
    }
    return t.prototype.render = function(e, o, a) {
      var p = o.getComponent("tooltip"), s = e.get("triggerOn") || p && p.get("triggerOn") || "mousemove|click";
      g("axisPointer", a, function(r, i, l) {
        s !== "none" && (r === "leave" || s.indexOf(r) >= 0) && l({
          type: "updateAxisPointer",
          currTrigger: r,
          x: i && i.offsetX,
          y: i && i.offsetY
        });
      });
    }, t.prototype.remove = function(e, o) {
      f("axisPointer", o);
    }, t.prototype.dispose = function(e, o) {
      f("axisPointer", o);
    }, t.type = "axisPointer", t;
  })(u)
);
export {
  y as default
};
