// echarts@6.0.0/lib/component/axisPointer/BaseAxisPointer.js
import { curry as E, clone as x, bind as H, isArray as P, isObject as v, each as w } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import * as z from "../../util/graphic.js";
import { createIcon as S } from "../../util/graphic.js";
import { getAxisInfo as T } from "./modelHelper.js";
import { stop as M } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/event.js";
import { createOrUpdate as I, clear as D } from "../../util/throttle.js";
import { makeInner as L } from "../../util/model.js";
import V from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import C from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Text.js";
import { updateProps as K } from "../../animation/basicTransition.js";
var u = L(), _ = x, p = H, Y = (
  /** @class */
  (function() {
    function r() {
      this._dragging = !1, this.animationThreshold = 15;
    }
    return r.prototype.render = function(e, t, a, i) {
      var o = t.get("value"), l = t.get("status");
      if (this._axisModel = e, this._axisPointerModel = t, this._api = a, !(!i && this._lastValue === o && this._lastStatus === l)) {
        this._lastValue = o, this._lastStatus = l;
        var n = this._group, s = this._handle;
        if (!l || l === "hide") {
          n && n.hide(), s && s.hide();
          return;
        }
        n && n.show(), s && s.show();
        var h = {};
        this.makeElOption(h, o, e, t, a);
        var f = h.graphicKey;
        f !== this._lastGraphicKey && this.clear(a), this._lastGraphicKey = f;
        var A = this._moveAnimation = this.determineAnimation(e, t);
        if (!n)
          n = this._group = new V(), this.createPointerEl(n, h, e, t), this.createLabelEl(n, h, e, t), a.getZr().add(n);
        else {
          var c = E(g, t, A);
          this.updatePointerEl(n, h, c), this.updateLabelEl(n, h, c, t);
        }
        y(n, t, !0), this._renderHandle(o);
      }
    }, r.prototype.remove = function(e) {
      this.clear(e);
    }, r.prototype.dispose = function(e) {
      this.clear(e);
    }, r.prototype.determineAnimation = function(e, t) {
      var a = t.get("animation"), i = e.axis, o = i.type === "category", l = t.get("snap");
      if (!l && !o)
        return !1;
      if (a === "auto" || a == null) {
        var n = this.animationThreshold;
        if (o && i.getBandWidth() > n)
          return !0;
        if (l) {
          var s = T(e).seriesDataCount, h = i.getExtent();
          return Math.abs(h[0] - h[1]) / s > n;
        }
        return !1;
      }
      return a === !0;
    }, r.prototype.makeElOption = function(e, t, a, i, o) {
    }, r.prototype.createPointerEl = function(e, t, a, i) {
      var o = t.pointer;
      if (o) {
        var l = u(e).pointerEl = new z[o.type](_(t.pointer));
        e.add(l);
      }
    }, r.prototype.createLabelEl = function(e, t, a, i) {
      if (t.label) {
        var o = u(e).labelEl = new C(_(t.label));
        e.add(o), m(o, i);
      }
    }, r.prototype.updatePointerEl = function(e, t, a) {
      var i = u(e).pointerEl;
      i && t.pointer && (i.setStyle(t.pointer.style), a(i, {
        shape: t.pointer.shape
      }));
    }, r.prototype.updateLabelEl = function(e, t, a, i) {
      var o = u(e).labelEl;
      o && (o.setStyle(t.label.style), a(o, {
        // Consider text length change in vertical axis, animation should
        // be used on shape, otherwise the effect will be weird.
        // TODOTODO
        // shape: elOption.label.shape,
        x: t.label.x,
        y: t.label.y
      }), m(o, i));
    }, r.prototype._renderHandle = function(e) {
      if (!(this._dragging || !this.updateHandleTransform)) {
        var t = this._axisPointerModel, a = this._api.getZr(), i = this._handle, o = t.getModel("handle"), l = t.get("status");
        if (!o.get("show") || !l || l === "hide") {
          i && a.remove(i), this._handle = null;
          return;
        }
        var n;
        this._handle || (n = !0, i = this._handle = S(o.get("icon"), {
          cursor: "move",
          draggable: !0,
          onmousemove: function(h) {
            M(h.event);
          },
          onmousedown: p(this._onHandleDragMove, this, 0, 0),
          drift: p(this._onHandleDragMove, this),
          ondragend: p(this._onHandleDragEnd, this)
        }), a.add(i)), y(i, t, !1), i.setStyle(o.getItemStyle(null, ["color", "borderColor", "borderWidth", "opacity", "shadowColor", "shadowBlur", "shadowOffsetX", "shadowOffsetY"]));
        var s = o.get("size");
        P(s) || (s = [s, s]), i.scaleX = s[0] / 2, i.scaleY = s[1] / 2, I(this, "_doDispatchAxisPointer", o.get("throttle") || 0, "fixRate"), this._moveHandleToValue(e, n);
      }
    }, r.prototype._moveHandleToValue = function(e, t) {
      g(this._axisPointerModel, !t && this._moveAnimation, this._handle, d(this.getHandleTransform(e, this._axisModel, this._axisPointerModel)));
    }, r.prototype._onHandleDragMove = function(e, t) {
      var a = this._handle;
      if (a) {
        this._dragging = !0;
        var i = this.updateHandleTransform(d(a), [e, t], this._axisModel, this._axisPointerModel);
        this._payloadInfo = i, a.stopAnimation(), a.attr(d(i)), u(a).lastProp = null, this._doDispatchAxisPointer();
      }
    }, r.prototype._doDispatchAxisPointer = function() {
      var e = this._handle;
      if (e) {
        var t = this._payloadInfo, a = this._axisModel;
        this._api.dispatchAction({
          type: "updateAxisPointer",
          x: t.cursorPoint[0],
          y: t.cursorPoint[1],
          tooltipOption: t.tooltipOption,
          axesInfo: [{
            axisDim: a.axis.dim,
            axisIndex: a.componentIndex
          }]
        });
      }
    }, r.prototype._onHandleDragEnd = function() {
      this._dragging = !1;
      var e = this._handle;
      if (e) {
        var t = this._axisPointerModel.get("value");
        this._moveHandleToValue(t), this._api.dispatchAction({
          type: "hideTip"
        });
      }
    }, r.prototype.clear = function(e) {
      this._lastValue = null, this._lastStatus = null;
      var t = e.getZr(), a = this._group, i = this._handle;
      t && a && (this._lastGraphicKey = null, a && t.remove(a), i && t.remove(i), this._group = null, this._handle = null, this._payloadInfo = null), D(this, "_doDispatchAxisPointer");
    }, r.prototype.doClear = function() {
    }, r.prototype.buildLabel = function(e, t, a) {
      return a = a || 0, {
        x: e[a],
        y: e[1 - a],
        width: t[a],
        height: t[1 - a]
      };
    }, r;
  })()
);
function g(r, e, t, a) {
  b(u(t).lastProp, a) || (u(t).lastProp = a, e ? K(t, a, r) : (t.stopAnimation(), t.attr(a)));
}
function b(r, e) {
  if (v(r) && v(e)) {
    var t = !0;
    return w(e, function(a, i) {
      t = t && b(r[i], a);
    }), !!t;
  } else
    return r === e;
}
function m(r, e) {
  r[e.get(["label", "show"]) ? "show" : "hide"]();
}
function d(r) {
  return {
    x: r.x || 0,
    y: r.y || 0,
    rotation: r.rotation || 0
  };
}
function y(r, e, t) {
  var a = e.get("z"), i = e.get("zlevel");
  r && r.traverse(function(o) {
    o.type !== "group" && (a != null && (o.z = a), i != null && (o.zlevel = i), o.silent = t);
  });
}
export {
  Y as default
};
