// echarts@6.0.0/lib/component/axisPointer/CartesianAxisPointer.js
import { __extends as h } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import y from "./BaseAxisPointer.js";
import { buildElStyle as b, buildCartesianSingleLabelElOption as P, getTransformedPosition as A, makeRectShape as C, makeLineShape as I } from "./viewHelper.js";
import { layout as m } from "../../coord/cartesian/cartesianAxisHelper.js";
var R = (
  /** @class */
  (function(e) {
    h(r, e);
    function r() {
      return e !== null && e.apply(this, arguments) || this;
    }
    return r.prototype.makeElOption = function(t, a, l, s, n) {
      var u = l.axis, g = u.grid, p = s.get("type"), i = c(g, u).getOtherAxis(u).getGlobalExtent(), o = u.toGlobalCoord(u.dataToCoord(a, !0));
      if (p && p !== "none") {
        var v = b(s), d = E[p](u, o, i);
        d.style = v, t.graphicKey = d.type, t.pointer = d;
      }
      var x = m(g.getRect(), l);
      P(a, t, x, l, s, n);
    }, r.prototype.getHandleTransform = function(t, a, l) {
      var s = m(a.axis.grid.getRect(), a, {
        labelInside: !1
      });
      s.labelMargin = l.get(["handle", "margin"]);
      var n = A(a.axis, t, s);
      return {
        x: n[0],
        y: n[1],
        rotation: s.rotation + (s.labelDirection < 0 ? Math.PI : 0)
      };
    }, r.prototype.updateHandleTransform = function(t, a, l, s) {
      var n = l.axis, u = n.grid, g = n.getGlobalExtent(!0), p = c(u, n).getOtherAxis(n).getGlobalExtent(), i = n.dim === "x" ? 0 : 1, o = [t.x, t.y];
      o[i] += a[i], o[i] = Math.min(g[1], o[i]), o[i] = Math.max(g[0], o[i]);
      var v = (p[1] + p[0]) / 2, d = [v, v];
      d[i] = o[i];
      var x = [{
        verticalAlign: "middle"
      }, {
        align: "center"
      }];
      return {
        x: o[0],
        y: o[1],
        rotation: t.rotation,
        cursorPoint: d,
        tooltipOption: x[i]
      };
    }, r;
  })(y)
);
function c(e, r) {
  var t = {};
  return t[r.dim + "AxisIndex"] = r.index, e.getCartesian(t);
}
var E = {
  line: function(e, r, t) {
    var a = I([r, t[0]], [r, t[1]], f(e));
    return {
      type: "Line",
      subPixelOptimize: !0,
      shape: a
    };
  },
  shadow: function(e, r, t) {
    var a = Math.max(1, e.getBandWidth()), l = t[1] - t[0];
    return {
      type: "Rect",
      shape: C([r - a / 2, t[0]], [a, l], f(e))
    };
  }
};
function f(e) {
  return e.dim === "x" ? 0 : 1;
}
export {
  R as default
};
