// echarts@6.0.0/lib/component/axisPointer/globalListener.js
import { curry as s, each as l } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import u from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/env.js";
import { makeInner as f } from "../../util/model.js";
var o = f(), p = l;
function b(i, r, e) {
  if (!u.node) {
    var n = r.getZr();
    o(n).records || (o(n).records = {}), v(n, r);
    var t = o(n).records[i] || (o(n).records[i] = {});
    t.handler = e;
  }
}
function v(i, r) {
  if (o(i).initialized)
    return;
  o(i).initialized = !0, e("click", s(d, "click")), e("mousemove", s(d, "mousemove")), e("globalout", g);
  function e(n, t) {
    i.on(n, function(h) {
      var c = A(r);
      p(o(i).records, function(a) {
        a && t(a, h, c.dispatchAction);
      }), m(c.pendings, r);
    });
  }
}
function m(i, r) {
  var e = i.showTip.length, n = i.hideTip.length, t;
  e ? t = i.showTip[e - 1] : n && (t = i.hideTip[n - 1]), t && (t.dispatchAction = null, r.dispatchAction(t));
}
function g(i, r, e) {
  i.handler("leave", null, e);
}
function d(i, r, e, n) {
  r.handler(i, e, n);
}
function A(i) {
  var r = {
    showTip: [],
    hideTip: []
  }, e = function(n) {
    var t = r[n.type];
    t ? t.push(n) : (n.dispatchAction = e, i.dispatchAction(n));
  };
  return {
    dispatchAction: e,
    pendings: r
  };
}
function k(i, r) {
  if (!u.node) {
    var e = r.getZr(), n = (o(e).records || {})[i];
    n && (o(e).records[i] = null);
  }
}
export {
  b as register,
  k as unregister
};
