// echarts@6.0.0/lib/component/axisPointer/modelHelper.js
import G from "../../model/Model.js";
import { each as g, curry as M, clone as U, defaults as z, isArray as B, indexOf as L } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
function Q(e, t) {
  var o = {
    /**
     * key: makeKey(axis.model)
     * value: {
     *      axis,
     *      coordSys,
     *      axisPointerModel,
     *      triggerTooltip,
     *      triggerEmphasis,
     *      involveSeries,
     *      snap,
     *      seriesModels,
     *      seriesDataCount
     * }
     */
    axesInfo: {},
    seriesInvolved: !1,
    /**
     * key: makeKey(coordSys.model)
     * value: Object: key makeKey(axis.model), value: axisInfo
     */
    coordSysAxesInfo: {},
    coordSysMap: {}
  };
  return N(o, e, t), o.seriesInvolved && j(o, e), o;
}
function N(e, t, o) {
  var i = t.getComponent("tooltip"), a = t.getComponent("axisPointer"), l = a.get("link", !0) || [], n = [];
  g(o.getCoordinateSystems(), function(s) {
    if (!s.axisPointerEnabled)
      return;
    var r = h(s.model), u = e.coordSysAxesInfo[r] = {};
    e.coordSysMap[r] = s;
    var p = s.model, f = p.getModel("tooltip", i);
    if (g(s.getAxes(), M(A, !1, null)), s.getTooltipAxes && i && f.get("show")) {
      var x = f.get("trigger") === "axis", I = f.get(["axisPointer", "type"]) === "cross", E = s.getTooltipAxes(f.get(["axisPointer", "axis"]));
      (x || I) && g(E.baseAxes, M(A, I ? "cross" : !0, x)), I && g(E.otherAxes, M(A, "cross", !1));
    }
    function A(S, c, d) {
      var v = d.model.getModel("axisPointer", a), D = v.get("show");
      if (!(!D || D === "auto" && !S && !C(v))) {
        c == null && (c = v.get("triggerTooltip")), v = S ? O(d, f, a, t, S, c) : v;
        var k = v.get("snap"), K = v.get("triggerEmphasis"), m = h(d.model), T = c || k || d.type === "category", P = e.axesInfo[m] = {
          key: m,
          axis: d,
          coordSys: s,
          axisPointerModel: v,
          triggerTooltip: c,
          triggerEmphasis: K,
          involveSeries: T,
          snap: k,
          useHandle: C(v),
          seriesModels: [],
          linkGroup: null
        };
        u[m] = P, e.seriesInvolved = e.seriesInvolved || T;
        var y = q(l, d);
        if (y != null) {
          var w = n[y] || (n[y] = {
            axesInfo: {}
          });
          w.axesInfo[m] = P, w.mapper = l[y].mapper, P.linkGroup = w;
        }
      }
    }
  });
}
function O(e, t, o, i, a, l) {
  var n = t.getModel("axisPointer"), s = ["type", "snap", "lineStyle", "shadowStyle", "label", "animation", "animationDurationUpdate", "animationEasingUpdate", "z"], r = {};
  g(s, function(x) {
    r[x] = U(n.get(x));
  }), r.snap = e.type !== "category" && !!l, n.get("type") === "cross" && (r.type = "line");
  var u = r.label || (r.label = {});
  if (u.show == null && (u.show = !1), a === "cross") {
    var p = n.get(["label", "show"]);
    if (u.show = p ?? !0, !l) {
      var f = r.lineStyle = n.get("crossStyle");
      f && z(u, f.textStyle);
    }
  }
  return e.model.getModel("axisPointer", new G(r, o, i));
}
function j(e, t) {
  t.eachSeries(function(o) {
    var i = o.coordinateSystem, a = o.get(["tooltip", "trigger"], !0), l = o.get(["tooltip", "show"], !0);
    !i || !i.model || a === "none" || a === !1 || a === "item" || l === !1 || o.get(["axisPointer", "show"], !0) === !1 || g(e.coordSysAxesInfo[h(i.model)], function(n) {
      var s = n.axis;
      i.getAxis(s.dim) === s && (n.seriesModels.push(o), n.seriesDataCount == null && (n.seriesDataCount = 0), n.seriesDataCount += o.getData().count());
    });
  });
}
function q(e, t) {
  for (var o = t.model, i = t.dim, a = 0; a < e.length; a++) {
    var l = e[a] || {};
    if (b(l[i + "AxisId"], o.id) || b(l[i + "AxisIndex"], o.componentIndex) || b(l[i + "AxisName"], o.name))
      return a;
  }
}
function b(e, t) {
  return e === "all" || B(e) && L(e, t) >= 0 || e === t;
}
function R(e) {
  var t = H(e);
  if (t) {
    var o = t.axisPointerModel, i = t.axis.scale, a = o.option, l = o.get("status"), n = o.get("value");
    n != null && (n = i.parse(n));
    var s = C(o);
    l == null && (a.status = s ? "show" : "hide");
    var r = i.getExtent().slice();
    r[0] > r[1] && r.reverse(), // Pick a value on axis when initializing.
    (n == null || n > r[1]) && (n = r[1]), n < r[0] && (n = r[0]), a.value = n, s && (a.status = t.axis.scale.isBlank() ? "hide" : "show");
  }
}
function H(e) {
  var t = (e.ecModel.getComponent("axisPointer") || {}).coordSysAxesInfo;
  return t && t.axesInfo[h(e)];
}
function W(e) {
  var t = H(e);
  return t && t.axisPointerModel;
}
function C(e) {
  return !!e.get(["handle", "show"]);
}
function h(e) {
  return e.type + "||" + e.id;
}
export {
  Q as collect,
  R as fixValue,
  H as getAxisInfo,
  W as getAxisPointerModel,
  h as makeKey
};
