// echarts@6.0.0/lib/component/axisPointer/PolarAxisPointer.js
import { __extends as u } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import A from "./BaseAxisPointer.js";
import { applyTransform as P } from "../../util/graphic.js";
import { buildElStyle as b, buildLabelElOption as T, makeSectorShape as f, makeLineShape as S } from "./viewHelper.js";
import { rotate as E, create as I, translate as L } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/matrix.js";
import B from "../axis/AxisBuilder.js";
var q = (
  /** @class */
  (function(n) {
    u(t, n);
    function t() {
      return n !== null && n.apply(this, arguments) || this;
    }
    return t.prototype.makeElOption = function(a, e, i, r, d) {
      var o = i.axis;
      o.dim === "angle" && (this.animationThreshold = Math.PI / 18);
      var c = o.polar, s = c.getOtherAxis(o), h = s.getExtent(), m = o.dataToCoord(e), v = r.get("type");
      if (v && v !== "none") {
        var y = b(r), g = C[v](o, c, m, h);
        g.style = y, a.graphicKey = g.type, a.pointer = g;
      }
      var x = r.get(["label", "margin"]), l = k(e, i, r, c, x);
      T(a, i, r, d, l);
    }, t;
  })(A)
);
function k(n, t, a, e, i) {
  var r = t.axis, d = r.dataToCoord(n), o = e.getAngleAxis().getExtent()[0];
  o = o / 180 * Math.PI;
  var c = e.getRadiusAxis().getExtent(), s, h, m;
  if (r.dim === "radius") {
    var v = I();
    E(v, v, o), L(v, v, [e.cx, e.cy]), s = P([d, -i], v);
    var y = t.getModel("axisLabel").get("rotate") || 0, g = B.innerTextLayout(o, y * Math.PI / 180, -1);
    h = g.textAlign, m = g.textVerticalAlign;
  } else {
    var x = c[1];
    s = e.coordToPoint([x + i, d]);
    var l = e.cx, p = e.cy;
    h = Math.abs(s[0] - l) / x < 0.3 ? "center" : s[0] > l ? "left" : "right", m = Math.abs(s[1] - p) / x < 0.3 ? "middle" : s[1] > p ? "top" : "bottom";
  }
  return {
    position: s,
    align: h,
    verticalAlign: m
  };
}
var C = {
  line: function(n, t, a, e) {
    return n.dim === "angle" ? {
      type: "Line",
      shape: S(t.coordToPoint([e[0], a]), t.coordToPoint([e[1], a]))
    } : {
      type: "Circle",
      shape: {
        cx: t.cx,
        cy: t.cy,
        r: a
      }
    };
  },
  shadow: function(n, t, a, e) {
    var i = Math.max(1, n.getBandWidth()), r = Math.PI / 180;
    return n.dim === "angle" ? {
      type: "Sector",
      shape: f(
        t.cx,
        t.cy,
        e[0],
        e[1],
        // In ECharts y is negative if angle is positive
        (-a - i / 2) * r,
        (-a + i / 2) * r
      )
    } : {
      type: "Sector",
      shape: f(t.cx, t.cy, a - i / 2, a + i / 2, 0, Math.PI * 2)
    };
  }
};
export {
  q as default
};
