// echarts@6.0.0/lib/component/axisPointer/SingleAxisPointer.js
import { __extends as x } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import S from "./BaseAxisPointer.js";
import { buildElStyle as P, buildCartesianSingleLabelElOption as b, getTransformedPosition as T, makeRectShape as A, makeLineShape as E } from "./viewHelper.js";
import { layout as f } from "../../coord/single/singleAxisHelper.js";
var h = ["x", "y"], H = ["width", "height"], W = (
  /** @class */
  (function(e) {
    x(r, e);
    function r() {
      return e !== null && e.apply(this, arguments) || this;
    }
    return r.prototype.makeElOption = function(t, n, o, l, u) {
      var s = o.axis, a = s.coordinateSystem, c = d(a, 1 - m(s)), i = a.dataToPoint(n)[0], p = l.get("type");
      if (p && p !== "none") {
        var y = P(l), v = I[p](s, i, c);
        v.style = y, t.graphicKey = v.type, t.pointer = v;
      }
      var g = f(o);
      b(n, t, g, o, l, u);
    }, r.prototype.getHandleTransform = function(t, n, o) {
      var l = f(n, {
        labelInside: !1
      });
      l.labelMargin = o.get(["handle", "margin"]);
      var u = T(n.axis, t, l);
      return {
        x: u[0],
        y: u[1],
        rotation: l.rotation + (l.labelDirection < 0 ? Math.PI : 0)
      };
    }, r.prototype.updateHandleTransform = function(t, n, o, l) {
      var u = o.axis, s = u.coordinateSystem, a = m(u), c = d(s, a), i = [t.x, t.y];
      i[a] += n[a], i[a] = Math.min(c[1], i[a]), i[a] = Math.max(c[0], i[a]);
      var p = d(s, 1 - a), y = (p[1] + p[0]) / 2, v = [y, y];
      return v[a] = i[a], {
        x: i[0],
        y: i[1],
        rotation: t.rotation,
        cursorPoint: v,
        tooltipOption: {
          verticalAlign: "middle"
        }
      };
    }, r;
  })(S)
), I = {
  line: function(e, r, t) {
    var n = E([r, t[0]], [r, t[1]], m(e));
    return {
      type: "Line",
      subPixelOptimize: !0,
      shape: n
    };
  },
  shadow: function(e, r, t) {
    var n = e.getBandWidth(), o = t[1] - t[0];
    return {
      type: "Rect",
      shape: A([r - n / 2, t[0]], [n, o], m(e))
    };
  }
};
function m(e) {
  return e.isHorizontal() ? 0 : 1;
}
function d(e, r) {
  var t = e.getRect();
  return [t[h[r]], t[h[r]] + t[H[r]]];
}
export {
  W as default
};
