// echarts@6.0.0/lib/component/axisPointer/viewHelper.js
import { each as h, isString as p, isFunction as y } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { applyTransform as S } from "../../util/graphic.js";
import { getBoundingRect as A } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/contain/text.js";
import { normalizeCssArray as L } from "../../util/format.js";
import { rotate as C, create as k, translate as T } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/matrix.js";
import { getAxisRawValue as w } from "../../coord/axisHelper.js";
import R from "../axis/AxisBuilder.js";
import { createTextStyle as B } from "../../label/labelStyle.js";
function K(t) {
  var r = t.get("type"), e = t.getModel(r + "Style"), a;
  return r === "line" ? (a = e.getLineStyle(), a.fill = null) : r === "shadow" && (a = e.getAreaStyle(), a.stroke = null), a;
}
function E(t, r, e, a, i) {
  var n = e.get("value"), l = V(n, r.axis, r.ecModel, e.get("seriesDataIndices"), {
    precision: e.get(["label", "precision"]),
    formatter: e.get(["label", "formatter"])
  }), o = e.getModel("label"), c = L(o.get("padding") || 0), s = o.getFont(), v = A(l, s), g = i.position, u = v.width + c[1] + c[3], m = v.height + c[0] + c[2], d = i.align;
  d === "right" && (g[0] -= u), d === "center" && (g[0] -= u / 2);
  var b = i.verticalAlign;
  b === "bottom" && (g[1] -= m), b === "middle" && (g[1] -= m / 2), O(g, u, m, a);
  var f = o.get("backgroundColor");
  (!f || f === "auto") && (f = r.get(["axisLine", "lineStyle", "color"])), t.label = {
    // shape: {x: 0, y: 0, width: width, height: height, r: labelModel.get('borderRadius')},
    x: g[0],
    y: g[1],
    style: B(o, {
      text: l,
      font: s,
      fill: o.getTextColor(),
      padding: c,
      backgroundColor: f
    }),
    // Label should be over axisPointer.
    z2: 10
  };
}
function O(t, r, e, a) {
  var i = a.getWidth(), n = a.getHeight();
  t[0] = Math.min(t[0] + r, i) - r, t[1] = Math.min(t[1] + e, n) - e, t[0] = Math.max(t[0], 0), t[1] = Math.max(t[1], 0);
}
function V(t, r, e, a, i) {
  t = r.scale.parse(t);
  var n = r.scale.getLabel({
    value: t
  }, {
    // If `precision` is set, width can be fixed (like '12.00500'), which
    // helps to debounce when when moving label.
    precision: i.precision
  }), l = i.formatter;
  if (l) {
    var o = {
      value: w(r, {
        value: t
      }),
      axisDimension: r.dim,
      axisIndex: r.index,
      seriesData: []
    };
    h(a, function(c) {
      var s = e.getSeriesByIndex(c.seriesIndex), v = c.dataIndexInside, g = s && s.getDataParams(v);
      g && o.seriesData.push(g);
    }), p(l) ? n = l.replace("{value}", n) : y(l) && (n = l(o));
  }
  return n;
}
function z(t, r, e) {
  var a = k();
  return C(a, a, e.rotation), T(a, a, e.position), S([t.dataToCoord(r), (e.labelOffset || 0) + (e.labelDirection || 1) * (e.labelMargin || 0)], a);
}
function N(t, r, e, a, i, n) {
  var l = R.innerTextLayout(e.rotation, 0, e.labelDirection);
  e.labelMargin = i.get(["label", "margin"]), E(r, a, i, n, {
    position: z(a.axis, t, e),
    align: l.textAlign,
    verticalAlign: l.textVerticalAlign
  });
}
function Q(t, r, e) {
  return e = e || 0, {
    x1: t[e],
    y1: t[1 - e],
    x2: r[e],
    y2: r[1 - e]
  };
}
function U(t, r, e) {
  return e = e || 0, {
    x: t[e],
    y: t[1 - e],
    width: r[e],
    height: r[1 - e]
  };
}
function X(t, r, e, a, i, n) {
  return {
    cx: t,
    cy: r,
    r0: e,
    r: a,
    startAngle: i,
    endAngle: n,
    clockwise: !0
  };
}
export {
  N as buildCartesianSingleLabelElOption,
  K as buildElStyle,
  E as buildLabelElOption,
  z as getTransformedPosition,
  V as getValueLabel,
  Q as makeLineShape,
  U as makeRectShape,
  X as makeSectorShape
};
