// motion@12.17.0/dist/es/framer-motion/dist/es/animation/sequence/create.js
import { resolveSubjects as J } from "../animate/resolve-subjects.js";
import { calculateRepeatDuration as Q } from "./utils/calc-repeat-duration.js";
import { calcNextTime as q } from "./utils/calc-time.js";
import { addKeyframes as U } from "./utils/edit.js";
import { normalizeTimes as W } from "./utils/normalize-times.js";
import { compareByTime as Y } from "./utils/sort.js";
import { isMotionValue as Z } from "../../../../../motion-dom/dist/es/value/utils/is-motion-value.js";
import { defaultOffset as $ } from "../../../../../motion-dom/dist/es/animation/keyframes/offsets/default.js";
import { isGenerator as ee } from "../../../../../motion-dom/dist/es/animation/generators/utils/is-generator.js";
import { createGeneratorEasing as te } from "../../../../../motion-dom/dist/es/animation/generators/utils/create-generator-easing.js";
import { fillOffset as ne } from "../../../../../motion-dom/dist/es/animation/keyframes/offsets/fill.js";
import { invariant as oe } from "../../../../../motion-utils/dist/es/errors.js";
import { getEasingForSegment as re } from "../../../../../motion-utils/dist/es/easing/utils/get-easing-for-segment.js";
import { progress as se } from "../../../../../motion-utils/dist/es/progress.js";
import { secondsToMilliseconds as ie } from "../../../../../motion-utils/dist/es/time-conversion.js";
const ae = "easeInOut", fe = 20;
function Ke(e, { defaultTransition: n = {}, ...B } = {}, C, P) {
  const X = n.duration || 0.3, M = /* @__PURE__ */ new Map(), k = /* @__PURE__ */ new Map(), _ = {}, v = /* @__PURE__ */ new Map();
  let O = 0, p = 0, x = 0;
  for (let h = 0; h < e.length; h++) {
    const s = e[h];
    if (typeof s == "string") {
      v.set(s, p);
      continue;
    } else if (!Array.isArray(s)) {
      v.set(s.name, q(p, s.at, O, v));
      continue;
    }
    let [g, i, o = {}] = s;
    o.at !== void 0 && (p = q(p, o.at, O, v));
    let r = 0;
    const y = (a, f, m, d = 0, b = 0) => {
      const t = ce(a), { delay: K = 0, times: c = $(t), type: I = "keyframes", repeat: A, repeatType: pe, repeatDelay: ge = 0, ...H } = f;
      let { ease: l = n.ease || "easeOut", duration: u } = f;
      const V = typeof K == "function" ? K(d, b) : K, w = t.length, G = ee(I) ? I : P?.[I];
      if (w <= 2 && G) {
        let S = 100;
        if (w === 2 && le(t)) {
          const T = t[1] - t[0];
          S = Math.abs(T);
        }
        const D = { ...H };
        u !== void 0 && (D.duration = ie(u));
        const j = te(D, S, G);
        l = j.ease, u = j.duration;
      }
      u ?? (u = X);
      const L = p + V;
      c.length === 1 && c[0] === 0 && (c[1] = 1);
      const N = c.length - t.length;
      if (N > 0 && ne(c, N), t.length === 1 && t.unshift(null), A) {
        oe(A < fe, "Repeat count too high, must be less than 20"), u = Q(u, A);
        const S = [...t], D = [...c];
        l = Array.isArray(l) ? [...l] : [l];
        const j = [...l];
        for (let T = 0; T < A; T++) {
          t.push(...S);
          for (let E = 0; E < S.length; E++)
            c.push(D[E] + (T + 1)), l.push(E === 0 ? "linear" : re(j, E - 1));
        }
        W(c, A);
      }
      const R = L + u;
      U(m, t, l, c, L, R), r = Math.max(V + u, r), x = Math.max(R, x);
    };
    if (Z(g)) {
      const a = F(g, k);
      y(i, o, z("default", a));
    } else {
      const a = J(g, i, C, _), f = a.length;
      for (let m = 0; m < f; m++) {
        i = i, o = o;
        const d = a[m], b = F(d, k);
        for (const t in i)
          y(i[t], ue(o, t), z(t, b), m, f);
      }
    }
    O = p, p += r;
  }
  return k.forEach((h, s) => {
    for (const g in h) {
      const i = h[g];
      i.sort(Y);
      const o = [], r = [], y = [];
      for (let f = 0; f < i.length; f++) {
        const { at: m, value: d, easing: b } = i[f];
        o.push(d), r.push(se(0, x, m)), y.push(b || "easeOut");
      }
      r[0] !== 0 && (r.unshift(0), o.unshift(o[0]), y.unshift(ae)), r[r.length - 1] !== 1 && (r.push(1), o.push(null)), M.has(s) || M.set(s, {
        keyframes: {},
        transition: {}
      });
      const a = M.get(s);
      a.keyframes[g] = o, a.transition[g] = {
        ...n,
        duration: x,
        ease: y,
        times: r,
        ...B
      };
    }
  }), M;
}
function F(e, n) {
  return !n.has(e) && n.set(e, {}), n.get(e);
}
function z(e, n) {
  return n[e] || (n[e] = []), n[e];
}
function ce(e) {
  return Array.isArray(e) ? e : [e];
}
function ue(e, n) {
  return e && e[n] ? {
    ...e,
    ...e[n]
  } : { ...e };
}
const me = (e) => typeof e == "number", le = (e) => e.every(me);
export {
  Ke as createAnimationsFromSequence,
  ue as getValueTransition
};
