// echarts@6.0.0/lib/chart/bar/BarSeries.js
import { __extends as l } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import i from "./BaseBarSeries.js";
import a from "../helper/createSeriesData.js";
import { inheritDefaultOption as n } from "../../util/component.js";
import d from "../../visual/tokens.js";
var c = (
  /** @class */
  (function(o) {
    l(e, o);
    function e() {
      var r = o !== null && o.apply(this, arguments) || this;
      return r.type = e.type, r;
    }
    return e.prototype.getInitialData = function() {
      return a(null, this, {
        useEncodeDefaulter: !0,
        createInvertedIndices: !!this.get("realtimeSort", !0) || null
      });
    }, e.prototype.getProgressive = function() {
      return this.get("large") ? this.get("progressive") : !1;
    }, e.prototype.getProgressiveThreshold = function() {
      var r = this.get("progressiveThreshold"), t = this.get("largeThreshold");
      return t > r && (r = t), r;
    }, e.prototype.brushSelector = function(r, t, s) {
      return s.rect(t.getItemLayout(r));
    }, e.type = "series.bar", e.dependencies = ["grid", "polar"], e.defaultOption = n(i.defaultOption, {
      // If clipped
      // Only available on cartesian2d
      clip: !0,
      roundCap: !1,
      showBackground: !1,
      backgroundStyle: {
        color: "rgba(180, 180, 180, 0.2)",
        borderColor: null,
        borderWidth: 0,
        borderType: "solid",
        borderRadius: 0,
        shadowBlur: 0,
        shadowColor: null,
        shadowOffsetX: 0,
        shadowOffsetY: 0,
        opacity: 1
      },
      select: {
        itemStyle: {
          borderColor: d.color.primary,
          borderWidth: 2
        }
      },
      realtimeSort: !1
    }), e;
  })(i)
);
export {
  c as default
};
