// echarts@6.0.0/lib/chart/bar/BarView.js
import { __extends as ee } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import ue from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Path.js";
import fe from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import { map as ge, extend as te, each as de } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { traverseElements as le } from "../../util/graphic.js";
import { getECData as F } from "../../util/innerStore.js";
import { toggleHoverEmphasis as ce, setStatesStylesFromModel as ve } from "../../util/states.js";
import { labelInner as me, getLabelStatesModels as _e, setLabelStyle as be, setLabelValueAnimation as ye } from "../../label/labelStyle.js";
import { throttle as Ae } from "../../util/throttle.js";
import { createClipPath as xe } from "../helper/createClipPathFromCoordSys.js";
import U from "../../util/shape/sausage.js";
import we from "../../view/Chart.js";
import { isCoordinateSystemType as re } from "../../coord/CoordinateSystem.js";
import { getDefaultLabel as Se, getDefaultInterpolatedLabel as De } from "../helper/labelHelper.js";
import { warn as B } from "../../util/log.js";
import { createSectorCalculateTextPosition as Le, setSectorTextRotation as Ce } from "../../label/sectorLabel.js";
import { initProps as I, updateProps as L, removeElementWithFadeOut as E, saveOldStyle as ke } from "../../animation/basicTransition.js";
import { getSectorCornerRadius as Ie } from "../helper/sectorHelper.js";
import ae from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Sector.js";
import ie from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Rect.js";
var R = Math.max, W = Math.min;
function Pe(i, r) {
  var e = i.getArea && i.getArea();
  if (re(i, "cartesian2d")) {
    var t = i.getBaseAxis();
    if (t.type !== "category" || !t.onBand) {
      var a = r.getLayout("bandWidth");
      t.isHorizontal() ? (e.x -= a, e.width += a * 2) : (e.y -= a, e.height += a * 2);
    }
  }
  return e;
}
var nt = (
  /** @class */
  (function(i) {
    ee(r, i);
    function r() {
      var e = i.call(this) || this;
      return e.type = r.type, e._isFirstFrame = !0, e;
    }
    return r.prototype.render = function(e, t, a, n) {
      this._model = e, this._removeOnRenderedListener(a), this._updateDrawMode(e);
      var o = e.get("coordinateSystem");
      o === "cartesian2d" || o === "polar" ? (this._progressiveEls = null, this._isLargeDraw ? this._renderLarge(e, t, a) : this._renderNormal(e, t, a, n)) : process.env.NODE_ENV !== "production" && B("Only cartesian2d and polar supported for bar.");
    }, r.prototype.incrementalPrepareRender = function(e) {
      this._clear(), this._updateDrawMode(e), this._updateLargeClip(e);
    }, r.prototype.incrementalRender = function(e, t) {
      this._progressiveEls = [], this._incrementalRenderLarge(e, t);
    }, r.prototype.eachRendered = function(e) {
      le(this._progressiveEls || this.group, e);
    }, r.prototype._updateDrawMode = function(e) {
      var t = e.pipelineContext.large;
      (this._isLargeDraw == null || t !== this._isLargeDraw) && (this._isLargeDraw = t, this._clear());
    }, r.prototype._renderNormal = function(e, t, a, n) {
      var o = this.group, s = e.getData(), u = this._data, h = e.coordinateSystem, p = h.getBaseAxis(), f;
      h.type === "cartesian2d" ? f = p.isHorizontal() : h.type === "polar" && (f = p.dim === "angle");
      var m = e.isAnimationEnabled() ? e : null, c = Ve(e, h);
      c && this._enableRealtimeSort(c, s, a);
      var A = e.get("clip", !0) || c, S = Pe(h, s);
      o.removeClipPath();
      var x = e.get("roundCap", !0), w = e.get("showBackground", !0), y = e.getModel("backgroundStyle"), C = y.get("borderRadius") || 0, D = [], G = this._backgroundEls, T = n && n.isInitSort, X = n && n.type === "changeAxisOrder";
      function z(g) {
        var _ = k[h.type](s, g);
        if (!_)
          return null;
        var l = Fe(h, f, _);
        return l.useStyle(y.getItemStyle()), h.type === "cartesian2d" ? l.setShape("r", C) : l.setShape("cornerRadius", C), D[g] = l, l;
      }
      s.diff(u).add(function(g) {
        var _ = s.getItemModel(g), l = k[h.type](s, g, _);
        if (l && (w && z(g), !(!s.hasValue(g) || !J[h.type](l)))) {
          var b = !1;
          A && (b = Y[h.type](S, l));
          var v = Z[h.type](e, s, g, l, f, m, p.model, !1, x);
          c && (v.forceLabelAnimation = !0), K(v, s, g, _, l, e, f, h.type === "polar"), T ? v.attr({
            shape: l
          }) : c ? j(c, m, v, l, g, f, !1, !1) : I(v, {
            shape: l
          }, e, g), s.setItemGraphicEl(g, v), o.add(v), v.ignore = b;
        }
      }).update(function(g, _) {
        var l = s.getItemModel(g), b = k[h.type](s, g, l);
        if (b) {
          if (w) {
            var v = void 0;
            G.length === 0 ? v = z(_) : (v = G[_], v.useStyle(y.getItemStyle()), h.type === "cartesian2d" ? v.setShape("r", C) : v.setShape("cornerRadius", C), D[g] = v);
            var oe = k[h.type](s, g), he = se(f, oe, h);
            L(v, {
              shape: he
            }, m, g);
          }
          var d = u.getItemGraphicEl(_);
          if (!s.hasValue(g) || !J[h.type](b)) {
            o.remove(d);
            return;
          }
          var N = !1;
          A && (N = Y[h.type](S, b), N && o.remove(d));
          var pe = d && (d.type === "sector" && x || d.type === "sausage" && !x);
          if (pe && (d && E(d, e, _), d = null), d ? ke(d) : d = Z[h.type](e, s, g, b, f, m, p.model, !0, x), c && (d.forceLabelAnimation = !0), X) {
            var M = d.getTextContent();
            if (M) {
              var O = me(M);
              O.prevValue != null && (O.prevValue = O.value);
            }
          } else
            K(d, s, g, l, b, e, f, h.type === "polar");
          T ? d.attr({
            shape: b
          }) : c ? j(c, m, d, b, g, f, !0, X) : L(d, {
            shape: b
          }, e, g, null), s.setItemGraphicEl(g, d), d.ignore = N, o.add(d);
        }
      }).remove(function(g) {
        var _ = u.getItemGraphicEl(g);
        _ && E(_, e, g);
      }).execute();
      var P = this._backgroundGroup || (this._backgroundGroup = new fe());
      P.removeAll();
      for (var V = 0; V < D.length; ++V)
        P.add(D[V]);
      o.add(P), this._backgroundEls = D, this._data = s;
    }, r.prototype._renderLarge = function(e, t, a) {
      this._clear(), $(e, this.group), this._updateLargeClip(e);
    }, r.prototype._incrementalRenderLarge = function(e, t) {
      this._removeBackground(), $(t, this.group, this._progressiveEls, !0);
    }, r.prototype._updateLargeClip = function(e) {
      var t = e.get("clip", !0) && xe(e.coordinateSystem, !1, e), a = this.group;
      t ? a.setClipPath(t) : a.removeClipPath();
    }, r.prototype._enableRealtimeSort = function(e, t, a) {
      var n = this;
      if (t.count()) {
        var o = e.baseAxis;
        if (this._isFirstFrame)
          this._dispatchInitSort(t, e, a), this._isFirstFrame = !1;
        else {
          var s = function(u) {
            var h = t.getItemGraphicEl(u), p = h && h.shape;
            return p && // The result should be consistent with the initial sort by data value.
            // Do not support the case that both positive and negative exist.
            Math.abs(o.isHorizontal() ? p.height : p.width) || 0;
          };
          this._onRendered = function() {
            n._updateSortWithinSameData(t, s, o, a);
          }, a.getZr().on("rendered", this._onRendered);
        }
      }
    }, r.prototype._dataSort = function(e, t, a) {
      var n = [];
      return e.each(e.mapDimension(t.dim), function(o, s) {
        var u = a(s);
        u = u ?? NaN, n.push({
          dataIndex: s,
          mappedValue: u,
          ordinalNumber: o
        });
      }), n.sort(function(o, s) {
        return s.mappedValue - o.mappedValue;
      }), {
        ordinalNumbers: ge(n, function(o) {
          return o.ordinalNumber;
        })
      };
    }, r.prototype._isOrderChangedWithinSameData = function(e, t, a) {
      for (var n = a.scale, o = e.mapDimension(a.dim), s = Number.MAX_VALUE, u = 0, h = n.getOrdinalMeta().categories.length; u < h; ++u) {
        var p = e.rawIndexOf(o, n.getRawOrdinalNumber(u)), f = p < 0 ? Number.MIN_VALUE : t(e.indexOfRawIndex(p));
        if (f > s)
          return !0;
        s = f;
      }
      return !1;
    }, r.prototype._isOrderDifferentInView = function(e, t) {
      for (var a = t.scale, n = a.getExtent(), o = Math.max(0, n[0]), s = Math.min(n[1], a.getOrdinalMeta().categories.length - 1); o <= s; ++o)
        if (e.ordinalNumbers[o] !== a.getRawOrdinalNumber(o))
          return !0;
    }, r.prototype._updateSortWithinSameData = function(e, t, a, n) {
      if (this._isOrderChangedWithinSameData(e, t, a)) {
        var o = this._dataSort(e, a, t);
        this._isOrderDifferentInView(o, a) && (this._removeOnRenderedListener(n), n.dispatchAction({
          type: "changeAxisOrder",
          componentType: a.dim + "Axis",
          axisId: a.index,
          sortInfo: o
        }));
      }
    }, r.prototype._dispatchInitSort = function(e, t, a) {
      var n = t.baseAxis, o = this._dataSort(e, n, function(s) {
        return e.get(e.mapDimension(t.otherAxis.dim), s);
      });
      a.dispatchAction({
        type: "changeAxisOrder",
        componentType: n.dim + "Axis",
        isInitSort: !0,
        axisId: n.index,
        sortInfo: o
      });
    }, r.prototype.remove = function(e, t) {
      this._clear(this._model), this._removeOnRenderedListener(t);
    }, r.prototype.dispose = function(e, t) {
      this._removeOnRenderedListener(t);
    }, r.prototype._removeOnRenderedListener = function(e) {
      this._onRendered && (e.getZr().off("rendered", this._onRendered), this._onRendered = null);
    }, r.prototype._clear = function(e) {
      var t = this.group, a = this._data;
      e && e.isAnimationEnabled() && a && !this._isLargeDraw ? (this._removeBackground(), this._backgroundEls = [], a.eachItemGraphicEl(function(n) {
        E(n, e, F(n).dataIndex);
      })) : t.removeAll(), this._data = null, this._isFirstFrame = !0;
    }, r.prototype._removeBackground = function() {
      this.group.remove(this._backgroundGroup), this._backgroundGroup = null;
    }, r.type = "bar", r;
  })(we)
), Y = {
  cartesian2d: function(i, r) {
    var e = r.width < 0 ? -1 : 1, t = r.height < 0 ? -1 : 1;
    e < 0 && (r.x += r.width, r.width = -r.width), t < 0 && (r.y += r.height, r.height = -r.height);
    var a = i.x + i.width, n = i.y + i.height, o = R(r.x, i.x), s = W(r.x + r.width, a), u = R(r.y, i.y), h = W(r.y + r.height, n), p = s < o, f = h < u;
    return r.x = p && o > a ? s : o, r.y = f && u > n ? h : u, r.width = p ? 0 : s - o, r.height = f ? 0 : h - u, e < 0 && (r.x += r.width, r.width = -r.width), t < 0 && (r.y += r.height, r.height = -r.height), p || f;
  },
  polar: function(i, r) {
    var e = r.r0 <= r.r ? 1 : -1;
    if (e < 0) {
      var t = r.r;
      r.r = r.r0, r.r0 = t;
    }
    var a = W(r.r, i.r), n = R(r.r0, i.r0);
    r.r = a, r.r0 = n;
    var o = a - n < 0;
    if (e < 0) {
      var t = r.r;
      r.r = r.r0, r.r0 = t;
    }
    return o;
  }
}, Z = {
  cartesian2d: function(i, r, e, t, a, n, o, s, u) {
    var h = new ie({
      shape: te({}, t),
      z2: 1
    });
    if (h.__dataIndex = e, h.name = "item", n) {
      var p = h.shape, f = a ? "height" : "width";
      p[f] = 0;
    }
    return h;
  },
  polar: function(i, r, e, t, a, n, o, s, u) {
    var h = !a && u ? U : ae, p = new h({
      shape: t,
      z2: 1
    });
    p.name = "item";
    var f = ne(a);
    if (p.calculateTextPosition = Le(f, {
      isRoundCap: h === U
    }), n) {
      var m = p.shape, c = a ? "r" : "endAngle", A = {};
      m[c] = a ? t.r0 : t.startAngle, A[c] = t[c], (s ? L : I)(p, {
        shape: A
        // __value: typeof dataValue === 'string' ? parseInt(dataValue, 10) : dataValue
      }, n);
    }
    return p;
  }
};
function Ve(i, r) {
  var e = i.get("realtimeSort", !0), t = r.getBaseAxis();
  if (process.env.NODE_ENV !== "production" && e && (t.type !== "category" && B("`realtimeSort` will not work because this bar series is not based on a category axis."), r.type !== "cartesian2d" && B("`realtimeSort` will not work because this bar series is not on cartesian2d.")), e && t.type === "category" && r.type === "cartesian2d")
    return {
      baseAxis: t,
      otherAxis: r.getOtherAxis(t)
    };
}
function j(i, r, e, t, a, n, o, s) {
  var u, h;
  n ? (h = {
    x: t.x,
    width: t.width
  }, u = {
    y: t.y,
    height: t.height
  }) : (h = {
    y: t.y,
    height: t.height
  }, u = {
    x: t.x,
    width: t.width
  }), s || (o ? L : I)(e, {
    shape: u
  }, r, a, null);
  var p = r ? i.baseAxis.model : null;
  (o ? L : I)(e, {
    shape: h
  }, p, a);
}
function q(i, r) {
  for (var e = 0; e < r.length; e++)
    if (!isFinite(i[r[e]]))
      return !0;
  return !1;
}
var Ne = ["x", "y", "width", "height"], Oe = ["cx", "cy", "r", "startAngle", "endAngle"], J = {
  cartesian2d: function(i) {
    return !q(i, Ne);
  },
  polar: function(i) {
    return !q(i, Oe);
  }
}, k = {
  // itemModel is only used to get borderWidth, which is not needed
  // when calculating bar background layout.
  cartesian2d: function(i, r, e) {
    var t = i.getItemLayout(r);
    if (!t)
      return null;
    var a = e ? Re(e, t) : 0, n = t.width > 0 ? 1 : -1, o = t.height > 0 ? 1 : -1;
    return {
      x: t.x + n * a / 2,
      y: t.y + o * a / 2,
      width: t.width - n * a,
      height: t.height - o * a
    };
  },
  polar: function(i, r, e) {
    var t = i.getItemLayout(r);
    return {
      cx: t.cx,
      cy: t.cy,
      r0: t.r0,
      r: t.r,
      startAngle: t.startAngle,
      endAngle: t.endAngle,
      clockwise: t.clockwise
    };
  }
};
function Ee(i) {
  return i.startAngle != null && i.endAngle != null && i.startAngle === i.endAngle;
}
function ne(i) {
  return /* @__PURE__ */ (function(r) {
    var e = r ? "Arc" : "Angle";
    return function(t) {
      switch (t) {
        case "start":
        case "insideStart":
        case "end":
        case "insideEnd":
          return t + e;
        default:
          return t;
      }
    };
  })(i);
}
function K(i, r, e, t, a, n, o, s) {
  var u = r.getItemVisual(e, "style");
  if (s) {
    if (!n.get("roundCap")) {
      var p = i.shape, f = Ie(t.getModel("itemStyle"), p, !0);
      te(p, f), i.setShape(p);
    }
  } else {
    var h = t.get(["itemStyle", "borderRadius"]) || 0;
    i.setShape("r", h);
  }
  i.useStyle(u);
  var m = t.getShallow("cursor");
  m && i.attr("cursor", m);
  var c = s ? o ? a.r >= a.r0 ? "endArc" : "startArc" : a.endAngle >= a.startAngle ? "endAngle" : "startAngle" : o ? a.height >= 0 ? "bottom" : "top" : a.width >= 0 ? "right" : "left", A = _e(t);
  be(i, A, {
    labelFetcher: n,
    labelDataIndex: e,
    defaultText: Se(n.getData(), e),
    inheritColor: u.fill,
    defaultOpacity: u.opacity,
    defaultOutsidePosition: c
  });
  var S = i.getTextContent();
  if (s && S) {
    var x = t.get(["label", "position"]);
    i.textConfig.inside = x === "middle" ? !0 : null, Ce(i, x === "outside" ? c : x, ne(o), t.get(["label", "rotate"]));
  }
  ye(S, A, n.getRawValue(e), function(y) {
    return De(r, y);
  });
  var w = t.getModel(["emphasis"]);
  ce(i, w.get("focus"), w.get("blurScope"), w.get("disabled")), ve(i, t), Ee(a) && (i.style.fill = "none", i.style.stroke = "none", de(i.states, function(y) {
    y.style && (y.style.fill = y.style.stroke = "none");
  }));
}
function Re(i, r) {
  var e = i.get(["itemStyle", "borderColor"]);
  if (!e || e === "none")
    return 0;
  var t = i.get(["itemStyle", "borderWidth"]) || 0, a = isNaN(r.width) ? Number.MAX_VALUE : Math.abs(r.width), n = isNaN(r.height) ? Number.MAX_VALUE : Math.abs(r.height);
  return Math.min(t, a, n);
}
var We = (
  /** @class */
  /* @__PURE__ */ (function() {
    function i() {
    }
    return i;
  })()
), Q = (
  /** @class */
  (function(i) {
    ee(r, i);
    function r(e) {
      var t = i.call(this, e) || this;
      return t.type = "largeBar", t;
    }
    return r.prototype.getDefaultShape = function() {
      return new We();
    }, r.prototype.buildPath = function(e, t) {
      for (var a = t.points, n = this.baseDimIdx, o = 1 - this.baseDimIdx, s = [], u = [], h = this.barWidth, p = 0; p < a.length; p += 3)
        u[n] = h, u[o] = a[p + 2], s[n] = a[p + n], s[o] = a[p + o], e.rect(s[0], s[1], u[0], u[1]);
    }, r;
  })(ue)
);
function $(i, r, e, t) {
  var a = i.getData(), n = a.getLayout("valueAxisHorizontal") ? 1 : 0, o = a.getLayout("largeDataIndices"), s = a.getLayout("size"), u = i.getModel("backgroundStyle"), h = a.getLayout("largeBackgroundPoints");
  if (h) {
    var p = new Q({
      shape: {
        points: h
      },
      incremental: !!t,
      silent: !0,
      z2: 0
    });
    p.baseDimIdx = n, p.largeDataIndices = o, p.barWidth = s, p.useStyle(u.getItemStyle()), r.add(p), e && e.push(p);
  }
  var f = new Q({
    shape: {
      points: a.getLayout("largePoints")
    },
    incremental: !!t,
    ignoreCoarsePointer: !0,
    z2: 1
  });
  f.baseDimIdx = n, f.largeDataIndices = o, f.barWidth = s, r.add(f), f.useStyle(a.getVisual("style")), f.style.stroke = null, F(f).seriesIndex = i.seriesIndex, i.get("silent") || (f.on("mousedown", H), f.on("mousemove", H)), e && e.push(f);
}
var H = Ae(function(i) {
  var r = this, e = Be(r, i.offsetX, i.offsetY);
  F(r).dataIndex = e >= 0 ? e : null;
}, 30, !1);
function Be(i, r, e) {
  for (var t = i.baseDimIdx, a = 1 - t, n = i.shape.points, o = i.largeDataIndices, s = [], u = [], h = i.barWidth, p = 0, f = n.length / 3; p < f; p++) {
    var m = p * 3;
    if (u[t] = h, u[a] = n[m + 2], s[t] = n[m + t], s[a] = n[m + a], u[a] < 0 && (s[a] += u[a], u[a] = -u[a]), r >= s[0] && r <= s[0] + u[0] && e >= s[1] && e <= s[1] + u[1])
      return o[p];
  }
  return -1;
}
function se(i, r, e) {
  if (re(e, "cartesian2d")) {
    var t = r, a = e.getArea();
    return {
      x: i ? t.x : a.x,
      y: i ? a.y : t.y,
      width: i ? t.width : a.width,
      height: i ? a.height : t.height
    };
  } else {
    var a = e.getArea(), n = r;
    return {
      cx: a.cx,
      cy: a.cy,
      r0: i ? a.r0 : n.r0,
      r: i ? a.r : n.r,
      startAngle: i ? n.startAngle : 0,
      endAngle: i ? n.endAngle : Math.PI * 2
    };
  }
}
function Fe(i, r, e) {
  var t = i.type === "polar" ? ae : ie;
  return new t({
    shape: se(r, e, i),
    silent: !0,
    z2: 0
  });
}
export {
  nt as default
};
