// echarts@6.0.0/lib/chart/bar/BaseBarSeries.js
import { __extends as A } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import S from "../../model/Series.js";
import G from "../helper/createSeriesData.js";
import { each as H } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
var P = (
  /** @class */
  (function(h) {
    A(a, h);
    function a() {
      var f = h !== null && h.apply(this, arguments) || this;
      return f.type = a.type, f;
    }
    return a.prototype.getInitialData = function(f, s) {
      return G(null, this, {
        useEncodeDefaulter: !0
      });
    }, a.prototype.getMarkerPosition = function(f, s, B) {
      var t = this.coordinateSystem;
      if (t && t.clampData) {
        var y = t.clampData(f), u = t.dataToPoint(y);
        if (B)
          H(t.getAxes(), function(n, i) {
            if (n.type === "category" && s != null) {
              var e = n.getTicksCoords(), z = n.getTickModel().get("alignWithLabel"), g = y[i], b = s[i] === "x1" || s[i] === "y1";
              if (b && !z && (g += 1), e.length < 2)
                return;
              if (e.length === 2) {
                u[i] = n.toGlobalCoord(n.getExtent()[b ? 1 : 0]);
                return;
              }
              for (var o = void 0, l = void 0, M = 1, r = 0; r < e.length; r++) {
                var p = e[r].coord, v = r === e.length - 1 ? e[r - 1].tickValue + M : e[r].tickValue;
                if (v === g) {
                  l = p;
                  break;
                } else if (v < g)
                  o = p;
                else if (o != null && v > g) {
                  l = (p + o) / 2;
                  break;
                }
                r === 1 && (M = v - e[0].tickValue);
              }
              l == null && (o ? o && (l = e[e.length - 1].coord) : l = e[0].coord), u[i] = n.toGlobalCoord(l);
            }
          });
        else {
          var m = this.getData(), D = m.getLayout("offset"), L = m.getLayout("size"), N = t.getBaseAxis().isHorizontal() ? 0 : 1;
          u[N] += D + L / 2;
        }
        return u;
      }
      return [NaN, NaN];
    }, a.type = "series.__base_bar__", a.defaultOption = {
      // zlevel: 0,
      z: 2,
      coordinateSystem: "cartesian2d",
      legendHoverLink: !0,
      // stack: null
      // Cartesian coordinate system
      // xAxisIndex: 0,
      // yAxisIndex: 0,
      barMinHeight: 0,
      barMinAngle: 0,
      // cursor: null,
      large: !1,
      largeThreshold: 400,
      progressive: 3e3,
      progressiveChunkMode: "mod",
      defaultBarGap: "10%"
    }, a;
  })(S)
);
S.registerClass(P);
export {
  P as default
};
