// echarts@6.0.0/lib/chart/bar/PictorialBarView.js
import { __extends as re } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { each as Q, isArray as U, retrieve as te, extend as z } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import * as Z from "../../util/graphic.js";
import { toggleHoverEmphasis as ae } from "../../util/states.js";
import { normalizeSymbolOffset as ne, createSymbol as oe } from "../../util/symbol.js";
import { parsePercent as L, isNumeric as le } from "../../util/number.js";
import ce from "../../view/Chart.js";
import { getDefaultLabel as se } from "../helper/labelHelper.js";
import { setLabelStyle as ve, getLabelStatesModels as ue } from "../../label/labelStyle.js";
import pe from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Image.js";
import { getECData as he } from "../../util/innerStore.js";
import { createClipPath as de } from "../helper/createClipPathFromCoordSys.js";
import X from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import fe from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Circle.js";
import q from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Rect.js";
import { updateProps as F, removeElement as me } from "../../animation/basicTransition.js";
var ge = ["itemStyle", "borderWidth"], V = [{
  xy: "x",
  wh: "width",
  index: 0,
  posDesc: ["left", "right"]
}, {
  xy: "y",
  wh: "height",
  index: 1,
  posDesc: ["top", "bottom"]
}], B = new fe(), He = (
  /** @class */
  (function(i) {
    re(r, i);
    function r() {
      var e = i !== null && i.apply(this, arguments) || this;
      return e.type = r.type, e;
    }
    return r.prototype.render = function(e, n, a) {
      var t = this.group, o = e.getData(), d = this._data, f = e.coordinateSystem, m = f.getBaseAxis(), u = m.isHorizontal(), p = f.master.getRect(), l = {
        ecSize: {
          width: a.getWidth(),
          height: a.getHeight()
        },
        seriesModel: e,
        coordSys: f,
        coordSysExtent: [[p.x, p.x + p.width], [p.y, p.y + p.height]],
        isHorizontal: u,
        valueDim: V[+u],
        categoryDim: V[1 - +u]
      };
      o.diff(d).add(function(c) {
        if (o.hasValue(c)) {
          var s = G(o, c), h = W(o, c, s, l), g = H(o, l, h);
          o.setItemGraphicEl(c, g), t.add(g), k(g, l, h);
        }
      }).update(function(c, s) {
        var h = d.getItemGraphicEl(s);
        if (!o.hasValue(c)) {
          t.remove(h);
          return;
        }
        var g = G(o, c), S = W(o, c, g, l), _ = ee(o, S);
        h && _ !== h.__pictorialShapeStr && (t.remove(h), o.setItemGraphicEl(c, null), h = null), h ? Re(h, l, S) : h = H(o, l, S, !0), o.setItemGraphicEl(c, h), h.__pictorialSymbolMeta = S, t.add(h), k(h, l, S);
      }).remove(function(c) {
        var s = d.getItemGraphicEl(c);
        s && b(d, c, s.__pictorialSymbolMeta.animationModel, s);
      }).execute();
      var v = e.get("clip", !0) ? de(e.coordinateSystem, !1, e) : null;
      return v ? t.setClipPath(v) : t.removeClipPath(), this._data = o, this.group;
    }, r.prototype.remove = function(e, n) {
      var a = this.group, t = this._data;
      e.get("animation") ? t && t.eachItemGraphicEl(function(o) {
        b(t, he(o).dataIndex, e, o);
      }) : a.removeAll();
    }, r.type = "pictorialBar", r;
  })(ce)
);
function W(i, r, e, n) {
  var a = i.getItemLayout(r), t = e.get("symbolRepeat"), o = e.get("symbolClip"), d = e.get("symbolPosition") || "start", f = e.get("symbolRotate"), m = (f || 0) * Math.PI / 180 || 0, u = e.get("symbolPatternSize") || 2, p = e.isAnimationEnabled(), l = {
    dataIndex: r,
    layout: a,
    itemModel: e,
    symbolType: i.getItemVisual(r, "symbol") || "circle",
    style: i.getItemVisual(r, "style"),
    symbolClip: o,
    symbolRepeat: t,
    symbolRepeatDirection: e.get("symbolRepeatDirection"),
    symbolPatternSize: u,
    rotation: m,
    animationModel: p ? e : null,
    hoverScale: p && e.get(["emphasis", "scale"]),
    z2: e.getShallow("z", !0) || 0
  };
  Se(e, t, a, n, l), _e(i, r, a, t, o, l.boundingLength, l.pxSign, u, n, l), xe(e, l.symbolScale, m, n, l);
  var v = l.symbolSize, c = ne(e.get("symbolOffset"), v);
  return Pe(e, v, a, t, o, c, d, l.valueLineWidth, l.boundingLength, l.repeatCutLength, n, l), l;
}
function Se(i, r, e, n, a) {
  var t = n.valueDim, o = i.get("symbolBoundingData"), d = n.coordSys.getOtherAxis(n.coordSys.getBaseAxis()), f = d.toGlobalCoord(d.dataToCoord(0)), m = 1 - +(e[t.wh] <= 0), u;
  if (U(o)) {
    var p = [A(d, o[0]) - f, A(d, o[1]) - f];
    p[1] < p[0] && p.reverse(), u = p[m];
  } else o != null ? u = A(d, o) - f : r ? u = n.coordSysExtent[t.index][m] - f : u = e[t.wh];
  a.boundingLength = u, r && (a.repeatCutLength = e[t.wh]);
  var l = t.xy === "x", v = d.inverse;
  a.pxSign = l && !v || !l && v ? u >= 0 ? 1 : -1 : u > 0 ? 1 : -1;
}
function A(i, r) {
  return i.toGlobalCoord(i.dataToCoord(i.scale.parse(r)));
}
function _e(i, r, e, n, a, t, o, d, f, m) {
  var u = f.valueDim, p = f.categoryDim, l = Math.abs(e[p.wh]), v = i.getItemVisual(r, "symbolSize"), c;
  U(v) ? c = v.slice() : v == null ? c = ["100%", "100%"] : c = [v, v], c[p.index] = L(c[p.index], l), c[u.index] = L(c[u.index], n ? l : Math.abs(t)), m.symbolSize = c;
  var s = m.symbolScale = [c[0] / d, c[1] / d];
  s[u.index] *= (f.isHorizontal ? -1 : 1) * o;
}
function xe(i, r, e, n, a) {
  var t = i.get(ge) || 0;
  t && (B.attr({
    scaleX: r[0],
    scaleY: r[1],
    rotation: e
  }), B.updateTransform(), t /= B.getLineScale(), t *= r[n.valueDim.index]), a.valueLineWidth = t || 0;
}
function Pe(i, r, e, n, a, t, o, d, f, m, u, p) {
  var l = u.categoryDim, v = u.valueDim, c = p.pxSign, s = Math.max(r[v.index] + d, 0), h = s;
  if (n) {
    var g = Math.abs(f), S = te(i.get("symbolMargin"), "15%") + "", _ = !1;
    S.lastIndexOf("!") === S.length - 1 && (_ = !0, S = S.slice(0, S.length - 1));
    var x = L(S, r[v.index]), R = Math.max(s + x * 2, 0), w = _ ? 0 : x * 2, Y = le(n), P = Y ? n : j((g + w) / R), ie = g - P * s;
    x = ie / 2 / (_ ? P : Math.max(P - 1, 1)), R = s + x * 2, w = _ ? 0 : x * 2, !Y && n !== "fixed" && (P = m ? j((Math.abs(m) + w) / R) : 0), h = P * R - w, p.repeatTimes = P, p.symbolMargin = x;
  }
  var I = c * (h / 2), D = p.pathPosition = [];
  D[l.index] = e[l.wh] / 2, D[v.index] = o === "start" ? I : o === "end" ? f - I : f / 2, t && (D[0] += t[0], D[1] += t[1]);
  var E = p.bundlePosition = [];
  E[l.index] = e[l.xy], E[v.index] = e[v.xy];
  var O = p.barRectShape = z({}, e);
  O[v.wh] = c * Math.max(Math.abs(e[v.wh]), Math.abs(D[v.index] + I)), O[l.wh] = e[l.wh];
  var C = p.clipShape = {};
  C[l.xy] = -e[l.xy], C[l.wh] = u.ecSize[l.wh], C[v.xy] = 0, C[v.wh] = e[v.wh];
}
function J(i) {
  var r = i.symbolPatternSize, e = oe(
    // Consider texture img, make a big size.
    i.symbolType,
    -r / 2,
    -r / 2,
    r,
    r
  );
  return e.attr({
    culling: !0
  }), e.type !== "image" && e.setStyle({
    strokeNoScale: !0
  }), e;
}
function K(i, r, e, n) {
  var a = i.__pictorialBundle, t = e.symbolSize, o = e.valueLineWidth, d = e.pathPosition, f = r.valueDim, m = e.repeatTimes || 0, u = 0, p = t[r.valueDim.index] + o + e.symbolMargin * 2;
  for (T(i, function(s) {
    s.__pictorialAnimationIndex = u, s.__pictorialRepeatTimes = m, u < m ? y(s, null, c(u), e, n) : y(s, null, {
      scaleX: 0,
      scaleY: 0
    }, e, n, function() {
      a.remove(s);
    }), u++;
  }); u < m; u++) {
    var l = J(e);
    l.__pictorialAnimationIndex = u, l.__pictorialRepeatTimes = m, a.add(l);
    var v = c(u);
    y(l, {
      x: v.x,
      y: v.y,
      scaleX: 0,
      scaleY: 0
    }, {
      scaleX: v.scaleX,
      scaleY: v.scaleY,
      rotation: v.rotation
    }, e, n);
  }
  function c(s) {
    var h = d.slice(), g = e.pxSign, S = s;
    return (e.symbolRepeatDirection === "start" ? g > 0 : g < 0) && (S = m - 1 - s), h[f.index] = p * (S - m / 2 + 0.5) + d[f.index], {
      x: h[0],
      y: h[1],
      scaleX: e.symbolScale[0],
      scaleY: e.symbolScale[1],
      rotation: e.rotation
    };
  }
}
function $(i, r, e, n) {
  var a = i.__pictorialBundle, t = i.__pictorialMainPath;
  t ? y(t, null, {
    x: e.pathPosition[0],
    y: e.pathPosition[1],
    scaleX: e.symbolScale[0],
    scaleY: e.symbolScale[1],
    rotation: e.rotation
  }, e, n) : (t = i.__pictorialMainPath = J(e), a.add(t), y(t, {
    x: e.pathPosition[0],
    y: e.pathPosition[1],
    scaleX: 0,
    scaleY: 0,
    rotation: e.rotation
  }, {
    scaleX: e.symbolScale[0],
    scaleY: e.symbolScale[1]
  }, e, n));
}
function N(i, r, e) {
  var n = z({}, r.barRectShape), a = i.__pictorialBarRect;
  a ? y(a, null, {
    shape: n
  }, r, e) : (a = i.__pictorialBarRect = new q({
    z2: 2,
    shape: n,
    silent: !0,
    style: {
      stroke: "transparent",
      fill: "transparent",
      lineWidth: 0
    }
  }), a.disableMorphing = !0, i.add(a));
}
function M(i, r, e, n) {
  if (e.symbolClip) {
    var a = i.__pictorialClipPath, t = z({}, e.clipShape), o = r.valueDim, d = e.animationModel, f = e.dataIndex;
    if (a)
      F(a, {
        shape: t
      }, d, f);
    else {
      t[o.wh] = 0, a = new q({
        shape: t
      }), i.__pictorialBundle.setClipPath(a), i.__pictorialClipPath = a;
      var m = {};
      m[o.wh] = e.clipShape[o.wh], Z[n ? "updateProps" : "initProps"](a, {
        shape: m
      }, d, f);
    }
  }
}
function G(i, r) {
  var e = i.getItemModel(r);
  return e.getAnimationDelayParams = ye, e.isAnimationEnabled = De, e;
}
function ye(i) {
  return {
    index: i.__pictorialAnimationIndex,
    count: i.__pictorialRepeatTimes
  };
}
function De() {
  return this.parentModel.isAnimationEnabled() && !!this.getShallow("animation");
}
function H(i, r, e, n) {
  var a = new X(), t = new X();
  return a.add(t), a.__pictorialBundle = t, t.x = e.bundlePosition[0], t.y = e.bundlePosition[1], e.symbolRepeat ? K(a, r, e) : $(a, r, e), N(a, e, n), M(a, r, e, n), a.__pictorialShapeStr = ee(i, e), a.__pictorialSymbolMeta = e, a;
}
function Re(i, r, e) {
  var n = e.animationModel, a = e.dataIndex, t = i.__pictorialBundle;
  F(t, {
    x: e.bundlePosition[0],
    y: e.bundlePosition[1]
  }, n, a), e.symbolRepeat ? K(i, r, e, !0) : $(i, r, e, !0), N(i, e, !0), M(i, r, e, !0);
}
function b(i, r, e, n) {
  var a = n.__pictorialBarRect;
  a && a.removeTextContent();
  var t = [];
  T(n, function(o) {
    t.push(o);
  }), n.__pictorialMainPath && t.push(n.__pictorialMainPath), n.__pictorialClipPath && (e = null), Q(t, function(o) {
    me(o, {
      scaleX: 0,
      scaleY: 0
    }, e, r, function() {
      n.parent && n.parent.remove(n);
    });
  }), i.setItemGraphicEl(r, null);
}
function ee(i, r) {
  return [i.getItemVisual(r.dataIndex, "symbol") || "none", !!r.symbolRepeat, !!r.symbolClip].join(":");
}
function T(i, r, e) {
  Q(i.__pictorialBundle.children(), function(n) {
    n !== i.__pictorialBarRect && r.call(e, n);
  });
}
function y(i, r, e, n, a, t) {
  r && i.attr(r), n.symbolClip && !a ? e && i.attr(e) : e && Z[a ? "updateProps" : "initProps"](i, e, n.animationModel, n.dataIndex, t);
}
function k(i, r, e) {
  var n = e.dataIndex, a = e.itemModel, t = a.getModel("emphasis"), o = t.getModel("itemStyle").getItemStyle(), d = a.getModel(["blur", "itemStyle"]).getItemStyle(), f = a.getModel(["select", "itemStyle"]).getItemStyle(), m = a.getShallow("cursor"), u = t.get("focus"), p = t.get("blurScope"), l = t.get("scale");
  T(i, function(s) {
    if (s instanceof pe) {
      var h = s.style;
      s.useStyle(z({
        // TODO other properties like dx, dy ?
        image: h.image,
        x: h.x,
        y: h.y,
        width: h.width,
        height: h.height
      }, e.style));
    } else
      s.useStyle(e.style);
    var g = s.ensureState("emphasis");
    g.style = o, l && (g.scaleX = s.scaleX * 1.1, g.scaleY = s.scaleY * 1.1), s.ensureState("blur").style = d, s.ensureState("select").style = f, m && (s.cursor = m), s.z2 = e.z2;
  });
  var v = r.valueDim.posDesc[+(e.boundingLength > 0)], c = i.__pictorialBarRect;
  c.ignoreClip = !0, ve(c, ue(a), {
    labelFetcher: r.seriesModel,
    labelDataIndex: n,
    defaultText: se(r.seriesModel.getData(), n),
    inheritColor: e.style.fill,
    defaultOpacity: e.style.opacity,
    defaultOutsidePosition: v
  }), ae(i, u, p, t.get("disabled"));
}
function j(i) {
  var r = Math.round(i);
  return Math.abs(i - r) < 1e-4 ? r : Math.ceil(i);
}
export {
  He as default
};
