// echarts@6.0.0/lib/chart/bar/PictorialBarSeries.js
import { __extends as i } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import o from "./BaseBarSeries.js";
import { inheritDefaultOption as r } from "../../util/component.js";
import a from "../../visual/tokens.js";
var u = (
  /** @class */
  (function(l) {
    i(e, l);
    function e() {
      var t = l !== null && l.apply(this, arguments) || this;
      return t.type = e.type, t.hasSymbolVisual = !0, t.defaultSymbol = "roundRect", t;
    }
    return e.prototype.getInitialData = function(t) {
      return t.stack = null, l.prototype.getInitialData.apply(this, arguments);
    }, e.type = "series.pictorialBar", e.dependencies = ["grid"], e.defaultOption = r(o.defaultOption, {
      symbol: "circle",
      symbolSize: null,
      symbolRotate: null,
      symbolPosition: null,
      symbolOffset: null,
      symbolMargin: null,
      symbolRepeat: !1,
      symbolRepeatDirection: "end",
      symbolClip: !1,
      symbolBoundingData: null,
      symbolPatternSize: 400,
      barGap: "-100%",
      // Pictorial bar do not clip by default because in many cases
      // xAxis and yAxis are not displayed and it's expected not to clip
      clip: !1,
      // z can be set in data item, which is z2 actually.
      // Disable progressive
      progressive: 0,
      emphasis: {
        // By default pictorialBar do not hover scale. Hover scale is not suitable
        // for the case that both has foreground and background.
        scale: !1
      },
      select: {
        itemStyle: {
          borderColor: a.color.primary
        }
      }
    }), e;
  })(o)
);
export {
  u as default
};
