// echarts@6.0.0/lib/coord/single/AxisModel.js
import { __extends as i } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import r from "../../model/Component.js";
import { AxisModelCommonMixin as n } from "../axisModelCommonMixin.js";
import { mixin as l } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
var p = (
  /** @class */
  (function(e) {
    i(t, e);
    function t() {
      var o = e !== null && e.apply(this, arguments) || this;
      return o.type = t.type, o;
    }
    return t.prototype.getCoordSysModel = function() {
      return this;
    }, t.type = "singleAxis", t.layoutMode = "box", t.defaultOption = {
      left: "5%",
      top: "5%",
      right: "5%",
      bottom: "5%",
      type: "value",
      position: "bottom",
      orient: "horizontal",
      axisLine: {
        show: !0,
        lineStyle: {
          width: 1,
          type: "solid"
        }
      },
      // Single coordinate system and single axis is the,
      // which is used as the parent tooltip model.
      // same model, so we set default tooltip show as true.
      tooltip: {
        show: !0
      },
      axisTick: {
        show: !0,
        length: 6,
        lineStyle: {
          width: 1
        }
      },
      axisLabel: {
        show: !0,
        interval: "auto"
      },
      splitLine: {
        show: !0,
        lineStyle: {
          type: "dashed",
          opacity: 0.2
        }
      },
      jitter: 0,
      jitterOverlap: !0,
      jitterMargin: 2
    }, t;
  })(r)
);
l(p, n.prototype);
export {
  p as default
};
