// echarts@6.0.0/lib/coord/single/Single.js
import c from "./SingleAxis.js";
import { createScaleByModel as h, niceScaleExtent as u } from "../axisHelper.js";
import { createBoxLayoutReference as p, getLayoutRect as y } from "../../util/layout.js";
import { each as l } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
var f = ["single"], m = (
  /** @class */
  (function() {
    function n(t, i, e) {
      this.type = "single", this.dimension = "single", this.dimensions = f, this.axisPointerEnabled = !0, this.model = t, this._init(t, i, e);
    }
    return n.prototype._init = function(t, i, e) {
      var o = this.dimension, r = new c(o, h(t), [0, 0], t.get("type"), t.get("position")), s = r.type === "category";
      r.onBand = s && t.get("boundaryGap"), r.inverse = t.get("inverse"), r.orient = t.get("orient"), t.axis = r, r.model = t, r.coordinateSystem = this, this._axis = r;
    }, n.prototype.update = function(t, i) {
      t.eachSeries(function(e) {
        if (e.coordinateSystem === this) {
          var o = e.getData();
          l(o.mapDimensionsAll(this.dimension), function(r) {
            this._axis.scale.unionExtentFromData(o, r);
          }, this), u(this._axis.scale, this._axis.model);
        }
      }, this);
    }, n.prototype.resize = function(t, i) {
      var e = p(t, i).refContainer;
      this._rect = y(t.getBoxLayoutParams(), e), this._adjustAxis();
    }, n.prototype.getRect = function() {
      return this._rect;
    }, n.prototype._adjustAxis = function() {
      var t = this._rect, i = this._axis, e = i.isHorizontal(), o = e ? [0, t.width] : [0, t.height], r = i.inverse ? 1 : 0;
      i.setExtent(o[r], o[1 - r]), this._updateAxisTransform(i, e ? t.x : t.y);
    }, n.prototype._updateAxisTransform = function(t, i) {
      var e = t.getExtent(), o = e[0] + e[1], r = t.isHorizontal();
      t.toGlobalCoord = r ? function(s) {
        return s + i;
      } : function(s) {
        return o - s + i;
      }, t.toLocalCoord = r ? function(s) {
        return s - i;
      } : function(s) {
        return o - s + i;
      };
    }, n.prototype.getAxis = function() {
      return this._axis;
    }, n.prototype.getBaseAxis = function() {
      return this._axis;
    }, n.prototype.getAxes = function() {
      return [this._axis];
    }, n.prototype.getTooltipAxes = function() {
      return {
        baseAxes: [this.getAxis()],
        // Empty otherAxes
        otherAxes: []
      };
    }, n.prototype.containPoint = function(t) {
      var i = this.getRect(), e = this.getAxis(), o = e.orient;
      return o === "horizontal" ? e.contain(e.toLocalCoord(t[0])) && t[1] >= i.y && t[1] <= i.y + i.height : e.contain(e.toLocalCoord(t[1])) && t[0] >= i.y && t[0] <= i.y + i.height;
    }, n.prototype.pointToData = function(t, i, e) {
      e = e || [];
      var o = this.getAxis();
      return e[0] = o.coordToData(o.toLocalCoord(t[o.orient === "horizontal" ? 0 : 1])), e;
    }, n.prototype.dataToPoint = function(t, i, e) {
      var o = this.getAxis(), r = this.getRect();
      e = e || [];
      var s = o.orient === "horizontal" ? 0 : 1;
      return t instanceof Array && (t = t[0]), e[s] = o.toGlobalCoord(o.dataToCoord(+t)), e[1 - s] = s === 0 ? r.y + r.height / 2 : r.x + r.width / 2, e;
    }, n.prototype.convertToPixel = function(t, i, e) {
      var o = a(i);
      return o === this ? this.dataToPoint(e) : null;
    }, n.prototype.convertFromPixel = function(t, i, e) {
      var o = a(i);
      return o === this ? this.pointToData(e) : null;
    }, n;
  })()
);
function a(n) {
  var t = n.seriesModel, i = n.singleAxisModel;
  return i && i.coordinateSystem || t && t.coordinateSystem;
}
export {
  m as default,
  f as singleDimensions
};
