// echarts@6.0.0/lib/coord/single/singleAxisHelper.js
import { retrieve as g } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
function D(e, a) {
  a = a || {};
  var s = e.coordinateSystem, l = e.axis, t = {}, o = l.position, n = l.orient, i = s.getRect(), r = [i.x, i.x + i.width, i.y, i.y + i.height], c = {
    horizontal: {
      top: r[2],
      bottom: r[3]
    },
    vertical: {
      left: r[0],
      right: r[1]
    }
  };
  t.position = [n === "vertical" ? c.vertical[o] : r[0], n === "horizontal" ? c.horizontal[o] : r[3]];
  var b = {
    horizontal: 0,
    vertical: 1
  };
  t.rotation = Math.PI / 2 * b[n];
  var h = {
    top: -1,
    bottom: 1,
    right: 1,
    left: -1
  };
  t.labelDirection = t.tickDirection = t.nameDirection = h[o], e.get(["axisTick", "inside"]) && (t.tickDirection = -t.tickDirection), g(a.labelInside, e.get(["axisLabel", "inside"])) && (t.labelDirection = -t.labelDirection);
  var v = e.get(["axisLabel", "rotate"]);
  return t.labelRotate = o === "top" ? -v : v, t.z2 = 1, t;
}
export {
  D as layout
};
