// echarts@6.0.0/lib/chart/tree/layoutHelper.js
function E(r) {
  var e = r;
  e.hierNode = {
    defaultAncestor: null,
    ancestor: e,
    prelim: 0,
    modifier: 0,
    change: 0,
    shift: 0,
    i: 0,
    thread: null
  };
  for (var a = [e], N, i; N = a.pop(); )
    if (i = N.children, N.isExpand && i.length)
      for (var h = i.length, l = h - 1; l >= 0; l--) {
        var o = i[l];
        o.hierNode = {
          defaultAncestor: null,
          ancestor: o,
          prelim: 0,
          modifier: 0,
          change: 0,
          shift: 0,
          i: l,
          thread: null
        }, a.push(o);
      }
}
function M(r, e) {
  var a = r.isExpand ? r.children : [], N = r.parentNode.children, i = r.hierNode.i ? N[r.hierNode.i - 1] : null;
  if (a.length) {
    t(r);
    var h = (a[0].hierNode.prelim + a[a.length - 1].hierNode.prelim) / 2;
    i ? (r.hierNode.prelim = i.hierNode.prelim + e(r, i), r.hierNode.modifier = r.hierNode.prelim - h) : r.hierNode.prelim = h;
  } else i && (r.hierNode.prelim = i.hierNode.prelim + e(r, i));
  r.parentNode.hierNode.defaultAncestor = x(r, i, r.parentNode.hierNode.defaultAncestor || N[0], e);
}
function S(r) {
  var e = r.hierNode.prelim + r.parentNode.hierNode.modifier;
  r.setLayout({
    x: e
  }, !0), r.hierNode.modifier += r.parentNode.hierNode.modifier;
}
function k(r) {
  return arguments.length ? r : A;
}
function y(r, e) {
  return r -= Math.PI / 2, {
    x: e * Math.cos(r),
    y: e * Math.sin(r)
  };
}
function t(r) {
  for (var e = r.children, a = e.length, N = 0, i = 0; --a >= 0; ) {
    var h = e[a];
    h.hierNode.prelim += N, h.hierNode.modifier += N, i += h.hierNode.change, N += h.hierNode.shift + i;
  }
}
function x(r, e, a, N) {
  if (e) {
    for (var i = r, h = r, l = h.parentNode.children[0], o = e, p = i.hierNode.modifier, d = h.hierNode.modifier, n = l.hierNode.modifier, v = o.hierNode.modifier; o = f(o), h = m(h), o && h; ) {
      i = f(i), l = m(l), i.hierNode.ancestor = r;
      var c = o.hierNode.prelim + v - h.hierNode.prelim - d + N(o, h);
      c > 0 && (u(s(o, r, a), r, c), d += c, p += c), v += o.hierNode.modifier, d += h.hierNode.modifier, p += i.hierNode.modifier, n += l.hierNode.modifier;
    }
    o && !f(i) && (i.hierNode.thread = o, i.hierNode.modifier += v - p), h && !m(l) && (l.hierNode.thread = h, l.hierNode.modifier += d - n, a = r);
  }
  return a;
}
function f(r) {
  var e = r.children;
  return e.length && r.isExpand ? e[e.length - 1] : r.hierNode.thread;
}
function m(r) {
  var e = r.children;
  return e.length && r.isExpand ? e[0] : r.hierNode.thread;
}
function s(r, e, a) {
  return r.hierNode.ancestor.parentNode === e.parentNode ? r.hierNode.ancestor : a;
}
function u(r, e, a) {
  var N = a / (e.hierNode.i - r.hierNode.i);
  e.hierNode.change -= N, e.hierNode.shift += a, e.hierNode.modifier += a, e.hierNode.prelim += a, r.hierNode.change += N;
}
function A(r, e) {
  return r.parentNode === e.parentNode ? 1 : 2;
}
export {
  M as firstWalk,
  E as init,
  y as radialCoordinate,
  S as secondWalk,
  k as separation
};
