// echarts@6.0.0/lib/chart/tree/treeAction.js
import { updateCenterAndZoomInAction as p } from "../../component/helper/roamHelper.js";
function d(r) {
  r.registerAction({
    type: "treeExpandAndCollapse",
    event: "treeExpandAndCollapse",
    update: "update"
  }, function(t, a) {
    a.eachComponent({
      mainType: "series",
      subType: "tree",
      query: t
    }, function(i) {
      var e = t.dataIndex, o = i.getData().tree, n = o.getNodeByDataIndex(e);
      n.isExpand = !n.isExpand;
    });
  }), r.registerAction({
    type: "treeRoam",
    event: "treeRoam",
    // Here we set 'none' instead of 'update', because roam action
    // just need to update the transform matrix without having to recalculate
    // the layout. So don't need to go through the whole update process, such
    // as 'dataPrcocess', 'coordSystemUpdate', 'layout' and so on.
    update: "none"
  }, function(t, a, i) {
    a.eachComponent({
      mainType: "series",
      subType: "tree",
      query: t
    }, function(e) {
      var o = e.coordinateSystem, n = p(o, t, e.get("scaleLimit"));
      e.setCenter(n.center), e.setZoom(n.zoom);
    });
  });
}
export {
  d as installTreeAction
};
