// echarts@6.0.0/lib/chart/tree/TreeView.js
import { __extends as V } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { concatArray as F, defaults as U } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { getECData as $ } from "../../util/innerStore.js";
import j from "../helper/Symbol.js";
import { radialCoordinate as w } from "./layoutHelper.js";
import { fromPoints as q } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/bbox.js";
import J from "../../coord/View.js";
import { updateController as K } from "../../component/helper/roamHelper.js";
import Q from "../../component/helper/RoamController.js";
import { parsePercent as W } from "../../util/number.js";
import b from "../../view/Chart.js";
import M from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Path.js";
import { HOVER_STATE_BLUR as ee, setStatesFlag as te, setStatesStylesFromModel as ae, setDefaultStateProxy as re } from "../../util/states.js";
import H from "../../visual/tokens.js";
import ne from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import oe from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/BezierCurve.js";
import { updateProps as I, removeElement as T } from "../../animation/basicTransition.js";
var ie = (
  /** @class */
  /* @__PURE__ */ (function() {
    function i() {
      this.parentPoint = [], this.childPoints = [];
    }
    return i;
  })()
), le = (
  /** @class */
  (function(i) {
    V(a, i);
    function a(e) {
      return i.call(this, e) || this;
    }
    return a.prototype.getDefaultStyle = function() {
      return {
        stroke: H.color.neutral99,
        fill: null
      };
    }, a.prototype.getDefaultShape = function() {
      return new ie();
    }, a.prototype.buildPath = function(e, n) {
      var c = n.childPoints, v = c.length, t = n.parentPoint, o = c[0], h = c[v - 1];
      if (v === 1) {
        e.moveTo(t[0], t[1]), e.lineTo(o[0], o[1]);
        return;
      }
      var p = n.orient, r = p === "TB" || p === "BT" ? 0 : 1, u = 1 - r, l = W(n.forkPosition, 1), d = [];
      d[r] = t[r], d[u] = t[u] + (h[u] - t[u]) * l, e.moveTo(t[0], t[1]), e.lineTo(d[0], d[1]), e.moveTo(o[0], o[1]), d[r] = o[r], e.lineTo(d[0], d[1]), d[r] = h[r], e.lineTo(d[0], d[1]), e.lineTo(h[0], h[1]);
      for (var y = 1; y < v - 1; y++) {
        var f = c[y];
        e.moveTo(f[0], f[1]), d[r] = f[r], e.lineTo(d[0], d[1]);
      }
    }, a;
  })(M)
), Te = (
  /** @class */
  (function(i) {
    V(a, i);
    function a() {
      var e = i !== null && i.apply(this, arguments) || this;
      return e.type = a.type, e._mainGroup = new ne(), e;
    }
    return a.prototype.init = function(e, n) {
      this._controller = new Q(n.getZr()), this._controllerHost = {
        target: this.group
      }, this.group.add(this._mainGroup);
    }, a.prototype.render = function(e, n, c) {
      var v = e.getData(), t = e.layoutInfo, o = this._mainGroup, h = e.get("layout");
      h === "radial" ? (o.x = t.x + t.width / 2, o.y = t.y + t.height / 2) : (o.x = t.x, o.y = t.y), this._updateViewCoordSys(e, c), this._updateController(e, null, n, c);
      var p = this._data;
      v.diff(p).add(function(r) {
        O(v, r) && Y(v, r, null, o, e);
      }).update(function(r, u) {
        var l = p.getItemGraphicEl(u);
        if (!O(v, r)) {
          l && E(p, u, l, o, e);
          return;
        }
        Y(v, r, l, o, e);
      }).remove(function(r) {
        var u = p.getItemGraphicEl(r);
        u && E(p, r, u, o, e);
      }).execute(), this._nodeScaleRatio = e.get("nodeScaleRatio"), this._updateNodeAndLinkScale(e), e.get("expandAndCollapse") === !0 && v.eachItemGraphicEl(function(r, u) {
        r.off("click").on("click", function() {
          c.dispatchAction({
            type: "treeExpandAndCollapse",
            seriesId: e.id,
            dataIndex: u
          });
        });
      }), this._data = v;
    }, a.prototype._updateViewCoordSys = function(e, n) {
      var c = e.getData(), v = [];
      c.each(function(u) {
        var l = c.getItemLayout(u);
        l && !isNaN(l.x) && !isNaN(l.y) && v.push([+l.x, +l.y]);
      });
      var t = [], o = [];
      q(v, t, o);
      var h = this._min, p = this._max;
      o[0] - t[0] === 0 && (t[0] = h ? h[0] : t[0] - 1, o[0] = p ? p[0] : o[0] + 1), o[1] - t[1] === 0 && (t[1] = h ? h[1] : t[1] - 1, o[1] = p ? p[1] : o[1] + 1);
      var r = e.coordinateSystem = new J(null, {
        api: n,
        ecModel: e.ecModel
      });
      r.zoomLimit = e.get("scaleLimit"), r.setBoundingRect(t[0], t[1], o[0] - t[0], o[1] - t[1]), r.setCenter(e.get("center")), r.setZoom(e.get("zoom")), this.group.attr({
        x: r.x,
        y: r.y,
        scaleX: r.scaleX,
        scaleY: r.scaleY
      }), this._min = t, this._max = o;
    }, a.prototype._updateController = function(e, n, c, v) {
      var t = this;
      K(e, v, this.group, this._controller, this._controllerHost, n), this._controller.on("zoom", function(o) {
        t._updateNodeAndLinkScale(e);
      });
    }, a.prototype._updateNodeAndLinkScale = function(e) {
      var n = e.getData(), c = this._getNodeGlobalScale(e);
      n.eachItemGraphicEl(function(v, t) {
        v.setSymbolScale(c);
      });
    }, a.prototype._getNodeGlobalScale = function(e) {
      var n = e.coordinateSystem;
      if (n.type !== "view")
        return 1;
      var c = this._nodeScaleRatio, v = n.scaleX || 1, t = n.getZoom(), o = (t - 1) * c + 1;
      return o / v;
    }, a.prototype.dispose = function() {
      this._controller && this._controller.dispose(), this._controllerHost = null;
    }, a.prototype.remove = function() {
      this._mainGroup.removeAll(), this._data = null;
    }, a.type = "tree", a;
  })(b)
);
function O(i, a) {
  var e = i.getItemLayout(a);
  return e && !isNaN(e.x) && !isNaN(e.y);
}
function Y(i, a, e, n, c) {
  var v = !e, t = i.tree.getNodeByDataIndex(a), o = t.getModel(), h = t.getVisual("style").fill, p = t.isExpand === !1 && t.children.length !== 0 ? h : H.color.neutral00, r = i.tree.root, u = t.parentNode === r ? t : t.parentNode || t, l = i.getItemGraphicEl(u.dataIndex), d = u.getLayout(), y = l ? {
    x: l.__oldX,
    y: l.__oldY,
    rawX: l.__radialOldRawX,
    rawY: l.__radialOldRawY
  } : d, f = t.getLayout();
  v ? (e = new j(i, a, null, {
    symbolInnerColor: p,
    useNameLabel: !0
  }), e.x = y.x, e.y = y.y) : e.updateData(i, a, null, {
    symbolInnerColor: p,
    useNameLabel: !0
  }), e.__radialOldRawX = e.__radialRawX, e.__radialOldRawY = e.__radialRawY, e.__radialRawX = f.rawX, e.__radialRawY = f.rawY, n.add(e), i.setItemGraphicEl(a, e), e.__oldX = e.x, e.__oldY = e.y, I(e, {
    x: f.x,
    y: f.y
  }, c);
  var s = e.getSymbolPath();
  if (c.get("layout") === "radial") {
    var _ = r.children[0], m = _.getLayout(), S = _.children.length, g = void 0, x = void 0;
    if (f.x === m.x && t.isExpand === !0 && _.children.length) {
      var N = {
        x: (_.children[0].getLayout().x + _.children[S - 1].getLayout().x) / 2,
        y: (_.children[0].getLayout().y + _.children[S - 1].getLayout().y) / 2
      };
      g = Math.atan2(N.y - m.y, N.x - m.x), g < 0 && (g = Math.PI * 2 + g), x = N.x < m.x, x && (g = g - Math.PI);
    } else
      g = Math.atan2(f.y - m.y, f.x - m.x), g < 0 && (g = Math.PI * 2 + g), t.children.length === 0 || t.children.length !== 0 && t.isExpand === !1 ? (x = f.x < m.x, x && (g = g - Math.PI)) : (x = f.x > m.x, x || (g = g - Math.PI));
    var Z = x ? "left" : "right", C = o.getModel("label"), G = C.get("rotate"), z = G * (Math.PI / 180), k = s.getTextContent();
    k && (s.setTextConfig({
      position: C.get("position") || Z,
      rotation: G == null ? -g : z,
      origin: "center"
    }), k.setStyle("verticalAlign", "middle"));
  }
  var P = o.get(["emphasis", "focus"]), R = P === "relative" ? F(t.getAncestorsIndices(), t.getDescendantIndices()) : P === "ancestor" ? t.getAncestorsIndices() : P === "descendant" ? t.getDescendantIndices() : null;
  R && ($(e).focus = R), ce(c, t, r, e, y, d, f, n), e.__edge && (e.onHoverStateChange = function(X) {
    if (X !== "blur") {
      var A = t.parentNode && i.getItemGraphicEl(t.parentNode.dataIndex);
      A && A.hoverState === ee || te(e.__edge, X);
    }
  });
}
function ce(i, a, e, n, c, v, t, o) {
  var h = a.getModel(), p = i.get("edgeShape"), r = i.get("layout"), u = i.getOrient(), l = i.get(["lineStyle", "curveness"]), d = i.get("edgeForkPosition"), y = h.getModel("lineStyle").getLineStyle(), f = n.__edge;
  if (p === "curve")
    a.parentNode && a.parentNode !== e && (f || (f = n.__edge = new oe({
      shape: D(r, u, l, c, c)
    })), I(f, {
      shape: D(r, u, l, v, t)
    }, i));
  else if (p === "polyline") {
    if (r === "orthogonal") {
      if (a !== e && a.children && a.children.length !== 0 && a.isExpand === !0) {
        for (var s = a.children, _ = [], m = 0; m < s.length; m++) {
          var S = s[m].getLayout();
          _.push([S.x, S.y]);
        }
        f || (f = n.__edge = new le({
          shape: {
            parentPoint: [t.x, t.y],
            childPoints: [[t.x, t.y]],
            orient: u,
            forkPosition: d
          }
        })), I(f, {
          shape: {
            parentPoint: [t.x, t.y],
            childPoints: _
          }
        }, i);
      }
    } else if (process.env.NODE_ENV !== "production")
      throw new Error("The polyline edgeShape can only be used in orthogonal layout");
  }
  f && !(p === "polyline" && !a.isExpand) && (f.useStyle(U({
    strokeNoScale: !0,
    fill: null
  }, y)), ae(f, h, "lineStyle"), re(f), o.add(f));
}
function B(i, a, e, n, c) {
  var v = a.tree.root, t = L(v, i), o = t.source, h = t.sourceLayout, p = a.getItemGraphicEl(i.dataIndex);
  if (p) {
    var r = a.getItemGraphicEl(o.dataIndex), u = r.__edge, l = p.__edge || (o.isExpand === !1 || o.children.length === 1 ? u : void 0), d = n.get("edgeShape"), y = n.get("layout"), f = n.get("orient"), s = n.get(["lineStyle", "curveness"]);
    l && (d === "curve" ? T(l, {
      shape: D(y, f, s, h, h),
      style: {
        opacity: 0
      }
    }, n, {
      cb: function() {
        e.remove(l);
      },
      removeOpt: c
    }) : d === "polyline" && n.get("layout") === "orthogonal" && T(l, {
      shape: {
        parentPoint: [h.x, h.y],
        childPoints: [[h.x, h.y]]
      },
      style: {
        opacity: 0
      }
    }, n, {
      cb: function() {
        e.remove(l);
      },
      removeOpt: c
    }));
  }
}
function L(i, a) {
  for (var e = a.parentNode === i ? a : a.parentNode || a, n; n = e.getLayout(), n == null; )
    e = e.parentNode === i ? e : e.parentNode || e;
  return {
    source: e,
    sourceLayout: n
  };
}
function E(i, a, e, n, c) {
  var v = i.tree.getNodeByDataIndex(a), t = i.tree.root, o = L(t, v).sourceLayout, h = {
    duration: c.get("animationDurationUpdate"),
    easing: c.get("animationEasingUpdate")
  };
  T(e, {
    x: o.x + 1,
    y: o.y + 1
  }, c, {
    cb: function() {
      n.remove(e), i.setItemGraphicEl(a, null);
    },
    removeOpt: h
  }), e.fadeOut(null, i.hostModel, {
    fadeLabel: !0,
    animation: h
  }), v.children.forEach(function(p) {
    B(p, i, n, c, h);
  }), B(v, i, n, c, h);
}
function D(i, a, e, n, c) {
  var v, t, o, h, p, r, u, l;
  if (i === "radial") {
    p = n.rawX, u = n.rawY, r = c.rawX, l = c.rawY;
    var d = w(p, u), y = w(p, u + (l - u) * e), f = w(r, l + (u - l) * e), s = w(r, l);
    return {
      x1: d.x || 0,
      y1: d.y || 0,
      x2: s.x || 0,
      y2: s.y || 0,
      cpx1: y.x || 0,
      cpy1: y.y || 0,
      cpx2: f.x || 0,
      cpy2: f.y || 0
    };
  } else
    p = n.x, u = n.y, r = c.x, l = c.y, (a === "LR" || a === "RL") && (v = p + (r - p) * e, t = u, o = r + (p - r) * e, h = l), (a === "TB" || a === "BT") && (v = p, t = u + (l - u) * e, o = r, h = l + (u - l) * e);
  return {
    x1: p,
    y1: u,
    x2: r,
    y2: l,
    cpx1: v,
    cpy1: t,
    cpx2: o,
    cpy2: h
  };
}
export {
  Te as default
};
