// echarts@6.0.0/lib/chart/tree/treeLayout.js
import { eachAfter as N, eachBefore as m } from "./traversalHelper.js";
import { init as T, separation as C, firstWalk as s, secondWalk as I, radialCoordinate as P } from "./layoutHelper.js";
import { createBoxLayoutReference as W, getLayoutRect as A } from "../../util/layout.js";
function j(e, R) {
  e.eachSeriesByType("tree", function(c) {
    D(c, R);
  });
}
function D(e, R) {
  var c = W(e, R).refContainer, g = A(e.getBoxLayoutParams(), c);
  e.layoutInfo = g;
  var w = e.get("layout"), p = 0, v = 0, x = null;
  w === "radial" ? (p = 2 * Math.PI, v = Math.min(g.height, g.width) / 2, x = C(function(a, y) {
    return (a.parentNode === y.parentNode ? 1 : 2) / a.depth;
  })) : (p = g.width, v = g.height, x = C());
  var d = e.getData().tree.root, t = d.children[0];
  if (t) {
    T(d), N(t, s, x), d.hierNode.modifier = -t.hierNode.prelim, m(t, I);
    var o = t, r = t, l = t;
    m(t, function(a) {
      var y = a.getLayout().x;
      y < o.getLayout().x && (o = a), y > r.getLayout().x && (r = a), a.depth > l.depth && (l = a);
    });
    var B = o === r ? 1 : x(o, r) / 2, L = B - o.getLayout().x, u = 0, i = 0, f = 0, h = 0;
    if (w === "radial")
      u = p / (r.getLayout().x + B + L), i = v / (l.depth - 1 || 1), m(t, function(a) {
        f = (a.getLayout().x + L) * u, h = (a.depth - 1) * i;
        var y = P(f, h);
        a.setLayout({
          x: y.x,
          y: y.y,
          rawX: f,
          rawY: h
        }, !0);
      });
    else {
      var n = e.getOrient();
      n === "RL" || n === "LR" ? (i = v / (r.getLayout().x + B + L), u = p / (l.depth - 1 || 1), m(t, function(a) {
        h = (a.getLayout().x + L) * i, f = n === "LR" ? (a.depth - 1) * u : p - (a.depth - 1) * u, a.setLayout({
          x: f,
          y: h
        }, !0);
      })) : (n === "TB" || n === "BT") && (u = p / (r.getLayout().x + B + L), i = v / (l.depth - 1 || 1), m(t, function(a) {
        f = (a.getLayout().x + L) * u, h = n === "TB" ? (a.depth - 1) * i : v - (a.depth - 1) * i, a.setLayout({
          x: f,
          y: h
        }, !0);
      }));
    }
  }
}
export {
  j as default
};
