// echarts@6.0.0/lib/chart/tree/TreeSeries.js
import { __extends as f } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import g from "../../model/Series.js";
import v from "../../data/Tree.js";
import y from "../../model/Model.js";
import { createTooltipMarkup as D } from "../../component/tooltip/tooltipMarkup.js";
import { wrapTreePathInfo as T } from "../helper/treeHelper.js";
import x from "../../visual/tokens.js";
var C = (
  /** @class */
  (function(s) {
    f(t, s);
    function t() {
      var e = s !== null && s.apply(this, arguments) || this;
      return e.hasSymbolVisual = !0, e.ignoreStyleOnData = !0, e;
    }
    return t.prototype.getInitialData = function(e) {
      var n = {
        name: e.name,
        children: e.data
      }, p = e.leaves || {}, m = new y(p, this, this.ecModel), i = v.createTree(n, this, r);
      function r(a) {
        a.wrapMethod("getItemModel", function(l, c) {
          var u = i.getNodeByDataIndex(c);
          return u && u.children.length && u.isExpand || (l.parentModel = m), l;
        });
      }
      var o = 0;
      i.eachNode("preorder", function(a) {
        a.depth > o && (o = a.depth);
      });
      var d = e.expandAndCollapse, h = d && e.initialTreeDepth >= 0 ? e.initialTreeDepth : o;
      return i.root.eachNode("preorder", function(a) {
        var l = a.hostTree.data.getRawDataItem(a.dataIndex);
        a.isExpand = l && l.collapsed != null ? !l.collapsed : a.depth <= h;
      }), i.data;
    }, t.prototype.getOrient = function() {
      var e = this.get("orient");
      return e === "horizontal" ? e = "LR" : e === "vertical" && (e = "TB"), e;
    }, t.prototype.setZoom = function(e) {
      this.option.zoom = e;
    }, t.prototype.setCenter = function(e) {
      this.option.center = e;
    }, t.prototype.formatTooltip = function(e, n, p) {
      for (var m = this.getData().tree, i = m.root.children[0], r = m.getNodeByDataIndex(e), o = r.getValue(), d = r.name; r && r !== i; )
        d = r.parentNode.name + "." + d, r = r.parentNode;
      return D("nameValue", {
        name: d,
        value: o,
        noValue: isNaN(o) || o == null
      });
    }, t.prototype.getDataParams = function(e) {
      var n = s.prototype.getDataParams.apply(this, arguments), p = this.getData().tree.getNodeByDataIndex(e);
      return n.treeAncestors = T(p, this), n.collapsed = !p.isExpand, n;
    }, t.type = "series.tree", t.layoutMode = "box", t.defaultOption = {
      // zlevel: 0,
      z: 2,
      // `coordinateSystem` can be declared as 'matrix', 'calendar',
      //  which provides box layout container.
      coordinateSystemUsage: "box",
      // the position of the whole view
      left: "12%",
      top: "12%",
      right: "12%",
      bottom: "12%",
      // the layout of the tree, two value can be selected, 'orthogonal' or 'radial'
      layout: "orthogonal",
      // value can be 'polyline'
      edgeShape: "curve",
      edgeForkPosition: "50%",
      // true | false | 'move' | 'scale', see module:component/helper/RoamController.
      roam: !1,
      roamTrigger: "global",
      // Symbol size scale ratio in roam
      nodeScaleRatio: 0.4,
      // Default on center of graph
      center: null,
      zoom: 1,
      orient: "LR",
      symbol: "emptyCircle",
      symbolSize: 7,
      expandAndCollapse: !0,
      initialTreeDepth: 2,
      lineStyle: {
        color: x.color.borderTint,
        width: 1.5,
        curveness: 0.5
      },
      itemStyle: {
        color: "lightsteelblue",
        // borderColor: '#c23531',
        borderWidth: 1.5
      },
      label: {
        show: !0
      },
      animationEasing: "linear",
      animationDuration: 700,
      animationDurationUpdate: 500
    }, t;
  })(g)
);
export {
  C as default
};
