// echarts@6.0.0/lib/chart/funnel/FunnelSeries.js
import { __extends as m } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { bind as n, curry as p } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import h from "../helper/createSeriesDataSimply.js";
import { defaultEmphasis as u } from "../../util/model.js";
import { makeSeriesEncodeForNameBased as d } from "../../data/helper/sourceHelper.js";
import c from "../../visual/LegendVisualProvider.js";
import f from "../../model/Series.js";
import s from "../../visual/tokens.js";
var x = (
  /** @class */
  (function(o) {
    m(t, o);
    function t() {
      var e = o !== null && o.apply(this, arguments) || this;
      return e.type = t.type, e;
    }
    return t.prototype.init = function(e) {
      o.prototype.init.apply(this, arguments), this.legendVisualProvider = new c(n(this.getData, this), n(this.getRawData, this)), this._defaultLabelLine(e);
    }, t.prototype.getInitialData = function(e, r) {
      return h(this, {
        coordDimensions: ["value"],
        encodeDefaulter: p(d, this)
      });
    }, t.prototype._defaultLabelLine = function(e) {
      u(e, "labelLine", ["show"]);
      var r = e.labelLine, i = e.emphasis.labelLine;
      r.show = r.show && e.label.show, i.show = i.show && e.emphasis.label.show;
    }, t.prototype.getDataParams = function(e) {
      var r = this.getData(), i = o.prototype.getDataParams.call(this, e), a = r.mapDimension("value"), l = r.getSum(a);
      return i.percent = l ? +(r.get(a, e) / l * 100).toFixed(2) : 0, i.$vars.push("percent"), i;
    }, t.type = "series.funnel", t.defaultOption = {
      coordinateSystemUsage: "box",
      // zlevel: 0,                  // 一级层叠
      z: 2,
      legendHoverLink: !0,
      colorBy: "data",
      left: 80,
      top: 60,
      right: 80,
      bottom: 65,
      // width: {totalWidth} - left - right,
      // height: {totalHeight} - top - bottom,
      // 默认取数据最小最大值
      // min: 0,
      // max: 100,
      minSize: "0%",
      maxSize: "100%",
      sort: "descending",
      orient: "vertical",
      gap: 0,
      funnelAlign: "center",
      label: {
        show: !0,
        position: "outer"
        // formatter: 标签文本格式器，同Tooltip.formatter，不支持异步回调
      },
      labelLine: {
        show: !0,
        length: 20,
        lineStyle: {
          // color: 各异,
          width: 1
        }
      },
      itemStyle: {
        // color: 各异,
        borderColor: s.color.neutral00,
        borderWidth: 1
      },
      emphasis: {
        label: {
          show: !0
        }
      },
      select: {
        itemStyle: {
          borderColor: s.color.primary
        }
      }
    }, t;
  })(f)
);
export {
  x as default
};
