// echarts@6.0.0/lib/chart/funnel/funnelLayout.js
import { createBoxLayoutReference as q, getLayoutRect as G } from "../../util/layout.js";
import { parsePercent as D, linearMap as X } from "../../util/number.js";
import { isFunction as J } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
function K(m, h) {
  for (var e = m.mapDimension("value"), s = m.mapArray(e, function(g) {
    return g;
  }), u = [], z = h === "ascending", r = 0, d = m.count(); r < d; r++)
    u[r] = r;
  return J(h) ? u.sort(h) : h !== "none" && u.sort(function(g, t) {
    return z ? s[g] - s[t] : s[t] - s[g];
  }), u;
}
function Q(m) {
  var h = m.hostModel, e = h.get("orient");
  m.each(function(s) {
    var u = m.getItemModel(s), z = u.getModel("label"), r = z.get("position"), d = u.getModel("labelLine"), g = m.getItemLayout(s), t = g.points, p = r === "inner" || r === "inside" || r === "center" || r === "insideLeft" || r === "insideRight", o, n, i, P;
    if (p)
      r === "insideLeft" ? (n = (t[0][0] + t[3][0]) / 2 + 5, i = (t[0][1] + t[3][1]) / 2, o = "left") : r === "insideRight" ? (n = (t[1][0] + t[2][0]) / 2 - 5, i = (t[1][1] + t[2][1]) / 2, o = "right") : (n = (t[0][0] + t[1][0] + t[2][0] + t[3][0]) / 4, i = (t[0][1] + t[1][1] + t[2][1] + t[3][1]) / 4, o = "center"), P = [[n, i], [n, i]];
    else {
      var l = void 0, a = void 0, v = void 0, c = void 0, f = d.get("length");
      process.env.NODE_ENV !== "production" && (e === "vertical" && ["top", "bottom"].indexOf(r) > -1 && (r = "left", console.warn("Position error: Funnel chart on vertical orient dose not support top and bottom.")), e === "horizontal" && ["left", "right"].indexOf(r) > -1 && (r = "bottom", console.warn("Position error: Funnel chart on horizontal orient dose not support left and right."))), r === "left" ? (l = (t[3][0] + t[0][0]) / 2, a = (t[3][1] + t[0][1]) / 2, v = l - f, n = v - 5, o = "right") : r === "right" ? (l = (t[1][0] + t[2][0]) / 2, a = (t[1][1] + t[2][1]) / 2, v = l + f, n = v + 5, o = "left") : r === "top" ? (l = (t[3][0] + t[0][0]) / 2, a = (t[3][1] + t[0][1]) / 2, c = a - f, i = c - 5, o = "center") : r === "bottom" ? (l = (t[1][0] + t[2][0]) / 2, a = (t[1][1] + t[2][1]) / 2, c = a + f, i = c + 5, o = "center") : r === "rightTop" ? (l = e === "horizontal" ? t[3][0] : t[1][0], a = e === "horizontal" ? t[3][1] : t[1][1], e === "horizontal" ? (c = a - f, i = c - 5, o = "center") : (v = l + f, n = v + 5, o = "top")) : r === "rightBottom" ? (l = t[2][0], a = t[2][1], e === "horizontal" ? (c = a + f, i = c + 5, o = "center") : (v = l + f, n = v + 5, o = "bottom")) : r === "leftTop" ? (l = t[0][0], a = e === "horizontal" ? t[0][1] : t[1][1], e === "horizontal" ? (c = a - f, i = c - 5, o = "center") : (v = l - f, n = v - 5, o = "right")) : r === "leftBottom" ? (l = e === "horizontal" ? t[1][0] : t[3][0], a = e === "horizontal" ? t[1][1] : t[2][1], e === "horizontal" ? (c = a + f, i = c + 5, o = "center") : (v = l - f, n = v - 5, o = "right")) : (l = (t[1][0] + t[2][0]) / 2, a = (t[1][1] + t[2][1]) / 2, e === "horizontal" ? (c = a + f, i = c + 5, o = "center") : (v = l + f, n = v + 5, o = "left")), e === "horizontal" ? (v = l, n = v) : (c = a, i = c), P = [[l, a], [v, c]];
    }
    g.label = {
      linePoints: P,
      x: n,
      y: i,
      verticalAlign: "middle",
      textAlign: o,
      inside: p
    };
  });
}
function H(m, h) {
  m.eachSeriesByType("funnel", function(e) {
    var s = e.getData(), u = s.mapDimension("value"), z = e.get("sort"), r = q(e, h), d = G(e.getBoxLayoutParams(), r.refContainer), g = e.get("orient"), t = d.width, p = d.height, o = K(s, z), n = d.x, i = d.y, P = g === "horizontal" ? [D(e.get("minSize"), p), D(e.get("maxSize"), p)] : [D(e.get("minSize"), t), D(e.get("maxSize"), t)], l = s.getDataExtent(u), a = e.get("min"), v = e.get("max");
    a == null && (a = Math.min(l[0], 0)), v == null && (v = l[1]);
    var c = e.get("funnelAlign"), f = e.get("gap"), Y = g === "horizontal" ? t : p, R = (Y - f * (s.count() - 1)) / s.count(), B = function(V, w) {
      if (g === "horizontal") {
        var _ = s.get(u, V) || 0, T = X(_, [a, v], P, !0), k = void 0;
        switch (c) {
          case "top":
            k = i;
            break;
          case "center":
            k = i + (p - T) / 2;
            break;
          case "bottom":
            k = i + (p - T);
            break;
        }
        return [[w, k], [w, k + T]];
      }
      var j = s.get(u, V) || 0, N = X(j, [a, v], P, !0), I;
      switch (c) {
        case "left":
          I = n;
          break;
        case "center":
          I = n + (t - N) / 2;
          break;
        case "right":
          I = n + t - N;
          break;
      }
      return [[I, w], [I + N, w]];
    };
    z === "ascending" && (R = -R, f = -f, g === "horizontal" ? n += t : i += p, o = o.reverse());
    for (var E = 0; E < o.length; E++) {
      var b = o[E], W = o[E + 1], C = s.getItemModel(b);
      if (g === "horizontal") {
        var S = C.get(["itemStyle", "width"]);
        S == null ? S = R : (S = D(S, t), z === "ascending" && (S = -S));
        var F = B(b, n), O = B(W, n + S);
        n += S + f, s.setItemLayout(b, {
          points: F.concat(O.slice().reverse())
        });
      } else {
        var A = C.get(["itemStyle", "height"]);
        A == null ? A = R : (A = D(A, p), z === "ascending" && (A = -A));
        var F = B(b, i), O = B(W, i + A);
        i += A + f, s.setItemLayout(b, {
          points: F.concat(O.slice().reverse())
        });
      }
    }
    Q(s);
  });
}
export {
  H as default
};
