// echarts@6.0.0/lib/chart/funnel/FunnelView.js
import { __extends as h } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { setStatesStylesFromModel as b, toggleHoverEmphasis as L } from "../../util/states.js";
import M from "../../view/Chart.js";
import { setLabelLineStyle as S, getLabelLineStatesModels as P } from "../../label/labelGuideHelper.js";
import { setLabelStyle as I, getLabelStatesModels as F } from "../../label/labelStyle.js";
import { removeElementWithFadeOut as T, saveOldStyle as C, initProps as G, updateProps as g } from "../../animation/basicTransition.js";
import w from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Polyline.js";
import D from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Text.js";
import E from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/Point.js";
import _ from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Polygon.js";
var x = ["itemStyle", "opacity"], A = (
  /** @class */
  (function(u) {
    h(r, u);
    function r(e, l) {
      var a = u.call(this) || this, t = a, n = new w(), p = new D();
      return t.setTextContent(p), a.setTextGuideLine(n), a.updateData(e, l, !0), a;
    }
    return r.prototype.updateData = function(e, l, a) {
      var t = this, n = e.hostModel, p = e.getItemModel(l), i = e.getItemLayout(l), s = p.getModel("emphasis"), o = p.get(x);
      o = o ?? 1, a || C(t), t.useStyle(e.getItemVisual(l, "style")), t.style.lineJoin = "round", a ? (t.setShape({
        points: i.points
      }), t.style.opacity = 0, G(t, {
        style: {
          opacity: o
        }
      }, n, l)) : g(t, {
        style: {
          opacity: o
        },
        shape: {
          points: i.points
        }
      }, n, l), b(t, p), this._updateLabel(e, l), L(this, s.get("focus"), s.get("blurScope"), s.get("disabled"));
    }, r.prototype._updateLabel = function(e, l) {
      var a = this, t = this.getTextGuideLine(), n = a.getTextContent(), p = e.hostModel, i = e.getItemModel(l), s = e.getItemLayout(l), o = s.label, y = e.getItemVisual(l, "style"), v = y.fill;
      I(
        // position will not be used in setLabelStyle
        n,
        F(i),
        {
          labelFetcher: e.hostModel,
          labelDataIndex: l,
          defaultOpacity: y.opacity,
          defaultText: e.getName(l)
        },
        {
          normal: {
            align: o.textAlign,
            verticalAlign: o.verticalAlign
          }
        }
      );
      var d = i.getModel("label"), f = d.get("color"), c = f === "inherit" ? v : null;
      a.setTextConfig({
        local: !0,
        inside: !!o.inside,
        insideStroke: c,
        outsideFill: c
      });
      var m = o.linePoints;
      t.setShape({
        points: m
      }), a.textGuideLineConfig = {
        anchor: m ? new E(m[0][0], m[0][1]) : null
      }, g(n, {
        style: {
          x: o.x,
          y: o.y
        }
      }, p, l), n.attr({
        rotation: o.rotation,
        originX: o.x,
        originY: o.y,
        z2: 10
      }), S(a, P(i), {
        // Default use item visual color
        stroke: v
      });
    }, r;
  })(_)
), X = (
  /** @class */
  (function(u) {
    h(r, u);
    function r() {
      var e = u !== null && u.apply(this, arguments) || this;
      return e.type = r.type, e.ignoreLabelLineUpdate = !0, e;
    }
    return r.prototype.render = function(e, l, a) {
      var t = e.getData(), n = this._data, p = this.group;
      t.diff(n).add(function(i) {
        var s = new A(t, i);
        t.setItemGraphicEl(i, s), p.add(s);
      }).update(function(i, s) {
        var o = n.getItemGraphicEl(s);
        o.updateData(t, i), p.add(o), t.setItemGraphicEl(i, o);
      }).remove(function(i) {
        var s = n.getItemGraphicEl(i);
        T(s, e, i);
      }).execute(), this._data = t;
    }, r.prototype.remove = function() {
      this.group.removeAll(), this._data = null;
    }, r.prototype.dispose = function() {
    }, r.type = "funnel", r;
  })(M)
);
export {
  X as default
};
