// recharts@2.15.0/es6/shape/Rectangle.js
import y, { useRef as W, useState as Z, useEffect as C } from "react";
import { clsx as H } from "../../../../../clsx@2.1.1/node_modules/clsx/dist/clsx.js";
import P from "../../../../../react-smooth@4.0.4_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/react-smooth/es6/Animate.js";
import { filterProps as M } from "../util/ReactUtils.js";
function b(r) {
  "@babel/helpers - typeof";
  return b = typeof Symbol == "function" && typeof Symbol.iterator == "symbol" ? function(n) {
    return typeof n;
  } : function(n) {
    return n && typeof Symbol == "function" && n.constructor === Symbol && n !== Symbol.prototype ? "symbol" : typeof n;
  }, b(r);
}
function g() {
  return g = Object.assign ? Object.assign.bind() : function(r) {
    for (var n = 1; n < arguments.length; n++) {
      var t = arguments[n];
      for (var c in t)
        Object.prototype.hasOwnProperty.call(t, c) && (r[c] = t[c]);
    }
    return r;
  }, g.apply(this, arguments);
}
function K(r, n) {
  return q(r) || $(r, n) || Y(r, n) || X();
}
function X() {
  throw new TypeError(`Invalid attempt to destructure non-iterable instance.
In order to be iterable, non-array objects must have a [Symbol.iterator]() method.`);
}
function Y(r, n) {
  if (r) {
    if (typeof r == "string") return E(r, n);
    var t = Object.prototype.toString.call(r).slice(8, -1);
    if (t === "Object" && r.constructor && (t = r.constructor.name), t === "Map" || t === "Set") return Array.from(r);
    if (t === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t)) return E(r, n);
  }
}
function E(r, n) {
  (n == null || n > r.length) && (n = r.length);
  for (var t = 0, c = new Array(n); t < n; t++) c[t] = r[t];
  return c;
}
function $(r, n) {
  var t = r == null ? null : typeof Symbol < "u" && r[Symbol.iterator] || r["@@iterator"];
  if (t != null) {
    var c, e, u, f, p = [], i = !0, l = !1;
    try {
      if (u = (t = t.call(r)).next, n !== 0) for (; !(i = (c = u.call(t)).done) && (p.push(c.value), p.length !== n); i = !0) ;
    } catch (o) {
      l = !0, e = o;
    } finally {
      try {
        if (!i && t.return != null && (f = t.return(), Object(f) !== f)) return;
      } finally {
        if (l) throw e;
      }
    }
    return p;
  }
}
function q(r) {
  if (Array.isArray(r)) return r;
}
function L(r, n) {
  var t = Object.keys(r);
  if (Object.getOwnPropertySymbols) {
    var c = Object.getOwnPropertySymbols(r);
    n && (c = c.filter(function(e) {
      return Object.getOwnPropertyDescriptor(r, e).enumerable;
    })), t.push.apply(t, c);
  }
  return t;
}
function D(r) {
  for (var n = 1; n < arguments.length; n++) {
    var t = arguments[n] != null ? arguments[n] : {};
    n % 2 ? L(Object(t), !0).forEach(function(c) {
      z(r, c, t[c]);
    }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(r, Object.getOwnPropertyDescriptors(t)) : L(Object(t)).forEach(function(c) {
      Object.defineProperty(r, c, Object.getOwnPropertyDescriptor(t, c));
    });
  }
  return r;
}
function z(r, n, t) {
  return n = F(n), n in r ? Object.defineProperty(r, n, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : r[n] = t, r;
}
function F(r) {
  var n = G(r, "string");
  return b(n) == "symbol" ? n : n + "";
}
function G(r, n) {
  if (b(r) != "object" || !r) return r;
  var t = r[Symbol.toPrimitive];
  if (t !== void 0) {
    var c = t.call(r, n);
    if (b(c) != "object") return c;
    throw new TypeError("@@toPrimitive must return a primitive value.");
  }
  return (n === "string" ? String : Number)(r);
}
var T = function(n, t, c, e, u) {
  var f = Math.min(Math.abs(c) / 2, Math.abs(e) / 2), p = e >= 0 ? 1 : -1, i = c >= 0 ? 1 : -1, l = e >= 0 && c >= 0 || e < 0 && c < 0 ? 1 : 0, o;
  if (f > 0 && u instanceof Array) {
    for (var a = [0, 0, 0, 0], s = 0, A = 4; s < A; s++)
      a[s] = u[s] > f ? f : u[s];
    o = "M".concat(n, ",").concat(t + p * a[0]), a[0] > 0 && (o += "A ".concat(a[0], ",").concat(a[0], ",0,0,").concat(l, ",").concat(n + i * a[0], ",").concat(t)), o += "L ".concat(n + c - i * a[1], ",").concat(t), a[1] > 0 && (o += "A ".concat(a[1], ",").concat(a[1], ",0,0,").concat(l, `,
        `).concat(n + c, ",").concat(t + p * a[1])), o += "L ".concat(n + c, ",").concat(t + e - p * a[2]), a[2] > 0 && (o += "A ".concat(a[2], ",").concat(a[2], ",0,0,").concat(l, `,
        `).concat(n + c - i * a[2], ",").concat(t + e)), o += "L ".concat(n + i * a[3], ",").concat(t + e), a[3] > 0 && (o += "A ".concat(a[3], ",").concat(a[3], ",0,0,").concat(l, `,
        `).concat(n, ",").concat(t + e - p * a[3])), o += "Z";
  } else if (f > 0 && u === +u && u > 0) {
    var m = Math.min(f, u);
    o = "M ".concat(n, ",").concat(t + p * m, `
            A `).concat(m, ",").concat(m, ",0,0,").concat(l, ",").concat(n + i * m, ",").concat(t, `
            L `).concat(n + c - i * m, ",").concat(t, `
            A `).concat(m, ",").concat(m, ",0,0,").concat(l, ",").concat(n + c, ",").concat(t + p * m, `
            L `).concat(n + c, ",").concat(t + e - p * m, `
            A `).concat(m, ",").concat(m, ",0,0,").concat(l, ",").concat(n + c - i * m, ",").concat(t + e, `
            L `).concat(n + i * m, ",").concat(t + e, `
            A `).concat(m, ",").concat(m, ",0,0,").concat(l, ",").concat(n, ",").concat(t + e - p * m, " Z");
  } else
    o = "M ".concat(n, ",").concat(t, " h ").concat(c, " v ").concat(e, " h ").concat(-c, " Z");
  return o;
}, d = function(n, t) {
  if (!n || !t)
    return !1;
  var c = n.x, e = n.y, u = t.x, f = t.y, p = t.width, i = t.height;
  if (Math.abs(p) > 0 && Math.abs(i) > 0) {
    var l = Math.min(u, u + p), o = Math.max(u, u + p), a = Math.min(f, f + i), s = Math.max(f, f + i);
    return c >= l && c <= o && e >= a && e <= s;
  }
  return !1;
}, J = {
  x: 0,
  y: 0,
  width: 0,
  height: 0,
  // The radius of border
  // The radius of four corners when radius is a number
  // The radius of left-top, right-top, right-bottom, left-bottom when radius is an array
  radius: 0,
  isAnimationActive: !1,
  isUpdateAnimationActive: !1,
  animationBegin: 0,
  animationDuration: 1500,
  animationEasing: "ease"
}, R = function(n) {
  var t = D(D({}, J), n), c = W(), e = Z(-1), u = K(e, 2), f = u[0], p = u[1];
  C(function() {
    if (c.current && c.current.getTotalLength)
      try {
        var v = c.current.getTotalLength();
        v && p(v);
      } catch {
      }
  }, []);
  var i = t.x, l = t.y, o = t.width, a = t.height, s = t.radius, A = t.className, m = t.animationEasing, h = t.animationDuration, I = t.animationBegin, x = t.isAnimationActive, O = t.isUpdateAnimationActive;
  if (i !== +i || l !== +l || o !== +o || a !== +a || o === 0 || a === 0)
    return null;
  var S = H("recharts-rectangle", A);
  return O ? /* @__PURE__ */ y.createElement(P, {
    canBegin: f > 0,
    from: {
      width: o,
      height: a,
      x: i,
      y: l
    },
    to: {
      width: o,
      height: a,
      x: i,
      y: l
    },
    duration: h,
    animationEasing: m,
    isActive: O
  }, function(v) {
    var j = v.width, N = v.height, B = v.x, U = v.y;
    return /* @__PURE__ */ y.createElement(P, {
      canBegin: f > 0,
      from: "0px ".concat(f === -1 ? 1 : f, "px"),
      to: "".concat(f, "px 0px"),
      attributeName: "strokeDasharray",
      begin: I,
      duration: h,
      isActive: x,
      easing: m
    }, /* @__PURE__ */ y.createElement("path", g({}, M(t, !0), {
      className: S,
      d: T(B, U, j, N, s),
      ref: c
    })));
  }) : /* @__PURE__ */ y.createElement("path", g({}, M(t, !0), {
    className: S,
    d: T(i, l, o, a, s)
  }));
};
export {
  R as Rectangle,
  d as isInRectangle
};
