// echarts@6.0.0/lib/loading/default.js
import { defaults as u } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import g from "../visual/tokens.js";
import v from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import c from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Rect.js";
import z from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Text.js";
import S from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Arc.js";
var a = Math.PI;
function R(l, e) {
  e = e || {}, u(e, {
    text: "loading",
    textColor: g.color.primary,
    fontSize: 12,
    fontWeight: "normal",
    fontStyle: "normal",
    fontFamily: "sans-serif",
    maskColor: "rgba(255,255,255,0.8)",
    showSpinner: !0,
    color: g.color.theme[0],
    spinnerRadius: 10,
    lineWidth: 5,
    zlevel: 0
  });
  var t = new v(), o = new c({
    style: {
      fill: e.maskColor
    },
    zlevel: e.zlevel,
    z: 1e4
  });
  t.add(o);
  var h = new z({
    style: {
      text: e.text,
      fill: e.textColor,
      fontSize: e.fontSize,
      fontWeight: e.fontWeight,
      fontStyle: e.fontStyle,
      fontFamily: e.fontFamily
    },
    zlevel: e.zlevel,
    z: 10001
  }), d = new c({
    style: {
      fill: "none"
    },
    textContent: h,
    textConfig: {
      position: "right",
      distance: 10
    },
    zlevel: e.zlevel,
    z: 10001
  });
  t.add(d);
  var i;
  return e.showSpinner && (i = new S({
    shape: {
      startAngle: -a / 2,
      endAngle: -a / 2 + 0.1,
      r: e.spinnerRadius
    },
    style: {
      stroke: e.color,
      lineCap: "round",
      lineWidth: e.lineWidth
    },
    zlevel: e.zlevel,
    z: 10001
  }), i.animateShape(!0).when(1e3, {
    endAngle: a * 3 / 2
  }).start("circularInOut"), i.animateShape(!0).when(1e3, {
    startAngle: a * 3 / 2
  }).delay(300).start("circularInOut"), t.add(i)), t.resize = function() {
    var n = h.getBoundingRect().width, r = e.showSpinner ? e.spinnerRadius : 0, f = (l.getWidth() - r * 2 - (e.showSpinner && n ? 10 : 0) - n) / 2 - (e.showSpinner && n ? 0 : 5 + n / 2) + (e.showSpinner ? 0 : n / 2) + (n ? 0 : r), m = l.getHeight() / 2;
    e.showSpinner && i.setShape({
      cx: f,
      cy: m
    }), d.setShape({
      x: f - r,
      y: m - r,
      width: r * 2,
      height: r * 2
    }), o.setShape({
      x: 0,
      y: 0,
      width: l.getWidth(),
      height: l.getHeight()
    });
  }, t.resize(), t;
}
export {
  R as default
};
