// wheel-gestures@2.2.48/dist/wheel-gestures.esm.js
function p() {
  return p = Object.assign || function(t) {
    for (var r = 1; r < arguments.length; r++) {
      var a = arguments[r];
      for (var o in a)
        Object.prototype.hasOwnProperty.call(a, o) && (t[o] = a[o]);
    }
    return t;
  }, p.apply(this, arguments);
}
var $ = 0.996, q = function(r, a) {
  return a === void 0 && (a = $), r * a / (1 - a);
};
function J(t) {
  return t[t.length - 1];
}
function K(t) {
  return t.reduce(function(r, a) {
    return r + a;
  }) / t.length;
}
var Q = function(r, a, o) {
  return Math.min(Math.max(a, r), o);
};
function S(t, r) {
  if (t.length !== r.length)
    throw new Error("vectors must be same length");
  return t.map(function(a, o) {
    return a + r[o];
  });
}
function y(t) {
  return Math.max.apply(Math, t.map(Math.abs));
}
function h(t) {
  return Object.freeze(t), Object.values(t).forEach(function(r) {
    r !== null && typeof r == "object" && !Object.isFrozen(r) && h(r);
  }), t;
}
function k() {
  var t = {};
  function r(l, c) {
    return t[l] = (t[l] || []).concat(c), function() {
      return a(l, c);
    };
  }
  function a(l, c) {
    t[l] = (t[l] || []).filter(function(e) {
      return e !== c;
    });
  }
  function o(l, c) {
    l in t && t[l].forEach(function(e) {
      return e(c);
    });
  }
  return h({
    on: r,
    off: a,
    dispatch: o
  });
}
function ee(t) {
  var r = [], a = function(e) {
    return e.addEventListener("wheel", t, {
      passive: !1
    }), r.push(e), function() {
      return o(e);
    };
  }, o = function(e) {
    e.removeEventListener("wheel", t), r = r.filter(function(M) {
      return M !== e;
    });
  }, l = function() {
    r.forEach(o);
  };
  return h({
    observe: a,
    unobserve: o,
    disconnect: l
  });
}
var te = 16 * 1.125, ne = typeof window < "u" && window.innerHeight || 800, T = [1, te, ne];
function re(t) {
  var r = t.deltaX * T[t.deltaMode], a = t.deltaY * T[t.deltaMode], o = (t.deltaZ || 0) * T[t.deltaMode];
  return {
    timeStamp: t.timeStamp,
    axisDelta: [r, a, o]
  };
}
var ae = [-1, -1, -1];
function ie(t, r) {
  if (!r)
    return t;
  var a = r === !0 ? ae : r.map(function(o) {
    return o ? -1 : 1;
  });
  return p({}, t, {
    axisDelta: t.axisDelta.map(function(o, l) {
      return o * a[l];
    })
  });
}
var j = 700, oe = function(r) {
  return p({}, r, {
    axisDelta: r.axisDelta.map(function(a) {
      return Q(a, -j, j);
    })
  });
}, D = process.env.NODE_ENV !== "production", ue = 0.6, se = 0.96, le = 2, w = 5, I = /* @__PURE__ */ h({
  preventWheelAction: !0,
  reverseSign: [!0, !0, !1]
}), ce = 400;
function F() {
  return {
    isStarted: !1,
    isStartPublished: !1,
    isMomentum: !1,
    startTime: 0,
    lastAbsDelta: 1 / 0,
    axisMovement: [0, 0, 0],
    axisVelocity: [0, 0, 0],
    accelerationFactors: [],
    scrollPoints: [],
    scrollPointsToMerge: [],
    willEndTimeout: ce
  };
}
function fe(t) {
  t === void 0 && (t = {});
  var r = k(), a = r.on, o = r.off, l = r.dispatch, c = I, e = F(), M, E = !1, g, A = function(n) {
    Array.isArray(n) ? n.forEach(function(i) {
      return P(i);
    }) : P(n);
  }, _ = function(n) {
    return n === void 0 && (n = {}), Object.values(n).some(function(i) {
      return i == null;
    }) ? (D && console.error("updateOptions ignored! undefined & null options not allowed"), c) : c = h(p({}, I, c, n));
  }, b = function(n) {
    var i = p({
      event: M,
      isStart: !1,
      isEnding: !1,
      isMomentumCancel: !1,
      isMomentum: e.isMomentum,
      axisDelta: [0, 0, 0],
      axisVelocity: e.axisVelocity,
      axisMovement: e.axisMovement,
      get axisMovementProjection() {
        return S(i.axisMovement, i.axisVelocity.map(function(s) {
          return q(s);
        }));
      }
    }, n);
    l("wheel", p({}, i, {
      previous: g
    })), g = i;
  }, L = function(n, i) {
    var s = c, f = s.preventWheelAction, v = i[0], m = i[1], d = i[2];
    if (typeof f == "boolean") return f;
    switch (f) {
      case "x":
        return Math.abs(v) >= n;
      case "y":
        return Math.abs(m) >= n;
      case "z":
        return Math.abs(d) >= n;
      default:
        return D && console.warn("unsupported preventWheelAction value: " + f, "warn"), !1;
    }
  }, P = function(n) {
    var i = oe(ie(re(n), c.reverseSign)), s = i.axisDelta, f = i.timeStamp, v = y(s);
    if (n.preventDefault && L(v, s) && n.preventDefault(), e.isStarted ? e.isMomentum && v > Math.max(2, e.lastAbsDelta * 2) && (V(!0), O()) : O(), v === 0 && Object.is && Object.is(n.deltaX, -0)) {
      E = !0;
      return;
    }
    M = n, e.axisMovement = S(e.axisMovement, s), e.lastAbsDelta = v, e.scrollPointsToMerge.push({
      axisDelta: s,
      timeStamp: f
    }), C(), b({
      axisDelta: s,
      isStart: !e.isStartPublished
    }), e.isStartPublished = !0, X();
  }, C = function() {
    e.scrollPointsToMerge.length === le ? (e.scrollPoints.unshift({
      axisDeltaSum: e.scrollPointsToMerge.map(function(n) {
        return n.axisDelta;
      }).reduce(S),
      timeStamp: K(e.scrollPointsToMerge.map(function(n) {
        return n.timeStamp;
      }))
    }), z(), e.scrollPointsToMerge.length = 0, e.scrollPoints.length = 1, e.isMomentum || R()) : e.isStartPublished || N();
  }, N = function() {
    e.axisVelocity = J(e.scrollPointsToMerge).axisDelta.map(function(n) {
      return n / e.willEndTimeout;
    });
  }, z = function() {
    var n = e.scrollPoints, i = n[0], s = n[1];
    if (!(!s || !i)) {
      var f = i.timeStamp - s.timeStamp;
      if (f <= 0) {
        D && console.warn("invalid deltaTime");
        return;
      }
      var v = i.axisDeltaSum.map(function(d) {
        return d / f;
      }), m = v.map(function(d, B) {
        return d / (e.axisVelocity[B] || 1);
      });
      e.axisVelocity = v, e.accelerationFactors.push(m), H(f);
    }
  }, H = function(n) {
    var i = Math.ceil(n / 10) * 10 * 1.2;
    e.isMomentum || (i = Math.max(100, i * 2)), e.willEndTimeout = Math.min(1e3, Math.round(i));
  }, G = function(n) {
    return n === 0 ? !0 : n <= se && n >= ue;
  }, R = function() {
    if (e.accelerationFactors.length >= w) {
      if (E && (E = !1, y(e.axisVelocity) >= 0.2)) {
        W();
        return;
      }
      var n = e.accelerationFactors.slice(w * -1), i = n.every(function(s) {
        var f = !!s.reduce(function(m, d) {
          return m && m < 1 && m === d ? 1 : 0;
        }), v = s.filter(G).length === s.length;
        return f || v;
      });
      i && W(), e.accelerationFactors = n;
    }
  }, W = function() {
    e.isMomentum = !0;
  }, O = function() {
    e = F(), e.isStarted = !0, e.startTime = Date.now(), g = void 0, E = !1;
  }, X = /* @__PURE__ */ (function() {
    var u;
    return function() {
      clearTimeout(u), u = setTimeout(V, e.willEndTimeout);
    };
  })(), V = function(n) {
    n === void 0 && (n = !1), e.isStarted && (e.isMomentum && n ? b({
      isEnding: !0,
      isMomentumCancel: !0
    }) : b({
      isEnding: !0
    }), e.isMomentum = !1, e.isStarted = !1);
  }, x = ee(A), Z = x.observe, U = x.unobserve, Y = x.disconnect;
  return _(t), h({
    on: a,
    off: o,
    observe: Z,
    unobserve: U,
    disconnect: Y,
    feedWheel: A,
    updateOptions: _
  });
}
export {
  fe as WheelGestures,
  y as absMax,
  S as addVectors,
  K as average,
  Q as clamp,
  I as configDefaults,
  h as deepFreeze,
  fe as default,
  J as lastOf,
  q as projection
};
