// @atlaskit/pragmatic-drag-and-drop@1.7.4/dist/esm/adapter/element-adapter.js
import P from "../../../../../../../@babel_runtime@7.28.6/node_modules/@babel/runtime/helpers/esm/slicedToArray.js";
import { d as k } from "../../../../../../../../../_virtual/index4.js";
import { getElementFromPointWithoutHoneypot as _ } from "../honey-pot-fix/get-element-from-point-without-honey-pot.js";
import { makeHoneyPotFix as w } from "../honey-pot-fix/make-honey-pot-fix.js";
import { makeAdapter as A } from "../make-adapter/make-adapter.js";
import { combine as H } from "../public-utils/combine.js";
import { once as R } from "../public-utils/once.js";
import { addAttribute as $ } from "../util/add-attribute.js";
import { isAndroid as M, androidFallbackText as O } from "../util/android.js";
import { getInput as S } from "../util/get-input.js";
import { textMediaType as h } from "../util/media-types/text-media-type.js";
import { URLMediaType as V } from "../util/media-types/url-media-type.js";
import { elementAdapterNativeDataKey as K } from "./element-adapter-native-data-key.js";
var i = /* @__PURE__ */ new WeakMap();
function L(t) {
  return i.set(t.element, t), function() {
    i.delete(t.element);
  };
}
var b = w(), v = A({
  typeKey: "element",
  defaultDropEffect: "move",
  mount: function(r) {
    return H(b.bindEvents(), k.bind(document, {
      type: "dragstart",
      listener: function(e) {
        var l, o, d, u, m, s;
        if (r.canStart(e) && !e.defaultPrevented) {
          if (!e.dataTransfer) {
            process.env.NODE_ENV !== "production" && console.warn(`
              It appears as though you have are not testing DragEvents correctly.

              - If you are unit testing, ensure you have polyfilled DragEvent.
              - If you are browser testing, ensure you are dispatching drag events correctly.

              Please see our testing guides for more information:
              https://atlassian.design/components/pragmatic-drag-and-drop/core-package/testing
            `.replace(/ {2}/g, ""));
            return;
          }
          var y = e.target;
          if (!(y instanceof HTMLElement))
            return null;
          var a = i.get(y);
          if (!a)
            return null;
          var g = S(e), p = {
            element: a.element,
            dragHandle: (l = a.dragHandle) !== null && l !== void 0 ? l : null,
            input: g
          };
          if (a.canDrag && !a.canDrag(p))
            return e.preventDefault(), null;
          if (a.dragHandle) {
            var I = _({
              x: g.clientX,
              y: g.clientY
            });
            if (!a.dragHandle.contains(I))
              return e.preventDefault(), null;
          }
          var D = (o = (d = a.getInitialDataForExternal) === null || d === void 0 ? void 0 : d.call(a, p)) !== null && o !== void 0 ? o : null;
          if (D)
            for (var c = 0, E = Object.entries(D); c < E.length; c++) {
              var T = P(E[c], 2), N = T[0], f = T[1];
              e.dataTransfer.setData(N, f ?? "");
            }
          M() && !e.dataTransfer.types.includes(h) && !e.dataTransfer.types.includes(V) && e.dataTransfer.setData(h, O), e.dataTransfer.setData(K, "");
          var x = {
            element: a.element,
            dragHandle: (u = a.dragHandle) !== null && u !== void 0 ? u : null,
            data: (m = (s = a.getInitialData) === null || s === void 0 ? void 0 : s.call(a, p)) !== null && m !== void 0 ? m : {}
          }, F = {
            type: "element",
            payload: x,
            startedFrom: "internal"
          };
          r.start({
            event: e,
            dragType: F
          });
        }
      }
    }));
  },
  dispatchEventToSource: function(r) {
    var n, e, l = r.eventName, o = r.payload;
    (n = i.get(o.source.element)) === null || n === void 0 || (e = n[l]) === null || e === void 0 || e.call(
      n,
      // I cannot seem to get the types right here.
      // TS doesn't seem to like that one event can need `nativeSetDragImage`
      // @ts-expect-error
      o
    );
  },
  onPostDispatch: b.getOnPostDispatch()
}), ee = v.dropTarget, te = v.monitor;
function ae(t) {
  if (process.env.NODE_ENV !== "production" && t.dragHandle && !t.element.contains(t.dragHandle) && console.warn("Drag handle element must be contained in draggable element", {
    element: t.element,
    dragHandle: t.dragHandle
  }), process.env.NODE_ENV !== "production") {
    var r = i.get(t.element);
    r && console.warn("You have already registered a `draggable` on the same element", {
      existing: r,
      proposed: t
    });
  }
  var n = H(
    // making the draggable register the adapter rather than drop targets
    // this is because you *must* have a draggable element to start a drag
    // but you _might_ not have any drop targets immediately
    // (You might create drop targets async)
    v.registerUsage(),
    L(t),
    $(t.element, {
      attribute: "draggable",
      value: "true"
    })
  );
  return R(n);
}
export {
  ae as draggable,
  ee as dropTargetForElements,
  te as monitorForElements
};
