// mdast-util-to-hast@13.2.0/lib/handlers/index.js
import { blockquote as r } from "./blockquote.js";
import { hardBreak as m } from "./break.js";
import { code as t } from "./code.js";
import { strikethrough as e } from "./delete.js";
import { emphasis as i } from "./emphasis.js";
import { footnoteReference as f } from "./footnote-reference.js";
import { heading as p } from "./heading.js";
import { html as n } from "./html.js";
import { imageReference as a } from "./image-reference.js";
import { image as l } from "./image.js";
import { inlineCode as d } from "./inline-code.js";
import { linkReference as h } from "./link-reference.js";
import { link as c } from "./link.js";
import { listItem as s } from "./list-item.js";
import { list as g } from "./list.js";
import { paragraph as k } from "./paragraph.js";
import { root as b } from "./root.js";
import { strong as u } from "./strong.js";
import { table as R } from "./table.js";
import { tableRow as x } from "./table-row.js";
import { tableCell as B } from "./table-cell.js";
import { text as C } from "./text.js";
import { thematicBreak as q } from "./thematic-break.js";
const V = {
  blockquote: r,
  break: m,
  code: t,
  delete: e,
  emphasis: i,
  footnoteReference: f,
  heading: p,
  html: n,
  imageReference: a,
  image: l,
  inlineCode: d,
  linkReference: h,
  link: c,
  listItem: s,
  list: g,
  paragraph: k,
  // @ts-expect-error: root is different, but hard to type.
  root: b,
  strong: u,
  table: R,
  tableCell: B,
  tableRow: x,
  text: C,
  thematicBreak: q,
  toml: o,
  yaml: o,
  definition: o,
  footnoteDefinition: o
};
function o() {
}
export {
  V as handlers
};
