// echarts@6.0.0/lib/coord/polar/Polar.js
import g from "./RadiusAxis.js";
import p from "./AngleAxis.js";
var v = ["radius", "angle"], _ = (
  /** @class */
  (function() {
    function e(i) {
      this.dimensions = v, this.type = "polar", this.cx = 0, this.cy = 0, this._radiusAxis = new g(), this._angleAxis = new p(), this.axisPointerEnabled = !0, this.name = i || "", this._radiusAxis.polar = this._angleAxis.polar = this;
    }
    return e.prototype.containPoint = function(i) {
      var t = this.pointToCoord(i);
      return this._radiusAxis.contain(t[0]) && this._angleAxis.contain(t[1]);
    }, e.prototype.containData = function(i) {
      return this._radiusAxis.containData(i[0]) && this._angleAxis.containData(i[1]);
    }, e.prototype.getAxis = function(i) {
      var t = "_" + i + "Axis";
      return this[t];
    }, e.prototype.getAxes = function() {
      return [this._radiusAxis, this._angleAxis];
    }, e.prototype.getAxesByScale = function(i) {
      var t = [], r = this._angleAxis, s = this._radiusAxis;
      return r.scale.type === i && t.push(r), s.scale.type === i && t.push(s), t;
    }, e.prototype.getAngleAxis = function() {
      return this._angleAxis;
    }, e.prototype.getRadiusAxis = function() {
      return this._radiusAxis;
    }, e.prototype.getOtherAxis = function(i) {
      var t = this._angleAxis;
      return i === t ? this._radiusAxis : t;
    }, e.prototype.getBaseAxis = function() {
      return this.getAxesByScale("ordinal")[0] || this.getAxesByScale("time")[0] || this.getAngleAxis();
    }, e.prototype.getTooltipAxes = function(i) {
      var t = i != null && i !== "auto" ? this.getAxis(i) : this.getBaseAxis();
      return {
        baseAxes: [t],
        otherAxes: [this.getOtherAxis(t)]
      };
    }, e.prototype.dataToPoint = function(i, t, r) {
      return this.coordToPoint([this._radiusAxis.dataToRadius(i[0], t), this._angleAxis.dataToAngle(i[1], t)], r);
    }, e.prototype.pointToData = function(i, t, r) {
      r = r || [];
      var s = this.pointToCoord(i);
      return r[0] = this._radiusAxis.radiusToData(s[0], t), r[1] = this._angleAxis.angleToData(s[1], t), r;
    }, e.prototype.pointToCoord = function(i) {
      var t = i[0] - this.cx, r = i[1] - this.cy, s = this.getAngleAxis(), a = s.getExtent(), o = Math.min(a[0], a[1]), h = Math.max(a[0], a[1]);
      s.inverse ? o = h - 360 : h = o + 360;
      var x = Math.sqrt(t * t + r * r);
      t /= x, r /= x;
      for (var n = Math.atan2(-r, t) / Math.PI * 180, u = n < o ? 1 : -1; n < o || n > h; )
        n += u * 360;
      return [x, n];
    }, e.prototype.coordToPoint = function(i, t) {
      t = t || [];
      var r = i[0], s = i[1] / 180 * Math.PI;
      return t[0] = Math.cos(s) * r + this.cx, t[1] = -Math.sin(s) * r + this.cy, t;
    }, e.prototype.getArea = function() {
      var i = this.getAngleAxis(), t = this.getRadiusAxis(), r = t.getExtent().slice();
      r[0] > r[1] && r.reverse();
      var s = i.getExtent(), a = Math.PI / 180, o = 1e-4;
      return {
        cx: this.cx,
        cy: this.cy,
        r0: r[0],
        r: r[1],
        startAngle: -s[0] * a,
        endAngle: -s[1] * a,
        clockwise: i.inverse,
        contain: function(h, x) {
          var n = h - this.cx, u = x - this.cy, l = n * n + u * u, A = this.r, c = this.r0;
          return A !== c && l - o <= A * A && l + o >= c * c;
        },
        // As the bounding box
        x: this.cx - r[1],
        y: this.cy - r[1],
        width: r[1] * 2,
        height: r[1] * 2
      };
    }, e.prototype.convertToPixel = function(i, t, r) {
      var s = d(t);
      return s === this ? this.dataToPoint(r) : null;
    }, e.prototype.convertFromPixel = function(i, t, r) {
      var s = d(t);
      return s === this ? this.pointToData(r) : null;
    }, e;
  })()
);
function d(e) {
  var i = e.seriesModel, t = e.polarModel;
  return t && t.coordinateSystem || i && i.coordinateSystem;
}
export {
  _ as default,
  v as polarDimensions
};
