// echarts@6.0.0/lib/coord/polar/polarCreator.js
import { retrieve as p, each as f, isArray as l } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import y, { polarDimensions as d } from "./Polar.js";
import { parsePercent as g } from "../../util/number.js";
import { getDataDimensionsOnAxis as v, niceScaleExtent as m, createScaleByModel as E } from "../axisHelper.js";
import { SINGLE_REFERRING as h } from "../../util/model.js";
import { createBoxLayoutReference as x } from "../../util/layout.js";
function S(t, r, a) {
  var e = r.get("center"), n = x(r, a).refContainer;
  t.cx = g(e[0], n.width) + n.x, t.cy = g(e[1], n.height) + n.y;
  var i = t.getRadiusAxis(), c = Math.min(n.width, n.height) / 2, o = r.get("radius");
  o == null ? o = [0, "100%"] : l(o) || (o = [0, o]);
  var s = [g(o[0], c), g(o[1], c)];
  i.inverse ? i.setExtent(s[1], s[0]) : i.setExtent(s[0], s[1]);
}
function R(t, r) {
  var a = this, e = a.getAngleAxis(), n = a.getRadiusAxis();
  if (e.scale.setExtent(1 / 0, -1 / 0), n.scale.setExtent(1 / 0, -1 / 0), t.eachSeries(function(o) {
    if (o.coordinateSystem === a) {
      var s = o.getData();
      f(v(s, "radius"), function(u) {
        n.scale.unionExtentFromData(s, u);
      }), f(v(s, "angle"), function(u) {
        e.scale.unionExtentFromData(s, u);
      });
    }
  }), m(e.scale, e.model), m(n.scale, n.model), e.type === "category" && !e.onBand) {
    var i = e.getExtent(), c = 360 / e.scale.count();
    e.inverse ? i[1] += c : i[1] -= c, e.setExtent(i[0], i[1]);
  }
}
function I(t) {
  return t.mainType === "angleAxis";
}
function A(t, r) {
  var a;
  if (t.type = r.get("type"), t.scale = E(r), t.onBand = r.get("boundaryGap") && t.type === "category", t.inverse = r.get("inverse"), I(r)) {
    t.inverse = t.inverse !== r.get("clockwise");
    var e = r.get("startAngle"), n = (a = r.get("endAngle")) !== null && a !== void 0 ? a : e + (t.inverse ? -360 : 360);
    t.setExtent(e, n);
  }
  r.axis = t, t.model = r;
}
var F = {
  dimensions: d,
  create: function(t, r) {
    var a = [];
    return t.eachComponent("polar", function(e, n) {
      var i = new y(n + "");
      i.update = R;
      var c = i.getRadiusAxis(), o = i.getAngleAxis(), s = e.findAxisModel("radiusAxis"), u = e.findAxisModel("angleAxis");
      A(c, s), A(o, u), S(i, e, r), a.push(i), e.coordinateSystem = i, i.model = e;
    }), t.eachSeries(function(e) {
      if (e.get("coordinateSystem") === "polar") {
        var n = e.getReferringComponents("polar", h).models[0];
        if (process.env.NODE_ENV !== "production" && !n)
          throw new Error('Polar "' + p(e.get("polarIndex"), e.get("polarId"), 0) + '" not found');
        e.coordinateSystem = n.coordinateSystem;
      }
    }), a;
  }
};
export {
  F as default
};
