// @tiptap/extension-node-range@2.24.0/dist/index.js
import { Extension as x } from "../../../../../@tiptap_core@2.24.0_@tiptap_pm@2.24.0/node_modules/@tiptap/core/dist/index.js";
import { SelectionRange as P, Selection as k, Plugin as A, PluginKey as D } from "../../../../../prosemirror-state@1.4.3/node_modules/prosemirror-state/dist/index.js";
import { DecorationSet as S, Decoration as z } from "../../../../../prosemirror-view@1.38.1/node_modules/prosemirror-view/dist/index.js";
import { NodeRange as E } from "../../../../../prosemirror-model@1.25.0/node_modules/prosemirror-model/dist/index.js";
function v(c) {
  if (!c.length)
    return S.empty;
  const t = [], e = c[0].$from.node(0);
  return c.forEach((n) => {
    const s = n.$from.pos, o = n.$from.nodeAfter;
    o && t.push(z.node(s, s + o.nodeSize, {
      class: "ProseMirror-selectednoderange"
    }));
  }), S.create(e, t);
}
function M(c, t, e) {
  const n = [], s = c.node(0);
  e = typeof e == "number" && e >= 0 ? e : c.sameParent(t) ? Math.max(0, c.sharedDepth(t.pos) - 1) : c.sharedDepth(t.pos);
  const o = new E(c, t, e), r = o.depth === 0 ? 0 : s.resolve(o.start).posAtIndex(0);
  return o.parent.forEach((h, l) => {
    const d = r + l, i = d + h.nodeSize;
    if (d < o.start || d >= o.end)
      return;
    const u = new P(s.resolve(d), s.resolve(i));
    n.push(u);
  }), n;
}
class g {
  constructor(t, e) {
    this.anchor = t, this.head = e;
  }
  map(t) {
    return new g(t.map(this.anchor), t.map(this.head));
  }
  resolve(t) {
    const e = t.resolve(this.anchor), n = t.resolve(this.head);
    return new a(e, n);
  }
}
class a extends k {
  constructor(t, e, n, s = 1) {
    const { doc: o } = t, r = t === e, h = t.pos === o.content.size && e.pos === o.content.size, l = r && !h ? o.resolve(e.pos + (s > 0 ? 1 : -1)) : e, d = r && h ? o.resolve(t.pos - (s > 0 ? 1 : -1)) : t, i = M(d.min(l), d.max(l), n), u = l.pos >= t.pos ? i[0].$from : i[i.length - 1].$to, f = l.pos >= t.pos ? i[i.length - 1].$to : i[0].$from;
    super(u, f, i), this.depth = n;
  }
  // we can safely ignore this TypeScript error: https://github.com/Microsoft/TypeScript/issues/338
  // @ts-ignore
  get $to() {
    return this.ranges[this.ranges.length - 1].$to;
  }
  eq(t) {
    return t instanceof a && t.$from.pos === this.$from.pos && t.$to.pos === this.$to.pos;
  }
  map(t, e) {
    const n = t.resolve(e.map(this.anchor)), s = t.resolve(e.map(this.head));
    return new a(n, s);
  }
  toJSON() {
    return {
      type: "nodeRange",
      anchor: this.anchor,
      head: this.head
    };
  }
  get isForwards() {
    return this.head >= this.anchor;
  }
  get isBackwards() {
    return !this.isForwards;
  }
  extendBackwards() {
    const { doc: t } = this.$from;
    if (this.isForwards && this.ranges.length > 1) {
      const s = this.ranges.slice(0, -1), o = s[0].$from, r = s[s.length - 1].$to;
      return new a(o, r, this.depth);
    }
    const e = this.ranges[0], n = t.resolve(Math.max(0, e.$from.pos - 1));
    return new a(this.$anchor, n, this.depth);
  }
  extendForwards() {
    const { doc: t } = this.$from;
    if (this.isBackwards && this.ranges.length > 1) {
      const s = this.ranges.slice(1), o = s[0].$from, r = s[s.length - 1].$to;
      return new a(r, o, this.depth);
    }
    const e = this.ranges[this.ranges.length - 1], n = t.resolve(Math.min(t.content.size, e.$to.pos + 1));
    return new a(this.$anchor, n, this.depth);
  }
  static fromJSON(t, e) {
    return new a(t.resolve(e.anchor), t.resolve(e.head));
  }
  static create(t, e, n, s, o = 1) {
    return new this(t.resolve(e), t.resolve(n), s, o);
  }
  getBookmark() {
    return new g(this.anchor, this.head);
  }
}
a.prototype.visible = !1;
function p(c) {
  return c instanceof a;
}
x.create({
  name: "nodeRange",
  addOptions() {
    return {
      depth: void 0,
      key: "Mod"
    };
  },
  addKeyboardShortcuts() {
    return {
      // extend NodeRangeSelection upwards
      "Shift-ArrowUp": ({ editor: c }) => {
        const { depth: t } = this.options, { view: e, state: n } = c, { doc: s, selection: o, tr: r } = n, { anchor: h, head: l } = o;
        if (!p(o)) {
          const i = a.create(s, h, l, t, -1);
          return r.setSelection(i), e.dispatch(r), !0;
        }
        const d = o.extendBackwards();
        return r.setSelection(d), e.dispatch(r), !0;
      },
      // extend NodeRangeSelection downwards
      "Shift-ArrowDown": ({ editor: c }) => {
        const { depth: t } = this.options, { view: e, state: n } = c, { doc: s, selection: o, tr: r } = n, { anchor: h, head: l } = o;
        if (!p(o)) {
          const i = a.create(s, h, l, t);
          return r.setSelection(i), e.dispatch(r), !0;
        }
        const d = o.extendForwards();
        return r.setSelection(d), e.dispatch(r), !0;
      },
      // add `NodeRangeSelection` to all nodes
      "Mod-a": ({ editor: c }) => {
        const { depth: t } = this.options, { view: e, state: n } = c, { doc: s, tr: o } = n, r = a.create(s, 0, s.content.size, t);
        return o.setSelection(r), e.dispatch(o), !0;
      }
    };
  },
  onSelectionUpdate() {
    const { selection: c } = this.editor.state;
    p(c) && this.editor.view.dom.classList.add("ProseMirror-noderangeselection");
  },
  addProseMirrorPlugins() {
    let c = !1, t = !1;
    return [
      new A({
        key: new D("nodeRange"),
        props: {
          attributes: () => c ? {
            class: "ProseMirror-noderangeselection"
          } : { class: "" },
          handleDOMEvents: {
            mousedown: (e, n) => {
              const { key: s } = this.options, o = /Mac/.test(navigator.platform), r = !!n.shiftKey, h = !!n.ctrlKey, l = !!n.altKey, d = !!n.metaKey, i = o ? d : h;
              return (s == null || s === "Shift" && r || s === "Control" && h || s === "Alt" && l || s === "Meta" && d || s === "Mod" && i) && (t = !0), t && document.addEventListener("mouseup", () => {
                t = !1;
                const { state: u } = e, { doc: f, selection: R, tr: m } = u, { $anchor: w, $head: $ } = R;
                if (w.sameParent($))
                  return;
                const y = a.create(f, w.pos, $.pos, this.options.depth);
                m.setSelection(y), e.dispatch(m);
              }, { once: !0 }), !1;
            }
          },
          // when selecting some text we want to render some decorations
          // to preview a `NodeRangeSelection`
          decorations: (e) => {
            const { selection: n } = e, s = p(n);
            if (c = !1, !t)
              return s ? (c = !0, v(n.ranges)) : null;
            const { $from: o, $to: r } = n;
            if (!s && o.sameParent(r))
              return null;
            const h = M(o, r, this.options.depth);
            return h.length ? (c = !0, v(h)) : null;
          }
        }
      })
    ];
  }
});
export {
  a as NodeRangeSelection,
  v as getNodeRangeDecorations,
  M as getSelectionRanges,
  p as isNodeRangeSelection
};
