// zrender@6.0.0/lib/graphic/Image.js
import { __extends as a } from "../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import s, { DEFAULT_COMMON_STYLE as p, DEFAULT_COMMON_ANIMATION_PROPS as y } from "./Displayable.js";
import f from "../core/BoundingRect.js";
import { createObject as l, defaults as g } from "../core/util.js";
var _ = g({
  x: 0,
  y: 0
}, p), m = {
  style: g({
    x: !0,
    y: !0,
    width: !0,
    height: !0,
    sx: !0,
    sy: !0,
    sWidth: !0,
    sHeight: !0
  }, y.style)
};
function c(e) {
  return !!(e && typeof e != "string" && e.width && e.height);
}
var A = (function(e) {
  a(r, e);
  function r() {
    return e !== null && e.apply(this, arguments) || this;
  }
  return r.prototype.createStyle = function(t) {
    return l(_, t);
  }, r.prototype._getSize = function(t) {
    var i = this.style, o = i[t];
    if (o != null)
      return o;
    var n = c(i.image) ? i.image : this.__image;
    if (!n)
      return 0;
    var h = t === "width" ? "height" : "width", u = i[h];
    return u == null ? n[t] : n[t] / n[h] * u;
  }, r.prototype.getWidth = function() {
    return this._getSize("width");
  }, r.prototype.getHeight = function() {
    return this._getSize("height");
  }, r.prototype.getAnimationStyleProps = function() {
    return m;
  }, r.prototype.getBoundingRect = function() {
    var t = this.style;
    return this._rect || (this._rect = new f(t.x || 0, t.y || 0, this.getWidth(), this.getHeight())), this._rect;
  }, r;
})(s);
A.prototype.type = "image";
export {
  m as DEFAULT_IMAGE_ANIMATION_PROPS,
  _ as DEFAULT_IMAGE_STYLE,
  A as default
};
