// zrender@6.0.0/lib/graphic/IncrementalDisplayable.js
import { __extends as l } from "../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import p from "./Displayable.js";
import y from "../core/BoundingRect.js";
var h = [], _ = (function(i) {
  l(a, i);
  function a() {
    var t = i !== null && i.apply(this, arguments) || this;
    return t.notClear = !0, t.incremental = !0, t._displayables = [], t._temporaryDisplayables = [], t._cursor = 0, t;
  }
  return a.prototype.traverse = function(t, r) {
    t.call(r, this);
  }, a.prototype.useStyle = function() {
    this.style = {};
  }, a.prototype.getCursor = function() {
    return this._cursor;
  }, a.prototype.innerAfterBrush = function() {
    this._cursor = this._displayables.length;
  }, a.prototype.clearDisplaybles = function() {
    this._displayables = [], this._temporaryDisplayables = [], this._cursor = 0, this.markRedraw(), this.notClear = !1;
  }, a.prototype.clearTemporalDisplayables = function() {
    this._temporaryDisplayables = [];
  }, a.prototype.addDisplayable = function(t, r) {
    r ? this._temporaryDisplayables.push(t) : this._displayables.push(t), this.markRedraw();
  }, a.prototype.addDisplayables = function(t, r) {
    r = r || !1;
    for (var e = 0; e < t.length; e++)
      this.addDisplayable(t[e], r);
  }, a.prototype.getDisplayables = function() {
    return this._displayables;
  }, a.prototype.getTemporalDisplayables = function() {
    return this._temporaryDisplayables;
  }, a.prototype.eachPendingDisplayable = function(t) {
    for (var r = this._cursor; r < this._displayables.length; r++)
      t && t(this._displayables[r]);
    for (var r = 0; r < this._temporaryDisplayables.length; r++)
      t && t(this._temporaryDisplayables[r]);
  }, a.prototype.update = function() {
    this.updateTransform();
    for (var t = this._cursor; t < this._displayables.length; t++) {
      var r = this._displayables[t];
      r.parent = this, r.update(), r.parent = null;
    }
    for (var t = 0; t < this._temporaryDisplayables.length; t++) {
      var r = this._temporaryDisplayables[t];
      r.parent = this, r.update(), r.parent = null;
    }
  }, a.prototype.getBoundingRect = function() {
    if (!this._rect) {
      for (var t = new y(1 / 0, 1 / 0, -1 / 0, -1 / 0), r = 0; r < this._displayables.length; r++) {
        var e = this._displayables[r], s = e.getBoundingRect().clone();
        e.needLocalTransform() && s.applyTransform(e.getLocalTransform(h)), t.union(s);
      }
      this._rect = t;
    }
    return this._rect;
  }, a.prototype.contain = function(t, r) {
    var e = this.transformCoordToLocal(t, r), s = this.getBoundingRect();
    if (s.contain(e[0], e[1]))
      for (var o = 0; o < this._displayables.length; o++) {
        var n = this._displayables[o];
        if (n.contain(t, r))
          return !0;
      }
    return !1;
  }, a;
})(p);
export {
  _ as default
};
