// zrender@6.0.0/lib/graphic/Path.js
import { __extends as v } from "../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import O, { DEFAULT_COMMON_ANIMATION_PROPS as R, DEFAULT_COMMON_STYLE as b } from "./Displayable.js";
import E from "../core/PathProxy.js";
import { containStroke as F, contain as N } from "../contain/path.js";
import { keys as m, extend as l, isString as g, createObject as B, defaults as L, clone as S } from "../core/util.js";
import { lum as P } from "../tool/color.js";
import { DARK_LABEL_COLOR as T, LIGHTER_LABEL_COLOR as x, LIGHT_LABEL_COLOR as A, DARK_MODE_THRESHOLD as I } from "../config.js";
import { REDRAW_BIT as D, SHAPE_CHANGED_BIT as c, STYLE_CHANGED_BIT as C } from "./constants.js";
import { TRANSFORMABLE_PROPS as M } from "../core/Transformable.js";
var H = L({
  fill: "#000",
  stroke: null,
  strokePercent: 1,
  fillOpacity: 1,
  strokeOpacity: 1,
  lineDashOffset: 0,
  lineWidth: 1,
  lineCap: "butt",
  miterLimit: 10,
  strokeNoScale: !1,
  strokeFirst: !1
}, b), U = {
  style: L({
    fill: !0,
    stroke: !0,
    strokePercent: !0,
    fillOpacity: !0,
    strokeOpacity: !0,
    lineDashOffset: !0,
    lineWidth: !0,
    miterLimit: !0
  }, R.style)
}, _ = M.concat([
  "invisible",
  "culling",
  "z",
  "z2",
  "zlevel",
  "parent"
]), Z = (function(p) {
  v(r, p);
  function r(t) {
    return p.call(this, t) || this;
  }
  return r.prototype.update = function() {
    var t = this;
    p.prototype.update.call(this);
    var e = this.style;
    if (e.decal) {
      var i = this._decalEl = this._decalEl || new r();
      i.buildPath === r.prototype.buildPath && (i.buildPath = function(n) {
        t.buildPath(n, t.shape);
      }), i.silent = !0;
      var s = i.style;
      for (var a in e)
        s[a] !== e[a] && (s[a] = e[a]);
      s.fill = e.fill ? e.decal : null, s.decal = null, s.shadowColor = null, e.strokeFirst && (s.stroke = null);
      for (var h = 0; h < _.length; ++h)
        i[_[h]] = this[_[h]];
      i.__dirty |= D;
    } else this._decalEl && (this._decalEl = null);
  }, r.prototype.getDecalElement = function() {
    return this._decalEl;
  }, r.prototype._init = function(t) {
    var e = m(t);
    this.shape = this.getDefaultShape();
    var i = this.getDefaultStyle();
    i && this.useStyle(i);
    for (var s = 0; s < e.length; s++) {
      var a = e[s], h = t[a];
      a === "style" ? this.style ? l(this.style, h) : this.useStyle(h) : a === "shape" ? l(this.shape, h) : p.prototype.attrKV.call(this, a, h);
    }
    this.style || this.useStyle({});
  }, r.prototype.getDefaultStyle = function() {
    return null;
  }, r.prototype.getDefaultShape = function() {
    return {};
  }, r.prototype.canBeInsideText = function() {
    return this.hasFill();
  }, r.prototype.getInsideTextFill = function() {
    var t = this.style.fill;
    if (t !== "none") {
      if (g(t)) {
        var e = P(t, 0);
        return e > 0.5 ? T : e > 0.2 ? x : A;
      } else if (t)
        return A;
    }
    return T;
  }, r.prototype.getInsideTextStroke = function(t) {
    var e = this.style.fill;
    if (g(e)) {
      var i = this.__zr, s = !!(i && i.isDarkMode()), a = P(t, 0) < I;
      if (s === a)
        return e;
    }
  }, r.prototype.buildPath = function(t, e, i) {
  }, r.prototype.pathUpdated = function() {
    this.__dirty &= ~c;
  }, r.prototype.getUpdatedPathProxy = function(t) {
    return !this.path && this.createPathProxy(), this.path.beginPath(), this.buildPath(this.path, this.shape, t), this.path;
  }, r.prototype.createPathProxy = function() {
    this.path = new E(!1);
  }, r.prototype.hasStroke = function() {
    var t = this.style, e = t.stroke;
    return !(e == null || e === "none" || !(t.lineWidth > 0));
  }, r.prototype.hasFill = function() {
    var t = this.style, e = t.fill;
    return e != null && e !== "none";
  }, r.prototype.getBoundingRect = function() {
    var t = this._rect, e = this.style, i = !t;
    if (i) {
      var s = !1;
      this.path || (s = !0, this.createPathProxy());
      var a = this.path;
      (s || this.__dirty & c) && (a.beginPath(), this.buildPath(a, this.shape, !1), this.pathUpdated()), t = a.getBoundingRect();
    }
    if (this._rect = t, this.hasStroke() && this.path && this.path.len() > 0) {
      var h = this._rectStroke || (this._rectStroke = t.clone());
      if (this.__dirty || i) {
        h.copy(t);
        var n = e.strokeNoScale ? this.getLineScale() : 1, o = e.lineWidth;
        if (!this.hasFill()) {
          var f = this.strokeContainThreshold;
          o = Math.max(o, f ?? 4);
        }
        n > 1e-10 && (h.width += o / n, h.height += o / n, h.x -= o / n / 2, h.y -= o / n / 2);
      }
      return h;
    }
    return t;
  }, r.prototype.contain = function(t, e) {
    var i = this.transformCoordToLocal(t, e), s = this.getBoundingRect(), a = this.style;
    if (t = i[0], e = i[1], s.contain(t, e)) {
      var h = this.path;
      if (this.hasStroke()) {
        var n = a.lineWidth, o = a.strokeNoScale ? this.getLineScale() : 1;
        if (o > 1e-10 && (this.hasFill() || (n = Math.max(n, this.strokeContainThreshold)), F(h, n / o, t, e)))
          return !0;
      }
      if (this.hasFill())
        return N(h, t, e);
    }
    return !1;
  }, r.prototype.dirtyShape = function() {
    this.__dirty |= c, this._rect && (this._rect = null), this._decalEl && this._decalEl.dirtyShape(), this.markRedraw();
  }, r.prototype.dirty = function() {
    this.dirtyStyle(), this.dirtyShape();
  }, r.prototype.animateShape = function(t) {
    return this.animate("shape", t);
  }, r.prototype.updateDuringAnimation = function(t) {
    t === "style" ? this.dirtyStyle() : t === "shape" ? this.dirtyShape() : this.markRedraw();
  }, r.prototype.attrKV = function(t, e) {
    t === "shape" ? this.setShape(e) : p.prototype.attrKV.call(this, t, e);
  }, r.prototype.setShape = function(t, e) {
    var i = this.shape;
    return i || (i = this.shape = {}), typeof t == "string" ? i[t] = e : l(i, t), this.dirtyShape(), this;
  }, r.prototype.shapeChanged = function() {
    return !!(this.__dirty & c);
  }, r.prototype.createStyle = function(t) {
    return B(H, t);
  }, r.prototype._innerSaveToNormal = function(t) {
    p.prototype._innerSaveToNormal.call(this, t);
    var e = this._normalState;
    t.shape && !e.shape && (e.shape = l({}, this.shape));
  }, r.prototype._applyStateObj = function(t, e, i, s, a, h) {
    p.prototype._applyStateObj.call(this, t, e, i, s, a, h);
    var n = !(e && s), o;
    if (e && e.shape ? a ? s ? o = e.shape : (o = l({}, i.shape), l(o, e.shape)) : (o = l({}, s ? this.shape : i.shape), l(o, e.shape)) : n && (o = i.shape), o)
      if (a) {
        this.shape = l({}, this.shape);
        for (var f = {}, d = m(o), y = 0; y < d.length; y++) {
          var u = d[y];
          typeof o[u] == "object" ? this.shape[u] = o[u] : f[u] = o[u];
        }
        this._transitionState(t, {
          shape: f
        }, h);
      } else
        this.shape = o, this.dirtyShape();
  }, r.prototype._mergeStates = function(t) {
    for (var e = p.prototype._mergeStates.call(this, t), i, s = 0; s < t.length; s++) {
      var a = t[s];
      a.shape && (i = i || {}, this._mergeStyle(i, a.shape));
    }
    return i && (e.shape = i), e;
  }, r.prototype.getAnimationStyleProps = function() {
    return U;
  }, r.prototype.isZeroArea = function() {
    return !1;
  }, r.extend = function(t) {
    var e = (function(s) {
      v(a, s);
      function a(h) {
        var n = s.call(this, h) || this;
        return t.init && t.init.call(n, h), n;
      }
      return a.prototype.getDefaultStyle = function() {
        return S(t.style);
      }, a.prototype.getDefaultShape = function() {
        return S(t.shape);
      }, a;
    })(r);
    for (var i in t)
      typeof t[i] == "function" && (e.prototype[i] = t[i]);
    return e;
  }, r.initDefaultProps = (function() {
    var t = r.prototype;
    t.type = "path", t.strokeContainThreshold = 5, t.segmentIgnoreThreshold = 0, t.subPixelOptimize = !1, t.autoBatch = !1, t.__dirty = D | C | c;
  })(), r;
})(O);
export {
  U as DEFAULT_PATH_ANIMATION_PROPS,
  H as DEFAULT_PATH_STYLE,
  Z as default
};
