// zrender@6.0.0/lib/graphic/Text.js
import { __extends as J } from "../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { calcInnerTextOverflowArea as N, parsePlainText as Q, tSpanCreateBoundingRect2 as X, parseRichText as $ } from "./helper/parseText.js";
import z from "./TSpan.js";
import { extend as P, keys as tt, defaults as rt, retrieve3 as R, retrieve2 as L, trim as et, each as at, normalizeCssArray as it } from "../core/util.js";
import { adjustTextX as I, adjustTextY as H } from "../contain/text.js";
import ot from "./Image.js";
import nt from "./shape/Rect.js";
import ht from "../core/BoundingRect.js";
import lt, { DEFAULT_COMMON_ANIMATION_PROPS as ft } from "./Displayable.js";
import { DEFAULT_FONT as k, DEFAULT_FONT_SIZE as dt } from "../core/platform.js";
var Z = {
  fill: "#000"
}, U = 2, O = {}, st = {
  style: rt({
    fill: !0,
    stroke: !0,
    fillOpacity: !0,
    strokeOpacity: !0,
    lineWidth: !0,
    fontSize: !0,
    lineHeight: !0,
    width: !0,
    height: !0,
    textShadowColor: !0,
    textShadowBlur: !0,
    textShadowOffsetX: !0,
    textShadowOffsetY: !0,
    backgroundColor: !0,
    padding: !0,
    borderColor: !0,
    borderWidth: !0,
    borderRadius: !0
  }, ft.style)
}, ut = (function(e) {
  J(a, e);
  function a(t) {
    var r = e.call(this) || this;
    return r.type = "text", r._children = [], r._defaultStyle = Z, r.attr(t), r;
  }
  return a.prototype.childrenRef = function() {
    return this._children;
  }, a.prototype.update = function() {
    e.prototype.update.call(this), this.styleChanged() && this._updateSubTexts();
    for (var t = 0; t < this._children.length; t++) {
      var r = this._children[t];
      r.zlevel = this.zlevel, r.z = this.z, r.z2 = this.z2, r.culling = this.culling, r.cursor = this.cursor, r.invisible = this.invisible;
    }
  }, a.prototype.updateTransform = function() {
    var t = this.innerTransformable;
    t ? (t.updateTransform(), t.transform && (this.transform = t.transform)) : e.prototype.updateTransform.call(this);
  }, a.prototype.getLocalTransform = function(t) {
    var r = this.innerTransformable;
    return r ? r.getLocalTransform(t) : e.prototype.getLocalTransform.call(this, t);
  }, a.prototype.getComputedTransform = function() {
    return this.__hostTarget && (this.__hostTarget.getComputedTransform(), this.__hostTarget.updateInnerText(!0)), e.prototype.getComputedTransform.call(this);
  }, a.prototype._updateSubTexts = function() {
    this._childCursor = 0, _t(this.style), this.style.rich ? this._updateRichTexts() : this._updatePlainTexts(), this._children.length = this._childCursor, this.styleUpdated();
  }, a.prototype.addSelfToZr = function(t) {
    e.prototype.addSelfToZr.call(this, t);
    for (var r = 0; r < this._children.length; r++)
      this._children[r].__zr = t;
  }, a.prototype.removeSelfFromZr = function(t) {
    e.prototype.removeSelfFromZr.call(this, t);
    for (var r = 0; r < this._children.length; r++)
      this._children[r].__zr = null;
  }, a.prototype.getBoundingRect = function() {
    if (this.styleChanged() && this._updateSubTexts(), !this._rect) {
      for (var t = new ht(0, 0, 0, 0), r = this._children, f = [], n = null, l = 0; l < r.length; l++) {
        var u = r[l], c = u.getBoundingRect(), i = u.getLocalTransform(f);
        i ? (t.copy(c), t.applyTransform(i), n = n || t.clone(), n.union(t)) : (n = n || c.clone(), n.union(c));
      }
      this._rect = n || t;
    }
    return this._rect;
  }, a.prototype.setDefaultTextStyle = function(t) {
    this._defaultStyle = t || Z;
  }, a.prototype.setTextContent = function(t) {
    if (process.env.NODE_ENV !== "production")
      throw new Error("Can't attach text on another text");
  }, a.prototype._mergeStyle = function(t, r) {
    if (!r)
      return t;
    var f = r.rich, n = t.rich || f && {};
    return P(t, r), f && n ? (this._mergeRich(n, f), t.rich = n) : n && (t.rich = n), t;
  }, a.prototype._mergeRich = function(t, r) {
    for (var f = tt(r), n = 0; n < f.length; n++) {
      var l = f[n];
      t[l] = t[l] || {}, P(t[l], r[l]);
    }
  }, a.prototype.getAnimationStyleProps = function() {
    return st;
  }, a.prototype._getOrCreateChild = function(t) {
    var r = this._children[this._childCursor];
    return (!r || !(r instanceof t)) && (r = new t()), this._children[this._childCursor++] = r, r.__zr = this.__zr, r.parent = this, r;
  }, a.prototype._updatePlainTexts = function() {
    var t = this.style, r = t.font || k, f = t.padding, n = this._defaultStyle, l = t.x || 0, u = t.y || 0, c = t.align || n.align || "left", i = t.verticalAlign || n.verticalAlign || "top";
    N(O, n.overflowRect, l, u, c, i), l = O.baseX, u = O.baseY;
    var T = q(t), s = Q(T, t, O.outerWidth, O.outerHeight), x = Y(t), C = !!t.backgroundColor, w = s.outerHeight, d = s.outerWidth, h = s.lines, p = s.lineHeight;
    this.isTruncated = !!s.isTruncated;
    var v = l, g = H(u, s.contentHeight, i);
    if (x || f) {
      var _ = I(l, d, c), D = H(u, w, i);
      x && this._renderBackground(t, t, _, D, d, w);
    }
    g += p / 2, f && (v = K(l, c, f), i === "top" ? g += f[0] : i === "bottom" && (g -= f[2]));
    for (var m = 0, W = !1, b = !1, y = E("fill" in t ? t.fill : (b = !0, n.fill)), S = G("stroke" in t ? t.stroke : !C && (!n.autoStroke || b) ? (m = U, W = !0, n.stroke) : null), B = t.textShadowBlur > 0, A = 0; A < h.length; A++) {
      var F = this._getOrCreateChild(z), o = F.createStyle();
      F.useStyle(o), o.text = h[A], o.x = v, o.y = g, o.textAlign = c, o.textBaseline = "middle", o.opacity = t.opacity, o.strokeFirst = !0, B && (o.shadowBlur = t.textShadowBlur || 0, o.shadowColor = t.textShadowColor || "transparent", o.shadowOffsetX = t.textShadowOffsetX || 0, o.shadowOffsetY = t.textShadowOffsetY || 0), o.stroke = S, o.fill = y, S && (o.lineWidth = t.lineWidth || m, o.lineDash = t.lineDash, o.lineDashOffset = t.lineDashOffset || 0), o.font = r, V(o, t), g += p, F.setBoundingRect(X(o, s.contentWidth, s.calculatedLineHeight, W ? 0 : null));
    }
  }, a.prototype._updateRichTexts = function() {
    var t = this.style, r = this._defaultStyle, f = t.align || r.align, n = t.verticalAlign || r.verticalAlign, l = t.x || 0, u = t.y || 0;
    N(O, r.overflowRect, l, u, f, n), l = O.baseX, u = O.baseY;
    var c = q(t), i = $(c, t, O.outerWidth, O.outerHeight, f), T = i.width, s = i.outerWidth, x = i.outerHeight, C = t.padding;
    this.isTruncated = !!i.isTruncated;
    var w = I(l, s, f), d = H(u, x, n), h = w, p = d;
    C && (h += C[3], p += C[0]);
    var v = h + T;
    Y(t) && this._renderBackground(t, t, w, d, s, x);
    for (var g = !!t.backgroundColor, _ = 0; _ < i.lines.length; _++) {
      for (var D = i.lines[_], m = D.tokens, W = m.length, b = D.lineHeight, y = D.width, S = 0, B = h, A = v, F = W - 1, o = void 0; S < W && (o = m[S], !o.align || o.align === "left"); )
        this._placeToken(o, t, b, p, B, "left", g), y -= o.width, B += o.width, S++;
      for (; F >= 0 && (o = m[F], o.align === "right"); )
        this._placeToken(o, t, b, p, A, "right", g), y -= o.width, A -= o.width, F--;
      for (B += (T - (B - h) - (v - A) - y) / 2; S <= F; )
        o = m[S], this._placeToken(o, t, b, p, B + o.width / 2, "center", g), B += o.width, S++;
      p += b;
    }
  }, a.prototype._placeToken = function(t, r, f, n, l, u, c) {
    var i = r.rich[t.styleName] || {};
    i.text = t.text;
    var T = t.verticalAlign, s = n + f / 2;
    T === "top" ? s = n + t.height / 2 : T === "bottom" && (s = n + f - t.height / 2);
    var x = !t.isLineHolder && Y(i);
    x && this._renderBackground(i, r, u === "right" ? l - t.width : u === "center" ? l - t.width / 2 : l, s - t.height / 2, t.width, t.height);
    var C = !!i.backgroundColor, w = t.textPadding;
    w && (l = K(l, u, w), s -= t.height / 2 - w[0] - t.innerHeight / 2);
    var d = this._getOrCreateChild(z), h = d.createStyle();
    d.useStyle(h);
    var p = this._defaultStyle, v = !1, g = 0, _ = !1, D = E("fill" in i ? i.fill : "fill" in r ? r.fill : (v = !0, p.fill)), m = G("stroke" in i ? i.stroke : "stroke" in r ? r.stroke : !C && !c && (!p.autoStroke || v) ? (g = U, _ = !0, p.stroke) : null), W = i.textShadowBlur > 0 || r.textShadowBlur > 0;
    h.text = t.text, h.x = l, h.y = s, W && (h.shadowBlur = i.textShadowBlur || r.textShadowBlur || 0, h.shadowColor = i.textShadowColor || r.textShadowColor || "transparent", h.shadowOffsetX = i.textShadowOffsetX || r.textShadowOffsetX || 0, h.shadowOffsetY = i.textShadowOffsetY || r.textShadowOffsetY || 0), h.textAlign = u, h.textBaseline = "middle", h.font = t.font || k, h.opacity = R(i.opacity, r.opacity, 1), V(h, i), m && (h.lineWidth = R(i.lineWidth, r.lineWidth, g), h.lineDash = L(i.lineDash, r.lineDash), h.lineDashOffset = r.lineDashOffset || 0, h.stroke = m), D && (h.fill = D), d.setBoundingRect(X(h, t.contentWidth, t.contentHeight, _ ? 0 : null));
  }, a.prototype._renderBackground = function(t, r, f, n, l, u) {
    var c = t.backgroundColor, i = t.borderWidth, T = t.borderColor, s = c && c.image, x = c && !s, C = t.borderRadius, w = this, d, h;
    if (x || t.lineHeight || i && T) {
      d = this._getOrCreateChild(nt), d.useStyle(d.createStyle()), d.style.fill = null;
      var p = d.shape;
      p.x = f, p.y = n, p.width = l, p.height = u, p.r = C, d.dirtyShape();
    }
    if (x) {
      var v = d.style;
      v.fill = c || null, v.fillOpacity = L(t.fillOpacity, 1);
    } else if (s) {
      h = this._getOrCreateChild(ot), h.onload = function() {
        w.dirtyStyle();
      };
      var g = h.style;
      g.image = c.image, g.x = f, g.y = n, g.width = l, g.height = u;
    }
    if (i && T) {
      var v = d.style;
      v.lineWidth = i, v.stroke = T, v.strokeOpacity = L(t.strokeOpacity, 1), v.lineDash = t.borderDash, v.lineDashOffset = t.borderDashOffset || 0, d.strokeContainThreshold = 0, d.hasFill() && d.hasStroke() && (v.strokeFirst = !0, v.lineWidth *= 2);
    }
    var _ = (d || h).style;
    _.shadowBlur = t.shadowBlur || 0, _.shadowColor = t.shadowColor || "transparent", _.shadowOffsetX = t.shadowOffsetX || 0, _.shadowOffsetY = t.shadowOffsetY || 0, _.opacity = R(t.opacity, r.opacity, 1);
  }, a.makeFont = function(t) {
    var r = "";
    return gt(t) && (r = [
      t.fontStyle,
      t.fontWeight,
      pt(t.fontSize),
      t.fontFamily || "sans-serif"
    ].join(" ")), r && et(r) || t.textFont || t.font;
  }, a;
})(lt), vt = { left: !0, right: 1, center: 1 }, ct = { top: 1, bottom: 1, middle: 1 }, M = ["fontStyle", "fontWeight", "fontSize", "fontFamily"];
function pt(e) {
  return typeof e == "string" && (e.indexOf("px") !== -1 || e.indexOf("rem") !== -1 || e.indexOf("em") !== -1) ? e : isNaN(+e) ? dt + "px" : e + "px";
}
function V(e, a) {
  for (var t = 0; t < M.length; t++) {
    var r = M[t], f = a[r];
    f != null && (e[r] = f);
  }
}
function gt(e) {
  return e.fontSize != null || e.fontFamily || e.fontWeight;
}
function _t(e) {
  return j(e), at(e.rich, j), e;
}
function j(e) {
  if (e) {
    e.font = ut.makeFont(e);
    var a = e.align;
    a === "middle" && (a = "center"), e.align = a == null || vt[a] ? a : "left";
    var t = e.verticalAlign;
    t === "center" && (t = "middle"), e.verticalAlign = t == null || ct[t] ? t : "top";
    var r = e.padding;
    r && (e.padding = it(e.padding));
  }
}
function G(e, a) {
  return e == null || a <= 0 || e === "transparent" || e === "none" ? null : e.image || e.colorStops ? "#000" : e;
}
function E(e) {
  return e == null || e === "none" ? null : e.image || e.colorStops ? "#000" : e;
}
function K(e, a, t) {
  return a === "right" ? e - t[1] : a === "center" ? e + t[3] / 2 - t[1] / 2 : e + t[3];
}
function q(e) {
  var a = e.text;
  return a != null && (a += ""), a;
}
function Y(e) {
  return !!(e.backgroundColor || e.lineHeight || e.borderWidth && e.borderColor);
}
export {
  st as DEFAULT_TEXT_ANIMATION_PROPS,
  ut as default,
  gt as hasSeparateFont,
  _t as normalizeTextStyle,
  pt as parseFontSize
};
