// zrender@6.0.0/lib/graphic/TSpan.js
import { __extends as n } from "../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import i from "./Displayable.js";
import { DEFAULT_PATH_STYLE as p } from "./Path.js";
import { createObject as a, defaults as s } from "../core/util.js";
import { DEFAULT_FONT as l } from "../core/platform.js";
import { tSpanHasStroke as f, tSpanCreateBoundingRect as u } from "./helper/parseText.js";
var c = s({
  strokeFirst: !0,
  font: l,
  x: 0,
  y: 0,
  textAlign: "left",
  textBaseline: "top",
  miterLimit: 2
}, p), y = (function(r) {
  n(t, r);
  function t() {
    return r !== null && r.apply(this, arguments) || this;
  }
  return t.prototype.hasStroke = function() {
    return f(this.style);
  }, t.prototype.hasFill = function() {
    var e = this.style, o = e.fill;
    return o != null && o !== "none";
  }, t.prototype.createStyle = function(e) {
    return a(c, e);
  }, t.prototype.setBoundingRect = function(e) {
    this._rect = e;
  }, t.prototype.getBoundingRect = function() {
    return this._rect || (this._rect = u(this.style)), this._rect;
  }, t.initDefaultProps = (function() {
    var e = t.prototype;
    e.dirtyRectTolerance = 10;
  })(), t;
})(i);
y.prototype.type = "tspan";
export {
  c as DEFAULT_TSPAN_STYLE,
  y as default
};
