// echarts@6.0.0/lib/scale/helper.js
import { getPrecision as v, nice as m, round as s, quantityExponent as p } from "../util/number.js";
import { bind as h } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
function S(r) {
  var a = Math.pow(10, p(Math.abs(r))), i = Math.abs(r / a);
  return i === 0 || i === 1 || i === 2 || i === 3 || i === 5;
}
function P(r) {
  return r.type === "interval" || r.type === "log";
}
function w(r, a, i, l, u) {
  var n = {}, o = n.interval = m(a / i);
  l != null && o < l && (o = n.interval = l), u != null && o > u && (o = n.interval = u);
  var f = n.intervalPrecision = g(o), M = n.niceTickExtent = [s(Math.ceil(r[0] / o) * o, f), s(Math.floor(r[1] / o) * o, f)];
  return z(M, r), n;
}
function B(r) {
  var a = Math.pow(10, p(r)), i = r / a;
  return i ? i === 2 ? i = 3 : i === 3 ? i = 5 : i *= 2 : i = 1, s(i * a);
}
function g(r) {
  return v(r) + 2;
}
function c(r, a, i) {
  r[a] = Math.max(Math.min(r[a], i[1]), i[0]);
}
function z(r, a) {
  !isFinite(r[0]) && (r[0] = a[0]), !isFinite(r[1]) && (r[1] = a[1]), c(r, 0, a), c(r, 1, a), r[0] > r[1] && (r[0] = r[1]);
}
function F(r, a) {
  return r >= a[0] && r <= a[1];
}
var q = (
  /** @class */
  (function() {
    function r() {
      this.normalize = t, this.scale = e;
    }
    return r.prototype.updateMethods = function(a) {
      a.hasBreaks() ? (this.normalize = h(a.normalize, a), this.scale = h(a.scale, a)) : (this.normalize = t, this.scale = e);
    }, r;
  })()
);
function t(r, a) {
  return a[1] === a[0] ? 0.5 : (r - a[0]) / (a[1] - a[0]);
}
function e(r, a) {
  return r * (a[1] - a[0]) + a[0];
}
function L(r, a, i) {
  var l = Math.log(r);
  return [
    // log(negative) is NaN, so safe guard here.
    // PENDING: But even getting a -Infinity still does not make sense in extent.
    //  Just keep it as is, getting a NaN to make some previous cases works by coincidence.
    Math.log(i ? a[0] : Math.max(0, a[0])) / l,
    Math.log(i ? a[1] : Math.max(0, a[1])) / l
  ];
}
export {
  q as ScaleCalculator,
  F as contain,
  z as fixExtent,
  g as getIntervalPrecision,
  B as increaseInterval,
  w as intervalScaleNiceTicks,
  P as isIntervalOrLogScale,
  S as isValueNice,
  L as logTransform
};
