// echarts@6.0.0/lib/scale/Interval.js
import { __extends as m } from "../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { round as T, getPrecision as E } from "../util/number.js";
import { addCommas as N } from "../util/format.js";
import y from "./Scale.js";
import { contain as d, getIntervalPrecision as p, intervalScaleNiceTicks as B } from "./helper.js";
import { getScaleBreakHelper as g } from "./break.js";
var o = T, P = (
  /** @class */
  (function(k) {
    m(s, k);
    function s() {
      var t = k !== null && k.apply(this, arguments) || this;
      return t.type = "interval", t._interval = 0, t._intervalPrecision = 2, t;
    }
    return s.prototype.parse = function(t) {
      return t == null || t === "" ? NaN : Number(t);
    }, s.prototype.contain = function(t) {
      return d(t, this._extent);
    }, s.prototype.normalize = function(t) {
      return this._calculator.normalize(t, this._extent);
    }, s.prototype.scale = function(t) {
      return this._calculator.scale(t, this._extent);
    }, s.prototype.getInterval = function() {
      return this._interval;
    }, s.prototype.setInterval = function(t) {
      this._interval = t, this._niceExtent = this._extent.slice(), this._intervalPrecision = p(t);
    }, s.prototype.getTicks = function(t) {
      t = t || {};
      var e = this._interval, n = this._extent, a = this._niceExtent, r = this._intervalPrecision, l = g(), i = [];
      if (!e)
        return i;
      if (t.breakTicks === "only_break" && l)
        return l.addBreaksToTicks(i, this._brkCtx.breaks, this._extent), i;
      var u = 1e4;
      n[0] < a[0] && (t.expandToNicedExtent ? i.push({
        value: o(a[0] - e, r)
      }) : i.push({
        value: n[0]
      }));
      for (var h = function(v, _) {
        return Math.round((_ - v) / e);
      }, c = a[0]; c <= a[1]; ) {
        if (i.push({
          value: c
        }), c = o(c + e, r), this._brkCtx) {
          var f = this._brkCtx.calcNiceTickMultiple(c, h);
          f >= 0 && (c = o(c + f * e, r));
        }
        if (i.length > 0 && c === i[i.length - 1].value)
          break;
        if (i.length > u)
          return [];
      }
      var x = i.length ? i[i.length - 1].value : a[1];
      return n[1] > x && (t.expandToNicedExtent ? i.push({
        value: o(x + e, r)
      }) : i.push({
        value: n[1]
      })), l && l.pruneTicksByBreak(t.pruneByBreak, i, this._brkCtx.breaks, function(v) {
        return v.value;
      }, this._interval, this._extent), t.breakTicks !== "none" && l && l.addBreaksToTicks(i, this._brkCtx.breaks, this._extent), i;
    }, s.prototype.getMinorTicks = function(t) {
      for (var e = this.getTicks({
        expandToNicedExtent: !0
      }), n = [], a = this.getExtent(), r = 1; r < e.length; r++) {
        var l = e[r], i = e[r - 1];
        if (!(i.break || l.break)) {
          for (var u = 0, h = [], c = l.value - i.value, f = c / t, x = p(f); u < t - 1; ) {
            var v = o(i.value + (u + 1) * f, x);
            v > a[0] && v < a[1] && h.push(v), u++;
          }
          var _ = g();
          _ && _.pruneTicksByBreak("auto", h, this._getNonTransBreaks(), function(b) {
            return b;
          }, this._interval, a), n.push(h);
        }
      }
      return n;
    }, s.prototype._getNonTransBreaks = function() {
      return this._brkCtx ? this._brkCtx.breaks : [];
    }, s.prototype.getLabel = function(t, e) {
      if (t == null)
        return "";
      var n = e && e.precision;
      n == null ? n = E(t.value) || 0 : n === "auto" && (n = this._intervalPrecision);
      var a = o(t.value, n, !0);
      return N(a);
    }, s.prototype.calcNiceTicks = function(t, e, n) {
      t = t || 5;
      var a = this._extent.slice(), r = this._getExtentSpanWithBreaks();
      if (isFinite(r)) {
        r < 0 && (r = -r, a.reverse(), this._innerSetExtent(a[0], a[1]), a = this._extent.slice());
        var l = B(a, r, t, e, n);
        this._intervalPrecision = l.intervalPrecision, this._interval = l.interval, this._niceExtent = l.niceTickExtent;
      }
    }, s.prototype.calcNiceExtent = function(t) {
      var e = this._extent.slice();
      if (e[0] === e[1])
        if (e[0] !== 0) {
          var n = Math.abs(e[0]);
          t.fixMax || (e[1] += n / 2), e[0] -= n / 2;
        } else
          e[1] = 1;
      var a = e[1] - e[0];
      isFinite(a) || (e[0] = 0, e[1] = 1), this._innerSetExtent(e[0], e[1]), e = this._extent.slice(), this.calcNiceTicks(t.splitNumber, t.minInterval, t.maxInterval);
      var r = this._interval, l = this._intervalPrecision;
      t.fixMin || (e[0] = o(Math.floor(e[0] / r) * r, l)), t.fixMax || (e[1] = o(Math.ceil(e[1] / r) * r, l)), this._innerSetExtent(e[0], e[1]);
    }, s.prototype.setNiceExtent = function(t, e) {
      this._niceExtent = [t, e];
    }, s.type = "interval", s;
  })(y)
);
y.registerClass(P);
export {
  P as default
};
