// echarts@6.0.0/lib/scale/Log.js
import { __extends as b } from "../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { map as S, bind as B } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import M from "./Scale.js";
import { round as d, quantity as F, getPrecision as L } from "../util/number.js";
import _ from "./Interval.js";
import { logTransform as E, getIntervalPrecision as w } from "./helper.js";
import { getScaleBreakHelper as k } from "./break.js";
var x = d, O = Math.floor, P = Math.ceil, l = Math.pow, c = Math.log, T = (
  /** @class */
  (function(a) {
    b(r, a);
    function r() {
      var t = a !== null && a.apply(this, arguments) || this;
      return t.type = "log", t.base = 10, t._originalScale = new _(), t;
    }
    return r.prototype.getTicks = function(t) {
      t = t || {};
      var e = this._extent.slice(), i = this._originalScale.getExtent(), n = a.prototype.getTicks.call(this, t), o = this.base, g = this._originalScale._innerGetBreaks(), v = k();
      return S(n, function(u) {
        var p = u.value, s = null, f = l(o, p);
        p === e[0] && this._fixMin ? s = i[0] : p === e[1] && this._fixMax && (s = i[1]);
        var m;
        if (v) {
          var y = v.getTicksLogTransformBreak(u, o, g, h);
          m = y.vBreak, s == null && (s = y.brkRoundingCriterion);
        }
        return s != null && (f = h(f, s)), {
          value: f,
          break: m
        };
      }, this);
    }, r.prototype._getNonTransBreaks = function() {
      return this._originalScale._innerGetBreaks();
    }, r.prototype.setExtent = function(t, e) {
      this._originalScale.setExtent(t, e);
      var i = E(this.base, [t, e]);
      a.prototype.setExtent.call(this, i[0], i[1]);
    }, r.prototype.getExtent = function() {
      var t = this.base, e = a.prototype.getExtent.call(this);
      e[0] = l(t, e[0]), e[1] = l(t, e[1]);
      var i = this._originalScale.getExtent();
      return this._fixMin && (e[0] = h(e[0], i[0])), this._fixMax && (e[1] = h(e[1], i[1])), e;
    }, r.prototype.unionExtentFromData = function(t, e) {
      this._originalScale.unionExtentFromData(t, e);
      var i = E(this.base, t.getApproximateExtent(e), !0);
      this._innerUnionExtent(i);
    }, r.prototype.calcNiceTicks = function(t) {
      t = t || 10;
      var e = this._extent.slice(), i = this._getExtentSpanWithBreaks();
      if (!(!isFinite(i) || i <= 0)) {
        var n = F(i), o = t / i * n;
        for (o <= 0.5 && (n *= 10); !isNaN(n) && Math.abs(n) < 1 && Math.abs(n) > 0; )
          n *= 10;
        var g = [x(P(e[0] / n) * n), x(O(e[1] / n) * n)];
        this._interval = n, this._intervalPrecision = w(n), this._niceExtent = g;
      }
    }, r.prototype.calcNiceExtent = function(t) {
      a.prototype.calcNiceExtent.call(this, t), this._fixMin = t.fixMin, this._fixMax = t.fixMax;
    }, r.prototype.contain = function(t) {
      return t = c(t) / c(this.base), a.prototype.contain.call(this, t);
    }, r.prototype.normalize = function(t) {
      return t = c(t) / c(this.base), a.prototype.normalize.call(this, t);
    }, r.prototype.scale = function(t) {
      return t = a.prototype.scale.call(this, t), l(this.base, t);
    }, r.prototype.setBreaksFromOption = function(t) {
      var e = k();
      if (e) {
        var i = e.logarithmicParseBreaksFromOption(t, this.base, B(this.parse, this)), n = i.parsedOriginal, o = i.parsedLogged;
        this._originalScale._innerSetBreak(n), this._innerSetBreak(o);
      }
    }, r.type = "log", r;
  })(_)
);
function h(a, r) {
  return x(a, L(r));
}
M.registerClass(T);
export {
  T as default
};
