// echarts@6.0.0/lib/scale/Ordinal.js
import { __extends as d } from "../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import h from "./Scale.js";
import c from "../data/OrdinalMeta.js";
import { contain as f } from "./helper.js";
import { isArray as g, map as y, isObject as _, isString as m } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
var N = (
  /** @class */
  (function(s) {
    d(e, s);
    function e(t) {
      var r = s.call(this, t) || this;
      r.type = "ordinal";
      var n = r.getSetting("ordinalMeta");
      return n || (n = new c({})), g(n) && (n = new c({
        categories: y(n, function(o) {
          return _(o) ? o.value : o;
        })
      })), r._ordinalMeta = n, r._extent = r.getSetting("extent") || [0, n.categories.length - 1], r;
    }
    return e.prototype.parse = function(t) {
      return t == null ? NaN : m(t) ? this._ordinalMeta.getOrdinal(t) : Math.round(t);
    }, e.prototype.contain = function(t) {
      return f(t, this._extent) && t >= 0 && t < this._ordinalMeta.categories.length;
    }, e.prototype.normalize = function(t) {
      return t = this._getTickNumber(t), this._calculator.normalize(t, this._extent);
    }, e.prototype.scale = function(t) {
      return t = Math.round(this._calculator.scale(t, this._extent)), this.getRawOrdinalNumber(t);
    }, e.prototype.getTicks = function() {
      for (var t = [], r = this._extent, n = r[0]; n <= r[1]; )
        t.push({
          value: n
        }), n++;
      return t;
    }, e.prototype.getMinorTicks = function(t) {
    }, e.prototype.setSortInfo = function(t) {
      if (t == null) {
        this._ordinalNumbersByTick = this._ticksByOrdinalNumber = null;
        return;
      }
      for (var r = t.ordinalNumbers, n = this._ordinalNumbersByTick = [], o = this._ticksByOrdinalNumber = [], i = 0, u = this._ordinalMeta.categories.length, p = Math.min(u, r.length); i < p; ++i) {
        var l = r[i];
        n[i] = l, o[l] = i;
      }
      for (var a = 0; i < u; ++i) {
        for (; o[a] != null; )
          a++;
        n.push(a), o[a] = i;
      }
    }, e.prototype._getTickNumber = function(t) {
      var r = this._ticksByOrdinalNumber;
      return r && t >= 0 && t < r.length ? r[t] : t;
    }, e.prototype.getRawOrdinalNumber = function(t) {
      var r = this._ordinalNumbersByTick;
      return r && t >= 0 && t < r.length ? r[t] : t;
    }, e.prototype.getLabel = function(t) {
      if (!this.isBlank()) {
        var r = this.getRawOrdinalNumber(t.value), n = this._ordinalMeta.categories[r];
        return n == null ? "" : n + "";
      }
    }, e.prototype.count = function() {
      return this._extent[1] - this._extent[0] + 1;
    }, e.prototype.isInExtentRange = function(t) {
      return t = this._getTickNumber(t), this._extent[0] <= t && this._extent[1] >= t;
    }, e.prototype.getOrdinalMeta = function() {
      return this._ordinalMeta;
    }, e.prototype.calcNiceTicks = function() {
    }, e.prototype.calcNiceExtent = function() {
    }, e.type = "ordinal", e;
  })(h)
);
h.registerClass(N);
export {
  N as default
};
