// echarts@6.0.0/lib/scale/Scale.js
import { enableClassManagement as s } from "../util/clazz.js";
import { ScaleCalculator as o } from "./helper.js";
import { bind as a } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { getScaleBreakHelper as r } from "./break.js";
var p = (
  /** @class */
  (function() {
    function n(t) {
      this._calculator = new o(), this._setting = t || {}, this._extent = [1 / 0, -1 / 0];
      var e = r();
      e && (this._brkCtx = e.createScaleBreakContext(), this._brkCtx.update(this._extent));
    }
    return n.prototype.getSetting = function(t) {
      return this._setting[t];
    }, n.prototype._innerUnionExtent = function(t) {
      var e = this._extent;
      this._innerSetExtent(t[0] < e[0] ? t[0] : e[0], t[1] > e[1] ? t[1] : e[1]);
    }, n.prototype.unionExtentFromData = function(t, e) {
      this._innerUnionExtent(t.getApproximateExtent(e));
    }, n.prototype.getExtent = function() {
      return this._extent.slice();
    }, n.prototype.setExtent = function(t, e) {
      this._innerSetExtent(t, e);
    }, n.prototype._innerSetExtent = function(t, e) {
      var i = this._extent;
      isNaN(t) || (i[0] = t), isNaN(e) || (i[1] = e), this._brkCtx && this._brkCtx.update(i);
    }, n.prototype.setBreaksFromOption = function(t) {
      var e = r();
      e && this._innerSetBreak(e.parseAxisBreakOption(t, a(this.parse, this)));
    }, n.prototype._innerSetBreak = function(t) {
      this._brkCtx && (this._brkCtx.setBreaks(t), this._calculator.updateMethods(this._brkCtx), this._brkCtx.update(this._extent));
    }, n.prototype._innerGetBreaks = function() {
      return this._brkCtx ? this._brkCtx.breaks : [];
    }, n.prototype.hasBreaks = function() {
      return this._brkCtx ? this._brkCtx.hasBreaks() : !1;
    }, n.prototype._getExtentSpanWithBreaks = function() {
      return this._brkCtx && this._brkCtx.hasBreaks() ? this._brkCtx.getExtentSpan() : this._extent[1] - this._extent[0];
    }, n.prototype.isInExtentRange = function(t) {
      return this._extent[0] <= t && this._extent[1] >= t;
    }, n.prototype.isBlank = function() {
      return this._isBlank;
    }, n.prototype.setBlank = function(t) {
      this._isBlank = t;
    }, n;
  })()
);
s(p);
export {
  p as default
};
