// echarts@6.0.0/lib/scale/Time.js
import { __extends as ne } from "../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { parseDate as ae, nice as ie } from "../util/number.js";
import { format as ue, fullLeveledFormatter as K, getDefaultFormatPrecisionOfInterval as se, getPrimaryTimeUnit as P, leveledFormat as le, getUnitFromValue as x, primaryTimeUnits as p, ONE_DAY as d, ONE_SECOND as C, ONE_MINUTE as ee, ONE_HOUR as H, ONE_YEAR as Q, timeUnits as oe, isPrimaryTimeUnit as ce, roundTime as V, millisecondsGetterName as fe, millisecondsSetterName as ve, secondsGetterName as he, secondsSetterName as me, minutesGetterName as ge, minutesSetterName as de, hoursGetterName as _e, hoursSetterName as pe, dateGetterName as ye, dateSetterName as ke, monthGetterName as be, monthSetterName as Se, fullYearGetterName as we, fullYearSetterName as Te } from "../util/time.js";
import { getIntervalPrecision as Ee, contain as Ue } from "./helper.js";
import Me from "./Interval.js";
import De from "./Scale.js";
import { warn as Oe } from "../util/log.js";
import { each as xe, indexOf as Y, isNumber as Le, filter as Z, map as Be } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { getScaleBreakHelper as F } from "./break.js";
var Ne = function(r, n, e, t) {
  for (; e < t; ) {
    var a = e + t >>> 1;
    r[a][1] < n ? e = a + 1 : t = a;
  }
  return e;
}, Ae = (
  /** @class */
  (function(r) {
    ne(n, r);
    function n(e) {
      var t = r.call(this, e) || this;
      return t.type = "time", t;
    }
    return n.prototype.getLabel = function(e) {
      var t = this.getSetting("useUTC");
      return ue(e.value, K[se(P(this._minLevelUnit))] || K.second, t, this.getSetting("locale"));
    }, n.prototype.getFormattedLabel = function(e, t, a) {
      var v = this.getSetting("useUTC"), u = this.getSetting("locale");
      return le(e, t, a, u, v);
    }, n.prototype.getTicks = function(e) {
      e = e || {};
      var t = this._interval, a = this._extent, v = F(), u = [];
      if (!t)
        return u;
      var f = this.getSetting("useUTC");
      if (v && e.breakTicks === "only_break")
        return F().addBreaksToTicks(u, this._brkCtx.breaks, this._extent), u;
      var L = x(a[1], f);
      u.push({
        value: a[0],
        time: {
          level: 0,
          upperTimeUnit: L,
          lowerTimeUnit: L
        }
      });
      var q = Ie(this._minLevelUnit, this._approxInterval, f, a, this._getExtentSpanWithBreaks(), this._brkCtx);
      u = u.concat(q);
      var B = x(a[1], f);
      u.push({
        value: a[1],
        time: {
          level: 0,
          upperTimeUnit: B,
          lowerTimeUnit: B
        }
      });
      var y = this.getSetting("useUTC"), h = p.length - 1, k = 0;
      return xe(u, function(m) {
        h = Math.min(h, Y(p, m.time.upperTimeUnit)), k = Math.max(k, m.time.level);
      }), v && F().pruneTicksByBreak(e.pruneByBreak, u, this._brkCtx.breaks, function(m) {
        return m.value;
      }, this._approxInterval, this._extent), v && e.breakTicks !== "none" && F().addBreaksToTicks(u, this._brkCtx.breaks, this._extent, function(m) {
        for (var i = Math.max(Y(p, x(m.vmin, y)), Y(p, x(m.vmax, y))), N = 0, S = 0; S < p.length; S++)
          if (!te(p[S], m.vmin, m.vmax, y)) {
            N = S;
            break;
          }
        var M = Math.min(N, h), b = Math.max(M, i);
        return {
          level: k,
          lowerTimeUnit: p[b],
          upperTimeUnit: p[M]
        };
      }), u;
    }, n.prototype.calcNiceExtent = function(e) {
      var t = this.getExtent();
      if (t[0] === t[1] && (t[0] -= d, t[1] += d), t[1] === -1 / 0 && t[0] === 1 / 0) {
        var a = /* @__PURE__ */ new Date();
        t[1] = +new Date(a.getFullYear(), a.getMonth(), a.getDate()), t[0] = t[1] - d;
      }
      this._innerSetExtent(t[0], t[1]), this.calcNiceTicks(e.splitNumber, e.minInterval, e.maxInterval);
    }, n.prototype.calcNiceTicks = function(e, t, a) {
      e = e || 10;
      var v = this._getExtentSpanWithBreaks();
      this._approxInterval = v / e, t != null && this._approxInterval < t && (this._approxInterval = t), a != null && this._approxInterval > a && (this._approxInterval = a);
      var u = G.length, f = Math.min(Ne(G, this._approxInterval, 0, u), u - 1);
      this._interval = G[f][1], this._intervalPrecision = Ee(this._interval), this._minLevelUnit = G[Math.max(f - 1, 0)][0];
    }, n.prototype.parse = function(e) {
      return Le(e) ? e : +ae(e);
    }, n.prototype.contain = function(e) {
      return Ue(e, this._extent);
    }, n.prototype.normalize = function(e) {
      return this._calculator.normalize(e, this._extent);
    }, n.prototype.scale = function(e) {
      return this._calculator.scale(e, this._extent);
    }, n.type = "time", n;
  })(Me)
), G = [
  // Format                           interval
  ["second", C],
  ["minute", ee],
  ["hour", H],
  ["quarter-day", H * 6],
  ["half-day", H * 12],
  ["day", d * 1.2],
  ["half-week", d * 3.5],
  ["week", d * 7],
  ["month", d * 31],
  ["quarter", d * 95],
  ["half-year", Q / 2],
  ["year", Q]
  // 1Y
];
function te(r, n, e, t) {
  return V(new Date(n), r, t).getTime() === V(new Date(e), r, t).getTime();
}
function Fe(r, n) {
  return r /= d, r > 16 ? 16 : r > 7.5 ? 7 : r > 3.5 ? 4 : r > 1.5 ? 2 : 1;
}
function Ge(r) {
  var n = 30 * d;
  return r /= n, r > 6 ? 6 : r > 3 ? 3 : r > 2 ? 2 : 1;
}
function Pe(r) {
  return r /= H, r > 12 ? 12 : r > 6 ? 6 : r > 3.5 ? 4 : r > 2 ? 2 : 1;
}
function $(r, n) {
  return r /= n ? ee : C, r > 30 ? 30 : r > 20 ? 20 : r > 15 ? 15 : r > 10 ? 10 : r > 5 ? 5 : r > 2 ? 2 : 1;
}
function He(r) {
  return ie(r);
}
function Ye(r, n, e) {
  var t = Math.max(0, Y(p, n) - 1);
  return V(new Date(r), p[t], e).getTime();
}
function qe(r, n) {
  var e = /* @__PURE__ */ new Date(0);
  e[r](1);
  var t = e.getTime();
  e[r](1 + n);
  var a = e.getTime() - t;
  return function(v, u) {
    return Math.max(0, Math.round((u - v) / a));
  };
}
function Ie(r, n, e, t, a, v) {
  var u = 1e4, f = oe, L = 0;
  function q(l, o, T, E, D, _, U) {
    for (var O = qe(D, l), s = o, c = new Date(s); s < T && s <= t[1]; ) {
      if (U.push({
        value: s
      }), L++ > u) {
        process.env.NODE_ENV !== "production" && Oe("Exceed safe limit in time scale.");
        break;
      }
      if (c[D](c[E]() + l), s = c.getTime(), v) {
        var g = v.calcNiceTickMultiple(s, O);
        g > 0 && (c[D](c[E]() + g * l), s = c.getTime());
      }
    }
    U.push({
      value: s,
      notAdd: !0
    });
  }
  function B(l, o, T) {
    var E = [], D = !o.length;
    if (!te(P(l), t[0], t[1], e)) {
      D && (o = [{
        value: Ye(t[0], l, e)
      }, {
        value: t[1]
      }]);
      for (var _ = 0; _ < o.length - 1; _++) {
        var U = o[_].value, O = o[_ + 1].value;
        if (U !== O) {
          var s = void 0, c = void 0, g = void 0, J = !1;
          switch (l) {
            case "year":
              s = Math.max(1, Math.round(n / d / 365)), c = we(e), g = Te(e);
              break;
            case "half-year":
            case "quarter":
            case "month":
              s = Ge(n), c = be(e), g = Se(e);
              break;
            case "week":
            // PENDING If week is added. Ignore day.
            case "half-week":
            case "day":
              s = Fe(n), c = ye(e), g = ke(e), J = !0;
              break;
            case "half-day":
            case "quarter-day":
            case "hour":
              s = Pe(n), c = _e(e), g = pe(e);
              break;
            case "minute":
              s = $(n, !0), c = ge(e), g = de(e);
              break;
            case "second":
              s = $(n, !1), c = he(e), g = me(e);
              break;
            case "millisecond":
              s = He(n), c = fe(e), g = ve(e);
              break;
          }
          O >= t[0] && U <= t[1] && q(s, U, O, c, g, J, E), l === "year" && T.length > 1 && _ === 0 && T.unshift({
            value: T[0].value - s
          });
        }
      }
      for (var _ = 0; _ < E.length; _++)
        T.push(E[_]);
    }
  }
  for (var y = [], h = [], k = 0, m = 0, i = 0; i < f.length; ++i) {
    var N = P(f[i]);
    if (ce(f[i])) {
      B(f[i], y[y.length - 1] || [], h);
      var S = f[i + 1] ? P(f[i + 1]) : null;
      if (N !== S) {
        if (h.length) {
          m = k, h.sort(function(l, o) {
            return l.value - o.value;
          });
          for (var M = [], b = 0; b < h.length; ++b) {
            var I = h[b].value;
            (b === 0 || h[b - 1].value !== I) && (M.push(h[b]), I >= t[0] && I <= t[1] && k++);
          }
          var z = a / n;
          if (k > z * 1.5 && m > z / 1.5 || (y.push(M), k > z || r === f[i]))
            break;
        }
        h = [];
      }
    }
  }
  for (var R = Z(Be(y, function(l) {
    return Z(l, function(o) {
      return o.value >= t[0] && o.value <= t[1] && !o.notAdd;
    });
  }), function(l) {
    return l.length > 0;
  }), w = [], re = R.length - 1, i = 0; i < R.length; ++i)
    for (var W = R[i], A = 0; A < W.length; ++A) {
      var X = x(W[A].value, e);
      w.push({
        value: W[A].value,
        time: {
          level: re - i,
          upperTimeUnit: X,
          lowerTimeUnit: X
        }
      });
    }
  w.sort(function(l, o) {
    return l.value - o.value;
  });
  for (var j = [], i = 0; i < w.length; ++i)
    (i === 0 || w[i].value !== w[i - 1].value) && j.push(w[i]);
  return j;
}
De.registerClass(Ae);
export {
  Ae as default
};
