// @tiptap/extension-character-count@2.11.7/dist/index.js
import { Extension as l } from "../../../../../@tiptap_core@2.24.0_@tiptap_pm@2.24.0/node_modules/@tiptap/core/dist/index.js";
import { Plugin as h, PluginKey as f } from "../../../../../prosemirror-state@1.4.3/node_modules/prosemirror-state/dist/index.js";
const x = l.create({
  name: "characterCount",
  addOptions() {
    return {
      limit: null,
      mode: "textSize",
      textCounter: (e) => e.length,
      wordCounter: (e) => e.split(" ").filter((t) => t !== "").length
    };
  },
  addStorage() {
    return {
      characters: () => 0,
      words: () => 0
    };
  },
  onBeforeCreate() {
    this.storage.characters = (e) => {
      const t = e?.node || this.editor.state.doc;
      if ((e?.mode || this.options.mode) === "textSize") {
        const r = t.textBetween(0, t.content.size, void 0, " ");
        return this.options.textCounter(r);
      }
      return t.nodeSize;
    }, this.storage.words = (e) => {
      const t = e?.node || this.editor.state.doc, i = t.textBetween(0, t.content.size, " ", " ");
      return this.options.wordCounter(i);
    };
  },
  addProseMirrorPlugins() {
    let e = !1;
    return [
      new h({
        key: new f("characterCount"),
        appendTransaction: (t, i, r) => {
          if (e)
            return;
          const n = this.options.limit;
          if (n == null || n === 0) {
            e = !0;
            return;
          }
          const o = this.storage.characters({ node: r.doc });
          if (o > n) {
            const c = o - n, s = 0, a = c;
            console.warn(`[CharacterCount] Initial content exceeded limit of ${n} characters. Content was automatically trimmed.`);
            const d = r.tr.deleteRange(s, a);
            return e = !0, d;
          }
          e = !0;
        },
        filterTransaction: (t, i) => {
          const r = this.options.limit;
          if (!t.docChanged || r === 0 || r === null || r === void 0)
            return !0;
          const n = this.storage.characters({ node: i.doc }), o = this.storage.characters({ node: t.doc });
          if (o <= r || n > r && o > r && o <= n)
            return !0;
          if (n > r && o > r && o > n || !t.getMeta("paste"))
            return !1;
          const s = t.selection.$head.pos, a = o - r, d = s - a, u = s;
          return t.deleteRange(d, u), !(this.storage.characters({ node: t.doc }) > r);
        }
      })
    ];
  }
});
export {
  x as CharacterCount,
  x as default
};
