// echarts@6.0.0/lib/component/marker/MarkAreaView.js
import { __extends as U } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { modifyAlpha as N } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/tool/color.js";
import M from "../../data/SeriesData.js";
import { asc as z, parsePercent as b } from "../../util/number.js";
import { setStatesStylesFromModel as X, toggleHoverEmphasis as Y } from "../../util/states.js";
import { dataTransform as F, zoneFilter as Z, dataFilter as L } from "./markerHelper.js";
import $ from "./MarkerView.js";
import { map as k, isString as w, retrieve2 as S, extend as P, curry as T, filter as _, retrieve as D, mergeAll as V } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { isCoordinateSystemType as q } from "../../coord/CoordinateSystem.js";
import tt from "./MarkerModel.js";
import { makeInner as rt } from "../../util/model.js";
import { getVisualFromData as et } from "../../visual/helper.js";
import { setLabelStyle as at, getLabelStatesModels as ot } from "../../label/labelStyle.js";
import { getECData as it } from "../../util/innerStore.js";
import { parseDataValue as H } from "../../data/helper/dataValueHelper.js";
import nt from "../../visual/tokens.js";
import lt from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import O from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Polygon.js";
import { updateProps as ft } from "../../animation/basicTransition.js";
var E = rt(), st = function(o, i, e, n) {
  var m = n[0], v = n[1];
  if (!(!m || !v)) {
    var u = F(o, m), l = F(o, v), g = u.coord, f = l.coord;
    g[0] = D(g[0], -1 / 0), g[1] = D(g[1], -1 / 0), f[0] = D(f[0], 1 / 0), f[1] = D(f[1], 1 / 0);
    var a = V([{}, u, l]);
    return a.coord = [u.coord, l.coord], a.x0 = u.x, a.y0 = u.y, a.x1 = l.x, a.y1 = l.y, a;
  }
};
function A(o) {
  return !isNaN(o) && !isFinite(o);
}
function K(o, i, e, n) {
  var m = 1 - o;
  return A(i[m]) && A(e[m]);
}
function mt(o, i) {
  var e = i.coord[0], n = i.coord[1], m = {
    coord: e,
    x: i.x0,
    y: i.y0
  }, v = {
    coord: n,
    x: i.x1,
    y: i.y1
  };
  return q(o, "cartesian2d") ? e && n && (K(1, e, n) || K(0, e, n)) ? !0 : Z(o, m, v) : L(o, m) || L(o, v);
}
function W(o, i, e, n, m) {
  var v = n.coordinateSystem, u = o.getItemModel(i), l, g = b(u.get(e[0]), m.getWidth()), f = b(u.get(e[1]), m.getHeight());
  if (!isNaN(g) && !isNaN(f))
    l = [g, f];
  else {
    if (n.getMarkerPosition) {
      var a = o.getValues(["x0", "y0"], i), r = o.getValues(["x1", "y1"], i), t = v.clampData(a), p = v.clampData(r), s = [];
      e[0] === "x0" ? s[0] = t[0] > p[0] ? r[0] : a[0] : s[0] = t[0] > p[0] ? a[0] : r[0], e[1] === "y0" ? s[1] = t[1] > p[1] ? r[1] : a[1] : s[1] = t[1] > p[1] ? a[1] : r[1], l = n.getMarkerPosition(s, e, !0);
    } else {
      var y = o.get(e[0], i), c = o.get(e[1], i), I = [y, c];
      v.clampData && v.clampData(I, I), l = v.dataToPoint(I, !0);
    }
    if (q(v, "cartesian2d")) {
      var x = v.getAxis("x"), d = v.getAxis("y"), y = o.get(e[0], i), c = o.get(e[1], i);
      A(y) ? l[0] = x.toGlobalCoord(x.getExtent()[e[0] === "x0" ? 0 : 1]) : A(c) && (l[1] = d.toGlobalCoord(d.getExtent()[e[1] === "y0" ? 0 : 1]));
    }
    isNaN(g) || (l[0] = g), isNaN(f) || (l[1] = f);
  }
  return l;
}
var j = [["x0", "y0"], ["x1", "y0"], ["x1", "y1"], ["x0", "y1"]], bt = (
  /** @class */
  (function(o) {
    U(i, o);
    function i() {
      var e = o !== null && o.apply(this, arguments) || this;
      return e.type = i.type, e;
    }
    return i.prototype.updateTransform = function(e, n, m) {
      n.eachSeries(function(v) {
        var u = tt.getMarkerModelFromSeries(v, "markArea");
        if (u) {
          var l = u.getData();
          l.each(function(g) {
            var f = k(j, function(r) {
              return W(l, g, r, v, m);
            });
            l.setItemLayout(g, f);
            var a = l.getItemGraphicEl(g);
            a.setShape("points", f);
          });
        }
      }, this);
    }, i.prototype.renderSeries = function(e, n, m, v) {
      var u = e.coordinateSystem, l = e.id, g = e.getData(), f = this.markerGroupMap, a = f.get(l) || f.set(l, {
        group: new lt()
      });
      this.group.add(a.group), this.markKeep(a);
      var r = vt(u, e, n);
      n.setData(r), r.each(function(t) {
        var p = k(j, function(R) {
          return W(r, t, R, e, v);
        }), s = u.getAxis("x").scale, y = u.getAxis("y").scale, c = s.getExtent(), I = y.getExtent(), x = [s.parse(r.get("x0", t)), s.parse(r.get("x1", t))], d = [y.parse(r.get("y0", t)), y.parse(r.get("y1", t))];
        z(x), z(d);
        var B = !(c[0] > x[1] || c[1] < x[0] || I[0] > d[1] || I[1] < d[0]), J = !B;
        r.setItemLayout(t, {
          points: p,
          allClipped: J
        });
        var C = r.getItemModel(t), h = C.getModel("itemStyle").getItemStyle(), Q = C.get("z2"), G = et(g, "color");
        h.fill || (h.fill = G, w(h.fill) && (h.fill = N(h.fill, 0.4))), h.stroke || (h.stroke = G), r.setItemVisual(t, "style", h), r.setItemVisual(t, "z2", S(Q, 0));
      }), r.diff(E(a).data).add(function(t) {
        var p = r.getItemLayout(t), s = r.getItemVisual(t, "z2");
        if (!p.allClipped) {
          var y = new O({
            z2: S(s, 0),
            shape: {
              points: p.points
            }
          });
          r.setItemGraphicEl(t, y), a.group.add(y);
        }
      }).update(function(t, p) {
        var s = E(a).data.getItemGraphicEl(p), y = r.getItemLayout(t), c = r.getItemVisual(t, "z2");
        y.allClipped ? s && a.group.remove(s) : (s ? ft(s, {
          z2: S(c, 0),
          shape: {
            points: y.points
          }
        }, n, t) : s = new O({
          shape: {
            points: y.points
          }
        }), r.setItemGraphicEl(t, s), a.group.add(s));
      }).remove(function(t) {
        var p = E(a).data.getItemGraphicEl(t);
        a.group.remove(p);
      }).execute(), r.eachItemGraphicEl(function(t, p) {
        var s = r.getItemModel(p), y = r.getItemVisual(p, "style");
        t.useStyle(r.getItemVisual(p, "style")), at(t, ot(s), {
          labelFetcher: n,
          labelDataIndex: p,
          defaultText: r.getName(p) || "",
          inheritColor: w(y.fill) ? N(y.fill, 1) : nt.color.neutral99
        }), X(t, s), Y(t, null, null, s.get(["emphasis", "disabled"])), it(t).dataModel = n;
      }), E(a).data = r, a.group.silent = n.get("silent") || e.get("silent");
    }, i.type = "markArea", i;
  })($)
);
function vt(o, i, e) {
  var n, m, v = ["x0", "y0", "x1", "y1"];
  if (o) {
    var u = k(o && o.dimensions, function(f) {
      var a = i.getData(), r = a.getDimensionInfo(a.mapDimension(f)) || {};
      return P(P({}, r), {
        name: f,
        // DON'T use ordinalMeta to parse and collect ordinal.
        ordinalMeta: null
      });
    });
    m = k(v, function(f, a) {
      return {
        name: f,
        type: u[a % 2].type
      };
    }), n = new M(m, e);
  } else
    m = [{
      name: "value",
      type: "float"
    }], n = new M(m, e);
  var l = k(e.get("data"), T(st, i, o, e));
  o && (l = _(l, T(mt, o)));
  var g = o ? function(f, a, r, t) {
    var p = f.coord[Math.floor(t / 2)][t % 2];
    return H(p, m[t]);
  } : function(f, a, r, t) {
    return H(f.value, m[t]);
  };
  return n.initData(l, null, g), n.hasItemOption = !0, n;
}
export {
  bt as default,
  j as dimPermutations
};
