// echarts@6.0.0/lib/component/marker/MarkLineModel.js
import { __extends as n } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import a from "./MarkerModel.js";
var p = (
  /** @class */
  (function(r) {
    n(e, r);
    function e() {
      var t = r !== null && r.apply(this, arguments) || this;
      return t.type = e.type, t;
    }
    return e.prototype.createMarkerModelFromSeries = function(t, i, o) {
      return new e(t, i, o);
    }, e.type = "markLine", e.defaultOption = {
      // zlevel: 0,
      z: 5,
      symbol: ["circle", "arrow"],
      symbolSize: [8, 16],
      // symbolRotate: 0,
      symbolOffset: 0,
      precision: 2,
      tooltip: {
        trigger: "item"
      },
      label: {
        show: !0,
        position: "end",
        distance: 5
      },
      lineStyle: {
        type: "dashed"
      },
      emphasis: {
        label: {
          show: !0
        },
        lineStyle: {
          width: 3
        }
      },
      animationEasing: "linear"
    }, e;
  })(a)
);
export {
  p as default
};
