// echarts@6.0.0/lib/component/marker/MarkerView.js
import { __extends as k } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import c from "../../view/Component.js";
import { createHashMap as h, each as s } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import f from "./MarkerModel.js";
import { makeInner as v } from "../../util/model.js";
import { enterBlur as l, leaveBlur as M } from "../../util/states.js";
import { retrieveZInfo as g, traverseUpdateZ as d } from "../../util/graphic.js";
var u = v(), Z = (
  /** @class */
  (function(m) {
    k(e, m);
    function e() {
      var t = m !== null && m.apply(this, arguments) || this;
      return t.type = e.type, t;
    }
    return e.prototype.init = function() {
      this.markerGroupMap = h();
    }, e.prototype.render = function(t, i, p) {
      var o = this, a = this.markerGroupMap;
      a.each(function(r) {
        u(r).keep = !1;
      }), i.eachSeries(function(r) {
        var n = f.getMarkerModelFromSeries(r, o.type);
        n && o.renderSeries(r, n, i, p);
      }), a.each(function(r) {
        !u(r).keep && o.group.remove(r.group);
      }), y(i, a, this.type);
    }, e.prototype.markKeep = function(t) {
      u(t).keep = !0;
    }, e.prototype.toggleBlurSeries = function(t, i) {
      var p = this;
      s(t, function(o) {
        var a = f.getMarkerModelFromSeries(o, p.type);
        if (a) {
          var r = a.getData();
          r.eachItemGraphicEl(function(n) {
            n && (i ? l(n) : M(n));
          });
        }
      });
    }, e.type = "marker", e;
  })(c)
);
function y(m, e, t) {
  m.eachSeries(function(i) {
    var p = f.getMarkerModelFromSeries(i, t), o = e.get(i.id);
    if (p && o && o.group) {
      var a = g(p), r = a.z, n = a.zlevel;
      d(o.group, r, n);
    }
  });
}
export {
  Z as default
};
