// echarts@6.0.0/lib/component/marker/MarkPointView.js
import { __extends as R } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import A from "../../chart/helper/SymbolDraw.js";
import { parsePercent as d } from "../../util/number.js";
import G from "../../data/SeriesData.js";
import { dataTransform as T, dataFilter as F, createMarkerDimValueGetter as K } from "./markerHelper.js";
import O from "./MarkerView.js";
import L from "./MarkerModel.js";
import { isFunction as u, retrieve2 as C, map as P, extend as V, curry as N, filter as E } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { getECData as H } from "../../util/innerStore.js";
import { getVisualFromData as W } from "../../visual/helper.js";
function z(e, i, a) {
  var t = i.coordinateSystem, p = a.getWidth(), n = a.getHeight(), r = t && t.getArea && t.getArea();
  e.each(function(l) {
    var y = e.getItemModel(l), c = y.get("relativeTo") === "coordinate", b = c ? r ? r.width : 0 : p, g = c ? r ? r.height : 0 : n, v = c && r ? r.x : 0, m = c && r ? r.y : 0, o, s = d(y.get("x"), b) + v, f = d(y.get("y"), g) + m;
    if (!isNaN(s) && !isNaN(f))
      o = [s, f];
    else if (i.getMarkerPosition)
      o = i.getMarkerPosition(e.getValues(e.dimensions, l));
    else if (t) {
      var h = e.get(t.dimensions[0], l), S = e.get(t.dimensions[1], l);
      o = t.dataToPoint([h, S]);
    }
    isNaN(s) || (o[0] = s), isNaN(f) || (o[1] = f), e.setItemLayout(l, o);
  });
}
var x = (
  /** @class */
  (function(e) {
    R(i, e);
    function i() {
      var a = e !== null && e.apply(this, arguments) || this;
      return a.type = i.type, a;
    }
    return i.prototype.updateTransform = function(a, t, p) {
      t.eachSeries(function(n) {
        var r = L.getMarkerModelFromSeries(n, "markPoint");
        r && (z(r.getData(), n, p), this.markerGroupMap.get(n.id).updateLayout());
      }, this);
    }, i.prototype.renderSeries = function(a, t, p, n) {
      var r = a.coordinateSystem, l = a.id, y = a.getData(), c = this.markerGroupMap, b = c.get(l) || c.set(l, new A()), g = _(r, a, t);
      t.setData(g), z(t.getData(), a, n), g.each(function(v) {
        var m = g.getItemModel(v), o = m.getShallow("symbol"), s = m.getShallow("symbolSize"), f = m.getShallow("symbolRotate"), h = m.getShallow("symbolOffset"), S = m.getShallow("symbolKeepAspect");
        if (u(o) || u(s) || u(f) || u(h)) {
          var D = t.getRawValue(v), k = t.getDataParams(v);
          u(o) && (o = o(D, k)), u(s) && (s = s(D, k)), u(f) && (f = f(D, k)), u(h) && (h = h(D, k));
        }
        var w = m.getModel("itemStyle").getItemStyle(), I = m.get("z2"), M = W(y, "color");
        w.fill || (w.fill = M), g.setItemVisual(v, {
          z2: C(I, 0),
          symbol: o,
          symbolSize: s,
          symbolRotate: f,
          symbolOffset: h,
          symbolKeepAspect: S,
          style: w
        });
      }), b.updateData(g), this.group.add(b.group), g.eachItemGraphicEl(function(v) {
        v.traverse(function(m) {
          H(m).dataModel = t;
        });
      }), this.markKeep(b), b.group.silent = t.get("silent") || a.get("silent");
    }, i.type = "markPoint", i;
  })(O)
);
function _(e, i, a) {
  var t;
  e ? t = P(e && e.dimensions, function(l) {
    var y = i.getData().getDimensionInfo(i.getData().mapDimension(l)) || {};
    return V(V({}, y), {
      name: l,
      // DON'T use ordinalMeta to parse and collect ordinal.
      ordinalMeta: null
    });
  }) : t = [{
    name: "value",
    type: "float"
  }];
  var p = new G(t, a), n = P(a.get("data"), N(T, i));
  e && (n = E(n, N(F, e)));
  var r = K(!!e, t);
  return p.initData(n, null, r), p;
}
export {
  x as default
};
