// echarts@6.0.0/lib/component/marker/MarkAreaModel.js
import { __extends as i } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import n from "./MarkerModel.js";
var s = (
  /** @class */
  (function(r) {
    i(t, r);
    function t() {
      var e = r !== null && r.apply(this, arguments) || this;
      return e.type = t.type, e;
    }
    return t.prototype.createMarkerModelFromSeries = function(e, o, a) {
      return new t(e, o, a);
    }, t.type = "markArea", t.defaultOption = {
      // zlevel: 0,
      // PENDING
      z: 1,
      tooltip: {
        trigger: "item"
      },
      // markArea should fixed on the coordinate system
      animation: !1,
      label: {
        show: !0,
        position: "top"
      },
      itemStyle: {
        // color and borderColor default to use color from series
        // color: 'auto'
        // borderColor: 'auto'
        borderWidth: 0
      },
      emphasis: {
        label: {
          show: !0,
          position: "top"
        }
      }
    }, t;
  })(n)
);
export {
  s as default
};
