// echarts@6.0.0/lib/component/marker/MarkLineView.js
import { __extends as B } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import w from "../../data/SeriesData.js";
import { parsePercent as K } from "../../util/number.js";
import { createMarkerDimValueGetter as J, getAxisInfo as Q, numCalculate as U, dataTransform as P, dataFilter as H } from "./markerHelper.js";
import X from "../../chart/helper/LineDraw.js";
import Y from "./MarkerView.js";
import { getStackedDimension as Z } from "../../data/helper/dataStackHelper.js";
import { isCoordinateSystemType as $ } from "../../coord/CoordinateSystem.js";
import { getECData as M } from "../../util/innerStore.js";
import T from "./MarkerModel.js";
import { isArray as A, retrieve2 as V, map as h, extend as R, curry as W, filter as C, retrieve as _, clone as F, isNumber as x, logError as d, merge as j } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { makeInner as tt } from "../../util/model.js";
import { getVisualFromData as et } from "../../visual/helper.js";
var N = tt(), at = function(e, a, r, t) {
  var g = e.getData(), i;
  if (A(t))
    i = t;
  else {
    var l = t.type;
    if (l === "min" || l === "max" || l === "average" || l === "median" || t.xAxis != null || t.yAxis != null) {
      var o = void 0, f = void 0;
      if (t.yAxis != null || t.xAxis != null)
        o = a.getAxis(t.yAxis != null ? "y" : "x"), f = _(t.yAxis, t.xAxis);
      else {
        var s = Q(t, g, a, e);
        o = s.valueAxis;
        var m = Z(g, s.valueDataDim);
        f = U(g, m, l);
      }
      var c = o.dim === "x" ? 0 : 1, y = 1 - c, u = F(t), v = {
        coord: []
      };
      u.type = null, u.coord = [], u.coord[y] = -1 / 0, v.coord[y] = 1 / 0;
      var D = r.get("precision");
      D >= 0 && x(f) && (f = +f.toFixed(Math.min(D, 20))), u.coord[c] = v.coord[c] = f, i = [u, v, {
        type: l,
        valueIndex: t.valueIndex,
        // Force to use the value of calculated value.
        value: f
      }];
    } else
      process.env.NODE_ENV !== "production" && d("Invalid markLine data."), i = [];
  }
  var p = [P(e, i[0]), P(e, i[1]), R({}, i[2])];
  return p[2].type = p[2].type || null, j(p[2], p[0]), j(p[2], p[1]), p;
};
function z(e) {
  return !isNaN(e) && !isFinite(e);
}
function q(e, a, r, t) {
  var g = 1 - e, i = t.dimensions[e];
  return z(a[g]) && z(r[g]) && a[e] === r[e] && t.getAxis(i).containData(a[e]);
}
function rt(e, a) {
  if (e.type === "cartesian2d") {
    var r = a[0].coord, t = a[1].coord;
    if (r && t && (q(1, r, t, e) || q(0, r, t, e)))
      return !0;
  }
  return H(e, a[0]) && H(e, a[1]);
}
function E(e, a, r, t, g) {
  var i = t.coordinateSystem, l = e.getItemModel(a), o, f = K(l.get("x"), g.getWidth()), s = K(l.get("y"), g.getHeight());
  if (!isNaN(f) && !isNaN(s))
    o = [f, s];
  else {
    if (t.getMarkerPosition)
      o = t.getMarkerPosition(e.getValues(e.dimensions, a));
    else {
      var m = i.dimensions, c = e.get(m[0], a), y = e.get(m[1], a);
      o = i.dataToPoint([c, y]);
    }
    if ($(i, "cartesian2d")) {
      var u = i.getAxis("x"), v = i.getAxis("y"), m = i.dimensions;
      z(e.get(m[0], a)) ? o[0] = u.toGlobalCoord(u.getExtent()[r ? 0 : 1]) : z(e.get(m[1], a)) && (o[1] = v.toGlobalCoord(v.getExtent()[r ? 0 : 1]));
    }
    isNaN(f) || (o[0] = f), isNaN(s) || (o[1] = s);
  }
  e.setItemLayout(a, o);
}
var Dt = (
  /** @class */
  (function(e) {
    B(a, e);
    function a() {
      var r = e !== null && e.apply(this, arguments) || this;
      return r.type = a.type, r;
    }
    return a.prototype.updateTransform = function(r, t, g) {
      t.eachSeries(function(i) {
        var l = T.getMarkerModelFromSeries(i, "markLine");
        if (l) {
          var o = l.getData(), f = N(l).from, s = N(l).to;
          f.each(function(m) {
            E(f, m, !0, i, g), E(s, m, !1, i, g);
          }), o.each(function(m) {
            o.setItemLayout(m, [f.getItemLayout(m), s.getItemLayout(m)]);
          }), this.markerGroupMap.get(i.id).updateLayout();
        }
      }, this);
    }, a.prototype.renderSeries = function(r, t, g, i) {
      var l = r.coordinateSystem, o = r.id, f = r.getData(), s = this.markerGroupMap, m = s.get(o) || s.set(o, new X());
      this.group.add(m.group);
      var c = nt(l, r, t), y = c.from, u = c.to, v = c.line;
      N(t).from = y, N(t).to = u, t.setData(v);
      var D = t.get("symbol"), p = t.get("symbolSize"), L = t.get("symbolRotate"), S = t.get("symbolOffset");
      A(D) || (D = [D, D]), A(p) || (p = [p, p]), A(L) || (L = [L, L]), A(S) || (S = [S, S]), c.from.each(function(n) {
        G(y, n, !0), G(u, n, !1);
      }), v.each(function(n) {
        var I = v.getItemModel(n), b = I.getModel("lineStyle").getLineStyle();
        v.setItemLayout(n, [y.getItemLayout(n), u.getItemLayout(n)]);
        var k = I.get("z2");
        b.stroke == null && (b.stroke = y.getItemVisual(n, "style").fill), v.setItemVisual(n, {
          z2: V(k, 0),
          fromSymbolKeepAspect: y.getItemVisual(n, "symbolKeepAspect"),
          fromSymbolOffset: y.getItemVisual(n, "symbolOffset"),
          fromSymbolRotate: y.getItemVisual(n, "symbolRotate"),
          fromSymbolSize: y.getItemVisual(n, "symbolSize"),
          fromSymbol: y.getItemVisual(n, "symbol"),
          toSymbolKeepAspect: u.getItemVisual(n, "symbolKeepAspect"),
          toSymbolOffset: u.getItemVisual(n, "symbolOffset"),
          toSymbolRotate: u.getItemVisual(n, "symbolRotate"),
          toSymbolSize: u.getItemVisual(n, "symbolSize"),
          toSymbol: u.getItemVisual(n, "symbol"),
          style: b
        });
      }), m.updateData(v), c.line.eachItemGraphicEl(function(n) {
        M(n).dataModel = t, n.traverse(function(I) {
          M(I).dataModel = t;
        });
      });
      function G(n, I, b) {
        var k = n.getItemModel(I);
        E(n, I, b, r, i);
        var O = k.getModel("itemStyle").getItemStyle();
        O.fill == null && (O.fill = et(f, "color")), n.setItemVisual(I, {
          symbolKeepAspect: k.get("symbolKeepAspect"),
          // `0` should be considered as a valid value, so use `retrieve2` instead of `||`
          symbolOffset: V(k.get("symbolOffset", !0), S[b ? 0 : 1]),
          symbolRotate: V(k.get("symbolRotate", !0), L[b ? 0 : 1]),
          // TODO: when 2d array is supported, it should ignore parent
          symbolSize: V(k.get("symbolSize"), p[b ? 0 : 1]),
          symbol: V(k.get("symbol", !0), D[b ? 0 : 1]),
          style: O
        });
      }
      this.markKeep(m), m.group.silent = t.get("silent") || r.get("silent");
    }, a.type = "markLine", a;
  })(Y)
);
function nt(e, a, r) {
  var t;
  e ? t = h(e && e.dimensions, function(s) {
    var m = a.getData().getDimensionInfo(a.getData().mapDimension(s)) || {};
    return R(R({}, m), {
      name: s,
      // DON'T use ordinalMeta to parse and collect ordinal.
      ordinalMeta: null
    });
  }) : t = [{
    name: "value",
    type: "float"
  }];
  var g = new w(t, r), i = new w(t, r), l = new w([], r), o = h(r.get("data"), W(at, a, e, r));
  e && (o = C(o, W(rt, e)));
  var f = J(!!e, t);
  return g.initData(h(o, function(s) {
    return s[0];
  }), null, f), i.initData(h(o, function(s) {
    return s[1];
  }), null, f), l.initData(h(o, function(s) {
    return s[2];
  })), l.hasItemOption = !0, {
    from: g,
    to: i,
    line: l
  };
}
export {
  Dt as default
};
