// echarts@6.0.0/lib/component/marker/markerHelper.js
import { getPrecision as y } from "../../util/number.js";
import { isDimensionStacked as F } from "../../data/helper/dataStackHelper.js";
import { isArray as p, clone as k, curry as A, indexOf as N } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { parseDataValue as b } from "../../data/helper/dataValueHelper.js";
function g(r) {
  return !(isNaN(parseFloat(r.x)) && isNaN(parseFloat(r.y)));
}
function O(r) {
  return !isNaN(parseFloat(r.x)) && !isNaN(parseFloat(r.y));
}
function D(r, a, n, o, e, i, s) {
  var l = [], c = F(
    a,
    e
    /* , otherDataDim */
  ), t = c ? a.getCalculationInfo("stackResultDimension") : e, x = m(a, t, r), v = a.hostModel, u = v.indicesOfNearest(n, t, x)[0];
  l[i] = a.get(o, u), l[s] = a.get(t, u);
  var h = a.get(e, u), f = y(a.get(e, u));
  return f = Math.min(f, 20), f >= 0 && (l[s] = +l[s].toFixed(f)), [l, h];
}
var d = {
  min: A(D, "min"),
  max: A(D, "max"),
  average: A(D, "average"),
  median: A(D, "median")
};
function Y(r, a) {
  if (a) {
    var n = r.getData(), o = r.coordinateSystem, e = o && o.dimensions;
    if (!O(a) && !p(a.coord) && p(e)) {
      var i = T(a, n, o, r);
      if (a = k(a), a.type && d[a.type] && i.baseAxis && i.valueAxis) {
        var s = N(e, i.baseAxis.dim), l = N(e, i.valueAxis.dim), c = d[a.type](n, i.valueAxis.dim, i.baseDataDim, i.valueDataDim, s, l);
        a.coord = c[0], a.value = c[1];
      } else
        a.coord = [a.xAxis != null ? a.xAxis : a.radiusAxis, a.yAxis != null ? a.yAxis : a.angleAxis];
    }
    if (a.coord == null || !p(e)) {
      a.coord = [];
      var t = r.getBaseAxis();
      if (t && a.type && d[a.type]) {
        var x = o.getOtherAxis(t);
        x && (a.value = m(n, n.mapDimension(x.dim), a.type));
      }
    } else
      for (var v = a.coord, u = 0; u < 2; u++)
        d[v[u]] && (v[u] = m(n, n.mapDimension(e[u]), v[u]));
    return a;
  }
}
function T(r, a, n, o) {
  var e = {};
  return r.valueIndex != null || r.valueDim != null ? (e.valueDataDim = r.valueIndex != null ? a.getDimension(r.valueIndex) : r.valueDim, e.valueAxis = n.getAxis(V(o, e.valueDataDim)), e.baseAxis = n.getOtherAxis(e.valueAxis), e.baseDataDim = a.mapDimension(e.baseAxis.dim)) : (e.baseAxis = o.getBaseAxis(), e.valueAxis = n.getOtherAxis(e.baseAxis), e.baseDataDim = a.mapDimension(e.baseAxis.dim), e.valueDataDim = a.mapDimension(e.valueAxis.dim)), e;
}
function V(r, a) {
  var n = r.getData().getDimensionInfo(a);
  return n && n.coordDim;
}
function Z(r, a) {
  return r && r.containData && a.coord && !g(a) ? r.containData(a.coord) : !0;
}
function w(r, a, n) {
  return r && r.containZone && a.coord && n.coord && !g(a) && !g(n) ? r.containZone(a.coord, n.coord) : !0;
}
function z(r, a) {
  return r ? function(n, o, e, i) {
    var s = i < 2 ? n.coord && n.coord[i] : n.value;
    return b(s, a[i]);
  } : function(n, o, e, i) {
    return b(n.value, a[i]);
  };
}
function m(r, a, n) {
  if (n === "average") {
    var o = 0, e = 0;
    return r.each(a, function(i, s) {
      isNaN(i) || (o += i, e++);
    }), o / e;
  } else return n === "median" ? r.getMedian(a) : r.getDataExtent(a)[n === "max" ? 1 : 0];
}
export {
  z as createMarkerDimValueGetter,
  Z as dataFilter,
  Y as dataTransform,
  T as getAxisInfo,
  m as numCalculate,
  w as zoneFilter
};
