// echarts@6.0.0/lib/component/marker/MarkerModel.js
import { __extends as y } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { each as d, extend as g, mixin as v } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import k from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/env.js";
import { DataFormatMixin as c } from "../../model/mixin/dataFormat.js";
import _ from "../../model/Component.js";
import { makeInner as S, defaultEmphasis as T } from "../../util/model.js";
import { createTooltipMarkup as h } from "../tooltip/tooltipMarkup.js";
function u(m) {
  T(m, "label", ["show"]);
}
var l = S(), D = (
  /** @class */
  (function(m) {
    y(t, m);
    function t() {
      var e = m !== null && m.apply(this, arguments) || this;
      return e.type = t.type, e.createdBySelf = !1, e.preventAutoZ = !0, e;
    }
    return t.prototype.init = function(e, r, a) {
      if (process.env.NODE_ENV !== "production" && this.type === "marker")
        throw new Error("Marker component is abstract component. Use markLine, markPoint, markArea instead.");
      this.mergeDefaultAndTheme(e, a), this._mergeOption(e, a, !1, !0);
    }, t.prototype.isAnimationEnabled = function() {
      if (k.node)
        return !1;
      var e = this.__hostSeries;
      return this.getShallow("animation") && e && e.isAnimationEnabled();
    }, t.prototype.mergeOption = function(e, r) {
      this._mergeOption(e, r, !1, !1);
    }, t.prototype._mergeOption = function(e, r, a, i) {
      var o = this.mainType;
      a || r.eachSeries(function(n) {
        var s = n.get(this.mainType, !0), p = l(n)[o];
        if (!s || !s.data) {
          l(n)[o] = null;
          return;
        }
        p ? p._mergeOption(s, r, !0) : (i && u(s), d(s.data, function(f) {
          f instanceof Array ? (u(f[0]), u(f[1])) : u(f);
        }), p = this.createMarkerModelFromSeries(s, this, r), g(p, {
          mainType: this.mainType,
          // Use the same series index and name
          seriesIndex: n.seriesIndex,
          name: n.name,
          createdBySelf: !0
        }), p.__hostSeries = n), l(n)[o] = p;
      }, this);
    }, t.prototype.formatTooltip = function(e, r, a) {
      var i = this.getData(), o = this.getRawValue(e), n = i.getName(e);
      return h("section", {
        header: this.name,
        blocks: [h("nameValue", {
          name: n,
          value: o,
          noName: !n,
          noValue: o == null
        })]
      });
    }, t.prototype.getData = function() {
      return this._data;
    }, t.prototype.setData = function(e) {
      this._data = e;
    }, t.prototype.getDataParams = function(e, r) {
      var a = c.prototype.getDataParams.call(this, e, r), i = this.__hostSeries;
      return i && (a.seriesId = i.id, a.seriesName = i.name, a.seriesType = i.subType), a;
    }, t.getMarkerModelFromSeries = function(e, r) {
      return l(e)[r];
    }, t.type = "marker", t.dependencies = ["series", "grid", "polar", "geo"], t;
  })(_)
);
v(D, c.prototype);
export {
  D as default
};
