// echarts@6.0.0/lib/visual/aria.js
import { clone as K, merge as A, createHashMap as Q, isFunction as W, extend as F, defaults as X, filter as Y, indexOf as $, isString as G, each as L } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { makeInner as _ } from "../util/model.js";
import { getDecalFromPalette as O } from "../model/mixin/palette.js";
var M = {
  label: {
    enabled: !0
  },
  decal: {
    show: !1
  }
}, R = _(), ee = {};
function ne(n, j) {
  var h = n.getModel("aria");
  if (!h.get("enabled"))
    return;
  var w = K(M);
  A(w.label, n.getLocaleModel().get("aria"), !1), A(h.option, w, !1), P(), E();
  function P() {
    var a = h.getModel("decal"), u = a.get("show");
    if (u) {
      var e = Q();
      n.eachSeries(function(t) {
        if (!t.isColorBySeries()) {
          var r = e.get(t.type);
          r || (r = {}, e.set(t.type, r)), R(t).scope = r;
        }
      }), n.eachRawSeries(function(t) {
        if (n.isSeriesFiltered(t))
          return;
        if (W(t.enableAriaDecal)) {
          t.enableAriaDecal();
          return;
        }
        var r = t.getData();
        if (t.isColorBySeries()) {
          var x = O(t.ecModel, t.name, ee, n.getSeriesCount()), C = r.getVisual("decal");
          r.setVisual("decal", S(C, x));
        } else {
          var m = t.getRawData(), d = {}, g = R(t).scope;
          r.each(function(i) {
            var l = r.getRawIndex(i);
            d[l] = i;
          });
          var D = m.count();
          m.each(function(i) {
            var l = d[i], c = m.getName(i) || i + "", v = O(t.ecModel, c, g, D), b = r.getItemVisual(l, "decal");
            r.setItemVisual(l, "decal", S(b, v));
          });
        }
        function S(i, l) {
          var c = i ? F(F({}, l), i) : l;
          return c.dirty = !0, c;
        }
      });
    }
  }
  function E() {
    var a = j.getZr().dom;
    if (a) {
      var u = n.getLocaleModel().get("aria"), e = h.getModel("label");
      if (e.option = X(e.option, u), !!e.get("enabled")) {
        if (a.setAttribute("role", "img"), e.get("description")) {
          a.setAttribute("aria-label", e.get("description"));
          return;
        }
        var t = n.getSeriesCount(), r = e.get(["data", "maxCount"]) || 10, m = e.get(["series", "maxCount"]) || 10, d = Math.min(t, m), g;
        if (!(t < 1)) {
          var D = k();
          if (D) {
            var x = e.get(["general", "withTitle"]);
            g = s(x, {
              title: D
            });
          } else
            g = e.get(["general", "withoutTitle"]);
          var C = [], S = t > 1 ? e.get(["series", "multiple", "prefix"]) : e.get(["series", "single", "prefix"]);
          g += s(S, {
            seriesCount: t
          }), n.eachSeries(function(v, b) {
            if (b < d) {
              var o = void 0, H = v.get("name"), y = H ? "withName" : "withoutName";
              o = t > 1 ? e.get(["series", "multiple", y]) : e.get(["series", "single", y]), o = s(o, {
                seriesId: v.seriesIndex,
                seriesName: v.get("name"),
                seriesType: B(v.subType)
              });
              var p = v.getData();
              if (p.count() > r) {
                var U = e.get(["data", "partialData"]);
                o += s(U, {
                  displayCnt: r
                });
              } else
                o += e.get(["data", "allData"]);
              for (var N = e.get(["data", "separator", "middle"]), Z = e.get(["data", "separator", "end"]), T = e.get(["data", "excludeDimensionId"]), V = [], f = 0; f < p.count(); f++)
                if (f < r) {
                  var I = p.getName(f), q = T ? Y(p.getValues(f), function(te, J) {
                    return $(T, J) === -1;
                  }) : p.getValues(f), z = e.get(["data", I ? "withName" : "withoutName"]);
                  V.push(s(z, {
                    name: I,
                    value: q.join(N)
                  }));
                }
              o += V.join(N) + Z, C.push(o);
            }
          });
          var i = e.getModel(["series", "multiple", "separator"]), l = i.get("middle"), c = i.get("end");
          g += C.join(l) + c, a.setAttribute("aria-label", g);
        }
      }
    }
  }
  function s(a, u) {
    if (!G(a))
      return a;
    var e = a;
    return L(u, function(t, r) {
      e = e.replace(new RegExp("\\{\\s*" + r + "\\s*\\}", "g"), t);
    }), e;
  }
  function k() {
    var a = n.get("title");
    return a && a.length && (a = a[0]), a && a.text;
  }
  function B(a) {
    var u = n.getLocaleModel().get(["series", "typeNames"]);
    return u[a] || u.chart;
  }
}
export {
  ne as default
};
