// echarts@6.0.0/lib/visual/helper.js
function l(e, s, t) {
  switch (t) {
    case "color":
      var o = e.getItemVisual(s, "style");
      return o[e.getVisual("drawType")];
    case "opacity":
      return e.getItemVisual(s, "style").opacity;
    case "symbol":
    case "symbolSize":
    case "liftZ":
      return e.getItemVisual(s, t);
    default:
      process.env.NODE_ENV !== "production" && console.warn("Unknown visual type " + t);
  }
}
function r(e, s) {
  switch (s) {
    case "color":
      var t = e.getVisual("style");
      return t[e.getVisual("drawType")];
    case "opacity":
      return e.getVisual("style").opacity;
    case "symbol":
    case "symbolSize":
    case "liftZ":
      return e.getVisual(s);
    default:
      process.env.NODE_ENV !== "production" && console.warn("Unknown visual type " + s);
  }
}
function c(e, s, t, o) {
  switch (t) {
    case "color":
      var i = e.ensureUniqueItemVisual(s, "style");
      i[e.getVisual("drawType")] = o, e.setItemVisual(s, "colorFromPalette", !1);
      break;
    case "opacity":
      e.ensureUniqueItemVisual(s, "style").opacity = o;
      break;
    case "symbol":
    case "symbolSize":
    case "liftZ":
      e.setItemVisual(s, t, o);
      break;
    default:
      process.env.NODE_ENV !== "production" && console.warn("Unknown visual type " + t);
  }
}
export {
  l as getItemVisualFromData,
  r as getVisualFromData,
  c as setItemVisualFromData
};
