// echarts@6.0.0/lib/visual/style.js
import { isFunction as S, extend as P, createHashMap as F } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import h from "../model/mixin/makeStyleMapper.js";
import { ITEM_STYLE_KEY_MAP as T } from "../model/mixin/itemStyle.js";
import { LINE_STYLE_KEY_MAP as A } from "../model/mixin/lineStyle.js";
import D from "../model/Model.js";
import { makeInner as E } from "../util/model.js";
var k = E(), w = {
  itemStyle: h(T, !0),
  lineStyle: h(A, !0)
}, K = {
  lineStyle: "stroke",
  itemStyle: "fill"
};
function I(e, r) {
  var t = e.visualStyleMapper || w[r];
  return t || (console.warn("Unknown style type '" + r + "'."), w.itemStyle);
}
function V(e, r) {
  var t = e.visualDrawType || K[r];
  return t || (console.warn("Unknown style type '" + r + "'."), "fill");
}
var L = {
  createOnAllSeries: !0,
  performRawSeries: !0,
  reset: function(e, r) {
    var t = e.getData(), l = e.visualStyleAccessPath || "itemStyle", u = e.getModel(l), o = I(e, l), a = o(u), c = u.getShallow("decal");
    c && (t.setVisual("decal", c), c.dirty = !0);
    var i = V(e, l), n = a[i], v = S(n) ? n : null, s = a.fill === "auto" || a.stroke === "auto";
    if (!a[i] || v || s) {
      var y = e.getColorFromPalette(
        // TODO series count changed.
        e.name,
        null,
        r.getSeriesCount()
      );
      a[i] || (a[i] = y, t.setVisual("colorFromPalette", !0)), a.fill = a.fill === "auto" || S(a.fill) ? y : a.fill, a.stroke = a.stroke === "auto" || S(a.stroke) ? y : a.stroke;
    }
    if (t.setVisual("style", a), t.setVisual("drawType", i), !r.isSeriesFiltered(e) && v)
      return t.setVisual("colorFromPalette", !1), {
        dataEach: function(p, m) {
          var C = e.getDataParams(m), g = P({}, a);
          g[i] = v(C), p.setItemVisual(m, "style", g);
        }
      };
  }
}, f = new D(), b = {
  createOnAllSeries: !0,
  performRawSeries: !0,
  reset: function(e, r) {
    if (!(e.ignoreStyleOnData || r.isSeriesFiltered(e))) {
      var t = e.getData(), l = e.visualStyleAccessPath || "itemStyle", u = I(e, l), o = t.getVisual("drawType");
      return {
        dataEach: t.hasItemOption ? function(a, c) {
          var i = a.getRawDataItem(c);
          if (i && i[l]) {
            f.option = i[l];
            var n = u(f), v = a.ensureUniqueItemVisual(c, "style");
            P(v, n), f.option.decal && (a.setItemVisual(c, "decal", f.option.decal), f.option.decal.dirty = !0), o in n && a.setItemVisual(c, "colorFromPalette", !1);
          }
        } : null
      };
    }
  }
}, q = {
  performRawSeries: !0,
  overallReset: function(e) {
    var r = F();
    e.eachSeries(function(t) {
      var l = t.getColorBy();
      if (!t.isColorBySeries()) {
        var u = t.type + "-" + l, o = r.get(u);
        o || (o = {}, r.set(u, o)), k(t).scope = o;
      }
    }), e.eachSeries(function(t) {
      if (!(t.isColorBySeries() || e.isSeriesFiltered(t))) {
        var l = t.getRawData(), u = {}, o = t.getData(), a = k(t).scope, c = t.visualStyleAccessPath || "itemStyle", i = V(t, c);
        o.each(function(n) {
          var v = o.getRawIndex(n);
          u[v] = n;
        }), l.each(function(n) {
          var v = u[n], s = o.getItemVisual(v, "colorFromPalette");
          if (s) {
            var y = o.ensureUniqueItemVisual(v, "style"), p = l.getName(n) || n + "", m = l.count();
            y[i] = t.getColorFromPalette(p, a, m);
          }
        });
      }
    });
  }
};
export {
  q as dataColorPaletteTask,
  b as dataStyleTask,
  L as seriesStyleTask
};
