// echarts@6.0.0/lib/visual/symbol.js
import { extend as p, keys as I, isFunction as O } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
var v = ["symbol", "symbolSize", "symbolRotate", "symbolOffset"], S = v.concat(["symbolKeepAspect"]), R = {
  createOnAllSeries: !0,
  // For legend.
  performRawSeries: !0,
  reset: function(a, c) {
    var o = a.getData();
    if (a.legendIcon && o.setVisual("legendIcon", a.legendIcon), !a.hasSymbolVisual)
      return;
    for (var t = {}, r = {}, l = !1, u = 0; u < v.length; u++) {
      var e = v[u], n = a.get(e);
      O(n) ? (l = !0, r[e] = n) : t[e] = n;
    }
    if (t.symbol = t.symbol || a.defaultSymbol, o.setVisual(p({
      legendIcon: a.legendIcon || t.symbol,
      symbolKeepAspect: a.get("symbolKeepAspect")
    }, t)), c.isSeriesFiltered(a))
      return;
    var s = I(r);
    function g(i, f) {
      for (var y = a.getRawValue(f), h = a.getDataParams(f), m = 0; m < s.length; m++) {
        var b = s[m];
        i.setItemVisual(f, b, r[b](y, h));
      }
    }
    return {
      dataEach: l ? g : null
    };
  }
}, w = {
  createOnAllSeries: !0,
  // For legend.
  performRawSeries: !0,
  reset: function(a, c) {
    if (!a.hasSymbolVisual || c.isSeriesFiltered(a))
      return;
    var o = a.getData();
    function t(r, l) {
      for (var u = r.getItemModel(l), e = 0; e < S.length; e++) {
        var n = S[e], s = u.getShallow(n, !0);
        s != null && r.setItemVisual(l, n, s);
      }
    }
    return {
      dataEach: o.hasItemOption ? t : null
    };
  }
};
export {
  w as dataSymbolTask,
  R as seriesSymbolTask
};
