// echarts@6.0.0/lib/visual/tokens.js
import { extend as l } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { modifyHSL as t } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/tool/color.js";
var n = {
  color: {},
  darkColor: {},
  size: {}
}, a = n.color = {
  theme: ["#5070dd", "#b6d634", "#505372", "#ff994d", "#0ca8df", "#ffd10a", "#fb628b", "#785db0", "#3fbe95"],
  neutral00: "#fff",
  neutral05: "#f4f7fd",
  neutral10: "#e8ebf0",
  neutral15: "#dbdee4",
  neutral20: "#cfd2d7",
  neutral25: "#c3c5cb",
  neutral30: "#b7b9be",
  neutral35: "#aaacb2",
  neutral40: "#9ea0a5",
  neutral45: "#929399",
  neutral50: "#86878c",
  neutral55: "#797b7f",
  neutral60: "#6d6e73",
  neutral65: "#616266",
  neutral70: "#54555a",
  neutral75: "#48494d",
  neutral80: "#3c3c41",
  neutral85: "#303034",
  neutral90: "#232328",
  neutral95: "#17171b",
  neutral99: "#000",
  accent05: "#eff1f9",
  accent10: "#e0e4f2",
  accent15: "#d0d6ec",
  accent20: "#c0c9e6",
  accent25: "#b1bbdf",
  accent30: "#a1aed9",
  accent35: "#91a0d3",
  accent40: "#8292cc",
  accent45: "#7285c6",
  accent50: "#6578ba",
  accent55: "#5c6da9",
  accent60: "#536298",
  accent65: "#4a5787",
  accent70: "#404c76",
  accent75: "#374165",
  accent80: "#2e3654",
  accent85: "#252b43",
  accent90: "#1b2032",
  accent95: "#121521",
  transparent: "rgba(0,0,0,0)",
  highlight: "rgba(255,231,130,0.8)"
};
l(a, {
  primary: a.neutral80,
  secondary: a.neutral70,
  tertiary: a.neutral60,
  quaternary: a.neutral50,
  disabled: a.neutral20,
  border: a.neutral30,
  borderTint: a.neutral20,
  borderShade: a.neutral40,
  background: a.neutral05,
  backgroundTint: "rgba(234,237,245,0.5)",
  backgroundTransparent: "rgba(255,255,255,0)",
  backgroundShade: a.neutral10,
  shadow: "rgba(0,0,0,0.2)",
  shadowTint: "rgba(129,130,136,0.2)",
  axisLine: a.neutral70,
  axisLineTint: a.neutral40,
  axisTick: a.neutral70,
  axisTickMinor: a.neutral60,
  axisLabel: a.neutral70,
  axisSplitLine: a.neutral15,
  axisMinorSplitLine: a.neutral05
});
for (var r in a)
  if (a.hasOwnProperty(r)) {
    var c = a[r];
    r === "theme" ? n.darkColor.theme = a.theme.slice() : r === "highlight" ? n.darkColor.highlight = "rgba(255,231,130,0.4)" : r.indexOf("accent") === 0 ? n.darkColor[r] = t(c, null, function(e) {
      return e * 0.5;
    }, function(e) {
      return Math.min(1, 1.3 - e);
    }) : n.darkColor[r] = t(c, null, function(e) {
      return e * 0.9;
    }, function(e) {
      return 1 - Math.pow(e, 1.5);
    });
  }
n.size = {
  xxs: 2,
  xs: 5,
  s: 10,
  m: 15,
  l: 20,
  xl: 30,
  xxl: 40,
  xxxl: 50
};
export {
  n as default
};
