// echarts@6.0.0/lib/visual/visualSolution.js
import { each as I, clone as T } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import y from "./VisualMapping.js";
import { getItemVisualFromData as F, setItemVisualFromData as P } from "./helper.js";
var D = I;
function d(e) {
  if (e) {
    for (var a in e)
      if (e.hasOwnProperty(a))
        return !0;
  }
}
function S(e, a, i) {
  var o = {};
  return D(a, function(t) {
    var p = o[t] = n();
    D(e[t], function(u, s) {
      if (y.isValidType(s)) {
        var l = {
          type: s,
          visual: u
        };
        i && i(l, t), p[s] = new y(l), s === "opacity" && (l = T(l), l.type = "colorAlpha", p.__hidden.__alphaForOpacity = new y(l));
      }
    });
  }), o;
  function n() {
    var t = function() {
    };
    t.prototype.__hidden = t.prototype;
    var p = new t();
    return p;
  }
}
function $(e, a, i) {
  var o;
  I(i, function(n) {
    a.hasOwnProperty(n) && d(a[n]) && (o = !0);
  }), o && I(i, function(n) {
    a.hasOwnProperty(n) && d(a[n]) ? e[n] = T(a[n]) : delete e[n];
  });
}
function q(e, a, i, o, n, t) {
  var p = {};
  I(e, function(r) {
    var f = y.prepareVisualTypes(a[r]);
    p[r] = f;
  });
  var u;
  function s(r) {
    return F(i, u, r);
  }
  function l(r, f) {
    P(i, u, r, f);
  }
  i.each(_);
  function _(r, f) {
    u = r;
    var v = i.getRawDataItem(u);
    if (!(v && v.visualMap === !1))
      for (var g = o.call(n, r), h = a[g], m = p[g], c = 0, V = m.length; c < V; c++) {
        var M = m[c];
        h[M] && h[M].applyVisual(r, s, l);
      }
  }
}
function z(e, a, i, o) {
  var n = {};
  return I(e, function(t) {
    var p = y.prepareVisualTypes(a[t]);
    n[t] = p;
  }), {
    progress: function(p, u) {
      var s;
      o != null && (s = u.getDimensionIndex(o));
      function l(w) {
        return F(u, r, w);
      }
      function _(w, A) {
        P(u, r, w, A);
      }
      for (var r, f = u.getStore(); (r = p.next()) != null; ) {
        var v = u.getRawDataItem(r);
        if (!(v && v.visualMap === !1))
          for (var g = o != null ? f.get(s, r) : r, h = i(g), m = a[h], c = n[h], V = 0, M = c.length; V < M; V++) {
            var x = c[V];
            m[x] && m[x].applyVisual(g, l, _);
          }
      }
    }
  };
}
export {
  q as applyVisual,
  S as createVisualMappings,
  z as incrementalApplyVisual,
  $ as replaceVisualOption
};
