// echarts@6.0.0/lib/visual/VisualMapping.js
import { clone as F, noop as O, assert as R, isObject as h, each as S, isArray as I, map as k, bind as D, keys as G, isString as U } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { parse as Z, stringify as M, fastLerp as x, modifyAlpha as $, modifyHSL as b } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/tool/color.js";
import { linearMap as d } from "../util/number.js";
import { warn as j } from "../util/log.js";
var V = S, B = h, v = -1, A = (
  /** @class */
  (function() {
    function a(r) {
      var i = r.mappingMethod, n = r.type, t = this.option = F(r);
      this.type = n, this.mappingMethod = i, this._normalizeData = q[i];
      var e = a.visualHandlers[n];
      this.applyVisual = e.applyVisual, this.getColorMapper = e.getColorMapper, this._normalizedToVisual = e._normalizedToVisual[i], i === "piecewise" ? (w(t), X(t)) : i === "category" ? t.categories ? Y(t) : w(t, !0) : (R(i !== "linear" || t.dataExtent), w(t));
    }
    return a.prototype.mapValueToVisual = function(r) {
      var i = this._normalizeData(r);
      return this._normalizedToVisual(i, r);
    }, a.prototype.getNormalizer = function() {
      return D(this._normalizeData, this);
    }, a.listVisualTypes = function() {
      return G(a.visualHandlers);
    }, a.isValidType = function(r) {
      return a.visualHandlers.hasOwnProperty(r);
    }, a.eachVisual = function(r, i, n) {
      h(r) ? S(r, i, n) : i.call(n, r);
    }, a.mapVisual = function(r, i, n) {
      var t, e = I(r) ? [] : h(r) ? {} : (t = !0, null);
      return a.eachVisual(r, function(o, u) {
        var l = i.call(n, o, u);
        t ? e = l : e[u] = l;
      }), e;
    }, a.retrieveVisuals = function(r) {
      var i = {}, n;
      return r && V(a.visualHandlers, function(t, e) {
        r.hasOwnProperty(e) && (i[e] = r[e], n = !0);
      }), n ? i : null;
    }, a.prepareVisualTypes = function(r) {
      if (I(r))
        r = r.slice();
      else if (B(r)) {
        var i = [];
        V(r, function(n, t) {
          i.push(t);
        }), r = i;
      } else
        return [];
      return r.sort(function(n, t) {
        return t === "color" && n !== "color" && n.indexOf("color") === 0 ? 1 : -1;
      }), r;
    }, a.dependsOn = function(r, i) {
      return i === "color" ? !!(r && r.indexOf(i) === 0) : r === i;
    }, a.findPieceIndex = function(r, i, n) {
      for (var t, e = 1 / 0, o = 0, u = i.length; o < u; o++) {
        var l = i[o].value;
        if (l != null) {
          if (l === r || U(l) && l === r + "")
            return o;
          n && z(l, o);
        }
      }
      for (var o = 0, u = i.length; o < u; o++) {
        var T = i[o], s = T.interval, g = T.close;
        if (s) {
          if (s[0] === -1 / 0) {
            if (y(g[1], r, s[1]))
              return o;
          } else if (s[1] === 1 / 0) {
            if (y(g[0], s[0], r))
              return o;
          } else if (y(g[0], s[0], r) && y(g[1], r, s[1]))
            return o;
          n && z(s[0], o), n && z(s[1], o);
        }
      }
      if (n)
        return r === 1 / 0 ? i.length - 1 : r === -1 / 0 ? 0 : t;
      function z(C, E) {
        var H = Math.abs(C - r);
        H < e && (e = H, t = E);
      }
    }, a.visualHandlers = {
      color: {
        applyVisual: f("color"),
        getColorMapper: function() {
          var r = this.option;
          return D(r.mappingMethod === "category" ? function(i, n) {
            return !n && (i = this._normalizeData(i)), p.call(this, i);
          } : function(i, n, t) {
            var e = !!t;
            return !n && (i = this._normalizeData(i)), t = x(i, r.parsedVisual, t), e ? t : M(t, "rgba");
          }, this);
        },
        _normalizedToVisual: {
          linear: function(r) {
            return M(x(r, this.option.parsedVisual), "rgba");
          },
          category: p,
          piecewise: function(r, i) {
            var n = P.call(this, i);
            return n == null && (n = M(x(r, this.option.parsedVisual), "rgba")), n;
          },
          fixed: c
        }
      },
      colorHue: m(function(r, i) {
        return b(r, i);
      }),
      colorSaturation: m(function(r, i) {
        return b(r, null, i);
      }),
      colorLightness: m(function(r, i) {
        return b(r, null, null, i);
      }),
      colorAlpha: m(function(r, i) {
        return $(r, i);
      }),
      decal: {
        applyVisual: f("decal"),
        _normalizedToVisual: {
          linear: null,
          category: p,
          piecewise: null,
          fixed: null
        }
      },
      opacity: {
        applyVisual: f("opacity"),
        _normalizedToVisual: _([0, 1])
      },
      liftZ: {
        applyVisual: f("liftZ"),
        _normalizedToVisual: {
          linear: c,
          category: c,
          piecewise: c,
          fixed: c
        }
      },
      symbol: {
        applyVisual: function(r, i, n) {
          var t = this.mapValueToVisual(r);
          n("symbol", t);
        },
        _normalizedToVisual: {
          linear: N,
          category: p,
          piecewise: function(r, i) {
            var n = P.call(this, i);
            return n == null && (n = N.call(this, r)), n;
          },
          fixed: c
        }
      },
      symbolSize: {
        applyVisual: f("symbolSize"),
        _normalizedToVisual: _([0, 1])
      }
    }, a;
  })()
);
function X(a) {
  var r = a.pieceList;
  a.hasSpecialVisual = !1, S(r, function(i, n) {
    i.originIndex = n, i.visual != null && (a.hasSpecialVisual = !0);
  });
}
function Y(a) {
  var r = a.categories, i = a.categoryMap = {}, n = a.visual;
  if (V(r, function(o, u) {
    i[o] = u;
  }), !I(n)) {
    var t = [];
    h(n) ? V(n, function(o, u) {
      var l = i[u];
      t[l ?? v] = o;
    }) : t[v] = n, n = L(a, t);
  }
  for (var e = r.length - 1; e >= 0; e--)
    n[e] == null && (delete i[r[e]], r.pop());
}
function w(a, r) {
  var i = a.visual, n = [];
  h(i) ? V(i, function(e) {
    n.push(e);
  }) : i != null && n.push(i);
  var t = {
    color: 1,
    symbol: 1
  };
  !r && n.length === 1 && !t.hasOwnProperty(a.type) && (n[1] = n[0]), L(a, n);
}
function m(a) {
  return {
    applyVisual: function(r, i, n) {
      var t = this.mapValueToVisual(r);
      n("color", a(i("color"), t));
    },
    _normalizedToVisual: _([0, 1])
  };
}
function N(a) {
  var r = this.option.visual;
  return r[Math.round(d(a, [0, 1], [0, r.length - 1], !0))] || {};
}
function f(a) {
  return function(r, i, n) {
    n(a, this.mapValueToVisual(r));
  };
}
function p(a) {
  var r = this.option.visual;
  return r[this.option.loop && a !== v ? a % r.length : a];
}
function c() {
  return this.option.visual[0];
}
function _(a) {
  return {
    linear: function(r) {
      return d(r, a, this.option.visual, !0);
    },
    category: p,
    piecewise: function(r, i) {
      var n = P.call(this, i);
      return n == null && (n = d(r, a, this.option.visual, !0)), n;
    },
    fixed: c
  };
}
function P(a) {
  var r = this.option, i = r.pieceList;
  if (r.hasSpecialVisual) {
    var n = A.findPieceIndex(a, i), t = i[n];
    if (t && t.visual)
      return t.visual[this.type];
  }
}
function L(a, r) {
  return a.visual = r, a.type === "color" && (a.parsedVisual = k(r, function(i) {
    var n = Z(i);
    return !n && process.env.NODE_ENV !== "production" && j("'" + i + "' is an illegal color, fallback to '#000000'", !0), n || [0, 0, 0, 1];
  })), r;
}
var q = {
  linear: function(a) {
    return d(a, this.option.dataExtent, [0, 1], !0);
  },
  piecewise: function(a) {
    var r = this.option.pieceList, i = A.findPieceIndex(a, r, !0);
    if (i != null)
      return d(i, [0, r.length - 1], [0, 1], !0);
  },
  category: function(a) {
    var r = this.option.categories ? this.option.categoryMap[a] : a;
    return r ?? v;
  },
  fixed: O
};
function y(a, r, i) {
  return a ? r <= i : r < i;
}
export {
  A as default
};
