// zrender@6.0.0/lib/dom/HandlerProxy.js
import { __extends as E } from "../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { normalizeEvent as u, removeEventListener as S, addEventListener as D, getNativeEvent as H } from "../core/event.js";
import { each as l, map as L, noop as h } from "../core/util.js";
import O from "../core/Eventful.js";
import a from "../core/env.js";
var b = 300, c = a.domSupported, p = (function() {
  var o = [
    "click",
    "dblclick",
    "mousewheel",
    "wheel",
    "mouseout",
    "mouseup",
    "mousedown",
    "mousemove",
    "contextmenu"
  ], t = [
    "touchstart",
    "touchend",
    "touchmove"
  ], r = {
    pointerdown: 1,
    pointerup: 1,
    pointermove: 1,
    pointerout: 1
  }, i = L(o, function(e) {
    var s = e.replace("mouse", "pointer");
    return r.hasOwnProperty(s) ? s : e;
  });
  return {
    mouse: o,
    touch: t,
    pointer: i
  };
})(), T = {
  mouse: ["mousemove", "mouseup"],
  pointer: ["pointermove", "pointerup"]
}, w = !1;
function g(o) {
  var t = o.pointerType;
  return t === "pen" || t === "touch";
}
function P(o) {
  o.touching = !0, o.touchTimer != null && (clearTimeout(o.touchTimer), o.touchTimer = null), o.touchTimer = setTimeout(function() {
    o.touching = !1, o.touchTimer = null;
  }, 700);
}
function f(o) {
  o && (o.zrByTouch = !0);
}
function M(o, t) {
  return u(o.dom, new z(o, t), !0);
}
function C(o, t) {
  for (var r = t, i = !1; r && r.nodeType !== 9 && !(i = r.domBelongToZr || r !== t && r === o.painterRoot); )
    r = r.parentNode;
  return i;
}
var z = /* @__PURE__ */ (function() {
  function o(t, r) {
    this.stopPropagation = h, this.stopImmediatePropagation = h, this.preventDefault = h, this.type = r.type, this.target = this.currentTarget = t.dom, this.pointerType = r.pointerType, this.clientX = r.clientX, this.clientY = r.clientY;
  }
  return o;
})(), n = {
  mousedown: function(o) {
    o = u(this.dom, o), this.__mayPointerCapture = [o.zrX, o.zrY], this.trigger("mousedown", o);
  },
  mousemove: function(o) {
    o = u(this.dom, o);
    var t = this.__mayPointerCapture;
    t && (o.zrX !== t[0] || o.zrY !== t[1]) && this.__togglePointerCapture(!0), this.trigger("mousemove", o);
  },
  mouseup: function(o) {
    o = u(this.dom, o), this.__togglePointerCapture(!1), this.trigger("mouseup", o);
  },
  mouseout: function(o) {
    o = u(this.dom, o);
    var t = o.toElement || o.relatedTarget;
    C(this, t) || (this.__pointerCapturing && (o.zrEventControl = "no_globalout"), this.trigger("mouseout", o));
  },
  wheel: function(o) {
    w = !0, o = u(this.dom, o), this.trigger("mousewheel", o);
  },
  mousewheel: function(o) {
    w || (o = u(this.dom, o), this.trigger("mousewheel", o));
  },
  touchstart: function(o) {
    o = u(this.dom, o), f(o), this.__lastTouchMoment = /* @__PURE__ */ new Date(), this.handler.processGesture(o, "start"), n.mousemove.call(this, o), n.mousedown.call(this, o);
  },
  touchmove: function(o) {
    o = u(this.dom, o), f(o), this.handler.processGesture(o, "change"), n.mousemove.call(this, o);
  },
  touchend: function(o) {
    o = u(this.dom, o), f(o), this.handler.processGesture(o, "end"), n.mouseup.call(this, o), +/* @__PURE__ */ new Date() - +this.__lastTouchMoment < b && n.click.call(this, o);
  },
  pointerdown: function(o) {
    n.mousedown.call(this, o);
  },
  pointermove: function(o) {
    g(o) || n.mousemove.call(this, o);
  },
  pointerup: function(o) {
    n.mouseup.call(this, o);
  },
  pointerout: function(o) {
    g(o) || n.mouseout.call(this, o);
  }
};
l(["click", "dblclick", "contextmenu"], function(o) {
  n[o] = function(t) {
    t = u(this.dom, t), this.trigger(o, t);
  };
});
var _ = {
  pointermove: function(o) {
    g(o) || _.mousemove.call(this, o);
  },
  pointerup: function(o) {
    _.mouseup.call(this, o);
  },
  mousemove: function(o) {
    this.trigger("mousemove", o);
  },
  mouseup: function(o) {
    var t = this.__pointerCapturing;
    this.__togglePointerCapture(!1), this.trigger("mouseup", o), t && (o.zrEventControl = "only_globalout", this.trigger("mouseout", o));
  }
};
function k(o, t) {
  var r = t.domHandlers;
  a.pointerEventsSupported ? l(p.pointer, function(i) {
    m(t, i, function(e) {
      r[i].call(o, e);
    });
  }) : (a.touchEventsSupported && l(p.touch, function(i) {
    m(t, i, function(e) {
      r[i].call(o, e), P(t);
    });
  }), l(p.mouse, function(i) {
    m(t, i, function(e) {
      e = H(e), t.touching || r[i].call(o, e);
    });
  }));
}
function G(o, t) {
  a.pointerEventsSupported ? l(T.pointer, r) : a.touchEventsSupported || l(T.mouse, r);
  function r(i) {
    function e(s) {
      s = H(s), C(o, s.target) || (s = M(o, s), t.domHandlers[i].call(o, s));
    }
    m(t, i, e, { capture: !0 });
  }
}
function m(o, t, r, i) {
  o.mounted[t] = r, o.listenerOpts[t] = i, D(o.domTarget, t, r, i);
}
function d(o) {
  var t = o.mounted;
  for (var r in t)
    t.hasOwnProperty(r) && S(o.domTarget, r, t[r], o.listenerOpts[r]);
  o.mounted = {};
}
var y = /* @__PURE__ */ (function() {
  function o(t, r) {
    this.mounted = {}, this.listenerOpts = {}, this.touching = !1, this.domTarget = t, this.domHandlers = r;
  }
  return o;
})(), v = (function(o) {
  E(t, o);
  function t(r, i) {
    var e = o.call(this) || this;
    return e.__pointerCapturing = !1, e.dom = r, e.painterRoot = i, e._localHandlerScope = new y(r, n), c && (e._globalHandlerScope = new y(document, _)), k(e, e._localHandlerScope), e;
  }
  return t.prototype.dispose = function() {
    d(this._localHandlerScope), c && d(this._globalHandlerScope);
  }, t.prototype.setCursor = function(r) {
    this.dom.style && (this.dom.style.cursor = r || "default");
  }, t.prototype.__togglePointerCapture = function(r) {
    if (this.__mayPointerCapture = null, c && +this.__pointerCapturing ^ +r) {
      this.__pointerCapturing = r;
      var i = this._globalHandlerScope;
      r ? G(this, i) : d(i);
    }
  }, t;
})(O);
export {
  v as default
};
