// echarts@6.0.0/lib/preprocessor/helper/compatStyle.js
import { each as x, isArray as S, isObject as C, isTypedArray as _, defaults as j, merge as k } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { TEXT_STYLE_OPTIONS as P } from "../../util/model.js";
import { deprecateLog as s, deprecateReplaceLog as L } from "../../util/log.js";
var o = x, m = C, V = ["areaStyle", "lineStyle", "nodeStyle", "linkStyle", "chordStyle", "label", "labelLine"];
function E(a) {
  var e = a && a.itemStyle;
  if (e)
    for (var f = 0, i = V.length; f < i; f++) {
      var n = V[f], c = e.normal, l = e.emphasis;
      c && c[n] && (process.env.NODE_ENV !== "production" && L("itemStyle.normal." + n, n), a[n] = a[n] || {}, a[n].normal ? k(a[n].normal, c[n]) : a[n].normal = c[n], c[n] = null), l && l[n] && (process.env.NODE_ENV !== "production" && L("itemStyle.emphasis." + n, "emphasis." + n), a[n] = a[n] || {}, a[n].emphasis ? k(a[n].emphasis, l[n]) : a[n].emphasis = l[n], l[n] = null);
    }
}
function u(a, e, f) {
  if (a && a[e] && (a[e].normal || a[e].emphasis)) {
    var i = a[e].normal, n = a[e].emphasis;
    i && (process.env.NODE_ENV !== "production" && s("'normal' hierarchy in " + e + " has been removed since 4.0. All style properties are configured in " + e + " directly now."), f ? (a[e].normal = a[e].emphasis = null, j(a[e], i)) : a[e] = i), n && (process.env.NODE_ENV !== "production" && s(e + ".emphasis has been changed to emphasis." + e + " since 4.0"), a.emphasis = a.emphasis || {}, a.emphasis[e] = n, n.focus && (a.emphasis.focus = n.focus), n.blurScope && (a.emphasis.blurScope = n.blurScope));
  }
}
function d(a) {
  u(a, "itemStyle"), u(a, "lineStyle"), u(a, "areaStyle"), u(a, "label"), u(a, "labelLine"), u(a, "upperLabel"), u(a, "edgeLabel");
}
function r(a, e) {
  var f = m(a) && a[e], i = m(f) && f.textStyle;
  if (i) {
    process.env.NODE_ENV !== "production" && s("textStyle hierarchy in " + e + " has been removed since 4.0. All textStyle properties are configured in " + e + " directly now.");
    for (var n = 0, c = P.length; n < c; n++) {
      var l = P[n];
      i.hasOwnProperty(l) && (f[l] = i[l]);
    }
  }
}
function v(a) {
  a && (d(a), r(a, "label"), a.emphasis && r(a.emphasis, "label"));
}
function y(a) {
  if (m(a)) {
    E(a), d(a), r(a, "label"), r(a, "upperLabel"), r(a, "edgeLabel"), a.emphasis && (r(a.emphasis, "label"), r(a.emphasis, "upperLabel"), r(a.emphasis, "edgeLabel"));
    var e = a.markPoint;
    e && (E(e), v(e));
    var f = a.markLine;
    f && (E(f), v(f));
    var i = a.markArea;
    i && v(i);
    var n = a.data;
    if (a.type === "graph") {
      n = n || a.nodes;
      var c = a.links || a.edges;
      if (c && !_(c))
        for (var l = 0; l < c.length; l++)
          v(c[l]);
      x(a.categories, function(g) {
        d(g);
      });
    }
    if (n && !_(n))
      for (var l = 0; l < n.length; l++)
        v(n[l]);
    if (e = a.markPoint, e && e.data)
      for (var A = e.data, l = 0; l < A.length; l++)
        v(A[l]);
    if (f = a.markLine, f && f.data)
      for (var h = f.data, l = 0; l < h.length; l++)
        S(h[l]) ? (v(h[l][0]), v(h[l][1])) : v(h[l]);
    a.type === "gauge" ? (r(a, "axisLabel"), r(a, "title"), r(a, "detail")) : a.type === "treemap" ? (u(a.breadcrumb, "itemStyle"), x(a.levels, function(g) {
      d(g);
    })) : a.type === "tree" && d(a.leaves);
  }
}
function b(a) {
  return S(a) ? a : a ? [a] : [];
}
function G(a) {
  return (S(a) ? a[0] : a) || {};
}
function $(a, e) {
  o(b(a.series), function(i) {
    m(i) && y(i);
  });
  var f = ["xAxis", "yAxis", "radiusAxis", "angleAxis", "singleAxis", "parallelAxis", "radar"];
  e && f.push("valueAxis", "categoryAxis", "logAxis", "timeAxis"), o(f, function(i) {
    o(b(a[i]), function(n) {
      n && (r(n, "axisLabel"), r(n.axisPointer, "label"));
    });
  }), o(b(a.parallel), function(i) {
    var n = i && i.parallelAxisDefault;
    r(n, "axisLabel"), r(n && n.axisPointer, "label");
  }), o(b(a.calendar), function(i) {
    u(i, "itemStyle"), r(i, "dayLabel"), r(i, "monthLabel"), r(i, "yearLabel");
  }), o(b(a.radar), function(i) {
    r(i, "name"), i.name && i.axisName == null && (i.axisName = i.name, delete i.name, process.env.NODE_ENV !== "production" && s("name property in radar component has been changed to axisName")), i.nameGap != null && i.axisNameGap == null && (i.axisNameGap = i.nameGap, delete i.nameGap, process.env.NODE_ENV !== "production" && s("nameGap property in radar component has been changed to axisNameGap")), process.env.NODE_ENV !== "production" && o(i.indicator, function(n) {
      n.text && L("text", "name", "radar.indicator");
    });
  }), o(b(a.geo), function(i) {
    m(i) && (v(i), o(b(i.regions), function(n) {
      v(n);
    }));
  }), o(b(a.timeline), function(i) {
    v(i), u(i, "label"), u(i, "itemStyle"), u(i, "controlStyle", !0);
    var n = i.data;
    S(n) && x(n, function(c) {
      C(c) && (u(c, "label"), u(c, "itemStyle"));
    });
  }), o(b(a.toolbox), function(i) {
    u(i, "iconStyle"), o(i.feature, function(n) {
      u(n, "iconStyle");
    });
  }), r(G(a.axisPointer), "label"), r(G(a.tooltip).axisPointer, "label");
}
export {
  $ as default
};
