// echarts@6.0.0/lib/preprocessor/backwardCompat.js
import { each as m, isObject as y, isTypedArray as v, defaults as E, isArray as b } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import p from "./helper/compatStyle.js";
import { normalizeToArray as N } from "../util/model.js";
import { deprecateReplaceLog as n, deprecateLog as _ } from "../util/log.js";
function A(l, c) {
  for (var a = c.split(","), e = l, r = 0; r < a.length && (e = e && e[a[r]], e != null); r++)
    ;
  return e;
}
function T(l, c, a, e) {
  for (var r = c.split(","), f = l, o, i = 0; i < r.length - 1; i++)
    o = r[i], f[o] == null && (f[o] = {}), f = f[o];
  f[r[i]] == null && (f[r[i]] = a);
}
function s(l) {
  l && m(D, function(c) {
    c[0] in l && !(c[1] in l) && (l[c[1]] = l[c[0]]);
  });
}
var D = [["x", "left"], ["y", "top"], ["x2", "right"], ["y2", "bottom"]], S = ["grid", "geo", "parallel", "legend", "toolbox", "title", "visualMap", "dataZoom", "timeline"], h = [["borderRadius", "barBorderRadius"], ["borderColor", "barBorderColor"], ["borderWidth", "barBorderWidth"]];
function u(l) {
  var c = l && l.itemStyle;
  if (c)
    for (var a = 0; a < h.length; a++) {
      var e = h[a][1], r = h[a][0];
      c[e] != null && (c[r] = c[e], process.env.NODE_ENV !== "production" && n(e, r));
    }
}
function d(l) {
  l && l.alignTo === "edge" && l.margin != null && l.edgeDistance == null && (process.env.NODE_ENV !== "production" && n("label.margin", "label.edgeDistance", "pie"), l.edgeDistance = l.margin);
}
function t(l) {
  l && l.downplay && !l.blur && (l.blur = l.downplay, process.env.NODE_ENV !== "production" && n("downplay", "blur", "sunburst"));
}
function k(l) {
  l && l.focusNodeAdjacency != null && (l.emphasis = l.emphasis || {}, l.emphasis.focus == null && (process.env.NODE_ENV !== "production" && n("focusNodeAdjacency", "emphasis: { focus: 'adjacency'}", "graph/sankey"), l.emphasis.focus = "adjacency"));
}
function g(l, c) {
  if (l)
    for (var a = 0; a < l.length; a++)
      c(l[a]), l[a] && g(l[a].children, c);
}
function P(l, c) {
  p(l, c), l.series = N(l.series), m(l.series, function(a) {
    if (y(a)) {
      var e = a.type;
      if (e === "line")
        a.clipOverflow != null && (a.clip = a.clipOverflow, process.env.NODE_ENV !== "production" && n("clipOverflow", "clip", "line"));
      else if (e === "pie" || e === "gauge") {
        a.clockWise != null && (a.clockwise = a.clockWise, process.env.NODE_ENV !== "production" && n("clockWise", "clockwise")), d(a.label);
        var r = a.data;
        if (r && !v(r))
          for (var f = 0; f < r.length; f++)
            d(r[f]);
        a.hoverOffset != null && (a.emphasis = a.emphasis || {}, (a.emphasis.scaleSize = null) && (process.env.NODE_ENV !== "production" && n("hoverOffset", "emphasis.scaleSize"), a.emphasis.scaleSize = a.hoverOffset));
      } else if (e === "gauge") {
        var o = A(a, "pointer.color");
        o != null && T(a, "itemStyle.color", o);
      } else if (e === "bar") {
        u(a), u(a.backgroundStyle), u(a.emphasis);
        var r = a.data;
        if (r && !v(r))
          for (var f = 0; f < r.length; f++)
            typeof r[f] == "object" && (u(r[f]), u(r[f] && r[f].emphasis));
      } else if (e === "sunburst") {
        var i = a.highlightPolicy;
        i && (a.emphasis = a.emphasis || {}, a.emphasis.focus || (a.emphasis.focus = i, process.env.NODE_ENV !== "production" && n("highlightPolicy", "emphasis.focus", "sunburst"))), t(a), g(a.data, t);
      } else e === "graph" || e === "sankey" ? k(a) : e === "map" && (a.mapType && !a.map && (process.env.NODE_ENV !== "production" && n("mapType", "map", "map"), a.map = a.mapType), a.mapLocation && (process.env.NODE_ENV !== "production" && _("`mapLocation` is not used anymore."), E(a, a.mapLocation)));
      a.hoverAnimation != null && (a.emphasis = a.emphasis || {}, a.emphasis && a.emphasis.scale == null && (process.env.NODE_ENV !== "production" && n("hoverAnimation", "emphasis.scale"), a.emphasis.scale = a.hoverAnimation)), s(a);
    }
  }), l.dataRange && (l.visualMap = l.dataRange), m(S, function(a) {
    var e = l[a];
    e && (b(e) || (e = [e]), m(e, function(r) {
      s(r);
    }));
  });
}
export {
  P as default
};
