// tippy.js@6.3.7/dist/tippy.esm.js
import { createPopper as bt } from "../../../../@popperjs_core@2.11.8/node_modules/@popperjs/core/lib/popper.js";
import Tt from "../../../../@popperjs_core@2.11.8/node_modules/@popperjs/core/lib/modifiers/applyStyles.js";
var Et = "tippy-box", ze = "tippy-content", wt = "tippy-backdrop", qe = "tippy-arrow", Ye = "tippy-svg-arrow", S = {
  passive: !0,
  capture: !0
}, Ke = function() {
  return document.body;
};
function Ot(e, i) {
  return {}.hasOwnProperty.call(e, i);
}
function ae(e, i, r) {
  if (Array.isArray(e)) {
    var a = e[i];
    return a ?? (Array.isArray(r) ? r[i] : r);
  }
  return e;
}
function le(e, i) {
  var r = {}.toString.call(e);
  return r.indexOf("[object") === 0 && r.indexOf(i + "]") > -1;
}
function Xe(e, i) {
  return typeof e == "function" ? e.apply(void 0, i) : e;
}
function _e(e, i) {
  if (i === 0)
    return e;
  var r;
  return function(a) {
    clearTimeout(r), r = setTimeout(function() {
      e(a);
    }, i);
  };
}
function Dt(e, i) {
  var r = Object.assign({}, e);
  return i.forEach(function(a) {
    delete r[a];
  }), r;
}
function At(e) {
  return e.split(/\s+/).filter(Boolean);
}
function I(e) {
  return [].concat(e);
}
function Ue(e, i) {
  e.indexOf(i) === -1 && e.push(i);
}
function Lt(e) {
  return e.filter(function(i, r) {
    return e.indexOf(i) === r;
  });
}
function Ct(e) {
  return e.split("-")[0];
}
function J(e) {
  return [].slice.call(e);
}
function Re(e) {
  return Object.keys(e).reduce(function(i, r) {
    return e[r] !== void 0 && (i[r] = e[r]), i;
  }, {});
}
function k() {
  return document.createElement("div");
}
function P(e) {
  return ["Element", "Fragment"].some(function(i) {
    return le(e, i);
  });
}
function St(e) {
  return le(e, "NodeList");
}
function Nt(e) {
  return le(e, "MouseEvent");
}
function xt(e) {
  return !!(e && e._tippy && e._tippy.reference === e);
}
function Mt(e) {
  return P(e) ? [e] : St(e) ? J(e) : Array.isArray(e) ? e : J(document.querySelectorAll(e));
}
function se(e, i) {
  e.forEach(function(r) {
    r && (r.style.transitionDuration = i + "ms");
  });
}
function ke(e, i) {
  e.forEach(function(r) {
    r && r.setAttribute("data-state", i);
  });
}
function Vt(e) {
  var i, r = I(e), a = r[0];
  return a != null && (i = a.ownerDocument) != null && i.body ? a.ownerDocument : document;
}
function It(e, i) {
  var r = i.clientX, a = i.clientY;
  return e.every(function(l) {
    var d = l.popperRect, u = l.popperState, h = l.props, v = h.interactiveBorder, y = Ct(u.placement), m = u.modifiersData.offset;
    if (!m)
      return !0;
    var H = y === "bottom" ? m.top.y : 0, F = y === "top" ? m.bottom.y : 0, j = y === "right" ? m.left.x : 0, D = y === "left" ? m.right.x : 0, W = d.top - a + H > v, Z = a - d.bottom - F > v, ee = d.left - r + j > v, $ = r - d.right - D > v;
    return W || Z || ee || $;
  });
}
function ue(e, i, r) {
  var a = i + "EventListener";
  ["transitionend", "webkitTransitionEnd"].forEach(function(l) {
    e[a](l, r);
  });
}
function Pe(e, i) {
  for (var r = i; r; ) {
    var a;
    if (e.contains(r))
      return !0;
    r = r.getRootNode == null || (a = r.getRootNode()) == null ? void 0 : a.host;
  }
  return !1;
}
var O = {
  isTouch: !1
}, Be = 0;
function jt() {
  O.isTouch || (O.isTouch = !0, window.performance && document.addEventListener("mousemove", Ge));
}
function Ge() {
  var e = performance.now();
  e - Be < 20 && (O.isTouch = !1, document.removeEventListener("mousemove", Ge)), Be = e;
}
function _t() {
  var e = document.activeElement;
  if (xt(e)) {
    var i = e._tippy;
    e.blur && !i.state.isVisible && e.blur();
  }
}
function Ut() {
  document.addEventListener("touchstart", jt, S), window.addEventListener("blur", _t);
}
var Rt = typeof window < "u" && typeof document < "u", kt = Rt ? (
  // @ts-ignore
  !!window.msCrypto
) : !1;
function V(e) {
  var i = e === "destroy" ? "n already-" : " ";
  return [e + "() was called on a" + i + "destroyed instance. This is a no-op but", "indicates a potential memory leak."].join(" ");
}
function He(e) {
  var i = /[ \t]{2,}/g, r = /^[ \t]*/gm;
  return e.replace(i, " ").replace(r, "").trim();
}
function Pt(e) {
  return He(`
  %ctippy.js

  %c` + He(e) + `

  %c👷‍ This is a development-only message. It will be removed in production.
  `);
}
function Je(e) {
  return [
    Pt(e),
    // title
    "color: #00C584; font-size: 1.3em; font-weight: bold;",
    // message
    "line-height: 1.5",
    // footer
    "color: #a6a095;"
  ];
}
var B;
process.env.NODE_ENV !== "production" && Bt();
function Bt() {
  B = /* @__PURE__ */ new Set();
}
function A(e, i) {
  if (e && !B.has(i)) {
    var r;
    B.add(i), (r = console).warn.apply(r, Je(i));
  }
}
function pe(e, i) {
  if (e && !B.has(i)) {
    var r;
    B.add(i), (r = console).error.apply(r, Je(i));
  }
}
function Ht(e) {
  var i = !e, r = Object.prototype.toString.call(e) === "[object Object]" && !e.addEventListener;
  pe(i, ["tippy() was passed", "`" + String(e) + "`", "as its targets (first) argument. Valid types are: String, Element,", "Element[], or NodeList."].join(" ")), pe(r, ["tippy() was passed a plain object which is not supported as an argument", "for virtual positioning. Use props.getReferenceClientRect instead."].join(" "));
}
var Qe = {
  animateFill: !1,
  followCursor: !1,
  inlinePositioning: !1,
  sticky: !1
}, Ft = {
  allowHTML: !1,
  animation: "fade",
  arrow: !0,
  content: "",
  inertia: !1,
  maxWidth: 350,
  role: "tooltip",
  theme: "",
  zIndex: 9999
}, b = Object.assign({
  appendTo: Ke,
  aria: {
    content: "auto",
    expanded: "auto"
  },
  delay: 0,
  duration: [300, 250],
  getReferenceClientRect: null,
  hideOnClick: !0,
  ignoreAttributes: !1,
  interactive: !1,
  interactiveBorder: 2,
  interactiveDebounce: 0,
  moveTransition: "",
  offset: [0, 10],
  onAfterUpdate: function() {
  },
  onBeforeUpdate: function() {
  },
  onCreate: function() {
  },
  onDestroy: function() {
  },
  onHidden: function() {
  },
  onHide: function() {
  },
  onMount: function() {
  },
  onShow: function() {
  },
  onShown: function() {
  },
  onTrigger: function() {
  },
  onUntrigger: function() {
  },
  onClickOutside: function() {
  },
  placement: "top",
  plugins: [],
  popperOptions: {},
  render: null,
  showOnCreate: !1,
  touch: !0,
  trigger: "mouseenter focus",
  triggerTarget: null
}, Qe, Ft), Wt = Object.keys(b), $t = function(i) {
  process.env.NODE_ENV !== "production" && et(i, []);
  var r = Object.keys(i);
  r.forEach(function(a) {
    b[a] = i[a];
  });
};
function Ze(e) {
  var i = e.plugins || [], r = i.reduce(function(a, l) {
    var d = l.name, u = l.defaultValue;
    if (d) {
      var h;
      a[d] = e[d] !== void 0 ? e[d] : (h = b[d]) != null ? h : u;
    }
    return a;
  }, {});
  return Object.assign({}, e, r);
}
function zt(e, i) {
  var r = i ? Object.keys(Ze(Object.assign({}, b, {
    plugins: i
  }))) : Wt, a = r.reduce(function(l, d) {
    var u = (e.getAttribute("data-tippy-" + d) || "").trim();
    if (!u)
      return l;
    if (d === "content")
      l[d] = u;
    else
      try {
        l[d] = JSON.parse(u);
      } catch {
        l[d] = u;
      }
    return l;
  }, {});
  return a;
}
function Fe(e, i) {
  var r = Object.assign({}, i, {
    content: Xe(i.content, [e])
  }, i.ignoreAttributes ? {} : zt(e, i.plugins));
  return r.aria = Object.assign({}, b.aria, r.aria), r.aria = {
    expanded: r.aria.expanded === "auto" ? i.interactive : r.aria.expanded,
    content: r.aria.content === "auto" ? i.interactive ? null : "describedby" : r.aria.content
  }, r;
}
function et(e, i) {
  e === void 0 && (e = {}), i === void 0 && (i = []);
  var r = Object.keys(e);
  r.forEach(function(a) {
    var l = Dt(b, Object.keys(Qe)), d = !Ot(l, a);
    d && (d = i.filter(function(u) {
      return u.name === a;
    }).length === 0), A(d, ["`" + a + "`", "is not a valid prop. You may have spelled it incorrectly, or if it's", "a plugin, forgot to pass it in an array as props.plugins.", `

`, `All props: https://atomiks.github.io/tippyjs/v6/all-props/
`, "Plugins: https://atomiks.github.io/tippyjs/v6/plugins/"].join(" "));
  });
}
var qt = function() {
  return "innerHTML";
};
function de(e, i) {
  e[qt()] = i;
}
function We(e) {
  var i = k();
  return e === !0 ? i.className = qe : (i.className = Ye, P(e) ? i.appendChild(e) : de(i, e)), i;
}
function $e(e, i) {
  P(i.content) ? (de(e, ""), e.appendChild(i.content)) : typeof i.content != "function" && (i.allowHTML ? de(e, i.content) : e.textContent = i.content);
}
function fe(e) {
  var i = e.firstElementChild, r = J(i.children);
  return {
    box: i,
    content: r.find(function(a) {
      return a.classList.contains(ze);
    }),
    arrow: r.find(function(a) {
      return a.classList.contains(qe) || a.classList.contains(Ye);
    }),
    backdrop: r.find(function(a) {
      return a.classList.contains(wt);
    })
  };
}
function tt(e) {
  var i = k(), r = k();
  r.className = Et, r.setAttribute("data-state", "hidden"), r.setAttribute("tabindex", "-1");
  var a = k();
  a.className = ze, a.setAttribute("data-state", "hidden"), $e(a, e.props), i.appendChild(r), r.appendChild(a), l(e.props, e.props);
  function l(d, u) {
    var h = fe(i), v = h.box, y = h.content, m = h.arrow;
    u.theme ? v.setAttribute("data-theme", u.theme) : v.removeAttribute("data-theme"), typeof u.animation == "string" ? v.setAttribute("data-animation", u.animation) : v.removeAttribute("data-animation"), u.inertia ? v.setAttribute("data-inertia", "") : v.removeAttribute("data-inertia"), v.style.maxWidth = typeof u.maxWidth == "number" ? u.maxWidth + "px" : u.maxWidth, u.role ? v.setAttribute("role", u.role) : v.removeAttribute("role"), (d.content !== u.content || d.allowHTML !== u.allowHTML) && $e(y, e.props), u.arrow ? m ? d.arrow !== u.arrow && (v.removeChild(m), v.appendChild(We(u.arrow))) : v.appendChild(We(u.arrow)) : m && v.removeChild(m);
  }
  return {
    popper: i,
    onUpdate: l
  };
}
tt.$$tippy = !0;
var Yt = 1, G = [], ce = [];
function Kt(e, i) {
  var r = Fe(e, Object.assign({}, b, Ze(Re(i)))), a, l, d, u = !1, h = !1, v = !1, y = !1, m, H, F, j = [], D = _e(Ce, r.interactiveDebounce), W, Z = Yt++, ee = null, $ = Lt(r.plugins), nt = {
    // Is the instance currently enabled?
    isEnabled: !0,
    // Is the tippy currently showing and not transitioning out?
    isVisible: !1,
    // Has the instance been destroyed?
    isDestroyed: !1,
    // Is the tippy currently mounted to the DOM?
    isMounted: !1,
    // Has the tippy finished transitioning in?
    isShown: !1
  }, t = {
    // properties
    id: Z,
    reference: e,
    popper: k(),
    popperInstance: ee,
    props: r,
    state: nt,
    plugins: $,
    // methods
    clearDelayTimeouts: pt,
    setProps: dt,
    setContent: ft,
    show: lt,
    hide: vt,
    hideWithInteractivity: mt,
    enable: ut,
    disable: ct,
    unmount: ht,
    destroy: gt
  };
  if (!r.render)
    return process.env.NODE_ENV !== "production" && pe(!0, "render() function has not been supplied."), t;
  var ve = r.render(t), p = ve.popper, me = ve.onUpdate;
  p.setAttribute("data-tippy-root", ""), p.id = "tippy-" + t.id, t.popper = p, e._tippy = t, p._tippy = t;
  var it = $.map(function(n) {
    return n.fn(t);
  }), rt = e.hasAttribute("aria-expanded");
  return De(), R(), z(), T("onCreate", [t]), r.showOnCreate && Ie(), p.addEventListener("mouseenter", function() {
    t.props.interactive && t.state.isVisible && t.clearDelayTimeouts();
  }), p.addEventListener("mouseleave", function() {
    t.props.interactive && t.props.trigger.indexOf("mouseenter") >= 0 && _().addEventListener("mousemove", D);
  }), t;
  function he() {
    var n = t.props.touch;
    return Array.isArray(n) ? n : [n, 0];
  }
  function ge() {
    return he()[0] === "hold";
  }
  function w() {
    var n;
    return !!((n = t.props.render) != null && n.$$tippy);
  }
  function L() {
    return W || e;
  }
  function _() {
    var n = L().parentNode;
    return n ? Vt(n) : document;
  }
  function U() {
    return fe(p);
  }
  function ye(n) {
    return t.state.isMounted && !t.state.isVisible || O.isTouch || m && m.type === "focus" ? 0 : ae(t.props.delay, n ? 0 : 1, b.delay);
  }
  function z(n) {
    n === void 0 && (n = !1), p.style.pointerEvents = t.props.interactive && !n ? "" : "none", p.style.zIndex = "" + t.props.zIndex;
  }
  function T(n, o, s) {
    if (s === void 0 && (s = !0), it.forEach(function(c) {
      c[n] && c[n].apply(c, o);
    }), s) {
      var f;
      (f = t.props)[n].apply(f, o);
    }
  }
  function be() {
    var n = t.props.aria;
    if (n.content) {
      var o = "aria-" + n.content, s = p.id, f = I(t.props.triggerTarget || e);
      f.forEach(function(c) {
        var g = c.getAttribute(o);
        if (t.state.isVisible)
          c.setAttribute(o, g ? g + " " + s : s);
        else {
          var E = g && g.replace(s, "").trim();
          E ? c.setAttribute(o, E) : c.removeAttribute(o);
        }
      });
    }
  }
  function R() {
    if (!(rt || !t.props.aria.expanded)) {
      var n = I(t.props.triggerTarget || e);
      n.forEach(function(o) {
        t.props.interactive ? o.setAttribute("aria-expanded", t.state.isVisible && o === L() ? "true" : "false") : o.removeAttribute("aria-expanded");
      });
    }
  }
  function te() {
    _().removeEventListener("mousemove", D), G = G.filter(function(n) {
      return n !== D;
    });
  }
  function q(n) {
    if (!(O.isTouch && (v || n.type === "mousedown"))) {
      var o = n.composedPath && n.composedPath()[0] || n.target;
      if (!(t.props.interactive && Pe(p, o))) {
        if (I(t.props.triggerTarget || e).some(function(s) {
          return Pe(s, o);
        })) {
          if (O.isTouch || t.state.isVisible && t.props.trigger.indexOf("click") >= 0)
            return;
        } else
          T("onClickOutside", [t, n]);
        t.props.hideOnClick === !0 && (t.clearDelayTimeouts(), t.hide(), h = !0, setTimeout(function() {
          h = !1;
        }), t.state.isMounted || ne());
      }
    }
  }
  function Te() {
    v = !0;
  }
  function Ee() {
    v = !1;
  }
  function we() {
    var n = _();
    n.addEventListener("mousedown", q, !0), n.addEventListener("touchend", q, S), n.addEventListener("touchstart", Ee, S), n.addEventListener("touchmove", Te, S);
  }
  function ne() {
    var n = _();
    n.removeEventListener("mousedown", q, !0), n.removeEventListener("touchend", q, S), n.removeEventListener("touchstart", Ee, S), n.removeEventListener("touchmove", Te, S);
  }
  function ot(n, o) {
    Oe(n, function() {
      !t.state.isVisible && p.parentNode && p.parentNode.contains(p) && o();
    });
  }
  function at(n, o) {
    Oe(n, o);
  }
  function Oe(n, o) {
    var s = U().box;
    function f(c) {
      c.target === s && (ue(s, "remove", f), o());
    }
    if (n === 0)
      return o();
    ue(s, "remove", H), ue(s, "add", f), H = f;
  }
  function N(n, o, s) {
    s === void 0 && (s = !1);
    var f = I(t.props.triggerTarget || e);
    f.forEach(function(c) {
      c.addEventListener(n, o, s), j.push({
        node: c,
        eventType: n,
        handler: o,
        options: s
      });
    });
  }
  function De() {
    ge() && (N("touchstart", Le, {
      passive: !0
    }), N("touchend", Se, {
      passive: !0
    })), At(t.props.trigger).forEach(function(n) {
      if (n !== "manual")
        switch (N(n, Le), n) {
          case "mouseenter":
            N("mouseleave", Se);
            break;
          case "focus":
            N(kt ? "focusout" : "blur", Ne);
            break;
          case "focusin":
            N("focusout", Ne);
            break;
        }
    });
  }
  function Ae() {
    j.forEach(function(n) {
      var o = n.node, s = n.eventType, f = n.handler, c = n.options;
      o.removeEventListener(s, f, c);
    }), j = [];
  }
  function Le(n) {
    var o, s = !1;
    if (!(!t.state.isEnabled || xe(n) || h)) {
      var f = ((o = m) == null ? void 0 : o.type) === "focus";
      m = n, W = n.currentTarget, R(), !t.state.isVisible && Nt(n) && G.forEach(function(c) {
        return c(n);
      }), n.type === "click" && (t.props.trigger.indexOf("mouseenter") < 0 || u) && t.props.hideOnClick !== !1 && t.state.isVisible ? s = !0 : Ie(n), n.type === "click" && (u = !s), s && !f && Y(n);
    }
  }
  function Ce(n) {
    var o = n.target, s = L().contains(o) || p.contains(o);
    if (!(n.type === "mousemove" && s)) {
      var f = ie().concat(p).map(function(c) {
        var g, E = c._tippy, x = (g = E.popperInstance) == null ? void 0 : g.state;
        return x ? {
          popperRect: c.getBoundingClientRect(),
          popperState: x,
          props: r
        } : null;
      }).filter(Boolean);
      It(f, n) && (te(), Y(n));
    }
  }
  function Se(n) {
    var o = xe(n) || t.props.trigger.indexOf("click") >= 0 && u;
    if (!o) {
      if (t.props.interactive) {
        t.hideWithInteractivity(n);
        return;
      }
      Y(n);
    }
  }
  function Ne(n) {
    t.props.trigger.indexOf("focusin") < 0 && n.target !== L() || t.props.interactive && n.relatedTarget && p.contains(n.relatedTarget) || Y(n);
  }
  function xe(n) {
    return O.isTouch ? ge() !== n.type.indexOf("touch") >= 0 : !1;
  }
  function Me() {
    Ve();
    var n = t.props, o = n.popperOptions, s = n.placement, f = n.offset, c = n.getReferenceClientRect, g = n.moveTransition, E = w() ? fe(p).arrow : null, x = c ? {
      getBoundingClientRect: c,
      contextElement: c.contextElement || L()
    } : e, je = {
      name: "$$tippy",
      enabled: !0,
      phase: "beforeWrite",
      requires: ["computeStyles"],
      fn: function(K) {
        var M = K.state;
        if (w()) {
          var yt = U(), oe = yt.box;
          ["placement", "reference-hidden", "escaped"].forEach(function(X) {
            X === "placement" ? oe.setAttribute("data-placement", M.placement) : M.attributes.popper["data-popper-" + X] ? oe.setAttribute("data-" + X, "") : oe.removeAttribute("data-" + X);
          }), M.attributes.popper = {};
        }
      }
    }, C = [{
      name: "offset",
      options: {
        offset: f
      }
    }, {
      name: "preventOverflow",
      options: {
        padding: {
          top: 2,
          bottom: 2,
          left: 5,
          right: 5
        }
      }
    }, {
      name: "flip",
      options: {
        padding: 5
      }
    }, {
      name: "computeStyles",
      options: {
        adaptive: !g
      }
    }, je];
    w() && E && C.push({
      name: "arrow",
      options: {
        element: E,
        padding: 3
      }
    }), C.push.apply(C, o?.modifiers || []), t.popperInstance = bt(x, p, Object.assign({}, o, {
      placement: s,
      onFirstUpdate: F,
      modifiers: C
    }));
  }
  function Ve() {
    t.popperInstance && (t.popperInstance.destroy(), t.popperInstance = null);
  }
  function st() {
    var n = t.props.appendTo, o, s = L();
    t.props.interactive && n === Ke || n === "parent" ? o = s.parentNode : o = Xe(n, [s]), o.contains(p) || o.appendChild(p), t.state.isMounted = !0, Me(), process.env.NODE_ENV !== "production" && A(t.props.interactive && n === b.appendTo && s.nextElementSibling !== p, ["Interactive tippy element may not be accessible via keyboard", "navigation because it is not directly after the reference element", "in the DOM source order.", `

`, "Using a wrapper <div> or <span> tag around the reference element", "solves this by creating a new parentNode context.", `

`, "Specifying `appendTo: document.body` silences this warning, but it", "assumes you are using a focus management solution to handle", "keyboard navigation.", `

`, "See: https://atomiks.github.io/tippyjs/v6/accessibility/#interactivity"].join(" "));
  }
  function ie() {
    return J(p.querySelectorAll("[data-tippy-root]"));
  }
  function Ie(n) {
    t.clearDelayTimeouts(), n && T("onTrigger", [t, n]), we();
    var o = ye(!0), s = he(), f = s[0], c = s[1];
    O.isTouch && f === "hold" && c && (o = c), o ? a = setTimeout(function() {
      t.show();
    }, o) : t.show();
  }
  function Y(n) {
    if (t.clearDelayTimeouts(), T("onUntrigger", [t, n]), !t.state.isVisible) {
      ne();
      return;
    }
    if (!(t.props.trigger.indexOf("mouseenter") >= 0 && t.props.trigger.indexOf("click") >= 0 && ["mouseleave", "mousemove"].indexOf(n.type) >= 0 && u)) {
      var o = ye(!1);
      o ? l = setTimeout(function() {
        t.state.isVisible && t.hide();
      }, o) : d = requestAnimationFrame(function() {
        t.hide();
      });
    }
  }
  function ut() {
    t.state.isEnabled = !0;
  }
  function ct() {
    t.hide(), t.state.isEnabled = !1;
  }
  function pt() {
    clearTimeout(a), clearTimeout(l), cancelAnimationFrame(d);
  }
  function dt(n) {
    if (process.env.NODE_ENV !== "production" && A(t.state.isDestroyed, V("setProps")), !t.state.isDestroyed) {
      T("onBeforeUpdate", [t, n]), Ae();
      var o = t.props, s = Fe(e, Object.assign({}, o, Re(n), {
        ignoreAttributes: !0
      }));
      t.props = s, De(), o.interactiveDebounce !== s.interactiveDebounce && (te(), D = _e(Ce, s.interactiveDebounce)), o.triggerTarget && !s.triggerTarget ? I(o.triggerTarget).forEach(function(f) {
        f.removeAttribute("aria-expanded");
      }) : s.triggerTarget && e.removeAttribute("aria-expanded"), R(), z(), me && me(o, s), t.popperInstance && (Me(), ie().forEach(function(f) {
        requestAnimationFrame(f._tippy.popperInstance.forceUpdate);
      })), T("onAfterUpdate", [t, n]);
    }
  }
  function ft(n) {
    t.setProps({
      content: n
    });
  }
  function lt() {
    process.env.NODE_ENV !== "production" && A(t.state.isDestroyed, V("show"));
    var n = t.state.isVisible, o = t.state.isDestroyed, s = !t.state.isEnabled, f = O.isTouch && !t.props.touch, c = ae(t.props.duration, 0, b.duration);
    if (!(n || o || s || f) && !L().hasAttribute("disabled") && (T("onShow", [t], !1), t.props.onShow(t) !== !1)) {
      if (t.state.isVisible = !0, w() && (p.style.visibility = "visible"), z(), we(), t.state.isMounted || (p.style.transition = "none"), w()) {
        var g = U(), E = g.box, x = g.content;
        se([E, x], 0);
      }
      F = function() {
        var C;
        if (!(!t.state.isVisible || y)) {
          if (y = !0, p.offsetHeight, p.style.transition = t.props.moveTransition, w() && t.props.animation) {
            var re = U(), K = re.box, M = re.content;
            se([K, M], c), ke([K, M], "visible");
          }
          be(), R(), Ue(ce, t), (C = t.popperInstance) == null || C.forceUpdate(), T("onMount", [t]), t.props.animation && w() && at(c, function() {
            t.state.isShown = !0, T("onShown", [t]);
          });
        }
      }, st();
    }
  }
  function vt() {
    process.env.NODE_ENV !== "production" && A(t.state.isDestroyed, V("hide"));
    var n = !t.state.isVisible, o = t.state.isDestroyed, s = !t.state.isEnabled, f = ae(t.props.duration, 1, b.duration);
    if (!(n || o || s) && (T("onHide", [t], !1), t.props.onHide(t) !== !1)) {
      if (t.state.isVisible = !1, t.state.isShown = !1, y = !1, u = !1, w() && (p.style.visibility = "hidden"), te(), ne(), z(!0), w()) {
        var c = U(), g = c.box, E = c.content;
        t.props.animation && (se([g, E], f), ke([g, E], "hidden"));
      }
      be(), R(), t.props.animation ? w() && ot(f, t.unmount) : t.unmount();
    }
  }
  function mt(n) {
    process.env.NODE_ENV !== "production" && A(t.state.isDestroyed, V("hideWithInteractivity")), _().addEventListener("mousemove", D), Ue(G, D), D(n);
  }
  function ht() {
    process.env.NODE_ENV !== "production" && A(t.state.isDestroyed, V("unmount")), t.state.isVisible && t.hide(), t.state.isMounted && (Ve(), ie().forEach(function(n) {
      n._tippy.unmount();
    }), p.parentNode && p.parentNode.removeChild(p), ce = ce.filter(function(n) {
      return n !== t;
    }), t.state.isMounted = !1, T("onHidden", [t]));
  }
  function gt() {
    process.env.NODE_ENV !== "production" && A(t.state.isDestroyed, V("destroy")), !t.state.isDestroyed && (t.clearDelayTimeouts(), t.unmount(), Ae(), delete e._tippy, t.state.isDestroyed = !0, T("onDestroy", [t]));
  }
}
function Q(e, i) {
  i === void 0 && (i = {});
  var r = b.plugins.concat(i.plugins || []);
  process.env.NODE_ENV !== "production" && (Ht(e), et(i, r)), Ut();
  var a = Object.assign({}, i, {
    plugins: r
  }), l = Mt(e);
  if (process.env.NODE_ENV !== "production") {
    var d = P(a.content), u = l.length > 1;
    A(d && u, ["tippy() was passed an Element as the `content` prop, but more than", "one tippy instance was created by this invocation. This means the", "content element will only be appended to the last tippy instance.", `

`, "Instead, pass the .innerHTML of the element, or use a function that", "returns a cloned version of the element instead.", `

`, `1) content: element.innerHTML
`, "2) content: () => element.cloneNode(true)"].join(" "));
  }
  var h = l.reduce(function(v, y) {
    var m = y && Kt(y, a);
    return m && v.push(m), v;
  }, []);
  return P(e) ? h[0] : h;
}
Q.defaultProps = b;
Q.setDefaultProps = $t;
Q.currentInput = O;
Object.assign({}, Tt, {
  effect: function(i) {
    var r = i.state, a = {
      popper: {
        position: r.options.strategy,
        left: "0",
        top: "0",
        margin: "0"
      },
      arrow: {
        position: "absolute"
      },
      reference: {}
    };
    Object.assign(r.elements.popper.style, a.popper), r.styles = a, r.elements.arrow && Object.assign(r.elements.arrow.style, a.arrow);
  }
});
Q.setDefaultProps({
  render: tt
});
export {
  Q as default
};
