// echarts@6.0.0/lib/chart/line/helper.js
import { isDimensionStacked as u } from "../../data/helper/dataStackHelper.js";
import { map as k, isNumber as x } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
function g(i, e, a) {
  var t = i.getBaseAxis(), s = i.getOtherAxis(t), m = p(s, a), r = t.dim, f = s.dim, l = e.mapDimension(f), o = e.mapDimension(r), c = f === "x" || f === "radius" ? 1 : 0, n = k(i.dimensions, function(b) {
    return e.mapDimension(b);
  }), v = !1, D = e.getCalculationInfo("stackResultDimension");
  return u(
    e,
    n[0]
    /* , dims[1] */
  ) && (v = !0, n[0] = D), u(
    e,
    n[1]
    /* , dims[0] */
  ) && (v = !0, n[1] = D), {
    dataDimsForPoint: n,
    valueStart: m,
    valueAxisDim: f,
    baseAxisDim: r,
    stacked: !!v,
    valueDim: l,
    baseDim: o,
    baseDataOffset: c,
    stackedOverDimension: e.getCalculationInfo("stackedOverDimension")
  };
}
function p(i, e) {
  var a = 0, t = i.scale.getExtent();
  return e === "start" ? a = t[0] : e === "end" ? a = t[1] : x(e) && !isNaN(e) ? a = e : t[0] > 0 ? a = t[0] : t[1] < 0 && (a = t[1]), a;
}
function O(i, e, a, t) {
  var s = NaN;
  i.stacked && (s = a.get(a.getCalculationInfo("stackedOverDimension"), t)), isNaN(s) && (s = i.valueStart);
  var m = i.baseDataOffset, r = [];
  return r[m] = a.get(i.baseDim, t), r[1 - m] = s, e.dataToPoint(r);
}
export {
  O as getStackedOnPoint,
  g as prepareDataCoordInfo
};
