// echarts@6.0.0/lib/chart/line/LineView.js
import { __extends as ut } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { bind as ht, defaults as Z, isFunction as T, each as it, map as ft, isNumber as pt } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import gt from "../helper/SymbolDraw.js";
import ot from "../helper/Symbol.js";
import mt from "./lineAnimationDiff.js";
import { queryDataIndex as q, interpolateRawValues as ct } from "../../util/model.js";
import { ECPolyline as yt, ECPolygon as dt } from "./poly.js";
import V from "../../view/Chart.js";
import { prepareDataCoordInfo as _t, getStackedOnPoint as bt } from "./helper.js";
import { createGridClipPath as St, createPolarClipPath as xt } from "../helper/createClipPathFromCoordSys.js";
import { isCoordinateSystemType as wt } from "../../coord/CoordinateSystem.js";
import { setStatesStylesFromModel as $, toggleHoverEmphasis as j, setStatesFlag as K, SPECIAL_STATES as Q } from "../../util/states.js";
import { setLabelStyle as It, getLabelStatesModels as Pt, labelInner as Dt } from "../../label/labelStyle.js";
import { getDefaultInterpolatedLabel as Lt, getDefaultLabel as Nt } from "../helper/labelHelper.js";
import { getECData as F } from "../../util/innerStore.js";
import { createFloat32Array as At } from "../../util/vendor.js";
import { convertToColorString as Y } from "../../util/format.js";
import { lerp as Et } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/tool/color.js";
import Ct from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import Ot from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Text.js";
import kt from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/LinearGradient.js";
import { initProps as Gt, updateProps as M } from "../../animation/basicTransition.js";
function tt(s, l) {
  if (s.length === l.length) {
    for (var t = 0; t < s.length; t++)
      if (s[t] !== l[t])
        return;
    return !0;
  }
}
function et(s) {
  for (var l = 1 / 0, t = 1 / 0, o = -1 / 0, n = -1 / 0, e = 0; e < s.length; ) {
    var a = s[e++], i = s[e++];
    isNaN(a) || (l = Math.min(a, l), o = Math.max(a, o)), isNaN(i) || (t = Math.min(i, t), n = Math.max(i, n));
  }
  return [[l, t], [o, n]];
}
function at(s, l) {
  var t = et(s), o = t[0], n = t[1], e = et(l), a = e[0], i = e[1];
  return Math.max(Math.abs(o[0] - a[0]), Math.abs(o[1] - a[1]), Math.abs(n[0] - i[0]), Math.abs(n[1] - i[1]));
}
function rt(s) {
  return pt(s) ? s : s ? 0.5 : 0;
}
function Tt(s, l, t) {
  if (!t.valueDim)
    return [];
  for (var o = l.count(), n = At(o * 2), e = 0; e < o; e++) {
    var a = bt(t, s, l, e);
    n[e * 2] = a[0], n[e * 2 + 1] = a[1];
  }
  return n;
}
function E(s, l, t, o, n) {
  var e = t.getBaseAxis(), a = e.dim === "x" || e.dim === "radius" ? 0 : 1, i = [], u = 0, r = [], v = [], f = [], h = [];
  if (n) {
    for (u = 0; u < s.length; u += 2) {
      var g = l || s;
      !isNaN(g[u]) && !isNaN(g[u + 1]) && h.push(s[u], s[u + 1]);
    }
    s = h;
  }
  for (u = 0; u < s.length - 2; u += 2)
    switch (f[0] = s[u + 2], f[1] = s[u + 3], v[0] = s[u], v[1] = s[u + 1], i.push(v[0], v[1]), o) {
      case "end":
        r[a] = f[a], r[1 - a] = v[1 - a], i.push(r[0], r[1]);
        break;
      case "middle":
        var m = (v[a] + f[a]) / 2, p = [];
        r[a] = p[a] = m, r[1 - a] = v[1 - a], p[1 - a] = f[1 - a], i.push(r[0], r[1]), i.push(p[0], p[1]);
        break;
      default:
        r[a] = v[a], r[1 - a] = f[1 - a], i.push(r[0], r[1]);
    }
  return i.push(s[u++], s[u++]), i;
}
function Vt(s, l) {
  var t = [], o = s.length, n, e;
  function a(v, f, h) {
    var g = v.coord, m = (h - g) / (f.coord - g), p = Et(m, [v.color, f.color]);
    return {
      coord: h,
      color: p
    };
  }
  for (var i = 0; i < o; i++) {
    var u = s[i], r = u.coord;
    if (r < 0)
      n = u;
    else if (r > l) {
      e ? t.push(a(e, u, l)) : n && t.push(a(n, u, 0), a(n, u, l));
      break;
    } else
      n && (t.push(a(n, u, 0)), n = null), t.push(u), e = u;
  }
  return t;
}
function Ft(s, l, t) {
  var o = s.getVisual("visualMeta");
  if (!(!o || !o.length || !s.count())) {
    if (l.type !== "cartesian2d") {
      process.env.NODE_ENV !== "production" && console.warn("Visual map on line style is only supported on cartesian2d.");
      return;
    }
    for (var n, e, a = o.length - 1; a >= 0; a--) {
      var i = s.getDimensionInfo(o[a].dimension);
      if (n = i && i.coordDim, n === "x" || n === "y") {
        e = o[a];
        break;
      }
    }
    if (!e) {
      process.env.NODE_ENV !== "production" && console.warn("Visual map on line style only support x or y dimension.");
      return;
    }
    var u = l.getAxis(n), r = ft(e.stops, function(c) {
      return {
        coord: u.toGlobalCoord(u.dataToCoord(c.value)),
        color: c.color
      };
    }), v = r.length, f = e.outerColors.slice();
    v && r[0].coord > r[v - 1].coord && (r.reverse(), f.reverse());
    var h = Vt(r, n === "x" ? t.getWidth() : t.getHeight()), g = h.length;
    if (!g && v)
      return r[0].coord < 0 ? f[1] ? f[1] : r[v - 1].color : f[0] ? f[0] : r[0].color;
    var m = 10, p = h[0].coord - m, b = h[g - 1].coord + m, d = b - p;
    if (d < 1e-3)
      return "transparent";
    it(h, function(c) {
      c.offset = (c.coord - p) / d;
    }), h.push({
      // NOTE: inRangeStopLen may still be 0 if stoplen is zero.
      offset: g ? h[g - 1].offset : 0.5,
      color: f[1] || "transparent"
    }), h.unshift({
      offset: g ? h[0].offset : 0.5,
      color: f[0] || "transparent"
    });
    var y = new kt(0, 0, 0, 0, h, !0);
    return y[n] = p, y[n + "2"] = b, y;
  }
}
function Ht(s, l, t) {
  var o = s.get("showAllSymbol"), n = o === "auto";
  if (!(o && !n)) {
    var e = t.getAxesByScale("ordinal")[0];
    if (e && !(n && zt(e, l))) {
      var a = l.mapDimension(e.dim), i = {};
      return it(e.getViewLabels(), function(u) {
        var r = e.scale.getRawOrdinalNumber(u.tickValue);
        i[r] = 1;
      }), function(u) {
        return !i.hasOwnProperty(l.get(a, u));
      };
    }
  }
}
function zt(s, l) {
  var t = s.getExtent(), o = Math.abs(t[1] - t[0]) / s.scale.count();
  isNaN(o) && (o = 0);
  for (var n = l.count(), e = Math.max(1, Math.round(n / 5)), a = 0; a < n; a += e)
    if (ot.getSymbolSize(
      l,
      a
      // Only for cartesian, where `isHorizontal` exists.
    )[s.isHorizontal() ? 1 : 0] * 1.5 > o)
      return !1;
  return !0;
}
function Bt(s, l) {
  return isNaN(s) || isNaN(l);
}
function Rt(s) {
  for (var l = s.length / 2; l > 0 && Bt(s[l * 2 - 2], s[l * 2 - 1]); l--)
    ;
  return l - 1;
}
function nt(s, l) {
  return [s[l * 2], s[l * 2 + 1]];
}
function Xt(s, l, t) {
  for (var o = s.length / 2, n = t === "x" ? 0 : 1, e, a, i = 0, u = -1, r = 0; r < o; r++)
    if (a = s[r * 2 + n], !(isNaN(a) || isNaN(s[r * 2 + 1 - n]))) {
      if (r === 0) {
        e = a;
        continue;
      }
      if (e <= l && a >= l || e >= l && a <= l) {
        u = r;
        break;
      }
      i = r, e = a;
    }
  return {
    range: [i, u],
    t: (l - e) / (a - e)
  };
}
function lt(s) {
  if (s.get(["endLabel", "show"]))
    return !0;
  for (var l = 0; l < Q.length; l++)
    if (s.get([Q[l], "endLabel", "show"]))
      return !0;
  return !1;
}
function H(s, l, t, o) {
  if (wt(l, "cartesian2d")) {
    var n = o.getModel("endLabel"), e = n.get("valueAnimation"), a = o.getData(), i = {
      lastFrameIndex: 0
    }, u = lt(o) ? function(g, m) {
      s._endLabelOnDuring(g, m, a, i, e, n, l);
    } : null, r = l.getBaseAxis().isHorizontal(), v = St(l, t, o, function() {
      var g = s._endLabel;
      g && t && i.originalX != null && g.attr({
        x: i.originalX,
        y: i.originalY
      });
    }, u);
    if (!o.get("clip", !0)) {
      var f = v.shape, h = Math.max(f.width, f.height);
      r ? (f.y -= h, f.height += h * 2) : (f.x -= h, f.width += h * 2);
    }
    return u && u(1, v), v;
  } else
    return process.env.NODE_ENV !== "production" && o.get(["endLabel", "show"]) && console.warn("endLabel is not supported for lines in polar systems."), xt(l, t, o);
}
function Ut(s, l) {
  var t = l.getBaseAxis(), o = t.isHorizontal(), n = t.inverse, e = o ? n ? "right" : "left" : "center", a = o ? "middle" : n ? "top" : "bottom";
  return {
    normal: {
      align: s.get("align") || e,
      verticalAlign: s.get("verticalAlign") || a
    }
  };
}
var fe = (
  /** @class */
  (function(s) {
    ut(l, s);
    function l() {
      return s !== null && s.apply(this, arguments) || this;
    }
    return l.prototype.init = function() {
      var t = new Ct(), o = new gt();
      this.group.add(o.group), this._symbolDraw = o, this._lineGroup = t, this._changePolyState = ht(this._changePolyState, this);
    }, l.prototype.render = function(t, o, n) {
      var e = t.coordinateSystem, a = this.group, i = t.getData(), u = t.getModel("lineStyle"), r = t.getModel("areaStyle"), v = i.getLayout("points") || [], f = e.type === "polar", h = this._coordSys, g = this._symbolDraw, m = this._polyline, p = this._polygon, b = this._lineGroup, d = !o.ssr && t.get("animation"), y = !r.isEmpty(), c = r.get("origin"), D = _t(e, i, c), _ = y && Tt(e, i, D), x = t.get("showSymbol"), P = t.get("connectNulls"), N = x && !f && Ht(t, i, e), S = this._data;
      S && S.eachItemGraphicEl(function(A, vt) {
        A.__temp && (a.remove(A), S.setItemGraphicEl(vt, null));
      }), x || g.remove(), a.add(b);
      var w = f ? !1 : t.get("step"), I;
      e && e.getArea && t.get("clip", !0) && (I = e.getArea(), I.width != null ? (I.x -= 0.1, I.y -= 0.1, I.width += 0.2, I.height += 0.2) : I.r0 && (I.r0 -= 0.5, I.r += 0.5)), this._clipShapeForSymbol = I;
      var L = Ft(i, e, n) || i.getVisual("style")[i.getVisual("drawType")];
      if (!(m && h.type === e.type && w === this._step))
        x && g.updateData(i, {
          isIgnore: N,
          clipShape: I,
          disableAnimation: !0,
          getSymbolPoint: function(A) {
            return [v[A * 2], v[A * 2 + 1]];
          }
        }), d && this._initSymbolLabelAnimation(i, e, I), w && (_ && (_ = E(_, v, e, w, P)), v = E(v, null, e, w, P)), m = this._newPolyline(v), y ? p = this._newPolygon(v, _) : p && (b.remove(p), p = this._polygon = null), f || this._initOrUpdateEndLabel(t, e, Y(L)), b.setClipPath(H(this, e, !0, t));
      else {
        y && !p ? p = this._newPolygon(v, _) : p && !y && (b.remove(p), p = this._polygon = null), f || this._initOrUpdateEndLabel(t, e, Y(L));
        var k = b.getClipPath();
        if (k) {
          var G = H(this, e, !1, t);
          Gt(k, {
            shape: G.shape
          }, t);
        } else
          b.setClipPath(H(this, e, !0, t));
        x && g.updateData(i, {
          isIgnore: N,
          clipShape: I,
          disableAnimation: !0,
          getSymbolPoint: function(A) {
            return [v[A * 2], v[A * 2 + 1]];
          }
        }), (!tt(this._stackedOnPoints, _) || !tt(this._points, v)) && (d ? this._doUpdateAnimation(i, _, e, n, w, c, P) : (w && (_ && (_ = E(_, v, e, w, P)), v = E(v, null, e, w, P)), m.setShape({
          points: v
        }), p && p.setShape({
          points: v,
          stackedOnPoints: _
        })));
      }
      var C = t.getModel("emphasis"), O = C.get("focus"), z = C.get("blurScope"), B = C.get("disabled");
      if (m.useStyle(Z(
        // Use color in lineStyle first
        u.getLineStyle(),
        {
          fill: "none",
          stroke: L,
          lineJoin: "bevel"
        }
      )), $(m, t, "lineStyle"), m.style.lineWidth > 0 && t.get(["emphasis", "lineStyle", "width"]) === "bolder") {
        var st = m.getState("emphasis").style;
        st.lineWidth = +m.style.lineWidth + 1;
      }
      F(m).seriesIndex = t.seriesIndex, j(m, O, z, B);
      var R = rt(t.get("smooth")), X = t.get("smoothMonotone");
      if (m.setShape({
        smooth: R,
        smoothMonotone: X,
        connectNulls: P
      }), p) {
        var U = i.getCalculationInfo("stackedOnSeries"), W = 0;
        p.useStyle(Z(r.getAreaStyle(), {
          fill: L,
          opacity: 0.7,
          lineJoin: "bevel",
          decal: i.getVisual("style").decal
        })), U && (W = rt(U.get("smooth"))), p.setShape({
          smooth: R,
          stackedOnSmooth: W,
          smoothMonotone: X,
          connectNulls: P
        }), $(p, t, "areaStyle"), F(p).seriesIndex = t.seriesIndex, j(p, O, z, B);
      }
      var J = this._changePolyState;
      i.eachItemGraphicEl(function(A) {
        A && (A.onHoverStateChange = J);
      }), this._polyline.onHoverStateChange = J, this._data = i, this._coordSys = e, this._stackedOnPoints = _, this._points = v, this._step = w, this._valueOrigin = c, t.get("triggerLineEvent") && (this.packEventData(t, m), p && this.packEventData(t, p));
    }, l.prototype.packEventData = function(t, o) {
      F(o).eventData = {
        componentType: "series",
        componentSubType: "line",
        componentIndex: t.componentIndex,
        seriesIndex: t.seriesIndex,
        seriesName: t.name,
        seriesType: "line"
      };
    }, l.prototype.highlight = function(t, o, n, e) {
      var a = t.getData(), i = q(a, e);
      if (this._changePolyState("emphasis"), !(i instanceof Array) && i != null && i >= 0) {
        var u = a.getLayout("points"), r = a.getItemGraphicEl(i);
        if (!r) {
          var v = u[i * 2], f = u[i * 2 + 1];
          if (isNaN(v) || isNaN(f) || this._clipShapeForSymbol && !this._clipShapeForSymbol.contain(v, f))
            return;
          var h = t.get("zlevel") || 0, g = t.get("z") || 0;
          r = new ot(a, i), r.x = v, r.y = f, r.setZ(h, g);
          var m = r.getSymbolPath().getTextContent();
          m && (m.zlevel = h, m.z = g, m.z2 = this._polyline.z2 + 1), r.__temp = !0, a.setItemGraphicEl(i, r), r.stopSymbolAnimation(!0), this.group.add(r);
        }
        r.highlight();
      } else
        V.prototype.highlight.call(this, t, o, n, e);
    }, l.prototype.downplay = function(t, o, n, e) {
      var a = t.getData(), i = q(a, e);
      if (this._changePolyState("normal"), i != null && i >= 0) {
        var u = a.getItemGraphicEl(i);
        u && (u.__temp ? (a.setItemGraphicEl(i, null), this.group.remove(u)) : u.downplay());
      } else
        V.prototype.downplay.call(this, t, o, n, e);
    }, l.prototype._changePolyState = function(t) {
      var o = this._polygon;
      K(this._polyline, t), o && K(o, t);
    }, l.prototype._newPolyline = function(t) {
      var o = this._polyline;
      return o && this._lineGroup.remove(o), o = new yt({
        shape: {
          points: t
        },
        segmentIgnoreThreshold: 2,
        z2: 10
      }), this._lineGroup.add(o), this._polyline = o, o;
    }, l.prototype._newPolygon = function(t, o) {
      var n = this._polygon;
      return n && this._lineGroup.remove(n), n = new dt({
        shape: {
          points: t,
          stackedOnPoints: o
        },
        segmentIgnoreThreshold: 2
      }), this._lineGroup.add(n), this._polygon = n, n;
    }, l.prototype._initSymbolLabelAnimation = function(t, o, n) {
      var e, a, i = o.getBaseAxis(), u = i.inverse;
      o.type === "cartesian2d" ? (e = i.isHorizontal(), a = !1) : o.type === "polar" && (e = i.dim === "angle", a = !0);
      var r = t.hostModel, v = r.get("animationDuration");
      T(v) && (v = v(null));
      var f = r.get("animationDelay") || 0, h = T(f) ? f(null) : f;
      t.eachItemGraphicEl(function(g, m) {
        var p = g;
        if (p) {
          var b = [g.x, g.y], d = void 0, y = void 0, c = void 0;
          if (n)
            if (a) {
              var D = n, _ = o.pointToCoord(b);
              e ? (d = D.startAngle, y = D.endAngle, c = -_[1] / 180 * Math.PI) : (d = D.r0, y = D.r, c = _[0]);
            } else {
              var x = n;
              e ? (d = x.x, y = x.x + x.width, c = g.x) : (d = x.y + x.height, y = x.y, c = g.y);
            }
          var P = y === d ? 0 : (c - d) / (y - d);
          u && (P = 1 - P);
          var N = T(f) ? f(m) : v * P + h, S = p.getSymbolPath(), w = S.getTextContent();
          p.attr({
            scaleX: 0,
            scaleY: 0
          }), p.animateTo({
            scaleX: 1,
            scaleY: 1
          }, {
            duration: 200,
            setToFinal: !0,
            delay: N
          }), w && w.animateFrom({
            style: {
              opacity: 0
            }
          }, {
            duration: 300,
            delay: N
          }), S.disableLabelAnimation = !0;
        }
      });
    }, l.prototype._initOrUpdateEndLabel = function(t, o, n) {
      var e = t.getModel("endLabel");
      if (lt(t)) {
        var a = t.getData(), i = this._polyline, u = a.getLayout("points");
        if (!u) {
          i.removeTextContent(), this._endLabel = null;
          return;
        }
        var r = this._endLabel;
        r || (r = this._endLabel = new Ot({
          z2: 200
          // should be higher than item symbol
        }), r.ignoreClip = !0, i.setTextContent(this._endLabel), i.disableLabelAnimation = !0);
        var v = Rt(u);
        v >= 0 && (It(i, Pt(t, "endLabel"), {
          inheritColor: n,
          labelFetcher: t,
          labelDataIndex: v,
          defaultText: function(f, h, g) {
            return g != null ? Lt(a, g) : Nt(a, f);
          },
          enableTextSetter: !0
        }, Ut(e, o)), i.textConfig.position = null);
      } else this._endLabel && (this._polyline.removeTextContent(), this._endLabel = null);
    }, l.prototype._endLabelOnDuring = function(t, o, n, e, a, i, u) {
      var r = this._endLabel, v = this._polyline;
      if (r) {
        t < 1 && e.originalX == null && (e.originalX = r.x, e.originalY = r.y);
        var f = n.getLayout("points"), h = n.hostModel, g = h.get("connectNulls"), m = i.get("precision"), p = i.get("distance") || 0, b = u.getBaseAxis(), d = b.isHorizontal(), y = b.inverse, c = o.shape, D = y ? d ? c.x : c.y + c.height : d ? c.x + c.width : c.y, _ = (d ? p : 0) * (y ? -1 : 1), x = (d ? 0 : -p) * (y ? -1 : 1), P = d ? "x" : "y", N = Xt(f, D, P), S = N.range, w = S[1] - S[0], I = void 0;
        if (w >= 1) {
          if (w > 1 && !g) {
            var L = nt(f, S[0]);
            r.attr({
              x: L[0] + _,
              y: L[1] + x
            }), a && (I = h.getRawValue(S[0]));
          } else {
            var L = v.getPointOn(D, P);
            L && r.attr({
              x: L[0] + _,
              y: L[1] + x
            });
            var k = h.getRawValue(S[0]), G = h.getRawValue(S[1]);
            a && (I = ct(n, m, k, G, N.t));
          }
          e.lastFrameIndex = S[0];
        } else {
          var C = t === 1 || e.lastFrameIndex > 0 ? S[0] : 0, L = nt(f, C);
          a && (I = h.getRawValue(C)), r.attr({
            x: L[0] + _,
            y: L[1] + x
          });
        }
        if (a) {
          var O = Dt(r);
          typeof O.setLabelText == "function" && O.setLabelText(I);
        }
      }
    }, l.prototype._doUpdateAnimation = function(t, o, n, e, a, i, u) {
      var r = this._polyline, v = this._polygon, f = t.hostModel, h = mt(this._data, t, this._stackedOnPoints, o, this._coordSys, n, this._valueOrigin), g = h.current, m = h.stackedOnCurrent, p = h.next, b = h.stackedOnNext;
      if (a && (m = E(h.stackedOnCurrent, h.current, n, a, u), g = E(h.current, null, n, a, u), b = E(h.stackedOnNext, h.next, n, a, u), p = E(h.next, null, n, a, u)), at(g, p) > 3e3 || v && at(m, b) > 3e3) {
        r.stopAnimation(), r.setShape({
          points: p
        }), v && (v.stopAnimation(), v.setShape({
          points: p,
          stackedOnPoints: b
        }));
        return;
      }
      r.shape.__points = h.current, r.shape.points = g;
      var d = {
        shape: {
          points: p
        }
      };
      h.current !== g && (d.shape.__points = h.next), r.stopAnimation(), M(r, d, f), v && (v.setShape({
        // Reuse the points with polyline.
        points: g,
        stackedOnPoints: m
      }), v.stopAnimation(), M(v, {
        shape: {
          stackedOnPoints: b
        }
      }, f), r.shape.points !== v.shape.points && (v.shape.points = r.shape.points));
      for (var y = [], c = h.status, D = 0; D < c.length; D++) {
        var _ = c[D].cmd;
        if (_ === "=") {
          var x = t.getItemGraphicEl(c[D].idx1);
          x && y.push({
            el: x,
            ptIdx: D
            // Index of points
          });
        }
      }
      r.animators && r.animators.length && r.animators[0].during(function() {
        v && v.dirtyShape();
        for (var P = r.shape.__points, N = 0; N < y.length; N++) {
          var S = y[N].el, w = y[N].ptIdx * 2;
          S.x = P[w], S.y = P[w + 1], S.markRedraw();
        }
      });
    }, l.prototype.remove = function(t) {
      var o = this.group, n = this._data;
      this._lineGroup.removeAll(), this._symbolDraw.remove(!0), n && n.eachItemGraphicEl(function(e, a) {
        e.__temp && (o.remove(e), n.setItemGraphicEl(a, null));
      }), this._polyline = this._polygon = this._coordSys = this._points = this._stackedOnPoints = this._endLabel = this._data = null;
    }, l.type = "line", l;
  })(V)
);
export {
  fe as default
};
