// echarts@6.0.0/lib/chart/line/LineSeries.js
import { __extends as d } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import h from "../helper/createSeriesData.js";
import c from "../../model/Series.js";
import { createSymbol as m } from "../../util/symbol.js";
import "../../util/graphic.js";
import f from "../../visual/tokens.js";
import p from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
var E = (
  /** @class */
  (function(o) {
    d(t, o);
    function t() {
      var e = o !== null && o.apply(this, arguments) || this;
      return e.type = t.type, e.hasSymbolVisual = !0, e;
    }
    return t.prototype.getInitialData = function(e) {
      if (process.env.NODE_ENV !== "production") {
        var l = e.coordinateSystem;
        if (l !== "polar" && l !== "cartesian2d")
          throw new Error("Line not support coordinateSystem besides cartesian and polar");
      }
      return h(null, this, {
        useEncodeDefaulter: !0
      });
    }, t.prototype.getLegendIcon = function(e) {
      var l = new p(), n = m("line", 0, e.itemHeight / 2, e.itemWidth, 0, e.lineStyle.stroke, !1);
      l.add(n), n.setStyle(e.lineStyle);
      var a = this.getData().getVisual("symbol"), y = this.getData().getVisual("symbolRotate"), s = a === "none" ? "circle" : a, r = e.itemHeight * 0.8, i = m(s, (e.itemWidth - r) / 2, (e.itemHeight - r) / 2, r, r, e.itemStyle.fill);
      l.add(i), i.setStyle(e.itemStyle);
      var u = e.iconRotate === "inherit" ? y : e.iconRotate || 0;
      return i.rotation = u * Math.PI / 180, i.setOrigin([e.itemWidth / 2, e.itemHeight / 2]), s.indexOf("empty") > -1 && (i.style.stroke = i.style.fill, i.style.fill = f.color.neutral00, i.style.lineWidth = 2), l;
    }, t.type = "series.line", t.dependencies = ["grid", "polar"], t.defaultOption = {
      // zlevel: 0,
      z: 3,
      coordinateSystem: "cartesian2d",
      legendHoverLink: !0,
      clip: !0,
      label: {
        position: "top"
      },
      // itemStyle: {
      // },
      endLabel: {
        show: !1,
        valueAnimation: !0,
        distance: 8
      },
      lineStyle: {
        width: 2,
        type: "solid"
      },
      emphasis: {
        scale: !0
      },
      // areaStyle: {
      // origin of areaStyle. Valid values:
      // `'auto'/null/undefined`: from axisLine to data
      // `'start'`: from min to data
      // `'end'`: from data to max
      // origin: 'auto'
      // },
      // false, 'start', 'end', 'middle'
      step: !1,
      // Disabled if step is true
      smooth: !1,
      smoothMonotone: null,
      symbol: "emptyCircle",
      symbolSize: 6,
      symbolRotate: null,
      showSymbol: !0,
      // `false`: follow the label interval strategy.
      // `true`: show all symbols.
      // `'auto'`: If possible, show all symbols, otherwise
      //           follow the label interval strategy.
      showAllSymbol: "auto",
      // Whether to connect break point.
      connectNulls: !1,
      // Sampling for large data. Can be: 'average', 'max', 'min', 'sum', 'lttb'.
      sampling: "none",
      animationEasing: "linear",
      // Disable progressive
      progressive: 0,
      hoverLayerThreshold: 1 / 0,
      universalTransition: {
        divideShape: "clone"
      },
      triggerLineEvent: !1
    }, t;
  })(c)
);
export {
  E as default
};
