// echarts@6.0.0/lib/chart/line/lineAnimationDiff.js
import { prepareDataCoordInfo as U, getStackedOnPoint as W } from "./helper.js";
import { createFloat32Array as C } from "../../util/vendor.js";
function Z(n, r) {
  var o = [];
  return r.diff(n).add(function(a) {
    o.push({
      cmd: "+",
      idx: a
    });
  }).update(function(a, i) {
    o.push({
      cmd: "=",
      idx: i,
      idx1: a
    });
  }).remove(function(a) {
    o.push({
      cmd: "-",
      idx: a
    });
  }).execute(), o;
}
function Y(n, r, o, a, i, _, M, $) {
  for (var S = Z(n, r), p = [], g = [], N = [], P = [], T = [], h = [], m = [], V = U(i, r, M), j = n.getLayout("points") || [], k = r.getLayout("points") || [], s = 0; s < S.length; s++) {
    var e = S[s], q = !0, c = void 0, t = void 0;
    switch (e.cmd) {
      case "=":
        c = e.idx * 2, t = e.idx1 * 2;
        var R = j[c], b = j[c + 1], z = k[t], B = k[t + 1];
        (isNaN(R) || isNaN(b)) && (R = z, b = B), p.push(R, b), g.push(z, B), N.push(o[c], o[c + 1]), P.push(a[t], a[t + 1]), m.push(r.getRawIndex(e.idx1));
        break;
      case "+":
        var f = e.idx, E = V.dataDimsForPoint, G = i.dataToPoint([r.get(E[0], f), r.get(E[1], f)]);
        t = f * 2, p.push(G[0], G[1]), g.push(k[t], k[t + 1]);
        var H = W(V, i, r, f);
        N.push(H[0], H[1]), P.push(a[t], a[t + 1]), m.push(r.getRawIndex(f));
        break;
      case "-":
        q = !1;
    }
    q && (T.push(e), h.push(h.length));
  }
  h.sort(function(O, Q) {
    return m[O] - m[Q];
  });
  for (var l = p.length, y = C(l), L = C(l), A = C(l), F = C(l), J = [], s = 0; s < h.length; s++) {
    var K = h[s], v = s * 2, u = K * 2;
    y[v] = p[u], y[v + 1] = p[u + 1], L[v] = g[u], L[v + 1] = g[u + 1], A[v] = N[u], A[v + 1] = N[u + 1], F[v] = P[u], F[v + 1] = P[u + 1], J[s] = T[K];
  }
  return {
    current: y,
    next: L,
    stackedOnCurrent: A,
    stackedOnNext: F,
    status: J
  };
}
export {
  Y as default
};
