// echarts@6.0.0/lib/chart/line/poly.js
import { __extends as L } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import X from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Path.js";
import _ from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/PathProxy.js";
import { cubicRootAt as N, cubicAt as d } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/curve.js";
import g from "../../visual/tokens.js";
var G = Math.min, H = Math.max;
function J(n, v) {
  return isNaN(n) || isNaN(v);
}
function K(n, v, u, f, s, o, i, t, M) {
  for (var h, w, b, Q, c, l, m = u, T = 0; T < f; T++) {
    var a = v[m * 2], r = v[m * 2 + 1];
    if (m >= s || m < 0)
      break;
    if (J(a, r)) {
      if (M) {
        m += o;
        continue;
      }
      break;
    }
    if (m === u)
      n[o > 0 ? "moveTo" : "lineTo"](a, r), b = a, Q = r;
    else {
      var z = a - h, j = r - w;
      if (z * z + j * j < 0.5) {
        m += o;
        continue;
      }
      if (i > 0) {
        for (var E = m + o, P = v[E * 2], k = v[E * 2 + 1]; P === a && k === r && T < f; )
          T++, E += o, m += o, P = v[E * 2], k = v[E * 2 + 1], a = v[m * 2], r = v[m * 2 + 1], z = a - h, j = r - w;
        var U = T + 1;
        if (M)
          for (; J(P, k) && U < f; )
            U++, E += o, P = v[E * 2], k = v[E * 2 + 1];
        var R = 0.5, B = 0, F = 0, q = void 0, A = void 0;
        if (U >= f || J(P, k))
          c = a, l = r;
        else {
          B = P - h, F = k - w;
          var V = a - h, W = P - a, Z = r - w, $ = k - r, p = void 0, D = void 0;
          if (t === "x") {
            p = Math.abs(V), D = Math.abs(W);
            var O = B > 0 ? 1 : -1;
            c = a - O * p * i, l = r, q = a + O * D * i, A = r;
          } else if (t === "y") {
            p = Math.abs(Z), D = Math.abs($);
            var C = F > 0 ? 1 : -1;
            c = a, l = r - C * p * i, q = a, A = r + C * D * i;
          } else
            p = Math.sqrt(V * V + Z * Z), D = Math.sqrt(W * W + $ * $), R = D / (D + p), c = a - B * i * (1 - R), l = r - F * i * (1 - R), q = a + B * i * R, A = r + F * i * R, q = G(q, H(P, a)), A = G(A, H(k, r)), q = H(q, G(P, a)), A = H(A, G(k, r)), B = q - a, F = A - r, c = a - B * p / D, l = r - F * p / D, c = G(c, H(h, a)), l = G(l, H(w, r)), c = H(c, G(h, a)), l = H(l, G(w, r)), B = a - c, F = r - l, q = a + B * D / p, A = r + F * D / p;
        }
        n.bezierCurveTo(b, Q, c, l, a, r), b = q, Q = A;
      } else
        n.lineTo(a, r);
    }
    h = a, w = r, m += o;
  }
  return T;
}
var I = (
  /** @class */
  /* @__PURE__ */ (function() {
    function n() {
      this.smooth = 0, this.smoothConstraint = !0;
    }
    return n;
  })()
), ra = (
  /** @class */
  (function(n) {
    L(v, n);
    function v(u) {
      var f = n.call(this, u) || this;
      return f.type = "ec-polyline", f;
    }
    return v.prototype.getDefaultStyle = function() {
      return {
        stroke: g.color.neutral99,
        fill: null
      };
    }, v.prototype.getDefaultShape = function() {
      return new I();
    }, v.prototype.buildPath = function(u, f) {
      var s = f.points, o = 0, i = s.length / 2;
      if (f.connectNulls) {
        for (; i > 0 && J(s[i * 2 - 2], s[i * 2 - 1]); i--)
          ;
        for (; o < i && J(s[o * 2], s[o * 2 + 1]); o++)
          ;
      }
      for (; o < i; )
        o += K(u, s, o, i, i, 1, f.smooth, f.smoothMonotone, f.connectNulls) + 1;
    }, v.prototype.getPointOn = function(u, f) {
      this.path || (this.createPathProxy(), this.buildPath(this.path, this.shape));
      for (var s = this.path, o = s.data, i = _.CMD, t, M, h = f === "x", w = [], b = 0; b < o.length; ) {
        var Q = o[b++], c = void 0, l = void 0, m = void 0, T = void 0, a = void 0, r = void 0, z = void 0;
        switch (Q) {
          case i.M:
            t = o[b++], M = o[b++];
            break;
          case i.L:
            if (c = o[b++], l = o[b++], z = h ? (u - t) / (c - t) : (u - M) / (l - M), z <= 1 && z >= 0) {
              var j = h ? (l - M) * z + M : (c - t) * z + t;
              return h ? [u, j] : [j, u];
            }
            t = c, M = l;
            break;
          case i.C:
            c = o[b++], l = o[b++], m = o[b++], T = o[b++], a = o[b++], r = o[b++];
            var E = h ? N(t, c, m, a, u, w) : N(M, l, T, r, u, w);
            if (E > 0)
              for (var P = 0; P < E; P++) {
                var k = w[P];
                if (k <= 1 && k >= 0) {
                  var j = h ? d(M, l, T, r, k) : d(t, c, m, a, k);
                  return h ? [u, j] : [j, u];
                }
              }
            t = a, M = r;
            break;
        }
      }
    }, v;
  })(X)
), S = (
  /** @class */
  (function(n) {
    L(v, n);
    function v() {
      return n !== null && n.apply(this, arguments) || this;
    }
    return v;
  })(I)
), oa = (
  /** @class */
  (function(n) {
    L(v, n);
    function v(u) {
      var f = n.call(this, u) || this;
      return f.type = "ec-polygon", f;
    }
    return v.prototype.getDefaultShape = function() {
      return new S();
    }, v.prototype.buildPath = function(u, f) {
      var s = f.points, o = f.stackedOnPoints, i = 0, t = s.length / 2, M = f.smoothMonotone;
      if (f.connectNulls) {
        for (; t > 0 && J(s[t * 2 - 2], s[t * 2 - 1]); t--)
          ;
        for (; i < t && J(s[i * 2], s[i * 2 + 1]); i++)
          ;
      }
      for (; i < t; ) {
        var h = K(u, s, i, t, t, 1, f.smooth, M, f.connectNulls);
        K(u, o, i + h - 1, h, t, -1, f.stackedOnSmooth, M, f.connectNulls), i += h + 1, u.closePath();
      }
    }, v;
  })(X)
);
export {
  oa as ECPolygon,
  ra as ECPolyline
};
