// @radix-ui/react-slot@1.1.1/dist/index.js
import * as o from "react";
import { composeRefs as u } from "../../../../../@radix-ui_react-compose-refs@1.1.1_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-compose-refs/dist/index.js";
import { jsx as c, Fragment as d } from "react/jsx-runtime";
var m = o.forwardRef((e, n) => {
  const { children: r, ...t } = e, l = o.Children.toArray(r), i = l.find(y);
  if (i) {
    const s = i.props.children, a = l.map((p) => p === i ? o.Children.count(s) > 1 ? o.Children.only(null) : o.isValidElement(s) ? s.props.children : null : p);
    return /* @__PURE__ */ c(f, { ...t, ref: n, children: o.isValidElement(s) ? o.cloneElement(s, void 0, a) : null });
  }
  return /* @__PURE__ */ c(f, { ...t, ref: n, children: r });
});
m.displayName = "Slot";
var f = o.forwardRef((e, n) => {
  const { children: r, ...t } = e;
  if (o.isValidElement(r)) {
    const l = C(r);
    return o.cloneElement(r, {
      ...h(t, r.props),
      // @ts-ignore
      ref: n ? u(n, l) : l
    });
  }
  return o.Children.count(r) > 1 ? o.Children.only(null) : null;
});
f.displayName = "SlotClone";
var g = ({ children: e }) => /* @__PURE__ */ c(d, { children: e });
function y(e) {
  return o.isValidElement(e) && e.type === g;
}
function h(e, n) {
  const r = { ...n };
  for (const t in n) {
    const l = e[t], i = n[t];
    /^on[A-Z]/.test(t) ? l && i ? r[t] = (...a) => {
      i(...a), l(...a);
    } : l && (r[t] = l) : t === "style" ? r[t] = { ...l, ...i } : t === "className" && (r[t] = [l, i].filter(Boolean).join(" "));
  }
  return { ...e, ...r };
}
function C(e) {
  let n = Object.getOwnPropertyDescriptor(e.props, "ref")?.get, r = n && "isReactWarning" in n && n.isReactWarning;
  return r ? e.ref : (n = Object.getOwnPropertyDescriptor(e, "ref")?.get, r = n && "isReactWarning" in n && n.isReactWarning, r ? e.props.ref : e.props.ref || e.ref);
}
export {
  m as Slot,
  g as Slottable
};
