// echarts@6.0.0/lib/chart/treemap/Breadcrumb.js
import { getECData as L } from "../../util/innerStore.js";
import { createBoxLayoutReference as w, getLayoutRect as _, positionElement as A } from "../../util/layout.js";
import { wrapTreePathInfo as C } from "../helper/treeHelper.js";
import { curry as E, defaults as D } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { convertOptionIdName as R } from "../../util/model.js";
import { Z2_EMPHASIS_LIFT as G, toggleHoverEmphasis as M } from "../../util/states.js";
import { createTextStyle as T } from "../../label/labelStyle.js";
import P from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import k from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Polygon.js";
import B from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Text.js";
var W = 8, x = 8, S = 5, U = (
  /** @class */
  (function() {
    function o(t) {
      this.group = new P(), t.add(this.group);
    }
    return o.prototype.render = function(t, e, n, a) {
      var r = t.getModel("breadcrumb"), p = this.group;
      if (p.removeAll(), !(!r.get("show") || !n)) {
        var m = r.getModel("itemStyle"), u = r.getModel("emphasis"), s = m.getModel("textStyle"), f = u.getModel(["itemStyle", "textStyle"]), d = w(t, e).refContainer, v = {
          left: r.get("left"),
          right: r.get("right"),
          top: r.get("top"),
          bottom: r.get("bottom")
        }, i = {
          emptyItemWidth: r.get("emptyItemWidth"),
          totalWidth: 0,
          renderList: []
        }, h = _(v, d);
        this._prepare(n, i, s), this._renderContent(t, i, h, m, u, s, f, a), A(p, v, d);
      }
    }, o.prototype._prepare = function(t, e, n) {
      for (var a = t; a; a = a.parentNode) {
        var r = R(a.getModel().get("name"), ""), p = n.getTextRect(r), m = Math.max(p.width + W * 2, e.emptyItemWidth);
        e.totalWidth += m + x, e.renderList.push({
          node: a,
          text: r,
          width: m
        });
      }
    }, o.prototype._renderContent = function(t, e, n, a, r, p, m, u) {
      for (var s = 0, f = e.emptyItemWidth, d = t.get(["breadcrumb", "height"]), v = e.totalWidth, i = e.renderList, h = r.getModel("itemStyle").getItemStyle(), g = i.length - 1; g >= 0; g--) {
        var y = i[g], b = y.node, c = y.width, I = y.text;
        v > n.width && (v -= c - f, c = f, I = null);
        var l = new k({
          shape: {
            points: z(s, 0, c, d, g === i.length - 1, g === 0)
          },
          style: D(a.getItemStyle(), {
            lineJoin: "bevel"
          }),
          textContent: new B({
            style: T(p, {
              text: I
            })
          }),
          textConfig: {
            position: "inside"
          },
          z2: G * 1e4,
          onclick: E(u, b)
        });
        l.disableLabelAnimation = !0, l.getTextContent().ensureState("emphasis").style = T(m, {
          text: I
        }), l.ensureState("emphasis").style = h, M(l, r.get("focus"), r.get("blurScope"), r.get("disabled")), this.group.add(l), X(l, t, b), s += c + x;
      }
    }, o.prototype.remove = function() {
      this.group.removeAll();
    }, o;
  })()
);
function z(o, t, e, n, a, r) {
  var p = [[a ? o : o - S, t], [o + e, t], [o + e, t + n], [a ? o : o - S, t + n]];
  return !r && p.splice(2, 0, [o + e + S, t + n / 2]), !a && p.push([o, t + n / 2]), p;
}
function X(o, t, e) {
  L(o).eventData = {
    componentType: "series",
    componentSubType: "treemap",
    componentIndex: t.componentIndex,
    seriesIndex: t.seriesIndex,
    seriesName: t.name,
    seriesType: "treemap",
    selfType: "breadcrumb",
    nodeData: {
      dataIndex: e && e.dataIndex,
      name: e && e.name
    },
    treePathInfo: e && C(e, t)
  };
}
export {
  U as default
};
