// echarts@6.0.0/lib/chart/treemap/treemapVisual.js
import T from "../../visual/VisualMapping.js";
import { each as V, extend as x, isArray as E } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { modifyHSL as I, modifyAlpha as L } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/tool/color.js";
import { makeInner as A } from "../../util/model.js";
var B = "itemStyle", S = A();
const U = {
  seriesType: "treemap",
  reset: function(t) {
    var r = t.getData().tree, a = r.root;
    a.isRemoved() || h(
      a,
      // Visual should calculate from tree root but not view root.
      {},
      t.getViewRoot().getAncestors(),
      t
    );
  }
};
function h(t, r, a, n) {
  var l = t.getModel(), e = t.getLayout(), o = t.hostTree.data;
  if (!(!e || e.invisible || !e.isInView)) {
    var i = l.getModel(B), u = D(i, r, n), v = o.ensureUniqueItemVisual(t.dataIndex, "style"), p = i.get("borderColor"), f = i.get("borderColorSaturation"), g;
    f != null && (g = y(u), p = k(f, g)), v.stroke = p;
    var s = t.viewChildren;
    if (!s || !s.length)
      g = y(u), v.fill = g;
    else {
      var C = w(t, l, e, i, u, s);
      V(s, function(c, M) {
        if (c.depth >= a.length || c === a[c.depth]) {
          var b = R(l, u, c, M, C, n);
          h(c, b, a, n);
        }
      });
    }
  }
}
function D(t, r, a) {
  var n = x({}, r), l = a.designatedVisualItemStyle;
  return V(["color", "colorAlpha", "colorSaturation"], function(e) {
    l[e] = r[e];
    var o = t.get(e);
    l[e] = null, o != null && (n[e] = o);
  }), n;
}
function y(t) {
  var r = d(t, "color");
  if (r) {
    var a = d(t, "colorAlpha"), n = d(t, "colorSaturation");
    return n && (r = I(r, null, null, n)), a && (r = L(r, a)), r;
  }
}
function k(t, r) {
  return r != null ? I(r, null, null, t) : null;
}
function d(t, r) {
  var a = t[r];
  if (a != null && a !== "none")
    return a;
}
function w(t, r, a, n, l, e) {
  if (!(!e || !e.length)) {
    var o = m(r, "color") || l.color != null && l.color !== "none" && (m(r, "colorAlpha") || m(r, "colorSaturation"));
    if (o) {
      var i = r.get("visualMin"), u = r.get("visualMax"), v = a.dataExtent.slice();
      i != null && i < v[0] && (v[0] = i), u != null && u > v[1] && (v[1] = u);
      var p = r.get("colorMappingBy"), f = {
        type: o.name,
        dataExtent: v,
        visual: o.range
      };
      f.type === "color" && (p === "index" || p === "id") ? (f.mappingMethod = "category", f.loop = !0) : f.mappingMethod = "linear";
      var g = new T(f);
      return S(g).drColorMappingBy = p, g;
    }
  }
}
function m(t, r) {
  var a = t.get(r);
  return E(a) && a.length ? {
    name: r,
    range: a
  } : null;
}
function R(t, r, a, n, l, e) {
  var o = x({}, r);
  if (l) {
    var i = l.type, u = i === "color" && S(l).drColorMappingBy, v = u === "index" ? n : u === "id" ? e.mapIdToIndex(a.getId()) : a.getValue(t.get("visualDimension"));
    o[i] = l.mapValueToVisual(v);
  }
  return o;
}
export {
  U as default
};
