// echarts@6.0.0/lib/chart/treemap/treemapLayout.js
import { retrieve as C, each as U, extend as W, filter as B } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import b from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/BoundingRect.js";
import { parsePercent as A, MAX_SAFE_INTEGER as N } from "../../util/number.js";
import { createBoxLayoutReference as S, getLayoutRect as z } from "../../util/layout.js";
import { retrieveTargetInfo as Z, getPathToRoot as G } from "../helper/treeHelper.js";
var I = Math.max, H = Math.min, D = C, P = U, d = ["itemStyle", "borderWidth"], X = ["itemStyle", "gapWidth"], $ = ["upperLabel", "show"], j = ["upperLabel", "height"];
const nt = {
  seriesType: "treemap",
  reset: function(a, n, r, v) {
    var i = a.option, u = S(a, r).refContainer, t = z(a.getBoxLayoutParams(), u), e = i.size || [], o = A(D(t.width, e[0]), u.width), f = A(D(t.height, e[1]), u.height), g = v && v.type, x = ["treemapZoomToNode", "treemapRootToNode"], y = Z(v, x, a), L = g === "treemapRender" || g === "treemapMove" ? v.rootRect : null, h = a.getViewRoot(), p = G(h);
    if (g !== "treemapMove") {
      var s = g === "treemapZoomToNode" ? F(a, y, h, o, f) : L ? [L.width, L.height] : [o, f], m = i.sort;
      m && m !== "asc" && m !== "desc" && (m = "desc");
      var w = {
        squareRatio: i.squareRatio,
        sort: m,
        leafDepth: i.leafDepth
      };
      h.hostTree.clearLayouts();
      var l = {
        x: 0,
        y: 0,
        width: s[0],
        height: s[1],
        area: s[0] * s[1]
      };
      h.setLayout(l), O(h, w, !1, 0), l = h.getLayout(), P(p, function(R, T) {
        var V = (p[T + 1] || h).getValue();
        R.setLayout(W({
          dataExtent: [V, V],
          borderWidth: 0,
          upperHeight: 0
        }, l));
      });
    }
    var c = a.getData().tree.root;
    c.setLayout(tt(t, L, y), !0), a.setLayoutInfo(t), _(
      c,
      // Transform to base element coordinate system.
      new b(-t.x, -t.y, r.getWidth(), r.getHeight()),
      p,
      h,
      0
    );
  }
};
function O(a, n, r, v) {
  var i, u;
  if (!a.isRemoved()) {
    var t = a.getLayout();
    i = t.width, u = t.height;
    var e = a.getModel(), o = e.get(d), f = e.get(X) / 2, g = q(e), x = Math.max(o, g), y = o - f, L = x - f;
    a.setLayout({
      borderWidth: o,
      upperHeight: x,
      upperLabelHeight: g
    }, !0), i = I(i - 2 * y, 0), u = I(u - y - L, 0);
    var h = i * u, p = k(a, e, h, n, r, v);
    if (p.length) {
      var s = {
        x: y,
        y: L,
        width: i,
        height: u
      }, m = H(i, u), w = 1 / 0, l = [];
      l.area = 0;
      for (var c = 0, R = p.length; c < R; ) {
        var T = p[c];
        l.push(T), l.area += T.getLayout().area;
        var V = Y(l, m, n.squareRatio);
        V <= w ? (c++, w = V) : (l.area -= l.pop().getLayout().area, M(l, m, s, f, !1), m = H(s.width, s.height), l.length = l.area = 0, w = 1 / 0);
      }
      if (l.length && M(l, m, s, f, !0), !r) {
        var E = e.get("childrenVisibleMin");
        E != null && h < E && (r = !0);
      }
      for (var c = 0, R = p.length; c < R; c++)
        O(p[c], n, r, v + 1);
    }
  }
}
function k(a, n, r, v, i, u) {
  var t = a.children || [], e = v.sort;
  e !== "asc" && e !== "desc" && (e = null);
  var o = v.leafDepth != null && v.leafDepth <= u;
  if (i && !o)
    return a.viewChildren = [];
  t = B(t, function(L) {
    return !L.isRemoved();
  }), K(t, e);
  var f = Q(n, t, e);
  if (f.sum === 0)
    return a.viewChildren = [];
  if (f.sum = J(n, r, f.sum, e, t), f.sum === 0)
    return a.viewChildren = [];
  for (var g = 0, x = t.length; g < x; g++) {
    var y = t[g].getValue() / f.sum * r;
    t[g].setLayout({
      area: y
    });
  }
  return o && (t.length && a.setLayout({
    isLeafRoot: !0
  }, !0), t.length = 0), a.viewChildren = t, a.setLayout({
    dataExtent: f.dataExtent
  }, !0), t;
}
function J(a, n, r, v, i) {
  if (!v)
    return r;
  for (var u = a.get("visibleMin"), t = i.length, e = t, o = t - 1; o >= 0; o--) {
    var f = i[v === "asc" ? t - o - 1 : o].getValue();
    f / r * n < u && (e = o, r -= f);
  }
  return v === "asc" ? i.splice(0, t - e) : i.splice(e, t - e), r;
}
function K(a, n) {
  return n && a.sort(function(r, v) {
    var i = n === "asc" ? r.getValue() - v.getValue() : v.getValue() - r.getValue();
    return i === 0 ? n === "asc" ? r.dataIndex - v.dataIndex : v.dataIndex - r.dataIndex : i;
  }), a;
}
function Q(a, n, r) {
  for (var v = 0, i = 0, u = n.length; i < u; i++)
    v += n[i].getValue();
  var t = a.get("visualDimension"), e;
  return !n || !n.length ? e = [NaN, NaN] : t === "value" && r ? (e = [n[n.length - 1].getValue(), n[0].getValue()], r === "asc" && e.reverse()) : (e = [1 / 0, -1 / 0], P(n, function(o) {
    var f = o.getValue(t);
    f < e[0] && (e[0] = f), f > e[1] && (e[1] = f);
  })), {
    sum: v,
    dataExtent: e
  };
}
function Y(a, n, r) {
  for (var v = 0, i = 1 / 0, u = 0, t = void 0, e = a.length; u < e; u++)
    t = a[u].getLayout().area, t && (t < i && (i = t), t > v && (v = t));
  var o = a.area * a.area, f = n * n * r;
  return o ? I(f * v / o, o / (f * i)) : 1 / 0;
}
function M(a, n, r, v, i) {
  var u = n === r.width ? 0 : 1, t = 1 - u, e = ["x", "y"], o = ["width", "height"], f = r[e[u]], g = n ? a.area / n : 0;
  (i || g > r[o[t]]) && (g = r[o[t]]);
  for (var x = 0, y = a.length; x < y; x++) {
    var L = a[x], h = {}, p = g ? L.getLayout().area / g : 0, s = h[o[t]] = I(g - 2 * v, 0), m = r[e[u]] + r[o[u]] - f, w = x === y - 1 || m < p ? m : p, l = h[o[u]] = I(w - 2 * v, 0);
    h[e[t]] = r[e[t]] + H(v, s / 2), h[e[u]] = f + H(v, l / 2), f += w, L.setLayout(h, !0);
  }
  r[e[t]] += g, r[o[t]] -= g;
}
function F(a, n, r, v, i) {
  var u = (n || {}).node, t = [v, i];
  if (!u || u === r)
    return t;
  for (var e, o = v * i, f = o * a.option.zoomToNodeRatio; e = u.parentNode; ) {
    for (var g = 0, x = e.children, y = 0, L = x.length; y < L; y++)
      g += x[y].getValue();
    var h = u.getValue();
    if (h === 0)
      return t;
    f *= g / h;
    var p = e.getModel(), s = p.get(d), m = Math.max(s, q(p));
    f += 4 * s * s + (3 * s + m) * Math.pow(f, 0.5), f > N && (f = N), u = e;
  }
  f < o && (f = o);
  var w = Math.pow(f / o, 0.5);
  return [v * w, i * w];
}
function tt(a, n, r) {
  if (n)
    return {
      x: n.x,
      y: n.y
    };
  var v = {
    x: 0,
    y: 0
  };
  if (!r)
    return v;
  var i = r.node, u = i.getLayout();
  if (!u)
    return v;
  for (var t = [u.width / 2, u.height / 2], e = i; e; ) {
    var o = e.getLayout();
    t[0] += o.x, t[1] += o.y, e = e.parentNode;
  }
  return {
    x: a.width / 2 - t[0],
    y: a.height / 2 - t[1]
  };
}
function _(a, n, r, v, i) {
  var u = a.getLayout(), t = r[i], e = t && t === a;
  if (!(t && !e || i === r.length && a !== v)) {
    a.setLayout({
      // isInView means: viewRoot sub tree + viewAbovePath
      isInView: !0,
      // invisible only means: outside view clip so that the node can not
      // see but still layout for animation preparation but not render.
      invisible: !e && !n.intersect(u),
      isAboveViewRoot: e
    }, !0);
    var o = new b(n.x - u.x, n.y - u.y, n.width, n.height);
    P(a.viewChildren || [], function(f) {
      _(f, o, r, v, i + 1);
    });
  }
}
function q(a) {
  return a.get($) ? a.get(j) : 0;
}
export {
  nt as default
};
