// echarts@6.0.0/lib/chart/treemap/TreemapView.js
import { __extends as Tt } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { indexOf as bt, each as E, curry as Dt, extend as _t, bind as q, isFunction as lt, normalizeCssArray as At } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { getECData as dt } from "../../util/innerStore.js";
import { isHighDownDispatcher as ht, setAsHighDownDispatcher as X, enableHoverFocus as ft, setDefaultStateProxy as ut, Z2_EMPHASIS_LIFT as j } from "../../util/states.js";
import Ct from "../../data/DataDiffer.js";
import { retrieveTargetInfo as Gt, aboveViewRoot as Mt } from "../helper/treeHelper.js";
import Rt from "./Breadcrumb.js";
import Ht from "../../component/helper/RoamController.js";
import Et from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/BoundingRect.js";
import { translate as vt, create as kt, scale as Nt } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/matrix.js";
import { createWrap as Wt } from "../../util/animation.js";
import zt from "../../model/mixin/makeStyleMapper.js";
import Ot from "../../view/Chart.js";
import Ft from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Displayable.js";
import { makeInner as Pt, convertOptionIdName as Bt } from "../../util/model.js";
import { windowOpen as Vt } from "../../util/format.js";
import { setLabelStyle as Zt, getLabelStatesModels as Xt } from "../../label/labelStyle.js";
import J from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import Lt from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Rect.js";
var K = J, pt = Lt, ct = 3, gt = "label", mt = "upperLabel", Yt = j * 10, Ut = j * 2, $t = j * 3, P = zt([
  ["fill", "color"],
  // `borderColor` and `borderWidth` has been occupied,
  // so use `stroke` to indicate the stroke of the rect.
  ["stroke", "strokeColor"],
  ["lineWidth", "strokeWidth"],
  ["shadowBlur"],
  ["shadowOffsetX"],
  ["shadowOffsetY"],
  ["shadowColor"]
  // Option decal is in `DecalObject` but style.decal is in `PatternObject`.
  // So do not transfer decal directly.
]), yt = function(A) {
  var d = P(A);
  return d.stroke = d.fill = d.lineWidth = null, d;
}, L = Pt(), ge = (
  /** @class */
  (function(A) {
    Tt(d, A);
    function d() {
      var t = A !== null && A.apply(this, arguments) || this;
      return t.type = d.type, t._state = "ready", t._storage = Z(), t;
    }
    return d.prototype.render = function(t, r, i, n) {
      var e = r.findComponents({
        mainType: "series",
        subType: "treemap",
        query: n
      });
      if (!(bt(e, t) < 0)) {
        this.seriesModel = t, this.api = i, this.ecModel = r;
        var u = ["treemapZoomToNode", "treemapRootToNode"], l = Gt(n, u, t), s = n && n.type, o = t.layoutInfo, S = !this._oldTree, m = this._storage, f = s === "treemapRootToNode" && l && m ? {
          rootNodeGroup: m.nodeGroup[l.node.getRawIndex()],
          direction: n.direction
        } : null, T = this._giveContainerGroup(o), v = t.get("animation"), h = this._doRender(T, t, f);
        v && !S && (!s || s === "treemapZoomToNode" || s === "treemapRootToNode") ? this._doAnimation(T, h, t, f) : h.renderFinally(), this._resetController(i), this._renderBreadcrumb(t, i, l);
      }
    }, d.prototype._giveContainerGroup = function(t) {
      var r = this._containerGroup;
      return r || (r = this._containerGroup = new K(), this._initEvents(r), this.group.add(r)), r.x = t.x, r.y = t.y, r;
    }, d.prototype._doRender = function(t, r, i) {
      var n = r.getData().tree, e = this._oldTree, u = Z(), l = Z(), s = this._storage, o = [];
      function S(g, x, F, k) {
        return qt(r, l, s, i, u, o, g, x, F, k);
      }
      v(n.root ? [n.root] : [], e && e.root ? [e.root] : [], t, n === e || !e, 0);
      var m = h(s);
      if (this._oldTree = n, this._storage = l, this._controllerHost) {
        var f = this.seriesModel.layoutInfo, T = n.root.getLayout();
        T.width === f.width && T.height === f.height && (this._controllerHost.zoom = 1);
      }
      return {
        lastsForAnimation: u,
        willDeleteEls: m,
        renderFinally: C
      };
      function v(g, x, F, k, W) {
        k ? (x = g, E(g, function(H, O) {
          !H.isRemoved() && z(O, O);
        })) : new Ct(x, g, R, R).add(z).update(z).remove(Dt(z, null)).execute();
        function R(H) {
          return H.getId();
        }
        function z(H, O) {
          var y = H != null ? g[H] : null, b = O != null ? x[O] : null, B = S(y, b, F, W);
          B && v(y && y.viewChildren || [], b && b.viewChildren || [], B, k, W + 1);
        }
      }
      function h(g) {
        var x = Z();
        return g && E(g, function(F, k) {
          var W = x[k];
          E(F, function(R) {
            R && (W.push(R), L(R).willDelete = !0);
          });
        }), x;
      }
      function C() {
        E(m, function(g) {
          E(g, function(x) {
            x.parent && x.parent.remove(x);
          });
        }), E(o, function(g) {
          g.invisible = !0, g.dirty();
        });
      }
    }, d.prototype._doAnimation = function(t, r, i, n) {
      var e = i.get("animationDurationUpdate"), u = i.get("animationEasing"), l = (lt(e) ? 0 : e) || 0, s = (lt(u) ? null : u) || "cubicOut", o = Wt();
      E(r.willDeleteEls, function(S, m) {
        E(S, function(f, T) {
          if (!f.invisible) {
            var v = f.parent, h, C = L(v);
            if (n && n.direction === "drillDown")
              h = v === n.rootNodeGroup ? {
                shape: {
                  x: 0,
                  y: 0,
                  width: C.nodeWidth,
                  height: C.nodeHeight
                },
                style: {
                  opacity: 0
                }
              } : {
                style: {
                  opacity: 0
                }
              };
            else {
              var g = 0, x = 0;
              C.willDelete || (g = C.nodeWidth / 2, x = C.nodeHeight / 2), h = m === "nodeGroup" ? {
                x: g,
                y: x,
                style: {
                  opacity: 0
                }
              } : {
                shape: {
                  x: g,
                  y: x,
                  width: 0,
                  height: 0
                },
                style: {
                  opacity: 0
                }
              };
            }
            h && o.add(f, h, l, 0, s);
          }
        });
      }), E(this._storage, function(S, m) {
        E(S, function(f, T) {
          var v = r.lastsForAnimation[m][T], h = {};
          v && (f instanceof J ? v.oldX != null && (h.x = f.x, h.y = f.y, f.x = v.oldX, f.y = v.oldY) : (v.oldShape && (h.shape = _t({}, f.shape), f.setShape(v.oldShape)), v.fadein ? (f.setStyle("opacity", 0), h.style = {
            opacity: 1
          }) : f.style.opacity !== 1 && (h.style = {
            opacity: 1
          })), o.add(f, h, l, 0, s));
        });
      }, this), this._state = "animating", o.finished(q(function() {
        this._state = "ready", r.renderFinally();
      }, this)).start();
    }, d.prototype._resetController = function(t) {
      var r = this, i = this._controller, n = this._controllerHost;
      n || (this._controllerHost = {
        target: this.group
      }, n = this._controllerHost);
      var e = this.seriesModel;
      i || (i = this._controller = new Ht(t.getZr()), i.on("pan", q(this._onPan, this)), i.on("zoom", q(this._onZoom, this))), i.enable(e.get("roam"), {
        api: t,
        zInfo: {
          component: e
        },
        triggerInfo: {
          roamTrigger: e.get("roamTrigger"),
          isInSelf: function(u, l, s) {
            var o = r._containerGroup;
            return o ? o.getBoundingRect().contain(l - o.x, s - o.y) : !1;
          }
        }
      }), n.zoomLimit = e.get("scaleLimit"), n.zoom = e.get("zoom");
    }, d.prototype._clearController = function() {
      var t = this._controller;
      this._controllerHost = null, t && (t.dispose(), t = null);
    }, d.prototype._onPan = function(t) {
      if (this._state !== "animating" && (Math.abs(t.dx) > ct || Math.abs(t.dy) > ct)) {
        var r = this.seriesModel.getData().tree.root;
        if (!r)
          return;
        var i = r.getLayout();
        if (!i)
          return;
        this.api.dispatchAction({
          type: "treemapMove",
          from: this.uid,
          seriesId: this.seriesModel.id,
          rootRect: {
            x: i.x + t.dx,
            y: i.y + t.dy,
            width: i.width,
            height: i.height
          }
        });
      }
    }, d.prototype._onZoom = function(t) {
      var r = t.originX, i = t.originY, n = t.scale;
      if (this._state !== "animating") {
        var e = this.seriesModel.getData().tree.root;
        if (!e)
          return;
        var u = e.getLayout();
        if (!u)
          return;
        var l = new Et(u.x, u.y, u.width, u.height), s = null, o = this._controllerHost;
        s = o.zoomLimit;
        var S = o.zoom = o.zoom || 1;
        if (S *= n, s) {
          var m = s.min || 0, f = s.max || 1 / 0;
          S = Math.max(Math.min(f, S), m);
        }
        var T = S / o.zoom;
        o.zoom = S;
        var v = this.seriesModel.layoutInfo;
        r -= v.x, i -= v.y;
        var h = kt();
        vt(h, h, [-r, -i]), Nt(h, h, [T, T]), vt(h, h, [r, i]), l.applyTransform(h), this.api.dispatchAction({
          type: "treemapRender",
          from: this.uid,
          seriesId: this.seriesModel.id,
          rootRect: {
            x: l.x,
            y: l.y,
            width: l.width,
            height: l.height
          }
        });
      }
    }, d.prototype._initEvents = function(t) {
      var r = this;
      t.on("click", function(i) {
        if (r._state === "ready") {
          var n = r.seriesModel.get("nodeClick", !0);
          if (n) {
            var e = r.findTarget(i.offsetX, i.offsetY);
            if (e) {
              var u = e.node;
              if (u.getLayout().isLeafRoot)
                r._rootToNode(e);
              else if (n === "zoomToNode")
                r._zoomToNode(e);
              else if (n === "link") {
                var l = u.hostTree.data.getItemModel(u.dataIndex), s = l.get("link", !0), o = l.get("target", !0) || "blank";
                s && Vt(s, o);
              }
            }
          }
        }
      }, this);
    }, d.prototype._renderBreadcrumb = function(t, r, i) {
      var n = this;
      i || (i = t.get("leafDepth", !0) != null ? {
        node: t.getViewRoot()
      } : this.findTarget(r.getWidth() / 2, r.getHeight() / 2), i || (i = {
        node: t.getData().tree.root
      })), (this._breadcrumb || (this._breadcrumb = new Rt(this.group))).render(t, r, i.node, function(e) {
        n._state !== "animating" && (Mt(t.getViewRoot(), e) ? n._rootToNode({
          node: e
        }) : n._zoomToNode({
          node: e
        }));
      });
    }, d.prototype.remove = function() {
      this._clearController(), this._containerGroup && this._containerGroup.removeAll(), this._storage = Z(), this._state = "ready", this._breadcrumb && this._breadcrumb.remove();
    }, d.prototype.dispose = function() {
      this._clearController();
    }, d.prototype._zoomToNode = function(t) {
      this.api.dispatchAction({
        type: "treemapZoomToNode",
        from: this.uid,
        seriesId: this.seriesModel.id,
        targetNode: t.node
      });
    }, d.prototype._rootToNode = function(t) {
      this.api.dispatchAction({
        type: "treemapRootToNode",
        from: this.uid,
        seriesId: this.seriesModel.id,
        targetNode: t.node
      });
    }, d.prototype.findTarget = function(t, r) {
      var i, n = this.seriesModel.getViewRoot();
      return n.eachNode({
        attr: "viewChildren",
        order: "preorder"
      }, function(e) {
        var u = this._storage.background[e.getRawIndex()];
        if (u) {
          var l = u.transformCoordToLocal(t, r), s = u.shape;
          if (s.x <= l[0] && l[0] <= s.x + s.width && s.y <= l[1] && l[1] <= s.y + s.height)
            i = {
              node: e,
              offsetX: l[0],
              offsetY: l[1]
            };
          else
            return !1;
        }
      }, this), i;
    }, d.type = "treemap", d;
  })(Ot)
);
function Z() {
  return {
    nodeGroup: [],
    background: [],
    content: []
  };
}
function qt(A, d, t, r, i, n, e, u, l, s) {
  if (!e)
    return;
  var o = e.getLayout(), S = A.getData(), m = e.getModel();
  if (S.setItemGraphicEl(e.dataIndex, null), !o || !o.isInView)
    return;
  var f = o.width, T = o.height, v = o.borderWidth, h = o.invisible, C = e.getRawIndex(), g = u && u.getRawIndex(), x = e.viewChildren, F = o.upperHeight, k = x && x.length, W = m.getModel("itemStyle"), R = m.getModel(["emphasis", "itemStyle"]), z = m.getModel(["blur", "itemStyle"]), H = m.getModel(["select", "itemStyle"]), O = W.get("borderRadius") || 0, y = $("nodeGroup", K);
  if (!y)
    return;
  if (l.add(y), y.x = o.x || 0, y.y = o.y || 0, y.markRedraw(), L(y).nodeWidth = f, L(y).nodeHeight = T, o.isAboveViewRoot)
    return y;
  var b = $("background", pt, s, Ut);
  b && xt(y, b, k && o.upperLabelHeight);
  var B = m.getModel("emphasis"), Y = B.get("focus"), Q = B.get("blurScope"), tt = B.get("disabled"), et = Y === "ancestor" ? e.getAncestorsIndices() : Y === "descendant" ? e.getDescendantIndices() : Y;
  if (k)
    ht(y) && X(y, !1), b && (X(b, !tt), S.setItemGraphicEl(e.dataIndex, b), ft(b, et, Q));
  else {
    var U = $("content", pt, s, $t);
    U && wt(y, U), b.disableMorphing = !0, b && ht(b) && X(b, !1), X(y, !tt), S.setItemGraphicEl(e.dataIndex, y);
    var rt = m.getShallow("cursor");
    rt && U.attr("cursor", rt), ft(y, et, Q);
  }
  return y;
  function xt(p, a, w) {
    var c = dt(a);
    if (c.dataIndex = e.dataIndex, c.seriesIndex = A.seriesIndex, a.setShape({
      x: 0,
      y: 0,
      width: f,
      height: T,
      r: O
    }), h)
      it(a);
    else {
      a.invisible = !1;
      var _ = e.getVisual("style"), D = _.stroke, G = yt(W);
      G.fill = D;
      var I = P(R);
      I.fill = R.get("borderColor");
      var M = P(z);
      M.fill = z.get("borderColor");
      var N = P(H);
      if (N.fill = H.get("borderColor"), w) {
        var V = f - 2 * v;
        at(
          // PENDING: convert ZRColor to ColorString for text.
          a,
          D,
          _.opacity,
          {
            x: v,
            y: 0,
            width: V,
            height: F
          }
        );
      } else
        a.removeTextContent();
      a.setStyle(G), a.ensureState("emphasis").style = I, a.ensureState("blur").style = M, a.ensureState("select").style = N, ut(a);
    }
    p.add(a);
  }
  function wt(p, a) {
    var w = dt(a);
    w.dataIndex = e.dataIndex, w.seriesIndex = A.seriesIndex;
    var c = Math.max(f - 2 * v, 0), _ = Math.max(T - 2 * v, 0);
    if (a.culling = !0, a.setShape({
      x: v,
      y: v,
      width: c,
      height: _,
      r: O
    }), h)
      it(a);
    else {
      a.invisible = !1;
      var D = e.getVisual("style"), G = D.fill, I = yt(W);
      I.fill = G, I.decal = D.decal;
      var M = P(R), N = P(z), V = P(H);
      at(a, G, D.opacity, null), a.setStyle(I), a.ensureState("emphasis").style = M, a.ensureState("blur").style = N, a.ensureState("select").style = V, ut(a);
    }
    p.add(a);
  }
  function it(p) {
    !p.invisible && n.push(p);
  }
  function at(p, a, w, c) {
    var _ = m.getModel(c ? mt : gt), D = Bt(m.get("name"), null), G = _.getShallow("show");
    Zt(p, Xt(m, c ? mt : gt), {
      defaultText: G ? D : null,
      inheritColor: a,
      defaultOpacity: w,
      labelFetcher: A,
      labelDataIndex: e.dataIndex
    });
    var I = p.getTextContent();
    if (I) {
      var M = I.style, N = At(M.padding || 0);
      c && (p.setTextConfig({
        layoutRect: c
      }), I.disableLabelLayout = !0), I.beforeUpdate = function() {
        var nt = Math.max((c ? c.width : p.shape.width) - N[1] - N[3], 0), st = Math.max((c ? c.height : p.shape.height) - N[0] - N[2], 0);
        (M.width !== nt || M.height !== st) && I.setStyle({
          width: nt,
          height: st
        });
      }, M.truncateMinChar = 2, M.lineOverflow = "truncate", ot(M, c, o);
      var V = I.getState("emphasis");
      ot(V ? V.style : null, c, o);
    }
  }
  function ot(p, a, w) {
    var c = p ? p.text : null;
    if (!a && w.isLeafRoot && c != null) {
      var _ = A.get("drillDownIcon", !0);
      p.text = _ ? _ + " " + c : c;
    }
  }
  function $(p, a, w, c) {
    var _ = g != null && t[p][g], D = i[p];
    return _ ? (t[p][g] = null, It(D, _)) : h || (_ = new a(), _ instanceof Ft && (_.z2 = Kt(w, c)), St(D, _)), d[p][C] = _;
  }
  function It(p, a) {
    var w = p[C] = {};
    a instanceof K ? (w.oldX = a.x, w.oldY = a.y) : w.oldShape = _t({}, a.shape);
  }
  function St(p, a) {
    var w = p[C] = {}, c = e.parentNode, _ = a instanceof J;
    if (c && (!r || r.direction === "drillDown")) {
      var D = 0, G = 0, I = i.background[c.getRawIndex()];
      !r && I && I.oldShape && (D = I.oldShape.width, G = I.oldShape.height), _ ? (w.oldX = 0, w.oldY = G) : w.oldShape = {
        x: D,
        y: G,
        width: 0,
        height: 0
      };
    }
    w.fadein = !_;
  }
}
function Kt(A, d) {
  return A * Yt + d;
}
export {
  ge as default
};
