// echarts@6.0.0/lib/chart/treemap/TreemapSeries.js
import { __extends as M } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { map as D, extend as w, createHashMap as S, each as y, isArray as d } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import T from "../../model/Series.js";
import V from "../../data/Tree.js";
import f from "../../model/Model.js";
import { wrapTreePathInfo as A } from "../helper/treeHelper.js";
import { normalizeToArray as g } from "../../util/model.js";
import { createTooltipMarkup as z } from "../../component/tooltip/tooltipMarkup.js";
import C from "../helper/enableAriaDecalForTree.js";
import i from "../../visual/tokens.js";
var O = (
  /** @class */
  (function(r) {
    M(o, r);
    function o() {
      var e = r !== null && r.apply(this, arguments) || this;
      return e.type = o.type, e.preventUsingHoverLayer = !0, e;
    }
    return o.prototype.getInitialData = function(e, t) {
      var a = {
        name: e.name,
        children: e.data
      };
      x(a);
      var l = e.levels || [], n = this.designatedVisualItemStyle = {}, s = new f({
        itemStyle: n
      }, this, t);
      l = e.levels = L(l, t);
      var u = D(l || [], function(m) {
        return new f(m, s, t);
      }, this), p = V.createTree(a, this, c);
      function c(m) {
        m.wrapMethod("getItemModel", function(h, b) {
          var v = p.getNodeByDataIndex(b), I = v ? u[v.depth] : null;
          return h.parentModel = I || s, h;
        });
      }
      return p.data;
    }, o.prototype.optionUpdated = function() {
      this.resetViewRoot();
    }, o.prototype.formatTooltip = function(e, t, a) {
      var l = this.getData(), n = this.getRawValue(e), s = l.getName(e);
      return z("nameValue", {
        name: s,
        value: n
      });
    }, o.prototype.getDataParams = function(e) {
      var t = r.prototype.getDataParams.apply(this, arguments), a = this.getData().tree.getNodeByDataIndex(e);
      return t.treeAncestors = A(a, this), t.treePathInfo = t.treeAncestors, t;
    }, o.prototype.setLayoutInfo = function(e) {
      this.layoutInfo = this.layoutInfo || {}, w(this.layoutInfo, e);
    }, o.prototype.mapIdToIndex = function(e) {
      var t = this._idIndexMap;
      t || (t = this._idIndexMap = S(), this._idIndexMapCount = 0);
      var a = t.get(e);
      return a == null && t.set(e, a = this._idIndexMapCount++), a;
    }, o.prototype.getViewRoot = function() {
      return this._viewRoot;
    }, o.prototype.resetViewRoot = function(e) {
      e ? this._viewRoot = e : e = this._viewRoot;
      var t = this.getRawData().tree.root;
      (!e || e !== t && !t.contains(e)) && (this._viewRoot = t);
    }, o.prototype.enableAriaDecal = function() {
      C(this);
    }, o.type = "series.treemap", o.layoutMode = "box", o.defaultOption = {
      // Disable progressive rendering
      progressive: 0,
      // size: ['80%', '80%'],            // deprecated, compatible with ec2.
      // `coordinateSystem` can be declared as 'matrix', 'calendar',
      //  which provides box layout container.
      coordinateSystemUsage: "box",
      left: i.size.l,
      top: i.size.xxxl,
      right: i.size.l,
      bottom: i.size.xxxl,
      sort: !0,
      clipWindow: "origin",
      squareRatio: 0.5 * (1 + Math.sqrt(5)),
      leafDepth: null,
      drillDownIcon: "▶",
      // to align specialized icon. ▷▶❒❐▼✚
      zoomToNodeRatio: 0.32 * 0.32,
      scaleLimit: {
        max: 5,
        min: 0.2
      },
      roam: !0,
      roamTrigger: "global",
      nodeClick: "zoomToNode",
      animation: !0,
      animationDurationUpdate: 900,
      animationEasing: "quinticInOut",
      breadcrumb: {
        show: !0,
        height: 22,
        left: "center",
        bottom: i.size.m,
        // right
        // bottom
        emptyItemWidth: 25,
        itemStyle: {
          color: i.color.backgroundShade,
          textStyle: {
            color: i.color.secondary
          }
        },
        emphasis: {
          itemStyle: {
            color: i.color.background
          }
        }
      },
      label: {
        show: !0,
        // Do not use textDistance, for ellipsis rect just the same as treemap node rect.
        distance: 0,
        padding: 5,
        position: "inside",
        // formatter: null,
        color: i.color.neutral00,
        overflow: "truncate"
        // align
        // verticalAlign
      },
      upperLabel: {
        show: !1,
        position: [0, "50%"],
        height: 20,
        // formatter: null,
        // color: '#fff',
        overflow: "truncate",
        // align: null,
        verticalAlign: "middle"
      },
      itemStyle: {
        color: null,
        colorAlpha: null,
        colorSaturation: null,
        borderWidth: 0,
        gapWidth: 0,
        borderColor: i.color.neutral00,
        borderColorSaturation: null
        // If specified, borderColor will be ineffective, and the
        // border color is evaluated by color of current node and
        // borderColorSaturation.
      },
      emphasis: {
        upperLabel: {
          show: !0,
          position: [0, "50%"],
          overflow: "truncate",
          verticalAlign: "middle"
        }
      },
      visualDimension: 0,
      visualMin: null,
      visualMax: null,
      color: [],
      // level[n].color (if necessary).
      // + Specify color list of each level. level[0].color would be global
      // color list if not specified. (see method `setDefault`).
      // + But set as a empty array to forbid fetch color from global palette
      // when using nodeModel.get('color'), otherwise nodes on deep level
      // will always has color palette set and are not able to inherit color
      // from parent node.
      // + TreemapSeries.color can not be set as 'none', otherwise effect
      // legend color fetching (see seriesColor.js).
      colorAlpha: null,
      colorSaturation: null,
      colorMappingBy: "index",
      visibleMin: 10,
      // be rendered. Only works when sort is 'asc' or 'desc'.
      childrenVisibleMin: null,
      // grandchildren will not show.
      // Why grandchildren? If not grandchildren but children,
      // some siblings show children and some not,
      // the appearance may be mess and not consistent,
      levels: []
      // Each item: {
      //     visibleMin, itemStyle, visualDimension, label
      // }
    }, o;
  })(T)
);
function x(r) {
  var o = 0;
  y(r.children, function(t) {
    x(t);
    var a = t.value;
    d(a) && (a = a[0]), o += a;
  });
  var e = r.value;
  d(e) && (e = e[0]), (e == null || isNaN(e)) && (e = o), e < 0 && (e = 0), d(r.value) ? r.value[0] = e : r.value = e;
}
function L(r, o) {
  var e = g(o.get("color")), t = g(o.get(["aria", "decal", "decals"]));
  if (e) {
    r = r || [];
    var a, l;
    y(r, function(s) {
      var u = new f(s), p = u.get("color"), c = u.get("decal");
      (u.get(["itemStyle", "color"]) || p && p !== "none") && (a = !0), (u.get(["itemStyle", "decal"]) || c && c !== "none") && (l = !0);
    });
    var n = r[0] || (r[0] = {});
    return a || (n.color = e.slice()), !l && t && (n.decal = t.slice()), r;
  }
}
export {
  O as default
};
