// echarts@6.0.0/lib/component/grid/installSimple.js
import { __extends as m } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import a from "../../view/Component.js";
import p from "../../coord/cartesian/GridModel.js";
import "../../util/graphic.js";
import { defaults as f } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { CartesianAxisModel as o } from "../../coord/cartesian/AxisModel.js";
import i from "../../coord/axisModelCreator.js";
import s from "../../coord/cartesian/Grid.js";
import { CartesianXAxisView as d, CartesianYAxisView as l } from "../axis/CartesianAxisView.js";
import c from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Rect.js";
var u = (
  /** @class */
  (function(t) {
    m(e, t);
    function e() {
      var r = t !== null && t.apply(this, arguments) || this;
      return r.type = "grid", r;
    }
    return e.prototype.render = function(r, x) {
      this.group.removeAll(), r.get("show") && this.group.add(new c({
        shape: r.coordinateSystem.getRect(),
        style: f({
          fill: r.get("backgroundColor")
        }, r.getItemStyle()),
        silent: !0,
        z2: -1
      }));
    }, e.type = "grid", e;
  })(a)
), n = {
  // gridIndex: 0,
  // gridId: '',
  offset: 0
};
function M(t) {
  t.registerComponentView(u), t.registerComponentModel(p), t.registerCoordinateSystem("cartesian2d", s), i(t, "x", o, n), i(t, "y", o, n), t.registerComponentView(d), t.registerComponentView(l), t.registerPreprocessor(function(e) {
    e.xAxis && e.yAxis && !e.grid && (e.grid = {});
  });
}
export {
  M as install
};
