// echarts@6.0.0/lib/component/legend/legendAction.js
import { curry as r, each as d, hasOwn as f } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
function a(e, l, t) {
  var s = e === "allSelect" || e === "inverseSelect", n = {}, i = [];
  t.eachComponent({
    mainType: "legend",
    query: l
  }, function(c) {
    s ? c[e]() : c[e](l.name), S(c, n), i.push(c.componentIndex);
  });
  var g = {};
  return t.eachComponent("legend", function(c) {
    d(n, function(u, o) {
      c[u ? "select" : "unSelect"](o);
    }), S(c, g);
  }), s ? {
    selected: g,
    // return legendIndex array to tell the developers which legends are allSelect / inverseSelect
    legendIndex: i
  } : {
    name: l.name,
    selected: g
  };
}
function S(e, l) {
  var t = l || {};
  return d(e.getData(), function(s) {
    var n = s.get("name");
    if (!(n === `
` || n === "")) {
      var i = e.isSelected(n);
      f(t, n) ? t[n] = t[n] && i : t[n] = i;
    }
  }), t;
}
function p(e) {
  e.registerAction("legendToggleSelect", "legendselectchanged", r(a, "toggleSelected")), e.registerAction("legendAllSelect", "legendselectall", r(a, "allSelect")), e.registerAction("legendInverseSelect", "legendinverseselect", r(a, "inverseSelect")), e.registerAction("legendSelect", "legendselected", r(a, "select")), e.registerAction("legendUnSelect", "legendunselected", r(a, "unSelect"));
}
export {
  p as installLegendAction
};
