// echarts@6.0.0/lib/component/legend/LegendModel.js
import { __extends as v } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { isArray as y, each as c, merge as m, isString as h, createHashMap as b, map as S, isNumber as M, filter as _, indexOf as w } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import O from "../../model/Model.js";
import { isNameSpecified as D } from "../../util/model.js";
import C from "../../model/Component.js";
import n from "../../visual/tokens.js";
var L = function(l, r) {
  if (r === "all")
    return {
      type: "all",
      title: l.getLocaleModel().get(["legend", "selector", "all"])
    };
  if (r === "inverse")
    return {
      type: "inverse",
      title: l.getLocaleModel().get(["legend", "selector", "inverse"])
    };
}, R = (
  /** @class */
  (function(l) {
    v(r, l);
    function r() {
      var e = l !== null && l.apply(this, arguments) || this;
      return e.type = r.type, e.layoutMode = {
        type: "box",
        // legend.width/height are maxWidth/maxHeight actually,
        // whereas real width/height is calculated by its content.
        // (Setting {left: 10, right: 10} does not make sense).
        // So consider the case:
        // `setOption({legend: {left: 10});`
        // then `setOption({legend: {right: 10});`
        // The previous `left` should be cleared by setting `ignoreSize`.
        ignoreSize: !0
      }, e;
    }
    return r.prototype.init = function(e, t, i) {
      this.mergeDefaultAndTheme(e, i), e.selected = e.selected || {}, this._updateSelector(e);
    }, r.prototype.mergeOption = function(e, t) {
      l.prototype.mergeOption.call(this, e, t), this._updateSelector(e);
    }, r.prototype._updateSelector = function(e) {
      var t = e.selector, i = this.ecModel;
      t === !0 && (t = e.selector = ["all", "inverse"]), y(t) && c(t, function(o, s) {
        h(o) && (o = {
          type: o
        }), t[s] = m(o, L(i, o.type));
      });
    }, r.prototype.optionUpdated = function() {
      this._updateData(this.ecModel);
      var e = this._data;
      if (e[0] && this.get("selectedMode") === "single") {
        for (var t = !1, i = 0; i < e.length; i++) {
          var o = e[i].get("name");
          if (this.isSelected(o)) {
            this.select(o), t = !0;
            break;
          }
        }
        !t && this.select(e[0].get("name"));
      }
    }, r.prototype._updateData = function(e) {
      var t = [], i = [];
      e.eachRawSeries(function(a) {
        var f = a.name;
        i.push(f);
        var d;
        if (a.legendVisualProvider) {
          var g = a.legendVisualProvider, p = g.getAllNames();
          e.isSeriesFiltered(a) || (i = i.concat(p)), p.length ? t = t.concat(p) : d = !0;
        } else
          d = !0;
        d && D(a) && t.push(a.name);
      }), this._availableNames = i;
      var o = this.get("data") || t, s = b(), u = S(o, function(a) {
        return (h(a) || M(a)) && (a = {
          name: a
        }), s.get(a.name) ? null : (s.set(a.name, !0), new O(a, this, this.ecModel));
      }, this);
      this._data = _(u, function(a) {
        return !!a;
      });
    }, r.prototype.getData = function() {
      return this._data;
    }, r.prototype.select = function(e) {
      var t = this.option.selected, i = this.get("selectedMode");
      if (i === "single") {
        var o = this._data;
        c(o, function(s) {
          t[s.get("name")] = !1;
        });
      }
      t[e] = !0;
    }, r.prototype.unSelect = function(e) {
      this.get("selectedMode") !== "single" && (this.option.selected[e] = !1);
    }, r.prototype.toggleSelected = function(e) {
      var t = this.option.selected;
      t.hasOwnProperty(e) || (t[e] = !0), this[t[e] ? "unSelect" : "select"](e);
    }, r.prototype.allSelect = function() {
      var e = this._data, t = this.option.selected;
      c(e, function(i) {
        t[i.get("name", !0)] = !0;
      });
    }, r.prototype.inverseSelect = function() {
      var e = this._data, t = this.option.selected;
      c(e, function(i) {
        var o = i.get("name", !0);
        t.hasOwnProperty(o) || (t[o] = !0), t[o] = !t[o];
      });
    }, r.prototype.isSelected = function(e) {
      var t = this.option.selected;
      return !(t.hasOwnProperty(e) && !t[e]) && w(this._availableNames, e) >= 0;
    }, r.prototype.getOrient = function() {
      return this.get("orient") === "vertical" ? {
        index: 1,
        name: "vertical"
      } : {
        index: 0,
        name: "horizontal"
      };
    }, r.type = "legend.plain", r.dependencies = ["series"], r.defaultOption = {
      // zlevel: 0,
      z: 4,
      show: !0,
      orient: "horizontal",
      left: "center",
      // right: 'center',
      // top: 0,
      bottom: n.size.m,
      align: "auto",
      backgroundColor: n.color.transparent,
      borderColor: n.color.border,
      borderRadius: 0,
      borderWidth: 0,
      padding: 5,
      itemGap: 8,
      itemWidth: 25,
      itemHeight: 14,
      symbolRotate: "inherit",
      symbolKeepAspect: !0,
      inactiveColor: n.color.disabled,
      inactiveBorderColor: n.color.disabled,
      inactiveBorderWidth: "auto",
      itemStyle: {
        color: "inherit",
        opacity: "inherit",
        borderColor: "inherit",
        borderWidth: "auto",
        borderCap: "inherit",
        borderJoin: "inherit",
        borderDashOffset: "inherit",
        borderMiterLimit: "inherit"
      },
      lineStyle: {
        width: "auto",
        color: "inherit",
        inactiveColor: n.color.disabled,
        inactiveWidth: 2,
        opacity: "inherit",
        type: "inherit",
        cap: "inherit",
        join: "inherit",
        dashOffset: "inherit",
        miterLimit: "inherit"
      },
      textStyle: {
        color: n.color.secondary
      },
      selectedMode: !0,
      selector: !1,
      selectorLabel: {
        show: !0,
        borderRadius: 10,
        padding: [3, 5, 3, 5],
        fontSize: 12,
        fontFamily: "sans-serif",
        color: n.color.tertiary,
        borderWidth: 1,
        borderColor: n.color.border
      },
      emphasis: {
        selectorLabel: {
          show: !0,
          color: n.color.quaternary
        }
      },
      selectorPosition: "auto",
      selectorItemGap: 7,
      selectorButtonGap: 10,
      tooltip: {
        show: !1
      },
      triggerEvent: !1
    }, r;
  })(C)
);
export {
  R as default
};
