// echarts@6.0.0/lib/component/legend/LegendView.js
import { __extends as q } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { defaults as J, createHashMap as Q, each as Y, curry as M, extend as V, isFunction as $, isString as tt } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { parse as et, stringify as it } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/tool/color.js";
import { setTooltipConfig as rt } from "../../util/graphic.js";
import { enableHoverEmphasis as z } from "../../util/states.js";
import { setLabelStyle as nt, createTextStyle as at } from "../../label/labelStyle.js";
import { makeBackground as ot } from "../helper/listComponent.js";
import { createBoxLayoutReference as st, getLayoutRect as U, box as Z } from "../../util/layout.js";
import ct from "../../view/Component.js";
import { createSymbol as lt } from "../../util/symbol.js";
import { createOrUpdatePatternFromDecal as ht } from "../../util/decal.js";
import { getECData as F } from "../../util/innerStore.js";
import ut from "../../visual/tokens.js";
import vt from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import N from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Text.js";
import ft from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Rect.js";
var _ = M, O = Y, T = vt, Tt = (
  /** @class */
  (function(a) {
    q(r, a);
    function r() {
      var t = a !== null && a.apply(this, arguments) || this;
      return t.type = r.type, t.newlineDisabled = !1, t;
    }
    return r.prototype.init = function() {
      this.group.add(this._contentGroup = new T()), this.group.add(this._selectorGroup = new T()), this._isFirstRender = !0;
    }, r.prototype.getContentGroup = function() {
      return this._contentGroup;
    }, r.prototype.getSelectorGroup = function() {
      return this._selectorGroup;
    }, r.prototype.render = function(t, e, h) {
      var c = this._isFirstRender;
      if (this._isFirstRender = !1, this.resetInner(), !!t.get("show", !0)) {
        var v = t.get("align"), y = t.get("orient");
        (!v || v === "auto") && (v = t.get("left") === "right" && y === "vertical" ? "right" : "left");
        var m = t.get("selector", !0), i = t.get("selectorPosition", !0);
        m && (!i || i === "auto") && (i = y === "horizontal" ? "end" : "start"), this.renderInner(v, t, e, h, m, y, i);
        var o = st(t, h).refContainer, f = t.getBoxLayoutParams(), p = t.get("padding"), s = U(f, o, p), u = this.layoutInner(t, v, s, c, m, i), l = U(J({
          width: u.width,
          height: u.height
        }, f), o, p);
        this.group.x = l.x - u.x, this.group.y = l.y - u.y, this.group.markRedraw(), this.group.add(this._backgroundEl = ot(
          u,
          // FXIME: most itemStyle options does not work in background because inherit is not handled yet.
          t
        ));
      }
    }, r.prototype.resetInner = function() {
      this.getContentGroup().removeAll(), this._backgroundEl && this.group.remove(this._backgroundEl), this.getSelectorGroup().removeAll();
    }, r.prototype.renderInner = function(t, e, h, c, v, y, m) {
      var i = this.getContentGroup(), o = Q(), f = e.get("selectedMode"), p = e.get("triggerEvent"), s = [];
      h.eachRawSeries(function(u) {
        !u.get("legendHoverLink") && s.push(u.id);
      }), O(e.getData(), function(u, l) {
        var d = this, n = u.get("name");
        if (!this.newlineDisabled && (n === "" || n === `
`)) {
          var b = new T();
          b.newline = !0, i.add(b);
          return;
        }
        var g = h.getSeriesByName(n)[0];
        if (!o.get(n)) {
          if (g) {
            var D = g.getData(), S = D.getVisual("legendLineStyle") || {}, A = D.getVisual("legendIcon"), E = D.getVisual("style"), L = this._createItem(g, n, l, u, e, t, S, E, A, f, c);
            L.on("click", _(X, n, null, c, s)).on("mouseover", _(P, g.name, null, c, s)).on("mouseout", _(K, g.name, null, c, s)), h.ssr && L.eachChild(function(x) {
              var k = F(x);
              k.seriesIndex = g.seriesIndex, k.dataIndex = l, k.ssrType = "legend";
            }), p && L.eachChild(function(x) {
              d.packEventData(x, e, g, l, n);
            }), o.set(n, !0);
          } else
            h.eachRawSeries(function(x) {
              var k = this;
              if (!o.get(n) && x.legendVisualProvider) {
                var w = x.legendVisualProvider;
                if (!w.containName(n))
                  return;
                var G = w.indexOfName(n), C = w.getItemVisual(G, "style"), W = w.getItemVisual(G, "legendIcon"), R = et(C.fill);
                R && R[3] === 0 && (R[3] = 0.2, C = V(V({}, C), {
                  fill: it(R, "rgba")
                }));
                var B = this._createItem(x, n, l, u, e, t, {}, C, W, f, c);
                B.on("click", _(X, null, n, c, s)).on("mouseover", _(P, null, n, c, s)).on("mouseout", _(K, null, n, c, s)), h.ssr && B.eachChild(function(H) {
                  var I = F(H);
                  I.seriesIndex = x.seriesIndex, I.dataIndex = l, I.ssrType = "legend";
                }), p && B.eachChild(function(H) {
                  k.packEventData(H, e, x, l, n);
                }), o.set(n, !0);
              }
            }, this);
          process.env.NODE_ENV !== "production" && (o.get(n) || console.warn(n + " series not exists. Legend data should be same with series name or data name."));
        }
      }, this), v && this._createSelector(v, e, c, y, m);
    }, r.prototype.packEventData = function(t, e, h, c, v) {
      var y = {
        componentType: "legend",
        componentIndex: e.componentIndex,
        dataIndex: c,
        value: v,
        seriesIndex: h.seriesIndex
      };
      F(t).eventData = y;
    }, r.prototype._createSelector = function(t, e, h, c, v) {
      var y = this.getSelectorGroup();
      O(t, function(i) {
        var o = i.type, f = new N({
          style: {
            x: 0,
            y: 0,
            align: "center",
            verticalAlign: "middle"
          },
          onclick: function() {
            h.dispatchAction({
              type: o === "all" ? "legendAllSelect" : "legendInverseSelect",
              legendId: e.id
            });
          }
        });
        y.add(f);
        var p = e.getModel("selectorLabel"), s = e.getModel(["emphasis", "selectorLabel"]);
        nt(f, {
          normal: p,
          emphasis: s
        }, {
          defaultText: i.title
        }), z(f);
      });
    }, r.prototype._createItem = function(t, e, h, c, v, y, m, i, o, f, p) {
      var s = t.visualDrawType, u = v.get("itemWidth"), l = v.get("itemHeight"), d = v.isSelected(e), n = c.get("symbolRotate"), b = c.get("symbolKeepAspect"), g = c.get("icon");
      o = g || o || "roundRect";
      var D = pt(o, c, m, i, s, d, p), S = new T(), A = c.getModel("textStyle");
      if ($(t.getLegendIcon) && (!g || g === "inherit"))
        S.add(t.getLegendIcon({
          itemWidth: u,
          itemHeight: l,
          icon: o,
          iconRotate: n,
          itemStyle: D.itemStyle,
          lineStyle: D.lineStyle,
          symbolKeepAspect: b
        }));
      else {
        var E = g === "inherit" && t.getData().getVisual("symbol") ? n === "inherit" ? t.getData().getVisual("symbolRotate") : n : 0;
        S.add(gt({
          itemWidth: u,
          itemHeight: l,
          icon: o,
          iconRotate: E,
          itemStyle: D.itemStyle,
          symbolKeepAspect: b
        }));
      }
      var L = y === "left" ? u + 5 : -5, x = y, k = v.get("formatter"), w = e;
      tt(k) && k ? w = k.replace("{name}", e ?? "") : $(k) && (w = k(e));
      var G = d ? A.getTextColor() : c.get("inactiveColor");
      S.add(new N({
        style: at(A, {
          text: w,
          x: L,
          y: l / 2,
          fill: G,
          align: x,
          verticalAlign: "middle"
        }, {
          inheritColor: G
        })
      }));
      var C = new ft({
        shape: S.getBoundingRect(),
        style: {
          // Cannot use 'invisible' because SVG SSR will miss the node
          fill: "transparent"
        }
      }), W = c.getModel("tooltip");
      return W.get("show") && rt({
        el: C,
        componentModel: v,
        itemName: e,
        itemTooltipOption: W.option
      }), S.add(C), S.eachChild(function(R) {
        R.silent = !0;
      }), C.silent = !f, this.getContentGroup().add(S), z(S), S.__legendDataIndex = h, S;
    }, r.prototype.layoutInner = function(t, e, h, c, v, y) {
      var m = this.getContentGroup(), i = this.getSelectorGroup();
      Z(t.get("orient"), m, t.get("itemGap"), h.width, h.height);
      var o = m.getBoundingRect(), f = [-o.x, -o.y];
      if (i.markRedraw(), m.markRedraw(), v) {
        Z(
          // Buttons in selectorGroup always layout horizontally
          "horizontal",
          i,
          t.get("selectorItemGap", !0)
        );
        var p = i.getBoundingRect(), s = [-p.x, -p.y], u = t.get("selectorButtonGap", !0), l = t.getOrient().index, d = l === 0 ? "width" : "height", n = l === 0 ? "height" : "width", b = l === 0 ? "y" : "x";
        y === "end" ? s[l] += o[d] + u : f[l] += p[d] + u, s[1 - l] += o[n] / 2 - p[n] / 2, i.x = s[0], i.y = s[1], m.x = f[0], m.y = f[1];
        var g = {
          x: 0,
          y: 0
        };
        return g[d] = o[d] + u + p[d], g[n] = Math.max(o[n], p[n]), g[b] = Math.min(0, p[b] + s[1 - l]), g;
      } else
        return m.x = f[0], m.y = f[1], this.group.getBoundingRect();
    }, r.prototype.remove = function() {
      this.getContentGroup().removeAll(), this._isFirstRender = !0;
    }, r.type = "legend.plain", r;
  })(ct)
);
function pt(a, r, t, e, h, c, v) {
  function y(d, n) {
    d.lineWidth === "auto" && (d.lineWidth = n.lineWidth > 0 ? 2 : 0), O(d, function(b, g) {
      d[g] === "inherit" && (d[g] = n[g]);
    });
  }
  var m = r.getModel("itemStyle"), i = m.getItemStyle(), o = a.lastIndexOf("empty", 0) === 0 ? "fill" : "stroke", f = m.getShallow("decal");
  i.decal = !f || f === "inherit" ? e.decal : ht(f, v), i.fill === "inherit" && (i.fill = e[h]), i.stroke === "inherit" && (i.stroke = e[o]), i.opacity === "inherit" && (i.opacity = (h === "fill" ? e : t).opacity), y(i, e);
  var p = r.getModel("lineStyle"), s = p.getLineStyle();
  if (y(s, t), i.fill === "auto" && (i.fill = e.fill), i.stroke === "auto" && (i.stroke = e.fill), s.stroke === "auto" && (s.stroke = e.fill), !c) {
    var u = r.get("inactiveBorderWidth"), l = i[o];
    i.lineWidth = u === "auto" ? e.lineWidth > 0 && l ? 2 : 0 : i.lineWidth, i.fill = r.get("inactiveColor"), i.stroke = r.get("inactiveBorderColor"), s.stroke = p.get("inactiveColor"), s.lineWidth = p.get("inactiveWidth");
  }
  return {
    itemStyle: i,
    lineStyle: s
  };
}
function gt(a) {
  var r = a.icon || "roundRect", t = lt(r, 0, 0, a.itemWidth, a.itemHeight, a.itemStyle.fill, a.symbolKeepAspect);
  return t.setStyle(a.itemStyle), t.rotation = (a.iconRotate || 0) * Math.PI / 180, t.setOrigin([a.itemWidth / 2, a.itemHeight / 2]), r.indexOf("empty") > -1 && (t.style.stroke = t.style.fill, t.style.fill = ut.color.neutral00, t.style.lineWidth = 2), t;
}
function X(a, r, t, e) {
  K(a, r, t, e), t.dispatchAction({
    type: "legendToggleSelect",
    name: a ?? r
  }), P(a, r, t, e);
}
function j(a) {
  for (var r = a.getZr().storage.getDisplayList(), t, e = 0, h = r.length; e < h && !(t = r[e].states.emphasis); )
    e++;
  return t && t.hoverLayer;
}
function P(a, r, t, e) {
  j(t) || t.dispatchAction({
    type: "highlight",
    seriesName: a,
    name: r,
    excludeSeriesId: e
  });
}
function K(a, r, t, e) {
  j(t) || t.dispatchAction({
    type: "downplay",
    seriesName: a,
    name: r,
    excludeSeriesId: e
  });
}
export {
  Tt as default
};
