// echarts@6.0.0/lib/component/legend/ScrollableLegendModel.js
import { __extends as c } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import i from "./LegendModel.js";
import { getLayoutParams as m, mergeLayoutParam as g } from "../../util/layout.js";
import { inheritDefaultOption as u } from "../../util/component.js";
import r from "../../visual/tokens.js";
var h = (
  /** @class */
  (function(e) {
    c(o, e);
    function o() {
      var t = e !== null && e.apply(this, arguments) || this;
      return t.type = o.type, t;
    }
    return o.prototype.setScrollDataIndex = function(t) {
      this.option.scrollDataIndex = t;
    }, o.prototype.init = function(t, n, a) {
      var p = m(t);
      e.prototype.init.call(this, t, n, a), l(this, t, p);
    }, o.prototype.mergeOption = function(t, n) {
      e.prototype.mergeOption.call(this, t, n), l(this, this.option, t);
    }, o.type = "legend.scroll", o.defaultOption = u(i.defaultOption, {
      scrollDataIndex: 0,
      pageButtonItemGap: 5,
      pageButtonGap: null,
      pageButtonPosition: "end",
      pageFormatter: "{current}/{total}",
      pageIcons: {
        horizontal: ["M0,0L12,-10L12,10z", "M0,0L-12,-10L-12,10z"],
        vertical: ["M0,0L20,0L10,-20z", "M0,0L20,0L10,20z"]
      },
      pageIconColor: r.color.accent50,
      pageIconInactiveColor: r.color.accent10,
      pageIconSize: 15,
      pageTextStyle: {
        color: r.color.tertiary
      },
      animationDurationUpdate: 800
    }), o;
  })(i)
);
function l(e, o, t) {
  var n = e.getOrient(), a = [1, 1];
  a[n.index] = 0, g(o, t, {
    type: "box",
    ignoreSize: !!a
  });
}
export {
  h as default
};
