// echarts@6.0.0/lib/component/geo/GeoView.js
import { __extends as h } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import m from "../helper/MapDraw.js";
import u from "../../view/Component.js";
import { getECData as s } from "../../util/innerStore.js";
import { findEventDispatcher as c } from "../../util/event.js";
var d = (
  /** @class */
  (function(n) {
    h(e, n);
    function e() {
      var t = n !== null && n.apply(this, arguments) || this;
      return t.type = e.type, t.focusBlurEnabled = !0, t;
    }
    return e.prototype.init = function(t, i) {
      this._api = i;
    }, e.prototype.render = function(t, i, r, o) {
      if (this._model = t, !t.get("show")) {
        this._mapDraw && this._mapDraw.remove(), this._mapDraw = null;
        return;
      }
      this._mapDraw || (this._mapDraw = new m(r));
      var a = this._mapDraw;
      a.draw(t, i, r, this, o), a.group.on("click", this._handleRegionClick, this), a.group.silent = t.get("silent"), this.group.add(a.group), this.updateSelectStatus(t, i, r);
    }, e.prototype._handleRegionClick = function(t) {
      var i;
      c(t.target, function(r) {
        return (i = s(r).eventData) != null;
      }, !0), i && this._api.dispatchAction({
        type: "geoToggleSelect",
        geoId: this._model.id,
        name: i.name
      });
    }, e.prototype.updateSelectStatus = function(t, i, r) {
      var o = this;
      this._mapDraw.group.traverse(function(a) {
        var p = s(a).eventData;
        if (p)
          return o._model.isSelected(p.name) ? r.enterSelect(a) : r.leaveSelect(a), !0;
      });
    }, e.prototype.findHighDownDispatchers = function(t) {
      return this._mapDraw && this._mapDraw.findHighDownDispatchers(t, this._model);
    }, e.prototype.dispose = function() {
      this._mapDraw && this._mapDraw.remove();
    }, e.type = "geo", e;
  })(u)
);
export {
  d as default
};
