// echarts@6.0.0/lib/component/geo/install.js
import S from "../../coord/geo/GeoModel.js";
import v from "../../coord/geo/geoCreator.js";
import { each as g } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { updateCenterAndZoomInAction as C } from "../helper/roamHelper.js";
import d from "./GeoView.js";
import l from "../../coord/geo/geoSourceManager.js";
function y(n, i, t) {
  l.registerMap(n, i, t);
}
function G(n) {
  n.registerCoordinateSystem("geo", v), n.registerComponentModel(S), n.registerComponentView(d), n.registerImpl("registerMap", y), n.registerImpl("getMap", function(t) {
    return l.getMapForUser(t);
  });
  function i(t, c) {
    c.update = "geo:updateSelectStatus", n.registerAction(c, function(s, o) {
      var e = {}, m = [];
      return o.eachComponent({
        mainType: "geo",
        query: s
      }, function(r) {
        r[t](s.name);
        var a = r.coordinateSystem;
        g(a.regions, function(u) {
          e[u.name] = r.isSelected(u.name) || !1;
        });
        var p = [];
        g(e, function(u, f) {
          e[f] && p.push(f);
        }), m.push({
          geoIndex: r.componentIndex,
          // Use singular, the same naming convention as the event `selectchanged`.
          name: p
        });
      }), {
        selected: e,
        allSelected: m,
        name: s.name
      };
    });
  }
  i("toggleSelected", {
    type: "geoToggleSelect",
    event: "geoselectchanged"
  }), i("select", {
    type: "geoSelect",
    event: "geoselected"
  }), i("unSelect", {
    type: "geoUnSelect",
    event: "geounselected"
  }), n.registerAction({
    type: "geoRoam",
    event: "geoRoam",
    update: "updateTransform"
  }, function(t, c, s) {
    var o = t.componentType;
    o || (t.geoId != null ? o = "geo" : t.seriesId != null && (o = "series")), o || (o = "series"), c.eachComponent({
      mainType: o,
      query: t
    }, function(e) {
      var m = e.coordinateSystem;
      if (m.type === "geo") {
        var r = C(m, t, e.get("scaleLimit"));
        e.setCenter && e.setCenter(r.center), e.setZoom && e.setZoom(r.zoom), o === "series" && g(e.seriesGroup, function(a) {
          a.setCenter(r.center), a.setZoom(r.zoom);
        });
      }
    });
  });
}
export {
  G as install
};
