// echarts@6.0.0/lib/chart/effectScatter/EffectScatterView.js
import { __extends as n } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import m from "../helper/SymbolDraw.js";
import f from "../helper/EffectSymbol.js";
import { clone as s } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/matrix.js";
import c from "../../layout/points.js";
import u from "../../view/Chart.js";
var v = (
  /** @class */
  (function(i) {
    n(o, i);
    function o() {
      var t = i !== null && i.apply(this, arguments) || this;
      return t.type = o.type, t;
    }
    return o.prototype.init = function() {
      this._symbolDraw = new m(f);
    }, o.prototype.render = function(t, r, p) {
      var e = t.getData(), a = this._symbolDraw;
      a.updateData(e, {
        clipShape: this._getClipShape(t)
      }), this.group.add(a.group);
    }, o.prototype._getClipShape = function(t) {
      var r = t.coordinateSystem, p = r && r.getArea && r.getArea();
      return t.get("clip", !0) ? p : null;
    }, o.prototype.updateTransform = function(t, r, p) {
      var e = t.getData();
      this.group.dirty();
      var a = c("").reset(t, r, p);
      a.progress && a.progress({
        start: 0,
        end: e.count(),
        count: e.count()
      }, e), this._symbolDraw.updateLayout();
    }, o.prototype._updateGroupTransform = function(t) {
      var r = t.coordinateSystem;
      r && r.getRoamTransform && (this.group.transform = s(r.getRoamTransform()), this.group.decomposeTransform());
    }, o.prototype.remove = function(t, r) {
      this._symbolDraw && this._symbolDraw.remove(!0);
    }, o.type = "effectScatter", o;
  })(u)
);
export {
  v as default
};
