// echarts@6.0.0/lib/component/axis/AngleAxisView.js
import { __extends as B } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { map as V, clone as D, each as R, defaults as L, isObject as N } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import * as O from "../../util/graphic.js";
import { mergePath as x, setTooltipConfig as j } from "../../util/graphic.js";
import { createTextStyle as Z } from "../../label/labelStyle.js";
import _ from "../../model/Model.js";
import q from "./AxisView.js";
import C from "./AxisBuilder.js";
import { getECData as E } from "../../util/innerStore.js";
import F from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Sector.js";
import A from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Line.js";
import G from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Text.js";
import H from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Ring.js";
var J = ["axisLine", "axisLabel", "axisTick", "minorTick", "splitLine", "minorSplitLine", "splitArea"];
function M(n, e, t) {
  e[1] > e[0] && (e = e.slice().reverse());
  var d = n.coordToPoint([e[0], t]), o = n.coordToPoint([e[1], t]);
  return {
    x1: d[0],
    y1: d[1],
    x2: o[0],
    y2: o[1]
  };
}
function T(n) {
  var e = n.getRadiusAxis();
  return e.inverse ? 0 : 1;
}
function P(n) {
  var e = n[0], t = n[n.length - 1];
  e && t && Math.abs(Math.abs(e.coord - t.coord) - 360) < 1e-4 && n.pop();
}
var ne = (
  /** @class */
  (function(n) {
    B(e, n);
    function e() {
      var t = n !== null && n.apply(this, arguments) || this;
      return t.type = e.type, t.axisPointerClass = "PolarAxisPointer", t;
    }
    return e.prototype.render = function(t, d) {
      if (this.group.removeAll(), !!t.get("show")) {
        var o = t.axis, s = o.polar, v = s.getRadiusAxis().getExtent(), c = o.getTicksCoords({
          breakTicks: "none"
        }), i = o.getMinorTicksCoords(), l = V(o.getViewLabels(), function(r) {
          r = D(r);
          var a = o.scale, g = a.type === "ordinal" ? a.getRawOrdinalNumber(r.tickValue) : r.tickValue;
          return r.coord = o.dataToCoord(g), r;
        });
        P(l), P(c), R(J, function(r) {
          t.get([r, "show"]) && (!o.scale.isBlank() || r === "axisLine") && K[r](this.group, t, s, c, i, v, l);
        }, this);
      }
    }, e.type = "angleAxis", e;
  })(q)
), K = {
  axisLine: function(n, e, t, d, o, s) {
    var v = e.getModel(["axisLine", "lineStyle"]), c = t.getAngleAxis(), i = Math.PI / 180, l = c.getExtent(), r = T(t), a = r ? 0 : 1, g, u = Math.abs(l[1] - l[0]) === 360 ? "Circle" : "Arc";
    s[a] === 0 ? g = new O[u]({
      shape: {
        cx: t.cx,
        cy: t.cy,
        r: s[r],
        startAngle: -l[0] * i,
        endAngle: -l[1] * i,
        clockwise: c.inverse
      },
      style: v.getLineStyle(),
      z2: 1,
      silent: !0
    }) : g = new H({
      shape: {
        cx: t.cx,
        cy: t.cy,
        r: s[r],
        r0: s[a]
      },
      style: v.getLineStyle(),
      z2: 1,
      silent: !0
    }), g.style.fill = null, n.add(g);
  },
  axisTick: function(n, e, t, d, o, s) {
    var v = e.getModel("axisTick"), c = (v.get("inside") ? -1 : 1) * v.get("length"), i = s[T(t)], l = V(d, function(r) {
      return new A({
        shape: M(t, [i, i + c], r.coord)
      });
    });
    n.add(x(l, {
      style: L(v.getModel("lineStyle").getLineStyle(), {
        stroke: e.get(["axisLine", "lineStyle", "color"])
      })
    }));
  },
  minorTick: function(n, e, t, d, o, s) {
    if (o.length) {
      for (var v = e.getModel("axisTick"), c = e.getModel("minorTick"), i = (v.get("inside") ? -1 : 1) * c.get("length"), l = s[T(t)], r = [], a = 0; a < o.length; a++)
        for (var g = 0; g < o[a].length; g++)
          r.push(new A({
            shape: M(t, [l, l + i], o[a][g].coord)
          }));
      n.add(x(r, {
        style: L(c.getModel("lineStyle").getLineStyle(), L(v.getLineStyle(), {
          stroke: e.get(["axisLine", "lineStyle", "color"])
        }))
      }));
    }
  },
  axisLabel: function(n, e, t, d, o, s, v) {
    var c = e.getCategories(!0), i = e.getModel("axisLabel"), l = i.get("margin"), r = e.get("triggerEvent");
    R(v, function(a, g) {
      var u = i, S = a.tickValue, h = s[T(t)], f = t.coordToPoint([h + l, a.coord]), y = t.cx, p = t.cy, m = Math.abs(f[0] - y) / h < 0.3 ? "center" : f[0] > y ? "left" : "right", z = Math.abs(f[1] - p) / h < 0.3 ? "middle" : f[1] > p ? "top" : "bottom";
      if (c && c[S]) {
        var w = c[S];
        N(w) && w.textStyle && (u = new _(w.textStyle, i, i.ecModel));
      }
      var k = new G({
        silent: C.isLabelSilent(e),
        style: Z(u, {
          x: f[0],
          y: f[1],
          fill: u.getTextColor() || e.get(["axisLine", "lineStyle", "color"]),
          text: a.formattedLabel,
          align: m,
          verticalAlign: z
        })
      });
      if (n.add(k), j({
        el: k,
        componentModel: e,
        itemName: a.formattedLabel,
        formatterParamsExtra: {
          isTruncated: function() {
            return k.isTruncated;
          },
          value: a.rawLabel,
          tickIndex: g
        }
      }), r) {
        var b = C.makeAxisEventDataBase(e);
        b.targetType = "axisLabel", b.value = a.rawLabel, E(k).eventData = b;
      }
    }, this);
  },
  splitLine: function(n, e, t, d, o, s) {
    var v = e.getModel("splitLine"), c = v.getModel("lineStyle"), i = c.get("color"), l = 0;
    i = i instanceof Array ? i : [i];
    for (var r = [], a = 0; a < d.length; a++) {
      var g = l++ % i.length;
      r[g] = r[g] || [], r[g].push(new A({
        shape: M(t, s, d[a].coord)
      }));
    }
    for (var a = 0; a < r.length; a++)
      n.add(x(r[a], {
        style: L({
          stroke: i[a % i.length]
        }, c.getLineStyle()),
        silent: !0,
        z: e.get("z")
      }));
  },
  minorSplitLine: function(n, e, t, d, o, s) {
    if (o.length) {
      for (var v = e.getModel("minorSplitLine"), c = v.getModel("lineStyle"), i = [], l = 0; l < o.length; l++)
        for (var r = 0; r < o[l].length; r++)
          i.push(new A({
            shape: M(t, s, o[l][r].coord)
          }));
      n.add(x(i, {
        style: c.getLineStyle(),
        silent: !0,
        z: e.get("z")
      }));
    }
  },
  splitArea: function(n, e, t, d, o, s) {
    if (d.length) {
      var v = e.getModel("splitArea"), c = v.getModel("areaStyle"), i = c.get("color"), l = 0;
      i = i instanceof Array ? i : [i];
      for (var r = [], a = Math.PI / 180, g = -d[0].coord * a, u = Math.min(s[0], s[1]), S = Math.max(s[0], s[1]), h = e.get("clockwise"), f = 1, y = d.length; f <= y; f++) {
        var p = f === y ? d[0].coord : d[f].coord, m = l++ % i.length;
        r[m] = r[m] || [], r[m].push(new F({
          shape: {
            cx: t.cx,
            cy: t.cy,
            r0: u,
            r: S,
            startAngle: g,
            endAngle: -p * a,
            clockwise: h
          },
          silent: !0
        })), g = -p * a;
      }
      for (var f = 0; f < r.length; f++)
        n.add(x(r[f], {
          style: L({
            fill: i[f % i.length]
          }, c.getAreaStyle()),
          silent: !0
        }));
    }
  }
};
export {
  ne as default
};
