// echarts@6.0.0/lib/component/axis/axisSplitHelper.js
import { createHashMap as h, isArray as R, defaults as H } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { makeInner as _ } from "../../util/model.js";
import x from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Rect.js";
var f = _();
function I(s, b, A, B) {
  var o = A.axis;
  if (!o.scale.isBlank()) {
    var C = A.getModel("splitArea"), g = C.getModel("areaStyle"), a = g.get("color"), i = B.coordinateSystem.getRect(), e = o.getTicksCoords({
      tickModel: C,
      clamp: !0,
      breakTicks: "none",
      pruneByBreak: "preserve_extent_bound"
    });
    if (e.length) {
      var u = a.length, m = f(s).splitAreaColors, S = h(), l = 0;
      if (m)
        for (var r = 0; r < e.length; r++) {
          var k = m.get(e[r].tickValue);
          if (k != null) {
            l = (k + (u - 1) * r) % u;
            break;
          }
        }
      var v = o.toGlobalCoord(e[0].coord), M = g.getAreaStyle();
      a = R(a) ? a : [a];
      for (var r = 1; r < e.length; r++) {
        var y = o.toGlobalCoord(e[r].coord), t = void 0, n = void 0, d = void 0, p = void 0;
        o.isHorizontal() ? (t = v, n = i.y, d = y - t, p = i.height, v = t + d) : (t = i.x, n = v, d = i.width, p = y - n, v = n + p);
        var c = e[r - 1].tickValue;
        c != null && S.set(c, l), b.add(new x({
          anid: c != null ? "area_" + c : null,
          shape: {
            x: t,
            y: n,
            width: d,
            height: p
          },
          style: H({
            fill: a[l]
          }, M),
          autoBatch: !0,
          silent: !0
        })), l = (l + 1) % u;
      }
      f(s).splitAreaColors = S;
    }
  }
}
function L(s) {
  f(s).splitAreaColors = null;
}
export {
  I as rectCoordAxisBuildSplitArea,
  L as rectCoordAxisHandleRemove
};
