// echarts@6.0.0/lib/component/axis/axisAction.js
import { parseFinder as E } from "../../util/model.js";
import { each as o, defaults as _ } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
var l = "expandAxisBreak", k = "collapseAxisBreak", B = "toggleAxisBreak", s = "axisbreakchanged", I = {
  type: l,
  event: s,
  update: "update",
  refineEvent: c
}, g = {
  type: k,
  event: s,
  update: "update",
  refineEvent: c
}, T = {
  type: B,
  event: s,
  update: "update",
  refineEvent: c
};
function c(n, t, a, d) {
  var e = [];
  return o(n, function(r) {
    e = e.concat(r.eventBreaks);
  }), {
    eventContent: {
      breaks: e
    }
  };
}
function P(n) {
  n.registerAction(I, t), n.registerAction(g, t), n.registerAction(T, t);
  function t(a, d) {
    var e = [], r = E(d, a);
    function i(x, p) {
      o(r[x], function(v) {
        var u = v.updateAxisBreaks(a);
        o(u.breaks, function(f) {
          var A;
          e.push(_((A = {}, A[p] = v.componentIndex, A), f));
        });
      });
    }
    return i("xAxisModels", "xAxisIndex"), i("yAxisModels", "yAxisIndex"), i("singleAxisModels", "singleAxisIndex"), {
      eventBreaks: e
    };
  }
}
export {
  k as AXIS_BREAK_COLLAPSE_ACTION_TYPE,
  l as AXIS_BREAK_EXPAND_ACTION_TYPE,
  B as AXIS_BREAK_TOGGLE_ACTION_TYPE,
  P as registerAction
};
