// echarts@6.0.0/lib/component/axis/axisBreakHelperImpl.js
import { makeInner as pr } from "../../util/model.js";
import { each as lr, find as $, assert as mr, map as J, extend as kr } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { getScaleBreakHelper as C } from "../../scale/break.js";
import { subPixelOptimizeLine as dr } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/helper/subPixelOptimize.js";
import { applyTransform as ar } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/vector.js";
import { rotate as xr, create as ir, mul as br } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/matrix.js";
import { AXIS_BREAK_EXPAND_ACTION_TYPE as vr, AXIS_BREAK_COLLAPSE_ACTION_TYPE as gr, AXIS_BREAK_TOGGLE_ACTION_TYPE as hr } from "./axisAction.js";
import { labelIntersect as Ar, labelLayoutApplyTranslation as nr } from "../../label/labelLayoutHelper.js";
import { registerAxisBreakHelperImpl as zr } from "./axisBreakHelper.js";
import { warn as yr } from "../../util/log.js";
import { WH as Q, XY as G } from "../../util/graphic.js";
import tr from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import _r from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Line.js";
import M from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/Point.js";
import or from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Polyline.js";
import Tr from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Polygon.js";
var sr = pr();
function Br(l, a) {
  var r = $(l, function(i) {
    return C().identifyAxisBreak(i.parsedBreak.breakOption, a.breakOption);
  });
  return r || l.push(r = {
    zigzagRandomList: [],
    parsedBreak: a,
    shouldRemove: !1
  }), r;
}
function Er(l) {
  lr(l, function(a) {
    return a.shouldRemove = !0;
  });
}
function Ir(l) {
  for (var a = l.length - 1; a >= 0; a--)
    l[a].shouldRemove && l.splice(a, 1);
}
function Or(l, a, r, i, e) {
  var n = r.axis;
  if (n.scale.isBlank() || !C())
    return;
  var s = C().retrieveAxisBreakPairs(n.scale.getTicks({
    breakTicks: "only_break"
  }), function(A) {
    return A.break;
  }, !1);
  if (!s.length)
    return;
  var v = r.getModel("breakArea"), u = v.get("zigzagAmplitude"), d = v.get("zigzagMinSpan"), p = v.get("zigzagMaxSpan");
  d = Math.max(2, d || 0), p = Math.max(d, p || 0);
  var T = v.get("expandOnClick"), f = v.get("zigzagZ"), E = v.getModel("itemStyle"), g = E.getItemStyle(), w = g.stroke, m = g.lineWidth, k = g.lineDash, t = g.fill, h = new tr({
    ignoreModelZ: !0
  }), z = n.isHorizontal(), I = sr(a).visualList || (sr(a).visualList = []);
  Er(I);
  for (var O = function(A) {
    var x = s[A][0].break.parsedBreak, o = [];
    o[0] = n.toGlobalCoord(n.dataToCoord(x.vmin, !0)), o[1] = n.toGlobalCoord(n.dataToCoord(x.vmax, !0)), o[1] < o[0] && o.reverse();
    var c = Br(I, x);
    c.shouldRemove = !1;
    var b = new tr();
    H(c.zigzagRandomList, b, o[0], o[1], z, x), T && b.on("click", function() {
      var S = {
        type: vr,
        breaks: [{
          start: x.breakOption.start,
          end: x.breakOption.end
        }]
      };
      S[n.dim + "AxisIndex"] = r.componentIndex, e.dispatchAction(S);
    }), b.silent = !T, h.add(b);
  }, y = 0; y < s.length; y++)
    O(y);
  l.add(h), Ir(I);
  function H(A, x, o, c, b, S) {
    var q = {
      stroke: w,
      lineWidth: m,
      lineDash: k,
      fill: "none"
    }, B = b ? 0 : 1, _ = 1 - B, Y = i[G[_]] + i[Q[_]];
    function rr(ur) {
      var P = [], Z = [];
      P[B] = Z[B] = ur, P[_] = i[G[_]], Z[_] = Y;
      var D = {
        x1: P[0],
        y1: P[1],
        x2: Z[0],
        y2: Z[1]
      };
      return dr(D, D, {
        lineWidth: 1
      }), P[0] = D.x1, P[1] = D.y1, P[B];
    }
    o = rr(o), c = rr(c);
    for (var K = [], V = [], N = !0, L = i[G[_]], j = 0; ; j++) {
      var fr = L === i[G[_]], F = L >= Y;
      F && (L = Y);
      var R = [], W = [];
      R[B] = o, W[B] = c, !fr && !F && (R[B] += N ? -u : u, W[B] -= N ? u : -u), R[_] = L, W[_] = L, K.push(R), V.push(W);
      var X = void 0;
      if (j < A.length ? X = A[j] : (X = Math.random(), A.push(X)), L += X * (p - d) + d, N = !N, F)
        break;
    }
    var U = C().serializeAxisBreakIdentifier(S.breakOption);
    if (x.add(new or({
      anid: "break_a_" + U,
      shape: {
        points: K
      },
      style: q,
      z: f
    })), S.gapReal !== 0) {
      x.add(new or({
        anid: "break_b_" + U,
        shape: {
          // Not reverse to keep the dash stable when dragging resizing.
          points: V
        },
        style: q,
        z: f
      }));
      var er = V.slice();
      er.reverse();
      var cr = K.concat(er);
      x.add(new Tr({
        anid: "break_c_" + U,
        shape: {
          points: cr
        },
        style: {
          fill: t,
          opacity: g.opacity
        },
        z: f
      }));
    }
  }
}
function Pr(l, a, r, i) {
  var e = l.axis, n = r.transform;
  mr(i.style);
  var s = e.getExtent();
  e.inverse && (s = s.slice(), s.reverse());
  var v = C().retrieveAxisBreakPairs(e.scale.getTicks({
    breakTicks: "only_break"
  }), function(m) {
    return m.break;
  }, !1), u = J(v, function(m) {
    var k = m[0].break.parsedBreak, t = [e.dataToCoord(k.vmin, !0), e.dataToCoord(k.vmax, !0)];
    return t[0] > t[1] && t.reverse(), {
      coordPair: t,
      brkId: C().serializeAxisBreakIdentifier(k.breakOption)
    };
  });
  u.sort(function(m, k) {
    return m.coordPair[0] - k.coordPair[0];
  });
  for (var d = s[0], p = null, T = 0; T < u.length; T++) {
    var f = u[T], E = Math.max(f.coordPair[0], s[0]), g = Math.min(f.coordPair[1], s[1]);
    d <= E && w(d, E, p, f), d = g, p = f;
  }
  d <= s[1] && w(d, s[1], p, null);
  function w(m, k, t, h) {
    function z(o, c) {
      n && (ar(o, o, n), ar(c, c, n));
    }
    function I(o, c) {
      var b = {
        x1: o[0],
        y1: o[1],
        x2: c[0],
        y2: c[1]
      };
      dr(b, b, i.style), o[0] = b.x1, o[1] = b.y1, c[0] = b.x2, c[1] = b.y2;
    }
    var O = [m, 0], y = [k, 0], H = [m, 5], A = [k, 5];
    z(O, H), I(O, H), z(y, A), I(y, A), I(O, y);
    var x = new _r(kr({
      shape: {
        x1: O[0],
        y1: O[1],
        x2: y[0],
        y2: y[1]
      }
    }, i));
    a.add(x), x.anid = "breakLine_" + (t ? t.brkId : "\0") + "_\0_" + (h ? h.brkId : "\0");
  }
}
function Mr(l, a, r) {
  if ($(r, function(t) {
    return !t;
  }))
    return;
  var i = new M();
  if (!Ar(r[0], r[1], i, {
    // Assert `labelPair` is `[break_min, break_max]`.
    // `axis.inverse: true` means a smaller scale value corresponds to a bigger value in axis.extent.
    // The axisRotation indicates mtv direction of OBB intersecting.
    direction: -(l ? a + Math.PI : a),
    touchThreshold: 0,
    // If need to resovle intersection align axis by moving labels according to MTV,
    // the direction must not be opposite, otherwise cause misleading.
    bidirectional: !1
  }))
    return;
  var e = ir();
  xr(e, e, -a);
  var n = J(r, function(t) {
    return t.transform ? br(ir(), e, t.transform) : e;
  });
  function s(t) {
    var h = r[0].localRect, z = new M(h[Q[t]] * n[0][0], h[Q[t]] * n[0][1]);
    return Math.abs(z.y) < 1e-5;
  }
  var v = 0.5;
  if (s(0) || s(1)) {
    var u = J(r, function(t, h) {
      var z = t.localRect.clone();
      return z.applyTransform(n[h]), z;
    }), d = new M();
    d.copy(r[0].label).add(r[1].label).scale(0.5), d.transform(e);
    var p = i.clone().transform(e), T = u[0].x + u[1].x + (p.x >= 0 ? u[0].width : u[1].width), f = (T + p.x) / 2 - d.x, E = Math.min(f, f - p.x), g = Math.max(f, f - p.x), w = g < 0 ? g : E > 0 ? E : 0;
    v = (f - w) / p.x;
  }
  var m = new M(), k = new M();
  M.scale(m, i, -v), M.scale(k, i, 1 - v), nr(r[0], m), nr(r[1], k);
}
function Cr(l, a) {
  var r = {
    breaks: []
  };
  return lr(a.breaks, function(i) {
    if (i) {
      var e = $(l.get("breaks", !0), function(v) {
        return C().identifyAxisBreak(v, i);
      });
      if (!e) {
        process.env.NODE_ENV !== "production" && yr("Can not find axis break by start: " + i.start + ", end: " + i.end);
        return;
      }
      var n = a.type, s = {
        isExpanded: !!e.isExpanded
      };
      e.isExpanded = n === vr ? !0 : n === gr ? !1 : n === hr ? !e.isExpanded : e.isExpanded, r.breaks.push({
        start: e.start,
        end: e.end,
        isExpanded: !!e.isExpanded,
        old: s
      });
    }
  }), r;
}
function Ur() {
  zr({
    adjustBreakLabelPair: Mr,
    buildAxisBreakLine: Pr,
    rectCoordBuildBreakAxis: Or,
    updateModelAxisBreak: Cr
  });
}
export {
  Ur as installAxisBreakHelper
};
