// echarts@6.0.0/lib/component/axis/CartesianAxisView.js
import { __extends as A } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { each as V, isArray as _, defaults as b } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { groupTransition as k, subPixelOptimizeLine as L } from "../../util/graphic.js";
import H from "./AxisView.js";
import { rectCoordAxisHandleRemove as R, rectCoordAxisBuildSplitArea as M } from "./axisSplitHelper.js";
import { getAxisBreakHelper as z } from "./axisBreakHelper.js";
import { shouldAxisShow as T } from "../../coord/axisHelper.js";
import P from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import B from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Line.js";
var I = ["splitArea", "splitLine", "minorSplitLine", "breakArea"], C = (
  /** @class */
  (function(t) {
    A(i, t);
    function i() {
      var e = t !== null && t.apply(this, arguments) || this;
      return e.type = i.type, e.axisPointerClass = "CartesianAxisPointer", e;
    }
    return i.prototype.render = function(e, u, v, r) {
      this.group.removeAll();
      var l = this._axisGroup;
      if (this._axisGroup = new P(), this.group.add(this._axisGroup), !!T(e)) {
        this._axisGroup.add(e.axis.axisBuilder.group), V(I, function(n) {
          e.get([n, "show"]) && O[n](this, this._axisGroup, e, e.getCoordSysModel(), v);
        }, this);
        var c = r && r.type === "changeAxisOrder" && r.isInitSort;
        c || k(l, this._axisGroup, e), t.prototype.render.call(this, e, u, v, r);
      }
    }, i.prototype.remove = function() {
      R(this);
    }, i.type = "cartesianAxis", i;
  })(H)
), O = {
  splitLine: function(t, i, e, u, v) {
    var r = e.axis;
    if (!r.scale.isBlank()) {
      var l = e.getModel("splitLine"), c = l.getModel("lineStyle"), n = c.get("color"), m = l.get("showMinLine") !== !1, x = l.get("showMaxLine") !== !1;
      n = _(n) ? n : [n];
      for (var a = u.coordinateSystem.getRect(), h = r.isHorizontal(), y = 0, p = r.getTicksCoords({
        tickModel: l,
        breakTicks: "none",
        pruneByBreak: "preserve_extent_bound"
      }), o = [], s = [], f = c.getLineStyle(), d = 0; d < p.length; d++) {
        var g = r.toGlobalCoord(p[d].coord);
        if (!(d === 0 && !m || d === p.length - 1 && !x)) {
          var w = p[d].tickValue;
          h ? (o[0] = g, o[1] = a.y, s[0] = g, s[1] = a.y + a.height) : (o[0] = a.x, o[1] = g, s[0] = a.x + a.width, s[1] = g);
          var G = y++ % n.length, S = new B({
            anid: w != null ? "line_" + w : null,
            autoBatch: !0,
            shape: {
              x1: o[0],
              y1: o[1],
              x2: s[0],
              y2: s[1]
            },
            style: b({
              stroke: n[G]
            }, f),
            silent: !0
          });
          L(S.shape, f.lineWidth), i.add(S);
        }
      }
    }
  },
  minorSplitLine: function(t, i, e, u, v) {
    var r = e.axis, l = e.getModel("minorSplitLine"), c = l.getModel("lineStyle"), n = u.coordinateSystem.getRect(), m = r.isHorizontal(), x = r.getMinorTicksCoords();
    if (x.length)
      for (var a = [], h = [], y = c.getLineStyle(), p = 0; p < x.length; p++)
        for (var o = 0; o < x[p].length; o++) {
          var s = r.toGlobalCoord(x[p][o].coord);
          m ? (a[0] = s, a[1] = n.y, h[0] = s, h[1] = n.y + n.height) : (a[0] = n.x, a[1] = s, h[0] = n.x + n.width, h[1] = s);
          var f = new B({
            anid: "minor_line_" + x[p][o].tickValue,
            autoBatch: !0,
            shape: {
              x1: a[0],
              y1: a[1],
              x2: h[0],
              y2: h[1]
            },
            style: y,
            silent: !0
          });
          L(f.shape, y.lineWidth), i.add(f);
        }
  },
  splitArea: function(t, i, e, u, v) {
    M(t, i, e, u);
  },
  breakArea: function(t, i, e, u, v) {
    var r = z(), l = e.axis.scale;
    r && l.type !== "ordinal" && r.rectCoordBuildBreakAxis(i, t, e, u.coordinateSystem.getRect(), v);
  }
}, W = (
  /** @class */
  (function(t) {
    A(i, t);
    function i() {
      var e = t !== null && t.apply(this, arguments) || this;
      return e.type = i.type, e;
    }
    return i.type = "xAxis", i;
  })(C)
), N = (
  /** @class */
  (function(t) {
    A(i, t);
    function i() {
      var e = t !== null && t.apply(this, arguments) || this;
      return e.type = W.type, e;
    }
    return i.type = "yAxis", i;
  })(C)
);
export {
  W as CartesianXAxisView,
  N as CartesianYAxisView,
  C as default
};
