// echarts@6.0.0/lib/component/axis/ParallelAxisView.js
import { __extends as x } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { bind as y, extend as v, map as d } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import g from "./AxisBuilder.js";
import C from "../helper/BrushController.js";
import { makeLinearBrushOtherExtent as A, makeRectIsTargetByCursor as B, makeRectPanelClipPath as b } from "../helper/brushHelper.js";
import { groupTransition as S } from "../../util/graphic.js";
import _ from "../../view/Component.js";
import T from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import w from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/BoundingRect.js";
var Z = (
  /** @class */
  (function(e) {
    x(t, e);
    function t() {
      var r = e !== null && e.apply(this, arguments) || this;
      return r.type = t.type, r;
    }
    return t.prototype.init = function(r, i) {
      e.prototype.init.apply(this, arguments), (this._brushController = new C(i.getZr())).on("brush", y(this._onBrush, this));
    }, t.prototype.render = function(r, i, o, s) {
      if (!G(r, i, s)) {
        this.axisModel = r, this.api = o, this.group.removeAll();
        var l = this._axisGroup;
        if (this._axisGroup = new T(), this.group.add(this._axisGroup), !!r.get("show")) {
          var n = L(r, i), u = n.coordinateSystem, p = r.getAreaSelectStyle(), h = p.width, a = r.axis.dim, f = u.getAxisLayout(a), m = v({
            strokeContainThreshold: h
          }, f), c = new g(r, o, m);
          c.build(), this._axisGroup.add(c.group), this._refreshBrushController(m, p, r, n, h, o), S(l, this._axisGroup, r);
        }
      }
    }, t.prototype._refreshBrushController = function(r, i, o, s, l, n) {
      var u = o.axis.getExtent(), p = u[1] - u[0], h = Math.min(30, Math.abs(p) * 0.1), a = w.create({
        x: u[0],
        y: -l / 2,
        width: p,
        height: l
      });
      a.x -= h, a.width += 2 * h, this._brushController.mount({
        enableGlobalPan: !0,
        rotation: r.rotation,
        x: r.position[0],
        y: r.position[1]
      }).setPanels([{
        panelId: "pl",
        clipPath: b(a),
        isTargetByCursor: B(a, n, s),
        getLinearBrushOtherExtent: A(a, 0)
      }]).enableBrush({
        brushType: "lineX",
        brushStyle: i,
        removeOnClick: !0
      }).updateCovers(I(o));
    }, t.prototype._onBrush = function(r) {
      var i = r.areas, o = this.axisModel, s = o.axis, l = d(i, function(n) {
        return [s.coordToData(n.range[0], !0), s.coordToData(n.range[1], !0)];
      });
      (!o.option.realtime === r.isEnd || r.removeOnClick) && this.api.dispatchAction({
        type: "axisAreaSelect",
        parallelAxisId: o.id,
        intervals: l
      });
    }, t.prototype.dispose = function() {
      this._brushController.dispose();
    }, t.type = "parallelAxis", t;
  })(_)
);
function G(e, t, r) {
  return r && r.type === "axisAreaSelect" && t.findComponents({
    mainType: "parallelAxis",
    query: r
  })[0] === e;
}
function I(e) {
  var t = e.axis;
  return d(e.activeIntervals, function(r) {
    return {
      brushType: "lineX",
      panelId: "pl",
      range: [t.dataToCoord(r[0], !0), t.dataToCoord(r[1], !0)]
    };
  });
}
function L(e, t) {
  return t.getComponent("parallel", e.get("parallelIndex"));
}
export {
  Z as default
};
