// echarts@6.0.0/lib/component/axis/RadiusAxisView.js
import { __extends as y } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { each as m, defaults as d } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import * as M from "../../util/graphic.js";
import { groupTransition as S, mergePath as A } from "../../util/graphic.js";
import w from "./AxisBuilder.js";
import L from "./AxisView.js";
import E from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import P from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Sector.js";
import b from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Circle.js";
var B = ["splitLine", "splitArea", "minorSplitLine"], q = (
  /** @class */
  (function(l) {
    y(a, l);
    function a() {
      var e = l !== null && l.apply(this, arguments) || this;
      return e.type = a.type, e.axisPointerClass = "PolarAxisPointer", e;
    }
    return a.prototype.render = function(e, x, f) {
      if (this.group.removeAll(), !!e.get("show")) {
        var s = this._axisGroup, o = this._axisGroup = new E();
        this.group.add(o);
        var n = e.axis, t = n.polar, u = t.getAngleAxis(), i = n.getTicksCoords(), g = n.getMinorTicksCoords(), r = u.getExtent()[0], v = n.getExtent(), c = R(t, e, r), p = new w(e, f, c);
        p.build(), o.add(p.group), S(s, o, e), m(B, function(h) {
          e.get([h, "show"]) && !n.scale.isBlank() && G[h](this.group, e, t, r, v, i, g);
        }, this);
      }
    }, a.type = "radiusAxis", a;
  })(L)
), G = {
  splitLine: function(l, a, e, x, f, s) {
    var o = a.getModel("splitLine"), n = o.getModel("lineStyle"), t = n.get("color"), u = 0, i = e.getAngleAxis(), g = Math.PI / 180, r = i.getExtent(), v = Math.abs(r[1] - r[0]) === 360 ? "Circle" : "Arc";
    t = t instanceof Array ? t : [t];
    for (var c = [], p = 0; p < s.length; p++) {
      var h = u++ % t.length;
      c[h] = c[h] || [], c[h].push(new M[v]({
        shape: {
          cx: e.cx,
          cy: e.cy,
          // ensure circle radius >= 0
          r: Math.max(s[p].coord, 0),
          startAngle: -r[0] * g,
          endAngle: -r[1] * g,
          clockwise: i.inverse
        }
      }));
    }
    for (var p = 0; p < c.length; p++)
      l.add(A(c[p], {
        style: d({
          stroke: t[p % t.length],
          fill: null
        }, n.getLineStyle()),
        silent: !0
      }));
  },
  minorSplitLine: function(l, a, e, x, f, s, o) {
    if (o.length) {
      for (var n = a.getModel("minorSplitLine"), t = n.getModel("lineStyle"), u = [], i = 0; i < o.length; i++)
        for (var g = 0; g < o[i].length; g++)
          u.push(new b({
            shape: {
              cx: e.cx,
              cy: e.cy,
              r: o[i][g].coord
            }
          }));
      l.add(A(u, {
        style: d({
          fill: null
        }, t.getLineStyle()),
        silent: !0
      }));
    }
  },
  splitArea: function(l, a, e, x, f, s) {
    if (s.length) {
      var o = a.getModel("splitArea"), n = o.getModel("areaStyle"), t = n.get("color"), u = 0;
      t = t instanceof Array ? t : [t];
      for (var i = [], g = s[0].coord, r = 1; r < s.length; r++) {
        var v = u++ % t.length;
        i[v] = i[v] || [], i[v].push(new P({
          shape: {
            cx: e.cx,
            cy: e.cy,
            r0: g,
            r: s[r].coord,
            startAngle: 0,
            endAngle: Math.PI * 2
          },
          silent: !0
        })), g = s[r].coord;
      }
      for (var r = 0; r < i.length; r++)
        l.add(A(i[r], {
          style: d({
            fill: t[r % t.length]
          }, n.getAreaStyle()),
          silent: !0
        }));
    }
  }
};
function R(l, a, e) {
  return {
    position: [l.cx, l.cy],
    rotation: e / 180 * Math.PI,
    labelDirection: -1,
    tickDirection: -1,
    nameDirection: 1,
    labelRotate: a.getModel("axisLabel").get("rotate"),
    // Over splitLine and splitArea
    z2: 1
  };
}
export {
  q as default
};
