// echarts@6.0.0/lib/component/axis/SingleAxisView.js
import { __extends as k } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { each as w, defaults as G } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import S from "./AxisBuilder.js";
import { groupTransition as L, subPixelOptimizeLine as b, mergePath as _ } from "../../util/graphic.js";
import { layout as C } from "../../coord/single/singleAxisHelper.js";
import H from "./AxisView.js";
import { rectCoordAxisHandleRemove as P, rectCoordAxisBuildSplitArea as R } from "./axisSplitHelper.js";
import { getAxisBreakHelper as V } from "./axisBreakHelper.js";
import z from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import T from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Line.js";
var E = ["splitArea", "splitLine", "breakArea"], U = (
  /** @class */
  (function(i) {
    k(t, i);
    function t() {
      var r = i !== null && i.apply(this, arguments) || this;
      return r.type = t.type, r.axisPointerClass = "SingleAxisPointer", r;
    }
    return t.prototype.render = function(r, e, p, o) {
      var n = this.group;
      n.removeAll();
      var h = this._axisGroup;
      this._axisGroup = new z();
      var a = C(r), g = new S(r, p, a);
      g.build(), n.add(this._axisGroup), n.add(g.group), w(E, function(s) {
        r.get([s, "show"]) && O[s](this, this.group, this._axisGroup, r, p);
      }, this), L(h, this._axisGroup, r), i.prototype.render.call(this, r, e, p, o);
    }, t.prototype.remove = function() {
      P(this);
    }, t.type = "singleAxis", t;
  })(H)
), O = {
  splitLine: function(i, t, r, e, p) {
    var o = e.axis;
    if (!o.scale.isBlank()) {
      var n = e.getModel("splitLine"), h = n.getModel("lineStyle"), a = h.get("color");
      a = a instanceof Array ? a : [a];
      for (var g = h.get("width"), s = e.coordinateSystem.getRect(), x = o.isHorizontal(), f = [], A = 0, c = o.getTicksCoords({
        tickModel: n,
        breakTicks: "none",
        pruneByBreak: "preserve_extent_bound"
      }), u = [], v = [], l = 0; l < c.length; ++l) {
        var m = o.toGlobalCoord(c[l].coord);
        x ? (u[0] = m, u[1] = s.y, v[0] = m, v[1] = s.y + s.height) : (u[0] = s.x, u[1] = m, v[0] = s.x + s.width, v[1] = m);
        var y = new T({
          shape: {
            x1: u[0],
            y1: u[1],
            x2: v[0],
            y2: v[1]
          },
          silent: !0
        });
        b(y.shape, g);
        var d = A++ % a.length;
        f[d] = f[d] || [], f[d].push(y);
      }
      for (var B = h.getLineStyle(["color"]), l = 0; l < f.length; ++l)
        t.add(_(f[l], {
          style: G({
            stroke: a[l % a.length]
          }, B),
          silent: !0
        }));
    }
  },
  splitArea: function(i, t, r, e, p) {
    R(i, r, e, e);
  },
  breakArea: function(i, t, r, e, p) {
    var o = V(), n = e.axis.scale;
    o && n.type !== "ordinal" && o.rectCoordBuildBreakAxis(t, i, e, e.coordinateSystem.getRect(), p);
  }
};
export {
  U as default
};
