// echarts@6.0.0/lib/component/toolbox/feature/DataView.js
import { __extends as N } from "../../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { registerAction as z } from "../../../core/echarts.js";
import { isFunction as A, isString as F, isDom as j, each as O, extend as G, defaults as R, filter as P, map as L, isObject as M, isArray as B } from "../../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { ToolboxFeature as V } from "../featureManager.js";
import { addEventListener as H } from "../../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/event.js";
import { warn as U } from "../../../util/log.js";
import g from "../../../visual/tokens.js";
var D = new Array(60).join("-"), h = "	";
function W(i) {
  var r = {}, e = [], t = [];
  return i.eachRawSeries(function(n) {
    var o = n.coordinateSystem;
    if (o && (o.type === "cartesian2d" || o.type === "polar")) {
      var a = o.getBaseAxis();
      if (a.type === "category") {
        var s = a.dim + "_" + a.index;
        r[s] || (r[s] = {
          categoryAxis: a,
          valueAxis: o.getOtherAxis(a),
          series: []
        }, t.push({
          axisDim: a.dim,
          axisIndex: a.index
        })), r[s].series.push(n);
      } else
        e.push(n);
    } else
      e.push(n);
  }), {
    seriesGroupByCategoryAxis: r,
    other: e,
    meta: t
  };
}
function $(i) {
  var r = [];
  return O(i, function(e, t) {
    var n = e.categoryAxis, o = e.valueAxis, a = o.dim, s = [" "].concat(L(e.series, function(v) {
      return v.name;
    })), l = [n.model.getCategories()];
    O(e.series, function(v) {
      var f = v.getRawData();
      l.push(v.getRawData().mapArray(f.mapDimension(a), function(m) {
        return m;
      }));
    });
    for (var u = [s.join(h)], c = 0; c < l[0].length; c++) {
      for (var p = [], d = 0; d < l.length; d++)
        p.push(l[d][c]);
      u.push(p.join(h));
    }
    r.push(u.join(`
`));
  }), r.join(`

` + D + `

`);
}
function q(i) {
  return L(i, function(r) {
    var e = r.getRawData(), t = [r.name], n = [];
    return e.each(e.dimensions, function() {
      for (var o = arguments.length, a = arguments[o - 1], s = e.getName(a), l = 0; l < o - 1; l++)
        n[l] = arguments[l];
      t.push((s ? s + h : "") + n.join(h));
    }), t.join(`
`);
  }).join(`

` + D + `

`);
}
function J(i) {
  var r = W(i);
  return {
    value: P([$(r.seriesGroupByCategoryAxis), q(r.other)], function(e) {
      return !!e.replace(/[\n\t\s]/g, "");
    }).join(`

` + D + `

`),
    meta: r.meta
  };
}
function w(i) {
  return i.replace(/^\s\s*/, "").replace(/\s\s*$/, "");
}
function Q(i) {
  var r = i.slice(0, i.indexOf(`
`));
  if (r.indexOf(h) >= 0)
    return !0;
}
var S = new RegExp("[" + h + "]+", "g");
function X(i) {
  for (var r = i.split(/\n+/g), e = w(r.shift()).split(S), t = [], n = L(e, function(l) {
    return {
      name: l,
      data: []
    };
  }), o = 0; o < r.length; o++) {
    var a = w(r[o]).split(S);
    t.push(a.shift());
    for (var s = 0; s < a.length; s++)
      n[s] && (n[s].data[o] = a[s]);
  }
  return {
    series: n,
    categories: t
  };
}
function Y(i) {
  for (var r = i.split(/\n+/g), e = w(r.shift()), t = [], n = 0; n < r.length; n++) {
    var o = w(r[n]);
    if (o) {
      var a = o.split(S), s = "", l = void 0, u = !1;
      isNaN(a[0]) ? (u = !0, s = a[0], a = a.slice(1), t[n] = {
        name: s,
        value: []
      }, l = t[n].value) : l = t[n] = [];
      for (var c = 0; c < a.length; c++)
        l.push(+a[c]);
      l.length === 1 && (u ? t[n].value = l[0] : t[n] = l[0]);
    }
  }
  return {
    name: e,
    data: t
  };
}
function Z(i, r) {
  var e = i.split(new RegExp(`
*` + D + `
*`, "g")), t = {
    series: []
  };
  return O(e, function(n, o) {
    if (Q(n)) {
      var a = X(n), s = r[o], l = s.axisDim + "Axis";
      s && (t[l] = t[l] || [], t[l][s.axisIndex] = {
        data: a.categories
      }, t.series = t.series.concat(a.series));
    } else {
      var a = Y(n);
      t.series.push(a);
    }
  }), t;
}
var se = (
  /** @class */
  (function(i) {
    N(r, i);
    function r() {
      return i !== null && i.apply(this, arguments) || this;
    }
    return r.prototype.onclick = function(e, t) {
      setTimeout(function() {
        t.dispatchAction({
          type: "hideTip"
        });
      });
      var n = t.getDom(), o = this.model;
      this._dom && n.removeChild(this._dom);
      var a = document.createElement("div");
      a.style.cssText = "position:absolute;top:0;bottom:0;left:0;right:0;padding:5px", a.style.backgroundColor = o.get("backgroundColor") || g.color.neutral00;
      var s = document.createElement("h4"), l = o.get("lang") || [];
      s.innerHTML = l[0] || o.get("title"), s.style.cssText = "margin:10px 20px", s.style.color = o.get("textColor");
      var u = document.createElement("div"), c = document.createElement("textarea");
      u.style.cssText = "overflow:auto";
      var p = o.get("optionToContent"), d = o.get("contentToOption"), v = J(e);
      if (A(p)) {
        var f = p(t.getOption());
        F(f) ? u.innerHTML = f : j(f) && u.appendChild(f);
      } else {
        c.readOnly = o.get("readOnly");
        var m = c.style;
        m.cssText = "display:block;width:100%;height:100%;font-family:monospace;font-size:14px;line-height:1.6rem;resize:none;box-sizing:border-box;outline:none", m.color = o.get("textColor"), m.borderColor = o.get("textareaBorderColor"), m.backgroundColor = o.get("textareaColor"), c.value = v.value, u.appendChild(c);
      }
      var _ = v.meta, x = document.createElement("div");
      x.style.cssText = "position:absolute;bottom:5px;left:0;right:0";
      var E = "float:right;margin-right:20px;border:none;cursor:pointer;padding:2px 5px;font-size:12px;border-radius:3px", C = document.createElement("div"), y = document.createElement("div");
      E += ";background-color:" + o.get("buttonColor"), E += ";color:" + o.get("buttonTextColor");
      var k = this;
      function T() {
        n.removeChild(a), k._dom = null;
      }
      H(C, "click", T), H(y, "click", function() {
        if (d == null && p != null || d != null && p == null) {
          process.env.NODE_ENV !== "production" && U("It seems you have just provided one of `contentToOption` and `optionToContent` functions but missed the other one. Data change is ignored."), T();
          return;
        }
        var b;
        try {
          A(d) ? b = d(u, t.getOption()) : b = Z(c.value, _);
        } catch (I) {
          throw T(), new Error("Data view format error " + I);
        }
        b && t.dispatchAction({
          type: "changeDataView",
          newOption: b
        }), T();
      }), C.innerHTML = l[1], y.innerHTML = l[2], y.style.cssText = C.style.cssText = E, !o.get("readOnly") && x.appendChild(y), x.appendChild(C), a.appendChild(s), a.appendChild(u), a.appendChild(x), u.style.height = n.clientHeight - 80 + "px", n.appendChild(a), this._dom = a;
    }, r.prototype.remove = function(e, t) {
      this._dom && t.getDom().removeChild(this._dom);
    }, r.prototype.dispose = function(e, t) {
      this.remove(e, t);
    }, r.getDefaultOption = function(e) {
      var t = {
        show: !0,
        readOnly: !1,
        optionToContent: null,
        contentToOption: null,
        // eslint-disable-next-line
        icon: "M17.5,17.3H33 M17.5,17.3H33 M45.4,29.5h-28 M11.5,2v56H51V14.8L38.4,2H11.5z M38.4,2.2v12.7H51 M45.4,41.7h-28",
        title: e.getLocaleModel().get(["toolbox", "dataView", "title"]),
        lang: e.getLocaleModel().get(["toolbox", "dataView", "lang"]),
        backgroundColor: g.color.background,
        textColor: g.color.primary,
        textareaColor: g.color.background,
        textareaBorderColor: g.color.border,
        buttonColor: g.color.accent50,
        buttonTextColor: g.color.neutral00
      };
      return t;
    }, r;
  })(V)
);
function K(i, r) {
  return L(i, function(e, t) {
    var n = r && r[t];
    if (M(n) && !B(n)) {
      var o = M(e) && !B(e);
      o || (e = {
        value: e
      });
      var a = n.name != null && e.name == null;
      return e = R(e, n), a && delete e.name, e;
    } else
      return e;
  });
}
z({
  type: "changeDataView",
  event: "dataViewChanged",
  update: "prepareAndUpdate"
}, function(i, r) {
  var e = [];
  O(i.newOption.series, function(t) {
    var n = r.getSeriesByName(t.name)[0];
    if (!n)
      e.push(G({
        // Default is scatter
        type: "scatter"
      }, t));
    else {
      var o = n.get("data");
      e.push({
        name: t.name,
        data: K(t.data, o)
      });
    }
  }), r.mergeOption(R({
    series: e
  }, i.newOption));
});
export {
  se as default
};
