// echarts@6.0.0/lib/component/toolbox/feature/DataZoom.js
import { __extends as b } from "../../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { bind as y, each as g, clone as I } from "../../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import Z from "../../helper/BrushController.js";
import v from "../../helper/BrushTargetManager.js";
import { push as _, pop as M, count as C } from "../../dataZoom/history.js";
import S from "../../helper/sliderMove.js";
import { ToolboxFeature as k } from "../featureManager.js";
import { parseFinder as T, makeInternalComponentId as D } from "../../../util/model.js";
import { registerInternalOptionCreator as B } from "../../../model/internalComponentCreator.js";
import O from "../../../visual/tokens.js";
var f = g, V = D("toolbox-dataZoom_"), j = (
  /** @class */
  (function(o) {
    b(t, o);
    function t() {
      return o !== null && o.apply(this, arguments) || this;
    }
    return t.prototype.render = function(n, e, r, i) {
      this._brushController || (this._brushController = new Z(r.getZr()), this._brushController.on("brush", y(this._onBrush, this)).mount()), E(n, e, this, i, r), w(n, e);
    }, t.prototype.onclick = function(n, e, r) {
      H[r].call(this);
    }, t.prototype.remove = function(n, e) {
      this._brushController && this._brushController.unmount();
    }, t.prototype.dispose = function(n, e) {
      this._brushController && this._brushController.dispose();
    }, t.prototype._onBrush = function(n) {
      var e = n.areas;
      if (!n.isEnd || !e.length)
        return;
      var r = {}, i = this.ecModel;
      this._brushController.updateCovers([]);
      var h = new v(A(this.model), i, {
        include: ["grid"]
      });
      h.matchOutputRanges(e, i, function(d, c, a) {
        if (a.type === "cartesian2d") {
          var l = d.brushType;
          l === "rect" ? (u("x", a, c[0]), u("y", a, c[1])) : u({
            lineX: "x",
            lineY: "y"
          }[l], a, c);
        }
      }), _(i, r), this._dispatchZoomAction(r);
      function u(d, c, a) {
        var l = c.getAxis(d), m = l.model, p = s(d, m, i), x = p.findRepresentativeAxisProxy(m).getMinMaxSpan();
        (x.minValueSpan != null || x.maxValueSpan != null) && (a = S(0, a.slice(), l.scale.getExtent(), 0, x.minValueSpan, x.maxValueSpan)), p && (r[p.id] = {
          dataZoomId: p.id,
          startValue: a[0],
          endValue: a[1]
        });
      }
      function s(d, c, a) {
        var l;
        return a.eachComponent({
          mainType: "dataZoom",
          subType: "select"
        }, function(m) {
          var p = m.getAxisModel(d, c.componentIndex);
          p && (l = m);
        }), l;
      }
    }, t.prototype._dispatchZoomAction = function(n) {
      var e = [];
      f(n, function(r, i) {
        e.push(I(r));
      }), e.length && this.api.dispatchAction({
        type: "dataZoom",
        from: this.uid,
        batch: e
      });
    }, t.getDefaultOption = function(n) {
      var e = {
        show: !0,
        filterMode: "filter",
        // Icon group
        icon: {
          zoom: "M0,13.5h26.9 M13.5,26.9V0 M32.1,13.5H58V58H13.5 V32.1",
          back: "M22,1.4L9.9,13.5l12.3,12.3 M10.3,13.5H54.9v44.6 H10.3v-26"
        },
        // `zoom`, `back`
        title: n.getLocaleModel().get(["toolbox", "dataZoom", "title"]),
        brushStyle: {
          borderWidth: 0,
          color: O.color.backgroundTint
        }
      };
      return e;
    }, t;
  })(k)
), H = {
  zoom: function() {
    var o = !this._isZoomActive;
    this.api.dispatchAction({
      type: "takeGlobalCursor",
      key: "dataZoomSelect",
      dataZoomSelectActive: o
    });
  },
  back: function() {
    this._dispatchZoomAction(M(this.ecModel));
  }
};
function A(o) {
  var t = {
    xAxisIndex: o.get("xAxisIndex", !0),
    yAxisIndex: o.get("yAxisIndex", !0),
    xAxisId: o.get("xAxisId", !0),
    yAxisId: o.get("yAxisId", !0)
  };
  return t.xAxisIndex == null && t.xAxisId == null && (t.xAxisIndex = "all"), t.yAxisIndex == null && t.yAxisId == null && (t.yAxisIndex = "all"), t;
}
function w(o, t) {
  o.setIconStatus("back", C(t) > 1 ? "emphasis" : "normal");
}
function E(o, t, n, e, r) {
  var i = n._isZoomActive;
  e && e.type === "takeGlobalCursor" && (i = e.key === "dataZoomSelect" ? e.dataZoomSelectActive : !1), n._isZoomActive = i, o.setIconStatus("zoom", i ? "emphasis" : "normal");
  var h = new v(A(o), t, {
    include: ["grid"]
  }), u = h.makePanelOpts(r, function(s) {
    return s.xAxisDeclared && !s.yAxisDeclared ? "lineX" : !s.xAxisDeclared && s.yAxisDeclared ? "lineY" : "rect";
  });
  n._brushController.setPanels(u).enableBrush(i && u.length ? {
    brushType: "auto",
    brushStyle: o.getModel("brushStyle").getItemStyle()
  } : !1);
}
B("dataZoom", function(o) {
  var t = o.getComponent("toolbox", 0), n = ["feature", "dataZoom"];
  if (!t || t.get(n) == null)
    return;
  var e = t.getModel(n), r = [], i = A(e), h = T(o, i);
  f(h.xAxisModels, function(s) {
    return u(s, "xAxis", "xAxisIndex");
  }), f(h.yAxisModels, function(s) {
    return u(s, "yAxis", "yAxisIndex");
  });
  function u(s, d, c) {
    var a = s.componentIndex, l = {
      type: "select",
      $fromToolbox: !0,
      // Default to be filter
      filterMode: e.get("filterMode", !0) || "filter",
      // Id for merge mapping.
      id: V + d + a
    };
    l[c] = a, r.push(l);
  }
  return r;
});
export {
  j as default
};
