// echarts@6.0.0/lib/component/toolbox/feature/MagicType.js
import { __extends as _ } from "../../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { registerAction as w } from "../../../core/echarts.js";
import { each as f, indexOf as z, merge as p, defaults as N } from "../../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { ToolboxFeature as R } from "../featureManager.js";
import { SINGLE_REFERRING as A } from "../../../util/model.js";
var b = "__ec_magicType_stack__", E = [["line", "bar"], ["stack"]], D = (
  /** @class */
  (function(r) {
    _(n, r);
    function n() {
      return r !== null && r.apply(this, arguments) || this;
    }
    return n.prototype.getIcons = function() {
      var t = this.model, a = t.get("icon"), e = {};
      return f(t.get("type"), function(i) {
        a[i] && (e[i] = a[i]);
      }), e;
    }, n.getDefaultOption = function(t) {
      var a = {
        show: !0,
        type: [],
        // Icon group
        icon: {
          line: "M4.1,28.9h7.1l9.3-22l7.4,38l9.7-19.7l3,12.8h14.9M4.1,58h51.4",
          bar: "M6.7,22.9h10V48h-10V22.9zM24.9,13h10v35h-10V13zM43.2,2h10v46h-10V2zM3.1,58h53.7",
          // eslint-disable-next-line
          stack: "M8.2,38.4l-8.4,4.1l30.6,15.3L60,42.5l-8.1-4.1l-21.5,11L8.2,38.4z M51.9,30l-8.1,4.2l-13.4,6.9l-13.9-6.9L8.2,30l-8.4,4.2l8.4,4.2l22.2,11l21.5-11l8.1-4.2L51.9,30z M51.9,21.7l-8.1,4.2L35.7,30l-5.3,2.8L24.9,30l-8.4-4.1l-8.3-4.2l-8.4,4.2L8.2,30l8.3,4.2l13.9,6.9l13.4-6.9l8.1-4.2l8.1-4.1L51.9,21.7zM30.4,2.2L-0.2,17.5l8.4,4.1l8.3,4.2l8.4,4.2l5.5,2.7l5.3-2.7l8.1-4.2l8.1-4.2l8.1-4.1L30.4,2.2z"
          // jshint ignore:line
        },
        // `line`, `bar`, `stack`, `tiled`
        title: t.getLocaleModel().get(["toolbox", "magicType", "title"]),
        option: {},
        seriesIndex: {}
      };
      return a;
    }, n.prototype.onclick = function(t, a, e) {
      var i = this.model, v = i.get(["seriesIndex", e]);
      if (T[e]) {
        var l = {
          series: []
        }, S = function(o) {
          var m = o.subType, y = o.id, s = T[e](m, y, o, i);
          s && (N(s, o.option), l.series.push(s));
          var g = o.coordinateSystem;
          if (g && g.type === "cartesian2d" && (e === "line" || e === "bar")) {
            var d = g.getAxesByScale("ordinal")[0];
            if (d) {
              var I = d.dim, c = I + "Axis", O = o.getReferringComponents(c, A).models[0], u = O.componentIndex;
              l[c] = l[c] || [];
              for (var L = 0; L <= u; L++)
                l[c][u] = l[c][u] || {};
              l[c][u].boundaryGap = e === "bar";
            }
          }
        };
        f(E, function(o) {
          z(o, e) >= 0 && f(o, function(m) {
            i.setIconStatus(m, "normal");
          });
        }), i.setIconStatus(e, "emphasis"), t.eachComponent({
          mainType: "series",
          query: v == null ? null : {
            seriesIndex: v
          }
        }, S);
        var h, k = e;
        e === "stack" && (h = p({
          stack: i.option.title.tiled,
          tiled: i.option.title.stack
        }, i.option.title), i.get(["iconStatus", e]) !== "emphasis" && (k = "tiled")), a.dispatchAction({
          type: "changeMagicType",
          currentType: k,
          newOption: l,
          newTitle: h,
          featureName: "magicType"
        });
      }
    }, n;
  })(R)
), T = {
  line: function(r, n, t, a) {
    if (r === "bar")
      return p({
        id: n,
        type: "line",
        // Preserve data related option
        data: t.get("data"),
        stack: t.get("stack"),
        markPoint: t.get("markPoint"),
        markLine: t.get("markLine")
      }, a.get(["option", "line"]) || {}, !0);
  },
  bar: function(r, n, t, a) {
    if (r === "line")
      return p({
        id: n,
        type: "bar",
        // Preserve data related option
        data: t.get("data"),
        stack: t.get("stack"),
        markPoint: t.get("markPoint"),
        markLine: t.get("markLine")
      }, a.get(["option", "bar"]) || {}, !0);
  },
  stack: function(r, n, t, a) {
    var e = t.get("stack") === b;
    if (r === "line" || r === "bar")
      return a.setIconStatus("stack", e ? "normal" : "emphasis"), p({
        id: n,
        stack: e ? "" : b
      }, a.get(["option", "stack"]) || {}, !0);
  }
};
w({
  type: "changeMagicType",
  event: "magicTypeChanged",
  update: "prepareAndUpdate"
}, function(r, n) {
  n.mergeOption(r.newOption);
});
export {
  D as default
};
