// echarts@6.0.0/lib/component/toolbox/feature/SaveAsImage.js
import { __extends as A } from "../../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import B from "../../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/env.js";
import { ToolboxFeature as S } from "../featureManager.js";
import C from "../../../visual/tokens.js";
var U = (
  /** @class */
  (function(d) {
    A(n, d);
    function n() {
      return d !== null && d.apply(this, arguments) || this;
    }
    return n.prototype.onclick = function(o, a) {
      var e = this.model, c = e.get("name") || o.get("title.0.text") || "echarts", g = a.getZr().painter.getType() === "svg", v = g ? "svg" : e.get("type", !0) || "png", u = a.getConnectedDataURL({
        type: v,
        backgroundColor: e.get("backgroundColor", !0) || o.get("backgroundColor") || C.color.neutral00,
        connectedBackgroundColor: e.get("connectedBackgroundColor"),
        excludeComponents: e.get("excludeComponents"),
        pixelRatio: e.get("pixelRatio")
      }), m = B.browser;
      if (typeof MouseEvent == "function" && (m.newEdge || !m.ie && !m.edge)) {
        var r = document.createElement("a");
        r.download = c + "." + v, r.target = "_blank", r.href = u;
        var h = new MouseEvent("click", {
          // some micro front-end framework， window maybe is a Proxy
          view: document.defaultView,
          bubbles: !0,
          cancelable: !1
        });
        r.dispatchEvent(h);
      } else if (window.navigator.msSaveOrOpenBlob || g) {
        var s = u.split(","), k = s[0].indexOf("base64") > -1, t = g ? decodeURIComponent(s[1]) : s[1];
        k && (t = window.atob(t));
        var w = c + "." + v;
        if (window.navigator.msSaveOrOpenBlob) {
          for (var l = t.length, b = new Uint8Array(l); l--; )
            b[l] = t.charCodeAt(l);
          var L = new Blob([b]);
          window.navigator.msSaveOrOpenBlob(L, w);
        } else {
          var p = document.createElement("iframe");
          document.body.appendChild(p);
          var f = p.contentWindow, i = f.document;
          i.open("image/svg+xml", "replace"), i.write(t), i.close(), f.focus(), i.execCommand("SaveAs", !0, w), document.body.removeChild(p);
        }
      } else {
        var x = e.get("lang"), O = '<body style="margin:0;"><img src="' + u + '" style="max-width:100%;" title="' + (x && x[0] || "") + '" /></body>', y = window.open();
        y.document.write(O), y.document.title = c;
      }
    }, n.getDefaultOption = function(o) {
      var a = {
        show: !0,
        icon: "M4.7,22.9L29.3,45.5L54.7,23.4M4.6,43.6L4.6,58L53.8,58L53.8,43.6M29.2,45.1L29.2,0",
        title: o.getLocaleModel().get(["toolbox", "saveAsImage", "title"]),
        type: "png",
        // Default use option.backgroundColor
        // backgroundColor: '#fff',
        connectedBackgroundColor: C.color.neutral00,
        name: "",
        excludeComponents: ["toolbox"],
        // use current pixel ratio of device by default
        // pixelRatio: 1,
        lang: o.getLocaleModel().get(["toolbox", "saveAsImage", "lang"])
      };
      return a;
    }, n;
  })(S)
);
export {
  U as default
};
