// zrender@6.0.0/lib/graphic/helper/image.js
import h from "../../core/LRU.js";
import { platformApi as u } from "../../core/platform.js";
var f = new h(50);
function p(t) {
  if (typeof t == "string") {
    var o = f.get(t);
    return o && o.image;
  } else
    return t;
}
function v(t, o, n, r, e) {
  if (t)
    if (typeof t == "string") {
      if (o && o.__zrImageSrc === t || !n)
        return o;
      var i = f.get(t), l = { hostEl: n, cb: r, cbPayload: e };
      return i ? (o = i.image, !d(o) && i.pending.push(l)) : (o = u.loadImage(t, s, s), o.__zrImageSrc = t, f.put(t, o.__cachedImgObj = {
        image: o,
        pending: [l]
      })), o;
    } else
      return t;
  else return o;
}
function s() {
  var t = this.__cachedImgObj;
  this.onload = this.onerror = this.__cachedImgObj = null;
  for (var o = 0; o < t.pending.length; o++) {
    var n = t.pending[o], r = n.cb;
    r && r(this, n.cbPayload), n.hostEl.dirty();
  }
  t.pending.length = 0;
}
function d(t) {
  return t && t.width && t.height;
}
export {
  v as createOrUpdateImage,
  p as findExistImage,
  d as isImageReady
};
