// zrender@6.0.0/lib/graphic/helper/parseText.js
import { findExistImage as lr, isImageReady as vr } from "./image.js";
import { retrieve2 as I, retrieve3 as cr, extend as fr, reduce as sr } from "../../core/util.js";
import { adjustTextX as K, adjustTextY as N, getLineHeight as Q, ensureFontMeasureInfo as L, measureWidth as M, measureCharWidth as tr, parsePercent as dr } from "../../contain/text.js";
import _ from "../../core/BoundingRect.js";
var D = /\{([a-zA-Z0-9_]+)\|([^}]*)\}/g;
function gr(a, r, v, u, e, t) {
  if (!v) {
    a.text = "", a.isTruncated = !1;
    return;
  }
  var i = (r + "").split(`
`);
  t = er(v, u, e, t);
  for (var n = !1, h = {}, f = 0, l = i.length; f < l; f++)
    ir(h, i[f], t), i[f] = h.textLine, n = n || h.isTruncated;
  a.text = i.join(`
`), a.isTruncated = n;
}
function er(a, r, v, u) {
  u = u || {};
  var e = fr({}, u);
  v = I(v, "..."), e.maxIterations = I(u.maxIterations, 2);
  var t = e.minChar = I(u.minChar, 0), i = e.fontMeasureInfo = L(r), n = i.asciiCharWidth;
  e.placeholder = I(u.placeholder, "");
  for (var h = a = Math.max(0, a - 1), f = 0; f < t && h >= n; f++)
    h -= n;
  var l = M(i, v);
  return l > h && (v = "", l = 0), h = a - l, e.ellipsis = v, e.ellipsisWidth = l, e.contentWidth = h, e.containerWidth = a, e;
}
function ir(a, r, v) {
  var u = v.containerWidth, e = v.contentWidth, t = v.fontMeasureInfo;
  if (!u) {
    a.textLine = "", a.isTruncated = !1;
    return;
  }
  var i = M(t, r);
  if (i <= u) {
    a.textLine = r, a.isTruncated = !1;
    return;
  }
  for (var n = 0; ; n++) {
    if (i <= e || n >= v.maxIterations) {
      r += v.ellipsis;
      break;
    }
    var h = n === 0 ? pr(r, e, t) : i > 0 ? Math.floor(r.length * e / i) : 0;
    r = r.substr(0, h), i = M(t, r);
  }
  r === "" && (r = v.placeholder), a.textLine = r, a.isTruncated = !0;
}
function pr(a, r, v) {
  for (var u = 0, e = 0, t = a.length; e < t && u < r; e++)
    u += tr(v, a.charCodeAt(e));
  return e;
}
function br(a, r, v, u) {
  var e = U(a), t = r.overflow, i = r.padding, n = i ? i[1] + i[3] : 0, h = i ? i[0] + i[2] : 0, f = r.font, l = t === "truncate", x = Q(f), o = I(r.lineHeight, x), p = r.lineOverflow === "truncate", s = !1, d = r.width;
  d == null && v != null && (d = v - n);
  var C = r.height;
  C == null && u != null && (C = u - h);
  var g;
  d != null && (t === "break" || t === "breakAll") ? g = e ? nr(e, r.font, d, t === "breakAll", 0).lines : [] : g = e ? e.split(`
`) : [];
  var m = g.length * o;
  if (C == null && (C = m), m > C && p) {
    var w = Math.floor(C / o);
    s = s || g.length > w, g = g.slice(0, w), m = g.length * o;
  }
  if (e && l && d != null)
    for (var A = er(d, f, r.ellipsis, {
      minChar: r.truncateMinChar,
      placeholder: r.placeholder
    }), R = {}, H = 0; H < g.length; H++)
      ir(R, g[H], A), g[H] = R.textLine, s = s || R.isTruncated;
  for (var Y = C, T = 0, b = L(f), H = 0; H < g.length; H++)
    T = Math.max(M(b, g[H]), T);
  d == null && (d = T);
  var B = d;
  return Y += h, B += n, {
    lines: g,
    height: C,
    outerWidth: B,
    outerHeight: Y,
    lineHeight: o,
    calculatedLineHeight: x,
    contentWidth: T,
    contentHeight: m,
    width: d,
    isTruncated: s
  };
}
var or = /* @__PURE__ */ (function() {
  function a() {
  }
  return a;
})(), y = /* @__PURE__ */ (function() {
  function a(r) {
    this.tokens = [], r && (this.tokens = r);
  }
  return a;
})(), wr = /* @__PURE__ */ (function() {
  function a() {
    this.width = 0, this.height = 0, this.contentWidth = 0, this.contentHeight = 0, this.outerWidth = 0, this.outerHeight = 0, this.lines = [], this.isTruncated = !1;
  }
  return a;
})();
function Ir(a, r, v, u, e) {
  var t = new wr(), i = U(a);
  if (!i)
    return t;
  var n = r.padding, h = n ? n[1] + n[3] : 0, f = n ? n[0] + n[2] : 0, l = r.width;
  l == null && v != null && (l = v - h);
  var x = r.height;
  x == null && u != null && (x = u - f);
  for (var o = r.overflow, p = (o === "break" || o === "breakAll") && l != null ? { width: l, accumWidth: 0, breakAll: o === "breakAll" } : null, s = D.lastIndex = 0, d; (d = D.exec(i)) != null; ) {
    var C = d.index;
    C > s && J(t, i.substring(s, C), r, p), J(t, d[2], r, p, d[1]), s = D.lastIndex;
  }
  s < i.length && J(t, i.substring(s, i.length), r, p);
  var g = [], m = 0, w = 0, A = o === "truncate", R = r.lineOverflow === "truncate", H = {};
  function Y(S, j, k) {
    S.width = j, S.lineHeight = k, m += k, w = Math.max(w, j);
  }
  r: for (var T = 0; T < t.lines.length; T++) {
    for (var b = t.lines[T], B = 0, V = 0, X = 0; X < b.tokens.length; X++) {
      var c = b.tokens[X], W = c.styleName && r.rich[c.styleName] || {}, F = c.textPadding = W.padding, q = F ? F[1] + F[3] : 0, z = c.font = W.font || r.font;
      c.contentHeight = Q(z);
      var G = I(W.height, c.contentHeight);
      if (c.innerHeight = G, F && (G += F[0] + F[2]), c.height = G, c.lineHeight = cr(W.lineHeight, r.lineHeight, G), c.align = W && W.align || e, c.verticalAlign = W && W.verticalAlign || "middle", R && x != null && m + c.lineHeight > x) {
        var hr = t.lines.length;
        X > 0 ? (b.tokens = b.tokens.slice(0, X), Y(b, V, B), t.lines = t.lines.slice(0, T + 1)) : t.lines = t.lines.slice(0, T), t.isTruncated = t.isTruncated || t.lines.length < hr;
        break r;
      }
      var E = W.width, $ = E == null || E === "auto";
      if (typeof E == "string" && E.charAt(E.length - 1) === "%")
        c.percentWidth = E, g.push(c), c.contentWidth = M(L(z), c.text);
      else {
        if ($) {
          var P = W.backgroundColor, O = P && P.image;
          O && (O = lr(O), vr(O) && (c.width = Math.max(c.width, O.width * G / O.height)));
        }
        var Z = A && l != null ? l - V : null;
        Z != null && Z < c.width ? !$ || Z < q ? (c.text = "", c.width = c.contentWidth = 0) : (gr(H, c.text, Z - q, z, r.ellipsis, { minChar: r.truncateMinChar }), c.text = H.text, t.isTruncated = t.isTruncated || H.isTruncated, c.width = c.contentWidth = M(L(z), c.text)) : c.contentWidth = M(L(z), c.text);
      }
      c.width += q, V += c.width, W && (B = Math.max(B, c.lineHeight));
    }
    Y(b, V, B);
  }
  t.outerWidth = t.width = I(l, w), t.outerHeight = t.height = I(x, m), t.contentHeight = m, t.contentWidth = w, t.outerWidth += h, t.outerHeight += f;
  for (var T = 0; T < g.length; T++) {
    var c = g[T], ur = c.percentWidth;
    c.width = parseInt(ur, 10) / 100 * t.width;
  }
  return t;
}
function J(a, r, v, u, e) {
  var t = r === "", i = e && v.rich[e] || {}, n = a.lines, h = i.font || v.font, f = !1, l, x;
  if (u) {
    var o = i.padding, p = o ? o[1] + o[3] : 0;
    if (i.width != null && i.width !== "auto") {
      var s = dr(i.width, u.width) + p;
      n.length > 0 && s + u.accumWidth > u.width && (l = r.split(`
`), f = !0), u.accumWidth = s;
    } else {
      var d = nr(r, h, u.width, u.breakAll, u.accumWidth);
      u.accumWidth = d.accumWidth + p, x = d.linesWidths, l = d.lines;
    }
  }
  l || (l = r.split(`
`));
  for (var C = L(h), g = 0; g < l.length; g++) {
    var m = l[g], w = new or();
    if (w.styleName = e, w.text = m, w.isLineHolder = !m && !t, typeof i.width == "number" ? w.width = i.width : w.width = x ? x[g] : M(C, m), !g && !f) {
      var A = (n[n.length - 1] || (n[0] = new y())).tokens, R = A.length;
      R === 1 && A[0].isLineHolder ? A[0] = w : (m || !R || t) && A.push(w);
    } else
      n.push(new y([w]));
  }
}
function xr(a) {
  var r = a.charCodeAt(0);
  return r >= 32 && r <= 591 || r >= 880 && r <= 4351 || r >= 4608 && r <= 5119 || r >= 7680 && r <= 8303;
}
var mr = sr(",&?/;] ".split(""), function(a, r) {
  return a[r] = !0, a;
}, {});
function Tr(a) {
  return xr(a) ? !!mr[a] : !0;
}
function nr(a, r, v, u, e) {
  for (var t = [], i = [], n = "", h = "", f = 0, l = 0, x = L(r), o = 0; o < a.length; o++) {
    var p = a.charAt(o);
    if (p === `
`) {
      h && (n += h, l += f), t.push(n), i.push(l), n = "", h = "", f = 0, l = 0;
      continue;
    }
    var s = tr(x, p.charCodeAt(0)), d = u ? !1 : !Tr(p);
    if (t.length ? l + s > v : e + l + s > v) {
      l ? (n || h) && (d ? (n || (n = h, h = "", f = 0, l = f), t.push(n), i.push(l - f), h += p, f += s, n = "", l = f) : (h && (n += h, h = "", f = 0), t.push(n), i.push(l), n = p, l = s)) : d ? (t.push(h), i.push(f), h = p, f = s) : (t.push(p), i.push(s));
      continue;
    }
    l += s, d ? (h += p, f += s) : (h && (n += h, h = "", f = 0), n += p);
  }
  return h && (n += h), n && (t.push(n), i.push(l)), t.length === 1 && (l += e), {
    accumWidth: l,
    lines: t,
    linesWidths: i
  };
}
function Lr(a, r, v, u, e, t) {
  if (a.baseX = v, a.baseY = u, a.outerWidth = a.outerHeight = null, !!r) {
    var i = r.width * 2, n = r.height * 2;
    _.set(rr, K(v, i, e), N(u, n, t), i, n), _.intersect(r, rr, null, ar);
    var h = ar.outIntersectRect;
    a.outerWidth = h.width, a.outerHeight = h.height, a.baseX = K(h.x, h.width, e, !0), a.baseY = N(h.y, h.height, t, !0);
  }
}
var rr = new _(0, 0, 0, 0), ar = { outIntersectRect: {}, clamp: !0 };
function U(a) {
  return a != null ? a += "" : a = "";
}
function Br(a) {
  var r = U(a.text), v = a.font, u = M(L(v), r), e = Q(v);
  return Cr(a, u, e, null);
}
function Cr(a, r, v, u) {
  var e = new _(K(a.x || 0, r, a.textAlign), N(a.y || 0, v, a.textBaseline), r, v), t = u ?? (Hr(a) ? a.lineWidth : 0);
  return t > 0 && (e.x -= t / 2, e.y -= t / 2, e.width += t, e.height += t), e;
}
function Hr(a) {
  var r = a.stroke;
  return r != null && r !== "none" && a.lineWidth > 0;
}
export {
  wr as RichTextContentBlock,
  Lr as calcInnerTextOverflowArea,
  br as parsePlainText,
  Ir as parseRichText,
  Br as tSpanCreateBoundingRect,
  Cr as tSpanCreateBoundingRect2,
  Hr as tSpanHasStroke
};
