// zrender@6.0.0/lib/graphic/helper/roundSector.js
import { isArray as yv } from "../../core/util.js";
var fv = Math.PI, vv = fv * 2, Q = Math.sin, Y = Math.cos, ev = Math.acos, h = Math.atan2, nv = Math.abs, S = Math.sqrt, p = Math.max, C = Math.min, s = 1e-4;
function uv(v, u, i, e, r, N, Z, G) {
  var d = i - v, T = e - u, o = Z - r, n = G - N, f = n * d - o * T;
  if (!(f * f < s))
    return f = (o * (u - N) - n * (v - r)) / f, [v + f * d, u + f * T];
}
function O(v, u, i, e, r, N, Z) {
  var G = v - i, d = u - e, T = (Z ? N : -N) / S(G * G + d * d), o = T * d, n = -T * G, f = v + o, l = u + n, U = i + o, H = e + n, V = (f + U) / 2, W = (l + H) / 2, g = U - f, A = H - l, B = g * g + A * A, J = r - N, b = f * H - U * l, k = (A < 0 ? -1 : 1) * S(p(0, J * J * B - b * b)), q = (b * A - g * k) / B, m = (-b * g - A * k) / B, P = (b * A + g * k) / B, D = (-b * g + A * k) / B, F = q - V, K = m - W, L = P - V, w = D - W;
  return F * F + K * K > L * L + w * w && (q = P, m = D), {
    cx: q,
    cy: m,
    x0: -o,
    y0: -n,
    x1: q * (r / J - 1),
    y1: m * (r / J - 1)
  };
}
function dv(v) {
  var u;
  if (yv(v)) {
    var i = v.length;
    if (!i)
      return v;
    i === 1 ? u = [v[0], v[0], 0, 0] : i === 2 ? u = [v[0], v[0], v[1], v[1]] : i === 3 ? u = v.concat(v[2]) : u = v;
  } else
    u = [v, v, v, v];
  return u;
}
function hv(v, u) {
  var i, e = p(u.r, 0), r = p(u.r0 || 0, 0), N = e > 0, Z = r > 0;
  if (!(!N && !Z)) {
    if (N || (e = r, r = 0), r > e) {
      var G = e;
      e = r, r = G;
    }
    var d = u.startAngle, T = u.endAngle;
    if (!(isNaN(d) || isNaN(T))) {
      var o = u.cx, n = u.cy, f = !!u.clockwise, l = nv(T - d), U = l > vv && l % vv;
      if (U > s && (l = U), !(e > s))
        v.moveTo(o, n);
      else if (l > vv - s)
        v.moveTo(o + e * Y(d), n + e * Q(d)), v.arc(o, n, e, d, T, !f), r > s && (v.moveTo(o + r * Y(T), n + r * Q(T)), v.arc(o, n, r, T, d, f));
      else {
        var H = void 0, V = void 0, W = void 0, g = void 0, A = void 0, B = void 0, J = void 0, b = void 0, k = void 0, q = void 0, m = void 0, P = void 0, D = void 0, F = void 0, K = void 0, L = void 0, w = e * Y(d), X = e * Q(d), $ = r * Y(T), I = r * Q(T), M = l > s;
        if (M) {
          var av = u.cornerRadius;
          av && (i = dv(av), H = i[0], V = i[1], W = i[2], g = i[3]);
          var E = nv(e - r) / 2;
          if (A = C(E, W), B = C(E, g), J = C(E, H), b = C(E, V), m = k = p(A, B), P = q = p(J, b), (k > s || q > s) && (D = e * Y(T), F = e * Q(T), K = r * Y(d), L = r * Q(d), l < fv)) {
            var z = uv(w, X, K, L, D, F, $, I);
            if (z) {
              var _ = w - z[0], x = X - z[1], t = D - z[0], c = F - z[1], rv = 1 / Q(ev((_ * t + x * c) / (S(_ * _ + x * x) * S(t * t + c * c))) / 2), ov = S(z[0] * z[0] + z[1] * z[1]);
              m = C(k, (e - ov) / (rv + 1)), P = C(q, (r - ov) / (rv - 1));
            }
          }
        }
        if (!M)
          v.moveTo(o + w, n + X);
        else if (m > s) {
          var R = C(W, m), j = C(g, m), a = O(K, L, w, X, e, R, f), y = O(D, F, $, I, e, j, f);
          v.moveTo(o + a.cx + a.x0, n + a.cy + a.y0), m < k && R === j ? v.arc(o + a.cx, n + a.cy, m, h(a.y0, a.x0), h(y.y0, y.x0), !f) : (R > 0 && v.arc(o + a.cx, n + a.cy, R, h(a.y0, a.x0), h(a.y1, a.x1), !f), v.arc(o, n, e, h(a.cy + a.y1, a.cx + a.x1), h(y.cy + y.y1, y.cx + y.x1), !f), j > 0 && v.arc(o + y.cx, n + y.cy, j, h(y.y1, y.x1), h(y.y0, y.x0), !f));
        } else
          v.moveTo(o + w, n + X), v.arc(o, n, e, d, T, !f);
        if (!(r > s) || !M)
          v.lineTo(o + $, n + I);
        else if (P > s) {
          var R = C(H, P), j = C(V, P), a = O($, I, D, F, r, -j, f), y = O(w, X, K, L, r, -R, f);
          v.lineTo(o + a.cx + a.x0, n + a.cy + a.y0), P < q && R === j ? v.arc(o + a.cx, n + a.cy, P, h(a.y0, a.x0), h(y.y0, y.x0), !f) : (j > 0 && v.arc(o + a.cx, n + a.cy, j, h(a.y0, a.x0), h(a.y1, a.x1), !f), v.arc(o, n, r, h(a.cy + a.y1, a.cx + a.x1), h(y.cy + y.y1, y.cx + y.x1), f), R > 0 && v.arc(o + y.cx, n + y.cy, R, h(y.y1, y.x1), h(y.y0, y.x0), !f));
        } else
          v.lineTo(o + $, n + I), v.arc(o, n, r, T, d, f);
      }
      v.closePath();
    }
  }
}
export {
  hv as buildPath
};
