// echarts@6.0.0/lib/label/labelStyle.js
import _ from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Text.js";
import { trim as Y, extend as H, isNumber as G, normalizeCssArray as J, keys as Z, retrieve2 as T, isFunction as j, retrieve3 as q } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { DISPLAY_STATES as A, SPECIAL_STATES as O } from "../util/states.js";
import { deprecateReplaceLog as p } from "../util/log.js";
import { makeInner as Q, interpolateRawValues as U } from "../util/model.js";
import "../util/graphic.js";
import { initProps as $, updateProps as x } from "../animation/basicTransition.js";
var D = {};
function z(r, a) {
  for (var n = 0; n < O.length; n++) {
    var o = O[n], i = a[o], e = r.ensureState(o);
    e.style = e.style || {}, e.style.text = i;
  }
  var f = r.currentStates.slice();
  r.clearStates(!0), r.setStyle({
    text: a.normal
  }), r.useStates(f, !0);
}
function P(r, a, n) {
  var o = r.labelFetcher, i = r.labelDataIndex, e = r.labelDimIndex, f = a.normal, u;
  o && (u = o.getFormattedLabel(i, "normal", null, e, f && f.get("formatter"), n != null ? {
    interpolatedValue: n
  } : null)), u == null && (u = j(r.defaultText) ? r.defaultText(i, r, n) : r.defaultText);
  for (var l = {
    normal: u
  }, c = 0; c < O.length; c++) {
    var v = O[c], t = a[v];
    l[v] = T(o ? o.getFormattedLabel(i, v, null, e, t && t.get("formatter")) : null, u);
  }
  return l;
}
function ir(r, a, n, o) {
  n = n || D;
  for (var i = r instanceof _, e = !1, f = 0; f < A.length; f++) {
    var u = a[A[f]];
    if (u && u.getShallow("show")) {
      e = !0;
      break;
    }
  }
  var l = i ? r : r.getTextContent();
  if (e) {
    i || (l || (l = new _(), r.setTextContent(l)), r.stateProxy && (l.stateProxy = r.stateProxy));
    var c = P(n, a), v = a.normal, t = !!v.getShallow("show"), s = V(v, o && o.normal, n, !1, !i);
    s.text = c.normal, i || r.setTextConfig(R(v, n, !1));
    for (var f = 0; f < O.length; f++) {
      var d = O[f], u = a[d];
      if (u) {
        var m = l.ensureState(d), C = !!T(u.getShallow("show"), t);
        if (C !== t && (m.ignore = !C), m.style = V(u, o && o[d], n, !0, !i), m.style.text = c[d], !i) {
          var w = r.ensureState(d);
          w.textConfig = R(u, n, !0);
        }
      }
    }
    l.silent = !!v.getShallow("silent"), l.style.x != null && (s.x = l.style.x), l.style.y != null && (s.y = l.style.y), l.ignore = !t, l.useStyle(s), l.dirty(), n.enableTextSetter && (B(l).setLabelText = function(h) {
      var g = P(n, a, h);
      z(l, g);
    });
  } else l && (l.ignore = !0);
  r.dirty();
}
function or(r, a) {
  a = a || "label";
  for (var n = {
    normal: r.getModel(a)
  }, o = 0; o < O.length; o++) {
    var i = O[o];
    n[i] = r.getModel([i, a]);
  }
  return n;
}
function V(r, a, n, o, i) {
  var e = {};
  return K(e, r, n, o, i), a && H(e, a), e;
}
function R(r, a, n) {
  a = a || {};
  var o = {}, i, e = r.getShallow("rotate"), f = T(r.getShallow("distance"), n ? null : 5), u = r.getShallow("offset");
  return i = r.getShallow("position") || (n ? null : "inside"), i === "outside" && (i = a.defaultOutsidePosition || "top"), i != null && (o.position = i), u != null && (o.offset = u), e != null && (e *= Math.PI / 180, o.rotation = e), f != null && (o.distance = f), o.outsideFill = r.get("color") === "inherit" ? a.inheritColor || null : "auto", a.autoOverflowArea != null && (o.autoOverflowArea = a.autoOverflowArea), a.layoutRect != null && (o.layoutRect = a.layoutRect), o;
}
function K(r, a, n, o, i) {
  n = n || D;
  var e = a.ecModel, f = e && e.option.textStyle, u = S(a), l;
  if (u) {
    l = {};
    var c = "richInheritPlainLabel", v = T(a.get(c), e ? e.get(c) : void 0);
    for (var t in u)
      if (u.hasOwnProperty(t)) {
        var s = a.getModel(["rich", t]);
        X(l[t] = {}, s, f, a, v, n, o, i, !1, !0);
      }
  }
  l && (r.rich = l);
  var d = a.get("overflow");
  d && (r.overflow = d);
  var m = a.get("lineOverflow");
  m && (r.lineOverflow = m);
  var C = r, w = a.get("minMargin");
  if (w != null)
    w = G(w) ? w / 2 : 0, C.margin = [w, w, w, w], C.__marginType = E.minMargin;
  else {
    var h = a.get("textMargin");
    h != null && (C.margin = J(h), C.__marginType = E.textMargin);
  }
  X(r, a, f, null, null, n, o, i, !0, !1);
}
function S(r) {
  for (var a; r && r !== r.ecModel; ) {
    var n = (r.option || D).rich;
    if (n) {
      a = a || {};
      for (var o = Z(n), i = 0; i < o.length; i++) {
        var e = o[i];
        a[e] = 1;
      }
    }
    r = r.parentModel;
  }
  return a;
}
var F = ["fontStyle", "fontWeight", "fontSize", "fontFamily", "textShadowColor", "textShadowBlur", "textShadowOffsetX", "textShadowOffsetY"], k = ["align", "lineHeight", "width", "height", "tag", "verticalAlign", "ellipsis"], W = ["padding", "borderWidth", "borderRadius", "borderDashOffset", "backgroundColor", "borderColor", "shadowColor", "shadowBlur", "shadowOffsetX", "shadowOffsetY"];
function X(r, a, n, o, i, e, f, u, l, c) {
  n = !f && n || D;
  var v = e && e.inheritColor, t = a.getShallow("color"), s = a.getShallow("textBorderColor"), d = T(a.getShallow("opacity"), n.opacity);
  (t === "inherit" || t === "auto") && (process.env.NODE_ENV !== "production" && t === "auto" && p("color: 'auto'", "color: 'inherit'"), v ? t = v : t = null), (s === "inherit" || s === "auto") && (process.env.NODE_ENV !== "production" && s === "auto" && p("color: 'auto'", "color: 'inherit'"), v ? s = v : s = null), u || (t = t || n.color, s = s || n.textBorderColor), t != null && (r.fill = t), s != null && (r.stroke = s);
  var m = T(a.getShallow("textBorderWidth"), n.textBorderWidth);
  m != null && (r.lineWidth = m);
  var C = T(a.getShallow("textBorderType"), n.textBorderType);
  C != null && (r.lineDash = C);
  var w = T(a.getShallow("textBorderDashOffset"), n.textBorderDashOffset);
  w != null && (r.lineDashOffset = w), !f && d == null && !c && (d = e && e.defaultOpacity), d != null && (r.opacity = d), !f && !u && r.fill == null && e.inheritColor && (r.fill = e.inheritColor);
  for (var h = 0; h < F.length; h++) {
    var g = F[h], b = i !== !1 && o ? q(a.getShallow(g), o.getShallow(g), n[g]) : T(a.getShallow(g), n[g]);
    b != null && (r[g] = b);
  }
  for (var h = 0; h < k.length; h++) {
    var g = k[h], b = a.getShallow(g);
    b != null && (r[g] = b);
  }
  if (r.verticalAlign == null) {
    var I = a.getShallow("baseline");
    I != null && (r.verticalAlign = I);
  }
  if (!l || !e.disableBox) {
    for (var h = 0; h < W.length; h++) {
      var g = W[h], b = a.getShallow(g);
      b != null && (r[g] = b);
    }
    var L = a.getShallow("borderType");
    L != null && (r.borderDash = L), (r.backgroundColor === "auto" || r.backgroundColor === "inherit") && v && (process.env.NODE_ENV !== "production" && r.backgroundColor === "auto" && p("backgroundColor: 'auto'", "backgroundColor: 'inherit'"), r.backgroundColor = v), (r.borderColor === "auto" || r.borderColor === "inherit") && v && (process.env.NODE_ENV !== "production" && r.borderColor === "auto" && p("borderColor: 'auto'", "borderColor: 'inherit'"), r.borderColor = v);
  }
}
function lr(r, a) {
  var n = a && a.getModel("textStyle");
  return Y([
    // FIXME in node-canvas fontWeight is before fontStyle
    r.fontStyle || n && n.getShallow("fontStyle") || "",
    r.fontWeight || n && n.getShallow("fontWeight") || "",
    (r.fontSize || n && n.getShallow("fontSize") || 12) + "px",
    r.fontFamily || n && n.getShallow("fontFamily") || "sans-serif"
  ].join(" "));
}
var B = Q();
function fr(r, a, n, o) {
  if (r) {
    var i = B(r);
    i.prevValue = i.value, i.value = n;
    var e = a.normal;
    i.valueAnimation = e.get("valueAnimation"), i.valueAnimation && (i.precision = e.get("precision"), i.defaultInterpolatedText = o, i.statesModels = a);
  }
}
function ur(r, a, n, o, i) {
  var e = B(r);
  if (!e.valueAnimation || e.prevValue === e.value)
    return;
  var f = e.defaultInterpolatedText, u = T(e.interpolatedValue, e.prevValue), l = e.value;
  function c(v) {
    var t = U(n, e.precision, u, l, v);
    e.interpolatedValue = v === 1 ? null : t;
    var s = P({
      labelDataIndex: a,
      labelFetcher: i,
      defaultText: f ? f(t) : t + ""
    }, e.statesModels, t);
    z(r, s);
  }
  r.percent = 0, (e.prevValue == null ? $ : x)(r, {
    // percent is used to prevent animation from being aborted #15916
    percent: 1
  }, o, a, null, c);
}
var E = {
  minMargin: 1,
  textMargin: 2
};
export {
  E as LabelMarginType,
  ur as animateLabelValue,
  R as createTextConfig,
  V as createTextStyle,
  lr as getFont,
  or as getLabelStatesModels,
  B as labelInner,
  ir as setLabelStyle,
  z as setLabelText,
  fr as setLabelValueAnimation
};
