// echarts@6.0.0/lib/label/labelGuideHelper.js
import "../util/graphic.js";
import J from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/PathProxy.js";
import { normalizeRadian as R } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/contain/util.js";
import { quadraticProjectPoint as K, cubicProjectPoint as V } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/curve.js";
import { retrieve2 as O, defaults as W } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { invert as X } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/matrix.js";
import { dist as _, lerp as z } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/vector.js";
import { DISPLAY_STATES as B, SPECIAL_STATES as Q } from "../util/states.js";
import x from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/Point.js";
import $ from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Polyline.js";
import E from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Path.js";
var G = Math.PI * 2, I = J.CMD, N = ["top", "right", "bottom", "left"];
function rr(e, n, v, r, o) {
  var s = v.width, f = v.height;
  switch (e) {
    case "top":
      r.set(v.x + s / 2, v.y - n), o.set(0, -1);
      break;
    case "bottom":
      r.set(v.x + s / 2, v.y + f + n), o.set(0, 1);
      break;
    case "left":
      r.set(v.x - n, v.y + f / 2), o.set(-1, 0);
      break;
    case "right":
      r.set(v.x + s + n, v.y + f / 2), o.set(1, 0);
      break;
  }
}
function ar(e, n, v, r, o, s, f, i, u) {
  f -= e, i -= n;
  var h = Math.sqrt(f * f + i * i);
  f /= h, i /= h;
  var a = f * v + e, m = i * v + n;
  if (Math.abs(r - o) % G < 1e-4)
    return u[0] = a, u[1] = m, h - v;
  if (s) {
    var p = r;
    r = R(o), o = R(p);
  } else
    r = R(r), o = R(o);
  r > o && (o += G);
  var t = Math.atan2(i, f);
  if (t < 0 && (t += G), t >= r && t <= o || t + G >= r && t + G <= o)
    return u[0] = a, u[1] = m, h - v;
  var b = v * Math.cos(r) + e, l = v * Math.sin(r) + n, M = v * Math.cos(o) + e, S = v * Math.sin(o) + n, k = (b - f) * (b - f) + (l - i) * (l - i), w = (M - f) * (M - f) + (S - i) * (S - i);
  return k < w ? (u[0] = b, u[1] = l, Math.sqrt(k)) : (u[0] = M, u[1] = S, Math.sqrt(w));
}
function q(e, n, v, r, o, s, f, i) {
  var u = o - e, h = s - n, a = v - e, m = r - n, p = Math.sqrt(a * a + m * m);
  a /= p, m /= p;
  var t = u * a + h * m, b = t / p;
  i && (b = Math.min(Math.max(b, 0), 1)), b *= p;
  var l = f[0] = e + b * a, M = f[1] = n + b * m;
  return Math.sqrt((l - o) * (l - o) + (M - s) * (M - s));
}
function U(e, n, v, r, o, s, f) {
  v < 0 && (e = e + v, v = -v), r < 0 && (n = n + r, r = -r);
  var i = e + v, u = n + r, h = f[0] = Math.min(Math.max(o, e), i), a = f[1] = Math.min(Math.max(s, n), u);
  return Math.sqrt((h - o) * (h - o) + (a - s) * (a - s));
}
var T = [];
function er(e, n, v) {
  var r = U(n.x, n.y, n.width, n.height, e.x, e.y, T);
  return v.set(T[0], T[1]), r;
}
function or(e, n, v) {
  for (var r = 0, o = 0, s = 0, f = 0, i, u, h = 1 / 0, a = n.data, m = e.x, p = e.y, t = 0; t < a.length; ) {
    var b = a[t++];
    t === 1 && (r = a[t], o = a[t + 1], s = r, f = o);
    var l = h;
    switch (b) {
      case I.M:
        s = a[t++], f = a[t++], r = s, o = f;
        break;
      case I.L:
        l = q(r, o, a[t], a[t + 1], m, p, T, !0), r = a[t++], o = a[t++];
        break;
      case I.C:
        l = V(r, o, a[t++], a[t++], a[t++], a[t++], a[t], a[t + 1], m, p, T), r = a[t++], o = a[t++];
        break;
      case I.Q:
        l = K(r, o, a[t++], a[t++], a[t], a[t + 1], m, p, T), r = a[t++], o = a[t++];
        break;
      case I.A:
        var M = a[t++], S = a[t++], k = a[t++], w = a[t++], j = a[t++], g = a[t++];
        t += 1;
        var Y = !!(1 - a[t++]);
        i = Math.cos(j) * k + M, u = Math.sin(j) * w + S, t <= 1 && (s = i, f = u);
        var Z = (m - M) * w / k + M;
        l = ar(M, S, w, j, j + g, Y, Z, p, T), r = Math.cos(j + g) * k + M, o = Math.sin(j + g) * w + S;
        break;
      case I.R:
        s = r = a[t++], f = o = a[t++];
        var F = a[t++], H = a[t++];
        l = U(s, f, F, H, m, p, T);
        break;
      case I.Z:
        l = q(r, o, s, f, m, p, T, !0), r = s, o = f;
        break;
    }
    l < h && (h = l, v.set(T[0], T[1]));
  }
  return h;
}
var L = new x(), y = new x(), c = new x(), d = new x(), P = new x();
function Mr(e, n) {
  if (e) {
    var v = e.getTextGuideLine(), r = e.getTextContent();
    if (r && v) {
      var o = e.textGuideLineConfig || {}, s = [[0, 0], [0, 0], [0, 0]], f = o.candidates || N, i = r.getBoundingRect().clone();
      i.applyTransform(r.getComputedTransform());
      var u = 1 / 0, h = o.anchor, a = e.getComputedTransform(), m = a && X([], a), p = n.get("length2") || 0;
      h && c.copy(h);
      for (var t = 0; t < f.length; t++) {
        var b = f[t];
        rr(b, 0, i, L, d), x.scaleAndAdd(y, L, d, p), y.transform(m);
        var l = e.getBoundingRect(), M = h ? h.distance(y) : e instanceof E ? or(y, e.path, c) : er(y, l, c);
        M < u && (u = M, y.transform(a), c.transform(a), c.toArray(s[0]), y.toArray(s[1]), L.toArray(s[2]));
      }
      vr(s, n.get("minTurnAngle")), v.setShape({
        points: s
      });
    }
  }
}
var A = [], C = new x();
function vr(e, n) {
  if (n <= 180 && n > 0) {
    n = n / 180 * Math.PI, L.fromArray(e[0]), y.fromArray(e[1]), c.fromArray(e[2]), x.sub(d, L, y), x.sub(P, c, y);
    var v = d.len(), r = P.len();
    if (!(v < 1e-3 || r < 1e-3)) {
      d.scale(1 / v), P.scale(1 / r);
      var o = d.dot(P), s = Math.cos(n);
      if (s < o) {
        var f = q(y.x, y.y, c.x, c.y, L.x, L.y, A, !1);
        C.fromArray(A), C.scaleAndAdd(P, f / Math.tan(Math.PI - n));
        var i = c.x !== y.x ? (C.x - y.x) / (c.x - y.x) : (C.y - y.y) / (c.y - y.y);
        if (isNaN(i))
          return;
        i < 0 ? x.copy(C, y) : i > 1 && x.copy(C, c), C.toArray(e[1]);
      }
    }
  }
}
function br(e, n, v) {
  if (v <= 180 && v > 0) {
    v = v / 180 * Math.PI, L.fromArray(e[0]), y.fromArray(e[1]), c.fromArray(e[2]), x.sub(d, y, L), x.sub(P, c, y);
    var r = d.len(), o = P.len();
    if (!(r < 1e-3 || o < 1e-3)) {
      d.scale(1 / r), P.scale(1 / o);
      var s = d.dot(n), f = Math.cos(v);
      if (s < f) {
        var i = q(y.x, y.y, c.x, c.y, L.x, L.y, A, !1);
        C.fromArray(A);
        var u = Math.PI / 2, h = Math.acos(P.dot(n)), a = u + h - v;
        if (a >= u)
          x.copy(C, c);
        else {
          C.scaleAndAdd(P, i / Math.tan(Math.PI / 2 - a));
          var m = c.x !== y.x ? (C.x - y.x) / (c.x - y.x) : (C.y - y.y) / (c.y - y.y);
          if (isNaN(m))
            return;
          m < 0 ? x.copy(C, y) : m > 1 && x.copy(C, c);
        }
        C.toArray(e[1]);
      }
    }
  }
}
function D(e, n, v, r) {
  var o = v === "normal", s = o ? e : e.ensureState(v);
  s.ignore = n;
  var f = r.get("smooth");
  f && f === !0 && (f = 0.3), s.shape = s.shape || {}, f > 0 && (s.shape.smooth = f);
  var i = r.getModel("lineStyle").getLineStyle();
  o ? e.useStyle(i) : s.style = i;
}
function tr(e, n) {
  var v = n.smooth, r = n.points;
  if (r)
    if (e.moveTo(r[0][0], r[0][1]), v > 0 && r.length >= 3) {
      var o = _(r[0], r[1]), s = _(r[1], r[2]);
      if (!o || !s) {
        e.lineTo(r[1][0], r[1][1]), e.lineTo(r[2][0], r[2][1]);
        return;
      }
      var f = Math.min(o, s) * v, i = z([], r[1], r[0], f / o), u = z([], r[1], r[2], f / s), h = z([], i, u, 0.5);
      e.bezierCurveTo(i[0], i[1], i[0], i[1], h[0], h[1]), e.bezierCurveTo(u[0], u[1], u[0], u[1], r[2][0], r[2][1]);
    } else
      for (var a = 1; a < r.length; a++)
        e.lineTo(r[a][0], r[a][1]);
}
function xr(e, n, v) {
  var r = e.getTextGuideLine(), o = e.getTextContent();
  if (!o) {
    r && e.removeTextGuideLine();
    return;
  }
  for (var s = n.normal, f = s.get("show"), i = o.ignore, u = 0; u < B.length; u++) {
    var h = B[u], a = n[h], m = h === "normal";
    if (a) {
      var p = a.get("show"), t = m ? i : O(o.states[h] && o.states[h].ignore, i);
      if (t || !O(p, f)) {
        var b = m ? r : r && r.states[h];
        b && (b.ignore = !0), r && D(r, !0, h, a);
        continue;
      }
      r || (r = new $(), e.setTextGuideLine(r), !m && (i || !f) && D(r, !0, "normal", n.normal), e.stateProxy && (r.stateProxy = e.stateProxy)), D(r, !1, h, a);
    }
  }
  if (r) {
    W(r.style, v), r.style.fill = null;
    var l = s.get("showAbove"), M = e.textGuideLineConfig = e.textGuideLineConfig || {};
    M.showAbove = l || !1, r.buildPath = tr;
  }
}
function Cr(e, n) {
  n = n || "labelLine";
  for (var v = {
    normal: e.getModel(n)
  }, r = 0; r < Q.length; r++) {
    var o = Q[r];
    v[o] = e.getModel([o, n]);
  }
  return v;
}
export {
  Cr as getLabelLineStatesModels,
  br as limitSurfaceAngle,
  vr as limitTurnAngle,
  xr as setLabelLineStyle,
  Mr as updateLabelLinePoints
};
