// echarts@6.0.0/lib/label/labelLayoutHelper.js
import { ensureCopyTransform as G, ensureCopyRect as O, expandOrShrinkRect as U, isBoundingRectAxisAligned as y, XY as K, WH as N } from "../util/graphic.js";
import { LabelMarginType as T } from "./labelStyle.js";
import Q from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/OrientedBoundingRect.js";
var q = ["label", "labelLine", "layoutOption", "priority", "defaultAttr", "marginForce", "minMarginForce", "marginDefault", "suggestIgnore"], V = 1, E = 2, H = V | E;
function S(r, n, e) {
  e = e || H, n ? r.dirty |= e : r.dirty &= ~e;
}
function W(r, n) {
  return n = n || H, r.dirty == null || !!(r.dirty & n);
}
function X(r) {
  if (r)
    return W(r) && Z(r, r.label, r), r;
}
function Z(r, n, e) {
  var a = n.getComputedTransform();
  r.transform = G(r.transform, a);
  var l = r.localRect = O(r.localRect, n.getBoundingRect()), t = n.style, c = t.margin, m = e && e.marginForce, h = e && e.minMarginForce, o = e && e.marginDefault, g = t.__marginType;
  g == null && o && (c = o, g = T.textMargin);
  for (var s = 0; s < 4; s++)
    L[s] = g === T.minMargin && h && h[s] != null ? h[s] : m && m[s] != null ? m[s] : c ? c[s] : 0;
  g === T.textMargin && U(l, L, !1, !1);
  var _ = r.rect = O(r.rect, l);
  return a && _.applyTransform(a), g === T.minMargin && U(_, L, !1, !1), r.axisAligned = y(a), (r.label = r.label || {}).ignore = n.ignore, S(r, !1), S(r, !0, E), r;
}
var L = [0, 0, 0, 0];
function rr(r, n, e) {
  return r.transform = G(r.transform, e), r.localRect = O(r.localRect, n), r.rect = O(r.rect, n), e && r.rect.applyTransform(e), r.axisAligned = y(e), r.obb = void 0, (r.label = r.label || {}).ignore = !1, r;
}
function nr(r, n) {
  if (r) {
    r.label.x += n.x, r.label.y += n.y, r.label.markRedraw();
    var e = r.transform;
    e && (e[4] += n.x, e[5] += n.y);
    var a = r.rect;
    a && (a.x += n.x, a.y += n.y);
    var l = r.obb;
    l && l.fromBoundingRect(r.localRect, e);
  }
}
function er(r, n) {
  for (var e = 0; e < q.length; e++) {
    var a = q[e];
    r[a] == null && (r[a] = n[a]);
  }
  return X(r);
}
function C(r) {
  var n = r.obb;
  return (!n || W(r, E)) && (r.obb = n = n || new Q(), n.fromBoundingRect(r.localRect, r.transform), S(r, !1, E)), n;
}
function ar(r, n, e, a, l) {
  var t = r.length, c = K[n], m = N[n];
  if (t < 2)
    return !1;
  r.sort(function(i, R) {
    return i.rect[c] - R.rect[c];
  });
  for (var h = 0, o, g = !1, s = 0; s < t; s++) {
    var _ = r[s], M = _.rect;
    o = M[c] - h, o < 0 && (M[c] -= o, _.label[c] -= o, g = !0), h = M[c] + M[m];
  }
  var J = r[0], D = r[t - 1], A, p;
  Y(), A < 0 && b(-A, 0.8), p < 0 && b(p, 0.8), Y(), F(A, p, 1), F(p, A, -1), Y(), A < 0 && k(-A), p < 0 && k(p);
  function Y() {
    A = J.rect[c] - e, p = a - D.rect[c] - D.rect[m];
  }
  function F(i, R, u) {
    if (i < 0) {
      var v = Math.min(R, -i);
      if (v > 0) {
        x(v * u, 0, t);
        var f = v + i;
        f < 0 && b(-f * u, 1);
      } else
        b(-i * u, 1);
    }
  }
  function x(i, R, u) {
    i !== 0 && (g = !0);
    for (var v = R; v < u; v++) {
      var f = r[v], d = f.rect;
      d[c] += i, f.label[c] += i;
    }
  }
  function b(i, R) {
    for (var u = [], v = 0, f = 1; f < t; f++) {
      var d = r[f - 1].rect, z = Math.max(r[f].rect[c] - d[c] - d[m], 0);
      u.push(z), v += z;
    }
    if (v) {
      var P = Math.min(Math.abs(i) / v, R);
      if (i > 0)
        for (var f = 0; f < t - 1; f++) {
          var B = u[f] * P;
          x(B, 0, f + 1);
        }
      else
        for (var f = t - 1; f > 0; f--) {
          var B = u[f - 1] * P;
          x(-B, f, t);
        }
    }
  }
  function k(i) {
    var R = i < 0 ? -1 : 1;
    i = Math.abs(i);
    for (var u = Math.ceil(i / (t - 1)), v = 0; v < t - 1; v++)
      if (R > 0 ? x(u, 0, v + 1) : x(-u, t - v - 1, t), i -= u, i <= 0)
        return;
  }
  return g;
}
function ir(r) {
  for (var n = 0; n < r.length; n++) {
    var e = r[n], a = e.defaultAttr, l = e.labelLine;
    e.label.attr("ignore", a.ignore), l && l.attr("ignore", a.labelGuideIgnore);
  }
}
function fr(r) {
  var n = [];
  r.sort(function(o, g) {
    return (g.suggestIgnore ? 1 : 0) - (o.suggestIgnore ? 1 : 0) || g.priority - o.priority;
  });
  function e(o) {
    if (!o.ignore) {
      var g = o.ensureState("emphasis");
      g.ignore == null && (g.ignore = !1);
    }
    o.ignore = !0;
  }
  for (var a = 0; a < r.length; a++) {
    var l = X(r[a]);
    if (!l.label.ignore) {
      for (var t = l.label, c = l.labelLine, m = !1, h = 0; h < n.length; h++)
        if ($(l, n[h], null, {
          touchThreshold: 0.05
        })) {
          m = !0;
          break;
        }
      m ? (e(t), c && e(c)) : n.push(l);
    }
  }
}
function $(r, n, e, a) {
  return !r || !n || r.label && r.label.ignore || n.label && n.label.ignore || !r.rect.intersect(n.rect, e, a) ? !1 : r.axisAligned && n.axisAligned ? !0 : C(r).intersect(C(n), e, a);
}
export {
  Z as computeLabelGeometry,
  rr as computeLabelGeometry2,
  X as ensureLabelLayoutWithGeometry,
  fr as hideOverlap,
  $ as labelIntersect,
  nr as labelLayoutApplyTranslation,
  er as newLabelLayoutWithGeometry,
  ir as restoreIgnore,
  S as setLabelLayoutDirty,
  ar as shiftLayoutOnXY
};
