// echarts@6.0.0/lib/label/LabelManager.js
import "../util/graphic.js";
import { getECData as m } from "../util/innerStore.js";
import { parsePercent as C } from "../util/number.js";
import H from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/Transformable.js";
import { setLabelLineStyle as N, getLabelLineStatesModels as W, updateLabelLinePoints as S } from "./labelGuideHelper.js";
import { makeInner as D } from "../util/model.js";
import { isFunction as R, keys as U, each as Y, filter as _, retrieve2 as V, indexOf as P } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { newLabelLayoutWithGeometry as j, shiftLayoutOnXY as A, restoreIgnore as w, hideOverlap as K } from "./labelLayoutHelper.js";
import { labelInner as F, animateLabelValue as I } from "./labelStyle.js";
import { normalizeRadian as q } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/contain/util.js";
import O from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/BoundingRect.js";
import { isElementRemoved as J, initProps as X, updateProps as E } from "../animation/basicTransition.js";
function Q(l) {
  if (l) {
    for (var i = [], n = 0; n < l.length; n++)
      i.push(l[n].slice());
    return i;
  }
}
function Z(l, i) {
  var n = l.label, t = i && i.getTextGuideLine();
  return {
    dataIndex: l.dataIndex,
    dataType: l.dataType,
    seriesIndex: l.seriesModel.seriesIndex,
    text: l.label.style.text,
    rect: l.hostRect,
    labelRect: l.rect,
    // x: labelAttr.x,
    // y: labelAttr.y,
    align: n.style.align,
    verticalAlign: n.style.verticalAlign,
    labelLinePoints: Q(t && t.shape.points)
  };
}
var G = ["align", "verticalAlign", "width", "height", "fontSize"], d = new H(), T = D(), $ = D();
function y(l, i, n) {
  for (var t = 0; t < n.length; t++) {
    var a = n[t];
    i[a] != null && (l[a] = i[a]);
  }
}
var x = ["x", "y", "rotation"], gt = (
  /** @class */
  (function() {
    function l() {
      this._labelList = [], this._chartViewList = [];
    }
    return l.prototype.clearLabels = function() {
      this._labelList = [], this._chartViewList = [];
    }, l.prototype._addLabel = function(i, n, t, a, v) {
      var e = a.style, r = a.__hostTarget, s = r.textConfig || {}, f = a.getComputedTransform(), o = a.getBoundingRect().plain();
      O.applyTransform(o, o, f), f ? d.setLocalTransform(f) : (d.x = d.y = d.rotation = d.originX = d.originY = 0, d.scaleX = d.scaleY = 1), d.rotation = q(d.rotation);
      var c = a.__hostTarget, g;
      if (c) {
        g = c.getBoundingRect().plain();
        var p = c.getComputedTransform();
        O.applyTransform(g, g, p);
      }
      var u = g && c.getTextGuideLine();
      this._labelList.push({
        label: a,
        labelLine: u,
        seriesModel: t,
        dataIndex: i,
        dataType: n,
        layoutOptionOrCb: v,
        layoutOption: null,
        rect: o,
        hostRect: g,
        // Label with lower priority will be hidden when overlapped
        // Use rect size as default priority
        priority: g ? g.width * g.height : 0,
        // Save default label attributes.
        // For restore if developers want get back to default value in callback.
        defaultAttr: {
          ignore: a.ignore,
          labelGuideIgnore: u && u.ignore,
          x: d.x,
          y: d.y,
          scaleX: d.scaleX,
          scaleY: d.scaleY,
          rotation: d.rotation,
          style: {
            x: e.x,
            y: e.y,
            align: e.align,
            verticalAlign: e.verticalAlign,
            width: e.width,
            height: e.height,
            fontSize: e.fontSize
          },
          cursor: a.cursor,
          attachedPos: s.position,
          attachedRot: s.rotation
        }
      });
    }, l.prototype.addLabelsOfSeries = function(i) {
      var n = this;
      this._chartViewList.push(i);
      var t = i.__model, a = t.get("labelLayout");
      (R(a) || U(a).length) && i.group.traverse(function(v) {
        if (v.ignore)
          return !0;
        var e = v.getTextContent(), r = m(v);
        e && !e.disableLabelLayout && n._addLabel(r.dataIndex, r.dataType, t, e, a);
      });
    }, l.prototype.updateLayoutConfig = function(i) {
      var n = i.getWidth(), t = i.getHeight();
      function a(B, z) {
        return function() {
          S(B, z);
        };
      }
      for (var v = 0; v < this._labelList.length; v++) {
        var e = this._labelList[v], r = e.label, s = r.__hostTarget, f = e.defaultAttr, o = void 0;
        R(e.layoutOptionOrCb) ? o = e.layoutOptionOrCb(Z(e, s)) : o = e.layoutOptionOrCb, o = o || {}, e.layoutOption = o;
        var c = Math.PI / 180;
        s && s.setTextConfig({
          // Force to set local false.
          local: !1,
          // Ignore position and rotation config on the host el if x or y is changed.
          position: o.x != null || o.y != null ? null : f.attachedPos,
          // Ignore rotation config on the host el if rotation is changed.
          rotation: o.rotate != null ? o.rotate * c : f.attachedRot,
          offset: [o.dx || 0, o.dy || 0]
        });
        var g = !1;
        if (o.x != null ? (r.x = C(o.x, n), r.setStyle("x", 0), g = !0) : (r.x = f.x, r.setStyle("x", f.style.x)), o.y != null ? (r.y = C(o.y, t), r.setStyle("y", 0), g = !0) : (r.y = f.y, r.setStyle("y", f.style.y)), o.labelLinePoints) {
          var p = s.getTextGuideLine();
          p && (p.setShape({
            points: o.labelLinePoints
          }), g = !1);
        }
        var u = T(r);
        u.needsUpdateLabelLine = g, r.rotation = o.rotate != null ? o.rotate * c : f.rotation, r.scaleX = f.scaleX, r.scaleY = f.scaleY;
        for (var h = 0; h < G.length; h++) {
          var L = G[h];
          r.setStyle(L, o[L] != null ? o[L] : f.style[L]);
        }
        if (o.draggable) {
          if (r.draggable = !0, r.cursor = "move", s) {
            var b = e.seriesModel;
            if (e.dataIndex != null) {
              var M = e.seriesModel.getData(e.dataType);
              b = M.getItemModel(e.dataIndex);
            }
            r.on("drag", a(s, b.getModel("labelLine")));
          }
        } else
          r.off("drag"), r.cursor = f.cursor;
      }
    }, l.prototype.layout = function(i) {
      var n = i.getWidth(), t = i.getHeight(), a = [];
      Y(this._labelList, function(s) {
        s.defaultAttr.ignore || a.push(j({}, s));
      });
      var v = _(a, function(s) {
        return s.layoutOption.moveOverlap === "shiftX";
      }), e = _(a, function(s) {
        return s.layoutOption.moveOverlap === "shiftY";
      });
      A(v, 0, 0, n), A(e, 1, 0, t);
      var r = _(a, function(s) {
        return s.layoutOption.hideOverlap;
      });
      w(r), K(r);
    }, l.prototype.processLabelsOverall = function() {
      var i = this;
      Y(this._chartViewList, function(n) {
        var t = n.__model, a = n.ignoreLabelLineUpdate, v = t.isAnimationEnabled();
        n.group.traverse(function(e) {
          if (e.ignore && !e.forceLabelAnimation)
            return !0;
          var r = !a, s = e.getTextContent();
          !r && s && (r = T(s).needsUpdateLabelLine), r && i._updateLabelLine(e, t), v && i._animateLabels(e, t);
        });
      });
    }, l.prototype._updateLabelLine = function(i, n) {
      var t = i.getTextContent(), a = m(i), v = a.dataIndex;
      if (t && v != null) {
        var e = n.getData(a.dataType), r = e.getItemModel(v), s = {}, f = e.getItemVisual(v, "style");
        if (f) {
          var o = e.getVisual("drawType");
          s.stroke = f[o];
        }
        var c = r.getModel("labelLine");
        N(i, W(r), s), S(i, c);
      }
    }, l.prototype._animateLabels = function(i, n) {
      var t = i.getTextContent(), a = i.getTextGuideLine();
      if (t && (i.forceLabelAnimation || !t.ignore && !t.invisible && !i.disableLabelAnimation && !J(i))) {
        var v = T(t), e = v.oldLayout, r = m(i), s = r.dataIndex, f = {
          x: t.x,
          y: t.y,
          rotation: t.rotation
        }, o = n.getData(r.dataType);
        if (e) {
          t.attr(e);
          var g = i.prevStates;
          g && (P(g, "select") >= 0 && t.attr(v.oldLayoutSelect), P(g, "emphasis") >= 0 && t.attr(v.oldLayoutEmphasis)), E(t, f, n, s);
        } else if (t.attr(f), !F(t).valueAnimation) {
          var c = V(t.style.opacity, 1);
          t.style.opacity = 0, X(t, {
            style: {
              opacity: c
            }
          }, n, s);
        }
        if (v.oldLayout = f, t.states.select) {
          var p = v.oldLayoutSelect = {};
          y(p, f, x), y(p, t.states.select, x);
        }
        if (t.states.emphasis) {
          var u = v.oldLayoutEmphasis = {};
          y(u, f, x), y(u, t.states.emphasis, x);
        }
        I(t, s, o, n, n);
      }
      if (a && !a.ignore && !a.invisible) {
        var v = $(a), e = v.oldLayout, h = {
          points: a.shape.points
        };
        e ? (a.attr({
          shape: e
        }), E(a, {
          shape: h
        }, n)) : (a.setShape(h), a.style.strokePercent = 0, X(a, {
          style: {
            strokePercent: 1
          }
        }, n)), v.oldLayout = h;
      }
    }, l;
  })()
);
export {
  gt as default
};
