// echarts@6.0.0/lib/label/sectorLabel.js
import { calculateTextPosition as y } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/contain/text.js";
import { isNumber as w, isArray as E } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
function R(n, s) {
  s = s || {};
  var k = s.isRoundCap;
  return function(r, t, S) {
    var p = t.position;
    if (!p || p instanceof Array)
      return y(r, t, S);
    var T = n(p), e = t.distance != null ? t.distance : 5, h = this.shape, a = h.cx, c = h.cy, b = h.r, M = h.r0, l = (b + M) / 2, i = h.startAngle, u = h.endAngle, d = (i + u) / 2, x = k ? Math.abs(b - M) / 2 : 0, o = Math.cos, A = Math.sin, g = a + b * o(i), v = c + b * A(i), f = "left", m = "top";
    switch (T) {
      case "startArc":
        g = a + (M - e) * o(d), v = c + (M - e) * A(d), f = "center", m = "top";
        break;
      case "insideStartArc":
        g = a + (M + e) * o(d), v = c + (M + e) * A(d), f = "center", m = "bottom";
        break;
      case "startAngle":
        g = a + l * o(i) + C(i, e + x, !1), v = c + l * A(i) + P(i, e + x, !1), f = "right", m = "middle";
        break;
      case "insideStartAngle":
        g = a + l * o(i) + C(i, -e + x, !1), v = c + l * A(i) + P(i, -e + x, !1), f = "left", m = "middle";
        break;
      case "middle":
        g = a + l * o(d), v = c + l * A(d), f = "center", m = "middle";
        break;
      case "endArc":
        g = a + (b + e) * o(d), v = c + (b + e) * A(d), f = "center", m = "bottom";
        break;
      case "insideEndArc":
        g = a + (b - e) * o(d), v = c + (b - e) * A(d), f = "center", m = "top";
        break;
      case "endAngle":
        g = a + l * o(u) + C(u, e + x, !0), v = c + l * A(u) + P(u, e + x, !0), f = "left", m = "middle";
        break;
      case "insideEndAngle":
        g = a + l * o(u) + C(u, -e + x, !0), v = c + l * A(u) + P(u, -e + x, !0), f = "right", m = "middle";
        break;
      default:
        return y(r, t, S);
    }
    return r = r || {}, r.x = g, r.y = v, r.align = f, r.verticalAlign = m, r;
  };
}
function j(n, s, k, r) {
  if (w(r)) {
    n.setTextConfig({
      rotation: r
    });
    return;
  } else if (E(s)) {
    n.setTextConfig({
      rotation: 0
    });
    return;
  }
  var t = n.shape, S = t.clockwise ? t.startAngle : t.endAngle, p = t.clockwise ? t.endAngle : t.startAngle, T = (S + p) / 2, e, h = k(s);
  switch (h) {
    case "startArc":
    case "insideStartArc":
    case "middle":
    case "insideEndArc":
    case "endArc":
      e = T;
      break;
    case "startAngle":
    case "insideStartAngle":
      e = S;
      break;
    case "endAngle":
    case "insideEndAngle":
      e = p;
      break;
    default:
      n.setTextConfig({
        rotation: 0
      });
      return;
  }
  var a = Math.PI * 1.5 - e;
  h === "middle" && a > Math.PI / 2 && a < Math.PI * 1.5 && (a -= Math.PI), n.setTextConfig({
    rotation: a
  });
}
function C(n, s, k) {
  return s * Math.sin(n) * (k ? -1 : 1);
}
function P(n, s, k) {
  return s * Math.cos(n) * (k ? 1 : -1);
}
export {
  R as createSectorCalculateTextPosition,
  j as setSectorTextRotation
};
