// echarts@6.0.0/lib/component/thumbnail/ThumbnailModel.js
import { __extends as l } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import p from "../../model/Component.js";
import { error as u } from "../../util/log.js";
import o from "../../visual/tokens.js";
import { injectThumbnailBridge as n } from "../helper/thumbnailBridge.js";
import { ThumbnailBridgeImpl as a } from "./ThumbnailBridgeImpl.js";
var c = (
  /** @class */
  (function(r) {
    l(t, r);
    function t() {
      var e = r !== null && r.apply(this, arguments) || this;
      return e.type = t.type, e.preventAutoZ = !0, e;
    }
    return t.prototype.optionUpdated = function(e, i) {
      this._updateBridge();
    }, t.prototype._updateBridge = function() {
      var e = this._birdge = this._birdge || new a(this);
      if (this._target = null, this.ecModel.eachSeries(function(s) {
        n(s, null);
      }), this.shouldShow()) {
        var i = this.getTarget();
        n(i.baseMapProvider, e);
      }
    }, t.prototype.shouldShow = function() {
      return this.getShallow("show", !0);
    }, t.prototype.getBridge = function() {
      return this._birdge;
    }, t.prototype.getTarget = function() {
      if (this._target)
        return this._target;
      var e = this.getReferringComponents("series", {
        useDefault: !1,
        enableAll: !1,
        enableNone: !1
      }).models[0];
      return e ? e.subType !== "graph" && (e = null, process.env.NODE_ENV !== "production" && u("series." + e.subType + " is not supported in thumbnail.", !0)) : e = this.ecModel.queryComponents({
        mainType: "series",
        subType: "graph"
      })[0], this._target = {
        baseMapProvider: e
      }, this._target;
    }, t.type = "thumbnail", t.layoutMode = "box", t.dependencies = ["series", "geo"], t.defaultOption = {
      show: !0,
      right: 1,
      bottom: 1,
      height: "25%",
      width: "25%",
      itemStyle: {
        // Use echarts option.backgorundColor by default.
        borderColor: o.color.border,
        borderWidth: 2
      },
      windowStyle: {
        borderWidth: 1,
        color: o.color.neutral30,
        borderColor: o.color.neutral40,
        opacity: 0.3
      },
      z: 10
    }, t;
  })(p)
);
export {
  c as ThumbnailModel
};
