// echarts@6.0.0/lib/component/thumbnail/ThumbnailView.js
import { __extends as C } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import x from "../../view/Component.js";
import V from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/BoundingRect.js";
import { invert as m, mul as z } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/matrix.js";
import I from "../helper/RoamController.js";
import W from "../../visual/tokens.js";
import { createBoxLayoutReference as E, getLayoutRect as u, getBoxLayoutParams as G } from "../../util/layout.js";
import { expandOrShrinkRect as B, retrieveZInfo as P, traverseUpdateZ as Z } from "../../util/graphic.js";
import { applyTransform as h } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/vector.js";
import A from "../../coord/View.js";
import { defaults as U, bind as _, extend as k } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import y from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import p from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Rect.js";
var Q = (
  /** @class */
  (function(a) {
    C(o, a);
    function o() {
      var e = a !== null && a.apply(this, arguments) || this;
      return e.type = o.type, e;
    }
    return o.prototype.render = function(e, t, i) {
      if (this._api = i, this._model = e, this._coordSys || (this._coordSys = new A()), !this._isEnabled()) {
        this._clear();
        return;
      }
      this._renderVersion = i.getMainProcessVersion();
      var r = this.group;
      r.removeAll();
      var s = e.getModel("itemStyle"), n = s.getItemStyle();
      n.fill == null && (n.fill = t.get("backgroundColor") || W.color.neutral00);
      var d = E(e, i).refContainer, f = u(G(e, !0), d), w = n.lineWidth || 0, b = this._contentRect = B(f.clone(), w / 2, !0, !0), l = new y();
      r.add(l), l.setClipPath(new p({
        shape: b.plain()
      }));
      var T = this._targetGroup = new y();
      l.add(T);
      var c = f.plain();
      c.r = s.getShallow("borderRadius", !0), r.add(this._bgRect = new p({
        style: n,
        shape: c,
        silent: !1,
        cursor: "grab"
      }));
      var g = e.getModel("windowStyle"), S = g.getShallow("borderRadius", !0);
      l.add(this._windowRect = new p({
        shape: {
          x: 0,
          y: 0,
          width: 0,
          height: 0,
          r: S
        },
        style: g.getItemStyle(),
        silent: !1,
        cursor: "grab"
      })), this._dealRenderContent(), this._dealUpdateWindow(), R(e, this);
    }, o.prototype.renderContent = function(e) {
      this._bridgeRendered = e, this._isEnabled() && (this._dealRenderContent(), this._dealUpdateWindow(), R(this._model, this));
    }, o.prototype._dealRenderContent = function() {
      var e = this._bridgeRendered;
      if (!(!e || e.renderVersion !== this._renderVersion)) {
        var t = this._targetGroup, i = this._coordSys, r = this._contentRect;
        if (t.removeAll(), !!e) {
          var s = e.group, n = s.getBoundingRect();
          t.add(s), this._bgRect.z2 = e.z2Range.min - 10, i.setBoundingRect(n.x, n.y, n.width, n.height);
          var d = u({
            left: "center",
            top: "center",
            aspect: n.width / n.height
          }, r);
          i.setViewRect(d.x, d.y, d.width, d.height), s.attr(i.getTransformInfo().raw), this._windowRect.z2 = e.z2Range.max + 10, this._resetRoamController(e.roamType);
        }
      }
    }, o.prototype.updateWindow = function(e) {
      var t = this._bridgeRendered;
      t && t.renderVersion === e.renderVersion && (t.targetTrans = e.targetTrans), this._isEnabled() && this._dealUpdateWindow();
    }, o.prototype._dealUpdateWindow = function() {
      var e = this._bridgeRendered;
      if (!(!e || e.renderVersion !== this._renderVersion)) {
        var t = m([], e.targetTrans), i = z([], this._coordSys.transform, t);
        this._transThisToTarget = m([], i);
        var r = e.viewportRect;
        r ? r = r.clone() : r = new V(0, 0, this._api.getWidth(), this._api.getHeight()), r.applyTransform(i);
        var s = this._windowRect, n = s.shape.r;
        s.setShape(U({
          r: n
        }, r));
      }
    }, o.prototype._resetRoamController = function(e) {
      var t = this, i = this._api, r = this._roamController;
      if (r || (r = this._roamController = new I(i.getZr())), !e || !this._isEnabled()) {
        r.disable();
        return;
      }
      r.enable(e, {
        api: i,
        zInfo: {
          component: this._model
        },
        triggerInfo: {
          roamTrigger: null,
          isInSelf: function(s, n, d) {
            return t._contentRect.contain(n, d);
          }
        }
      }), r.off("pan").off("zoom").on("pan", _(this._onPan, this)).on("zoom", _(this._onZoom, this));
    }, o.prototype._onPan = function(e) {
      var t = this._transThisToTarget;
      if (!(!this._isEnabled() || !t)) {
        var i = h([], [e.oldX, e.oldY], t), r = h([], [e.oldX - e.dx, e.oldY - e.dy], t);
        this._api.dispatchAction(v(this._model.getTarget().baseMapProvider, {
          dx: r[0] - i[0],
          dy: r[1] - i[1]
        }));
      }
    }, o.prototype._onZoom = function(e) {
      var t = this._transThisToTarget;
      if (!(!this._isEnabled() || !t)) {
        var i = h([], [e.originX, e.originY], t);
        this._api.dispatchAction(v(this._model.getTarget().baseMapProvider, {
          zoom: 1 / e.scale,
          originX: i[0],
          originY: i[1]
        }));
      }
    }, o.prototype._isEnabled = function() {
      var e = this._model;
      if (!e || !e.shouldShow())
        return !1;
      var t = e.getTarget().baseMapProvider;
      return !!t;
    }, o.prototype._clear = function() {
      this.group.removeAll(), this._bridgeRendered = null, this._roamController && this._roamController.disable();
    }, o.prototype.remove = function() {
      this._clear();
    }, o.prototype.dispose = function() {
      this._clear();
    }, o.type = "thumbnail", o;
  })(x)
);
function v(a, o) {
  var e = a.mainType === "series" ? a.subType + "Roam" : a.mainType + "Roam", t = {
    type: e
  };
  return t[a.mainType + "Id"] = a.id, k(t, o), t;
}
function R(a, o) {
  var e = P(a);
  Z(o.group, e.z, e.zlevel);
}
export {
  Q as ThumbnailView
};
