// hast-util-to-html@9.0.5/lib/omission/closing.js
import { siblingAfter as n } from "./util/siblings.js";
import { omission as N } from "./omission.js";
import { whitespace as c } from "../../../../../hast-util-whitespace@3.0.0/node_modules/hast-util-whitespace/lib/index.js";
const C = N({
  body: u,
  caption: o,
  colgroup: o,
  dd: s,
  dt: d,
  head: o,
  html: i,
  li: p,
  optgroup: f,
  option: y,
  p: l,
  rp: g,
  rt: g,
  tbody: x,
  td: r,
  tfoot: _,
  th: r,
  thead: h,
  tr: b
});
function o(m, a, e) {
  const t = n(e, a, !0);
  return !t || t.type !== "comment" && !(t.type === "text" && c(t.value.charAt(0)));
}
function i(m, a, e) {
  const t = n(e, a);
  return !t || t.type !== "comment";
}
function u(m, a, e) {
  const t = n(e, a);
  return !t || t.type !== "comment";
}
function l(m, a, e) {
  const t = n(e, a);
  return t ? t.type === "element" && (t.tagName === "address" || t.tagName === "article" || t.tagName === "aside" || t.tagName === "blockquote" || t.tagName === "details" || t.tagName === "div" || t.tagName === "dl" || t.tagName === "fieldset" || t.tagName === "figcaption" || t.tagName === "figure" || t.tagName === "footer" || t.tagName === "form" || t.tagName === "h1" || t.tagName === "h2" || t.tagName === "h3" || t.tagName === "h4" || t.tagName === "h5" || t.tagName === "h6" || t.tagName === "header" || t.tagName === "hgroup" || t.tagName === "hr" || t.tagName === "main" || t.tagName === "menu" || t.tagName === "nav" || t.tagName === "ol" || t.tagName === "p" || t.tagName === "pre" || t.tagName === "section" || t.tagName === "table" || t.tagName === "ul") : !e || // Confusing parent.
  !(e.type === "element" && (e.tagName === "a" || e.tagName === "audio" || e.tagName === "del" || e.tagName === "ins" || e.tagName === "map" || e.tagName === "noscript" || e.tagName === "video"));
}
function p(m, a, e) {
  const t = n(e, a);
  return !t || t.type === "element" && t.tagName === "li";
}
function d(m, a, e) {
  const t = n(e, a);
  return !!(t && t.type === "element" && (t.tagName === "dt" || t.tagName === "dd"));
}
function s(m, a, e) {
  const t = n(e, a);
  return !t || t.type === "element" && (t.tagName === "dt" || t.tagName === "dd");
}
function g(m, a, e) {
  const t = n(e, a);
  return !t || t.type === "element" && (t.tagName === "rp" || t.tagName === "rt");
}
function f(m, a, e) {
  const t = n(e, a);
  return !t || t.type === "element" && t.tagName === "optgroup";
}
function y(m, a, e) {
  const t = n(e, a);
  return !t || t.type === "element" && (t.tagName === "option" || t.tagName === "optgroup");
}
function h(m, a, e) {
  const t = n(e, a);
  return !!(t && t.type === "element" && (t.tagName === "tbody" || t.tagName === "tfoot"));
}
function x(m, a, e) {
  const t = n(e, a);
  return !t || t.type === "element" && (t.tagName === "tbody" || t.tagName === "tfoot");
}
function _(m, a, e) {
  return !n(e, a);
}
function b(m, a, e) {
  const t = n(e, a);
  return !t || t.type === "element" && t.tagName === "tr";
}
function r(m, a, e) {
  const t = n(e, a);
  return !t || t.type === "element" && (t.tagName === "td" || t.tagName === "th");
}
export {
  C as closing
};
