// @atlaskit/pragmatic-drag-and-drop@1.7.4/dist/esm/ledger/dispatch-consumer-event.js
import p from "../../../../../../../raf-schd@4.0.3/node_modules/raf-schd/dist/raf-schd.esm.js";
var s = p(function(a) {
  return a();
}), d = /* @__PURE__ */ (function() {
  var a = null;
  function n(l) {
    var o = requestAnimationFrame(function() {
      a = null, l();
    });
    a = {
      frameId: o,
      fn: l
    };
  }
  function t() {
    a && (cancelAnimationFrame(a.frameId), a.fn(), a = null);
  }
  return {
    schedule: n,
    flush: t
  };
})();
function g(a) {
  var n = a.source, t = a.initial, l = a.dispatchEvent, o = {
    dropTargets: []
  };
  function u(c) {
    l(c), o = {
      dropTargets: c.payload.location.current.dropTargets
    };
  }
  var v = {
    start: function(r) {
      var i = r.nativeSetDragImage, e = {
        current: t,
        previous: o,
        initial: t
      };
      u({
        eventName: "onGenerateDragPreview",
        payload: {
          source: n,
          location: e,
          nativeSetDragImage: i
        }
      }), d.schedule(function() {
        u({
          eventName: "onDragStart",
          payload: {
            source: n,
            location: e
          }
        });
      });
    },
    dragUpdate: function(r) {
      var i = r.current;
      d.flush(), s.cancel(), u({
        eventName: "onDropTargetChange",
        payload: {
          source: n,
          location: {
            initial: t,
            previous: o,
            current: i
          }
        }
      });
    },
    drag: function(r) {
      var i = r.current;
      s(function() {
        d.flush();
        var e = {
          initial: t,
          previous: o,
          current: i
        };
        u({
          eventName: "onDrag",
          payload: {
            source: n,
            location: e
          }
        });
      });
    },
    drop: function(r) {
      var i = r.current, e = r.updatedSourcePayload;
      d.flush(), s.cancel(), u({
        eventName: "onDrop",
        payload: {
          source: e ?? n,
          location: {
            current: i,
            previous: o,
            initial: t
          }
        }
      });
    }
  };
  return v;
}
export {
  g as makeDispatch
};
