// @atlaskit/pragmatic-drag-and-drop@1.7.4/dist/esm/ledger/lifecycle-manager.js
import E from "../../../../../../../@babel_runtime@7.28.6/node_modules/@babel/runtime/helpers/esm/toConsumableArray.js";
import { d as x } from "../../../../../../../../../_virtual/index4.js";
import { getElementFromPointWithoutHoneypot as S } from "../honey-pot-fix/get-element-from-point-without-honey-pot.js";
import { isHoneyPotElement as A } from "../honey-pot-fix/is-honey-pot-element.js";
import { isLeavingWindow as O } from "../util/changing-window/is-leaving-window.js";
import { getBindingsForBrokenDrags as b } from "../util/detect-broken-drag.js";
import { getInput as d } from "../util/get-input.js";
import { makeDispatch as P } from "./dispatch-consumer-event.js";
var v = {
  isActive: !1
};
function y() {
  return !v.isActive;
}
function F(r) {
  return r.dataTransfer ? r.dataTransfer.setDragImage.bind(r.dataTransfer) : null;
}
function I(r) {
  var a = r.current, n = r.next;
  if (a.length !== n.length)
    return !0;
  for (var i = 0; i < a.length; i++)
    if (a[i].element !== n[i].element)
      return !0;
  return !1;
}
function w(r) {
  var a = r.event, n = r.dragType, i = r.getDropTargetsOver, u = r.dispatchEvent;
  if (!y())
    return;
  var p = C({
    event: a,
    dragType: n,
    getDropTargetsOver: i
  });
  v.isActive = !0;
  var e = {
    current: p
  };
  f({
    event: a,
    current: p.dropTargets
  });
  var c = P({
    source: n.payload,
    dispatchEvent: u,
    initial: p
  });
  function g(o) {
    var t = I({
      current: e.current.dropTargets,
      next: o.dropTargets
    });
    e.current = o, t && c.dragUpdate({
      current: e.current
    });
  }
  function T(o) {
    var t = d(o), D = A(o.target) ? S({
      x: t.clientX,
      y: t.clientY
    }) : o.target, l = i({
      target: D,
      input: t,
      source: n.payload,
      current: e.current.dropTargets
    });
    l.length && (o.preventDefault(), f({
      event: o,
      current: l
    })), g({
      dropTargets: l,
      input: t
    });
  }
  function s() {
    e.current.dropTargets.length && g({
      dropTargets: [],
      input: e.current.input
    }), c.drop({
      current: e.current,
      updatedSourcePayload: null
    }), m();
  }
  function m() {
    v.isActive = !1, h();
  }
  var h = x.bindAll(
    window,
    [{
      // 👋 Note: we are repurposing the `dragover` event as our `drag` event
      // this is because firefox does not publish pointer coordinates during
      // a `drag` event, but does for every other type of drag event
      // `dragover` fires on all elements that are being dragged over
      // Because we are binding to `window` - our `dragover` is effectively the same as a `drag`
      // 🦊😤
      type: "dragover",
      listener: function(t) {
        T(t), c.drag({
          current: e.current
        });
      }
    }, {
      type: "dragenter",
      listener: T
    }, {
      type: "dragleave",
      listener: function(t) {
        O({
          dragLeave: t
        }) && (g({
          input: e.current.input,
          dropTargets: []
        }), n.startedFrom === "external" && s());
      }
    }, {
      // A "drop" can only happen if the browser allowed the drop
      type: "drop",
      listener: function(t) {
        if (e.current = {
          dropTargets: e.current.dropTargets,
          input: d(t)
        }, !e.current.dropTargets.length) {
          s();
          return;
        }
        t.preventDefault(), f({
          event: t,
          current: e.current.dropTargets
        }), c.drop({
          current: e.current,
          // When dropping something native, we need to extract the latest
          // `.items` from the "drop" event as it is now accessible
          updatedSourcePayload: n.type === "external" ? n.getDropPayload(t) : null
        }), m();
      }
    }, {
      // "dragend" fires when on the drag source (eg a draggable element)
      // when the drag is finished.
      // "dragend" will fire after "drop" (if there was a successful drop)
      // "dragend" does not fire if the draggable source has been removed during the drag
      // or for external drag sources (eg files)
      // This "dragend" listener will not fire if there was a successful drop
      // as we will have already removed the event listener
      type: "dragend",
      listener: function(t) {
        e.current = {
          dropTargets: e.current.dropTargets,
          input: d(t)
        }, s();
      }
    }].concat(E(b({
      onDragEnd: s
    }))),
    // Once we have started a managed drag operation it is important that we see / own all drag events
    // We got one adoption bug pop up where some code was stopping (`event.stopPropagation()`)
    // all "drop" events in the bubble phase on the `document.body`.
    // This meant that we never saw the "drop" event.
    {
      capture: !0
    }
  );
  c.start({
    nativeSetDragImage: F(a)
  });
}
function f(r) {
  var a, n = r.event, i = r.current, u = (a = i[0]) === null || a === void 0 ? void 0 : a.dropEffect;
  u != null && n.dataTransfer && (n.dataTransfer.dropEffect = u);
}
function C(r) {
  var a = r.event, n = r.dragType, i = r.getDropTargetsOver, u = d(a);
  if (n.startedFrom === "external")
    return {
      input: u,
      dropTargets: []
    };
  var p = i({
    input: u,
    source: n.payload,
    target: a.target,
    current: []
  });
  return {
    input: u,
    dropTargets: p
  };
}
var X = {
  canStart: y,
  start: w
};
export {
  X as lifecycle
};
