// @tiptap/extension-typography@2.11.7/dist/index.js
import { Extension as o, textInputRule as t } from "../../../../../@tiptap_core@2.24.0_@tiptap_pm@2.24.0/node_modules/@tiptap/core/dist/index.js";
const i = (s) => t({
  find: /--$/,
  replace: s ?? "—"
}), n = (s) => t({
  find: /\.\.\.$/,
  replace: s ?? "…"
}), l = (s) => t({
  find: /(?:^|[\s{[(<'"\u2018\u201C])(")$/,
  replace: s ?? "“"
}), p = (s) => t({
  find: /"$/,
  replace: s ?? "”"
}), u = (s) => t({
  find: /(?:^|[\s{[(<'"\u2018\u201C])(')$/,
  replace: s ?? "‘"
}), e = (s) => t({
  find: /'$/,
  replace: s ?? "’"
}), a = (s) => t({
  find: /<-$/,
  replace: s ?? "←"
}), h = (s) => t({
  find: /->$/,
  replace: s ?? "→"
}), c = (s) => t({
  find: /\(c\)$/,
  replace: s ?? "©"
}), f = (s) => t({
  find: /\(tm\)$/,
  replace: s ?? "™"
}), r = (s) => t({
  find: /\(sm\)$/,
  replace: s ?? "℠"
}), m = (s) => t({
  find: /\(r\)$/,
  replace: s ?? "®"
}), Q = (s) => t({
  find: /(?:^|\s)(1\/2)\s$/,
  replace: s ?? "½"
}), g = (s) => t({
  find: /\+\/-$/,
  replace: s ?? "±"
}), $ = (s) => t({
  find: /!=$/,
  replace: s ?? "≠"
}), d = (s) => t({
  find: /<<$/,
  replace: s ?? "«"
}), T = (s) => t({
  find: />>$/,
  replace: s ?? "»"
}), k = (s) => t({
  find: /\d+\s?([*x])\s?\d+$/,
  replace: s ?? "×"
}), q = (s) => t({
  find: /\^2$/,
  replace: s ?? "²"
}), w = (s) => t({
  find: /\^3$/,
  replace: s ?? "³"
}), D = (s) => t({
  find: /(?:^|\s)(1\/4)\s$/,
  replace: s ?? "¼"
}), b = (s) => t({
  find: /(?:^|\s)(3\/4)\s$/,
  replace: s ?? "¾"
}), A = o.create({
  name: "typography",
  addOptions() {
    return {
      closeDoubleQuote: "”",
      closeSingleQuote: "’",
      copyright: "©",
      ellipsis: "…",
      emDash: "—",
      laquo: "«",
      leftArrow: "←",
      multiplication: "×",
      notEqual: "≠",
      oneHalf: "½",
      oneQuarter: "¼",
      openDoubleQuote: "“",
      openSingleQuote: "‘",
      plusMinus: "±",
      raquo: "»",
      registeredTrademark: "®",
      rightArrow: "→",
      servicemark: "℠",
      superscriptThree: "³",
      superscriptTwo: "²",
      threeQuarters: "¾",
      trademark: "™"
    };
  },
  addInputRules() {
    const s = [];
    return this.options.emDash !== !1 && s.push(i(this.options.emDash)), this.options.ellipsis !== !1 && s.push(n(this.options.ellipsis)), this.options.openDoubleQuote !== !1 && s.push(l(this.options.openDoubleQuote)), this.options.closeDoubleQuote !== !1 && s.push(p(this.options.closeDoubleQuote)), this.options.openSingleQuote !== !1 && s.push(u(this.options.openSingleQuote)), this.options.closeSingleQuote !== !1 && s.push(e(this.options.closeSingleQuote)), this.options.leftArrow !== !1 && s.push(a(this.options.leftArrow)), this.options.rightArrow !== !1 && s.push(h(this.options.rightArrow)), this.options.copyright !== !1 && s.push(c(this.options.copyright)), this.options.trademark !== !1 && s.push(f(this.options.trademark)), this.options.servicemark !== !1 && s.push(r(this.options.servicemark)), this.options.registeredTrademark !== !1 && s.push(m(this.options.registeredTrademark)), this.options.oneHalf !== !1 && s.push(Q(this.options.oneHalf)), this.options.plusMinus !== !1 && s.push(g(this.options.plusMinus)), this.options.notEqual !== !1 && s.push($(this.options.notEqual)), this.options.laquo !== !1 && s.push(d(this.options.laquo)), this.options.raquo !== !1 && s.push(T(this.options.raquo)), this.options.multiplication !== !1 && s.push(k(this.options.multiplication)), this.options.superscriptTwo !== !1 && s.push(q(this.options.superscriptTwo)), this.options.superscriptThree !== !1 && s.push(w(this.options.superscriptThree)), this.options.oneQuarter !== !1 && s.push(D(this.options.oneQuarter)), this.options.threeQuarters !== !1 && s.push(b(this.options.threeQuarters)), s;
  }
});
export {
  A as Typography,
  p as closeDoubleQuote,
  e as closeSingleQuote,
  c as copyright,
  A as default,
  n as ellipsis,
  i as emDash,
  d as laquo,
  a as leftArrow,
  k as multiplication,
  $ as notEqual,
  Q as oneHalf,
  D as oneQuarter,
  l as openDoubleQuote,
  u as openSingleQuote,
  g as plusMinus,
  T as raquo,
  m as registeredTrademark,
  h as rightArrow,
  r as servicemark,
  w as superscriptThree,
  q as superscriptTwo,
  b as threeQuarters,
  f as trademark
};
