// echarts@6.0.0/lib/chart/graph/circularLayoutHelper.js
import { normalize as C, scale as z, clone as M } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/vector.js";
import { getNodeGlobalScale as D, getSymbolSize as H } from "./graphHelper.js";
import { retrieve3 as w, extend as E } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { getCurvenessForEdge as T } from "../helper/multipleGraphEdgeHelper.js";
var x = Math.PI, p = [];
function _(i, h, f, c) {
  var l = i.coordinateSystem;
  if (!(l && l.type !== "view")) {
    var v = l.getBoundingRect(), u = i.getData(), o = u.graph, e = v.width / 2 + v.x, t = v.height / 2 + v.y, n = Math.min(v.width, v.height) / 2, r = u.count();
    if (u.setLayout({
      cx: e,
      cy: t
    }), !!r) {
      if (f) {
        var a = l.pointToData(c), s = a[0], S = a[1], y = [s - e, S - t];
        C(y, y), z(y, y, n), f.setLayout([e + y[0], t + y[1]], !0);
        var I = i.get(["circular", "rotateLabel"]);
        G(f, I, e, t);
      }
      B[h](i, o, u, n, e, t, r), o.eachEdge(function(m, N) {
        var g = w(m.getModel().get(["lineStyle", "curveness"]), T(m, i, N), 0), b = M(m.node1.getLayout()), d = M(m.node2.getLayout()), L, P = (b[0] + d[0]) / 2, R = (b[1] + d[1]) / 2;
        +g && (g *= 3, L = [e * g + P * (1 - g), t * g + R * (1 - g)]), m.setLayout([b, d, L]);
      });
    }
  }
}
var B = {
  value: function(i, h, f, c, l, v, u) {
    var o = 0, e = f.getSum("value"), t = Math.PI * 2 / (e || u);
    h.eachNode(function(n) {
      var r = n.getValue("value"), a = t * (e ? r : 1) / 2;
      o += a, n.setLayout([c * Math.cos(o) + l, c * Math.sin(o) + v]), o += a;
    });
  },
  symbolSize: function(i, h, f, c, l, v, u) {
    var o = 0;
    p.length = u;
    var e = D(i);
    h.eachNode(function(r) {
      var a = H(r);
      isNaN(a) && (a = 2), a < 0 && (a = 0), a *= e;
      var s = Math.asin(a / 2 / c);
      isNaN(s) && (s = x / 2), p[r.dataIndex] = s, o += s * 2;
    });
    var t = (2 * x - o) / u / 2, n = 0;
    h.eachNode(function(r) {
      var a = t + p[r.dataIndex];
      n += a, (!r.getLayout() || !r.getLayout().fixed) && r.setLayout([c * Math.cos(n) + l, c * Math.sin(n) + v]), n += a;
    });
  }
};
function G(i, h, f, c) {
  var l = i.getGraphicEl();
  if (l) {
    var v = i.getModel(), u = v.get(["label", "rotate"]) || 0, o = l.getSymbolPath();
    if (h) {
      var e = i.getLayout(), t = Math.atan2(e[1] - c, e[0] - f);
      t < 0 && (t = Math.PI * 2 + t);
      var n = e[0] < f;
      n && (t = t - Math.PI);
      var r = n ? "left" : "right";
      o.setTextConfig({
        rotation: -t,
        position: r,
        origin: "center"
      });
      var a = o.ensureState("emphasis");
      E(a.textConfig || (a.textConfig = {}), {
        position: r
      });
    } else
      o.setTextConfig({
        rotation: u *= Math.PI / 180
      });
  }
}
export {
  _ as circularLayout,
  G as rotateNodeLabel
};
