// echarts@6.0.0/lib/chart/graph/createView.js
import d from "../../coord/View.js";
import { createBoxLayoutReference as h, getLayoutRect as s, applyPreserveAspect as x } from "../../util/layout.js";
import { fromPoints as R } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/bbox.js";
import { extend as S } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { injectCoordSysByOption as C } from "../../core/CoordinateSystem.js";
function L(i, c, n) {
  var o = h(i, c), v = S(i.getBoxLayoutParams(), {
    aspect: n
  }), m = s(v, o.refContainer);
  return x(i, m, n);
}
function D(i, c) {
  var n = [];
  return i.eachSeriesByType("graph", function(o) {
    C({
      targetModel: o,
      coordSysType: "view",
      coordSysProvider: v,
      isDefaultDataCoordSys: !0
    });
    function v() {
      var m = o.getData(), u = m.mapArray(function(w) {
        var f = m.getItemModel(w);
        return [+f.get("x"), +f.get("y")];
      }), t = [], e = [];
      R(u, t, e), e[0] - t[0] === 0 && (e[0] += 1, t[0] -= 1), e[1] - t[1] === 0 && (e[1] += 1, t[1] -= 1);
      var y = (e[0] - t[0]) / (e[1] - t[1]), r = L(o, c, y);
      isNaN(y) && (t = [r.x, r.y], e = [r.x + r.width, r.y + r.height]);
      var p = e[0] - t[0], g = e[1] - t[1], a = new d(null, {
        api: c,
        ecModel: i
      });
      return a.zoomLimit = o.get("scaleLimit"), a.setBoundingRect(t[0], t[1], p, g), a.setViewRect(r.x, r.y, r.width, r.height), a.setCenter(o.get("center")), a.setZoom(o.get("zoom")), n.push(a), a;
    }
  }), n;
}
export {
  D as default
};
