// echarts@6.0.0/lib/chart/graph/forceLayout.js
import { forceLayout as k } from "./forceHelper.js";
import { simpleLayout as q } from "./simpleLayoutHelper.js";
import { circularLayout as s } from "./circularLayoutHelper.js";
import { linearMap as b } from "../../util/number.js";
import { copy as F } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/vector.js";
import { isArray as C, retrieve3 as z } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { getCurvenessForEdge as H } from "../helper/multipleGraphEdgeHelper.js";
function X(G) {
  G.eachSeriesByType("graph", function(r) {
    var d = r.coordinateSystem;
    if (!(d && d.type !== "view"))
      if (r.get("layout") === "force") {
        var I = r.preservedPoints || {}, y = r.getGraph(), f = y.data, w = y.edgeData, l = r.getModel("force"), B = l.get("initLayout");
        r.preservedPoints ? f.each(function(a) {
          var v = f.getId(a);
          f.setItemLayout(a, I[v] || [NaN, NaN]);
        }) : !B || B === "none" ? q(r) : B === "circular" && s(r, "value");
        var P = f.getDataExtent("value"), R = w.getDataExtent("value"), N = l.get("repulsion"), g = l.get("edgeLength"), L = C(N) ? N : [N, N], c = C(g) ? g : [g, g];
        c = [c[1], c[0]];
        var E = f.mapArray("value", function(a, v) {
          var e = f.getItemLayout(v), t = b(a, P, L);
          return isNaN(t) && (t = (L[0] + L[1]) / 2), {
            w: t,
            rep: t,
            fixed: f.getItemModel(v).get("fixed"),
            p: !e || isNaN(e[0]) || isNaN(e[1]) ? null : e
          };
        }), T = w.mapArray("value", function(a, v) {
          var e = y.getEdgeByIndex(v), t = b(a, R, c);
          isNaN(t) && (t = (c[0] + c[1]) / 2);
          var m = e.getModel(), o = z(e.getModel().get(["lineStyle", "curveness"]), -H(e, r, v, !0), 0);
          return {
            n1: E[e.node1.dataIndex],
            n2: E[e.node2.dataIndex],
            d: t,
            curveness: o,
            ignoreForceLayout: m.get("ignoreForceLayout")
          };
        }), j = d.getBoundingRect(), x = k(E, T, {
          rect: j,
          gravity: l.get("gravity"),
          friction: l.get("friction")
        });
        x.beforeStep(function(a, v) {
          for (var e = 0, t = a.length; e < t; e++)
            a[e].fixed && F(a[e].p, y.getNodeByIndex(e).getLayout());
        }), x.afterStep(function(a, v, e) {
          for (var t = 0, m = a.length; t < m; t++)
            a[t].fixed || y.getNodeByIndex(t).setLayout(a[t].p), I[f.getId(t)] = a[t].p;
          for (var t = 0, m = v.length; t < m; t++) {
            var o = v[t], M = y.getEdgeByIndex(t), i = o.n1.p, n = o.n2.p, u = M.getLayout();
            u = u ? u.slice() : [], u[0] = u[0] || [], u[1] = u[1] || [], F(u[0], i), F(u[1], n), +o.curveness && (u[2] = [(i[0] + n[0]) / 2 - (i[1] - n[1]) * o.curveness, (i[1] + n[1]) / 2 - (n[0] - i[0]) * o.curveness]), M.setLayout(u);
          }
        }), r.forceLayout = x, r.preservedPoints = I, x.step();
      } else
        r.forceLayout = null;
  });
}
export {
  X as default
};
