// echarts@6.0.0/lib/chart/graph/forceHelper.js
import { create as q, clone as B, sub as s, len as N, normalize as D, scaleAndAdd as E, set as G, copy as H } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/vector.js";
var o = E;
function J(U, z, d) {
  for (var t = U, l = z, u = d.rect, S = u.width, w = u.height, m = [u.x + S / 2, u.y + w / 2], $ = d.gravity == null ? 0.1 : d.gravity, y = 0; y < t.length; y++) {
    var c = t[y];
    c.p || (c.p = q(S * (Math.random() - 0.5) + m[0], w * (Math.random() - 0.5) + m[1])), c.pp = B(c.p), c.edges = null;
  }
  var L = d.friction == null ? 0.6 : d.friction, p = L, A, b;
  return {
    warmUp: function() {
      p = L * 0.8;
    },
    setFixed: function(v) {
      t[v].fixed = !0;
    },
    setUnfixed: function(v) {
      t[v].fixed = !1;
    },
    /**
     * Before step hook
     */
    beforeStep: function(v) {
      A = v;
    },
    /**
     * After step hook
     */
    afterStep: function(v) {
      b = v;
    },
    /**
     * Some formulas were originally copied from "d3.js"
     * https://github.com/d3/d3/blob/b516d77fb8566b576088e73410437494717ada26/src/layout/force.js
     * with some modifications made for this project.
     * See the license statement at the head of this file.
     */
    step: function(v) {
      A && A(t, l);
      for (var a = [], h = t.length, r = 0; r < l.length; r++) {
        var x = l[r];
        if (!x.ignoreForceLayout) {
          var f = x.n1, e = x.n2;
          s(a, e.p, f.p);
          var n = N(a) - x.d, g = e.w / (f.w + e.w);
          isNaN(g) && (g = 0), D(a, a), !f.fixed && o(f.p, f.p, a, g * n * p), !e.fixed && o(e.p, e.p, a, -(1 - g) * n * p);
        }
      }
      for (var r = 0; r < h; r++) {
        var i = t[r];
        i.fixed || (s(a, m, i.p), o(i.p, i.p, a, $ * p));
      }
      for (var r = 0; r < h; r++)
        for (var f = t[r], M = r + 1; M < h; M++) {
          var e = t[M];
          s(a, e.p, f.p);
          var n = N(a);
          n === 0 && (G(a, Math.random() - 0.5, Math.random() - 0.5), n = 1);
          var k = (f.rep + e.rep) / n / n;
          !f.fixed && o(f.pp, f.pp, a, k), !e.fixed && o(e.pp, e.pp, a, -k);
        }
      for (var C = [], r = 0; r < h; r++) {
        var i = t[r];
        i.fixed || (s(C, i.p, i.pp), o(i.p, i.p, C, p), H(i.pp, i.p));
      }
      p = p * 0.992;
      var F = p < 0.01;
      b && b(t, l, F), v && v(F);
    }
  };
}
export {
  J as forceLayout
};
