// echarts@6.0.0/lib/chart/graph/GraphSeries.js
import { __extends as M } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import b from "../../data/SeriesData.js";
import { each as D, map as S, extend as L } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { defaultEmphasis as x } from "../../util/model.js";
import w from "../../model/Model.js";
import C from "../helper/createGraphFromNodeEdge.js";
import k from "../../visual/LegendVisualProvider.js";
import I from "../../model/Series.js";
import { createTooltipMarkup as E } from "../../component/tooltip/tooltipMarkup.js";
import { defaultSeriesFormatTooltip as G } from "../../component/tooltip/seriesFormatTooltip.js";
import { initCurvenessList as P, createEdgeMapForCurveness as T } from "../helper/multipleGraphEdgeHelper.js";
import y from "../../visual/tokens.js";
var J = (
  /** @class */
  (function(n) {
    M(t, n);
    function t() {
      var e = n !== null && n.apply(this, arguments) || this;
      return e.type = t.type, e.hasSymbolVisual = !0, e;
    }
    return t.prototype.init = function(e) {
      n.prototype.init.apply(this, arguments);
      var r = this;
      function a() {
        return r._categoriesData;
      }
      this.legendVisualProvider = new k(a, a), this.fillDataTextStyle(e.edges || e.links), this._updateCategoriesData();
    }, t.prototype.mergeOption = function(e) {
      n.prototype.mergeOption.apply(this, arguments), this.fillDataTextStyle(e.edges || e.links), this._updateCategoriesData();
    }, t.prototype.mergeDefaultAndTheme = function(e) {
      n.prototype.mergeDefaultAndTheme.apply(this, arguments), x(e, "edgeLabel", ["show"]);
    }, t.prototype.getInitialData = function(e, r) {
      var a = e.edges || e.links || [], s = e.data || e.nodes || [], d = this;
      if (s && a) {
        P(this);
        var u = C(s, a, this, !0, g);
        return D(u.edges, function(i) {
          T(i.node1, i.node2, this, i.dataIndex);
        }, this), u.data;
      }
      function g(i, p) {
        i.wrapMethod("getItemModel", function(o) {
          var l = d._categoriesModels, c = o.getShallow("category"), m = l[c];
          return m && (m.parentModel = o.parentModel, o.parentModel = m), o;
        });
        var f = w.prototype.getModel;
        function v(o, l) {
          var c = f.call(this, o, l);
          return c.resolveParentPath = h, c;
        }
        p.wrapMethod("getItemModel", function(o) {
          return o.resolveParentPath = h, o.getModel = v, o;
        });
        function h(o) {
          if (o && (o[0] === "label" || o[1] === "label")) {
            var l = o.slice();
            return o[0] === "label" ? l[0] = "edgeLabel" : o[1] === "label" && (l[1] = "edgeLabel"), l;
          }
          return o;
        }
      }
    }, t.prototype.getGraph = function() {
      return this.getData().graph;
    }, t.prototype.getEdgeData = function() {
      return this.getGraph().edgeData;
    }, t.prototype.getCategoriesData = function() {
      return this._categoriesData;
    }, t.prototype.formatTooltip = function(e, r, a) {
      if (a === "edge") {
        var s = this.getData(), d = this.getDataParams(e, a), u = s.graph.getEdgeByIndex(e), g = s.getName(u.node1.dataIndex), i = s.getName(u.node2.dataIndex), p = [];
        return g != null && p.push(g), i != null && p.push(i), E("nameValue", {
          name: p.join(" > "),
          value: d.value,
          noValue: d.value == null
        });
      }
      var f = G({
        series: this,
        dataIndex: e,
        multipleSeries: r
      });
      return f;
    }, t.prototype._updateCategoriesData = function() {
      var e = S(this.option.categories || [], function(a) {
        return a.value != null ? a : L({
          value: 0
        }, a);
      }), r = new b(["value"], this);
      r.initData(e), this._categoriesData = r, this._categoriesModels = r.mapArray(function(a) {
        return r.getItemModel(a);
      });
    }, t.prototype.setZoom = function(e) {
      this.option.zoom = e;
    }, t.prototype.setCenter = function(e) {
      this.option.center = e;
    }, t.prototype.isAnimationEnabled = function() {
      return n.prototype.isAnimationEnabled.call(this) && !(this.get("layout") === "force" && this.get(["force", "layoutAnimation"]));
    }, t.type = "series.graph", t.dependencies = ["grid", "polar", "geo", "singleAxis", "calendar"], t.defaultOption = {
      // zlevel: 0,
      z: 2,
      coordinateSystem: "view",
      // Default option for all coordinate systems
      // xAxisIndex: 0,
      // yAxisIndex: 0,
      // polarIndex: 0,
      // geoIndex: 0,
      legendHoverLink: !0,
      layout: null,
      // Configuration of circular layout
      circular: {
        rotateLabel: !1
      },
      // Configuration of force directed layout
      force: {
        initLayout: null,
        // Node repulsion. Can be an array to represent range.
        repulsion: [0, 50],
        gravity: 0.1,
        // Initial friction
        friction: 0.6,
        // Edge length. Can be an array to represent range.
        edgeLength: 30,
        layoutAnimation: !0
      },
      left: "center",
      top: "center",
      // right: null,
      // bottom: null,
      // width: '80%',
      // height: '80%',
      symbol: "circle",
      symbolSize: 10,
      edgeSymbol: ["none", "none"],
      edgeSymbolSize: 10,
      edgeLabel: {
        position: "middle",
        distance: 5
      },
      draggable: !1,
      roam: !1,
      // Default on center of graph
      center: null,
      zoom: 1,
      // Symbol size scale ratio in roam
      nodeScaleRatio: 0.6,
      // cursor: null,
      // categories: [],
      // data: []
      // Or
      // nodes: []
      //
      // links: []
      // Or
      // edges: []
      label: {
        show: !1,
        formatter: "{b}"
      },
      itemStyle: {},
      lineStyle: {
        // Don't use tokens.color.border because of the opacity
        color: y.color.neutral50,
        width: 1,
        opacity: 0.5
      },
      emphasis: {
        scale: !0,
        label: {
          show: !0
        }
      },
      select: {
        itemStyle: {
          borderColor: y.color.primary
        }
      }
    }, t;
  })(I)
);
export {
  J as default
};
