// echarts@6.0.0/lib/chart/graph/GraphView.js
import { __extends as N } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import R from "../helper/SymbolDraw.js";
import V from "../helper/LineDraw.js";
import X from "../../component/helper/RoamController.js";
import { updateViewOnPan as Y, updateViewOnZoom as k } from "../../component/helper/roamHelper.js";
import S from "./adjustEdge.js";
import { getNodeGlobalScale as b } from "./graphHelper.js";
import H from "../../view/Chart.js";
import { getECData as C } from "../../util/innerStore.js";
import { simpleLayoutEdge as F } from "./simpleLayoutHelper.js";
import { circularLayout as j, rotateNodeLabel as O } from "./circularLayoutHelper.js";
import { clone as L, extend as P } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import W from "../helper/LinePath.js";
import { getThumbnailBridge as Z } from "../../component/helper/thumbnailBridge.js";
import D from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import { updateProps as U } from "../../animation/basicTransition.js";
function E(y) {
  return y.type === "view";
}
var ct = (
  /** @class */
  (function(y) {
    N(n, y);
    function n() {
      var t = y !== null && y.apply(this, arguments) || this;
      return t.type = n.type, t;
    }
    return n.prototype.init = function(t, o) {
      var a = new R(), r = new V(), i = this.group, e = new D();
      this._controller = new X(o.getZr()), this._controllerHost = {
        target: e
      }, e.add(a.group), e.add(r.group), i.add(e), this._symbolDraw = a, this._lineDraw = r, this._mainGroup = e, this._firstRender = !0;
    }, n.prototype.render = function(t, o, a) {
      var r = this, i = t.coordinateSystem, e = !1;
      this._model = t, this._api = a, this._active = !0;
      var u = this._getThumbnailInfo();
      u && u.bridge.reset(a);
      var f = this._symbolDraw, h = this._lineDraw;
      if (E(i)) {
        var g = {
          x: i.x,
          y: i.y,
          scaleX: i.scaleX,
          scaleY: i.scaleY
        };
        this._firstRender ? this._mainGroup.attr(g) : U(this._mainGroup, g, t);
      }
      S(t.getGraph(), b(t));
      var l = t.getData();
      f.updateData(l);
      var m = t.getEdgeData();
      h.updateData(m), this._updateNodeAndLinkScale(), this._updateController(null, t, a), clearTimeout(this._layoutTimeout);
      var s = t.forceLayout, w = t.get(["force", "layoutAnimation"]);
      s && (e = !0, this._startForceLayoutIteration(s, a, w));
      var x = t.get("layout");
      l.graph.eachNode(function(c) {
        var p = c.dataIndex, d = c.getGraphicEl(), T = c.getModel();
        if (d) {
          d.off("drag").off("dragend");
          var G = T.get("draggable");
          G && d.on("drag", function(z) {
            switch (x) {
              case "force":
                s.warmUp(), !r._layouting && r._startForceLayoutIteration(s, a, w), s.setFixed(p), l.setItemLayout(p, [d.x, d.y]);
                break;
              case "circular":
                l.setItemLayout(p, [d.x, d.y]), c.setLayout({
                  fixed: !0
                }, !0), j(t, "symbolSize", c, [z.offsetX, z.offsetY]), r.updateLayout(t);
                break;
              default:
                l.setItemLayout(p, [d.x, d.y]), F(t.getGraph(), t), r.updateLayout(t);
                break;
            }
          }).on("dragend", function() {
            s && s.setUnfixed(p);
          }), d.setDraggable(G, !!T.get("cursor"));
          var A = T.get(["emphasis", "focus"]);
          A === "adjacency" && (C(d).focus = c.getAdjacentDataIndices());
        }
      }), l.graph.eachEdge(function(c) {
        var p = c.getGraphicEl(), d = c.getModel().get(["emphasis", "focus"]);
        p && d === "adjacency" && (C(p).focus = {
          edge: [c.dataIndex],
          node: [c.node1.dataIndex, c.node2.dataIndex]
        });
      });
      var I = t.get("layout") === "circular" && t.get(["circular", "rotateLabel"]), v = l.getLayout("cx"), _ = l.getLayout("cy");
      l.graph.eachNode(function(c) {
        O(c, I, v, _);
      }), this._firstRender = !1, e || this._renderThumbnail(t, a, this._symbolDraw, this._lineDraw);
    }, n.prototype.dispose = function() {
      this.remove(), this._controller && this._controller.dispose(), this._controllerHost = null;
    }, n.prototype._startForceLayoutIteration = function(t, o, a) {
      var r = this, i = !1;
      (function e() {
        t.step(function(u) {
          r.updateLayout(r._model), (u || !i) && (i = !0, r._renderThumbnail(r._model, o, r._symbolDraw, r._lineDraw)), (r._layouting = !u) && (a ? r._layoutTimeout = setTimeout(e, 16) : e());
        });
      })();
    }, n.prototype._updateController = function(t, o, a) {
      var r = this._controller, i = this._controllerHost, e = o.coordinateSystem;
      if (!E(e)) {
        r.disable();
        return;
      }
      r.enable(o.get("roam"), {
        api: a,
        zInfo: {
          component: o
        },
        triggerInfo: {
          roamTrigger: o.get("roamTrigger"),
          isInSelf: function(u, f, h) {
            return e.containPoint([f, h]);
          },
          isInClip: function(u, f, h) {
            return !t || t.contain(f, h);
          }
        }
      }), i.zoomLimit = o.get("scaleLimit"), i.zoom = e.getZoom(), r.off("pan").off("zoom").on("pan", function(u) {
        a.dispatchAction({
          seriesId: o.id,
          type: "graphRoam",
          dx: u.dx,
          dy: u.dy
        });
      }).on("zoom", function(u) {
        a.dispatchAction({
          seriesId: o.id,
          type: "graphRoam",
          zoom: u.scale,
          originX: u.originX,
          originY: u.originY
        });
      });
    }, n.prototype.updateViewOnPan = function(t, o, a) {
      this._active && (Y(this._controllerHost, a.dx, a.dy), this._updateThumbnailWindow());
    }, n.prototype.updateViewOnZoom = function(t, o, a) {
      this._active && (k(this._controllerHost, a.zoom, a.originX, a.originY), this._updateNodeAndLinkScale(), S(t.getGraph(), b(t)), this._lineDraw.updateLayout(), o.updateLabelLayout(), this._updateThumbnailWindow());
    }, n.prototype._updateNodeAndLinkScale = function() {
      var t = this._model, o = t.getData(), a = b(t);
      o.eachItemGraphicEl(function(r, i) {
        r && r.setSymbolScale(a);
      });
    }, n.prototype.updateLayout = function(t) {
      this._active && (S(t.getGraph(), b(t)), this._symbolDraw.updateLayout(), this._lineDraw.updateLayout());
    }, n.prototype.remove = function() {
      this._active = !1, clearTimeout(this._layoutTimeout), this._layouting = !1, this._layoutTimeout = null, this._symbolDraw && this._symbolDraw.remove(), this._lineDraw && this._lineDraw.remove(), this._controller && this._controller.disable();
    }, n.prototype._getThumbnailInfo = function() {
      var t = this._model, o = t.coordinateSystem;
      if (o.type === "view") {
        var a = Z(t);
        if (a)
          return {
            bridge: a,
            coordSys: o
          };
      }
    }, n.prototype._updateThumbnailWindow = function() {
      var t = this._getThumbnailInfo();
      t && t.bridge.updateWindow(t.coordSys.transform, this._api);
    }, n.prototype._renderThumbnail = function(t, o, a, r) {
      var i = this._getThumbnailInfo();
      if (i) {
        var e = new D(), u = a.group.children(), f = r.group.children(), h = new D(), g = new D();
        e.add(g), e.add(h);
        for (var l = 0; l < u.length; l++) {
          var m = u[l], s = m.children()[0], w = m.x, x = m.y, I = L(s.shape), v = P(I, {
            width: s.scaleX,
            height: s.scaleY,
            x: w - s.scaleX / 2,
            y: x - s.scaleY / 2
          }), _ = L(s.style), c = new s.constructor({
            shape: v,
            style: _,
            z2: 151
          });
          g.add(c);
        }
        for (var l = 0; l < f.length; l++) {
          var m = f[l], p = m.children()[0], _ = L(p.style), v = L(p.shape), d = new W({
            style: _,
            shape: v,
            z2: 151
          });
          h.add(d);
        }
        i.bridge.renderContent({
          api: o,
          roamType: t.get("roam"),
          viewportRect: null,
          group: e,
          targetTrans: i.coordSys.transform
        });
      }
    }, n.type = "graph", n;
  })(H)
);
export {
  ct as default
};
