// echarts@6.0.0/lib/chart/graph/install.js
import a from "./categoryFilter.js";
import p from "./categoryVisual.js";
import f from "./edgeVisual.js";
import y from "./simpleLayout.js";
import d from "./circularLayout.js";
import g from "./forceLayout.js";
import A from "./createView.js";
import V from "../../coord/View.js";
import h from "./GraphView.js";
import S from "./GraphSeries.js";
import { updateCenterAndZoomInAction as w } from "../../component/helper/roamHelper.js";
import { noop as m } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
function Z(o) {
  o.registerChartView(h), o.registerSeriesModel(S), o.registerProcessor(a), o.registerVisual(p), o.registerVisual(f), o.registerLayout(y), o.registerLayout(o.PRIORITY.VISUAL.POST_CHART_LAYOUT, d), o.registerLayout(g), o.registerCoordinateSystem("graphView", {
    dimensions: V.dimensions,
    create: A
  }), o.registerAction({
    type: "focusNodeAdjacency",
    event: "focusNodeAdjacency",
    update: "series:focusNodeAdjacency"
  }, m), o.registerAction({
    type: "unfocusNodeAdjacency",
    event: "unfocusNodeAdjacency",
    update: "series:unfocusNodeAdjacency"
  }, m), o.registerAction({
    type: "graphRoam",
    event: "graphRoam",
    update: "none"
  }, function(e, c, r) {
    c.eachComponent({
      mainType: "series",
      query: e
    }, function(t) {
      var n = r.getViewOfSeriesModel(t);
      n && (e.dx != null && e.dy != null && n.updateViewOnPan(t, r, e), e.zoom != null && e.originX != null && e.originY != null && n.updateViewOnZoom(t, r, e));
      var u = t.coordinateSystem, i = w(u, e, t.get("scaleLimit"));
      t.setCenter && t.setCenter(i.center), t.setZoom && t.setZoom(i.zoom);
    });
  });
}
export {
  Z as install
};
