// @tiptap/react@2.24.0/dist/index.js
import { BubbleMenuPlugin as se } from "../../../../../@tiptap_extension-bubble-menu@2.24.0_@tiptap_core@2.24.0_@tiptap_pm@2.24.0__@tiptap_pm@2.24.0/node_modules/@tiptap/extension-bubble-menu/dist/index.js";
import v, { useRef as ue, useState as W, useDebugValue as Z, useEffect as z, forwardRef as ae, useLayoutEffect as ce, useContext as ee, createContext as te, createRef as de, memo as le, createElement as fe, version as G } from "react";
import he from "react-dom";
import { Editor as pe, NodeView as me, getRenderedAttributes as ve } from "../../../../../@tiptap_core@2.24.0_@tiptap_pm@2.24.0/node_modules/@tiptap/core/dist/index.js";
import { CommandManager as Xe, Extension as Ze, InputRule as et, Mark as tt, Node as nt, NodePos as rt, PasteRule as it, callOrReturn as ot, combineTransactionSteps as st, createChainableState as ut, createDocument as at, createNodeFromContent as ct, createStyleTag as dt, defaultBlockAt as lt, deleteProps as ft, elementFromString as ht, escapeForRegEx as pt, findChildren as mt, findChildrenInRange as vt, findDuplicates as St, findParentNode as Et, findParentNodeClosestToPos as _t, fromString as yt, generateHTML as gt, getAttributes as Ot, getAttributesFromExtensions as bt, getChangedRanges as Rt, getExtensionField as wt, getHTMLFromFragment as Dt, getMarkAttributes as Ct, getMarkRange as Tt, getMarkType as Mt, getMarksBetween as xt, getNodeAttributes as Nt, getNodeType as At, getSchema as Lt, getSchemaByResolvedExtensions as Vt, getSchemaTypeByName as It, getSchemaTypeNameByName as Pt, getSplittedAttributes as Bt, getText as jt, getTextBetween as Ft, getTextContentFromNodes as Ut, getTextSerializersFromSchema as $t, injectExtensionAttributesToParseRule as Wt, inputRulesPlugin as zt, isActive as Ht, isEmptyObject as Gt, isExtensionRulesEnabled as Kt, isFunction as qt, isList as kt, isMacOS as Yt, isMarkActive as Jt, isNodeActive as Qt, isNodeEmpty as Xt, isNodeSelection as Zt, isNumber as en, isPlainObject as tn, isRegExp as nn, isTextSelection as rn, isiOS as on, markInputRule as sn, markPasteRule as un, mergeAttributes as an, mergeDeep as cn, minMax as dn, nodeInputRule as ln, nodePasteRule as fn, objectIncludes as hn, pasteRulesPlugin as pn, posToDOMRect as mn, removeDuplicates as vn, resolveFocusPosition as Sn, selectionToInsertionEnd as En, splitExtensions as _n, textInputRule as yn, textblockTypeInputRule as gn, wrappingInputRule as On } from "../../../../../@tiptap_core@2.24.0_@tiptap_pm@2.24.0/node_modules/@tiptap/core/dist/index.js";
function Se(n) {
  return n && n.__esModule && Object.prototype.hasOwnProperty.call(n, "default") ? n.default : n;
}
var F = { exports: {} }, I = {};
var K;
function Ee() {
  if (K) return I;
  K = 1;
  var n = v;
  function e(s, f) {
    return s === f && (s !== 0 || 1 / s === 1 / f) || s !== s && f !== f;
  }
  var t = typeof Object.is == "function" ? Object.is : e, r = n.useState, i = n.useEffect, o = n.useLayoutEffect, l = n.useDebugValue;
  function S(s, f) {
    var p = f(), E = r({ inst: { value: p, getSnapshot: f } }), m = E[0].inst, _ = E[1];
    return o(function() {
      m.value = p, m.getSnapshot = f, O(m) && _({ inst: m });
    }, [s, p, f]), i(function() {
      return O(m) && _({ inst: m }), s(function() {
        O(m) && _({ inst: m });
      });
    }, [s]), l(p), p;
  }
  function O(s) {
    var f = s.getSnapshot;
    s = s.value;
    try {
      var p = f();
      return !t(s, p);
    } catch {
      return !0;
    }
  }
  function d(s, f) {
    return f();
  }
  var a = typeof window > "u" || typeof window.document > "u" || typeof window.document.createElement > "u" ? d : S;
  return I.useSyncExternalStore = n.useSyncExternalStore !== void 0 ? n.useSyncExternalStore : a, I;
}
var P = {};
var q;
function _e() {
  return q || (q = 1, process.env.NODE_ENV !== "production" && (function() {
    typeof __REACT_DEVTOOLS_GLOBAL_HOOK__ < "u" && typeof __REACT_DEVTOOLS_GLOBAL_HOOK__.registerInternalModuleStart == "function" && __REACT_DEVTOOLS_GLOBAL_HOOK__.registerInternalModuleStart(new Error());
    var n = v, e = n.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED;
    function t(h) {
      {
        for (var u = arguments.length, g = new Array(u > 1 ? u - 1 : 0), c = 1; c < u; c++)
          g[c - 1] = arguments[c];
        r("error", h, g);
      }
    }
    function r(h, u, g) {
      {
        var c = e.ReactDebugCurrentFrame, y = c.getStackAddendum();
        y !== "" && (u += "%s", g = g.concat([y]));
        var R = g.map(function(b) {
          return String(b);
        });
        R.unshift("Warning: " + u), Function.prototype.apply.call(console[h], console, R);
      }
    }
    function i(h, u) {
      return h === u && (h !== 0 || 1 / h === 1 / u) || h !== h && u !== u;
    }
    var o = typeof Object.is == "function" ? Object.is : i, l = n.useState, S = n.useEffect, O = n.useLayoutEffect, d = n.useDebugValue, a = !1, s = !1;
    function f(h, u, g) {
      a || n.startTransition !== void 0 && (a = !0, t("You are using an outdated, pre-release alpha of React 18 that does not support useSyncExternalStore. The use-sync-external-store shim will not work correctly. Upgrade to a newer pre-release."));
      var c = u();
      if (!s) {
        var y = u();
        o(c, y) || (t("The result of getSnapshot should be cached to avoid an infinite loop"), s = !0);
      }
      var R = l({
        inst: {
          value: c,
          getSnapshot: u
        }
      }), b = R[0].inst, C = R[1];
      return O(function() {
        b.value = c, b.getSnapshot = u, p(b) && C({
          inst: b
        });
      }, [h, c, u]), S(function() {
        p(b) && C({
          inst: b
        });
        var x = function() {
          p(b) && C({
            inst: b
          });
        };
        return h(x);
      }, [h]), d(c), c;
    }
    function p(h) {
      var u = h.getSnapshot, g = h.value;
      try {
        var c = u();
        return !o(g, c);
      } catch {
        return !0;
      }
    }
    function E(h, u, g) {
      return u();
    }
    var m = typeof window < "u" && typeof window.document < "u" && typeof window.document.createElement < "u", _ = !m, w = _ ? E : f, D = n.useSyncExternalStore !== void 0 ? n.useSyncExternalStore : w;
    P.useSyncExternalStore = D, typeof __REACT_DEVTOOLS_GLOBAL_HOOK__ < "u" && typeof __REACT_DEVTOOLS_GLOBAL_HOOK__.registerInternalModuleStop == "function" && __REACT_DEVTOOLS_GLOBAL_HOOK__.registerInternalModuleStop(new Error());
  })()), P;
}
process.env.NODE_ENV === "production" ? F.exports = Ee() : F.exports = _e();
var M = F.exports;
const ye = (...n) => (e) => {
  n.forEach((t) => {
    typeof t == "function" ? t(e) : t && (t.current = e);
  });
}, ge = ({ contentComponent: n }) => {
  const e = M.useSyncExternalStore(n.subscribe, n.getSnapshot, n.getServerSnapshot);
  return v.createElement(v.Fragment, null, Object.values(e));
};
function Oe() {
  const n = /* @__PURE__ */ new Set();
  let e = {};
  return {
    /**
     * Subscribe to the editor instance's changes.
     */
    subscribe(t) {
      return n.add(t), () => {
        n.delete(t);
      };
    },
    getSnapshot() {
      return e;
    },
    getServerSnapshot() {
      return e;
    },
    /**
     * Adds a new NodeView Renderer to the editor.
     */
    setRenderer(t, r) {
      e = {
        ...e,
        [t]: he.createPortal(r.reactElement, r.element, t)
      }, n.forEach((i) => i());
    },
    /**
     * Removes a NodeView Renderer from the editor.
     */
    removeRenderer(t) {
      const r = { ...e };
      delete r[t], e = r, n.forEach((i) => i());
    }
  };
}
class be extends v.Component {
  constructor(e) {
    var t;
    super(e), this.editorContentRef = v.createRef(), this.initialized = !1, this.state = {
      hasContentComponentInitialized: !!(!((t = e.editor) === null || t === void 0) && t.contentComponent)
    };
  }
  componentDidMount() {
    this.init();
  }
  componentDidUpdate() {
    this.init();
  }
  init() {
    const e = this.props.editor;
    if (e && !e.isDestroyed && e.options.element) {
      if (e.contentComponent)
        return;
      const t = this.editorContentRef.current;
      t.append(...e.options.element.childNodes), e.setOptions({
        element: t
      }), e.contentComponent = Oe(), this.state.hasContentComponentInitialized || (this.unsubscribeToContentComponent = e.contentComponent.subscribe(() => {
        this.setState((r) => r.hasContentComponentInitialized ? r : {
          hasContentComponentInitialized: !0
        }), this.unsubscribeToContentComponent && this.unsubscribeToContentComponent();
      })), e.createNodeViews(), this.initialized = !0;
    }
  }
  componentWillUnmount() {
    const e = this.props.editor;
    if (!e || (this.initialized = !1, e.isDestroyed || e.view.setProps({
      nodeViews: {}
    }), this.unsubscribeToContentComponent && this.unsubscribeToContentComponent(), e.contentComponent = null, !e.options.element.firstChild))
      return;
    const t = document.createElement("div");
    t.append(...e.options.element.childNodes), e.setOptions({
      element: t
    });
  }
  render() {
    const { editor: e, innerRef: t, ...r } = this.props;
    return v.createElement(
      v.Fragment,
      null,
      v.createElement("div", { ref: ye(t, this.editorContentRef), ...r }),
      e?.contentComponent && v.createElement(ge, { contentComponent: e.contentComponent })
    );
  }
}
const Re = ae((n, e) => {
  const t = v.useMemo(() => Math.floor(Math.random() * 4294967295).toString(), [n.editor]);
  return v.createElement(be, {
    key: t,
    innerRef: e,
    ...n
  });
}), He = v.memo(Re);
var we = function n(e, t) {
  if (e === t) return !0;
  if (e && t && typeof e == "object" && typeof t == "object") {
    if (e.constructor !== t.constructor) return !1;
    var r, i, o;
    if (Array.isArray(e)) {
      if (r = e.length, r != t.length) return !1;
      for (i = r; i-- !== 0; )
        if (!n(e[i], t[i])) return !1;
      return !0;
    }
    if (e instanceof Map && t instanceof Map) {
      if (e.size !== t.size) return !1;
      for (i of e.entries())
        if (!t.has(i[0])) return !1;
      for (i of e.entries())
        if (!n(i[1], t.get(i[0]))) return !1;
      return !0;
    }
    if (e instanceof Set && t instanceof Set) {
      if (e.size !== t.size) return !1;
      for (i of e.entries())
        if (!t.has(i[0])) return !1;
      return !0;
    }
    if (ArrayBuffer.isView(e) && ArrayBuffer.isView(t)) {
      if (r = e.length, r != t.length) return !1;
      for (i = r; i-- !== 0; )
        if (e[i] !== t[i]) return !1;
      return !0;
    }
    if (e.constructor === RegExp) return e.source === t.source && e.flags === t.flags;
    if (e.valueOf !== Object.prototype.valueOf) return e.valueOf() === t.valueOf();
    if (e.toString !== Object.prototype.toString) return e.toString() === t.toString();
    if (o = Object.keys(e), r = o.length, r !== Object.keys(t).length) return !1;
    for (i = r; i-- !== 0; )
      if (!Object.prototype.hasOwnProperty.call(t, o[i])) return !1;
    for (i = r; i-- !== 0; ) {
      var l = o[i];
      if (!(l === "_owner" && e.$$typeof) && !n(e[l], t[l]))
        return !1;
    }
    return !0;
  }
  return e !== e && t !== t;
}, De = /* @__PURE__ */ Se(we), U = { exports: {} }, B = {};
var k;
function Ce() {
  if (k) return B;
  k = 1;
  var n = v, e = M;
  function t(d, a) {
    return d === a && (d !== 0 || 1 / d === 1 / a) || d !== d && a !== a;
  }
  var r = typeof Object.is == "function" ? Object.is : t, i = e.useSyncExternalStore, o = n.useRef, l = n.useEffect, S = n.useMemo, O = n.useDebugValue;
  return B.useSyncExternalStoreWithSelector = function(d, a, s, f, p) {
    var E = o(null);
    if (E.current === null) {
      var m = { hasValue: !1, value: null };
      E.current = m;
    } else m = E.current;
    E = S(function() {
      function w(c) {
        if (!D) {
          if (D = !0, h = c, c = f(c), p !== void 0 && m.hasValue) {
            var y = m.value;
            if (p(y, c)) return u = y;
          }
          return u = c;
        }
        if (y = u, r(h, c)) return y;
        var R = f(c);
        return p !== void 0 && p(y, R) ? y : (h = c, u = R);
      }
      var D = !1, h, u, g = s === void 0 ? null : s;
      return [function() {
        return w(a());
      }, g === null ? void 0 : function() {
        return w(g());
      }];
    }, [a, s, f, p]);
    var _ = i(d, E[0], E[1]);
    return l(function() {
      m.hasValue = !0, m.value = _;
    }, [_]), O(_), _;
  }, B;
}
var j = {};
var Y;
function Te() {
  return Y || (Y = 1, process.env.NODE_ENV !== "production" && (function() {
    typeof __REACT_DEVTOOLS_GLOBAL_HOOK__ < "u" && typeof __REACT_DEVTOOLS_GLOBAL_HOOK__.registerInternalModuleStart == "function" && __REACT_DEVTOOLS_GLOBAL_HOOK__.registerInternalModuleStart(new Error());
    var n = v, e = M;
    function t(a, s) {
      return a === s && (a !== 0 || 1 / a === 1 / s) || a !== a && s !== s;
    }
    var r = typeof Object.is == "function" ? Object.is : t, i = e.useSyncExternalStore, o = n.useRef, l = n.useEffect, S = n.useMemo, O = n.useDebugValue;
    function d(a, s, f, p, E) {
      var m = o(null), _;
      m.current === null ? (_ = {
        hasValue: !1,
        value: null
      }, m.current = _) : _ = m.current;
      var w = S(function() {
        var g = !1, c, y, R = function(T) {
          if (!g) {
            g = !0, c = T;
            var N = p(T);
            if (E !== void 0 && _.hasValue) {
              var A = _.value;
              if (E(A, N))
                return y = A, A;
            }
            return y = N, N;
          }
          var oe = c, L = y;
          if (r(oe, T))
            return L;
          var V = p(T);
          return E !== void 0 && E(L, V) ? L : (c = T, y = V, V);
        }, b = f === void 0 ? null : f, C = function() {
          return R(s());
        }, x = b === null ? void 0 : function() {
          return R(b());
        };
        return [C, x];
      }, [s, f, p, E]), D = w[0], h = w[1], u = i(a, D, h);
      return l(function() {
        _.hasValue = !0, _.value = u;
      }, [u]), O(u), u;
    }
    j.useSyncExternalStoreWithSelector = d, typeof __REACT_DEVTOOLS_GLOBAL_HOOK__ < "u" && typeof __REACT_DEVTOOLS_GLOBAL_HOOK__.registerInternalModuleStop == "function" && __REACT_DEVTOOLS_GLOBAL_HOOK__.registerInternalModuleStop(new Error());
  })()), j;
}
process.env.NODE_ENV === "production" ? U.exports = Ce() : U.exports = Te();
var Me = U.exports;
const xe = typeof window < "u" ? ce : z;
class Ne {
  constructor(e) {
    this.transactionNumber = 0, this.lastTransactionNumber = 0, this.subscribers = /* @__PURE__ */ new Set(), this.editor = e, this.lastSnapshot = { editor: e, transactionNumber: 0 }, this.getSnapshot = this.getSnapshot.bind(this), this.getServerSnapshot = this.getServerSnapshot.bind(this), this.watch = this.watch.bind(this), this.subscribe = this.subscribe.bind(this);
  }
  /**
   * Get the current editor instance.
   */
  getSnapshot() {
    return this.transactionNumber === this.lastTransactionNumber ? this.lastSnapshot : (this.lastTransactionNumber = this.transactionNumber, this.lastSnapshot = { editor: this.editor, transactionNumber: this.transactionNumber }, this.lastSnapshot);
  }
  /**
   * Always disable the editor on the server-side.
   */
  getServerSnapshot() {
    return { editor: null, transactionNumber: 0 };
  }
  /**
   * Subscribe to the editor instance's changes.
   */
  subscribe(e) {
    return this.subscribers.add(e), () => {
      this.subscribers.delete(e);
    };
  }
  /**
   * Watch the editor instance for changes.
   */
  watch(e) {
    if (this.editor = e, this.editor) {
      const t = () => {
        this.transactionNumber += 1, this.subscribers.forEach((i) => i());
      }, r = this.editor;
      return r.on("transaction", t), () => {
        r.off("transaction", t);
      };
    }
  }
}
function Ae(n) {
  var e;
  const [t] = W(() => new Ne(n.editor)), r = Me.useSyncExternalStoreWithSelector(t.subscribe, t.getSnapshot, t.getServerSnapshot, n.selector, (e = n.equalityFn) !== null && e !== void 0 ? e : De);
  return xe(() => t.watch(n.editor), [n.editor, t]), Z(r), r;
}
const J = process.env.NODE_ENV !== "production", $ = typeof window > "u", Le = $ || !!(typeof window < "u" && window.next);
class H {
  constructor(e) {
    this.editor = null, this.subscriptions = /* @__PURE__ */ new Set(), this.isComponentMounted = !1, this.previousDeps = null, this.instanceId = "", this.options = e, this.subscriptions = /* @__PURE__ */ new Set(), this.setEditor(this.getInitialEditor()), this.scheduleDestroy(), this.getEditor = this.getEditor.bind(this), this.getServerSnapshot = this.getServerSnapshot.bind(this), this.subscribe = this.subscribe.bind(this), this.refreshEditorInstance = this.refreshEditorInstance.bind(this), this.scheduleDestroy = this.scheduleDestroy.bind(this), this.onRender = this.onRender.bind(this), this.createEditor = this.createEditor.bind(this);
  }
  setEditor(e) {
    this.editor = e, this.instanceId = Math.random().toString(36).slice(2, 9), this.subscriptions.forEach((t) => t());
  }
  getInitialEditor() {
    if (this.options.current.immediatelyRender === void 0)
      return $ || Le ? (J && console.warn("Tiptap Error: SSR has been detected, please set `immediatelyRender` explicitly to `false` to avoid hydration mismatches."), null) : this.createEditor();
    if (this.options.current.immediatelyRender && $ && J)
      throw new Error("Tiptap Error: SSR has been detected, and `immediatelyRender` has been set to `true` this is an unsupported configuration that may result in errors, explicitly set `immediatelyRender` to `false` to avoid hydration mismatches.");
    return this.options.current.immediatelyRender ? this.createEditor() : null;
  }
  /**
   * Create a new editor instance. And attach event listeners.
   */
  createEditor() {
    const e = {
      ...this.options.current,
      // Always call the most recent version of the callback function by default
      onBeforeCreate: (...r) => {
        var i, o;
        return (o = (i = this.options.current).onBeforeCreate) === null || o === void 0 ? void 0 : o.call(i, ...r);
      },
      onBlur: (...r) => {
        var i, o;
        return (o = (i = this.options.current).onBlur) === null || o === void 0 ? void 0 : o.call(i, ...r);
      },
      onCreate: (...r) => {
        var i, o;
        return (o = (i = this.options.current).onCreate) === null || o === void 0 ? void 0 : o.call(i, ...r);
      },
      onDestroy: (...r) => {
        var i, o;
        return (o = (i = this.options.current).onDestroy) === null || o === void 0 ? void 0 : o.call(i, ...r);
      },
      onFocus: (...r) => {
        var i, o;
        return (o = (i = this.options.current).onFocus) === null || o === void 0 ? void 0 : o.call(i, ...r);
      },
      onSelectionUpdate: (...r) => {
        var i, o;
        return (o = (i = this.options.current).onSelectionUpdate) === null || o === void 0 ? void 0 : o.call(i, ...r);
      },
      onTransaction: (...r) => {
        var i, o;
        return (o = (i = this.options.current).onTransaction) === null || o === void 0 ? void 0 : o.call(i, ...r);
      },
      onUpdate: (...r) => {
        var i, o;
        return (o = (i = this.options.current).onUpdate) === null || o === void 0 ? void 0 : o.call(i, ...r);
      },
      onContentError: (...r) => {
        var i, o;
        return (o = (i = this.options.current).onContentError) === null || o === void 0 ? void 0 : o.call(i, ...r);
      },
      onDrop: (...r) => {
        var i, o;
        return (o = (i = this.options.current).onDrop) === null || o === void 0 ? void 0 : o.call(i, ...r);
      },
      onPaste: (...r) => {
        var i, o;
        return (o = (i = this.options.current).onPaste) === null || o === void 0 ? void 0 : o.call(i, ...r);
      }
    };
    return new pe(e);
  }
  /**
   * Get the current editor instance.
   */
  getEditor() {
    return this.editor;
  }
  /**
   * Always disable the editor on the server-side.
   */
  getServerSnapshot() {
    return null;
  }
  /**
   * Subscribe to the editor instance's changes.
   */
  subscribe(e) {
    return this.subscriptions.add(e), () => {
      this.subscriptions.delete(e);
    };
  }
  static compareOptions(e, t) {
    return Object.keys(e).every((r) => ["onCreate", "onBeforeCreate", "onDestroy", "onUpdate", "onTransaction", "onFocus", "onBlur", "onSelectionUpdate", "onContentError", "onDrop", "onPaste"].includes(r) ? !0 : r === "extensions" && e.extensions && t.extensions ? e.extensions.length !== t.extensions.length ? !1 : e.extensions.every((i, o) => {
      var l;
      return i === ((l = t.extensions) === null || l === void 0 ? void 0 : l[o]);
    }) : e[r] === t[r]);
  }
  /**
   * On each render, we will create, update, or destroy the editor instance.
   * @param deps The dependencies to watch for changes
   * @returns A cleanup function
   */
  onRender(e) {
    return () => (this.isComponentMounted = !0, clearTimeout(this.scheduledDestructionTimeout), this.editor && !this.editor.isDestroyed && e.length === 0 ? H.compareOptions(this.options.current, this.editor.options) || this.editor.setOptions({
      ...this.options.current,
      editable: this.editor.isEditable
    }) : this.refreshEditorInstance(e), () => {
      this.isComponentMounted = !1, this.scheduleDestroy();
    });
  }
  /**
   * Recreate the editor instance if the dependencies have changed.
   */
  refreshEditorInstance(e) {
    if (this.editor && !this.editor.isDestroyed) {
      if (this.previousDeps === null) {
        this.previousDeps = e;
        return;
      }
      if (this.previousDeps.length === e.length && this.previousDeps.every((r, i) => r === e[i]))
        return;
    }
    this.editor && !this.editor.isDestroyed && this.editor.destroy(), this.setEditor(this.createEditor()), this.previousDeps = e;
  }
  /**
   * Schedule the destruction of the editor instance.
   * This will only destroy the editor if it was not mounted on the next tick.
   * This is to avoid destroying the editor instance when it's actually still mounted.
   */
  scheduleDestroy() {
    const e = this.instanceId, t = this.editor;
    this.scheduledDestructionTimeout = setTimeout(() => {
      if (this.isComponentMounted && this.instanceId === e) {
        t && t.setOptions(this.options.current);
        return;
      }
      t && !t.isDestroyed && (t.destroy(), this.instanceId === e && this.setEditor(null));
    }, 1);
  }
}
function Ge(n = {}, e = []) {
  const t = ue(n);
  t.current = n;
  const [r] = W(() => new H(t)), i = M.useSyncExternalStore(r.subscribe, r.getEditor, r.getServerSnapshot);
  return Z(i), z(r.onRender(e)), Ae({
    editor: i,
    selector: ({ transactionNumber: o }) => n.shouldRerenderOnTransaction === !1 ? null : n.immediatelyRender && o === 0 ? 0 : o + 1
  }), i;
}
const ne = te({
  editor: null
});
ne.Consumer;
const Ve = () => ee(ne), Ke = (n) => {
  const [e, t] = W(null), { editor: r } = Ve();
  return z(() => {
    var i;
    if (!e || !((i = n.editor) === null || i === void 0) && i.isDestroyed || r?.isDestroyed)
      return;
    const { pluginKey: o = "bubbleMenu", editor: l, tippyOptions: S = {}, updateDelay: O, shouldShow: d = null } = n, a = l || r;
    if (!a) {
      console.warn("BubbleMenu component is not rendered inside of an editor component or does not have editor prop.");
      return;
    }
    const s = se({
      updateDelay: O,
      editor: a,
      element: e,
      pluginKey: o,
      shouldShow: d,
      tippyOptions: S
    });
    return a.registerPlugin(s), () => {
      a.unregisterPlugin(o);
    };
  }, [n.editor, r, e]), v.createElement("div", { ref: t, className: n.className, style: { visibility: "hidden" } }, n.children);
}, re = te({
  onDragStart: void 0
}), ie = () => ee(re), qe = (n) => {
  const e = n.as || "div", { nodeViewContentRef: t } = ie();
  return (
    // @ts-ignore
    v.createElement(e, { ...n, ref: t, "data-node-view-content": "", style: {
      whiteSpace: "pre-wrap",
      ...n.style
    } })
  );
}, ke = v.forwardRef((n, e) => {
  const { onDragStart: t } = ie(), r = n.as || "div";
  return (
    // @ts-ignore
    v.createElement(r, { ...n, ref: e, "data-node-view-wrapper": "", onDragStart: t, style: {
      whiteSpace: "normal",
      ...n.style
    } })
  );
});
function Q(n) {
  return !!(typeof n == "function" && n.prototype && n.prototype.isReactComponent);
}
function X(n) {
  return !!(typeof n == "object" && n.$$typeof && (n.$$typeof.toString() === "Symbol(react.forward_ref)" || n.$$typeof.description === "react.forward_ref"));
}
function Ie(n) {
  return !!(typeof n == "object" && n.$$typeof && (n.$$typeof.toString() === "Symbol(react.memo)" || n.$$typeof.description === "react.memo"));
}
function Pe(n) {
  if (Q(n) || X(n))
    return !0;
  if (Ie(n)) {
    const e = n.type;
    if (e)
      return Q(e) || X(e);
  }
  return !1;
}
function Be() {
  try {
    if (G)
      return parseInt(G.split(".")[0], 10) >= 19;
  } catch {
  }
  return !1;
}
class je {
  /**
   * Immediately creates element and renders the provided React component.
   */
  constructor(e, { editor: t, props: r = {}, as: i = "div", className: o = "" }) {
    this.ref = null, this.id = Math.floor(Math.random() * 4294967295).toString(), this.component = e, this.editor = t, this.props = r, this.element = document.createElement(i), this.element.classList.add("react-renderer"), o && this.element.classList.add(...o.split(" ")), queueMicrotask(() => {
      this.render();
    });
  }
  /**
   * Render the React component.
   */
  render() {
    var e;
    const t = this.component, r = this.props, i = this.editor, o = Be(), l = Pe(t), S = { ...r };
    S.ref && !(o || l) && delete S.ref, !S.ref && (o || l) && (S.ref = (O) => {
      this.ref = O;
    }), this.reactElement = v.createElement(t, { ...S }), (e = i?.contentComponent) === null || e === void 0 || e.setRenderer(this.id, this);
  }
  /**
   * Re-renders the React component with new props.
   */
  updateProps(e = {}) {
    this.props = {
      ...this.props,
      ...e
    }, this.render();
  }
  /**
   * Destroy the React component.
   */
  destroy() {
    var e;
    const t = this.editor;
    (e = t?.contentComponent) === null || e === void 0 || e.removeRenderer(this.id);
  }
  /**
   * Update the attributes of the element that holds the React component.
   */
  updateAttributes(e) {
    Object.keys(e).forEach((t) => {
      this.element.setAttribute(t, e[t]);
    });
  }
}
class Fe extends me {
  constructor(e, t, r) {
    super(e, t, r), this.node.isLeaf || (this.options.contentDOMElementTag ? this.contentDOMElement = document.createElement(this.options.contentDOMElementTag) : this.contentDOMElement = document.createElement(this.node.isInline ? "span" : "div"), this.contentDOMElement.dataset.nodeViewContentReact = "", this.contentDOMElement.dataset.nodeViewWrapper = "", this.contentDOMElement.style.whiteSpace = "inherit", this.dom.appendChild(this.contentDOMElement));
  }
  /**
   * Setup the React component.
   * Called on initialization.
   */
  mount() {
    const e = {
      editor: this.editor,
      node: this.node,
      decorations: this.decorations,
      innerDecorations: this.innerDecorations,
      view: this.view,
      selected: !1,
      extension: this.extension,
      HTMLAttributes: this.HTMLAttributes,
      getPos: () => this.getPos(),
      updateAttributes: (d = {}) => this.updateAttributes(d),
      deleteNode: () => this.deleteNode(),
      ref: de()
    };
    if (!this.component.displayName) {
      const d = (a) => a.charAt(0).toUpperCase() + a.substring(1);
      this.component.displayName = d(this.extension.name);
    }
    const i = { onDragStart: this.onDragStart.bind(this), nodeViewContentRef: (d) => {
      d && this.contentDOMElement && d.firstChild !== this.contentDOMElement && (d.hasAttribute("data-node-view-wrapper") && d.removeAttribute("data-node-view-wrapper"), d.appendChild(this.contentDOMElement));
    } }, o = this.component, l = le((d) => v.createElement(re.Provider, { value: i }, fe(o, d)));
    l.displayName = "ReactNodeView";
    let S = this.node.isInline ? "span" : "div";
    this.options.as && (S = this.options.as);
    const { className: O = "" } = this.options;
    this.handleSelectionUpdate = this.handleSelectionUpdate.bind(this), this.renderer = new je(l, {
      editor: this.editor,
      props: e,
      as: S,
      className: `node-${this.node.type.name} ${O}`.trim()
    }), this.editor.on("selectionUpdate", this.handleSelectionUpdate), this.updateElementAttributes();
  }
  /**
   * Return the DOM element.
   * This is the element that will be used to display the node view.
   */
  get dom() {
    var e;
    if (this.renderer.element.firstElementChild && !(!((e = this.renderer.element.firstElementChild) === null || e === void 0) && e.hasAttribute("data-node-view-wrapper")))
      throw Error("Please use the NodeViewWrapper component for your node view.");
    return this.renderer.element;
  }
  /**
   * Return the content DOM element.
   * This is the element that will be used to display the rich-text content of the node.
   */
  get contentDOM() {
    return this.node.isLeaf ? null : this.contentDOMElement;
  }
  /**
   * On editor selection update, check if the node is selected.
   * If it is, call `selectNode`, otherwise call `deselectNode`.
   */
  handleSelectionUpdate() {
    const { from: e, to: t } = this.editor.state.selection, r = this.getPos();
    if (typeof r == "number")
      if (e <= r && t >= r + this.node.nodeSize) {
        if (this.renderer.props.selected)
          return;
        this.selectNode();
      } else {
        if (!this.renderer.props.selected)
          return;
        this.deselectNode();
      }
  }
  /**
   * On update, update the React component.
   * To prevent unnecessary updates, the `update` option can be used.
   */
  update(e, t, r) {
    const i = (o) => {
      this.renderer.updateProps(o), typeof this.options.attrs == "function" && this.updateElementAttributes();
    };
    if (e.type !== this.node.type)
      return !1;
    if (typeof this.options.update == "function") {
      const o = this.node, l = this.decorations, S = this.innerDecorations;
      return this.node = e, this.decorations = t, this.innerDecorations = r, this.options.update({
        oldNode: o,
        oldDecorations: l,
        newNode: e,
        newDecorations: t,
        oldInnerDecorations: S,
        innerDecorations: r,
        updateProps: () => i({ node: e, decorations: t, innerDecorations: r })
      });
    }
    return e === this.node && this.decorations === t && this.innerDecorations === r || (this.node = e, this.decorations = t, this.innerDecorations = r, i({ node: e, decorations: t, innerDecorations: r })), !0;
  }
  /**
   * Select the node.
   * Add the `selected` prop and the `ProseMirror-selectednode` class.
   */
  selectNode() {
    this.renderer.updateProps({
      selected: !0
    }), this.renderer.element.classList.add("ProseMirror-selectednode");
  }
  /**
   * Deselect the node.
   * Remove the `selected` prop and the `ProseMirror-selectednode` class.
   */
  deselectNode() {
    this.renderer.updateProps({
      selected: !1
    }), this.renderer.element.classList.remove("ProseMirror-selectednode");
  }
  /**
   * Destroy the React component instance.
   */
  destroy() {
    this.renderer.destroy(), this.editor.off("selectionUpdate", this.handleSelectionUpdate), this.contentDOMElement = null;
  }
  /**
   * Update the attributes of the top-level element that holds the React component.
   * Applying the attributes defined in the `attrs` option.
   */
  updateElementAttributes() {
    if (this.options.attrs) {
      let e = {};
      if (typeof this.options.attrs == "function") {
        const t = this.editor.extensionManager.attributes, r = ve(this.node, t);
        e = this.options.attrs({ node: this.node, HTMLAttributes: r });
      } else
        e = this.options.attrs;
      this.renderer.updateAttributes(e);
    }
  }
}
function Ye(n, e) {
  return (t) => t.editor.contentComponent ? new Fe(n, t, e) : {};
}
export {
  Ke as BubbleMenu,
  Xe as CommandManager,
  pe as Editor,
  He as EditorContent,
  ne as EditorContext,
  Ze as Extension,
  et as InputRule,
  tt as Mark,
  nt as Node,
  rt as NodePos,
  me as NodeView,
  qe as NodeViewContent,
  ke as NodeViewWrapper,
  it as PasteRule,
  be as PureEditorContent,
  Fe as ReactNodeView,
  re as ReactNodeViewContext,
  Ye as ReactNodeViewRenderer,
  je as ReactRenderer,
  ot as callOrReturn,
  st as combineTransactionSteps,
  ut as createChainableState,
  at as createDocument,
  ct as createNodeFromContent,
  dt as createStyleTag,
  lt as defaultBlockAt,
  ft as deleteProps,
  ht as elementFromString,
  pt as escapeForRegEx,
  mt as findChildren,
  vt as findChildrenInRange,
  St as findDuplicates,
  Et as findParentNode,
  _t as findParentNodeClosestToPos,
  yt as fromString,
  gt as generateHTML,
  Ot as getAttributes,
  bt as getAttributesFromExtensions,
  Rt as getChangedRanges,
  wt as getExtensionField,
  Dt as getHTMLFromFragment,
  Ct as getMarkAttributes,
  Tt as getMarkRange,
  Mt as getMarkType,
  xt as getMarksBetween,
  Nt as getNodeAttributes,
  At as getNodeType,
  ve as getRenderedAttributes,
  Lt as getSchema,
  Vt as getSchemaByResolvedExtensions,
  It as getSchemaTypeByName,
  Pt as getSchemaTypeNameByName,
  Bt as getSplittedAttributes,
  jt as getText,
  Ft as getTextBetween,
  Ut as getTextContentFromNodes,
  $t as getTextSerializersFromSchema,
  Wt as injectExtensionAttributesToParseRule,
  zt as inputRulesPlugin,
  Ht as isActive,
  Gt as isEmptyObject,
  Kt as isExtensionRulesEnabled,
  qt as isFunction,
  kt as isList,
  Yt as isMacOS,
  Jt as isMarkActive,
  Qt as isNodeActive,
  Xt as isNodeEmpty,
  Zt as isNodeSelection,
  en as isNumber,
  tn as isPlainObject,
  nn as isRegExp,
  rn as isTextSelection,
  on as isiOS,
  sn as markInputRule,
  un as markPasteRule,
  an as mergeAttributes,
  cn as mergeDeep,
  dn as minMax,
  ln as nodeInputRule,
  fn as nodePasteRule,
  hn as objectIncludes,
  pn as pasteRulesPlugin,
  mn as posToDOMRect,
  vn as removeDuplicates,
  Sn as resolveFocusPosition,
  En as selectionToInsertionEnd,
  _n as splitExtensions,
  yn as textInputRule,
  gn as textblockTypeInputRule,
  Ve as useCurrentEditor,
  Ge as useEditor,
  Ae as useEditorState,
  ie as useReactNodeView,
  On as wrappingInputRule
};
