// echarts@6.0.0/lib/component/dataset/install.js
import { __extends as r } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import p from "../../model/Component.js";
import s from "../../view/Component.js";
import { SERIES_LAYOUT_BY_COLUMN as m } from "../../util/types.js";
import { SourceManager as u, disableTransformOptionMerge as i } from "../../data/helper/sourceManager.js";
var l = (
  /** @class */
  (function(t) {
    r(e, t);
    function e() {
      var o = t !== null && t.apply(this, arguments) || this;
      return o.type = "dataset", o;
    }
    return e.prototype.init = function(o, n, a) {
      t.prototype.init.call(this, o, n, a), this._sourceManager = new u(this), i(this);
    }, e.prototype.mergeOption = function(o, n) {
      t.prototype.mergeOption.call(this, o, n), i(this);
    }, e.prototype.optionUpdated = function() {
      this._sourceManager.dirty();
    }, e.prototype.getSourceManager = function() {
      return this._sourceManager;
    }, e.type = "dataset", e.defaultOption = {
      seriesLayoutBy: m
    }, e;
  })(p)
), c = (
  /** @class */
  (function(t) {
    r(e, t);
    function e() {
      var o = t !== null && t.apply(this, arguments) || this;
      return o.type = "dataset", o;
    }
    return e.type = "dataset", e;
  })(s)
);
function M(t) {
  t.registerComponentModel(l), t.registerComponentView(c);
}
export {
  l as DatasetModel,
  M as install
};
