// @atlaskit/pragmatic-drag-and-drop-react-drop-indicator@3.2.11/dist/esm/internal/line.js
import './line.compiled.css';/* empty css                  */
import * as _ from "react";
import { presetStrokeWidth as p, presetStrokeColors as f } from "../presets.js";
import h from "../../../../../../../@compiled_react@0.18.6_react@18.3.1/node_modules/@compiled/react/dist/esm/runtime/ax.js";
var k = {
  top: "horizontal",
  bottom: "horizontal",
  left: "vertical",
  right: "vertical"
}, u = {
  root: "_1e0c1ule _kqswstnw _1pbykb7n _lcxvglyw _bfhkys7w _rfx31ssb _3l8810ly _kzdanqa1 _15m6ys7w _cfu11ld9 _1kt9b3bt _1cs8stnw _13y0usvi _1mp4vjfa _kfgtvjfa"
}, y = {
  horizontal: "_4t3i10ly _1e02fghn _rjxpidpf _z5wtuj5p",
  vertical: "_1bsb10ly _154ifghn _94n5idpf _1aukuj5p"
}, b = {
  top: "_154ihv0e _1auk70hn",
  right: "_1xi2hv0e _ooun70hn",
  bottom: "_94n5hv0e _19wo70hn",
  left: "_1ltvhv0e _qnec70hn"
}, g = {
  // - half the terminal bleeding out the containing element
  // - half the terminal inside the containing element (we need to position the line next to this)
  terminal: function(e) {
    var r = e.indent;
    return "calc(var(--terminal-radius) + ".concat(r, ")");
  },
  // The full terminal is inside the containing element (we need to position the line next to this)
  "terminal-no-bleed": function(e) {
    var r = e.indent;
    return "calc(var(--terminal-diameter) + ".concat(r, ")");
  },
  // No terminal to worry about, line should take up all the space
  "no-terminal": function(e) {
    var r = e.indent;
    return r;
  }
};
function j(t) {
  var e = t.edge, r = t.gap, s = r === void 0 ? "0px" : r, a = t.indent, d = a === void 0 ? "0px" : a, i = t.strokeColor, c = i === void 0 ? f.default : i, n = t.strokeWidth, m = n === void 0 ? p : n, o = t.type, l = o === void 0 ? "terminal" : o, v = k[e];
  return /* @__PURE__ */ _.createElement("div", {
    style: {
      // ## All
      "--stroke-color": c,
      "--stroke-width": m,
      // Shift line and terminal on the main access to account for gaps between items
      "--main-axis-offset": "calc(-0.5 * (".concat(s, " + var(--stroke-width)))"),
      // ## Line
      // If there is a terminal, we want the line to start from next to it
      "--line-main-axis-start": g[l]({
        indent: d
      }),
      // ## Terminal
      "--terminal-display": l === "no-terminal" ? "none" : "block",
      "--terminal-diameter": "calc(var(--stroke-width) * 4)",
      "--terminal-radius": "calc(var(--terminal-diameter) / 2)",
      // The line is positioned to account for the the terminal (--line-main-axis-start).
      // The terminal is rendered relative to the line (it's a `::before`)
      // We need to pull the terminal backwards so it sits before the start of the line
      "--terminal-main-axis-start": "calc(-1 * var(--terminal-diameter))",
      // Pull the terminal backwards on the cross axis (eg "up" on "vertical")
      // so the center of the terminal lines up with the center of the line
      "--terminal-cross-axis-offset": "calc(calc(var(--stroke-width) - var(--terminal-diameter)) / 2)"
    },
    className: h([u.root, y[v], b[e]])
  });
}
export {
  j as Line,
  j as default
};
