// echarts@6.0.0/lib/coord/parallel/Parallel.js
import { createHashMap as C, each as E, clone as b } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { rotate as M, create as P, translate as S } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/matrix.js";
import { createBoxLayoutReference as B, getLayoutRect as I } from "../../util/layout.js";
import { createScaleByModel as D, niceScaleExtent as T } from "../axisHelper.js";
import k from "./ParallelAxis.js";
import { applyTransform as N } from "../../util/graphic.js";
import { round as F } from "../../util/number.js";
import z from "../../component/helper/sliderMove.js";
var f = E, _ = Math.min, L = Math.max, g = Math.floor, R = Math.ceil, W = F, $ = Math.PI, Y = (
  /** @class */
  (function() {
    function l(a, t, x) {
      this.type = "parallel", this._axesMap = C(), this._axesLayout = {}, this.dimensions = a.dimensions, this._model = a, this._init(a, t, x);
    }
    return l.prototype._init = function(a, t, x) {
      var i = a.dimensions, e = a.parallelAxisIndex;
      f(i, function(n, o) {
        var r = e[o], h = t.getComponent("parallelAxis", r), s = this._axesMap.set(n, new k(n, D(h), [0, 0], h.get("type"), r)), v = s.type === "category";
        s.onBand = v && h.get("boundaryGap"), s.inverse = h.get("inverse"), h.axis = s, s.model = h, s.coordinateSystem = h.coordinateSystem = this;
      }, this);
    }, l.prototype.update = function(a, t) {
      this._updateAxesFromSeries(this._model, a);
    }, l.prototype.containPoint = function(a) {
      var t = this._makeLayoutInfo(), x = t.axisBase, i = t.layoutBase, e = t.pixelDimIndex, n = a[1 - e], o = a[e];
      return n >= x && n <= x + t.axisLength && o >= i && o <= i + t.layoutLength;
    }, l.prototype.getModel = function() {
      return this._model;
    }, l.prototype._updateAxesFromSeries = function(a, t) {
      t.eachSeries(function(x) {
        if (a.contains(x, t)) {
          var i = x.getData();
          f(this.dimensions, function(e) {
            var n = this._axesMap.get(e);
            n.scale.unionExtentFromData(i, i.mapDimension(e)), T(n.scale, n.model);
          }, this);
        }
      }, this);
    }, l.prototype.resize = function(a, t) {
      var x = B(a, t).refContainer;
      this._rect = I(a.getBoxLayoutParams(), x), this._layoutAxes();
    }, l.prototype.getRect = function() {
      return this._rect;
    }, l.prototype._makeLayoutInfo = function() {
      var a = this._model, t = this._rect, x = ["x", "y"], i = ["width", "height"], e = a.get("layout"), n = e === "horizontal" ? 0 : 1, o = t[i[n]], r = [0, o], h = this.dimensions.length, s = m(a.get("axisExpandWidth"), r), v = m(a.get("axisExpandCount") || 0, [0, h]), u = a.get("axisExpandable") && h > 3 && h > v && v > 1 && s > 0 && o > 0, p = a.get("axisExpandWindow"), d;
      if (p)
        d = m(p[1] - p[0], r), p[1] = p[0] + d;
      else {
        d = m(s * (v - 1), r);
        var y = a.get("axisExpandCenter") || g(h / 2);
        p = [s * y - d / 2], p[1] = p[0] + d;
      }
      var c = (o - d) / (h - v);
      c < 3 && (c = 0);
      var A = [g(W(p[0] / s, 1)) + 1, R(W(p[1] / s, 1)) - 1], w = c / s * p[0];
      return {
        layout: e,
        pixelDimIndex: n,
        layoutBase: t[x[n]],
        layoutLength: o,
        axisBase: t[x[1 - n]],
        axisLength: t[i[1 - n]],
        axisExpandable: u,
        axisExpandWidth: s,
        axisCollapseWidth: c,
        axisExpandWindow: p,
        axisCount: h,
        winInnerIndices: A,
        axisExpandWindow0Pos: w
      };
    }, l.prototype._layoutAxes = function() {
      var a = this._rect, t = this._axesMap, x = this.dimensions, i = this._makeLayoutInfo(), e = i.layout;
      t.each(function(n) {
        var o = [0, i.axisLength], r = n.inverse ? 1 : 0;
        n.setExtent(o[r], o[1 - r]);
      }), f(x, function(n, o) {
        var r = (i.axisExpandable ? H : G)(o, i), h = {
          horizontal: {
            x: r.position,
            y: i.axisLength
          },
          vertical: {
            x: 0,
            y: r.position
          }
        }, s = {
          horizontal: $ / 2,
          vertical: 0
        }, v = [h[e].x + a.x, h[e].y + a.y], u = s[e], p = P();
        M(p, p, u), S(p, p, v), this._axesLayout[n] = {
          position: v,
          rotation: u,
          transform: p,
          axisNameAvailableWidth: r.axisNameAvailableWidth,
          axisLabelShow: r.axisLabelShow,
          nameTruncateMaxWidth: r.nameTruncateMaxWidth,
          tickDirection: 1,
          labelDirection: 1
        };
      }, this);
    }, l.prototype.getAxis = function(a) {
      return this._axesMap.get(a);
    }, l.prototype.dataToPoint = function(a, t) {
      return this.axisCoordToPoint(this._axesMap.get(t).dataToCoord(a), t);
    }, l.prototype.eachActiveState = function(a, t, x, i) {
      x == null && (x = 0), i == null && (i = a.count());
      var e = this._axesMap, n = this.dimensions, o = [], r = [];
      E(n, function(c) {
        o.push(a.mapDimension(c)), r.push(e.get(c).model);
      });
      for (var h = this.hasAxisBrushed(), s = x; s < i; s++) {
        var v = void 0;
        if (!h)
          v = "normal";
        else {
          v = "active";
          for (var u = a.getValues(o, s), p = 0, d = n.length; p < d; p++) {
            var y = r[p].getActiveState(u[p]);
            if (y === "inactive") {
              v = "inactive";
              break;
            }
          }
        }
        t(v, s);
      }
    }, l.prototype.hasAxisBrushed = function() {
      for (var a = this.dimensions, t = this._axesMap, x = !1, i = 0, e = a.length; i < e; i++)
        t.get(a[i]).model.getActiveState() !== "normal" && (x = !0);
      return x;
    }, l.prototype.axisCoordToPoint = function(a, t) {
      var x = this._axesLayout[t];
      return N([a, 0], x.transform);
    }, l.prototype.getAxisLayout = function(a) {
      return b(this._axesLayout[a]);
    }, l.prototype.getSlidedAxisExpandWindow = function(a) {
      var t = this._makeLayoutInfo(), x = t.pixelDimIndex, i = t.axisExpandWindow.slice(), e = i[1] - i[0], n = [0, t.axisExpandWidth * (t.axisCount - 1)];
      if (!this.containPoint(a))
        return {
          behavior: "none",
          axisExpandWindow: i
        };
      var o = a[x] - t.layoutBase - t.axisExpandWindow0Pos, r, h = "slide", s = t.axisCollapseWidth, v = this._model.get("axisExpandSlideTriggerArea"), u = v[0] != null;
      if (s)
        u && s && o < e * v[0] ? (h = "jump", r = o - e * v[2]) : u && s && o > e * (1 - v[0]) ? (h = "jump", r = o - e * (1 - v[2])) : (r = o - e * v[1]) >= 0 && (r = o - e * (1 - v[1])) <= 0 && (r = 0), r *= t.axisExpandWidth / s, r ? z(r, i, n, "all") : h = "none";
      else {
        var p = i[1] - i[0], d = n[1] * o / p;
        i = [L(0, d - p / 2)], i[1] = _(n[1], i[0] + p), i[0] = i[1] - p;
      }
      return {
        axisExpandWindow: i,
        behavior: h
      };
    }, l;
  })()
);
function m(l, a) {
  return _(L(l, a[0]), a[1]);
}
function G(l, a) {
  var t = a.layoutLength / (a.axisCount - 1);
  return {
    position: t * l,
    axisNameAvailableWidth: t,
    axisLabelShow: !0
  };
}
function H(l, a) {
  var t = a.layoutLength, x = a.axisExpandWidth, i = a.axisCount, e = a.axisCollapseWidth, n = a.winInnerIndices, o, r = e, h = !1, s;
  return l < n[0] ? (o = l * e, s = e) : l <= n[1] ? (o = a.axisExpandWindow0Pos + l * x - a.axisExpandWindow[0], r = x, h = !0) : (o = t - (i - 1 - l) * e, s = e), {
    position: o,
    axisNameAvailableWidth: r,
    axisLabelShow: h,
    nameTruncateMaxWidth: s
  };
}
export {
  Y as default
};
