// echarts@6.0.0/lib/coord/parallel/ParallelModel.js
import { __extends as p } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { merge as r, each as l, filter as s } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import x from "../../model/Component.js";
var m = (
  /** @class */
  (function(i) {
    p(t, i);
    function t() {
      var n = i !== null && i.apply(this, arguments) || this;
      return n.type = t.type, n;
    }
    return t.prototype.init = function() {
      i.prototype.init.apply(this, arguments), this.mergeOption({});
    }, t.prototype.mergeOption = function(n) {
      var e = this.option;
      n && r(e, n, !0), this._initDimensions();
    }, t.prototype.contains = function(n, e) {
      var a = n.get("parallelIndex");
      return a != null && e.getComponent("parallel", a) === this;
    }, t.prototype.setAxisExpand = function(n) {
      l(["axisExpandable", "axisExpandCenter", "axisExpandCount", "axisExpandWidth", "axisExpandWindow"], function(e) {
        n.hasOwnProperty(e) && (this.option[e] = n[e]);
      }, this);
    }, t.prototype._initDimensions = function() {
      var n = this.dimensions = [], e = this.parallelAxisIndex = [], a = s(this.ecModel.queryComponents({
        mainType: "parallelAxis"
      }), function(o) {
        return (o.get("parallelIndex") || 0) === this.componentIndex;
      }, this);
      l(a, function(o) {
        n.push("dim" + o.get("dim")), e.push(o.componentIndex);
      });
    }, t.type = "parallel", t.dependencies = ["parallelAxis"], t.layoutMode = "box", t.defaultOption = {
      // zlevel: 0,
      z: 0,
      left: 80,
      top: 60,
      right: 80,
      bottom: 60,
      // width: {totalWidth} - left - right,
      // height: {totalHeight} - top - bottom,
      layout: "horizontal",
      // FIXME
      // naming?
      axisExpandable: !1,
      axisExpandCenter: null,
      axisExpandCount: 0,
      axisExpandWidth: 50,
      axisExpandRate: 17,
      axisExpandDebounce: 50,
      // [out, in, jumpTarget]. In percentage. If use [null, 0.05], null means full.
      // Do not doc to user until necessary.
      axisExpandSlideTriggerArea: [-0.15, 0.05, 0.4],
      axisExpandTriggerOn: "click",
      parallelAxisDefault: null
    }, t;
  })(x)
);
export {
  m as default
};
