// echarts@6.0.0/lib/component/tooltip/helper.js
import { toCamelCase as o } from "../../util/format.js";
import u from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/env.js";
function m(n) {
  var t = n.get("confine");
  return t != null ? !!t : n.get("renderMode") === "richText";
}
function e(n) {
  if (u.domSupported) {
    for (var t = document.documentElement.style, r = 0, i = n.length; r < i; r++)
      if (n[r] in t)
        return n[r];
  }
}
var c = e(["transform", "webkitTransform", "OTransform", "MozTransform", "msTransform"]), T = e(["webkitTransition", "transition", "OTransition", "MozTransition", "msTransition"]);
function d(n, t) {
  if (!n)
    return t;
  t = o(t, !0);
  var r = n.indexOf(t);
  return n = r === -1 ? t : "-" + n.slice(0, r) + "-" + t, n.toLowerCase();
}
function S(n, t) {
  var r = n.currentStyle || document.defaultView && document.defaultView.getComputedStyle(n);
  return r ? r[t] : null;
}
export {
  c as TRANSFORM_VENDOR,
  T as TRANSITION_VENDOR,
  S as getComputedStyle,
  m as shouldTooltipConfine,
  d as toCSSVendorPrefix
};
