// echarts@6.0.0/lib/component/tooltip/seriesFormatTooltip.js
import { isArray as N, trim as D, reduce as b, each as k } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { retrieveVisualColorForTooltipMarker as S, createTooltipMarkup as d } from "./tooltipMarkup.js";
import { retrieveRawValue as I } from "../../data/helper/dataProvider.js";
import { isNameSpecified as x } from "../../util/model.js";
function A(s) {
  var r = s.series, i = s.dataIndex, u = s.multipleSeries, m = r.getData(), l = m.mapDimensionsAll("defaultedTooltip"), p = l.length, t = r.getRawValue(i), v = N(t), f = S(r, i), o, e, c, a;
  if (p > 1 || v && !p) {
    var n = C(t, r, i, l, f);
    o = n.inlineValues, e = n.inlineValueTypes, c = n.blocks, a = n.inlineValues[0];
  } else if (p) {
    var g = m.getDimensionInfo(l[0]);
    a = o = I(m, i, l[0]), e = g.type;
  } else
    a = o = v ? t[0] : t;
  var V = x(r), y = V && r.name || "", h = m.getName(i), T = u ? y : h;
  return d("section", {
    header: y,
    // When series name is not specified, do not show a header line with only '-'.
    // This case always happens in tooltip.trigger: 'item'.
    noHeader: u || !V,
    sortParam: a,
    blocks: [d("nameValue", {
      markerType: "item",
      markerColor: f,
      // Do not mix display seriesName and itemName in one tooltip,
      // which might confuses users.
      name: T,
      // name dimension might be auto assigned, where the name might
      // be not readable. So we check trim here.
      noName: !D(T),
      value: o,
      valueType: e,
      dataIndex: i
    })].concat(c || [])
  });
}
function C(s, r, i, u, m) {
  var l = r.getData(), p = b(s, function(e, c, a) {
    var n = l.getDimensionInfo(a);
    return e = e || n && n.tooltip !== !1 && n.displayName != null;
  }, !1), t = [], v = [], f = [];
  u.length ? k(u, function(e) {
    o(I(l, i, e), e);
  }) : k(s, o);
  function o(e, c) {
    var a = l.getDimensionInfo(c);
    !a || a.otherDims.tooltip === !1 || (p ? f.push(d("nameValue", {
      markerType: "subItem",
      markerColor: m,
      name: a.displayName,
      value: e,
      valueType: a.type
    })) : (t.push(e), v.push(a.type)));
  }
  return {
    inlineValues: t,
    inlineValueTypes: v,
    blocks: f
  };
}
export {
  A as defaultSeriesFormatTooltip
};
