// echarts@6.0.0/lib/component/tooltip/tooltipMarkup.js
import { convertToColorString as W, makeValueReadable as C, getTooltipMarker as A } from "../../util/format.js";
import { assert as _, isArray as m, hasOwn as D, each as I, extend as g, map as w, isString as G } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { SortOrderComparator as O } from "../../data/helper/dataValueHelper.js";
import { getRandomIdBase as U } from "../../util/number.js";
import H from "../../visual/tokens.js";
import { encodeHTML as u } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/dom.js";
var j = "line-height:1";
function x(e) {
  var r = e.lineHeight;
  return r == null ? j : "line-height:" + u(r + "") + "px";
}
function V(e, r) {
  var a = e.color || H.color.tertiary, n = e.fontSize || 12, o = e.fontWeight || "400", t = e.color || H.color.secondary, l = e.fontSize || 14, i = e.fontWeight || "900";
  return r === "html" ? {
    // eslint-disable-next-line max-len
    nameStyle: "font-size:" + u(n + "") + "px;color:" + u(a) + ";font-weight:" + u(o + ""),
    // eslint-disable-next-line max-len
    valueStyle: "font-size:" + u(l + "") + "px;color:" + u(t) + ";font-weight:" + u(i + "")
  } : {
    nameStyle: {
      fontSize: n,
      fill: a,
      fontWeight: o
    },
    valueStyle: {
      fontSize: l,
      fill: t,
      fontWeight: i
    }
  };
}
var B = [0, 10, 20, 30], X = ["", `
`, `

`, `


`];
function oe(e, r) {
  return r.type = e, r;
}
function F(e) {
  return e.type === "section";
}
function R(e) {
  return F(e) ? q : J;
}
function z(e) {
  if (F(e)) {
    var r = 0, a = e.blocks.length, n = a > 1 || a > 0 && !e.noHeader;
    return I(e.blocks, function(o) {
      var t = z(o);
      t >= r && (r = t + +(n && // 0 always can not be readable gap level.
      (!t || F(o) && !o.noHeader)));
    }), r;
  }
  return 0;
}
function q(e, r, a, n) {
  var o = r.noHeader, t = K(z(r)), l = [], i = r.blocks || [];
  _(!i || m(i)), i = i || [];
  var p = e.orderMode;
  if (r.sortBlocks && p) {
    i = i.slice();
    var h = {
      valueAsc: "asc",
      valueDesc: "desc"
    };
    if (D(h, p)) {
      var M = new O(h[p], null);
      i.sort(function(v, c) {
        return M.evaluate(v.sortParam, c.sortParam);
      });
    } else p === "seriesDesc" && i.reverse();
  }
  I(i, function(v, c) {
    var T = r.valueFormatter, y = R(v)(
      // Inherit valueFormatter
      T ? g(g({}, e), {
        valueFormatter: T
      }) : e,
      v,
      c > 0 ? t.html : 0,
      n
    );
    y != null && l.push(y);
  });
  var s = e.renderMode === "richText" ? l.join(t.richText) : b(n, l.join(""), o ? a : t.html);
  if (o)
    return s;
  var f = C(r.header, "ordinal", e.useUTC), d = V(n, e.renderMode).nameStyle, k = x(n);
  return e.renderMode === "richText" ? L(e, f, d) + t.richText + s : b(n, '<div style="' + d + ";" + k + ';">' + u(f) + "</div>" + s, a);
}
function J(e, r, a, n) {
  var o = e.renderMode, t = r.noName, l = r.noValue, i = !r.markerType, p = r.name, h = e.useUTC, M = r.valueFormatter || e.valueFormatter || function(S) {
    return S = m(S) ? S : [S], w(S, function(E, P) {
      return C(E, m(d) ? d[P] : d, h);
    });
  };
  if (!(t && l)) {
    var s = i ? "" : e.markupStyleCreator.makeTooltipMarker(r.markerType, r.markerColor || H.color.secondary, o), f = t ? "" : C(p, "ordinal", h), d = r.valueType, k = l ? [] : M(r.value, r.dataIndex), v = !i || !t, c = !i && t, T = V(n, o), y = T.nameStyle, N = T.valueStyle;
    return o === "richText" ? (i ? "" : s) + (t ? "" : L(e, f, y)) + (l ? "" : Z(e, k, v, c, N)) : b(n, (i ? "" : s) + (t ? "" : Q(f, !i, y)) + (l ? "" : Y(k, v, c, N)), a);
  }
}
function ie(e, r, a, n, o, t) {
  if (e) {
    var l = R(e), i = {
      useUTC: o,
      renderMode: a,
      orderMode: n,
      markupStyleCreator: r,
      valueFormatter: e.valueFormatter
    };
    return l(i, e, 0, t);
  }
}
function K(e) {
  return {
    html: B[e],
    richText: X[e]
  };
}
function b(e, r, a) {
  var n = '<div style="clear:both"></div>', o = "margin: " + a + "px 0 0", t = x(e);
  return '<div style="' + o + ";" + t + ';">' + r + n + "</div>";
}
function Q(e, r, a) {
  var n = r ? "margin-left:2px" : "";
  return '<span style="' + a + ";" + n + '">' + u(e) + "</span>";
}
function Y(e, r, a, n) {
  var o = a ? "10px" : "20px", t = r ? "float:right;margin-left:" + o : "";
  return e = m(e) ? e : [e], '<span style="' + t + ";" + n + '">' + w(e, function(l) {
    return u(l);
  }).join("&nbsp;&nbsp;") + "</span>";
}
function L(e, r, a) {
  return e.markupStyleCreator.wrapRichTextStyle(r, a);
}
function Z(e, r, a, n, o) {
  var t = [o], l = n ? 10 : 20;
  return a && t.push({
    padding: [0, 0, 0, l],
    align: "right"
  }), e.markupStyleCreator.wrapRichTextStyle(m(r) ? r.join("  ") : r, t);
}
function le(e, r) {
  var a = e.getData().getItemVisual(r, "style"), n = a[e.visualDrawType];
  return W(n);
}
function ue(e, r) {
  var a = e.get("padding");
  return a ?? (r === "richText" ? [8, 10] : 10);
}
var ve = (
  /** @class */
  (function() {
    function e() {
      this.richTextStyles = {}, this._nextStyleNameId = U();
    }
    return e.prototype._generateStyleName = function() {
      return "__EC_aUTo_" + this._nextStyleNameId++;
    }, e.prototype.makeTooltipMarker = function(r, a, n) {
      var o = n === "richText" ? this._generateStyleName() : null, t = A({
        color: a,
        type: r,
        renderMode: n,
        markerId: o
      });
      return G(t) ? t : (process.env.NODE_ENV !== "production" && _(o), this.richTextStyles[o] = t.style, t.content);
    }, e.prototype.wrapRichTextStyle = function(r, a) {
      var n = {};
      m(a) ? I(a, function(t) {
        return g(n, t);
      }) : g(n, a);
      var o = this._generateStyleName();
      return this.richTextStyles[o] = n, "{" + o + "|" + r + "}";
    }, e;
  })()
);
export {
  ve as TooltipMarkupStyleCreator,
  ie as buildTooltipMarkup,
  oe as createTooltipMarkup,
  ue as getPaddingFromTooltipModel,
  le as retrieveVisualColorForTooltipMarker
};
