// echarts@6.0.0/lib/component/tooltip/TooltipView.js
import { __extends as K } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { bind as B, each as w, trim as $, isString as R, clone as F, isArray as D, isFunction as V, isObject as E, extend as N } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import A from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/env.js";
import tt from "./TooltipHTMLContent.js";
import et from "./TooltipRichContent.js";
import { convertToColorString as U, formatTpl as ot } from "../../util/format.js";
import { parsePercent as W } from "../../util/number.js";
import "../../util/graphic.js";
import it from "../axisPointer/findPointFromSeries.js";
import { getLayoutRect as rt } from "../../util/layout.js";
import P from "../../model/Model.js";
import { register as nt, unregister as at } from "../axisPointer/globalListener.js";
import { getAxisRawValue as st } from "../../coord/axisHelper.js";
import { getValueLabel as lt } from "../axisPointer/viewHelper.js";
import { getTooltipRenderMode as ft, preParseFinder as ut, queryReferringComponents as ht } from "../../util/model.js";
import dt from "../../view/Component.js";
import { format as vt } from "../../util/time.js";
import { getECData as S } from "../../util/innerStore.js";
import { shouldTooltipConfine as pt } from "./helper.js";
import { normalizeTooltipFormatResult as q } from "../../model/mixin/dataFormat.js";
import { createTooltipMarkup as j, TooltipMarkupStyleCreator as Y, buildTooltipMarkup as G } from "./tooltipMarkup.js";
import { findEventDispatcher as mt } from "../../util/event.js";
import { createOrUpdate as gt, clear as Q } from "../../util/throttle.js";
import ct from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Rect.js";
import { encodeHTML as _t } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/dom.js";
var Tt = new ct({
  shape: {
    x: -1,
    y: -1,
    width: 2,
    height: 2
  }
}), Zt = (
  /** @class */
  (function(v) {
    K(a, v);
    function a() {
      var i = v !== null && v.apply(this, arguments) || this;
      return i.type = a.type, i;
    }
    return a.prototype.init = function(i, r) {
      if (!(A.node || !r.getDom())) {
        var e = i.getComponent("tooltip"), t = this._renderMode = ft(e.get("renderMode"));
        this._tooltipContent = t === "richText" ? new et(r) : new tt(r, {
          appendTo: e.get("appendToBody", !0) ? "body" : e.get("appendTo", !0)
        });
      }
    }, a.prototype.render = function(i, r, e) {
      if (!(A.node || !e.getDom())) {
        this.group.removeAll(), this._tooltipModel = i, this._ecModel = r, this._api = e;
        var t = this._tooltipContent;
        t.update(i), t.setEnterable(i.get("enterable")), this._initGlobalListener(), this._keepShow(), this._renderMode !== "richText" && i.get("transitionDuration") ? gt(this, "_updatePosition", 50, "fixRate") : Q(this, "_updatePosition");
      }
    }, a.prototype._initGlobalListener = function() {
      var i = this._tooltipModel, r = i.get("triggerOn");
      nt("itemTooltip", this._api, B(function(e, t, o) {
        r !== "none" && (r.indexOf(e) >= 0 ? this._tryShow(t, o) : e === "leave" && this._hide(o));
      }, this));
    }, a.prototype._keepShow = function() {
      var i = this._tooltipModel, r = this._ecModel, e = this._api, t = i.get("triggerOn");
      if (this._lastX != null && this._lastY != null && t !== "none" && t !== "click") {
        var o = this;
        clearTimeout(this._refreshUpdateTimeout), this._refreshUpdateTimeout = setTimeout(function() {
          !e.isDisposed() && o.manuallyShowTip(i, r, e, {
            x: o._lastX,
            y: o._lastY,
            dataByCoordSys: o._lastDataByCoordSys
          });
        });
      }
    }, a.prototype.manuallyShowTip = function(i, r, e, t) {
      if (!(t.from === this.uid || A.node || !e.getDom())) {
        var o = Z(t, e);
        this._ticket = "";
        var f = t.dataByCoordSys, n = Mt(t, r, e);
        if (n) {
          var s = n.el.getBoundingRect().clone();
          s.applyTransform(n.el.transform), this._tryShow({
            offsetX: s.x + s.width / 2,
            offsetY: s.y + s.height / 2,
            target: n.el,
            position: t.position,
            // When manully trigger, the mouse is not on the el, so we'd better to
            // position tooltip on the bottom of the el and display arrow is possible.
            positionDefault: "bottom"
          }, o);
        } else if (t.tooltip && t.x != null && t.y != null) {
          var l = Tt;
          l.x = t.x, l.y = t.y, l.update(), S(l).tooltipConfig = {
            name: null,
            option: t.tooltip
          }, this._tryShow({
            offsetX: t.x,
            offsetY: t.y,
            target: l
          }, o);
        } else if (f)
          this._tryShow({
            offsetX: t.x,
            offsetY: t.y,
            position: t.position,
            dataByCoordSys: f,
            tooltipOption: t.tooltipOption
          }, o);
        else if (t.seriesIndex != null) {
          if (this._manuallyAxisShowTip(i, r, e, t))
            return;
          var u = it(t, r), h = u.point[0], d = u.point[1];
          h != null && d != null && this._tryShow({
            offsetX: h,
            offsetY: d,
            target: u.el,
            position: t.position,
            // When manully trigger, the mouse is not on the el, so we'd better to
            // position tooltip on the bottom of the el and display arrow is possible.
            positionDefault: "bottom"
          }, o);
        } else t.x != null && t.y != null && (e.dispatchAction({
          type: "updateAxisPointer",
          x: t.x,
          y: t.y
        }), this._tryShow({
          offsetX: t.x,
          offsetY: t.y,
          position: t.position,
          target: e.getZr().findHover(t.x, t.y).target
        }, o));
      }
    }, a.prototype.manuallyHideTip = function(i, r, e, t) {
      var o = this._tooltipContent;
      this._tooltipModel && o.hideLater(this._tooltipModel.get("hideDelay")), this._lastX = this._lastY = this._lastDataByCoordSys = null, t.from !== this.uid && this._hide(Z(t, e));
    }, a.prototype._manuallyAxisShowTip = function(i, r, e, t) {
      var o = t.seriesIndex, f = t.dataIndex, n = r.getComponent("axisPointer").coordSysAxesInfo;
      if (!(o == null || f == null || n == null)) {
        var s = r.getSeriesByIndex(o);
        if (s) {
          var l = s.getData(), u = I([l.getItemModel(f), s, (s.coordinateSystem || {}).model], this._tooltipModel);
          if (u.get("trigger") === "axis")
            return e.dispatchAction({
              type: "updateAxisPointer",
              seriesIndex: o,
              dataIndex: f,
              position: t.position
            }), !0;
        }
      }
    }, a.prototype._tryShow = function(i, r) {
      var e = i.target, t = this._tooltipModel;
      if (t) {
        this._lastX = i.offsetX, this._lastY = i.offsetY;
        var o = i.dataByCoordSys;
        if (o && o.length)
          this._showAxisTooltip(o, i);
        else if (e) {
          var f = S(e);
          if (f.ssrType === "legend")
            return;
          this._lastDataByCoordSys = null;
          var n, s;
          mt(e, function(l) {
            if (l.tooltipDisabled)
              return n = s = null, !0;
            n || s || (S(l).dataIndex != null ? n = l : S(l).tooltipConfig != null && (s = l));
          }, !0), n ? this._showSeriesItemTooltip(i, n, r) : s ? this._showComponentItemTooltip(i, s, r) : this._hide(r);
        } else
          this._lastDataByCoordSys = null, this._hide(r);
      }
    }, a.prototype._showOrMove = function(i, r) {
      var e = i.get("showDelay");
      r = B(r, this), clearTimeout(this._showTimout), e > 0 ? this._showTimout = setTimeout(r, e) : r();
    }, a.prototype._showAxisTooltip = function(i, r) {
      var e = this._ecModel, t = this._tooltipModel, o = [r.offsetX, r.offsetY], f = I([r.tooltipOption], t), n = this._renderMode, s = [], l = j("section", {
        blocks: [],
        noHeader: !0
      }), u = [], h = new Y();
      w(i, function(T) {
        w(T.dataByAxis, function(g) {
          var x = e.getComponent(g.axisDim + "Axis", g.axisIndex), C = g.value;
          if (!(!x || C == null)) {
            var M = lt(C, x.axis, e, g.seriesDataIndices, g.valueLabelOpt), k = j("section", {
              header: M,
              noHeader: !$(M),
              sortBlocks: !0,
              blocks: []
            });
            l.blocks.push(k), w(g.seriesDataIndices, function(b) {
              var L = e.getSeriesByIndex(b.seriesIndex), H = b.dataIndexInside, y = L.getDataParams(H);
              if (!(y.dataIndex < 0)) {
                y.axisDim = g.axisDim, y.axisIndex = g.axisIndex, y.axisType = g.axisType, y.axisId = g.axisId, y.axisValue = st(x.axis, {
                  value: C
                }), y.axisValueLabel = M, y.marker = h.makeTooltipMarker("item", U(y.color), n);
                var O = q(L.formatTooltip(H, !0, null)), X = O.frag;
                if (X) {
                  var z = I([L], t).get("valueFormatter");
                  k.blocks.push(z ? N({
                    valueFormatter: z
                  }, X) : X);
                }
                O.text && u.push(O.text), s.push(y);
              }
            });
          }
        });
      }), l.blocks.reverse(), u.reverse();
      var d = r.position, c = f.get("order"), m = G(l, h, n, c, e.get("useUTC"), f.get("textStyle"));
      m && u.unshift(m);
      var _ = n === "richText" ? `

` : "<br/>", p = u.join(_);
      this._showOrMove(f, function() {
        this._updateContentNotChangedOnAxis(i, s) ? this._updatePosition(f, d, o[0], o[1], this._tooltipContent, s) : this._showTooltipContent(f, p, s, Math.random() + "", o[0], o[1], d, null, h);
      });
    }, a.prototype._showSeriesItemTooltip = function(i, r, e) {
      var t = this._ecModel, o = S(r), f = o.seriesIndex, n = t.getSeriesByIndex(f), s = o.dataModel || n, l = o.dataIndex, u = o.dataType, h = s.getData(u), d = this._renderMode, c = i.positionDefault, m = I([h.getItemModel(l), s, n && (n.coordinateSystem || {}).model], this._tooltipModel, c ? {
        position: c
      } : null), _ = m.get("trigger");
      if (!(_ != null && _ !== "item")) {
        var p = s.getDataParams(l, u), T = new Y();
        p.marker = T.makeTooltipMarker("item", U(p.color), d);
        var g = q(s.formatTooltip(l, !1, u)), x = m.get("order"), C = m.get("valueFormatter"), M = g.frag, k = M ? G(C ? N({
          valueFormatter: C
        }, M) : M, T, d, x, t.get("useUTC"), m.get("textStyle")) : g.text, b = "item_" + s.name + "_" + l;
        this._showOrMove(m, function() {
          this._showTooltipContent(m, k, p, b, i.offsetX, i.offsetY, i.position, i.target, T);
        }), e({
          type: "showTip",
          dataIndexInside: l,
          dataIndex: h.getRawIndex(l),
          seriesIndex: f,
          from: this.uid
        });
      }
    }, a.prototype._showComponentItemTooltip = function(i, r, e) {
      var t = this._renderMode === "html", o = S(r), f = o.tooltipConfig, n = f.option || {}, s = n.encodeHTMLContent;
      if (R(n)) {
        var l = n;
        n = {
          content: l,
          // Fixed formatter
          formatter: l
        }, s = !0;
      }
      s && t && n.content && (n = F(n), n.content = _t(n.content));
      var u = [n], h = this._ecModel.getComponent(o.componentMainType, o.componentIndex);
      h && u.push(h), u.push({
        formatter: n.content
      });
      var d = i.positionDefault, c = I(u, this._tooltipModel, d ? {
        position: d
      } : null), m = c.get("content"), _ = Math.random() + "", p = new Y();
      this._showOrMove(c, function() {
        var T = F(c.get("formatterParams") || {});
        this._showTooltipContent(c, m, T, _, i.offsetX, i.offsetY, i.position, r, p);
      }), e({
        type: "showTip",
        from: this.uid
      });
    }, a.prototype._showTooltipContent = function(i, r, e, t, o, f, n, s, l) {
      if (this._ticket = "", !(!i.get("showContent") || !i.get("show"))) {
        var u = this._tooltipContent;
        u.setEnterable(i.get("enterable"));
        var h = i.get("formatter");
        n = n || i.get("position");
        var d = r, c = this._getNearestPoint([o, f], e, i.get("trigger"), i.get("borderColor"), i.get("defaultBorderColor", !0)), m = c.color;
        if (h)
          if (R(h)) {
            var _ = i.ecModel.get("useUTC"), p = D(e) ? e[0] : e, T = p && p.axisType && p.axisType.indexOf("time") >= 0;
            d = h, T && (d = vt(p.axisValue, d, _)), d = ot(d, e, !0);
          } else if (V(h)) {
            var g = B(function(x, C) {
              x === this._ticket && (u.setContent(C, l, i, m, n), this._updatePosition(i, n, o, f, u, e, s));
            }, this);
            this._ticket = t, d = h(e, t, g);
          } else
            d = h;
        u.setContent(d, l, i, m, n), u.show(i, m), this._updatePosition(i, n, o, f, u, e, s);
      }
    }, a.prototype._getNearestPoint = function(i, r, e, t, o) {
      if (e === "axis" || D(r))
        return {
          color: t || o
        };
      if (!D(r))
        return {
          color: t || r.color || r.borderColor
        };
    }, a.prototype._updatePosition = function(i, r, e, t, o, f, n) {
      var s = this._api.getWidth(), l = this._api.getHeight();
      r = r || i.get("position");
      var u = o.getSize(), h = i.get("align"), d = i.get("verticalAlign"), c = n && n.getBoundingRect().clone();
      if (n && c.applyTransform(n.transform), V(r) && (r = r([e, t], f, o.el, c, {
        viewSize: [s, l],
        contentSize: u.slice()
      })), D(r))
        e = W(r[0], s), t = W(r[1], l);
      else if (E(r)) {
        var m = r;
        m.width = u[0], m.height = u[1];
        var _ = rt(m, {
          width: s,
          height: l
        });
        e = _.x, t = _.y, h = null, d = null;
      } else if (R(r) && n) {
        var p = Ct(r, c, u, i.get("borderWidth"));
        e = p[0], t = p[1];
      } else {
        var p = xt(e, t, o, s, l, h ? null : 20, d ? null : 20);
        e = p[0], t = p[1];
      }
      if (h && (e -= J(h) ? u[0] / 2 : h === "right" ? u[0] : 0), d && (t -= J(d) ? u[1] / 2 : d === "bottom" ? u[1] : 0), pt(i)) {
        var p = yt(e, t, o, s, l);
        e = p[0], t = p[1];
      }
      o.moveTo(e, t);
    }, a.prototype._updateContentNotChangedOnAxis = function(i, r) {
      var e = this._lastDataByCoordSys, t = this._cbParamsList, o = !!e && e.length === i.length;
      return o && w(e, function(f, n) {
        var s = f.dataByAxis || [], l = i[n] || {}, u = l.dataByAxis || [];
        o = o && s.length === u.length, o && w(s, function(h, d) {
          var c = u[d] || {}, m = h.seriesDataIndices || [], _ = c.seriesDataIndices || [];
          o = o && h.value === c.value && h.axisType === c.axisType && h.axisId === c.axisId && m.length === _.length, o && w(m, function(p, T) {
            var g = _[T];
            o = o && p.seriesIndex === g.seriesIndex && p.dataIndex === g.dataIndex;
          }), t && w(h.seriesDataIndices, function(p) {
            var T = p.seriesIndex, g = r[T], x = t[T];
            g && x && x.data !== g.data && (o = !1);
          });
        });
      }), this._lastDataByCoordSys = i, this._cbParamsList = r, !!o;
    }, a.prototype._hide = function(i) {
      this._lastDataByCoordSys = null, i({
        type: "hideTip",
        from: this.uid
      });
    }, a.prototype.dispose = function(i, r) {
      A.node || !r.getDom() || (Q(this, "_updatePosition"), this._tooltipContent.dispose(), at("itemTooltip", r));
    }, a.type = "tooltip", a;
  })(dt)
);
function I(v, a, i) {
  var r = a.ecModel, e;
  i ? (e = new P(i, r, r), e = new P(a.option, e, r)) : e = a;
  for (var t = v.length - 1; t >= 0; t--) {
    var o = v[t];
    o && (o instanceof P && (o = o.get("tooltip", !0)), R(o) && (o = {
      formatter: o
    }), o && (e = new P(o, e, r)));
  }
  return e;
}
function Z(v, a) {
  return v.dispatchAction || B(a.dispatchAction, a);
}
function xt(v, a, i, r, e, t, o) {
  var f = i.getSize(), n = f[0], s = f[1];
  return t != null && (v + n + t + 2 > r ? v -= n + t : v += t), o != null && (a + s + o > e ? a -= s + o : a += o), [v, a];
}
function yt(v, a, i, r, e) {
  var t = i.getSize(), o = t[0], f = t[1];
  return v = Math.min(v + o, r) - o, a = Math.min(a + f, e) - f, v = Math.max(v, 0), a = Math.max(a, 0), [v, a];
}
function Ct(v, a, i, r) {
  var e = i[0], t = i[1], o = Math.ceil(Math.SQRT2 * r) + 8, f = 0, n = 0, s = a.width, l = a.height;
  switch (v) {
    case "inside":
      f = a.x + s / 2 - e / 2, n = a.y + l / 2 - t / 2;
      break;
    case "top":
      f = a.x + s / 2 - e / 2, n = a.y - t - o;
      break;
    case "bottom":
      f = a.x + s / 2 - e / 2, n = a.y + l + o;
      break;
    case "left":
      f = a.x - e - o, n = a.y + l / 2 - t / 2;
      break;
    case "right":
      f = a.x + s + o, n = a.y + l / 2 - t / 2;
  }
  return [f, n];
}
function J(v) {
  return v === "center" || v === "middle";
}
function Mt(v, a, i) {
  var r = ut(v).queryOptionMap, e = r.keys()[0];
  if (!(!e || e === "series")) {
    var t = ht(a, e, r.get(e), {
      useDefault: !1,
      enableAll: !1,
      enableNone: !1
    }), o = t.models[0];
    if (o) {
      var f = i.getViewOfComponentModel(o), n;
      if (f.group.traverse(function(s) {
        var l = S(s).tooltipConfig;
        if (l && l.name === v.name)
          return n = s, !0;
      }), n)
        return {
          componentMainType: e,
          componentIndex: o.componentIndex,
          el: n
        };
    }
  }
}
export {
  Zt as default
};
