// echarts@6.0.0/lib/component/tooltip/TooltipHTMLContent.js
import { isString as m, isDom as C, isFunction as D, isArray as R, each as _, bind as N, indexOf as z, retrieve2 as L } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { normalizeEvent as k } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/event.js";
import { transformLocalCoord as F, transformLocalCoordClear as V } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/dom.js";
import u from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/env.js";
import { convertToColorString as x, toCamelCase as W, normalizeCssArray as A } from "../../util/format.js";
import { toCSSVendorPrefix as S, getComputedStyle as E, shouldTooltipConfine as I, TRANSFORM_VENDOR as H, TRANSITION_VENDOR as B } from "./helper.js";
import { getPaddingFromTooltipModel as X } from "./tooltipMarkup.js";
var Y = S(B, "transition"), w = S(H, "transform"), j = "position:absolute;display:block;border-style:solid;white-space:nowrap;z-index:9999999;" + (u.transform3dSupported ? "will-change:transform;" : "");
function P(e) {
  return e = e === "left" ? "right" : e === "right" ? "left" : e === "top" ? "bottom" : "top", e;
}
function Q(e, t, a) {
  if (!m(a) || a === "inside")
    return "";
  var o = e.get("backgroundColor"), r = e.get("borderWidth");
  t = x(t);
  var i = P(a), s = Math.max(Math.round(r) * 1.5, 6), n = "", h = w + ":", l;
  z(["left", "right"], i) > -1 ? (n += "top:50%", h += "translateY(-50%) rotate(" + (l = i === "left" ? -225 : -45) + "deg)") : (n += "left:50%", h += "translateX(-50%) rotate(" + (l = i === "top" ? 225 : 45) + "deg)");
  var f = l * Math.PI / 180, d = s + r, v = d * Math.abs(Math.cos(f)) + d * Math.abs(Math.sin(f)), c = Math.round(((v - Math.SQRT2 * r) / 2 + Math.SQRT2 * r - (v - d) / 2) * 100) / 100;
  n += ";" + i + ":-" + c + "px";
  var p = t + " solid " + r + "px;", g = ["position:absolute;width:" + s + "px;height:" + s + "px;z-index:-1;", n + ";" + h + ";", "border-bottom:" + p, "border-right:" + p, "background-color:" + o + ";"];
  return '<div style="' + g.join("") + '"></div>';
}
function Z(e, t, a) {
  var o = "cubic-bezier(0.23,1,0.32,1)", r = "", i = "";
  return a && (r = " " + e / 2 + "s " + o, i = "opacity" + r + ",visibility" + r), t || (r = " " + e + "s " + o, i += (i.length ? "," : "") + (u.transformSupported ? "" + w + r : ",left" + r + ",top" + r)), Y + ":" + i;
}
function y(e, t, a) {
  var o = e.toFixed(0) + "px", r = t.toFixed(0) + "px";
  if (!u.transformSupported)
    return a ? "top:" + r + ";left:" + o + ";" : [["top", r], ["left", o]];
  var i = u.transform3dSupported, s = "translate" + (i ? "3d" : "") + "(" + o + "," + r + (i ? ",0" : "") + ")";
  return a ? "top:0;left:0;" + w + ":" + s + ";" : [["top", 0], ["left", 0], [H, s]];
}
function q(e) {
  var t = [], a = e.get("fontSize"), o = e.getTextColor();
  o && t.push("color:" + o), t.push("font:" + e.getFont());
  var r = L(e.get("lineHeight"), Math.round(a * 3 / 2));
  a && t.push("line-height:" + r + "px");
  var i = e.get("textShadowColor"), s = e.get("textShadowBlur") || 0, n = e.get("textShadowOffsetX") || 0, h = e.get("textShadowOffsetY") || 0;
  return i && s && t.push("text-shadow:" + n + "px " + h + "px " + s + "px " + i), _(["decoration", "align"], function(l) {
    var f = e.get(l);
    f && t.push("text-" + l + ":" + f);
  }), t.join(";");
}
function G(e, t, a, o) {
  var r = [], i = e.get("transitionDuration"), s = e.get("backgroundColor"), n = e.get("shadowBlur"), h = e.get("shadowColor"), l = e.get("shadowOffsetX"), f = e.get("shadowOffsetY"), d = e.getModel("textStyle"), v = X(e, "html"), c = l + "px " + f + "px " + n + "px " + h;
  return r.push("box-shadow:" + c), t && i > 0 && r.push(Z(i, a, o)), s && r.push("background-color:" + s), _(["width", "color", "radius"], function(p) {
    var g = "border-" + p, O = W(g), T = e.get(O);
    T != null && r.push(g + ":" + T + (p === "color" ? "" : "px"));
  }), r.push(q(d)), v != null && r.push("padding:" + A(v).join("px ") + "px"), r.join(";") + ";";
}
function b(e, t, a, o, r) {
  var i = t && t.painter;
  if (a) {
    var s = i && i.getViewportRoot();
    s && F(e, s, a, o, r);
  } else {
    e[0] = o, e[1] = r;
    var n = i && i.getViewportRootOffset();
    n && (e[0] += n.offsetLeft, e[1] += n.offsetTop);
  }
  e[2] = e[0] / t.getWidth(), e[3] = e[1] / t.getHeight();
}
var rt = (
  /** @class */
  (function() {
    function e(t, a) {
      if (this._show = !1, this._styleCoord = [0, 0, 0, 0], this._enterable = !0, this._alwaysShowContent = !1, this._firstShow = !0, this._longHide = !0, u.wxa)
        return null;
      var o = document.createElement("div");
      o.domBelongToZr = !0, this.el = o;
      var r = this._zr = t.getZr(), i = a.appendTo, s = i && (m(i) ? document.querySelector(i) : C(i) ? i : D(i) && i(t.getDom()));
      b(this._styleCoord, r, s, t.getWidth() / 2, t.getHeight() / 2), (s || t.getDom()).appendChild(o), this._api = t, this._container = s;
      var n = this;
      o.onmouseenter = function() {
        n._enterable && (clearTimeout(n._hideTimeout), n._show = !0), n._inContent = !0;
      }, o.onmousemove = function(h) {
        if (h = h || window.event, !n._enterable) {
          var l = r.handler, f = r.painter.getViewportRoot();
          k(f, h, !0), l.dispatch("mousemove", h);
        }
      }, o.onmouseleave = function() {
        n._inContent = !1, n._enterable && n._show && n.hideLater(n._hideDelay);
      };
    }
    return e.prototype.update = function(t) {
      if (!this._container) {
        var a = this._api.getDom(), o = E(a, "position"), r = a.style;
        r.position !== "absolute" && o !== "absolute" && (r.position = "relative");
      }
      var i = t.get("alwaysShowContent");
      i && this._moveIfResized(), this._alwaysShowContent = i, this._enableDisplayTransition = t.get("displayTransition") && t.get("transitionDuration") > 0, this.el.className = t.get("className") || "";
    }, e.prototype.show = function(t, a) {
      clearTimeout(this._hideTimeout), clearTimeout(this._longHideTimeout);
      var o = this.el, r = o.style, i = this._styleCoord;
      o.innerHTML ? r.cssText = j + G(t, !this._firstShow, this._longHide, this._enableDisplayTransition) + y(i[0], i[1], !0) + ("border-color:" + x(a) + ";") + (t.get("extraCssText") || "") + (";pointer-events:" + (this._enterable ? "auto" : "none")) : r.display = "none", this._show = !0, this._firstShow = !1, this._longHide = !1;
    }, e.prototype.setContent = function(t, a, o, r, i) {
      var s = this.el;
      if (t == null) {
        s.innerHTML = "";
        return;
      }
      var n = "";
      if (m(i) && o.get("trigger") === "item" && !I(o) && (n = Q(o, r, i)), m(t))
        s.innerHTML = t + n;
      else if (t) {
        s.innerHTML = "", R(t) || (t = [t]);
        for (var h = 0; h < t.length; h++)
          C(t[h]) && t[h].parentNode !== s && s.appendChild(t[h]);
        if (n && s.childNodes.length) {
          var l = document.createElement("div");
          l.innerHTML = n, s.appendChild(l);
        }
      }
    }, e.prototype.setEnterable = function(t) {
      this._enterable = t;
    }, e.prototype.getSize = function() {
      var t = this.el;
      return t ? [t.offsetWidth, t.offsetHeight] : [0, 0];
    }, e.prototype.moveTo = function(t, a) {
      if (this.el) {
        var o = this._styleCoord;
        if (b(o, this._zr, this._container, t, a), o[0] != null && o[1] != null) {
          var r = this.el.style, i = y(o[0], o[1]);
          _(i, function(s) {
            r[s[0]] = s[1];
          });
        }
      }
    }, e.prototype._moveIfResized = function() {
      var t = this._styleCoord[2], a = this._styleCoord[3];
      this.moveTo(t * this._zr.getWidth(), a * this._zr.getHeight());
    }, e.prototype.hide = function() {
      var t = this, a = this.el.style;
      this._enableDisplayTransition ? (a.visibility = "hidden", a.opacity = "0") : a.display = "none", u.transform3dSupported && (a.willChange = ""), this._show = !1, this._longHideTimeout = setTimeout(function() {
        return t._longHide = !0;
      }, 500);
    }, e.prototype.hideLater = function(t) {
      this._show && !(this._inContent && this._enterable) && !this._alwaysShowContent && (t ? (this._hideDelay = t, this._show = !1, this._hideTimeout = setTimeout(N(this.hide, this), t)) : this.hide());
    }, e.prototype.isShow = function() {
      return this._show;
    }, e.prototype.dispose = function() {
      clearTimeout(this._hideTimeout), clearTimeout(this._longHideTimeout);
      var t = this._zr;
      V(t && t.painter && t.painter.getViewportRoot(), this._container);
      var a = this.el;
      if (a) {
        a.onmouseenter = a.onmousemove = a.onmouseleave = null;
        var o = a.parentNode;
        o && o.removeChild(a);
      }
      this.el = this._container = null;
    }, e;
  })()
);
export {
  rt as default
};
