// date-fns@3.6.0/parse/_lib/parsers/ISODayParser.js
import { setISODay as n } from "../../../setISODay.js";
import { Parser as s } from "../Parser.js";
import { mapValue as e, parseNDigits as w } from "../utils.js";
class c extends s {
  priority = 90;
  parse(r, t, i) {
    const o = (a) => a === 0 ? 7 : a;
    switch (t) {
      // 2
      case "i":
      case "ii":
        return w(t.length, r);
      // 2nd
      case "io":
        return i.ordinalNumber(r, { unit: "day" });
      // Tue
      case "iii":
        return e(
          i.day(r, {
            width: "abbreviated",
            context: "formatting"
          }) || i.day(r, {
            width: "short",
            context: "formatting"
          }) || i.day(r, {
            width: "narrow",
            context: "formatting"
          }),
          o
        );
      // T
      case "iiiii":
        return e(
          i.day(r, {
            width: "narrow",
            context: "formatting"
          }),
          o
        );
      // Tu
      case "iiiiii":
        return e(
          i.day(r, {
            width: "short",
            context: "formatting"
          }) || i.day(r, {
            width: "narrow",
            context: "formatting"
          }),
          o
        );
      default:
        return e(
          i.day(r, {
            width: "wide",
            context: "formatting"
          }) || i.day(r, {
            width: "abbreviated",
            context: "formatting"
          }) || i.day(r, {
            width: "short",
            context: "formatting"
          }) || i.day(r, {
            width: "narrow",
            context: "formatting"
          }),
          o
        );
    }
  }
  validate(r, t) {
    return t >= 1 && t <= 7;
  }
  set(r, t, i) {
    return r = n(r, i), r.setHours(0, 0, 0, 0), r;
  }
  incompatibleTokens = [
    "y",
    "Y",
    "u",
    "q",
    "Q",
    "M",
    "L",
    "w",
    "d",
    "D",
    "E",
    "e",
    "c",
    "t",
    "T"
  ];
}
export {
  c as ISODayParser
};
