// date-fns@3.6.0/parse/_lib/parsers/LocalWeekYearParser.js
import { getWeekYear as u } from "../../../getWeekYear.js";
import { startOfWeek as o } from "../../../startOfWeek.js";
import { Parser as c } from "../Parser.js";
import { mapValue as s, parseNDigits as n, normalizeTwoDigitYear as l } from "../utils.js";
class w extends c {
  priority = 130;
  parse(r, e, i) {
    const a = (t) => ({
      year: t,
      isTwoDigitYear: e === "YY"
    });
    switch (e) {
      case "Y":
        return s(n(4, r), a);
      case "Yo":
        return s(
          i.ordinalNumber(r, {
            unit: "year"
          }),
          a
        );
      default:
        return s(n(e.length, r), a);
    }
  }
  validate(r, e) {
    return e.isTwoDigitYear || e.year > 0;
  }
  set(r, e, i, a) {
    const t = u(r, a);
    if (i.isTwoDigitYear) {
      const m = l(
        i.year,
        t
      );
      return r.setFullYear(
        m,
        0,
        a.firstWeekContainsDate
      ), r.setHours(0, 0, 0, 0), o(r, a);
    }
    const Y = !("era" in e) || e.era === 1 ? i.year : 1 - i.year;
    return r.setFullYear(Y, 0, a.firstWeekContainsDate), r.setHours(0, 0, 0, 0), o(r, a);
  }
  incompatibleTokens = [
    "y",
    "R",
    "u",
    "Q",
    "q",
    "M",
    "L",
    "I",
    "d",
    "D",
    "i",
    "t",
    "T"
  ];
}
export {
  w as LocalWeekYearParser
};
