// date-fns@3.6.0/parse/_lib/parsers/LocalDayParser.js
import { setDay as w } from "../../../setDay.js";
import { Parser as d } from "../Parser.js";
import { mapValue as s, parseNDigits as f } from "../utils.js";
class x extends d {
  priority = 90;
  parse(e, r, t, o) {
    const a = (n) => {
      const i = Math.floor((n - 1) / 7) * 7;
      return (n + o.weekStartsOn + 6) % 7 + i;
    };
    switch (r) {
      // 3
      case "e":
      case "ee":
        return s(f(r.length, e), a);
      // 3rd
      case "eo":
        return s(
          t.ordinalNumber(e, {
            unit: "day"
          }),
          a
        );
      // Tue
      case "eee":
        return t.day(e, {
          width: "abbreviated",
          context: "formatting"
        }) || t.day(e, { width: "short", context: "formatting" }) || t.day(e, { width: "narrow", context: "formatting" });
      // T
      case "eeeee":
        return t.day(e, {
          width: "narrow",
          context: "formatting"
        });
      // Tu
      case "eeeeee":
        return t.day(e, { width: "short", context: "formatting" }) || t.day(e, { width: "narrow", context: "formatting" });
      default:
        return t.day(e, { width: "wide", context: "formatting" }) || t.day(e, {
          width: "abbreviated",
          context: "formatting"
        }) || t.day(e, { width: "short", context: "formatting" }) || t.day(e, { width: "narrow", context: "formatting" });
    }
  }
  validate(e, r) {
    return r >= 0 && r <= 6;
  }
  set(e, r, t, o) {
    return e = w(e, t, o), e.setHours(0, 0, 0, 0), e;
  }
  incompatibleTokens = [
    "y",
    "R",
    "u",
    "q",
    "Q",
    "M",
    "L",
    "I",
    "d",
    "D",
    "E",
    "i",
    "c",
    "t",
    "T"
  ];
}
export {
  x as LocalDayParser
};
