// date-fns@3.6.0/parse/_lib/parsers/MonthParser.js
import { numericPatterns as i } from "../constants.js";
import { Parser as a } from "../Parser.js";
import { mapValue as n, parseNDigits as m, parseNumericPattern as M } from "../utils.js";
class h extends a {
  incompatibleTokens = [
    "Y",
    "R",
    "q",
    "Q",
    "L",
    "w",
    "I",
    "D",
    "i",
    "e",
    "c",
    "t",
    "T"
  ];
  priority = 110;
  parse(t, o, r) {
    const e = (s) => s - 1;
    switch (o) {
      // 1, 2, ..., 12
      case "M":
        return n(
          M(i.month, t),
          e
        );
      // 01, 02, ..., 12
      case "MM":
        return n(m(2, t), e);
      // 1st, 2nd, ..., 12th
      case "Mo":
        return n(
          r.ordinalNumber(t, {
            unit: "month"
          }),
          e
        );
      // Jan, Feb, ..., Dec
      case "MMM":
        return r.month(t, {
          width: "abbreviated",
          context: "formatting"
        }) || r.month(t, { width: "narrow", context: "formatting" });
      // J, F, ..., D
      case "MMMMM":
        return r.month(t, {
          width: "narrow",
          context: "formatting"
        });
      default:
        return r.month(t, { width: "wide", context: "formatting" }) || r.month(t, {
          width: "abbreviated",
          context: "formatting"
        }) || r.month(t, { width: "narrow", context: "formatting" });
    }
  }
  validate(t, o) {
    return o >= 0 && o <= 11;
  }
  set(t, o, r) {
    return t.setMonth(r, 1), t.setHours(0, 0, 0, 0), t;
  }
}
export {
  h as MonthParser
};
