// date-fns@3.6.0/parse/_lib/parsers/QuarterParser.js
import { Parser as o } from "../Parser.js";
import { parseNDigits as a } from "../utils.js";
class u extends o {
  priority = 120;
  parse(r, e, t) {
    switch (e) {
      // 1, 2, 3, 4
      case "Q":
      case "QQ":
        return a(e.length, r);
      // 1st, 2nd, 3rd, 4th
      case "Qo":
        return t.ordinalNumber(r, { unit: "quarter" });
      // Q1, Q2, Q3, Q4
      case "QQQ":
        return t.quarter(r, {
          width: "abbreviated",
          context: "formatting"
        }) || t.quarter(r, {
          width: "narrow",
          context: "formatting"
        });
      // 1, 2, 3, 4 (narrow quarter; could be not numerical)
      case "QQQQQ":
        return t.quarter(r, {
          width: "narrow",
          context: "formatting"
        });
      default:
        return t.quarter(r, {
          width: "wide",
          context: "formatting"
        }) || t.quarter(r, {
          width: "abbreviated",
          context: "formatting"
        }) || t.quarter(r, {
          width: "narrow",
          context: "formatting"
        });
    }
  }
  validate(r, e) {
    return e >= 1 && e <= 4;
  }
  set(r, e, t) {
    return r.setMonth((t - 1) * 3, 1), r.setHours(0, 0, 0, 0), r;
  }
  incompatibleTokens = [
    "Y",
    "R",
    "q",
    "M",
    "L",
    "w",
    "I",
    "d",
    "D",
    "i",
    "e",
    "c",
    "t",
    "T"
  ];
}
export {
  u as QuarterParser
};
