// date-fns@3.6.0/parse/_lib/parsers/StandAloneMonthParser.js
import { numericPatterns as a } from "../constants.js";
import { Parser as i } from "../Parser.js";
import { mapValue as r, parseNDigits as l, parseNumericPattern as L } from "../utils.js";
class d extends i {
  priority = 110;
  parse(e, n, t) {
    const o = (s) => s - 1;
    switch (n) {
      // 1, 2, ..., 12
      case "L":
        return r(
          L(a.month, e),
          o
        );
      // 01, 02, ..., 12
      case "LL":
        return r(l(2, e), o);
      // 1st, 2nd, ..., 12th
      case "Lo":
        return r(
          t.ordinalNumber(e, {
            unit: "month"
          }),
          o
        );
      // Jan, Feb, ..., Dec
      case "LLL":
        return t.month(e, {
          width: "abbreviated",
          context: "standalone"
        }) || t.month(e, { width: "narrow", context: "standalone" });
      // J, F, ..., D
      case "LLLLL":
        return t.month(e, {
          width: "narrow",
          context: "standalone"
        });
      default:
        return t.month(e, { width: "wide", context: "standalone" }) || t.month(e, {
          width: "abbreviated",
          context: "standalone"
        }) || t.month(e, { width: "narrow", context: "standalone" });
    }
  }
  validate(e, n) {
    return n >= 0 && n <= 11;
  }
  set(e, n, t) {
    return e.setMonth(t, 1), e.setHours(0, 0, 0, 0), e;
  }
  incompatibleTokens = [
    "Y",
    "R",
    "q",
    "Q",
    "M",
    "w",
    "I",
    "D",
    "i",
    "e",
    "c",
    "t",
    "T"
  ];
}
export {
  d as StandAloneMonthParser
};
