// date-fns@3.6.0/parse/_lib/parsers/StandAloneQuarterParser.js
import { Parser as a } from "../Parser.js";
import { parseNDigits as n } from "../utils.js";
class u extends a {
  priority = 120;
  parse(r, t, e) {
    switch (t) {
      // 1, 2, 3, 4
      case "q":
      case "qq":
        return n(t.length, r);
      // 1st, 2nd, 3rd, 4th
      case "qo":
        return e.ordinalNumber(r, { unit: "quarter" });
      // Q1, Q2, Q3, Q4
      case "qqq":
        return e.quarter(r, {
          width: "abbreviated",
          context: "standalone"
        }) || e.quarter(r, {
          width: "narrow",
          context: "standalone"
        });
      // 1, 2, 3, 4 (narrow quarter; could be not numerical)
      case "qqqqq":
        return e.quarter(r, {
          width: "narrow",
          context: "standalone"
        });
      default:
        return e.quarter(r, {
          width: "wide",
          context: "standalone"
        }) || e.quarter(r, {
          width: "abbreviated",
          context: "standalone"
        }) || e.quarter(r, {
          width: "narrow",
          context: "standalone"
        });
    }
  }
  validate(r, t) {
    return t >= 1 && t <= 4;
  }
  set(r, t, e) {
    return r.setMonth((e - 1) * 3, 1), r.setHours(0, 0, 0, 0), r;
  }
  incompatibleTokens = [
    "Y",
    "R",
    "Q",
    "M",
    "L",
    "w",
    "I",
    "d",
    "D",
    "i",
    "e",
    "c",
    "t",
    "T"
  ];
}
export {
  u as StandAloneQuarterParser
};
