// date-fns@3.6.0/parse/_lib/parsers/StandAloneLocalDayParser.js
import { setDay as d } from "../../../setDay.js";
import { Parser as l } from "../Parser.js";
import { mapValue as s, parseNDigits as w } from "../utils.js";
class p extends l {
  priority = 90;
  parse(e, a, o, n) {
    const t = (r) => {
      const c = Math.floor((r - 1) / 7) * 7;
      return (r + n.weekStartsOn + 6) % 7 + c;
    };
    switch (a) {
      // 3
      case "c":
      case "cc":
        return s(w(a.length, e), t);
      // 3rd
      case "co":
        return s(
          o.ordinalNumber(e, {
            unit: "day"
          }),
          t
        );
      // Tue
      case "ccc":
        return o.day(e, {
          width: "abbreviated",
          context: "standalone"
        }) || o.day(e, { width: "short", context: "standalone" }) || o.day(e, { width: "narrow", context: "standalone" });
      // T
      case "ccccc":
        return o.day(e, {
          width: "narrow",
          context: "standalone"
        });
      // Tu
      case "cccccc":
        return o.day(e, { width: "short", context: "standalone" }) || o.day(e, { width: "narrow", context: "standalone" });
      default:
        return o.day(e, { width: "wide", context: "standalone" }) || o.day(e, {
          width: "abbreviated",
          context: "standalone"
        }) || o.day(e, { width: "short", context: "standalone" }) || o.day(e, { width: "narrow", context: "standalone" });
    }
  }
  validate(e, a) {
    return a >= 0 && a <= 6;
  }
  set(e, a, o, n) {
    return e = d(e, o, n), e.setHours(0, 0, 0, 0), e;
  }
  incompatibleTokens = [
    "y",
    "R",
    "u",
    "q",
    "Q",
    "M",
    "L",
    "I",
    "d",
    "D",
    "E",
    "i",
    "e",
    "t",
    "T"
  ];
}
export {
  p as StandAloneLocalDayParser
};
