// date-fns@3.6.0/parse/_lib/parsers/YearParser.js
import { Parser as u } from "../Parser.js";
import { mapValue as t, parseNDigits as o, normalizeTwoDigitYear as l } from "../utils.js";
class m extends u {
  priority = 130;
  incompatibleTokens = ["Y", "R", "u", "w", "I", "i", "e", "c", "t", "T"];
  parse(r, e, i) {
    const a = (s) => ({
      year: s,
      isTwoDigitYear: e === "yy"
    });
    switch (e) {
      case "y":
        return t(o(4, r), a);
      case "yo":
        return t(
          i.ordinalNumber(r, {
            unit: "year"
          }),
          a
        );
      default:
        return t(o(e.length, r), a);
    }
  }
  validate(r, e) {
    return e.isTwoDigitYear || e.year > 0;
  }
  set(r, e, i) {
    const a = r.getFullYear();
    if (i.isTwoDigitYear) {
      const n = l(
        i.year,
        a
      );
      return r.setFullYear(n, 0, 1), r.setHours(0, 0, 0, 0), r;
    }
    const s = !("era" in e) || e.era === 1 ? i.year : 1 - i.year;
    return r.setFullYear(s, 0, 1), r.setHours(0, 0, 0, 0), r;
  }
}
export {
  m as YearParser
};
