// zod-to-json-schema@3.25.1/dist/esm/parsers/object.js
import { parseDef as c } from "../parseDef.js";
function P(r, e) {
  const s = e.target === "openAi", n = {
    type: "object",
    properties: {}
  }, o = [], p = r.shape();
  for (const i in p) {
    let t = p[i];
    if (t === void 0 || t._def === void 0)
      continue;
    let a = f(t);
    a && s && (t._def.typeName === "ZodOptional" && (t = t._def.innerType), t.isNullable() || (t = t.nullable()), a = !1);
    const l = c(t._def, {
      ...e,
      currentPath: [...e.currentPath, "properties", i],
      propertyPath: [...e.currentPath, "properties", i]
    });
    l !== void 0 && (n.properties[i] = l, a || o.push(i));
  }
  o.length && (n.required = o);
  const d = u(r, e);
  return d !== void 0 && (n.additionalProperties = d), n;
}
function u(r, e) {
  if (r.catchall._def.typeName !== "ZodNever")
    return c(r.catchall._def, {
      ...e,
      currentPath: [...e.currentPath, "additionalProperties"]
    });
  switch (r.unknownKeys) {
    case "passthrough":
      return e.allowedAdditionalProperties;
    case "strict":
      return e.rejectedAdditionalProperties;
    case "strip":
      return e.removeAdditionalStrategy === "strict" ? e.allowedAdditionalProperties : e.rejectedAdditionalProperties;
  }
}
function f(r) {
  try {
    return r.isOptional();
  } catch {
    return !0;
  }
}
export {
  P as parseObjectDef
};
