// zod-to-json-schema@3.25.1/dist/esm/parsers/record.js
import { parseDef as a } from "../parseDef.js";
import { parseStringDef as n } from "./string.js";
import { parseBrandedDef as i } from "./branded.js";
import { parseAnyDef as d } from "./any.js";
import { ZodFirstPartyTypeKind as y } from "../../../../../../zod@3.25.76/node_modules/zod/v3/types.js";
function T(e, r) {
  if (r.target === "openAi" && console.warn("Warning: OpenAI may not support records in schemas! Try an array of key-value pairs instead."), r.target === "openApi3" && e.keyType?._def.typeName === y.ZodEnum)
    return {
      type: "object",
      required: e.keyType._def.values,
      properties: e.keyType._def.values.reduce((o, t) => ({
        ...o,
        [t]: a(e.valueType._def, {
          ...r,
          currentPath: [...r.currentPath, "properties", t]
        }) ?? d(r)
      }), {}),
      additionalProperties: r.rejectedAdditionalProperties
    };
  const p = {
    type: "object",
    additionalProperties: a(e.valueType._def, {
      ...r,
      currentPath: [...r.currentPath, "additionalProperties"]
    }) ?? r.allowedAdditionalProperties
  };
  if (r.target === "openApi3")
    return p;
  if (e.keyType?._def.typeName === y.ZodString && e.keyType._def.checks?.length) {
    const { type: o, ...t } = n(e.keyType._def, r);
    return {
      ...p,
      propertyNames: t
    };
  } else {
    if (e.keyType?._def.typeName === y.ZodEnum)
      return {
        ...p,
        propertyNames: {
          enum: e.keyType._def.values
        }
      };
    if (e.keyType?._def.typeName === y.ZodBranded && e.keyType._def.type._def.typeName === y.ZodString && e.keyType._def.type._def.checks?.length) {
      const { type: o, ...t } = i(e.keyType._def, r);
      return {
        ...p,
        propertyNames: t
      };
    }
  }
  return p;
}
export {
  T as parseRecordDef
};
