// echarts@6.0.0/lib/component/helper/BrushTargetManager.js
import { each as d, indexOf as f, createHashMap as _, curry as l, assert as x, map as v } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { getTransform as b } from "../../util/graphic.js";
import { makeLinearBrushOtherExtent as P, makeRectIsTargetByCursor as D, makeRectPanelClipPath as R } from "./brushHelper.js";
import { parseFinder as B } from "../../util/model.js";
var E = ["grid", "xAxis", "yAxis", "geo", "graph", "polar", "radiusAxis", "angleAxis", "bmap"], O = (
  /** @class */
  (function() {
    function i(e, r, t) {
      var n = this;
      this._targetInfoList = [];
      var s = M(r, e);
      d(L, function(o, a) {
        (!t || !t.include || f(t.include, a) >= 0) && o(s, n._targetInfoList);
      });
    }
    return i.prototype.setOutputRanges = function(e, r) {
      return this.matchOutputRanges(e, r, function(t, n, s) {
        if ((t.coordRanges || (t.coordRanges = [])).push(n), !t.coordRange) {
          t.coordRange = n;
          var o = y[t.brushType](0, s, n);
          t.__rangeOffset = {
            offset: A[t.brushType](o.values, t.range, [1, 1]),
            xyMinMax: o.xyMinMax
          };
        }
      }), e;
    }, i.prototype.matchOutputRanges = function(e, r, t) {
      d(e, function(n) {
        var s = this.findTargetInfo(n, r);
        s && s !== !0 && d(s.coordSyses, function(o) {
          var a = y[n.brushType](1, o, n.range, !0);
          t(n, a.values, o, r);
        });
      }, this);
    }, i.prototype.setInputRanges = function(e, r) {
      d(e, function(t) {
        var n = this.findTargetInfo(t, r);
        if (process.env.NODE_ENV !== "production" && (x(!n || n === !0 || t.coordRange, "coordRange must be specified when coord index specified."), x(!n || n !== !0 || t.range, "range must be specified in global brush.")), t.range = t.range || [], n && n !== !0) {
          t.panelId = n.panelId;
          var s = y[t.brushType](0, n.coordSys, t.coordRange), o = t.__rangeOffset;
          t.range = o ? A[t.brushType](s.values, o.offset, k(s.xyMinMax, o.xyMinMax)) : s.values;
        }
      }, this);
    }, i.prototype.makePanelOpts = function(e, r) {
      return v(this._targetInfoList, function(t) {
        var n = t.getPanelRect();
        return {
          panelId: t.panelId,
          defaultBrushType: r ? r(t) : null,
          clipPath: R(n),
          isTargetByCursor: D(n, e, t.coordSysModel),
          getLinearBrushOtherExtent: P(n)
        };
      });
    }, i.prototype.controlSeries = function(e, r, t) {
      var n = this.findTargetInfo(e, t);
      return n === !0 || n && f(n.coordSyses, r.coordinateSystem) >= 0;
    }, i.prototype.findTargetInfo = function(e, r) {
      for (var t = this._targetInfoList, n = M(r, e), s = 0; s < t.length; s++) {
        var o = t[s], a = e.panelId;
        if (a) {
          if (o.panelId === a)
            return o;
        } else
          for (var u = 0; u < m.length; u++)
            if (m[u](n, o))
              return o;
      }
      return !0;
    }, i;
  })()
);
function h(i) {
  return i[0] > i[1] && i.reverse(), i;
}
function M(i, e) {
  return B(i, e, {
    includeMainTypes: E
  });
}
var L = {
  grid: function(i, e) {
    var r = i.xAxisModels, t = i.yAxisModels, n = i.gridModels, s = _(), o = {}, a = {};
    !r && !t && !n || (d(r, function(u) {
      var c = u.axis.grid.model;
      s.set(c.id, c), o[c.id] = !0;
    }), d(t, function(u) {
      var c = u.axis.grid.model;
      s.set(c.id, c), a[c.id] = !0;
    }), d(n, function(u) {
      s.set(u.id, u), o[u.id] = !0, a[u.id] = !0;
    }), s.each(function(u) {
      var c = u.coordinateSystem, p = [];
      d(c.getCartesians(), function(g, F) {
        (f(r, g.getAxis("x").model) >= 0 || f(t, g.getAxis("y").model) >= 0) && p.push(g);
      }), e.push({
        panelId: "grid--" + u.id,
        gridModel: u,
        coordSysModel: u,
        // Use the first one as the representitive coordSys.
        coordSys: p[0],
        coordSyses: p,
        getPanelRect: T.grid,
        xAxisDeclared: o[u.id],
        yAxisDeclared: a[u.id]
      });
    }));
  },
  geo: function(i, e) {
    d(i.geoModels, function(r) {
      var t = r.coordinateSystem;
      e.push({
        panelId: "geo--" + r.id,
        geoModel: r,
        coordSysModel: r,
        coordSys: t,
        coordSyses: [t],
        getPanelRect: T.geo
      });
    });
  }
}, m = [
  // grid
  function(i, e) {
    var r = i.xAxisModel, t = i.yAxisModel, n = i.gridModel;
    return !n && r && (n = r.axis.grid.model), !n && t && (n = t.axis.grid.model), n && n === e.gridModel;
  },
  // geo
  function(i, e) {
    var r = i.geoModel;
    return r && r === e.geoModel;
  }
], T = {
  grid: function() {
    return this.coordSys.master.getRect().clone();
  },
  geo: function() {
    var i = this.coordSys, e = i.getBoundingRect().clone();
    return e.applyTransform(b(i)), e;
  }
}, y = {
  lineX: l(I, 0),
  lineY: l(I, 1),
  rect: function(i, e, r, t) {
    var n = i ? e.pointToData([r[0][0], r[1][0]], t) : e.dataToPoint([r[0][0], r[1][0]], t), s = i ? e.pointToData([r[0][1], r[1][1]], t) : e.dataToPoint([r[0][1], r[1][1]], t), o = [h([n[0], s[0]]), h([n[1], s[1]])];
    return {
      values: o,
      xyMinMax: o
    };
  },
  polygon: function(i, e, r, t) {
    var n = [[1 / 0, -1 / 0], [1 / 0, -1 / 0]], s = v(r, function(o) {
      var a = i ? e.pointToData(o, t) : e.dataToPoint(o, t);
      return n[0][0] = Math.min(n[0][0], a[0]), n[1][0] = Math.min(n[1][0], a[1]), n[0][1] = Math.max(n[0][1], a[0]), n[1][1] = Math.max(n[1][1], a[1]), a;
    });
    return {
      values: s,
      xyMinMax: n
    };
  }
};
function I(i, e, r, t) {
  process.env.NODE_ENV !== "production" && x(r.type === "cartesian2d", "lineX/lineY brush is available only in cartesian2d.");
  var n = r.getAxis(["x", "y"][i]), s = h(v([0, 1], function(a) {
    return e ? n.coordToData(n.toLocalCoord(t[a]), !0) : n.toGlobalCoord(n.dataToCoord(t[a]));
  })), o = [];
  return o[i] = s, o[1 - i] = [NaN, NaN], {
    values: s,
    xyMinMax: o
  };
}
var A = {
  lineX: l(N, 0),
  lineY: l(N, 1),
  rect: function(i, e, r) {
    return [[i[0][0] - r[0] * e[0][0], i[0][1] - r[0] * e[0][1]], [i[1][0] - r[1] * e[1][0], i[1][1] - r[1] * e[1][1]]];
  },
  polygon: function(i, e, r) {
    return v(i, function(t, n) {
      return [t[0] - r[0] * e[n][0], t[1] - r[1] * e[n][1]];
    });
  }
};
function N(i, e, r, t) {
  return [e[0] - t[i] * r[0], e[1] - t[i] * r[1]];
}
function k(i, e) {
  var r = S(i), t = S(e), n = [r[0] / t[0], r[1] / t[1]];
  return isNaN(n[0]) && (n[0] = 1), isNaN(n[1]) && (n[1] = 1), n;
}
function S(i) {
  return i ? [i[0][1] - i[0][0], i[1][1] - i[1][0]] : [NaN, NaN];
}
export {
  O as default
};
