// echarts@6.0.0/lib/component/helper/MapDraw.js
import { createHashMap as T, each as G, extend as A } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import Z from "./RoamController.js";
import { updateViewOnPan as U, updateViewOnZoom as C } from "./roamHelper.js";
import { setTooltipConfig as q } from "../../util/graphic.js";
import { setDefaultStateProxy as O, toggleHoverEmphasis as K, enableComponentHighDownFeatures as Q } from "../../util/states.js";
import E from "../../coord/geo/geoSourceManager.js";
import { getUID as W } from "../../util/component.js";
import { setLabelStyle as $, getLabelStatesModels as B } from "../../label/labelStyle.js";
import { getECData as ee } from "../../util/innerStore.js";
import { createOrUpdatePatternFromDecal as te } from "../../util/decal.js";
import w from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Displayable.js";
import { makeInner as oe } from "../../util/model.js";
import D from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import re from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Rect.js";
import ae from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Polygon.js";
import ne from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Polyline.js";
import ie from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/CompoundPath.js";
import { updateProps as se } from "../../animation/basicTransition.js";
var V = ["rect", "circle", "line", "ellipse", "polygon", "polyline", "path"], le = T(V), ue = T(V.concat(["g"])), pe = T(V.concat(["g"])), X = oe();
function R(u) {
  var e = u.getItemStyle(), t = u.get("areaColor");
  return t != null && (e.fill = t), e;
}
function F(u) {
  var e = u.style;
  e && (e.stroke = e.stroke || e.fill, e.fill = null);
}
var Ne = (
  /** @class */
  (function() {
    function u(e) {
      var t = this.group = new D(), o = this._transformGroup = new D();
      t.add(o), this.uid = W("ec_map_draw"), this._controller = new Z(e.getZr()), this._controllerHost = {
        target: o
      }, o.add(this._regionsGroup = new D()), o.add(this._svgGroup = new D());
    }
    return u.prototype.draw = function(e, t, o, n, a) {
      var p = e.mainType === "geo", r = e.getData && e.getData();
      p && t.eachComponent({
        mainType: "series",
        subType: "map"
      }, function(I) {
        !r && I.getHostGeoModel() === e && (r = I.getData());
      });
      var s = e.coordinateSystem, g = this._regionsGroup, l = this._transformGroup, c = s.getTransformInfo(), h = c.raw, i = c.roam, f = !g.childAt(0) || a, m = e.getShallow("clip", !0), v;
      m ? (v = s.getViewRect().clone(), this.group.setClipPath(new re({
        shape: v.clone()
      }))) : this.group.removeClipPath(), f ? (l.x = i.x, l.y = i.y, l.scaleX = i.scaleX, l.scaleY = i.scaleY, l.dirty()) : se(l, i, e);
      var d = r && r.getVisual("visualMeta") && r.getVisual("visualMeta").length > 0, _ = {
        api: o,
        geo: s,
        mapOrGeoModel: e,
        data: r,
        isVisualEncodedByVisualMap: d,
        isGeo: p,
        transformInfoRaw: h
      };
      s.resourceType === "geoJSON" ? this._buildGeoJSON(_) : s.resourceType === "geoSVG" && this._buildSVG(_), this._updateController(e, v, t, o), this._updateMapSelectHandler(e, g, o, n);
    }, u.prototype._buildGeoJSON = function(e) {
      var t = this._regionsGroupByName = T(), o = T(), n = this._regionsGroup, a = e.transformInfoRaw, p = e.mapOrGeoModel, r = e.data, s = e.geo.projection, g = s && s.stream;
      function l(i, f) {
        return f && (i = f(i)), i && [i[0] * a.scaleX + a.x, i[1] * a.scaleY + a.y];
      }
      function c(i) {
        for (var f = [], m = !g && s && s.project, v = 0; v < i.length; ++v) {
          var d = l(i[v], m);
          d && f.push(d);
        }
        return f;
      }
      function h(i) {
        return {
          shape: {
            points: c(i)
          }
        };
      }
      n.removeAll(), G(e.geo.regions, function(i) {
        var f = i.name, m = t.get(f), v = o.get(f) || {}, d = v.dataIdx, _ = v.regionModel;
        if (!m) {
          m = t.set(f, new D()), n.add(m), d = r ? r.indexOfName(f) : null, _ = e.isGeo ? p.getRegionModel(f) : r ? r.getItemModel(d) : null;
          var I = _.get("silent", !0);
          I != null && (m.silent = I), o.set(f, {
            dataIdx: d,
            regionModel: _
          });
        }
        var N = [], P = [];
        G(i.geometries, function(S) {
          if (S.type === "polygon") {
            var M = [S.exterior].concat(S.interiors || []);
            g && (M = x(M, g)), G(M, function(b) {
              N.push(new ae(h(b)));
            });
          } else {
            var y = S.points;
            g && (y = x(y, g, !0)), G(y, function(b) {
              P.push(new ne(h(b)));
            });
          }
        });
        var J = l(i.getCenter(), s && s.project);
        function L(S, M) {
          if (S.length) {
            var y = new ie({
              culling: !0,
              segmentIgnoreThreshold: 1,
              shape: {
                paths: S
              }
            });
            m.add(y), H(e, y, d, _), z(e, y, f, _, p, d, J), M && (F(y), G(y.states, F));
          }
        }
        L(N), L(P, !0);
      }), t.each(function(i, f) {
        var m = o.get(f), v = m.dataIdx, d = m.regionModel;
        k(e, i, f, d, p, v), Y(e, i, f, d, p), j(e, i, f, d, p);
      }, this);
    }, u.prototype._buildSVG = function(e) {
      var t = e.geo.map, o = e.transformInfoRaw;
      this._svgGroup.x = o.x, this._svgGroup.y = o.y, this._svgGroup.scaleX = o.scaleX, this._svgGroup.scaleY = o.scaleY, this._svgResourceChanged(t) && (this._freeSVG(), this._useSVG(t));
      var n = this._svgDispatcherMap = T(), a = !1;
      G(this._svgGraphicRecord.named, function(p) {
        var r = p.name, s = e.mapOrGeoModel, g = e.data, l = p.svgNodeTagLower, c = p.el, h = g ? g.indexOfName(r) : null, i = s.getRegionModel(r);
        le.get(l) != null && c instanceof w && H(e, c, h, i), c instanceof w && (c.culling = !0);
        var f = i.get("silent", !0);
        if (f != null && (c.silent = f), c.z2EmphasisLift = 0, !p.namedFrom && (pe.get(l) != null && z(e, c, r, i, s, h, null), k(e, c, r, i, s, h), Y(e, c, r, i, s), ue.get(l) != null)) {
          var m = j(e, c, r, i, s);
          m === "self" && (a = !0);
          var v = n.get(r) || n.set(r, []);
          v.push(c);
        }
      }, this), this._enableBlurEntireSVG(a, e);
    }, u.prototype._enableBlurEntireSVG = function(e, t) {
      if (e && t.isGeo) {
        var o = t.mapOrGeoModel.getModel(["blur", "itemStyle"]).getItemStyle(), n = o.opacity;
        this._svgGraphicRecord.root.traverse(function(a) {
          if (!a.isGroup) {
            O(a);
            var p = a.ensureState("blur").style || {};
            p.opacity == null && n != null && (p.opacity = n), a.ensureState("emphasis");
          }
        });
      }
    }, u.prototype.remove = function() {
      this._regionsGroup.removeAll(), this._regionsGroupByName = null, this._svgGroup.removeAll(), this._freeSVG(), this._controller.dispose(), this._controllerHost = null;
    }, u.prototype.findHighDownDispatchers = function(e, t) {
      if (e == null)
        return [];
      var o = t.coordinateSystem;
      if (o.resourceType === "geoJSON") {
        var n = this._regionsGroupByName;
        if (n) {
          var a = n.get(e);
          return a ? [a] : [];
        }
      } else if (o.resourceType === "geoSVG")
        return this._svgDispatcherMap && this._svgDispatcherMap.get(e) || [];
    }, u.prototype._svgResourceChanged = function(e) {
      return this._svgMapName !== e;
    }, u.prototype._useSVG = function(e) {
      var t = E.getGeoResource(e);
      if (t && t.type === "geoSVG") {
        var o = t.useGraphic(this.uid);
        this._svgGroup.add(o.root), this._svgGraphicRecord = o, this._svgMapName = e;
      }
    }, u.prototype._freeSVG = function() {
      var e = this._svgMapName;
      if (e != null) {
        var t = E.getGeoResource(e);
        t && t.type === "geoSVG" && t.freeGraphic(this.uid), this._svgGraphicRecord = null, this._svgDispatcherMap = null, this._svgGroup.removeAll(), this._svgMapName = null;
      }
    }, u.prototype._updateController = function(e, t, o, n) {
      var a = e.coordinateSystem, p = this._controller, r = this._controllerHost;
      r.zoomLimit = e.get("scaleLimit"), r.zoom = a.getZoom(), p.enable(e.get("roam") || !1, {
        api: n,
        zInfo: {
          component: e
        },
        triggerInfo: {
          roamTrigger: e.get("roamTrigger"),
          isInSelf: function(l, c, h) {
            return a.containPoint([c, h]);
          },
          isInClip: function(l, c, h) {
            return !t || t.contain(c, h);
          }
        }
      });
      var s = e.mainType;
      function g() {
        var l = {
          type: "geoRoam",
          componentType: s
        };
        return l[s + "Id"] = e.id, l;
      }
      p.off("pan").on("pan", function(l) {
        this._mouseDownFlag = !1, U(r, l.dx, l.dy), n.dispatchAction(A(g(), {
          dx: l.dx,
          dy: l.dy,
          animation: {
            duration: 0
          }
        }));
      }, this), p.off("zoom").on("zoom", function(l) {
        this._mouseDownFlag = !1, C(r, l.scale, l.originX, l.originY), n.dispatchAction(A(g(), {
          totalZoom: r.zoom,
          zoom: l.scale,
          originX: l.originX,
          originY: l.originY,
          animation: {
            duration: 0
          }
        }));
      }, this);
    }, u.prototype.resetForLabelLayout = function() {
      this.group.traverse(function(e) {
        var t = e.getTextContent();
        t && (t.ignore = X(t).ignore);
      });
    }, u.prototype._updateMapSelectHandler = function(e, t, o, n) {
      var a = this;
      t.off("mousedown"), t.off("click"), e.get("selectedMode") && (t.on("mousedown", function() {
        a._mouseDownFlag = !0;
      }), t.on("click", function(p) {
        a._mouseDownFlag && (a._mouseDownFlag = !1);
      }));
    }, u;
  })()
);
function H(u, e, t, o) {
  var n = o.getModel("itemStyle"), a = o.getModel(["emphasis", "itemStyle"]), p = o.getModel(["blur", "itemStyle"]), r = o.getModel(["select", "itemStyle"]), s = R(n), g = R(a), l = R(r), c = R(p), h = u.data;
  if (h) {
    var i = h.getItemVisual(t, "style"), f = h.getItemVisual(t, "decal");
    u.isVisualEncodedByVisualMap && i.fill && (s.fill = i.fill), f && (s.decal = te(f, u.api));
  }
  e.setStyle(s), e.style.strokeNoScale = !0, e.ensureState("emphasis").style = g, e.ensureState("select").style = l, e.ensureState("blur").style = c, O(e);
}
function z(u, e, t, o, n, a, p) {
  var r = u.data, s = u.isGeo, g = r && isNaN(r.get(r.mapDimension("value"), a)), l = r && r.getItemLayout(a);
  if (s || g || l && l.showLabel) {
    var c = s ? t : a, h = void 0;
    (!r || a >= 0) && (h = n);
    var i = p ? {
      normal: {
        align: "center",
        verticalAlign: "middle"
      }
    } : null;
    $(e, B(o), {
      labelFetcher: h,
      labelDataIndex: c,
      defaultText: t
    }, i);
    var f = e.getTextContent();
    if (f && (X(f).ignore = f.ignore, e.textConfig && p)) {
      var m = e.getBoundingRect().clone();
      e.textConfig.layoutRect = m, e.textConfig.position = [(p[0] - m.x) / m.width * 100 + "%", (p[1] - m.y) / m.height * 100 + "%"];
    }
    e.disableLabelAnimation = !0;
  } else
    e.removeTextContent(), e.removeTextConfig(), e.disableLabelAnimation = null;
}
function k(u, e, t, o, n, a) {
  u.data ? u.data.setItemGraphicEl(a, e) : ee(e).eventData = {
    componentType: "geo",
    componentIndex: n.componentIndex,
    geoIndex: n.componentIndex,
    name: t,
    region: o && o.option || {}
  };
}
function Y(u, e, t, o, n) {
  u.data || q({
    el: e,
    componentModel: n,
    itemName: t,
    // @ts-ignore FIXME:TS fix the "compatible with each other"?
    itemTooltipOption: o.get("tooltip")
  });
}
function j(u, e, t, o, n) {
  e.highDownSilentOnTouch = !!n.get("selectedMode");
  var a = o.getModel("emphasis"), p = a.get("focus");
  return K(e, p, a.get("blurScope"), a.get("disabled")), u.isGeo && Q(e, n, t), p;
}
function x(u, e, t) {
  var o = [], n;
  function a() {
    n = [];
  }
  function p() {
    n.length && (o.push(n), n = []);
  }
  var r = e({
    polygonStart: a,
    polygonEnd: p,
    lineStart: a,
    lineEnd: p,
    point: function(s, g) {
      isFinite(s) && isFinite(g) && n.push([s, g]);
    },
    sphere: function() {
    }
  });
  return !t && r.polygonStart(), G(u, function(s) {
    r.lineStart();
    for (var g = 0; g < s.length; g++)
      r.point(s[g][0], s[g][1]);
    r.lineEnd();
  }), !t && r.polygonEnd(), o;
}
export {
  Ne as default
};
