// echarts@6.0.0/lib/component/helper/RoamController.js
import { __extends as k } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import P from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/Eventful.js";
import { isMiddleOrRightButtonOnMouseUpDown as b, stop as z } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/event.js";
import { isTaken as H } from "./interactionMutex.js";
import { bind as h, retrieve2 as O, extend as I, defaults as S, isString as D } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { makeInner as R } from "../../util/model.js";
import { retrieveZInfo as Z } from "../../util/graphic.js";
import { onIrrelevantElement as x } from "./cursorHelper.js";
var G = (
  /** @class */
  (function(n) {
    k(o, n);
    function o(e) {
      var r = n.call(this) || this;
      r._zr = e;
      var t = h(r._mousedownHandler, r), i = h(r._mousemoveHandler, r), a = h(r._mouseupHandler, r), l = h(r._mousewheelHandler, r), u = h(r._pinchHandler, r);
      return r.enable = function(s, f) {
        var v = f.zInfo, c = Z(v.component), C = c.z, X = c.zlevel, m = {
          component: v.component,
          z: C,
          zlevel: X,
          // By default roam controller is the lowest z2 comparing to other elememts in a component.
          z2: O(v.z2, -1 / 0)
        }, Y = I({}, f.triggerInfo);
        this._opt = S(I({}, f), {
          zoomOnMouseWheel: !0,
          moveOnMouseMove: !0,
          // By default, wheel do not trigger move.
          moveOnMouseWheel: !1,
          preventDefaultMouseMove: !0,
          zInfoParsed: m,
          triggerInfo: Y
        }), s == null && (s = !0), (!this._enabled || this._controlType !== s) && (this._enabled = !0, this.disable(), (s === !0 || s === "move" || s === "pan") && (d(e, "mousedown", t, m), d(e, "mousemove", i, m), d(e, "mouseup", a, m)), (s === !0 || s === "scale" || s === "zoom") && (d(e, "mousewheel", l, m), d(e, "pinch", u, m)));
      }, r.disable = function() {
        this._enabled = !1, _(e, "mousedown", t), _(e, "mousemove", i), _(e, "mouseup", a), _(e, "mousewheel", l), _(e, "pinch", u);
      }, r;
    }
    return o.prototype.isDragging = function() {
      return this._dragging;
    }, o.prototype.isPinching = function() {
      return this._pinching;
    }, o.prototype._checkPointer = function(e, r, t) {
      var i = this._opt, a = i.zInfoParsed;
      if (x(e, i.api, a.component))
        return !1;
      var l = i.triggerInfo, u = l.roamTrigger, s = !1;
      return u === "global" && (s = !0), s || (s = l.isInSelf(e, r, t)), s && l.isInClip && !l.isInClip(e, r, t) && (s = !1), s;
    }, o.prototype._decideCursorStyle = function(e, r, t, i) {
      var a = e.target;
      if (!a && this._checkPointer(e, r, t))
        return "grab";
      if (i)
        return a && a.cursor || "default";
    }, o.prototype.dispose = function() {
      this.disable();
    }, o.prototype._mousedownHandler = function(e) {
      if (!(b(e) || g(e))) {
        for (var r = e.target; r; ) {
          if (r.draggable)
            return;
          r = r.__hostTarget || r.parent;
        }
        var t = e.offsetX, i = e.offsetY;
        this._checkPointer(e, t, i) && (this._x = t, this._y = i, this._dragging = !0);
      }
    }, o.prototype._mousemoveHandler = function(e) {
      var r = this._zr;
      if (!(e.gestureEvent === "pinch" || H(r, "globalPan") || g(e))) {
        var t = e.offsetX, i = e.offsetY;
        if (!this._dragging || !p("moveOnMouseMove", e, this._opt)) {
          var a = this._decideCursorStyle(e, t, i, !1);
          a && r.setCursorStyle(a);
          return;
        }
        r.setCursorStyle("grabbing");
        var l = this._x, u = this._y, s = t - l, f = i - u;
        this._x = t, this._y = i, this._opt.preventDefaultMouseMove && z(e.event), e.__ecRoamConsumed = !0, w(this, "pan", "moveOnMouseMove", e, {
          dx: s,
          dy: f,
          oldX: l,
          oldY: u,
          newX: t,
          newY: i,
          isAvailableBehavior: null
        });
      }
    }, o.prototype._mouseupHandler = function(e) {
      if (!g(e)) {
        var r = this._zr;
        if (!b(e)) {
          this._dragging = !1;
          var t = this._decideCursorStyle(e, e.offsetX, e.offsetY, !0);
          t && r.setCursorStyle(t);
        }
      }
    }, o.prototype._mousewheelHandler = function(e) {
      if (!g(e)) {
        var r = p("zoomOnMouseWheel", e, this._opt), t = p("moveOnMouseWheel", e, this._opt), i = e.wheelDelta, a = Math.abs(i), l = e.offsetX, u = e.offsetY;
        if (!(i === 0 || !r && !t)) {
          if (r) {
            var s = a > 3 ? 1.4 : a > 1 ? 1.2 : 1.1, f = i > 0 ? s : 1 / s;
            this._checkTriggerMoveZoom(this, "zoom", "zoomOnMouseWheel", e, {
              scale: f,
              originX: l,
              originY: u,
              isAvailableBehavior: null
            });
          }
          if (t) {
            var v = Math.abs(i), c = (i > 0 ? 1 : -1) * (v > 3 ? 0.4 : v > 1 ? 0.15 : 0.05);
            this._checkTriggerMoveZoom(this, "scrollMove", "moveOnMouseWheel", e, {
              scrollDelta: c,
              originX: l,
              originY: u,
              isAvailableBehavior: null
            });
          }
        }
      }
    }, o.prototype._pinchHandler = function(e) {
      if (!(H(this._zr, "globalPan") || g(e))) {
        var r = e.pinchScale > 1 ? 1.1 : 1 / 1.1;
        this._checkTriggerMoveZoom(this, "zoom", null, e, {
          scale: r,
          originX: e.pinchX,
          originY: e.pinchY,
          isAvailableBehavior: null
        });
      }
    }, o.prototype._checkTriggerMoveZoom = function(e, r, t, i, a) {
      e._checkPointer(i, a.originX, a.originY) && (z(i.event), i.__ecRoamConsumed = !0, w(e, r, t, i, a));
    }, o;
  })(P)
);
function g(n) {
  return n.__ecRoamConsumed;
}
var y = R();
function M(n) {
  var o = y(n);
  return o.roam = o.roam || {}, o.uniform = o.uniform || {}, o;
}
function d(n, o, e, r) {
  for (var t = M(n), i = t.roam, a = i[o] = i[o] || [], l = 0; l < a.length; l++) {
    var u = a[l].zInfoParsed;
    if ((u.zlevel - r.zlevel || u.z - r.z || u.z2 - r.z2) <= 0)
      break;
  }
  a.splice(l, 0, {
    listener: e,
    zInfoParsed: r
  }), A(n, o);
}
function _(n, o, e) {
  for (var r = M(n), t = r.roam[o] || [], i = 0; i < t.length; i++)
    if (t[i].listener === e) {
      t.splice(i, 1), t.length || B(n, o);
      return;
    }
}
function A(n, o) {
  var e = M(n);
  e.uniform[o] || n.on(o, e.uniform[o] = function(r) {
    var t = e.roam[o];
    if (t)
      for (var i = 0; i < t.length; i++)
        t[i].listener(r);
  });
}
function B(n, o) {
  var e = M(n), r = e.uniform;
  r[o] && (n.off(o, r[o]), r[o] = null);
}
function w(n, o, e, r, t) {
  t.isAvailableBehavior = h(p, null, e, r), n.trigger(o, t);
}
function p(n, o, e) {
  var r = e[n];
  return !n || r && (!D(r) || o.event[r + "Key"]);
}
export {
  G as default
};
