// echarts@6.0.0/lib/component/helper/roamHelper.js
import "../../util/graphic.js";
import T from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/BoundingRect.js";
function z(n, o, m) {
  var t = n.target;
  t.x += o, t.y += m, t.dirty();
}
function x(n, o, m, t) {
  var a = n.target, u = n.zoomLimit, i = n.zoom = n.zoom || 1;
  i *= o, i = y(i, u);
  var e = i / n.zoom;
  n.zoom = i, p(a, m, t, e), a.dirty();
}
function Z(n, o, m, t, a, u) {
  var i = new T(0, 0, 0, 0);
  t.enable(n.get("roam"), {
    api: o,
    zInfo: {
      component: n
    },
    triggerInfo: {
      roamTrigger: n.get("roamTrigger"),
      isInSelf: function(r, f, g) {
        return i.copy(m.getBoundingRect()), i.applyTransform(m.getComputedTransform()), i.contain(f, g);
      },
      isInClip: function(r, f, g) {
        return !u || u.contain(f, g);
      }
    }
  }), a.zoomLimit = n.get("scaleLimit");
  var e = n.coordinateSystem;
  a.zoom = e ? e.getZoom() : 1;
  var c = n.subType + "Roam";
  t.off("pan").off("zoom").on("pan", function(r) {
    z(a, r.dx, r.dy), o.dispatchAction({
      seriesId: n.id,
      type: c,
      dx: r.dx,
      dy: r.dy
    });
  }).on("zoom", function(r) {
    x(a, r.scale, r.originX, r.originY), o.dispatchAction({
      seriesId: n.id,
      type: c,
      zoom: r.scale,
      originX: r.originX,
      originY: r.originY
    }), o.updateLabelLayout();
  });
}
function d(n, o) {
  return n.pointToProjected ? n.pointToProjected(o) : n.pointToData(o);
}
function P(n, o, m) {
  var t = n.getZoom(), a = n.getCenter(), u = o.zoom, i = n.projectedToPoint ? n.projectedToPoint(a) : n.dataToPoint(a);
  return o.dx != null && o.dy != null && (i[0] -= o.dx, i[1] -= o.dy, n.setCenter(d(n, i))), u != null && (u = y(t * u, m) / t, p(n, o.originX, o.originY, u), n.updateTransform(), n.setCenter(d(n, i)), n.setZoom(u * t)), {
    center: n.getCenter(),
    zoom: n.getZoom()
  };
}
function p(n, o, m, t) {
  n.x -= (o - n.x) * (t - 1), n.y -= (m - n.y) * (t - 1), n.scaleX *= t, n.scaleY *= t;
}
function y(n, o) {
  if (o) {
    var m = o.min || 0, t = o.max || 1 / 0;
    n = Math.max(Math.min(t, n), m);
  }
  return n;
}
export {
  y as clampByZoomLimit,
  P as updateCenterAndZoomInAction,
  Z as updateController,
  z as updateViewOnPan,
  x as updateViewOnZoom
};
