// @radix-ui/react-dropdown-menu@2.1.16/dist/index.js
import * as u from "react";
import { composeEventHandlers as m } from "../../../../../@radix-ui_primitive@1.1.3/node_modules/@radix-ui/primitive/dist/index.js";
import { composeRefs as L } from "../../../../../@radix-ui_react-compose-refs@1.1.2_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-compose-refs/dist/index.js";
import { createContextScope as K } from "../../../../../@radix-ui_react-context@1.1.2_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-context/dist/index.js";
import { useControllableState as g } from "../../../../../@radix-ui_react-use-controllable-state@1.2.2_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-use-controllable-state/dist/index.js";
import { Primitive as U } from "../../../../../@radix-ui_react-primitive@2.1.3_@types_react-dom@18.3.1_@types_react@18.3.18_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/@radix-ui/react-primitive/dist/index.js";
import { createMenuScope as D, Root as B, Anchor as F, Portal as H, Sub as W, Content as $, Item as j, Separator as X, SubTrigger as q, SubContent as z, Label as J, Group as Q, CheckboxItem as V, RadioGroup as Y, RadioItem as Z, ItemIndicator as oo, Arrow as eo } from "../../../../../@radix-ui_react-menu@2.1.16_@types_react-dom@18.3.1_@types_react@18.3.18_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/@radix-ui/react-menu/dist/index.js";
import { useId as v } from "../../../../../@radix-ui_react-id@1.1.1_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-id/dist/index.js";
import { jsx as t } from "react/jsx-runtime";
var f = "DropdownMenu", [ro] = K(
  f,
  [D]
), i = D(), [no, _] = ro(f), h = (o) => {
  const {
    __scopeDropdownMenu: n,
    children: r,
    dir: e,
    open: a,
    defaultOpen: d,
    onOpenChange: s,
    modal: p = !0
  } = o, c = i(n), M = u.useRef(null), [l, w] = g({
    prop: a,
    defaultProp: d ?? !1,
    onChange: s,
    caller: f
  });
  return /* @__PURE__ */ t(
    no,
    {
      scope: n,
      triggerId: v(),
      triggerRef: M,
      contentId: v(),
      open: l,
      onOpenChange: w,
      onOpenToggle: u.useCallback(() => w((k) => !k), [w]),
      modal: p,
      children: /* @__PURE__ */ t(B, { ...c, open: l, onOpenChange: w, dir: e, modal: p, children: r })
    }
  );
};
h.displayName = f;
var R = "DropdownMenuTrigger", b = u.forwardRef(
  (o, n) => {
    const { __scopeDropdownMenu: r, disabled: e = !1, ...a } = o, d = _(R, r), s = i(r);
    return /* @__PURE__ */ t(F, { asChild: !0, ...s, children: /* @__PURE__ */ t(
      U.button,
      {
        type: "button",
        id: d.triggerId,
        "aria-haspopup": "menu",
        "aria-expanded": d.open,
        "aria-controls": d.open ? d.contentId : void 0,
        "data-state": d.open ? "open" : "closed",
        "data-disabled": e ? "" : void 0,
        disabled: e,
        ...a,
        ref: L(n, d.triggerRef),
        onPointerDown: m(o.onPointerDown, (p) => {
          !e && p.button === 0 && p.ctrlKey === !1 && (d.onOpenToggle(), d.open || p.preventDefault());
        }),
        onKeyDown: m(o.onKeyDown, (p) => {
          e || (["Enter", " "].includes(p.key) && d.onOpenToggle(), p.key === "ArrowDown" && d.onOpenChange(!0), ["Enter", " ", "ArrowDown"].includes(p.key) && p.preventDefault());
        })
      }
    ) });
  }
);
b.displayName = R;
var ao = "DropdownMenuPortal", I = (o) => {
  const { __scopeDropdownMenu: n, ...r } = o, e = i(n);
  return /* @__PURE__ */ t(H, { ...e, ...r });
};
I.displayName = ao;
var S = "DropdownMenuContent", C = u.forwardRef(
  (o, n) => {
    const { __scopeDropdownMenu: r, ...e } = o, a = _(S, r), d = i(r), s = u.useRef(!1);
    return /* @__PURE__ */ t(
      $,
      {
        id: a.contentId,
        "aria-labelledby": a.triggerId,
        ...d,
        ...e,
        ref: n,
        onCloseAutoFocus: m(o.onCloseAutoFocus, (p) => {
          s.current || a.triggerRef.current?.focus(), s.current = !1, p.preventDefault();
        }),
        onInteractOutside: m(o.onInteractOutside, (p) => {
          const c = p.detail.originalEvent, M = c.button === 0 && c.ctrlKey === !0, l = c.button === 2 || M;
          (!a.modal || l) && (s.current = !0);
        }),
        style: {
          ...o.style,
          "--radix-dropdown-menu-content-transform-origin": "var(--radix-popper-transform-origin)",
          "--radix-dropdown-menu-content-available-width": "var(--radix-popper-available-width)",
          "--radix-dropdown-menu-content-available-height": "var(--radix-popper-available-height)",
          "--radix-dropdown-menu-trigger-width": "var(--radix-popper-anchor-width)",
          "--radix-dropdown-menu-trigger-height": "var(--radix-popper-anchor-height)"
        }
      }
    );
  }
);
C.displayName = S;
var to = "DropdownMenuGroup", N = u.forwardRef(
  (o, n) => {
    const { __scopeDropdownMenu: r, ...e } = o, a = i(r);
    return /* @__PURE__ */ t(Q, { ...a, ...e, ref: n });
  }
);
N.displayName = to;
var po = "DropdownMenuLabel", x = u.forwardRef(
  (o, n) => {
    const { __scopeDropdownMenu: r, ...e } = o, a = i(r);
    return /* @__PURE__ */ t(J, { ...a, ...e, ref: n });
  }
);
x.displayName = po;
var uo = "DropdownMenuItem", A = u.forwardRef(
  (o, n) => {
    const { __scopeDropdownMenu: r, ...e } = o, a = i(r);
    return /* @__PURE__ */ t(j, { ...a, ...e, ref: n });
  }
);
A.displayName = uo;
var io = "DropdownMenuCheckboxItem", P = u.forwardRef((o, n) => {
  const { __scopeDropdownMenu: r, ...e } = o, a = i(r);
  return /* @__PURE__ */ t(V, { ...a, ...e, ref: n });
});
P.displayName = io;
var so = "DropdownMenuRadioGroup", co = u.forwardRef((o, n) => {
  const { __scopeDropdownMenu: r, ...e } = o, a = i(r);
  return /* @__PURE__ */ t(Y, { ...a, ...e, ref: n });
});
co.displayName = so;
var lo = "DropdownMenuRadioItem", E = u.forwardRef((o, n) => {
  const { __scopeDropdownMenu: r, ...e } = o, a = i(r);
  return /* @__PURE__ */ t(Z, { ...a, ...e, ref: n });
});
E.displayName = lo;
var wo = "DropdownMenuItemIndicator", O = u.forwardRef((o, n) => {
  const { __scopeDropdownMenu: r, ...e } = o, a = i(r);
  return /* @__PURE__ */ t(oo, { ...a, ...e, ref: n });
});
O.displayName = wo;
var mo = "DropdownMenuSeparator", y = u.forwardRef((o, n) => {
  const { __scopeDropdownMenu: r, ...e } = o, a = i(r);
  return /* @__PURE__ */ t(X, { ...a, ...e, ref: n });
});
y.displayName = mo;
var fo = "DropdownMenuArrow", Mo = u.forwardRef(
  (o, n) => {
    const { __scopeDropdownMenu: r, ...e } = o, a = i(r);
    return /* @__PURE__ */ t(eo, { ...a, ...e, ref: n });
  }
);
Mo.displayName = fo;
var vo = (o) => {
  const { __scopeDropdownMenu: n, children: r, open: e, onOpenChange: a, defaultOpen: d } = o, s = i(n), [p, c] = g({
    prop: e,
    defaultProp: d ?? !1,
    onChange: a,
    caller: "DropdownMenuSub"
  });
  return /* @__PURE__ */ t(W, { ...s, open: p, onOpenChange: c, children: r });
}, go = "DropdownMenuSubTrigger", T = u.forwardRef((o, n) => {
  const { __scopeDropdownMenu: r, ...e } = o, a = i(r);
  return /* @__PURE__ */ t(q, { ...a, ...e, ref: n });
});
T.displayName = go;
var Do = "DropdownMenuSubContent", G = u.forwardRef((o, n) => {
  const { __scopeDropdownMenu: r, ...e } = o, a = i(r);
  return /* @__PURE__ */ t(
    z,
    {
      ...a,
      ...e,
      ref: n,
      style: {
        ...o.style,
        "--radix-dropdown-menu-content-transform-origin": "var(--radix-popper-transform-origin)",
        "--radix-dropdown-menu-content-available-width": "var(--radix-popper-available-width)",
        "--radix-dropdown-menu-content-available-height": "var(--radix-popper-available-height)",
        "--radix-dropdown-menu-trigger-width": "var(--radix-popper-anchor-width)",
        "--radix-dropdown-menu-trigger-height": "var(--radix-popper-anchor-height)"
      }
    }
  );
});
G.displayName = Do;
var xo = h, Ao = b, Po = I, Eo = C, Oo = N, yo = x, To = A, Go = P, ko = E, Lo = O, Ko = y, Uo = vo, Bo = T, Fo = G;
export {
  Go as CheckboxItem,
  Eo as Content,
  h as DropdownMenu,
  Mo as DropdownMenuArrow,
  P as DropdownMenuCheckboxItem,
  C as DropdownMenuContent,
  N as DropdownMenuGroup,
  A as DropdownMenuItem,
  O as DropdownMenuItemIndicator,
  x as DropdownMenuLabel,
  I as DropdownMenuPortal,
  co as DropdownMenuRadioGroup,
  E as DropdownMenuRadioItem,
  y as DropdownMenuSeparator,
  vo as DropdownMenuSub,
  G as DropdownMenuSubContent,
  T as DropdownMenuSubTrigger,
  b as DropdownMenuTrigger,
  Oo as Group,
  To as Item,
  Lo as ItemIndicator,
  yo as Label,
  Po as Portal,
  ko as RadioItem,
  xo as Root,
  Ko as Separator,
  Uo as Sub,
  Fo as SubContent,
  Bo as SubTrigger,
  Ao as Trigger
};
