// @tiptap/extension-details-content@2.23.0/dist/index.js
import { Node as I, findParentNode as L, defaultBlockAt as H, mergeAttributes as C } from "../../../../../@tiptap_core@2.24.0_@tiptap_pm@2.24.0/node_modules/@tiptap/core/dist/index.js";
import { Selection as S } from "../../../../../prosemirror-state@1.4.3/node_modules/prosemirror-state/dist/index.js";
const O = I.create({
  name: "detailsContent",
  content: "block+",
  defining: !0,
  selectable: !1,
  addOptions() {
    return {
      HTMLAttributes: {}
    };
  },
  parseHTML() {
    return [
      {
        tag: `div[data-type="${this.name}"]`
      }
    ];
  },
  renderHTML({ HTMLAttributes: r }) {
    return [
      "div",
      C(this.options.HTMLAttributes, r, { "data-type": this.name }),
      0
    ];
  },
  addNodeView() {
    return ({ HTMLAttributes: r }) => {
      const t = document.createElement("div"), d = C(this.options.HTMLAttributes, r, {
        "data-type": this.name,
        hidden: "hidden"
      });
      return Object.entries(d).forEach(([e, o]) => t.setAttribute(e, o)), t.addEventListener("toggleDetailsContent", () => {
        t.toggleAttribute("hidden");
      }), {
        dom: t,
        contentDOM: t,
        ignoreMutation(e) {
          return e.type === "selection" ? !1 : !t.contains(e.target) || t === e.target;
        },
        update: (e) => e.type === this.type
      };
    };
  },
  addKeyboardShortcuts() {
    return {
      // Escape node on double enter
      Enter: ({ editor: r }) => {
        const { state: t, view: d } = r, { selection: e } = t, { $from: o, empty: b } = e, n = L((x) => x.type === this.type)(e);
        if (!b || !n || !n.node.childCount)
          return !1;
        const g = o.index(n.depth), { childCount: f } = n.node;
        if (!(f === g + 1))
          return !1;
        const i = n.node.type.contentMatch.defaultType, p = i?.createAndFill();
        if (!p)
          return !1;
        const v = t.doc.resolve(n.pos + 1), h = f - 1, m = n.node.child(h), y = v.posAtIndex(h, n.depth);
        if (!m.eq(p))
          return !1;
        const a = o.node(-3);
        if (!a)
          return !1;
        const l = o.indexAfter(-3), c = H(a.contentMatchAt(l));
        if (!c || !a.canReplaceWith(l, l, c))
          return !1;
        const A = c.createAndFill();
        if (!A)
          return !1;
        const { tr: s } = t, u = o.after(-2);
        s.replaceWith(u, u, A);
        const M = s.doc.resolve(u), E = S.near(M, 1);
        s.setSelection(E);
        const T = y, N = y + m.nodeSize;
        return s.delete(T, N), s.scrollIntoView(), d.dispatch(s), !0;
      }
    };
  }
});
export {
  O as DetailsContent,
  O as default
};
