// @number-flow/react@0.5.10/dist/NumberFlow-client-48rw3j0J.js
import * as i from "react";
import { f as b, r as A, N as w, d as E } from "../../../../../number-flow@0.5.8/node_modules/number-flow/dist/lite-BTIaQdTe.js";
import { BROWSER as N } from "../../../../../esm-env-runtime@0.1.1/node_modules/esm-env-runtime/browser.js";
const F = parseInt(i.version.match(/^(\d+)\./)?.[1]), u = F >= 19, _ = [
  "data",
  "digits"
];
class c extends w {
  attributeChangedCallback(t, n, e) {
    this[t] = JSON.parse(e);
  }
}
c.observedAttributes = u ? [] : _;
E("number-flow-react", c);
const C = {}, f = u ? (o) => o : JSON.stringify;
function g(o) {
  const { transformTiming: t, spinTiming: n, opacityTiming: e, animated: s, respectMotionPreference: a, trend: p, plugins: d, ...r } = o;
  return [
    {
      transformTiming: t,
      spinTiming: n,
      opacityTiming: e,
      animated: s,
      respectMotionPreference: a,
      trend: p,
      plugins: d
    },
    r
  ];
}
class O extends i.Component {
  // Update the non-`data` props to avoid JSON serialization
  // Data needs to be set in render still:
  updateProperties(t) {
    if (!this.el) return;
    this.el.batched = !this.props.isolate;
    const [n] = g(this.props);
    Object.entries(n).forEach(([e, s]) => {
      this.el[e] = s ?? c.defaultProps[e];
    }), t?.onAnimationsStart && this.el.removeEventListener("animationsstart", t.onAnimationsStart), this.props.onAnimationsStart && this.el.addEventListener("animationsstart", this.props.onAnimationsStart), t?.onAnimationsFinish && this.el.removeEventListener("animationsfinish", t.onAnimationsFinish), this.props.onAnimationsFinish && this.el.addEventListener("animationsfinish", this.props.onAnimationsFinish);
  }
  componentDidMount() {
    this.updateProperties(), u && this.el && (this.el.digits = this.props.digits, this.el.data = this.props.data);
  }
  getSnapshotBeforeUpdate(t) {
    if (this.updateProperties(t), t.data !== this.props.data) {
      if (this.props.group)
        return this.props.group.willUpdate(), () => this.props.group?.didUpdate();
      if (!this.props.isolate)
        return this.el?.willUpdate(), () => this.el?.didUpdate();
    }
    return null;
  }
  componentDidUpdate(t, n, e) {
    e?.();
  }
  handleRef(t) {
    this.props.innerRef && (this.props.innerRef.current = t), this.el = t;
  }
  render() {
    const [t, { innerRef: n, className: e, data: s, willChange: a, isolate: p, group: d, digits: r, onAnimationsStart: l, onAnimationsFinish: m, ...h }] = g(this.props);
    return (
      // @ts-expect-error missing types
      /* @__PURE__ */ i.createElement("number-flow-react", {
        ref: this.handleRef,
        "data-will-change": a ? "" : void 0,
        // Have to rename this:
        class: e,
        ...h,
        dangerouslySetInnerHTML: {
          __html: N ? "" : A(s)
        },
        suppressHydrationWarning: !0,
        digits: f(r),
        // Make sure data is set last, everything else is updated:
        data: f(s)
      })
    );
  }
  constructor(t) {
    super(t), this.handleRef = this.handleRef.bind(this);
  }
}
const y = /* @__PURE__ */ i.forwardRef(function({ value: t, locales: n, format: e, prefix: s, suffix: a, ...p }, d) {
  i.useImperativeHandle(d, () => r.current, []);
  const r = i.useRef(), l = i.useContext(T);
  l?.useRegister(r);
  const m = i.useMemo(() => n ? JSON.stringify(n) : "", [
    n
  ]), h = i.useMemo(() => e ? JSON.stringify(e) : "", [
    e
  ]), R = i.useMemo(() => {
    const S = C[`${m}:${h}`] ??= new Intl.NumberFormat(n, e);
    return b(t, S, s, a);
  }, [
    t,
    m,
    h,
    s,
    a
  ]);
  return /* @__PURE__ */ i.createElement(O, {
    ...p,
    group: l,
    data: R,
    innerRef: r
  });
}), T = /* @__PURE__ */ i.createContext(void 0);
export {
  y as N,
  c as a
};
