// zod@3.25.76/v3/types.js
import { ZodIssueCode as c, ZodError as M } from "./ZodError.js";
import { errorUtil as u } from "./helpers/errorUtil.js";
import { addIssueToContext as o, INVALID as p, ParseStatus as w, isAsync as U, isValid as N, OK as k, DIRTY as z, isAborted as se, isDirty as ae } from "./helpers/parseUtil.js";
import { util as v, ZodParsedType as h, getParsedType as C } from "./helpers/util.js";
class A {
  constructor(e, t, s, a) {
    this._cachedPath = [], this.parent = e, this.data = t, this._path = s, this._key = a;
  }
  get path() {
    return this._cachedPath.length || (Array.isArray(this._key) ? this._cachedPath.push(...this._path, ...this._key) : this._cachedPath.push(...this._path, this._key)), this._cachedPath;
  }
}
const re = (n, e) => {
  if (N(e))
    return { success: !0, data: e.value };
  if (!n.common.issues.length)
    throw new Error("Validation failed but no issues detected.");
  return {
    success: !1,
    get error() {
      if (this._error)
        return this._error;
      const t = new M(n.common.issues);
      return this._error = t, this._error;
    }
  };
};
function _(n) {
  if (!n)
    return {};
  const { errorMap: e, invalid_type_error: t, required_error: s, description: a } = n;
  if (e && (t || s))
    throw new Error(`Can't use "invalid_type_error" or "required_error" in conjunction with custom error map.`);
  return e ? { errorMap: e, description: a } : { errorMap: (d, i) => {
    const { message: l } = n;
    return d.code === "invalid_enum_value" ? { message: l ?? i.defaultError } : typeof i.data > "u" ? { message: l ?? s ?? i.defaultError } : d.code !== "invalid_type" ? { message: i.defaultError } : { message: l ?? t ?? i.defaultError };
  }, description: a };
}
class g {
  get description() {
    return this._def.description;
  }
  _getType(e) {
    return C(e.data);
  }
  _getOrReturnCtx(e, t) {
    return t || {
      common: e.parent.common,
      data: e.data,
      parsedType: C(e.data),
      schemaErrorMap: this._def.errorMap,
      path: e.path,
      parent: e.parent
    };
  }
  _processInputParams(e) {
    return {
      status: new w(),
      ctx: {
        common: e.parent.common,
        data: e.data,
        parsedType: C(e.data),
        schemaErrorMap: this._def.errorMap,
        path: e.path,
        parent: e.parent
      }
    };
  }
  _parseSync(e) {
    const t = this._parse(e);
    if (U(t))
      throw new Error("Synchronous parse encountered promise.");
    return t;
  }
  _parseAsync(e) {
    const t = this._parse(e);
    return Promise.resolve(t);
  }
  parse(e, t) {
    const s = this.safeParse(e, t);
    if (s.success)
      return s.data;
    throw s.error;
  }
  safeParse(e, t) {
    const s = {
      common: {
        issues: [],
        async: t?.async ?? !1,
        contextualErrorMap: t?.errorMap
      },
      path: t?.path || [],
      schemaErrorMap: this._def.errorMap,
      parent: null,
      data: e,
      parsedType: C(e)
    }, a = this._parseSync({ data: e, path: s.path, parent: s });
    return re(s, a);
  }
  "~validate"(e) {
    const t = {
      common: {
        issues: [],
        async: !!this["~standard"].async
      },
      path: [],
      schemaErrorMap: this._def.errorMap,
      parent: null,
      data: e,
      parsedType: C(e)
    };
    if (!this["~standard"].async)
      try {
        const s = this._parseSync({ data: e, path: [], parent: t });
        return N(s) ? {
          value: s.value
        } : {
          issues: t.common.issues
        };
      } catch (s) {
        s?.message?.toLowerCase()?.includes("encountered") && (this["~standard"].async = !0), t.common = {
          issues: [],
          async: !0
        };
      }
    return this._parseAsync({ data: e, path: [], parent: t }).then((s) => N(s) ? {
      value: s.value
    } : {
      issues: t.common.issues
    });
  }
  async parseAsync(e, t) {
    const s = await this.safeParseAsync(e, t);
    if (s.success)
      return s.data;
    throw s.error;
  }
  async safeParseAsync(e, t) {
    const s = {
      common: {
        issues: [],
        contextualErrorMap: t?.errorMap,
        async: !0
      },
      path: t?.path || [],
      schemaErrorMap: this._def.errorMap,
      parent: null,
      data: e,
      parsedType: C(e)
    }, a = this._parse({ data: e, path: s.path, parent: s }), r = await (U(a) ? a : Promise.resolve(a));
    return re(s, r);
  }
  refine(e, t) {
    const s = (a) => typeof t == "string" || typeof t > "u" ? { message: t } : typeof t == "function" ? t(a) : t;
    return this._refinement((a, r) => {
      const d = e(a), i = () => r.addIssue({
        code: c.custom,
        ...s(a)
      });
      return typeof Promise < "u" && d instanceof Promise ? d.then((l) => l ? !0 : (i(), !1)) : d ? !0 : (i(), !1);
    });
  }
  refinement(e, t) {
    return this._refinement((s, a) => e(s) ? !0 : (a.addIssue(typeof t == "function" ? t(s, a) : t), !1));
  }
  _refinement(e) {
    return new E({
      schema: this,
      typeName: f.ZodEffects,
      effect: { type: "refinement", refinement: e }
    });
  }
  superRefine(e) {
    return this._refinement(e);
  }
  constructor(e) {
    this.spa = this.safeParseAsync, this._def = e, this.parse = this.parse.bind(this), this.safeParse = this.safeParse.bind(this), this.parseAsync = this.parseAsync.bind(this), this.safeParseAsync = this.safeParseAsync.bind(this), this.spa = this.spa.bind(this), this.refine = this.refine.bind(this), this.refinement = this.refinement.bind(this), this.superRefine = this.superRefine.bind(this), this.optional = this.optional.bind(this), this.nullable = this.nullable.bind(this), this.nullish = this.nullish.bind(this), this.array = this.array.bind(this), this.promise = this.promise.bind(this), this.or = this.or.bind(this), this.and = this.and.bind(this), this.transform = this.transform.bind(this), this.brand = this.brand.bind(this), this.default = this.default.bind(this), this.catch = this.catch.bind(this), this.describe = this.describe.bind(this), this.pipe = this.pipe.bind(this), this.readonly = this.readonly.bind(this), this.isNullable = this.isNullable.bind(this), this.isOptional = this.isOptional.bind(this), this["~standard"] = {
      version: 1,
      vendor: "zod",
      validate: (t) => this["~validate"](t)
    };
  }
  optional() {
    return Z.create(this, this._def);
  }
  nullable() {
    return L.create(this, this._def);
  }
  nullish() {
    return this.nullable().optional();
  }
  array() {
    return b.create(this);
  }
  promise() {
    return W.create(this, this._def);
  }
  or(e) {
    return B.create([this, e], this._def);
  }
  and(e) {
    return F.create(this, e, this._def);
  }
  transform(e) {
    return new E({
      ..._(this._def),
      schema: this,
      typeName: f.ZodEffects,
      effect: { type: "transform", transform: e }
    });
  }
  default(e) {
    const t = typeof e == "function" ? e : () => e;
    return new X({
      ..._(this._def),
      innerType: this,
      defaultValue: t,
      typeName: f.ZodDefault
    });
  }
  brand() {
    return new Pe({
      typeName: f.ZodBranded,
      type: this,
      ..._(this._def)
    });
  }
  catch(e) {
    const t = typeof e == "function" ? e : () => e;
    return new K({
      ..._(this._def),
      innerType: this,
      catchValue: t,
      typeName: f.ZodCatch
    });
  }
  describe(e) {
    const t = this.constructor;
    return new t({
      ...this._def,
      description: e
    });
  }
  pipe(e) {
    return te.create(this, e);
  }
  readonly() {
    return ee.create(this);
  }
  isOptional() {
    return this.safeParse(void 0).success;
  }
  isNullable() {
    return this.safeParse(null).success;
  }
}
const ge = /^c[^\s-]{8,}$/i, ye = /^[0-9a-z]+$/, ve = /^[0-9A-HJKMNP-TV-Z]{26}$/i, xe = /^[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}$/i, ke = /^[a-z0-9_-]{21}$/i, we = /^[A-Za-z0-9-_]+\.[A-Za-z0-9-_]+\.[A-Za-z0-9-_]*$/, be = /^[-+]?P(?!$)(?:(?:[-+]?\d+Y)|(?:[-+]?\d+[.,]\d+Y$))?(?:(?:[-+]?\d+M)|(?:[-+]?\d+[.,]\d+M$))?(?:(?:[-+]?\d+W)|(?:[-+]?\d+[.,]\d+W$))?(?:(?:[-+]?\d+D)|(?:[-+]?\d+[.,]\d+D$))?(?:T(?=[\d+-])(?:(?:[-+]?\d+H)|(?:[-+]?\d+[.,]\d+H$))?(?:(?:[-+]?\d+M)|(?:[-+]?\d+[.,]\d+M$))?(?:[-+]?\d+(?:[.,]\d+)?S)?)??$/, Ce = /^(?!\.)(?!.*\.\.)([A-Z0-9_'+\-\.]*)[A-Z0-9_+-]@([A-Z0-9][A-Z0-9\-]*\.)+[A-Z]{2,}$/i, Te = "^(\\p{Extended_Pictographic}|\\p{Emoji_Component})+$";
let J;
const Ze = /^(?:(?:25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9][0-9]|[0-9])\.){3}(?:25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9][0-9]|[0-9])$/, Ae = /^(?:(?:25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9][0-9]|[0-9])\.){3}(?:25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9][0-9]|[0-9])\/(3[0-2]|[12]?[0-9])$/, Se = /^(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9]))$/, Oe = /^(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9]))\/(12[0-8]|1[01][0-9]|[1-9]?[0-9])$/, Re = /^([0-9a-zA-Z+/]{4})*(([0-9a-zA-Z+/]{2}==)|([0-9a-zA-Z+/]{3}=))?$/, Ne = /^([0-9a-zA-Z-_]{4})*(([0-9a-zA-Z-_]{2}(==)?)|([0-9a-zA-Z-_]{3}(=)?))?$/, fe = "((\\d\\d[2468][048]|\\d\\d[13579][26]|\\d\\d0[48]|[02468][048]00|[13579][26]00)-02-29|\\d{4}-((0[13578]|1[02])-(0[1-9]|[12]\\d|3[01])|(0[469]|11)-(0[1-9]|[12]\\d|30)|(02)-(0[1-9]|1\\d|2[0-8])))", Ie = new RegExp(`^${fe}$`);
function me(n) {
  let e = "[0-5]\\d";
  n.precision ? e = `${e}\\.\\d{${n.precision}}` : n.precision == null && (e = `${e}(\\.\\d+)?`);
  const t = n.precision ? "+" : "?";
  return `([01]\\d|2[0-3]):[0-5]\\d(:${e})${t}`;
}
function je(n) {
  return new RegExp(`^${me(n)}$`);
}
function Ee(n) {
  let e = `${fe}T${me(n)}`;
  const t = [];
  return t.push(n.local ? "Z?" : "Z"), n.offset && t.push("([+-]\\d{2}:?\\d{2})"), e = `${e}(${t.join("|")})`, new RegExp(`^${e}$`);
}
function Le(n, e) {
  return !!((e === "v4" || !e) && Ze.test(n) || (e === "v6" || !e) && Se.test(n));
}
function Ve(n, e) {
  if (!we.test(n))
    return !1;
  try {
    const [t] = n.split(".");
    if (!t)
      return !1;
    const s = t.replace(/-/g, "+").replace(/_/g, "/").padEnd(t.length + (4 - t.length % 4) % 4, "="), a = JSON.parse(atob(s));
    return !(typeof a != "object" || a === null || "typ" in a && a?.typ !== "JWT" || !a.alg || e && a.alg !== e);
  } catch {
    return !1;
  }
}
function ze(n, e) {
  return !!((e === "v4" || !e) && Ae.test(n) || (e === "v6" || !e) && Oe.test(n));
}
class T extends g {
  _parse(e) {
    if (this._def.coerce && (e.data = String(e.data)), this._getType(e) !== h.string) {
      const r = this._getOrReturnCtx(e);
      return o(r, {
        code: c.invalid_type,
        expected: h.string,
        received: r.parsedType
      }), p;
    }
    const s = new w();
    let a;
    for (const r of this._def.checks)
      if (r.kind === "min")
        e.data.length < r.value && (a = this._getOrReturnCtx(e, a), o(a, {
          code: c.too_small,
          minimum: r.value,
          type: "string",
          inclusive: !0,
          exact: !1,
          message: r.message
        }), s.dirty());
      else if (r.kind === "max")
        e.data.length > r.value && (a = this._getOrReturnCtx(e, a), o(a, {
          code: c.too_big,
          maximum: r.value,
          type: "string",
          inclusive: !0,
          exact: !1,
          message: r.message
        }), s.dirty());
      else if (r.kind === "length") {
        const d = e.data.length > r.value, i = e.data.length < r.value;
        (d || i) && (a = this._getOrReturnCtx(e, a), d ? o(a, {
          code: c.too_big,
          maximum: r.value,
          type: "string",
          inclusive: !0,
          exact: !0,
          message: r.message
        }) : i && o(a, {
          code: c.too_small,
          minimum: r.value,
          type: "string",
          inclusive: !0,
          exact: !0,
          message: r.message
        }), s.dirty());
      } else if (r.kind === "email")
        Ce.test(e.data) || (a = this._getOrReturnCtx(e, a), o(a, {
          validation: "email",
          code: c.invalid_string,
          message: r.message
        }), s.dirty());
      else if (r.kind === "emoji")
        J || (J = new RegExp(Te, "u")), J.test(e.data) || (a = this._getOrReturnCtx(e, a), o(a, {
          validation: "emoji",
          code: c.invalid_string,
          message: r.message
        }), s.dirty());
      else if (r.kind === "uuid")
        xe.test(e.data) || (a = this._getOrReturnCtx(e, a), o(a, {
          validation: "uuid",
          code: c.invalid_string,
          message: r.message
        }), s.dirty());
      else if (r.kind === "nanoid")
        ke.test(e.data) || (a = this._getOrReturnCtx(e, a), o(a, {
          validation: "nanoid",
          code: c.invalid_string,
          message: r.message
        }), s.dirty());
      else if (r.kind === "cuid")
        ge.test(e.data) || (a = this._getOrReturnCtx(e, a), o(a, {
          validation: "cuid",
          code: c.invalid_string,
          message: r.message
        }), s.dirty());
      else if (r.kind === "cuid2")
        ye.test(e.data) || (a = this._getOrReturnCtx(e, a), o(a, {
          validation: "cuid2",
          code: c.invalid_string,
          message: r.message
        }), s.dirty());
      else if (r.kind === "ulid")
        ve.test(e.data) || (a = this._getOrReturnCtx(e, a), o(a, {
          validation: "ulid",
          code: c.invalid_string,
          message: r.message
        }), s.dirty());
      else if (r.kind === "url")
        try {
          new URL(e.data);
        } catch {
          a = this._getOrReturnCtx(e, a), o(a, {
            validation: "url",
            code: c.invalid_string,
            message: r.message
          }), s.dirty();
        }
      else r.kind === "regex" ? (r.regex.lastIndex = 0, r.regex.test(e.data) || (a = this._getOrReturnCtx(e, a), o(a, {
        validation: "regex",
        code: c.invalid_string,
        message: r.message
      }), s.dirty())) : r.kind === "trim" ? e.data = e.data.trim() : r.kind === "includes" ? e.data.includes(r.value, r.position) || (a = this._getOrReturnCtx(e, a), o(a, {
        code: c.invalid_string,
        validation: { includes: r.value, position: r.position },
        message: r.message
      }), s.dirty()) : r.kind === "toLowerCase" ? e.data = e.data.toLowerCase() : r.kind === "toUpperCase" ? e.data = e.data.toUpperCase() : r.kind === "startsWith" ? e.data.startsWith(r.value) || (a = this._getOrReturnCtx(e, a), o(a, {
        code: c.invalid_string,
        validation: { startsWith: r.value },
        message: r.message
      }), s.dirty()) : r.kind === "endsWith" ? e.data.endsWith(r.value) || (a = this._getOrReturnCtx(e, a), o(a, {
        code: c.invalid_string,
        validation: { endsWith: r.value },
        message: r.message
      }), s.dirty()) : r.kind === "datetime" ? Ee(r).test(e.data) || (a = this._getOrReturnCtx(e, a), o(a, {
        code: c.invalid_string,
        validation: "datetime",
        message: r.message
      }), s.dirty()) : r.kind === "date" ? Ie.test(e.data) || (a = this._getOrReturnCtx(e, a), o(a, {
        code: c.invalid_string,
        validation: "date",
        message: r.message
      }), s.dirty()) : r.kind === "time" ? je(r).test(e.data) || (a = this._getOrReturnCtx(e, a), o(a, {
        code: c.invalid_string,
        validation: "time",
        message: r.message
      }), s.dirty()) : r.kind === "duration" ? be.test(e.data) || (a = this._getOrReturnCtx(e, a), o(a, {
        validation: "duration",
        code: c.invalid_string,
        message: r.message
      }), s.dirty()) : r.kind === "ip" ? Le(e.data, r.version) || (a = this._getOrReturnCtx(e, a), o(a, {
        validation: "ip",
        code: c.invalid_string,
        message: r.message
      }), s.dirty()) : r.kind === "jwt" ? Ve(e.data, r.alg) || (a = this._getOrReturnCtx(e, a), o(a, {
        validation: "jwt",
        code: c.invalid_string,
        message: r.message
      }), s.dirty()) : r.kind === "cidr" ? ze(e.data, r.version) || (a = this._getOrReturnCtx(e, a), o(a, {
        validation: "cidr",
        code: c.invalid_string,
        message: r.message
      }), s.dirty()) : r.kind === "base64" ? Re.test(e.data) || (a = this._getOrReturnCtx(e, a), o(a, {
        validation: "base64",
        code: c.invalid_string,
        message: r.message
      }), s.dirty()) : r.kind === "base64url" ? Ne.test(e.data) || (a = this._getOrReturnCtx(e, a), o(a, {
        validation: "base64url",
        code: c.invalid_string,
        message: r.message
      }), s.dirty()) : v.assertNever(r);
    return { status: s.value, value: e.data };
  }
  _regex(e, t, s) {
    return this.refinement((a) => e.test(a), {
      validation: t,
      code: c.invalid_string,
      ...u.errToObj(s)
    });
  }
  _addCheck(e) {
    return new T({
      ...this._def,
      checks: [...this._def.checks, e]
    });
  }
  email(e) {
    return this._addCheck({ kind: "email", ...u.errToObj(e) });
  }
  url(e) {
    return this._addCheck({ kind: "url", ...u.errToObj(e) });
  }
  emoji(e) {
    return this._addCheck({ kind: "emoji", ...u.errToObj(e) });
  }
  uuid(e) {
    return this._addCheck({ kind: "uuid", ...u.errToObj(e) });
  }
  nanoid(e) {
    return this._addCheck({ kind: "nanoid", ...u.errToObj(e) });
  }
  cuid(e) {
    return this._addCheck({ kind: "cuid", ...u.errToObj(e) });
  }
  cuid2(e) {
    return this._addCheck({ kind: "cuid2", ...u.errToObj(e) });
  }
  ulid(e) {
    return this._addCheck({ kind: "ulid", ...u.errToObj(e) });
  }
  base64(e) {
    return this._addCheck({ kind: "base64", ...u.errToObj(e) });
  }
  base64url(e) {
    return this._addCheck({
      kind: "base64url",
      ...u.errToObj(e)
    });
  }
  jwt(e) {
    return this._addCheck({ kind: "jwt", ...u.errToObj(e) });
  }
  ip(e) {
    return this._addCheck({ kind: "ip", ...u.errToObj(e) });
  }
  cidr(e) {
    return this._addCheck({ kind: "cidr", ...u.errToObj(e) });
  }
  datetime(e) {
    return typeof e == "string" ? this._addCheck({
      kind: "datetime",
      precision: null,
      offset: !1,
      local: !1,
      message: e
    }) : this._addCheck({
      kind: "datetime",
      precision: typeof e?.precision > "u" ? null : e?.precision,
      offset: e?.offset ?? !1,
      local: e?.local ?? !1,
      ...u.errToObj(e?.message)
    });
  }
  date(e) {
    return this._addCheck({ kind: "date", message: e });
  }
  time(e) {
    return typeof e == "string" ? this._addCheck({
      kind: "time",
      precision: null,
      message: e
    }) : this._addCheck({
      kind: "time",
      precision: typeof e?.precision > "u" ? null : e?.precision,
      ...u.errToObj(e?.message)
    });
  }
  duration(e) {
    return this._addCheck({ kind: "duration", ...u.errToObj(e) });
  }
  regex(e, t) {
    return this._addCheck({
      kind: "regex",
      regex: e,
      ...u.errToObj(t)
    });
  }
  includes(e, t) {
    return this._addCheck({
      kind: "includes",
      value: e,
      position: t?.position,
      ...u.errToObj(t?.message)
    });
  }
  startsWith(e, t) {
    return this._addCheck({
      kind: "startsWith",
      value: e,
      ...u.errToObj(t)
    });
  }
  endsWith(e, t) {
    return this._addCheck({
      kind: "endsWith",
      value: e,
      ...u.errToObj(t)
    });
  }
  min(e, t) {
    return this._addCheck({
      kind: "min",
      value: e,
      ...u.errToObj(t)
    });
  }
  max(e, t) {
    return this._addCheck({
      kind: "max",
      value: e,
      ...u.errToObj(t)
    });
  }
  length(e, t) {
    return this._addCheck({
      kind: "length",
      value: e,
      ...u.errToObj(t)
    });
  }
  /**
   * Equivalent to `.min(1)`
   */
  nonempty(e) {
    return this.min(1, u.errToObj(e));
  }
  trim() {
    return new T({
      ...this._def,
      checks: [...this._def.checks, { kind: "trim" }]
    });
  }
  toLowerCase() {
    return new T({
      ...this._def,
      checks: [...this._def.checks, { kind: "toLowerCase" }]
    });
  }
  toUpperCase() {
    return new T({
      ...this._def,
      checks: [...this._def.checks, { kind: "toUpperCase" }]
    });
  }
  get isDatetime() {
    return !!this._def.checks.find((e) => e.kind === "datetime");
  }
  get isDate() {
    return !!this._def.checks.find((e) => e.kind === "date");
  }
  get isTime() {
    return !!this._def.checks.find((e) => e.kind === "time");
  }
  get isDuration() {
    return !!this._def.checks.find((e) => e.kind === "duration");
  }
  get isEmail() {
    return !!this._def.checks.find((e) => e.kind === "email");
  }
  get isURL() {
    return !!this._def.checks.find((e) => e.kind === "url");
  }
  get isEmoji() {
    return !!this._def.checks.find((e) => e.kind === "emoji");
  }
  get isUUID() {
    return !!this._def.checks.find((e) => e.kind === "uuid");
  }
  get isNANOID() {
    return !!this._def.checks.find((e) => e.kind === "nanoid");
  }
  get isCUID() {
    return !!this._def.checks.find((e) => e.kind === "cuid");
  }
  get isCUID2() {
    return !!this._def.checks.find((e) => e.kind === "cuid2");
  }
  get isULID() {
    return !!this._def.checks.find((e) => e.kind === "ulid");
  }
  get isIP() {
    return !!this._def.checks.find((e) => e.kind === "ip");
  }
  get isCIDR() {
    return !!this._def.checks.find((e) => e.kind === "cidr");
  }
  get isBase64() {
    return !!this._def.checks.find((e) => e.kind === "base64");
  }
  get isBase64url() {
    return !!this._def.checks.find((e) => e.kind === "base64url");
  }
  get minLength() {
    let e = null;
    for (const t of this._def.checks)
      t.kind === "min" && (e === null || t.value > e) && (e = t.value);
    return e;
  }
  get maxLength() {
    let e = null;
    for (const t of this._def.checks)
      t.kind === "max" && (e === null || t.value < e) && (e = t.value);
    return e;
  }
}
T.create = (n) => new T({
  checks: [],
  typeName: f.ZodString,
  coerce: n?.coerce ?? !1,
  ..._(n)
});
function Me(n, e) {
  const t = (n.toString().split(".")[1] || "").length, s = (e.toString().split(".")[1] || "").length, a = t > s ? t : s, r = Number.parseInt(n.toFixed(a).replace(".", "")), d = Number.parseInt(e.toFixed(a).replace(".", ""));
  return r % d / 10 ** a;
}
class I extends g {
  constructor() {
    super(...arguments), this.min = this.gte, this.max = this.lte, this.step = this.multipleOf;
  }
  _parse(e) {
    if (this._def.coerce && (e.data = Number(e.data)), this._getType(e) !== h.number) {
      const r = this._getOrReturnCtx(e);
      return o(r, {
        code: c.invalid_type,
        expected: h.number,
        received: r.parsedType
      }), p;
    }
    let s;
    const a = new w();
    for (const r of this._def.checks)
      r.kind === "int" ? v.isInteger(e.data) || (s = this._getOrReturnCtx(e, s), o(s, {
        code: c.invalid_type,
        expected: "integer",
        received: "float",
        message: r.message
      }), a.dirty()) : r.kind === "min" ? (r.inclusive ? e.data < r.value : e.data <= r.value) && (s = this._getOrReturnCtx(e, s), o(s, {
        code: c.too_small,
        minimum: r.value,
        type: "number",
        inclusive: r.inclusive,
        exact: !1,
        message: r.message
      }), a.dirty()) : r.kind === "max" ? (r.inclusive ? e.data > r.value : e.data >= r.value) && (s = this._getOrReturnCtx(e, s), o(s, {
        code: c.too_big,
        maximum: r.value,
        type: "number",
        inclusive: r.inclusive,
        exact: !1,
        message: r.message
      }), a.dirty()) : r.kind === "multipleOf" ? Me(e.data, r.value) !== 0 && (s = this._getOrReturnCtx(e, s), o(s, {
        code: c.not_multiple_of,
        multipleOf: r.value,
        message: r.message
      }), a.dirty()) : r.kind === "finite" ? Number.isFinite(e.data) || (s = this._getOrReturnCtx(e, s), o(s, {
        code: c.not_finite,
        message: r.message
      }), a.dirty()) : v.assertNever(r);
    return { status: a.value, value: e.data };
  }
  gte(e, t) {
    return this.setLimit("min", e, !0, u.toString(t));
  }
  gt(e, t) {
    return this.setLimit("min", e, !1, u.toString(t));
  }
  lte(e, t) {
    return this.setLimit("max", e, !0, u.toString(t));
  }
  lt(e, t) {
    return this.setLimit("max", e, !1, u.toString(t));
  }
  setLimit(e, t, s, a) {
    return new I({
      ...this._def,
      checks: [
        ...this._def.checks,
        {
          kind: e,
          value: t,
          inclusive: s,
          message: u.toString(a)
        }
      ]
    });
  }
  _addCheck(e) {
    return new I({
      ...this._def,
      checks: [...this._def.checks, e]
    });
  }
  int(e) {
    return this._addCheck({
      kind: "int",
      message: u.toString(e)
    });
  }
  positive(e) {
    return this._addCheck({
      kind: "min",
      value: 0,
      inclusive: !1,
      message: u.toString(e)
    });
  }
  negative(e) {
    return this._addCheck({
      kind: "max",
      value: 0,
      inclusive: !1,
      message: u.toString(e)
    });
  }
  nonpositive(e) {
    return this._addCheck({
      kind: "max",
      value: 0,
      inclusive: !0,
      message: u.toString(e)
    });
  }
  nonnegative(e) {
    return this._addCheck({
      kind: "min",
      value: 0,
      inclusive: !0,
      message: u.toString(e)
    });
  }
  multipleOf(e, t) {
    return this._addCheck({
      kind: "multipleOf",
      value: e,
      message: u.toString(t)
    });
  }
  finite(e) {
    return this._addCheck({
      kind: "finite",
      message: u.toString(e)
    });
  }
  safe(e) {
    return this._addCheck({
      kind: "min",
      inclusive: !0,
      value: Number.MIN_SAFE_INTEGER,
      message: u.toString(e)
    })._addCheck({
      kind: "max",
      inclusive: !0,
      value: Number.MAX_SAFE_INTEGER,
      message: u.toString(e)
    });
  }
  get minValue() {
    let e = null;
    for (const t of this._def.checks)
      t.kind === "min" && (e === null || t.value > e) && (e = t.value);
    return e;
  }
  get maxValue() {
    let e = null;
    for (const t of this._def.checks)
      t.kind === "max" && (e === null || t.value < e) && (e = t.value);
    return e;
  }
  get isInt() {
    return !!this._def.checks.find((e) => e.kind === "int" || e.kind === "multipleOf" && v.isInteger(e.value));
  }
  get isFinite() {
    let e = null, t = null;
    for (const s of this._def.checks) {
      if (s.kind === "finite" || s.kind === "int" || s.kind === "multipleOf")
        return !0;
      s.kind === "min" ? (t === null || s.value > t) && (t = s.value) : s.kind === "max" && (e === null || s.value < e) && (e = s.value);
    }
    return Number.isFinite(t) && Number.isFinite(e);
  }
}
I.create = (n) => new I({
  checks: [],
  typeName: f.ZodNumber,
  coerce: n?.coerce || !1,
  ..._(n)
});
class P extends g {
  constructor() {
    super(...arguments), this.min = this.gte, this.max = this.lte;
  }
  _parse(e) {
    if (this._def.coerce)
      try {
        e.data = BigInt(e.data);
      } catch {
        return this._getInvalidInput(e);
      }
    if (this._getType(e) !== h.bigint)
      return this._getInvalidInput(e);
    let s;
    const a = new w();
    for (const r of this._def.checks)
      r.kind === "min" ? (r.inclusive ? e.data < r.value : e.data <= r.value) && (s = this._getOrReturnCtx(e, s), o(s, {
        code: c.too_small,
        type: "bigint",
        minimum: r.value,
        inclusive: r.inclusive,
        message: r.message
      }), a.dirty()) : r.kind === "max" ? (r.inclusive ? e.data > r.value : e.data >= r.value) && (s = this._getOrReturnCtx(e, s), o(s, {
        code: c.too_big,
        type: "bigint",
        maximum: r.value,
        inclusive: r.inclusive,
        message: r.message
      }), a.dirty()) : r.kind === "multipleOf" ? e.data % r.value !== BigInt(0) && (s = this._getOrReturnCtx(e, s), o(s, {
        code: c.not_multiple_of,
        multipleOf: r.value,
        message: r.message
      }), a.dirty()) : v.assertNever(r);
    return { status: a.value, value: e.data };
  }
  _getInvalidInput(e) {
    const t = this._getOrReturnCtx(e);
    return o(t, {
      code: c.invalid_type,
      expected: h.bigint,
      received: t.parsedType
    }), p;
  }
  gte(e, t) {
    return this.setLimit("min", e, !0, u.toString(t));
  }
  gt(e, t) {
    return this.setLimit("min", e, !1, u.toString(t));
  }
  lte(e, t) {
    return this.setLimit("max", e, !0, u.toString(t));
  }
  lt(e, t) {
    return this.setLimit("max", e, !1, u.toString(t));
  }
  setLimit(e, t, s, a) {
    return new P({
      ...this._def,
      checks: [
        ...this._def.checks,
        {
          kind: e,
          value: t,
          inclusive: s,
          message: u.toString(a)
        }
      ]
    });
  }
  _addCheck(e) {
    return new P({
      ...this._def,
      checks: [...this._def.checks, e]
    });
  }
  positive(e) {
    return this._addCheck({
      kind: "min",
      value: BigInt(0),
      inclusive: !1,
      message: u.toString(e)
    });
  }
  negative(e) {
    return this._addCheck({
      kind: "max",
      value: BigInt(0),
      inclusive: !1,
      message: u.toString(e)
    });
  }
  nonpositive(e) {
    return this._addCheck({
      kind: "max",
      value: BigInt(0),
      inclusive: !0,
      message: u.toString(e)
    });
  }
  nonnegative(e) {
    return this._addCheck({
      kind: "min",
      value: BigInt(0),
      inclusive: !0,
      message: u.toString(e)
    });
  }
  multipleOf(e, t) {
    return this._addCheck({
      kind: "multipleOf",
      value: e,
      message: u.toString(t)
    });
  }
  get minValue() {
    let e = null;
    for (const t of this._def.checks)
      t.kind === "min" && (e === null || t.value > e) && (e = t.value);
    return e;
  }
  get maxValue() {
    let e = null;
    for (const t of this._def.checks)
      t.kind === "max" && (e === null || t.value < e) && (e = t.value);
    return e;
  }
}
P.create = (n) => new P({
  checks: [],
  typeName: f.ZodBigInt,
  coerce: n?.coerce ?? !1,
  ..._(n)
});
class Y extends g {
  _parse(e) {
    if (this._def.coerce && (e.data = !!e.data), this._getType(e) !== h.boolean) {
      const s = this._getOrReturnCtx(e);
      return o(s, {
        code: c.invalid_type,
        expected: h.boolean,
        received: s.parsedType
      }), p;
    }
    return k(e.data);
  }
}
Y.create = (n) => new Y({
  typeName: f.ZodBoolean,
  coerce: n?.coerce || !1,
  ..._(n)
});
class $ extends g {
  _parse(e) {
    if (this._def.coerce && (e.data = new Date(e.data)), this._getType(e) !== h.date) {
      const r = this._getOrReturnCtx(e);
      return o(r, {
        code: c.invalid_type,
        expected: h.date,
        received: r.parsedType
      }), p;
    }
    if (Number.isNaN(e.data.getTime())) {
      const r = this._getOrReturnCtx(e);
      return o(r, {
        code: c.invalid_date
      }), p;
    }
    const s = new w();
    let a;
    for (const r of this._def.checks)
      r.kind === "min" ? e.data.getTime() < r.value && (a = this._getOrReturnCtx(e, a), o(a, {
        code: c.too_small,
        message: r.message,
        inclusive: !0,
        exact: !1,
        minimum: r.value,
        type: "date"
      }), s.dirty()) : r.kind === "max" ? e.data.getTime() > r.value && (a = this._getOrReturnCtx(e, a), o(a, {
        code: c.too_big,
        message: r.message,
        inclusive: !0,
        exact: !1,
        maximum: r.value,
        type: "date"
      }), s.dirty()) : v.assertNever(r);
    return {
      status: s.value,
      value: new Date(e.data.getTime())
    };
  }
  _addCheck(e) {
    return new $({
      ...this._def,
      checks: [...this._def.checks, e]
    });
  }
  min(e, t) {
    return this._addCheck({
      kind: "min",
      value: e.getTime(),
      message: u.toString(t)
    });
  }
  max(e, t) {
    return this._addCheck({
      kind: "max",
      value: e.getTime(),
      message: u.toString(t)
    });
  }
  get minDate() {
    let e = null;
    for (const t of this._def.checks)
      t.kind === "min" && (e === null || t.value > e) && (e = t.value);
    return e != null ? new Date(e) : null;
  }
  get maxDate() {
    let e = null;
    for (const t of this._def.checks)
      t.kind === "max" && (e === null || t.value < e) && (e = t.value);
    return e != null ? new Date(e) : null;
  }
}
$.create = (n) => new $({
  checks: [],
  coerce: n?.coerce || !1,
  typeName: f.ZodDate,
  ..._(n)
});
class ne extends g {
  _parse(e) {
    if (this._getType(e) !== h.symbol) {
      const s = this._getOrReturnCtx(e);
      return o(s, {
        code: c.invalid_type,
        expected: h.symbol,
        received: s.parsedType
      }), p;
    }
    return k(e.data);
  }
}
ne.create = (n) => new ne({
  typeName: f.ZodSymbol,
  ..._(n)
});
class ie extends g {
  _parse(e) {
    if (this._getType(e) !== h.undefined) {
      const s = this._getOrReturnCtx(e);
      return o(s, {
        code: c.invalid_type,
        expected: h.undefined,
        received: s.parsedType
      }), p;
    }
    return k(e.data);
  }
}
ie.create = (n) => new ie({
  typeName: f.ZodUndefined,
  ..._(n)
});
class de extends g {
  _parse(e) {
    if (this._getType(e) !== h.null) {
      const s = this._getOrReturnCtx(e);
      return o(s, {
        code: c.invalid_type,
        expected: h.null,
        received: s.parsedType
      }), p;
    }
    return k(e.data);
  }
}
de.create = (n) => new de({
  typeName: f.ZodNull,
  ..._(n)
});
class q extends g {
  constructor() {
    super(...arguments), this._any = !0;
  }
  _parse(e) {
    return k(e.data);
  }
}
q.create = (n) => new q({
  typeName: f.ZodAny,
  ..._(n)
});
class H extends g {
  constructor() {
    super(...arguments), this._unknown = !0;
  }
  _parse(e) {
    return k(e.data);
  }
}
H.create = (n) => new H({
  typeName: f.ZodUnknown,
  ..._(n)
});
class S extends g {
  _parse(e) {
    const t = this._getOrReturnCtx(e);
    return o(t, {
      code: c.invalid_type,
      expected: h.never,
      received: t.parsedType
    }), p;
  }
}
S.create = (n) => new S({
  typeName: f.ZodNever,
  ..._(n)
});
class ce extends g {
  _parse(e) {
    if (this._getType(e) !== h.undefined) {
      const s = this._getOrReturnCtx(e);
      return o(s, {
        code: c.invalid_type,
        expected: h.void,
        received: s.parsedType
      }), p;
    }
    return k(e.data);
  }
}
ce.create = (n) => new ce({
  typeName: f.ZodVoid,
  ..._(n)
});
class b extends g {
  _parse(e) {
    const { ctx: t, status: s } = this._processInputParams(e), a = this._def;
    if (t.parsedType !== h.array)
      return o(t, {
        code: c.invalid_type,
        expected: h.array,
        received: t.parsedType
      }), p;
    if (a.exactLength !== null) {
      const d = t.data.length > a.exactLength.value, i = t.data.length < a.exactLength.value;
      (d || i) && (o(t, {
        code: d ? c.too_big : c.too_small,
        minimum: i ? a.exactLength.value : void 0,
        maximum: d ? a.exactLength.value : void 0,
        type: "array",
        inclusive: !0,
        exact: !0,
        message: a.exactLength.message
      }), s.dirty());
    }
    if (a.minLength !== null && t.data.length < a.minLength.value && (o(t, {
      code: c.too_small,
      minimum: a.minLength.value,
      type: "array",
      inclusive: !0,
      exact: !1,
      message: a.minLength.message
    }), s.dirty()), a.maxLength !== null && t.data.length > a.maxLength.value && (o(t, {
      code: c.too_big,
      maximum: a.maxLength.value,
      type: "array",
      inclusive: !0,
      exact: !1,
      message: a.maxLength.message
    }), s.dirty()), t.common.async)
      return Promise.all([...t.data].map((d, i) => a.type._parseAsync(new A(t, d, t.path, i)))).then((d) => w.mergeArray(s, d));
    const r = [...t.data].map((d, i) => a.type._parseSync(new A(t, d, t.path, i)));
    return w.mergeArray(s, r);
  }
  get element() {
    return this._def.type;
  }
  min(e, t) {
    return new b({
      ...this._def,
      minLength: { value: e, message: u.toString(t) }
    });
  }
  max(e, t) {
    return new b({
      ...this._def,
      maxLength: { value: e, message: u.toString(t) }
    });
  }
  length(e, t) {
    return new b({
      ...this._def,
      exactLength: { value: e, message: u.toString(t) }
    });
  }
  nonempty(e) {
    return this.min(1, e);
  }
}
b.create = (n, e) => new b({
  type: n,
  minLength: null,
  maxLength: null,
  exactLength: null,
  typeName: f.ZodArray,
  ..._(e)
});
function R(n) {
  if (n instanceof x) {
    const e = {};
    for (const t in n.shape) {
      const s = n.shape[t];
      e[t] = Z.create(R(s));
    }
    return new x({
      ...n._def,
      shape: () => e
    });
  } else return n instanceof b ? new b({
    ...n._def,
    type: R(n.element)
  }) : n instanceof Z ? Z.create(R(n.unwrap())) : n instanceof L ? L.create(R(n.unwrap())) : n instanceof O ? O.create(n.items.map((e) => R(e))) : n;
}
class x extends g {
  constructor() {
    super(...arguments), this._cached = null, this.nonstrict = this.passthrough, this.augment = this.extend;
  }
  _getCached() {
    if (this._cached !== null)
      return this._cached;
    const e = this._def.shape(), t = v.objectKeys(e);
    return this._cached = { shape: e, keys: t }, this._cached;
  }
  _parse(e) {
    if (this._getType(e) !== h.object) {
      const m = this._getOrReturnCtx(e);
      return o(m, {
        code: c.invalid_type,
        expected: h.object,
        received: m.parsedType
      }), p;
    }
    const { status: s, ctx: a } = this._processInputParams(e), { shape: r, keys: d } = this._getCached(), i = [];
    if (!(this._def.catchall instanceof S && this._def.unknownKeys === "strip"))
      for (const m in a.data)
        d.includes(m) || i.push(m);
    const l = [];
    for (const m of d) {
      const y = r[m], V = a.data[m];
      l.push({
        key: { status: "valid", value: m },
        value: y._parse(new A(a, V, a.path, m)),
        alwaysSet: m in a.data
      });
    }
    if (this._def.catchall instanceof S) {
      const m = this._def.unknownKeys;
      if (m === "passthrough")
        for (const y of i)
          l.push({
            key: { status: "valid", value: y },
            value: { status: "valid", value: a.data[y] }
          });
      else if (m === "strict")
        i.length > 0 && (o(a, {
          code: c.unrecognized_keys,
          keys: i
        }), s.dirty());
      else if (m !== "strip") throw new Error("Internal ZodObject error: invalid unknownKeys value.");
    } else {
      const m = this._def.catchall;
      for (const y of i) {
        const V = a.data[y];
        l.push({
          key: { status: "valid", value: y },
          value: m._parse(
            new A(a, V, a.path, y)
            //, ctx.child(key), value, getParsedType(value)
          ),
          alwaysSet: y in a.data
        });
      }
    }
    return a.common.async ? Promise.resolve().then(async () => {
      const m = [];
      for (const y of l) {
        const V = await y.key, _e = await y.value;
        m.push({
          key: V,
          value: _e,
          alwaysSet: y.alwaysSet
        });
      }
      return m;
    }).then((m) => w.mergeObjectSync(s, m)) : w.mergeObjectSync(s, l);
  }
  get shape() {
    return this._def.shape();
  }
  strict(e) {
    return u.errToObj, new x({
      ...this._def,
      unknownKeys: "strict",
      ...e !== void 0 ? {
        errorMap: (t, s) => {
          const a = this._def.errorMap?.(t, s).message ?? s.defaultError;
          return t.code === "unrecognized_keys" ? {
            message: u.errToObj(e).message ?? a
          } : {
            message: a
          };
        }
      } : {}
    });
  }
  strip() {
    return new x({
      ...this._def,
      unknownKeys: "strip"
    });
  }
  passthrough() {
    return new x({
      ...this._def,
      unknownKeys: "passthrough"
    });
  }
  // const AugmentFactory =
  //   <Def extends ZodObjectDef>(def: Def) =>
  //   <Augmentation extends ZodRawShape>(
  //     augmentation: Augmentation
  //   ): ZodObject<
  //     extendShape<ReturnType<Def["shape"]>, Augmentation>,
  //     Def["unknownKeys"],
  //     Def["catchall"]
  //   > => {
  //     return new ZodObject({
  //       ...def,
  //       shape: () => ({
  //         ...def.shape(),
  //         ...augmentation,
  //       }),
  //     }) as any;
  //   };
  extend(e) {
    return new x({
      ...this._def,
      shape: () => ({
        ...this._def.shape(),
        ...e
      })
    });
  }
  /**
   * Prior to zod@1.0.12 there was a bug in the
   * inferred type of merged objects. Please
   * upgrade if you are experiencing issues.
   */
  merge(e) {
    return new x({
      unknownKeys: e._def.unknownKeys,
      catchall: e._def.catchall,
      shape: () => ({
        ...this._def.shape(),
        ...e._def.shape()
      }),
      typeName: f.ZodObject
    });
  }
  // merge<
  //   Incoming extends AnyZodObject,
  //   Augmentation extends Incoming["shape"],
  //   NewOutput extends {
  //     [k in keyof Augmentation | keyof Output]: k extends keyof Augmentation
  //       ? Augmentation[k]["_output"]
  //       : k extends keyof Output
  //       ? Output[k]
  //       : never;
  //   },
  //   NewInput extends {
  //     [k in keyof Augmentation | keyof Input]: k extends keyof Augmentation
  //       ? Augmentation[k]["_input"]
  //       : k extends keyof Input
  //       ? Input[k]
  //       : never;
  //   }
  // >(
  //   merging: Incoming
  // ): ZodObject<
  //   extendShape<T, ReturnType<Incoming["_def"]["shape"]>>,
  //   Incoming["_def"]["unknownKeys"],
  //   Incoming["_def"]["catchall"],
  //   NewOutput,
  //   NewInput
  // > {
  //   const merged: any = new ZodObject({
  //     unknownKeys: merging._def.unknownKeys,
  //     catchall: merging._def.catchall,
  //     shape: () =>
  //       objectUtil.mergeShapes(this._def.shape(), merging._def.shape()),
  //     typeName: ZodFirstPartyTypeKind.ZodObject,
  //   }) as any;
  //   return merged;
  // }
  setKey(e, t) {
    return this.augment({ [e]: t });
  }
  // merge<Incoming extends AnyZodObject>(
  //   merging: Incoming
  // ): //ZodObject<T & Incoming["_shape"], UnknownKeys, Catchall> = (merging) => {
  // ZodObject<
  //   extendShape<T, ReturnType<Incoming["_def"]["shape"]>>,
  //   Incoming["_def"]["unknownKeys"],
  //   Incoming["_def"]["catchall"]
  // > {
  //   // const mergedShape = objectUtil.mergeShapes(
  //   //   this._def.shape(),
  //   //   merging._def.shape()
  //   // );
  //   const merged: any = new ZodObject({
  //     unknownKeys: merging._def.unknownKeys,
  //     catchall: merging._def.catchall,
  //     shape: () =>
  //       objectUtil.mergeShapes(this._def.shape(), merging._def.shape()),
  //     typeName: ZodFirstPartyTypeKind.ZodObject,
  //   }) as any;
  //   return merged;
  // }
  catchall(e) {
    return new x({
      ...this._def,
      catchall: e
    });
  }
  pick(e) {
    const t = {};
    for (const s of v.objectKeys(e))
      e[s] && this.shape[s] && (t[s] = this.shape[s]);
    return new x({
      ...this._def,
      shape: () => t
    });
  }
  omit(e) {
    const t = {};
    for (const s of v.objectKeys(this.shape))
      e[s] || (t[s] = this.shape[s]);
    return new x({
      ...this._def,
      shape: () => t
    });
  }
  /**
   * @deprecated
   */
  deepPartial() {
    return R(this);
  }
  partial(e) {
    const t = {};
    for (const s of v.objectKeys(this.shape)) {
      const a = this.shape[s];
      e && !e[s] ? t[s] = a : t[s] = a.optional();
    }
    return new x({
      ...this._def,
      shape: () => t
    });
  }
  required(e) {
    const t = {};
    for (const s of v.objectKeys(this.shape))
      if (e && !e[s])
        t[s] = this.shape[s];
      else {
        let r = this.shape[s];
        for (; r instanceof Z; )
          r = r._def.innerType;
        t[s] = r;
      }
    return new x({
      ...this._def,
      shape: () => t
    });
  }
  keyof() {
    return pe(v.objectKeys(this.shape));
  }
}
x.create = (n, e) => new x({
  shape: () => n,
  unknownKeys: "strip",
  catchall: S.create(),
  typeName: f.ZodObject,
  ..._(e)
});
x.strictCreate = (n, e) => new x({
  shape: () => n,
  unknownKeys: "strict",
  catchall: S.create(),
  typeName: f.ZodObject,
  ..._(e)
});
x.lazycreate = (n, e) => new x({
  shape: n,
  unknownKeys: "strip",
  catchall: S.create(),
  typeName: f.ZodObject,
  ..._(e)
});
class B extends g {
  _parse(e) {
    const { ctx: t } = this._processInputParams(e), s = this._def.options;
    function a(r) {
      for (const i of r)
        if (i.result.status === "valid")
          return i.result;
      for (const i of r)
        if (i.result.status === "dirty")
          return t.common.issues.push(...i.ctx.common.issues), i.result;
      const d = r.map((i) => new M(i.ctx.common.issues));
      return o(t, {
        code: c.invalid_union,
        unionErrors: d
      }), p;
    }
    if (t.common.async)
      return Promise.all(s.map(async (r) => {
        const d = {
          ...t,
          common: {
            ...t.common,
            issues: []
          },
          parent: null
        };
        return {
          result: await r._parseAsync({
            data: t.data,
            path: t.path,
            parent: d
          }),
          ctx: d
        };
      })).then(a);
    {
      let r;
      const d = [];
      for (const l of s) {
        const m = {
          ...t,
          common: {
            ...t.common,
            issues: []
          },
          parent: null
        }, y = l._parseSync({
          data: t.data,
          path: t.path,
          parent: m
        });
        if (y.status === "valid")
          return y;
        y.status === "dirty" && !r && (r = { result: y, ctx: m }), m.common.issues.length && d.push(m.common.issues);
      }
      if (r)
        return t.common.issues.push(...r.ctx.common.issues), r.result;
      const i = d.map((l) => new M(l));
      return o(t, {
        code: c.invalid_union,
        unionErrors: i
      }), p;
    }
  }
  get options() {
    return this._def.options;
  }
}
B.create = (n, e) => new B({
  options: n,
  typeName: f.ZodUnion,
  ..._(e)
});
function G(n, e) {
  const t = C(n), s = C(e);
  if (n === e)
    return { valid: !0, data: n };
  if (t === h.object && s === h.object) {
    const a = v.objectKeys(e), r = v.objectKeys(n).filter((i) => a.indexOf(i) !== -1), d = { ...n, ...e };
    for (const i of r) {
      const l = G(n[i], e[i]);
      if (!l.valid)
        return { valid: !1 };
      d[i] = l.data;
    }
    return { valid: !0, data: d };
  } else if (t === h.array && s === h.array) {
    if (n.length !== e.length)
      return { valid: !1 };
    const a = [];
    for (let r = 0; r < n.length; r++) {
      const d = n[r], i = e[r], l = G(d, i);
      if (!l.valid)
        return { valid: !1 };
      a.push(l.data);
    }
    return { valid: !0, data: a };
  } else return t === h.date && s === h.date && +n == +e ? { valid: !0, data: n } : { valid: !1 };
}
class F extends g {
  _parse(e) {
    const { status: t, ctx: s } = this._processInputParams(e), a = (r, d) => {
      if (se(r) || se(d))
        return p;
      const i = G(r.value, d.value);
      return i.valid ? ((ae(r) || ae(d)) && t.dirty(), { status: t.value, value: i.data }) : (o(s, {
        code: c.invalid_intersection_types
      }), p);
    };
    return s.common.async ? Promise.all([
      this._def.left._parseAsync({
        data: s.data,
        path: s.path,
        parent: s
      }),
      this._def.right._parseAsync({
        data: s.data,
        path: s.path,
        parent: s
      })
    ]).then(([r, d]) => a(r, d)) : a(this._def.left._parseSync({
      data: s.data,
      path: s.path,
      parent: s
    }), this._def.right._parseSync({
      data: s.data,
      path: s.path,
      parent: s
    }));
  }
}
F.create = (n, e, t) => new F({
  left: n,
  right: e,
  typeName: f.ZodIntersection,
  ..._(t)
});
class O extends g {
  _parse(e) {
    const { status: t, ctx: s } = this._processInputParams(e);
    if (s.parsedType !== h.array)
      return o(s, {
        code: c.invalid_type,
        expected: h.array,
        received: s.parsedType
      }), p;
    if (s.data.length < this._def.items.length)
      return o(s, {
        code: c.too_small,
        minimum: this._def.items.length,
        inclusive: !0,
        exact: !1,
        type: "array"
      }), p;
    !this._def.rest && s.data.length > this._def.items.length && (o(s, {
      code: c.too_big,
      maximum: this._def.items.length,
      inclusive: !0,
      exact: !1,
      type: "array"
    }), t.dirty());
    const r = [...s.data].map((d, i) => {
      const l = this._def.items[i] || this._def.rest;
      return l ? l._parse(new A(s, d, s.path, i)) : null;
    }).filter((d) => !!d);
    return s.common.async ? Promise.all(r).then((d) => w.mergeArray(t, d)) : w.mergeArray(t, r);
  }
  get items() {
    return this._def.items;
  }
  rest(e) {
    return new O({
      ...this._def,
      rest: e
    });
  }
}
O.create = (n, e) => {
  if (!Array.isArray(n))
    throw new Error("You must pass an array of schemas to z.tuple([ ... ])");
  return new O({
    items: n,
    typeName: f.ZodTuple,
    rest: null,
    ..._(e)
  });
};
class oe extends g {
  get keySchema() {
    return this._def.keyType;
  }
  get valueSchema() {
    return this._def.valueType;
  }
  _parse(e) {
    const { status: t, ctx: s } = this._processInputParams(e);
    if (s.parsedType !== h.map)
      return o(s, {
        code: c.invalid_type,
        expected: h.map,
        received: s.parsedType
      }), p;
    const a = this._def.keyType, r = this._def.valueType, d = [...s.data.entries()].map(([i, l], m) => ({
      key: a._parse(new A(s, i, s.path, [m, "key"])),
      value: r._parse(new A(s, l, s.path, [m, "value"]))
    }));
    if (s.common.async) {
      const i = /* @__PURE__ */ new Map();
      return Promise.resolve().then(async () => {
        for (const l of d) {
          const m = await l.key, y = await l.value;
          if (m.status === "aborted" || y.status === "aborted")
            return p;
          (m.status === "dirty" || y.status === "dirty") && t.dirty(), i.set(m.value, y.value);
        }
        return { status: t.value, value: i };
      });
    } else {
      const i = /* @__PURE__ */ new Map();
      for (const l of d) {
        const m = l.key, y = l.value;
        if (m.status === "aborted" || y.status === "aborted")
          return p;
        (m.status === "dirty" || y.status === "dirty") && t.dirty(), i.set(m.value, y.value);
      }
      return { status: t.value, value: i };
    }
  }
}
oe.create = (n, e, t) => new oe({
  valueType: e,
  keyType: n,
  typeName: f.ZodMap,
  ..._(t)
});
class D extends g {
  _parse(e) {
    const { status: t, ctx: s } = this._processInputParams(e);
    if (s.parsedType !== h.set)
      return o(s, {
        code: c.invalid_type,
        expected: h.set,
        received: s.parsedType
      }), p;
    const a = this._def;
    a.minSize !== null && s.data.size < a.minSize.value && (o(s, {
      code: c.too_small,
      minimum: a.minSize.value,
      type: "set",
      inclusive: !0,
      exact: !1,
      message: a.minSize.message
    }), t.dirty()), a.maxSize !== null && s.data.size > a.maxSize.value && (o(s, {
      code: c.too_big,
      maximum: a.maxSize.value,
      type: "set",
      inclusive: !0,
      exact: !1,
      message: a.maxSize.message
    }), t.dirty());
    const r = this._def.valueType;
    function d(l) {
      const m = /* @__PURE__ */ new Set();
      for (const y of l) {
        if (y.status === "aborted")
          return p;
        y.status === "dirty" && t.dirty(), m.add(y.value);
      }
      return { status: t.value, value: m };
    }
    const i = [...s.data.values()].map((l, m) => r._parse(new A(s, l, s.path, m)));
    return s.common.async ? Promise.all(i).then((l) => d(l)) : d(i);
  }
  min(e, t) {
    return new D({
      ...this._def,
      minSize: { value: e, message: u.toString(t) }
    });
  }
  max(e, t) {
    return new D({
      ...this._def,
      maxSize: { value: e, message: u.toString(t) }
    });
  }
  size(e, t) {
    return this.min(e, t).max(e, t);
  }
  nonempty(e) {
    return this.min(1, e);
  }
}
D.create = (n, e) => new D({
  valueType: n,
  minSize: null,
  maxSize: null,
  typeName: f.ZodSet,
  ..._(e)
});
class ue extends g {
  get schema() {
    return this._def.getter();
  }
  _parse(e) {
    const { ctx: t } = this._processInputParams(e);
    return this._def.getter()._parse({ data: t.data, path: t.path, parent: t });
  }
}
ue.create = (n, e) => new ue({
  getter: n,
  typeName: f.ZodLazy,
  ..._(e)
});
class Q extends g {
  _parse(e) {
    if (e.data !== this._def.value) {
      const t = this._getOrReturnCtx(e);
      return o(t, {
        received: t.data,
        code: c.invalid_literal,
        expected: this._def.value
      }), p;
    }
    return { status: "valid", value: e.data };
  }
  get value() {
    return this._def.value;
  }
}
Q.create = (n, e) => new Q({
  value: n,
  typeName: f.ZodLiteral,
  ..._(e)
});
function pe(n, e) {
  return new j({
    values: n,
    typeName: f.ZodEnum,
    ..._(e)
  });
}
class j extends g {
  _parse(e) {
    if (typeof e.data != "string") {
      const t = this._getOrReturnCtx(e), s = this._def.values;
      return o(t, {
        expected: v.joinValues(s),
        received: t.parsedType,
        code: c.invalid_type
      }), p;
    }
    if (this._cache || (this._cache = new Set(this._def.values)), !this._cache.has(e.data)) {
      const t = this._getOrReturnCtx(e), s = this._def.values;
      return o(t, {
        received: t.data,
        code: c.invalid_enum_value,
        options: s
      }), p;
    }
    return k(e.data);
  }
  get options() {
    return this._def.values;
  }
  get enum() {
    const e = {};
    for (const t of this._def.values)
      e[t] = t;
    return e;
  }
  get Values() {
    const e = {};
    for (const t of this._def.values)
      e[t] = t;
    return e;
  }
  get Enum() {
    const e = {};
    for (const t of this._def.values)
      e[t] = t;
    return e;
  }
  extract(e, t = this._def) {
    return j.create(e, {
      ...this._def,
      ...t
    });
  }
  exclude(e, t = this._def) {
    return j.create(this.options.filter((s) => !e.includes(s)), {
      ...this._def,
      ...t
    });
  }
}
j.create = pe;
class le extends g {
  _parse(e) {
    const t = v.getValidEnumValues(this._def.values), s = this._getOrReturnCtx(e);
    if (s.parsedType !== h.string && s.parsedType !== h.number) {
      const a = v.objectValues(t);
      return o(s, {
        expected: v.joinValues(a),
        received: s.parsedType,
        code: c.invalid_type
      }), p;
    }
    if (this._cache || (this._cache = new Set(v.getValidEnumValues(this._def.values))), !this._cache.has(e.data)) {
      const a = v.objectValues(t);
      return o(s, {
        received: s.data,
        code: c.invalid_enum_value,
        options: a
      }), p;
    }
    return k(e.data);
  }
  get enum() {
    return this._def.values;
  }
}
le.create = (n, e) => new le({
  values: n,
  typeName: f.ZodNativeEnum,
  ..._(e)
});
class W extends g {
  unwrap() {
    return this._def.type;
  }
  _parse(e) {
    const { ctx: t } = this._processInputParams(e);
    if (t.parsedType !== h.promise && t.common.async === !1)
      return o(t, {
        code: c.invalid_type,
        expected: h.promise,
        received: t.parsedType
      }), p;
    const s = t.parsedType === h.promise ? t.data : Promise.resolve(t.data);
    return k(s.then((a) => this._def.type.parseAsync(a, {
      path: t.path,
      errorMap: t.common.contextualErrorMap
    })));
  }
}
W.create = (n, e) => new W({
  type: n,
  typeName: f.ZodPromise,
  ..._(e)
});
class E extends g {
  innerType() {
    return this._def.schema;
  }
  sourceType() {
    return this._def.schema._def.typeName === f.ZodEffects ? this._def.schema.sourceType() : this._def.schema;
  }
  _parse(e) {
    const { status: t, ctx: s } = this._processInputParams(e), a = this._def.effect || null, r = {
      addIssue: (d) => {
        o(s, d), d.fatal ? t.abort() : t.dirty();
      },
      get path() {
        return s.path;
      }
    };
    if (r.addIssue = r.addIssue.bind(r), a.type === "preprocess") {
      const d = a.transform(s.data, r);
      if (s.common.async)
        return Promise.resolve(d).then(async (i) => {
          if (t.value === "aborted")
            return p;
          const l = await this._def.schema._parseAsync({
            data: i,
            path: s.path,
            parent: s
          });
          return l.status === "aborted" ? p : l.status === "dirty" ? z(l.value) : t.value === "dirty" ? z(l.value) : l;
        });
      {
        if (t.value === "aborted")
          return p;
        const i = this._def.schema._parseSync({
          data: d,
          path: s.path,
          parent: s
        });
        return i.status === "aborted" ? p : i.status === "dirty" ? z(i.value) : t.value === "dirty" ? z(i.value) : i;
      }
    }
    if (a.type === "refinement") {
      const d = (i) => {
        const l = a.refinement(i, r);
        if (s.common.async)
          return Promise.resolve(l);
        if (l instanceof Promise)
          throw new Error("Async refinement encountered during synchronous parse operation. Use .parseAsync instead.");
        return i;
      };
      if (s.common.async === !1) {
        const i = this._def.schema._parseSync({
          data: s.data,
          path: s.path,
          parent: s
        });
        return i.status === "aborted" ? p : (i.status === "dirty" && t.dirty(), d(i.value), { status: t.value, value: i.value });
      } else
        return this._def.schema._parseAsync({ data: s.data, path: s.path, parent: s }).then((i) => i.status === "aborted" ? p : (i.status === "dirty" && t.dirty(), d(i.value).then(() => ({ status: t.value, value: i.value }))));
    }
    if (a.type === "transform")
      if (s.common.async === !1) {
        const d = this._def.schema._parseSync({
          data: s.data,
          path: s.path,
          parent: s
        });
        if (!N(d))
          return p;
        const i = a.transform(d.value, r);
        if (i instanceof Promise)
          throw new Error("Asynchronous transform encountered during synchronous parse operation. Use .parseAsync instead.");
        return { status: t.value, value: i };
      } else
        return this._def.schema._parseAsync({ data: s.data, path: s.path, parent: s }).then((d) => N(d) ? Promise.resolve(a.transform(d.value, r)).then((i) => ({
          status: t.value,
          value: i
        })) : p);
    v.assertNever(a);
  }
}
E.create = (n, e, t) => new E({
  schema: n,
  typeName: f.ZodEffects,
  effect: e,
  ..._(t)
});
E.createWithPreprocess = (n, e, t) => new E({
  schema: e,
  effect: { type: "preprocess", transform: n },
  typeName: f.ZodEffects,
  ..._(t)
});
class Z extends g {
  _parse(e) {
    return this._getType(e) === h.undefined ? k(void 0) : this._def.innerType._parse(e);
  }
  unwrap() {
    return this._def.innerType;
  }
}
Z.create = (n, e) => new Z({
  innerType: n,
  typeName: f.ZodOptional,
  ..._(e)
});
class L extends g {
  _parse(e) {
    return this._getType(e) === h.null ? k(null) : this._def.innerType._parse(e);
  }
  unwrap() {
    return this._def.innerType;
  }
}
L.create = (n, e) => new L({
  innerType: n,
  typeName: f.ZodNullable,
  ..._(e)
});
class X extends g {
  _parse(e) {
    const { ctx: t } = this._processInputParams(e);
    let s = t.data;
    return t.parsedType === h.undefined && (s = this._def.defaultValue()), this._def.innerType._parse({
      data: s,
      path: t.path,
      parent: t
    });
  }
  removeDefault() {
    return this._def.innerType;
  }
}
X.create = (n, e) => new X({
  innerType: n,
  typeName: f.ZodDefault,
  defaultValue: typeof e.default == "function" ? e.default : () => e.default,
  ..._(e)
});
class K extends g {
  _parse(e) {
    const { ctx: t } = this._processInputParams(e), s = {
      ...t,
      common: {
        ...t.common,
        issues: []
      }
    }, a = this._def.innerType._parse({
      data: s.data,
      path: s.path,
      parent: {
        ...s
      }
    });
    return U(a) ? a.then((r) => ({
      status: "valid",
      value: r.status === "valid" ? r.value : this._def.catchValue({
        get error() {
          return new M(s.common.issues);
        },
        input: s.data
      })
    })) : {
      status: "valid",
      value: a.status === "valid" ? a.value : this._def.catchValue({
        get error() {
          return new M(s.common.issues);
        },
        input: s.data
      })
    };
  }
  removeCatch() {
    return this._def.innerType;
  }
}
K.create = (n, e) => new K({
  innerType: n,
  typeName: f.ZodCatch,
  catchValue: typeof e.catch == "function" ? e.catch : () => e.catch,
  ..._(e)
});
class he extends g {
  _parse(e) {
    if (this._getType(e) !== h.nan) {
      const s = this._getOrReturnCtx(e);
      return o(s, {
        code: c.invalid_type,
        expected: h.nan,
        received: s.parsedType
      }), p;
    }
    return { status: "valid", value: e.data };
  }
}
he.create = (n) => new he({
  typeName: f.ZodNaN,
  ..._(n)
});
class Pe extends g {
  _parse(e) {
    const { ctx: t } = this._processInputParams(e), s = t.data;
    return this._def.type._parse({
      data: s,
      path: t.path,
      parent: t
    });
  }
  unwrap() {
    return this._def.type;
  }
}
class te extends g {
  _parse(e) {
    const { status: t, ctx: s } = this._processInputParams(e);
    if (s.common.async)
      return (async () => {
        const r = await this._def.in._parseAsync({
          data: s.data,
          path: s.path,
          parent: s
        });
        return r.status === "aborted" ? p : r.status === "dirty" ? (t.dirty(), z(r.value)) : this._def.out._parseAsync({
          data: r.value,
          path: s.path,
          parent: s
        });
      })();
    {
      const a = this._def.in._parseSync({
        data: s.data,
        path: s.path,
        parent: s
      });
      return a.status === "aborted" ? p : a.status === "dirty" ? (t.dirty(), {
        status: "dirty",
        value: a.value
      }) : this._def.out._parseSync({
        data: a.value,
        path: s.path,
        parent: s
      });
    }
  }
  static create(e, t) {
    return new te({
      in: e,
      out: t,
      typeName: f.ZodPipeline
    });
  }
}
class ee extends g {
  _parse(e) {
    const t = this._def.innerType._parse(e), s = (a) => (N(a) && (a.value = Object.freeze(a.value)), a);
    return U(t) ? t.then((a) => s(a)) : s(t);
  }
  unwrap() {
    return this._def.innerType;
  }
}
ee.create = (n, e) => new ee({
  innerType: n,
  typeName: f.ZodReadonly,
  ..._(e)
});
var f;
(function(n) {
  n.ZodString = "ZodString", n.ZodNumber = "ZodNumber", n.ZodNaN = "ZodNaN", n.ZodBigInt = "ZodBigInt", n.ZodBoolean = "ZodBoolean", n.ZodDate = "ZodDate", n.ZodSymbol = "ZodSymbol", n.ZodUndefined = "ZodUndefined", n.ZodNull = "ZodNull", n.ZodAny = "ZodAny", n.ZodUnknown = "ZodUnknown", n.ZodNever = "ZodNever", n.ZodVoid = "ZodVoid", n.ZodArray = "ZodArray", n.ZodObject = "ZodObject", n.ZodUnion = "ZodUnion", n.ZodDiscriminatedUnion = "ZodDiscriminatedUnion", n.ZodIntersection = "ZodIntersection", n.ZodTuple = "ZodTuple", n.ZodRecord = "ZodRecord", n.ZodMap = "ZodMap", n.ZodSet = "ZodSet", n.ZodFunction = "ZodFunction", n.ZodLazy = "ZodLazy", n.ZodLiteral = "ZodLiteral", n.ZodEnum = "ZodEnum", n.ZodEffects = "ZodEffects", n.ZodNativeEnum = "ZodNativeEnum", n.ZodOptional = "ZodOptional", n.ZodNullable = "ZodNullable", n.ZodDefault = "ZodDefault", n.ZodCatch = "ZodCatch", n.ZodPromise = "ZodPromise", n.ZodBranded = "ZodBranded", n.ZodPipeline = "ZodPipeline", n.ZodReadonly = "ZodReadonly";
})(f || (f = {}));
const Fe = T.create, We = I.create, Je = Y.create, Ye = $.create, qe = q.create, He = H.create;
S.create;
const Ge = b.create, Qe = x.create;
B.create;
F.create;
O.create;
const Xe = Q.create, Ke = j.create;
W.create;
Z.create;
L.create;
export {
  g as Schema,
  q as ZodAny,
  b as ZodArray,
  P as ZodBigInt,
  Y as ZodBoolean,
  Pe as ZodBranded,
  K as ZodCatch,
  $ as ZodDate,
  X as ZodDefault,
  E as ZodEffects,
  j as ZodEnum,
  f as ZodFirstPartyTypeKind,
  F as ZodIntersection,
  ue as ZodLazy,
  Q as ZodLiteral,
  oe as ZodMap,
  he as ZodNaN,
  le as ZodNativeEnum,
  S as ZodNever,
  de as ZodNull,
  L as ZodNullable,
  I as ZodNumber,
  x as ZodObject,
  Z as ZodOptional,
  te as ZodPipeline,
  W as ZodPromise,
  ee as ZodReadonly,
  g as ZodSchema,
  D as ZodSet,
  T as ZodString,
  ne as ZodSymbol,
  E as ZodTransformer,
  O as ZodTuple,
  g as ZodType,
  ie as ZodUndefined,
  B as ZodUnion,
  H as ZodUnknown,
  ce as ZodVoid,
  qe as any,
  Ge as array,
  Je as boolean,
  Ye as date,
  Ee as datetimeRegex,
  Ke as enum,
  Xe as literal,
  We as number,
  Qe as object,
  Fe as string,
  He as unknown
};
