// zod@3.25.76/v3/ZodError.js
import { util as h } from "./helpers/util.js";
const f = h.arrayToEnum([
  "invalid_type",
  "invalid_literal",
  "custom",
  "invalid_union",
  "invalid_union_discriminator",
  "invalid_enum_value",
  "unrecognized_keys",
  "invalid_arguments",
  "invalid_return_type",
  "invalid_date",
  "invalid_string",
  "too_small",
  "too_big",
  "invalid_intersection_types",
  "not_multiple_of",
  "not_finite"
]);
class l extends Error {
  get errors() {
    return this.issues;
  }
  constructor(t) {
    super(), this.issues = [], this.addIssue = (e) => {
      this.issues = [...this.issues, e];
    }, this.addIssues = (e = []) => {
      this.issues = [...this.issues, ...e];
    };
    const s = new.target.prototype;
    Object.setPrototypeOf ? Object.setPrototypeOf(this, s) : this.__proto__ = s, this.name = "ZodError", this.issues = t;
  }
  format(t) {
    const s = t || function(o) {
      return o.message;
    }, e = { _errors: [] }, i = (o) => {
      for (const r of o.issues)
        if (r.code === "invalid_union")
          r.unionErrors.map(i);
        else if (r.code === "invalid_return_type")
          i(r.returnTypeError);
        else if (r.code === "invalid_arguments")
          i(r.argumentsError);
        else if (r.path.length === 0)
          e._errors.push(s(r));
        else {
          let n = e, u = 0;
          for (; u < r.path.length; ) {
            const a = r.path[u];
            u === r.path.length - 1 ? (n[a] = n[a] || { _errors: [] }, n[a]._errors.push(s(r))) : n[a] = n[a] || { _errors: [] }, n = n[a], u++;
          }
        }
    };
    return i(this), e;
  }
  static assert(t) {
    if (!(t instanceof l))
      throw new Error(`Not a ZodError: ${t}`);
  }
  toString() {
    return this.message;
  }
  get message() {
    return JSON.stringify(this.issues, h.jsonStringifyReplacer, 2);
  }
  get isEmpty() {
    return this.issues.length === 0;
  }
  flatten(t = (s) => s.message) {
    const s = {}, e = [];
    for (const i of this.issues)
      if (i.path.length > 0) {
        const o = i.path[0];
        s[o] = s[o] || [], s[o].push(t(i));
      } else
        e.push(t(i));
    return { formErrors: e, fieldErrors: s };
  }
  get formErrors() {
    return this.flatten();
  }
}
l.create = (c) => new l(c);
export {
  l as ZodError,
  f as ZodIssueCode
};
