// property-information@7.1.0/lib/aria.js
import { create as l } from "./util/create.js";
import { number as r, booleanish as a, spaceSeparated as i } from "./util/types.js";
const u = l({
  properties: {
    ariaActiveDescendant: null,
    ariaAtomic: a,
    ariaAutoComplete: null,
    ariaBusy: a,
    ariaChecked: a,
    ariaColCount: r,
    ariaColIndex: r,
    ariaColSpan: r,
    ariaControls: i,
    ariaCurrent: null,
    ariaDescribedBy: i,
    ariaDetails: null,
    ariaDisabled: a,
    ariaDropEffect: i,
    ariaErrorMessage: null,
    ariaExpanded: a,
    ariaFlowTo: i,
    ariaGrabbed: a,
    ariaHasPopup: null,
    ariaHidden: a,
    ariaInvalid: null,
    ariaKeyShortcuts: null,
    ariaLabel: null,
    ariaLabelledBy: i,
    ariaLevel: r,
    ariaLive: null,
    ariaModal: a,
    ariaMultiLine: a,
    ariaMultiSelectable: a,
    ariaOrientation: null,
    ariaOwns: i,
    ariaPlaceholder: null,
    ariaPosInSet: r,
    ariaPressed: a,
    ariaReadOnly: a,
    ariaRelevant: null,
    ariaRequired: a,
    ariaRoleDescription: i,
    ariaRowCount: r,
    ariaRowIndex: r,
    ariaRowSpan: r,
    ariaSelected: a,
    ariaSetSize: r,
    ariaSort: null,
    ariaValueMax: r,
    ariaValueMin: r,
    ariaValueNow: r,
    ariaValueText: null,
    role: null
  },
  transform(n, e) {
    return e === "role" ? e : "aria-" + e.slice(4).toLowerCase();
  }
});
export {
  u as aria
};
