// mdast-util-to-markdown@2.1.2/lib/handle/index.js
import { blockquote as o } from "./blockquote.js";
import { hardBreak as r } from "./break.js";
import { code as m } from "./code.js";
import { definition as i } from "./definition.js";
import { emphasis as t } from "./emphasis.js";
import { heading as e } from "./heading.js";
import { html as p } from "./html.js";
import { image as f } from "./image.js";
import { imageReference as a } from "./image-reference.js";
import { inlineCode as n } from "./inline-code.js";
import { link as l } from "./link.js";
import { linkReference as h } from "./link-reference.js";
import { list as c } from "./list.js";
import { listItem as d } from "./list-item.js";
import { paragraph as k } from "./paragraph.js";
import { root as s } from "./root.js";
import { strong as g } from "./strong.js";
import { text as b } from "./text.js";
import { thematicBreak as x } from "./thematic-break.js";
const L = {
  blockquote: o,
  break: r,
  code: m,
  definition: i,
  emphasis: t,
  hardBreak: r,
  heading: e,
  html: p,
  image: f,
  imageReference: a,
  inlineCode: n,
  link: l,
  linkReference: h,
  list: c,
  listItem: d,
  paragraph: k,
  root: s,
  strong: g,
  text: b,
  thematicBreak: x
};
export {
  L as handle
};
