// echarts@6.0.0/lib/component/graphic/GraphicView.js
import { __extends as Q } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { createHashMap as F, each as S, assert as U, indexOf as X, extend as Y, hasOwn as Z, retrieve2 as k, isFunction as P, keys as O } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import tt from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Displayable.js";
import { convertOptionIdName as D, makeInner as et } from "../../util/model.js";
import { setTooltipConfig as rt, getShapeClass as nt } from "../../util/graphic.js";
import { positionElement as at, LOCATION_PARAMS as it } from "../../util/layout.js";
import { parsePercent as W } from "../../util/number.js";
import ot from "../../view/Component.js";
import { getECData as B } from "../../util/innerStore.js";
import { isEC4CompatibleStyle as ft, convertFromEC4CompatibleStyle as vt } from "../../util/styleCompat.js";
import { applyUpdateTransition as L, updateLeaveTo as st, isTransitionAll as mt, applyLeaveTransition as ut } from "../../animation/customGraphicTransition.js";
import { updateProps as pt } from "../../animation/basicTransition.js";
import { stopPreviousKeyframeAnimationAndRestore as dt, applyKeyframeAnimation as M } from "../../animation/customGraphicKeyframeAnimation.js";
import H from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Text.js";
import lt from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Image.js";
import ht from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
var R = {
  // Reserved but not supported in graphic component.
  path: null,
  compoundPath: null,
  // Supported in graphic component.
  group: ht,
  image: lt,
  text: H
}, u = et(), Lt = (
  /** @class */
  (function(t) {
    Q(r, t);
    function r() {
      var e = t !== null && t.apply(this, arguments) || this;
      return e.type = r.type, e;
    }
    return r.prototype.init = function() {
      this._elMap = F();
    }, r.prototype.render = function(e, i, n) {
      e !== this._lastGraphicModel && this._clear(), this._lastGraphicModel = e, this._updateElements(e), this._relocate(e, n);
    }, r.prototype._updateElements = function(e) {
      var i = e.useElOptionsToUpdate();
      if (i) {
        var n = this._elMap, v = this.group, b = e.get("z"), _ = e.get("zlevel");
        S(i, function(a) {
          var y = D(a.id, null), s = y != null ? n.get(y) : null, d = D(a.parentId, null), g = d != null ? n.get(d) : v, o = a.type, m = a.style;
          o === "text" && m && a.hv && a.hv[1] && (m.textVerticalAlign = m.textBaseline = m.verticalAlign = m.align = null);
          var l = a.textContent, C = a.textConfig;
          if (m && ft(m, o, !!C, !!l)) {
            var h = vt(m, o, !0);
            !C && h.textConfig && (C = a.textConfig = h.textConfig), !l && h.textContent && (l = h.textContent);
          }
          var G = ct(a);
          process.env.NODE_ENV !== "production" && s && U(g === s.parent, "Changing parent is not supported.");
          var w = a.$action || "merge", A = w === "merge", E = w === "replace";
          if (A) {
            var c = !s, p = s;
            c ? p = j(y, g, a.type, n) : (p && (u(p).isNew = !1), dt(p)), p && (L(p, G, e, {
              isInit: c
            }), q(p, a, b, _));
          } else if (E) {
            I(s, a, n, e);
            var x = j(y, g, a.type, n);
            x && (L(x, G, e, {
              isInit: !0
            }), q(x, a, b, _));
          } else w === "remove" && (st(s, a), I(s, a, n, e));
          var f = n.get(y);
          if (f && l)
            if (A) {
              var $ = f.getTextContent();
              $ ? $.attr(l) : f.setTextContent(new H(l));
            } else E && f.setTextContent(new H(l));
          if (f) {
            var T = a.clipPath;
            if (T) {
              var V = T.type, N = void 0, c = !1;
              if (A) {
                var z = f.getClipPath();
                c = !z || u(z).type !== V, N = c ? K(V) : z;
              } else E && (c = !0, N = K(V));
              f.setClipPath(N), L(N, T, e, {
                isInit: c
              }), M(N, T.keyframeAnimation, e);
            }
            var J = u(f);
            f.setTextConfig(C), J.option = a, gt(f, e, a), rt({
              el: f,
              componentModel: e,
              itemName: f.name,
              itemTooltipOption: a.tooltip
            }), M(f, a.keyframeAnimation, e);
          }
        });
      }
    }, r.prototype._relocate = function(e, i) {
      for (var n = e.option.elements, v = this.group, b = this._elMap, _ = i.getWidth(), a = i.getHeight(), y = ["x", "y"], s = 0; s < n.length; s++) {
        var d = n[s], g = D(d.id, null), o = g != null ? b.get(g) : null;
        if (!(!o || !o.isGroup)) {
          var m = o.parent, l = m === v, C = u(o), h = u(m);
          C.width = W(C.option.width, l ? _ : h.width) || 0, C.height = W(C.option.height, l ? a : h.height) || 0;
        }
      }
      for (var s = n.length - 1; s >= 0; s--) {
        var d = n[s], g = D(d.id, null), o = g != null ? b.get(g) : null;
        if (o) {
          var m = o.parent, h = u(m), G = m === v ? {
            width: _,
            height: a
          } : {
            width: h.width,
            height: h.height
          }, w = {}, A = at(o, d, G, null, {
            hv: d.hv,
            boundingMode: d.bounding
          }, w);
          if (!u(o).isNew && A) {
            for (var E = d.transition, c = {}, p = 0; p < y.length; p++) {
              var x = y[p], f = w[x];
              E && (mt(E) || X(E, x) >= 0) ? c[x] = f : o[x] = f;
            }
            pt(o, c, e, 0);
          } else
            o.attr(w);
        }
      }
    }, r.prototype._clear = function() {
      var e = this, i = this._elMap;
      i.each(function(n) {
        I(n, u(n).option, i, e._lastGraphicModel);
      }), this._elMap = F();
    }, r.prototype.dispose = function() {
      this._clear();
    }, r.type = "graphic", r;
  })(ot)
);
function K(t) {
  process.env.NODE_ENV !== "production" && U(t, "graphic type MUST be set");
  var r = Z(R, t) ? R[t] : nt(t);
  process.env.NODE_ENV !== "production" && U(r, "graphic type " + t + " can not be found");
  var e = new r({});
  return u(e).type = t, e;
}
function j(t, r, e, i) {
  var n = K(e);
  return r.add(n), i.set(t, n), u(n).id = t, u(n).isNew = !0, n;
}
function I(t, r, e, i) {
  var n = t && t.parent;
  n && (t.type === "group" && t.traverse(function(v) {
    I(v, r, e, i);
  }), ut(t, r, i), e.removeKey(u(t).id));
}
function q(t, r, e, i) {
  t.isGroup || S([
    ["cursor", tt.prototype.cursor],
    // We should not support configure z and zlevel in the element level.
    // But seems we didn't limit it previously. So here still use it to avoid breaking.
    ["zlevel", i || 0],
    ["z", e || 0],
    // z2 must not be null/undefined, otherwise sort error may occur.
    ["z2", 0]
  ], function(n) {
    var v = n[0];
    Z(r, v) ? t[v] = k(r[v], n[1]) : t[v] == null && (t[v] = n[1]);
  }), S(O(r), function(n) {
    if (n.indexOf("on") === 0) {
      var v = r[n];
      t[n] = P(v) ? v : null;
    }
  }), Z(r, "draggable") && (t.draggable = r.draggable), r.name != null && (t.name = r.name), r.id != null && (t.id = r.id);
}
function ct(t) {
  return t = Y({}, t), S(["id", "parentId", "$action", "hv", "bounding", "textContent", "clipPath"].concat(it), function(r) {
    delete t[r];
  }), t;
}
function gt(t, r, e) {
  var i = B(t).eventData;
  !t.silent && !t.ignore && !i && (i = B(t).eventData = {
    componentType: "graphic",
    componentIndex: r.componentIndex,
    name: t.name
  }), i && (i.info = e.info);
}
export {
  Lt as GraphicComponentView,
  u as inner
};
