// echarts@6.0.0/lib/component/graphic/GraphicModel.js
import { __extends as T } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { each as p, assert as m, isObject as _, filter as N, extend as g, merge as O } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { mappingToExists as S } from "../../util/model.js";
import A from "../../model/Component.js";
import { mergeLayoutParam as C, copyLayoutParams as R } from "../../util/layout.js";
function U(n, e) {
  var t = n.existing;
  if (e.id = n.keyInfo.id, !e.type && t && (e.type = t.type), e.parentId == null) {
    var o = e.parentOption;
    o ? e.parentId = o.id : t && (e.parentId = t.parentId);
  }
  e.parentOption = null;
}
function c(n, e) {
  var t;
  return p(e, function(o) {
    n[o] != null && n[o] !== "auto" && (t = !0);
  }), t;
}
function I(n, e, t) {
  var o = g({}, t), i = n[e], a = t.$action || "merge";
  if (a === "merge")
    if (i) {
      if (process.env.NODE_ENV !== "production") {
        var r = t.type;
        m(!r || i.type === r, 'Please set $action: "replace" to change `type`');
      }
      O(i, o, !0), C(i, o, {
        ignoreSize: !0
      }), R(t, i), s(t, i), s(t, i, "shape"), s(t, i, "style"), s(t, i, "extra"), t.clipPath = i.clipPath;
    } else
      n[e] = o;
  else a === "replace" ? n[e] = o : a === "remove" && i && (n[e] = null);
}
var v = ["transition", "enterFrom", "leaveTo"], L = v.concat(["enterAnimation", "updateAnimation", "leaveAnimation"]);
function s(n, e, t) {
  if (t && (!n[t] && e[t] && (n[t] = {}), n = n[t], e = e[t]), !(!n || !e))
    for (var o = t ? v : L, i = 0; i < o.length; i++) {
      var a = o[i];
      n[a] == null && e[a] != null && (n[a] = e[a]);
    }
}
function M(n, e) {
  if (n && (n.hv = e.hv = [
    // Rigid body, don't care about `width`.
    c(e, ["left", "right"]),
    // Rigid body, don't care about `height`.
    c(e, ["top", "bottom"])
  ], n.type === "group")) {
    var t = n, o = e;
    t.width == null && (t.width = o.width = 0), t.height == null && (t.height = o.height = 0);
  }
}
var b = (
  /** @class */
  (function(n) {
    T(e, n);
    function e() {
      var t = n !== null && n.apply(this, arguments) || this;
      return t.type = e.type, t.preventAutoZ = !0, t;
    }
    return e.prototype.mergeOption = function(t, o) {
      var i = this.option.elements;
      this.option.elements = null, n.prototype.mergeOption.call(this, t, o), this.option.elements = i;
    }, e.prototype.optionUpdated = function(t, o) {
      var i = this.option, a = (o ? i : t).elements, r = i.elements = o ? [] : i.elements, h = [];
      this._flatten(a, h, null);
      var d = S(r, h, "normalMerge"), y = this._elOptionsToUpdate = [];
      p(d, function(l, u) {
        var f = l.newOption;
        process.env.NODE_ENV !== "production" && m(_(f) || l.existing, "Empty graphic option definition"), f && (y.push(f), U(l, f), I(r, u, f), M(r[u], f));
      }, this), i.elements = N(r, function(l) {
        return l && delete l.$action, l != null;
      });
    }, e.prototype._flatten = function(t, o, i) {
      p(t, function(a) {
        if (a) {
          i && (a.parentOption = i), o.push(a);
          var r = a.children;
          r && r.length && this._flatten(r, o, a), delete a.children;
        }
      }, this);
    }, e.prototype.useElOptionsToUpdate = function() {
      var t = this._elOptionsToUpdate;
      return this._elOptionsToUpdate = null, t;
    }, e.type = "graphic", e.defaultOption = {
      elements: []
      // parentId: null
    }, e;
  })(A)
);
export {
  b as GraphicComponentModel,
  U as setKeyInfoToNewElOption
};
