// echarts@6.0.0/lib/component/brush/BrushModel.js
import { __extends as l } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { assert as u, isArray as p, each as a, map as f, merge as d } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { replaceVisualOption as b } from "../../visual/visualSolution.js";
import c from "../../model/Model.js";
import m from "../../model/Component.js";
import i from "../../visual/tokens.js";
var x = (
  /** @class */
  (function(e) {
    l(t, e);
    function t() {
      var r = e !== null && e.apply(this, arguments) || this;
      return r.type = t.type, r.areas = [], r.brushOption = {}, r;
    }
    return t.prototype.optionUpdated = function(r, o) {
      var s = this.option;
      !o && b(s, r, ["inBrush", "outOfBrush"]);
      var n = s.inBrush = s.inBrush || {};
      s.outOfBrush = s.outOfBrush || {
        color: this.option.defaultOutOfBrushColor
      }, n.hasOwnProperty("liftZ") || (n.liftZ = 5);
    }, t.prototype.setAreas = function(r) {
      process.env.NODE_ENV !== "production" && (u(p(r)), a(r, function(o) {
        u(o.brushType, "Illegal areas");
      })), r && (this.areas = f(r, function(o) {
        return h(this.option, o);
      }, this));
    }, t.prototype.setBrushOption = function(r) {
      this.brushOption = h(this.option, r), this.brushType = this.brushOption.brushType;
    }, t.type = "brush", t.dependencies = ["geo", "grid", "xAxis", "yAxis", "parallel", "series"], t.defaultOption = {
      seriesIndex: "all",
      brushType: "rect",
      brushMode: "single",
      transformable: !0,
      brushStyle: {
        borderWidth: 1,
        color: i.color.backgroundTint,
        borderColor: i.color.borderTint
      },
      throttleType: "fixRate",
      throttleDelay: 0,
      removeOnClick: !0,
      z: 1e4,
      defaultOutOfBrushColor: i.color.disabled
    }, t;
  })(m)
);
function h(e, t) {
  return d({
    brushType: e.brushType,
    brushMode: e.brushMode,
    transformable: e.transformable,
    brushStyle: new c(e.brushStyle).getItemStyle(),
    removeOnClick: e.removeOnClick,
    z: e.z
  }, t, !0);
}
export {
  x as default
};
