// echarts@6.0.0/lib/component/brush/BrushView.js
import { __extends as p } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { bind as u, clone as n } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import a from "../helper/BrushController.js";
import { layoutCovers as h } from "./visualEncoding.js";
import l from "../../view/Component.js";
var C = (
  /** @class */
  (function(i) {
    p(r, i);
    function r() {
      var t = i !== null && i.apply(this, arguments) || this;
      return t.type = r.type, t;
    }
    return r.prototype.init = function(t, o) {
      this.ecModel = t, this.api = o, this.model, (this._brushController = new a(o.getZr())).on("brush", u(this._onBrush, this)).mount();
    }, r.prototype.render = function(t, o, e, s) {
      this.model = t, this._updateController(t, o, e, s);
    }, r.prototype.updateTransform = function(t, o, e, s) {
      h(o), this._updateController(t, o, e, s);
    }, r.prototype.updateVisual = function(t, o, e, s) {
      this.updateTransform(t, o, e, s);
    }, r.prototype.updateView = function(t, o, e, s) {
      this._updateController(t, o, e, s);
    }, r.prototype._updateController = function(t, o, e, s) {
      (!s || s.$from !== t.id) && this._brushController.setPanels(t.brushTargetManager.makePanelOpts(e)).enableBrush(t.brushOption).updateCovers(t.areas.slice());
    }, r.prototype.dispose = function() {
      this._brushController.dispose();
    }, r.prototype._onBrush = function(t) {
      var o = this.model.id, e = this.model.brushTargetManager.setOutputRanges(t.areas, this.ecModel);
      (!t.isEnd || t.removeOnClick) && this.api.dispatchAction({
        type: "brush",
        brushId: o,
        areas: n(e),
        $from: o
      }), t.isEnd && this.api.dispatchAction({
        type: "brushEnd",
        brushId: o,
        areas: n(e),
        $from: o
      });
    }, r.type = "brush", r;
  })(l)
);
export {
  C as default
};
