// echarts@6.0.0/lib/component/brush/selector.js
import { contain as f } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/contain/polygon.js";
import g from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/BoundingRect.js";
import { linePolygonIntersect as h } from "../../util/graphic.js";
function S(r) {
  var s = r.brushType, t = {
    point: function(n) {
      return l[s].point(n, t, r);
    },
    rect: function(n) {
      return l[s].rect(n, t, r);
    }
  };
  return t;
}
var l = {
  lineX: u(0),
  lineY: u(1),
  rect: {
    point: function(r, s, t) {
      return r && t.boundingRect.contain(r[0], r[1]);
    },
    rect: function(r, s, t) {
      return r && t.boundingRect.intersect(r);
    }
  },
  polygon: {
    point: function(r, s, t) {
      return r && t.boundingRect.contain(r[0], r[1]) && f(t.range, r[0], r[1]);
    },
    rect: function(r, s, t) {
      var n = t.range;
      if (!r || n.length <= 1)
        return !1;
      var i = r.x, e = r.y, o = r.width, c = r.height, p = n[0];
      if (f(n, i, e) || f(n, i + o, e) || f(n, i, e + c) || f(n, i + o, e + c) || g.create(r).contain(p[0], p[1]) || h(i, e, i + o, e, n) || h(i, e, i, e + c, n) || h(i + o, e, i + o, e + c, n) || h(i, e + c, i + o, e + c, n))
        return !0;
    }
  }
};
function u(r) {
  var s = ["x", "y"], t = ["width", "height"];
  return {
    point: function(n, i, e) {
      if (n) {
        var o = e.range, c = n[r];
        return v(c, o);
      }
    },
    rect: function(n, i, e) {
      if (n) {
        var o = e.range, c = [n[s[r]], n[s[r]] + n[t[r]]];
        return c[1] < c[0] && c.reverse(), v(c[0], o) || v(c[1], o) || v(o[0], c) || v(o[1], c);
      }
    }
  };
}
function v(r, s) {
  return s[0] <= r && r <= s[1];
}
export {
  l as default,
  S as makeBrushCommonSelectorForSeries
};
