// echarts@6.0.0/lib/component/brush/visualEncoding.js
import { clone as H, map as L, defaults as N, isArray as _, each as I, indexOf as P } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import G from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/BoundingRect.js";
import { createVisualMappings as Z, applyVisual as U } from "../../visual/visualSolution.js";
import { makeBrushCommonSelectorForSeries as j } from "./selector.js";
import { createOrUpdate as q } from "../../util/throttle.js";
import x from "../helper/BrushTargetManager.js";
var k = ["inBrush", "outOfBrush"], y = "__ecBrushSelect", B = "__ecInBrushSelectEvent";
function z(t) {
  t.eachComponent({
    mainType: "brush"
  }, function(r) {
    var e = r.brushTargetManager = new x(r.option, t);
    e.setInputRanges(r.areas, t);
  });
}
function nt(t, r, e) {
  var c = [], h, a;
  t.eachComponent({
    mainType: "brush"
  }, function(o) {
    e && e.type === "takeGlobalCursor" && o.setBrushOption(e.key === "brush" ? e.brushOption : {
      brushType: !1
    });
  }), z(t), t.eachComponent({
    mainType: "brush"
  }, function(o, T) {
    var A = {
      brushId: o.id,
      brushIndex: T,
      brushName: o.name,
      areas: H(o.areas),
      selected: []
    };
    c.push(A);
    var g = o.option, m = g.brushLink, D = [], S = [], O = [], p = !1;
    T || (h = g.throttleType, a = g.throttleDelay);
    var w = L(o.areas, function(n) {
      var i = W[n.brushType], u = N({
        boundingRect: i ? i(n) : void 0
      }, n);
      return u.selectors = j(u), u;
    }), E = Z(o.option, k, function(n) {
      n.mappingMethod = "fixed";
    });
    _(m) && I(m, function(n) {
      D[n] = 1;
    });
    function l(n) {
      return m === "all" || !!D[n];
    }
    function b(n) {
      return !!n.length;
    }
    t.eachSeries(function(n, i) {
      var u = O[i] = [];
      n.subType === "parallel" ? F(n, i) : V(n, i, u);
    });
    function F(n, i) {
      var u = n.coordinateSystem;
      p = p || u.hasAxisBrushed(), l(i) && u.eachActiveState(n.getData(), function(f, s) {
        f === "active" && (S[s] = 1);
      });
    }
    function V(n, i, u) {
      if (!(!n.brushSelector || Q(o, i)) && (I(w, function(s) {
        o.brushTargetManager.controlSeries(s, n, t) && u.push(s), p = p || b(u);
      }), l(i) && b(u))) {
        var f = n.getData();
        f.each(function(s) {
          C(n, u, f, s) && (S[s] = 1);
        });
      }
    }
    t.eachSeries(function(n, i) {
      var u = {
        seriesId: n.id,
        seriesIndex: i,
        seriesName: n.name,
        dataIndex: []
      };
      A.selected.push(u);
      var f = O[i], s = n.getData(), d = l(i) ? function(v) {
        return S[v] ? (u.dataIndex.push(s.getRawIndex(v)), "inBrush") : "outOfBrush";
      } : function(v) {
        return C(n, f, s, v) ? (u.dataIndex.push(s.getRawIndex(v)), "inBrush") : "outOfBrush";
      };
      (l(i) ? p : b(f)) && U(k, E, s, d);
    });
  }), J(r, h, a, c, e);
}
function J(t, r, e, c, h) {
  if (h) {
    var a = t.getZr();
    if (!a[B]) {
      a[y] || (a[y] = K);
      var o = q(a, y, e, r);
      o(t, c);
    }
  }
}
function K(t, r) {
  if (!t.isDisposed()) {
    var e = t.getZr();
    e[B] = !0, t.dispatchAction({
      type: "brushSelect",
      batch: r
    }), e[B] = !1;
  }
}
function C(t, r, e, c) {
  for (var h = 0, a = r.length; h < a; h++) {
    var o = r[h];
    if (t.brushSelector(c, e, o.selectors, o))
      return !0;
  }
}
function Q(t, r) {
  var e = t.option.seriesIndex;
  return e != null && e !== "all" && (_(e) ? P(e, r) < 0 : r !== e);
}
var W = {
  rect: function(t) {
    return R(t.range);
  },
  polygon: function(t) {
    for (var r, e = t.range, c = 0, h = e.length; c < h; c++) {
      r = r || [[1 / 0, -1 / 0], [1 / 0, -1 / 0]];
      var a = e[c];
      a[0] < r[0][0] && (r[0][0] = a[0]), a[0] > r[0][1] && (r[0][1] = a[0]), a[1] < r[1][0] && (r[1][0] = a[1]), a[1] > r[1][1] && (r[1][1] = a[1]);
    }
    return r && R(r);
  }
};
function R(t) {
  return new G(t[0][0], t[1][0], t[0][1] - t[0][0], t[1][1] - t[1][0]);
}
export {
  nt as default,
  z as layoutCovers
};
