// motion@12.17.0/dist/es/motion-dom/dist/es/animation/GroupAnimation.js
class a {
  constructor(t) {
    this.stop = () => this.runAll("stop"), this.animations = t.filter(Boolean);
  }
  get finished() {
    return Promise.all(this.animations.map((t) => t.finished));
  }
  /**
   * TODO: Filter out cancelled or stopped animations before returning
   */
  getAll(t) {
    return this.animations[0][t];
  }
  setAll(t, i) {
    for (let e = 0; e < this.animations.length; e++)
      this.animations[e][t] = i;
  }
  attachTimeline(t) {
    const i = this.animations.map((e) => e.attachTimeline(t));
    return () => {
      i.forEach((e, s) => {
        e && e(), this.animations[s].stop();
      });
    };
  }
  get time() {
    return this.getAll("time");
  }
  set time(t) {
    this.setAll("time", t);
  }
  get speed() {
    return this.getAll("speed");
  }
  set speed(t) {
    this.setAll("speed", t);
  }
  get state() {
    return this.getAll("state");
  }
  get startTime() {
    return this.getAll("startTime");
  }
  get duration() {
    let t = 0;
    for (let i = 0; i < this.animations.length; i++)
      t = Math.max(t, this.animations[i].duration);
    return t;
  }
  runAll(t) {
    this.animations.forEach((i) => i[t]());
  }
  play() {
    this.runAll("play");
  }
  pause() {
    this.runAll("pause");
  }
  cancel() {
    this.runAll("cancel");
  }
  complete() {
    this.runAll("complete");
  }
}
export {
  a as GroupAnimation
};
