// micromark@4.0.2/lib/constructs.js
import { resolver as f } from "./initialize/text.js";
import { blockQuote as p } from "../../../../micromark-core-commonmark@2.0.3/node_modules/micromark-core-commonmark/lib/block-quote.js";
import { list as o } from "../../../../micromark-core-commonmark@2.0.3/node_modules/micromark-core-commonmark/lib/list.js";
import { definition as c } from "../../../../micromark-core-commonmark@2.0.3/node_modules/micromark-core-commonmark/lib/definition.js";
import { codeIndented as t } from "../../../../micromark-core-commonmark@2.0.3/node_modules/micromark-core-commonmark/lib/code-indented.js";
import { codeFenced as i } from "../../../../micromark-core-commonmark@2.0.3/node_modules/micromark-core-commonmark/lib/code-fenced.js";
import { thematicBreak as r } from "../../../../micromark-core-commonmark@2.0.3/node_modules/micromark-core-commonmark/lib/thematic-break.js";
import { setextUnderline as n } from "../../../../micromark-core-commonmark@2.0.3/node_modules/micromark-core-commonmark/lib/setext-underline.js";
import { htmlFlow as s } from "../../../../micromark-core-commonmark@2.0.3/node_modules/micromark-core-commonmark/lib/html-flow.js";
import { headingAtx as d } from "../../../../micromark-core-commonmark@2.0.3/node_modules/micromark-core-commonmark/lib/heading-atx.js";
import { characterEscape as l } from "../../../../micromark-core-commonmark@2.0.3/node_modules/micromark-core-commonmark/lib/character-escape.js";
import { characterReference as a } from "../../../../micromark-core-commonmark@2.0.3/node_modules/micromark-core-commonmark/lib/character-reference.js";
import { codeText as h } from "../../../../micromark-core-commonmark@2.0.3/node_modules/micromark-core-commonmark/lib/code-text.js";
import { attention as e } from "../../../../micromark-core-commonmark@2.0.3/node_modules/micromark-core-commonmark/lib/attention.js";
import { labelEnd as k } from "../../../../micromark-core-commonmark@2.0.3/node_modules/micromark-core-commonmark/lib/label-end.js";
import { hardBreakEscape as u } from "../../../../micromark-core-commonmark@2.0.3/node_modules/micromark-core-commonmark/lib/hard-break-escape.js";
import { labelStartLink as x } from "../../../../micromark-core-commonmark@2.0.3/node_modules/micromark-core-commonmark/lib/label-start-link.js";
import { autolink as b } from "../../../../micromark-core-commonmark@2.0.3/node_modules/micromark-core-commonmark/lib/autolink.js";
import { htmlText as g } from "../../../../micromark-core-commonmark@2.0.3/node_modules/micromark-core-commonmark/lib/html-text.js";
import { labelStartImage as E } from "../../../../micromark-core-commonmark@2.0.3/node_modules/micromark-core-commonmark/lib/label-start-image.js";
import { lineEnding as m } from "../../../../micromark-core-commonmark@2.0.3/node_modules/micromark-core-commonmark/lib/line-ending.js";
const J = {
  42: o,
  43: o,
  45: o,
  48: o,
  49: o,
  50: o,
  51: o,
  52: o,
  53: o,
  54: o,
  55: o,
  56: o,
  57: o,
  62: p
}, K = {
  91: c
}, N = {
  [-2]: t,
  [-1]: t,
  32: t
}, O = {
  35: d,
  42: r,
  45: [n, r],
  60: s,
  61: n,
  95: r,
  96: i,
  126: i
}, P = {
  38: a,
  92: l
}, V = {
  [-5]: m,
  [-4]: m,
  [-3]: m,
  33: E,
  38: a,
  42: e,
  60: [b, g],
  91: x,
  92: [u, l],
  93: k,
  95: e,
  96: h
}, W = {
  null: [e, f]
}, X = {
  null: [42, 95]
}, Y = {
  null: []
};
export {
  X as attentionMarkers,
  K as contentInitial,
  Y as disable,
  J as document,
  O as flow,
  N as flowInitial,
  W as insideSpan,
  P as string,
  V as text
};
