// echarts@6.0.0/lib/component/timeline/preprocessor.js
import { isArray as t, each as f, isObject as c } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
function h(e) {
  var i = e && e.timeline;
  t(i) || (i = i ? [i] : []), f(i, function(s) {
    s && v(s);
  });
}
function v(e) {
  var i = e.type, s = {
    number: "value",
    time: "time"
  };
  if (s[i] && (e.axisType = s[i], delete e.type), o(e), r(e, "controlPosition")) {
    var l = e.controlStyle || (e.controlStyle = {});
    r(l, "position") || (l.position = e.controlPosition), l.position === "none" && !r(l, "show") && (l.show = !1, delete l.position), delete e.controlPosition;
  }
  f(e.data || [], function(n) {
    c(n) && !t(n) && (!r(n, "value") && r(n, "name") && (n.value = n.name), o(n));
  });
}
function o(e) {
  var i = e.itemStyle || (e.itemStyle = {}), s = i.emphasis || (i.emphasis = {}), l = e.label || e.label || {}, n = l.normal || (l.normal = {}), m = {
    normal: 1,
    emphasis: 1
  };
  f(l, function(u, a) {
    !m[a] && !r(n, a) && (n[a] = u);
  }), s.label && !r(l, "emphasis") && (l.emphasis = s.label, delete s.label);
}
function r(e, i) {
  return e.hasOwnProperty(i);
}
export {
  h as default
};
