// echarts@6.0.0/lib/component/timeline/SliderTimelineView.js
import { __extends as V } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import N from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/BoundingRect.js";
import { translate as R, create as H, rotate as U } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/matrix.js";
import { applyTransform as X, createIcon as Y } from "../../util/graphic.js";
import { createTextStyle as L } from "../../label/labelStyle.js";
import { getLayoutRect as W, createBoxLayoutReference as Z } from "../../util/layout.js";
import j from "./TimelineView.js";
import q from "./TimelineAxis.js";
import { createSymbol as F, normalizeSymbolSize as J, normalizeSymbolOffset as K } from "../../util/symbol.js";
import { asc as Q } from "../../util/number.js";
import { each as B, isString as D, extend as $, defaults as tt, bind as b, retrieve2 as et, merge as rt } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import at from "../../scale/Ordinal.js";
import it from "../../scale/Time.js";
import ot from "../../scale/Interval.js";
import { parsePercent as nt } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/contain/text.js";
import { makeInner as lt } from "../../util/model.js";
import { getECData as st } from "../../util/innerStore.js";
import { enableHoverEmphasis as A } from "../../util/states.js";
import { createTooltipMarkup as ct } from "../tooltip/tooltipMarkup.js";
import vt from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import M from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Line.js";
import ut from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Text.js";
var z = Math.PI, E = lt(), Gt = (
  /** @class */
  (function(u) {
    V(n, u);
    function n() {
      var t = u !== null && u.apply(this, arguments) || this;
      return t.type = n.type, t;
    }
    return n.prototype.init = function(t, a) {
      this.api = a;
    }, n.prototype.render = function(t, a, e) {
      if (this.model = t, this.api = e, this.ecModel = a, this.group.removeAll(), t.get("show", !0)) {
        var r = this._layout(t, e), i = this._createGroup("_mainGroup"), o = this._createGroup("_labelGroup"), s = this._axis = this._createAxis(r, t);
        t.formatTooltip = function(l) {
          var c = s.scale.getLabel({
            value: l
          });
          return ct("nameValue", {
            noName: !0,
            value: c
          });
        }, B(["AxisLine", "AxisTick", "Control", "CurrentPointer"], function(l) {
          this["_render" + l](r, i, s, t);
        }, this), this._renderAxisLabel(r, o, s, t), this._position(r, t);
      }
      this._doPlayStop(), this._updateTicksStatus();
    }, n.prototype.remove = function() {
      this._clearTimer(), this.group.removeAll();
    }, n.prototype.dispose = function() {
      this._clearTimer();
    }, n.prototype._layout = function(t, a) {
      var e = t.get(["label", "position"]), r = t.get("orient"), i = pt(t, a), o;
      e == null || e === "auto" ? o = r === "horizontal" ? i.y + i.height / 2 < a.getHeight() / 2 ? "-" : "+" : i.x + i.width / 2 < a.getWidth() / 2 ? "+" : "-" : D(e) ? o = {
        horizontal: {
          top: "-",
          bottom: "+"
        },
        vertical: {
          left: "-",
          right: "+"
        }
      }[r][e] : o = e;
      var s = {
        horizontal: "center",
        vertical: o >= 0 || o === "+" ? "left" : "right"
      }, l = {
        horizontal: o >= 0 || o === "+" ? "top" : "bottom",
        vertical: "middle"
      }, c = {
        horizontal: 0,
        vertical: z / 2
      }, v = r === "vertical" ? i.height : i.width, h = t.getModel("controlStyle"), d = h.get("show", !0), f = d ? h.get("itemSize") : 0, y = d ? h.get("itemGap") : 0, p = f + y, g = t.get(["label", "rotate"]) || 0;
      g = g * z / 180;
      var _, x, m, C = h.get("position", !0), k = d && h.get("showPlayBtn", !0), P = d && h.get("showPrevBtn", !0), T = d && h.get("showNextBtn", !0), w = 0, S = v;
      C === "left" || C === "bottom" ? (k && (_ = [0, 0], w += p), P && (x = [w, 0], w += p), T && (m = [S - f, 0], S -= p)) : (k && (_ = [S - f, 0], S -= p), P && (x = [0, 0], w += p), T && (m = [S - f, 0], S -= p));
      var I = [w, S];
      return t.get("inverse") && I.reverse(), {
        viewRect: i,
        mainLength: v,
        orient: r,
        rotation: c[r],
        labelRotation: g,
        labelPosOpt: o,
        labelAlign: t.get(["label", "align"]) || s[r],
        labelBaseline: t.get(["label", "verticalAlign"]) || t.get(["label", "baseline"]) || l[r],
        // Based on mainGroup.
        playPosition: _,
        prevBtnPosition: x,
        nextBtnPosition: m,
        axisExtent: I,
        controlSize: f,
        controlGap: y
      };
    }, n.prototype._position = function(t, a) {
      var e = this._mainGroup, r = this._labelGroup, i = t.viewRect;
      if (t.orient === "vertical") {
        var o = H(), s = i.x, l = i.y + i.height;
        R(o, o, [-s, -l]), U(o, o, -z / 2), R(o, o, [s, l]), i = i.clone(), i.applyTransform(o);
      }
      var c = _(i), v = _(e.getBoundingRect()), h = _(r.getBoundingRect()), d = [e.x, e.y], f = [r.x, r.y];
      f[0] = d[0] = c[0][0];
      var y = t.labelPosOpt;
      if (y == null || D(y)) {
        var p = y === "+" ? 0 : 1;
        x(d, v, c, 1, p), x(f, h, c, 1, 1 - p);
      } else {
        var p = y >= 0 ? 0 : 1;
        x(d, v, c, 1, p), f[1] = d[1] + y;
      }
      e.setPosition(d), r.setPosition(f), e.rotation = r.rotation = t.rotation, g(e), g(r);
      function g(m) {
        m.originX = c[0][0] - m.x, m.originY = c[1][0] - m.y;
      }
      function _(m) {
        return [[m.x, m.x + m.width], [m.y, m.y + m.height]];
      }
      function x(m, C, k, P, T) {
        m[P] += k[P][T] - C[P][T];
      }
    }, n.prototype._createAxis = function(t, a) {
      var e = a.getData(), r = a.get("axisType"), i = gt(a, r);
      i.getTicks = function() {
        return e.mapArray(["value"], function(l) {
          return {
            value: l
          };
        });
      };
      var o = e.getDataExtent("value");
      i.setExtent(o[0], o[1]), i.calcNiceTicks();
      var s = new q("value", i, t.axisExtent, r);
      return s.model = a, s;
    }, n.prototype._createGroup = function(t) {
      var a = this[t] = new vt();
      return this.group.add(a), a;
    }, n.prototype._renderAxisLine = function(t, a, e, r) {
      var i = e.getExtent();
      if (r.get(["lineStyle", "show"])) {
        var o = new M({
          shape: {
            x1: i[0],
            y1: 0,
            x2: i[1],
            y2: 0
          },
          style: $({
            lineCap: "round"
          }, r.getModel("lineStyle").getLineStyle()),
          silent: !0,
          z2: 1
        });
        a.add(o);
        var s = this._progressLine = new M({
          shape: {
            x1: i[0],
            x2: this._currentPointer ? this._currentPointer.x : i[0],
            y1: 0,
            y2: 0
          },
          style: tt({
            lineCap: "round",
            lineWidth: o.style.lineWidth
          }, r.getModel(["progress", "lineStyle"]).getLineStyle()),
          silent: !0,
          z2: 1
        });
        a.add(s);
      }
    }, n.prototype._renderAxisTick = function(t, a, e, r) {
      var i = this, o = r.getData(), s = e.scale.getTicks();
      this._tickSymbols = [], B(s, function(l) {
        var c = e.dataToCoord(l.value), v = o.getItemModel(l.value), h = v.getModel("itemStyle"), d = v.getModel(["emphasis", "itemStyle"]), f = v.getModel(["progress", "itemStyle"]), y = {
          x: c,
          y: 0,
          onclick: b(i._changeTimeline, i, l.value)
        }, p = G(v, h, a, y);
        p.ensureState("emphasis").style = d.getItemStyle(), p.ensureState("progress").style = f.getItemStyle(), A(p);
        var g = st(p);
        v.get("tooltip") ? (g.dataIndex = l.value, g.dataModel = r) : g.dataIndex = g.dataModel = null, i._tickSymbols.push(p);
      });
    }, n.prototype._renderAxisLabel = function(t, a, e, r) {
      var i = this, o = e.getLabelModel();
      if (o.get("show")) {
        var s = r.getData(), l = e.getViewLabels();
        this._tickLabels = [], B(l, function(c) {
          var v = c.tickValue, h = s.getItemModel(v), d = h.getModel("label"), f = h.getModel(["emphasis", "label"]), y = h.getModel(["progress", "label"]), p = e.dataToCoord(c.tickValue), g = new ut({
            x: p,
            y: 0,
            rotation: t.labelRotation - t.rotation,
            onclick: b(i._changeTimeline, i, v),
            silent: !1,
            style: L(d, {
              text: c.formattedLabel,
              align: t.labelAlign,
              verticalAlign: t.labelBaseline
            })
          });
          g.ensureState("emphasis").style = L(f), g.ensureState("progress").style = L(y), a.add(g), A(g), E(g).dataIndex = v, i._tickLabels.push(g);
        });
      }
    }, n.prototype._renderControl = function(t, a, e, r) {
      var i = t.controlSize, o = t.rotation, s = r.getModel("controlStyle").getItemStyle(), l = r.getModel(["emphasis", "controlStyle"]).getItemStyle(), c = r.getPlayState(), v = r.get("inverse", !0);
      h(t.nextBtnPosition, "next", b(this._changeTimeline, this, v ? "-" : "+")), h(t.prevBtnPosition, "prev", b(this._changeTimeline, this, v ? "+" : "-")), h(t.playPosition, c ? "stop" : "play", b(this._handlePlayClick, this, !c), !0);
      function h(d, f, y, p) {
        if (d) {
          var g = nt(et(r.get(["controlStyle", f + "BtnSize"]), i), i), _ = [0, -g / 2, g, g], x = ht(r, f + "Icon", _, {
            x: d[0],
            y: d[1],
            originX: i / 2,
            originY: 0,
            rotation: p ? -o : 0,
            rectHover: !0,
            style: s,
            onclick: y
          });
          x.ensureState("emphasis").style = l, a.add(x), A(x);
        }
      }
    }, n.prototype._renderCurrentPointer = function(t, a, e, r) {
      var i = r.getData(), o = r.getCurrentIndex(), s = i.getItemModel(o).getModel("checkpointStyle"), l = this, c = {
        onCreate: function(v) {
          v.draggable = !0, v.drift = b(l._handlePointerDrag, l), v.ondragend = b(l._handlePointerDragend, l), O(v, l._progressLine, o, e, r, !0);
        },
        onUpdate: function(v) {
          O(v, l._progressLine, o, e, r);
        }
      };
      this._currentPointer = G(s, s, this._mainGroup, {}, this._currentPointer, c);
    }, n.prototype._handlePlayClick = function(t) {
      this._clearTimer(), this.api.dispatchAction({
        type: "timelinePlayChange",
        playState: t,
        from: this.uid
      });
    }, n.prototype._handlePointerDrag = function(t, a, e) {
      this._clearTimer(), this._pointerChangeTimeline([e.offsetX, e.offsetY]);
    }, n.prototype._handlePointerDragend = function(t) {
      this._pointerChangeTimeline([t.offsetX, t.offsetY], !0);
    }, n.prototype._pointerChangeTimeline = function(t, a) {
      var e = this._toAxisCoord(t)[0], r = this._axis, i = Q(r.getExtent().slice());
      e > i[1] && (e = i[1]), e < i[0] && (e = i[0]), this._currentPointer.x = e, this._currentPointer.markRedraw();
      var o = this._progressLine;
      o && (o.shape.x2 = e, o.dirty());
      var s = this._findNearestTick(e), l = this.model;
      (a || s !== l.getCurrentIndex() && l.get("realtime")) && this._changeTimeline(s);
    }, n.prototype._doPlayStop = function() {
      var t = this;
      this._clearTimer(), this.model.getPlayState() && (this._timer = setTimeout(function() {
        var a = t.model;
        t._changeTimeline(a.getCurrentIndex() + (a.get("rewind", !0) ? -1 : 1));
      }, this.model.get("playInterval")));
    }, n.prototype._toAxisCoord = function(t) {
      var a = this._mainGroup.getLocalTransform();
      return X(t, a, !0);
    }, n.prototype._findNearestTick = function(t) {
      var a = this.model.getData(), e = 1 / 0, r, i = this._axis;
      return a.each(["value"], function(o, s) {
        var l = i.dataToCoord(o), c = Math.abs(l - t);
        c < e && (e = c, r = s);
      }), r;
    }, n.prototype._clearTimer = function() {
      this._timer && (clearTimeout(this._timer), this._timer = null);
    }, n.prototype._changeTimeline = function(t) {
      var a = this.model.getCurrentIndex();
      t === "+" ? t = a + 1 : t === "-" && (t = a - 1), this.api.dispatchAction({
        type: "timelineChange",
        currentIndex: t,
        from: this.uid
      });
    }, n.prototype._updateTicksStatus = function() {
      var t = this.model.getCurrentIndex(), a = this._tickSymbols, e = this._tickLabels;
      if (a)
        for (var r = 0; r < a.length; r++)
          a && a[r] && a[r].toggleState("progress", r < t);
      if (e)
        for (var r = 0; r < e.length; r++)
          e && e[r] && e[r].toggleState("progress", E(e[r]).dataIndex <= t);
    }, n.type = "timeline.slider", n;
  })(j)
);
function gt(u, n) {
  if (n = n || u.get("type"), n)
    switch (n) {
      // Buildin scale
      case "category":
        return new at({
          ordinalMeta: u.getCategories(),
          extent: [1 / 0, -1 / 0]
        });
      case "time":
        return new it({
          locale: u.ecModel.getLocaleModel(),
          useUTC: u.ecModel.get("useUTC")
        });
      default:
        return new ot();
    }
}
function pt(u, n) {
  return W(u.getBoxLayoutParams(), Z(u, n).refContainer, u.get("padding"));
}
function ht(u, n, t, a) {
  var e = a.style, r = Y(u.get(["controlStyle", n]), a || {}, new N(t[0], t[1], t[2], t[3]));
  return e && r.setStyle(e), r;
}
function G(u, n, t, a, e, r) {
  var i = n.get("color");
  if (e)
    e.setColor(i), t.add(e), r && r.onUpdate(e);
  else {
    var o = u.get("symbol");
    e = F(o, -1, -1, 2, 2, i), e.setStyle("strokeNoScale", !0), t.add(e), r && r.onCreate(e);
  }
  var s = n.getItemStyle(["color"]);
  e.setStyle(s), a = rt({
    rectHover: !0,
    z2: 100
  }, a, !0);
  var l = J(u.get("symbolSize"));
  a.scaleX = l[0] / 2, a.scaleY = l[1] / 2;
  var c = K(u.get("symbolOffset"), l);
  c && (a.x = (a.x || 0) + c[0], a.y = (a.y || 0) + c[1]);
  var v = u.get("symbolRotate");
  return a.rotation = (v || 0) * Math.PI / 180 || 0, e.attr(a), e.updateTransform(), e;
}
function O(u, n, t, a, e, r) {
  if (!u.dragging) {
    var i = e.getModel("checkpointStyle"), o = a.dataToCoord(e.getData().get("value", t));
    if (r || !i.get("animation", !0))
      u.attr({
        x: o,
        y: 0
      }), n && n.attr({
        shape: {
          x2: o
        }
      });
    else {
      var s = {
        duration: i.get("animationDuration", !0),
        easing: i.get("animationEasing", !0)
      };
      u.stopAnimation(null, !0), u.animateTo({
        x: o,
        y: 0
      }, s), n && n.animateTo({
        shape: {
          x2: o
        }
      }, s);
    }
  }
}
export {
  Gt as default
};
