// echarts@6.0.0/lib/component/timeline/TimelineModel.js
import { __extends as f } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import c from "../../model/Component.js";
import d from "../../data/SeriesData.js";
import { each as g, isObject as v, clone as D } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { convertOptionIdName as _, getDataItemValue as T } from "../../util/model.js";
import u from "../../visual/tokens.js";
var C = (
  /** @class */
  (function(i) {
    f(e, i);
    function e() {
      var t = i !== null && i.apply(this, arguments) || this;
      return t.type = e.type, t.layoutMode = "box", t;
    }
    return e.prototype.init = function(t, o, a) {
      this.mergeDefaultAndTheme(t, a), this._initData();
    }, e.prototype.mergeOption = function(t) {
      i.prototype.mergeOption.apply(this, arguments), this._initData();
    }, e.prototype.setCurrentIndex = function(t) {
      t == null && (t = this.option.currentIndex);
      var o = this._data.count();
      this.option.loop ? t = (t % o + o) % o : (t >= o && (t = o - 1), t < 0 && (t = 0)), this.option.currentIndex = t;
    }, e.prototype.getCurrentIndex = function() {
      return this.option.currentIndex;
    }, e.prototype.isIndexMax = function() {
      return this.getCurrentIndex() >= this._data.count() - 1;
    }, e.prototype.setPlayState = function(t) {
      this.option.autoPlay = !!t;
    }, e.prototype.getPlayState = function() {
      return !!this.option.autoPlay;
    }, e.prototype._initData = function() {
      var t = this.option, o = t.data || [], a = t.axisType, s = this._names = [], n;
      a === "category" ? (n = [], g(o, function(p, l) {
        var h = _(T(p), ""), r;
        v(p) ? (r = D(p), r.value = l) : r = l, n.push(r), s.push(h);
      })) : n = o;
      var m = {
        category: "ordinal",
        time: "time",
        value: "number"
      }[a] || "number", y = this._data = new d([{
        name: "value",
        type: m
      }], this);
      y.initData(n, s);
    }, e.prototype.getData = function() {
      return this._data;
    }, e.prototype.getCategories = function() {
      if (this.get("axisType") === "category")
        return this._names.slice();
    }, e.type = "timeline", e.defaultOption = {
      // zlevel: 0,                  // 一级层叠
      z: 4,
      show: !0,
      axisType: "time",
      realtime: !0,
      left: "20%",
      top: null,
      right: "20%",
      bottom: 0,
      width: null,
      height: 40,
      padding: u.size.m,
      controlPosition: "left",
      autoPlay: !1,
      rewind: !1,
      loop: !0,
      playInterval: 2e3,
      currentIndex: 0,
      itemStyle: {},
      label: {
        color: u.color.secondary
      },
      data: []
    }, e;
  })(c)
);
export {
  C as default
};
