// echarts@6.0.0/lib/chart/parallel/ParallelView.js
import { __extends as w } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { setStatesStylesFromModel as z, toggleHoverEmphasis as C } from "../../util/states.js";
import M from "../../view/Chart.js";
import { numericToNumber as T } from "../../util/number.js";
import { eqNaN as V } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { updateProps as A, saveOldStyle as R, initProps as b } from "../../animation/basicTransition.js";
import N from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import O from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Polyline.js";
import F from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Rect.js";
var H = 0.3, X = (
  /** @class */
  (function(r) {
    w(t, r);
    function t() {
      var e = r !== null && r.apply(this, arguments) || this;
      return e.type = t.type, e._dataGroup = new N(), e._initialized = !1, e;
    }
    return t.prototype.init = function() {
      this.group.add(this._dataGroup);
    }, t.prototype.render = function(e, n, i, o) {
      this._progressiveEls = null;
      var a = this._dataGroup, p = e.getData(), m = this._data, u = e.coordinateSystem, l = u.dimensions, h = g(e);
      p.diff(m).add(_).update(S).remove(G).execute();
      function _(s) {
        var c = f(p, a, s, l, u);
        v(c, p, s, h);
      }
      function S(s, c) {
        var d = m.getItemGraphicEl(c), P = y(p, s, l, u);
        p.setItemGraphicEl(s, d), A(d, {
          shape: {
            points: P
          }
        }, e, s), R(d), v(d, p, s, h);
      }
      function G(s) {
        var c = m.getItemGraphicEl(s);
        a.remove(c);
      }
      if (!this._initialized) {
        this._initialized = !0;
        var E = L(u, e, function() {
          setTimeout(function() {
            a.removeClipPath();
          });
        });
        a.setClipPath(E);
      }
      this._data = p;
    }, t.prototype.incrementalPrepareRender = function(e, n, i) {
      this._initialized = !0, this._data = null, this._dataGroup.removeAll();
    }, t.prototype.incrementalRender = function(e, n, i) {
      for (var o = n.getData(), a = n.coordinateSystem, p = a.dimensions, m = g(n), u = this._progressiveEls = [], l = e.start; l < e.end; l++) {
        var h = f(o, this._dataGroup, l, p, a);
        h.incremental = !0, v(h, o, l, m), u.push(h);
      }
    }, t.prototype.remove = function() {
      this._dataGroup && this._dataGroup.removeAll(), this._data = null;
    }, t.type = "parallel", t;
  })(M)
);
function L(r, t, e) {
  var n = r.model, i = r.getRect(), o = new F({
    shape: {
      x: i.x,
      y: i.y,
      width: i.width,
      height: i.height
    }
  }), a = n.get("layout") === "horizontal" ? "width" : "height";
  return o.setShape(a, 0), b(o, {
    shape: {
      width: i.width,
      height: i.height
    }
  }, t, e), o;
}
function y(r, t, e, n) {
  for (var i = [], o = 0; o < e.length; o++) {
    var a = e[o], p = r.get(r.mapDimension(a), t);
    q(p, n.getAxis(a).type) || i.push(n.dataToPoint(p, a));
  }
  return i;
}
function f(r, t, e, n, i) {
  var o = y(r, e, n, i), a = new O({
    shape: {
      points: o
    },
    // silent: true,
    z2: 10
  });
  return t.add(a), r.setItemGraphicEl(e, a), a;
}
function g(r) {
  var t = r.get("smooth", !0);
  return t === !0 && (t = H), t = T(t), V(t) && (t = 0), {
    smooth: t
  };
}
function v(r, t, e, n) {
  r.useStyle(t.getItemVisual(e, "style")), r.style.fill = null, r.setShape("smooth", n.smooth);
  var i = t.getItemModel(e), o = i.getModel("emphasis");
  z(r, i, "lineStyle"), C(r, o.get("focus"), o.get("blurScope"), o.get("disabled"));
}
function q(r, t) {
  return t === "category" ? r == null : r == null || isNaN(r);
}
export {
  X as default
};
