// echarts@6.0.0/lib/chart/parallel/ParallelSeries.js
import { __extends as s } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { bind as c, each as p } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import d from "../../model/Series.js";
import u from "../helper/createSeriesData.js";
var g = (
  /** @class */
  (function(a) {
    s(e, a);
    function e() {
      var t = a !== null && a.apply(this, arguments) || this;
      return t.type = e.type, t.visualStyleAccessPath = "lineStyle", t.visualDrawType = "stroke", t;
    }
    return e.prototype.getInitialData = function(t, l) {
      return u(null, this, {
        useEncodeDefaulter: c(f, null, this)
      });
    }, e.prototype.getRawIndicesByActiveState = function(t) {
      var l = this.coordinateSystem, i = this.getData(), n = [];
      return l.eachActiveState(i, function(r, o) {
        t === r && n.push(i.getRawIndex(o));
      }), n;
    }, e.type = "series.parallel", e.dependencies = ["parallel"], e.defaultOption = {
      // zlevel: 0,
      z: 2,
      coordinateSystem: "parallel",
      parallelIndex: 0,
      label: {
        show: !1
      },
      inactiveOpacity: 0.05,
      activeOpacity: 1,
      lineStyle: {
        width: 1,
        opacity: 0.45,
        type: "solid"
      },
      emphasis: {
        label: {
          show: !1
        }
      },
      progressive: 500,
      smooth: !1,
      animationEasing: "linear"
    }, e;
  })(d)
);
function f(a) {
  var e = a.ecModel.getComponent("parallel", a.get("parallelIndex"));
  if (e) {
    var t = {};
    return p(e.dimensions, function(l) {
      var i = m(l);
      t[l] = i;
    }), t;
  }
}
function m(a) {
  return +a.replace("dim", "");
}
export {
  g as default
};
