// echarts@6.0.0/lib/chart/helper/createGraphFromNodeEdge.js
import { retrieve as l, indexOf as N } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import S from "../../data/SeriesData.js";
import w from "../../data/Graph.js";
import x from "../../data/helper/linkSeriesData.js";
import y from "../../data/helper/createDimensions.js";
import A from "../../core/CoordinateSystem.js";
import G from "./createSeriesData.js";
import { convertOptionIdName as O } from "../../util/model.js";
function J(r, v, e, s, p) {
  for (var i = new w(s), a = 0; a < r.length; a++)
    i.addNode(l(
      // Id, name, dataIndex
      r[a].id,
      r[a].name,
      a
    ), a);
  for (var f = [], g = [], u = 0, a = 0; a < v.length; a++) {
    var m = v[a], c = m.source, D = m.target;
    i.addEdge(c, D, u) && (g.push(m), f.push(l(O(m.id, null), c + " > " + D)), u++);
  }
  var o = e.get("coordinateSystem"), t;
  if (o === "cartesian2d" || o === "polar" || o === "matrix")
    t = G(r, e);
  else {
    var h = A.get(o), n = h ? h.dimensions || [] : [];
    N(n, "value") < 0 && n.concat(["value"]);
    var E = y(r, {
      coordDimensions: n,
      encodeDefine: e.getEncode()
    }).dimensions;
    t = new S(E, e), t.initData(r);
  }
  var d = new S(["value"], e);
  return d.initData(g, f), p && p(t, d), x({
    mainData: t,
    struct: i,
    structAttr: "graph",
    datas: {
      node: t,
      edge: d
    },
    datasAttr: {
      node: "data",
      edge: "edgeData"
    }
  }), i.update(), i;
}
export {
  J as default
};
