// echarts@6.0.0/lib/chart/helper/createSeriesData.js
import { isFunction as O, curry as x, map as N, each as A, isArray as I } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import E from "../../data/SeriesData.js";
import F from "../../data/helper/createDimensions.js";
import { getDimensionTypeByAxis as M } from "../../data/helper/dimensionHelper.js";
import { getDataItemValue as k } from "../../util/model.js";
import G from "../../core/CoordinateSystem.js";
import { getCoordSysInfoBySeries as V } from "../../model/referHelper.js";
import { createSourceFromSeriesDataOption as B } from "../../data/Source.js";
import { enableDataStack as R } from "../../data/helper/dataStackHelper.js";
import { makeSeriesEncodeForAxisCoordSys as T } from "../../data/helper/sourceHelper.js";
import { SOURCE_FORMAT_ORIGINAL as g } from "../../util/types.js";
function U(r, e) {
  var o = r.get("coordinateSystem"), a = G.get(o), t;
  return e && e.coordSysDims && (t = N(e.coordSysDims, function(i) {
    var m = {
      name: i
    }, s = e.axisMap.get(i);
    if (s) {
      var n = s.get("type");
      m.type = M(n);
    }
    return m;
  })), t || (t = a && (a.getDimensionsInfo ? a.getDimensionsInfo() : a.dimensions.slice()) || ["x", "y"]), t;
}
function _(r, e, o) {
  var a, t;
  return o && A(r, function(i, m) {
    var s = i.coordDim, n = o.categoryAxisMap.get(s);
    n && (a == null && (a = m), i.ordinalMeta = n.getOrdinalMeta(), e && (i.createInvertedIndices = !0)), i.otherDims.itemName != null && (t = !0);
  }), !t && a != null && (r[a].otherDims.itemName = 0), a;
}
function Y(r, e, o) {
  o = o || {};
  var a = e.getSourceManager(), t, i = !1;
  r ? (i = !0, t = B(r)) : (t = a.getSource(), i = t.sourceFormat === g);
  var m = V(e), s = U(e, m), n = o.useEncodeDefaulter, v = O(n) ? n : n ? x(T, s, e) : null, S = {
    coordDimensions: s,
    generateCoord: o.generateCoord,
    encodeDefine: e.getEncode(),
    encodeDefaulter: v,
    canOmitUnusedDimensions: !i
  }, c = F(t, S), f = _(c.dimensions, o.createInvertedIndices, m), l = i ? null : a.getSharedDataStore(c), p = R(e, {
    schema: c,
    store: l
  }), u = new E(c, e);
  u.setCalculationInfo(p);
  var y = f != null && b(t) ? function(C, h, D, d) {
    return d === f ? D : this.defaultDimValueGetter(C, h, D, d);
  } : null;
  return u.hasItemOption = !1, u.initData(
    // Try to reuse the data store in sourceManager if using dataset.
    i ? t : l,
    null,
    y
  ), u;
}
function b(r) {
  if (r.sourceFormat === g) {
    var e = j(r.data || []);
    return !I(k(e));
  }
}
function j(r) {
  for (var e = 0; e < r.length && r[e] == null; )
    e++;
  return r[e];
}
export {
  Y as default
};
