// echarts@6.0.0/lib/chart/helper/createClipPathFromCoordSys.js
import { round as o } from "../../util/number.js";
import { isFunction as d } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import x from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Rect.js";
import P from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Sector.js";
import { initProps as s } from "../../animation/basicTransition.js";
function y(e, p, i, r, a) {
  var n = e.getArea(), t = n.x, f = n.y, h = n.width, c = n.height, g = i.get(["lineStyle", "width"]) || 0;
  t -= g / 2, f -= g / 2, h += g, c += g, h = Math.ceil(h), t !== Math.floor(t) && (t = Math.floor(t), h++);
  var l = new x({
    shape: {
      x: t,
      y: f,
      width: h,
      height: c
    }
  });
  if (p) {
    var v = e.getBaseAxis(), m = v.isHorizontal(), u = v.inverse;
    m ? (u && (l.shape.x += h), l.shape.width = 0) : (u || (l.shape.y += c), l.shape.height = 0);
    var w = d(a) ? function(A) {
      a(A, l);
    } : null;
    s(l, {
      shape: {
        width: h,
        height: c,
        x: t,
        y: f
      }
    }, i, null, r, w);
  }
  return l;
}
function C(e, p, i) {
  var r = e.getArea(), a = o(r.r0, 1), n = o(r.r, 1), t = new P({
    shape: {
      cx: o(e.cx, 1),
      cy: o(e.cy, 1),
      r0: a,
      r: n,
      startAngle: r.startAngle,
      endAngle: r.endAngle,
      clockwise: r.clockwise
    }
  });
  if (p) {
    var f = e.getBaseAxis().dim === "angle";
    f ? t.shape.endAngle = r.startAngle : t.shape.r = a, s(t, {
      shape: {
        endAngle: r.endAngle,
        r: n
      }
    }, i);
  }
  return t;
}
function H(e, p, i, r, a) {
  if (e) {
    if (e.type === "polar")
      return C(e, p, i);
    if (e.type === "cartesian2d")
      return y(e, p, i, r, a);
  } else return null;
  return null;
}
export {
  H as createClipPath,
  y as createGridClipPath,
  C as createPolarClipPath
};
