// echarts@6.0.0/lib/chart/helper/EffectSymbol.js
import { __extends as d } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { createSymbol as S, normalizeSymbolSize as g, normalizeSymbolOffset as A } from "../../util/symbol.js";
import "../../util/graphic.js";
import { enterEmphasis as z, leaveEmphasis as w, toggleHoverEmphasis as O } from "../../util/states.js";
import T from "./Symbol.js";
import v from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
function b(f, o) {
  var e = o.rippleEffectColor || o.color;
  f.eachChild(function(r) {
    r.attr({
      z: o.z,
      zlevel: o.zlevel,
      style: {
        stroke: o.brushType === "stroke" ? e : null,
        fill: o.brushType === "fill" ? e : null
      }
    });
  });
}
var N = (
  /** @class */
  (function(f) {
    d(o, f);
    function o(e, r) {
      var s = f.call(this) || this, p = new T(e, r), l = new v();
      return s.add(p), s.add(l), s.updateData(e, r), s;
    }
    return o.prototype.stopEffectAnimation = function() {
      this.childAt(1).removeAll();
    }, o.prototype.startEffectAnimation = function(e) {
      for (var r = e.symbolType, s = e.color, p = e.rippleNumber, l = this.childAt(1), i = 0; i < p; i++) {
        var a = S(r, -1, -1, 2, 2, s);
        a.attr({
          style: {
            strokeNoScale: !0
          },
          z2: 99,
          silent: !0,
          scaleX: 0.5,
          scaleY: 0.5
        });
        var n = -i / p * e.period + e.effectOffset;
        a.animate("", !0).when(e.period, {
          scaleX: e.rippleScale / 2,
          scaleY: e.rippleScale / 2
        }).delay(n).start(), a.animateStyle(!0).when(e.period, {
          opacity: 0
        }).delay(n).start(), l.add(a);
      }
      b(l, e);
    }, o.prototype.updateEffectAnimation = function(e) {
      for (var r = this._effectCfg, s = this.childAt(1), p = ["symbolType", "period", "rippleScale", "rippleNumber"], l = 0; l < p.length; l++) {
        var i = p[l];
        if (r[i] !== e[i]) {
          this.stopEffectAnimation(), this.startEffectAnimation(e);
          return;
        }
      }
      b(s, e);
    }, o.prototype.highlight = function() {
      z(this);
    }, o.prototype.downplay = function() {
      w(this);
    }, o.prototype.getSymbolType = function() {
      var e = this.childAt(0);
      return e && e.getSymbolType();
    }, o.prototype.updateData = function(e, r) {
      var s = this, p = e.hostModel;
      this.childAt(0).updateData(e, r);
      var l = this.childAt(1), i = e.getItemModel(r), a = e.getItemVisual(r, "symbol"), n = g(e.getItemVisual(r, "symbolSize")), u = e.getItemVisual(r, "style"), y = u && u.fill, c = i.getModel("emphasis");
      l.setScale(n), l.traverse(function(m) {
        m.setStyle("fill", y);
      });
      var h = A(e.getItemVisual(r, "symbolOffset"), n);
      h && (l.x = h[0], l.y = h[1]);
      var E = e.getItemVisual(r, "symbolRotate");
      l.rotation = (E || 0) * Math.PI / 180 || 0;
      var t = {};
      t.showEffectOn = p.get("showEffectOn"), t.rippleScale = i.get(["rippleEffect", "scale"]), t.brushType = i.get(["rippleEffect", "brushType"]), t.period = i.get(["rippleEffect", "period"]) * 1e3, t.effectOffset = r / e.count(), t.z = p.getShallow("z") || 0, t.zlevel = p.getShallow("zlevel") || 0, t.symbolType = a, t.color = y, t.rippleEffectColor = i.get(["rippleEffect", "color"]), t.rippleNumber = i.get(["rippleEffect", "number"]), t.showEffectOn === "render" ? (this._effectCfg ? this.updateEffectAnimation(t) : this.startEffectAnimation(t), this._effectCfg = t) : (this._effectCfg = null, this.stopEffectAnimation(), this.onHoverStateChange = function(m) {
        m === "emphasis" ? t.showEffectOn !== "render" && s.startEffectAnimation(t) : m === "normal" && t.showEffectOn !== "render" && s.stopEffectAnimation();
      }), this._effectCfg = t, O(this, c.get("focus"), c.get("blurScope"), c.get("disabled"));
    }, o.prototype.fadeOut = function(e) {
      e && e();
    }, o;
  })(v)
);
export {
  N as default
};
