// echarts@6.0.0/lib/chart/helper/LargeLineDraw.js
import { __extends as _ } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { containStroke as g } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/contain/line.js";
import { containStroke as I } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/contain/quadratic.js";
import { getECData as S } from "../../util/innerStore.js";
import L from "../../visual/tokens.js";
import D from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import P from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/BoundingRect.js";
import w from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Path.js";
var A = (
  /** @class */
  /* @__PURE__ */ (function() {
    function v() {
      this.polyline = !1, this.curveness = 0, this.segs = [];
    }
    return v;
  })()
), C = (
  /** @class */
  (function(v) {
    _(e, v);
    function e(o) {
      var n = v.call(this, o) || this;
      return n._off = 0, n.hoverDataIdx = -1, n;
    }
    return e.prototype.reset = function() {
      this.notClear = !1, this._off = 0;
    }, e.prototype.getDefaultStyle = function() {
      return {
        stroke: L.color.neutral99,
        fill: null
      };
    }, e.prototype.getDefaultShape = function() {
      return new A();
    }, e.prototype.buildPath = function(o, n) {
      var t = n.segs, a = n.curveness, r;
      if (n.polyline)
        for (r = this._off; r < t.length; ) {
          var f = t[r++];
          if (f > 0) {
            o.moveTo(t[r++], t[r++]);
            for (var s = 1; s < f; s++)
              o.lineTo(t[r++], t[r++]);
          }
        }
      else
        for (r = this._off; r < t.length; ) {
          var i = t[r++], u = t[r++], h = t[r++], l = t[r++];
          if (o.moveTo(i, u), a > 0) {
            var c = (i + h) / 2 - (u - l) * a, p = (u + l) / 2 - (h - i) * a;
            o.quadraticCurveTo(c, p, h, l);
          } else
            o.lineTo(h, l);
        }
      this.incremental && (this._off = r, this.notClear = !0);
    }, e.prototype.findDataIndex = function(o, n) {
      var t = this.shape, a = t.segs, r = t.curveness, f = this.style.lineWidth;
      if (t.polyline)
        for (var s = 0, i = 0; i < a.length; ) {
          var u = a[i++];
          if (u > 0)
            for (var h = a[i++], l = a[i++], c = 1; c < u; c++) {
              var p = a[i++], d = a[i++];
              if (g(h, l, p, d, f, o, n))
                return s;
            }
          s++;
        }
      else
        for (var s = 0, i = 0; i < a.length; ) {
          var h = a[i++], l = a[i++], p = a[i++], d = a[i++];
          if (r > 0) {
            var m = (h + p) / 2 - (l - d) * r, y = (l + d) / 2 - (p - h) * r;
            if (I(h, l, m, y, p, d, f, o, n))
              return s;
          } else if (g(h, l, p, d, f, o, n))
            return s;
          s++;
        }
      return -1;
    }, e.prototype.contain = function(o, n) {
      var t = this.transformCoordToLocal(o, n), a = this.getBoundingRect();
      if (o = t[0], n = t[1], a.contain(o, n)) {
        var r = this.hoverDataIdx = this.findDataIndex(o, n);
        return r >= 0;
      }
      return this.hoverDataIdx = -1, !1;
    }, e.prototype.getBoundingRect = function() {
      var o = this._rect;
      if (!o) {
        for (var n = this.shape, t = n.segs, a = 1 / 0, r = 1 / 0, f = -1 / 0, s = -1 / 0, i = 0; i < t.length; ) {
          var u = t[i++], h = t[i++];
          a = Math.min(u, a), f = Math.max(u, f), r = Math.min(h, r), s = Math.max(h, s);
        }
        o = this._rect = new P(a, r, f, s);
      }
      return o;
    }, e;
  })(w)
), q = (
  /** @class */
  (function() {
    function v() {
      this.group = new D();
    }
    return v.prototype.updateData = function(e) {
      this._clear();
      var o = this._create();
      o.setShape({
        segs: e.getLayout("linesPoints")
      }), this._setCommon(o, e);
    }, v.prototype.incrementalPrepareUpdate = function(e) {
      this.group.removeAll(), this._clear();
    }, v.prototype.incrementalUpdate = function(e, o) {
      var n = this._newAdded[0], t = o.getLayout("linesPoints"), a = n && n.shape.segs;
      if (a && a.length < 2e4) {
        var r = a.length, f = new Float32Array(r + t.length);
        f.set(a), f.set(t, r), n.setShape({
          segs: f
        });
      } else {
        this._newAdded = [];
        var s = this._create();
        s.incremental = !0, s.setShape({
          segs: t
        }), this._setCommon(s, o), s.__startIndex = e.start;
      }
    }, v.prototype.remove = function() {
      this._clear();
    }, v.prototype.eachRendered = function(e) {
      this._newAdded[0] && e(this._newAdded[0]);
    }, v.prototype._create = function() {
      var e = new C({
        cursor: "default",
        ignoreCoarsePointer: !0
      });
      return this._newAdded.push(e), this.group.add(e), e;
    }, v.prototype._setCommon = function(e, o, n) {
      var t = o.hostModel;
      e.setShape({
        polyline: t.get("polyline"),
        curveness: t.get(["lineStyle", "curveness"])
      }), e.useStyle(t.getModel("lineStyle").getLineStyle()), e.style.strokeNoScale = !0;
      var a = o.getVisual("style");
      a && a.stroke && e.setStyle("stroke", a.stroke), e.setStyle("fill", null);
      var r = S(e);
      r.seriesIndex = t.seriesIndex, e.on("mousemove", function(f) {
        r.dataIndex = null;
        var s = e.hoverDataIdx;
        s > 0 && (r.dataIndex = s + e.__startIndex);
      });
    }, v.prototype._clear = function() {
      this._newAdded = [], this.group.removeAll();
    }, v;
  })()
);
export {
  q as default
};
