// echarts@6.0.0/lib/chart/helper/LargeSymbolDraw.js
import { __extends as g } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { createSymbol as S } from "../../util/symbol.js";
import { getECData as _ } from "../../util/innerStore.js";
import y from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import m from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/BoundingRect.js";
import I from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Path.js";
var l = 4, x = (
  /** @class */
  /* @__PURE__ */ (function() {
    function p() {
    }
    return p;
  })()
), C = (
  /** @class */
  (function(p) {
    g(t, p);
    function t(r) {
      var e = p.call(this, r) || this;
      return e._off = 0, e.hoverDataIdx = -1, e;
    }
    return t.prototype.getDefaultShape = function() {
      return new x();
    }, t.prototype.reset = function() {
      this.notClear = !1, this._off = 0;
    }, t.prototype.buildPath = function(r, e) {
      var a = e.points, o = e.size, h = this.symbolProxy, i = h.shape, s = r.getContext ? r.getContext() : r, n = s && o[0] < l, u = this.softClipShape, f;
      if (n) {
        this._ctx = s;
        return;
      }
      for (this._ctx = null, f = this._off; f < a.length; ) {
        var v = a[f++], c = a[f++];
        isNaN(v) || isNaN(c) || u && !u.contain(v, c) || (i.x = v - o[0] / 2, i.y = c - o[1] / 2, i.width = o[0], i.height = o[1], h.buildPath(r, i, !0));
      }
      this.incremental && (this._off = f, this.notClear = !0);
    }, t.prototype.afterBrush = function() {
      var r = this.shape, e = r.points, a = r.size, o = this._ctx, h = this.softClipShape, i;
      if (o) {
        for (i = this._off; i < e.length; ) {
          var s = e[i++], n = e[i++];
          isNaN(s) || isNaN(n) || h && !h.contain(s, n) || o.fillRect(s - a[0] / 2, n - a[1] / 2, a[0], a[1]);
        }
        this.incremental && (this._off = i, this.notClear = !0);
      }
    }, t.prototype.findDataIndex = function(r, e) {
      for (var a = this.shape, o = a.points, h = a.size, i = Math.max(h[0], 4), s = Math.max(h[1], 4), n = o.length / 2 - 1; n >= 0; n--) {
        var u = n * 2, f = o[u] - i / 2, v = o[u + 1] - s / 2;
        if (r >= f && e >= v && r <= f + i && e <= v + s)
          return n;
      }
      return -1;
    }, t.prototype.contain = function(r, e) {
      var a = this.transformCoordToLocal(r, e), o = this.getBoundingRect();
      if (r = a[0], e = a[1], o.contain(r, e)) {
        var h = this.hoverDataIdx = this.findDataIndex(r, e);
        return h >= 0;
      }
      return this.hoverDataIdx = -1, !1;
    }, t.prototype.getBoundingRect = function() {
      var r = this._rect;
      if (!r) {
        for (var e = this.shape, a = e.points, o = e.size, h = o[0], i = o[1], s = 1 / 0, n = 1 / 0, u = -1 / 0, f = -1 / 0, v = 0; v < a.length; ) {
          var c = a[v++], d = a[v++];
          s = Math.min(c, s), u = Math.max(c, u), n = Math.min(d, n), f = Math.max(d, f);
        }
        r = this._rect = new m(s - h / 2, n - i / 2, u - s + h, f - n + i);
      }
      return r;
    }, t;
  })(I)
), M = (
  /** @class */
  (function() {
    function p() {
      this.group = new y();
    }
    return p.prototype.updateData = function(t, r) {
      this._clear();
      var e = this._create();
      e.setShape({
        points: t.getLayout("points")
      }), this._setCommon(e, t, r);
    }, p.prototype.updateLayout = function(t) {
      var r = t.getLayout("points");
      this.group.eachChild(function(e) {
        if (e.startIndex != null) {
          var a = (e.endIndex - e.startIndex) * 2, o = e.startIndex * 4 * 2;
          r = new Float32Array(r.buffer, o, a);
        }
        e.setShape("points", r), e.reset();
      });
    }, p.prototype.incrementalPrepareUpdate = function(t) {
      this._clear();
    }, p.prototype.incrementalUpdate = function(t, r, e) {
      var a = this._newAdded[0], o = r.getLayout("points"), h = a && a.shape.points;
      if (h && h.length < 2e4) {
        var i = h.length, s = new Float32Array(i + o.length);
        s.set(h), s.set(o, i), a.endIndex = t.end, a.setShape({
          points: s
        });
      } else {
        this._newAdded = [];
        var n = this._create();
        n.startIndex = t.start, n.endIndex = t.end, n.incremental = !0, n.setShape({
          points: o
        }), this._setCommon(n, r, e);
      }
    }, p.prototype.eachRendered = function(t) {
      this._newAdded[0] && t(this._newAdded[0]);
    }, p.prototype._create = function() {
      var t = new C({
        cursor: "default"
      });
      return t.ignoreCoarsePointer = !0, this.group.add(t), this._newAdded.push(t), t;
    }, p.prototype._setCommon = function(t, r, e) {
      var a = r.hostModel;
      e = e || {};
      var o = r.getVisual("symbolSize");
      t.setShape("size", o instanceof Array ? o : [o, o]), t.softClipShape = e.clipShape || null, t.symbolProxy = S(r.getVisual("symbol"), 0, 0, 0, 0), t.setColor = t.symbolProxy.setColor;
      var h = t.shape.size[0] < l;
      t.useStyle(
        // Draw shadow when doing fillRect is extremely slow.
        a.getModel("itemStyle").getItemStyle(h ? ["color", "shadowBlur", "shadowColor"] : ["color"])
      );
      var i = r.getVisual("style"), s = i && i.fill;
      s && t.setColor(s);
      var n = _(t);
      n.seriesIndex = a.seriesIndex, t.on("mousemove", function(u) {
        n.dataIndex = null;
        var f = t.hoverDataIdx;
        f >= 0 && (n.dataIndex = f + (t.startIndex || 0));
      });
    }, p.prototype.remove = function() {
      this._clear();
    }, p.prototype._clear = function() {
      this._newAdded = [], this.group.removeAll();
    }, p;
  })()
);
export {
  M as default
};
