// echarts@6.0.0/lib/chart/helper/LinePath.js
import { __extends as l } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { normalize as h } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/vector.js";
import s from "../../visual/tokens.js";
import f from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Line.js";
import c from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/BezierCurve.js";
import y from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Path.js";
var u = f.prototype, n = c.prototype, p = (
  /** @class */
  /* @__PURE__ */ (function() {
    function r() {
      this.x1 = 0, this.y1 = 0, this.x2 = 0, this.y2 = 0, this.percent = 1;
    }
    return r;
  })()
);
(function(r) {
  l(i, r);
  function i() {
    return r !== null && r.apply(this, arguments) || this;
  }
  return i;
})(p);
function o(r) {
  return isNaN(+r.cpx1) || isNaN(+r.cpy1);
}
var x = (
  /** @class */
  (function(r) {
    l(i, r);
    function i(e) {
      var t = r.call(this, e) || this;
      return t.type = "ec-line", t;
    }
    return i.prototype.getDefaultStyle = function() {
      return {
        stroke: s.color.neutral99,
        fill: null
      };
    }, i.prototype.getDefaultShape = function() {
      return new p();
    }, i.prototype.buildPath = function(e, t) {
      o(t) ? u.buildPath.call(this, e, t) : n.buildPath.call(this, e, t);
    }, i.prototype.pointAt = function(e) {
      return o(this.shape) ? u.pointAt.call(this, e) : n.pointAt.call(this, e);
    }, i.prototype.tangentAt = function(e) {
      var t = this.shape, a = o(t) ? [t.x2 - t.x1, t.y2 - t.y1] : n.tangentAt.call(this, e);
      return h(a, a);
    }, i;
  })(y)
);
export {
  x as default
};
