// echarts@6.0.0/lib/chart/helper/Line.js
import { __extends as K } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { each as z, isArray as Y, retrieve2 as G } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { sub as X, normalize as H } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/vector.js";
import { normalizeSymbolSize as C, normalizeSymbolOffset as F, createSymbol as U } from "../../util/symbol.js";
import $ from "./LinePath.js";
import { SPECIAL_STATES as A, toggleHoverEmphasis as j, enterEmphasis as q, leaveEmphasis as J } from "../../util/states.js";
import { getLabelStatesModels as Q, setLabelStyle as W } from "../../label/labelStyle.js";
import { round as Z } from "../../util/number.js";
import x from "../../visual/tokens.js";
import D from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import { initProps as ee, updateProps as te } from "../../animation/basicTransition.js";
var E = ["fromSymbol", "toSymbol"];
function w(a) {
  return "_" + a + "Type";
}
function V(a, i, t) {
  var o = i.getItemVisual(t, a);
  if (!o || o === "none")
    return o;
  var r = i.getItemVisual(t, a + "Size"), e = i.getItemVisual(t, a + "Rotate"), l = i.getItemVisual(t, a + "Offset"), v = i.getItemVisual(t, a + "KeepAspect"), s = C(r), n = F(l || 0, s);
  return o + s + n + (e || "") + (v || "");
}
function B(a, i, t) {
  var o = i.getItemVisual(t, a);
  if (!(!o || o === "none")) {
    var r = i.getItemVisual(t, a + "Size"), e = i.getItemVisual(t, a + "Rotate"), l = i.getItemVisual(t, a + "Offset"), v = i.getItemVisual(t, a + "KeepAspect"), s = C(r), n = F(l || 0, s), m = U(o, -s[0] / 2 + n[0], -s[1] / 2 + n[1], s[0], s[1], null, v);
    return m.__specifiedRotation = e == null || isNaN(e) ? void 0 : +e * Math.PI / 180 || 0, m.name = a, m;
  }
}
function oe(a) {
  var i = new $({
    name: "line",
    subPixelOptimize: !0
  });
  return R(i.shape, a), i;
}
function R(a, i) {
  a.x1 = i[0][0], a.y1 = i[0][1], a.x2 = i[1][0], a.y2 = i[1][1], a.percent = 1;
  var t = i[2];
  t ? (a.cpx1 = t[0], a.cpy1 = t[1]) : (a.cpx1 = NaN, a.cpy1 = NaN);
}
var ce = (
  /** @class */
  (function(a) {
    K(i, a);
    function i(t, o, r) {
      var e = a.call(this) || this;
      return e._createLine(t, o, r), e;
    }
    return i.prototype._createLine = function(t, o, r) {
      var e = t.hostModel, l = t.getItemLayout(o), v = t.getItemVisual(o, "z2"), s = oe(l);
      s.shape.percent = 0, ee(s, {
        z2: G(v, 0),
        shape: {
          percent: 1
        }
      }, e, o), this.add(s), z(E, function(n) {
        var m = B(n, t, o);
        this.add(m), this[w(n)] = V(n, t, o);
      }, this), this._updateCommonStl(t, o, r);
    }, i.prototype.updateData = function(t, o, r) {
      var e = t.hostModel, l = this.childOfName("line"), v = t.getItemLayout(o), s = {
        shape: {}
      };
      R(s.shape, v), te(l, s, e, o), z(E, function(n) {
        var m = V(n, t, o), u = w(n);
        if (this[u] !== m) {
          this.remove(this.childOfName(n));
          var f = B(n, t, o);
          this.add(f);
        }
        this[u] = m;
      }, this), this._updateCommonStl(t, o, r);
    }, i.prototype.getLinePath = function() {
      return this.childAt(0);
    }, i.prototype._updateCommonStl = function(t, o, r) {
      var e = t.hostModel, l = this.childOfName("line"), v = r && r.emphasisLineStyle, s = r && r.blurLineStyle, n = r && r.selectLineStyle, m = r && r.labelStatesModels, u = r && r.emphasisDisabled, f = r && r.focus, P = r && r.blurScope;
      if (!r || t.hasItemOption) {
        var y = t.getItemModel(o), h = y.getModel("emphasis");
        v = h.getModel("lineStyle").getLineStyle(), s = y.getModel(["blur", "lineStyle"]).getLineStyle(), n = y.getModel(["select", "lineStyle"]).getLineStyle(), u = h.get("disabled"), f = h.get("focus"), P = h.get("blurScope"), m = Q(y);
      }
      var L = t.getItemVisual(o, "style"), g = L.stroke;
      l.useStyle(L), l.style.fill = null, l.style.strokeNoScale = !0, l.ensureState("emphasis").style = v, l.ensureState("blur").style = s, l.ensureState("select").style = n, z(E, function(O) {
        var p = this.childOfName(O);
        if (p) {
          p.setColor(g), p.style.opacity = L.opacity;
          for (var I = 0; I < A.length; I++) {
            var c = A[I], k = l.getState(c);
            if (k) {
              var T = k.style || {}, M = p.ensureState(c), N = M.style || (M.style = {});
              T.stroke != null && (N[p.__isEmptyBrush ? "stroke" : "fill"] = T.stroke), T.opacity != null && (N.opacity = T.opacity);
            }
          }
          p.markRedraw();
        }
      }, this);
      var S = e.getRawValue(o);
      W(this, m, {
        labelDataIndex: o,
        labelFetcher: {
          getFormattedLabel: function(O, p) {
            return e.getFormattedLabel(O, p, t.dataType);
          }
        },
        inheritColor: g || x.color.neutral99,
        defaultOpacity: L.opacity,
        defaultText: (S == null ? t.getName(o) : isFinite(S) ? Z(S) : S) + ""
      });
      var b = this.getTextContent();
      if (b) {
        var _ = m.normal;
        b.__align = b.style.align, b.__verticalAlign = b.style.verticalAlign, b.__position = _.get("position") || "middle";
        var d = _.get("distance");
        Y(d) || (d = [d, d]), b.__labelDistance = d;
      }
      this.setTextConfig({
        position: null,
        local: !0,
        inside: !1
        // Can't be inside for stroke element.
      }), j(this, f, P, u);
    }, i.prototype.highlight = function() {
      q(this);
    }, i.prototype.downplay = function() {
      J(this);
    }, i.prototype.updateLayout = function(t, o) {
      this.setLinePoints(t.getItemLayout(o));
    }, i.prototype.setLinePoints = function(t) {
      var o = this.childOfName("line");
      R(o.shape, t), o.dirty();
    }, i.prototype.beforeUpdate = function() {
      var t = this, o = t.childOfName("fromSymbol"), r = t.childOfName("toSymbol"), e = t.getTextContent();
      if (!o && !r && (!e || e.ignore))
        return;
      for (var l = 1, v = this.parent; v; )
        v.scaleX && (l /= v.scaleX), v = v.parent;
      var s = t.childOfName("line");
      if (!this.__dirty && !s.__dirty)
        return;
      var n = s.shape.percent, m = s.pointAt(0), u = s.pointAt(n), f = X([], u, m);
      H(f, f);
      function P(k, T) {
        var M = k.__specifiedRotation;
        if (M == null) {
          var N = s.tangentAt(T);
          k.attr("rotation", (T === 1 ? -1 : 1) * Math.PI / 2 - Math.atan2(N[1], N[0]));
        } else
          k.attr("rotation", M);
      }
      if (o && (o.setPosition(m), P(o, 0), o.scaleX = o.scaleY = l * n, o.markRedraw()), r && (r.setPosition(u), P(r, 1), r.scaleX = r.scaleY = l * n, r.markRedraw()), e && !e.ignore) {
        e.x = e.y = 0, e.originX = e.originY = 0;
        var y = void 0, h = void 0, L = e.__labelDistance, g = L[0] * l, S = L[1] * l, b = n / 2, _ = s.tangentAt(b), d = [_[1], -_[0]], O = s.pointAt(b);
        d[1] > 0 && (d[0] = -d[0], d[1] = -d[1]);
        var p = _[0] < 0 ? -1 : 1;
        if (e.__position !== "start" && e.__position !== "end") {
          var I = -Math.atan2(_[1], _[0]);
          u[0] < m[0] && (I = Math.PI + I), e.rotation = I;
        }
        var c = void 0;
        switch (e.__position) {
          case "insideStartTop":
          case "insideMiddleTop":
          case "insideEndTop":
          case "middle":
            c = -S, h = "bottom";
            break;
          case "insideStartBottom":
          case "insideMiddleBottom":
          case "insideEndBottom":
            c = S, h = "top";
            break;
          default:
            c = 0, h = "middle";
        }
        switch (e.__position) {
          case "end":
            e.x = f[0] * g + u[0], e.y = f[1] * S + u[1], y = f[0] > 0.8 ? "left" : f[0] < -0.8 ? "right" : "center", h = f[1] > 0.8 ? "top" : f[1] < -0.8 ? "bottom" : "middle";
            break;
          case "start":
            e.x = -f[0] * g + m[0], e.y = -f[1] * S + m[1], y = f[0] > 0.8 ? "right" : f[0] < -0.8 ? "left" : "center", h = f[1] > 0.8 ? "bottom" : f[1] < -0.8 ? "top" : "middle";
            break;
          case "insideStartTop":
          case "insideStart":
          case "insideStartBottom":
            e.x = g * p + m[0], e.y = m[1] + c, y = _[0] < 0 ? "right" : "left", e.originX = -g * p, e.originY = -c;
            break;
          case "insideMiddleTop":
          case "insideMiddle":
          case "insideMiddleBottom":
          case "middle":
            e.x = O[0], e.y = O[1] + c, y = "center", e.originY = -c;
            break;
          case "insideEndTop":
          case "insideEnd":
          case "insideEndBottom":
            e.x = -g * p + u[0], e.y = u[1] + c, y = _[0] >= 0 ? "right" : "left", e.originX = g * p, e.originY = -c;
            break;
        }
        e.scaleX = e.scaleY = l, e.setStyle({
          // Use the user specified text align and baseline first
          verticalAlign: e.__verticalAlign || h,
          align: e.__align || y
        });
      }
    }, i;
  })(D)
);
export {
  ce as default
};
