// echarts@6.0.0/lib/chart/helper/LineDraw.js
import { traverseElements as h } from "../../util/graphic.js";
import d from "./Line.js";
import { getLabelStatesModels as f } from "../../label/labelStyle.js";
import m from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
var _ = (
  /** @class */
  (function() {
    function r(e) {
      this.group = new m(), this._LineCtor = e || d;
    }
    return r.prototype.updateData = function(e) {
      var t = this;
      this._progressiveEls = null;
      var s = this, o = s.group, n = s._lineData;
      s._lineData = e, n || o.removeAll();
      var i = p(e);
      e.diff(n).add(function(a) {
        t._doAdd(e, a, i);
      }).update(function(a, c) {
        t._doUpdate(n, e, c, a, i);
      }).remove(function(a) {
        o.remove(n.getItemGraphicEl(a));
      }).execute();
    }, r.prototype.updateLayout = function() {
      var e = this._lineData;
      e && e.eachItemGraphicEl(function(t, s) {
        t.updateLayout(e, s);
      }, this);
    }, r.prototype.incrementalPrepareUpdate = function(e) {
      this._seriesScope = p(e), this._lineData = null, this.group.removeAll();
    }, r.prototype.incrementalUpdate = function(e, t) {
      this._progressiveEls = [];
      function s(a) {
        !a.isGroup && !g(a) && (a.incremental = !0, a.ensureState("emphasis").hoverLayer = !0);
      }
      for (var o = e.start; o < e.end; o++) {
        var n = t.getItemLayout(o);
        if (u(n)) {
          var i = new this._LineCtor(t, o, this._seriesScope);
          i.traverse(s), this.group.add(i), t.setItemGraphicEl(o, i), this._progressiveEls.push(i);
        }
      }
    }, r.prototype.remove = function() {
      this.group.removeAll();
    }, r.prototype.eachRendered = function(e) {
      h(this._progressiveEls || this.group, e);
    }, r.prototype._doAdd = function(e, t, s) {
      var o = e.getItemLayout(t);
      if (u(o)) {
        var n = new this._LineCtor(e, t, s);
        e.setItemGraphicEl(t, n), this.group.add(n);
      }
    }, r.prototype._doUpdate = function(e, t, s, o, n) {
      var i = e.getItemGraphicEl(s);
      if (!u(t.getItemLayout(o))) {
        this.group.remove(i);
        return;
      }
      i ? i.updateData(t, o, n) : i = new this._LineCtor(t, o, n), t.setItemGraphicEl(o, i), this.group.add(i);
    }, r;
  })()
);
function g(r) {
  return r.animators && r.animators.length > 0;
}
function p(r) {
  var e = r.hostModel, t = e.getModel("emphasis");
  return {
    lineStyle: e.getModel("lineStyle").getLineStyle(),
    emphasisLineStyle: t.getModel(["lineStyle"]).getLineStyle(),
    blurLineStyle: e.getModel(["blur", "lineStyle"]).getLineStyle(),
    selectLineStyle: e.getModel(["select", "lineStyle"]).getLineStyle(),
    emphasisDisabled: t.get("disabled"),
    blurScope: t.get("blurScope"),
    focus: t.get("focus"),
    labelStatesModels: f(e)
  };
}
function l(r) {
  return isNaN(r[0]) || isNaN(r[1]);
}
function u(r) {
  return r && !l(r[0]) && !l(r[1]);
}
export {
  _ as default
};
