// echarts@6.0.0/lib/chart/helper/multipleGraphEdgeHelper.js
import { isArray as E, isNumber as w } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
var y = "-->", p = function(r) {
  return r.get("autoCurveness") || null;
}, d = function(r, n) {
  var t = p(r), u = 20, a = [];
  if (w(t))
    u = t;
  else if (E(t)) {
    r.__curvenessList = t;
    return;
  }
  n > u && (u = n);
  var v = u % 2 ? u + 2 : u + 3;
  a = [];
  for (var e = 0; e < v; e++)
    a.push((e % 2 ? e + 1 : e) / 10 * (e % 2 ? -1 : 1));
  r.__curvenessList = a;
}, s = function(r, n, t) {
  var u = [r.id, r.dataIndex].join("."), a = [n.id, n.dataIndex].join(".");
  return [t.uid, u, a].join(y);
}, m = function(r) {
  var n = r.split(y);
  return [n[0], n[2], n[1]].join(y);
}, I = function(r, n) {
  var t = s(r.node1, r.node2, n);
  return n.__edgeMap[t];
}, K = function(r, n) {
  var t = L(s(r.node1, r.node2, n), n), u = L(s(r.node2, r.node1, n), n);
  return t + u;
}, L = function(r, n) {
  var t = n.__edgeMap;
  return t[r] ? t[r].length : 0;
};
function x(r) {
  p(r) && (r.__curvenessList = [], r.__edgeMap = {}, d(r));
}
function A(r, n, t, u) {
  if (p(t)) {
    var a = s(r, n, t), v = t.__edgeMap, e = v[m(a)];
    v[a] && !e ? v[a].isForward = !0 : e && v[a] && (e.isForward = !0, v[a].isForward = !1), v[a] = v[a] || [], v[a].push(u);
  }
}
function P(r, n, t, u) {
  var a = p(n), v = E(a);
  if (!a)
    return null;
  var e = I(r, n);
  if (!e)
    return null;
  for (var c = -1, g = 0; g < e.length; g++)
    if (e[g] === t) {
      c = g;
      break;
    }
  var C = K(r, n);
  d(n, C), r.lineStyle = r.lineStyle || {};
  var h = s(r.node1, r.node2, n), _ = n.__curvenessList, i = v || C % 2 ? 0 : 1;
  if (e.isForward)
    return _[i + c];
  var F = m(h), f = L(F, n), o = _[c + f + i];
  return u ? v ? a && a[0] === 0 ? (f + i) % 2 ? o : -o : ((f % 2 ? 0 : 1) + i) % 2 ? o : -o : (f + i) % 2 ? o : -o : _[c + f + i];
}
export {
  A as createEdgeMapForCurveness,
  P as getCurvenessForEdge,
  x as initCurvenessList
};
