// echarts@6.0.0/lib/chart/helper/Symbol.js
import { __extends as P } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { createSymbol as E, normalizeSymbolOffset as Z, normalizeSymbolSize as w } from "../../util/symbol.js";
import { getECData as x } from "../../util/innerStore.js";
import { enterEmphasis as N, leaveEmphasis as R, toggleHoverEmphasis as F } from "../../util/states.js";
import { getDefaultLabel as k } from "./labelHelper.js";
import { retrieve2 as B, extend as D } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { getLabelStatesModels as G, setLabelStyle as H } from "../../label/labelStyle.js";
import K from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Image.js";
import { updateProps as $, saveOldStyle as j, initProps as q, removeElement as T } from "../../animation/basicTransition.js";
import J from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
var mt = (
  /** @class */
  (function(S) {
    P(i, S);
    function i(t, e, o, l) {
      var s = S.call(this) || this;
      return s.updateData(t, e, o, l), s;
    }
    return i.prototype._createSymbol = function(t, e, o, l, s, a) {
      this.removeAll();
      var m = E(t, -1, -1, 2, 2, null, a);
      m.attr({
        z2: B(s, 100),
        culling: !0,
        scaleX: l[0] / 2,
        scaleY: l[1] / 2
      }), m.drift = Q, this._symbolType = t, this.add(m);
    }, i.prototype.stopSymbolAnimation = function(t) {
      this.childAt(0).stopAnimation(null, t);
    }, i.prototype.getSymbolType = function() {
      return this._symbolType;
    }, i.prototype.getSymbolPath = function() {
      return this.childAt(0);
    }, i.prototype.highlight = function() {
      N(this.childAt(0));
    }, i.prototype.downplay = function() {
      R(this.childAt(0));
    }, i.prototype.setZ = function(t, e) {
      var o = this.childAt(0);
      o.zlevel = t, o.z = e;
    }, i.prototype.setDraggable = function(t, e) {
      var o = this.childAt(0);
      o.draggable = t, o.cursor = !e && t ? "move" : o.cursor;
    }, i.prototype.updateData = function(t, e, o, l) {
      this.silent = !1;
      var s = t.getItemVisual(e, "symbol") || "circle", a = t.hostModel, m = i.getSymbolSize(t, e), c = i.getSymbolZ2(t, e), h = s !== this._symbolType, y = l && l.disableAnimation;
      if (h) {
        var f = t.getItemVisual(e, "symbolKeepAspect");
        this._createSymbol(s, t, e, m, c, f);
      } else {
        var r = this.childAt(0);
        r.silent = !1;
        var n = {
          scaleX: m[0] / 2,
          scaleY: m[1] / 2
        };
        y ? r.attr(n) : $(r, n, a, e), j(r);
      }
      if (this._updateCommon(t, e, m, o, l), h) {
        var r = this.childAt(0);
        if (!y) {
          var n = {
            scaleX: this._sizeX,
            scaleY: this._sizeY,
            style: {
              // Always fadeIn. Because it has fadeOut animation when symbol is removed..
              opacity: r.style.opacity
            }
          };
          r.scaleX = r.scaleY = 0, r.style.opacity = 0, q(r, n, a, e);
        }
      }
      y && this.childAt(0).stopAnimation("leave");
    }, i.prototype._updateCommon = function(t, e, o, l, s) {
      var a = this.childAt(0), m = t.hostModel, c, h, y, f, r, n, I, u, d;
      if (l && (c = l.emphasisItemStyle, h = l.blurItemStyle, y = l.selectItemStyle, f = l.focus, r = l.blurScope, I = l.labelStatesModels, u = l.hoverScale, d = l.cursorStyle, n = l.emphasisDisabled), !l || t.hasItemOption) {
        var v = l && l.itemModel ? l.itemModel : t.getItemModel(e), b = v.getModel("emphasis");
        c = b.getModel("itemStyle").getItemStyle(), y = v.getModel(["select", "itemStyle"]).getItemStyle(), h = v.getModel(["blur", "itemStyle"]).getItemStyle(), f = b.get("focus"), r = b.get("blurScope"), n = b.get("disabled"), I = G(v), u = b.getShallow("scale"), d = v.getShallow("cursor");
      }
      var V = t.getItemVisual(e, "symbolRotate");
      a.attr("rotation", (V || 0) * Math.PI / 180 || 0);
      var z = Z(t.getItemVisual(e, "symbolOffset"), o);
      z && (a.x = z[0], a.y = z[1]), d && a.attr("cursor", d);
      var p = t.getItemVisual(e, "style"), O = p.fill;
      if (a instanceof K) {
        var g = a.style;
        a.useStyle(D({
          // TODO other properties like x, y ?
          image: g.image,
          x: g.x,
          y: g.y,
          width: g.width,
          height: g.height
        }, p));
      } else
        a.__isEmptyBrush ? a.useStyle(D({}, p)) : a.useStyle(p), a.style.decal = null, a.setColor(O, s && s.symbolInnerColor), a.style.strokeNoScale = !0;
      var Y = t.getItemVisual(e, "liftZ"), _ = this._z2;
      Y != null ? _ == null && (this._z2 = a.z2, a.z2 += Y) : _ != null && (a.z2 = _, this._z2 = null);
      var L = s && s.useNameLabel;
      H(a, I, {
        labelFetcher: m,
        labelDataIndex: e,
        defaultText: M,
        inheritColor: O,
        defaultOpacity: p.opacity
      });
      function M(C) {
        return L ? t.getName(C) : k(t, C);
      }
      this._sizeX = o[0] / 2, this._sizeY = o[1] / 2;
      var A = a.ensureState("emphasis");
      A.style = c, a.ensureState("select").style = y, a.ensureState("blur").style = h;
      var X = u == null || u === !0 ? Math.max(1.1, 3 / this._sizeY) : isFinite(u) && u > 0 ? +u : 1;
      A.scaleX = this._sizeX * X, A.scaleY = this._sizeY * X, this.setSymbolScale(1), F(this, f, r, n);
    }, i.prototype.setSymbolScale = function(t) {
      this.scaleX = this.scaleY = t;
    }, i.prototype.fadeOut = function(t, e, o) {
      var l = this.childAt(0), s = x(this).dataIndex, a = o && o.animation;
      if (this.silent = l.silent = !0, o && o.fadeLabel) {
        var m = l.getTextContent();
        m && T(m, {
          style: {
            opacity: 0
          }
        }, e, {
          dataIndex: s,
          removeOpt: a,
          cb: function() {
            l.removeTextContent();
          }
        });
      } else
        l.removeTextContent();
      T(l, {
        style: {
          opacity: 0
        },
        scaleX: 0,
        scaleY: 0
      }, e, {
        dataIndex: s,
        cb: t,
        removeOpt: a
      });
    }, i.getSymbolSize = function(t, e) {
      return w(t.getItemVisual(e, "symbolSize"));
    }, i.getSymbolZ2 = function(t, e) {
      return t.getItemVisual(e, "z2");
    }, i;
  })(J)
);
function Q(S, i) {
  this.parent.drift(S, i);
}
export {
  mt as default
};
