// echarts@6.0.0/lib/chart/helper/SymbolDraw.js
import { traverseElements as _ } from "../../util/graphic.js";
import I from "./Symbol.js";
import { isObject as M } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { getLabelStatesModels as E } from "../../label/labelStyle.js";
import G from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import { updateProps as P } from "../../animation/basicTransition.js";
function f(o, e, t, r) {
  return e && !isNaN(e[0]) && !isNaN(e[1]) && !(r.isIgnore && r.isIgnore(t)) && !(r.clipShape && !r.clipShape.contain(e[0], e[1])) && o.getItemVisual(t, "symbol") !== "none";
}
function S(o) {
  return o != null && !M(o) && (o = {
    isIgnore: o
  }), o || {};
}
function b(o) {
  var e = o.hostModel, t = e.getModel("emphasis");
  return {
    emphasisItemStyle: t.getModel("itemStyle").getItemStyle(),
    blurItemStyle: e.getModel(["blur", "itemStyle"]).getItemStyle(),
    selectItemStyle: e.getModel(["select", "itemStyle"]).getItemStyle(),
    focus: t.get("focus"),
    blurScope: t.get("blurScope"),
    emphasisDisabled: t.get("disabled"),
    hoverScale: t.get("scale"),
    labelStatesModels: E(e),
    cursorStyle: e.get("cursor")
  };
}
var T = (
  /** @class */
  (function() {
    function o(e) {
      this.group = new G(), this._SymbolCtor = e || I;
    }
    return o.prototype.updateData = function(e, t) {
      this._progressiveEls = null, t = S(t);
      var r = this.group, l = e.hostModel, a = this._data, p = this._SymbolCtor, n = t.disableAnimation, u = b(e), h = {
        disableAnimation: n
      }, v = t.getSymbolPoint || function(s) {
        return e.getItemLayout(s);
      };
      a || r.removeAll(), e.diff(a).add(function(s) {
        var m = v(s);
        if (f(e, m, s, t)) {
          var i = new p(e, s, u, h);
          i.setPosition(m), e.setItemGraphicEl(s, i), r.add(i);
        }
      }).update(function(s, m) {
        var i = a.getItemGraphicEl(m), c = v(s);
        if (!f(e, c, s, t)) {
          r.remove(i);
          return;
        }
        var d = e.getItemVisual(s, "symbol") || "circle", y = i && i.getSymbolType && i.getSymbolType();
        if (!i || y && y !== d)
          r.remove(i), i = new p(e, s, u, h), i.setPosition(c);
        else {
          i.updateData(e, s, u, h);
          var g = {
            x: c[0],
            y: c[1]
          };
          n ? i.attr(g) : P(i, g, l);
        }
        r.add(i), e.setItemGraphicEl(s, i);
      }).remove(function(s) {
        var m = a.getItemGraphicEl(s);
        m && m.fadeOut(function() {
          r.remove(m);
        }, l);
      }).execute(), this._getSymbolPoint = v, this._data = e;
    }, o.prototype.updateLayout = function() {
      var e = this, t = this._data;
      t && t.eachItemGraphicEl(function(r, l) {
        var a = e._getSymbolPoint(l);
        r.setPosition(a), r.markRedraw();
      });
    }, o.prototype.incrementalPrepareUpdate = function(e) {
      this._seriesScope = b(e), this._data = null, this.group.removeAll();
    }, o.prototype.incrementalUpdate = function(e, t, r) {
      this._progressiveEls = [], r = S(r);
      function l(u) {
        u.isGroup || (u.incremental = !0, u.ensureState("emphasis").hoverLayer = !0);
      }
      for (var a = e.start; a < e.end; a++) {
        var p = t.getItemLayout(a);
        if (f(t, p, a, r)) {
          var n = new this._SymbolCtor(t, a, this._seriesScope);
          n.traverse(l), n.setPosition(p), this.group.add(n), t.setItemGraphicEl(a, n), this._progressiveEls.push(n);
        }
      }
    }, o.prototype.eachRendered = function(e) {
      _(this._progressiveEls || this.group, e);
    }, o.prototype.remove = function(e) {
      var t = this.group, r = this._data;
      r && e ? r.eachItemGraphicEl(function(l) {
        l.fadeOut(function() {
          t.remove(l);
        }, r.hostModel);
      }) : t.removeAll();
    }, o;
  })()
);
export {
  T as default
};
