// echarts@6.0.0/lib/chart/helper/whiskerBoxCommon.js
import I from "./createSeriesDataSimply.js";
import { each as E, isArray as f, extend as T, curry as B } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { getDimensionTypeByAxis as p } from "../../data/helper/dimensionHelper.js";
import { makeSeriesEncodeForAxisCoordSys as R } from "../../data/helper/sourceHelper.js";
var O = (
  /** @class */
  (function() {
    function s() {
    }
    return s.prototype._hasEncodeRule = function(e) {
      var a = this.getEncode();
      return a && a.get(e) != null;
    }, s.prototype.getInitialData = function(e, a) {
      var o, n = a.getComponent("xAxis", this.get("xAxisIndex")), l = a.getComponent("yAxis", this.get("yAxisIndex")), A = n.get("type"), g = l.get("type"), x;
      A === "category" ? (e.layout = "horizontal", o = n.getOrdinalMeta(), x = !this._hasEncodeRule("x")) : g === "category" ? (e.layout = "vertical", o = l.getOrdinalMeta(), x = !this._hasEncodeRule("y")) : e.layout = e.layout || "horizontal";
      var d = ["x", "y"], r = e.layout === "horizontal" ? 0 : 1, D = this._baseAxisDim = d[r], M = d[1 - r], u = [n, l], C = u[r].get("type"), b = u[1 - r].get("type"), m = e.data;
      if (m && x) {
        var y = [];
        E(m, function(i, c) {
          var t;
          f(i) ? (t = i.slice(), i.unshift(c)) : f(i.value) ? (t = T({}, i), t.value = t.value.slice(), i.value.unshift(c)) : t = i, y.push(t);
        }), e.data = y;
      }
      var h = this.defaultValueDimensions, v = [{
        name: D,
        type: p(C),
        ordinalMeta: o,
        otherDims: {
          tooltip: !1,
          itemName: 0
        },
        dimsDef: ["base"]
      }, {
        name: M,
        type: p(b),
        dimsDef: h.slice()
      }];
      return I(this, {
        coordDimensions: v,
        dimensionsCount: h.length + 1,
        encodeDefaulter: B(R, v, this)
      });
    }, s.prototype.getBaseAxis = function() {
      var e = this._baseAxisDim;
      return this.ecModel.getComponent(e + "Axis", this.get(e + "AxisIndex")).axis;
    }, s;
  })()
);
export {
  O as WhiskerBoxCommonMixin
};
