// motion@12.17.0/dist/es/framer-motion/dist/es/components/Reorder/Group.js
import { jsx as d } from "react/jsx-runtime";
import { forwardRef as x, useRef as g, useEffect as h } from "react";
import { ReorderContext as y } from "../../context/ReorderContext.js";
import { motion as C } from "../../render/components/motion/proxy.js";
import { useConstant as G } from "../../utils/use-constant.js";
import { checkReorder as O } from "./utils/check-reorder.js";
import { invariant as w } from "../../../../../motion-utils/dist/es/errors.js";
function I({ children: e, as: u = "ul", axis: f = "y", onReorder: p, values: c, ...s }, a) {
  const l = G(() => C[u]), r = [], m = g(!1);
  w(!!c, "Reorder.Group must be provided a values prop");
  const R = {
    axis: f,
    registerItem: (o, t) => {
      const n = r.findIndex((i) => o === i.value);
      n !== -1 ? r[n].layout = t[f] : r.push({ value: o, layout: t[f] }), r.sort(j);
    },
    updateOrder: (o, t, n) => {
      if (m.current)
        return;
      const i = O(r, o, t, n);
      r !== i && (m.current = !0, p(i.map(b).filter((v) => c.indexOf(v) !== -1)));
    }
  };
  return h(() => {
    m.current = !1;
  }), d(l, { ...s, ref: a, ignoreStrict: !0, children: d(y.Provider, { value: R, children: e }) });
}
const q = /* @__PURE__ */ x(I);
function b(e) {
  return e.value;
}
function j(e, u) {
  return e.layout.min - u.layout.min;
}
export {
  q as ReorderGroup,
  I as ReorderGroupComponent
};
