// motion@12.17.0/dist/es/framer-motion/dist/es/components/Reorder/Item.js
import { jsx as C } from "react/jsx-runtime";
import { forwardRef as I, useContext as M } from "react";
import { ReorderContext as V } from "../../context/ReorderContext.js";
import { motion as h } from "../../render/components/motion/proxy.js";
import { useConstant as O } from "../../utils/use-constant.js";
import { useMotionValue as S } from "../../value/use-motion-value.js";
import { useTransform as T } from "../../value/use-transform.js";
import { invariant as b } from "../../../../../motion-utils/dist/es/errors.js";
import { isMotionValue as j } from "../../../../../motion-dom/dist/es/value/utils/is-motion-value.js";
function f(e, o = 0) {
  return j(e) ? e : S(o);
}
function w({ children: e, style: o = {}, value: m, as: p = "li", onDrag: u, layout: c = !0, ...d }, x) {
  const l = O(() => h[p]), s = M(V), r = {
    x: f(o.x),
    y: f(o.y)
  }, y = T([r.x, r.y], ([t, i]) => t || i ? 1 : "unset");
  b(!!s, "Reorder.Item must be a child of Reorder.Group");
  const { axis: n, registerItem: R, updateOrder: g } = s;
  return C(l, { drag: n, ...d, dragSnapToOrigin: !0, style: { ...o, x: r.x, y: r.y, zIndex: y }, layout: c, onDrag: (t, i) => {
    const { velocity: a } = i;
    a[n] && g(m, r[n].get(), a[n]), u && u(t, i);
  }, onLayoutMeasure: (t) => R(m, t), ref: x, ignoreStrict: !0, children: e });
}
const E = /* @__PURE__ */ I(w);
export {
  E as ReorderItem,
  w as ReorderItemComponent
};
