// recharts@2.15.0/es6/cartesian/Line.js
import y, { PureComponent as it } from "react";
import at from "../../../../../react-smooth@4.0.4_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/react-smooth/es6/Animate.js";
import ot from "../../../../../../../_virtual/isFunction.js";
import V from "../../../../../../../_virtual/isNil.js";
import st from "../../../../../../../_virtual/isEqual.js";
import { clsx as G } from "../../../../../clsx@2.1.1/node_modules/clsx/dist/clsx.js";
import { Curve as lt } from "../shape/Curve.js";
import { Dot as ut } from "../shape/Dot.js";
import { Layer as z } from "../container/Layer.js";
import { LabelList as ct } from "../component/LabelList.js";
import { ErrorBar as pt } from "./ErrorBar.js";
import { interpolateNumber as N, uniqueId as ft } from "../util/DataUtils.js";
import { findAllByType as mt, filterProps as W, hasClipDot as ht } from "../util/ReactUtils.js";
import { Global as vt } from "../util/Global.js";
import { getValueByDataKey as Q, getCateCoordinateOfLine as H } from "../util/ChartUtils.js";
var yt = ["type", "layout", "connectNulls", "ref"], dt = ["key"];
function j(t) {
  "@babel/helpers - typeof";
  return j = typeof Symbol == "function" && typeof Symbol.iterator == "symbol" ? function(r) {
    return typeof r;
  } : function(r) {
    return r && typeof Symbol == "function" && r.constructor === Symbol && r !== Symbol.prototype ? "symbol" : typeof r;
  }, j(t);
}
function X(t, r) {
  if (t == null) return {};
  var n = gt(t, r), e, i;
  if (Object.getOwnPropertySymbols) {
    var o = Object.getOwnPropertySymbols(t);
    for (i = 0; i < o.length; i++)
      e = o[i], !(r.indexOf(e) >= 0) && Object.prototype.propertyIsEnumerable.call(t, e) && (n[e] = t[e]);
  }
  return n;
}
function gt(t, r) {
  if (t == null) return {};
  var n = {};
  for (var e in t)
    if (Object.prototype.hasOwnProperty.call(t, e)) {
      if (r.indexOf(e) >= 0) continue;
      n[e] = t[e];
    }
  return n;
}
function B() {
  return B = Object.assign ? Object.assign.bind() : function(t) {
    for (var r = 1; r < arguments.length; r++) {
      var n = arguments[r];
      for (var e in n)
        Object.prototype.hasOwnProperty.call(n, e) && (t[e] = n[e]);
    }
    return t;
  }, B.apply(this, arguments);
}
function Y(t, r) {
  var n = Object.keys(t);
  if (Object.getOwnPropertySymbols) {
    var e = Object.getOwnPropertySymbols(t);
    r && (e = e.filter(function(i) {
      return Object.getOwnPropertyDescriptor(t, i).enumerable;
    })), n.push.apply(n, e);
  }
  return n;
}
function g(t) {
  for (var r = 1; r < arguments.length; r++) {
    var n = arguments[r] != null ? arguments[r] : {};
    r % 2 ? Y(Object(n), !0).forEach(function(e) {
      A(t, e, n[e]);
    }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(t, Object.getOwnPropertyDescriptors(n)) : Y(Object(n)).forEach(function(e) {
      Object.defineProperty(t, e, Object.getOwnPropertyDescriptor(n, e));
    });
  }
  return t;
}
function _(t) {
  return xt(t) || Pt(t) || At(t) || bt();
}
function bt() {
  throw new TypeError(`Invalid attempt to spread non-iterable instance.
In order to be iterable, non-array objects must have a [Symbol.iterator]() method.`);
}
function At(t, r) {
  if (t) {
    if (typeof t == "string") return M(t, r);
    var n = Object.prototype.toString.call(t).slice(8, -1);
    if (n === "Object" && t.constructor && (n = t.constructor.name), n === "Map" || n === "Set") return Array.from(t);
    if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return M(t, r);
  }
}
function Pt(t) {
  if (typeof Symbol < "u" && t[Symbol.iterator] != null || t["@@iterator"] != null) return Array.from(t);
}
function xt(t) {
  if (Array.isArray(t)) return M(t);
}
function M(t, r) {
  (r == null || r > t.length) && (r = t.length);
  for (var n = 0, e = new Array(r); n < r; n++) e[n] = t[n];
  return e;
}
function Ot(t, r) {
  if (!(t instanceof r))
    throw new TypeError("Cannot call a class as a function");
}
function J(t, r) {
  for (var n = 0; n < r.length; n++) {
    var e = r[n];
    e.enumerable = e.enumerable || !1, e.configurable = !0, "value" in e && (e.writable = !0), Object.defineProperty(t, tt(e.key), e);
  }
}
function St(t, r, n) {
  return r && J(t.prototype, r), n && J(t, n), Object.defineProperty(t, "prototype", { writable: !1 }), t;
}
function wt(t, r, n) {
  return r = R(r), Dt(t, Z() ? Reflect.construct(r, n || [], R(t).constructor) : r.apply(t, n));
}
function Dt(t, r) {
  if (r && (j(r) === "object" || typeof r == "function"))
    return r;
  if (r !== void 0)
    throw new TypeError("Derived constructors may only return object or undefined");
  return _t(t);
}
function _t(t) {
  if (t === void 0)
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  return t;
}
function Z() {
  try {
    var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function() {
    }));
  } catch {
  }
  return (Z = function() {
    return !!t;
  })();
}
function R(t) {
  return R = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function(n) {
    return n.__proto__ || Object.getPrototypeOf(n);
  }, R(t);
}
function jt(t, r) {
  if (typeof r != "function" && r !== null)
    throw new TypeError("Super expression must either be null or a function");
  t.prototype = Object.create(r && r.prototype, { constructor: { value: t, writable: !0, configurable: !0 } }), Object.defineProperty(t, "prototype", { writable: !1 }), r && U(t, r);
}
function U(t, r) {
  return U = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function(e, i) {
    return e.__proto__ = i, e;
  }, U(t, r);
}
function A(t, r, n) {
  return r = tt(r), r in t ? Object.defineProperty(t, r, { value: n, enumerable: !0, configurable: !0, writable: !0 }) : t[r] = n, t;
}
function tt(t) {
  var r = Et(t, "string");
  return j(r) == "symbol" ? r : r + "";
}
function Et(t, r) {
  if (j(t) != "object" || !t) return t;
  var n = t[Symbol.toPrimitive];
  if (n !== void 0) {
    var e = n.call(t, r);
    if (j(e) != "object") return e;
    throw new TypeError("@@toPrimitive must return a primitive value.");
  }
  return String(t);
}
var q = /* @__PURE__ */ (function(t) {
  function r() {
    var n;
    Ot(this, r);
    for (var e = arguments.length, i = new Array(e), o = 0; o < e; o++)
      i[o] = arguments[o];
    return n = wt(this, r, [].concat(i)), A(n, "state", {
      isAnimationFinished: !0,
      totalLength: 0
    }), A(n, "generateSimpleStrokeDasharray", function(a, s) {
      return "".concat(s, "px ").concat(a - s, "px");
    }), A(n, "getStrokeDasharray", function(a, s, l) {
      var p = l.reduce(function(P, O) {
        return P + O;
      });
      if (!p)
        return n.generateSimpleStrokeDasharray(s, a);
      for (var f = Math.floor(a / p), m = a % p, c = s - a, h = [], u = 0, v = 0; u < l.length; v += l[u], ++u)
        if (v + l[u] > m) {
          h = [].concat(_(l.slice(0, u)), [m - v]);
          break;
        }
      var b = h.length % 2 === 0 ? [0, c] : [c];
      return [].concat(_(r.repeat(l, f)), _(h), b).map(function(P) {
        return "".concat(P, "px");
      }).join(", ");
    }), A(n, "id", ft("recharts-line-")), A(n, "pathRef", function(a) {
      n.mainCurve = a;
    }), A(n, "handleAnimationEnd", function() {
      n.setState({
        isAnimationFinished: !0
      }), n.props.onAnimationEnd && n.props.onAnimationEnd();
    }), A(n, "handleAnimationStart", function() {
      n.setState({
        isAnimationFinished: !1
      }), n.props.onAnimationStart && n.props.onAnimationStart();
    }), n;
  }
  return jt(r, t), St(r, [{
    key: "componentDidMount",
    value: function() {
      if (this.props.isAnimationActive) {
        var e = this.getTotalLength();
        this.setState({
          totalLength: e
        });
      }
    }
  }, {
    key: "componentDidUpdate",
    value: function() {
      if (this.props.isAnimationActive) {
        var e = this.getTotalLength();
        e !== this.state.totalLength && this.setState({
          totalLength: e
        });
      }
    }
  }, {
    key: "getTotalLength",
    value: function() {
      var e = this.mainCurve;
      try {
        return e && e.getTotalLength && e.getTotalLength() || 0;
      } catch {
        return 0;
      }
    }
  }, {
    key: "renderErrorBar",
    value: function(e, i) {
      if (this.props.isAnimationActive && !this.state.isAnimationFinished)
        return null;
      var o = this.props, a = o.points, s = o.xAxis, l = o.yAxis, p = o.layout, f = o.children, m = mt(f, pt);
      if (!m)
        return null;
      var c = function(v, b) {
        return {
          x: v.x,
          y: v.y,
          value: v.value,
          errorVal: Q(v.payload, b)
        };
      }, h = {
        clipPath: e ? "url(#clipPath-".concat(i, ")") : null
      };
      return /* @__PURE__ */ y.createElement(z, h, m.map(function(u) {
        return /* @__PURE__ */ y.cloneElement(u, {
          key: "bar-".concat(u.props.dataKey),
          data: a,
          xAxis: s,
          yAxis: l,
          layout: p,
          dataPointFormatter: c
        });
      }));
    }
  }, {
    key: "renderDots",
    value: function(e, i, o) {
      var a = this.props.isAnimationActive;
      if (a && !this.state.isAnimationFinished)
        return null;
      var s = this.props, l = s.dot, p = s.points, f = s.dataKey, m = W(this.props, !1), c = W(l, !0), h = p.map(function(v, b) {
        var P = g(g(g({
          key: "dot-".concat(b),
          r: 3
        }, m), c), {}, {
          value: v.value,
          dataKey: f,
          cx: v.x,
          cy: v.y,
          index: b,
          payload: v.payload
        });
        return r.renderDotItem(l, P);
      }), u = {
        clipPath: e ? "url(#clipPath-".concat(i ? "" : "dots-").concat(o, ")") : null
      };
      return /* @__PURE__ */ y.createElement(z, B({
        className: "recharts-line-dots",
        key: "dots"
      }, u), h);
    }
  }, {
    key: "renderCurveStatically",
    value: function(e, i, o, a) {
      var s = this.props, l = s.type, p = s.layout, f = s.connectNulls;
      s.ref;
      var m = X(s, yt), c = g(g(g({}, W(m, !0)), {}, {
        fill: "none",
        className: "recharts-line-curve",
        clipPath: i ? "url(#clipPath-".concat(o, ")") : null,
        points: e
      }, a), {}, {
        type: l,
        layout: p,
        connectNulls: f
      });
      return /* @__PURE__ */ y.createElement(lt, B({}, c, {
        pathRef: this.pathRef
      }));
    }
  }, {
    key: "renderCurveWithAnimation",
    value: function(e, i) {
      var o = this, a = this.props, s = a.points, l = a.strokeDasharray, p = a.isAnimationActive, f = a.animationBegin, m = a.animationDuration, c = a.animationEasing, h = a.animationId, u = a.animateNewValues, v = a.width, b = a.height, P = this.state, O = P.prevPoints, E = P.totalLength;
      return /* @__PURE__ */ y.createElement(at, {
        begin: f,
        duration: m,
        isActive: p,
        easing: c,
        from: {
          t: 0
        },
        to: {
          t: 1
        },
        key: "line-".concat(h),
        onAnimationEnd: this.handleAnimationEnd,
        onAnimationStart: this.handleAnimationStart
      }, function(w) {
        var x = w.t;
        if (O) {
          var k = O.length / s.length, S = s.map(function(d, K) {
            var T = Math.floor(K * k);
            if (O[T]) {
              var C = O[T], D = N(C.x, d.x), et = N(C.y, d.y);
              return g(g({}, d), {}, {
                x: D(x),
                y: et(x)
              });
            }
            if (u) {
              var rt = N(v * 2, d.x), nt = N(b / 2, d.y);
              return g(g({}, d), {}, {
                x: rt(x),
                y: nt(x)
              });
            }
            return g(g({}, d), {}, {
              x: d.x,
              y: d.y
            });
          });
          return o.renderCurveStatically(S, e, i);
        }
        var F = N(0, E), L = F(x), I;
        if (l) {
          var $ = "".concat(l).split(/[,\s]+/gim).map(function(d) {
            return parseFloat(d);
          });
          I = o.getStrokeDasharray(L, E, $);
        } else
          I = o.generateSimpleStrokeDasharray(E, L);
        return o.renderCurveStatically(s, e, i, {
          strokeDasharray: I
        });
      });
    }
  }, {
    key: "renderCurve",
    value: function(e, i) {
      var o = this.props, a = o.points, s = o.isAnimationActive, l = this.state, p = l.prevPoints, f = l.totalLength;
      return s && a && a.length && (!p && f > 0 || !st(p, a)) ? this.renderCurveWithAnimation(e, i) : this.renderCurveStatically(a, e, i);
    }
  }, {
    key: "render",
    value: function() {
      var e, i = this.props, o = i.hide, a = i.dot, s = i.points, l = i.className, p = i.xAxis, f = i.yAxis, m = i.top, c = i.left, h = i.width, u = i.height, v = i.isAnimationActive, b = i.id;
      if (o || !s || !s.length)
        return null;
      var P = this.state.isAnimationFinished, O = s.length === 1, E = G("recharts-line", l), w = p && p.allowDataOverflow, x = f && f.allowDataOverflow, k = w || x, S = V(b) ? this.id : b, F = (e = W(a, !1)) !== null && e !== void 0 ? e : {
        r: 3,
        strokeWidth: 2
      }, L = F.r, I = L === void 0 ? 3 : L, $ = F.strokeWidth, d = $ === void 0 ? 2 : $, K = ht(a) ? a : {}, T = K.clipDot, C = T === void 0 ? !0 : T, D = I * 2 + d;
      return /* @__PURE__ */ y.createElement(z, {
        className: E
      }, w || x ? /* @__PURE__ */ y.createElement("defs", null, /* @__PURE__ */ y.createElement("clipPath", {
        id: "clipPath-".concat(S)
      }, /* @__PURE__ */ y.createElement("rect", {
        x: w ? c : c - h / 2,
        y: x ? m : m - u / 2,
        width: w ? h : h * 2,
        height: x ? u : u * 2
      })), !C && /* @__PURE__ */ y.createElement("clipPath", {
        id: "clipPath-dots-".concat(S)
      }, /* @__PURE__ */ y.createElement("rect", {
        x: c - D / 2,
        y: m - D / 2,
        width: h + D,
        height: u + D
      }))) : null, !O && this.renderCurve(k, S), this.renderErrorBar(k, S), (O || a) && this.renderDots(k, C, S), (!v || P) && ct.renderCallByParent(this.props, s));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function(e, i) {
      return e.animationId !== i.prevAnimationId ? {
        prevAnimationId: e.animationId,
        curPoints: e.points,
        prevPoints: i.curPoints
      } : e.points !== i.curPoints ? {
        curPoints: e.points
      } : null;
    }
  }, {
    key: "repeat",
    value: function(e, i) {
      for (var o = e.length % 2 !== 0 ? [].concat(_(e), [0]) : e, a = [], s = 0; s < i; ++s)
        a = [].concat(_(a), _(o));
      return a;
    }
  }, {
    key: "renderDotItem",
    value: function(e, i) {
      var o;
      if (/* @__PURE__ */ y.isValidElement(e))
        o = /* @__PURE__ */ y.cloneElement(e, i);
      else if (ot(e))
        o = e(i);
      else {
        var a = i.key, s = X(i, dt), l = G("recharts-line-dot", typeof e != "boolean" ? e.className : "");
        o = /* @__PURE__ */ y.createElement(ut, B({
          key: a
        }, s, {
          className: l
        }));
      }
      return o;
    }
  }]);
})(it);
A(q, "displayName", "Line");
A(q, "defaultProps", {
  xAxisId: 0,
  yAxisId: 0,
  connectNulls: !1,
  activeDot: !0,
  dot: !0,
  legendType: "line",
  stroke: "#3182bd",
  strokeWidth: 1,
  fill: "#fff",
  points: [],
  isAnimationActive: !vt.isSsr,
  animateNewValues: !0,
  animationBegin: 0,
  animationDuration: 1500,
  animationEasing: "ease",
  hide: !1,
  label: !1
});
A(q, "getComposedData", function(t) {
  var r = t.props, n = t.xAxis, e = t.yAxis, i = t.xAxisTicks, o = t.yAxisTicks, a = t.dataKey, s = t.bandSize, l = t.displayedData, p = t.offset, f = r.layout, m = l.map(function(c, h) {
    var u = Q(c, a);
    return f === "horizontal" ? {
      x: H({
        axis: n,
        ticks: i,
        bandSize: s,
        entry: c,
        index: h
      }),
      y: V(u) ? null : e.scale(u),
      value: u,
      payload: c
    } : {
      x: V(u) ? null : n.scale(u),
      y: H({
        axis: e,
        ticks: o,
        bandSize: s,
        entry: c,
        index: h
      }),
      value: u,
      payload: c
    };
  });
  return g({
    points: m,
    layout: f
  }, p);
});
export {
  q as Line
};
