// recharts@2.15.0/es6/cartesian/ReferenceLine.js
import m from "react";
import F from "../../../../../../../_virtual/isFunction.js";
import K from "../../../../../../../_virtual/some.js";
import { clsx as M } from "../../../../../clsx@2.1.1/node_modules/clsx/dist/clsx.js";
import { Layer as V } from "../container/Layer.js";
import { Label as k } from "../component/Label.js";
import { ifOverflowMatches as x } from "../util/IfOverflowMatches.js";
import { isNumOrStr as L } from "../util/DataUtils.js";
import { createLabeledScales as z, rectWithCoords as H } from "../util/CartesianUtils.js";
import { warn as U } from "../util/LogUtils.js";
import { filterProps as q } from "../util/ReactUtils.js";
import { useClipPathId as G, useXAxisOrThrow as J, useYAxisOrThrow as Q, useViewBox as Z } from "../context/chartLayoutContext.js";
function d(e) {
  "@babel/helpers - typeof";
  return d = typeof Symbol == "function" && typeof Symbol.iterator == "symbol" ? function(t) {
    return typeof t;
  } : function(t) {
    return t && typeof Symbol == "function" && t.constructor === Symbol && t !== Symbol.prototype ? "symbol" : typeof t;
  }, d(e);
}
function ee(e, t) {
  if (!(e instanceof t))
    throw new TypeError("Cannot call a class as a function");
}
function te(e, t) {
  for (var r = 0; r < t.length; r++) {
    var n = t[r];
    n.enumerable = n.enumerable || !1, n.configurable = !0, "value" in n && (n.writable = !0), Object.defineProperty(e, Y(n.key), n);
  }
}
function re(e, t, r) {
  return t && te(e.prototype, t), Object.defineProperty(e, "prototype", { writable: !1 }), e;
}
function ne(e, t, r) {
  return t = _(t), ie(e, X() ? Reflect.construct(t, r || [], _(e).constructor) : t.apply(e, r));
}
function ie(e, t) {
  if (t && (d(t) === "object" || typeof t == "function"))
    return t;
  if (t !== void 0)
    throw new TypeError("Derived constructors may only return object or undefined");
  return oe(e);
}
function oe(e) {
  if (e === void 0)
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  return e;
}
function X() {
  try {
    var e = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function() {
    }));
  } catch {
  }
  return (X = function() {
    return !!e;
  })();
}
function _(e) {
  return _ = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function(r) {
    return r.__proto__ || Object.getPrototypeOf(r);
  }, _(e);
}
function ae(e, t) {
  if (typeof t != "function" && t !== null)
    throw new TypeError("Super expression must either be null or a function");
  e.prototype = Object.create(t && t.prototype, { constructor: { value: e, writable: !0, configurable: !0 } }), Object.defineProperty(e, "prototype", { writable: !1 }), t && A(e, t);
}
function A(e, t) {
  return A = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function(n, o) {
    return n.__proto__ = o, n;
  }, A(e, t);
}
function N(e, t) {
  var r = Object.keys(e);
  if (Object.getOwnPropertySymbols) {
    var n = Object.getOwnPropertySymbols(e);
    t && (n = n.filter(function(o) {
      return Object.getOwnPropertyDescriptor(e, o).enumerable;
    })), r.push.apply(r, n);
  }
  return r;
}
function D(e) {
  for (var t = 1; t < arguments.length; t++) {
    var r = arguments[t] != null ? arguments[t] : {};
    t % 2 ? N(Object(r), !0).forEach(function(n) {
      E(e, n, r[n]);
    }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(r)) : N(Object(r)).forEach(function(n) {
      Object.defineProperty(e, n, Object.getOwnPropertyDescriptor(r, n));
    });
  }
  return e;
}
function E(e, t, r) {
  return t = Y(t), t in e ? Object.defineProperty(e, t, { value: r, enumerable: !0, configurable: !0, writable: !0 }) : e[t] = r, e;
}
function Y(e) {
  var t = fe(e, "string");
  return d(t) == "symbol" ? t : t + "";
}
function fe(e, t) {
  if (d(e) != "object" || !e) return e;
  var r = e[Symbol.toPrimitive];
  if (r !== void 0) {
    var n = r.call(e, t);
    if (d(n) != "object") return n;
    throw new TypeError("@@toPrimitive must return a primitive value.");
  }
  return String(e);
}
function ue(e, t) {
  return se(e) || ye(e, t) || ce(e, t) || le();
}
function le() {
  throw new TypeError(`Invalid attempt to destructure non-iterable instance.
In order to be iterable, non-array objects must have a [Symbol.iterator]() method.`);
}
function ce(e, t) {
  if (e) {
    if (typeof e == "string") return W(e, t);
    var r = Object.prototype.toString.call(e).slice(8, -1);
    if (r === "Object" && e.constructor && (r = e.constructor.name), r === "Map" || r === "Set") return Array.from(e);
    if (r === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(r)) return W(e, t);
  }
}
function W(e, t) {
  (t == null || t > e.length) && (t = e.length);
  for (var r = 0, n = new Array(t); r < t; r++) n[r] = e[r];
  return n;
}
function ye(e, t) {
  var r = e == null ? null : typeof Symbol < "u" && e[Symbol.iterator] || e["@@iterator"];
  if (r != null) {
    var n, o, a, f, u = [], l = !0, i = !1;
    try {
      if (a = (r = r.call(e)).next, t !== 0) for (; !(l = (n = a.call(r)).done) && (u.push(n.value), u.length !== t); l = !0) ;
    } catch (c) {
      i = !0, o = c;
    } finally {
      try {
        if (!l && r.return != null && (f = r.return(), Object(f) !== f)) return;
      } finally {
        if (i) throw o;
      }
    }
    return u;
  }
}
function se(e) {
  if (Array.isArray(e)) return e;
}
function I() {
  return I = Object.assign ? Object.assign.bind() : function(e) {
    for (var t = 1; t < arguments.length; t++) {
      var r = arguments[t];
      for (var n in r)
        Object.prototype.hasOwnProperty.call(r, n) && (e[n] = r[n]);
    }
    return e;
  }, I.apply(this, arguments);
}
var me = function(t, r) {
  var n;
  return /* @__PURE__ */ m.isValidElement(t) ? n = /* @__PURE__ */ m.cloneElement(t, r) : F(t) ? n = t(r) : n = /* @__PURE__ */ m.createElement("line", I({}, r, {
    className: "recharts-reference-line-line"
  })), n;
}, de = function(t, r, n, o, a, f, u, l, i) {
  var c = a.x, b = a.y, O = a.width, j = a.height;
  if (n) {
    var S = i.y, p = t.y.apply(S, {
      position: f
    });
    if (x(i, "discard") && !t.y.isInRange(p))
      return null;
    var g = [{
      x: c + O,
      y: p
    }, {
      x: c,
      y: p
    }];
    return l === "left" ? g.reverse() : g;
  }
  if (r) {
    var P = i.x, y = t.x.apply(P, {
      position: f
    });
    if (x(i, "discard") && !t.x.isInRange(y))
      return null;
    var v = [{
      x: y,
      y: b + j
    }, {
      x: y,
      y: b
    }];
    return u === "top" ? v.reverse() : v;
  }
  if (o) {
    var w = i.segment, h = w.map(function(s) {
      return t.apply(s, {
        position: f
      });
    });
    return x(i, "discard") && K(h, function(s) {
      return !t.isInRange(s);
    }) ? null : h;
  }
  return null;
};
function be(e) {
  var t = e.x, r = e.y, n = e.segment, o = e.xAxisId, a = e.yAxisId, f = e.shape, u = e.className, l = e.alwaysShow, i = G(), c = J(o), b = Q(a), O = Z();
  if (!i || !O)
    return null;
  U(l === void 0, 'The alwaysShow prop is deprecated. Please use ifOverflow="extendDomain" instead.');
  var j = z({
    x: c.scale,
    y: b.scale
  }), S = L(t), p = L(r), g = n && n.length === 2, P = de(j, S, p, g, O, e.position, c.orientation, b.orientation, e);
  if (!P)
    return null;
  var y = ue(P, 2), v = y[0], w = v.x, h = v.y, s = y[1], R = s.x, T = s.y, B = x(e, "hidden") ? "url(#".concat(i, ")") : void 0, C = D(D({
    clipPath: B
  }, q(e, !0)), {}, {
    x1: w,
    y1: h,
    x2: R,
    y2: T
  });
  return /* @__PURE__ */ m.createElement(V, {
    className: M("recharts-reference-line", u)
  }, me(f, C), k.renderCallByParent(e, H({
    x1: w,
    y1: h,
    x2: R,
    y2: T
  })));
}
var $ = /* @__PURE__ */ (function(e) {
  function t() {
    return ee(this, t), ne(this, t, arguments);
  }
  return ae(t, e), re(t, [{
    key: "render",
    value: function() {
      return /* @__PURE__ */ m.createElement(be, this.props);
    }
  }]);
})(m.Component);
E($, "displayName", "ReferenceLine");
E($, "defaultProps", {
  isFront: !1,
  ifOverflow: "discard",
  xAxisId: 0,
  yAxisId: 0,
  fill: "none",
  stroke: "#ccc",
  fillOpacity: 1,
  strokeWidth: 1,
  position: "middle"
});
export {
  $ as ReferenceLine,
  de as getEndPoints
};
