// recharts@2.15.0/es6/cartesian/ReferenceDot.js
import a from "react";
import I from "../../../../../../../_virtual/isFunction.js";
import { clsx as T } from "../../../../../clsx@2.1.1/node_modules/clsx/dist/clsx.js";
import { Layer as B } from "../container/Layer.js";
import { Dot as L } from "../shape/Dot.js";
import { Label as F } from "../component/Label.js";
import { isNumOrStr as v } from "../util/DataUtils.js";
import { ifOverflowMatches as _ } from "../util/IfOverflowMatches.js";
import { createLabeledScales as K } from "../util/CartesianUtils.js";
import { warn as $ } from "../util/LogUtils.js";
import { filterProps as z } from "../util/ReactUtils.js";
function y() {
  return y = Object.assign ? Object.assign.bind() : function(t) {
    for (var e = 1; e < arguments.length; e++) {
      var r = arguments[e];
      for (var n in r)
        Object.prototype.hasOwnProperty.call(r, n) && (t[n] = r[n]);
    }
    return t;
  }, y.apply(this, arguments);
}
function c(t) {
  "@babel/helpers - typeof";
  return c = typeof Symbol == "function" && typeof Symbol.iterator == "symbol" ? function(e) {
    return typeof e;
  } : function(e) {
    return e && typeof Symbol == "function" && e.constructor === Symbol && e !== Symbol.prototype ? "symbol" : typeof e;
  }, c(t);
}
function P(t, e) {
  var r = Object.keys(t);
  if (Object.getOwnPropertySymbols) {
    var n = Object.getOwnPropertySymbols(t);
    e && (n = n.filter(function(o) {
      return Object.getOwnPropertyDescriptor(t, o).enumerable;
    })), r.push.apply(r, n);
  }
  return r;
}
function w(t) {
  for (var e = 1; e < arguments.length; e++) {
    var r = arguments[e] != null ? arguments[e] : {};
    e % 2 ? P(Object(r), !0).forEach(function(n) {
      l(t, n, r[n]);
    }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(t, Object.getOwnPropertyDescriptors(r)) : P(Object(r)).forEach(function(n) {
      Object.defineProperty(t, n, Object.getOwnPropertyDescriptor(r, n));
    });
  }
  return t;
}
function C(t, e) {
  if (!(t instanceof e))
    throw new TypeError("Cannot call a class as a function");
}
function M(t, e) {
  for (var r = 0; r < e.length; r++) {
    var n = e[r];
    n.enumerable = n.enumerable || !1, n.configurable = !0, "value" in n && (n.writable = !0), Object.defineProperty(t, j(n.key), n);
  }
}
function V(t, e, r) {
  return e && M(t.prototype, e), Object.defineProperty(t, "prototype", { writable: !1 }), t;
}
function W(t, e, r) {
  return e = u(e), X(t, g() ? Reflect.construct(e, r || [], u(t).constructor) : e.apply(t, r));
}
function X(t, e) {
  if (e && (c(e) === "object" || typeof e == "function"))
    return e;
  if (e !== void 0)
    throw new TypeError("Derived constructors may only return object or undefined");
  return Y(t);
}
function Y(t) {
  if (t === void 0)
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  return t;
}
function g() {
  try {
    var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function() {
    }));
  } catch {
  }
  return (g = function() {
    return !!t;
  })();
}
function u(t) {
  return u = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function(r) {
    return r.__proto__ || Object.getPrototypeOf(r);
  }, u(t);
}
function k(t, e) {
  if (typeof e != "function" && e !== null)
    throw new TypeError("Super expression must either be null or a function");
  t.prototype = Object.create(e && e.prototype, { constructor: { value: t, writable: !0, configurable: !0 } }), Object.defineProperty(t, "prototype", { writable: !1 }), e && m(t, e);
}
function m(t, e) {
  return m = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function(n, o) {
    return n.__proto__ = o, n;
  }, m(t, e);
}
function l(t, e, r) {
  return e = j(e), e in t ? Object.defineProperty(t, e, { value: r, enumerable: !0, configurable: !0, writable: !0 }) : t[e] = r, t;
}
function j(t) {
  var e = q(t, "string");
  return c(e) == "symbol" ? e : e + "";
}
function q(t, e) {
  if (c(t) != "object" || !t) return t;
  var r = t[Symbol.toPrimitive];
  if (r !== void 0) {
    var n = r.call(t, e);
    if (c(n) != "object") return n;
    throw new TypeError("@@toPrimitive must return a primitive value.");
  }
  return String(t);
}
var G = function(e) {
  var r = e.x, n = e.y, o = e.xAxis, s = e.yAxis, i = K({
    x: o.scale,
    y: s.scale
  }), f = i.apply({
    x: r,
    y: n
  }, {
    bandAware: !0
  });
  return _(e, "discard") && !i.isInRange(f) ? null : f;
}, b = /* @__PURE__ */ (function(t) {
  function e() {
    return C(this, e), W(this, e, arguments);
  }
  return k(e, t), V(e, [{
    key: "render",
    value: function() {
      var n = this.props, o = n.x, s = n.y, i = n.r, f = n.alwaysShow, x = n.clipPathId, S = v(o), D = v(s);
      if ($(f === void 0, 'The alwaysShow prop is deprecated. Please use ifOverflow="extendDomain" instead.'), !S || !D)
        return null;
      var p = G(this.props);
      if (!p)
        return null;
      var d = p.x, h = p.y, O = this.props, E = O.shape, R = O.className, N = _(this.props, "hidden") ? "url(#".concat(x, ")") : void 0, A = w(w({
        clipPath: N
      }, z(this.props, !0)), {}, {
        cx: d,
        cy: h
      });
      return /* @__PURE__ */ a.createElement(B, {
        className: T("recharts-reference-dot", R)
      }, e.renderDot(E, A), F.renderCallByParent(this.props, {
        x: d - i,
        y: h - i,
        width: 2 * i,
        height: 2 * i
      }));
    }
  }]);
})(a.Component);
l(b, "displayName", "ReferenceDot");
l(b, "defaultProps", {
  isFront: !1,
  ifOverflow: "discard",
  xAxisId: 0,
  yAxisId: 0,
  r: 10,
  fill: "#fff",
  stroke: "#ccc",
  fillOpacity: 1,
  strokeWidth: 1
});
l(b, "renderDot", function(t, e) {
  var r;
  return /* @__PURE__ */ a.isValidElement(t) ? r = /* @__PURE__ */ a.cloneElement(t, e) : I(t) ? r = t(e) : r = /* @__PURE__ */ a.createElement(L, y({}, e, {
    cx: e.cx,
    cy: e.cy,
    className: "recharts-reference-dot-dot"
  })), r;
});
export {
  b as ReferenceDot
};
