// recharts@2.15.0/es6/cartesian/Scatter.js
import d, { PureComponent as G } from "react";
import X from "../../../../../react-smooth@4.0.4_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/react-smooth/es6/Animate.js";
import w from "../../../../../../../_virtual/isNil.js";
import Y from "../../../../../../../_virtual/isEqual.js";
import H from "../../../../../../../_virtual/isFunction.js";
import { clsx as Q } from "../../../../../clsx@2.1.1/node_modules/clsx/dist/clsx.js";
import { Layer as K } from "../container/Layer.js";
import { LabelList as U } from "../component/LabelList.js";
import { filterProps as F, findAllByType as V } from "../util/ReactUtils.js";
import { Global as tt } from "../util/Global.js";
import { ZAxis as et } from "./ZAxis.js";
import { Curve as it } from "../shape/Curve.js";
import { ErrorBar as rt } from "./ErrorBar.js";
import { Cell as nt } from "../component/Cell.js";
import { interpolateNumber as D, getLinearRegression as at, uniqueId as ot } from "../util/DataUtils.js";
import { getValueByDataKey as T, getCateCoordinateOfLine as q } from "../util/ChartUtils.js";
import { adaptEventsOfChild as st } from "../util/types.js";
import { ScatterSymbol as lt } from "../util/ScatterUtils.js";
function E(i) {
  "@babel/helpers - typeof";
  return E = typeof Symbol == "function" && typeof Symbol.iterator == "symbol" ? function(t) {
    return typeof t;
  } : function(t) {
    return t && typeof Symbol == "function" && t.constructor === Symbol && t !== Symbol.prototype ? "symbol" : typeof t;
  }, E(i);
}
function _() {
  return _ = Object.assign ? Object.assign.bind() : function(i) {
    for (var t = 1; t < arguments.length; t++) {
      var r = arguments[t];
      for (var e in r)
        Object.prototype.hasOwnProperty.call(r, e) && (i[e] = r[e]);
    }
    return i;
  }, _.apply(this, arguments);
}
function J(i, t) {
  var r = Object.keys(i);
  if (Object.getOwnPropertySymbols) {
    var e = Object.getOwnPropertySymbols(i);
    t && (e = e.filter(function(n) {
      return Object.getOwnPropertyDescriptor(i, n).enumerable;
    })), r.push.apply(r, e);
  }
  return r;
}
function x(i) {
  for (var t = 1; t < arguments.length; t++) {
    var r = arguments[t] != null ? arguments[t] : {};
    t % 2 ? J(Object(r), !0).forEach(function(e) {
      O(i, e, r[e]);
    }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(i, Object.getOwnPropertyDescriptors(r)) : J(Object(r)).forEach(function(e) {
      Object.defineProperty(i, e, Object.getOwnPropertyDescriptor(r, e));
    });
  }
  return i;
}
function ct(i, t) {
  if (!(i instanceof t))
    throw new TypeError("Cannot call a class as a function");
}
function M(i, t) {
  for (var r = 0; r < t.length; r++) {
    var e = t[r];
    e.enumerable = e.enumerable || !1, e.configurable = !0, "value" in e && (e.writable = !0), Object.defineProperty(i, Z(e.key), e);
  }
}
function pt(i, t, r) {
  return t && M(i.prototype, t), r && M(i, r), Object.defineProperty(i, "prototype", { writable: !1 }), i;
}
function ut(i, t, r) {
  return t = I(t), mt(i, W() ? Reflect.construct(t, r || [], I(i).constructor) : t.apply(i, r));
}
function mt(i, t) {
  if (t && (E(t) === "object" || typeof t == "function"))
    return t;
  if (t !== void 0)
    throw new TypeError("Derived constructors may only return object or undefined");
  return ft(i);
}
function ft(i) {
  if (i === void 0)
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  return i;
}
function W() {
  try {
    var i = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function() {
    }));
  } catch {
  }
  return (W = function() {
    return !!i;
  })();
}
function I(i) {
  return I = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function(r) {
    return r.__proto__ || Object.getPrototypeOf(r);
  }, I(i);
}
function yt(i, t) {
  if (typeof t != "function" && t !== null)
    throw new TypeError("Super expression must either be null or a function");
  i.prototype = Object.create(t && t.prototype, { constructor: { value: i, writable: !0, configurable: !0 } }), Object.defineProperty(i, "prototype", { writable: !1 }), t && N(i, t);
}
function N(i, t) {
  return N = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function(e, n) {
    return e.__proto__ = n, e;
  }, N(i, t);
}
function O(i, t, r) {
  return t = Z(t), t in i ? Object.defineProperty(i, t, { value: r, enumerable: !0, configurable: !0, writable: !0 }) : i[t] = r, i;
}
function Z(i) {
  var t = dt(i, "string");
  return E(t) == "symbol" ? t : t + "";
}
function dt(i, t) {
  if (E(i) != "object" || !i) return i;
  var r = i[Symbol.toPrimitive];
  if (r !== void 0) {
    var e = r.call(i, t);
    if (E(e) != "object") return e;
    throw new TypeError("@@toPrimitive must return a primitive value.");
  }
  return String(i);
}
var C = /* @__PURE__ */ (function(i) {
  function t() {
    var r;
    ct(this, t);
    for (var e = arguments.length, n = new Array(e), a = 0; a < e; a++)
      n[a] = arguments[a];
    return r = ut(this, t, [].concat(n)), O(r, "state", {
      isAnimationFinished: !1
    }), O(r, "handleAnimationEnd", function() {
      r.setState({
        isAnimationFinished: !0
      });
    }), O(r, "handleAnimationStart", function() {
      r.setState({
        isAnimationFinished: !1
      });
    }), O(r, "id", ot("recharts-scatter-")), r;
  }
  return yt(t, i), pt(t, [{
    key: "renderSymbolsStatically",
    value: function(e) {
      var n = this, a = this.props, u = a.shape, P = a.activeShape, f = a.activeIndex, m = F(this.props, !1);
      return e.map(function(o, s) {
        var p = f === s, l = p ? P : u, y = x(x({}, m), o);
        return /* @__PURE__ */ d.createElement(K, _({
          className: "recharts-scatter-symbol",
          key: "symbol-".concat(o?.cx, "-").concat(o?.cy, "-").concat(o?.size, "-").concat(s)
        }, st(n.props, o, s), {
          role: "img"
        }), /* @__PURE__ */ d.createElement(lt, _({
          option: l,
          isActive: p,
          key: "symbol-".concat(s)
        }, y)));
      });
    }
  }, {
    key: "renderSymbolsWithAnimation",
    value: function() {
      var e = this, n = this.props, a = n.points, u = n.isAnimationActive, P = n.animationBegin, f = n.animationDuration, m = n.animationEasing, o = n.animationId, s = this.state.prevPoints;
      return /* @__PURE__ */ d.createElement(X, {
        begin: P,
        duration: f,
        isActive: u,
        easing: m,
        from: {
          t: 0
        },
        to: {
          t: 1
        },
        key: "pie-".concat(o),
        onAnimationEnd: this.handleAnimationEnd,
        onAnimationStart: this.handleAnimationStart
      }, function(p) {
        var l = p.t, y = a.map(function(v, h) {
          var b = s && s[h];
          if (b) {
            var A = D(b.cx, v.cx), c = D(b.cy, v.cy), g = D(b.size, v.size);
            return x(x({}, v), {}, {
              cx: A(l),
              cy: c(l),
              size: g(l)
            });
          }
          var S = D(0, v.size);
          return x(x({}, v), {}, {
            size: S(l)
          });
        });
        return /* @__PURE__ */ d.createElement(K, null, e.renderSymbolsStatically(y));
      });
    }
  }, {
    key: "renderSymbols",
    value: function() {
      var e = this.props, n = e.points, a = e.isAnimationActive, u = this.state.prevPoints;
      return a && n && n.length && (!u || !Y(u, n)) ? this.renderSymbolsWithAnimation() : this.renderSymbolsStatically(n);
    }
  }, {
    key: "renderErrorBar",
    value: function() {
      var e = this.props.isAnimationActive;
      if (e && !this.state.isAnimationFinished)
        return null;
      var n = this.props, a = n.points, u = n.xAxis, P = n.yAxis, f = n.children, m = V(f, rt);
      return m ? m.map(function(o, s) {
        var p = o.props, l = p.direction, y = p.dataKey;
        return /* @__PURE__ */ d.cloneElement(o, {
          key: "".concat(l, "-").concat(y, "-").concat(a[s]),
          data: a,
          xAxis: u,
          yAxis: P,
          layout: l === "x" ? "vertical" : "horizontal",
          dataPointFormatter: function(h, b) {
            return {
              x: h.cx,
              y: h.cy,
              value: l === "x" ? +h.node.x : +h.node.y,
              errorVal: T(h, b)
            };
          }
        });
      }) : null;
    }
  }, {
    key: "renderLine",
    value: function() {
      var e = this.props, n = e.points, a = e.line, u = e.lineType, P = e.lineJointType, f = F(this.props, !1), m = F(a, !1), o, s;
      if (u === "joint")
        o = n.map(function(c) {
          return {
            x: c.cx,
            y: c.cy
          };
        });
      else if (u === "fitting") {
        var p = at(n), l = p.xmin, y = p.xmax, v = p.a, h = p.b, b = function(g) {
          return v * g + h;
        };
        o = [{
          x: l,
          y: b(l)
        }, {
          x: y,
          y: b(y)
        }];
      }
      var A = x(x(x({}, f), {}, {
        fill: "none",
        stroke: f && f.fill
      }, m), {}, {
        points: o
      });
      return /* @__PURE__ */ d.isValidElement(a) ? s = /* @__PURE__ */ d.cloneElement(a, A) : H(a) ? s = a(A) : s = /* @__PURE__ */ d.createElement(it, _({}, A, {
        type: P
      })), /* @__PURE__ */ d.createElement(K, {
        className: "recharts-scatter-line",
        key: "recharts-scatter-line"
      }, s);
    }
  }, {
    key: "render",
    value: function() {
      var e = this.props, n = e.hide, a = e.points, u = e.line, P = e.className, f = e.xAxis, m = e.yAxis, o = e.left, s = e.top, p = e.width, l = e.height, y = e.id, v = e.isAnimationActive;
      if (n || !a || !a.length)
        return null;
      var h = this.state.isAnimationFinished, b = Q("recharts-scatter", P), A = f && f.allowDataOverflow, c = m && m.allowDataOverflow, g = A || c, S = w(y) ? this.id : y;
      return /* @__PURE__ */ d.createElement(K, {
        className: b,
        clipPath: g ? "url(#clipPath-".concat(S, ")") : null
      }, A || c ? /* @__PURE__ */ d.createElement("defs", null, /* @__PURE__ */ d.createElement("clipPath", {
        id: "clipPath-".concat(S)
      }, /* @__PURE__ */ d.createElement("rect", {
        x: A ? o : o - p / 2,
        y: c ? s : s - l / 2,
        width: A ? p : p * 2,
        height: c ? l : l * 2
      }))) : null, u && this.renderLine(), this.renderErrorBar(), /* @__PURE__ */ d.createElement(K, {
        key: "recharts-scatter-symbols"
      }, this.renderSymbols()), (!v || h) && U.renderCallByParent(this.props, a));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function(e, n) {
      return e.animationId !== n.prevAnimationId ? {
        prevAnimationId: e.animationId,
        curPoints: e.points,
        prevPoints: n.curPoints
      } : e.points !== n.curPoints ? {
        curPoints: e.points
      } : null;
    }
  }]);
})(G);
O(C, "displayName", "Scatter");
O(C, "defaultProps", {
  xAxisId: 0,
  yAxisId: 0,
  zAxisId: 0,
  legendType: "circle",
  lineType: "joint",
  lineJointType: "linear",
  data: [],
  shape: "circle",
  hide: !1,
  isAnimationActive: !tt.isSsr,
  animationBegin: 0,
  animationDuration: 400,
  animationEasing: "linear"
});
O(C, "getComposedData", function(i) {
  var t = i.xAxis, r = i.yAxis, e = i.zAxis, n = i.item, a = i.displayedData, u = i.xAxisTicks, P = i.yAxisTicks, f = i.offset, m = n.props.tooltipType, o = V(n.props.children, nt), s = w(t.dataKey) ? n.props.dataKey : t.dataKey, p = w(r.dataKey) ? n.props.dataKey : r.dataKey, l = e && e.dataKey, y = e ? e.range : et.defaultProps.range, v = y && y[0], h = t.scale.bandwidth ? t.scale.bandwidth() : 0, b = r.scale.bandwidth ? r.scale.bandwidth() : 0, A = a.map(function(c, g) {
    var S = T(c, s), $ = T(c, p), j = !w(l) && T(c, l) || "-", L = [{
      name: w(t.dataKey) ? n.props.name : t.name || t.dataKey,
      unit: t.unit || "",
      value: S,
      payload: c,
      dataKey: s,
      type: m
    }, {
      name: w(r.dataKey) ? n.props.name : r.name || r.dataKey,
      unit: r.unit || "",
      value: $,
      payload: c,
      dataKey: p,
      type: m
    }];
    j !== "-" && L.push({
      name: e.name || e.dataKey,
      unit: e.unit || "",
      value: j,
      payload: c,
      dataKey: l,
      type: m
    });
    var B = q({
      axis: t,
      ticks: u,
      bandSize: h,
      entry: c,
      index: g,
      dataKey: s
    }), k = q({
      axis: r,
      ticks: P,
      bandSize: b,
      entry: c,
      index: g,
      dataKey: p
    }), R = j !== "-" ? e.scale(j) : v, z = Math.sqrt(Math.max(R, 0) / Math.PI);
    return x(x({}, c), {}, {
      cx: B,
      cy: k,
      x: B - z,
      y: k - z,
      xAxis: t,
      yAxis: r,
      zAxis: e,
      width: 2 * z,
      height: 2 * z,
      size: R,
      node: {
        x: S,
        y: $,
        z: j
      },
      tooltipPayload: L,
      tooltipPosition: {
        x: B,
        y: k
      },
      payload: c
    }, o && o[g] && o[g].props);
  });
  return x({
    points: A
  }, f);
});
export {
  C as Scatter
};
