// recharts@2.15.0/es6/cartesian/YAxis.js
import f from "react";
import { clsx as h } from "../../../../../clsx@2.1.1/node_modules/clsx/dist/clsx.js";
import { useChartWidth as v, useChartHeight as d, useYAxisOrThrow as O } from "../context/chartLayoutContext.js";
import { CartesianAxis as _ } from "./CartesianAxis.js";
import { getTicksOfAxis as w } from "../util/ChartUtils.js";
function o(e) {
  "@babel/helpers - typeof";
  return o = typeof Symbol == "function" && typeof Symbol.iterator == "symbol" ? function(t) {
    return typeof t;
  } : function(t) {
    return t && typeof Symbol == "function" && t.constructor === Symbol && t !== Symbol.prototype ? "symbol" : typeof t;
  }, o(e);
}
function g(e, t) {
  if (!(e instanceof t))
    throw new TypeError("Cannot call a class as a function");
}
function P(e, t) {
  for (var r = 0; r < t.length; r++) {
    var n = t[r];
    n.enumerable = n.enumerable || !1, n.configurable = !0, "value" in n && (n.writable = !0), Object.defineProperty(e, y(n.key), n);
  }
}
function x(e, t, r) {
  return t && P(e.prototype, t), Object.defineProperty(e, "prototype", { writable: !1 }), e;
}
function j(e, t, r) {
  return t = u(t), A(e, s() ? Reflect.construct(t, r || [], u(e).constructor) : t.apply(e, r));
}
function A(e, t) {
  if (t && (o(t) === "object" || typeof t == "function"))
    return t;
  if (t !== void 0)
    throw new TypeError("Derived constructors may only return object or undefined");
  return S(e);
}
function S(e) {
  if (e === void 0)
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  return e;
}
function s() {
  try {
    var e = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function() {
    }));
  } catch {
  }
  return (s = function() {
    return !!e;
  })();
}
function u(e) {
  return u = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function(r) {
    return r.__proto__ || Object.getPrototypeOf(r);
  }, u(e);
}
function T(e, t) {
  if (typeof t != "function" && t !== null)
    throw new TypeError("Super expression must either be null or a function");
  e.prototype = Object.create(t && t.prototype, { constructor: { value: e, writable: !0, configurable: !0 } }), Object.defineProperty(e, "prototype", { writable: !1 }), t && c(e, t);
}
function c(e, t) {
  return c = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function(n, a) {
    return n.__proto__ = a, n;
  }, c(e, t);
}
function p(e, t, r) {
  return t = y(t), t in e ? Object.defineProperty(e, t, { value: r, enumerable: !0, configurable: !0, writable: !0 }) : e[t] = r, e;
}
function y(e) {
  var t = E(e, "string");
  return o(t) == "symbol" ? t : t + "";
}
function E(e, t) {
  if (o(e) != "object" || !e) return e;
  var r = e[Symbol.toPrimitive];
  if (r !== void 0) {
    var n = r.call(e, t);
    if (o(n) != "object") return n;
    throw new TypeError("@@toPrimitive must return a primitive value.");
  }
  return String(e);
}
function l() {
  return l = Object.assign ? Object.assign.bind() : function(e) {
    for (var t = 1; t < arguments.length; t++) {
      var r = arguments[t];
      for (var n in r)
        Object.prototype.hasOwnProperty.call(r, n) && (e[n] = r[n]);
    }
    return e;
  }, l.apply(this, arguments);
}
var R = function(t) {
  var r = t.yAxisId, n = v(), a = d(), i = O(r);
  return i == null ? null : (
    // @ts-expect-error the axisOptions type is not exactly what CartesianAxis is expecting.
    /* @__PURE__ */ f.createElement(_, l({}, i, {
      className: h("recharts-".concat(i.axisType, " ").concat(i.axisType), i.className),
      viewBox: {
        x: 0,
        y: 0,
        width: n,
        height: a
      },
      ticksGenerator: function(b) {
        return w(b, !0);
      }
    }))
  );
}, m = /* @__PURE__ */ (function(e) {
  function t() {
    return g(this, t), j(this, t, arguments);
  }
  return T(t, e), x(t, [{
    key: "render",
    value: function() {
      return /* @__PURE__ */ f.createElement(R, this.props);
    }
  }]);
})(f.Component);
p(m, "displayName", "YAxis");
p(m, "defaultProps", {
  allowDuplicatedCategory: !0,
  allowDecimals: !0,
  hide: !1,
  orientation: "left",
  width: 60,
  height: 0,
  mirror: !1,
  yAxisId: 0,
  tickCount: 5,
  type: "number",
  padding: {
    top: 0,
    bottom: 0
  },
  allowDataOverflow: !1,
  scale: "auto",
  reversed: !1
});
export {
  m as YAxis
};
