// echarts@6.0.0/lib/chart/boxplot/boxplotLayout.js
import { each as P, indexOf as D, isArray as T } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { parsePercent as E } from "../../util/number.js";
var N = P;
function k(a) {
  var v = z(a);
  N(v, function(r) {
    var n = r.seriesModels;
    n.length && (G(r), N(n, function(t, i) {
      R(t, r.boxOffsetList[i], r.boxWidthList[i]);
    }));
  });
}
function z(a) {
  var v = [], r = [];
  return a.eachSeriesByType("boxplot", function(n) {
    var t = n.getBaseAxis(), i = D(r, t);
    i < 0 && (i = r.length, r[i] = t, v[i] = {
      axis: t,
      seriesModels: []
    }), v[i].seriesModels.push(n);
  }), v;
}
function G(a) {
  var v = a.axis, r = a.seriesModels, n = r.length, t = a.boxWidthList = [], i = a.boxOffsetList = [], s = [], f;
  if (v.type === "category")
    f = v.getBandWidth();
  else {
    var L = 0;
    N(r, function(l) {
      L = Math.max(L, l.getData().count());
    });
    var W = v.getExtent();
    f = Math.abs(W[1] - W[0]) / L;
  }
  N(r, function(l) {
    var e = l.get("boxWidth");
    T(e) || (e = [e, e]), s.push([E(e[0], f) || 0, E(e[1], f) || 0]);
  });
  var u = f * 0.8 - 2, o = u / n * 0.3, c = (u - o * (n - 1)) / n, B = c / 2 - u / 2;
  N(r, function(l, e) {
    i.push(B), B += o + c, t.push(Math.min(Math.max(c, s[e][0]), s[e][1]));
  });
}
function R(a, v, r) {
  var n = a.coordinateSystem, t = a.getData(), i = r / 2, s = a.get("layout") === "horizontal" ? 0 : 1, f = 1 - s, L = ["x", "y"], W = t.mapDimension(L[s]), u = t.mapDimensionsAll(L[f]);
  if (W == null || u.length < 5)
    return;
  for (var o = 0; o < t.count(); o++) {
    var c = t.get(W, o), B = M(c, u[2], o), l = M(c, u[0], o), e = M(c, u[1], o), m = M(c, u[3], o), O = M(c, u[4], o), b = [];
    d(b, e, !1), d(b, m, !0), b.push(l, e, O, m), S(b, l), S(b, O), S(b, B), t.setItemLayout(o, {
      initBaseline: B[f],
      ends: b
    });
  }
  function M(p, y, x) {
    var h = t.get(y, x), g = [];
    g[s] = p, g[f] = h;
    var A;
    return isNaN(p) || isNaN(h) ? A = [NaN, NaN] : (A = n.dataToPoint(g), A[s] += v), A;
  }
  function d(p, y, x) {
    var h = y.slice(), g = y.slice();
    h[s] += i, g[s] -= i, x ? p.push(h, g) : p.push(g, h);
  }
  function S(p, y) {
    var x = y.slice(), h = y.slice();
    x[s] -= i, h[s] += i, p.push(x, h);
  }
}
export {
  k as default
};
