// echarts@6.0.0/lib/chart/boxplot/BoxplotSeries.js
import { __extends as r } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import a from "../../model/Series.js";
import { WhiskerBoxCommonMixin as l } from "../helper/whiskerBoxCommon.js";
import { mixin as n } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import i from "../../visual/tokens.js";
var s = (
  /** @class */
  (function(t) {
    r(e, t);
    function e() {
      var o = t !== null && t.apply(this, arguments) || this;
      return o.type = e.type, o.defaultValueDimensions = [{
        name: "min",
        defaultTooltip: !0
      }, {
        name: "Q1",
        defaultTooltip: !0
      }, {
        name: "median",
        defaultTooltip: !0
      }, {
        name: "Q3",
        defaultTooltip: !0
      }, {
        name: "max",
        defaultTooltip: !0
      }], o.visualDrawType = "stroke", o;
    }
    return e.type = "series.boxplot", e.dependencies = ["xAxis", "yAxis", "grid"], e.defaultOption = {
      // zlevel: 0,
      z: 2,
      coordinateSystem: "cartesian2d",
      legendHoverLink: !0,
      layout: null,
      boxWidth: [7, 50],
      itemStyle: {
        color: i.color.neutral00,
        borderWidth: 1
      },
      emphasis: {
        scale: !0,
        itemStyle: {
          borderWidth: 2,
          shadowBlur: 5,
          shadowOffsetX: 1,
          shadowOffsetY: 1,
          shadowColor: i.color.shadow
        }
      },
      animationDuration: 800
    }, e;
  })(a)
);
n(s, l, !0);
export {
  s as default
};
