// echarts@6.0.0/lib/chart/boxplot/BoxplotView.js
import { __extends as v } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { map as d } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import g from "../../view/Chart.js";
import * as y from "../../util/graphic.js";
import { setStatesStylesFromModel as B, toggleHoverEmphasis as P } from "../../util/states.js";
import M from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Path.js";
import { saveOldStyle as S } from "../../animation/basicTransition.js";
var k = (
  /** @class */
  (function(i) {
    v(e, i);
    function e() {
      var o = i !== null && i.apply(this, arguments) || this;
      return o.type = e.type, o;
    }
    return e.prototype.render = function(o, r, a) {
      var t = o.getData(), s = this.group, p = this._data;
      this._data || s.removeAll();
      var u = o.get("layout") === "horizontal" ? 1 : 0;
      t.diff(p).add(function(n) {
        if (t.hasValue(n)) {
          var h = t.getItemLayout(n), l = f(h, t, n, u, !0);
          t.setItemGraphicEl(n, l), s.add(l);
        }
      }).update(function(n, h) {
        var l = p.getItemGraphicEl(h);
        if (!t.hasValue(n)) {
          s.remove(l);
          return;
        }
        var m = t.getItemLayout(n);
        l ? (S(l), c(m, l, t, n)) : l = f(m, t, n, u), s.add(l), t.setItemGraphicEl(n, l);
      }).remove(function(n) {
        var h = p.getItemGraphicEl(n);
        h && s.remove(h);
      }).execute(), this._data = t;
    }, e.prototype.remove = function(o) {
      var r = this.group, a = this._data;
      this._data = null, a && a.eachItemGraphicEl(function(t) {
        t && r.remove(t);
      });
    }, e.type = "boxplot", e;
  })(g)
), E = (
  /** @class */
  /* @__PURE__ */ (function() {
    function i() {
    }
    return i;
  })()
), V = (
  /** @class */
  (function(i) {
    v(e, i);
    function e(o) {
      var r = i.call(this, o) || this;
      return r.type = "boxplotBoxPath", r;
    }
    return e.prototype.getDefaultShape = function() {
      return new E();
    }, e.prototype.buildPath = function(o, r) {
      var a = r.points, t = 0;
      for (o.moveTo(a[t][0], a[t][1]), t++; t < 4; t++)
        o.lineTo(a[t][0], a[t][1]);
      for (o.closePath(); t < a.length; t++)
        o.moveTo(a[t][0], a[t][1]), t++, o.lineTo(a[t][0], a[t][1]);
    }, e;
  })(M)
);
function f(i, e, o, r, a) {
  var t = i.ends, s = new V({
    shape: {
      points: a ? _(t, r, i) : t
    }
  });
  return c(i, s, e, o, a), s;
}
function c(i, e, o, r, a) {
  var t = o.hostModel, s = y[a ? "initProps" : "updateProps"];
  s(e, {
    shape: {
      points: i.ends
    }
  }, t, r), e.useStyle(o.getItemVisual(r, "style")), e.style.strokeNoScale = !0, e.z2 = 100;
  var p = o.getItemModel(r), u = p.getModel("emphasis");
  B(e, p), P(e, u.get("focus"), u.get("blurScope"), u.get("disabled"));
}
function _(i, e, o) {
  return d(i, function(r) {
    return r = r.slice(), r[e] = o.initBaseline, r;
  });
}
export {
  k as default
};
