// echarts@6.0.0/lib/component/parallel/ParallelView.js
import { __extends as h } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import p from "../../view/Component.js";
import { each as d, bind as l, extend as f } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { createOrUpdate as c, clear as u } from "../../util/throttle.js";
var x = 5, g = (
  /** @class */
  (function(o) {
    h(i, o);
    function i() {
      var t = o !== null && o.apply(this, arguments) || this;
      return t.type = i.type, t;
    }
    return i.prototype.render = function(t, n, e) {
      this._model = t, this._api = e, this._handlers || (this._handlers = {}, d(m, function(a, r) {
        e.getZr().on(r, this._handlers[r] = l(a, this));
      }, this)), c(this, "_throttledDispatchExpand", t.get("axisExpandRate"), "fixRate");
    }, i.prototype.dispose = function(t, n) {
      u(this, "_throttledDispatchExpand"), d(this._handlers, function(e, a) {
        n.getZr().off(a, e);
      }), this._handlers = null;
    }, i.prototype._throttledDispatchExpand = function(t) {
      this._dispatchExpand(t);
    }, i.prototype._dispatchExpand = function(t) {
      t && this._api.dispatchAction(f({
        type: "parallelAxisExpand"
      }, t));
    }, i.type = "parallel", i;
  })(p)
), m = {
  mousedown: function(o) {
    s(this, "click") && (this._mouseDownPoint = [o.offsetX, o.offsetY]);
  },
  mouseup: function(o) {
    var i = this._mouseDownPoint;
    if (s(this, "click") && i) {
      var t = [o.offsetX, o.offsetY], n = Math.pow(i[0] - t[0], 2) + Math.pow(i[1] - t[1], 2);
      if (n > x)
        return;
      var e = this._model.coordinateSystem.getSlidedAxisExpandWindow([o.offsetX, o.offsetY]);
      e.behavior !== "none" && this._dispatchExpand({
        axisExpandWindow: e.axisExpandWindow
      });
    }
    this._mouseDownPoint = null;
  },
  mousemove: function(o) {
    if (!(this._mouseDownPoint || !s(this, "mousemove"))) {
      var i = this._model, t = i.coordinateSystem.getSlidedAxisExpandWindow([o.offsetX, o.offsetY]), n = t.behavior;
      n === "jump" && this._throttledDispatchExpand.debounceNextCall(i.get("axisExpandDebounce")), this._throttledDispatchExpand(n === "none" ? null : {
        axisExpandWindow: t.axisExpandWindow,
        // Jumping uses animation, and sliding suppresses animation.
        animation: n === "jump" ? null : {
          duration: 0
          // Disable animation.
        }
      });
    }
  }
};
function s(o, i) {
  var t = o._model;
  return t.get("axisExpandable") && t.get("axisExpandTriggerOn") === i;
}
export {
  g as default
};
