// echarts@6.0.0/lib/component/calendar/CalendarView.js
import { __extends as N } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { isString as w, isFunction as O, extend as T, map as z } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { createTextStyle as L } from "../../label/labelStyle.js";
import { formatTplSimple as W } from "../../util/format.js";
import { parsePercent as A } from "../../util/number.js";
import I from "../../view/Component.js";
import { getLocaleModel as C } from "../../core/locale.js";
import D from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Rect.js";
import R from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Polyline.js";
import P from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Text.js";
var J = (
  /** @class */
  (function(c) {
    N(f, c);
    function f() {
      var e = c !== null && c.apply(this, arguments) || this;
      return e.type = f.type, e;
    }
    return f.prototype.render = function(e, r, l) {
      var i = this.group;
      i.removeAll();
      var t = e.coordinateSystem, n = t.getRangeInfo(), o = t.getOrient(), a = r.getLocaleModel();
      this._renderDayRect(e, n, i), this._renderLines(e, n, o, i), this._renderYearText(e, n, o, i), this._renderMonthText(e, a, o, i), this._renderWeekText(e, a, n, o, i);
    }, f.prototype._renderDayRect = function(e, r, l) {
      for (var i = e.coordinateSystem, t = e.getModel("itemStyle").getItemStyle(), n = i.getCellWidth(), o = i.getCellHeight(), a = r.start.time; a <= r.end.time; a = i.getNextNDay(a, 1).time) {
        var s = i.dataToCalendarLayout([a], !1).tl, v = new D({
          shape: {
            x: s[0],
            y: s[1],
            width: n,
            height: o
          },
          cursor: "default",
          style: t
        });
        l.add(v);
      }
    }, f.prototype._renderLines = function(e, r, l, i) {
      var t = this, n = e.coordinateSystem, o = e.getModel(["splitLine", "lineStyle"]).getLineStyle(), a = e.get(["splitLine", "show"]), s = o.lineWidth;
      this._tlpoints = [], this._blpoints = [], this._firstDayOfMonth = [], this._firstDayPoints = [];
      for (var v = r.start, h = 0; v.time <= r.end.time; h++) {
        p(v.formatedDate), h === 0 && (v = n.getDateInfo(r.start.y + "-" + r.start.m));
        var m = v.date;
        m.setMonth(m.getMonth() + 1), v = n.getDateInfo(m);
      }
      p(n.getNextNDay(r.end.time, 1).formatedDate);
      function p(g) {
        t._firstDayOfMonth.push(n.getDateInfo(g)), t._firstDayPoints.push(n.dataToCalendarLayout([g], !1).tl);
        var y = t._getLinePointsOfOneWeek(e, g, l);
        t._tlpoints.push(y[0]), t._blpoints.push(y[y.length - 1]), a && t._drawSplitline(y, o, i);
      }
      a && this._drawSplitline(t._getEdgesPoints(t._tlpoints, s, l), o, i), a && this._drawSplitline(t._getEdgesPoints(t._blpoints, s, l), o, i);
    }, f.prototype._getEdgesPoints = function(e, r, l) {
      var i = [e[0].slice(), e[e.length - 1].slice()], t = l === "horizontal" ? 0 : 1;
      return i[0][t] = i[0][t] - r / 2, i[1][t] = i[1][t] + r / 2, i;
    }, f.prototype._drawSplitline = function(e, r, l) {
      var i = new R({
        z2: 20,
        shape: {
          points: e
        },
        style: r
      });
      l.add(i);
    }, f.prototype._getLinePointsOfOneWeek = function(e, r, l) {
      for (var i = e.coordinateSystem, t = i.getDateInfo(r), n = [], o = 0; o < 7; o++) {
        var a = i.getNextNDay(t.time, o), s = i.dataToCalendarLayout([a.time], !1);
        n[2 * a.day] = s.tl, n[2 * a.day + 1] = s[l === "horizontal" ? "bl" : "tr"];
      }
      return n;
    }, f.prototype._formatterLabel = function(e, r) {
      return w(e) && e ? W(e, r) : O(e) ? e(r) : r.nameMap;
    }, f.prototype._yearTextPositionControl = function(e, r, l, i, t) {
      var n = r[0], o = r[1], a = ["center", "bottom"];
      i === "bottom" ? (o += t, a = ["center", "top"]) : i === "left" ? n -= t : i === "right" ? (n += t, a = ["center", "top"]) : o -= t;
      var s = 0;
      return (i === "left" || i === "right") && (s = Math.PI / 2), {
        rotation: s,
        x: n,
        y: o,
        style: {
          align: a[0],
          verticalAlign: a[1]
        }
      };
    }, f.prototype._renderYearText = function(e, r, l, i) {
      var t = e.getModel("yearLabel");
      if (t.get("show")) {
        var n = t.get("margin"), o = t.get("position");
        o || (o = l !== "horizontal" ? "top" : "left");
        var a = [this._tlpoints[this._tlpoints.length - 1], this._blpoints[0]], s = (a[0][0] + a[1][0]) / 2, v = (a[0][1] + a[1][1]) / 2, h = l === "horizontal" ? 0 : 1, m = {
          top: [s, a[h][1]],
          bottom: [s, a[1 - h][1]],
          left: [a[1 - h][0], v],
          right: [a[h][0], v]
        }, p = r.start.y;
        +r.end.y > +r.start.y && (p = p + "-" + r.end.y);
        var g = t.get("formatter"), y = {
          start: r.start.y,
          end: r.end.y,
          nameMap: p
        }, _ = this._formatterLabel(g, y), d = new P({
          z2: 30,
          style: L(t, {
            text: _
          }),
          silent: t.get("silent")
        });
        d.attr(this._yearTextPositionControl(d, m[o], l, o, n)), i.add(d);
      }
    }, f.prototype._monthTextPositionControl = function(e, r, l, i, t) {
      var n = "left", o = "top", a = e[0], s = e[1];
      return l === "horizontal" ? (s = s + t, r && (n = "center"), i === "start" && (o = "bottom")) : (a = a + t, r && (o = "middle"), i === "start" && (n = "right")), {
        x: a,
        y: s,
        align: n,
        verticalAlign: o
      };
    }, f.prototype._renderMonthText = function(e, r, l, i) {
      var t = e.getModel("monthLabel");
      if (t.get("show")) {
        var n = t.get("nameMap"), o = t.get("margin"), a = t.get("position"), s = t.get("align"), v = [this._tlpoints, this._blpoints];
        (!n || w(n)) && (n && (r = C(n) || r), n = r.get(["time", "monthAbbr"]) || []);
        var h = a === "start" ? 0 : 1, m = l === "horizontal" ? 0 : 1;
        o = a === "start" ? -o : o;
        for (var p = s === "center", g = t.get("silent"), y = 0; y < v[h].length - 1; y++) {
          var _ = v[h][y].slice(), d = this._firstDayOfMonth[y];
          if (p) {
            var x = this._firstDayPoints[y];
            _[m] = (x[m] + v[0][y + 1][m]) / 2;
          }
          var u = t.get("formatter"), b = n[+d.m - 1], S = {
            yyyy: d.y,
            yy: (d.y + "").slice(2),
            MM: d.m,
            M: +d.m,
            nameMap: b
          }, M = this._formatterLabel(u, S), k = new P({
            z2: 30,
            style: T(L(t, {
              text: M
            }), this._monthTextPositionControl(_, p, l, a, o)),
            silent: g
          });
          i.add(k);
        }
      }
    }, f.prototype._weekTextPositionControl = function(e, r, l, i, t) {
      var n = "center", o = "middle", a = e[0], s = e[1], v = l === "start";
      return r === "horizontal" ? (a = a + i + (v ? 1 : -1) * t[0] / 2, n = v ? "right" : "left") : (s = s + i + (v ? 1 : -1) * t[1] / 2, o = v ? "bottom" : "top"), {
        x: a,
        y: s,
        align: n,
        verticalAlign: o
      };
    }, f.prototype._renderWeekText = function(e, r, l, i, t) {
      var n = e.getModel("dayLabel");
      if (n.get("show")) {
        var o = e.coordinateSystem, a = n.get("position"), s = n.get("nameMap"), v = n.get("margin"), h = o.getFirstDayOfWeek();
        if (!s || w(s)) {
          s && (r = C(s) || r);
          var m = r.get(["time", "dayOfWeekShort"]);
          s = m || z(r.get(["time", "dayOfWeekAbbr"]), function(S) {
            return S[0];
          });
        }
        var p = o.getNextNDay(l.end.time, 7 - l.lweek).time, g = [o.getCellWidth(), o.getCellHeight()];
        v = A(v, Math.min(g[1], g[0])), a === "start" && (p = o.getNextNDay(l.start.time, -(7 + l.fweek)).time, v = -v);
        for (var y = n.get("silent"), _ = 0; _ < 7; _++) {
          var d = o.getNextNDay(p, _), x = o.dataToCalendarLayout([d.time], !1).center, u = _;
          u = Math.abs((_ + h) % 7);
          var b = new P({
            z2: 30,
            style: T(L(n, {
              text: s[u]
            }), this._weekTextPositionControl(x, i, a, v, g)),
            silent: y
          });
          t.add(b);
        }
      }
    }, f.type = "calendar", f;
  })(I)
);
export {
  J as default
};
