// echarts@6.0.0/lib/chart/heatmap/HeatmapLayer.js
import { platformApi as S } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/platform.js";
import L from "../../visual/tokens.js";
var A = 256, B = (
  /** @class */
  (function() {
    function u() {
      this.blurSize = 30, this.pointSize = 20, this.maxOpacity = 1, this.minOpacity = 0, this._gradientPixels = {
        inRange: null,
        outOfRange: null
      };
      var t = S.createCanvas();
      this.canvas = t;
    }
    return u.prototype.update = function(t, e, r, a, h, v) {
      var l = this._getBrush(), b = this._getGradient(h, "inRange"), w = this._getGradient(h, "outOfRange"), d = this.pointSize + this.blurSize, i = this.canvas, p = i.getContext("2d"), _ = t.length;
      i.width = e, i.height = r;
      for (var f = 0; f < _; ++f) {
        var m = t[f], C = m[0], O = m[1], z = m[2], n = a(z);
        p.globalAlpha = n, p.drawImage(l, C - d, O - d);
      }
      if (!i.width || !i.height)
        return i;
      for (var x = p.getImageData(0, 0, i.width, i.height), o = x.data, s = 0, R = o.length, y = this.minOpacity, I = this.maxOpacity, P = I - y; s < R; ) {
        var n = o[s + 3] / 256, g = Math.floor(n * (A - 1)) * 4;
        if (n > 0) {
          var c = v(n) ? b : w;
          n > 0 && (n = n * P + y), o[s++] = c[g], o[s++] = c[g + 1], o[s++] = c[g + 2], o[s++] = c[g + 3] * n * 256;
        } else
          s += 4;
      }
      return p.putImageData(x, 0, 0), i;
    }, u.prototype._getBrush = function() {
      var t = this._brushCanvas || (this._brushCanvas = S.createCanvas()), e = this.pointSize + this.blurSize, r = e * 2;
      t.width = r, t.height = r;
      var a = t.getContext("2d");
      return a.clearRect(0, 0, r, r), a.shadowOffsetX = r, a.shadowBlur = this.blurSize, a.shadowColor = L.color.neutral99, a.beginPath(), a.arc(-e, e, this.pointSize, 0, Math.PI * 2, !0), a.closePath(), a.fill(), t;
    }, u.prototype._getGradient = function(t, e) {
      for (var r = this._gradientPixels, a = r[e] || (r[e] = new Uint8ClampedArray(256 * 4)), h = [0, 0, 0, 0], v = 0, l = 0; l < 256; l++)
        t[e](l / 255, !0, h), a[v++] = h[0], a[v++] = h[1], a[v++] = h[2], a[v++] = h[3];
      return a;
    }, u;
  })()
);
export {
  B as default
};
