// echarts@6.0.0/lib/chart/heatmap/HeatmapView.js
import { __extends as j } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { traverseElements as J } from "../../util/graphic.js";
import { toggleHoverEmphasis as K } from "../../util/states.js";
import Q from "./HeatmapLayer.js";
import { eqNaN as k, map as U } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import $ from "../../view/Chart.js";
import { isCoordinateSystemType as A } from "../../coord/CoordinateSystem.js";
import { getLabelStatesModels as F, setLabelStyle as X } from "../../label/labelStyle.js";
import P from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Rect.js";
import Y from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Image.js";
function M(p, a, t) {
  var n = p[1] - p[0];
  a = U(a, function(r) {
    return {
      interval: [(r.interval[0] - p[0]) / n, (r.interval[1] - p[0]) / n]
    };
  });
  var i = a.length, s = 0;
  return function(r) {
    var e;
    for (e = s; e < i; e++) {
      var m = a[e].interval;
      if (m[0] <= r && r <= m[1]) {
        s = e;
        break;
      }
    }
    if (e === i)
      for (e = s - 1; e >= 0; e--) {
        var m = a[e].interval;
        if (m[0] <= r && r <= m[1]) {
          s = e;
          break;
        }
      }
    return e >= 0 && e < i && t[e];
  };
}
function tt(p, a) {
  var t = p[1] - p[0];
  return a = [(a[0] - p[0]) / t, (a[1] - p[0]) / t], function(n) {
    return n >= a[0] && n <= a[1];
  };
}
function Z(p) {
  var a = p.dimensions;
  return a[0] === "lng" && a[1] === "lat";
}
var ut = (
  /** @class */
  (function(p) {
    j(a, p);
    function a() {
      var t = p !== null && p.apply(this, arguments) || this;
      return t.type = a.type, t;
    }
    return a.prototype.render = function(t, n, i) {
      var s;
      if (n.eachComponent("visualMap", function(e) {
        e.eachTargetSeries(function(m) {
          m === t && (s = e);
        });
      }), process.env.NODE_ENV !== "production" && !s)
        throw new Error("Heatmap must use with visualMap");
      this._progressiveEls = null, this.group.removeAll();
      var r = t.coordinateSystem;
      r.type === "cartesian2d" || r.type === "calendar" || r.type === "matrix" ? this._renderOnGridLike(t, i, 0, t.getData().count()) : Z(r) && this._renderOnGeo(r, t, s, i);
    }, a.prototype.incrementalPrepareRender = function(t, n, i) {
      this.group.removeAll();
    }, a.prototype.incrementalRender = function(t, n, i, s) {
      var r = n.coordinateSystem;
      r && (Z(r) ? this.render(n, i, s) : (this._progressiveEls = [], this._renderOnGridLike(n, s, t.start, t.end, !0)));
    }, a.prototype.eachRendered = function(t) {
      J(this._progressiveEls || this.group, t);
    }, a.prototype._renderOnGridLike = function(t, n, i, s, r) {
      var e = t.coordinateSystem, m = A(e, "cartesian2d"), l = A(e, "matrix"), v, b, h, y;
      if (m) {
        var S = e.getAxis("x"), d = e.getAxis("y");
        if (process.env.NODE_ENV !== "production") {
          if (!(S.type === "category" && d.type === "category"))
            throw new Error("Heatmap on cartesian must have two category axes");
          if (!(S.onBand && d.onBand))
            throw new Error("Heatmap on cartesian must have two axes with boundaryGap true");
        }
        v = S.getBandWidth() + 0.5, b = d.getBandWidth() + 0.5, h = S.scale.getExtent(), y = d.scale.getExtent();
      }
      for (var E = this.group, o = t.getData(), N = t.getModel(["emphasis", "itemStyle"]).getItemStyle(), _ = t.getModel(["blur", "itemStyle"]).getItemStyle(), R = t.getModel(["select", "itemStyle"]).getItemStyle(), D = t.get(["itemStyle", "borderRadius"]), T = F(t), I = t.getModel("emphasis"), V = I.get("focus"), L = I.get("blurScope"), f = I.get("disabled"), c = m || l ? [o.mapDimension("x"), o.mapDimension("y"), o.mapDimension("value")] : [o.mapDimension("time"), o.mapDimension("value")], g = i; g < s; g++) {
        var u = void 0, C = o.getItemVisual(g, "style");
        if (m) {
          var G = o.get(c[0], g), H = o.get(c[1], g);
          if (isNaN(o.get(c[2], g)) || isNaN(G) || isNaN(H) || G < h[0] || G > h[1] || H < y[0] || H > y[1])
            continue;
          var B = e.dataToPoint([G, H]);
          u = new P({
            shape: {
              x: B[0] - v / 2,
              y: B[1] - b / 2,
              width: v,
              height: b
            },
            style: C
          });
        } else if (l) {
          var w = e.dataToLayout([o.get(c[0], g), o.get(c[1], g)]).rect;
          if (k(w.x))
            continue;
          u = new P({
            z2: 1,
            shape: w,
            style: C
          });
        } else {
          if (isNaN(o.get(c[1], g)))
            continue;
          var W = e.dataToLayout([o.get(c[0], g)]), w = W.contentRect || W.rect;
          if (k(w.x) || k(w.y))
            continue;
          u = new P({
            z2: 1,
            shape: w,
            style: C
          });
        }
        if (o.hasItemOption) {
          var O = o.getItemModel(g), z = O.getModel("emphasis");
          N = z.getModel("itemStyle").getItemStyle(), _ = O.getModel(["blur", "itemStyle"]).getItemStyle(), R = O.getModel(["select", "itemStyle"]).getItemStyle(), D = O.get(["itemStyle", "borderRadius"]), V = z.get("focus"), L = z.get("blurScope"), f = z.get("disabled"), T = F(O);
        }
        u.shape.r = D;
        var x = t.getRawValue(g), q = "-";
        x && x[2] != null && (q = x[2] + ""), X(u, T, {
          labelFetcher: t,
          labelDataIndex: g,
          defaultOpacity: C.opacity,
          defaultText: q
        }), u.ensureState("emphasis").style = N, u.ensureState("blur").style = _, u.ensureState("select").style = R, K(u, V, L, f), u.incremental = r, r && (u.states.emphasis.hoverLayer = !0), E.add(u), o.setItemGraphicEl(g, u), this._progressiveEls && this._progressiveEls.push(u);
      }
    }, a.prototype._renderOnGeo = function(t, n, i, s) {
      var r = i.targetVisuals.inRange, e = i.targetVisuals.outOfRange, m = n.getData(), l = this._hmLayer || this._hmLayer || new Q();
      l.blurSize = n.get("blurSize"), l.pointSize = n.get("pointSize"), l.minOpacity = n.get("minOpacity"), l.maxOpacity = n.get("maxOpacity");
      var v = t.getViewRect().clone(), b = t.getRoamTransform();
      v.applyTransform(b);
      var h = Math.max(v.x, 0), y = Math.max(v.y, 0), S = Math.min(v.width + v.x, s.getWidth()), d = Math.min(v.height + v.y, s.getHeight()), E = S - h, o = d - y, N = [m.mapDimension("lng"), m.mapDimension("lat"), m.mapDimension("value")], _ = m.mapArray(N, function(I, V, L) {
        var f = t.dataToPoint([I, V]);
        return f[0] -= h, f[1] -= y, f.push(L), f;
      }), R = i.getExtent(), D = i.type === "visualMap.continuous" ? tt(R, i.option.range) : M(R, i.getPieceList(), i.option.selected);
      l.update(_, E, o, r.color.getNormalizer(), {
        inRange: r.color.getColorMapper(),
        outOfRange: e.color.getColorMapper()
      }, D);
      var T = new Y({
        style: {
          width: E,
          height: o,
          x: h,
          y,
          image: l.canvas
        },
        silent: !0
      });
      this.group.add(T);
    }, a.type = "heatmap", a;
  })($)
);
export {
  ut as default
};
