// echarts@6.0.0/lib/chart/heatmap/HeatmapSeries.js
import { __extends as o } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import i from "../../model/Series.js";
import n from "../helper/createSeriesData.js";
import a from "../../core/CoordinateSystem.js";
import m from "../../visual/tokens.js";
var u = (
  /** @class */
  (function(r) {
    o(e, r);
    function e() {
      var t = r !== null && r.apply(this, arguments) || this;
      return t.type = e.type, t;
    }
    return e.prototype.getInitialData = function(t, s) {
      return n(null, this, {
        generateCoord: "value"
      });
    }, e.prototype.preventIncremental = function() {
      var t = a.get(this.get("coordinateSystem"));
      if (t && t.dimensions)
        return t.dimensions[0] === "lng" && t.dimensions[1] === "lat";
    }, e.type = "series.heatmap", e.dependencies = ["grid", "geo", "calendar", "matrix"], e.defaultOption = {
      coordinateSystem: "cartesian2d",
      // zlevel: 0,
      z: 2,
      // Cartesian coordinate system
      // xAxisIndex: 0,
      // yAxisIndex: 0,
      // Geo coordinate system
      geoIndex: 0,
      blurSize: 30,
      pointSize: 20,
      maxOpacity: 1,
      minOpacity: 0,
      select: {
        itemStyle: {
          borderColor: m.color.primary
        }
      }
    }, e;
  })(i)
);
export {
  u as default
};
