// @atlaskit/pragmatic-drag-and-drop@1.7.4/dist/esm/honey-pot-fix/make-honey-pot-fix.js
import b from "../../../../../../../@babel_runtime@7.28.6/node_modules/@babel/runtime/helpers/esm/defineProperty.js";
import { d as y } from "../../../../../../../../../_virtual/index4.js";
import { maxZIndex as h } from "../util/max-z-index.js";
import { honeyPotDataAttribute as w } from "./honey-pot-data-attribute.js";
function d(t, n) {
  var e = Object.keys(t);
  if (Object.getOwnPropertySymbols) {
    var r = Object.getOwnPropertySymbols(t);
    n && (r = r.filter(function(o) {
      return Object.getOwnPropertyDescriptor(t, o).enumerable;
    })), e.push.apply(e, r);
  }
  return e;
}
function v(t) {
  for (var n = 1; n < arguments.length; n++) {
    var e = arguments[n] != null ? arguments[n] : {};
    n % 2 ? d(Object(e), !0).forEach(function(r) {
      b(t, r, e[r]);
    }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(t, Object.getOwnPropertyDescriptors(e)) : d(Object(e)).forEach(function(r) {
      Object.defineProperty(t, r, Object.getOwnPropertyDescriptor(e, r));
    });
  }
  return t;
}
var l = 2, m = l / 2;
function O(t) {
  return {
    x: Math.floor(t.x),
    y: Math.floor(t.y)
  };
}
function P(t) {
  return {
    x: t.x - m,
    y: t.y - m
  };
}
function j(t) {
  return {
    x: Math.max(t.x, 0),
    y: Math.max(t.y, 0)
  };
}
function R(t) {
  return {
    x: Math.min(t.x, window.innerWidth - l),
    y: Math.min(t.y, window.innerHeight - l)
  };
}
function x(t) {
  var n = t.client, e = R(j(P(O(n))));
  return DOMRect.fromRect({
    x: e.x,
    y: e.y,
    width: l,
    height: l
  });
}
function g(t) {
  var n = t.clientRect;
  return {
    left: "".concat(n.left, "px"),
    top: "".concat(n.top, "px"),
    width: "".concat(n.width, "px"),
    height: "".concat(n.height, "px")
  };
}
function D(t) {
  var n = t.client, e = t.clientRect;
  return (
    // is within horizontal bounds
    n.x >= e.x && n.x <= e.x + e.width && // is within vertical bounds
    n.y >= e.y && n.y <= e.y + e.height
  );
}
function S(t) {
  var n = t.initial, e = document.createElement("div");
  e.setAttribute(w, "true");
  var r = x({
    client: n
  });
  Object.assign(e.style, v(v({
    // Setting a background color explicitly to avoid any inherited styles.
    // Looks like this could be `opacity: 0`, but worried that _might_
    // cause the element to be ignored on some platforms.
    // When debugging, set backgroundColor to something like "red".
    backgroundColor: "transparent",
    position: "fixed",
    // Being explicit to avoid inheriting styles
    padding: 0,
    margin: 0,
    boxSizing: "border-box"
  }, g({
    clientRect: r
  })), {}, {
    // We want this element to absorb pointer events,
    // it's kind of the whole point 😉
    pointerEvents: "auto",
    // Want to make sure the honey pot is top of everything else.
    // Don't need to worry about native drag previews, as they will
    // have been rendered (and removed) before the honey pot is rendered
    zIndex: h
  })), document.body.appendChild(e);
  var o = y.bind(window, {
    type: "pointermove",
    listener: function(c) {
      var a = {
        x: c.clientX,
        y: c.clientY
      };
      r = x({
        client: a
      }), Object.assign(e.style, g({
        clientRect: r
      }));
    },
    // using capture so we are less likely to be impacted by event stopping
    options: {
      capture: !0
    }
  });
  return function(c) {
    var a = c.current;
    if (o(), D({
      client: a,
      clientRect: r
    })) {
      e.remove();
      return;
    }
    function i() {
      s(), e.remove();
    }
    var s = y.bindAll(window, [
      {
        type: "pointerdown",
        listener: i
      },
      {
        type: "pointermove",
        listener: i
      },
      {
        type: "focusin",
        listener: i
      },
      {
        type: "focusout",
        listener: i
      },
      // a 'pointerdown' should happen before 'dragstart', but just being super safe
      {
        type: "dragstart",
        listener: i
      },
      // if the user has dragged something out of the window
      // and then is dragging something back into the window
      // the first events we will see are "dragenter" (and then "dragover").
      // So if we see any of these we need to clear the post drag fix.
      {
        type: "dragenter",
        listener: i
      },
      {
        type: "dragover",
        listener: i
      }
      // Not adding a "wheel" event listener, as "wheel" by itself does not
      // resolve the bug.
    ], {
      // Using `capture` so less likely to be impacted by other code stopping events
      capture: !0
    });
  };
}
function z() {
  var t = null;
  function n() {
    return t = null, y.bind(window, {
      type: "pointermove",
      listener: function(o) {
        t = {
          x: o.clientX,
          y: o.clientY
        };
      },
      // listening for pointer move in capture phase
      // so we are less likely to be impacted by events being stopped.
      options: {
        capture: !0
      }
    });
  }
  function e() {
    var r = null;
    return function(u) {
      var c = u.eventName, a = u.payload;
      if (c === "onDragStart") {
        var i = a.location.initial.input, s = t ?? {
          x: i.clientX,
          y: i.clientY
        };
        r = S({
          initial: s
        });
      }
      if (c === "onDrop") {
        var p, f = a.location.current.input;
        (p = r) === null || p === void 0 || p({
          current: {
            x: f.clientX,
            y: f.clientY
          }
        }), r = null, t = null;
      }
    };
  }
  return {
    bindEvents: n,
    getOnPostDispatch: e
  };
}
export {
  z as makeHoneyPotFix
};
