// @atlaskit/pragmatic-drag-and-drop-hitbox@1.1.0/dist/esm/closest-edge.js
import l from "../../../../../../@babel_runtime@7.28.6/node_modules/@babel/runtime/helpers/esm/defineProperty.js";
function a(r, t) {
  var e = Object.keys(r);
  if (Object.getOwnPropertySymbols) {
    var n = Object.getOwnPropertySymbols(r);
    t && (n = n.filter(function(u) {
      return Object.getOwnPropertyDescriptor(r, u).enumerable;
    })), e.push.apply(e, n);
  }
  return e;
}
function c(r) {
  for (var t = 1; t < arguments.length; t++) {
    var e = arguments[t] != null ? arguments[t] : {};
    t % 2 ? a(Object(e), !0).forEach(function(n) {
      l(r, n, e[n]);
    }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(r, Object.getOwnPropertyDescriptors(e)) : a(Object(e)).forEach(function(n) {
      Object.defineProperty(r, n, Object.getOwnPropertyDescriptor(e, n));
    });
  }
  return r;
}
var y = {
  top: function(t, e) {
    return Math.abs(e.y - t.top);
  },
  right: function(t, e) {
    return Math.abs(t.right - e.x);
  },
  bottom: function(t, e) {
    return Math.abs(t.bottom - e.y);
  },
  left: function(t, e) {
    return Math.abs(e.x - t.left);
  }
}, s = /* @__PURE__ */ Symbol("closestEdge");
function m(r, t) {
  var e, n, u = t.element, i = t.input, g = t.allowedEdges, p = {
    x: i.clientX,
    y: i.clientY
  }, f = u.getBoundingClientRect(), b = g.map(function(o) {
    return {
      edge: o,
      value: y[o](f, p)
    };
  }), d = (e = (n = b.sort(function(o, v) {
    return o.value - v.value;
  })[0]) === null || n === void 0 ? void 0 : n.edge) !== null && e !== void 0 ? e : null;
  return c(c({}, r), {}, l({}, s, d));
}
function h(r) {
  var t;
  return (t = r[s]) !== null && t !== void 0 ? t : null;
}
export {
  m as attachClosestEdge,
  h as extractClosestEdge
};
