// react-day-picker@8.10.1/dist/index.esm.js
import { jsx as s, jsxs as P, Fragment as L } from "react/jsx-runtime";
import { useState as B, useEffect as V, createContext as I, useContext as T, useLayoutEffect as Ze, forwardRef as Ge, useRef as qe } from "react";
import { endOfMonth as ie } from "../../../../date-fns@3.6.0/node_modules/date-fns/endOfMonth.js";
import { enUS as Xe } from "../../../../date-fns@3.6.0/node_modules/date-fns/locale/en-US.js";
import { addMonths as S } from "../../../../date-fns@3.6.0/node_modules/date-fns/addMonths.js";
import { differenceInCalendarMonths as Y } from "../../../../date-fns@3.6.0/node_modules/date-fns/differenceInCalendarMonths.js";
import { isSameMonth as le } from "../../../../date-fns@3.6.0/node_modules/date-fns/isSameMonth.js";
import { isSameDay as N } from "../../../../date-fns@3.6.0/node_modules/date-fns/isSameDay.js";
import { addDays as k } from "../../../../date-fns@3.6.0/node_modules/date-fns/addDays.js";
import { subDays as Q } from "../../../../date-fns@3.6.0/node_modules/date-fns/subDays.js";
import { differenceInCalendarDays as W } from "../../../../date-fns@3.6.0/node_modules/date-fns/differenceInCalendarDays.js";
import { getWeeksInMonth as Je } from "../../../../date-fns@3.6.0/node_modules/date-fns/getWeeksInMonth.js";
import { addWeeks as ae } from "../../../../date-fns@3.6.0/node_modules/date-fns/addWeeks.js";
import { getUnixTime as Qe } from "../../../../date-fns@3.6.0/node_modules/date-fns/getUnixTime.js";
import { addYears as $e } from "../../../../date-fns@3.6.0/node_modules/date-fns/addYears.js";
import { max as en } from "../../../../date-fns@3.6.0/node_modules/date-fns/max.js";
import { min as nn } from "../../../../date-fns@3.6.0/node_modules/date-fns/min.js";
import { isSameYear as rn } from "../../../../date-fns@3.6.0/node_modules/date-fns/isSameYear.js";
import { setMonth as $ } from "../../../../date-fns@3.6.0/node_modules/date-fns/setMonth.js";
import { setYear as pe } from "../../../../date-fns@3.6.0/node_modules/date-fns/setYear.js";
import { endOfISOWeek as Me } from "../../../../date-fns@3.6.0/node_modules/date-fns/endOfISOWeek.js";
import { endOfWeek as Ne } from "../../../../date-fns@3.6.0/node_modules/date-fns/endOfWeek.js";
import { startOfISOWeek as se } from "../../../../date-fns@3.6.0/node_modules/date-fns/startOfISOWeek.js";
import { getISOWeek as tn } from "../../../../date-fns@3.6.0/node_modules/date-fns/getISOWeek.js";
import { getWeek as an } from "../../../../date-fns@3.6.0/node_modules/date-fns/getWeek.js";
import { isDate as ke } from "../../../../date-fns@3.6.0/node_modules/date-fns/isDate.js";
import { startOfMonth as C } from "../../../../date-fns@3.6.0/node_modules/date-fns/startOfMonth.js";
import { startOfDay as be } from "../../../../date-fns@3.6.0/node_modules/date-fns/startOfDay.js";
import { isBefore as Ce } from "../../../../date-fns@3.6.0/node_modules/date-fns/isBefore.js";
import { formatDate as E } from "../../../../date-fns@3.6.0/node_modules/date-fns/format.js";
import { isAfter as oe } from "../../../../date-fns@3.6.0/node_modules/date-fns/isAfter.js";
import { startOfYear as on } from "../../../../date-fns@3.6.0/node_modules/date-fns/startOfYear.js";
import { startOfWeek as de } from "../../../../date-fns@3.6.0/node_modules/date-fns/startOfWeek.js";
var y = function() {
  return y = Object.assign || function(r) {
    for (var n, t = 1, a = arguments.length; t < a; t++) {
      n = arguments[t];
      for (var o in n) Object.prototype.hasOwnProperty.call(n, o) && (r[o] = n[o]);
    }
    return r;
  }, y.apply(this, arguments);
};
function ln(e, r) {
  var n = {};
  for (var t in e) Object.prototype.hasOwnProperty.call(e, t) && r.indexOf(t) < 0 && (n[t] = e[t]);
  if (e != null && typeof Object.getOwnPropertySymbols == "function")
    for (var a = 0, t = Object.getOwnPropertySymbols(e); a < t.length; a++)
      r.indexOf(t[a]) < 0 && Object.prototype.propertyIsEnumerable.call(e, t[a]) && (n[t[a]] = e[t[a]]);
  return n;
}
function xe(e, r, n) {
  for (var t = 0, a = r.length, o; t < a; t++)
    (o || !(t in r)) && (o || (o = Array.prototype.slice.call(r, 0, t)), o[t] = r[t]);
  return e.concat(o || Array.prototype.slice.call(r));
}
function A(e) {
  return e.mode === "multiple";
}
function j(e) {
  return e.mode === "range";
}
function z(e) {
  return e.mode === "single";
}
var sn = {
  root: "rdp",
  multiple_months: "rdp-multiple_months",
  with_weeknumber: "rdp-with_weeknumber",
  vhidden: "rdp-vhidden",
  button_reset: "rdp-button_reset",
  button: "rdp-button",
  caption: "rdp-caption",
  caption_start: "rdp-caption_start",
  caption_end: "rdp-caption_end",
  caption_between: "rdp-caption_between",
  caption_label: "rdp-caption_label",
  caption_dropdowns: "rdp-caption_dropdowns",
  dropdown: "rdp-dropdown",
  dropdown_month: "rdp-dropdown_month",
  dropdown_year: "rdp-dropdown_year",
  dropdown_icon: "rdp-dropdown_icon",
  months: "rdp-months",
  month: "rdp-month",
  table: "rdp-table",
  tbody: "rdp-tbody",
  tfoot: "rdp-tfoot",
  head: "rdp-head",
  head_row: "rdp-head_row",
  head_cell: "rdp-head_cell",
  nav: "rdp-nav",
  nav_button: "rdp-nav_button",
  nav_button_previous: "rdp-nav_button_previous",
  nav_button_next: "rdp-nav_button_next",
  nav_icon: "rdp-nav_icon",
  row: "rdp-row",
  weeknumber: "rdp-weeknumber",
  cell: "rdp-cell",
  day: "rdp-day",
  day_today: "rdp-day_today",
  day_outside: "rdp-day_outside",
  day_selected: "rdp-day_selected",
  day_disabled: "rdp-day_disabled",
  day_hidden: "rdp-day_hidden",
  day_range_start: "rdp-day_range_start",
  day_range_end: "rdp-day_range_end",
  day_range_middle: "rdp-day_range_middle"
};
function dn(e, r) {
  return E(e, "LLLL y", r);
}
function un(e, r) {
  return E(e, "d", r);
}
function cn(e, r) {
  return E(e, "LLLL", r);
}
function fn(e) {
  return "".concat(e);
}
function vn(e, r) {
  return E(e, "cccccc", r);
}
function mn(e, r) {
  return E(e, "yyyy", r);
}
var hn = /* @__PURE__ */ Object.freeze({
  __proto__: null,
  formatCaption: dn,
  formatDay: un,
  formatMonthCaption: cn,
  formatWeekNumber: fn,
  formatWeekdayName: vn,
  formatYearCaption: mn
}), yn = function(e, r, n) {
  return E(e, "do MMMM (EEEE)", n);
}, pn = function() {
  return "Month: ";
}, bn = function() {
  return "Go to next month";
}, _n = function() {
  return "Go to previous month";
}, gn = function(e, r) {
  return E(e, "cccc", r);
}, Dn = function(e) {
  return "Week n. ".concat(e);
}, wn = function() {
  return "Year: ";
}, Mn = /* @__PURE__ */ Object.freeze({
  __proto__: null,
  labelDay: yn,
  labelMonthDropdown: pn,
  labelNext: bn,
  labelPrevious: _n,
  labelWeekNumber: Dn,
  labelWeekday: gn,
  labelYearDropdown: wn
});
function Nn() {
  var e = "buttons", r = sn, n = Xe, t = {}, a = {}, o = 1, i = {}, l = /* @__PURE__ */ new Date();
  return {
    captionLayout: e,
    classNames: r,
    formatters: hn,
    labels: Mn,
    locale: n,
    modifiersClassNames: t,
    modifiers: a,
    numberOfMonths: o,
    styles: i,
    today: l,
    mode: "default"
  };
}
function kn(e) {
  var r = e.fromYear, n = e.toYear, t = e.fromMonth, a = e.toMonth, o = e.fromDate, i = e.toDate;
  return t ? o = C(t) : r && (o = new Date(r, 0, 1)), a ? i = ie(a) : n && (i = new Date(n, 11, 31)), {
    fromDate: o ? be(o) : void 0,
    toDate: i ? be(i) : void 0
  };
}
var Pe = I(void 0);
function Cn(e) {
  var r, n = e.initialProps, t = Nn(), a = kn(n), o = a.fromDate, i = a.toDate, l = (r = n.captionLayout) !== null && r !== void 0 ? r : t.captionLayout;
  l !== "buttons" && (!o || !i) && (l = "buttons");
  var d;
  (z(n) || A(n) || j(n)) && (d = n.onSelect);
  var c = y(y(y({}, t), n), { captionLayout: l, classNames: y(y({}, t.classNames), n.classNames), components: y({}, n.components), formatters: y(y({}, t.formatters), n.formatters), fromDate: o, labels: y(y({}, t.labels), n.labels), mode: n.mode || t.mode, modifiers: y(y({}, t.modifiers), n.modifiers), modifiersClassNames: y(y({}, t.modifiersClassNames), n.modifiersClassNames), onSelect: d, styles: y(y({}, t.styles), n.styles), toDate: i });
  return s(Pe.Provider, { value: c, children: e.children });
}
function D() {
  var e = T(Pe);
  if (!e)
    throw new Error("useDayPicker must be used within a DayPickerProvider.");
  return e;
}
function Se(e) {
  var r = D(), n = r.locale, t = r.classNames, a = r.styles, o = r.formatters.formatCaption;
  return s("div", { className: t.caption_label, style: a.caption_label, "aria-live": "polite", role: "presentation", id: e.id, children: o(e.displayMonth, { locale: n }) });
}
function xn(e) {
  return s("svg", y({ width: "8px", height: "8px", viewBox: "0 0 120 120", "data-testid": "iconDropdown" }, e, { children: s("path", { d: "M4.22182541,48.2218254 C8.44222828,44.0014225 15.2388494,43.9273804 19.5496459,47.9996989 L19.7781746,48.2218254 L60,88.443 L100.221825,48.2218254 C104.442228,44.0014225 111.238849,43.9273804 115.549646,47.9996989 L115.778175,48.2218254 C119.998577,52.4422283 120.07262,59.2388494 116.000301,63.5496459 L115.778175,63.7781746 L67.7781746,111.778175 C63.5577717,115.998577 56.7611506,116.07262 52.4503541,112.000301 L52.2218254,111.778175 L4.22182541,63.7781746 C-0.0739418023,59.4824074 -0.0739418023,52.5175926 4.22182541,48.2218254 Z", fill: "currentColor", fillRule: "nonzero" }) }));
}
function Oe(e) {
  var r, n, t = e.onChange, a = e.value, o = e.children, i = e.caption, l = e.className, d = e.style, c = D(), u = (n = (r = c.components) === null || r === void 0 ? void 0 : r.IconDropdown) !== null && n !== void 0 ? n : xn;
  return P("div", { className: l, style: d, children: [s("span", { className: c.classNames.vhidden, children: e["aria-label"] }), s("select", { name: e.name, "aria-label": e["aria-label"], className: c.classNames.dropdown, style: c.styles.dropdown, value: a, onChange: t, children: o }), P("div", { className: c.classNames.caption_label, style: c.styles.caption_label, "aria-hidden": "true", children: [i, s(u, { className: c.classNames.dropdown_icon, style: c.styles.dropdown_icon })] })] });
}
function Pn(e) {
  var r, n = D(), t = n.fromDate, a = n.toDate, o = n.styles, i = n.locale, l = n.formatters.formatMonthCaption, d = n.classNames, c = n.components, u = n.labels.labelMonthDropdown;
  if (!t)
    return s(L, {});
  if (!a)
    return s(L, {});
  var f = [];
  if (rn(t, a))
    for (var m = C(t), h = t.getMonth(); h <= a.getMonth(); h++)
      f.push($(m, h));
  else
    for (var m = C(/* @__PURE__ */ new Date()), h = 0; h <= 11; h++)
      f.push($(m, h));
  var b = function(_) {
    var M = Number(_.target.value), w = $(C(e.displayMonth), M);
    e.onChange(w);
  }, g = (r = c?.Dropdown) !== null && r !== void 0 ? r : Oe;
  return s(g, { name: "months", "aria-label": u(), className: d.dropdown_month, style: o.dropdown_month, onChange: b, value: e.displayMonth.getMonth(), caption: l(e.displayMonth, { locale: i }), children: f.map(function(_) {
    return s("option", { value: _.getMonth(), children: l(_, { locale: i }) }, _.getMonth());
  }) });
}
function Sn(e) {
  var r, n = e.displayMonth, t = D(), a = t.fromDate, o = t.toDate, i = t.locale, l = t.styles, d = t.classNames, c = t.components, u = t.formatters.formatYearCaption, f = t.labels.labelYearDropdown, m = [];
  if (!a)
    return s(L, {});
  if (!o)
    return s(L, {});
  for (var h = a.getFullYear(), b = o.getFullYear(), g = h; g <= b; g++)
    m.push(pe(on(/* @__PURE__ */ new Date()), g));
  var _ = function(w) {
    var F = pe(C(n), Number(w.target.value));
    e.onChange(F);
  }, M = (r = c?.Dropdown) !== null && r !== void 0 ? r : Oe;
  return s(M, { name: "years", "aria-label": f(), className: d.dropdown_year, style: l.dropdown_year, onChange: _, value: n.getFullYear(), caption: u(n, { locale: i }), children: m.map(function(w) {
    return s("option", { value: w.getFullYear(), children: u(w, { locale: i }) }, w.getFullYear());
  }) });
}
function On(e, r) {
  var n = B(e), t = n[0], a = n[1], o = r === void 0 ? t : r;
  return [o, a];
}
function Wn(e) {
  var r = e.month, n = e.defaultMonth, t = e.today, a = r || n || t || /* @__PURE__ */ new Date(), o = e.toDate, i = e.fromDate, l = e.numberOfMonths, d = l === void 0 ? 1 : l;
  if (o && Y(o, a) < 0) {
    var c = -1 * (d - 1);
    a = S(o, c);
  }
  return i && Y(a, i) < 0 && (a = i), C(a);
}
function Ln() {
  var e = D(), r = Wn(e), n = On(r, e.month), t = n[0], a = n[1], o = function(i) {
    var l;
    if (!e.disableNavigation) {
      var d = C(i);
      a(d), (l = e.onMonthChange) === null || l === void 0 || l.call(e, d);
    }
  };
  return [t, o];
}
function In(e, r) {
  for (var n = r.reverseMonths, t = r.numberOfMonths, a = C(e), o = C(S(a, t)), i = Y(o, a), l = [], d = 0; d < i; d++) {
    var c = S(a, d);
    l.push(c);
  }
  return n && (l = l.reverse()), l;
}
function Tn(e, r) {
  if (!r.disableNavigation) {
    var n = r.toDate, t = r.pagedNavigation, a = r.numberOfMonths, o = a === void 0 ? 1 : a, i = t ? o : 1, l = C(e);
    if (!n)
      return S(l, i);
    var d = Y(n, e);
    if (!(d < o))
      return S(l, i);
  }
}
function En(e, r) {
  if (!r.disableNavigation) {
    var n = r.fromDate, t = r.pagedNavigation, a = r.numberOfMonths, o = a === void 0 ? 1 : a, i = t ? o : 1, l = C(e);
    if (!n)
      return S(l, -i);
    var d = Y(l, n);
    if (!(d <= 0))
      return S(l, -i);
  }
}
var We = I(void 0);
function Fn(e) {
  var r = D(), n = Ln(), t = n[0], a = n[1], o = In(t, r), i = Tn(t, r), l = En(t, r), d = function(f) {
    return o.some(function(m) {
      return le(f, m);
    });
  }, c = function(f, m) {
    d(f) || (m && Ce(f, m) ? a(S(f, 1 + r.numberOfMonths * -1)) : a(f));
  }, u = {
    currentMonth: t,
    displayMonths: o,
    goToMonth: a,
    goToDate: c,
    previousMonth: l,
    nextMonth: i,
    isDateDisplayed: d
  };
  return s(We.Provider, { value: u, children: e.children });
}
function H() {
  var e = T(We);
  if (!e)
    throw new Error("useNavigation must be used within a NavigationProvider");
  return e;
}
function _e(e) {
  var r, n = D(), t = n.classNames, a = n.styles, o = n.components, i = H().goToMonth, l = function(u) {
    i(S(u, e.displayIndex ? -e.displayIndex : 0));
  }, d = (r = o?.CaptionLabel) !== null && r !== void 0 ? r : Se, c = s(d, { id: e.id, displayMonth: e.displayMonth });
  return P("div", { className: t.caption_dropdowns, style: a.caption_dropdowns, children: [s("div", { className: t.vhidden, children: c }), s(Pn, { onChange: l, displayMonth: e.displayMonth }), s(Sn, { onChange: l, displayMonth: e.displayMonth })] });
}
function Rn(e) {
  return s("svg", y({ width: "16px", height: "16px", viewBox: "0 0 120 120" }, e, { children: s("path", { d: "M69.490332,3.34314575 C72.6145263,0.218951416 77.6798462,0.218951416 80.8040405,3.34314575 C83.8617626,6.40086786 83.9268205,11.3179931 80.9992143,14.4548388 L80.8040405,14.6568542 L35.461,60 L80.8040405,105.343146 C83.8617626,108.400868 83.9268205,113.317993 80.9992143,116.454839 L80.8040405,116.656854 C77.7463184,119.714576 72.8291931,119.779634 69.6923475,116.852028 L69.490332,116.656854 L18.490332,65.6568542 C15.4326099,62.5991321 15.367552,57.6820069 18.2951583,54.5451612 L18.490332,54.3431458 L69.490332,3.34314575 Z", fill: "currentColor", fillRule: "nonzero" }) }));
}
function Bn(e) {
  return s("svg", y({ width: "16px", height: "16px", viewBox: "0 0 120 120" }, e, { children: s("path", { d: "M49.8040405,3.34314575 C46.6798462,0.218951416 41.6145263,0.218951416 38.490332,3.34314575 C35.4326099,6.40086786 35.367552,11.3179931 38.2951583,14.4548388 L38.490332,14.6568542 L83.8333725,60 L38.490332,105.343146 C35.4326099,108.400868 35.367552,113.317993 38.2951583,116.454839 L38.490332,116.656854 C41.5480541,119.714576 46.4651794,119.779634 49.602025,116.852028 L49.8040405,116.656854 L100.804041,65.6568542 C103.861763,62.5991321 103.926821,57.6820069 100.999214,54.5451612 L100.804041,54.3431458 L49.8040405,3.34314575 Z", fill: "currentColor" }) }));
}
var U = Ge(function(e, r) {
  var n = D(), t = n.classNames, a = n.styles, o = [t.button_reset, t.button];
  e.className && o.push(e.className);
  var i = o.join(" "), l = y(y({}, a.button_reset), a.button);
  return e.style && Object.assign(l, e.style), s("button", y({}, e, { ref: r, type: "button", className: i, style: l }));
});
function Yn(e) {
  var r, n, t = D(), a = t.dir, o = t.locale, i = t.classNames, l = t.styles, d = t.labels, c = d.labelPrevious, u = d.labelNext, f = t.components;
  if (!e.nextMonth && !e.previousMonth)
    return s(L, {});
  var m = c(e.previousMonth, { locale: o }), h = [
    i.nav_button,
    i.nav_button_previous
  ].join(" "), b = u(e.nextMonth, { locale: o }), g = [
    i.nav_button,
    i.nav_button_next
  ].join(" "), _ = (r = f?.IconRight) !== null && r !== void 0 ? r : Bn, M = (n = f?.IconLeft) !== null && n !== void 0 ? n : Rn;
  return P("div", { className: i.nav, style: l.nav, children: [!e.hidePrevious && s(U, { name: "previous-month", "aria-label": m, className: h, style: l.nav_button_previous, disabled: !e.previousMonth, onClick: e.onPreviousClick, children: a === "rtl" ? s(_, { className: i.nav_icon, style: l.nav_icon }) : s(M, { className: i.nav_icon, style: l.nav_icon }) }), !e.hideNext && s(U, { name: "next-month", "aria-label": b, className: g, style: l.nav_button_next, disabled: !e.nextMonth, onClick: e.onNextClick, children: a === "rtl" ? s(M, { className: i.nav_icon, style: l.nav_icon }) : s(_, { className: i.nav_icon, style: l.nav_icon }) })] });
}
function ge(e) {
  var r = D().numberOfMonths, n = H(), t = n.previousMonth, a = n.nextMonth, o = n.goToMonth, i = n.displayMonths, l = i.findIndex(function(b) {
    return le(e.displayMonth, b);
  }), d = l === 0, c = l === i.length - 1, u = r > 1 && (d || !c), f = r > 1 && (c || !d), m = function() {
    t && o(t);
  }, h = function() {
    a && o(a);
  };
  return s(Yn, { displayMonth: e.displayMonth, hideNext: u, hidePrevious: f, nextMonth: a, previousMonth: t, onPreviousClick: m, onNextClick: h });
}
function An(e) {
  var r, n = D(), t = n.classNames, a = n.disableNavigation, o = n.styles, i = n.captionLayout, l = n.components, d = (r = l?.CaptionLabel) !== null && r !== void 0 ? r : Se, c;
  return a ? c = s(d, { id: e.id, displayMonth: e.displayMonth }) : i === "dropdown" ? c = s(_e, { displayMonth: e.displayMonth, id: e.id }) : i === "dropdown-buttons" ? c = P(L, { children: [s(_e, { displayMonth: e.displayMonth, displayIndex: e.displayIndex, id: e.id }), s(ge, { displayMonth: e.displayMonth, displayIndex: e.displayIndex, id: e.id })] }) : c = P(L, { children: [s(d, { id: e.id, displayMonth: e.displayMonth, displayIndex: e.displayIndex }), s(ge, { displayMonth: e.displayMonth, id: e.id })] }), s("div", { className: t.caption, style: o.caption, children: c });
}
function jn(e) {
  var r = D(), n = r.footer, t = r.styles, a = r.classNames.tfoot;
  return n ? s("tfoot", { className: a, style: t.tfoot, children: s("tr", { children: s("td", { colSpan: 8, children: n }) }) }) : s(L, {});
}
function Hn(e, r, n) {
  for (var t = n ? se(/* @__PURE__ */ new Date()) : de(/* @__PURE__ */ new Date(), { locale: e, weekStartsOn: r }), a = [], o = 0; o < 7; o++) {
    var i = k(t, o);
    a.push(i);
  }
  return a;
}
function Kn() {
  var e = D(), r = e.classNames, n = e.styles, t = e.showWeekNumber, a = e.locale, o = e.weekStartsOn, i = e.ISOWeek, l = e.formatters.formatWeekdayName, d = e.labels.labelWeekday, c = Hn(a, o, i);
  return P("tr", { style: n.head_row, className: r.head_row, children: [t && s("td", { style: n.head_cell, className: r.head_cell }), c.map(function(u, f) {
    return s("th", { scope: "col", className: r.head_cell, style: n.head_cell, "aria-label": d(u, { locale: a }), children: l(u, { locale: a }) }, f);
  })] });
}
function Un() {
  var e, r = D(), n = r.classNames, t = r.styles, a = r.components, o = (e = a?.HeadRow) !== null && e !== void 0 ? e : Kn;
  return s("thead", { style: t.head, className: n.head, children: s(o, {}) });
}
function Vn(e) {
  var r = D(), n = r.locale, t = r.formatters.formatDay;
  return s(L, { children: t(e.date, { locale: n }) });
}
var ue = I(void 0);
function zn(e) {
  if (!A(e.initialProps)) {
    var r = {
      selected: void 0,
      modifiers: {
        disabled: []
      }
    };
    return s(ue.Provider, { value: r, children: e.children });
  }
  return s(Zn, { initialProps: e.initialProps, children: e.children });
}
function Zn(e) {
  var r = e.initialProps, n = e.children, t = r.selected, a = r.min, o = r.max, i = function(c, u, f) {
    var m, h;
    (m = r.onDayClick) === null || m === void 0 || m.call(r, c, u, f);
    var b = !!(u.selected && a && t?.length === a);
    if (!b) {
      var g = !!(!u.selected && o && t?.length === o);
      if (!g) {
        var _ = t ? xe([], t) : [];
        if (u.selected) {
          var M = _.findIndex(function(w) {
            return N(c, w);
          });
          _.splice(M, 1);
        } else
          _.push(c);
        (h = r.onSelect) === null || h === void 0 || h.call(r, _, c, u, f);
      }
    }
  }, l = {
    disabled: []
  };
  t && l.disabled.push(function(c) {
    var u = o && t.length > o - 1, f = t.some(function(m) {
      return N(m, c);
    });
    return !!(u && !f);
  });
  var d = {
    selected: t,
    onDayClick: i,
    modifiers: l
  };
  return s(ue.Provider, { value: d, children: n });
}
function ce() {
  var e = T(ue);
  if (!e)
    throw new Error("useSelectMultiple must be used within a SelectMultipleProvider");
  return e;
}
function Gn(e, r) {
  var n = r || {}, t = n.from, a = n.to;
  return t && a ? N(a, e) && N(t, e) ? void 0 : N(a, e) ? { from: a, to: void 0 } : N(t, e) ? void 0 : oe(t, e) ? { from: e, to: a } : { from: t, to: e } : a ? oe(e, a) ? { from: a, to: e } : { from: e, to: a } : t ? Ce(e, t) ? { from: e, to: t } : { from: t, to: e } : { from: e, to: void 0 };
}
var fe = I(void 0);
function qn(e) {
  if (!j(e.initialProps)) {
    var r = {
      selected: void 0,
      modifiers: {
        range_start: [],
        range_end: [],
        range_middle: [],
        disabled: []
      }
    };
    return s(fe.Provider, { value: r, children: e.children });
  }
  return s(Xn, { initialProps: e.initialProps, children: e.children });
}
function Xn(e) {
  var r = e.initialProps, n = e.children, t = r.selected, a = t || {}, o = a.from, i = a.to, l = r.min, d = r.max, c = function(h, b, g) {
    var _, M;
    (_ = r.onDayClick) === null || _ === void 0 || _.call(r, h, b, g);
    var w = Gn(h, t);
    (M = r.onSelect) === null || M === void 0 || M.call(r, w, h, b, g);
  }, u = {
    range_start: [],
    range_end: [],
    range_middle: [],
    disabled: []
  };
  if (o ? (u.range_start = [o], i ? (u.range_end = [i], N(o, i) || (u.range_middle = [
    {
      after: o,
      before: i
    }
  ])) : u.range_end = [o]) : i && (u.range_start = [i], u.range_end = [i]), l && (o && !i && u.disabled.push({
    after: Q(o, l - 1),
    before: k(o, l - 1)
  }), o && i && u.disabled.push({
    after: o,
    before: k(o, l - 1)
  }), !o && i && u.disabled.push({
    after: Q(i, l - 1),
    before: k(i, l - 1)
  })), d) {
    if (o && !i && (u.disabled.push({
      before: k(o, -d + 1)
    }), u.disabled.push({
      after: k(o, d - 1)
    })), o && i) {
      var f = W(i, o) + 1, m = d - f;
      u.disabled.push({
        before: Q(o, m)
      }), u.disabled.push({
        after: k(i, m)
      });
    }
    !o && i && (u.disabled.push({
      before: k(i, -d + 1)
    }), u.disabled.push({
      after: k(i, d - 1)
    }));
  }
  return s(fe.Provider, { value: { selected: t, onDayClick: c, modifiers: u }, children: n });
}
function ve() {
  var e = T(fe);
  if (!e)
    throw new Error("useSelectRange must be used within a SelectRangeProvider");
  return e;
}
function K(e) {
  return Array.isArray(e) ? xe([], e) : e !== void 0 ? [e] : [];
}
function Jn(e) {
  var r = {};
  return Object.entries(e).forEach(function(n) {
    var t = n[0], a = n[1];
    r[t] = K(a);
  }), r;
}
var x;
(function(e) {
  e.Outside = "outside", e.Disabled = "disabled", e.Selected = "selected", e.Hidden = "hidden", e.Today = "today", e.RangeStart = "range_start", e.RangeEnd = "range_end", e.RangeMiddle = "range_middle";
})(x || (x = {}));
var Qn = x.Selected, O = x.Disabled, $n = x.Hidden, er = x.Today, ee = x.RangeEnd, ne = x.RangeMiddle, re = x.RangeStart, nr = x.Outside;
function rr(e, r, n) {
  var t, a = (t = {}, t[Qn] = K(e.selected), t[O] = K(e.disabled), t[$n] = K(e.hidden), t[er] = [e.today], t[ee] = [], t[ne] = [], t[re] = [], t[nr] = [], t);
  return e.fromDate && a[O].push({ before: e.fromDate }), e.toDate && a[O].push({ after: e.toDate }), A(e) ? a[O] = a[O].concat(r.modifiers[O]) : j(e) && (a[O] = a[O].concat(n.modifiers[O]), a[re] = n.modifiers[re], a[ne] = n.modifiers[ne], a[ee] = n.modifiers[ee]), a;
}
var Le = I(void 0);
function tr(e) {
  var r = D(), n = ce(), t = ve(), a = rr(r, n, t), o = Jn(r.modifiers), i = y(y({}, a), o);
  return s(Le.Provider, { value: i, children: e.children });
}
function Ie() {
  var e = T(Le);
  if (!e)
    throw new Error("useModifiers must be used within a ModifiersProvider");
  return e;
}
function ar(e) {
  return !!(e && typeof e == "object" && "before" in e && "after" in e);
}
function or(e) {
  return !!(e && typeof e == "object" && "from" in e);
}
function ir(e) {
  return !!(e && typeof e == "object" && "after" in e);
}
function lr(e) {
  return !!(e && typeof e == "object" && "before" in e);
}
function sr(e) {
  return !!(e && typeof e == "object" && "dayOfWeek" in e);
}
function dr(e, r) {
  var n, t = r.from, a = r.to;
  if (t && a) {
    var o = W(a, t) < 0;
    o && (n = [a, t], t = n[0], a = n[1]);
    var i = W(e, t) >= 0 && W(a, e) >= 0;
    return i;
  }
  return a ? N(a, e) : t ? N(t, e) : !1;
}
function ur(e) {
  return ke(e);
}
function cr(e) {
  return Array.isArray(e) && e.every(ke);
}
function fr(e, r) {
  return r.some(function(n) {
    if (typeof n == "boolean")
      return n;
    if (ur(n))
      return N(e, n);
    if (cr(n))
      return n.includes(e);
    if (or(n))
      return dr(e, n);
    if (sr(n))
      return n.dayOfWeek.includes(e.getDay());
    if (ar(n)) {
      var t = W(n.before, e), a = W(n.after, e), o = t > 0, i = a < 0, l = oe(n.before, n.after);
      return l ? i && o : o || i;
    }
    return ir(n) ? W(e, n.after) > 0 : lr(n) ? W(n.before, e) > 0 : typeof n == "function" ? n(e) : !1;
  });
}
function me(e, r, n) {
  var t = Object.keys(r).reduce(function(o, i) {
    var l = r[i];
    return fr(e, l) && o.push(i), o;
  }, []), a = {};
  return t.forEach(function(o) {
    return a[o] = !0;
  }), n && !le(e, n) && (a.outside = !0), a;
}
function vr(e, r) {
  for (var n = C(e[0]), t = ie(e[e.length - 1]), a, o, i = n; i <= t; ) {
    var l = me(i, r), d = !l.disabled && !l.hidden;
    if (!d) {
      i = k(i, 1);
      continue;
    }
    if (l.selected)
      return i;
    l.today && !o && (o = i), a || (a = i), i = k(i, 1);
  }
  return o || a;
}
var mr = 365;
function Te(e, r) {
  var n = r.moveBy, t = r.direction, a = r.context, o = r.modifiers, i = r.retry, l = i === void 0 ? { count: 0, lastFocused: e } : i, d = a.weekStartsOn, c = a.fromDate, u = a.toDate, f = a.locale, m = {
    day: k,
    week: ae,
    month: S,
    year: $e,
    startOfWeek: function(_) {
      return a.ISOWeek ? se(_) : de(_, { locale: f, weekStartsOn: d });
    },
    endOfWeek: function(_) {
      return a.ISOWeek ? Me(_) : Ne(_, { locale: f, weekStartsOn: d });
    }
  }, h = m[n](e, t === "after" ? 1 : -1);
  t === "before" && c ? h = en([c, h]) : t === "after" && u && (h = nn([u, h]));
  var b = !0;
  if (o) {
    var g = me(h, o);
    b = !g.disabled && !g.hidden;
  }
  return b ? h : l.count > mr ? l.lastFocused : Te(h, {
    moveBy: n,
    direction: t,
    context: a,
    modifiers: o,
    retry: y(y({}, l), { count: l.count + 1 })
  });
}
var Ee = I(void 0);
function hr(e) {
  var r = H(), n = Ie(), t = B(), a = t[0], o = t[1], i = B(), l = i[0], d = i[1], c = vr(r.displayMonths, n), u = a ?? (l && r.isDateDisplayed(l)) ? l : c, f = function() {
    d(a), o(void 0);
  }, m = function(_) {
    o(_);
  }, h = D(), b = function(_, M) {
    if (a) {
      var w = Te(a, {
        moveBy: _,
        direction: M,
        context: h,
        modifiers: n
      });
      N(a, w) || (r.goToDate(w, a), m(w));
    }
  }, g = {
    focusedDay: a,
    focusTarget: u,
    blur: f,
    focus: m,
    focusDayAfter: function() {
      return b("day", "after");
    },
    focusDayBefore: function() {
      return b("day", "before");
    },
    focusWeekAfter: function() {
      return b("week", "after");
    },
    focusWeekBefore: function() {
      return b("week", "before");
    },
    focusMonthBefore: function() {
      return b("month", "before");
    },
    focusMonthAfter: function() {
      return b("month", "after");
    },
    focusYearBefore: function() {
      return b("year", "before");
    },
    focusYearAfter: function() {
      return b("year", "after");
    },
    focusStartOfWeek: function() {
      return b("startOfWeek", "before");
    },
    focusEndOfWeek: function() {
      return b("endOfWeek", "after");
    }
  };
  return s(Ee.Provider, { value: g, children: e.children });
}
function he() {
  var e = T(Ee);
  if (!e)
    throw new Error("useFocusContext must be used within a FocusProvider");
  return e;
}
function yr(e, r) {
  var n = Ie(), t = me(e, n, r);
  return t;
}
var ye = I(void 0);
function pr(e) {
  if (!z(e.initialProps)) {
    var r = {
      selected: void 0
    };
    return s(ye.Provider, { value: r, children: e.children });
  }
  return s(br, { initialProps: e.initialProps, children: e.children });
}
function br(e) {
  var r = e.initialProps, n = e.children, t = function(o, i, l) {
    var d, c, u;
    if ((d = r.onDayClick) === null || d === void 0 || d.call(r, o, i, l), i.selected && !r.required) {
      (c = r.onSelect) === null || c === void 0 || c.call(r, void 0, o, i, l);
      return;
    }
    (u = r.onSelect) === null || u === void 0 || u.call(r, o, o, i, l);
  }, a = {
    selected: r.selected,
    onDayClick: t
  };
  return s(ye.Provider, { value: a, children: n });
}
function Fe() {
  var e = T(ye);
  if (!e)
    throw new Error("useSelectSingle must be used within a SelectSingleProvider");
  return e;
}
function _r(e, r) {
  var n = D(), t = Fe(), a = ce(), o = ve(), i = he(), l = i.focusDayAfter, d = i.focusDayBefore, c = i.focusWeekAfter, u = i.focusWeekBefore, f = i.blur, m = i.focus, h = i.focusMonthBefore, b = i.focusMonthAfter, g = i.focusYearBefore, _ = i.focusYearAfter, M = i.focusStartOfWeek, w = i.focusEndOfWeek, F = function(p) {
    var v, q, X, J;
    z(n) ? (v = t.onDayClick) === null || v === void 0 || v.call(t, e, r, p) : A(n) ? (q = a.onDayClick) === null || q === void 0 || q.call(a, e, r, p) : j(n) ? (X = o.onDayClick) === null || X === void 0 || X.call(o, e, r, p) : (J = n.onDayClick) === null || J === void 0 || J.call(n, e, r, p);
  }, Z = function(p) {
    var v;
    m(e), (v = n.onDayFocus) === null || v === void 0 || v.call(n, e, r, p);
  }, G = function(p) {
    var v;
    f(), (v = n.onDayBlur) === null || v === void 0 || v.call(n, e, r, p);
  }, R = function(p) {
    var v;
    (v = n.onDayMouseEnter) === null || v === void 0 || v.call(n, e, r, p);
  }, Re = function(p) {
    var v;
    (v = n.onDayMouseLeave) === null || v === void 0 || v.call(n, e, r, p);
  }, Be = function(p) {
    var v;
    (v = n.onDayPointerEnter) === null || v === void 0 || v.call(n, e, r, p);
  }, Ye = function(p) {
    var v;
    (v = n.onDayPointerLeave) === null || v === void 0 || v.call(n, e, r, p);
  }, Ae = function(p) {
    var v;
    (v = n.onDayTouchCancel) === null || v === void 0 || v.call(n, e, r, p);
  }, je = function(p) {
    var v;
    (v = n.onDayTouchEnd) === null || v === void 0 || v.call(n, e, r, p);
  }, He = function(p) {
    var v;
    (v = n.onDayTouchMove) === null || v === void 0 || v.call(n, e, r, p);
  }, Ke = function(p) {
    var v;
    (v = n.onDayTouchStart) === null || v === void 0 || v.call(n, e, r, p);
  }, Ue = function(p) {
    var v;
    (v = n.onDayKeyUp) === null || v === void 0 || v.call(n, e, r, p);
  }, Ve = function(p) {
    var v;
    switch (p.key) {
      case "ArrowLeft":
        p.preventDefault(), p.stopPropagation(), n.dir === "rtl" ? l() : d();
        break;
      case "ArrowRight":
        p.preventDefault(), p.stopPropagation(), n.dir === "rtl" ? d() : l();
        break;
      case "ArrowDown":
        p.preventDefault(), p.stopPropagation(), c();
        break;
      case "ArrowUp":
        p.preventDefault(), p.stopPropagation(), u();
        break;
      case "PageUp":
        p.preventDefault(), p.stopPropagation(), p.shiftKey ? g() : h();
        break;
      case "PageDown":
        p.preventDefault(), p.stopPropagation(), p.shiftKey ? _() : b();
        break;
      case "Home":
        p.preventDefault(), p.stopPropagation(), M();
        break;
      case "End":
        p.preventDefault(), p.stopPropagation(), w();
        break;
    }
    (v = n.onDayKeyDown) === null || v === void 0 || v.call(n, e, r, p);
  }, ze = {
    onClick: F,
    onFocus: Z,
    onBlur: G,
    onKeyDown: Ve,
    onKeyUp: Ue,
    onMouseEnter: R,
    onMouseLeave: Re,
    onPointerEnter: Be,
    onPointerLeave: Ye,
    onTouchCancel: Ae,
    onTouchEnd: je,
    onTouchMove: He,
    onTouchStart: Ke
  };
  return ze;
}
function gr() {
  var e = D(), r = Fe(), n = ce(), t = ve(), a = z(e) ? r.selected : A(e) ? n.selected : j(e) ? t.selected : void 0;
  return a;
}
function Dr(e) {
  return Object.values(x).includes(e);
}
function wr(e, r) {
  var n = [e.classNames.day];
  return Object.keys(r).forEach(function(t) {
    var a = e.modifiersClassNames[t];
    if (a)
      n.push(a);
    else if (Dr(t)) {
      var o = e.classNames["day_".concat(t)];
      o && n.push(o);
    }
  }), n;
}
function Mr(e, r) {
  var n = y({}, e.styles.day);
  return Object.keys(r).forEach(function(t) {
    var a;
    n = y(y({}, n), (a = e.modifiersStyles) === null || a === void 0 ? void 0 : a[t]);
  }), n;
}
function Nr(e, r, n) {
  var t, a, o, i = D(), l = he(), d = yr(e, r), c = _r(e, d), u = gr(), f = !!(i.onDayClick || i.mode !== "default");
  V(function() {
    var R;
    d.outside || l.focusedDay && f && N(l.focusedDay, e) && ((R = n.current) === null || R === void 0 || R.focus());
  }, [
    l.focusedDay,
    e,
    n,
    f,
    d.outside
  ]);
  var m = wr(i, d).join(" "), h = Mr(i, d), b = !!(d.outside && !i.showOutsideDays || d.hidden), g = (o = (a = i.components) === null || a === void 0 ? void 0 : a.DayContent) !== null && o !== void 0 ? o : Vn, _ = s(g, { date: e, displayMonth: r, activeModifiers: d }), M = {
    style: h,
    className: m,
    children: _,
    role: "gridcell"
  }, w = l.focusTarget && N(l.focusTarget, e) && !d.outside, F = l.focusedDay && N(l.focusedDay, e), Z = y(y(y({}, M), (t = { disabled: d.disabled, role: "gridcell" }, t["aria-selected"] = d.selected, t.tabIndex = F || w ? 0 : -1, t)), c), G = {
    isButton: f,
    isHidden: b,
    activeModifiers: d,
    selectedDays: u,
    buttonProps: Z,
    divProps: M
  };
  return G;
}
function kr(e) {
  var r = qe(null), n = Nr(e.date, e.displayMonth, r);
  return n.isHidden ? s("div", { role: "gridcell" }) : n.isButton ? s(U, y({ name: "day", ref: r }, n.buttonProps)) : s("div", y({}, n.divProps));
}
function Cr(e) {
  var r = e.number, n = e.dates, t = D(), a = t.onWeekNumberClick, o = t.styles, i = t.classNames, l = t.locale, d = t.labels.labelWeekNumber, c = t.formatters.formatWeekNumber, u = c(Number(r), { locale: l });
  if (!a)
    return s("span", { className: i.weeknumber, style: o.weeknumber, children: u });
  var f = d(Number(r), { locale: l }), m = function(h) {
    a(r, n, h);
  };
  return s(U, { name: "week-number", "aria-label": f, className: i.weeknumber, style: o.weeknumber, onClick: m, children: u });
}
function xr(e) {
  var r, n, t = D(), a = t.styles, o = t.classNames, i = t.showWeekNumber, l = t.components, d = (r = l?.Day) !== null && r !== void 0 ? r : kr, c = (n = l?.WeekNumber) !== null && n !== void 0 ? n : Cr, u;
  return i && (u = s("td", { className: o.cell, style: a.cell, children: s(c, { number: e.weekNumber, dates: e.dates }) })), P("tr", { className: o.row, style: a.row, children: [u, e.dates.map(function(f) {
    return s("td", { className: o.cell, style: a.cell, role: "presentation", children: s(d, { displayMonth: e.displayMonth, date: f }) }, Qe(f));
  })] });
}
function De(e, r, n) {
  for (var t = n?.ISOWeek ? Me(r) : Ne(r, n), a = n?.ISOWeek ? se(e) : de(e, n), o = W(t, a), i = [], l = 0; l <= o; l++)
    i.push(k(a, l));
  var d = i.reduce(function(c, u) {
    var f = n?.ISOWeek ? tn(u) : an(u, n), m = c.find(function(h) {
      return h.weekNumber === f;
    });
    return m ? (m.dates.push(u), c) : (c.push({
      weekNumber: f,
      dates: [u]
    }), c);
  }, []);
  return d;
}
function Pr(e, r) {
  var n = De(C(e), ie(e), r);
  if (r?.useFixedWeeks) {
    var t = Je(e, r);
    if (t < 6) {
      var a = n[n.length - 1], o = a.dates[a.dates.length - 1], i = ae(o, 6 - t), l = De(ae(o, 1), i, r);
      n.push.apply(n, l);
    }
  }
  return n;
}
function Sr(e) {
  var r, n, t, a = D(), o = a.locale, i = a.classNames, l = a.styles, d = a.hideHead, c = a.fixedWeeks, u = a.components, f = a.weekStartsOn, m = a.firstWeekContainsDate, h = a.ISOWeek, b = Pr(e.displayMonth, {
    useFixedWeeks: !!c,
    ISOWeek: h,
    locale: o,
    weekStartsOn: f,
    firstWeekContainsDate: m
  }), g = (r = u?.Head) !== null && r !== void 0 ? r : Un, _ = (n = u?.Row) !== null && n !== void 0 ? n : xr, M = (t = u?.Footer) !== null && t !== void 0 ? t : jn;
  return P("table", { id: e.id, className: i.table, style: l.table, role: "grid", "aria-labelledby": e["aria-labelledby"], children: [!d && s(g, {}), s("tbody", { className: i.tbody, style: l.tbody, children: b.map(function(w) {
    return s(_, { displayMonth: e.displayMonth, dates: w.dates, weekNumber: w.weekNumber }, w.weekNumber);
  }) }), s(M, { displayMonth: e.displayMonth })] });
}
function Or() {
  return !!(typeof window < "u" && window.document && window.document.createElement);
}
var Wr = Or() ? Ze : V, te = !1, Lr = 0;
function we() {
  return "react-day-picker-".concat(++Lr);
}
function Ir(e) {
  var r, n = e ?? (te ? we() : null), t = B(n), a = t[0], o = t[1];
  return Wr(function() {
    a === null && o(we());
  }, []), V(function() {
    te === !1 && (te = !0);
  }, []), (r = e ?? a) !== null && r !== void 0 ? r : void 0;
}
function Tr(e) {
  var r, n, t = D(), a = t.dir, o = t.classNames, i = t.styles, l = t.components, d = H().displayMonths, c = Ir(t.id ? "".concat(t.id, "-").concat(e.displayIndex) : void 0), u = t.id ? "".concat(t.id, "-grid-").concat(e.displayIndex) : void 0, f = [o.month], m = i.month, h = e.displayIndex === 0, b = e.displayIndex === d.length - 1, g = !h && !b;
  a === "rtl" && (r = [h, b], b = r[0], h = r[1]), h && (f.push(o.caption_start), m = y(y({}, m), i.caption_start)), b && (f.push(o.caption_end), m = y(y({}, m), i.caption_end)), g && (f.push(o.caption_between), m = y(y({}, m), i.caption_between));
  var _ = (n = l?.Caption) !== null && n !== void 0 ? n : An;
  return P("div", { className: f.join(" "), style: m, children: [s(_, { id: c, displayMonth: e.displayMonth, displayIndex: e.displayIndex }), s(Sr, { id: u, "aria-labelledby": c, displayMonth: e.displayMonth })] }, e.displayIndex);
}
function Er(e) {
  var r = D(), n = r.classNames, t = r.styles;
  return s("div", { className: n.months, style: t.months, children: e.children });
}
function Fr(e) {
  var r, n, t = e.initialProps, a = D(), o = he(), i = H(), l = B(!1), d = l[0], c = l[1];
  V(function() {
    a.initialFocus && o.focusTarget && (d || (o.focus(o.focusTarget), c(!0)));
  }, [
    a.initialFocus,
    d,
    o.focus,
    o.focusTarget,
    o
  ]);
  var u = [a.classNames.root, a.className];
  a.numberOfMonths > 1 && u.push(a.classNames.multiple_months), a.showWeekNumber && u.push(a.classNames.with_weeknumber);
  var f = y(y({}, a.styles.root), a.style), m = Object.keys(t).filter(function(b) {
    return b.startsWith("data-");
  }).reduce(function(b, g) {
    var _;
    return y(y({}, b), (_ = {}, _[g] = t[g], _));
  }, {}), h = (n = (r = t.components) === null || r === void 0 ? void 0 : r.Months) !== null && n !== void 0 ? n : Er;
  return s("div", y({ className: u.join(" "), style: f, dir: a.dir, id: a.id, nonce: t.nonce, title: t.title, lang: t.lang }, m, { children: s(h, { children: i.displayMonths.map(function(b, g) {
    return s(Tr, { displayIndex: g, displayMonth: b }, g);
  }) }) }));
}
function Rr(e) {
  var r = e.children, n = ln(e, ["children"]);
  return s(Cn, { initialProps: n, children: s(Fn, { children: s(pr, { initialProps: n, children: s(zn, { initialProps: n, children: s(qn, { initialProps: n, children: s(tr, { children: s(hr, { children: r }) }) }) }) }) }) });
}
function pt(e) {
  return s(Rr, y({}, e, { children: s(Fr, { initialProps: e }) }));
}
export {
  U as Button,
  An as Caption,
  _e as CaptionDropdowns,
  Se as CaptionLabel,
  ge as CaptionNavigation,
  kr as Day,
  Vn as DayContent,
  pt as DayPicker,
  Pe as DayPickerContext,
  Cn as DayPickerProvider,
  Oe as Dropdown,
  Ee as FocusContext,
  hr as FocusProvider,
  jn as Footer,
  Un as Head,
  Kn as HeadRow,
  xn as IconDropdown,
  Rn as IconLeft,
  Bn as IconRight,
  x as InternalModifier,
  Er as Months,
  We as NavigationContext,
  Fn as NavigationProvider,
  Rr as RootProvider,
  xr as Row,
  ue as SelectMultipleContext,
  zn as SelectMultipleProvider,
  Zn as SelectMultipleProviderInternal,
  fe as SelectRangeContext,
  qn as SelectRangeProvider,
  Xn as SelectRangeProviderInternal,
  ye as SelectSingleContext,
  pr as SelectSingleProvider,
  br as SelectSingleProviderInternal,
  Cr as WeekNumber,
  Gn as addToRange,
  ir as isDateAfterType,
  lr as isDateBeforeType,
  ar as isDateInterval,
  or as isDateRange,
  sr as isDayOfWeekType,
  A as isDayPickerMultiple,
  j as isDayPickerRange,
  z as isDayPickerSingle,
  fr as isMatch,
  yr as useActiveModifiers,
  D as useDayPicker,
  Nr as useDayRender,
  he as useFocusContext,
  H as useNavigation,
  ce as useSelectMultiple,
  ve as useSelectRange,
  Fe as useSelectSingle
};
