// zen-observable-ts@1.1.0/module.js
function I(t, r) {
  var e = typeof Symbol < "u" && t[Symbol.iterator] || t["@@iterator"];
  if (e) return (e = e.call(t)).next.bind(e);
  if (Array.isArray(t) || (e = M(t)) || r) {
    e && (t = e);
    var n = 0;
    return function() {
      return n >= t.length ? { done: !0 } : { done: !1, value: t[n++] };
    };
  }
  throw new TypeError(`Invalid attempt to iterate non-iterable instance.
In order to be iterable, non-array objects must have a [Symbol.iterator]() method.`);
}
function M(t, r) {
  if (t) {
    if (typeof t == "string") return T(t, r);
    var e = Object.prototype.toString.call(t).slice(8, -1);
    if (e === "Object" && t.constructor && (e = t.constructor.name), e === "Map" || e === "Set") return Array.from(t);
    if (e === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(e)) return T(t, r);
  }
}
function T(t, r) {
  (r == null || r > t.length) && (r = t.length);
  for (var e = 0, n = new Array(r); e < r; e++)
    n[e] = t[e];
  return n;
}
function A(t, r) {
  for (var e = 0; e < r.length; e++) {
    var n = r[e];
    n.enumerable = n.enumerable || !1, n.configurable = !0, "value" in n && (n.writable = !0), Object.defineProperty(t, n.key, n);
  }
}
function g(t, r, e) {
  return r && A(t.prototype, r), e && A(t, e), t;
}
var S = function() {
  return typeof Symbol == "function";
}, E = function(t) {
  return S() && !!Symbol[t];
}, x = function(t) {
  return E(t) ? Symbol[t] : "@@" + t;
};
S() && !E("observable") && (Symbol.observable = /* @__PURE__ */ Symbol("observable"));
var k = x("iterator"), w = x("observable"), q = x("species");
function m(t, r) {
  var e = t[r];
  if (e != null) {
    if (typeof e != "function") throw new TypeError(e + " is not a function");
    return e;
  }
}
function d(t) {
  var r = t.constructor;
  return r !== void 0 && (r = r[q], r === null && (r = void 0)), r !== void 0 ? r : O;
}
function z(t) {
  return t instanceof O;
}
function y(t) {
  y.log ? y.log(t) : setTimeout(function() {
    throw t;
  });
}
function b(t) {
  Promise.resolve().then(function() {
    try {
      t();
    } catch (r) {
      y(r);
    }
  });
}
function C(t) {
  var r = t._cleanup;
  if (r !== void 0 && (t._cleanup = void 0, !!r))
    try {
      if (typeof r == "function")
        r();
      else {
        var e = m(r, "unsubscribe");
        e && e.call(r);
      }
    } catch (n) {
      y(n);
    }
}
function _(t) {
  t._observer = void 0, t._queue = void 0, t._state = "closed";
}
function P(t) {
  var r = t._queue;
  if (r) {
    t._queue = void 0, t._state = "ready";
    for (var e = 0; e < r.length && (j(t, r[e].type, r[e].value), t._state !== "closed"); ++e)
      ;
  }
}
function j(t, r, e) {
  t._state = "running";
  var n = t._observer;
  try {
    var u = m(n, r);
    switch (r) {
      case "next":
        u && u.call(n, e);
        break;
      case "error":
        if (_(t), u) u.call(n, e);
        else throw e;
        break;
      case "complete":
        _(t), u && u.call(n);
        break;
    }
  } catch (f) {
    y(f);
  }
  t._state === "closed" ? C(t) : t._state === "running" && (t._state = "ready");
}
function v(t, r, e) {
  if (t._state !== "closed") {
    if (t._state === "buffering") {
      t._queue.push({
        type: r,
        value: e
      });
      return;
    }
    if (t._state !== "ready") {
      t._state = "buffering", t._queue = [{
        type: r,
        value: e
      }], b(function() {
        return P(t);
      });
      return;
    }
    j(t, r, e);
  }
}
var N = /* @__PURE__ */ (function() {
  function t(e, n) {
    this._cleanup = void 0, this._observer = e, this._queue = void 0, this._state = "initializing";
    var u = new R(this);
    try {
      this._cleanup = n.call(void 0, u);
    } catch (f) {
      u.error(f);
    }
    this._state === "initializing" && (this._state = "ready");
  }
  var r = t.prototype;
  return r.unsubscribe = function() {
    this._state !== "closed" && (_(this), C(this));
  }, g(t, [{
    key: "closed",
    get: function() {
      return this._state === "closed";
    }
  }]), t;
})(), R = /* @__PURE__ */ (function() {
  function t(e) {
    this._subscription = e;
  }
  var r = t.prototype;
  return r.next = function(n) {
    v(this._subscription, "next", n);
  }, r.error = function(n) {
    v(this._subscription, "error", n);
  }, r.complete = function() {
    v(this._subscription, "complete");
  }, g(t, [{
    key: "closed",
    get: function() {
      return this._subscription._state === "closed";
    }
  }]), t;
})(), O = /* @__PURE__ */ (function() {
  function t(e) {
    if (!(this instanceof t)) throw new TypeError("Observable cannot be called as a function");
    if (typeof e != "function") throw new TypeError("Observable initializer must be a function");
    this._subscriber = e;
  }
  var r = t.prototype;
  return r.subscribe = function(n) {
    return (typeof n != "object" || n === null) && (n = {
      next: n,
      error: arguments[1],
      complete: arguments[2]
    }), new N(n, this._subscriber);
  }, r.forEach = function(n) {
    var u = this;
    return new Promise(function(f, o) {
      if (typeof n != "function") {
        o(new TypeError(n + " is not a function"));
        return;
      }
      function i() {
        c.unsubscribe(), f();
      }
      var c = u.subscribe({
        next: function(s) {
          try {
            n(s, i);
          } catch (a) {
            o(a), c.unsubscribe();
          }
        },
        error: o,
        complete: f
      });
    });
  }, r.map = function(n) {
    var u = this;
    if (typeof n != "function") throw new TypeError(n + " is not a function");
    var f = d(this);
    return new f(function(o) {
      return u.subscribe({
        next: function(i) {
          try {
            i = n(i);
          } catch (c) {
            return o.error(c);
          }
          o.next(i);
        },
        error: function(i) {
          o.error(i);
        },
        complete: function() {
          o.complete();
        }
      });
    });
  }, r.filter = function(n) {
    var u = this;
    if (typeof n != "function") throw new TypeError(n + " is not a function");
    var f = d(this);
    return new f(function(o) {
      return u.subscribe({
        next: function(i) {
          try {
            if (!n(i)) return;
          } catch (c) {
            return o.error(c);
          }
          o.next(i);
        },
        error: function(i) {
          o.error(i);
        },
        complete: function() {
          o.complete();
        }
      });
    });
  }, r.reduce = function(n) {
    var u = this;
    if (typeof n != "function") throw new TypeError(n + " is not a function");
    var f = d(this), o = arguments.length > 1, i = !1, c = arguments[1], s = c;
    return new f(function(a) {
      return u.subscribe({
        next: function(h) {
          var l = !i;
          if (i = !0, !l || o)
            try {
              s = n(s, h);
            } catch (p) {
              return a.error(p);
            }
          else
            s = h;
        },
        error: function(h) {
          a.error(h);
        },
        complete: function() {
          if (!i && !o) return a.error(new TypeError("Cannot reduce an empty sequence"));
          a.next(s), a.complete();
        }
      });
    });
  }, r.concat = function() {
    for (var n = this, u = arguments.length, f = new Array(u), o = 0; o < u; o++)
      f[o] = arguments[o];
    var i = d(this);
    return new i(function(c) {
      var s, a = 0;
      function h(l) {
        s = l.subscribe({
          next: function(p) {
            c.next(p);
          },
          error: function(p) {
            c.error(p);
          },
          complete: function() {
            a === f.length ? (s = void 0, c.complete()) : h(i.from(f[a++]));
          }
        });
      }
      return h(n), function() {
        s && (s.unsubscribe(), s = void 0);
      };
    });
  }, r.flatMap = function(n) {
    var u = this;
    if (typeof n != "function") throw new TypeError(n + " is not a function");
    var f = d(this);
    return new f(function(o) {
      var i = [], c = u.subscribe({
        next: function(a) {
          if (n)
            try {
              a = n(a);
            } catch (l) {
              return o.error(l);
            }
          var h = f.from(a).subscribe({
            next: function(l) {
              o.next(l);
            },
            error: function(l) {
              o.error(l);
            },
            complete: function() {
              var l = i.indexOf(h);
              l >= 0 && i.splice(l, 1), s();
            }
          });
          i.push(h);
        },
        error: function(a) {
          o.error(a);
        },
        complete: function() {
          s();
        }
      });
      function s() {
        c.closed && i.length === 0 && o.complete();
      }
      return function() {
        i.forEach(function(a) {
          return a.unsubscribe();
        }), c.unsubscribe();
      };
    });
  }, r[w] = function() {
    return this;
  }, t.from = function(n) {
    var u = typeof this == "function" ? this : t;
    if (n == null) throw new TypeError(n + " is not an object");
    var f = m(n, w);
    if (f) {
      var o = f.call(n);
      if (Object(o) !== o) throw new TypeError(o + " is not an object");
      return z(o) && o.constructor === u ? o : new u(function(i) {
        return o.subscribe(i);
      });
    }
    if (E("iterator") && (f = m(n, k), f))
      return new u(function(i) {
        b(function() {
          if (!i.closed) {
            for (var c = I(f.call(n)), s; !(s = c()).done; ) {
              var a = s.value;
              if (i.next(a), i.closed) return;
            }
            i.complete();
          }
        });
      });
    if (Array.isArray(n))
      return new u(function(i) {
        b(function() {
          if (!i.closed) {
            for (var c = 0; c < n.length; ++c)
              if (i.next(n[c]), i.closed) return;
            i.complete();
          }
        });
      });
    throw new TypeError(n + " is not observable");
  }, t.of = function() {
    for (var n = arguments.length, u = new Array(n), f = 0; f < n; f++)
      u[f] = arguments[f];
    var o = typeof this == "function" ? this : t;
    return new o(function(i) {
      b(function() {
        if (!i.closed) {
          for (var c = 0; c < u.length; ++c)
            if (i.next(u[c]), i.closed) return;
          i.complete();
        }
      });
    });
  }, g(t, null, [{
    key: q,
    get: function() {
      return this;
    }
  }]), t;
})();
S() && Object.defineProperty(O, /* @__PURE__ */ Symbol("extensions"), {
  value: {
    symbol: w,
    hostReportError: y
  },
  configurable: !0
});
export {
  O as Observable
};
