// echarts@6.0.0/lib/coord/geo/Geo.js
import { __extends as S } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { createHashMap as M, retrieve2 as T, isString as j, mixin as _ } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import G from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/BoundingRect.js";
import y from "../View.js";
import v from "./geoSourceManager.js";
import { SINGLE_REFERRING as P } from "../../util/model.js";
import { warn as m } from "../../util/log.js";
var R = {
  geoJSON: {
    aspectScale: 0.75,
    invertLongitute: !0
  },
  geoSVG: {
    aspectScale: 1,
    invertLongitute: !1
  }
}, E = ["lng", "lat"], N = (
  /** @class */
  (function(c) {
    S(i, c);
    function i(e, t, r) {
      var o = c.call(this, e, {
        api: r.api,
        ecModel: r.ecModel
      }) || this;
      o.dimensions = E, o.type = "geo", o._nameCoordMap = M(), o.map = t;
      var n = r.projection, p = v.load(t, r.nameMap, r.nameProperty), a = v.getGeoResource(t), u = o.resourceType = a ? a.type : null, l = o.regions = p.regions, f = R[a.type];
      o._regionsMap = p.regionsMap, o.regions = p.regions, process.env.NODE_ENV !== "production" && n && (u === "geoSVG" && (process.env.NODE_ENV !== "production" && m("Map " + t + " with SVG source can't use projection. Only GeoJSON source supports projection."), n = null), n.project && n.unproject || (process.env.NODE_ENV !== "production" && m("project and unproject must be both provided in the projeciton."), n = null)), o.projection = n;
      var s;
      if (n)
        for (var g = 0; g < l.length; g++) {
          var d = l[g].getBoundingRect(n);
          s = s || d.clone(), s.union(d);
        }
      else
        s = p.boundingRect;
      return o.setBoundingRect(s.x, s.y, s.width, s.height), o.aspectScale = n ? 1 : T(r.aspectScale, f.aspectScale), o._invertLongitute = n ? !1 : f.invertLongitute, o;
    }
    return i.prototype._transformTo = function(e, t, r, o) {
      var n = this.getBoundingRect(), p = this._invertLongitute;
      n = n.clone(), p && (n.y = -n.y - n.height);
      var a = this._rawTransformable;
      a.transform = n.calculateTransform(new G(e, t, r, o));
      var u = a.parent;
      a.parent = null, a.decomposeTransform(), a.parent = u, p && (a.scaleY = -a.scaleY), this._updateTransform();
    }, i.prototype.getRegion = function(e) {
      return this._regionsMap.get(e);
    }, i.prototype.getRegionByCoord = function(e) {
      for (var t = this.regions, r = 0; r < t.length; r++) {
        var o = t[r];
        if (o.type === "geoJSON" && o.contain(e))
          return t[r];
      }
    }, i.prototype.addGeoCoord = function(e, t) {
      this._nameCoordMap.set(e, t);
    }, i.prototype.getGeoCoord = function(e) {
      var t = this._regionsMap.get(e);
      return this._nameCoordMap.get(e) || t && t.getCenter();
    }, i.prototype.dataToPoint = function(e, t, r) {
      if (j(e) && (e = this.getGeoCoord(e)), e) {
        var o = this.projection;
        return o && (e = o.project(e)), e && this.projectedToPoint(e, t, r);
      }
    }, i.prototype.pointToData = function(e, t, r) {
      var o = this.projection;
      return o && (e = o.unproject(e)), e && this.pointToProjected(e, r);
    }, i.prototype.pointToProjected = function(e, t) {
      return c.prototype.pointToData.call(this, e, 0, t);
    }, i.prototype.projectedToPoint = function(e, t, r) {
      return c.prototype.dataToPoint.call(this, e, t, r);
    }, i.prototype.convertToPixel = function(e, t, r) {
      var o = h(t);
      return o === this ? o.dataToPoint(r) : null;
    }, i.prototype.convertFromPixel = function(e, t, r) {
      var o = h(t);
      return o === this ? o.pointToData(r) : null;
    }, i;
  })(y)
);
_(N, y);
function h(c) {
  var i = c.geoModel, e = c.seriesModel;
  return i ? i.coordinateSystem : e ? e.coordinateSystem || (e.getReferringComponents("geo", P).models[0] || {}).coordinateSystem : null;
}
export {
  N as default,
  E as geo2DDimensions
};
