// echarts@6.0.0/lib/coord/geo/geoCreator.js
import { extend as I, each as C, map as P, mergeAll as H, retrieve as V, createHashMap as J, merge as U } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import L, { geo2DDimensions as Z } from "./Geo.js";
import { createBoxLayoutReference as k, getLayoutRect as q, applyPreserveAspect as K } from "../../util/layout.js";
import { parsePercent as w } from "../../util/number.js";
import Q from "./geoSourceManager.js";
import { min as W, max as X } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/vector.js";
import { injectCoordSysByOption as Y } from "../../core/CoordinateSystem.js";
import { SINGLE_REFERRING as $ } from "../../util/model.js";
function O(a, f) {
  var u = a.get("boundingCoords");
  if (u != null) {
    var n = u[0], o = u[1];
    if (!(isFinite(n[0]) && isFinite(n[1]) && isFinite(o[0]) && isFinite(o[1])))
      process.env.NODE_ENV !== "production" && console.error("Invalid boundingCoords");
    else {
      var c = this.projection;
      if (c) {
        var e = n[0], t = n[1], v = o[0], r = o[1];
        n = [1 / 0, 1 / 0], o = [-1 / 0, -1 / 0];
        var s = function(R, l, j, A) {
          for (var D = j - R, b = A - l, N = 0; N <= 100; N++) {
            var E = N / 100, F = c.project([R + D * E, l + b * E]);
            W(n, n, F), X(o, o, F);
          }
        };
        s(e, t, v, t), s(v, t, v, r), s(v, r, e, r), s(e, r, v, t);
      }
      this.setBoundingRect(n[0], n[1], o[0] - n[0], o[1] - n[1]);
    }
  }
  var p = this.getBoundingRect(), h = a.get("layoutCenter"), S = a.get("layoutSize"), m = k(a, f).refContainer, d = p.width / p.height * this.aspectScale, G = !1, y, g;
  h && S && (y = [w(h[0], m.width) + m.x, w(h[1], m.height) + m.y], g = w(S, Math.min(m.width, m.height)), !isNaN(y[0]) && !isNaN(y[1]) && !isNaN(g) ? G = !0 : process.env.NODE_ENV !== "production" && console.warn("Given layoutCenter or layoutSize data are invalid. Use left/top/width/height instead."));
  var i;
  if (G)
    i = {}, d > 1 ? (i.width = g, i.height = g / d) : (i.height = g, i.width = g * d), i.y = y[1] - i.height / 2, i.x = y[0] - i.width / 2;
  else {
    var z = a.getBoxLayoutParams();
    z.aspect = d, i = q(z, m), i = K(a, i, d);
  }
  this.setViewRect(i.x, i.y, i.width, i.height), this.setCenter(a.get("center")), this.setZoom(a.get("zoom"));
}
function B(a, f) {
  C(f.get("geoCoord"), function(u, n) {
    a.addGeoCoord(n, u);
  });
}
var x = (
  /** @class */
  (function() {
    function a() {
      this.dimensions = Z;
    }
    return a.prototype.create = function(f, u) {
      var n = [];
      function o(e) {
        return {
          nameProperty: e.get("nameProperty"),
          aspectScale: e.get("aspectScale"),
          projection: e.get("projection")
        };
      }
      f.eachComponent("geo", function(e, t) {
        var v = e.get("map"), r = new L(v + t, v, I({
          nameMap: e.get("nameMap"),
          api: u,
          ecModel: f
        }, o(e)));
        r.zoomLimit = e.get("scaleLimit"), n.push(r), e.coordinateSystem = r, r.model = e, r.resize = O, r.resize(e, u);
      }), f.eachSeries(function(e) {
        Y({
          targetModel: e,
          coordSysType: "geo",
          coordSysProvider: function() {
            var t = e.subType === "map" ? e.getHostGeoModel() : e.getReferringComponents("geo", $).models[0];
            return t && t.coordinateSystem;
          },
          allowNotFound: !0
        });
      });
      var c = {};
      return f.eachSeriesByType("map", function(e) {
        if (!e.getHostGeoModel()) {
          var t = e.getMapType();
          c[t] = c[t] || [], c[t].push(e);
        }
      }), C(c, function(e, t) {
        var v = P(e, function(s) {
          return s.get("nameMap");
        }), r = new L(t, t, I({
          nameMap: H(v),
          api: u,
          ecModel: f
        }, o(e[0])));
        r.zoomLimit = V.apply(null, P(e, function(s) {
          return s.get("scaleLimit");
        })), n.push(r), r.resize = O, r.resize(e[0], u), C(e, function(s) {
          s.coordinateSystem = r, B(r, s);
        });
      }), n;
    }, a.prototype.getFilledRegions = function(f, u, n, o) {
      for (var c = (f || []).slice(), e = J(), t = 0; t < c.length; t++)
        e.set(c[t].name, c[t]);
      var v = Q.load(u, n, o);
      return C(v.regions, function(r) {
        var s = r.name, p = e.get(s), h = r.properties && r.properties.echartsStyle;
        p || (p = {
          name: s
        }, c.push(p)), h && U(p, h);
      }), c;
    }, a;
  })()
), ae = new x();
export {
  ae as default
};
