// echarts@6.0.0/lib/coord/geo/GeoJSONResource.js
import { createHashMap as c, isString as h, each as f, hasOwn as u } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import g from "./parseGeoJson.js";
import m from "./fix/nanhai.js";
import N from "./fix/textCoord.js";
import l from "./fix/diaoyuIsland.js";
import d from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/BoundingRect.js";
var _ = "name", T = (
  /** @class */
  (function() {
    function t(a, e, r) {
      this.type = "geoJSON", this._parsedMap = c(), this._mapName = a, this._specialAreas = r, this._geoJSON = J(e);
    }
    return t.prototype.load = function(a, e) {
      e = e || _;
      var r = this._parsedMap.get(e);
      if (!r) {
        var n = this._parseToRegions(e);
        r = this._parsedMap.set(e, {
          regions: n,
          boundingRect: v(n)
        });
      }
      var i = c(), s = [];
      return f(r.regions, function(o) {
        var p = o.name;
        a && u(a, p) && (o = o.cloneShallow(p = a[p])), s.push(o), i.set(p, o);
      }), {
        regions: s,
        boundingRect: r.boundingRect || new d(0, 0, 0, 0),
        regionsMap: i
      };
    }, t.prototype._parseToRegions = function(a) {
      var e = this._mapName, r = this._geoJSON, n;
      try {
        n = r ? g(r, a) : [];
      } catch (i) {
        throw new Error(`Invalid geoJson format
` + i.message);
      }
      return m(e, n), f(n, function(i) {
        var s = i.name;
        N(e, i), l(e, i);
        var o = this._specialAreas && this._specialAreas[s];
        o && i.transformTo(o.left, o.top, o.width, o.height);
      }, this), n;
    }, t.prototype.getMapForUser = function() {
      return {
        // For backward compatibility, use geoJson
        // PENDING: it has been returning them without clone.
        // do we need to avoid outsite modification?
        geoJson: this._geoJSON,
        geoJSON: this._geoJSON,
        specialAreas: this._specialAreas
      };
    }, t;
  })()
);
function v(t) {
  for (var a, e = 0; e < t.length; e++) {
    var r = t[e].getBoundingRect();
    a = a || r.clone(), a.union(r);
  }
  return a;
}
function J(t) {
  return h(t) ? typeof JSON < "u" && JSON.parse ? JSON.parse(t) : new Function("return (" + t + ");")() : t;
}
export {
  T as GeoJSONResource
};
