// echarts@6.0.0/lib/coord/geo/GeoModel.js
import { __extends as c } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { reduce as d, createHashMap as u, isFunction as m, isString as f } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { defaultEmphasis as h } from "../../util/model.js";
import g from "../../model/Component.js";
import p from "../../model/Model.js";
import M from "./geoCreator.js";
import y from "./geoSourceManager.js";
import n from "../../visual/tokens.js";
var z = (
  /** @class */
  (function(a) {
    c(o, a);
    function o() {
      var e = a !== null && a.apply(this, arguments) || this;
      return e.type = o.type, e;
    }
    return o.prototype.init = function(e, t, l) {
      this.mergeDefaultAndTheme(e, l);
      var r = y.getGeoResource(e.map);
      if (r && r.type === "geoJSON") {
        var i = e.itemStyle = e.itemStyle || {};
        "color" in i || (i.color = e.defaultItemStyleColor || n.color.backgroundTint);
      }
      h(e, "label", ["show"]);
    }, o.prototype.optionUpdated = function() {
      var e = this, t = this.option;
      t.regions = M.getFilledRegions(t.regions, t.map, t.nameMap, t.nameProperty);
      var l = {};
      this._optionModelMap = d(t.regions || [], function(r, i) {
        var s = i.name;
        return s && (r.set(s, new p(i, e, e.ecModel)), i.selected && (l[s] = !0)), r;
      }, u()), t.selectedMap || (t.selectedMap = l);
    }, o.prototype.getRegionModel = function(e) {
      return this._optionModelMap.get(e) || new p(null, this, this.ecModel);
    }, o.prototype.getFormattedLabel = function(e, t) {
      var l = this.getRegionModel(e), r = t === "normal" ? l.get(["label", "formatter"]) : l.get(["emphasis", "label", "formatter"]), i = {
        name: e
      };
      if (m(r))
        return i.status = t, r(i);
      if (f(r))
        return r.replace("{a}", e ?? "");
    }, o.prototype.setZoom = function(e) {
      this.option.zoom = e;
    }, o.prototype.setCenter = function(e) {
      this.option.center = e;
    }, o.prototype.select = function(e) {
      var t = this.option, l = t.selectedMode;
      if (l) {
        l !== "multiple" && (t.selectedMap = null);
        var r = t.selectedMap || (t.selectedMap = {});
        r[e] = !0;
      }
    }, o.prototype.unSelect = function(e) {
      var t = this.option.selectedMap;
      t && (t[e] = !1);
    }, o.prototype.toggleSelected = function(e) {
      this[this.isSelected(e) ? "unSelect" : "select"](e);
    }, o.prototype.isSelected = function(e) {
      var t = this.option.selectedMap;
      return !!(t && t[e]);
    }, o.type = "geo", o.layoutMode = "box", o.defaultOption = {
      // zlevel: 0,
      z: 0,
      show: !0,
      left: "center",
      top: "center",
      // Default value:
      // for geoSVG source: 1,
      // for geoJSON source: 0.75.
      aspectScale: null,
      // /// Layout with center and size
      // If you want to put map in a fixed size box with right aspect ratio
      // This two properties may be more convenient
      // layoutCenter: [50%, 50%]
      // layoutSize: 100
      silent: !1,
      // Map type
      map: "",
      // Define left-top, right-bottom coords to control view
      // For example, [ [180, 90], [-180, -90] ]
      boundingCoords: null,
      // Default on center of map
      center: null,
      zoom: 1,
      scaleLimit: null,
      // selectedMode: false
      label: {
        show: !1,
        color: n.color.tertiary
      },
      itemStyle: {
        borderWidth: 0.5,
        borderColor: n.color.border
      },
      emphasis: {
        label: {
          show: !0,
          color: n.color.primary
        },
        itemStyle: {
          color: n.color.highlight
        }
      },
      select: {
        label: {
          show: !0,
          color: n.color.primary
        },
        itemStyle: {
          color: n.color.highlight
        }
      },
      regions: []
      // tooltip: {
      //     show: false
      // }
    }, o;
  })(g)
);
export {
  z as default
};
