// echarts@6.0.0/lib/coord/geo/GeoSVGResource.js
import { parseSVG as m, makeViewBoxTransform as M } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/tool/parseSVG.js";
import w from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import y from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Rect.js";
import { createHashMap as v, assert as V, each as _ } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import x from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/BoundingRect.js";
import { parseXML as S } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/tool/parseXML.js";
import { GeoSVGRegion as L } from "./Region.js";
var A = v([
  "rect",
  "circle",
  "line",
  "ellipse",
  "polygon",
  "polyline",
  "path",
  // <text> <tspan> are also enabled because some SVG might paint text itself,
  // but still need to trigger events or tooltip.
  "text",
  "tspan",
  // <g> is also enabled because this case: if multiple tags share one name
  // and need label displayed, every tags will display the name, which is not
  // expected. So we can put them into a <g name="xxx">. Thereby only one label
  // displayed and located based on the bounding rect of the <g>.
  "g"
]), T = (
  /** @class */
  (function() {
    function o(i, e) {
      this.type = "geoSVG", this._usedGraphicMap = v(), this._freedGraphics = [], this._mapName = i, this._parsedXML = S(e);
    }
    return o.prototype.load = function() {
      var i = this._firstGraphic;
      if (!i) {
        i = this._firstGraphic = this._buildGraphic(this._parsedXML), this._freedGraphics.push(i), this._boundingRect = this._firstGraphic.boundingRect.clone();
        var e = N(i.named), r = e.regions, s = e.regionsMap;
        this._regions = r, this._regionsMap = s;
      }
      return {
        boundingRect: this._boundingRect,
        regions: this._regions,
        regionsMap: this._regionsMap
      };
    }, o.prototype._buildGraphic = function(i) {
      var e, r;
      try {
        e = i && m(i, {
          ignoreViewBox: !0,
          ignoreRootClip: !0
        }) || {}, r = e.root, V(r != null);
      } catch (h) {
        throw new Error(`Invalid svg format
` + h.message);
      }
      var s = new w();
      s.add(r), s.isGeoSVGGraphicRoot = !0;
      var G = e.width, c = e.height, t = e.viewBoxRect, p = this._boundingRect;
      if (!p) {
        var n = void 0, a = void 0, u = void 0, f = void 0;
        if (G != null ? (n = 0, u = G) : t && (n = t.x, u = t.width), c != null ? (a = 0, f = c) : t && (a = t.y, f = t.height), n == null || a == null) {
          var g = r.getBoundingRect();
          n == null && (n = g.x, u = g.width), a == null && (a = g.y, f = g.height);
        }
        p = this._boundingRect = new x(n, a, u, f);
      }
      if (t) {
        var l = M(t, p);
        r.scaleX = r.scaleY = l.scale, r.x = l.x, r.y = l.y;
      }
      s.setClipPath(new y({
        shape: p.plain()
      }));
      var d = [];
      return _(e.named, function(h) {
        A.get(h.svgNodeTagLower) != null && (d.push(h), E(h.el));
      }), {
        root: s,
        boundingRect: p,
        named: d
      };
    }, o.prototype.useGraphic = function(i) {
      var e = this._usedGraphicMap, r = e.get(i);
      return r || (r = this._freedGraphics.pop() || this._buildGraphic(this._parsedXML), e.set(i, r), r);
    }, o.prototype.freeGraphic = function(i) {
      var e = this._usedGraphicMap, r = e.get(i);
      r && (e.removeKey(i), this._freedGraphics.push(r));
    }, o;
  })()
);
function E(o) {
  o.silent = !1, o.isGroup && o.traverse(function(i) {
    i.silent = !1;
  });
}
function N(o) {
  var i = [], e = v();
  return _(o, function(r) {
    if (r.namedFrom == null) {
      var s = new L(r.name, r.el);
      i.push(s), e.set(r.name, s);
    }
  }), {
    regions: i,
    regionsMap: e
  };
}
export {
  T as GeoSVGResource
};
