// echarts@6.0.0/lib/coord/geo/Region.js
import { __extends as R } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import y from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/BoundingRect.js";
import { applyTransform as S, min as x, max as F } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/vector.js";
import { contain as G } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/contain/polygon.js";
import { identity as O, mul as B, invert as N } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/matrix.js";
import { each as l } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
var d = [];
function m(o, n) {
  for (var r = 0; r < o.length; r++)
    S(o[r], o[r], n);
}
function p(o, n, r, i) {
  for (var e = 0; e < o.length; e++) {
    var t = o[e];
    i && (t = i.project(t)), t && isFinite(t[0]) && isFinite(t[1]) && (x(n, n, t), F(r, r, t));
  }
}
function w(o) {
  for (var n = 0, r = 0, i = 0, e = o.length, t = o[e - 1][0], a = o[e - 1][1], f = 0; f < e; f++) {
    var s = o[f][0], c = o[f][1], u = t * c - s * a;
    n += u, r += (t + s) * u, i += (a + c) * u, t = s, a = c;
  }
  return n ? [r / n / 3, i / n / 3, n] : [o[0][0] || 0, o[0][1] || 0];
}
var _ = (
  /** @class */
  (function() {
    function o(n) {
      this.name = n;
    }
    return o.prototype.setCenter = function(n) {
      this._center = n;
    }, o.prototype.getCenter = function() {
      var n = this._center;
      return n || (n = this._center = this.calcCenter()), n;
    }, o;
  })()
), L = (
  /** @class */
  /* @__PURE__ */ (function() {
    function o(n, r) {
      this.type = "polygon", this.exterior = n, this.interiors = r;
    }
    return o;
  })()
), A = (
  /** @class */
  /* @__PURE__ */ (function() {
    function o(n) {
      this.type = "linestring", this.points = n;
    }
    return o;
  })()
), M = (
  /** @class */
  (function(o) {
    R(n, o);
    function n(r, i, e) {
      var t = o.call(this, r) || this;
      return t.type = "geoJSON", t.geometries = i, t._center = e && [e[0], e[1]], t;
    }
    return n.prototype.calcCenter = function() {
      for (var r = this.geometries, i, e = 0, t = 0; t < r.length; t++) {
        var a = r[t], f = a.exterior, s = f && f.length;
        s > e && (i = a, e = s);
      }
      if (i)
        return w(i.exterior);
      var c = this.getBoundingRect();
      return [c.x + c.width / 2, c.y + c.height / 2];
    }, n.prototype.getBoundingRect = function(r) {
      var i = this._rect;
      if (i && !r)
        return i;
      var e = [1 / 0, 1 / 0], t = [-1 / 0, -1 / 0], a = this.geometries;
      return l(a, function(f) {
        f.type === "polygon" ? p(f.exterior, e, t, r) : l(f.points, function(s) {
          p(s, e, t, r);
        });
      }), isFinite(e[0]) && isFinite(e[1]) && isFinite(t[0]) && isFinite(t[1]) || (e[0] = e[1] = t[0] = t[1] = 0), i = new y(e[0], e[1], t[0] - e[0], t[1] - e[1]), r || (this._rect = i), i;
    }, n.prototype.contain = function(r) {
      var i = this.getBoundingRect(), e = this.geometries;
      if (!i.contain(r[0], r[1]))
        return !1;
      t: for (var t = 0, a = e.length; t < a; t++) {
        var f = e[t];
        if (f.type === "polygon") {
          var s = f.exterior, c = f.interiors;
          if (G(s, r[0], r[1])) {
            for (var u = 0; u < (c ? c.length : 0); u++)
              if (G(c[u], r[0], r[1]))
                continue t;
            return !0;
          }
        }
      }
      return !1;
    }, n.prototype.transformTo = function(r, i, e, t) {
      var a = this.getBoundingRect(), f = a.width / a.height;
      e ? t || (t = e / f) : e = f * t;
      for (var s = new y(r, i, e, t), c = a.calculateTransform(s), u = this.geometries, g = 0; g < u.length; g++) {
        var v = u[g];
        v.type === "polygon" ? (m(v.exterior, c), l(v.interiors, function(h) {
          m(h, c);
        })) : l(v.points, function(h) {
          m(h, c);
        });
      }
      a = this._rect, a.copy(s), this._center = [a.x + a.width / 2, a.y + a.height / 2];
    }, n.prototype.cloneShallow = function(r) {
      r == null && (r = this.name);
      var i = new n(r, this.geometries, this._center);
      return i._rect = this._rect, i.transformTo = null, i;
    }, n;
  })(_)
), b = (
  /** @class */
  (function(o) {
    R(n, o);
    function n(r, i) {
      var e = o.call(this, r) || this;
      return e.type = "geoSVG", e._elOnlyForCalculate = i, e;
    }
    return n.prototype.calcCenter = function() {
      for (var r = this._elOnlyForCalculate, i = r.getBoundingRect(), e = [i.x + i.width / 2, i.y + i.height / 2], t = O(d), a = r; a && !a.isGeoSVGGraphicRoot; )
        B(t, a.getLocalTransform(), t), a = a.parent;
      return N(t, t), S(e, e, t), e;
    }, n;
  })(_)
);
export {
  A as GeoJSONLineStringGeometry,
  L as GeoJSONPolygonGeometry,
  M as GeoJSONRegion,
  b as GeoSVGRegion,
  _ as Region
};
