// echarts@6.0.0/lib/component/toolbox/ToolboxModel.js
import { __extends as p } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { each as a, merge as d } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { getFeature as u } from "./featureManager.js";
import f from "../../model/Component.js";
import e from "../../visual/tokens.js";
var y = (
  /** @class */
  (function(r) {
    p(o, r);
    function o() {
      var i = r !== null && r.apply(this, arguments) || this;
      return i.type = o.type, i;
    }
    return o.prototype.optionUpdated = function() {
      r.prototype.optionUpdated.apply(this, arguments);
      var i = this.ecModel;
      a(this.option.feature, function(n, l) {
        var t = u(l);
        t && (t.getDefaultOption && (t.defaultOption = t.getDefaultOption(i)), d(n, t.defaultOption));
      });
    }, o.type = "toolbox", o.layoutMode = {
      type: "box",
      ignoreSize: !0
    }, o.defaultOption = {
      show: !0,
      z: 6,
      // zlevel: 0,
      orient: "horizontal",
      left: "right",
      top: "top",
      // right
      // bottom
      backgroundColor: "transparent",
      borderColor: e.color.border,
      borderRadius: 0,
      borderWidth: 0,
      padding: e.size.m,
      itemSize: 15,
      itemGap: e.size.s,
      showTitle: !0,
      iconStyle: {
        borderColor: e.color.accent50,
        color: "none"
      },
      emphasis: {
        iconStyle: {
          borderColor: e.color.accent50
        }
      },
      // textStyle: {},
      // feature
      tooltip: {
        show: !1,
        position: "bottom"
      }
    }, o;
  })(f)
);
export {
  y as default
};
