// echarts@6.0.0/lib/component/toolbox/ToolboxView.js
import { __extends as O } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { each as F, curry as Z, isString as L, bind as j, isFunction as q } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { getBoundingRect as J } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/contain/text.js";
import { createIcon as K, setTooltipConfig as Q } from "../../util/graphic.js";
import { enterEmphasis as A, leaveEmphasis as D } from "../../util/states.js";
import Y from "../../model/Model.js";
import $ from "../../data/DataDiffer.js";
import { makeBackground as b } from "../helper/listComponent.js";
import N from "../../view/Component.js";
import { getFeature as X, ToolboxFeature as x } from "./featureManager.js";
import { getUID as M } from "../../util/component.js";
import W from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Text.js";
import { getFont as tt } from "../../label/labelStyle.js";
import { createBoxLayoutReference as et, getLayoutRect as it, box as rt, positionElement as nt } from "../../util/layout.js";
import ot from "../../visual/tokens.js";
var Ct = (
  /** @class */
  (function(w) {
    O(c, w);
    function c() {
      return w !== null && w.apply(this, arguments) || this;
    }
    return c.prototype.render = function(e, u, i, l) {
      var g = this.group;
      if (g.removeAll(), !e.get("show"))
        return;
      var S = +e.get("itemSize"), _ = e.get("orient") === "vertical", P = e.get("feature") || {}, I = this._features || (this._features = {}), C = [];
      F(P, function(n, h) {
        C.push(h);
      }), new $(this._featureNames || [], C).add(k).update(k).remove(Z(k, null)).execute(), this._featureNames = C;
      function k(n, h) {
        var o = C[n], v = C[h], r = P[o], s = new Y(r, e, e.ecModel), t;
        if (l && l.newTitle != null && l.featureName === o && (r.title = l.newTitle), o && !v) {
          if (st(o))
            t = {
              onclick: s.option.onclick,
              featureName: o
            };
          else {
            var p = X(o);
            if (!p)
              return;
            t = new p();
          }
          I[o] = t;
        } else if (t = I[v], !t)
          return;
        t.uid = M("toolbox-feature"), t.model = s, t.ecModel = u, t.api = i;
        var m = t instanceof x;
        if (!o && v) {
          m && t.dispose && t.dispose(u, i);
          return;
        }
        if (!s.get("show") || m && t.unusable) {
          m && t.remove && t.remove(u, i);
          return;
        }
        U(s, t, o), s.setIconStatus = function(d, y) {
          var a = this.option, f = this.iconPaths;
          a.iconStatus = a.iconStatus || {}, a.iconStatus[d] = y, f[d] && (y === "emphasis" ? A : D)(f[d]);
        }, t instanceof x && t.render && t.render(s, u, i, l);
      }
      function U(n, h, o) {
        var v = n.getModel("iconStyle"), r = n.getModel(["emphasis", "iconStyle"]), s = h instanceof x && h.getIcons ? h.getIcons() : n.get("icon"), t = n.get("title") || {}, p, m;
        L(s) ? (p = {}, p[o] = s) : p = s, L(t) ? (m = {}, m[o] = t) : m = t;
        var d = n.iconPaths = {};
        F(p, function(y, a) {
          var f = K(y, {}, {
            x: -S / 2,
            y: -S / 2,
            width: S,
            height: S
          });
          f.setStyle(v.getItemStyle());
          var G = f.ensureState("emphasis");
          G.style = r.getItemStyle();
          var T = new W({
            style: {
              text: m[a],
              align: r.get("textAlign"),
              borderRadius: r.get("textBorderRadius"),
              padding: r.get("textPadding"),
              fill: null,
              font: tt({
                fontStyle: r.get("textFontStyle"),
                fontFamily: r.get("textFontFamily"),
                fontSize: r.get("textFontSize"),
                fontWeight: r.get("textFontWeight")
              }, u)
            },
            ignore: !0
          });
          f.setTextContent(T), Q({
            el: f,
            componentModel: e,
            itemName: a,
            formatterParamsExtra: {
              title: m[a]
            }
          }), f.__title = m[a], f.on("mouseover", function() {
            var z = r.getItemStyle(), H = _ ? e.get("right") == null && e.get("left") !== "right" ? "right" : "left" : e.get("bottom") == null && e.get("top") !== "bottom" ? "bottom" : "top";
            T.setStyle({
              fill: r.get("textFill") || z.fill || z.stroke || ot.color.neutral99,
              backgroundColor: r.get("textBackgroundColor")
            }), f.setTextConfig({
              position: r.get("textPosition") || H
            }), T.ignore = !e.get("showTitle"), i.enterEmphasis(this);
          }).on("mouseout", function() {
            n.get(["iconStatus", a]) !== "emphasis" && i.leaveEmphasis(this), T.hide();
          }), (n.get(["iconStatus", a]) === "emphasis" ? A : D)(f), g.add(f), f.on("click", j(h.onclick, h, u, i, a)), d[a] = f;
        });
      }
      var E = et(e, i).refContainer, R = e.getBoxLayoutParams(), B = e.get("padding"), V = it(R, E, B);
      rt(e.get("orient"), g, e.get("itemGap"), V.width, V.height), nt(g, R, E, B), g.add(b(g.getBoundingRect(), e)), _ || g.eachChild(function(n) {
        var h = n.__title, o = n.ensureState("emphasis"), v = o.textConfig || (o.textConfig = {}), r = n.getTextContent(), s = r && r.ensureState("emphasis");
        if (s && !q(s) && h) {
          var t = s.style || (s.style = {}), p = J(h, W.makeFont(t)), m = n.x + g.x, d = n.y + g.y + S, y = !1;
          d + p.height > i.getHeight() && (v.position = "top", y = !0);
          var a = y ? -5 - p.height : S + 10;
          m + p.width / 2 > i.getWidth() ? (v.position = ["100%", a], t.align = "right") : m - p.width / 2 < 0 && (v.position = [0, a], t.align = "left");
        }
      });
    }, c.prototype.updateView = function(e, u, i, l) {
      F(this._features, function(g) {
        g instanceof x && g.updateView && g.updateView(g.model, u, i, l);
      });
    }, c.prototype.remove = function(e, u) {
      F(this._features, function(i) {
        i instanceof x && i.remove && i.remove(e, u);
      }), this.group.removeAll();
    }, c.prototype.dispose = function(e, u) {
      F(this._features, function(i) {
        i instanceof x && i.dispose && i.dispose(e, u);
      });
    }, c.type = "toolbox", c;
  })(N)
);
function st(w) {
  return w.indexOf("my") === 0;
}
export {
  Ct as default
};
