// echarts@6.0.0/lib/coord/cartesian/cartesianAxisHelper.js
import { retrieve as h, each as d, retrieve3 as A, assert as w, keys as y } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { SINGLE_REFERRING as b } from "../../util/model.js";
import D from "../../component/axis/AxisBuilder.js";
import { isIntervalOrLogScale as O } from "../../scale/helper.js";
function m(e, o, a) {
  a = a || {};
  var r = o.axis, t = {}, i = r.getAxesOnZeroOf()[0], s = r.position, u = i ? "onZero" : s, f = r.dim, n = [e.x, e.x + e.width, e.y, e.y + e.height], v = {
    left: 0,
    right: 1,
    top: 0,
    bottom: 1,
    onZero: 2
  }, c = o.get("offset") || 0, l = f === "x" ? [n[2] - c, n[3] + c] : [n[0] - c, n[1] + c];
  if (i) {
    var g = i.toGlobalCoord(i.dataToCoord(0));
    l[v.onZero] = Math.max(Math.min(g, l[1]), l[0]);
  }
  t.position = [f === "y" ? l[v[u]] : n[0], f === "x" ? l[v[u]] : n[3]], t.rotation = Math.PI / 2 * (f === "x" ? 0 : 1);
  var p = {
    top: -1,
    bottom: 1,
    left: -1,
    right: 1
  };
  t.labelDirection = t.tickDirection = t.nameDirection = p[s], t.labelOffset = i ? l[v[s]] - l[v.onZero] : 0, o.get(["axisTick", "inside"]) && (t.tickDirection = -t.tickDirection), h(a.labelInside, o.get(["axisLabel", "inside"])) && (t.labelDirection = -t.labelDirection);
  var x = o.get(["axisLabel", "rotate"]);
  return t.labelRotate = u === "top" ? -x : x, t.z2 = 1, t;
}
function I(e) {
  return e.coordinateSystem && e.coordinateSystem.type === "cartesian2d";
}
function N(e) {
  var o = {
    xAxisModel: null,
    yAxisModel: null
  };
  return d(o, function(a, r) {
    var t = r.replace(/Model$/, ""), i = e.getReferringComponents(t, b).models[0];
    if (process.env.NODE_ENV !== "production" && !i)
      throw new Error(t + ' "' + A(e.get(t + "Index"), e.get(t + "Id"), 0) + '" not found');
    o[r] = i;
  }), o;
}
function k(e, o, a, r, t, i) {
  for (var s = m(e, a), u = !1, f = !1, n = 0; n < o.length; n++)
    O(o[n].getOtherAxis(a.axis).scale) && (u = f = !0, a.axis.type === "category" && a.axis.onBand && (f = !1));
  return s.axisLineAutoShow = u, s.axisTickAutoShow = f, s.defaultNameMoveOverlap = i, new D(a, r, s, t);
}
function L(e, o, a) {
  var r = m(o, a);
  if (process.env.NODE_ENV !== "production") {
    var t = e.__getRawCfg();
    d(y(r), function(i) {
      i !== "position" && i !== "labelOffset" && w(r[i] === t[i]);
    });
  }
  e.updateCfg(r);
}
export {
  k as createCartesianAxisViewCommonPartBuilder,
  N as findAxisModels,
  I as isCartesian2DInjectedAsDataCoordSys,
  m as layout,
  L as updateCartesianAxisViewCommonPartBuilder
};
