// echarts@6.0.0/lib/coord/cartesian/Cartesian2D.js
import { __extends as A } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import l from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/BoundingRect.js";
import d from "./Cartesian.js";
import { invert as u } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/matrix.js";
import { applyTransform as m } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/vector.js";
var T = ["x", "y"];
function f(x) {
  return (x.type === "interval" || x.type === "time") && !x.hasBreaks();
}
var _ = (
  /** @class */
  (function(x) {
    A(s, x);
    function s() {
      var a = x !== null && x.apply(this, arguments) || this;
      return a.type = "cartesian2d", a.dimensions = T, a;
    }
    return s.prototype.calcAffineTransform = function() {
      this._transform = this._invTransform = null;
      var a = this.getAxis("x").scale, r = this.getAxis("y").scale;
      if (!(!f(a) || !f(r))) {
        var t = a.getExtent(), i = r.getExtent(), n = this.dataToPoint([t[0], i[0]]), e = this.dataToPoint([t[1], i[1]]), o = t[1] - t[0], h = i[1] - i[0];
        if (!(!o || !h)) {
          var v = (e[0] - n[0]) / o, c = (e[1] - n[1]) / h, y = n[0] - t[0] * v, g = n[1] - i[0] * c, p = this._transform = [v, 0, 0, c, y, g];
          this._invTransform = u([], p);
        }
      }
    }, s.prototype.getBaseAxis = function() {
      return this.getAxesByScale("ordinal")[0] || this.getAxesByScale("time")[0] || this.getAxis("x");
    }, s.prototype.containPoint = function(a) {
      var r = this.getAxis("x"), t = this.getAxis("y");
      return r.contain(r.toLocalCoord(a[0])) && t.contain(t.toLocalCoord(a[1]));
    }, s.prototype.containData = function(a) {
      return this.getAxis("x").containData(a[0]) && this.getAxis("y").containData(a[1]);
    }, s.prototype.containZone = function(a, r) {
      var t = this.dataToPoint(a), i = this.dataToPoint(r), n = this.getArea(), e = new l(t[0], t[1], i[0] - t[0], i[1] - t[1]);
      return n.intersect(e);
    }, s.prototype.dataToPoint = function(a, r, t) {
      t = t || [];
      var i = a[0], n = a[1];
      if (this._transform && i != null && isFinite(i) && n != null && isFinite(n))
        return m(t, a, this._transform);
      var e = this.getAxis("x"), o = this.getAxis("y");
      return t[0] = e.toGlobalCoord(e.dataToCoord(i, r)), t[1] = o.toGlobalCoord(o.dataToCoord(n, r)), t;
    }, s.prototype.clampData = function(a, r) {
      var t = this.getAxis("x").scale, i = this.getAxis("y").scale, n = t.getExtent(), e = i.getExtent(), o = t.parse(a[0]), h = i.parse(a[1]);
      return r = r || [], r[0] = Math.min(Math.max(Math.min(n[0], n[1]), o), Math.max(n[0], n[1])), r[1] = Math.min(Math.max(Math.min(e[0], e[1]), h), Math.max(e[0], e[1])), r;
    }, s.prototype.pointToData = function(a, r, t) {
      if (t = t || [], this._invTransform)
        return m(t, a, this._invTransform);
      var i = this.getAxis("x"), n = this.getAxis("y");
      return t[0] = i.coordToData(i.toLocalCoord(a[0]), r), t[1] = n.coordToData(n.toLocalCoord(a[1]), r), t;
    }, s.prototype.getOtherAxis = function(a) {
      return this.getAxis(a.dim === "x" ? "y" : "x");
    }, s.prototype.getArea = function(a) {
      a = a || 0;
      var r = this.getAxis("x").getGlobalExtent(), t = this.getAxis("y").getGlobalExtent(), i = Math.min(r[0], r[1]) - a, n = Math.min(t[0], t[1]) - a, e = Math.max(r[0], r[1]) - i + a, o = Math.max(t[0], t[1]) - n + a;
      return new l(i, n, e, o);
    }, s;
  })(d)
);
export {
  T as cartesian2DDimensions,
  _ as default
};
