// echarts@6.0.0/lib/coord/cartesian/Grid.js
import { each as h, isObject as W, indexOf as C, retrieve3 as Y, keys as q, assert as R, find as J, retrieve2 as N, eqNaN as Q } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { createBoxLayoutReference as $, getLayoutRect as X } from "../../util/layout.js";
import { createScaleByModel as ee, getDataDimensionsOnAxis as te, niceScaleExtent as D, shouldAxisShow as L, isNameLocationCenter as H, ifAxisCrossZero as re } from "../axisHelper.js";
import ne, { cartesian2DDimensions as T } from "./Cartesian2D.js";
import ie from "./Axis2D.js";
import { SINGLE_REFERRING as G } from "../../util/model.js";
import { OUTER_BOUNDS_DEFAULT as oe, OUTER_BOUNDS_CLAMP_DEFAULT as P } from "./GridModel.js";
import { isCartesian2DInjectedAsDataCoordSys as ae, findAxisModels as M, createCartesianAxisViewCommonPartBuilder as se, updateCartesianAxisViewCommonPartBuilder as ue } from "./cartesianAxisHelper.js";
import { isIntervalOrLogScale as w } from "../../scale/helper.js";
import { alignScaleTicks as fe } from "../axisAlignTicks.js";
import { expandOrShrinkRect as le, XY as A, WH as _ } from "../../util/graphic.js";
import { AxisBuilderSharedContext as ce, getLabelInner as de, resolveAxisNameOverlapDefault as ve, moveIfOverlapByLinearLabels as pe } from "../../component/axis/AxisBuilder.js";
import { log as me, error as I } from "../../util/log.js";
import { AxisTickLabelComputingKind as E } from "../axisTickLabelBuilder.js";
import { injectCoordSysByOption as he } from "../../core/CoordinateSystem.js";
import { mathMax as k, parsePositionSizeOption as V } from "../../util/number.js";
var Z = [
  [3, 1],
  [0, 2]
  // xyIdx 1 => 'y'
], ke = (
  /** @class */
  (function() {
    function i(t, r, n) {
      this.type = "grid", this._coordsMap = {}, this._coordsList = [], this._axesMap = {}, this._axesList = [], this.axisPointerEnabled = !0, this.dimensions = T, this._initCartesian(t, r, n), this.model = t;
    }
    return i.prototype.getRect = function() {
      return this._rect;
    }, i.prototype.update = function(t, r) {
      var n = this._axesMap;
      this._updateScale(t, this.model);
      function e(a) {
        var s, c = q(a), l = c.length;
        if (l) {
          for (var u = [], d = l - 1; d >= 0; d--) {
            var v = +c[d], f = a[v], p = f.model, m = f.scale;
            // Only value and log axis without interval support alignTicks.
            w(m) && p.get("alignTicks") && p.get("interval") == null ? u.push(f) : (D(m, p), w(m) && (s = f));
          }
          u.length && (s || (s = u.pop(), D(s.scale, s.model)), h(u, function(g) {
            fe(g.scale, g.model, s.scale);
          }));
        }
      }
      e(n.x), e(n.y);
      var o = {};
      h(n.x, function(a) {
        U(n, "y", a, o);
      }), h(n.y, function(a) {
        U(n, "x", a, o);
      }), this.resize(this.model, r);
    }, i.prototype.resize = function(t, r, n) {
      var e = $(t, r), o = this._rect = X(t.getBoxLayoutParams(), e.refContainer), a = this._axesMap, s = this._coordsList, c = t.get("containLabel");
      if (B(a, o), !n) {
        var l = xe(o, s, a, c, r), u = void 0;
        if (c)
          S ? (S(this._axesList, o), B(a, o)) : (process.env.NODE_ENV !== "production" && me("Specified `grid.containLabel` but no `use(LegacyGridContainLabel)`;use `grid.outerBounds` instead.", !0), u = j(o.clone(), "axisLabel", null, o, a, l, e));
        else {
          var d = ye(t, o, e), v = d.outerBoundsRect, f = d.parsedOuterBoundsContain, p = d.outerBoundsClamp;
          v && (u = j(v, f, p, o, a, l, e));
        }
        K(o, a, E.determine, null, u, e);
      }
      h(this._coordsList, function(m) {
        m.calcAffineTransform();
      });
    }, i.prototype.getAxis = function(t, r) {
      var n = this._axesMap[t];
      if (n != null)
        return n[r || 0];
    }, i.prototype.getAxes = function() {
      return this._axesList.slice();
    }, i.prototype.getCartesian = function(t, r) {
      if (t != null && r != null) {
        var n = "x" + t + "y" + r;
        return this._coordsMap[n];
      }
      W(t) && (r = t.yAxisIndex, t = t.xAxisIndex);
      for (var e = 0, o = this._coordsList; e < o.length; e++)
        if (o[e].getAxis("x").index === t || o[e].getAxis("y").index === r)
          return o[e];
    }, i.prototype.getCartesians = function() {
      return this._coordsList.slice();
    }, i.prototype.convertToPixel = function(t, r, n) {
      var e = this._findConvertTarget(r);
      return e.cartesian ? e.cartesian.dataToPoint(n) : e.axis ? e.axis.toGlobalCoord(e.axis.dataToCoord(n)) : null;
    }, i.prototype.convertFromPixel = function(t, r, n) {
      var e = this._findConvertTarget(r);
      return e.cartesian ? e.cartesian.pointToData(n) : e.axis ? e.axis.coordToData(e.axis.toLocalCoord(n)) : null;
    }, i.prototype._findConvertTarget = function(t) {
      var r = t.seriesModel, n = t.xAxisModel || r && r.getReferringComponents("xAxis", G).models[0], e = t.yAxisModel || r && r.getReferringComponents("yAxis", G).models[0], o = t.gridModel, a = this._coordsList, s, c;
      if (r)
        s = r.coordinateSystem, C(a, s) < 0 && (s = null);
      else if (n && e)
        s = this.getCartesian(n.componentIndex, e.componentIndex);
      else if (n)
        c = this.getAxis("x", n.componentIndex);
      else if (e)
        c = this.getAxis("y", e.componentIndex);
      else if (o) {
        var l = o.coordinateSystem;
        l === this && (s = this._coordsList[0]);
      }
      return {
        cartesian: s,
        axis: c
      };
    }, i.prototype.containPoint = function(t) {
      var r = this._coordsList[0];
      if (r)
        return r.containPoint(t);
    }, i.prototype._initCartesian = function(t, r, n) {
      var e = this, o = this, a = {
        left: !1,
        right: !1,
        top: !1,
        bottom: !1
      }, s = {
        x: {},
        y: {}
      }, c = {
        x: 0,
        y: 0
      };
      if (r.eachComponent("xAxis", l("x"), this), r.eachComponent("yAxis", l("y"), this), !c.x || !c.y) {
        this._axesMap = {}, this._axesList = [];
        return;
      }
      this._axesMap = s, h(s.x, function(u, d) {
        h(s.y, function(v, f) {
          var p = "x" + d + "y" + f, m = new ne(p);
          m.master = e, m.model = t, e._coordsMap[p] = m, e._coordsList.push(m), m.addAxis(u), m.addAxis(v);
        });
      });
      function l(u) {
        return function(d, v) {
          if (b(d, t)) {
            var f = d.get("position");
            u === "x" ? f !== "top" && f !== "bottom" && (f = a.bottom ? "top" : "bottom") : f !== "left" && f !== "right" && (f = a.left ? "right" : "left"), a[f] = !0;
            var p = new ie(u, ee(d), [0, 0], d.get("type"), f), m = p.type === "category";
            p.onBand = m && d.get("boundaryGap"), p.inverse = d.get("inverse"), d.axis = p, p.model = d, p.grid = o, p.index = v, o._axesList.push(p), s[u][v] = p, c[u]++;
          }
        };
      }
    }, i.prototype._updateScale = function(t, r) {
      h(this._axesList, function(e) {
        if (e.scale.setExtent(1 / 0, -1 / 0), e.type === "category") {
          var o = e.model.get("categorySortInfo");
          e.scale.setSortInfo(o);
        }
      }), t.eachSeries(function(e) {
        if (ae(e)) {
          var o = M(e), a = o.xAxisModel, s = o.yAxisModel;
          if (!b(a, r) || !b(s, r))
            return;
          var c = this.getCartesian(a.componentIndex, s.componentIndex), l = e.getData(), u = c.getAxis("x"), d = c.getAxis("y");
          n(l, u), n(l, d);
        }
      }, this);
      function n(e, o) {
        h(te(e, o.dim), function(a) {
          o.scale.unionExtentFromData(e, a);
        });
      }
    }, i.prototype.getTooltipAxes = function(t) {
      var r = [], n = [];
      return h(this.getCartesians(), function(e) {
        var o = t != null && t !== "auto" ? e.getAxis(t) : e.getBaseAxis(), a = e.getOtherAxis(o);
        C(r, o) < 0 && r.push(o), C(n, a) < 0 && n.push(a);
      }), {
        baseAxes: r,
        otherAxes: n
      };
    }, i.create = function(t, r) {
      var n = [];
      return t.eachComponent("grid", function(e, o) {
        var a = new i(e, t, r);
        a.name = "grid_" + o, a.resize(e, r, !0), e.coordinateSystem = a, n.push(a);
      }), t.eachSeries(function(e) {
        he({
          targetModel: e,
          coordSysType: "cartesian2d",
          coordSysProvider: o
        });
        function o() {
          var a = M(e), s = a.xAxisModel, c = a.yAxisModel, l = s.getCoordSysModel();
          if (process.env.NODE_ENV !== "production") {
            if (!l)
              throw new Error('Grid "' + Y(s.get("gridIndex"), s.get("gridId"), 0) + '" not found');
            if (s.getCoordSysModel() !== c.getCoordSysModel())
              throw new Error("xAxis and yAxis must use the same grid");
          }
          var u = l.coordinateSystem;
          return u.getCartesian(s.componentIndex, c.componentIndex);
        }
      }), n;
    }, i.dimensions = T, i;
  })()
);
function b(i, t) {
  return i.getCoordSysModel() === t;
}
function U(i, t, r, n) {
  r.getAxesOnZeroOf = function() {
    return o ? [o] : [];
  };
  var e = i[t], o, a = r.model, s = a.get(["axisLine", "onZero"]), c = a.get(["axisLine", "onZeroAxisIndex"]);
  if (!s)
    return;
  if (c != null)
    F(e[c]) && (o = e[c]);
  else
    for (var l in e)
      if (e.hasOwnProperty(l) && F(e[l]) && !n[u(e[l])]) {
        o = e[l];
        break;
      }
  o && (n[u(o)] = !0);
  function u(d) {
    return d.dim + "_" + d.index;
  }
}
function F(i) {
  return i && i.type !== "category" && i.type !== "time" && re(i);
}
function ge(i, t) {
  var r = i.getExtent(), n = r[0] + r[1];
  i.toGlobalCoord = i.dim === "x" ? function(e) {
    return e + t;
  } : function(e) {
    return n - e + t;
  }, i.toLocalCoord = i.dim === "x" ? function(e) {
    return e - t;
  } : function(e) {
    return n - e + t;
  };
}
function B(i, t) {
  h(i.x, function(r) {
    return z(r, t.x, t.width);
  }), h(i.y, function(r) {
    return z(r, t.y, t.height);
  });
}
function z(i, t, r) {
  var n = [0, r], e = i.inverse ? 1 : 0;
  i.setExtent(n[e], n[1 - e]), ge(i, t);
}
var S;
function Ve(i) {
  S = i;
}
function j(i, t, r, n, e, o, a) {
  process.env.NODE_ENV !== "production" && R(t === "all" || t === "axisLabel"), K(n, e, E.estimate, t, !1, a);
  var s = [0, 0, 0, 0];
  l(0), l(1), u(n, 0, NaN), u(n, 1, NaN);
  var c = J(s, function(v) {
    return v > 0;
  }) == null;
  return le(n, s, !0, !0, r), B(e, n), c;
  function l(v) {
    h(e[A[v]], function(f) {
      if (L(f.model)) {
        var p = o.ensureRecord(f.model), m = p.labelInfoList;
        if (m)
          for (var g = 0; g < m.length; g++) {
            var y = m[g], x = f.scale.normalize(de(y.label).tickValue);
            x = v === 1 ? 1 - x : x, u(y.rect, v, x), u(y.rect, 1 - v, NaN);
          }
        var O = p.nameLayout;
        if (O) {
          var x = H(p.nameLocation) ? 0.5 : NaN;
          u(O.rect, v, x), u(O.rect, 1 - v, NaN);
        }
      }
    });
  }
  function u(v, f, p) {
    var m = i[A[f]] - v[A[f]], g = v[_[f]] + v[A[f]] - (i[_[f]] + i[A[f]]);
    m = d(m, 1 - p), g = d(g, p);
    var y = Z[f][0], x = Z[f][1];
    s[y] = k(s[y], m), s[x] = k(s[x], g);
  }
  function d(v, f) {
    return v > 0 && !Q(f) && f > 1e-4 && (v /= f), v;
  }
}
function xe(i, t, r, n, e) {
  var o = new ce(Ae);
  return h(r, function(a) {
    return h(a, function(s) {
      if (L(s.model)) {
        var c = !n;
        s.axisBuilder = se(i, t, s.model, e, o, c);
      }
    });
  }), o;
}
function K(i, t, r, n, e, o) {
  var a = r === E.determine;
  h(t, function(l) {
    return h(l, function(u) {
      L(u.model) && (ue(u.axisBuilder, i, u.model), u.axisBuilder.build(a ? {
        axisTickLabelDetermine: !0
      } : {
        axisTickLabelEstimate: !0
      }, {
        noPxChange: e
      }));
    });
  });
  var s = {
    x: 0,
    y: 0
  };
  c(0), c(1);
  function c(l) {
    s[A[1 - l]] = i[_[l]] <= o.refContainer[_[l]] * 0.5 ? 0 : 1 - l === 1 ? 2 : 1;
  }
  h(t, function(l, u) {
    return h(l, function(d) {
      L(d.model) && ((n === "all" || a) && d.axisBuilder.build({
        axisName: !0
      }, {
        nameMarginLevel: s[u]
      }), a && d.axisBuilder.build({
        axisLine: !0
      }));
    });
  });
}
function ye(i, t, r) {
  var n, e = i.get("outerBoundsMode", !0);
  e === "same" ? n = t.clone() : e == null || e === "auto" ? n = X(i.get("outerBounds", !0) || oe, r.refContainer) : e !== "none" && process.env.NODE_ENV !== "production" && I("Invalid grid[" + i.componentIndex + "].outerBoundsMode.");
  var o = i.get("outerBoundsContain", !0), a;
  o == null || o === "auto" ? a = "all" : C(["all", "axisLabel"], o) < 0 ? (process.env.NODE_ENV !== "production" && I("Invalid grid[" + i.componentIndex + "].outerBoundsContain."), a = "all") : a = o;
  var s = [V(N(i.get("outerBoundsClampWidth", !0), P[0]), t.width), V(N(i.get("outerBoundsClampHeight", !0), P[1]), t.height)];
  return {
    outerBoundsRect: n,
    parsedOuterBoundsContain: a,
    outerBoundsClamp: s
  };
}
var Ae = function(i, t, r, n, e, o) {
  var a = r.axis.dim === "x" ? "y" : "x";
  ve(i, t, r, n, e, o), H(i.nameLocation) || h(t.recordMap[a], function(s) {
    s && s.labelInfoList && s.dirVec && pe(s.labelInfoList, s.dirVec, n, e);
  });
};
export {
  ke as default,
  Ve as registerLegacyGridContainLabelImpl
};
