// echarts@6.0.0/lib/coord/cartesian/GridModel.js
import { __extends as i } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import l from "../../model/Component.js";
import { getLayoutParams as s, mergeLayoutParam as n } from "../../util/layout.js";
import u from "../../visual/tokens.js";
var p = {
  left: 0,
  right: 0,
  top: 0,
  bottom: 0
}, a = ["25%", "25%"], h = (
  /** @class */
  (function(e) {
    i(o, e);
    function o() {
      return e !== null && e.apply(this, arguments) || this;
    }
    return o.prototype.mergeDefaultAndTheme = function(t, d) {
      var r = s(t.outerBounds);
      e.prototype.mergeDefaultAndTheme.apply(this, arguments), r && t.outerBounds && n(t.outerBounds, r);
    }, o.prototype.mergeOption = function(t, d) {
      e.prototype.mergeOption.apply(this, arguments), this.option.outerBounds && t.outerBounds && n(this.option.outerBounds, t.outerBounds);
    }, o.type = "grid", o.dependencies = ["xAxis", "yAxis"], o.layoutMode = "box", o.defaultOption = {
      show: !1,
      // zlevel: 0,
      z: 0,
      left: "15%",
      top: 65,
      right: "10%",
      bottom: 80,
      // If grid size contain label
      containLabel: !1,
      outerBoundsMode: "auto",
      outerBounds: p,
      outerBoundsContain: "all",
      outerBoundsClampWidth: a[0],
      outerBoundsClampHeight: a[1],
      // width: {totalWidth} - left - right,
      // height: {totalHeight} - top - bottom,
      backgroundColor: u.color.transparent,
      borderWidth: 1,
      borderColor: u.color.neutral30
    }, o;
  })(l)
);
export {
  a as OUTER_BOUNDS_CLAMP_DEFAULT,
  p as OUTER_BOUNDS_DEFAULT,
  h as default
};
