// echarts@6.0.0/lib/coord/cartesian/legacyContainLabel.js
import C from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/BoundingRect.js";
import { each as T } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { registerLegacyGridContainLabelImpl as B } from "./Grid.js";
import G from "../../scale/Ordinal.js";
import { makeLabelFormatter as H } from "../axisHelper.js";
function z() {
  B(S);
}
function S(r, o) {
  T(r, function(a) {
    if (!a.model.get(["axisLabel", "inside"])) {
      var t = E(a);
      if (t) {
        var e = a.isHorizontal() ? "height" : "width", i = a.model.get(["axisLabel", "margin"]);
        o[e] -= t[e] + i, a.position === "top" ? o.y += t.height + i : a.position === "left" && (o.x += t.width + i);
      }
    }
  });
}
function E(r) {
  var o = r.model, a = r.scale;
  if (!o.get(["axisLabel", "show"]) || a.isBlank())
    return;
  var t, e, i = a.getExtent();
  a instanceof G ? e = a.count() : (t = a.getTicks(), e = t.length);
  var v = r.getLabelModel(), g = H(r), l, s = 1;
  e > 40 && (s = Math.ceil(e / 40));
  for (var n = 0; n < e; n += s) {
    var u = t ? t[n] : {
      value: i[0] + n
    }, b = g(u, n), M = v.getTextRect(b), c = L(M, v.get("rotate") || 0);
    l ? l.union(c) : l = c;
  }
  return l;
  function L(h, p) {
    var f = p * Math.PI / 180, d = h.width, m = h.height, y = d * Math.abs(Math.cos(f)) + Math.abs(m * Math.sin(f)), k = d * Math.abs(Math.sin(f)) + Math.abs(m * Math.cos(f)), w = new C(h.x, h.y, y, k);
    return w;
  }
}
export {
  z as installLegacyGridContainLabel
};
