// zrender@6.0.0/lib/canvas/graphic.js
import { DEFAULT_COMMON_STYLE as p } from "../graphic/Displayable.js";
import $ from "../core/PathProxy.js";
import { createOrUpdateImage as V, isImageReady as j } from "../graphic/helper/image.js";
import { isClipPathChanged as x, getCanvasGradient as W } from "./helper.js";
import I from "../graphic/Path.js";
import c from "../graphic/Image.js";
import rr from "../graphic/TSpan.js";
import { RADIAN_TO_DEGREE as ar } from "../core/util.js";
import { getLineDash as q } from "./dashStyle.js";
import { REDRAW_BIT as ir, SHAPE_CHANGED_BIT as er } from "../graphic/constants.js";
import { DEFAULT_FONT as fr } from "../core/platform.js";
var nr = new $(!0);
function D(a) {
  var r = a.stroke;
  return !(r == null || r === "none" || !(a.lineWidth > 0));
}
function L(a) {
  return typeof a == "string" && a !== "none";
}
function P(a) {
  var r = a.fill;
  return r != null && r !== "none";
}
function M(a, r) {
  if (r.fillOpacity != null && r.fillOpacity !== 1) {
    var i = a.globalAlpha;
    a.globalAlpha = r.fillOpacity * r.opacity, a.fill(), a.globalAlpha = i;
  } else
    a.fill();
}
function Y(a, r) {
  if (r.strokeOpacity != null && r.strokeOpacity !== 1) {
    var i = a.globalAlpha;
    a.globalAlpha = r.strokeOpacity * r.opacity, a.stroke(), a.globalAlpha = i;
  } else
    a.stroke();
}
function U(a, r, i) {
  var f = V(r.image, r.__image, i);
  if (j(f)) {
    var n = a.createPattern(f, r.repeat || "repeat");
    if (typeof DOMMatrix == "function" && n && n.setTransform) {
      var e = new DOMMatrix();
      e.translateSelf(r.x || 0, r.y || 0), e.rotateSelf(0, 0, (r.rotation || 0) * ar), e.scaleSelf(r.scaleX || 1, r.scaleY || 1), n.setTransform(e);
    }
    return n;
  }
}
function tr(a, r, i, f) {
  var n, e = D(i), o = P(i), h = i.strokePercent, t = h < 1, l = !r.path;
  (!r.silent || t) && l && r.createPathProxy();
  var v = r.path || nr, u = r.__dirty;
  if (!f) {
    var d = i.fill, _ = i.stroke, C = o && !!d.colorStops, w = e && !!_.colorStops, O = o && !!d.image, A = e && !!_.image, S = void 0, k = void 0, b = void 0, m = void 0, B = void 0;
    (C || w) && (B = r.getBoundingRect()), C && (S = u ? W(a, d, B) : r.__canvasFillGradient, r.__canvasFillGradient = S), w && (k = u ? W(a, _, B) : r.__canvasStrokeGradient, r.__canvasStrokeGradient = k), O && (b = u || !r.__canvasFillPattern ? U(a, d, r) : r.__canvasFillPattern, r.__canvasFillPattern = b), A && (m = u || !r.__canvasStrokePattern ? U(a, _, r) : r.__canvasStrokePattern, r.__canvasStrokePattern = m), C ? a.fillStyle = S : O && (b ? a.fillStyle = b : o = !1), w ? a.strokeStyle = k : A && (m ? a.strokeStyle = m : e = !1);
  }
  var H = r.getGlobalScale();
  v.setScale(H[0], H[1], r.segmentIgnoreThreshold);
  var T, R;
  a.setLineDash && i.lineDash && (n = q(r), T = n[0], R = n[1]);
  var F = !0;
  (l || u & er) && (v.setDPR(a.dpr), t ? v.setContext(null) : (v.setContext(a), F = !1), v.reset(), r.buildPath(v, r.shape, f), v.toStatic(), r.pathUpdated()), F && v.rebuildPath(a, t ? h : 1), T && (a.setLineDash(T), a.lineDashOffset = R), f || (i.strokeFirst ? (e && Y(a, i), o && M(a, i)) : (o && M(a, i), e && Y(a, i))), T && a.setLineDash([]);
}
function or(a, r, i) {
  var f = r.__image = V(i.image, r.__image, r, r.onload);
  if (!(!f || !j(f))) {
    var n = i.x || 0, e = i.y || 0, o = r.getWidth(), h = r.getHeight(), t = f.width / f.height;
    if (o == null && h != null ? o = h * t : h == null && o != null ? h = o / t : o == null && h == null && (o = f.width, h = f.height), i.sWidth && i.sHeight) {
      var l = i.sx || 0, v = i.sy || 0;
      a.drawImage(f, l, v, i.sWidth, i.sHeight, n, e, o, h);
    } else if (i.sx && i.sy) {
      var l = i.sx, v = i.sy, u = o - l, d = h - v;
      a.drawImage(f, l, v, u, d, n, e, o, h);
    } else
      a.drawImage(f, n, e, o, h);
  }
}
function hr(a, r, i) {
  var f, n = i.text;
  if (n != null && (n += ""), n) {
    a.font = i.font || fr, a.textAlign = i.textAlign, a.textBaseline = i.textBaseline;
    var e = void 0, o = void 0;
    a.setLineDash && i.lineDash && (f = q(r), e = f[0], o = f[1]), e && (a.setLineDash(e), a.lineDashOffset = o), i.strokeFirst ? (D(i) && a.strokeText(n, i.x, i.y), P(i) && a.fillText(n, i.x, i.y)) : (P(i) && a.fillText(n, i.x, i.y), D(i) && a.strokeText(n, i.x, i.y)), e && a.setLineDash([]);
  }
}
var N = ["shadowBlur", "shadowOffsetX", "shadowOffsetY"], y = [
  ["lineCap", "butt"],
  ["lineJoin", "miter"],
  ["miterLimit", 10]
];
function z(a, r, i, f, n) {
  var e = !1;
  if (!f && (i = i || {}, r === i))
    return !1;
  if (f || r.opacity !== i.opacity) {
    s(a, n), e = !0;
    var o = Math.max(Math.min(r.opacity, 1), 0);
    a.globalAlpha = isNaN(o) ? p.opacity : o;
  }
  (f || r.blend !== i.blend) && (e || (s(a, n), e = !0), a.globalCompositeOperation = r.blend || p.blend);
  for (var h = 0; h < N.length; h++) {
    var t = N[h];
    (f || r[t] !== i[t]) && (e || (s(a, n), e = !0), a[t] = a.dpr * (r[t] || 0));
  }
  return (f || r.shadowColor !== i.shadowColor) && (e || (s(a, n), e = !0), a.shadowColor = r.shadowColor || p.shadowColor), e;
}
function X(a, r, i, f, n) {
  var e = g(r, n.inHover), o = f ? null : i && g(i, n.inHover) || {};
  if (e === o)
    return !1;
  var h = z(a, e, o, f, n);
  if ((f || e.fill !== o.fill) && (h || (s(a, n), h = !0), L(e.fill) && (a.fillStyle = e.fill)), (f || e.stroke !== o.stroke) && (h || (s(a, n), h = !0), L(e.stroke) && (a.strokeStyle = e.stroke)), (f || e.opacity !== o.opacity) && (h || (s(a, n), h = !0), a.globalAlpha = e.opacity == null ? 1 : e.opacity), r.hasStroke()) {
    var t = e.lineWidth, l = t / (e.strokeNoScale && r.getLineScale ? r.getLineScale() : 1);
    a.lineWidth !== l && (h || (s(a, n), h = !0), a.lineWidth = l);
  }
  for (var v = 0; v < y.length; v++) {
    var u = y[v], d = u[0];
    (f || e[d] !== o[d]) && (h || (s(a, n), h = !0), a[d] = e[d] || u[1]);
  }
  return h;
}
function vr(a, r, i, f, n) {
  return z(a, g(r, n.inHover), i && g(i, n.inHover), f, n);
}
function Q(a, r) {
  var i = r.transform, f = a.dpr || 1;
  i ? a.setTransform(f * i[0], f * i[1], f * i[2], f * i[3], f * i[4], f * i[5]) : a.setTransform(f, 0, 0, f, 0, 0);
}
function lr(a, r, i) {
  for (var f = !1, n = 0; n < a.length; n++) {
    var e = a[n];
    f = f || e.isZeroArea(), Q(r, e), r.beginPath(), e.buildPath(r, e.shape), r.clip();
  }
  i.allClipped = f;
}
function ur(a, r) {
  return a && r ? a[0] !== r[0] || a[1] !== r[1] || a[2] !== r[2] || a[3] !== r[3] || a[4] !== r[4] || a[5] !== r[5] : !(!a && !r);
}
var Z = 1, G = 2, J = 3, K = 4;
function sr(a) {
  var r = P(a), i = D(a);
  return !(a.lineDash || !(+r ^ +i) || r && typeof a.fill != "string" || i && typeof a.stroke != "string" || a.strokePercent < 1 || a.strokeOpacity < 1 || a.fillOpacity < 1);
}
function s(a, r) {
  r.batchFill && a.fill(), r.batchStroke && a.stroke(), r.batchFill = "", r.batchStroke = "";
}
function g(a, r) {
  return r && a.__hoverStyle || a.style;
}
function Br(a, r) {
  E(a, r, { inHover: !1, viewWidth: 0, viewHeight: 0 }, !0);
}
function E(a, r, i, f) {
  var n = r.transform;
  if (!r.shouldBePainted(i.viewWidth, i.viewHeight, !1, !1)) {
    r.__dirty &= ~ir, r.__isRendered = !1;
    return;
  }
  var e = r.__clipPaths, o = i.prevElClipPaths, h = !1, t = !1;
  if ((!o || x(e, o)) && (o && o.length && (s(a, i), a.restore(), t = h = !0, i.prevElClipPaths = null, i.allClipped = !1, i.prevEl = null), e && e.length && (s(a, i), a.save(), lr(e, a, i), h = !0), i.prevElClipPaths = e), i.allClipped) {
    r.__isRendered = !1;
    return;
  }
  r.beforeBrush && r.beforeBrush(), r.innerBeforeBrush();
  var l = i.prevEl;
  l || (t = h = !0);
  var v = r instanceof I && r.autoBatch && sr(r.style);
  h || ur(n, l.transform) ? (s(a, i), Q(a, r)) : v || s(a, i);
  var u = g(r, i.inHover);
  r instanceof I ? (i.lastDrawType !== Z && (t = !0, i.lastDrawType = Z), X(a, r, l, t, i), (!v || !i.batchFill && !i.batchStroke) && a.beginPath(), tr(a, r, u, v), v && (i.batchFill = u.fill || "", i.batchStroke = u.stroke || "")) : r instanceof rr ? (i.lastDrawType !== J && (t = !0, i.lastDrawType = J), X(a, r, l, t, i), hr(a, r, u)) : r instanceof c ? (i.lastDrawType !== G && (t = !0, i.lastDrawType = G), vr(a, r, l, t, i), or(a, r, u)) : r.getTemporalDisplayables && (i.lastDrawType !== K && (t = !0, i.lastDrawType = K), dr(a, r, i)), v && f && s(a, i), r.innerAfterBrush(), r.afterBrush && r.afterBrush(), i.prevEl = r, r.__dirty = 0, r.__isRendered = !0;
}
function dr(a, r, i) {
  var f = r.getDisplayables(), n = r.getTemporalDisplayables();
  a.save();
  var e = {
    prevElClipPaths: null,
    prevEl: null,
    allClipped: !1,
    viewWidth: i.viewWidth,
    viewHeight: i.viewHeight,
    inHover: i.inHover
  }, o, h;
  for (o = r.getCursor(), h = f.length; o < h; o++) {
    var t = f[o];
    t.beforeBrush && t.beforeBrush(), t.innerBeforeBrush(), E(a, t, e, o === h - 1), t.innerAfterBrush(), t.afterBrush && t.afterBrush(), e.prevEl = t;
  }
  for (var l = 0, v = n.length; l < v; l++) {
    var t = n[l];
    t.beforeBrush && t.beforeBrush(), t.innerBeforeBrush(), E(a, t, e, l === v - 1), t.innerAfterBrush(), t.afterBrush && t.afterBrush(), e.prevEl = t;
  }
  r.clearTemporalDisplayables(), r.notClear = !0, a.restore();
}
export {
  E as brush,
  Br as brushSingle,
  U as createCanvasPattern
};
