// zrender@6.0.0/lib/canvas/helper.js
function f(r) {
  return isFinite(r);
}
function h(r, n, a) {
  var t = n.x == null ? 0 : n.x, i = n.x2 == null ? 1 : n.x2, l = n.y == null ? 0 : n.y, e = n.y2 == null ? 0 : n.y2;
  n.global || (t = t * a.width + a.x, i = i * a.width + a.x, l = l * a.height + a.y, e = e * a.height + a.y), t = f(t) ? t : 0, i = f(i) ? i : 1, l = f(l) ? l : 0, e = f(e) ? e : 0;
  var u = r.createLinearGradient(t, l, i, e);
  return u;
}
function y(r, n, a) {
  var t = a.width, i = a.height, l = Math.min(t, i), e = n.x == null ? 0.5 : n.x, u = n.y == null ? 0.5 : n.y, d = n.r == null ? 0.5 : n.r;
  n.global || (e = e * t + a.x, u = u * i + a.y, d = d * l), e = f(e) ? e : 0.5, u = f(u) ? u : 0.5, d = d >= 0 && f(d) ? d : 0.5;
  var v = r.createRadialGradient(e, u, 0, e, u, d);
  return v;
}
function x(r, n, a) {
  for (var t = n.type === "radial" ? y(r, n, a) : h(r, n, a), i = n.colorStops, l = 0; l < i.length; l++)
    t.addColorStop(i[l].offset, i[l].color);
  return t;
}
function o(r, n) {
  if (r === n || !r && !n)
    return !1;
  if (!r || !n || r.length !== n.length)
    return !0;
  for (var a = 0; a < r.length; a++)
    if (r[a] !== n[a])
      return !0;
  return !1;
}
function g(r) {
  return parseInt(r, 10);
}
function G(r, n, a) {
  var t = ["width", "height"][n], i = ["clientWidth", "clientHeight"][n], l = ["paddingLeft", "paddingTop"][n], e = ["paddingRight", "paddingBottom"][n];
  if (a[t] != null && a[t] !== "auto")
    return parseFloat(a[t]);
  var u = document.defaultView.getComputedStyle(r);
  return (r[i] || g(u[t]) || g(r.style[t])) - (g(u[l]) || 0) - (g(u[e]) || 0) | 0;
}
export {
  h as createLinearGradient,
  y as createRadialGradient,
  x as getCanvasGradient,
  G as getSize,
  o as isClipPathChanged
};
