// zrender@6.0.0/lib/canvas/Layer.js
import { __extends as F } from "../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { isObject as D, disableUserSelect as L, each as U, isGradientObject as G, isImagePatternObject as W } from "../core/util.js";
import { devicePixelRatio as X } from "../config.js";
import Y from "../core/Eventful.js";
import { getCanvasGradient as Z } from "./helper.js";
import { createCanvasPattern as O } from "./graphic.js";
import R from "../core/BoundingRect.js";
import { REDRAW_BIT as S } from "../graphic/constants.js";
import { platformApi as q } from "../core/platform.js";
function E(I, h, a) {
  var i = q.createCanvas(), r = h.getWidth(), e = h.getHeight(), t = i.style;
  return t && (t.position = "absolute", t.left = "0", t.top = "0", t.width = r + "px", t.height = e + "px", i.setAttribute("data-zr-dom-id", I)), i.width = r * a, i.height = e * a, i;
}
var C = (function(I) {
  F(h, I);
  function h(a, i, r) {
    var e = I.call(this) || this;
    e.motionBlur = !1, e.lastFrameAlpha = 0.7, e.dpr = 1, e.virtual = !1, e.config = {}, e.incremental = !1, e.zlevel = 0, e.maxRepaintRectCount = 5, e.__dirty = !0, e.__firstTimePaint = !0, e.__used = !1, e.__drawIndex = 0, e.__startIndex = 0, e.__endIndex = 0, e.__prevStartIndex = null, e.__prevEndIndex = null;
    var t;
    r = r || X, typeof a == "string" ? t = E(a, i, r) : D(a) && (t = a, a = t.id), e.id = a, e.dom = t;
    var o = t.style;
    return o && (L(t), t.onselectstart = function() {
      return !1;
    }, o.padding = "0", o.margin = "0", o.borderWidth = "0"), e.painter = i, e.dpr = r, e;
  }
  return h.prototype.getElementCount = function() {
    return this.__endIndex - this.__startIndex;
  }, h.prototype.afterBrush = function() {
    this.__prevStartIndex = this.__startIndex, this.__prevEndIndex = this.__endIndex;
  }, h.prototype.initContext = function() {
    this.ctx = this.dom.getContext("2d"), this.ctx.dpr = this.dpr;
  }, h.prototype.setUnpainted = function() {
    this.__firstTimePaint = !0;
  }, h.prototype.createBackBuffer = function() {
    var a = this.dpr;
    this.domBack = E("back-" + this.id, this.painter, a), this.ctxBack = this.domBack.getContext("2d"), a !== 1 && this.ctxBack.scale(a, a);
  }, h.prototype.createRepaintRects = function(a, i, r, e) {
    if (this.__firstTimePaint)
      return this.__firstTimePaint = !1, null;
    var t = [], o = this.maxRepaintRectCount, m = !1, _ = new R(0, 0, 0, 0);
    function x(f) {
      if (!(!f.isFinite() || f.isZero()))
        if (t.length === 0) {
          var v = new R(0, 0, 0, 0);
          v.copy(f), t.push(v);
        } else {
          for (var g = !1, w = 1 / 0, P = 0, y = 0; y < t.length; ++y) {
            var B = t[y];
            if (B.intersect(f)) {
              var k = new R(0, 0, 0, 0);
              k.copy(B), k.union(f), t[y] = k, g = !0;
              break;
            } else if (m) {
              _.copy(f), _.union(B);
              var A = f.width * f.height, T = B.width * B.height, z = _.width * _.height, b = z - A - T;
              b < w && (w = b, P = y);
            }
          }
          if (m && (t[P].union(f), g = !0), !g) {
            var v = new R(0, 0, 0, 0);
            v.copy(f), t.push(v);
          }
          m || (m = t.length >= o);
        }
    }
    for (var n = this.__startIndex; n < this.__endIndex; ++n) {
      var s = a[n];
      if (s) {
        var l = s.shouldBePainted(r, e, !0, !0), u = s.__isRendered && (s.__dirty & S || !l) ? s.getPrevPaintRect() : null;
        u && x(u);
        var p = l && (s.__dirty & S || !s.__isRendered) ? s.getPaintRect() : null;
        p && x(p);
      }
    }
    for (var n = this.__prevStartIndex; n < this.__prevEndIndex; ++n) {
      var s = i[n], l = s && s.shouldBePainted(r, e, !0, !0);
      if (s && (!l || !s.__zr) && s.__isRendered) {
        var u = s.getPrevPaintRect();
        u && x(u);
      }
    }
    var c;
    do {
      c = !1;
      for (var n = 0; n < t.length; ) {
        if (t[n].isZero()) {
          t.splice(n, 1);
          continue;
        }
        for (var d = n + 1; d < t.length; )
          t[n].intersect(t[d]) ? (c = !0, t[n].union(t[d]), t.splice(d, 1)) : d++;
        n++;
      }
    } while (c);
    return this._paintRects = t, t;
  }, h.prototype.debugGetPaintRects = function() {
    return (this._paintRects || []).slice();
  }, h.prototype.resize = function(a, i) {
    var r = this.dpr, e = this.dom, t = e.style, o = this.domBack;
    t && (t.width = a + "px", t.height = i + "px"), e.width = a * r, e.height = i * r, o && (o.width = a * r, o.height = i * r, r !== 1 && this.ctxBack.scale(r, r));
  }, h.prototype.clear = function(a, i, r) {
    var e = this.dom, t = this.ctx, o = e.width, m = e.height;
    i = i || this.clearColor;
    var _ = this.motionBlur && !a, x = this.lastFrameAlpha, n = this.dpr, s = this;
    _ && (this.domBack || this.createBackBuffer(), this.ctxBack.globalCompositeOperation = "copy", this.ctxBack.drawImage(e, 0, 0, o / n, m / n));
    var l = this.domBack;
    function u(p, c, d, f) {
      if (t.clearRect(p, c, d, f), i && i !== "transparent") {
        var v = void 0;
        if (G(i)) {
          var g = i.global || i.__width === d && i.__height === f;
          v = g && i.__canvasGradient || Z(t, i, {
            x: 0,
            y: 0,
            width: d,
            height: f
          }), i.__canvasGradient = v, i.__width = d, i.__height = f;
        } else W(i) && (i.scaleX = i.scaleX || n, i.scaleY = i.scaleY || n, v = O(t, i, {
          dirty: function() {
            s.setUnpainted(), s.painter.refresh();
          }
        }));
        t.save(), t.fillStyle = v || i, t.fillRect(p, c, d, f), t.restore();
      }
      _ && (t.save(), t.globalAlpha = x, t.drawImage(l, p, c, d, f), t.restore());
    }
    !r || _ ? u(0, 0, o, m) : r.length && U(r, function(p) {
      u(p.x * n, p.y * n, p.width * n, p.height * n);
    });
  }, h;
})(Y);
export {
  C as default
};
