// zrender@6.0.0/lib/canvas/Painter.js
import { devicePixelRatio as F } from "../config.js";
import { each as k, merge as C, logError as P, indexOf as U, extend as z, disableUserSelect as j } from "../core/util.js";
import D from "./Layer.js";
import q from "../animation/requestAnimationFrame.js";
import Y from "../core/env.js";
import { brush as I, brushSingle as G } from "./graphic.js";
import { REDRAW_BIT as J } from "../graphic/constants.js";
import { getSize as R } from "./helper.js";
var O = 1e5, y = 314159, b = 0.01, K = 1e-3;
function Q(a) {
  return a ? a.__builtin__ ? !0 : !(typeof a.resize != "function" || typeof a.refresh != "function") : !1;
}
function X(a, t) {
  var i = document.createElement("div");
  return i.style.cssText = [
    "position:relative",
    "width:" + a + "px",
    "height:" + t + "px",
    "padding:0",
    "margin:0",
    "border-width:0"
  ].join(";") + ";", i;
}
var ot = (function() {
  function a(t, i, e, r) {
    this.type = "canvas", this._zlevelList = [], this._prevDisplayList = [], this._layers = {}, this._layerConfig = {}, this._needsManuallyCompositing = !1, this.type = "canvas";
    var s = !t.nodeName || t.nodeName.toUpperCase() === "CANVAS";
    this._opts = e = z({}, e || {}), this.dpr = e.devicePixelRatio || F, this._singleCanvas = s, this.root = t;
    var n = t.style;
    n && (j(t), t.innerHTML = ""), this.storage = i;
    var f = this._zlevelList;
    this._prevDisplayList = [];
    var o = this._layers;
    if (s) {
      var h = t, _ = h.width, p = h.height;
      e.width != null && (_ = e.width), e.height != null && (p = e.height), this.dpr = e.devicePixelRatio || 1, h.width = _ * this.dpr, h.height = p * this.dpr, this._width = _, this._height = p;
      var u = new D(h, this, this.dpr);
      u.__builtin__ = !0, u.initContext(), o[y] = u, u.zlevel = y, f.push(y), this._domRoot = t;
    } else {
      this._width = R(t, 0, e), this._height = R(t, 1, e);
      var d = this._domRoot = X(this._width, this._height);
      t.appendChild(d);
    }
  }
  return a.prototype.getType = function() {
    return "canvas";
  }, a.prototype.isSingleCanvas = function() {
    return this._singleCanvas;
  }, a.prototype.getViewportRoot = function() {
    return this._domRoot;
  }, a.prototype.getViewportRootOffset = function() {
    var t = this.getViewportRoot();
    if (t)
      return {
        offsetLeft: t.offsetLeft || 0,
        offsetTop: t.offsetTop || 0
      };
  }, a.prototype.refresh = function(t) {
    var i = this.storage.getDisplayList(!0), e = this._prevDisplayList, r = this._zlevelList;
    this._redrawId = Math.random(), this._paintList(i, e, t, this._redrawId);
    for (var s = 0; s < r.length; s++) {
      var n = r[s], f = this._layers[n];
      if (!f.__builtin__ && f.refresh) {
        var o = s === 0 ? this._backgroundColor : null;
        f.refresh(o);
      }
    }
    return this._opts.useDirtyRect && (this._prevDisplayList = i.slice()), this;
  }, a.prototype.refreshHover = function() {
    this._paintHoverList(this.storage.getDisplayList(!1));
  }, a.prototype._paintHoverList = function(t) {
    var i = t.length, e = this._hoverlayer;
    if (e && e.clear(), !!i) {
      for (var r = {
        inHover: !0,
        viewWidth: this._width,
        viewHeight: this._height
      }, s, n = 0; n < i; n++) {
        var f = t[n];
        f.__inHover && (e || (e = this._hoverlayer = this.getLayer(O)), s || (s = e.ctx, s.save()), I(s, f, r, n === i - 1));
      }
      s && s.restore();
    }
  }, a.prototype.getHoverLayer = function() {
    return this.getLayer(O);
  }, a.prototype.paintOne = function(t, i) {
    G(t, i);
  }, a.prototype._paintList = function(t, i, e, r) {
    if (this._redrawId === r) {
      e = e || !1, this._updateLayerStatus(t);
      var s = this._doPaintList(t, i, e), n = s.finished, f = s.needsRefreshHover;
      if (this._needsManuallyCompositing && this._compositeManually(), f && this._paintHoverList(t), n)
        this.eachLayer(function(d) {
          d.afterBrush && d.afterBrush();
        });
      else {
        var o = this;
        q(function() {
          o._paintList(t, i, e, r);
        });
      }
    }
  }, a.prototype._compositeManually = function() {
    var t = this.getLayer(y).ctx, i = this._domRoot.width, e = this._domRoot.height;
    t.clearRect(0, 0, i, e), this.eachBuiltinLayer(function(r) {
      r.virtual && t.drawImage(r.dom, 0, 0, i, e);
    });
  }, a.prototype._doPaintList = function(t, i, e) {
    for (var r = this, s = [], n = this._opts.useDirtyRect, f = 0; f < this._zlevelList.length; f++) {
      var o = this._zlevelList[f], d = this._layers[o];
      d.__builtin__ && d !== this._hoverlayer && (d.__dirty || e) && s.push(d);
    }
    for (var h = !0, _ = !1, p = function(g) {
      var v = s[g], l = v.ctx, L = n && v.createRepaintRects(t, i, u._width, u._height), m = e ? v.__startIndex : v.__drawIndex, M = !e && v.incremental && Date.now, W = M && Date.now(), N = v.zlevel === u._zlevelList[0] ? u._backgroundColor : null;
      if (v.__startIndex === v.__endIndex)
        v.clear(!1, N, L);
      else if (m === v.__startIndex) {
        var T = t[m];
        (!T.incremental || !T.notClear || e) && v.clear(!1, N, L);
      }
      m === -1 && (console.error("For some unknown reason. drawIndex is -1"), m = v.__startIndex);
      var c, B = function(Z) {
        var S = {
          inHover: !1,
          allClipped: !1,
          prevEl: null,
          viewWidth: r._width,
          viewHeight: r._height
        };
        for (c = m; c < v.__endIndex; c++) {
          var V = t[c];
          if (V.__inHover && (_ = !0), r._doPaintEl(V, v, n, Z, S, c === v.__endIndex - 1), M) {
            var A = Date.now() - W;
            if (A > 15)
              break;
          }
        }
        S.prevElClipPaths && l.restore();
      };
      if (L)
        if (L.length === 0)
          c = v.__endIndex;
        else
          for (var w = u.dpr, H = 0; H < L.length; ++H) {
            var x = L[H];
            l.save(), l.beginPath(), l.rect(x.x * w, x.y * w, x.width * w, x.height * w), l.clip(), B(x), l.restore();
          }
      else
        l.save(), B(), l.restore();
      v.__drawIndex = c, v.__drawIndex < v.__endIndex && (h = !1);
    }, u = this, E = 0; E < s.length; E++)
      p(E);
    return Y.wxa && k(this._layers, function(g) {
      g && g.ctx && g.ctx.draw && g.ctx.draw();
    }), {
      finished: h,
      needsRefreshHover: _
    };
  }, a.prototype._doPaintEl = function(t, i, e, r, s, n) {
    var f = i.ctx;
    if (e) {
      var o = t.getPaintRect();
      (!r || o && o.intersect(r)) && (I(f, t, s, n), t.setPrevPaintRect(o));
    } else
      I(f, t, s, n);
  }, a.prototype.getLayer = function(t, i) {
    this._singleCanvas && !this._needsManuallyCompositing && (t = y);
    var e = this._layers[t];
    return e || (e = new D("zr_" + t, this, this.dpr), e.zlevel = t, e.__builtin__ = !0, this._layerConfig[t] ? C(e, this._layerConfig[t], !0) : this._layerConfig[t - b] && C(e, this._layerConfig[t - b], !0), i && (e.virtual = i), this.insertLayer(t, e), e.initContext()), e;
  }, a.prototype.insertLayer = function(t, i) {
    var e = this._layers, r = this._zlevelList, s = r.length, n = this._domRoot, f = null, o = -1;
    if (e[t]) {
      process.env.NODE_ENV !== "production" && P("ZLevel " + t + " has been used already");
      return;
    }
    if (!Q(i)) {
      process.env.NODE_ENV !== "production" && P("Layer of zlevel " + t + " is not valid");
      return;
    }
    if (s > 0 && t > r[0]) {
      for (o = 0; o < s - 1 && !(r[o] < t && r[o + 1] > t); o++)
        ;
      f = e[r[o]];
    }
    if (r.splice(o + 1, 0, t), e[t] = i, !i.virtual)
      if (f) {
        var d = f.dom;
        d.nextSibling ? n.insertBefore(i.dom, d.nextSibling) : n.appendChild(i.dom);
      } else
        n.firstChild ? n.insertBefore(i.dom, n.firstChild) : n.appendChild(i.dom);
    i.painter || (i.painter = this);
  }, a.prototype.eachLayer = function(t, i) {
    for (var e = this._zlevelList, r = 0; r < e.length; r++) {
      var s = e[r];
      t.call(i, this._layers[s], s);
    }
  }, a.prototype.eachBuiltinLayer = function(t, i) {
    for (var e = this._zlevelList, r = 0; r < e.length; r++) {
      var s = e[r], n = this._layers[s];
      n.__builtin__ && t.call(i, n, s);
    }
  }, a.prototype.eachOtherLayer = function(t, i) {
    for (var e = this._zlevelList, r = 0; r < e.length; r++) {
      var s = e[r], n = this._layers[s];
      n.__builtin__ || t.call(i, n, s);
    }
  }, a.prototype.getLayers = function() {
    return this._layers;
  }, a.prototype._updateLayerStatus = function(t) {
    this.eachBuiltinLayer(function(_, p) {
      _.__dirty = _.__used = !1;
    });
    function i(_) {
      s && (s.__endIndex !== _ && (s.__dirty = !0), s.__endIndex = _);
    }
    if (this._singleCanvas)
      for (var e = 1; e < t.length; e++) {
        var r = t[e];
        if (r.zlevel !== t[e - 1].zlevel || r.incremental) {
          this._needsManuallyCompositing = !0;
          break;
        }
      }
    var s = null, n = 0, f, o;
    for (o = 0; o < t.length; o++) {
      var r = t[o], d = r.zlevel, h = void 0;
      f !== d && (f = d, n = 0), r.incremental ? (h = this.getLayer(d + K, this._needsManuallyCompositing), h.incremental = !0, n = 1) : h = this.getLayer(d + (n > 0 ? b : 0), this._needsManuallyCompositing), h.__builtin__ || P("ZLevel " + d + " has been used by unkown layer " + h.id), h !== s && (h.__used = !0, h.__startIndex !== o && (h.__dirty = !0), h.__startIndex = o, h.incremental ? h.__drawIndex = -1 : h.__drawIndex = o, i(o), s = h), r.__dirty & J && !r.__inHover && (h.__dirty = !0, h.incremental && h.__drawIndex < 0 && (h.__drawIndex = o));
    }
    i(o), this.eachBuiltinLayer(function(_, p) {
      !_.__used && _.getElementCount() > 0 && (_.__dirty = !0, _.__startIndex = _.__endIndex = _.__drawIndex = 0), _.__dirty && _.__drawIndex < 0 && (_.__drawIndex = _.__startIndex);
    });
  }, a.prototype.clear = function() {
    return this.eachBuiltinLayer(this._clearLayer), this;
  }, a.prototype._clearLayer = function(t) {
    t.clear();
  }, a.prototype.setBackgroundColor = function(t) {
    this._backgroundColor = t, k(this._layers, function(i) {
      i.setUnpainted();
    });
  }, a.prototype.configLayer = function(t, i) {
    if (i) {
      var e = this._layerConfig;
      e[t] ? C(e[t], i, !0) : e[t] = i;
      for (var r = 0; r < this._zlevelList.length; r++) {
        var s = this._zlevelList[r];
        if (s === t || s === t + b) {
          var n = this._layers[s];
          C(n, e[t], !0);
        }
      }
    }
  }, a.prototype.delLayer = function(t) {
    var i = this._layers, e = this._zlevelList, r = i[t];
    r && (r.dom.parentNode.removeChild(r.dom), delete i[t], e.splice(U(e, t), 1));
  }, a.prototype.resize = function(t, i) {
    if (this._domRoot.style) {
      var e = this._domRoot;
      e.style.display = "none";
      var r = this._opts, s = this.root;
      if (t != null && (r.width = t), i != null && (r.height = i), t = R(s, 0, r), i = R(s, 1, r), e.style.display = "", this._width !== t || i !== this._height) {
        e.style.width = t + "px", e.style.height = i + "px";
        for (var n in this._layers)
          this._layers.hasOwnProperty(n) && this._layers[n].resize(t, i);
        this.refresh(!0);
      }
      this._width = t, this._height = i;
    } else {
      if (t == null || i == null)
        return;
      this._width = t, this._height = i, this.getLayer(y).resize(t, i);
    }
    return this;
  }, a.prototype.clearLayer = function(t) {
    var i = this._layers[t];
    i && i.clear();
  }, a.prototype.dispose = function() {
    this.root.innerHTML = "", this.root = this.storage = this._domRoot = this._layers = null;
  }, a.prototype.getRenderedCanvas = function(t) {
    if (t = t || {}, this._singleCanvas && !this._compositeManually)
      return this._layers[y].dom;
    var i = new D("image", this, t.pixelRatio || this.dpr);
    i.initContext(), i.clear(!1, t.backgroundColor || this._backgroundColor);
    var e = i.ctx;
    if (t.pixelRatio <= this.dpr) {
      this.refresh();
      var r = i.dom.width, s = i.dom.height;
      this.eachLayer(function(_) {
        _.__builtin__ ? e.drawImage(_.dom, 0, 0, r, s) : _.renderToCanvas && (e.save(), _.renderToCanvas(e), e.restore());
      });
    } else
      for (var n = {
        inHover: !1,
        viewWidth: this._width,
        viewHeight: this._height
      }, f = this.storage.getDisplayList(!0), o = 0, d = f.length; o < d; o++) {
        var h = f[o];
        I(e, h, n, o === d - 1);
      }
    return i.dom;
  }, a.prototype.getWidth = function() {
    return this._width;
  }, a.prototype.getHeight = function() {
    return this._height;
  }, a;
})();
export {
  ot as default
};
