// echarts@6.0.0/lib/model/mixin/dataFormat.js
import { isArray as h, isFunction as D, isString as N, isObject as V } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { retrieveRawValue as T } from "../../data/helper/dataProvider.js";
import { formatTpl as b } from "../../util/format.js";
import { error as d, makePrintable as w } from "../../util/log.js";
var I = /\{@(.+?)\}/g, k = (
  /** @class */
  (function() {
    function a() {
    }
    return a.prototype.getDataParams = function(e, r) {
      var i = this.getData(r), p = this.getRawValue(e, r), o = i.getRawIndex(e), n = i.getName(e), m = i.getRawDataItem(e), t = i.getItemVisual(e, "style"), v = t && t[i.getItemVisual(e, "drawType") || "fill"], f = t && t.stroke, c = this.mainType, l = c === "series", s = i.userOutput && i.userOutput.get();
      return {
        componentType: c,
        componentSubType: this.subType,
        componentIndex: this.componentIndex,
        seriesType: l ? this.subType : null,
        seriesIndex: this.seriesIndex,
        seriesId: l ? this.id : null,
        seriesName: l ? this.name : null,
        name: n,
        dataIndex: o,
        data: m,
        dataType: r,
        value: p,
        color: v,
        borderColor: f,
        dimensionNames: s ? s.fullDimensions : null,
        encode: s ? s.encode : null,
        // Param name list for mapping `a`, `b`, `c`, `d`, `e`
        $vars: ["seriesName", "name", "value"]
      };
    }, a.prototype.getFormattedLabel = function(e, r, i, p, o, n) {
      r = r || "normal";
      var m = this.getData(i), t = this.getDataParams(e, i);
      if (n && (t.value = n.interpolatedValue), p != null && h(t.value) && (t.value = t.value[p]), !o) {
        var v = m.getItemModel(e);
        o = v.get(r === "normal" ? ["label", "formatter"] : [r, "label", "formatter"]);
      }
      if (D(o))
        return t.status = r, t.dimensionIndex = p, o(t);
      if (N(o)) {
        var f = b(o, t);
        return f.replace(I, function(c, l) {
          var s = l.length, u = l;
          u.charAt(0) === "[" && u.charAt(s - 1) === "]" && (u = +u.slice(1, s - 1), process.env.NODE_ENV !== "production" && isNaN(u) && d("Invalide label formatter: @" + l + ", only support @[0], @[1], @[2], ..."));
          var g = T(m, e, u);
          if (n && h(n.interpolatedValue)) {
            var y = m.getDimensionIndex(u);
            y >= 0 && (g = n.interpolatedValue[y]);
          }
          return g != null ? g + "" : "";
        });
      }
    }, a.prototype.getRawValue = function(e, r) {
      return T(this.getData(r), e);
    }, a.prototype.formatTooltip = function(e, r, i) {
    }, a;
  })()
);
function A(a) {
  var e, r;
  return V(a) ? a.type ? r = a : process.env.NODE_ENV !== "production" && console.warn("The return type of `formatTooltip` is not supported: " + w(a)) : e = a, {
    text: e,
    // markers: markers || markersExisting,
    frag: r
  };
}
export {
  k as DataFormatMixin,
  A as normalizeTooltipFormatResult
};
