// echarts@6.0.0/lib/model/mixin/palette.js
import { makeInner as g, normalizeToArray as P } from "../../util/model.js";
var v = g(), d = g(), y = (
  /** @class */
  (function() {
    function t() {
    }
    return t.prototype.getColorFromPalette = function(r, a, e) {
      var l = P(this.get("color", !0)), o = this.get("colorLayer", !0);
      return m(this, v, l, o, r, a, e);
    }, t.prototype.clearColorPalette = function() {
      x(this, v);
    }, t;
  })()
);
function M(t, r, a, e) {
  var l = P(t.get(["aria", "decal", "decals"]));
  return m(t, d, l, null, r, a, e);
}
function h(t, r) {
  for (var a = t.length, e = 0; e < a; e++)
    if (t[e].length > r)
      return t[e];
  return t[a - 1];
}
function m(t, r, a, e, l, o, p) {
  o = o || t;
  var i = r(o), c = i.paletteIdx || 0, u = i.paletteNameMap = i.paletteNameMap || {};
  if (u.hasOwnProperty(l))
    return u[l];
  var n = p == null || !e ? a : h(e, p);
  if (n = n || a, !(!n || !n.length)) {
    var f = n[c];
    return l && (u[l] = f), i.paletteIdx = (c + 1) % n.length, f;
  }
}
function x(t, r) {
  r(t).paletteIdx = 0, r(t).paletteNameMap = {};
}
export {
  y as PaletteMixin,
  M as getDecalFromPalette
};
