// echarts@6.0.0/lib/model/mixin/textStyle.js
import { getFont as r } from "../../label/labelStyle.js";
import a from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Text.js";
var h = ["textStyle", "color"], n = ["fontStyle", "fontWeight", "fontSize", "fontFamily", "padding", "lineHeight", "rich", "width", "height", "overflow"], i = new a(), S = (
  /** @class */
  (function() {
    function e() {
    }
    return e.prototype.getTextColor = function(l) {
      var t = this.ecModel;
      return this.getShallow("color") || (!l && t ? t.get(h) : null);
    }, e.prototype.getFont = function() {
      return r({
        fontStyle: this.getShallow("fontStyle"),
        fontWeight: this.getShallow("fontWeight"),
        fontSize: this.getShallow("fontSize"),
        fontFamily: this.getShallow("fontFamily")
      }, this.ecModel);
    }, e.prototype.getTextRect = function(l) {
      for (var t = {
        text: l,
        verticalAlign: this.getShallow("verticalAlign") || this.getShallow("baseline")
      }, o = 0; o < n.length; o++)
        t[n[o]] = this.getShallow(n[o]);
      return i.useStyle(t), i.update(), i.getBoundingRect();
    }, e;
  })()
);
export {
  S as default
};
