// echarts@6.0.0/lib/coord/calendar/Calendar.js
import { each as y, isArray as d, logError as p } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { getLayoutRect as v } from "../../util/layout.js";
import { parseDate as _ } from "../../util/number.js";
import D from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/BoundingRect.js";
import { expandOrShrinkRect as w } from "../../util/graphic.js";
import { injectCoordSysByOption as I, simpleCoordSysInjectionProvider as k } from "../../core/CoordinateSystem.js";
var u = 864e5, O = (
  /** @class */
  (function() {
    function o(r, t, e) {
      this.type = "calendar", this.dimensions = o.dimensions, this.getDimensionsInfo = o.getDimensionsInfo, this._model = r, this._update(t, e);
    }
    return o.getDimensionsInfo = function() {
      return [{
        name: "time",
        type: "time"
      }, "value"];
    }, o.prototype.getRangeInfo = function() {
      return this._rangeInfo;
    }, o.prototype.getModel = function() {
      return this._model;
    }, o.prototype.getRect = function() {
      return this._rect;
    }, o.prototype.getCellWidth = function() {
      return this._sw;
    }, o.prototype.getCellHeight = function() {
      return this._sh;
    }, o.prototype.getOrient = function() {
      return this._orient;
    }, o.prototype.getFirstDayOfWeek = function() {
      return this._firstDayOfWeek;
    }, o.prototype.getDateInfo = function(r) {
      r = _(r);
      var t = r.getFullYear(), e = r.getMonth() + 1, i = e < 10 ? "0" + e : "" + e, n = r.getDate(), a = n < 10 ? "0" + n : "" + n, s = r.getDay();
      return s = Math.abs((s + 7 - this.getFirstDayOfWeek()) % 7), {
        y: t + "",
        m: i,
        d: a,
        day: s,
        time: r.getTime(),
        formatedDate: t + "-" + i + "-" + a,
        date: r
      };
    }, o.prototype.getNextNDay = function(r, t) {
      return t = t || 0, t === 0 ? this.getDateInfo(r) : (r = new Date(this.getDateInfo(r).time), r.setDate(r.getDate() + t), this.getDateInfo(r));
    }, o.prototype._update = function(r, t) {
      this._firstDayOfWeek = +this._model.getModel("dayLabel").get("firstDay"), this._orient = this._model.get("orient"), this._lineWidth = this._model.getModel("itemStyle").getItemStyle().lineWidth || 0, this._rangeInfo = this._getRangeInfo(this._initRangeOption());
      var e = this._rangeInfo.weeks || 1, i = ["width", "height"], n = this._model.getCellSize().slice(), a = this._model.getBoxLayoutParams(), s = this._orient === "horizontal" ? [e, 7] : [7, e];
      y([0, 1], function(h) {
        l(n, h) && (a[i[h]] = n[h] * s[h]);
      });
      var f = {
        width: t.getWidth(),
        height: t.getHeight()
      }, c = this._rect = v(a, f);
      y([0, 1], function(h) {
        l(n, h) || (n[h] = c[i[h]] / s[h]);
      });
      function l(h, g) {
        return h[g] != null && h[g] !== "auto";
      }
      this._sw = n[0], this._sh = n[1];
    }, o.prototype.dataToPoint = function(r, t, e) {
      e = e || [], d(r) && (r = r[0]), t == null && (t = !0);
      var i = this.getDateInfo(r), n = this._rangeInfo, a = i.formatedDate;
      if (t && !(i.time >= n.start.time && i.time < n.end.time + u))
        return e[0] = e[1] = NaN, e;
      var s = i.day, f = this._getRangeInfo([n.start.time, a]).nthWeek;
      return this._orient === "vertical" ? (e[0] = this._rect.x + s * this._sw + this._sw / 2, e[1] = this._rect.y + f * this._sh + this._sh / 2) : (e[0] = this._rect.x + f * this._sw + this._sw / 2, e[1] = this._rect.y + s * this._sh + this._sh / 2), e;
    }, o.prototype.pointToData = function(r) {
      var t = this.pointToDate(r);
      return t && t.time;
    }, o.prototype.dataToLayout = function(r, t, e) {
      e = e || {};
      var i = e.rect = e.rect || {}, n = e.contentRect = e.contentRect || {}, a = this.dataToPoint(r, t);
      return i.x = a[0] - this._sw / 2, i.y = a[1] - this._sh / 2, i.width = this._sw, i.height = this._sh, D.copy(n, i), w(n, this._lineWidth / 2, !0, !0), e;
    }, o.prototype.dataToCalendarLayout = function(r, t) {
      var e = this.dataToPoint(r, t);
      return {
        center: e,
        tl: [e[0] - this._sw / 2, e[1] - this._sh / 2],
        tr: [e[0] + this._sw / 2, e[1] - this._sh / 2],
        br: [e[0] + this._sw / 2, e[1] + this._sh / 2],
        bl: [e[0] - this._sw / 2, e[1] + this._sh / 2]
      };
    }, o.prototype.pointToDate = function(r) {
      var t = Math.floor((r[0] - this._rect.x) / this._sw) + 1, e = Math.floor((r[1] - this._rect.y) / this._sh) + 1, i = this._rangeInfo.range;
      return this._orient === "vertical" ? this._getDateByWeeksAndDay(e, t - 1, i) : this._getDateByWeeksAndDay(t, e - 1, i);
    }, o.prototype.convertToPixel = function(r, t, e) {
      var i = m(t);
      return i === this ? i.dataToPoint(e) : null;
    }, o.prototype.convertToLayout = function(r, t, e) {
      var i = m(t);
      return i === this ? i.dataToLayout(e) : null;
    }, o.prototype.convertFromPixel = function(r, t, e) {
      var i = m(t);
      return i === this ? i.pointToData(e) : null;
    }, o.prototype.containPoint = function(r) {
      return console.warn("Not implemented."), !1;
    }, o.prototype._initRangeOption = function() {
      var r = this._model.get("range"), t;
      if (d(r) && r.length === 1 && (r = r[0]), d(r))
        t = r;
      else {
        var e = r.toString();
        if (/^\d{4}$/.test(e) && (t = [e + "-01-01", e + "-12-31"]), /^\d{4}[\/|-]\d{1,2}$/.test(e)) {
          var i = this.getDateInfo(e), n = i.date;
          n.setMonth(n.getMonth() + 1);
          var a = this.getNextNDay(n, -1);
          t = [i.formatedDate, a.formatedDate];
        }
        /^\d{4}[\/|-]\d{1,2}[\/|-]\d{1,2}$/.test(e) && (t = [e, e]);
      }
      if (!t)
        return process.env.NODE_ENV !== "production" && p("Invalid date range."), r;
      var s = this._getRangeInfo(t);
      return s.start.time > s.end.time && t.reverse(), t;
    }, o.prototype._getRangeInfo = function(r) {
      var t = [this.getDateInfo(r[0]), this.getDateInfo(r[1])], e;
      t[0].time > t[1].time && (e = !0, t.reverse());
      var i = Math.floor(t[1].time / u) - Math.floor(t[0].time / u) + 1, n = new Date(t[0].time), a = n.getDate(), s = t[1].date.getDate();
      n.setDate(a + i - 1);
      var f = n.getDate();
      if (f !== s)
        for (var c = n.getTime() - t[1].time > 0 ? 1 : -1; (f = n.getDate()) !== s && (n.getTime() - t[1].time) * c > 0; )
          i -= c, n.setDate(f - c);
      var l = Math.floor((i + t[0].day + 6) / 7), h = e ? -l + 1 : l - 1;
      return e && t.reverse(), {
        range: [t[0].formatedDate, t[1].formatedDate],
        start: t[0],
        end: t[1],
        allDay: i,
        weeks: l,
        // From 0.
        nthWeek: h,
        fweek: t[0].day,
        lweek: t[1].day
      };
    }, o.prototype._getDateByWeeksAndDay = function(r, t, e) {
      var i = this._getRangeInfo(e);
      if (r > i.weeks || r === 0 && t < i.fweek || r === i.weeks && t > i.lweek)
        return null;
      var n = (r - 1) * 7 - i.fweek + t, a = new Date(i.start.time);
      return a.setDate(+i.start.d + n), this.getDateInfo(a);
    }, o.create = function(r, t) {
      var e = [];
      return r.eachComponent("calendar", function(i) {
        var n = new o(i, r, t);
        e.push(n), i.coordinateSystem = n;
      }), r.eachComponent(function(i, n) {
        I({
          targetModel: n,
          coordSysType: "calendar",
          coordSysProvider: k
        });
      }), e;
    }, o.dimensions = ["time", "value"], o;
  })()
);
function m(o) {
  var r = o.calendarModel, t = o.seriesModel, e = r ? r.coordinateSystem : t ? t.coordinateSystem : null;
  return e;
}
export {
  O as default
};
