// echarts@6.0.0/lib/coord/calendar/CalendarModel.js
import { __extends as s } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { isArray as p, map as m } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import u from "../../model/Component.js";
import { getLayoutParams as c, sizeCalculable as y, mergeLayoutParam as f } from "../../util/layout.js";
import i from "../../visual/tokens.js";
var b = (
  /** @class */
  (function(t) {
    s(o, t);
    function o() {
      var e = t !== null && t.apply(this, arguments) || this;
      return e.type = o.type, e;
    }
    return o.prototype.init = function(e, r, n) {
      var l = c(e);
      t.prototype.init.apply(this, arguments), a(e, l);
    }, o.prototype.mergeOption = function(e) {
      t.prototype.mergeOption.apply(this, arguments), a(this.option, e);
    }, o.prototype.getCellSize = function() {
      return this.option.cellSize;
    }, o.type = "calendar", o.layoutMode = "box", o.defaultOption = {
      // zlevel: 0,
      // TODO: theoretically, the z of the calendar should be lower
      // than series, but we don't want the series to be displayed
      // on top of the borders like month split line. To align with
      // the effect of previous versions, we set the z to 2 for now
      // until better solution is found.
      z: 2,
      left: 80,
      top: 60,
      cellSize: 20,
      // horizontal vertical
      orient: "horizontal",
      // month separate line style
      splitLine: {
        show: !0,
        lineStyle: {
          color: i.color.axisLine,
          width: 1,
          type: "solid"
        }
      },
      // rect style  temporarily unused emphasis
      itemStyle: {
        color: i.color.neutral00,
        borderWidth: 1,
        borderColor: i.color.neutral10
      },
      // week text style
      dayLabel: {
        show: !0,
        firstDay: 0,
        // start end
        position: "start",
        margin: i.size.s,
        color: i.color.secondary
      },
      // month text style
      monthLabel: {
        show: !0,
        // start end
        position: "start",
        margin: i.size.s,
        // center or left
        align: "center",
        formatter: null,
        color: i.color.secondary
      },
      // year text style
      yearLabel: {
        show: !0,
        // top bottom left right
        position: null,
        margin: i.size.xl,
        formatter: null,
        color: i.color.quaternary,
        fontFamily: "sans-serif",
        fontWeight: "bolder",
        fontSize: 20
      }
    }, o;
  })(u)
);
function a(t, o) {
  var e = t.cellSize, r;
  p(e) ? r = e : r = t.cellSize = [e, e], r.length === 1 && (r[1] = r[0]);
  var n = m([0, 1], function(l) {
    return y(o, l) && (r[l] = "auto"), r[l] != null && r[l] !== "auto";
  });
  f(t, o, {
    type: "box",
    ignoreSize: n
  });
}
export {
  b as default
};
