// @radix-ui/react-dialog@1.1.5/dist/index.js
import * as s from "react";
import { composeEventHandlers as p } from "../../../../../@radix-ui_primitive@1.1.1/node_modules/@radix-ui/primitive/dist/index.js";
import { useComposedRefs as R } from "../../../../../@radix-ui_react-compose-refs@1.1.1_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-compose-refs/dist/index.js";
import { createContextScope as H, createContext as V } from "../../../../../@radix-ui_react-context@1.1.1_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-context/dist/index.js";
import { useId as C } from "../../../../../@radix-ui_react-id@1.1.0_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-id/dist/index.js";
import { useControllableState as q } from "../../../../../@radix-ui_react-use-controllable-state@1.1.0_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-use-controllable-state/dist/index.js";
import { DismissableLayer as K } from "../../../../../@radix-ui_react-dismissable-layer@1.1.4_@types_react-dom@18.3.1_@types_react@18.3.18_react-do_4tz2ji7xudyyxfit74kdiyd35a/node_modules/@radix-ui/react-dismissable-layer/dist/index.js";
import { FocusScope as U } from "../../../../../@radix-ui_react-focus-scope@1.1.1_@types_react-dom@18.3.1_@types_react@18.3.18_react-dom@18.3_5sl5w2z6hfwxfunfgzrt77oufa/node_modules/@radix-ui/react-focus-scope/dist/index.js";
import { Portal as Y } from "../../../../../@radix-ui_react-portal@1.1.3_@types_react-dom@18.3.1_@types_react@18.3.18_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/@radix-ui/react-portal/dist/index.js";
import { Presence as _ } from "../../../../../@radix-ui_react-presence@1.1.2_@types_react-dom@18.3.1_@types_react@18.3.18_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/@radix-ui/react-presence/dist/index.js";
import { Primitive as g } from "../../../../../@radix-ui_react-primitive@2.0.1_@types_react-dom@18.3.1_@types_react@18.3.18_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/@radix-ui/react-primitive/dist/index.js";
import { useFocusGuards as Z } from "../../../../../@radix-ui_react-focus-guards@1.1.1_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-focus-guards/dist/index.js";
import z from "../../../../../react-remove-scroll@2.7.1_@types_react@18.3.18_react@18.3.1/node_modules/react-remove-scroll/dist/es2015/Combination.js";
import { hideOthers as J } from "../../../../../aria-hidden@1.2.6/node_modules/aria-hidden/dist/es2015/index.js";
import { Slot as Q } from "../../../../../@radix-ui_react-slot@1.1.1_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-slot/dist/index.js";
import { jsx as i, jsxs as y, Fragment as I } from "react/jsx-runtime";
var h = "Dialog", [O] = H(h), [X, u] = O(h), x = (e) => {
  const {
    __scopeDialog: t,
    children: n,
    open: a,
    defaultOpen: r,
    onOpenChange: o,
    modal: l = !0
  } = e, c = s.useRef(null), f = s.useRef(null), [D = !1, v] = q({
    prop: a,
    defaultProp: r,
    onChange: o
  });
  return /* @__PURE__ */ i(
    X,
    {
      scope: t,
      triggerRef: c,
      contentRef: f,
      contentId: C(),
      titleId: C(),
      descriptionId: C(),
      open: D,
      onOpenChange: v,
      onOpenToggle: s.useCallback(() => v((j) => !j), [v]),
      modal: l,
      children: n
    }
  );
};
x.displayName = h;
var A = "DialogTrigger", T = s.forwardRef(
  (e, t) => {
    const { __scopeDialog: n, ...a } = e, r = u(A, n), o = R(t, r.triggerRef);
    return /* @__PURE__ */ i(
      g.button,
      {
        type: "button",
        "aria-haspopup": "dialog",
        "aria-expanded": r.open,
        "aria-controls": r.contentId,
        "data-state": P(r.open),
        ...a,
        ref: o,
        onClick: p(e.onClick, r.onOpenToggle)
      }
    );
  }
);
T.displayName = A;
var E = "DialogPortal", [ee, b] = O(E, {
  forceMount: void 0
}), M = (e) => {
  const { __scopeDialog: t, forceMount: n, children: a, container: r } = e, o = u(E, t);
  return /* @__PURE__ */ i(ee, { scope: t, forceMount: n, children: s.Children.map(a, (l) => /* @__PURE__ */ i(_, { present: n || o.open, children: /* @__PURE__ */ i(Y, { asChild: !0, container: r, children: l }) })) });
};
M.displayName = E;
var m = "DialogOverlay", w = s.forwardRef(
  (e, t) => {
    const n = b(m, e.__scopeDialog), { forceMount: a = n.forceMount, ...r } = e, o = u(m, e.__scopeDialog);
    return o.modal ? /* @__PURE__ */ i(_, { present: a || o.open, children: /* @__PURE__ */ i(oe, { ...r, ref: t }) }) : null;
  }
);
w.displayName = m;
var oe = s.forwardRef(
  (e, t) => {
    const { __scopeDialog: n, ...a } = e, r = u(m, n);
    return (
      // Make sure `Content` is scrollable even when it doesn't live inside `RemoveScroll`
      // ie. when `Overlay` and `Content` are siblings
      /* @__PURE__ */ i(z, { as: Q, allowPinchZoom: !0, shards: [r.contentRef], children: /* @__PURE__ */ i(
        g.div,
        {
          "data-state": P(r.open),
          ...a,
          ref: t,
          style: { pointerEvents: "auto", ...a.style }
        }
      ) })
    );
  }
), d = "DialogContent", F = s.forwardRef(
  (e, t) => {
    const n = b(d, e.__scopeDialog), { forceMount: a = n.forceMount, ...r } = e, o = u(d, e.__scopeDialog);
    return /* @__PURE__ */ i(_, { present: a || o.open, children: o.modal ? /* @__PURE__ */ i(te, { ...r, ref: t }) : /* @__PURE__ */ i(re, { ...r, ref: t }) });
  }
);
F.displayName = d;
var te = s.forwardRef(
  (e, t) => {
    const n = u(d, e.__scopeDialog), a = s.useRef(null), r = R(t, n.contentRef, a);
    return s.useEffect(() => {
      const o = a.current;
      if (o) return J(o);
    }, []), /* @__PURE__ */ i(
      S,
      {
        ...e,
        ref: r,
        trapFocus: n.open,
        disableOutsidePointerEvents: !0,
        onCloseAutoFocus: p(e.onCloseAutoFocus, (o) => {
          o.preventDefault(), n.triggerRef.current?.focus();
        }),
        onPointerDownOutside: p(e.onPointerDownOutside, (o) => {
          const l = o.detail.originalEvent, c = l.button === 0 && l.ctrlKey === !0;
          (l.button === 2 || c) && o.preventDefault();
        }),
        onFocusOutside: p(
          e.onFocusOutside,
          (o) => o.preventDefault()
        )
      }
    );
  }
), re = s.forwardRef(
  (e, t) => {
    const n = u(d, e.__scopeDialog), a = s.useRef(!1), r = s.useRef(!1);
    return /* @__PURE__ */ i(
      S,
      {
        ...e,
        ref: t,
        trapFocus: !1,
        disableOutsidePointerEvents: !1,
        onCloseAutoFocus: (o) => {
          e.onCloseAutoFocus?.(o), o.defaultPrevented || (a.current || n.triggerRef.current?.focus(), o.preventDefault()), a.current = !1, r.current = !1;
        },
        onInteractOutside: (o) => {
          e.onInteractOutside?.(o), o.defaultPrevented || (a.current = !0, o.detail.originalEvent.type === "pointerdown" && (r.current = !0));
          const l = o.target;
          n.triggerRef.current?.contains(l) && o.preventDefault(), o.detail.originalEvent.type === "focusin" && r.current && o.preventDefault();
        }
      }
    );
  }
), S = s.forwardRef(
  (e, t) => {
    const { __scopeDialog: n, trapFocus: a, onOpenAutoFocus: r, onCloseAutoFocus: o, ...l } = e, c = u(d, n), f = s.useRef(null), D = R(t, f);
    return Z(), /* @__PURE__ */ y(I, { children: [
      /* @__PURE__ */ i(
        U,
        {
          asChild: !0,
          loop: !0,
          trapped: a,
          onMountAutoFocus: r,
          onUnmountAutoFocus: o,
          children: /* @__PURE__ */ i(
            K,
            {
              role: "dialog",
              id: c.contentId,
              "aria-describedby": c.descriptionId,
              "aria-labelledby": c.titleId,
              "data-state": P(c.open),
              ...l,
              ref: D,
              onDismiss: () => c.onOpenChange(!1)
            }
          )
        }
      ),
      /* @__PURE__ */ y(I, { children: [
        /* @__PURE__ */ i(ae, { titleId: c.titleId }),
        /* @__PURE__ */ i(se, { contentRef: f, descriptionId: c.descriptionId })
      ] })
    ] });
  }
), N = "DialogTitle", W = s.forwardRef(
  (e, t) => {
    const { __scopeDialog: n, ...a } = e, r = u(N, n);
    return /* @__PURE__ */ i(g.h2, { id: r.titleId, ...a, ref: t });
  }
);
W.displayName = N;
var k = "DialogDescription", G = s.forwardRef(
  (e, t) => {
    const { __scopeDialog: n, ...a } = e, r = u(k, n);
    return /* @__PURE__ */ i(g.p, { id: r.descriptionId, ...a, ref: t });
  }
);
G.displayName = k;
var L = "DialogClose", ne = s.forwardRef(
  (e, t) => {
    const { __scopeDialog: n, ...a } = e, r = u(L, n);
    return /* @__PURE__ */ i(
      g.button,
      {
        type: "button",
        ...a,
        ref: t,
        onClick: p(e.onClick, () => r.onOpenChange(!1))
      }
    );
  }
);
ne.displayName = L;
function P(e) {
  return e ? "open" : "closed";
}
var $ = "DialogTitleWarning", [Ne, B] = V($, {
  contentName: d,
  titleName: N,
  docsSlug: "dialog"
}), ae = ({ titleId: e }) => {
  const t = B($), n = `\`${t.contentName}\` requires a \`${t.titleName}\` for the component to be accessible for screen reader users.

If you want to hide the \`${t.titleName}\`, you can wrap it with our VisuallyHidden component.

For more information, see https://radix-ui.com/primitives/docs/components/${t.docsSlug}`;
  return s.useEffect(() => {
    e && (document.getElementById(e) || console.error(n));
  }, [n, e]), null;
}, ie = "DialogDescriptionWarning", se = ({ contentRef: e, descriptionId: t }) => {
  const a = `Warning: Missing \`Description\` or \`aria-describedby={undefined}\` for {${B(ie).contentName}}.`;
  return s.useEffect(() => {
    const r = e.current?.getAttribute("aria-describedby");
    t && r && (document.getElementById(t) || console.warn(a));
  }, [a, e, t]), null;
}, Pe = x, ye = T, Ie = M, Oe = w, xe = F, Ae = W, Te = G;
export {
  xe as Content,
  Te as Description,
  x as Dialog,
  ne as DialogClose,
  F as DialogContent,
  G as DialogDescription,
  w as DialogOverlay,
  M as DialogPortal,
  W as DialogTitle,
  T as DialogTrigger,
  Oe as Overlay,
  Ie as Portal,
  Pe as Root,
  Ae as Title,
  ye as Trigger,
  Ne as WarningProvider
};
