// zrender@6.0.0/lib/Element.js
import W, { TRANSFORMABLE_PROPS as Y } from "./core/Transformable.js";
import X, { cloneValue as j } from "./animation/Animator.js";
import B from "./core/BoundingRect.js";
import K from "./core/Eventful.js";
import { calculateTextPosition as k, parsePercent as M } from "./contain/text.js";
import { extend as x, isObject as F, keys as q, indexOf as R, logError as V, reduce as tt, guid as it, isArrayLike as y, isGradientObject as rt, filter as U, isTypedArray as ot, mixin as J } from "./core/util.js";
import { LIGHT_LABEL_COLOR as et, DARK_LABEL_COLOR as nt } from "./config.js";
import { parse as st, stringify as at } from "./tool/color.js";
import { REDRAW_BIT as A } from "./graphic/constants.js";
import { invert as ht } from "./core/matrix.js";
var z = "__zr_normal__", D = Y.concat(["ignore"]), ft = tt(Y, function(o, t) {
  return o[t] = !0, o;
}, { ignore: !1 }), P = {}, ut = new B(0, 0, 0, 0), L = [], Q = (function() {
  function o(t) {
    this.id = it(), this.animators = [], this.currentStates = [], this.states = {}, this._init(t);
  }
  return o.prototype._init = function(t) {
    this.attr(t);
  }, o.prototype.drift = function(t, i, r) {
    switch (this.draggable) {
      case "horizontal":
        i = 0;
        break;
      case "vertical":
        t = 0;
        break;
    }
    var n = this.transform;
    n || (n = this.transform = [1, 0, 0, 1, 0, 0]), n[4] += t, n[5] += i, this.decomposeTransform(), this.markRedraw();
  }, o.prototype.beforeUpdate = function() {
  }, o.prototype.afterUpdate = function() {
  }, o.prototype.update = function() {
    this.updateTransform(), this.__dirty && this.updateInnerText();
  }, o.prototype.updateInnerText = function(t) {
    var i = this._textContent;
    if (i && (!i.ignore || t)) {
      this.textConfig || (this.textConfig = {});
      var r = this.textConfig, n = r.local, e = i.innerTransformable, s = void 0, h = void 0, a = !1;
      e.parent = n ? this : null;
      var v = !1;
      e.copyTransform(i);
      var f = r.position != null, _ = r.autoOverflowArea, u = void 0;
      if ((_ || f) && (u = ut, r.layoutRect ? u.copy(r.layoutRect) : u.copy(this.getBoundingRect()), n || u.applyTransform(this.transform)), f) {
        this.calculateTextPosition ? this.calculateTextPosition(P, r, u) : k(P, r, u), e.x = P.x, e.y = P.y, s = P.align, h = P.verticalAlign;
        var g = r.origin;
        if (g && r.rotation != null) {
          var m = void 0, c = void 0;
          g === "center" ? (m = u.width * 0.5, c = u.height * 0.5) : (m = M(g[0], u.width), c = M(g[1], u.height)), v = !0, e.originX = -e.x + m + (n ? 0 : u.x), e.originY = -e.y + c + (n ? 0 : u.y);
        }
      }
      r.rotation != null && (e.rotation = r.rotation);
      var T = r.offset;
      T && (e.x += T[0], e.y += T[1], v || (e.originX = -T[0], e.originY = -T[1]));
      var p = this._innerTextDefaultStyle || (this._innerTextDefaultStyle = {});
      if (_) {
        var S = p.overflowRect = p.overflowRect || new B(0, 0, 0, 0);
        e.getLocalTransform(L), ht(L, L), B.copy(S, u), S.applyTransform(L);
      } else
        p.overflowRect = null;
      var C = r.inside == null ? typeof r.position == "string" && r.position.indexOf("inside") >= 0 : r.inside, d = void 0, l = void 0, w = void 0;
      C && this.canBeInsideText() ? (d = r.insideFill, l = r.insideStroke, (d == null || d === "auto") && (d = this.getInsideTextFill()), (l == null || l === "auto") && (l = this.getInsideTextStroke(d), w = !0)) : (d = r.outsideFill, l = r.outsideStroke, (d == null || d === "auto") && (d = this.getOutsideFill()), (l == null || l === "auto") && (l = this.getOutsideStroke(d), w = !0)), d = d || "#000", (d !== p.fill || l !== p.stroke || w !== p.autoStroke || s !== p.align || h !== p.verticalAlign) && (a = !0, p.fill = d, p.stroke = l, p.autoStroke = w, p.align = s, p.verticalAlign = h, i.setDefaultTextStyle(p)), i.__dirty |= A, a && i.dirtyStyle(!0);
    }
  }, o.prototype.canBeInsideText = function() {
    return !0;
  }, o.prototype.getInsideTextFill = function() {
    return "#fff";
  }, o.prototype.getInsideTextStroke = function(t) {
    return "#000";
  }, o.prototype.getOutsideFill = function() {
    return this.__zr && this.__zr.isDarkMode() ? et : nt;
  }, o.prototype.getOutsideStroke = function(t) {
    var i = this.__zr && this.__zr.getBackgroundColor(), r = typeof i == "string" && st(i);
    r || (r = [255, 255, 255, 1]);
    for (var n = r[3], e = this.__zr.isDarkMode(), s = 0; s < 3; s++)
      r[s] = r[s] * n + (e ? 0 : 255) * (1 - n);
    return r[3] = 1, at(r, "rgba");
  }, o.prototype.traverse = function(t, i) {
  }, o.prototype.attrKV = function(t, i) {
    t === "textConfig" ? this.setTextConfig(i) : t === "textContent" ? this.setTextContent(i) : t === "clipPath" ? this.setClipPath(i) : t === "extra" ? (this.extra = this.extra || {}, x(this.extra, i)) : this[t] = i;
  }, o.prototype.hide = function() {
    this.ignore = !0, this.markRedraw();
  }, o.prototype.show = function() {
    this.ignore = !1, this.markRedraw();
  }, o.prototype.attr = function(t, i) {
    if (typeof t == "string")
      this.attrKV(t, i);
    else if (F(t))
      for (var r = t, n = q(r), e = 0; e < n.length; e++) {
        var s = n[e];
        this.attrKV(s, t[s]);
      }
    return this.markRedraw(), this;
  }, o.prototype.saveCurrentToNormalState = function(t) {
    this._innerSaveToNormal(t);
    for (var i = this._normalState, r = 0; r < this.animators.length; r++) {
      var n = this.animators[r], e = n.__fromStateTransition;
      if (!(n.getLoop() || e && e !== z)) {
        var s = n.targetName, h = s ? i[s] : i;
        n.saveTo(h);
      }
    }
  }, o.prototype._innerSaveToNormal = function(t) {
    var i = this._normalState;
    i || (i = this._normalState = {}), t.textConfig && !i.textConfig && (i.textConfig = this.textConfig), this._savePrimaryToNormal(t, i, D);
  }, o.prototype._savePrimaryToNormal = function(t, i, r) {
    for (var n = 0; n < r.length; n++) {
      var e = r[n];
      t[e] != null && !(e in i) && (i[e] = this[e]);
    }
  }, o.prototype.hasState = function() {
    return this.currentStates.length > 0;
  }, o.prototype.getState = function(t) {
    return this.states[t];
  }, o.prototype.ensureState = function(t) {
    var i = this.states;
    return i[t] || (i[t] = {}), i[t];
  }, o.prototype.clearStates = function(t) {
    this.useState(z, !1, t);
  }, o.prototype.useState = function(t, i, r, n) {
    var e = t === z, s = this.hasState();
    if (!(!s && e)) {
      var h = this.currentStates, a = this.stateTransition;
      if (!(R(h, t) >= 0 && (i || h.length === 1))) {
        var v;
        if (this.stateProxy && !e && (v = this.stateProxy(t)), v || (v = this.states && this.states[t]), !v && !e) {
          V("State " + t + " not exists.");
          return;
        }
        e || this.saveCurrentToNormalState(v);
        var f = !!(v && v.hoverLayer || n);
        f && this._toggleHoverLayerFlag(!0), this._applyStateObj(t, v, this._normalState, i, !r && !this.__inHover && a && a.duration > 0, a);
        var _ = this._textContent, u = this._textGuide;
        return _ && _.useState(t, i, r, f), u && u.useState(t, i, r, f), e ? (this.currentStates = [], this._normalState = {}) : i ? this.currentStates.push(t) : this.currentStates = [t], this._updateAnimationTargets(), this.markRedraw(), !f && this.__inHover && (this._toggleHoverLayerFlag(!1), this.__dirty &= ~A), v;
      }
    }
  }, o.prototype.useStates = function(t, i, r) {
    if (!t.length)
      this.clearStates();
    else {
      var n = [], e = this.currentStates, s = t.length, h = s === e.length;
      if (h) {
        for (var a = 0; a < s; a++)
          if (t[a] !== e[a]) {
            h = !1;
            break;
          }
      }
      if (h)
        return;
      for (var a = 0; a < s; a++) {
        var v = t[a], f = void 0;
        this.stateProxy && (f = this.stateProxy(v, t)), f || (f = this.states[v]), f && n.push(f);
      }
      var _ = n[s - 1], u = !!(_ && _.hoverLayer || r);
      u && this._toggleHoverLayerFlag(!0);
      var g = this._mergeStates(n), m = this.stateTransition;
      this.saveCurrentToNormalState(g), this._applyStateObj(t.join(","), g, this._normalState, !1, !i && !this.__inHover && m && m.duration > 0, m);
      var c = this._textContent, T = this._textGuide;
      c && c.useStates(t, i, u), T && T.useStates(t, i, u), this._updateAnimationTargets(), this.currentStates = t.slice(), this.markRedraw(), !u && this.__inHover && (this._toggleHoverLayerFlag(!1), this.__dirty &= ~A);
    }
  }, o.prototype.isSilent = function() {
    for (var t = this; t; ) {
      if (t.silent)
        return !0;
      var i = t.__hostTarget;
      t = i ? t.ignoreHostSilent ? null : i : t.parent;
    }
    return !1;
  }, o.prototype._updateAnimationTargets = function() {
    for (var t = 0; t < this.animators.length; t++) {
      var i = this.animators[t];
      i.targetName && i.changeTarget(this[i.targetName]);
    }
  }, o.prototype.removeState = function(t) {
    var i = R(this.currentStates, t);
    if (i >= 0) {
      var r = this.currentStates.slice();
      r.splice(i, 1), this.useStates(r);
    }
  }, o.prototype.replaceState = function(t, i, r) {
    var n = this.currentStates.slice(), e = R(n, t), s = R(n, i) >= 0;
    e >= 0 ? s ? n.splice(e, 1) : n[e] = i : r && !s && n.push(i), this.useStates(n);
  }, o.prototype.toggleState = function(t, i) {
    i ? this.useState(t, !0) : this.removeState(t);
  }, o.prototype._mergeStates = function(t) {
    for (var i = {}, r, n = 0; n < t.length; n++) {
      var e = t[n];
      x(i, e), e.textConfig && (r = r || {}, x(r, e.textConfig));
    }
    return r && (i.textConfig = r), i;
  }, o.prototype._applyStateObj = function(t, i, r, n, e, s) {
    var h = !(i && n);
    i && i.textConfig ? (this.textConfig = x({}, n ? this.textConfig : r.textConfig), x(this.textConfig, i.textConfig)) : h && r.textConfig && (this.textConfig = r.textConfig);
    for (var a = {}, v = !1, f = 0; f < D.length; f++) {
      var _ = D[f], u = e && ft[_];
      i && i[_] != null ? u ? (v = !0, a[_] = i[_]) : this[_] = i[_] : h && r[_] != null && (u ? (v = !0, a[_] = r[_]) : this[_] = r[_]);
    }
    if (!e)
      for (var f = 0; f < this.animators.length; f++) {
        var g = this.animators[f], m = g.targetName;
        g.getLoop() || g.__changeFinalValue(m ? (i || r)[m] : i || r);
      }
    v && this._transitionState(t, a, s);
  }, o.prototype._attachComponent = function(t) {
    if (t.__zr && !t.__hostTarget) {
      if (process.env.NODE_ENV !== "production")
        throw new Error("Text element has been added to zrender.");
      return;
    }
    if (t === this) {
      if (process.env.NODE_ENV !== "production")
        throw new Error("Recursive component attachment.");
      return;
    }
    var i = this.__zr;
    i && t.addSelfToZr(i), t.__zr = i, t.__hostTarget = this;
  }, o.prototype._detachComponent = function(t) {
    t.__zr && t.removeSelfFromZr(t.__zr), t.__zr = null, t.__hostTarget = null;
  }, o.prototype.getClipPath = function() {
    return this._clipPath;
  }, o.prototype.setClipPath = function(t) {
    this._clipPath && this._clipPath !== t && this.removeClipPath(), this._attachComponent(t), this._clipPath = t, this.markRedraw();
  }, o.prototype.removeClipPath = function() {
    var t = this._clipPath;
    t && (this._detachComponent(t), this._clipPath = null, this.markRedraw());
  }, o.prototype.getTextContent = function() {
    return this._textContent;
  }, o.prototype.setTextContent = function(t) {
    var i = this._textContent;
    if (i !== t) {
      if (i && i !== t && this.removeTextContent(), process.env.NODE_ENV !== "production" && t.__zr && !t.__hostTarget)
        throw new Error("Text element has been added to zrender.");
      t.innerTransformable = new W(), this._attachComponent(t), this._textContent = t, this.markRedraw();
    }
  }, o.prototype.setTextConfig = function(t) {
    this.textConfig || (this.textConfig = {}), x(this.textConfig, t), this.markRedraw();
  }, o.prototype.removeTextConfig = function() {
    this.textConfig = null, this.markRedraw();
  }, o.prototype.removeTextContent = function() {
    var t = this._textContent;
    t && (t.innerTransformable = null, this._detachComponent(t), this._textContent = null, this._innerTextDefaultStyle = null, this.markRedraw());
  }, o.prototype.getTextGuideLine = function() {
    return this._textGuide;
  }, o.prototype.setTextGuideLine = function(t) {
    this._textGuide && this._textGuide !== t && this.removeTextGuideLine(), this._attachComponent(t), this._textGuide = t, this.markRedraw();
  }, o.prototype.removeTextGuideLine = function() {
    var t = this._textGuide;
    t && (this._detachComponent(t), this._textGuide = null, this.markRedraw());
  }, o.prototype.markRedraw = function() {
    this.__dirty |= A;
    var t = this.__zr;
    t && (this.__inHover ? t.refreshHover() : t.refresh()), this.__hostTarget && this.__hostTarget.markRedraw();
  }, o.prototype.dirty = function() {
    this.markRedraw();
  }, o.prototype._toggleHoverLayerFlag = function(t) {
    this.__inHover = t;
    var i = this._textContent, r = this._textGuide;
    i && (i.__inHover = t), r && (r.__inHover = t);
  }, o.prototype.addSelfToZr = function(t) {
    if (this.__zr !== t) {
      this.__zr = t;
      var i = this.animators;
      if (i)
        for (var r = 0; r < i.length; r++)
          t.animation.addAnimator(i[r]);
      this._clipPath && this._clipPath.addSelfToZr(t), this._textContent && this._textContent.addSelfToZr(t), this._textGuide && this._textGuide.addSelfToZr(t);
    }
  }, o.prototype.removeSelfFromZr = function(t) {
    if (this.__zr) {
      this.__zr = null;
      var i = this.animators;
      if (i)
        for (var r = 0; r < i.length; r++)
          t.animation.removeAnimator(i[r]);
      this._clipPath && this._clipPath.removeSelfFromZr(t), this._textContent && this._textContent.removeSelfFromZr(t), this._textGuide && this._textGuide.removeSelfFromZr(t);
    }
  }, o.prototype.animate = function(t, i, r) {
    var n = t ? this[t] : this;
    if (process.env.NODE_ENV !== "production" && !n) {
      V('Property "' + t + '" is not existed in element ' + this.id);
      return;
    }
    var e = new X(n, i, r);
    return t && (e.targetName = t), this.addAnimator(e, t), e;
  }, o.prototype.addAnimator = function(t, i) {
    var r = this.__zr, n = this;
    t.during(function() {
      n.updateDuringAnimation(i);
    }).done(function() {
      var e = n.animators, s = R(e, t);
      s >= 0 && e.splice(s, 1);
    }), this.animators.push(t), r && r.animation.addAnimator(t), r && r.wakeUp();
  }, o.prototype.updateDuringAnimation = function(t) {
    this.markRedraw();
  }, o.prototype.stopAnimation = function(t, i) {
    for (var r = this.animators, n = r.length, e = [], s = 0; s < n; s++) {
      var h = r[s];
      !t || t === h.scope ? h.stop(i) : e.push(h);
    }
    return this.animators = e, this;
  }, o.prototype.animateTo = function(t, i, r) {
    H(this, t, i, r);
  }, o.prototype.animateFrom = function(t, i, r) {
    H(this, t, i, r, !0);
  }, o.prototype._transitionState = function(t, i, r, n) {
    for (var e = H(this, i, r, n), s = 0; s < e.length; s++)
      e[s].__fromStateTransition = t;
  }, o.prototype.getBoundingRect = function() {
    return null;
  }, o.prototype.getPaintRect = function() {
    return null;
  }, o.initDefaultProps = (function() {
    var t = o.prototype;
    t.type = "element", t.name = "", t.ignore = t.silent = t.ignoreHostSilent = t.isGroup = t.draggable = t.dragging = t.ignoreClip = t.__inHover = !1, t.__dirty = A;
    var i = {};
    function r(e, s, h) {
      i[e + s + h] || (console.warn("DEPRECATED: '" + e + "' has been deprecated. use '" + s + "', '" + h + "' instead"), i[e + s + h] = !0);
    }
    function n(e, s, h, a) {
      Object.defineProperty(t, e, {
        get: function() {
          if (process.env.NODE_ENV !== "production" && r(e, h, a), !this[s]) {
            var f = this[s] = [];
            v(this, f);
          }
          return this[s];
        },
        set: function(f) {
          process.env.NODE_ENV !== "production" && r(e, h, a), this[h] = f[0], this[a] = f[1], this[s] = f, v(this, f);
        }
      });
      function v(f, _) {
        Object.defineProperty(_, 0, {
          get: function() {
            return f[h];
          },
          set: function(u) {
            f[h] = u;
          }
        }), Object.defineProperty(_, 1, {
          get: function() {
            return f[a];
          },
          set: function(u) {
            f[a] = u;
          }
        });
      }
    }
    Object.defineProperty && (n("position", "_legacyPos", "x", "y"), n("scale", "_legacyScale", "scaleX", "scaleY"), n("origin", "_legacyOrigin", "originX", "originY"));
  })(), o;
})();
J(Q, K);
J(Q, W);
function H(o, t, i, r, n) {
  i = i || {};
  var e = [];
  $(o, "", o, t, i, r, e, n);
  var s = e.length, h = !1, a = i.done, v = i.aborted, f = function() {
    h = !0, s--, s <= 0 && (h ? a && a() : v && v());
  }, _ = function() {
    s--, s <= 0 && (h ? a && a() : v && v());
  };
  s || a && a(), e.length > 0 && i.during && e[0].during(function(m, c) {
    i.during(c);
  });
  for (var u = 0; u < e.length; u++) {
    var g = e[u];
    f && g.done(f), _ && g.aborted(_), i.force && g.duration(i.duration), g.start(i.easing);
  }
  return e;
}
function I(o, t, i) {
  for (var r = 0; r < i; r++)
    o[r] = t[r];
}
function pt(o) {
  return y(o[0]);
}
function vt(o, t, i) {
  if (y(t[i]))
    if (y(o[i]) || (o[i] = []), ot(t[i])) {
      var r = t[i].length;
      o[i].length !== r && (o[i] = new t[i].constructor(r), I(o[i], t[i], r));
    } else {
      var n = t[i], e = o[i], s = n.length;
      if (pt(n))
        for (var h = n[0].length, a = 0; a < s; a++)
          e[a] ? I(e[a], n[a], h) : e[a] = Array.prototype.slice.call(n[a]);
      else
        I(e, n, s);
      e.length = n.length;
    }
  else
    o[i] = t[i];
}
function _t(o, t) {
  return o === t || y(o) && y(t) && dt(o, t);
}
function dt(o, t) {
  var i = o.length;
  if (i !== t.length)
    return !1;
  for (var r = 0; r < i; r++)
    if (o[r] !== t[r])
      return !1;
  return !0;
}
function $(o, t, i, r, n, e, s, h) {
  for (var a = q(r), v = n.duration, f = n.delay, _ = n.additive, u = n.setToFinal, g = !F(e), m = o.animators, c = [], T = 0; T < a.length; T++) {
    var p = a[T], S = r[p];
    if (S != null && i[p] != null && (g || e[p]))
      if (F(S) && !y(S) && !rt(S)) {
        if (t) {
          h || (i[p] = S, o.updateDuringAnimation(t));
          continue;
        }
        $(o, p, i[p], S, n, e && e[p], s, h);
      } else
        c.push(p);
    else h || (i[p] = S, o.updateDuringAnimation(t), c.push(p));
  }
  var C = c.length;
  if (!_ && C)
    for (var d = 0; d < m.length; d++) {
      var l = m[d];
      if (l.targetName === t) {
        var w = l.stopTracks(c);
        if (w) {
          var O = R(m, l);
          m.splice(O, 1);
        }
      }
    }
  if (n.force || (c = U(c, function(Z) {
    return !_t(r[Z], i[Z]);
  }), C = c.length), C > 0 || n.force && !s.length) {
    var E = void 0, N = void 0, G = void 0;
    if (h) {
      N = {}, u && (E = {});
      for (var d = 0; d < C; d++) {
        var p = c[d];
        N[p] = i[p], u ? E[p] = r[p] : i[p] = r[p];
      }
    } else if (u) {
      G = {};
      for (var d = 0; d < C; d++) {
        var p = c[d];
        G[p] = j(i[p]), vt(i, r, p);
      }
    }
    var l = new X(i, !1, !1, _ ? U(m, function(b) {
      return b.targetName === t;
    }) : null);
    l.targetName = t, n.scope && (l.scope = n.scope), u && E && l.whenWithKeys(0, E, c), G && l.whenWithKeys(0, G, c), l.whenWithKeys(v ?? 500, h ? N : r, c).delay(f || 0), o.addAnimator(l, t), s.push(l);
  }
}
export {
  z as PRESERVED_NORMAL_STATE,
  Q as default
};
