// zrender@6.0.0/lib/Handler.js
import { __extends as y } from "../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { each as _ } from "./core/util.js";
import { dist as M } from "./core/vector.js";
import k from "./mixin/Draggable.js";
import w from "./core/Eventful.js";
import { stop as B } from "./core/event.js";
import { GestureMgr as X } from "./core/GestureMgr.js";
import T from "./core/BoundingRect.js";
var E = "silent";
function Y(a, o, t) {
  return {
    type: a,
    event: t,
    target: o.target,
    topTarget: o.topTarget,
    cancelBubble: !1,
    offsetX: t.zrX,
    offsetY: t.zrY,
    gestureEvent: t.gestureEvent,
    pinchX: t.pinchX,
    pinchY: t.pinchY,
    pinchScale: t.pinchScale,
    wheelDelta: t.zrDelta,
    zrByTouch: t.zrByTouch,
    which: t.which,
    stop: O
  };
}
function O() {
  B(this.event);
}
var R = (function(a) {
  y(o, a);
  function o() {
    var t = a !== null && a.apply(this, arguments) || this;
    return t.handler = null, t;
  }
  return o.prototype.dispose = function() {
  }, o.prototype.setCursor = function() {
  }, o;
})(w), f = /* @__PURE__ */ (function() {
  function a(o, t) {
    this.x = o, this.y = t;
  }
  return a;
})(), D = [
  "click",
  "dblclick",
  "mousewheel",
  "mouseout",
  "mouseup",
  "mousedown",
  "mousemove",
  "contextmenu"
], d = new T(0, 0, 0, 0), L = (function(a) {
  y(o, a);
  function o(t, r, s, n, e) {
    var i = a.call(this) || this;
    return i._hovered = new f(0, 0), i.storage = t, i.painter = r, i.painterRoot = n, i._pointerSize = e, s = s || new R(), i.proxy = null, i.setHandlerProxy(s), i._draggingMgr = new k(i), i;
  }
  return o.prototype.setHandlerProxy = function(t) {
    this.proxy && this.proxy.dispose(), t && (_(D, function(r) {
      t.on && t.on(r, this[r], this);
    }, this), t.handler = this), this.proxy = t;
  }, o.prototype.mousemove = function(t) {
    var r = t.zrX, s = t.zrY, n = z(this, r, s), e = this._hovered, i = e.target;
    i && !i.__zr && (e = this.findHover(e.x, e.y), i = e.target);
    var u = this._hovered = n ? new f(r, s) : this.findHover(r, s), h = u.target, c = this.proxy;
    c.setCursor && c.setCursor(h ? h.cursor : "default"), i && h !== i && this.dispatchToElement(e, "mouseout", t), this.dispatchToElement(u, "mousemove", t), h && h !== i && this.dispatchToElement(u, "mouseover", t);
  }, o.prototype.mouseout = function(t) {
    var r = t.zrEventControl;
    r !== "only_globalout" && this.dispatchToElement(this._hovered, "mouseout", t), r !== "no_globalout" && this.trigger("globalout", { type: "globalout", event: t });
  }, o.prototype.resize = function() {
    this._hovered = new f(0, 0);
  }, o.prototype.dispatch = function(t, r) {
    var s = this[t];
    s && s.call(this, r);
  }, o.prototype.dispose = function() {
    this.proxy.dispose(), this.storage = null, this.proxy = null, this.painter = null;
  }, o.prototype.setCursorStyle = function(t) {
    var r = this.proxy;
    r.setCursor && r.setCursor(t);
  }, o.prototype.dispatchToElement = function(t, r, s) {
    t = t || {};
    var n = t.target;
    if (!(n && n.silent)) {
      for (var e = "on" + r, i = Y(r, t, s); n && (n[e] && (i.cancelBubble = !!n[e].call(n, i)), n.trigger(r, i), n = n.__hostTarget ? n.__hostTarget : n.parent, !i.cancelBubble); )
        ;
      i.cancelBubble || (this.trigger(r, i), this.painter && this.painter.eachOtherLayer && this.painter.eachOtherLayer(function(u) {
        typeof u[e] == "function" && u[e].call(u, i), u.trigger && u.trigger(r, i);
      }));
    }
  }, o.prototype.findHover = function(t, r, s) {
    var n = this.storage.getDisplayList(), e = new f(t, r);
    if (m(n, e, t, r, s), this._pointerSize && !e.target) {
      for (var i = [], u = this._pointerSize, h = u / 2, c = new T(t - h, r - h, u, u), v = n.length - 1; v >= 0; v--) {
        var p = n[v];
        p !== s && !p.ignore && !p.ignoreCoarsePointer && (!p.parent || !p.parent.ignoreCoarsePointer) && (d.copy(p.getBoundingRect()), p.transform && d.applyTransform(p.transform), d.intersect(c) && i.push(p));
      }
      if (i.length)
        for (var H = 4, b = Math.PI / 12, P = Math.PI * 2, l = 0; l < h; l += H)
          for (var g = 0; g < P; g += b) {
            var C = t + l * Math.cos(g), x = r + l * Math.sin(g);
            if (m(i, e, C, x, s), e.target)
              return e;
          }
    }
    return e;
  }, o.prototype.processGesture = function(t, r) {
    this._gestureMgr || (this._gestureMgr = new X());
    var s = this._gestureMgr;
    r === "start" && s.clear();
    var n = s.recognize(t, this.findHover(t.zrX, t.zrY, null).target, this.proxy.dom);
    if (r === "end" && s.clear(), n) {
      var e = n.type;
      t.gestureEvent = e;
      var i = new f();
      i.target = n.target, this.dispatchToElement(i, e, n.event);
    }
  }, o;
})(w);
_(["click", "mousedown", "mouseup", "mousewheel", "dblclick", "contextmenu"], function(a) {
  L.prototype[a] = function(o) {
    var t = o.zrX, r = o.zrY, s = z(this, t, r), n, e;
    if ((a !== "mouseup" || !s) && (n = this.findHover(t, r), e = n.target), a === "mousedown")
      this._downEl = e, this._downPoint = [o.zrX, o.zrY], this._upEl = e;
    else if (a === "mouseup")
      this._upEl = e;
    else if (a === "click") {
      if (this._downEl !== this._upEl || !this._downPoint || M(this._downPoint, [o.zrX, o.zrY]) > 4)
        return;
      this._downPoint = null;
    }
    this.dispatchToElement(n, a, o);
  };
});
function S(a, o, t) {
  if (a[a.rectHover ? "rectContain" : "contain"](o, t)) {
    for (var r = a, s = void 0, n = !1; r; ) {
      if (r.ignoreClip && (n = !0), !n) {
        var e = r.getClipPath();
        if (e && !e.contain(o, t))
          return !1;
      }
      r.silent && (s = !0);
      var i = r.__hostTarget;
      r = i ? r.ignoreHostSilent ? null : i : r.parent;
    }
    return s ? E : !0;
  }
  return !1;
}
function m(a, o, t, r, s) {
  for (var n = a.length - 1; n >= 0; n--) {
    var e = a[n], i = void 0;
    if (e !== s && !e.ignore && (i = S(e, t, r)) && (!o.topTarget && (o.topTarget = e), i !== E)) {
      o.target = e;
      break;
    }
  }
}
function z(a, o, t) {
  var r = a.painter;
  return o < 0 || o > r.getWidth() || t < 0 || t > r.getHeight();
}
export {
  L as default
};
