// zrender@6.0.0/lib/Storage.js
import { indexOf as A } from "./core/util.js";
import D from "./core/timsort.js";
import { REDRAW_BIT as m } from "./graphic/constants.js";
var g = !1;
function h() {
  g || (g = !0, console.warn("z / z2 / zlevel of displayable is invalid, which may cause unexpected errors"));
}
function c(r, t) {
  return r.zlevel === t.zlevel ? r.z === t.z ? r.z2 - t.z2 : r.z - t.z : r.zlevel - t.zlevel;
}
var N = (function() {
  function r() {
    this._roots = [], this._displayList = [], this._displayListLen = 0, this.displayableSortFunc = c;
  }
  return r.prototype.traverse = function(t, i) {
    for (var s = 0; s < this._roots.length; s++)
      this._roots[s].traverse(t, i);
  }, r.prototype.getDisplayList = function(t, i) {
    i = i || !1;
    var s = this._displayList;
    return (t || !s.length) && this.updateDisplayList(i), s;
  }, r.prototype.updateDisplayList = function(t) {
    this._displayListLen = 0;
    for (var i = this._roots, s = this._displayList, o = 0, l = i.length; o < l; o++)
      this._updateAndAddDisplayable(i[o], null, t);
    s.length = this._displayListLen, D(s, c);
  }, r.prototype._updateAndAddDisplayable = function(t, i, s) {
    if (!(t.ignore && !s)) {
      t.beforeUpdate(), t.update(), t.afterUpdate();
      var o = t.getClipPath(), l = i && i.length, p = 0, e = t.__clipPaths;
      if (!t.ignoreClip && (l || o)) {
        if (e || (e = t.__clipPaths = []), l)
          for (var d = 0; d < i.length; d++)
            e[p++] = i[d];
        for (var a = o, _ = t; a; )
          a.parent = _, a.updateTransform(), e[p++] = a, _ = a, a = a.getClipPath();
      }
      if (e && (e.length = p), t.childrenRef) {
        for (var y = t.childrenRef(), f = 0; f < y.length; f++) {
          var u = y[f];
          t.__dirty && (u.__dirty |= m), this._updateAndAddDisplayable(u, e, s);
        }
        t.__dirty = 0;
      } else {
        var n = t;
        isNaN(n.z) && (h(), n.z = 0), isNaN(n.z2) && (h(), n.z2 = 0), isNaN(n.zlevel) && (h(), n.zlevel = 0), this._displayList[this._displayListLen++] = n;
      }
      var v = t.getDecalElement && t.getDecalElement();
      v && this._updateAndAddDisplayable(v, e, s);
      var L = t.getTextGuideLine();
      L && this._updateAndAddDisplayable(L, e, s);
      var z = t.getTextContent();
      z && this._updateAndAddDisplayable(z, e, s);
    }
  }, r.prototype.addRoot = function(t) {
    t.__zr && t.__zr.storage === this || this._roots.push(t);
  }, r.prototype.delRoot = function(t) {
    if (t instanceof Array) {
      for (var i = 0, s = t.length; i < s; i++)
        this.delRoot(t[i]);
      return;
    }
    var o = A(this._roots, t);
    o >= 0 && this._roots.splice(o, 1);
  }, r.prototype.delAllRoots = function() {
    this._roots = [], this._displayList = [], this._displayListLen = 0;
  }, r.prototype.getRoots = function() {
    return this._roots;
  }, r.prototype.dispose = function() {
    this._displayList = null, this._roots = null;
  }, r;
})();
export {
  N as default
};
