// echarts@6.0.0/lib/chart/pie/labelLayout.js
import { parsePercent as W } from "../../util/number.js";
import "../../util/graphic.js";
import be from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/BoundingRect.js";
import { each as j, isNumber as he } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { limitTurnAngle as ue, limitSurfaceAngle as de } from "../../label/labelGuideHelper.js";
import { computeLabelGeometry as me, shiftLayoutOnXY as pe } from "../../label/labelLayoutHelper.js";
import ie from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/Point.js";
var Me = Math.PI / 180;
function oe(r, t, p, g, s, C, f, D, A, m) {
  if (r.length < 2)
    return;
  function M(n) {
    for (var i = n.rB, c = i * i, b = 0; b < n.list.length; b++) {
      var v = n.list[b], x = Math.abs(v.label.y - p), h = g + v.len, u = h * h, S = Math.sqrt(Math.abs((1 - x * x / c) * u)), o = t + (S + v.len2) * s, d = o - v.label.x, w = v.targetTextWidth - d * s;
      ge(v, w, !0), v.label.x = o;
    }
  }
  function N(n) {
    for (var i = {
      list: [],
      maxY: 0
    }, c = {
      list: [],
      maxY: 0
    }, b = 0; b < n.length; b++)
      if (n[b].labelAlignTo === "none") {
        var v = n[b], x = v.label.y > p ? c : i, h = Math.abs(v.label.y - p);
        if (h >= x.maxY) {
          var u = v.label.x - t - v.len2 * s, S = g + v.len, o = Math.abs(u) < S ? Math.sqrt(h * h / (1 - u * u / S / S)) : S;
          x.rB = o, x.maxY = h;
        }
        x.list.push(v);
      }
    M(i), M(c);
  }
  for (var l = r.length, a = 0; a < l; a++)
    if (r[a].position === "outer" && r[a].labelAlignTo === "labelLine") {
      var e = r[a].label.x - m;
      r[a].linePoints[1][0] += e, r[a].label.x = m;
    }
  pe(r, 1, A, A + f) && N(r);
}
function Ae(r, t, p, g, s, C, f, D) {
  for (var A = [], m = [], M = Number.MAX_VALUE, N = -Number.MAX_VALUE, l = 0; l < r.length; l++) {
    var a = r[l].label;
    ee(r[l]) || (a.x < t ? (M = Math.min(M, a.x), A.push(r[l])) : (N = Math.max(N, a.x), m.push(r[l])));
  }
  for (var l = 0; l < r.length; l++) {
    var e = r[l];
    if (!ee(e) && e.linePoints) {
      if (e.labelStyleWidth != null)
        continue;
      var a = e.label, n = e.linePoints, i = void 0;
      e.labelAlignTo === "edge" ? a.x < t ? i = n[2][0] - e.labelDistance - f - e.edgeDistance : i = f + s - e.edgeDistance - n[2][0] - e.labelDistance : e.labelAlignTo === "labelLine" ? a.x < t ? i = M - f - e.bleedMargin : i = f + s - N - e.bleedMargin : a.x < t ? i = a.x - f - e.bleedMargin : i = f + s - a.x - e.bleedMargin, e.targetTextWidth = i, ge(e, i, !1);
    }
  }
  oe(m, t, p, g, 1, s, C, f, D, N), oe(A, t, p, g, -1, s, C, f, D, M);
  for (var l = 0; l < r.length; l++) {
    var e = r[l];
    if (!ee(e) && e.linePoints) {
      var a = e.label, n = e.linePoints, c = e.labelAlignTo === "edge", b = a.style.padding, v = b ? b[1] + b[3] : 0, x = a.style.backgroundColor ? 0 : v, h = e.rect.width + x, u = n[1][0] - n[2][0];
      c ? a.x < t ? n[2][0] = f + e.edgeDistance + h + e.labelDistance : n[2][0] = f + s - e.edgeDistance - h - e.labelDistance : (a.x < t ? n[2][0] = a.x + e.labelDistance : n[2][0] = a.x - e.labelDistance, n[1][0] = n[2][0] + u), n[1][1] = n[2][1] = a.y;
    }
  }
}
function ge(r, t, p) {
  if (r.labelStyleWidth == null) {
    var g = r.label, s = g.style, C = r.rect, f = s.backgroundColor, D = s.padding, A = D ? D[1] + D[3] : 0, m = s.overflow, M = C.width + (f ? 0 : A);
    if (t < M || p) {
      if (m && m.match("break")) {
        g.setStyle("backgroundColor", null), g.setStyle("width", t - A);
        var N = g.getBoundingRect();
        g.setStyle("width", Math.ceil(N.width)), g.setStyle("backgroundColor", f);
      } else {
        var l = t - A, a = t < M ? l : (
          // Current available width is enough, but the text may have
          // already been wrapped with a smaller available width.
          p ? l > r.unconstrainedWidth ? null : l : null
        );
        g.setStyle("width", a);
      }
      se(C, g);
    }
  }
}
function se(r, t) {
  ve.rect = r, me(ve, t, xe);
}
var xe = {
  minMarginForce: [null, 0, null, 0],
  marginDefault: [1, 0, 1, 0]
}, ve = {};
function ee(r) {
  return r.position === "center";
}
function we(r) {
  var t = r.getData(), p = [], g, s, C = !1, f = (r.get("minShowLabelAngle") || 0) * Me, D = t.getLayout("viewRect"), A = t.getLayout("r"), m = D.width, M = D.x, N = D.y, l = D.height;
  function a(u) {
    u.ignore = !0;
  }
  function e(u) {
    if (!u.ignore)
      return !0;
    for (var S in u.states)
      if (u.states[S].ignore === !1)
        return !0;
    return !1;
  }
  t.each(function(u) {
    var S = t.getItemGraphicEl(u), o = S.shape, d = S.getTextContent(), w = S.getTextGuideLine(), F = t.getItemModel(u), G = F.getModel("label"), P = G.get("position") || F.get(["emphasis", "label", "position"]), H = G.get("distanceToLabelLine"), z = G.get("alignTo"), J = W(G.get("edgeDistance"), m), K = G.get("bleedMargin");
    K == null && (K = Math.min(m, l) > 200 ? 10 : 2);
    var I = F.getModel("labelLine"), X = I.get("length");
    X = W(X, m);
    var E = I.get("length2");
    if (E = W(E, m), Math.abs(o.endAngle - o.startAngle) < f) {
      j(d.states, a), d.ignore = !0, w && (j(w.states, a), w.ignore = !0);
      return;
    }
    if (e(d)) {
      var q = (o.startAngle + o.endAngle) / 2, T = Math.cos(q), L = Math.sin(q), Y, U, re, V;
      g = o.cx, s = o.cy;
      var O = P === "inside" || P === "inner";
      if (P === "center")
        Y = o.cx, U = o.cy, V = "center";
      else {
        var Q = (O ? (o.r + o.r0) / 2 * T : o.r * T) + g, Z = (O ? (o.r + o.r0) / 2 * L : o.r * L) + s;
        if (Y = Q + T * 3, U = Z + L * 3, !O) {
          var ae = Q + T * (X + A - o.r), ne = Z + L * (X + A - o.r), te = ae + (T < 0 ? -1 : 1) * E, le = ne;
          z === "edge" ? Y = T < 0 ? M + J : M + m - J : Y = te + (T < 0 ? -H : H), U = le, re = [[Q, Z], [ae, ne], [te, le]];
        }
        V = O ? "center" : z === "edge" ? T > 0 ? "right" : "left" : T > 0 ? "left" : "right";
      }
      var _ = Math.PI, R = 0, k = G.get("rotate");
      if (he(k))
        R = k * (_ / 180);
      else if (P === "center")
        R = 0;
      else if (k === "radial" || k === !0) {
        var fe = T < 0 ? -q + _ : -q;
        R = fe;
      } else if (k === "tangential" && P !== "outside" && P !== "outer") {
        var B = Math.atan2(T, L);
        B < 0 && (B = _ * 2 + B);
        var ce = L > 0;
        ce && (B = _ + B), R = B - _;
      }
      if (C = !!R, d.x = Y, d.y = U, d.rotation = R, d.setStyle({
        verticalAlign: "middle"
      }), O) {
        d.setStyle({
          align: V
        });
        var y = d.states.select;
        y && (y.x += d.x, y.y += d.y);
      } else {
        var $ = new be(0, 0, 0, 0);
        se($, d), p.push({
          label: d,
          labelLine: w,
          position: P,
          len: X,
          len2: E,
          minTurnAngle: I.get("minTurnAngle"),
          maxSurfaceAngle: I.get("maxSurfaceAngle"),
          surfaceNormal: new ie(T, L),
          linePoints: re,
          textAlign: V,
          labelDistance: H,
          labelAlignTo: z,
          edgeDistance: J,
          bleedMargin: K,
          rect: $,
          unconstrainedWidth: $.width,
          labelStyleWidth: d.style.width
        });
      }
      S.setTextConfig({
        inside: O
      });
    }
  }), !C && r.get("avoidLabelOverlap") && Ae(p, g, s, A, m, l, M, N);
  for (var n = 0; n < p.length; n++) {
    var i = p[n], c = i.label, b = i.labelLine, v = isNaN(c.x) || isNaN(c.y);
    if (c) {
      c.setStyle({
        align: i.textAlign
      }), v && (j(c.states, a), c.ignore = !0);
      var x = c.states.select;
      x && (x.x += c.x, x.y += c.y);
    }
    if (b) {
      var h = i.linePoints;
      v || !h ? (j(b.states, a), b.ignore = !0) : (ue(h, i.minTurnAngle), de(h, i.surfaceNormal, i.maxSurfaceAngle), b.setShape({
        points: h
      }), c.__hostTarget.textGuideLineConfig = {
        anchor: new ie(h[0][0], h[0][1])
      });
    }
  }
}
export {
  we as default
};
