// echarts@6.0.0/lib/chart/pie/pieLayout.js
import { linearMap as J } from "../../util/number.js";
import { normalizeArcAngles as K } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/PathProxy.js";
import { makeInner as O } from "../../util/model.js";
import { getCircleLayout as Q } from "../../util/layout.js";
var z = Math.PI * 2, P = Math.PI / 180;
function $(B, j, q) {
  j.eachSeriesByType(B, function(v) {
    var r = v.getData(), s = r.mapDimension("value"), N = Q(v, q), R = N.cx, x = N.cy, y = N.r, L = N.r0, F = N.viewRect, i = -v.get("startAngle") * P, m = v.get("endAngle"), h = v.get("padAngle") * P;
    m = m === "auto" ? i - z : -m * P;
    var G = v.get("minAngle") * P, p = G + h, u = 0;
    r.each(s, function(n) {
      !isNaN(n) && u++;
    });
    var C = r.getSum(s), D = Math.PI / (C || u) * 2, w = v.get("clockwise"), T = v.get("roseType"), H = v.get("stillShowZeroSum"), Z = r.getDataExtent(s);
    Z[0] = 0;
    var g = w ? 1 : -1, S = [i, m], l = g * h / 2;
    K(S, !w), i = S[0], m = S[1];
    var f = U(v);
    f.startAngle = i, f.endAngle = m, f.clockwise = w, f.cx = R, f.cy = x, f.r = y, f.r0 = L;
    var d = Math.abs(m - i), k = d, b = 0, c = i;
    if (r.setLayout({
      viewRect: F,
      r: y
    }), r.each(s, function(n, o) {
      var a;
      if (isNaN(n)) {
        r.setItemLayout(o, {
          angle: NaN,
          startAngle: NaN,
          endAngle: NaN,
          clockwise: w,
          cx: R,
          cy: x,
          r0: L,
          r: T ? NaN : y
        });
        return;
      }
      T !== "area" ? a = C === 0 && H ? D : n * D : a = d / u, a < p ? (a = p, k -= p) : b += n;
      var t = c + g * a, e = 0, A = 0;
      h > a ? (e = c + g * a / 2, A = e) : (e = c + l, A = t - l), r.setItemLayout(o, {
        angle: a,
        startAngle: e,
        endAngle: A,
        clockwise: w,
        cx: R,
        cy: x,
        r0: L,
        r: T ? J(n, Z, [L, y]) : y
      }), c = t;
    }), k < z && u)
      if (k <= 1e-3) {
        var I = d / u;
        r.each(s, function(n, o) {
          if (!isNaN(n)) {
            var a = r.getItemLayout(o);
            a.angle = I;
            var t = 0, e = 0;
            I < h ? (t = i + g * (o + 1 / 2) * I, e = t) : (t = i + g * o * I + l, e = i + g * (o + 1) * I - l), a.startAngle = t, a.endAngle = e;
          }
        });
      } else
        D = k / b, c = i, r.each(s, function(n, o) {
          if (!isNaN(n)) {
            var a = r.getItemLayout(o), t = a.angle === p ? p : n * D, e = 0, A = 0;
            t < h ? (e = c + g * t / 2, A = e) : (e = c + l, A = c + g * t - l), a.startAngle = e, a.endAngle = A, c += g * t;
          }
        });
  });
}
var U = O();
export {
  $ as default,
  U as getSeriesLayoutData
};
