// echarts@6.0.0/lib/chart/pie/PieSeries.js
import { __extends as m } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import u from "../helper/createSeriesDataSimply.js";
import { bind as s, curry as h } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { makeInner as c, defaultEmphasis as d } from "../../util/model.js";
import { getPercentSeats as g } from "../../util/number.js";
import { makeSeriesEncodeForNameBased as f } from "../../data/helper/sourceHelper.js";
import y from "../../visual/LegendVisualProvider.js";
import b from "../../model/Series.js";
import { registerLayOutOnCoordSysUsage as w } from "../../core/CoordinateSystem.js";
var v = c(), S = (
  /** @class */
  (function(a) {
    m(t, a);
    function t() {
      return a !== null && a.apply(this, arguments) || this;
    }
    return t.prototype.init = function(e) {
      a.prototype.init.apply(this, arguments), this.legendVisualProvider = new y(s(this.getData, this), s(this.getRawData, this)), this._defaultLabelLine(e);
    }, t.prototype.mergeOption = function() {
      a.prototype.mergeOption.apply(this, arguments);
    }, t.prototype.getInitialData = function() {
      return u(this, {
        coordDimensions: ["value"],
        encodeDefaulter: h(f, this)
      });
    }, t.prototype.getDataParams = function(e) {
      var i = this.getData(), r = v(i), o = r.seats;
      if (!o) {
        var l = [];
        i.each(i.mapDimension("value"), function(p) {
          l.push(p);
        }), o = r.seats = g(l, i.hostModel.get("percentPrecision"));
      }
      var n = a.prototype.getDataParams.call(this, e);
      return n.percent = o[e] || 0, n.$vars.push("percent"), n;
    }, t.prototype._defaultLabelLine = function(e) {
      d(e, "labelLine", ["show"]);
      var i = e.labelLine, r = e.emphasis.labelLine;
      i.show = i.show && e.label.show, r.show = r.show && e.emphasis.label.show;
    }, t.type = "series.pie", t.defaultOption = {
      // zlevel: 0,
      z: 2,
      legendHoverLink: !0,
      colorBy: "data",
      // 默认全局居中
      center: ["50%", "50%"],
      radius: [0, "50%"],
      // 默认顺时针
      clockwise: !0,
      startAngle: 90,
      endAngle: "auto",
      padAngle: 0,
      // 最小角度改为0
      minAngle: 0,
      // If the angle of a sector less than `minShowLabelAngle`,
      // the label will not be displayed.
      minShowLabelAngle: 0,
      // 选中时扇区偏移量
      selectedOffset: 10,
      // 选择模式，默认关闭，可选single，multiple
      // selectedMode: false,
      // 南丁格尔玫瑰图模式，'radius'（半径） | 'area'（面积）
      // roseType: null,
      percentPrecision: 2,
      // If still show when all data zero.
      stillShowZeroSum: !0,
      // cursor: null,
      coordinateSystemUsage: "box",
      left: 0,
      top: 0,
      right: 0,
      bottom: 0,
      width: null,
      height: null,
      label: {
        // color: 'inherit',
        // If rotate around circle
        rotate: 0,
        show: !0,
        overflow: "truncate",
        // 'outer', 'inside', 'center'
        position: "outer",
        // 'none', 'labelLine', 'edge'. Works only when position is 'outer'
        alignTo: "none",
        // Closest distance between label and chart edge.
        // Works only position is 'outer' and alignTo is 'edge'.
        edgeDistance: "25%",
        // Works only position is 'outer' and alignTo is not 'edge'.
        // The default `bleedMargin` is auto determined according to view rect size.
        // bleedMargin: 10,
        // Distance between text and label line.
        distanceToLabelLine: 5
        // formatter: 标签文本格式器，同 tooltip.formatter，不支持异步回调
        // 默认使用全局文本样式，详见 textStyle
        // distance: 当position为inner时有效，为label位置到圆心的距离与圆半径(环状图为内外半径和)的比例系数
      },
      // Enabled when label.normal.position is 'outer'
      labelLine: {
        show: !0,
        // 引导线两段中的第一段长度
        length: 15,
        // 引导线两段中的第二段长度
        length2: 30,
        smooth: !1,
        minTurnAngle: 90,
        maxSurfaceAngle: 90,
        lineStyle: {
          // color: 各异,
          width: 1,
          type: "solid"
        }
      },
      itemStyle: {
        borderWidth: 1,
        borderJoin: "round"
      },
      showEmptyCircle: !0,
      emptyCircleStyle: {
        color: "lightgray",
        opacity: 1
      },
      labelLayout: {
        // Hide the overlapped label.
        hideOverlap: !0
      },
      emphasis: {
        scale: !0,
        scaleSize: 5
      },
      // If use strategy to avoid label overlapping
      avoidLabelOverlap: !0,
      // Animation type. Valid values: expansion, scale
      animationType: "expansion",
      animationDuration: 1e3,
      // Animation type when update. Valid values: transition, expansion
      animationTypeUpdate: "transition",
      animationEasingUpdate: "cubicInOut",
      animationDurationUpdate: 500,
      animationEasing: "cubicInOut"
    }, t;
  })(b)
);
w({
  fullType: S.type,
  getCoord2: function(a) {
    return a.getShallow("center");
  }
});
export {
  S as default
};
