// echarts@6.0.0/lib/chart/pie/PieView.js
import { __extends as A } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { clone as T, extend as y, retrieve3 as C } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { setStatesStylesFromModel as x, toggleHoverEmphasis as w } from "../../util/states.js";
import D from "../../view/Chart.js";
import G from "./labelLayout.js";
import { setLabelLineStyle as _, getLabelLineStatesModels as E } from "../../label/labelGuideHelper.js";
import { setLabelStyle as F, getLabelStatesModels as N } from "../../label/labelStyle.js";
import { getSectorCornerRadius as d } from "../helper/sectorHelper.js";
import { removeElementWithFadeOut as O, initProps as S, updateProps as M, saveOldStyle as V } from "../../animation/basicTransition.js";
import { getSeriesLayoutData as z } from "./pieLayout.js";
import I from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Sector.js";
import k from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Text.js";
import R from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Polyline.js";
var U = (
  /** @class */
  (function(v) {
    A(u, v);
    function u(t, r, n) {
      var l = v.call(this) || this;
      l.z2 = 2;
      var e = new k();
      return l.setTextContent(e), l.updateData(t, r, n, !0), l;
    }
    return u.prototype.updateData = function(t, r, n, l) {
      var e = this, o = t.hostModel, i = t.getItemModel(r), s = i.getModel("emphasis"), a = t.getItemLayout(r), p = y(d(i.getModel("itemStyle"), a, !0), a);
      if (isNaN(p.startAngle)) {
        e.setShape(p);
        return;
      }
      if (l) {
        e.setShape(p);
        var h = o.getShallow("animationType");
        o.ecModel.ssr ? (S(e, {
          scaleX: 0,
          scaleY: 0
        }, o, {
          dataIndex: r,
          isFrom: !0
        }), e.originX = p.cx, e.originY = p.cy) : h === "scale" ? (e.shape.r = a.r0, S(e, {
          shape: {
            r: a.r
          }
        }, o, r)) : n != null ? (e.setShape({
          startAngle: n,
          endAngle: n
        }), S(e, {
          shape: {
            startAngle: a.startAngle,
            endAngle: a.endAngle
          }
        }, o, r)) : (e.shape.endAngle = a.startAngle, M(e, {
          shape: {
            endAngle: a.endAngle
          }
        }, o, r));
      } else
        V(e), M(e, {
          shape: p
        }, o, r);
      e.useStyle(t.getItemVisual(r, "style")), x(e, i);
      var m = (a.startAngle + a.endAngle) / 2, c = o.get("selectedOffset"), g = Math.cos(m) * c, f = Math.sin(m) * c, L = i.getShallow("cursor");
      L && e.attr("cursor", L), this._updateLabel(o, t, r), e.ensureState("emphasis").shape = y({
        r: a.r + (s.get("scale") && s.get("scaleSize") || 0)
      }, d(s.getModel("itemStyle"), a)), y(e.ensureState("select"), {
        x: g,
        y: f,
        shape: d(i.getModel(["select", "itemStyle"]), a)
      }), y(e.ensureState("blur"), {
        shape: d(i.getModel(["blur", "itemStyle"]), a)
      });
      var b = e.getTextGuideLine(), P = e.getTextContent();
      b && y(b.ensureState("select"), {
        x: g,
        y: f
      }), y(P.ensureState("select"), {
        x: g,
        y: f
      }), w(this, s.get("focus"), s.get("blurScope"), s.get("disabled"));
    }, u.prototype._updateLabel = function(t, r, n) {
      var l = this, e = r.getItemModel(n), o = e.getModel("labelLine"), i = r.getItemVisual(n, "style"), s = i && i.fill, a = i && i.opacity;
      F(l, N(e), {
        labelFetcher: r.hostModel,
        labelDataIndex: n,
        inheritColor: s,
        defaultOpacity: a,
        defaultText: t.getFormattedLabel(n, "normal") || r.getName(n)
      });
      var p = l.getTextContent();
      l.setTextConfig({
        // reset position, rotation
        position: null,
        rotation: null
      }), p.attr({
        z2: 10
      });
      var h = e.get(["label", "position"]);
      if (h !== "outside" && h !== "outer")
        l.removeTextGuideLine();
      else {
        var m = this.getTextGuideLine();
        m || (m = new R(), this.setTextGuideLine(m)), _(this, E(e), {
          stroke: s,
          opacity: C(o.get(["lineStyle", "opacity"]), a, 1)
        });
      }
    }, u;
  })(I)
), te = (
  /** @class */
  (function(v) {
    A(u, v);
    function u() {
      var t = v !== null && v.apply(this, arguments) || this;
      return t.ignoreLabelLineUpdate = !0, t;
    }
    return u.prototype.render = function(t, r, n, l) {
      var e = t.getData(), o = this._data, i = this.group, s;
      if (!o && e.count() > 0) {
        for (var a = e.getItemLayout(0), p = 1; isNaN(a && a.startAngle) && p < e.count(); ++p)
          a = e.getItemLayout(p);
        a && (s = a.startAngle);
      }
      if (this._emptyCircleSector && i.remove(this._emptyCircleSector), e.count() === 0 && t.get("showEmptyCircle")) {
        var h = z(t), m = new I({
          shape: T(h)
        });
        m.useStyle(t.getModel("emptyCircleStyle").getItemStyle()), this._emptyCircleSector = m, i.add(m);
      }
      e.diff(o).add(function(c) {
        var g = new U(e, c, s);
        e.setItemGraphicEl(c, g), i.add(g);
      }).update(function(c, g) {
        var f = o.getItemGraphicEl(g);
        f.updateData(e, c, s), f.off("click"), i.add(f), e.setItemGraphicEl(c, f);
      }).remove(function(c) {
        var g = o.getItemGraphicEl(c);
        O(g, t, c);
      }).execute(), G(t), t.get("animationTypeUpdate") !== "expansion" && (this._data = e);
    }, u.prototype.dispose = function() {
    }, u.prototype.containPoint = function(t, r) {
      var n = r.getData(), l = n.getItemLayout(0);
      if (l) {
        var e = t[0] - l.cx, o = t[1] - l.cy, i = Math.sqrt(e * e + o * o);
        return i <= l.r && i >= l.r0;
      }
    }, u.type = "pie", u;
  })(D)
);
export {
  te as default
};
