// echarts@6.0.0/lib/chart/custom/CustomSeries.js
import { __extends as n } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import s from "../helper/createSeriesData.js";
import { makeInner as l } from "../../util/model.js";
import p from "../../model/Series.js";
var y = {
  color: "fill",
  borderColor: "stroke"
}, S = {
  symbol: 1,
  symbolSize: 1,
  symbolKeepAspect: 1,
  legendIcon: 1,
  visualMeta: 1,
  liftZ: 1,
  decal: 1
}, m = l(), g = (
  /** @class */
  (function(r) {
    n(e, r);
    function e() {
      var t = r !== null && r.apply(this, arguments) || this;
      return t.type = e.type, t;
    }
    return e.prototype.optionUpdated = function() {
      this.currentZLevel = this.get("zlevel", !0), this.currentZ = this.get("z", !0);
    }, e.prototype.getInitialData = function(t, o) {
      return s(null, this);
    }, e.prototype.getDataParams = function(t, o, i) {
      var a = r.prototype.getDataParams.call(this, t, o);
      return i && (a.info = m(i).info), a;
    }, e.type = "series.custom", e.dependencies = ["grid", "polar", "geo", "singleAxis", "calendar", "matrix"], e.defaultOption = {
      coordinateSystem: "cartesian2d",
      // zlevel: 0,
      z: 2,
      legendHoverLink: !0,
      // Custom series will not clip by default.
      // Some case will use custom series to draw label
      // For example https://echarts.apache.org/examples/en/editor.html?c=custom-gantt-flight
      clip: !1
      // Cartesian coordinate system
      // xAxisIndex: 0,
      // yAxisIndex: 0,
      // Polar coordinate system
      // polarIndex: 0,
      // Geo coordinate system
      // geoIndex: 0,
    }, e;
  })(p)
);
export {
  S as NON_STYLE_VISUAL_PROPS,
  y as STYLE_VISUAL_TYPE,
  m as customInnerStore,
  g as default
};
