// echarts@6.0.0/lib/chart/custom/CustomView.js
import { __extends as Le } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { assert as w, defaults as M, retrieve2 as _e, retrieve3 as we, hasOwn as D, each as Ae, indexOf as Ne, isString as Te, map as Fe } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { traverseElements as Ve, makePath as ee, getShapeClass as re } from "../../util/graphic.js";
import { toggleHoverEmphasis as Re, setDefaultStateProxy as Me } from "../../util/states.js";
import { createTextStyle as te, createTextConfig as ne, getFont as ze } from "../../label/labelStyle.js";
import { getDefaultLabel as oe } from "../helper/labelHelper.js";
import { getLayoutOnAxis as Ue } from "../../layout/barGrid.js";
import Ge from "../../data/DataDiffer.js";
import Ze from "../../view/Chart.js";
import { createClipPath as ke } from "../helper/createClipPathFromCoordSys.js";
import He from "../../coord/cartesian/prepareCustom.js";
import Be from "../../coord/geo/prepareCustom.js";
import xe from "../../coord/single/prepareCustom.js";
import Oe from "../../coord/polar/prepareCustom.js";
import Ye from "../../coord/calendar/prepareCustom.js";
import Ke from "../../coord/matrix/prepareCustom.js";
import We from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Displayable.js";
import { warnDeprecated as ie, convertToEC4StyleForCustomSerise as le, isEC4CompatibleStyle as qe, convertFromEC4CompatibleStyle as Xe } from "../../util/styleCompat.js";
import { throwError as z } from "../../util/log.js";
import { createOrUpdatePatternFromDecal as $e } from "../../util/decal.js";
import { customInnerStore as E, STYLE_VISUAL_TYPE as ae, NON_STYLE_VISUAL_PROPS as Je } from "./CustomSeries.js";
import { applyLeaveTransition as H, applyUpdateTransition as Qe } from "../../animation/customGraphicTransition.js";
import { stopPreviousKeyframeAnimationAndRestore as je, applyKeyframeAnimation as Ie } from "../../animation/customGraphicKeyframeAnimation.js";
import { getCustomSeries as er } from "./customSeriesRegister.js";
import rr from "../../visual/tokens.js";
import tr from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Image.js";
import nr from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Text.js";
import or from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import ir from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/CompoundPath.js";
import lr from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Path.js";
var P = "emphasis", b = "normal", B = "blur", x = "select", L = [b, P, B, x], U = {
  normal: ["itemStyle"],
  emphasis: [P, "itemStyle"],
  blur: [B, "itemStyle"],
  select: [x, "itemStyle"]
}, G = {
  normal: ["label"],
  emphasis: [P, "label"],
  blur: [B, "label"],
  select: [x, "label"]
}, ar = ["x", "y"], ur = "e\0\0", C = {
  normal: {},
  emphasis: {},
  blur: {},
  select: {}
}, ue = {
  cartesian2d: He,
  geo: Be,
  single: xe,
  polar: Oe,
  calendar: Ye,
  matrix: Ke
};
function T(r) {
  return r instanceof lr;
}
function k(r) {
  return r instanceof We;
}
function fr(r, e) {
  e.copyTransform(r), k(e) && k(r) && (e.setStyle(r.style), e.z = r.z, e.z2 = r.z2, e.zlevel = r.zlevel, e.invisible = r.invisible, e.ignore = r.ignore, T(e) && T(r) && e.setShape(r.shape));
}
var Jr = (
  /** @class */
  (function(r) {
    Le(e, r);
    function e() {
      var n = r !== null && r.apply(this, arguments) || this;
      return n.type = e.type, n;
    }
    return e.prototype.render = function(n, t, i, l) {
      this._progressiveEls = null;
      var o = this._data, p = n.getData(), m = this.group, v = fe(n, p, t, i);
      o || m.removeAll(), p.diff(o).add(function(s) {
        Z(i, null, s, v(s, l), n, m, p);
      }).remove(function(s) {
        var c = o.getItemGraphicEl(s);
        c && H(c, E(c).option, n);
      }).update(function(s, c) {
        var h = o.getItemGraphicEl(c);
        Z(i, h, s, v(s, l), n, m, p);
      }).execute();
      var u = n.get("clip", !0) ? ke(n.coordinateSystem, !1, n) : null;
      u ? m.setClipPath(u) : m.removeClipPath(), this._data = p;
    }, e.prototype.incrementalPrepareRender = function(n, t, i) {
      this.group.removeAll(), this._data = null;
    }, e.prototype.incrementalRender = function(n, t, i, l, o) {
      var p = t.getData(), m = fe(t, p, i, l), v = this._progressiveEls = [];
      function u(h) {
        h.isGroup || (h.incremental = !0, h.ensureState("emphasis").hoverLayer = !0);
      }
      for (var s = n.start; s < n.end; s++) {
        var c = Z(null, null, s, m(s, o), t, this.group, p);
        c && (c.traverse(u), v.push(c));
      }
    }, e.prototype.eachRendered = function(n) {
      Ve(this._progressiveEls || this.group, n);
    }, e.prototype.filterForExposedEvent = function(n, t, i, l) {
      var o = t.element;
      if (o == null || i.name === o)
        return !0;
      for (; (i = i.__hostTarget || i.parent) && i !== this.group; )
        if (i.name === o)
          return !0;
      return !1;
    }, e.type = "custom", e;
  })(Ze)
);
function O(r) {
  var e = r.type, n;
  if (e === "path") {
    var t = r.shape, i = t.width != null && t.height != null ? {
      x: t.x || 0,
      y: t.y || 0,
      width: t.width,
      height: t.height
    } : null, l = ge(t);
    n = ee(l, null, i, t.layout || "center"), E(n).customPathData = l;
  } else if (e === "image")
    n = new tr({}), E(n).customImagePath = r.style.image;
  else if (e === "text")
    n = new nr({});
  else if (e === "group")
    n = new or();
  else if (e === "compoundPath") {
    var t = r.shape;
    if (!t || !t.paths) {
      var o = "";
      process.env.NODE_ENV !== "production" && (o = "shape.paths must be specified in compoundPath"), z(o);
    }
    var p = Fe(t.paths, function(u) {
      if (u.type === "path")
        return ee(u.shape.pathData, u, null);
      var s = re(u.type);
      if (!s) {
        var c = "";
        process.env.NODE_ENV !== "production" && (c = 'graphic type "' + e + '" can not be found.'), z(c);
      }
      return new s();
    });
    n = new ir({
      shape: {
        paths: p
      }
    });
  } else {
    var m = re(e);
    if (!m) {
      var o = "";
      process.env.NODE_ENV !== "production" && (o = 'graphic type "' + e + '" can not be found.'), z(o);
    }
    n = new m();
  }
  return E(n).customGraphicType = e, n.name = r.name, n.z2EmphasisLift = 1, n.z2SelectLift = 1, n;
}
function Y(r, e, n, t, i, l, o) {
  je(e);
  var p = i && i.normal.cfg;
  p && e.setTextConfig(p), t && t.transition == null && (t.transition = ar);
  var m = t && t.style;
  if (m) {
    if (e.type === "text") {
      var v = m;
      D(v, "textFill") && (v.fill = v.textFill), D(v, "textStroke") && (v.stroke = v.textStroke);
    }
    var u = void 0, s = T(e) ? m.decal : null;
    r && s && (s.dirty = !0, u = $e(s, r)), m.__decalPattern = u;
  }
  if (k(e) && m) {
    var u = m.__decalPattern;
    u && (m.decal = u);
  }
  Qe(e, t, l, {
    dataIndex: n,
    isInit: o,
    clearStyle: !0
  }), Ie(e, t.keyframeAnimation, l);
}
function se(r, e, n, t, i) {
  var l = e.isGroup ? null : e, o = i && i[r].cfg;
  if (l) {
    var p = l.ensureState(r);
    if (t === !1) {
      var m = l.getState(r);
      m && (m.style = null);
    } else
      p.style = t || null;
    o && (p.textConfig = o), Me(l);
  }
}
function pr(r, e, n) {
  if (!r.isGroup) {
    var t = r, i = n.currentZ, l = n.currentZLevel;
    t.z = i, t.zlevel = l;
    var o = e.z2;
    o != null && (t.z2 = o || 0);
    for (var p = 0; p < L.length; p++)
      vr(t, e, L[p]);
  }
}
function vr(r, e, n) {
  var t = n === b, i = t ? e : F(e, n), l = i ? i.z2 : null, o;
  l != null && (o = t ? r : r.ensureState(n), o.z2 = l || 0);
}
function fe(r, e, n, t) {
  var i = r.get("renderItem");
  if (typeof i == "string") {
    var l = er(i);
    l ? i = l : process.env.NODE_ENV !== "production" && console.warn("Custom series renderItem '" + i + `' not found.
                Call 'echarts.registerCustomSeries' to register it.`);
  }
  var o = r.coordinateSystem, p = {};
  o && (process.env.NODE_ENV !== "production" && (w(i, "series.render is required."), w(o.prepareCustoms || ue[o.type], "This coordSys does not support custom series.")), p = o.prepareCustoms ? o.prepareCustoms(o) : ue[o.type](o));
  for (var m = M({
    getWidth: t.getWidth,
    getHeight: t.getHeight,
    getZr: t.getZr,
    getDevicePixelRatio: t.getDevicePixelRatio,
    value: he,
    style: Ce,
    ordinalRawValue: ye,
    styleEmphasis: de,
    visual: De,
    barLayout: Pe,
    currentSeriesIndices: Se,
    font: be
  }, p.api || {}), v = {
    // The life cycle of context: current round of rendering.
    // The global life cycle is probably not necessary, because
    // user can store global status by themselves.
    context: {},
    seriesId: r.id,
    seriesName: r.name,
    seriesIndex: r.seriesIndex,
    coordSys: p.coordSys,
    dataInsideLength: e.count(),
    encode: mr(r.getData()),
    itemPayload: r.get("itemPayload") || {}
  }, u, s, c = {}, h = {}, _ = {}, q = {}, V = 0; V < L.length; V++) {
    var A = L[V];
    _[A] = r.getModel(U[A]), q[A] = r.getModel(G[A]);
  }
  function N(f) {
    return f === u ? s || (s = e.getItemModel(f)) : e.getItemModel(f);
  }
  function X(f, a) {
    return e.hasItemOption ? f === u ? c[a] || (c[a] = N(f).getModel(U[a])) : N(f).getModel(U[a]) : _[a];
  }
  function $(f, a) {
    return e.hasItemOption ? f === u ? h[a] || (h[a] = N(f).getModel(G[a])) : N(f).getModel(G[a]) : q[a];
  }
  return function(f, a) {
    return u = f, s = null, c = {}, h = {}, i && i(M({
      dataIndexInside: f,
      dataIndex: e.getRawIndex(f),
      // Can be used for optimization when zoom or roam.
      actionType: a ? a.type : null
    }, v), m);
  };
  function he(f, a) {
    return a == null && (a = u), e.getStore().get(e.getDimensionIndex(f || 0), a);
  }
  function ye(f, a) {
    a == null && (a = u), f = f || 0;
    var g = e.getDimensionInfo(f);
    if (!g) {
      var d = e.getDimensionIndex(f);
      return d >= 0 ? e.getStore().get(d, a) : void 0;
    }
    var S = e.get(g.name, a), y = g && g.ordinalMeta;
    return y ? y.categories[S] : S;
  }
  function Ce(f, a) {
    process.env.NODE_ENV !== "production" && ie("api.style", "Please write literal style directly instead."), a == null && (a = u);
    var g = e.getItemVisual(a, "style"), d = g && g.fill, S = g && g.opacity, y = X(a, b).getItemStyle();
    d != null && (y.fill = d), S != null && (y.opacity = S);
    var j = {
      inheritColor: Te(d) ? d : rr.color.neutral99
    }, R = $(a, b), I = te(R, null, j, !1, !0);
    I.text = R.getShallow("show") ? _e(r.getFormattedLabel(a, b), oe(e, a)) : null;
    var Ee = ne(R, j, !1);
    return Q(f, y), y = le(y, I, Ee), f && J(y, f), y.legacy = !0, y;
  }
  function de(f, a) {
    process.env.NODE_ENV !== "production" && ie("api.styleEmphasis", "Please write literal style directly instead."), a == null && (a = u);
    var g = X(a, P).getItemStyle(), d = $(a, P), S = te(d, null, null, !0, !0);
    S.text = d.getShallow("show") ? we(r.getFormattedLabel(a, P), r.getFormattedLabel(a, b), oe(e, a)) : null;
    var y = ne(d, null, !0);
    return Q(f, g), g = le(g, S, y), f && J(g, f), g.legacy = !0, g;
  }
  function J(f, a) {
    for (var g in a)
      D(a, g) && (f[g] = a[g]);
  }
  function Q(f, a) {
    f && (f.textFill && (a.textFill = f.textFill), f.textPosition && (a.textPosition = f.textPosition));
  }
  function De(f, a) {
    if (a == null && (a = u), D(ae, f)) {
      var g = e.getItemVisual(a, "style");
      return g ? g[ae[f]] : null;
    }
    if (D(Je, f))
      return e.getItemVisual(a, f);
  }
  function Pe(f) {
    if (o.type === "cartesian2d") {
      var a = o.getBaseAxis();
      return Ue(M({
        axis: a
      }, f));
    }
  }
  function Se() {
    return n.getCurrentSeriesIndices();
  }
  function be(f) {
    return ze(f, n);
  }
}
function mr(r) {
  var e = {};
  return Ae(r.dimensions, function(n) {
    var t = r.getDimensionInfo(n);
    if (!t.isExtraCoord) {
      var i = t.coordDim, l = e[i] = e[i] || [];
      l[t.coordDimIndex] = r.getDimensionIndex(n);
    }
  }), e;
}
function Z(r, e, n, t, i, l, o) {
  if (!t) {
    l.remove(e);
    return;
  }
  var p = K(r, e, n, t, i, l);
  return p && o.setItemGraphicEl(n, p), p && Re(p, t.focus, t.blurScope, t.emphasisDisabled), p;
}
function K(r, e, n, t, i, l) {
  process.env.NODE_ENV !== "production" && w(t, "should not have an null/undefined element setting");
  var o = -1, p = e;
  e && ce(e, t, i) && (o = Ne(l.childrenRef(), e), e = null);
  var m = !e, v = e;
  v ? v.clearStates() : (v = O(t), p && fr(p, v)), t.morph === !1 ? v.disableMorphing = !0 : v.disableMorphing && (v.disableMorphing = !1), t.tooltipDisabled && (v.tooltipDisabled = !0), C.normal.cfg = C.normal.conOpt = C.emphasis.cfg = C.emphasis.conOpt = C.blur.cfg = C.blur.conOpt = C.select.cfg = C.select.conOpt = null, C.isLegacy = !1, cr(v, n, t, i, m, C), sr(v, n, t, i, m), Y(r, v, n, t, C, i, m), D(t, "info") && (E(v).info = t.info);
  for (var u = 0; u < L.length; u++) {
    var s = L[u];
    if (s !== b) {
      var c = F(t, s), h = W(t, c, s);
      se(s, v, c, h, C);
    }
  }
  return pr(v, t, i), t.type === "group" && gr(r, v, n, t, i), o >= 0 ? l.replaceAt(v, o) : l.add(v), v;
}
function ce(r, e, n) {
  var t = E(r), i = e.type, l = e.shape, o = e.style;
  return (
    // Always create new if universal transition is enabled.
    // Because we do transition after render. It needs to know what old element is. Replacement will loose it.
    n.isUniversalTransitionEnabled() || i != null && i !== t.customGraphicType || i === "path" && dr(l) && ge(l) !== t.customPathData || i === "image" && D(o, "image") && o.image !== t.customImagePath
  );
}
function sr(r, e, n, t, i) {
  var l = n.clipPath;
  if (l === !1)
    r && r.getClipPath() && r.removeClipPath();
  else if (l) {
    var o = r.getClipPath();
    o && ce(o, l, t) && (o = null), o || (o = O(l), process.env.NODE_ENV !== "production" && w(T(o), "Only any type of `path` can be used in `clipPath`, rather than " + o.type + "."), r.setClipPath(o)), Y(null, o, e, l, null, t, i);
  }
}
function cr(r, e, n, t, i, l) {
  if (!(r.isGroup || r.type === "compoundPath")) {
    pe(n, null, l), pe(n, P, l);
    var o = l.normal.conOpt, p = l.emphasis.conOpt, m = l.blur.conOpt, v = l.select.conOpt;
    if (o != null || p != null || v != null || m != null) {
      var u = r.getTextContent();
      if (o === !1)
        u && r.removeTextContent();
      else {
        o = l.normal.conOpt = o || {
          type: "text"
        }, u ? u.clearStates() : (u = O(o), r.setTextContent(u)), Y(null, u, e, o, null, t, i);
        for (var s = o && o.style, c = 0; c < L.length; c++) {
          var h = L[c];
          if (h !== b) {
            var _ = l[h].conOpt;
            se(h, u, _, W(o, _, h), null);
          }
        }
        s ? u.dirty() : u.markRedraw();
      }
    }
  }
}
function pe(r, e, n) {
  var t = e ? F(r, e) : r, i = e ? W(r, t, P) : r.style, l = r.type, o = t ? t.textConfig : null, p = r.textContent, m = p ? e ? F(p, e) : p : null;
  if (i && // Because emphasis style has little info to detect legacy,
  // if normal is legacy, emphasis is trade as legacy.
  (n.isLegacy || qe(i, l, !!o, !!m))) {
    n.isLegacy = !0;
    var v = Xe(i, l, !e);
    !o && v.textConfig && (o = v.textConfig), !m && v.textContent && (m = v.textContent);
  }
  if (!e && m) {
    var u = m;
    !u.type && (u.type = "text"), process.env.NODE_ENV !== "production" && w(u.type === "text", 'textContent.type must be "text"');
  }
  var s = e ? n[e] : n.normal;
  s.cfg = o, s.conOpt = m;
}
function F(r, e) {
  return e ? r ? r[e] : null : r;
}
function W(r, e, n) {
  var t = e && e.style;
  return t == null && n === P && r && (t = r.styleEmphasis), t;
}
function gr(r, e, n, t, i) {
  var l = t.children, o = l ? l.length : 0, p = t.$mergeChildren, m = p === "byName" || t.diffChildrenByName, v = p === !1;
  if (!(!o && !m && !v)) {
    if (m) {
      yr({
        api: r,
        oldChildren: e.children() || [],
        newChildren: l || [],
        dataIndex: n,
        seriesModel: i,
        group: e
      });
      return;
    }
    v && e.removeAll();
    for (var u = 0; u < o; u++) {
      var s = l[u], c = e.childAt(u);
      s ? (s.ignore == null && (s.ignore = !1), K(r, c, n, s, i, e)) : (process.env.NODE_ENV !== "production" && w(c, "renderItem should not return a group containing elements as null/undefined/{} if they do not exist before."), c.ignore = !0);
    }
    for (var h = e.childCount() - 1; h >= u; h--) {
      var _ = e.childAt(h);
      hr(e, _, i);
    }
  }
}
function hr(r, e, n) {
  e && H(e, E(r).option, n);
}
function yr(r) {
  new Ge(r.oldChildren, r.newChildren, ve, ve, r).add(me).update(me).remove(Cr).execute();
}
function ve(r, e) {
  var n = r && r.name;
  return n ?? ur + e;
}
function me(r, e) {
  var n = this.context, t = r != null ? n.newChildren[r] : null, i = e != null ? n.oldChildren[e] : null;
  K(n.api, i, n.dataIndex, t, n.seriesModel, n.group);
}
function Cr(r) {
  var e = this.context, n = e.oldChildren[r];
  n && H(n, E(n).option, e.seriesModel);
}
function ge(r) {
  return r && (r.pathData || r.d);
}
function dr(r) {
  return r && (D(r, "pathData") || D(r, "d"));
}
export {
  Jr as default
};
