// zrender@6.0.0/lib/tool/color.js
import O from "../core/LRU.js";
import { isFunction as w, isString as j, isGradientObject as B, extend as G, map as N } from "../core/util.js";
var L = {
  transparent: [0, 0, 0, 0],
  aliceblue: [240, 248, 255, 1],
  antiquewhite: [250, 235, 215, 1],
  aqua: [0, 255, 255, 1],
  aquamarine: [127, 255, 212, 1],
  azure: [240, 255, 255, 1],
  beige: [245, 245, 220, 1],
  bisque: [255, 228, 196, 1],
  black: [0, 0, 0, 1],
  blanchedalmond: [255, 235, 205, 1],
  blue: [0, 0, 255, 1],
  blueviolet: [138, 43, 226, 1],
  brown: [165, 42, 42, 1],
  burlywood: [222, 184, 135, 1],
  cadetblue: [95, 158, 160, 1],
  chartreuse: [127, 255, 0, 1],
  chocolate: [210, 105, 30, 1],
  coral: [255, 127, 80, 1],
  cornflowerblue: [100, 149, 237, 1],
  cornsilk: [255, 248, 220, 1],
  crimson: [220, 20, 60, 1],
  cyan: [0, 255, 255, 1],
  darkblue: [0, 0, 139, 1],
  darkcyan: [0, 139, 139, 1],
  darkgoldenrod: [184, 134, 11, 1],
  darkgray: [169, 169, 169, 1],
  darkgreen: [0, 100, 0, 1],
  darkgrey: [169, 169, 169, 1],
  darkkhaki: [189, 183, 107, 1],
  darkmagenta: [139, 0, 139, 1],
  darkolivegreen: [85, 107, 47, 1],
  darkorange: [255, 140, 0, 1],
  darkorchid: [153, 50, 204, 1],
  darkred: [139, 0, 0, 1],
  darksalmon: [233, 150, 122, 1],
  darkseagreen: [143, 188, 143, 1],
  darkslateblue: [72, 61, 139, 1],
  darkslategray: [47, 79, 79, 1],
  darkslategrey: [47, 79, 79, 1],
  darkturquoise: [0, 206, 209, 1],
  darkviolet: [148, 0, 211, 1],
  deeppink: [255, 20, 147, 1],
  deepskyblue: [0, 191, 255, 1],
  dimgray: [105, 105, 105, 1],
  dimgrey: [105, 105, 105, 1],
  dodgerblue: [30, 144, 255, 1],
  firebrick: [178, 34, 34, 1],
  floralwhite: [255, 250, 240, 1],
  forestgreen: [34, 139, 34, 1],
  fuchsia: [255, 0, 255, 1],
  gainsboro: [220, 220, 220, 1],
  ghostwhite: [248, 248, 255, 1],
  gold: [255, 215, 0, 1],
  goldenrod: [218, 165, 32, 1],
  gray: [128, 128, 128, 1],
  green: [0, 128, 0, 1],
  greenyellow: [173, 255, 47, 1],
  grey: [128, 128, 128, 1],
  honeydew: [240, 255, 240, 1],
  hotpink: [255, 105, 180, 1],
  indianred: [205, 92, 92, 1],
  indigo: [75, 0, 130, 1],
  ivory: [255, 255, 240, 1],
  khaki: [240, 230, 140, 1],
  lavender: [230, 230, 250, 1],
  lavenderblush: [255, 240, 245, 1],
  lawngreen: [124, 252, 0, 1],
  lemonchiffon: [255, 250, 205, 1],
  lightblue: [173, 216, 230, 1],
  lightcoral: [240, 128, 128, 1],
  lightcyan: [224, 255, 255, 1],
  lightgoldenrodyellow: [250, 250, 210, 1],
  lightgray: [211, 211, 211, 1],
  lightgreen: [144, 238, 144, 1],
  lightgrey: [211, 211, 211, 1],
  lightpink: [255, 182, 193, 1],
  lightsalmon: [255, 160, 122, 1],
  lightseagreen: [32, 178, 170, 1],
  lightskyblue: [135, 206, 250, 1],
  lightslategray: [119, 136, 153, 1],
  lightslategrey: [119, 136, 153, 1],
  lightsteelblue: [176, 196, 222, 1],
  lightyellow: [255, 255, 224, 1],
  lime: [0, 255, 0, 1],
  limegreen: [50, 205, 50, 1],
  linen: [250, 240, 230, 1],
  magenta: [255, 0, 255, 1],
  maroon: [128, 0, 0, 1],
  mediumaquamarine: [102, 205, 170, 1],
  mediumblue: [0, 0, 205, 1],
  mediumorchid: [186, 85, 211, 1],
  mediumpurple: [147, 112, 219, 1],
  mediumseagreen: [60, 179, 113, 1],
  mediumslateblue: [123, 104, 238, 1],
  mediumspringgreen: [0, 250, 154, 1],
  mediumturquoise: [72, 209, 204, 1],
  mediumvioletred: [199, 21, 133, 1],
  midnightblue: [25, 25, 112, 1],
  mintcream: [245, 255, 250, 1],
  mistyrose: [255, 228, 225, 1],
  moccasin: [255, 228, 181, 1],
  navajowhite: [255, 222, 173, 1],
  navy: [0, 0, 128, 1],
  oldlace: [253, 245, 230, 1],
  olive: [128, 128, 0, 1],
  olivedrab: [107, 142, 35, 1],
  orange: [255, 165, 0, 1],
  orangered: [255, 69, 0, 1],
  orchid: [218, 112, 214, 1],
  palegoldenrod: [238, 232, 170, 1],
  palegreen: [152, 251, 152, 1],
  paleturquoise: [175, 238, 238, 1],
  palevioletred: [219, 112, 147, 1],
  papayawhip: [255, 239, 213, 1],
  peachpuff: [255, 218, 185, 1],
  peru: [205, 133, 63, 1],
  pink: [255, 192, 203, 1],
  plum: [221, 160, 221, 1],
  powderblue: [176, 224, 230, 1],
  purple: [128, 0, 128, 1],
  red: [255, 0, 0, 1],
  rosybrown: [188, 143, 143, 1],
  royalblue: [65, 105, 225, 1],
  saddlebrown: [139, 69, 19, 1],
  salmon: [250, 128, 114, 1],
  sandybrown: [244, 164, 96, 1],
  seagreen: [46, 139, 87, 1],
  seashell: [255, 245, 238, 1],
  sienna: [160, 82, 45, 1],
  silver: [192, 192, 192, 1],
  skyblue: [135, 206, 235, 1],
  slateblue: [106, 90, 205, 1],
  slategray: [112, 128, 144, 1],
  slategrey: [112, 128, 144, 1],
  snow: [255, 250, 250, 1],
  springgreen: [0, 255, 127, 1],
  steelblue: [70, 130, 180, 1],
  tan: [210, 180, 140, 1],
  teal: [0, 128, 128, 1],
  thistle: [216, 191, 216, 1],
  tomato: [255, 99, 71, 1],
  turquoise: [64, 224, 208, 1],
  violet: [238, 130, 238, 1],
  wheat: [245, 222, 179, 1],
  white: [255, 255, 255, 1],
  whitesmoke: [245, 245, 245, 1],
  yellow: [255, 255, 0, 1],
  yellowgreen: [154, 205, 50, 1]
};
function o(n) {
  return n = Math.round(n), n < 0 ? 0 : n > 255 ? 255 : n;
}
function U(n) {
  return n = Math.round(n), n < 0 ? 0 : n > 360 ? 360 : n;
}
function y(n) {
  return n < 0 ? 0 : n > 1 ? 1 : n;
}
function x(n) {
  var e = n;
  return e.length && e.charAt(e.length - 1) === "%" ? o(parseFloat(e) / 100 * 255) : o(parseInt(e, 10));
}
function v(n) {
  var e = n;
  return e.length && e.charAt(e.length - 1) === "%" ? y(parseFloat(e) / 100) : y(parseFloat(e));
}
function b(n, e, t) {
  return t < 0 ? t += 1 : t > 1 && (t -= 1), t * 6 < 1 ? n + (e - n) * t * 6 : t * 2 < 1 ? e : t * 3 < 2 ? n + (e - n) * (2 / 3 - t) * 6 : n;
}
function h(n, e, t) {
  return n + (e - n) * t;
}
function d(n, e, t, r, i) {
  return n[0] = e, n[1] = t, n[2] = r, n[3] = i, n;
}
function I(n, e) {
  return n[0] = e[0], n[1] = e[1], n[2] = e[2], n[3] = e[3], n;
}
var T = new O(20), k = null;
function g(n, e) {
  k && I(k, e), k = T.put(n, k || e.slice());
}
function c(n, e) {
  if (n) {
    e = e || [];
    var t = T.get(n);
    if (t)
      return I(e, t);
    n = n + "";
    var r = n.replace(/ /g, "").toLowerCase();
    if (r in L)
      return I(e, L[r]), g(n, e), e;
    var i = r.length;
    if (r.charAt(0) === "#") {
      if (i === 4 || i === 5) {
        var l = parseInt(r.slice(1, 4), 16);
        if (!(l >= 0 && l <= 4095)) {
          d(e, 0, 0, 0, 1);
          return;
        }
        return d(e, (l & 3840) >> 4 | (l & 3840) >> 8, l & 240 | (l & 240) >> 4, l & 15 | (l & 15) << 4, i === 5 ? parseInt(r.slice(4), 16) / 15 : 1), g(n, e), e;
      } else if (i === 7 || i === 9) {
        var l = parseInt(r.slice(1, 7), 16);
        if (!(l >= 0 && l <= 16777215)) {
          d(e, 0, 0, 0, 1);
          return;
        }
        return d(e, (l & 16711680) >> 16, (l & 65280) >> 8, l & 255, i === 9 ? parseInt(r.slice(7), 16) / 255 : 1), g(n, e), e;
      }
      return;
    }
    var a = r.indexOf("("), s = r.indexOf(")");
    if (a !== -1 && s + 1 === i) {
      var f = r.substr(0, a), u = r.substr(a + 1, s - (a + 1)).split(","), p = 1;
      switch (f) {
        case "rgba":
          if (u.length !== 4)
            return u.length === 3 ? d(e, +u[0], +u[1], +u[2], 1) : d(e, 0, 0, 0, 1);
          p = v(u.pop());
        case "rgb":
          if (u.length >= 3)
            return d(e, x(u[0]), x(u[1]), x(u[2]), u.length === 3 ? p : v(u[3])), g(n, e), e;
          d(e, 0, 0, 0, 1);
          return;
        case "hsla":
          if (u.length !== 4) {
            d(e, 0, 0, 0, 1);
            return;
          }
          return u[3] = v(u[3]), M(u, e), g(n, e), e;
        case "hsl":
          if (u.length !== 3) {
            d(e, 0, 0, 0, 1);
            return;
          }
          return M(u, e), g(n, e), e;
        default:
          return;
      }
    }
    d(e, 0, 0, 0, 1);
  }
}
function M(n, e) {
  var t = (parseFloat(n[0]) % 360 + 360) % 360 / 360, r = v(n[1]), i = v(n[2]), l = i <= 0.5 ? i * (r + 1) : i + r - i * r, a = i * 2 - l;
  return e = e || [], d(e, o(b(a, l, t + 1 / 3) * 255), o(b(a, l, t) * 255), o(b(a, l, t - 1 / 3) * 255), 1), n.length === 4 && (e[3] = n[3]), e;
}
function D(n) {
  if (n) {
    var e = n[0] / 255, t = n[1] / 255, r = n[2] / 255, i = Math.min(e, t, r), l = Math.max(e, t, r), a = l - i, s = (l + i) / 2, f, u;
    if (a === 0)
      f = 0, u = 0;
    else {
      s < 0.5 ? u = a / (l + i) : u = a / (2 - l - i);
      var p = ((l - e) / 6 + a / 2) / a, q = ((l - t) / 6 + a / 2) / a, C = ((l - r) / 6 + a / 2) / a;
      e === l ? f = C - q : t === l ? f = 1 / 3 + p - C : r === l && (f = 2 / 3 + q - p), f < 0 && (f += 1), f > 1 && (f -= 1);
    }
    var F = [f * 360, u, s];
    return n[3] != null && F.push(n[3]), F;
  }
}
function R(n, e) {
  var t = c(n);
  if (t) {
    for (var r = 0; r < 3; r++)
      e < 0 ? t[r] = t[r] * (1 - e) | 0 : t[r] = (255 - t[r]) * e + t[r] | 0, t[r] > 255 ? t[r] = 255 : t[r] < 0 && (t[r] = 0);
    return m(t, t.length === 4 ? "rgba" : "rgb");
  }
}
function K(n, e, t) {
  if (!(!(e && e.length) || !(n >= 0 && n <= 1))) {
    t = t || [];
    var r = n * (e.length - 1), i = Math.floor(r), l = Math.ceil(r), a = e[i], s = e[l], f = r - i;
    return t[0] = o(h(a[0], s[0], f)), t[1] = o(h(a[1], s[1], f)), t[2] = o(h(a[2], s[2], f)), t[3] = y(h(a[3], s[3], f)), t;
  }
}
function P(n, e, t) {
  if (!(!(e && e.length) || !(n >= 0 && n <= 1))) {
    var r = n * (e.length - 1), i = Math.floor(r), l = Math.ceil(r), a = c(e[i]), s = c(e[l]), f = r - i, u = m([
      o(h(a[0], s[0], f)),
      o(h(a[1], s[1], f)),
      o(h(a[2], s[2], f)),
      y(h(a[3], s[3], f))
    ], "rgba");
    return t ? {
      color: u,
      leftIndex: i,
      rightIndex: l,
      value: r
    } : u;
  }
}
function Q(n, e, t, r) {
  var i = c(n);
  if (n)
    return i = D(i), e != null && (i[0] = U(w(e) ? e(i[0]) : e)), t != null && (i[1] = v(w(t) ? t(i[1]) : t)), r != null && (i[2] = v(w(r) ? r(i[2]) : r)), m(M(i), "rgba");
}
function W(n, e) {
  var t = c(n);
  if (t && e != null)
    return t[3] = y(e), m(t, "rgba");
}
function m(n, e) {
  if (!(!n || !n.length)) {
    var t = n[0] + "," + n[1] + "," + n[2];
    return (e === "rgba" || e === "hsva" || e === "hsla") && (t += "," + n[3]), e + "(" + t + ")";
  }
}
function X(n, e) {
  var t = c(n);
  return t ? (0.299 * t[0] + 0.587 * t[1] + 0.114 * t[2]) * t[3] / 255 + (1 - t[3]) * e : 0;
}
var H = new O(100);
function Y(n) {
  if (j(n)) {
    var e = H.get(n);
    return e || (e = R(n, -0.1), H.put(n, e)), e;
  } else if (B(n)) {
    var t = G({}, n);
    return t.colorStops = N(n.colorStops, function(r) {
      return {
        offset: r.offset,
        color: R(r.color, -0.1)
      };
    }), t;
  }
  return n;
}
export {
  K as fastLerp,
  P as lerp,
  R as lift,
  Y as liftColor,
  X as lum,
  W as modifyAlpha,
  Q as modifyHSL,
  c as parse,
  v as parseCssFloat,
  x as parseCssInt,
  m as stringify
};
