// zrender@6.0.0/lib/tool/convertPath.js
import { cubicSubdivide as z } from "../core/curve.js";
import y from "../core/PathProxy.js";
var C = y.CMD;
function R(d, a) {
  return Math.abs(d - a) < 1e-5;
}
function l(d) {
  var a = d.data, m = d.len(), b = [], o, v = 0, h = 0, u = 0, t = 0;
  function M(q, P) {
    o && o.length > 2 && b.push(o), o = [q, P];
  }
  function c(q, P, B, G) {
    R(q, B) && R(P, G) || o.push(q, P, B, G, B, G);
  }
  function E(q, P, B, G, Y, Z) {
    var i = Math.abs(P - q), H = Math.tan(i / 4) * 4 / 3, J = P < q ? -1 : 1, K = Math.cos(q), O = Math.sin(q), U = Math.cos(P), V = Math.sin(P), g = K * Y + B, x = O * Z + G, W = U * Y + B, $ = V * Z + G, k = Y * H * J, F = Z * H * J;
    o.push(g - k * O, x + F * K, W + k * V, $ - F * U, W, $);
  }
  for (var p, f, L, n, r = 0; r < m; ) {
    var I = a[r++], N = r === 1;
    switch (N && (v = a[r], h = a[r + 1], u = v, t = h, (I === C.L || I === C.C || I === C.Q) && (o = [u, t])), I) {
      case C.M:
        v = u = a[r++], h = t = a[r++], M(u, t);
        break;
      case C.L:
        p = a[r++], f = a[r++], c(v, h, p, f), v = p, h = f;
        break;
      case C.C:
        o.push(a[r++], a[r++], a[r++], a[r++], v = a[r++], h = a[r++]);
        break;
      case C.Q:
        p = a[r++], f = a[r++], L = a[r++], n = a[r++], o.push(v + 2 / 3 * (p - v), h + 2 / 3 * (f - h), L + 2 / 3 * (p - L), n + 2 / 3 * (f - n), L, n), v = L, h = n;
        break;
      case C.A:
        var Q = a[r++], X = a[r++], T = a[r++], e = a[r++], _ = a[r++], D = a[r++] + _;
        r += 1;
        var s = !a[r++];
        p = Math.cos(_) * T + Q, f = Math.sin(_) * e + X, N ? (u = p, t = f, M(u, t)) : c(v, h, p, f), v = Math.cos(D) * T + Q, h = Math.sin(D) * e + X;
        for (var S = (s ? -1 : 1) * Math.PI / 2, j = _; s ? j > D : j < D; j += S) {
          var A = s ? Math.max(j + S, D) : Math.min(j + S, D);
          E(j, A, Q, X, T, e);
        }
        break;
      case C.R:
        u = v = a[r++], t = h = a[r++], p = u + a[r++], f = t + a[r++], M(p, t), c(p, t, p, f), c(p, f, u, f), c(u, f, u, t), c(u, t, p, t);
        break;
      case C.Z:
        o && c(v, h, u, t), v = u, h = t;
        break;
    }
  }
  return o && o.length > 2 && b.push(o), b;
}
function w(d, a, m, b, o, v, h, u, t, M) {
  if (R(d, m) && R(a, b) && R(o, h) && R(v, u)) {
    t.push(h, u);
    return;
  }
  var c = 2 / M, E = c * c, p = h - d, f = u - a, L = Math.sqrt(p * p + f * f);
  p /= L, f /= L;
  var n = m - d, r = b - a, I = o - h, N = v - u, Q = n * n + r * r, X = I * I + N * N;
  if (Q < E && X < E) {
    t.push(h, u);
    return;
  }
  var T = p * n + f * r, e = -p * I - f * N, _ = Q - T * T, D = X - e * e;
  if (_ < E && T >= 0 && D < E && e >= 0) {
    t.push(h, u);
    return;
  }
  var s = [], S = [];
  z(d, m, o, h, 0.5, s), z(a, b, v, u, 0.5, S), w(s[0], S[0], s[1], S[1], s[2], S[2], s[3], S[3], t, M), w(s[4], S[4], s[5], S[5], s[6], S[6], s[7], S[7], t, M);
}
function va(d, a) {
  var m = l(d), b = [];
  a = a || 1;
  for (var o = 0; o < m.length; o++) {
    var v = m[o], h = [], u = v[0], t = v[1];
    h.push(u, t);
    for (var M = 2; M < v.length; ) {
      var c = v[M++], E = v[M++], p = v[M++], f = v[M++], L = v[M++], n = v[M++];
      w(u, t, c, E, p, f, L, n, h, a), u = L, t = n;
    }
    b.push(h);
  }
  return b;
}
export {
  l as pathToBezierCurves,
  va as pathToPolygons
};
