// zrender@6.0.0/lib/tool/dividePath.js
import { fromPoints as L } from "../core/bbox.js";
import q from "../core/BoundingRect.js";
import P from "../core/Point.js";
import { map as R } from "../core/util.js";
import I from "../graphic/shape/Polygon.js";
import O from "../graphic/shape/Rect.js";
import k from "../graphic/shape/Sector.js";
import { pathToPolygons as B } from "./convertPath.js";
import { clonePath as C } from "./path.js";
function T(r, t, v) {
  var n = r[t], e = r[1 - t], a = Math.abs(n / e), d = Math.ceil(Math.sqrt(a * v)), l = Math.floor(v / d);
  l === 0 && (l = 1, d = v);
  for (var s = [], i = 0; i < d; i++)
    s.push(l);
  var f = d * l, h = v - f;
  if (h > 0)
    for (var i = 0; i < h; i++)
      s[i % d] += 1;
  return s;
}
function p(r, t, v) {
  for (var n = r.r0, e = r.r, a = r.startAngle, d = r.endAngle, l = Math.abs(d - a), s = l * e, i = e - n, f = s > Math.abs(i), h = T([s, i], f ? 0 : 1, t), g = (f ? l : i) / h.length, u = 0; u < h.length; u++)
    for (var x = (f ? i : l) / h[u], o = 0; o < h[u]; o++) {
      var y = {};
      f ? (y.startAngle = a + g * u, y.endAngle = a + g * (u + 1), y.r0 = n + x * o, y.r = n + x * (o + 1)) : (y.startAngle = a + x * o, y.endAngle = a + x * (o + 1), y.r0 = n + g * u, y.r = n + g * (u + 1)), y.clockwise = r.clockwise, y.cx = r.cx, y.cy = r.cy, v.push(y);
    }
}
function G(r, t, v) {
  for (var n = r.width, e = r.height, a = n > e, d = T([n, e], a ? 0 : 1, t), l = a ? "width" : "height", s = a ? "height" : "width", i = a ? "x" : "y", f = a ? "y" : "x", h = r[l] / d.length, g = 0; g < d.length; g++)
    for (var u = r[s] / d[g], x = 0; x < d[g]; x++) {
      var o = {};
      o[i] = g * h, o[f] = x * u, o[l] = h, o[s] = u, o.x += r.x, o.y += r.y, v.push(o);
    }
}
function D(r, t, v, n) {
  return r * n - v * t;
}
function H(r, t, v, n, e, a, d, l) {
  var s = v - r, i = n - t, f = d - e, h = l - a, g = D(f, h, s, i);
  if (Math.abs(g) < 1e-6)
    return null;
  var u = r - e, x = t - a, o = D(u, x, f, h) / g;
  return o < 0 || o > 1 ? null : new P(o * s + r, o * i + t);
}
function U(r, t, v) {
  var n = new P();
  P.sub(n, v, t), n.normalize();
  var e = new P();
  P.sub(e, r, t);
  var a = e.dot(n);
  return a;
}
function m(r, t) {
  var v = r[r.length - 1];
  v && v[0] === t[0] && v[1] === t[1] || r.push(t);
}
function E(r, t, v) {
  for (var n = r.length, e = [], a = 0; a < n; a++) {
    var d = r[a], l = r[(a + 1) % n], s = H(d[0], d[1], l[0], l[1], t.x, t.y, v.x, v.y);
    s && e.push({
      projPt: U(s, t, v),
      pt: s,
      idx: a
    });
  }
  if (e.length < 2)
    return [{ points: r }, { points: r }];
  e.sort(function(y, c) {
    return y.projPt - c.projPt;
  });
  var i = e[0], f = e[e.length - 1];
  if (f.idx < i.idx) {
    var h = i;
    i = f, f = h;
  }
  for (var g = [i.pt.x, i.pt.y], u = [f.pt.x, f.pt.y], x = [g], o = [u], a = i.idx + 1; a <= f.idx; a++)
    m(x, r[a].slice());
  m(x, u), m(x, g);
  for (var a = f.idx + 1; a <= i.idx + n; a++)
    m(o, r[a % n].slice());
  return m(o, g), m(o, u), [{
    points: x
  }, {
    points: o
  }];
}
function j(r) {
  var t = r.points, v = [], n = [];
  L(t, v, n);
  var e = new q(v[0], v[1], n[0] - v[0], n[1] - v[1]), a = e.width, d = e.height, l = e.x, s = e.y, i = new P(), f = new P();
  return a > d ? (i.x = f.x = l + a / 2, i.y = s, f.y = s + d) : (i.y = f.y = s + d / 2, i.x = l, f.x = l + a), E(t, i, f);
}
function b(r, t, v, n) {
  if (v === 1)
    n.push(t);
  else {
    var e = Math.floor(v / 2), a = r(t);
    b(r, a[0], e, n), b(r, a[1], v - e, n);
  }
  return n;
}
function F(r, t) {
  for (var v = [], n = 0; n < t; n++)
    v.push(C(r));
  return v;
}
function J(r, t) {
  t.setStyle(r.style), t.z = r.z, t.z2 = r.z2, t.zlevel = r.zlevel;
}
function K(r) {
  for (var t = [], v = 0; v < r.length; )
    t.push([r[v++], r[v++]]);
  return t;
}
function w(r, t) {
  var v = [], n = r.shape, e;
  switch (r.type) {
    case "rect":
      G(n, t, v), e = O;
      break;
    case "sector":
      p(n, t, v), e = k;
      break;
    case "circle":
      p({
        r0: 0,
        r: n.r,
        startAngle: 0,
        endAngle: Math.PI * 2,
        cx: n.cx,
        cy: n.cy
      }, t, v), e = k;
      break;
    default:
      var a = r.getComputedTransform(), d = a ? Math.sqrt(Math.max(a[0] * a[0] + a[1] * a[1], a[2] * a[2] + a[3] * a[3])) : 1, l = R(B(r.getUpdatedPathProxy(), d), function(c) {
        return K(c);
      }), s = l.length;
      if (s === 0)
        b(j, {
          points: l[0]
        }, t, v);
      else if (s === t)
        for (var i = 0; i < s; i++)
          v.push({
            points: l[i]
          });
      else {
        var f = 0, h = R(l, function(c) {
          var M = [], A = [];
          L(c, M, A);
          var z = (A[1] - M[1]) * (A[0] - M[0]);
          return f += z, { poly: c, area: z };
        });
        h.sort(function(c, M) {
          return M.area - c.area;
        });
        for (var g = t, i = 0; i < s; i++) {
          var u = h[i];
          if (g <= 0)
            break;
          var x = i === s - 1 ? g : Math.ceil(u.area / f * t);
          x < 0 || (b(j, {
            points: u.poly
          }, x, v), g -= x);
        }
      }
      e = I;
      break;
  }
  if (!e)
    return F(r, t);
  for (var o = [], i = 0; i < v.length; i++) {
    var y = new e();
    y.setShape(v[i]), J(r, y), o.push(y);
  }
  return o;
}
export {
  F as clone,
  w as split
};
