// zrender@6.0.0/lib/tool/parseSVG.js
import b from "../graphic/Group.js";
import G from "../graphic/Image.js";
import W from "../graphic/shape/Circle.js";
import P from "../graphic/shape/Rect.js";
import V from "../graphic/shape/Ellipse.js";
import j from "../graphic/shape/Line.js";
import H from "../graphic/shape/Polygon.js";
import J from "../graphic/shape/Polyline.js";
import { mul as E, rotate as K, scale as m, translate as q, create as Z } from "../core/matrix.js";
import { createFromString as z } from "./path.js";
import { keys as x, hasOwn as _, defaults as Q, each as T, map as $, trim as B } from "../core/util.js";
import N from "../graphic/LinearGradient.js";
import tt from "../graphic/RadialGradient.js";
import et from "../graphic/TSpan.js";
import { parseXML as rt } from "./parseXML.js";
import { parse as it, parseCssFloat as at, stringify as st } from "./color.js";
var w, d = {
  fill: "fill",
  stroke: "stroke",
  "stroke-width": "lineWidth",
  opacity: "opacity",
  "fill-opacity": "fillOpacity",
  "stroke-opacity": "strokeOpacity",
  "stroke-dasharray": "lineDash",
  "stroke-dashoffset": "lineDashOffset",
  "stroke-linecap": "lineCap",
  "stroke-linejoin": "lineJoin",
  "stroke-miterlimit": "miterLimit",
  "font-family": "fontFamily",
  "font-size": "fontSize",
  "font-style": "fontStyle",
  "font-weight": "fontWeight",
  "text-anchor": "textAlign",
  visibility: "visibility",
  display: "display"
}, U = x(d), F = {
  "alignment-baseline": "textBaseline",
  "stop-color": "stopColor"
}, L = x(F), nt = (function() {
  function s() {
    this._defs = {}, this._root = null;
  }
  return s.prototype.parse = function(e, i) {
    i = i || {};
    var t = rt(e);
    if (process.env.NODE_ENV !== "production" && !t)
      throw new Error("Illegal svg");
    this._defsUsePending = [];
    var r = new b();
    this._root = r;
    var a = [], f = t.getAttribute("viewBox") || "", l = parseFloat(t.getAttribute("width") || i.width), n = parseFloat(t.getAttribute("height") || i.height);
    isNaN(l) && (l = null), isNaN(n) && (n = null), o(t, r, null, !0, !1);
    for (var p = t.firstChild; p; )
      this._parseNode(p, r, a, null, !1, !1), p = p.nextSibling;
    ot(this._defs, this._defsUsePending), this._defsUsePending = [];
    var v, y;
    if (f) {
      var h = S(f);
      h.length >= 4 && (v = {
        x: parseFloat(h[0] || 0),
        y: parseFloat(h[1] || 0),
        width: parseFloat(h[2]),
        height: parseFloat(h[3])
      });
    }
    if (v && l != null && n != null && (y = vt(v, { x: 0, y: 0, width: l, height: n }), !i.ignoreViewBox)) {
      var g = r;
      r = new b(), r.add(g), g.scaleX = g.scaleY = y.scale, g.x = y.x, g.y = y.y;
    }
    return !i.ignoreRootClip && l != null && n != null && r.setClipPath(new P({
      shape: { x: 0, y: 0, width: l, height: n }
    })), {
      root: r,
      width: l,
      height: n,
      viewBoxRect: v,
      viewBoxTransform: y,
      named: a
    };
  }, s.prototype._parseNode = function(e, i, t, r, a, f) {
    var l = e.nodeName.toLowerCase(), n, p = r;
    if (l === "defs" && (a = !0), l === "text" && (f = !0), l === "defs" || l === "switch")
      n = i;
    else {
      if (!a) {
        var v = w[l];
        if (v && _(w, l)) {
          n = v.call(this, e, i);
          var y = e.getAttribute("name");
          if (y) {
            var h = {
              name: y,
              namedFrom: null,
              svgNodeTagLower: l,
              el: n
            };
            t.push(h), l === "g" && (p = h);
          } else r && t.push({
            name: r.name,
            namedFrom: r,
            svgNodeTagLower: l,
            el: n
          });
          i.add(n);
        }
      }
      var g = I[l];
      if (g && _(I, l)) {
        var R = g.call(this, e), k = e.getAttribute("id");
        k && (this._defs[k] = R);
      }
    }
    if (n && n.isGroup)
      for (var c = e.firstChild; c; )
        c.nodeType === 1 ? this._parseNode(c, n, t, p, a, f) : c.nodeType === 3 && f && this._parseText(c, n), c = c.nextSibling;
  }, s.prototype._parseText = function(e, i) {
    var t = new et({
      style: {
        text: e.textContent
      },
      silent: !0,
      x: this._textX || 0,
      y: this._textY || 0
    });
    u(i, t), o(e, t, this._defsUsePending, !1, !1), lt(t, i);
    var r = t.style, a = r.fontSize;
    a && a < 9 && (r.fontSize = 9, t.scaleX *= a / 9, t.scaleY *= a / 9);
    var f = (r.fontSize || r.fontFamily) && [
      r.fontStyle,
      r.fontWeight,
      (r.fontSize || 12) + "px",
      r.fontFamily || "sans-serif"
    ].join(" ");
    r.font = f;
    var l = t.getBoundingRect();
    return this._textX += l.width, i.add(t), t;
  }, s.internalField = (function() {
    w = {
      g: function(e, i) {
        var t = new b();
        return u(i, t), o(e, t, this._defsUsePending, !1, !1), t;
      },
      rect: function(e, i) {
        var t = new P();
        return u(i, t), o(e, t, this._defsUsePending, !1, !1), t.setShape({
          x: parseFloat(e.getAttribute("x") || "0"),
          y: parseFloat(e.getAttribute("y") || "0"),
          width: parseFloat(e.getAttribute("width") || "0"),
          height: parseFloat(e.getAttribute("height") || "0")
        }), t.silent = !0, t;
      },
      circle: function(e, i) {
        var t = new W();
        return u(i, t), o(e, t, this._defsUsePending, !1, !1), t.setShape({
          cx: parseFloat(e.getAttribute("cx") || "0"),
          cy: parseFloat(e.getAttribute("cy") || "0"),
          r: parseFloat(e.getAttribute("r") || "0")
        }), t.silent = !0, t;
      },
      line: function(e, i) {
        var t = new j();
        return u(i, t), o(e, t, this._defsUsePending, !1, !1), t.setShape({
          x1: parseFloat(e.getAttribute("x1") || "0"),
          y1: parseFloat(e.getAttribute("y1") || "0"),
          x2: parseFloat(e.getAttribute("x2") || "0"),
          y2: parseFloat(e.getAttribute("y2") || "0")
        }), t.silent = !0, t;
      },
      ellipse: function(e, i) {
        var t = new V();
        return u(i, t), o(e, t, this._defsUsePending, !1, !1), t.setShape({
          cx: parseFloat(e.getAttribute("cx") || "0"),
          cy: parseFloat(e.getAttribute("cy") || "0"),
          rx: parseFloat(e.getAttribute("rx") || "0"),
          ry: parseFloat(e.getAttribute("ry") || "0")
        }), t.silent = !0, t;
      },
      polygon: function(e, i) {
        var t = e.getAttribute("points"), r;
        t && (r = X(t));
        var a = new H({
          shape: {
            points: r || []
          },
          silent: !0
        });
        return u(i, a), o(e, a, this._defsUsePending, !1, !1), a;
      },
      polyline: function(e, i) {
        var t = e.getAttribute("points"), r;
        t && (r = X(t));
        var a = new J({
          shape: {
            points: r || []
          },
          silent: !0
        });
        return u(i, a), o(e, a, this._defsUsePending, !1, !1), a;
      },
      image: function(e, i) {
        var t = new G();
        return u(i, t), o(e, t, this._defsUsePending, !1, !1), t.setStyle({
          image: e.getAttribute("xlink:href") || e.getAttribute("href"),
          x: +e.getAttribute("x"),
          y: +e.getAttribute("y"),
          width: +e.getAttribute("width"),
          height: +e.getAttribute("height")
        }), t.silent = !0, t;
      },
      text: function(e, i) {
        var t = e.getAttribute("x") || "0", r = e.getAttribute("y") || "0", a = e.getAttribute("dx") || "0", f = e.getAttribute("dy") || "0";
        this._textX = parseFloat(t) + parseFloat(a), this._textY = parseFloat(r) + parseFloat(f);
        var l = new b();
        return u(i, l), o(e, l, this._defsUsePending, !1, !0), l;
      },
      tspan: function(e, i) {
        var t = e.getAttribute("x"), r = e.getAttribute("y");
        t != null && (this._textX = parseFloat(t)), r != null && (this._textY = parseFloat(r));
        var a = e.getAttribute("dx") || "0", f = e.getAttribute("dy") || "0", l = new b();
        return u(i, l), o(e, l, this._defsUsePending, !1, !0), this._textX += parseFloat(a), this._textY += parseFloat(f), l;
      },
      path: function(e, i) {
        var t = e.getAttribute("d") || "", r = z(t);
        return u(i, r), o(e, r, this._defsUsePending, !1, !1), r.silent = !0, r;
      }
    };
  })(), s;
})(), I = {
  lineargradient: function(s) {
    var e = parseInt(s.getAttribute("x1") || "0", 10), i = parseInt(s.getAttribute("y1") || "0", 10), t = parseInt(s.getAttribute("x2") || "10", 10), r = parseInt(s.getAttribute("y2") || "0", 10), a = new N(e, i, t, r);
    return Y(s, a), C(s, a), a;
  },
  radialgradient: function(s) {
    var e = parseInt(s.getAttribute("cx") || "0", 10), i = parseInt(s.getAttribute("cy") || "0", 10), t = parseInt(s.getAttribute("r") || "0", 10), r = new tt(e, i, t);
    return Y(s, r), C(s, r), r;
  }
};
function Y(s, e) {
  var i = s.getAttribute("gradientUnits");
  i === "userSpaceOnUse" && (e.global = !0);
}
function C(s, e) {
  for (var i = s.firstChild; i; ) {
    if (i.nodeType === 1 && i.nodeName.toLocaleLowerCase() === "stop") {
      var t = i.getAttribute("offset"), r = void 0;
      t && t.indexOf("%") > 0 ? r = parseInt(t, 10) / 100 : t ? r = parseFloat(t) : r = 0;
      var a = {};
      D(i, a, a);
      var f = a.stopColor || i.getAttribute("stop-color") || "#000000", l = a.stopOpacity || i.getAttribute("stop-opacity");
      if (l) {
        var n = it(f), p = n && n[3];
        p && (n[3] *= at(l), f = st(n, "rgba"));
      }
      e.colorStops.push({
        offset: r,
        color: f
      });
    }
    i = i.nextSibling;
  }
}
function u(s, e) {
  s && s.__inheritedStyle && (e.__inheritedStyle || (e.__inheritedStyle = {}), Q(e.__inheritedStyle, s.__inheritedStyle));
}
function X(s) {
  for (var e = S(s), i = [], t = 0; t < e.length; t += 2) {
    var r = parseFloat(e[t]), a = parseFloat(e[t + 1]);
    i.push([r, a]);
  }
  return i;
}
function o(s, e, i, t, r) {
  var a = e, f = a.__inheritedStyle = a.__inheritedStyle || {}, l = {};
  s.nodeType === 1 && (ht(s, e), D(s, f, l), t || gt(s, f, l)), a.style = a.style || {}, f.fill != null && (a.style.fill = M(a, "fill", f.fill, i)), f.stroke != null && (a.style.stroke = M(a, "stroke", f.stroke, i)), T([
    "lineWidth",
    "opacity",
    "fillOpacity",
    "strokeOpacity",
    "miterLimit",
    "fontSize"
  ], function(n) {
    f[n] != null && (a.style[n] = parseFloat(f[n]));
  }), T([
    "lineDashOffset",
    "lineCap",
    "lineJoin",
    "fontWeight",
    "fontFamily",
    "fontStyle",
    "textAlign"
  ], function(n) {
    f[n] != null && (a.style[n] = f[n]);
  }), r && (a.__selfStyle = l), f.lineDash && (a.style.lineDash = $(S(f.lineDash), function(n) {
    return parseFloat(n);
  })), (f.visibility === "hidden" || f.visibility === "collapse") && (a.invisible = !0), f.display === "none" && (a.ignore = !0);
}
function lt(s, e) {
  var i = e.__selfStyle;
  if (i) {
    var t = i.textBaseline, r = t;
    !t || t === "auto" || t === "baseline" ? r = "alphabetic" : t === "before-edge" || t === "text-before-edge" ? r = "top" : t === "after-edge" || t === "text-after-edge" ? r = "bottom" : (t === "central" || t === "mathematical") && (r = "middle"), s.style.textBaseline = r;
  }
  var a = e.__inheritedStyle;
  if (a) {
    var f = a.textAlign, l = f;
    f && (f === "middle" && (l = "center"), s.style.textAlign = l);
  }
}
var ft = /^url\(\s*#(.*?)\)/;
function M(s, e, i, t) {
  var r = i && i.match(ft);
  if (r) {
    var a = B(r[1]);
    t.push([s, e, a]);
    return;
  }
  return i === "none" && (i = null), i;
}
function ot(s, e) {
  for (var i = 0; i < e.length; i++) {
    var t = e[i];
    t[0].style[t[1]] = s[t[2]];
  }
}
var ut = /-?([0-9]*\.)?[0-9]+([eE]-?[0-9]+)?/g;
function S(s) {
  return s.match(ut) || [];
}
var pt = /(translate|scale|rotate|skewX|skewY|matrix)\(([\-\s0-9\.eE,]*)\)/g, A = Math.PI / 180;
function ht(s, e) {
  var i = s.getAttribute("transform");
  if (i) {
    i = i.replace(/,/g, " ");
    var t = [], r = null;
    i.replace(pt, function(y, h, g) {
      return t.push(h, g), "";
    });
    for (var a = t.length - 1; a > 0; a -= 2) {
      var f = t[a], l = t[a - 1], n = S(f);
      switch (r = r || Z(), l) {
        case "translate":
          q(r, r, [parseFloat(n[0]), parseFloat(n[1] || "0")]);
          break;
        case "scale":
          m(r, r, [parseFloat(n[0]), parseFloat(n[1] || n[0])]);
          break;
        case "rotate":
          K(r, r, -parseFloat(n[0]) * A, [
            parseFloat(n[1] || "0"),
            parseFloat(n[2] || "0")
          ]);
          break;
        case "skewX":
          var p = Math.tan(parseFloat(n[0]) * A);
          E(r, [1, 0, p, 1, 0, 0], r);
          break;
        case "skewY":
          var v = Math.tan(parseFloat(n[0]) * A);
          E(r, [1, v, 0, 1, 0, 0], r);
          break;
        case "matrix":
          r[0] = parseFloat(n[0]), r[1] = parseFloat(n[1]), r[2] = parseFloat(n[2]), r[3] = parseFloat(n[3]), r[4] = parseFloat(n[4]), r[5] = parseFloat(n[5]);
          break;
      }
    }
    e.setLocalTransform(r);
  }
}
var O = /([^\s:;]+)\s*:\s*([^:;]+)/g;
function D(s, e, i) {
  var t = s.getAttribute("style");
  if (t) {
    O.lastIndex = 0;
    for (var r; (r = O.exec(t)) != null; ) {
      var a = r[1], f = _(d, a) ? d[a] : null;
      f && (e[f] = r[2]);
      var l = _(F, a) ? F[a] : null;
      l && (i[l] = r[2]);
    }
  }
}
function gt(s, e, i) {
  for (var t = 0; t < U.length; t++) {
    var r = U[t], a = s.getAttribute(r);
    a != null && (e[d[r]] = a);
  }
  for (var t = 0; t < L.length; t++) {
    var r = L[t], a = s.getAttribute(r);
    a != null && (i[F[r]] = a);
  }
}
function vt(s, e) {
  var i = e.width / s.width, t = e.height / s.height, r = Math.min(i, t);
  return {
    scale: r,
    x: -(s.x + s.width / 2) * r + (e.x + e.width / 2),
    y: -(s.y + s.height / 2) * r + (e.y + e.height / 2)
  };
}
function Yt(s, e) {
  var i = new nt();
  return i.parse(s, e);
}
export {
  vt as makeViewBoxTransform,
  Yt as parseSVG,
  rt as parseXML
};
