// zrender@6.0.0/lib/tool/path.js
import { __extends as U } from "../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import Z from "../graphic/Path.js";
import m from "../core/PathProxy.js";
import X from "./transformPath.js";
import { extend as J } from "../core/util.js";
var G = Math.sqrt, R = Math.sin, V = Math.cos, F = Math.PI;
function B(s) {
  return Math.sqrt(s[0] * s[0] + s[1] * s[1]);
}
function I(s, t) {
  return (s[0] * t[0] + s[1] * t[1]) / (B(s) * B(t));
}
function H(s, t) {
  return (s[0] * t[1] < s[1] * t[0] ? -1 : 1) * Math.acos(I(s, t));
}
function P(s, t, a, n, b, g, u, v, L, A, O) {
  var d = L * (F / 180), i = V(d) * (s - a) / 2 + R(d) * (t - n) / 2, r = -1 * R(d) * (s - a) / 2 + V(d) * (t - n) / 2, Q = i * i / (u * u) + r * r / (v * v);
  Q > 1 && (u *= G(Q), v *= G(Q));
  var z = (b === g ? -1 : 1) * G((u * u * (v * v) - u * u * (r * r) - v * v * (i * i)) / (u * u * (r * r) + v * v * (i * i))) || 0, e = z * u * r / v, h = z * -v * i / u, k = (s + a) / 2 + V(d) * e - R(d) * h, q = (t + n) / 2 + R(d) * e + V(d) * h, w = H([1, 0], [(i - e) / u, (r - h) / v]), S = [(i - e) / u, (r - h) / v], D = [(-1 * i - e) / u, (-1 * r - h) / v], T = H(S, D);
  if (I(S, D) <= -1 && (T = F), I(S, D) >= 1 && (T = 0), T < 0) {
    var C = Math.round(T / F * 1e6) / 1e6;
    T = F * 2 + C % 2 * F;
  }
  O.addData(A, k, q, u, v, w, T, d, g);
}
var K = /([mlvhzcqtsa])([^mlvhzcqtsa]*)/ig, N = /-?([0-9]*\.)?[0-9]+([eE]-?[0-9]+)?/g;
function W(s) {
  var t = new m();
  if (!s)
    return t;
  var a = 0, n = 0, b = a, g = n, u, v = m.CMD, L = s.match(K);
  if (!L)
    return t;
  for (var A = 0; A < L.length; A++) {
    for (var O = L[A], d = O.charAt(0), i = void 0, r = O.match(N) || [], Q = r.length, z = 0; z < Q; z++)
      r[z] = parseFloat(r[z]);
    for (var e = 0; e < Q; ) {
      var h = void 0, k = void 0, q = void 0, w = void 0, S = void 0, D = void 0, T = void 0, C = a, l = n, o = void 0, M = void 0;
      switch (d) {
        case "l":
          a += r[e++], n += r[e++], i = v.L, t.addData(i, a, n);
          break;
        case "L":
          a = r[e++], n = r[e++], i = v.L, t.addData(i, a, n);
          break;
        case "m":
          a += r[e++], n += r[e++], i = v.M, t.addData(i, a, n), b = a, g = n, d = "l";
          break;
        case "M":
          a = r[e++], n = r[e++], i = v.M, t.addData(i, a, n), b = a, g = n, d = "L";
          break;
        case "h":
          a += r[e++], i = v.L, t.addData(i, a, n);
          break;
        case "H":
          a = r[e++], i = v.L, t.addData(i, a, n);
          break;
        case "v":
          n += r[e++], i = v.L, t.addData(i, a, n);
          break;
        case "V":
          n = r[e++], i = v.L, t.addData(i, a, n);
          break;
        case "C":
          i = v.C, t.addData(i, r[e++], r[e++], r[e++], r[e++], r[e++], r[e++]), a = r[e - 2], n = r[e - 1];
          break;
        case "c":
          i = v.C, t.addData(i, r[e++] + a, r[e++] + n, r[e++] + a, r[e++] + n, r[e++] + a, r[e++] + n), a += r[e - 2], n += r[e - 1];
          break;
        case "S":
          h = a, k = n, o = t.len(), M = t.data, u === v.C && (h += a - M[o - 4], k += n - M[o - 3]), i = v.C, C = r[e++], l = r[e++], a = r[e++], n = r[e++], t.addData(i, h, k, C, l, a, n);
          break;
        case "s":
          h = a, k = n, o = t.len(), M = t.data, u === v.C && (h += a - M[o - 4], k += n - M[o - 3]), i = v.C, C = a + r[e++], l = n + r[e++], a += r[e++], n += r[e++], t.addData(i, h, k, C, l, a, n);
          break;
        case "Q":
          C = r[e++], l = r[e++], a = r[e++], n = r[e++], i = v.Q, t.addData(i, C, l, a, n);
          break;
        case "q":
          C = r[e++] + a, l = r[e++] + n, a += r[e++], n += r[e++], i = v.Q, t.addData(i, C, l, a, n);
          break;
        case "T":
          h = a, k = n, o = t.len(), M = t.data, u === v.Q && (h += a - M[o - 4], k += n - M[o - 3]), a = r[e++], n = r[e++], i = v.Q, t.addData(i, h, k, a, n);
          break;
        case "t":
          h = a, k = n, o = t.len(), M = t.data, u === v.Q && (h += a - M[o - 4], k += n - M[o - 3]), a += r[e++], n += r[e++], i = v.Q, t.addData(i, h, k, a, n);
          break;
        case "A":
          q = r[e++], w = r[e++], S = r[e++], D = r[e++], T = r[e++], C = a, l = n, a = r[e++], n = r[e++], i = v.A, P(C, l, a, n, D, T, q, w, S, i, t);
          break;
        case "a":
          q = r[e++], w = r[e++], S = r[e++], D = r[e++], T = r[e++], C = a, l = n, a += r[e++], n += r[e++], i = v.A, P(C, l, a, n, D, T, q, w, S, i, t);
          break;
      }
    }
    (d === "z" || d === "Z") && (i = v.Z, t.addData(i), a = b, n = g), u = i;
  }
  return t.toStatic(), t;
}
var Y = (function(s) {
  U(t, s);
  function t() {
    return s !== null && s.apply(this, arguments) || this;
  }
  return t.prototype.applyTransform = function(a) {
  }, t;
})(Z);
function j(s) {
  return s.setData != null;
}
function E(s, t) {
  var a = W(s), n = J({}, t);
  return n.buildPath = function(b) {
    var g = j(b);
    if (g && b.canSave()) {
      b.appendPath(a);
      var u = b.getContext();
      u && b.rebuildPath(u, 1);
    } else {
      var u = g ? b.getContext() : b;
      u && a.rebuildPath(u, 1);
    }
  }, n.applyTransform = function(b) {
    X(a, b), this.dirtyShape();
  }, n;
}
function p(s, t) {
  return new Y(E(s, t));
}
function f(s, t) {
  var a = E(s, t), n = (function(b) {
    U(g, b);
    function g(u) {
      var v = b.call(this, u) || this;
      return v.applyTransform = a.applyTransform, v.buildPath = a.buildPath, v;
    }
    return g;
  })(Y);
  return n;
}
function aa(s, t) {
  for (var a = [], n = s.length, b = 0; b < n; b++) {
    var g = s[b];
    a.push(g.getUpdatedPathProxy(!0));
  }
  var u = new Z(t);
  return u.createPathProxy(), u.buildPath = function(v) {
    if (j(v)) {
      v.appendPath(a);
      var L = v.getContext();
      L && v.rebuildPath(L, 1);
    }
  }, u;
}
function ra(s, t) {
  t = t || {};
  var a = new Z();
  return s.shape && a.setShape(s.shape), a.setStyle(s.style), t.bakeTransform ? X(a.path, s.getComputedTransform()) : t.toLocal ? a.setLocalTransform(s.getComputedTransform()) : a.copyTransform(s), a.buildPath = s.buildPath, a.applyTransform = a.applyTransform, a.z = s.z, a.z2 = s.z2, a.zlevel = s.zlevel, a;
}
export {
  ra as clonePath,
  p as createFromString,
  f as extendFromString,
  aa as mergePath
};
