// zrender@6.0.0/lib/core/bbox.js
import { min as H, max as J, create as j } from "./vector.js";
import { quadraticAt as K, cubicExtrema as N, cubicAt as O, quadraticExtremum as R } from "./curve.js";
var b = Math.min, I = Math.max, Q = Math.sin, S = Math.cos, p = Math.PI * 2, C = j(), D = j(), L = j();
function V(t, c, u) {
  if (t.length !== 0) {
    for (var a = t[0], r = a[0], v = a[0], M = a[1], i = a[1], f = 1; f < t.length; f++)
      a = t[f], r = b(r, a[0]), v = I(v, a[0]), M = b(M, a[1]), i = I(i, a[1]);
    c[0] = r, c[1] = M, u[0] = v, u[1] = i;
  }
}
function W(t, c, u, a, r, v) {
  r[0] = b(t, u), r[1] = b(c, a), v[0] = I(t, u), v[1] = I(c, a);
}
var F = [], G = [];
function X(t, c, u, a, r, v, M, i, f, o) {
  var E = N, P = O, q = E(t, u, r, M, F);
  f[0] = 1 / 0, f[1] = 1 / 0, o[0] = -1 / 0, o[1] = -1 / 0;
  for (var h = 0; h < q; h++) {
    var z = P(t, u, r, M, F[h]);
    f[0] = b(z, f[0]), o[0] = I(z, o[0]);
  }
  q = E(c, a, v, i, G);
  for (var h = 0; h < q; h++) {
    var B = P(c, a, v, i, G[h]);
    f[1] = b(B, f[1]), o[1] = I(B, o[1]);
  }
  f[0] = b(t, f[0]), o[0] = I(t, o[0]), f[0] = b(M, f[0]), o[0] = I(M, o[0]), f[1] = b(c, f[1]), o[1] = I(c, o[1]), f[1] = b(i, f[1]), o[1] = I(i, o[1]);
}
function Y(t, c, u, a, r, v, M, i) {
  var f = R, o = K, E = I(b(f(t, u, r), 1), 0), P = I(b(f(c, a, v), 1), 0), q = o(t, u, r, E), h = o(c, a, v, P);
  M[0] = b(t, r, q), M[1] = b(c, v, h), i[0] = I(t, r, q), i[1] = I(c, v, h);
}
function Z(t, c, u, a, r, v, M, i, f) {
  var o = H, E = J, P = Math.abs(r - v);
  if (P % p < 1e-4 && P > 1e-4) {
    i[0] = t - u, i[1] = c - a, f[0] = t + u, f[1] = c + a;
    return;
  }
  if (C[0] = S(r) * u + t, C[1] = Q(r) * a + c, D[0] = S(v) * u + t, D[1] = Q(v) * a + c, o(i, C, D), E(f, C, D), r = r % p, r < 0 && (r = r + p), v = v % p, v < 0 && (v = v + p), r > v && !M ? v += p : r < v && M && (r += p), M) {
    var q = v;
    v = r, r = q;
  }
  for (var h = 0; h < v; h += Math.PI / 2)
    h > r && (L[0] = S(h) * u + t, L[1] = Q(h) * a + c, o(i, L, i), E(f, L, f));
}
export {
  Z as fromArc,
  X as fromCubic,
  W as fromLine,
  V as fromPoints,
  Y as fromQuadratic
};
