// zrender@6.0.0/lib/core/BoundingRect.js
import { translate as V, scale as X, create as Y } from "./matrix.js";
import x from "./Point.js";
var T = Math.min, M = Math.max, z = Math.abs, A = ["x", "y"], Z = ["width", "height"], c = new x(), u = new x(), w = new x(), p = new x(), v = H(), D = v.minTv, B = v.maxTv, _ = [0, 0], P = (function() {
  function r(i, n, t, e) {
    r.set(this, i, n, t, e);
  }
  return r.set = function(i, n, t, e, h) {
    return e < 0 && (n = n + e, e = -e), h < 0 && (t = t + h, h = -h), i.x = n, i.y = t, i.width = e, i.height = h, i;
  }, r.prototype.union = function(i) {
    var n = T(i.x, this.x), t = T(i.y, this.y);
    isFinite(this.x) && isFinite(this.width) ? this.width = M(i.x + i.width, this.x + this.width) - n : this.width = i.width, isFinite(this.y) && isFinite(this.height) ? this.height = M(i.y + i.height, this.y + this.height) - t : this.height = i.height, this.x = n, this.y = t;
  }, r.prototype.applyTransform = function(i) {
    r.applyTransform(this, this, i);
  }, r.prototype.calculateTransform = function(i) {
    var n = this, t = i.width / n.width, e = i.height / n.height, h = Y();
    return V(h, h, [-n.x, -n.y]), X(h, h, [t, e]), V(h, h, [i.x, i.y]), h;
  }, r.prototype.intersect = function(i, n, t) {
    return r.intersect(this, i, n, t);
  }, r.intersect = function(i, n, t, e) {
    t && x.set(t, 0, 0);
    var h = e && e.outIntersectRect || null, s = e && e.clamp;
    if (h && (h.x = h.y = h.width = h.height = NaN), !i || !n)
      return !1;
    i instanceof r || (i = r.set(k, i.x, i.y, i.width, i.height)), n instanceof r || (n = r.set(q, n.x, n.y, n.width, n.height));
    var y = !!t;
    v.reset(e, y);
    var a = v.touchThreshold, f = i.x + a, d = i.x + i.width - a, o = i.y + a, m = i.y + i.height - a, l = n.x + a, F = n.x + n.width - a, g = n.y + a, C = n.y + n.height - a;
    if (f > d || o > m || l > F || g > C)
      return !1;
    var S = !(d < l || F < f || m < g || C < o);
    return (y || h) && (_[0] = 1 / 0, _[1] = 0, N(f, d, l, F, 0, y, h, s), N(o, m, g, C, 1, y, h, s), y && x.copy(t, S ? v.useDir ? v.dirMinTv : D : B)), S;
  }, r.contain = function(i, n, t) {
    return n >= i.x && n <= i.x + i.width && t >= i.y && t <= i.y + i.height;
  }, r.prototype.contain = function(i, n) {
    return r.contain(this, i, n);
  }, r.prototype.clone = function() {
    return new r(this.x, this.y, this.width, this.height);
  }, r.prototype.copy = function(i) {
    r.copy(this, i);
  }, r.prototype.plain = function() {
    return {
      x: this.x,
      y: this.y,
      width: this.width,
      height: this.height
    };
  }, r.prototype.isFinite = function() {
    return isFinite(this.x) && isFinite(this.y) && isFinite(this.width) && isFinite(this.height);
  }, r.prototype.isZero = function() {
    return this.width === 0 || this.height === 0;
  }, r.create = function(i) {
    return new r(i.x, i.y, i.width, i.height);
  }, r.copy = function(i, n) {
    return i.x = n.x, i.y = n.y, i.width = n.width, i.height = n.height, i;
  }, r.applyTransform = function(i, n, t) {
    if (!t) {
      i !== n && r.copy(i, n);
      return;
    }
    if (t[1] < 1e-5 && t[1] > -1e-5 && t[2] < 1e-5 && t[2] > -1e-5) {
      var e = t[0], h = t[3], s = t[4], y = t[5];
      i.x = n.x * e + s, i.y = n.y * h + y, i.width = n.width * e, i.height = n.height * h, i.width < 0 && (i.x += i.width, i.width = -i.width), i.height < 0 && (i.y += i.height, i.height = -i.height);
      return;
    }
    c.x = w.x = n.x, c.y = p.y = n.y, u.x = p.x = n.x + n.width, u.y = w.y = n.y + n.height, c.transform(t), p.transform(t), u.transform(t), w.transform(t), i.x = T(c.x, u.x, w.x, p.x), i.y = T(c.y, u.y, w.y, p.y);
    var a = M(c.x, u.x, w.x, p.x), f = M(c.y, u.y, w.y, p.y);
    i.width = a - i.x, i.height = f - i.y;
  }, r;
})(), k = new P(0, 0, 0, 0), q = new P(0, 0, 0, 0);
function N(r, i, n, t, e, h, s, y) {
  var a = z(i - n), f = z(t - r), d = T(a, f), o = A[e], m = A[1 - e], l = Z[e];
  i < n || t < r ? a < f ? (h && (B[o] = -a), y && (s[o] = i, s[l] = 0)) : (h && (B[o] = f), y && (s[o] = r, s[l] = 0)) : (s && (s[o] = M(r, n), s[l] = T(i, t) - s[o]), h && (d < _[0] || v.useDir) && (_[0] = T(d, _[0]), (a < f || !v.bidirectional) && (D[o] = a, D[m] = 0, v.useDir && v.calcDirMTV()), (a >= f || !v.bidirectional) && (D[o] = -f, D[m] = 0, v.useDir && v.calcDirMTV())));
}
function H() {
  var r = 0, i = new x(), n = new x(), t = {
    minTv: new x(),
    maxTv: new x(),
    useDir: !1,
    dirMinTv: new x(),
    touchThreshold: 0,
    bidirectional: !0,
    negativeSize: !1,
    reset: function(h, s) {
      t.touchThreshold = 0, h && h.touchThreshold != null && (t.touchThreshold = M(0, h.touchThreshold)), t.negativeSize = !1, s && (t.minTv.set(1 / 0, 1 / 0), t.maxTv.set(0, 0), t.useDir = !1, h && h.direction != null && (t.useDir = !0, t.dirMinTv.copy(t.minTv), n.copy(t.minTv), r = h.direction, t.bidirectional = h.bidirectional == null || !!h.bidirectional, t.bidirectional || i.set(Math.cos(r), Math.sin(r))));
    },
    calcDirMTV: function() {
      var h = t.minTv, s = t.dirMinTv, y = h.y * h.y + h.x * h.x, a = Math.sin(r), f = Math.cos(r), d = a * h.y + f * h.x;
      if (e(d)) {
        e(h.x) && e(h.y) && s.set(0, 0);
        return;
      }
      if (n.x = y * f / d, n.y = y * a / d, e(n.x) && e(n.y)) {
        s.set(0, 0);
        return;
      }
      (t.bidirectional || i.dot(n) > 0) && n.len() < s.len() && s.copy(n);
    }
  };
  function e(h) {
    return z(h) < 1e-10;
  }
  return t;
}
export {
  H as createIntersectContext,
  P as default
};
