// zrender@6.0.0/lib/core/curve.js
import { distSquare as T, create as Z } from "./vector.js";
var j = Math.pow, D = Math.sqrt, m = 1e-8, Q = 1e-4, B = D(3), k = 1 / 3, R = Z(), q = Z(), g = Z();
function L(a) {
  return a > -m && a < m;
}
function U(a) {
  return a > m || a < -m;
}
function A(a, i, u, e, v) {
  var r = 1 - v;
  return r * r * (r * a + 3 * v * i) + v * v * (v * e + 3 * r * u);
}
function G(a, i, u, e, v) {
  var r = 1 - v;
  return 3 * (((i - a) * r + 2 * (u - i) * v) * r + (e - u) * v * v);
}
function J(a, i, u, e, v, r) {
  var n = e + 3 * (i - u) - a, l = 3 * (u - i * 2 + a), h = 3 * (i - a), f = a - v, b = l * l - 3 * n * h, c = l * h - 9 * n * f, s = h * h - 3 * l * f, d = 0;
  if (L(b) && L(c))
    if (L(l))
      r[0] = 0;
    else {
      var M = -h / l;
      M >= 0 && M <= 1 && (r[d++] = M);
    }
  else {
    var E = c * c - 4 * b * s;
    if (L(E)) {
      var I = c / b, M = -l / n + I, o = -I / 2;
      M >= 0 && M <= 1 && (r[d++] = M), o >= 0 && o <= 1 && (r[d++] = o);
    } else if (E > 0) {
      var t = D(E), N = b * l + 1.5 * n * (-c + t), O = b * l + 1.5 * n * (-c - t);
      N < 0 ? N = -j(-N, k) : N = j(N, k), O < 0 ? O = -j(-O, k) : O = j(O, k);
      var M = (-l - (N + O)) / (3 * n);
      M >= 0 && M <= 1 && (r[d++] = M);
    } else {
      var z = (2 * b * l - 3 * n * c) / (2 * D(b * b * b)), w = Math.acos(z) / 3, C = D(b), H = Math.cos(w), M = (-l - 2 * C * H) / (3 * n), o = (-l + C * (H + B * Math.sin(w))) / (3 * n), S = (-l + C * (H - B * Math.sin(w))) / (3 * n);
      M >= 0 && M <= 1 && (r[d++] = M), o >= 0 && o <= 1 && (r[d++] = o), S >= 0 && S <= 1 && (r[d++] = S);
    }
  }
  return d;
}
function V(a, i, u, e, v) {
  var r = 6 * u - 12 * i + 6 * a, n = 9 * i + 3 * e - 3 * a - 9 * u, l = 3 * i - 3 * a, h = 0;
  if (L(n)) {
    if (U(r)) {
      var f = -l / r;
      f >= 0 && f <= 1 && (v[h++] = f);
    }
  } else {
    var b = r * r - 4 * n * l;
    if (L(b))
      v[0] = -r / (2 * n);
    else if (b > 0) {
      var c = D(b), f = (-r + c) / (2 * n), s = (-r - c) / (2 * n);
      f >= 0 && f <= 1 && (v[h++] = f), s >= 0 && s <= 1 && (v[h++] = s);
    }
  }
  return h;
}
function W(a, i, u, e, v, r) {
  var n = (i - a) * v + a, l = (u - i) * v + i, h = (e - u) * v + u, f = (l - n) * v + n, b = (h - l) * v + l, c = (b - f) * v + f;
  r[0] = a, r[1] = n, r[2] = f, r[3] = c, r[4] = c, r[5] = b, r[6] = h, r[7] = e;
}
function X(a, i, u, e, v, r, n, l, h, f, b) {
  var c, s = 5e-3, d = 1 / 0, M, E, I, o;
  R[0] = h, R[1] = f;
  for (var t = 0; t < 1; t += 0.05)
    q[0] = A(a, u, v, n, t), q[1] = A(i, e, r, l, t), I = T(R, q), I < d && (c = t, d = I);
  d = 1 / 0;
  for (var N = 0; N < 32 && !(s < Q); N++)
    M = c - s, E = c + s, q[0] = A(a, u, v, n, M), q[1] = A(i, e, r, l, M), I = T(q, R), M >= 0 && I < d ? (c = M, d = I) : (g[0] = A(a, u, v, n, E), g[1] = A(i, e, r, l, E), o = T(g, R), E <= 1 && o < d ? (c = E, d = o) : s *= 0.5);
  return b && (b[0] = A(a, u, v, n, c), b[1] = A(i, e, r, l, c)), D(d);
}
function $(a, i, u, e, v, r, n, l, h) {
  for (var f = a, b = i, c = 0, s = 1 / h, d = 1; d <= h; d++) {
    var M = d * s, E = A(a, u, v, n, M), I = A(i, e, r, l, M), o = E - f, t = I - b;
    c += Math.sqrt(o * o + t * t), f = E, b = I;
  }
  return c;
}
function P(a, i, u, e) {
  var v = 1 - e;
  return v * (v * a + 2 * e * i) + e * e * u;
}
function K(a, i, u, e) {
  return 2 * ((1 - e) * (i - a) + e * (u - i));
}
function _(a, i, u, e, v) {
  var r = a - 2 * i + u, n = 2 * (i - a), l = a - e, h = 0;
  if (L(r)) {
    if (U(n)) {
      var f = -l / n;
      f >= 0 && f <= 1 && (v[h++] = f);
    }
  } else {
    var b = n * n - 4 * r * l;
    if (L(b)) {
      var f = -n / (2 * r);
      f >= 0 && f <= 1 && (v[h++] = f);
    } else if (b > 0) {
      var c = D(b), f = (-n + c) / (2 * r), s = (-n - c) / (2 * r);
      f >= 0 && f <= 1 && (v[h++] = f), s >= 0 && s <= 1 && (v[h++] = s);
    }
  }
  return h;
}
function Y(a, i, u) {
  var e = a + u - 2 * i;
  return e === 0 ? 0.5 : (a - i) / e;
}
function p(a, i, u, e, v) {
  var r = (i - a) * e + a, n = (u - i) * e + i, l = (n - r) * e + r;
  v[0] = a, v[1] = r, v[2] = l, v[3] = l, v[4] = n, v[5] = u;
}
function y(a, i, u, e, v, r, n, l, h) {
  var f, b = 5e-3, c = 1 / 0;
  R[0] = n, R[1] = l;
  for (var s = 0; s < 1; s += 0.05) {
    q[0] = P(a, u, v, s), q[1] = P(i, e, r, s);
    var d = T(R, q);
    d < c && (f = s, c = d);
  }
  c = 1 / 0;
  for (var M = 0; M < 32 && !(b < Q); M++) {
    var E = f - b, I = f + b;
    q[0] = P(a, u, v, E), q[1] = P(i, e, r, E);
    var d = T(q, R);
    if (E >= 0 && d < c)
      f = E, c = d;
    else {
      g[0] = P(a, u, v, I), g[1] = P(i, e, r, I);
      var o = T(g, R);
      I <= 1 && o < c ? (f = I, c = o) : b *= 0.5;
    }
  }
  return h && (h[0] = P(a, u, v, f), h[1] = P(i, e, r, f)), D(c);
}
function x(a, i, u, e, v, r, n) {
  for (var l = a, h = i, f = 0, b = 1 / n, c = 1; c <= n; c++) {
    var s = c * b, d = P(a, u, v, s), M = P(i, e, r, s), E = d - l, I = M - h;
    f += Math.sqrt(E * E + I * I), l = d, h = M;
  }
  return f;
}
export {
  A as cubicAt,
  G as cubicDerivativeAt,
  V as cubicExtrema,
  $ as cubicLength,
  X as cubicProjectPoint,
  J as cubicRootAt,
  W as cubicSubdivide,
  P as quadraticAt,
  K as quadraticDerivativeAt,
  Y as quadraticExtremum,
  x as quadraticLength,
  y as quadraticProjectPoint,
  _ as quadraticRootAt,
  p as quadraticSubdivide
};
