// zrender@6.0.0/lib/core/dom.js
import g from "./env.js";
import { buildTransformer as C } from "./fourPointsTransform.js";
import { each as E } from "./util.js";
var c = "___zrEVENTSAVED", d = [];
function L(t, r, o, a, e) {
  return h(d, r, a, e, !0) && h(t, o, d[0], d[1]);
}
function R(t, r) {
  t && o(t), r && o(r);
  function o(a) {
    var e = a[c];
    e && (e.clearMarkers && e.clearMarkers(), delete a[c]);
  }
}
function h(t, r, o, a, e) {
  if (r.getBoundingClientRect && g.domSupported && !M(r)) {
    var n = r[c] || (r[c] = {}), u = T(r, n), f = N(u, n, e);
    if (f)
      return f(t, o, a), !0;
  }
  return !1;
}
function T(t, r) {
  var o = r.markers;
  if (o)
    return o;
  o = r.markers = [];
  for (var a = ["left", "right"], e = ["top", "bottom"], n = 0; n < 4; n++) {
    var u = document.createElement("div"), f = u.style, i = n % 2, p = (n >> 1) % 2;
    f.cssText = [
      "position: absolute",
      "visibility: hidden",
      "padding: 0",
      "margin: 0",
      "border-width: 0",
      "user-select: none",
      "width:0",
      "height:0",
      a[i] + ":0",
      e[p] + ":0",
      a[1 - i] + ":auto",
      e[1 - p] + ":auto",
      ""
    ].join("!important;"), t.appendChild(u), o.push(u);
  }
  return r.clearMarkers = function() {
    E(o, function(s) {
      s.parentNode && s.parentNode.removeChild(s);
    });
  }, o;
}
function N(t, r, o) {
  for (var a = o ? "invTrans" : "trans", e = r[a], n = r.srcCoords, u = [], f = [], i = !0, p = 0; p < 4; p++) {
    var s = t[p].getBoundingClientRect(), v = 2 * p, l = s.left, m = s.top;
    u.push(l, m), i = i && n && l === n[v] && m === n[v + 1], f.push(t[p].offsetLeft, t[p].offsetTop);
  }
  return i && e ? e : (r.srcCoords = u, r[a] = o ? C(f, u) : C(u, f));
}
function M(t) {
  return t.nodeName.toUpperCase() === "CANVAS";
}
var S = /([&<>"'])/g, V = {
  "&": "&amp;",
  "<": "&lt;",
  ">": "&gt;",
  '"': "&quot;",
  "'": "&#39;"
};
function w(t) {
  return t == null ? "" : (t + "").replace(S, function(r, o) {
    return V[o];
  });
}
export {
  w as encodeHTML,
  M as isCanvasEl,
  h as transformCoordWithViewport,
  L as transformLocalCoord,
  R as transformLocalCoordClear
};
