// zrender@6.0.0/lib/core/env.js
var p = /* @__PURE__ */ (function() {
  function t() {
    this.firefox = !1, this.ie = !1, this.edge = !1, this.newEdge = !1, this.weChat = !1;
  }
  return t;
})(), u = /* @__PURE__ */ (function() {
  function t() {
    this.browser = new p(), this.node = !1, this.wxa = !1, this.worker = !1, this.svgSupported = !1, this.touchEventsSupported = !1, this.pointerEventsSupported = !1, this.domSupported = !1, this.transformSupported = !1, this.transform3dSupported = !1, this.hasGlobalWindow = typeof window < "u";
  }
  return t;
})(), o = new u();
typeof wx == "object" && typeof wx.getSystemInfoSync == "function" ? (o.wxa = !0, o.touchEventsSupported = !0) : typeof document > "u" && typeof self < "u" ? o.worker = !0 : !o.hasGlobalWindow || "Deno" in window || typeof navigator < "u" && typeof navigator.userAgent == "string" && navigator.userAgent.indexOf("Node.js") > -1 ? (o.node = !0, o.svgSupported = !0) : w(navigator.userAgent, o);
function w(t, r) {
  var e = r.browser, s = t.match(/Firefox\/([\d.]+)/), d = t.match(/MSIE\s([\d.]+)/) || t.match(/Trident\/.+?rv:(([\d.]+))/), i = t.match(/Edge?\/([\d.]+)/), f = /micromessenger/i.test(t);
  s && (e.firefox = !0, e.version = s[1]), d && (e.ie = !0, e.version = d[1]), i && (e.edge = !0, e.version = i[1], e.newEdge = +i[1].split(".")[0] > 18), f && (e.weChat = !0), r.svgSupported = typeof SVGRect < "u", r.touchEventsSupported = "ontouchstart" in window && !e.ie && !e.edge, r.pointerEventsSupported = "onpointerdown" in window && (e.edge || e.ie && +e.version >= 11);
  var a = r.domSupported = typeof document < "u";
  if (a) {
    var n = document.documentElement.style;
    r.transform3dSupported = (e.ie && "transition" in n || e.edge || "WebKitCSSMatrix" in window && "m11" in new WebKitCSSMatrix() || "MozPerspective" in n) && !("OTransition" in n), r.transformSupported = r.transform3dSupported || e.ie && +e.version >= 9;
  }
}
export {
  o as default
};
