// zrender@6.0.0/lib/core/event.js
import f from "./env.js";
import { isCanvasEl as u, transformCoordWithViewport as d } from "./dom.js";
var p = /^(?:mouse|pointer|contextmenu|drag|drop)|click/, o = [], h = f.browser.firefox && +f.browser.version.split(".")[0] < 39;
function v(t, r, n, e) {
  return n = n || {}, e ? c(t, r, n) : h && r.layerX != null && r.layerX !== r.offsetX ? (n.zrX = r.layerX, n.zrY = r.layerY) : r.offsetX != null ? (n.zrX = r.offsetX, n.zrY = r.offsetY) : c(t, r, n), n;
}
function c(t, r, n) {
  if (f.domSupported && t.getBoundingClientRect) {
    var e = r.clientX, a = r.clientY;
    if (u(t)) {
      var i = t.getBoundingClientRect();
      n.zrX = e - i.left, n.zrY = a - i.top;
      return;
    } else if (d(o, t, e, a)) {
      n.zrX = o[0], n.zrY = o[1];
      return;
    }
  }
  n.zrX = n.zrY = 0;
}
function X(t) {
  return t || window.event;
}
function m(t, r, n) {
  if (r = X(r), r.zrX != null)
    return r;
  var e = r.type, a = e && e.indexOf("touch") >= 0;
  if (a) {
    var s = e !== "touchend" ? r.targetTouches[0] : r.changedTouches[0];
    s && v(t, s, r, n);
  } else {
    v(t, r, r, n);
    var i = z(r);
    r.zrDelta = i ? i / 120 : -(r.detail || 0) / 3;
  }
  var l = r.button;
  return r.which == null && l !== void 0 && p.test(r.type) && (r.which = l & 1 ? 1 : l & 2 ? 3 : l & 4 ? 2 : 0), r;
}
function z(t) {
  var r = t.wheelDelta;
  if (r)
    return r;
  var n = t.deltaX, e = t.deltaY;
  if (n == null || e == null)
    return r;
  var a = Math.abs(e !== 0 ? e : n), i = e > 0 ? -1 : e < 0 ? 1 : n > 0 ? -1 : 1;
  return 3 * a * i;
}
function w(t, r, n, e) {
  t.addEventListener(r, n, e);
}
function E(t, r, n, e) {
  t.removeEventListener(r, n, e);
}
var y = function(t) {
  t.preventDefault(), t.stopPropagation(), t.cancelBubble = !0;
};
function b(t) {
  return t.which === 2 || t.which === 3;
}
export {
  w as addEventListener,
  v as clientToLocal,
  X as getNativeEvent,
  b as isMiddleOrRightButtonOnMouseUpDown,
  m as normalizeEvent,
  E as removeEventListener,
  y as stop
};
