// zrender@6.0.0/lib/core/GestureMgr.js
import { clientToLocal as p } from "./event.js";
var _ = (function() {
  function r() {
    this._track = [];
  }
  return r.prototype.recognize = function(t, n, e) {
    return this._doTrack(t, n, e), this._recognize(t);
  }, r.prototype.clear = function() {
    return this._track.length = 0, this;
  }, r.prototype._doTrack = function(t, n, e) {
    var o = t.touches;
    if (o) {
      for (var c = {
        points: [],
        touches: [],
        target: n,
        event: t
      }, i = 0, f = o.length; i < f; i++) {
        var h = o[i], a = p(e, h, {});
        c.points.push([a.zrX, a.zrY]), c.touches.push(h);
      }
      this._track.push(c);
    }
  }, r.prototype._recognize = function(t) {
    for (var n in u)
      if (u.hasOwnProperty(n)) {
        var e = u[n](this._track, t);
        if (e)
          return e;
      }
  }, r;
})();
function s(r) {
  var t = r[1][0] - r[0][0], n = r[1][1] - r[0][1];
  return Math.sqrt(t * t + n * n);
}
function g(r) {
  return [
    (r[0][0] + r[1][0]) / 2,
    (r[0][1] + r[1][1]) / 2
  ];
}
var u = {
  pinch: function(r, t) {
    var n = r.length;
    if (n) {
      var e = (r[n - 1] || {}).points, o = (r[n - 2] || {}).points || e;
      if (o && o.length > 1 && e && e.length > 1) {
        var c = s(e) / s(o);
        !isFinite(c) && (c = 1), t.pinchScale = c;
        var i = g(e);
        return t.pinchX = i[0], t.pinchY = i[1], {
          type: "pinch",
          target: r[0].target,
          event: t
        };
      }
    }
  }
};
export {
  _ as GestureMgr
};
