// zrender@6.0.0/lib/core/LRU.js
var h = /* @__PURE__ */ (function() {
  function e(t) {
    this.value = t;
  }
  return e;
})(), u = (function() {
  function e() {
    this._len = 0;
  }
  return e.prototype.insert = function(t) {
    var i = new h(t);
    return this.insertEntry(i), i;
  }, e.prototype.insertEntry = function(t) {
    this.head ? (this.tail.next = t, t.prev = this.tail, t.next = null, this.tail = t) : this.head = this.tail = t, this._len++;
  }, e.prototype.remove = function(t) {
    var i = t.prev, n = t.next;
    i ? i.next = n : this.head = n, n ? n.prev = i : this.tail = i, t.next = t.prev = null, this._len--;
  }, e.prototype.len = function() {
    return this._len;
  }, e.prototype.clear = function() {
    this.head = this.tail = null, this._len = 0;
  }, e;
})(), p = (function() {
  function e(t) {
    this._list = new u(), this._maxSize = 10, this._map = {}, this._maxSize = t;
  }
  return e.prototype.put = function(t, i) {
    var n = this._list, l = this._map, a = null;
    if (l[t] == null) {
      var o = n.len(), r = this._lastRemovedEntry;
      if (o >= this._maxSize && o > 0) {
        var s = n.head;
        n.remove(s), delete l[s.key], a = s.value, this._lastRemovedEntry = s;
      }
      r ? r.value = i : r = new h(i), r.key = t, n.insertEntry(r), l[t] = r;
    }
    return a;
  }, e.prototype.get = function(t) {
    var i = this._map[t], n = this._list;
    if (i != null)
      return i !== n.tail && (n.remove(i), n.insertEntry(i)), i.value;
  }, e.prototype.clear = function() {
    this._list.clear(), this._map = {};
  }, e.prototype.len = function() {
    return this._list.len();
  }, e;
})();
export {
  h as Entry,
  u as LinkedList,
  p as default
};
