// zrender@6.0.0/lib/core/matrix.js
function x() {
  return [1, 0, 0, 1, 0, 0];
}
function h(r) {
  return r[0] = 1, r[1] = 0, r[2] = 0, r[3] = 1, r[4] = 0, r[5] = 0, r;
}
function b(r, n) {
  return r[0] = n[0], r[1] = n[1], r[2] = n[2], r[3] = n[3], r[4] = n[4], r[5] = n[5], r;
}
function M(r, n, a) {
  var e = n[0] * a[0] + n[2] * a[1], c = n[1] * a[0] + n[3] * a[1], f = n[0] * a[2] + n[2] * a[3], i = n[1] * a[2] + n[3] * a[3], l = n[0] * a[4] + n[2] * a[5] + n[4], v = n[1] * a[4] + n[3] * a[5] + n[5];
  return r[0] = e, r[1] = c, r[2] = f, r[3] = i, r[4] = l, r[5] = v, r;
}
function g(r, n, a) {
  return r[0] = n[0], r[1] = n[1], r[2] = n[2], r[3] = n[3], r[4] = n[4] + a[0], r[5] = n[5] + a[1], r;
}
function j(r, n, a, e) {
  e === void 0 && (e = [0, 0]);
  var c = n[0], f = n[2], i = n[4], l = n[1], v = n[3], d = n[5], s = Math.sin(a), y = Math.cos(a);
  return r[0] = c * y + l * s, r[1] = -c * s + l * y, r[2] = f * y + v * s, r[3] = -f * s + y * v, r[4] = y * (i - e[0]) + s * (d - e[1]) + e[0], r[5] = y * (d - e[1]) - s * (i - e[0]) + e[1], r;
}
function k(r, n, a) {
  var e = a[0], c = a[1];
  return r[0] = n[0] * e, r[1] = n[1] * c, r[2] = n[2] * e, r[3] = n[3] * c, r[4] = n[4] * e, r[5] = n[5] * c, r;
}
function q(r, n) {
  var a = n[0], e = n[2], c = n[4], f = n[1], i = n[3], l = n[5], v = a * i - f * e;
  return v ? (v = 1 / v, r[0] = i * v, r[1] = -f * v, r[2] = -e * v, r[3] = a * v, r[4] = (e * l - i * c) * v, r[5] = (f * c - a * l) * v, r) : null;
}
function w(r) {
  var n = x();
  return b(n, r), n;
}
export {
  w as clone,
  b as copy,
  x as create,
  h as identity,
  q as invert,
  M as mul,
  j as rotate,
  k as scale,
  g as translate
};
