// zrender@6.0.0/lib/core/PathProxy.js
import { min as bt, max as mt } from "./vector.js";
import yt from "./BoundingRect.js";
import { devicePixelRatio as ut } from "../config.js";
import { fromLine as _t, fromArc as Dt, fromQuadratic as Lt, fromCubic as Tt } from "./bbox.js";
import { quadraticSubdivide as pt, cubicSubdivide as ct, quadraticLength as Ct, cubicLength as Mt } from "./curve.js";
var h = {
  M: 1,
  L: 2,
  C: 3,
  Q: 4,
  A: 5,
  Z: 6,
  R: 7
}, X = [], Z = [], M = [], q = [], g = [], k = [], it = Math.min, et = Math.max, V = Math.cos, Y = Math.sin, S = Math.abs, vt = Math.PI, w = vt * 2, st = typeof Float32Array < "u", K = [];
function nt(v) {
  var t = Math.round(v / vt * 1e8) / 1e8;
  return t % 2 * vt;
}
function gt(v, t) {
  var r = nt(v[0]);
  r < 0 && (r += w);
  var a = r - v[0], e = v[1];
  e += a, !t && e - r >= w ? e = r + w : t && r - e >= w ? e = r - w : !t && r > e ? e = r + (w - nt(r - e)) : t && r < e && (e = r - (w - nt(e - r))), v[0] = r, v[1] = e;
}
var wt = (function() {
  function v(t) {
    this.dpr = 1, this._xi = 0, this._yi = 0, this._x0 = 0, this._y0 = 0, this._len = 0, t && (this._saveData = !1), this._saveData && (this.data = []);
  }
  return v.prototype.increaseVersion = function() {
    this._version++;
  }, v.prototype.getVersion = function() {
    return this._version;
  }, v.prototype.setScale = function(t, r, a) {
    a = a || 0, a > 0 && (this._ux = S(a / ut / t) || 0, this._uy = S(a / ut / r) || 0);
  }, v.prototype.setDPR = function(t) {
    this.dpr = t;
  }, v.prototype.setContext = function(t) {
    this._ctx = t;
  }, v.prototype.getContext = function() {
    return this._ctx;
  }, v.prototype.beginPath = function() {
    return this._ctx && this._ctx.beginPath(), this.reset(), this;
  }, v.prototype.reset = function() {
    this._saveData && (this._len = 0), this._pathSegLen && (this._pathSegLen = null, this._pathLen = 0), this._version++;
  }, v.prototype.moveTo = function(t, r) {
    return this._drawPendingPt(), this.addData(h.M, t, r), this._ctx && this._ctx.moveTo(t, r), this._x0 = t, this._y0 = r, this._xi = t, this._yi = r, this;
  }, v.prototype.lineTo = function(t, r) {
    var a = S(t - this._xi), e = S(r - this._yi), s = a > this._ux || e > this._uy;
    if (this.addData(h.L, t, r), this._ctx && s && this._ctx.lineTo(t, r), s)
      this._xi = t, this._yi = r, this._pendingPtDist = 0;
    else {
      var i = a * a + e * e;
      i > this._pendingPtDist && (this._pendingPtX = t, this._pendingPtY = r, this._pendingPtDist = i);
    }
    return this;
  }, v.prototype.bezierCurveTo = function(t, r, a, e, s, i) {
    return this._drawPendingPt(), this.addData(h.C, t, r, a, e, s, i), this._ctx && this._ctx.bezierCurveTo(t, r, a, e, s, i), this._xi = s, this._yi = i, this;
  }, v.prototype.quadraticCurveTo = function(t, r, a, e) {
    return this._drawPendingPt(), this.addData(h.Q, t, r, a, e), this._ctx && this._ctx.quadraticCurveTo(t, r, a, e), this._xi = a, this._yi = e, this;
  }, v.prototype.arc = function(t, r, a, e, s, i) {
    this._drawPendingPt(), K[0] = e, K[1] = s, gt(K, i), e = K[0], s = K[1];
    var u = s - e;
    return this.addData(h.A, t, r, a, a, e, u, 0, i ? 0 : 1), this._ctx && this._ctx.arc(t, r, a, e, s, i), this._xi = V(s) * a + t, this._yi = Y(s) * a + r, this;
  }, v.prototype.arcTo = function(t, r, a, e, s) {
    return this._drawPendingPt(), this._ctx && this._ctx.arcTo(t, r, a, e, s), this;
  }, v.prototype.rect = function(t, r, a, e) {
    return this._drawPendingPt(), this._ctx && this._ctx.rect(t, r, a, e), this.addData(h.R, t, r, a, e), this;
  }, v.prototype.closePath = function() {
    this._drawPendingPt(), this.addData(h.Z);
    var t = this._ctx, r = this._x0, a = this._y0;
    return t && t.closePath(), this._xi = r, this._yi = a, this;
  }, v.prototype.fill = function(t) {
    t && t.fill(), this.toStatic();
  }, v.prototype.stroke = function(t) {
    t && t.stroke(), this.toStatic();
  }, v.prototype.len = function() {
    return this._len;
  }, v.prototype.setData = function(t) {
    if (this._saveData) {
      var r = t.length;
      !(this.data && this.data.length === r) && st && (this.data = new Float32Array(r));
      for (var a = 0; a < r; a++)
        this.data[a] = t[a];
      this._len = r;
    }
  }, v.prototype.appendPath = function(t) {
    if (this._saveData) {
      t instanceof Array || (t = [t]);
      for (var r = t.length, a = 0, e = this._len, s = 0; s < r; s++)
        a += t[s].len();
      var i = this.data;
      if (st && (i instanceof Float32Array || !i) && (this.data = new Float32Array(e + a), e > 0 && i))
        for (var u = 0; u < e; u++)
          this.data[u] = i[u];
      for (var s = 0; s < r; s++)
        for (var l = t[s].data, u = 0; u < l.length; u++)
          this.data[e++] = l[u];
      this._len = e;
    }
  }, v.prototype.addData = function(t, r, a, e, s, i, u, l, d) {
    if (this._saveData) {
      var _ = this.data;
      this._len + arguments.length > _.length && (this._expandData(), _ = this.data);
      for (var c = 0; c < arguments.length; c++)
        _[this._len++] = arguments[c];
    }
  }, v.prototype._drawPendingPt = function() {
    this._pendingPtDist > 0 && (this._ctx && this._ctx.lineTo(this._pendingPtX, this._pendingPtY), this._pendingPtDist = 0);
  }, v.prototype._expandData = function() {
    if (!(this.data instanceof Array)) {
      for (var t = [], r = 0; r < this._len; r++)
        t[r] = this.data[r];
      this.data = t;
    }
  }, v.prototype.toStatic = function() {
    if (this._saveData) {
      this._drawPendingPt();
      var t = this.data;
      t instanceof Array && (t.length = this._len, st && this._len > 11 && (this.data = new Float32Array(t)));
    }
  }, v.prototype.getBoundingRect = function() {
    M[0] = M[1] = g[0] = g[1] = Number.MAX_VALUE, q[0] = q[1] = k[0] = k[1] = -Number.MAX_VALUE;
    var t = this.data, r = 0, a = 0, e = 0, s = 0, i;
    for (i = 0; i < this._len; ) {
      var u = t[i++], l = i === 1;
      switch (l && (r = t[i], a = t[i + 1], e = r, s = a), u) {
        case h.M:
          r = e = t[i++], a = s = t[i++], g[0] = e, g[1] = s, k[0] = e, k[1] = s;
          break;
        case h.L:
          _t(r, a, t[i], t[i + 1], g, k), r = t[i++], a = t[i++];
          break;
        case h.C:
          Tt(r, a, t[i++], t[i++], t[i++], t[i++], t[i], t[i + 1], g, k), r = t[i++], a = t[i++];
          break;
        case h.Q:
          Lt(r, a, t[i++], t[i++], t[i], t[i + 1], g, k), r = t[i++], a = t[i++];
          break;
        case h.A:
          var d = t[i++], _ = t[i++], c = t[i++], n = t[i++], y = t[i++], D = t[i++] + y;
          i += 1;
          var L = !t[i++];
          l && (e = V(y) * c + d, s = Y(y) * n + _), Dt(d, _, c, n, y, D, L, g, k), r = V(D) * c + d, a = Y(D) * n + _;
          break;
        case h.R:
          e = r = t[i++], s = a = t[i++];
          var f = t[i++], m = t[i++];
          _t(e, s, e + f, s + m, g, k);
          break;
        case h.Z:
          r = e, a = s;
          break;
      }
      bt(M, M, g), mt(q, q, k);
    }
    return i === 0 && (M[0] = M[1] = q[0] = q[1] = 0), new yt(M[0], M[1], q[0] - M[0], q[1] - M[1]);
  }, v.prototype._calculateLength = function() {
    var t = this.data, r = this._len, a = this._ux, e = this._uy, s = 0, i = 0, u = 0, l = 0;
    this._pathSegLen || (this._pathSegLen = []);
    for (var d = this._pathSegLen, _ = 0, c = 0, n = 0; n < r; ) {
      var y = t[n++], D = n === 1;
      D && (s = t[n], i = t[n + 1], u = s, l = i);
      var L = -1;
      switch (y) {
        case h.M:
          s = u = t[n++], i = l = t[n++];
          break;
        case h.L: {
          var f = t[n++], m = t[n++], P = f - s, T = m - i;
          (S(P) > a || S(T) > e || n === r - 1) && (L = Math.sqrt(P * P + T * T), s = f, i = m);
          break;
        }
        case h.C: {
          var E = t[n++], U = t[n++], f = t[n++], m = t[n++], o = t[n++], A = t[n++];
          L = Mt(s, i, E, U, f, m, o, A, 10), s = o, i = A;
          break;
        }
        case h.Q: {
          var E = t[n++], U = t[n++], f = t[n++], m = t[n++];
          L = Ct(s, i, E, U, f, m, 10), s = f, i = m;
          break;
        }
        case h.A:
          var G = t[n++], F = t[n++], z = t[n++], p = t[n++], b = t[n++], H = t[n++], x = H + b;
          n += 1, D && (u = V(b) * z + G, l = Y(b) * p + F), L = et(z, p) * it(w, Math.abs(H)), s = V(x) * z + G, i = Y(x) * p + F;
          break;
        case h.R: {
          u = s = t[n++], l = i = t[n++];
          var Q = t[n++], R = t[n++];
          L = Q * 2 + R * 2;
          break;
        }
        case h.Z: {
          var P = u - s, T = l - i;
          L = Math.sqrt(P * P + T * T), s = u, i = l;
          break;
        }
      }
      L >= 0 && (d[c++] = L, _ += L);
    }
    return this._pathLen = _, _;
  }, v.prototype.rebuildPath = function(t, r) {
    var a = this.data, e = this._ux, s = this._uy, i = this._len, u, l, d, _, c, n, y = r < 1, D, L, f = 0, m = 0, P, T = 0, E, U;
    if (!(y && (this._pathSegLen || this._calculateLength(), D = this._pathSegLen, L = this._pathLen, P = r * L, !P)))
      t: for (var o = 0; o < i; ) {
        var A = a[o++], G = o === 1;
        switch (G && (d = a[o], _ = a[o + 1], u = d, l = _), A !== h.L && T > 0 && (t.lineTo(E, U), T = 0), A) {
          case h.M:
            u = d = a[o++], l = _ = a[o++], t.moveTo(d, _);
            break;
          case h.L: {
            c = a[o++], n = a[o++];
            var F = S(c - d), z = S(n - _);
            if (F > e || z > s) {
              if (y) {
                var p = D[m++];
                if (f + p > P) {
                  var b = (P - f) / p;
                  t.lineTo(d * (1 - b) + c * b, _ * (1 - b) + n * b);
                  break t;
                }
                f += p;
              }
              t.lineTo(c, n), d = c, _ = n, T = 0;
            } else {
              var H = F * F + z * z;
              H > T && (E = c, U = n, T = H);
            }
            break;
          }
          case h.C: {
            var x = a[o++], Q = a[o++], R = a[o++], N = a[o++], at = a[o++], rt = a[o++];
            if (y) {
              var p = D[m++];
              if (f + p > P) {
                var b = (P - f) / p;
                ct(d, x, R, at, b, X), ct(_, Q, N, rt, b, Z), t.bezierCurveTo(X[1], Z[1], X[2], Z[2], X[3], Z[3]);
                break t;
              }
              f += p;
            }
            t.bezierCurveTo(x, Q, R, N, at, rt), d = at, _ = rt;
            break;
          }
          case h.Q: {
            var x = a[o++], Q = a[o++], R = a[o++], N = a[o++];
            if (y) {
              var p = D[m++];
              if (f + p > P) {
                var b = (P - f) / p;
                pt(d, x, R, b, X), pt(_, Q, N, b, Z), t.quadraticCurveTo(X[1], Z[1], X[2], Z[2]);
                break t;
              }
              f += p;
            }
            t.quadraticCurveTo(x, Q, R, N), d = R, _ = N;
            break;
          }
          case h.A:
            var W = a[o++], tt = a[o++], O = a[o++], $ = a[o++], B = a[o++], ot = a[o++], dt = a[o++], ht = !a[o++], Pt = O > $ ? O : $, lt = S(O - $) > 1e-3, I = B + ot, ft = !1;
            if (y) {
              var p = D[m++];
              f + p > P && (I = B + ot * (P - f) / p, ft = !0), f += p;
            }
            if (lt && t.ellipse ? t.ellipse(W, tt, O, $, dt, B, I, ht) : t.arc(W, tt, Pt, B, I, ht), ft)
              break t;
            G && (u = V(B) * O + W, l = Y(B) * $ + tt), d = V(I) * O + W, _ = Y(I) * $ + tt;
            break;
          case h.R:
            u = d = a[o], l = _ = a[o + 1], c = a[o++], n = a[o++];
            var j = a[o++], J = a[o++];
            if (y) {
              var p = D[m++];
              if (f + p > P) {
                var C = P - f;
                t.moveTo(c, n), t.lineTo(c + it(C, j), n), C -= j, C > 0 && t.lineTo(c + j, n + it(C, J)), C -= J, C > 0 && t.lineTo(c + et(j - C, 0), n + J), C -= j, C > 0 && t.lineTo(c, n + et(J - C, 0));
                break t;
              }
              f += p;
            }
            t.rect(c, n, j, J);
            break;
          case h.Z:
            if (y) {
              var p = D[m++];
              if (f + p > P) {
                var b = (P - f) / p;
                t.lineTo(d * (1 - b) + u * b, _ * (1 - b) + l * b);
                break t;
              }
              f += p;
            }
            t.closePath(), d = u, _ = l;
        }
      }
  }, v.prototype.clone = function() {
    var t = new v(), r = this.data;
    return t.data = r.slice ? r.slice() : Array.prototype.slice.call(r), t._len = this._len, t;
  }, v.prototype.canSave = function() {
    return !!this._saveData;
  }, v.CMD = h, v.initDefaultProps = (function() {
    var t = v.prototype;
    t._saveData = !0, t._ux = 0, t._uy = 0, t._pendingPtDist = 0, t._version = 0;
  })(), v;
})();
export {
  wt as default,
  gt as normalizeArcAngles
};
