// zrender@6.0.0/lib/core/platform.js
var c = 12, f = "sans-serif", d = c + "px " + f, T = 20, m = 100, W = "007LLmW'55;N0500LLLLLLLLLL00NNNLzWW\\\\WQb\\0FWLg\\bWb\\WQ\\WrWWQ000CL5LLFLL0LL**F*gLLLL5F0LF\\FFF5.5N";
function g(a) {
  var n = {};
  if (typeof JSON > "u")
    return n;
  for (var r = 0; r < a.length; r++) {
    var e = String.fromCharCode(r + 32), L = (a.charCodeAt(r) - T) / m;
    n[e] = L;
  }
  return n;
}
var l = g(W), i = {
  createCanvas: function() {
    return typeof document < "u" && document.createElement("canvas");
  },
  measureText: /* @__PURE__ */ (function() {
    var a, n;
    return function(r, e) {
      if (!a) {
        var L = i.createCanvas();
        a = L && L.getContext("2d");
      }
      if (a)
        return n !== e && (n = a.font = e || d), a.measureText(r);
      r = r || "", e = e || d;
      var o = /((?:\d+)?\.?\d*)px/.exec(e), v = o && +o[1] || c, t = 0;
      if (e.indexOf("mono") >= 0)
        t = v * r.length;
      else
        for (var u = 0; u < r.length; u++) {
          var F = l[r[u]];
          t += F == null ? v : F * v;
        }
      return { width: t };
    };
  })(),
  loadImage: function(a, n, r) {
    var e = new Image();
    return e.onload = n, e.onerror = r, e.src = a, e;
  }
};
export {
  d as DEFAULT_FONT,
  f as DEFAULT_FONT_FAMILY,
  c as DEFAULT_FONT_SIZE,
  l as DEFAULT_TEXT_WIDTH_MAP,
  i as platformApi
};
