// zrender@6.0.0/lib/core/timsort.js
var j = 32, H = 7;
function J(f) {
  for (var w = 0; f >= j; )
    w |= f & 1, f >>= 1;
  return f + w;
}
function C(f, w, s, E) {
  var u = w + 1;
  if (u === s)
    return 1;
  if (E(f[u++], f[w]) < 0) {
    for (; u < s && E(f[u], f[u - 1]) < 0; )
      u++;
    K(f, w, u);
  } else
    for (; u < s && E(f[u], f[u - 1]) >= 0; )
      u++;
  return u - w;
}
function K(f, w, s) {
  for (s--; w < s; ) {
    var E = f[w];
    f[w++] = f[s], f[s--] = E;
  }
}
function P(f, w, s, E, u) {
  for (E === w && E++; E < s; E++) {
    for (var d = f[E], v = w, G = E, e; v < G; )
      e = v + G >>> 1, u(d, f[e]) < 0 ? G = e : v = e + 1;
    var D = E - v;
    switch (D) {
      case 3:
        f[v + 3] = f[v + 2];
      case 2:
        f[v + 2] = f[v + 1];
      case 1:
        f[v + 1] = f[v];
        break;
      default:
        for (; D > 0; )
          f[v + D] = f[v + D - 1], D--;
    }
    f[v] = d;
  }
}
function U(f, w, s, E, u, d) {
  var v = 0, G = 0, e = 1;
  if (d(f, w[s + u]) > 0) {
    for (G = E - u; e < G && d(f, w[s + u + e]) > 0; )
      v = e, e = (e << 1) + 1, e <= 0 && (e = G);
    e > G && (e = G), v += u, e += u;
  } else {
    for (G = u + 1; e < G && d(f, w[s + u - e]) <= 0; )
      v = e, e = (e << 1) + 1, e <= 0 && (e = G);
    e > G && (e = G);
    var D = v;
    v = u - e, e = u - D;
  }
  for (v++; v < e; ) {
    var N = v + (e - v >>> 1);
    d(f, w[s + N]) > 0 ? v = N + 1 : e = N;
  }
  return e;
}
function z(f, w, s, E, u, d) {
  var v = 0, G = 0, e = 1;
  if (d(f, w[s + u]) < 0) {
    for (G = u + 1; e < G && d(f, w[s + u - e]) < 0; )
      v = e, e = (e << 1) + 1, e <= 0 && (e = G);
    e > G && (e = G);
    var D = v;
    v = u - e, e = u - D;
  } else {
    for (G = E - u; e < G && d(f, w[s + u + e]) >= 0; )
      v = e, e = (e << 1) + 1, e <= 0 && (e = G);
    e > G && (e = G), v += u, e += u;
  }
  for (v++; v < e; ) {
    var N = v + (e - v >>> 1);
    d(f, w[s + N]) < 0 ? e = N : v = N + 1;
  }
  return e;
}
function Q(f, w) {
  var s = H, E, u, d = 0, v = [];
  E = [], u = [];
  function G(k, b) {
    E[d] = k, u[d] = b, d += 1;
  }
  function e() {
    for (; d > 1; ) {
      var k = d - 2;
      if (k >= 1 && u[k - 1] <= u[k] + u[k + 1] || k >= 2 && u[k - 2] <= u[k] + u[k - 1])
        u[k - 1] < u[k + 1] && k--;
      else if (u[k] > u[k + 1])
        break;
      N(k);
    }
  }
  function D() {
    for (; d > 1; ) {
      var k = d - 2;
      k > 0 && u[k - 1] < u[k + 1] && k--, N(k);
    }
  }
  function N(k) {
    var b = E[k], S = u[k], R = E[k + 1], i = u[k + 1];
    u[k] = S + i, k === d - 3 && (E[k + 1] = E[k + 2], u[k + 1] = u[k + 2]), d--;
    var M = z(f[R], f, b, S, 0, w);
    b += M, S -= M, S !== 0 && (i = U(f[b + S - 1], f, R, i, i - 1, w), i !== 0 && (S <= i ? q(b, S, R, i) : B(b, S, R, i)));
  }
  function q(k, b, S, R) {
    var i = 0;
    for (i = 0; i < b; i++)
      v[i] = f[k + i];
    var M = 0, A = S, L = k;
    if (f[L++] = f[A++], --R === 0) {
      for (i = 0; i < b; i++)
        f[L + i] = v[M + i];
      return;
    }
    if (b === 1) {
      for (i = 0; i < R; i++)
        f[L + i] = f[A + i];
      f[L + R] = v[M];
      return;
    }
    for (var O = s, _, I, o; ; ) {
      _ = 0, I = 0, o = !1;
      do
        if (w(f[A], v[M]) < 0) {
          if (f[L++] = f[A++], I++, _ = 0, --R === 0) {
            o = !0;
            break;
          }
        } else if (f[L++] = v[M++], _++, I = 0, --b === 1) {
          o = !0;
          break;
        }
      while ((_ | I) < O);
      if (o)
        break;
      do {
        if (_ = z(f[A], v, M, b, 0, w), _ !== 0) {
          for (i = 0; i < _; i++)
            f[L + i] = v[M + i];
          if (L += _, M += _, b -= _, b <= 1) {
            o = !0;
            break;
          }
        }
        if (f[L++] = f[A++], --R === 0) {
          o = !0;
          break;
        }
        if (I = U(v[M], f, A, R, 0, w), I !== 0) {
          for (i = 0; i < I; i++)
            f[L + i] = f[A + i];
          if (L += I, A += I, R -= I, R === 0) {
            o = !0;
            break;
          }
        }
        if (f[L++] = v[M++], --b === 1) {
          o = !0;
          break;
        }
        O--;
      } while (_ >= H || I >= H);
      if (o)
        break;
      O < 0 && (O = 0), O += 2;
    }
    if (s = O, s < 1 && (s = 1), b === 1) {
      for (i = 0; i < R; i++)
        f[L + i] = f[A + i];
      f[L + R] = v[M];
    } else {
      if (b === 0)
        throw new Error();
      for (i = 0; i < b; i++)
        f[L + i] = v[M + i];
    }
  }
  function B(k, b, S, R) {
    var i = 0;
    for (i = 0; i < R; i++)
      v[i] = f[S + i];
    var M = k + b - 1, A = R - 1, L = S + R - 1, O = 0, _ = 0;
    if (f[L--] = f[M--], --b === 0) {
      for (O = L - (R - 1), i = 0; i < R; i++)
        f[O + i] = v[i];
      return;
    }
    if (R === 1) {
      for (L -= b, M -= b, _ = L + 1, O = M + 1, i = b - 1; i >= 0; i--)
        f[_ + i] = f[O + i];
      f[L] = v[A];
      return;
    }
    for (var I = s; ; ) {
      var o = 0, T = 0, F = !1;
      do
        if (w(v[A], f[M]) < 0) {
          if (f[L--] = f[M--], o++, T = 0, --b === 0) {
            F = !0;
            break;
          }
        } else if (f[L--] = v[A--], T++, o = 0, --R === 1) {
          F = !0;
          break;
        }
      while ((o | T) < I);
      if (F)
        break;
      do {
        if (o = b - z(v[A], f, k, b, b - 1, w), o !== 0) {
          for (L -= o, M -= o, b -= o, _ = L + 1, O = M + 1, i = o - 1; i >= 0; i--)
            f[_ + i] = f[O + i];
          if (b === 0) {
            F = !0;
            break;
          }
        }
        if (f[L--] = v[A--], --R === 1) {
          F = !0;
          break;
        }
        if (T = R - U(f[M], v, 0, R, R - 1, w), T !== 0) {
          for (L -= T, A -= T, R -= T, _ = L + 1, O = A + 1, i = 0; i < T; i++)
            f[_ + i] = v[O + i];
          if (R <= 1) {
            F = !0;
            break;
          }
        }
        if (f[L--] = f[M--], --b === 0) {
          F = !0;
          break;
        }
        I--;
      } while (o >= H || T >= H);
      if (F)
        break;
      I < 0 && (I = 0), I += 2;
    }
    if (s = I, s < 1 && (s = 1), R === 1) {
      for (L -= b, M -= b, _ = L + 1, O = M + 1, i = b - 1; i >= 0; i--)
        f[_ + i] = f[O + i];
      f[L] = v[A];
    } else {
      if (R === 0)
        throw new Error();
      for (O = L - (R - 1), i = 0; i < R; i++)
        f[O + i] = v[i];
    }
  }
  return {
    mergeRuns: e,
    forceMergeRuns: D,
    pushRun: G
  };
}
function V(f, w, s, E) {
  s || (s = 0), E || (E = f.length);
  var u = E - s;
  if (!(u < 2)) {
    var d = 0;
    if (u < j) {
      d = C(f, s, E, w), P(f, s, E, s + d, w);
      return;
    }
    var v = Q(f, w), G = J(u);
    do {
      if (d = C(f, s, E, w), d < G) {
        var e = u;
        e > G && (e = G), P(f, s, s + e, s + d, w), d = e;
      }
      v.pushRun(s, d), v.mergeRuns(), u -= d, s += d;
    } while (u !== 0);
    v.forceMergeRuns();
  }
}
export {
  V as default
};
