// zrender@6.0.0/lib/core/util.js
var O = b([
  "Function",
  "RegExp",
  "Date",
  "Error",
  "CanvasGradient",
  "CanvasPattern",
  "Image",
  "Canvas"
], function(n, t) {
  return n["[object " + t + "]"] = !0, n;
}, {}), w = b([
  "Int8",
  "Uint8",
  "Uint8Clamped",
  "Int16",
  "Uint16",
  "Int32",
  "Uint32",
  "Float32",
  "Float64"
], function(n, t) {
  return n["[object " + t + "Array]"] = !0, n;
}, {}), u = Object.prototype.toString, s = Array.prototype, P = s.forEach, S = s.filter, y = s.slice, I = s.map, m = (function() {
}).constructor, l = m ? m.prototype : null, g = "__proto__", x = 2311;
function B() {
  return x++;
}
function G() {
  for (var n = [], t = 0; t < arguments.length; t++)
    n[t] = arguments[t];
  typeof console < "u" && console.error.apply(console, n);
}
function p(n) {
  if (n == null || typeof n != "object")
    return n;
  var t = n, r = u.call(n);
  if (r === "[object Array]") {
    if (!a(n)) {
      t = [];
      for (var e = 0, i = n.length; e < i; e++)
        t[e] = p(n[e]);
    }
  } else if (w[r]) {
    if (!a(n)) {
      var f = n.constructor;
      if (f.from)
        t = f.from(n);
      else {
        t = new f(n.length);
        for (var e = 0, i = n.length; e < i; e++)
          t[e] = n[e];
      }
    }
  } else if (!O[r] && !a(n) && !v(n)) {
    t = {};
    for (var o in n)
      n.hasOwnProperty(o) && o !== g && (t[o] = p(n[o]));
  }
  return t;
}
function A(n, t, r) {
  if (!c(t) || !c(n))
    return r ? p(t) : n;
  for (var e in t)
    if (t.hasOwnProperty(e) && e !== g) {
      var i = n[e], f = t[e];
      c(f) && c(i) && !h(f) && !h(i) && !v(f) && !v(i) && !d(f) && !d(i) && !a(f) && !a(i) ? A(i, f, r) : (r || !(e in n)) && (n[e] = p(t[e]));
    }
  return n;
}
function L(n, t) {
  for (var r = n[0], e = 1, i = n.length; e < i; e++)
    r = A(r, n[e], t);
  return r;
}
function N(n, t) {
  if (Object.assign)
    Object.assign(n, t);
  else
    for (var r in t)
      t.hasOwnProperty(r) && r !== g && (n[r] = t[r]);
  return n;
}
function T(n, t, r) {
  for (var e = E(t), i = 0, f = e.length; i < f; i++) {
    var o = e[i];
    (r ? t[o] != null : n[o] == null) && (n[o] = t[o]);
  }
  return n;
}
function Y(n, t) {
  if (n) {
    if (n.indexOf)
      return n.indexOf(t);
    for (var r = 0, e = n.length; r < e; r++)
      if (n[r] === t)
        return r;
  }
  return -1;
}
function q(n, t) {
  var r = n.prototype;
  function e() {
  }
  e.prototype = t.prototype, n.prototype = new e();
  for (var i in r)
    r.hasOwnProperty(i) && (n.prototype[i] = r[i]);
  n.prototype.constructor = n, n.superClass = t;
}
function J(n, t, r) {
  if (n = "prototype" in n ? n.prototype : n, t = "prototype" in t ? t.prototype : t, Object.getOwnPropertyNames)
    for (var e = Object.getOwnPropertyNames(t), i = 0; i < e.length; i++) {
      var f = e[i];
      f !== "constructor" && (r ? t[f] != null : n[f] == null) && (n[f] = t[f]);
    }
  else
    T(n, t, r);
}
function $(n) {
  return !n || typeof n == "string" ? !1 : typeof n.length == "number";
}
function R(n, t, r) {
  if (n && t)
    if (n.forEach && n.forEach === P)
      n.forEach(t, r);
    else if (n.length === +n.length)
      for (var e = 0, i = n.length; e < i; e++)
        t.call(r, n[e], e, n);
    else
      for (var f in n)
        n.hasOwnProperty(f) && t.call(r, n[f], f, n);
}
function Q(n, t, r) {
  if (!n)
    return [];
  if (!t)
    return j(n);
  if (n.map && n.map === I)
    return n.map(t, r);
  for (var e = [], i = 0, f = n.length; i < f; i++)
    e.push(t.call(r, n[i], i, n));
  return e;
}
function b(n, t, r, e) {
  if (n && t) {
    for (var i = 0, f = n.length; i < f; i++)
      r = t.call(e, r, n[i], i, n);
    return r;
  }
}
function V(n, t, r) {
  if (!n)
    return [];
  if (!t)
    return j(n);
  if (n.filter && n.filter === S)
    return n.filter(t, r);
  for (var e = [], i = 0, f = n.length; i < f; i++)
    t.call(r, n[i], i, n) && e.push(n[i]);
  return e;
}
function W(n, t, r) {
  if (n && t) {
    for (var e = 0, i = n.length; e < i; e++)
      if (t.call(r, n[e], e, n))
        return n[e];
  }
}
function E(n) {
  if (!n)
    return [];
  if (Object.keys)
    return Object.keys(n);
  var t = [];
  for (var r in n)
    n.hasOwnProperty(r) && t.push(r);
  return t;
}
function U(n, t) {
  for (var r = [], e = 2; e < arguments.length; e++)
    r[e - 2] = arguments[e];
  return function() {
    return n.apply(t, r.concat(y.call(arguments)));
  };
}
var X = l && D(l.bind) ? l.call.bind(l.bind) : U;
function Z(n) {
  for (var t = [], r = 1; r < arguments.length; r++)
    t[r - 1] = arguments[r];
  return function() {
    return n.apply(this, t.concat(y.call(arguments)));
  };
}
function h(n) {
  return Array.isArray ? Array.isArray(n) : u.call(n) === "[object Array]";
}
function D(n) {
  return typeof n == "function";
}
function _(n) {
  return typeof n == "string";
}
function z(n) {
  return u.call(n) === "[object String]";
}
function k(n) {
  return typeof n == "number";
}
function c(n) {
  var t = typeof n;
  return t === "function" || !!n && t === "object";
}
function d(n) {
  return !!O[u.call(n)];
}
function nn(n) {
  return !!w[u.call(n)];
}
function v(n) {
  return typeof n == "object" && typeof n.nodeType == "number" && typeof n.ownerDocument == "object";
}
function tn(n) {
  return n.colorStops != null;
}
function rn(n) {
  return n.image != null;
}
function en(n) {
  return u.call(n) === "[object RegExp]";
}
function fn(n) {
  return n !== n;
}
function on() {
  for (var n = [], t = 0; t < arguments.length; t++)
    n[t] = arguments[t];
  for (var r = 0, e = n.length; r < e; r++)
    if (n[r] != null)
      return n[r];
}
function un(n, t) {
  return n ?? t;
}
function an(n, t, r) {
  return n ?? t ?? r;
}
function j(n) {
  for (var t = [], r = 1; r < arguments.length; r++)
    t[r - 1] = arguments[r];
  return y.apply(n, t);
}
function ln(n) {
  if (typeof n == "number")
    return [n, n, n, n];
  var t = n.length;
  return t === 2 ? [n[0], n[1], n[0], n[1]] : t === 3 ? [n[0], n[1], n[2], n[1]] : n;
}
function cn(n, t) {
  if (!n)
    throw new Error(t);
}
function pn(n) {
  return n == null ? null : typeof n.trim == "function" ? n.trim() : n.replace(/^[\s\uFEFF\xA0]+|[\s\uFEFF\xA0]+$/g, "");
}
var F = "__ec_primitive__";
function sn(n) {
  n[F] = !0;
}
function a(n) {
  return n[F];
}
var C = (function() {
  function n() {
    this.data = {};
  }
  return n.prototype.delete = function(t) {
    var r = this.has(t);
    return r && delete this.data[t], r;
  }, n.prototype.has = function(t) {
    return this.data.hasOwnProperty(t);
  }, n.prototype.get = function(t) {
    return this.data[t];
  }, n.prototype.set = function(t, r) {
    return this.data[t] = r, this;
  }, n.prototype.keys = function() {
    return E(this.data);
  }, n.prototype.forEach = function(t) {
    var r = this.data;
    for (var e in r)
      r.hasOwnProperty(e) && t(r[e], e);
  }, n;
})(), M = typeof Map == "function";
function H() {
  return M ? /* @__PURE__ */ new Map() : new C();
}
var K = (function() {
  function n(t) {
    var r = h(t);
    this.data = H();
    var e = this;
    t instanceof n ? t.each(i) : t && R(t, i);
    function i(f, o) {
      r ? e.set(f, o) : e.set(o, f);
    }
  }
  return n.prototype.hasKey = function(t) {
    return this.data.has(t);
  }, n.prototype.get = function(t) {
    return this.data.get(t);
  }, n.prototype.set = function(t, r) {
    return this.data.set(t, r), r;
  }, n.prototype.each = function(t, r) {
    this.data.forEach(function(e, i) {
      t.call(r, e, i);
    });
  }, n.prototype.keys = function() {
    var t = this.data.keys();
    return M ? Array.from(t) : t;
  }, n.prototype.removeKey = function(t) {
    this.data.delete(t);
  }, n;
})();
function hn(n) {
  return new K(n);
}
function vn(n, t) {
  for (var r = new n.constructor(n.length + t.length), e = 0; e < n.length; e++)
    r[e] = n[e];
  for (var i = n.length, e = 0; e < t.length; e++)
    r[e + i] = t[e];
  return r;
}
function yn(n, t) {
  var r;
  if (Object.create)
    r = Object.create(n);
  else {
    var e = function() {
    };
    e.prototype = n, r = new e();
  }
  return t && N(r, t), r;
}
function gn(n) {
  var t = n.style;
  t.webkitUserSelect = "none", t.userSelect = "none", t.webkitTapHighlightColor = "rgba(0,0,0,0)", t["-webkit-touch-callout"] = "none";
}
function mn(n, t) {
  return n.hasOwnProperty(t);
}
function dn() {
}
var On = 180 / Math.PI;
export {
  K as HashMap,
  On as RADIAN_TO_DEGREE,
  cn as assert,
  X as bind,
  p as clone,
  vn as concatArray,
  hn as createHashMap,
  yn as createObject,
  Z as curry,
  T as defaults,
  gn as disableUserSelect,
  R as each,
  fn as eqNaN,
  N as extend,
  V as filter,
  W as find,
  B as guid,
  mn as hasOwn,
  Y as indexOf,
  q as inherits,
  h as isArray,
  $ as isArrayLike,
  d as isBuiltInObject,
  v as isDom,
  D as isFunction,
  tn as isGradientObject,
  rn as isImagePatternObject,
  k as isNumber,
  c as isObject,
  a as isPrimitive,
  en as isRegExp,
  _ as isString,
  z as isStringSafe,
  nn as isTypedArray,
  E as keys,
  G as logError,
  Q as map,
  A as merge,
  L as mergeAll,
  J as mixin,
  dn as noop,
  ln as normalizeCssArray,
  b as reduce,
  on as retrieve,
  un as retrieve2,
  an as retrieve3,
  sn as setAsPrimitive,
  j as slice,
  pn as trim
};
