// zrender@6.0.0/lib/core/vector.js
function l(n, r) {
  return n == null && (n = 0), r == null && (r = 0), [n, r];
}
function s(n, r) {
  return n[0] = r[0], n[1] = r[1], n;
}
function d(n) {
  return [n[0], n[1]];
}
function h(n, r, e) {
  return n[0] = r, n[1] = e, n;
}
function M(n, r, e) {
  return n[0] = r[0] + e[0], n[1] = r[1] + e[1], n;
}
function p(n, r, e, i) {
  return n[0] = r[0] + e[0] * i, n[1] = r[1] + e[1] * i, n;
}
function q(n, r, e) {
  return n[0] = r[0] - e[0], n[1] = r[1] - e[1], n;
}
function c(n) {
  return Math.sqrt(t(n));
}
function t(n) {
  return n[0] * n[0] + n[1] * n[1];
}
function m(n, r, e) {
  return n[0] = r[0] * e, n[1] = r[1] * e, n;
}
function x(n, r) {
  var e = c(r);
  return e === 0 ? (n[0] = 0, n[1] = 0) : (n[0] = r[0] / e, n[1] = r[1] / e), n;
}
function f(n, r) {
  return Math.sqrt((n[0] - r[0]) * (n[0] - r[0]) + (n[1] - r[1]) * (n[1] - r[1]));
}
var S = f;
function u(n, r) {
  return (n[0] - r[0]) * (n[0] - r[0]) + (n[1] - r[1]) * (n[1] - r[1]);
}
var A = u;
function b(n, r, e, i) {
  return n[0] = r[0] + i * (e[0] - r[0]), n[1] = r[1] + i * (e[1] - r[1]), n;
}
function y(n, r, e) {
  var i = r[0], a = r[1];
  return n[0] = e[0] * i + e[2] * a + e[4], n[1] = e[1] * i + e[3] * a + e[5], n;
}
function z(n, r, e) {
  return n[0] = Math.min(r[0], e[0]), n[1] = Math.min(r[1], e[1]), n;
}
function T(n, r, e) {
  return n[0] = Math.max(r[0], e[0]), n[1] = Math.max(r[1], e[1]), n;
}
export {
  M as add,
  y as applyTransform,
  d as clone,
  s as copy,
  l as create,
  S as dist,
  A as distSquare,
  f as distance,
  u as distanceSquare,
  c as len,
  t as lenSquare,
  b as lerp,
  T as max,
  z as min,
  x as normalize,
  m as scale,
  p as scaleAndAdd,
  h as set,
  q as sub
};
