// motion@12.17.0/dist/es/framer-motion/dist/es/animation/utils/create-visual-element.js
import { HTMLVisualElement as r } from "../../render/html/HTMLVisualElement.js";
import { ObjectVisualElement as s } from "../../render/object/ObjectVisualElement.js";
import { visualElementStore as n } from "../../render/store.js";
import { SVGVisualElement as a } from "../../render/svg/SVGVisualElement.js";
import { isSVGElement as i } from "../../../../../motion-dom/dist/es/utils/is-svg-element.js";
import { isSVGSVGElement as l } from "../../../../../motion-dom/dist/es/utils/is-svg-svg-element.js";
function S(t) {
  const o = {
    presenceContext: null,
    props: {},
    visualState: {
      renderState: {
        transform: {},
        transformOrigin: {},
        style: {},
        vars: {},
        attrs: {}
      },
      latestValues: {}
    }
  }, e = i(t) && !l(t) ? new a(o) : new r(o);
  e.mount(t), n.set(t, e);
}
function E(t) {
  const o = {
    presenceContext: null,
    props: {},
    visualState: {
      renderState: {
        output: {}
      },
      latestValues: {}
    }
  }, e = new s(o);
  e.mount(t), n.set(t, e);
}
export {
  S as createDOMVisualElement,
  E as createObjectVisualElement
};
