// echarts@6.0.0/lib/chart/map/MapSeries.js
import { __extends as h } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { curry as v, createHashMap as S, each as M, merge as D } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import b from "../helper/createSeriesDataSimply.js";
import N from "../../model/Series.js";
import C from "../../coord/geo/geoSourceManager.js";
import { makeSeriesEncodeForNameBased as k } from "../../data/helper/sourceHelper.js";
import { createTooltipMarkup as y } from "../../component/tooltip/tooltipMarkup.js";
import { createSymbol as w } from "../../util/symbol.js";
import { decideCoordSysUsageKind as R, CoordinateSystemUsageKind as T } from "../../core/CoordinateSystem.js";
import l from "../../visual/tokens.js";
var A = (
  /** @class */
  (function(f) {
    h(o, f);
    function o() {
      var e = f !== null && f.apply(this, arguments) || this;
      return e.type = o.type, e.needsDrawMap = !1, e.seriesGroup = [], e.getTooltipPosition = function(t) {
        if (t != null) {
          var r = this.getData().getName(t), n = this.coordinateSystem, a = n.getRegion(r);
          return a && n.dataToPoint(a.getCenter());
        }
      }, e;
    }
    return o.prototype.getInitialData = function(e) {
      for (var t = b(this, {
        coordDimensions: ["value"],
        encodeDefaulter: v(k, this)
      }), r = S(), n = [], a = 0, m = t.count(); a < m; a++) {
        var s = t.getName(a);
        r.set(s, a);
      }
      var p = C.load(this.getMapType(), this.option.nameMap, this.option.nameProperty);
      return M(p.regions, function(i) {
        var u = i.name, c = r.get(u), g = i.properties && i.properties.echartsStyle, d;
        c == null ? (d = {
          name: u
        }, n.push(d)) : d = t.getRawDataItem(c), g && D(d, g);
      }), t.appendData(n), t;
    }, o.prototype.getHostGeoModel = function() {
      if (R(this).kind !== T.boxCoordSys)
        return this.getReferringComponents("geo", {
          useDefault: !1,
          enableAll: !1,
          enableNone: !1
        }).models[0];
    }, o.prototype.getMapType = function() {
      return (this.getHostGeoModel() || this).option.map;
    }, o.prototype.getRawValue = function(e) {
      var t = this.getData();
      return t.get(t.mapDimension("value"), e);
    }, o.prototype.getRegionModel = function(e) {
      var t = this.getData();
      return t.getItemModel(t.indexOfName(e));
    }, o.prototype.formatTooltip = function(e, t, r) {
      for (var n = this.getData(), a = this.getRawValue(e), m = n.getName(e), s = this.seriesGroup, p = [], i = 0; i < s.length; i++) {
        var u = s[i].originalData.indexOfName(m), c = n.mapDimension("value");
        isNaN(s[i].originalData.get(c, u)) || p.push(s[i].name);
      }
      return y("section", {
        header: p.join(", "),
        noHeader: !p.length,
        blocks: [y("nameValue", {
          name: m,
          value: a
        })]
      });
    }, o.prototype.setZoom = function(e) {
      this.option.zoom = e;
    }, o.prototype.setCenter = function(e) {
      this.option.center = e;
    }, o.prototype.getLegendIcon = function(e) {
      var t = e.icon || "roundRect", r = w(t, 0, 0, e.itemWidth, e.itemHeight, e.itemStyle.fill);
      return r.setStyle(e.itemStyle), r.style.stroke = "none", t.indexOf("empty") > -1 && (r.style.stroke = r.style.fill, r.style.fill = l.color.neutral00, r.style.lineWidth = 2), r;
    }, o.type = "series.map", o.dependencies = ["geo"], o.layoutMode = "box", o.defaultOption = {
      // 一级层叠
      // zlevel: 0,
      // 二级层叠
      z: 2,
      coordinateSystem: "geo",
      // map should be explicitly specified since ec3.
      map: "",
      // If `geoIndex` is not specified, a exclusive geo will be
      // created. Otherwise use the specified geo component, and
      // `map` and `mapType` are ignored.
      // geoIndex: 0,
      // 'center' | 'left' | 'right' | 'x%' | {number}
      left: "center",
      // 'center' | 'top' | 'bottom' | 'x%' | {number}
      top: "center",
      // right
      // bottom
      // width:
      // height
      // Aspect is width / height. Inited to be geoJson bbox aspect
      // This parameter is used for scale this aspect
      // Default value:
      // for geoSVG source: 1,
      // for geoJSON source: 0.75.
      aspectScale: null,
      // Layout with center and size
      // If you want to put map in a fixed size box with right aspect ratio
      // This two properties may be more convenient.
      // layoutCenter: [50%, 50%]
      // layoutSize: 100
      showLegendSymbol: !0,
      // Define left-top, right-bottom coords to control view
      // For example, [ [180, 90], [-180, -90] ],
      // higher priority than center and zoom
      boundingCoords: null,
      // Default on center of map
      center: null,
      zoom: 1,
      scaleLimit: null,
      selectedMode: !0,
      label: {
        show: !1,
        color: l.color.tertiary
      },
      // scaleLimit: null,
      itemStyle: {
        borderWidth: 0.5,
        borderColor: l.color.border,
        areaColor: l.color.background
      },
      emphasis: {
        label: {
          show: !0,
          color: l.color.primary
        },
        itemStyle: {
          areaColor: l.color.highlight
        }
      },
      select: {
        label: {
          show: !0,
          color: l.color.primary
        },
        itemStyle: {
          color: l.color.highlight
        }
      },
      nameProperty: "name"
    }, o;
  })(N)
);
export {
  A as default
};
