// echarts@6.0.0/lib/chart/map/MapView.js
import { __extends as y } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import S from "../../component/helper/MapDraw.js";
import L from "../../view/Chart.js";
import { setLabelStyle as d, getLabelStatesModels as F } from "../../label/labelStyle.js";
import { Z2_EMPHASIS_LIFT as C, setStatesFlag as T } from "../../util/states.js";
import x from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Circle.js";
var G = (
  /** @class */
  (function(p) {
    y(r, p);
    function r() {
      var e = p !== null && p.apply(this, arguments) || this;
      return e.type = r.type, e;
    }
    return r.prototype.render = function(e, s, n, t) {
      if (!(t && t.type === "mapToggleSelect" && t.from === this.uid)) {
        var a = this.group;
        if (a.removeAll(), !e.getHostGeoModel()) {
          if (this._mapDraw && t && t.type === "geoRoam" && this._mapDraw.resetForLabelLayout(), t && t.type === "geoRoam" && t.componentType === "series" && t.seriesId === e.id) {
            var i = this._mapDraw;
            i && a.add(i.group);
          } else if (e.needsDrawMap) {
            var i = this._mapDraw || new S(n);
            a.add(i.group), i.draw(e, s, n, this, t), this._mapDraw = i;
          } else
            this._mapDraw && this._mapDraw.remove(), this._mapDraw = null;
          e.get("showLegendSymbol") && s.getComponent("legend") && this._renderSymbols(e, s, n);
        }
      }
    }, r.prototype.remove = function() {
      this._mapDraw && this._mapDraw.remove(), this._mapDraw = null, this.group.removeAll();
    }, r.prototype.dispose = function() {
      this._mapDraw && this._mapDraw.remove(), this._mapDraw = null;
    }, r.prototype._renderSymbols = function(e, s, n) {
      var t = e.originalData, a = this.group;
      t.each(t.mapDimension("value"), function(i, u) {
        if (!isNaN(i)) {
          var m = t.getItemLayout(u);
          if (!(!m || !m.point)) {
            var g = m.point, f = m.offset, o = new x({
              style: {
                // Because the special of map draw.
                // Which needs statistic of multiple series and draw on one map.
                // And each series also need a symbol with legend color
                //
                // Layout and visual are put one the different data
                // TODO
                fill: e.getData().getVisual("style").fill
              },
              shape: {
                cx: g[0] + f * 9,
                cy: g[1],
                r: 3
              },
              silent: !0,
              // Do not overlap the first series, on which labels are displayed.
              z2: 8 + (f ? 0 : C + 1)
            });
            if (!f) {
              var h = e.mainSeries.getData(), l = t.getName(u), v = h.indexOfName(l), w = t.getItemModel(u), D = w.getModel("label"), _ = h.getItemGraphicEl(v);
              d(o, F(w), {
                labelFetcher: {
                  getFormattedLabel: function(c, b) {
                    return e.getFormattedLabel(v, b);
                  }
                },
                defaultText: l
              }), o.disableLabelAnimation = !0, D.get("position") || o.setTextConfig({
                position: "bottom"
              }), _.onHoverStateChange = function(c) {
                T(o, c);
              };
            }
            a.add(o);
          }
        }
      });
    }, r.type = "map", r;
  })(L)
);
export {
  G as default
};
