// echarts@6.0.0/lib/model/Component.js
import { __extends as c } from "../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { merge as u, each as y, map as h, indexOf as g } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import l from "./Model.js";
import { getUID as v, enableSubTypeDefaulter as M, enableTopologicalTravel as x } from "../util/component.js";
import { isExtendedClass as C, mountExtend as O, enableClassManagement as T, parseClassType as D } from "../util/clazz.js";
import { makeInner as I, queryReferringComponents as P } from "../util/model.js";
import { fetchLayoutMode as d, getLayoutParams as b, mergeLayoutParam as m, getBoxLayoutParams as L } from "../util/layout.js";
var _ = I(), a = (
  /** @class */
  (function(i) {
    c(t, i);
    function t(e, n, o) {
      var r = i.call(this, e, n, o) || this;
      return r.uid = v("ec_cpt_model"), r;
    }
    return t.prototype.init = function(e, n, o) {
      this.mergeDefaultAndTheme(e, o);
    }, t.prototype.mergeDefaultAndTheme = function(e, n) {
      var o = d(this), r = o ? b(e) : {}, p = n.getTheme();
      u(e, p.get(this.mainType)), u(e, this.getDefaultOption()), o && m(e, r, o);
    }, t.prototype.mergeOption = function(e, n) {
      u(this.option, e, !0);
      var o = d(this);
      o && m(this.option, e, o);
    }, t.prototype.optionUpdated = function(e, n) {
    }, t.prototype.getDefaultOption = function() {
      var e = this.constructor;
      if (!C(e))
        return e.defaultOption;
      var n = _(this);
      if (!n.defaultOption) {
        for (var o = [], r = e; r; ) {
          var p = r.prototype.defaultOption;
          p && o.push(p), r = r.superClass;
        }
        for (var s = {}, f = o.length - 1; f >= 0; f--)
          s = u(s, o[f], !0);
        n.defaultOption = s;
      }
      return n.defaultOption;
    }, t.prototype.getReferringComponents = function(e, n) {
      var o = e + "Index", r = e + "Id";
      return P(this.ecModel, e, {
        index: this.get(o, !0),
        id: this.get(r, !0)
      }, n);
    }, t.prototype.getBoxLayoutParams = function() {
      return L(this, !1);
    }, t.prototype.getZLevelKey = function() {
      return "";
    }, t.prototype.setZLevel = function(e) {
      this.option.zlevel = e;
    }, t.protoInitialize = (function() {
      var e = t.prototype;
      e.type = "component", e.id = "", e.name = "", e.mainType = "", e.subType = "", e.componentIndex = 0;
    })(), t;
  })(l)
);
O(a, l);
T(a);
M(a);
x(a, B);
function B(i) {
  var t = [];
  return y(a.getClassesByMainType(i), function(e) {
    t = t.concat(e.dependencies || e.prototype.dependencies || []);
  }), t = h(t, function(e) {
    return D(e).main;
  }), i !== "dataset" && g(t, "dataset") <= 0 && t.unshift("dataset"), t;
}
export {
  a as default
};
