// echarts@6.0.0/lib/model/Global.js
import { __extends as X } from "../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { assert as k, each as h, createHashMap as m, clone as T, filter as g, isFunction as $, isString as ee, isObject as H, merge as D, extend as j, isArray as te, mixin as ne } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { normalizeToArray as A, mappingToExists as re, setComponentTypeToKeyInfo as oe, convertOptionIdName as L, isComponentIdInternal as ie } from "../util/model.js";
import S from "./Model.js";
import v from "./Component.js";
import se from "./globalDefault.js";
import { resetSourceDefaulter as ae } from "../data/helper/sourceHelper.js";
import { concatInternalOptions as le } from "./internalComponentCreator.js";
import { PaletteMixin as ue } from "./mixin/palette.js";
import { error as G, warn as pe } from "../util/log.js";
var I, x, K, b = "\0_ec_inner", Z = 1, ce = {
  grid: "GridComponent",
  polar: "PolarComponent",
  geo: "GeoComponent",
  singleAxis: "SingleAxisComponent",
  parallel: "ParallelComponent",
  calendar: "CalendarComponent",
  matrix: "MatrixComponent",
  graphic: "GraphicComponent",
  toolbox: "ToolboxComponent",
  tooltip: "TooltipComponent",
  axisPointer: "AxisPointerComponent",
  brush: "BrushComponent",
  title: "TitleComponent",
  timeline: "TimelineComponent",
  markPoint: "MarkPointComponent",
  markLine: "MarkLineComponent",
  markArea: "MarkAreaComponent",
  legend: "LegendComponent",
  dataZoom: "DataZoomComponent",
  visualMap: "VisualMapComponent",
  // aria: 'AriaComponent',
  // dataset: 'DatasetComponent',
  // Dependencies
  xAxis: "GridComponent",
  yAxis: "GridComponent",
  angleAxis: "PolarComponent",
  radiusAxis: "PolarComponent"
}, fe = {
  line: "LineChart",
  bar: "BarChart",
  pie: "PieChart",
  scatter: "ScatterChart",
  radar: "RadarChart",
  map: "MapChart",
  tree: "TreeChart",
  treemap: "TreemapChart",
  graph: "GraphChart",
  chord: "ChordChart",
  gauge: "GaugeChart",
  funnel: "FunnelChart",
  parallel: "ParallelChart",
  sankey: "SankeyChart",
  boxplot: "BoxplotChart",
  candlestick: "CandlestickChart",
  effectScatter: "EffectScatterChart",
  lines: "LinesChart",
  heatmap: "HeatmapChart",
  pictorialBar: "PictorialBarChart",
  themeRiver: "ThemeRiverChart",
  sunburst: "SunburstChart",
  custom: "CustomChart"
}, E = {};
function he(a) {
  h(a, function(o, e) {
    if (!v.hasClass(e)) {
      var n = ce[e];
      n && !E[n] && (G("Component " + e + ` is used but not imported.
import { ` + n + ` } from 'echarts/components';
echarts.use([` + n + "]);"), E[n] = !0);
    }
  });
}
var ve = (
  /** @class */
  (function(a) {
    X(o, a);
    function o() {
      return a !== null && a.apply(this, arguments) || this;
    }
    return o.prototype.init = function(e, n, t, r, i, p) {
      r = r || {}, this.option = null, this._theme = new S(r), this._locale = new S(i), this._optionManager = p;
    }, o.prototype.setOption = function(e, n, t) {
      process.env.NODE_ENV !== "production" && (k(e != null, "option is null/undefined"), k(e[b] !== Z, "please use chart.getOption()"));
      var r = W(n);
      this._optionManager.setOption(e, t, r), this._resetOption(null, r);
    }, o.prototype.resetOption = function(e, n) {
      return this._resetOption(e, W(n));
    }, o.prototype._resetOption = function(e, n) {
      var t = !1, r = this._optionManager;
      if (!e || e === "recreate") {
        var i = r.mountOption(e === "recreate");
        process.env.NODE_ENV !== "production" && he(i), !this.option || e === "recreate" ? K(this, i) : (this.restoreData(), this._mergeOption(i, n)), t = !0;
      }
      if ((e === "timeline" || e === "media") && this.restoreData(), !e || e === "recreate" || e === "timeline") {
        var p = r.getTimelineOption(this);
        p && (t = !0, this._mergeOption(p, n));
      }
      if (!e || e === "recreate" || e === "media") {
        var u = r.getMediaOption(this);
        u.length && h(u, function(s) {
          t = !0, this._mergeOption(s, n);
        }, this);
      }
      return t;
    }, o.prototype.mergeOption = function(e) {
      this._mergeOption(e, null);
    }, o.prototype._mergeOption = function(e, n) {
      var t = this.option, r = this._componentsMap, i = this._componentsCount, p = [], u = m(), s = n && n.replaceMergeMainTypeMap;
      ae(this), h(e, function(l, c) {
        l != null && (v.hasClass(c) ? c && (p.push(c), u.set(c, !0)) : t[c] = t[c] == null ? T(l) : D(t[c], l, !0));
      }), s && s.each(function(l, c) {
        v.hasClass(c) && !u.get(c) && (p.push(c), u.set(c, !0));
      }), v.topologicalTravel(p, v.getAllClassMainTypes(), C, this);
      function C(l) {
        var c = le(this, l, A(e[l])), _ = r.get(l), M = (
          // `!oldCmptList` means init. See the comment in `mappingToExists`
          _ ? s && s.get(l) ? "replaceMerge" : "normalMerge" : "replaceAll"
        ), U = re(_, c, M);
        oe(U, l, v), t[l] = null, r.set(l, null), i.set(l, 0);
        var P = [], B = [], V = 0, F, R;
        h(U, function(d, Y) {
          var f = d.existing, y = d.newOption;
          if (!y)
            f && (f.mergeOption({}, this), f.optionUpdated({}, !1));
          else {
            var J = l === "series", N = v.getClass(
              l,
              d.keyInfo.subType,
              !J
              // Give a more detailed warn later if series don't exists
            );
            if (!N) {
              if (process.env.NODE_ENV !== "production") {
                var O = d.keyInfo.subType, w = fe[O];
                E[O] || (E[O] = !0, w ? G("Series " + O + ` is used but not imported.
import { ` + w + ` } from 'echarts/charts';
echarts.use([` + w + "]);") : G("Unknown series " + O));
              }
              return;
            }
            if (l === "tooltip") {
              if (F) {
                process.env.NODE_ENV !== "production" && (R || (pe("Currently only one tooltip component is allowed."), R = !0));
                return;
              }
              F = !0;
            }
            if (f && f.constructor === N)
              f.name = d.keyInfo.name, f.mergeOption(y, this), f.optionUpdated(y, !1);
            else {
              var z = j({
                componentIndex: Y
              }, d.keyInfo);
              f = new N(y, this, this, z), j(f, z), d.brandNew && (f.__requireNewView = !0), f.init(y, this, this), f.optionUpdated(null, !0);
            }
          }
          f ? (P.push(f.option), B.push(f), V++) : (P.push(void 0), B.push(void 0));
        }, this), t[l] = P, r.set(l, B), i.set(l, V), l === "series" && I(this);
      }
      this._seriesIndices || I(this);
    }, o.prototype.getOption = function() {
      var e = T(this.option);
      return h(e, function(n, t) {
        if (v.hasClass(t)) {
          for (var r = A(n), i = r.length, p = !1, u = i - 1; u >= 0; u--)
            r[u] && !ie(r[u]) ? p = !0 : (r[u] = null, !p && i--);
          r.length = i, e[t] = r;
        }
      }), delete e[b], e;
    }, o.prototype.setTheme = function(e) {
      this._theme = new S(e), this._resetOption("recreate", null);
    }, o.prototype.getTheme = function() {
      return this._theme;
    }, o.prototype.getLocaleModel = function() {
      return this._locale;
    }, o.prototype.setUpdatePayload = function(e) {
      this._payload = e;
    }, o.prototype.getUpdatePayload = function() {
      return this._payload;
    }, o.prototype.getComponent = function(e, n) {
      var t = this._componentsMap.get(e);
      if (t) {
        var r = t[n || 0];
        if (r)
          return r;
        if (n == null) {
          for (var i = 0; i < t.length; i++)
            if (t[i])
              return t[i];
        }
      }
    }, o.prototype.queryComponents = function(e) {
      var n = e.mainType;
      if (!n)
        return [];
      var t = e.index, r = e.id, i = e.name, p = this._componentsMap.get(n);
      if (!p || !p.length)
        return [];
      var u;
      return t != null ? (u = [], h(A(t), function(s) {
        p[s] && u.push(p[s]);
      })) : r != null ? u = q("id", r, p) : i != null ? u = q("name", i, p) : u = g(p, function(s) {
        return !!s;
      }), Q(u, e);
    }, o.prototype.findComponents = function(e) {
      var n = e.query, t = e.mainType, r = p(n), i = r ? this.queryComponents(r) : g(this._componentsMap.get(t), function(s) {
        return !!s;
      });
      return u(Q(i, e));
      function p(s) {
        var C = t + "Index", l = t + "Id", c = t + "Name";
        return s && (s[C] != null || s[l] != null || s[c] != null) ? {
          mainType: t,
          // subType will be filtered finally.
          index: s[C],
          id: s[l],
          name: s[c]
        } : null;
      }
      function u(s) {
        return e.filter ? g(s, e.filter) : s;
      }
    }, o.prototype.eachComponent = function(e, n, t) {
      var r = this._componentsMap;
      if ($(e)) {
        var i = n, p = e;
        r.each(function(l, c) {
          for (var _ = 0; l && _ < l.length; _++) {
            var M = l[_];
            M && p.call(i, c, M, M.componentIndex);
          }
        });
      } else
        for (var u = ee(e) ? r.get(e) : H(e) ? this.findComponents(e) : null, s = 0; u && s < u.length; s++) {
          var C = u[s];
          C && n.call(t, C, C.componentIndex);
        }
    }, o.prototype.getSeriesByName = function(e) {
      var n = L(e, null);
      return g(this._componentsMap.get("series"), function(t) {
        return !!t && n != null && t.name === n;
      });
    }, o.prototype.getSeriesByIndex = function(e) {
      return this._componentsMap.get("series")[e];
    }, o.prototype.getSeriesByType = function(e) {
      return g(this._componentsMap.get("series"), function(n) {
        return !!n && n.subType === e;
      });
    }, o.prototype.getSeries = function() {
      return g(this._componentsMap.get("series"), function(e) {
        return !!e;
      });
    }, o.prototype.getSeriesCount = function() {
      return this._componentsCount.get("series");
    }, o.prototype.eachSeries = function(e, n) {
      x(this), h(this._seriesIndices, function(t) {
        var r = this._componentsMap.get("series")[t];
        e.call(n, r, t);
      }, this);
    }, o.prototype.eachRawSeries = function(e, n) {
      h(this._componentsMap.get("series"), function(t) {
        t && e.call(n, t, t.componentIndex);
      });
    }, o.prototype.eachSeriesByType = function(e, n, t) {
      x(this), h(this._seriesIndices, function(r) {
        var i = this._componentsMap.get("series")[r];
        i.subType === e && n.call(t, i, r);
      }, this);
    }, o.prototype.eachRawSeriesByType = function(e, n, t) {
      return h(this.getSeriesByType(e), n, t);
    }, o.prototype.isSeriesFiltered = function(e) {
      return x(this), this._seriesIndicesMap.get(e.componentIndex) == null;
    }, o.prototype.getCurrentSeriesIndices = function() {
      return (this._seriesIndices || []).slice();
    }, o.prototype.filterSeries = function(e, n) {
      x(this);
      var t = [];
      h(this._seriesIndices, function(r) {
        var i = this._componentsMap.get("series")[r];
        e.call(n, i, r) && t.push(r);
      }, this), this._seriesIndices = t, this._seriesIndicesMap = m(t);
    }, o.prototype.restoreData = function(e) {
      I(this);
      var n = this._componentsMap, t = [];
      n.each(function(r, i) {
        v.hasClass(i) && t.push(i);
      }), v.topologicalTravel(t, v.getAllClassMainTypes(), function(r) {
        h(n.get(r), function(i) {
          i && (r !== "series" || !ge(i, e)) && i.restoreData();
        });
      });
    }, o.internalField = (function() {
      I = function(e) {
        var n = e._seriesIndices = [];
        h(e._componentsMap.get("series"), function(t) {
          t && n.push(t.componentIndex);
        }), e._seriesIndicesMap = m(n);
      }, x = function(e) {
        if (process.env.NODE_ENV !== "production" && !e._seriesIndices)
          throw new Error("Option should contains series.");
      }, K = function(e, n) {
        e.option = {}, e.option[b] = Z, e._componentsMap = m({
          series: []
        }), e._componentsCount = m();
        var t = n.aria;
        H(t) && t.enabled == null && (t.enabled = !0), Ce(n, e._theme.option), D(n, se, !1), e._mergeOption(n, null);
      };
    })(), o;
  })(S)
);
function ge(a, o) {
  if (o) {
    var e = o.seriesIndex, n = o.seriesId, t = o.seriesName;
    return e != null && a.componentIndex !== e || n != null && a.id !== n || t != null && a.name !== t;
  }
}
function Ce(a, o) {
  var e = a.color && !a.colorLayer;
  h(o, function(n, t) {
    t === "colorLayer" && e || t === "color" && a.color || v.hasClass(t) || (typeof n == "object" ? a[t] = a[t] ? D(a[t], n, !1) : T(n) : a[t] == null && (a[t] = n));
  });
}
function q(a, o, e) {
  if (te(o)) {
    var n = m();
    return h(o, function(r) {
      if (r != null) {
        var i = L(r, null);
        i != null && n.set(r, !0);
      }
    }), g(e, function(r) {
      return r && n.get(r[a]);
    });
  } else {
    var t = L(o, null);
    return g(e, function(r) {
      return r && t != null && r[a] === t;
    });
  }
}
function Q(a, o) {
  return o.hasOwnProperty("subType") ? g(a, function(e) {
    return e && e.subType === o.subType;
  }) : a;
}
function W(a) {
  var o = m();
  return a && h(A(a.replaceMerge), function(e) {
    process.env.NODE_ENV !== "production" && k(v.hasClass(e), '"' + e + '" is not valid component main type in "replaceMerge"'), o.set(e, !0);
  }), {
    replaceMergeMainTypeMap: o
  };
}
ne(ve, ue);
export {
  ve as default
};
