// echarts@6.0.0/lib/model/Model.js
import s from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/env.js";
import { enableClassExtend as a, enableClassCheck as u } from "../util/clazz.js";
import { AreaStyleMixin as f } from "./mixin/areaStyle.js";
import h from "./mixin/textStyle.js";
import { LineStyleMixin as c } from "./mixin/lineStyle.js";
import { ItemStyleMixin as m } from "./mixin/itemStyle.js";
import { merge as d, clone as y, mixin as p } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
var r = (
  /** @class */
  (function() {
    function n(t, e, o) {
      this.parentModel = e, this.ecModel = o, this.option = t;
    }
    return n.prototype.init = function(t, e, o) {
    }, n.prototype.mergeOption = function(t, e) {
      d(this.option, t, !0);
    }, n.prototype.get = function(t, e) {
      return t == null ? this.option : this._doGet(this.parsePath(t), !e && this.parentModel);
    }, n.prototype.getShallow = function(t, e) {
      var o = this.option, i = o == null ? o : o[t];
      if (i == null && !e) {
        var l = this.parentModel;
        l && (i = l.getShallow(t));
      }
      return i;
    }, n.prototype.getModel = function(t, e) {
      var o = t != null, i = o ? this.parsePath(t) : null, l = o ? this._doGet(i) : this.option;
      return e = e || this.parentModel && this.parentModel.getModel(this.resolveParentPath(i)), new n(l, e, this.ecModel);
    }, n.prototype.isEmpty = function() {
      return this.option == null;
    }, n.prototype.restoreData = function() {
    }, n.prototype.clone = function() {
      var t = this.constructor;
      return new t(y(this.option));
    }, n.prototype.parsePath = function(t) {
      return typeof t == "string" ? t.split(".") : t;
    }, n.prototype.resolveParentPath = function(t) {
      return t;
    }, n.prototype.isAnimationEnabled = function() {
      if (!s.node && this.option) {
        if (this.option.animation != null)
          return !!this.option.animation;
        if (this.parentModel)
          return this.parentModel.isAnimationEnabled();
      }
    }, n.prototype._doGet = function(t, e) {
      var o = this.option;
      if (!t)
        return o;
      for (var i = 0; i < t.length && !(t[i] && (o = o && typeof o == "object" ? o[t[i]] : null, o == null)); i++)
        ;
      return o == null && e && (o = e._doGet(this.resolveParentPath(t), e.parentModel)), o;
    }, n;
  })()
);
a(r);
u(r);
p(r, c);
p(r, m);
p(r, f);
p(r, h);
export {
  r as default
};
