// echarts@6.0.0/lib/model/OptionManager.js
import { normalizeToArray as h } from "../util/model.js";
import { each as m, isTypedArray as v, setAsPrimitive as _, clone as l, map as D, isArray as b, isObject as O } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { error as y } from "../util/log.js";
var B = /^(min|max)?(.+)$/, N = (
  /** @class */
  (function() {
    function e(i) {
      this._timelineOptions = [], this._mediaList = [], this._currentMediaIndices = [], this._api = i;
    }
    return e.prototype.setOption = function(i, a, u) {
      i && (m(h(i.series), function(t) {
        t && t.data && v(t.data) && _(t.data);
      }), m(h(i.dataset), function(t) {
        t && t.source && v(t.source) && _(t.source);
      })), i = l(i);
      var o = this._optionBackup, n = k(i, a, !o);
      this._newBaseOption = n.baseOption, o ? (n.timelineOptions.length && (o.timelineOptions = n.timelineOptions), n.mediaList.length && (o.mediaList = n.mediaList), n.mediaDefault && (o.mediaDefault = n.mediaDefault)) : this._optionBackup = n;
    }, e.prototype.mountOption = function(i) {
      var a = this._optionBackup;
      return this._timelineOptions = a.timelineOptions, this._mediaList = a.mediaList, this._mediaDefault = a.mediaDefault, this._currentMediaIndices = [], l(i ? a.baseOption : this._newBaseOption);
    }, e.prototype.getTimelineOption = function(i) {
      var a, u = this._timelineOptions;
      if (u.length) {
        var o = i.getComponent("timeline");
        o && (a = l(
          // FIXME:TS as TimelineModel or quivlant interface
          u[o.getCurrentIndex()]
        ));
      }
      return a;
    }, e.prototype.getMediaOption = function(i) {
      var a = this._api.getWidth(), u = this._api.getHeight(), o = this._mediaList, n = this._mediaDefault, t = [], s = [];
      if (!o.length && !n)
        return s;
      for (var p = 0, f = o.length; p < f; p++)
        E(o[p].query, a, u) && t.push(p);
      return !t.length && n && (t = [-1]), t.length && !q(t, this._currentMediaIndices) && (s = D(t, function(c) {
        return l(c === -1 ? n.option : o[c].option);
      })), this._currentMediaIndices = t, s;
    }, e;
  })()
);
function k(e, i, a) {
  var u = [], o, n, t = e.baseOption, s = e.timeline, p = e.options, f = e.media, c = !!e.media, L = !!(p || s || t && t.timeline);
  t ? (n = t, n.timeline || (n.timeline = s)) : ((L || c) && (e.options = e.media = null), n = e), c && (b(f) ? m(f, function(r) {
    process.env.NODE_ENV !== "production" && r && !r.option && O(r.query) && O(r.query.option) && y("Illegal media option. Must be like { media: [ { query: {}, option: {} } ] }"), r && r.option && (r.query ? u.push(r) : o || (o = r));
  }) : process.env.NODE_ENV !== "production" && y("Illegal media option. Must be an array. Like { media: [ {...}, {...} ] }")), d(n), m(p, function(r) {
    return d(r);
  }), m(u, function(r) {
    return d(r.option);
  });
  function d(r) {
    m(i, function(g) {
      g(r, a);
    });
  }
  return {
    baseOption: n,
    timelineOptions: p || [],
    mediaDefault: o,
    mediaList: u
  };
}
function E(e, i, a) {
  var u = {
    width: i,
    height: a,
    aspectratio: i / a
    // lower case for convenience.
  }, o = !0;
  return m(e, function(n, t) {
    var s = t.match(B);
    if (!(!s || !s[1] || !s[2])) {
      var p = s[1], f = s[2].toLowerCase();
      I(u[f], n, p) || (o = !1);
    }
  }), o;
}
function I(e, i, a) {
  return a === "min" ? e >= i : a === "max" ? e <= i : e === i;
}
function q(e, i) {
  return e.join(",") === i.join(",");
}
export {
  N as default
};
