// echarts@6.0.0/lib/model/referHelper.js
import { each as y, retrieve as g, createHashMap as m } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { SINGLE_REFERRING as d } from "../util/model.js";
var p = (
  /** @class */
  /* @__PURE__ */ (function() {
    function o(e) {
      this.coordSysDims = [], this.axisMap = m(), this.categoryAxisMap = m(), this.coordSysName = e;
    }
    return o;
  })()
);
function D(o) {
  var e = o.get("coordinateSystem"), r = new p(e), n = u[e];
  if (n)
    return n(o, r, r.axisMap, r.categoryAxisMap), r;
}
var u = {
  cartesian2d: function(o, e, r, n) {
    var i = o.getReferringComponents("xAxis", d).models[0], s = o.getReferringComponents("yAxis", d).models[0];
    if (process.env.NODE_ENV !== "production") {
      if (!i)
        throw new Error('xAxis "' + g(o.get("xAxisIndex"), o.get("xAxisId"), 0) + '" not found');
      if (!s)
        throw new Error('yAxis "' + g(o.get("xAxisIndex"), o.get("yAxisId"), 0) + '" not found');
    }
    e.coordSysDims = ["x", "y"], r.set("x", i), r.set("y", s), a(i) && (n.set("x", i), e.firstCategoryDimIndex = 0), a(s) && (n.set("y", s), e.firstCategoryDimIndex == null && (e.firstCategoryDimIndex = 1));
  },
  singleAxis: function(o, e, r, n) {
    var i = o.getReferringComponents("singleAxis", d).models[0];
    if (process.env.NODE_ENV !== "production" && !i)
      throw new Error("singleAxis should be specified.");
    e.coordSysDims = ["single"], r.set("single", i), a(i) && (n.set("single", i), e.firstCategoryDimIndex = 0);
  },
  polar: function(o, e, r, n) {
    var i = o.getReferringComponents("polar", d).models[0], s = i.findAxisModel("radiusAxis"), t = i.findAxisModel("angleAxis");
    if (process.env.NODE_ENV !== "production") {
      if (!t)
        throw new Error("angleAxis option not found");
      if (!s)
        throw new Error("radiusAxis option not found");
    }
    e.coordSysDims = ["radius", "angle"], r.set("radius", s), r.set("angle", t), a(s) && (n.set("radius", s), e.firstCategoryDimIndex = 0), a(t) && (n.set("angle", t), e.firstCategoryDimIndex == null && (e.firstCategoryDimIndex = 1));
  },
  geo: function(o, e, r, n) {
    e.coordSysDims = ["lng", "lat"];
  },
  parallel: function(o, e, r, n) {
    var i = o.ecModel, s = i.getComponent("parallel", o.get("parallelIndex")), t = e.coordSysDims = s.dimensions.slice();
    y(s.parallelAxisIndex, function(c, x) {
      var f = i.getComponent("parallelAxis", c), l = t[x];
      r.set(l, f), a(f) && (n.set(l, f), e.firstCategoryDimIndex == null && (e.firstCategoryDimIndex = x));
    });
  },
  matrix: function(o, e, r, n) {
    var i = o.getReferringComponents("matrix", d).models[0];
    if (process.env.NODE_ENV !== "production" && !i)
      throw new Error("matrix coordinate system should be specified.");
    e.coordSysDims = ["x", "y"];
    var s = i.getDimensionModel("x"), t = i.getDimensionModel("y");
    r.set("x", s), r.set("y", t), n.set("x", s), n.set("y", t);
  }
};
function a(o) {
  return o.get("type") === "category";
}
export {
  D as getCoordSysInfoBySeries
};
