// echarts@6.0.0/lib/model/Series.js
import { __extends as N } from "../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { assert as x, merge as D, isTypedArray as B, createHashMap as L, keys as O, isObject as F, each as P, curry as H, concatArray as V, mixin as b } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import j from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/env.js";
import { makeInner as U, defaultEmphasis as T, isNameSpecified as z } from "../util/model.js";
import y from "./Component.js";
import { PaletteMixin as C } from "./mixin/palette.js";
import { DataFormatMixin as G } from "./mixin/dataFormat.js";
import { fetchLayoutMode as I, getLayoutParams as K, mergeLayoutParam as _ } from "../util/layout.js";
import { createTask as W } from "../core/task.js";
import { mountExtend as J } from "../util/clazz.js";
import { SourceManager as Q } from "../data/helper/sourceManager.js";
import { defaultSeriesFormatTooltip as X } from "../component/tooltip/seriesFormatTooltip.js";
var c = U();
function v(n, e) {
  return n.getName(e) || n.getId(e);
}
var Y = "__universalTransitionEnabled", S = (
  /** @class */
  (function(n) {
    N(e, n);
    function e() {
      var t = n !== null && n.apply(this, arguments) || this;
      return t._selectedDataIndicesMap = {}, t;
    }
    return e.prototype.init = function(t, a, r) {
      this.seriesIndex = this.componentIndex, this.dataTask = W({
        count: $,
        reset: q
      }), this.dataTask.context = {
        model: this
      }, this.mergeDefaultAndTheme(t, r);
      var i = c(this).sourceManager = new Q(this);
      i.prepareSource();
      var o = this.getInitialData(t, r);
      k(o, this), this.dataTask.context.data = o, process.env.NODE_ENV !== "production" && x(o, "getInitialData returned invalid data."), c(this).dataBeforeProcessed = o, A(this), this._initSelectedMapFromData(o);
    }, e.prototype.mergeDefaultAndTheme = function(t, a) {
      var r = I(this), i = r ? K(t) : {}, o = this.subType;
      y.hasClass(o) && (o += "Series"), D(t, a.getTheme().get(this.subType)), D(t, this.getDefaultOption()), T(t, "label", ["show"]), this.fillDataTextStyle(t.data), r && _(t, i, r);
    }, e.prototype.mergeOption = function(t, a) {
      t = D(this.option, t, !0), this.fillDataTextStyle(t.data);
      var r = I(this);
      r && _(this.option, t, r);
      var i = c(this).sourceManager;
      i.dirty(), i.prepareSource();
      var o = this.getInitialData(t, a);
      k(o, this), this.dataTask.dirty(), this.dataTask.context.data = o, c(this).dataBeforeProcessed = o, A(this), this._initSelectedMapFromData(o);
    }, e.prototype.fillDataTextStyle = function(t) {
      if (t && !B(t))
        for (var a = ["show"], r = 0; r < t.length; r++)
          t[r] && t[r].label && T(t[r], "label", a);
    }, e.prototype.getInitialData = function(t, a) {
    }, e.prototype.appendData = function(t) {
      var a = this.getRawData();
      a.appendData(t.data);
    }, e.prototype.getData = function(t) {
      var a = M(this);
      if (a) {
        var r = a.context.data;
        return t == null || !r.getLinkedData ? r : r.getLinkedData(t);
      } else
        return c(this).data;
    }, e.prototype.getAllData = function() {
      var t = this.getData();
      return t && t.getLinkedDataAll ? t.getLinkedDataAll() : [{
        data: t
      }];
    }, e.prototype.setData = function(t) {
      var a = M(this);
      if (a) {
        var r = a.context;
        r.outputData = t, a !== this.dataTask && (r.data = t);
      }
      c(this).data = t;
    }, e.prototype.getEncode = function() {
      var t = this.get("encode", !0);
      if (t)
        return L(t);
    }, e.prototype.getSourceManager = function() {
      return c(this).sourceManager;
    }, e.prototype.getSource = function() {
      return this.getSourceManager().getSource();
    }, e.prototype.getRawData = function() {
      return c(this).dataBeforeProcessed;
    }, e.prototype.getColorBy = function() {
      var t = this.get("colorBy");
      return t || "series";
    }, e.prototype.isColorBySeries = function() {
      return this.getColorBy() === "series";
    }, e.prototype.getBaseAxis = function() {
      var t = this.coordinateSystem;
      return t && t.getBaseAxis && t.getBaseAxis();
    }, e.prototype.indicesOfNearest = function(t, a, r, i) {
      var o = this.getData(), s = this.coordinateSystem, l = s && s.getAxis(t);
      if (!s || !l)
        return [];
      var f = l.dataToCoord(r);
      i == null && (i = 1 / 0);
      var p = [], d = 1 / 0, u = -1, h = 0;
      return o.each(a, function(E, w) {
        var R = l.dataToCoord(E), g = f - R, m = Math.abs(g);
        m <= i && ((m < d || m === d && g >= 0 && u < 0) && (d = m, u = g, h = 0), g === u && (p[h++] = w));
      }), p.length = h, p;
    }, e.prototype.formatTooltip = function(t, a, r) {
      return X({
        series: this,
        dataIndex: t,
        multipleSeries: a
      });
    }, e.prototype.isAnimationEnabled = function() {
      var t = this.ecModel;
      if (j.node && !(t && t.ssr))
        return !1;
      var a = this.getShallow("animation");
      return a && this.getData().count() > this.getShallow("animationThreshold") && (a = !1), !!a;
    }, e.prototype.restoreData = function() {
      this.dataTask.dirty();
    }, e.prototype.getColorFromPalette = function(t, a, r) {
      var i = this.ecModel, o = C.prototype.getColorFromPalette.call(this, t, a, r);
      return o || (o = i.getColorFromPalette(t, a, r)), o;
    }, e.prototype.coordDimToDataDim = function(t) {
      return this.getRawData().mapDimensionsAll(t);
    }, e.prototype.getProgressive = function() {
      return this.get("progressive");
    }, e.prototype.getProgressiveThreshold = function() {
      return this.get("progressiveThreshold");
    }, e.prototype.select = function(t, a) {
      this._innerSelect(this.getData(a), t);
    }, e.prototype.unselect = function(t, a) {
      var r = this.option.selectedMap;
      if (r) {
        var i = this.option.selectedMode, o = this.getData(a);
        if (i === "series" || r === "all") {
          this.option.selectedMap = {}, this._selectedDataIndicesMap = {};
          return;
        }
        for (var s = 0; s < t.length; s++) {
          var l = t[s], f = v(o, l);
          r[f] = !1, this._selectedDataIndicesMap[f] = -1;
        }
      }
    }, e.prototype.toggleSelect = function(t, a) {
      for (var r = [], i = 0; i < t.length; i++)
        r[0] = t[i], this.isSelected(t[i], a) ? this.unselect(r, a) : this.select(r, a);
    }, e.prototype.getSelectedDataIndices = function() {
      if (this.option.selectedMap === "all")
        return [].slice.call(this.getData().getIndices());
      for (var t = this._selectedDataIndicesMap, a = O(t), r = [], i = 0; i < a.length; i++) {
        var o = t[a[i]];
        o >= 0 && r.push(o);
      }
      return r;
    }, e.prototype.isSelected = function(t, a) {
      var r = this.option.selectedMap;
      if (!r)
        return !1;
      var i = this.getData(a);
      return (r === "all" || r[v(i, t)]) && !i.getItemModel(t).get(["select", "disabled"]);
    }, e.prototype.isUniversalTransitionEnabled = function() {
      if (this[Y])
        return !0;
      var t = this.option.universalTransition;
      return t ? t === !0 ? !0 : t && t.enabled : !1;
    }, e.prototype._innerSelect = function(t, a) {
      var r, i, o = this.option, s = o.selectedMode, l = a.length;
      if (!(!s || !l)) {
        if (s === "series")
          o.selectedMap = "all";
        else if (s === "multiple") {
          F(o.selectedMap) || (o.selectedMap = {});
          for (var f = o.selectedMap, p = 0; p < l; p++) {
            var d = a[p], u = v(t, d);
            f[u] = !0, this._selectedDataIndicesMap[u] = t.getRawIndex(d);
          }
        } else if (s === "single" || s === !0) {
          var h = a[l - 1], u = v(t, h);
          o.selectedMap = (r = {}, r[u] = !0, r), this._selectedDataIndicesMap = (i = {}, i[u] = t.getRawIndex(h), i);
        }
      }
    }, e.prototype._initSelectedMapFromData = function(t) {
      if (!this.option.selectedMap) {
        var a = [];
        t.hasItemOption && t.each(function(r) {
          var i = t.getRawDataItem(r);
          i && i.selected && a.push(r);
        }), a.length > 0 && this._innerSelect(t, a);
      }
    }, e.registerClass = function(t) {
      return y.registerClass(t);
    }, e.protoInitialize = (function() {
      var t = e.prototype;
      t.type = "series.__base__", t.seriesIndex = 0, t.ignoreStyleOnData = !1, t.hasSymbolVisual = !1, t.defaultSymbol = "circle", t.visualStyleAccessPath = "itemStyle", t.visualDrawType = "fill";
    })(), e;
  })(y)
);
b(S, G);
b(S, C);
J(S, y);
function A(n) {
  var e = n.name;
  z(n) || (n.name = Z(n) || e);
}
function Z(n) {
  var e = n.getRawData(), t = e.mapDimensionsAll("seriesName"), a = [];
  return P(t, function(r) {
    var i = e.getDimensionInfo(r);
    i.displayName && a.push(i.displayName);
  }), a.join(" ");
}
function $(n) {
  return n.model.getRawData().count();
}
function q(n) {
  var e = n.model;
  return e.setData(e.getRawData().cloneShallow()), tt;
}
function tt(n, e) {
  e.outputData && n.end > e.outputData.count() && e.model.getRawData().cloneShallow(e.outputData);
}
function k(n, e) {
  P(V(n.CHANGABLE_METHODS, n.DOWNSAMPLE_METHODS), function(t) {
    n.wrapMethod(t, H(et, e));
  });
}
function et(n, e) {
  var t = M(n);
  return t && t.setOutputEnd((e || this).count()), e;
}
function M(n) {
  var e = (n.ecModel || {}).scheduler, t = e && e.getPipeline(n.uid);
  if (t) {
    var a = t.currentTask;
    if (a) {
      var r = a.agentStubMap;
      r && (a = r.get(n.uid));
    }
    return a;
  }
}
export {
  Y as SERIES_UNIVERSAL_TRANSITION_PROP,
  S as default
};
