// @tiptap/extension-task-item@2.11.7/dist/index.js
import { Node as k, wrappingInputRule as m, mergeAttributes as f } from "../../../../../@tiptap_core@2.24.0_@tiptap_pm@2.24.0/node_modules/@tiptap/core/dist/index.js";
const b = /^\s*(\[([( |x])?\])\s$/, E = k.create({
  name: "taskItem",
  addOptions() {
    return {
      nested: !1,
      HTMLAttributes: {},
      taskListTypeName: "taskList"
    };
  },
  content() {
    return this.options.nested ? "paragraph block*" : "paragraph+";
  },
  defining: !0,
  addAttributes() {
    return {
      checked: {
        default: !1,
        keepOnSplit: !1,
        parseHTML: (e) => {
          const c = e.getAttribute("data-checked");
          return c === "" || c === "true";
        },
        renderHTML: (e) => ({
          "data-checked": e.checked
        })
      }
    };
  },
  parseHTML() {
    return [
      {
        tag: `li[data-type="${this.name}"]`,
        priority: 51
      }
    ];
  },
  renderHTML({ node: e, HTMLAttributes: c }) {
    return [
      "li",
      f(this.options.HTMLAttributes, c, {
        "data-type": this.name
      }),
      [
        "label",
        [
          "input",
          {
            type: "checkbox",
            checked: e.attrs.checked ? "checked" : null
          }
        ],
        ["span"]
      ],
      ["div", 0]
    ];
  },
  addKeyboardShortcuts() {
    const e = {
      Enter: () => this.editor.commands.splitListItem(this.name),
      "Shift-Tab": () => this.editor.commands.liftListItem(this.name)
    };
    return this.options.nested ? {
      ...e,
      Tab: () => this.editor.commands.sinkListItem(this.name)
    } : e;
  },
  addNodeView() {
    return ({ node: e, HTMLAttributes: c, getPos: h, editor: a }) => {
      const s = document.createElement("li"), i = document.createElement("label"), l = document.createElement("span"), n = document.createElement("input"), u = document.createElement("div");
      return i.contentEditable = "false", n.type = "checkbox", n.addEventListener("mousedown", (t) => t.preventDefault()), n.addEventListener("change", (t) => {
        if (!a.isEditable && !this.options.onReadOnlyChecked) {
          n.checked = !n.checked;
          return;
        }
        const { checked: r } = t.target;
        a.isEditable && typeof h == "function" && a.chain().focus(void 0, { scrollIntoView: !1 }).command(({ tr: p }) => {
          const d = h();
          if (typeof d != "number")
            return !1;
          const o = p.doc.nodeAt(d);
          return p.setNodeMarkup(d, void 0, {
            ...o?.attrs,
            checked: r
          }), !0;
        }).run(), !a.isEditable && this.options.onReadOnlyChecked && (this.options.onReadOnlyChecked(e, r) || (n.checked = !n.checked));
      }), Object.entries(this.options.HTMLAttributes).forEach(([t, r]) => {
        s.setAttribute(t, r);
      }), s.dataset.checked = e.attrs.checked, n.checked = e.attrs.checked, i.append(n, l), s.append(i, u), Object.entries(c).forEach(([t, r]) => {
        s.setAttribute(t, r);
      }), {
        dom: s,
        contentDOM: u,
        update: (t) => t.type !== this.type ? !1 : (s.dataset.checked = t.attrs.checked, n.checked = t.attrs.checked, !0)
      };
    };
  },
  addInputRules() {
    return [
      m({
        find: b,
        type: this.type,
        getAttributes: (e) => ({
          checked: e[e.length - 1] === "x"
        })
      })
    ];
  }
});
export {
  E as TaskItem,
  E as default,
  b as inputRegex
};
