// echarts@6.0.0/lib/component/title/install.js
import { __extends as C } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { retrieve2 as V } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { getECData as w } from "../../util/innerStore.js";
import { createTextStyle as S } from "../../label/labelStyle.js";
import { createBoxLayoutReference as E, getLayoutRect as O } from "../../util/layout.js";
import D from "../../model/Component.js";
import I from "../../view/Component.js";
import { windowOpen as R } from "../../util/format.js";
import m from "../../visual/tokens.js";
import k from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Text.js";
import L from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Rect.js";
var G = (
  /** @class */
  (function(o) {
    C(e, o);
    function e() {
      var t = o !== null && o.apply(this, arguments) || this;
      return t.type = e.type, t.layoutMode = {
        type: "box",
        ignoreSize: !0
      }, t;
    }
    return e.type = "title", e.defaultOption = {
      // zlevel: 0,
      z: 6,
      show: !0,
      text: "",
      target: "blank",
      subtext: "",
      subtarget: "blank",
      left: "center",
      top: m.size.m,
      backgroundColor: m.color.transparent,
      borderColor: m.color.primary,
      borderWidth: 0,
      padding: 5,
      itemGap: 10,
      textStyle: {
        fontSize: 18,
        fontWeight: "bold",
        color: m.color.primary
      },
      subtextStyle: {
        fontSize: 12,
        color: m.color.quaternary
      }
    }, e;
  })(D)
), P = (
  /** @class */
  (function(o) {
    C(e, o);
    function e() {
      var t = o !== null && o.apply(this, arguments) || this;
      return t.type = e.type, t;
    }
    return e.prototype.render = function(t, W, T) {
      if (this.group.removeAll(), !!t.get("show")) {
        var i = this.group, x = t.getModel("textStyle"), h = t.getModel("subtextStyle"), a = t.get("textAlign"), n = V(t.get("textBaseline"), t.get("textVerticalAlign")), l = new k({
          style: S(x, {
            text: t.get("text"),
            fill: x.getTextColor()
          }, {
            disableBox: !0
          }),
          z2: 10
        }), z = l.getBoundingRect(), v = t.get("subtext"), p = new k({
          style: S(h, {
            text: v,
            fill: h.getTextColor(),
            y: z.height + t.get("itemGap"),
            verticalAlign: "top"
          }, {
            disableBox: !0
          }),
          z2: 10
        }), c = t.get("link"), y = t.get("sublink"), f = t.get("triggerEvent", !0);
        l.silent = !c && !f, p.silent = !y && !f, c && l.on("click", function() {
          R(c, "_" + t.get("target"));
        }), y && p.on("click", function() {
          R(y, "_" + t.get("subtarget"));
        }), w(l).eventData = w(p).eventData = f ? {
          componentType: "title",
          componentIndex: t.componentIndex
        } : null, i.add(l), v && i.add(p);
        var g = i.getBoundingRect(), s = t.getBoxLayoutParams();
        s.width = g.width, s.height = g.height;
        var B = E(t, T), r = O(s, B.refContainer, t.get("padding"));
        a || (a = t.get("left") || t.get("right"), a === "middle" && (a = "center"), a === "right" ? r.x += r.width : a === "center" && (r.x += r.width / 2)), n || (n = t.get("top") || t.get("bottom"), n === "center" && (n = "middle"), n === "bottom" ? r.y += r.height : n === "middle" && (r.y += r.height / 2), n = n || "top"), i.x = r.x, i.y = r.y, i.markRedraw();
        var d = {
          align: a,
          verticalAlign: n
        };
        l.setStyle(d), p.setStyle(d), g = i.getBoundingRect();
        var u = r.margin, b = t.getItemStyle(["color", "opacity"]);
        b.fill = t.get("backgroundColor");
        var A = new L({
          shape: {
            x: g.x - u[3],
            y: g.y - u[0],
            width: g.width + u[1] + u[3],
            height: g.height + u[0] + u[2],
            r: t.get("borderRadius")
          },
          style: b,
          subPixelOptimize: !0,
          silent: !0
        });
        i.add(A);
      }
    }, e.type = "title", e;
  })(I)
);
function X(o) {
  o.registerComponentModel(G), o.registerComponentView(P);
}
export {
  X as install
};
