// echarts@6.0.0/lib/animation/basicTransition.js
import { isFunction as c, isObject as O, retrieve2 as y } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { makeInner as A } from "../util/model.js";
var D = A();
function F(i, n, o, e, t) {
  var r;
  if (n && n.ecModel) {
    var u = n.ecModel.getUpdatePayload();
    r = u && u.animation;
  }
  var m = n && n.isAnimationEnabled(), s = i === "update";
  if (m) {
    var v = void 0, a = void 0, f = void 0;
    e ? (v = y(e.duration, 200), a = y(e.easing, "cubicOut"), f = 0) : (v = n.getShallow(s ? "animationDurationUpdate" : "animationDuration"), a = n.getShallow(s ? "animationEasingUpdate" : "animationEasing"), f = n.getShallow(s ? "animationDelayUpdate" : "animationDelay")), r && (r.duration != null && (v = r.duration), r.easing != null && (a = r.easing), r.delay != null && (f = r.delay)), c(f) && (f = f(o, t)), c(v) && (v = v(o));
    var l = {
      duration: v || 0,
      delay: f,
      easing: a
    };
    return l;
  } else
    return null;
}
function p(i, n, o, e, t, r, u) {
  var m = !1, s;
  c(t) ? (u = r, r = t, t = null) : O(t) && (r = t.cb, u = t.during, m = t.isFrom, s = t.removeOpt, t = t.dataIndex);
  var v = i === "leave";
  v || n.stopAnimation("leave");
  var a = F(i, e, t, v ? s || {} : null, e && e.getAnimationDelayParams ? e.getAnimationDelayParams(n, t) : null);
  if (a && a.duration > 0) {
    var f = a.duration, l = a.delay, E = a.easing, g = {
      duration: f,
      delay: l || 0,
      easing: E,
      done: r,
      force: !!r || !!u,
      // Set to final state in update/init animation.
      // So the post processing based on the path shape can be done correctly.
      setToFinal: !v,
      scope: i,
      during: u
    };
    m ? n.animateFrom(o, g) : n.animateTo(o, g);
  } else
    n.stopAnimation(), !m && n.attr(o), u && u(1), r && r();
}
function G(i, n, o, e, t, r) {
  p("update", i, n, o, e, t, r);
}
function R(i, n, o, e, t, r) {
  p("enter", i, n, o, e, t, r);
}
function P(i) {
  if (!i.__zr)
    return !0;
  for (var n = 0; n < i.animators.length; n++) {
    var o = i.animators[n];
    if (o.scope === "leave")
      return !0;
  }
  return !1;
}
function h(i, n, o, e, t, r) {
  P(i) || p("leave", i, n, o, e, t, r);
}
function S(i, n, o, e) {
  i.removeTextContent(), i.removeTextGuideLine(), h(i, {
    style: {
      opacity: 0
    }
  }, n, o, e);
}
function C(i, n, o) {
  function e() {
    i.parent && i.parent.remove(i);
  }
  i.isGroup ? i.traverse(function(t) {
    t.isGroup || S(t, n, o, e);
  }) : S(i, n, o, e);
}
function T(i) {
  D(i).oldStyle = i.style;
}
function _(i) {
  return D(i).oldStyle;
}
export {
  F as getAnimationConfig,
  _ as getOldStyle,
  R as initProps,
  P as isElementRemoved,
  h as removeElement,
  C as removeElementWithFadeOut,
  T as saveOldStyle,
  D as transitionStore,
  G as updateProps
};
