// echarts@6.0.0/lib/animation/customGraphicTransition.js
import { makeInner as U, normalizeToArray as V } from "../util/model.js";
import { each as X, extend as x, bind as j, keys as A, indexOf as B, hasOwn as p, isArrayLike as q, eqNaN as z, assert as L, reduce as H } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { cloneValue as J } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/animation/Animator.js";
import Q from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Displayable.js";
import { getAnimationConfig as W } from "./basicTransition.js";
import "../util/graphic.js";
import { warn as C } from "../util/log.js";
import { TRANSFORMABLE_PROPS as N } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/Transformable.js";
import Z from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Path.js";
var O = {
  position: ["x", "y"],
  scale: ["scaleX", "scaleY"],
  origin: ["originX", "originY"]
}, F = A(O), y = H(N, function(r, n) {
  return r[n] = 1, r;
}, {}), M = N.join(", "), _ = ["", "style", "shape", "extra"], D = U();
function I(r, n, a, t, s) {
  var i = r + "Animation", e = W(r, t, s) || {}, o = D(n).userDuring;
  return e.duration > 0 && (e.during = o ? j(tr, {
    el: n,
    userDuring: o
  }) : null, e.setToFinal = !0, e.scope = r), x(e, a[i]), e;
}
function Er(r, n, a, t) {
  t = t || {};
  var s = t.dataIndex, i = t.isInit, e = t.clearStyle, o = a.isAnimationEnabled(), d = D(r), u = n.style;
  d.userDuring = n.during;
  var v = {}, f = {};
  if (er(r, n, f), r.type === "compound")
    for (var h = r.shape.paths, g = n.shape.paths, l = 0; l < g.length; l++) {
      var G = g[l];
      P("shape", G, h[l]);
    }
  else
    P("shape", n, f), P("extra", n, f);
  if (!i && o && (ar(r, n, v), w("shape", r, n, v), w("extra", r, n, v), ir(r, n, u, v)), f.style = u, K(r, f, e), rr(r, n), o)
    if (i) {
      var T = {};
      X(_, function(E) {
        var R = E ? n[E] : n;
        R && R.enterFrom && (E && (T[E] = T[E] || {}), x(E ? T[E] : T, R.enterFrom));
      });
      var b = I("enter", r, n, a, s);
      b.duration > 0 && r.animateFrom(T, b);
    } else
      k(r, n, s || 0, a, v);
  $(r, n), u ? r.dirty() : r.markRedraw();
}
function $(r, n) {
  for (var a = D(r).leaveToProps, t = 0; t < _.length; t++) {
    var s = _[t], i = s ? n[s] : n;
    i && i.leaveTo && (a || (a = D(r).leaveToProps = {}), s && (a[s] = a[s] || {}), x(s ? a[s] : a, i.leaveTo));
  }
}
function pr(r, n, a, t) {
  if (r) {
    var s = r.parent, i = D(r).leaveToProps;
    if (i) {
      var e = I("update", r, n, a, 0);
      e.done = function() {
        s && s.remove(r);
      }, r.animateTo(i, e);
    } else
      s && s.remove(r);
  }
}
function m(r) {
  return r === "all";
}
function K(r, n, a) {
  var t = n.style;
  if (!r.isGroup && t) {
    if (a) {
      r.useStyle({});
      for (var s = r.animators, i = 0; i < s.length; i++) {
        var e = s[i];
        e.targetName === "style" && e.changeTarget(r.style);
      }
    }
    r.setStyle(t);
  }
  n && (n.style = null, n && r.attr(n), n.style = t);
}
function k(r, n, a, t, s) {
  if (s) {
    var i = I("update", r, n, t, a);
    i.duration > 0 && r.animateFrom(s, i);
  }
}
function rr(r, n) {
  p(n, "silent") && (r.silent = n.silent), p(n, "ignore") && (r.ignore = n.ignore), r instanceof Q && p(n, "invisible") && (r.invisible = n.invisible), r instanceof Z && p(n, "autoBatch") && (r.autoBatch = n.autoBatch);
}
var c = {}, nr = {
  // Usually other props do not need to be changed in animation during.
  setTransform: function(r, n) {
    return process.env.NODE_ENV !== "production" && L(p(y, r), "Only " + M + " available in `setTransform`."), c.el[r] = n, this;
  },
  getTransform: function(r) {
    return process.env.NODE_ENV !== "production" && L(p(y, r), "Only " + M + " available in `getTransform`."), c.el[r];
  },
  setShape: function(r, n) {
    process.env.NODE_ENV !== "production" && S(r);
    var a = c.el, t = a.shape || (a.shape = {});
    return t[r] = n, a.dirtyShape && a.dirtyShape(), this;
  },
  getShape: function(r) {
    process.env.NODE_ENV !== "production" && S(r);
    var n = c.el.shape;
    if (n)
      return n[r];
  },
  setStyle: function(r, n) {
    process.env.NODE_ENV !== "production" && S(r);
    var a = c.el, t = a.style;
    return t && (process.env.NODE_ENV !== "production" && z(n) && C("style." + r + " must not be assigned with NaN."), t[r] = n, a.dirtyStyle && a.dirtyStyle()), this;
  },
  getStyle: function(r) {
    process.env.NODE_ENV !== "production" && S(r);
    var n = c.el.style;
    if (n)
      return n[r];
  },
  setExtra: function(r, n) {
    process.env.NODE_ENV !== "production" && S(r);
    var a = c.el.extra || (c.el.extra = {});
    return a[r] = n, this;
  },
  getExtra: function(r) {
    process.env.NODE_ENV !== "production" && S(r);
    var n = c.el.extra;
    if (n)
      return n[r];
  }
};
function S(r) {
  if (process.env.NODE_ENV !== "production" && (r === "transition" || r === "enterFrom" || r === "leaveTo"))
    throw new Error('key must not be "' + r + '"');
}
function tr() {
  var r = this, n = r.el;
  if (n) {
    var a = D(n).userDuring, t = r.userDuring;
    if (a !== t) {
      r.el = r.userDuring = null;
      return;
    }
    c.el = n, t(nr);
  }
}
function w(r, n, a, t) {
  var s = a[r];
  if (s) {
    var i = n[r], e;
    if (i) {
      var o = a.transition, d = s.transition;
      if (d)
        if (!e && (e = t[r] = {}), m(d))
          x(e, i);
        else
          for (var u = V(d), v = 0; v < u.length; v++) {
            var f = u[v], h = i[f];
            e[f] = h;
          }
      else if (m(o) || B(o, r) >= 0) {
        !e && (e = t[r] = {});
        for (var g = A(i), v = 0; v < g.length; v++) {
          var f = g[v], h = i[f];
          sr(s[f], h) && (e[f] = h);
        }
      }
    }
  }
}
function P(r, n, a) {
  var t = n[r];
  if (t)
    for (var s = a[r] = {}, i = A(t), e = 0; e < i.length; e++) {
      var o = i[e];
      s[o] = J(t[o]);
    }
}
function ar(r, n, a) {
  for (var t = n.transition, s = m(t) ? N : V(t || []), i = 0; i < s.length; i++) {
    var e = s[i];
    if (!(e === "style" || e === "shape" || e === "extra")) {
      var o = r[e];
      process.env.NODE_ENV !== "production" && Y(e, "el.transition"), a[e] = o;
    }
  }
}
function er(r, n, a) {
  for (var t = 0; t < F.length; t++) {
    var s = F[t], i = O[s], e = n[s];
    e && (a[i[0]] = e[0], a[i[1]] = e[1]);
  }
  for (var t = 0; t < N.length; t++) {
    var o = N[t];
    n[o] != null && (a[o] = n[o]);
  }
}
function ir(r, n, a, t) {
  if (a) {
    var s = r.style, i;
    if (s) {
      var e = a.transition, o = n.transition;
      if (e && !m(e)) {
        var d = V(e);
        !i && (i = t.style = {});
        for (var u = 0; u < d.length; u++) {
          var v = d[u], f = s[v];
          i[v] = f;
        }
      } else if (r.getAnimationStyleProps && (m(o) || m(e) || B(o, "style") >= 0)) {
        var h = r.getAnimationStyleProps(), g = h ? h.style : null;
        if (g) {
          !i && (i = t.style = {});
          for (var l = A(a), u = 0; u < l.length; u++) {
            var v = l[u];
            if (g[v]) {
              var f = s[v];
              i[v] = f;
            }
          }
        }
      }
    }
  }
}
function sr(r, n) {
  return q(r) ? r !== n : r != null && isFinite(r);
}
var Y;
process.env.NODE_ENV !== "production" && (Y = function(r, n) {
  p(y, r) || C("Prop `" + r + "` is not a permitted in `" + n + "`. Only `" + A(y).join("`, `") + "` are permitted.");
});
export {
  _ as ELEMENT_ANIMATABLE_PROPS,
  pr as applyLeaveTransition,
  Er as applyUpdateTransition,
  m as isTransitionAll,
  $ as updateLeaveTo
};
