// echarts@6.0.0/lib/animation/morphTransitionHelper.js
import { isCombineMorphing as Q, morphPath as R, combineMorph as U, separateMorph as V } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/tool/morphPath.js";
import "../util/graphic.js";
import { isArray as H, defaults as O } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { getAnimationConfig as W } from "./basicTransition.js";
import { clonePath as X } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/tool/path.js";
import Y from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Path.js";
function E(r) {
  return H(r[0]);
}
function G(r, a) {
  for (var t = [], i = r.length, n = 0; n < i; n++)
    t.push({
      one: r[n],
      many: []
    });
  for (var n = 0; n < a.length; n++) {
    var u = a[n].length, h = void 0;
    for (h = 0; h < u; h++)
      t[h % i].many.push(a[n][h]);
  }
  for (var e = 0, n = i - 1; n >= 0; n--)
    if (!t[n].many.length) {
      var v = t[e].many;
      if (v.length <= 1)
        if (e)
          e = 0;
        else
          return t;
      var u = v.length, l = Math.ceil(u / 2);
      t[n].many = v.slice(l, u), t[e].many = v.slice(0, l), e++;
    }
  return t;
}
var Z = {
  clone: function(r) {
    for (var a = [], t = 1 - Math.pow(1 - r.path.style.opacity, 1 / r.count), i = 0; i < r.count; i++) {
      var n = X(r.path);
      n.setStyle("opacity", t), a.push(n);
    }
    return a;
  },
  // Use the default divider
  split: null
};
function ar(r, a, t, i, n, u) {
  if (!r.length || !a.length)
    return;
  var h = W("update", i, n);
  if (!(h && h.duration > 0))
    return;
  var e = i.getModel("universalTransition").get("delay"), v = Object.assign({
    // Need to setToFinal so the further calculation based on the style can be correct.
    // Like emphasis color.
    setToFinal: !0
  }, h), l, A;
  E(r) && (l = r, A = a), E(a) && (l = a, A = r);
  function j(c, s, C, D, J) {
    var p = c.many, d = c.one;
    if (p.length === 1 && !J) {
      var y = s ? p[0] : d, M = s ? d : p[0];
      if (Q(y))
        j({
          many: [y],
          one: M
        }, !0, C, D, !0);
      else {
        var P = e ? O({
          delay: e(C, D)
        }, v) : v;
        R(y, M, P), u(y, M, y, M, P);
      }
    } else
      for (var L = O({
        dividePath: Z[t],
        individualDelay: e && function(N, b, m, T) {
          return e(N + C, D);
        }
      }, v), S = s ? U(p, d, L) : V(d, p, L), q = S.fromIndividuals, K = S.toIndividuals, z = q.length, f = 0; f < z; f++) {
        var P = e ? O({
          delay: e(f, z)
        }, v) : v;
        u(q[f], K[f], s ? p[f] : c.one, s ? c.one : p[f], P);
      }
  }
  for (var B = l ? l === r : r.length > a.length, g = l ? G(A, l) : G(B ? a : r, [B ? r : a]), w = 0, o = 0; o < g.length; o++)
    w += g[o].many.length;
  for (var x = 0, o = 0; o < g.length; o++)
    j(g[o], B, x, w), x += g[o].many.length;
}
function $(r) {
  if (!r)
    return [];
  if (H(r)) {
    for (var a = [], t = 0; t < r.length; t++)
      a.push($(r[t]));
    return a;
  }
  var i = [];
  return r.traverse(function(n) {
    n instanceof Y && !n.disableMorphing && !n.invisible && !n.ignore && i.push(n);
  }), i;
}
export {
  ar as applyMorphAnimation,
  $ as getPathList
};
