// echarts@6.0.0/lib/animation/universalTransition.js
import { SERIES_UNIVERSAL_TRANSITION_PROP as P } from "../model/Series.js";
import { each as m, createHashMap as x, isArray as L, filter as B, map as H, extend as _ } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { applyMorphAnimation as R, getPathList as N } from "./morphTransitionHelper.js";
import X from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Path.js";
import "../util/graphic.js";
import b from "../data/DataDiffer.js";
import { makeInner as ee, normalizeToArray as k } from "../util/model.js";
import { warn as Y } from "../util/log.js";
import { getAnimationConfig as Z, initProps as te, getOldStyle as re } from "./basicTransition.js";
import ae from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Displayable.js";
var $ = 1e4, ie = 0, j = 1, q = 2, ne = ee();
function oe(t, a) {
  for (var i = t.dimensions, e = 0; e < i.length; e++) {
    var r = t.getDimensionInfo(i[e]);
    if (r && r.otherDims[a] === 0)
      return i[e];
  }
}
function de(t, a, i) {
  var e = t.getDimensionInfo(i), r = e && e.ordinalMeta;
  if (e) {
    var u = t.get(e.name, a);
    return r && r.categories[u] || u + "";
  }
}
function z(t, a, i, e) {
  var r = e ? "itemChildGroupId" : "itemGroupId", u = oe(t, r);
  if (u) {
    var n = de(t, a, u);
    return n;
  }
  var d = t.getRawDataItem(a), p = e ? "childGroupId" : "groupId";
  if (d && d[p])
    return d[p] + "";
  if (!e)
    return i || t.getId(a);
}
function w(t) {
  var a = [];
  return m(t, function(i) {
    var e = i.data, r = i.dataGroupId;
    if (e.count() > $) {
      process.env.NODE_ENV !== "production" && Y("Universal transition is disabled on large data > 10k.");
      return;
    }
    for (var u = e.getIndices(), n = 0; n < u.length; n++)
      a.push({
        data: e,
        groupId: z(e, n, r, !1),
        childGroupId: z(e, n, r, !0),
        divide: i.divide,
        dataIndex: n
      });
  }), a;
}
function C(t, a, i) {
  t.traverse(function(e) {
    e instanceof X && te(e, {
      style: {
        opacity: 0
      }
    }, a, {
      dataIndex: i,
      isFrom: !0
    });
  });
}
function U(t) {
  if (t.parent) {
    var a = t.getComputedTransform();
    t.setLocalTransform(a), t.parent.remove(t);
  }
}
function O(t) {
  t.stopAnimation(), t.isGroup && t.traverse(function(a) {
    a.stopAnimation();
  });
}
function ue(t, a, i) {
  var e = Z("update", i, a);
  e && t.traverse(function(r) {
    if (r instanceof ae) {
      var u = re(r);
      u && r.animateFrom({
        style: u
      }, e);
    }
  });
}
function se(t, a) {
  var i = t.length;
  if (i !== a.length)
    return !1;
  for (var e = 0; e < i; e++) {
    var r = t[e], u = a[e];
    if (r.data.getId(r.dataIndex) !== u.data.getId(u.dataIndex))
      return !1;
  }
  return !0;
}
function K(t, a, i) {
  var e = w(t), r = w(a);
  function u(s, G, o, f, I) {
    (o || s) && G.animateFrom({
      style: o && o !== s ? _(_({}, o.style), s.style) : s.style
    }, I);
  }
  var n = !1, d = ie, p = x(), S = x();
  e.forEach(function(s) {
    s.groupId && p.set(s.groupId, !0), s.childGroupId && S.set(s.childGroupId, !0);
  });
  for (var l = 0; l < r.length; l++) {
    var h = r[l].groupId;
    if (S.get(h)) {
      d = j;
      break;
    }
    var T = r[l].childGroupId;
    if (T && p.get(T)) {
      d = q;
      break;
    }
  }
  function g(s, G) {
    return function(o) {
      var f = o.data, I = o.dataIndex;
      return G ? f.getId(I) : s ? d === j ? o.childGroupId : o.groupId : d === q ? o.childGroupId : o.groupId;
    };
  }
  var D = se(e, r), A = {};
  if (!D)
    for (var l = 0; l < r.length; l++) {
      var M = r[l], V = M.data.getItemGraphicEl(M.dataIndex);
      V && (A[V.id] = !0);
    }
  function F(s, G) {
    var o = e[G], f = r[s], I = f.data.hostModel, c = o.data.getItemGraphicEl(o.dataIndex), v = f.data.getItemGraphicEl(f.dataIndex);
    if (c === v) {
      v && ue(v, f.dataIndex, I);
      return;
    }
    // We can't use the elements that already being morphed
    c && A[c.id] || v && (O(v), c ? (O(c), U(c), n = !0, R(N(c), N(v), f.divide, I, s, u)) : C(v, I, s));
  }
  new b(e, r, g(!0, D), g(!1, D), null, "multiple").update(F).updateManyToOne(function(s, G) {
    var o = r[s], f = o.data, I = f.hostModel, c = f.getItemGraphicEl(o.dataIndex), v = B(H(G, function(E) {
      return e[E].data.getItemGraphicEl(e[E].dataIndex);
    }), function(E) {
      return E && E !== c && !A[E.id];
    });
    c && (O(c), v.length ? (m(v, function(E) {
      O(E), U(E);
    }), n = !0, R(N(v), N(c), o.divide, I, s, u)) : C(c, I, o.dataIndex));
  }).updateOneToMany(function(s, G) {
    var o = e[G], f = o.data.getItemGraphicEl(o.dataIndex);
    if (!(f && A[f.id])) {
      var I = B(H(s, function(v) {
        return r[v].data.getItemGraphicEl(r[v].dataIndex);
      }), function(v) {
        return v && v !== f;
      }), c = r[s[0]].data.hostModel;
      I.length && (m(I, function(v) {
        return O(v);
      }), f ? (O(f), U(f), n = !0, R(
        N(f),
        N(I),
        o.divide,
        // Use divide on old.
        c,
        s[0],
        u
      )) : m(I, function(v) {
        return C(v, c, s[0]);
      }));
    }
  }).updateManyToMany(function(s, G) {
    new b(G, s, function(o) {
      return e[o].data.getId(e[o].dataIndex);
    }, function(o) {
      return r[o].data.getId(r[o].dataIndex);
    }).update(function(o, f) {
      F(s[o], G[f]);
    }).execute();
  }).execute(), n && m(a, function(s) {
    var G = s.data, o = G.hostModel, f = o && i.getViewOfSeriesModel(o), I = Z("update", o, 0);
    f && o.isAnimationEnabled() && I && I.duration > 0 && f.group.traverse(function(c) {
      c instanceof X && !c.animators.length && c.animateFrom({
        style: {
          opacity: 0
        }
      }, I);
    });
  });
}
function J(t) {
  var a = t.getModel("universalTransition").get("seriesKey");
  return a || t.id;
}
function Q(t) {
  return L(t) ? t.sort().join(",") : t;
}
function y(t) {
  if (t.hostModel)
    return t.hostModel.getModel("universalTransition").get("divideShape");
}
function fe(t, a) {
  var i = x(), e = x(), r = x();
  m(t.oldSeries, function(n, d) {
    var p = t.oldDataGroupIds[d], S = t.oldData[d], l = J(n), h = Q(l);
    e.set(h, {
      dataGroupId: p,
      data: S
    }), L(l) && m(l, function(T) {
      r.set(T, {
        key: h,
        dataGroupId: p,
        data: S
      });
    });
  });
  function u(n) {
    i.get(n) && Y("Duplicated seriesKey in universalTransition " + n);
  }
  return m(a.updatedSeries, function(n) {
    if (n.isUniversalTransitionEnabled() && n.isAnimationEnabled()) {
      var d = n.get("dataGroupId"), p = n.getData(), S = J(n), l = Q(S), h = e.get(l);
      if (h)
        process.env.NODE_ENV !== "production" && u(l), i.set(l, {
          oldSeries: [{
            dataGroupId: h.dataGroupId,
            divide: y(h.data),
            data: h.data
          }],
          newSeries: [{
            dataGroupId: d,
            divide: y(p),
            data: p
          }]
        });
      else if (L(S)) {
        process.env.NODE_ENV !== "production" && u(l);
        var T = [];
        m(S, function(A) {
          var M = e.get(A);
          M.data && T.push({
            dataGroupId: M.dataGroupId,
            divide: y(M.data),
            data: M.data
          });
        }), T.length && i.set(l, {
          oldSeries: T,
          newSeries: [{
            dataGroupId: d,
            data: p,
            divide: y(p)
          }]
        });
      } else {
        var g = r.get(S);
        if (g) {
          var D = i.get(g.key);
          D || (D = {
            oldSeries: [{
              dataGroupId: g.dataGroupId,
              data: g.data,
              divide: y(g.data)
            }],
            newSeries: []
          }, i.set(g.key, D)), D.newSeries.push({
            dataGroupId: d,
            data: p,
            divide: y(p)
          });
        }
      }
    }
  }), i;
}
function W(t, a) {
  for (var i = 0; i < t.length; i++) {
    var e = a.seriesIndex != null && a.seriesIndex === t[i].seriesIndex || a.seriesId != null && a.seriesId === t[i].id;
    if (e)
      return i;
  }
}
function ve(t, a, i, e) {
  var r = [], u = [];
  m(k(t.from), function(n) {
    var d = W(a.oldSeries, n);
    d >= 0 && r.push({
      dataGroupId: a.oldDataGroupIds[d],
      data: a.oldData[d],
      // TODO can specify divideShape in transition.
      divide: y(a.oldData[d]),
      groupIdDim: n.dimension
    });
  }), m(k(t.to), function(n) {
    var d = W(i.updatedSeries, n);
    if (d >= 0) {
      var p = i.updatedSeries[d].getData();
      u.push({
        dataGroupId: a.oldDataGroupIds[d],
        data: p,
        divide: y(p),
        groupIdDim: n.dimension
      });
    }
  }), r.length > 0 && u.length > 0 && K(r, u, e);
}
function De(t) {
  t.registerUpdateLifecycle("series:beforeupdate", function(a, i, e) {
    m(k(e.seriesTransition), function(r) {
      m(k(r.to), function(u) {
        for (var n = e.updatedSeries, d = 0; d < n.length; d++)
          (u.seriesIndex != null && u.seriesIndex === n[d].seriesIndex || u.seriesId != null && u.seriesId === n[d].id) && (n[d][P] = !0);
      });
    });
  }), t.registerUpdateLifecycle("series:transition", function(a, i, e) {
    var r = ne(i);
    if (r.oldSeries && e.updatedSeries && e.optionChanged) {
      var u = e.seriesTransition;
      if (u)
        m(k(u), function(g) {
          ve(g, r, e, i);
        });
      else {
        var n = fe(r, e);
        m(n.keys(), function(g) {
          var D = n.get(g);
          K(D.oldSeries, D.newSeries, i);
        });
      }
      m(e.updatedSeries, function(g) {
        g[P] && (g[P] = !1);
      });
    }
    for (var d = a.getSeries(), p = r.oldSeries = [], S = r.oldDataGroupIds = [], l = r.oldData = [], h = 0; h < d.length; h++) {
      var T = d[h].getData();
      T.count() < $ && (p.push(d[h]), S.push(d[h].get("dataGroupId")), l.push(T));
    }
  });
}
export {
  De as installUniversalTransition
};
