// echarts@6.0.0/lib/component/polar/install.js
import { __extends as e } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { use as a } from "../../extension.js";
import m from "../axis/AxisView.js";
import n from "../axisPointer/PolarAxisPointer.js";
import { install as l } from "../axisPointer/install.js";
import p from "../../coord/polar/PolarModel.js";
import t from "../../coord/axisModelCreator.js";
import { AngleAxisModel as s, RadiusAxisModel as f } from "../../coord/polar/AxisModel.js";
import u from "../../coord/polar/polarCreator.js";
import x from "../axis/AngleAxisView.js";
import A from "../axis/RadiusAxisView.js";
import d from "../../view/Component.js";
import { curry as w } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import C from "../../layout/barPolar.js";
var P = {
  startAngle: 90,
  clockwise: !0,
  splitNumber: 12,
  axisLabel: {
    rotate: 0
  }
}, V = {
  splitNumber: 5
}, y = (
  /** @class */
  (function(o) {
    e(r, o);
    function r() {
      var i = o !== null && o.apply(this, arguments) || this;
      return i.type = r.type, i;
    }
    return r.type = "polar", r;
  })(d)
);
function $(o) {
  a(l), m.registerAxisPointerClass("PolarAxisPointer", n), o.registerCoordinateSystem("polar", u), o.registerComponentModel(p), o.registerComponentView(y), t(o, "angle", s, P), t(o, "radius", f, V), o.registerComponentView(x), o.registerComponentView(A), o.registerLayout(w(C, "bar"));
}
export {
  $ as install
};
