// echarts@6.0.0/lib/component/dataZoom/AxisProxy.js
import { clone as D, each as E, map as N } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { linearMap as g, asc as A, getPixelPrecision as I } from "../../util/number.js";
import P from "../helper/sliderMove.js";
import { unionAxisExtentFromData as R } from "../../coord/axisHelper.js";
import { ensureScaleRawExtentInfo as V } from "../../coord/scaleRawExtentInfo.js";
import { isCoordSupported as F, getAxisMainType as Z } from "./helper.js";
import { SINGLE_REFERRING as T } from "../../util/model.js";
var h = E, y = A, j = (
  /** @class */
  (function() {
    function u(n, e, r, t) {
      this._dimName = n, this._axisIndex = e, this.ecModel = t, this._dataZoomModel = r;
    }
    return u.prototype.hostedBy = function(n) {
      return this._dataZoomModel === n;
    }, u.prototype.getDataValueWindow = function() {
      return this._valueWindow.slice();
    }, u.prototype.getDataPercentWindow = function() {
      return this._percentWindow.slice();
    }, u.prototype.getTargetSeriesModels = function() {
      var n = [];
      return this.ecModel.eachSeries(function(e) {
        if (F(e)) {
          var r = Z(this._dimName), t = e.getReferringComponents(r, T).models[0];
          t && this._axisIndex === t.componentIndex && n.push(e);
        }
      }, this), n;
    }, u.prototype.getAxisModel = function() {
      return this.ecModel.getComponent(this._dimName + "Axis", this._axisIndex);
    }, u.prototype.getMinMaxSpan = function() {
      return D(this._minMaxSpan);
    }, u.prototype.calculateDataWindow = function(n) {
      var e = this._dataExtent, r = this.getAxisModel(), t = r.axis.scale, o = this._dataZoomModel.getRangePropMode(), a = [0, 100], v = [], d = [], p;
      h(["start", "end"], function(m, i) {
        var s = n[m], l = n[m + "Value"];
        o[i] === "percent" ? (s == null && (s = a[i]), l = t.parse(g(s, a, e))) : (p = !0, l = l == null ? e[i] : t.parse(l), s = g(l, e, a)), d[i] = l == null || isNaN(l) ? e[i] : l, v[i] = s == null || isNaN(s) ? a[i] : s;
      }), y(d), y(v);
      var f = this._minMaxSpan;
      p ? _(d, v, e, a, !1) : _(v, d, a, e, !0);
      function _(m, i, s, l, M) {
        var x = M ? "Span" : "ValueSpan";
        P(0, m, s, "all", f["min" + x], f["max" + x]);
        for (var c = 0; c < 2; c++)
          i[c] = g(m[c], s, l, !0), M && (i[c] = t.parse(i[c]));
      }
      return {
        valueWindow: d,
        percentWindow: v
      };
    }, u.prototype.reset = function(n) {
      if (n === this._dataZoomModel) {
        var e = this.getTargetSeriesModels();
        this._dataExtent = C(this, this._dimName, e), this._updateMinMaxSpan();
        var r = this.calculateDataWindow(n.settledOption);
        this._valueWindow = r.valueWindow, this._percentWindow = r.percentWindow, this._setAxisModel();
      }
    }, u.prototype.filterData = function(n, e) {
      if (n !== this._dataZoomModel)
        return;
      var r = this._dimName, t = this.getTargetSeriesModels(), o = n.get("filterMode"), a = this._valueWindow;
      if (o === "none")
        return;
      h(t, function(d) {
        var p = d.getData(), f = p.mapDimensionsAll(r);
        if (f.length) {
          if (o === "weakFilter") {
            var _ = p.getStore(), m = N(f, function(i) {
              return p.getDimensionIndex(i);
            }, p);
            p.filterSelf(function(i) {
              for (var s, l, M, x = 0; x < f.length; x++) {
                var c = _.get(m[x], i), S = !isNaN(c), w = c < a[0], W = c > a[1];
                if (S && !w && !W)
                  return !0;
                S && (M = !0), w && (s = !0), W && (l = !0);
              }
              return M && s && l;
            });
          } else
            h(f, function(i) {
              if (o === "empty")
                d.setData(p = p.map(i, function(l) {
                  return v(l) ? l : NaN;
                }));
              else {
                var s = {};
                s[i] = a, p.selectRange(s);
              }
            });
          h(f, function(i) {
            p.setApproximateExtent(a, i);
          });
        }
      });
      function v(d) {
        return d >= a[0] && d <= a[1];
      }
    }, u.prototype._updateMinMaxSpan = function() {
      var n = this._minMaxSpan = {}, e = this._dataZoomModel, r = this._dataExtent;
      h(["min", "max"], function(t) {
        var o = e.get(t + "Span"), a = e.get(t + "ValueSpan");
        a != null && (a = this.getAxisModel().axis.scale.parse(a)), a != null ? o = g(r[0] + a, r, [0, 100], !0) : o != null && (a = g(o, [0, 100], r, !0) - r[0]), n[t + "Span"] = o, n[t + "ValueSpan"] = a;
      }, this);
    }, u.prototype._setAxisModel = function() {
      var n = this.getAxisModel(), e = this._percentWindow, r = this._valueWindow;
      if (e) {
        var t = I(r, [0, 500]);
        t = Math.min(t, 20);
        var o = n.axis.scale.rawExtentInfo;
        e[0] !== 0 && o.setDeterminedMinMax("min", +r[0].toFixed(t)), e[1] !== 100 && o.setDeterminedMinMax("max", +r[1].toFixed(t)), o.freeze();
      }
    }, u;
  })()
);
function C(u, n, e) {
  var r = [1 / 0, -1 / 0];
  h(e, function(a) {
    R(r, a.getData(), n);
  });
  var t = u.getAxisModel(), o = V(t.axis.scale, t, r).calculate();
  return [o.min, o.max];
}
export {
  j as default
};
