// echarts@6.0.0/lib/component/dataZoom/dataZoomProcessor.js
import { createHashMap as f, each as x } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { getAxisMainType as u } from "./helper.js";
import p from "./AxisProxy.js";
var P = {
  // `dataZoomProcessor` will only be performed in needed series. Consider if
  // there is a line series and a pie series, it is better not to update the
  // line series if only pie series is needed to be updated.
  getTargetSeries: function(a) {
    function s(t) {
      a.eachComponent("dataZoom", function(e) {
        e.eachTargetAxis(function(i, o) {
          var c = a.getComponent(u(i), o);
          t(i, o, c, e);
        });
      });
    }
    s(function(t, e, i, o) {
      i.__dzAxisProxy = null;
    });
    var n = [];
    s(function(t, e, i, o) {
      i.__dzAxisProxy || (i.__dzAxisProxy = new p(t, e, o, a), n.push(i.__dzAxisProxy));
    });
    var r = f();
    return x(n, function(t) {
      x(t.getTargetSeriesModels(), function(e) {
        r.set(e.uid, e);
      });
    }), r;
  },
  // Consider appendData, where filter should be performed. Because data process is
  // in block mode currently, it is not need to worry about that the overallProgress
  // execute every frame.
  overallReset: function(a, s) {
    a.eachComponent("dataZoom", function(n) {
      n.eachTargetAxis(function(r, t) {
        n.getAxisProxy(r, t).reset(n);
      }), n.eachTargetAxis(function(r, t) {
        n.getAxisProxy(r, t).filterData(n, s);
      });
    }), a.eachComponent("dataZoom", function(n) {
      var r = n.findRepresentativeAxisProxy();
      if (r) {
        var t = r.getDataPercentWindow(), e = r.getDataValueWindow();
        n.setCalculatedRange({
          start: t[0],
          end: t[1],
          startValue: e[0],
          endValue: e[1]
        });
      }
    });
  }
};
export {
  P as default
};
