// echarts@6.0.0/lib/component/dataZoom/DataZoomModel.js
import { __extends as m } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { merge as A, each as s, createHashMap as O, assert as _ } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import R from "../../model/Component.js";
import { getAxisMainType as c, DATA_ZOOM_AXIS_DIMENSIONS as x } from "./helper.js";
import { MULTIPLE_REFERRING as P, SINGLE_REFERRING as M } from "../../util/model.js";
var v = (
  /** @class */
  (function() {
    function f() {
      this.indexList = [], this.indexMap = [];
    }
    return f.prototype.add = function(r) {
      this.indexMap[r] || (this.indexList.push(r), this.indexMap[r] = !0);
    }, f;
  })()
), w = (
  /** @class */
  (function(f) {
    m(r, f);
    function r() {
      var t = f !== null && f.apply(this, arguments) || this;
      return t.type = r.type, t._autoThrottle = !0, t._noTarget = !0, t._rangePropMode = ["percent", "percent"], t;
    }
    return r.prototype.init = function(t, e, i) {
      var n = T(t);
      this.settledOption = n, this.mergeDefaultAndTheme(t, i), this._doInit(n);
    }, r.prototype.mergeOption = function(t) {
      var e = T(t);
      A(this.option, t, !0), A(this.settledOption, e, !0), this._doInit(e);
    }, r.prototype._doInit = function(t) {
      var e = this.option;
      this._setDefaultThrottle(t), this._updateRangeUse(t);
      var i = this.settledOption;
      s([["start", "startValue"], ["end", "endValue"]], function(n, o) {
        this._rangePropMode[o] === "value" && (e[n[0]] = i[n[0]] = null);
      }, this), this._resetTarget();
    }, r.prototype._resetTarget = function() {
      var t = this.get("orient", !0), e = this._targetAxisInfoMap = O(), i = this._fillSpecifiedTargetAxis(e);
      i ? this._orient = t || this._makeAutoOrientByTargetAxis() : (this._orient = t || "horizontal", this._fillAutoTargetAxisByOrient(e, this._orient)), this._noTarget = !0, e.each(function(n) {
        n.indexList.length && (this._noTarget = !1);
      }, this);
    }, r.prototype._fillSpecifiedTargetAxis = function(t) {
      var e = !1;
      return s(x, function(i) {
        var n = this.getReferringComponents(c(i), P);
        if (n.specified) {
          e = !0;
          var o = new v();
          s(n.models, function(a) {
            o.add(a.componentIndex);
          }), t.set(i, o);
        }
      }, this), e;
    }, r.prototype._fillAutoTargetAxisByOrient = function(t, e) {
      var i = this.ecModel, n = !0;
      if (n) {
        var o = e === "vertical" ? "y" : "x", a = i.findComponents({
          mainType: o + "Axis"
        });
        p(a, o);
      }
      if (n) {
        var a = i.findComponents({
          mainType: "singleAxis",
          filter: function(l) {
            return l.get("orient", !0) === e;
          }
        });
        p(a, "single");
      }
      function p(u, l) {
        var d = u[0];
        if (d) {
          var h = new v();
          if (h.add(d.componentIndex), t.set(l, h), n = !1, l === "x" || l === "y") {
            var y = d.getReferringComponents("grid", M).models[0];
            y && s(u, function(g) {
              d.componentIndex !== g.componentIndex && y === g.getReferringComponents("grid", M).models[0] && h.add(g.componentIndex);
            });
          }
        }
      }
      n && s(x, function(u) {
        if (n) {
          var l = i.findComponents({
            mainType: c(u),
            filter: function(h) {
              return h.get("type", !0) === "category";
            }
          });
          if (l[0]) {
            var d = new v();
            d.add(l[0].componentIndex), t.set(u, d), n = !1;
          }
        }
      }, this);
    }, r.prototype._makeAutoOrientByTargetAxis = function() {
      var t;
      return this.eachTargetAxis(function(e) {
        !t && (t = e);
      }, this), t === "y" ? "vertical" : "horizontal";
    }, r.prototype._setDefaultThrottle = function(t) {
      if (t.hasOwnProperty("throttle") && (this._autoThrottle = !1), this._autoThrottle) {
        var e = this.ecModel.option;
        this.option.throttle = e.animation && e.animationDurationUpdate > 0 ? 100 : 20;
      }
    }, r.prototype._updateRangeUse = function(t) {
      var e = this._rangePropMode, i = this.get("rangeMode");
      s([["start", "startValue"], ["end", "endValue"]], function(n, o) {
        var a = t[n[0]] != null, p = t[n[1]] != null;
        a && !p ? e[o] = "percent" : !a && p ? e[o] = "value" : i ? e[o] = i[o] : a && (e[o] = "percent");
      });
    }, r.prototype.noTarget = function() {
      return this._noTarget;
    }, r.prototype.getFirstTargetAxisModel = function() {
      var t;
      return this.eachTargetAxis(function(e, i) {
        t == null && (t = this.ecModel.getComponent(c(e), i));
      }, this), t;
    }, r.prototype.eachTargetAxis = function(t, e) {
      this._targetAxisInfoMap.each(function(i, n) {
        s(i.indexList, function(o) {
          t.call(e, n, o);
        });
      });
    }, r.prototype.getAxisProxy = function(t, e) {
      var i = this.getAxisModel(t, e);
      if (i)
        return i.__dzAxisProxy;
    }, r.prototype.getAxisModel = function(t, e) {
      process.env.NODE_ENV !== "production" && _(t && e != null);
      var i = this._targetAxisInfoMap.get(t);
      if (i && i.indexMap[e])
        return this.ecModel.getComponent(c(t), e);
    }, r.prototype.setRawRange = function(t) {
      var e = this.option, i = this.settledOption;
      s([["start", "startValue"], ["end", "endValue"]], function(n) {
        (t[n[0]] != null || t[n[1]] != null) && (e[n[0]] = i[n[0]] = t[n[0]], e[n[1]] = i[n[1]] = t[n[1]]);
      }, this), this._updateRangeUse(t);
    }, r.prototype.setCalculatedRange = function(t) {
      var e = this.option;
      s(["start", "startValue", "end", "endValue"], function(i) {
        e[i] = t[i];
      });
    }, r.prototype.getPercentRange = function() {
      var t = this.findRepresentativeAxisProxy();
      if (t)
        return t.getDataPercentWindow();
    }, r.prototype.getValueRange = function(t, e) {
      if (t == null && e == null) {
        var i = this.findRepresentativeAxisProxy();
        if (i)
          return i.getDataValueWindow();
      } else
        return this.getAxisProxy(t, e).getDataValueWindow();
    }, r.prototype.findRepresentativeAxisProxy = function(t) {
      if (t)
        return t.__dzAxisProxy;
      for (var e, i = this._targetAxisInfoMap.keys(), n = 0; n < i.length; n++)
        for (var o = i[n], a = this._targetAxisInfoMap.get(o), p = 0; p < a.indexList.length; p++) {
          var u = this.getAxisProxy(o, a.indexList[p]);
          if (u.hostedBy(this))
            return u;
          e || (e = u);
        }
      return e;
    }, r.prototype.getRangePropMode = function() {
      return this._rangePropMode.slice();
    }, r.prototype.getOrient = function() {
      return process.env.NODE_ENV !== "production" && _(this._orient), this._orient;
    }, r.type = "dataZoom", r.dependencies = ["xAxis", "yAxis", "radiusAxis", "angleAxis", "singleAxis", "series", "toolbox"], r.defaultOption = {
      // zlevel: 0,
      z: 4,
      filterMode: "filter",
      start: 0,
      end: 100
    }, r;
  })(R)
);
function T(f) {
  var r = {};
  return s(["start", "end", "startValue", "endValue", "throttle"], function(t) {
    f.hasOwnProperty(t) && (r[t] = f[t]);
  }), r;
}
export {
  w as default
};
