// echarts@6.0.0/lib/component/dataZoom/helper.js
import { createHashMap as p, assert as S, indexOf as x } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
var h = ["x", "y", "radius", "angle", "single"], M = ["cartesian2d", "polar", "singleAxis"];
function C(r) {
  var t = r.get("coordinateSystem");
  return x(M, t) >= 0;
}
function y(r) {
  return process.env.NODE_ENV !== "production" && S(r), r + "Axis";
}
function I(r, t) {
  var n = p(), u = [], s = p();
  r.eachComponent({
    mainType: "dataZoom",
    query: t
  }, function(e) {
    s.get(e.uid) || a(e);
  });
  var o;
  do
    o = !1, r.eachComponent("dataZoom", f);
  while (o);
  function f(e) {
    !s.get(e.uid) && i(e) && (a(e), o = !0);
  }
  function a(e) {
    s.set(e.uid, !0), u.push(e), g(e);
  }
  function i(e) {
    var c = !1;
    return e.eachTargetAxis(function(d, v) {
      var l = n.get(d);
      l && l[v] && (c = !0);
    }), c;
  }
  function g(e) {
    e.eachTargetAxis(function(c, d) {
      (n.get(c) || n.set(c, []))[d] = !0;
    });
  }
  return u;
}
function T(r) {
  var t = r.ecModel, n = {
    infoList: [],
    infoMap: p()
  };
  return r.eachTargetAxis(function(u, s) {
    var o = t.getComponent(y(u), s);
    if (o) {
      var f = o.getCoordSysModel();
      if (f) {
        var a = f.uid, i = n.infoMap.get(a);
        i || (i = {
          model: f,
          axisModels: []
        }, n.infoList.push(i), n.infoMap.set(a, i)), i.axisModels.push(o);
      }
    }
  }), n;
}
export {
  h as DATA_ZOOM_AXIS_DIMENSIONS,
  T as collectReferCoordSysModelInfo,
  I as findEffectedDataZooms,
  y as getAxisMainType,
  C as isCoordSupported
};
