// echarts@6.0.0/lib/component/dataZoom/history.js
import { each as v } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { makeInner as c } from "../../util/model.js";
var i = v, u = c();
function k(r, n) {
  var e = p(r);
  i(n, function(o, t) {
    for (var s = e.length - 1; s >= 0; s--) {
      var a = e[s];
      if (a[t])
        break;
    }
    if (s < 0) {
      var h = r.queryComponents({
        mainType: "dataZoom",
        subType: "select",
        id: t
      })[0];
      if (h) {
        var f = h.getPercentRange();
        e[0][t] = {
          dataZoomId: t,
          start: f[0],
          end: f[1]
        };
      }
    }
  }), e.push(n);
}
function m(r) {
  var n = p(r), e = n[n.length - 1];
  n.length > 1 && n.pop();
  var o = {};
  return i(e, function(t, s) {
    for (var a = n.length - 1; a >= 0; a--)
      if (t = n[a][s], t) {
        o[s] = t;
        break;
      }
  }), o;
}
function y(r) {
  u(r).snapshots = null;
}
function S(r) {
  return p(r).length;
}
function p(r) {
  var n = u(r);
  return n.snapshots || (n.snapshots = [{}]), n.snapshots;
}
export {
  y as clear,
  S as count,
  m as pop,
  k as push
};
