// echarts@6.0.0/lib/component/dataZoom/InsideZoomView.js
import { __extends as m } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import f from "./DataZoomView.js";
import d from "../helper/sliderMove.js";
import { setViewInfoToCoordSysRecord as u, disposeCoordSysRecordIfNeeded as y } from "./roams.js";
import { bind as x } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
var D = (
  /** @class */
  (function(t) {
    m(a, t);
    function a() {
      var n = t !== null && t.apply(this, arguments) || this;
      return n.type = "dataZoom.inside", n;
    }
    return a.prototype.render = function(n, o, l) {
      if (t.prototype.render.apply(this, arguments), n.noTarget()) {
        this._clear();
        return;
      }
      this.range = n.getPercentRange(), u(l, n, {
        pan: x(g.pan, this),
        zoom: x(g.zoom, this),
        scrollMove: x(g.scrollMove, this)
      });
    }, a.prototype.dispose = function() {
      this._clear(), t.prototype.dispose.apply(this, arguments);
    }, a.prototype._clear = function() {
      y(this.api, this.dataZoomModel), this.range = null;
    }, a.type = "dataZoom.inside", a;
  })(f)
), g = {
  zoom: function(t, a, n, o) {
    var l = this.range, r = l.slice(), e = t.axisModels[0];
    if (e) {
      var i = c[a](null, [o.originX, o.originY], e, n, t), s = (i.signal > 0 ? i.pixelStart + i.pixelLength - i.pixel : i.pixel - i.pixelStart) / i.pixelLength * (r[1] - r[0]) + r[0], p = Math.max(1 / o.scale, 0);
      r[0] = (r[0] - s) * p + s, r[1] = (r[1] - s) * p + s;
      var v = this.dataZoomModel.findRepresentativeAxisProxy().getMinMaxSpan();
      if (d(0, r, [0, 100], 0, v.minSpan, v.maxSpan), this.range = r, l[0] !== r[0] || l[1] !== r[1])
        return r;
    }
  },
  pan: h(function(t, a, n, o, l, r) {
    var e = c[o]([r.oldX, r.oldY], [r.newX, r.newY], a, l, n);
    return e.signal * (t[1] - t[0]) * e.pixel / e.pixelLength;
  }),
  scrollMove: h(function(t, a, n, o, l, r) {
    var e = c[o]([0, 0], [r.scrollDelta, r.scrollDelta], a, l, n);
    return e.signal * (t[1] - t[0]) * r.scrollDelta;
  })
};
function h(t) {
  return function(a, n, o, l) {
    var r = this.range, e = r.slice(), i = a.axisModels[0];
    if (i) {
      var s = t(e, i, a, n, o, l);
      if (d(s, e, [0, 100], "all"), this.range = e, r[0] !== e[0] || r[1] !== e[1])
        return e;
    }
  };
}
var c = {
  grid: function(t, a, n, o, l) {
    var r = n.axis, e = {}, i = l.model.coordinateSystem.getRect();
    return t = t || [0, 0], r.dim === "x" ? (e.pixel = a[0] - t[0], e.pixelLength = i.width, e.pixelStart = i.x, e.signal = r.inverse ? 1 : -1) : (e.pixel = a[1] - t[1], e.pixelLength = i.height, e.pixelStart = i.y, e.signal = r.inverse ? -1 : 1), e;
  },
  polar: function(t, a, n, o, l) {
    var r = n.axis, e = {}, i = l.model.coordinateSystem, s = i.getRadiusAxis().getExtent(), p = i.getAngleAxis().getExtent();
    return t = t ? i.pointToCoord(t) : [0, 0], a = i.pointToCoord(a), n.mainType === "radiusAxis" ? (e.pixel = a[0] - t[0], e.pixelLength = s[1] - s[0], e.pixelStart = s[0], e.signal = r.inverse ? 1 : -1) : (e.pixel = a[1] - t[1], e.pixelLength = p[1] - p[0], e.pixelStart = p[0], e.signal = r.inverse ? -1 : 1), e;
  },
  singleAxis: function(t, a, n, o, l) {
    var r = n.axis, e = l.model.coordinateSystem.getRect(), i = {};
    return t = t || [0, 0], r.orient === "horizontal" ? (i.pixel = a[0] - t[0], i.pixelLength = e.width, i.pixelStart = e.x, i.signal = r.inverse ? 1 : -1) : (i.pixel = a[1] - t[1], i.pixelLength = e.height, i.pixelStart = e.y, i.signal = r.inverse ? -1 : 1), i;
  }
};
export {
  D as default
};
