// echarts@6.0.0/lib/component/dataZoom/roams.js
import y from "../helper/RoamController.js";
import { createOrUpdate as g } from "../../util/throttle.js";
import { makeInner as h } from "../../util/model.js";
import { createHashMap as s, each as p, curry as m } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { collectReferCoordSysModelInfo as M } from "./helper.js";
var d = h();
function A(t, n, o) {
  d(t).coordSysRecordMap.each(function(c) {
    var a = c.dataZoomInfoMap.get(n.uid);
    a && (a.getRange = o);
  });
}
function k(t, n) {
  for (var o = d(t).coordSysRecordMap, c = o.keys(), a = 0; a < c.length; a++) {
    var r = c[a], l = o.get(r), e = l.dataZoomInfoMap;
    if (e) {
      var i = n.uid, u = e.get(i);
      u && (e.removeKey(i), e.keys().length || v(o, l));
    }
  }
}
function v(t, n) {
  if (n) {
    t.removeKey(n.model.uid);
    var o = n.controller;
    o && o.dispose();
  }
}
function R(t, n) {
  var o = {
    model: n,
    containsPoint: m(P, n),
    dispatchAction: m(I, t),
    dataZoomInfoMap: null,
    controller: null
  }, c = o.controller = new y(t.getZr());
  return p(["pan", "zoom", "scrollMove"], function(a) {
    c.on(a, function(r) {
      var l = [];
      o.dataZoomInfoMap.each(function(e) {
        if (r.isAvailableBehavior(e.model.option)) {
          var i = (e.getRange || {})[a], u = i && i(e.dzReferCoordSysInfo, o.model.mainType, o.controller, r);
          !e.model.get("disabled", !0) && u && l.push({
            dataZoomId: e.model.id,
            start: u[0],
            end: u[1]
          });
        }
      }), l.length && o.dispatchAction(l);
    });
  }), o;
}
function I(t, n) {
  t.isDisposed() || t.dispatchAction({
    type: "dataZoom",
    animation: {
      easing: "cubicOut",
      duration: 100
    },
    batch: n
  });
}
function P(t, n, o, c) {
  return t.coordinateSystem.containPoint([o, c]);
}
function S(t, n, o) {
  var c, a = "type_", r = {
    type_true: 2,
    type_move: 1,
    type_false: 0,
    type_undefined: -1
  }, l = !0;
  return t.each(function(e) {
    var i = e.model, u = i.get("disabled", !0) ? !1 : i.get("zoomLock", !0) ? "move" : !0;
    r[a + u] > r[a + c] && (c = u), l = l && i.get("preventDefaultMouseMove", !0);
  }), {
    controlType: c,
    opt: {
      // RoamController will enable all of these functionalities,
      // and the final behavior is determined by its event listener
      // provided by each inside zoom.
      zoomOnMouseWheel: !0,
      moveOnMouseMove: !0,
      moveOnMouseWheel: !0,
      preventDefaultMouseMove: !!l,
      api: o,
      zInfo: {
        component: n.model
      },
      triggerInfo: {
        roamTrigger: null,
        isInSelf: n.containsPoint
      }
    }
  };
}
function D(t) {
  t.registerProcessor(t.PRIORITY.PROCESSOR.FILTER, function(n, o) {
    var c = d(o), a = c.coordSysRecordMap || (c.coordSysRecordMap = s());
    a.each(function(r) {
      r.dataZoomInfoMap = null;
    }), n.eachComponent({
      mainType: "dataZoom",
      subType: "inside"
    }, function(r) {
      var l = M(r);
      p(l.infoList, function(e) {
        var i = e.model.uid, u = a.get(i) || a.set(i, R(o, e.model)), f = u.dataZoomInfoMap || (u.dataZoomInfoMap = s());
        f.set(r.uid, {
          dzReferCoordSysInfo: e,
          model: r,
          getRange: null
        });
      });
    }), a.each(function(r) {
      var l = r.controller, e, i = r.dataZoomInfoMap;
      if (i) {
        var u = i.keys()[0];
        u != null && (e = i.get(u));
      }
      if (!e) {
        v(a, r);
        return;
      }
      var f = S(i, r, o);
      l.enable(f.controlType, f.opt), g(r, "dispatchAction", e.model.get("throttle", !0), "fixRate");
    });
  });
}
export {
  k as disposeCoordSysRecordIfNeeded,
  D as installDataZoomRoamProcessor,
  A as setViewInfoToCoordSysRecord
};
