// echarts@6.0.0/lib/component/dataZoom/SliderZoomView.js
import { __extends as j } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { bind as y, each as P, indexOf as q, isFunction as J, isString as K } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { stop as V } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/event.js";
import { getTransform as Q, transformDirection as tt, applyTransform as F } from "../../util/graphic.js";
import { createOrUpdate as et, clear as at } from "../../util/throttle.js";
import rt from "./DataZoomView.js";
import { linearMap as _, parsePercent as N, asc as C } from "../../util/number.js";
import { createBoxLayoutReference as it, getLayoutParams as ot, getLayoutRect as st } from "../../util/layout.js";
import W from "../helper/sliderMove.js";
import { getAxisMainType as nt, collectReferCoordSysModelInfo as ht } from "./helper.js";
import { enableHoverEmphasis as lt } from "../../util/states.js";
import { symbolBuildProxies as dt, createSymbol as Y } from "../../util/symbol.js";
import { deprecateLog as ut } from "../../util/log.js";
import { createTextStyle as pt } from "../../label/labelStyle.js";
import X from "../../visual/tokens.js";
import U from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import O from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Rect.js";
import vt from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Text.js";
import gt from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/Point.js";
import mt from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Polygon.js";
import ft from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Polyline.js";
var I = O, ct = 1, H = 30, yt = 7, z = "horizontal", $ = "vertical", _t = 5, wt = ["line", "bar", "candlestick", "scatter"], St = {
  easing: "cubicOut",
  duration: 100,
  delay: 0
}, Yt = (
  /** @class */
  (function(w) {
    j(h, w);
    function h() {
      var t = w !== null && w.apply(this, arguments) || this;
      return t.type = h.type, t._displayables = {}, t;
    }
    return h.prototype.init = function(t, e) {
      this.api = e, this._onBrush = y(this._onBrush, this), this._onBrushEnd = y(this._onBrushEnd, this);
    }, h.prototype.render = function(t, e, o, s) {
      if (w.prototype.render.apply(this, arguments), et(this, "_dispatchZoomAction", t.get("throttle"), "fixRate"), this._orient = t.getOrient(), t.get("show") === !1) {
        this.group.removeAll();
        return;
      }
      if (t.noTarget()) {
        this._clear(), this.group.removeAll();
        return;
      }
      (!s || s.type !== "dataZoom" || s.from !== this.uid) && this._buildView(), this._updateView();
    }, h.prototype.dispose = function() {
      this._clear(), w.prototype.dispose.apply(this, arguments);
    }, h.prototype._clear = function() {
      at(this, "_dispatchZoomAction");
      var t = this.api.getZr();
      t.off("mousemove", this._onBrush), t.off("mouseup", this._onBrushEnd);
    }, h.prototype._buildView = function() {
      var t = this.group;
      t.removeAll(), this._brushing = !1, this._displayables.brushRect = null, this._resetLocation(), this._resetInterval();
      var e = this._displayables.sliderGroup = new U();
      this._renderBackground(), this._renderHandle(), this._renderDataShadow(), t.add(e), this._positionGroup();
    }, h.prototype._resetLocation = function() {
      var t = this.dataZoomModel, e = this.api, o = t.get("brushSelect"), s = o ? yt : 0, a = it(t, e).refContainer, r = this._findCoordRect(), i = t.get("defaultLocationEdgeGap", !0) || 0, n = this._orient === z ? {
        // Why using 'right', because right should be used in vertical,
        // and it is better to be consistent for dealing with position param merge.
        right: a.width - r.x - r.width,
        top: a.height - H - i - s,
        width: r.width,
        height: H
      } : {
        right: i,
        top: r.y,
        width: H,
        height: r.height
      }, l = ot(t.option);
      P(["right", "top", "width", "height"], function(u) {
        l[u] === "ph" && (l[u] = n[u]);
      });
      var d = st(l, a);
      this._location = {
        x: d.x,
        y: d.y
      }, this._size = [d.width, d.height], this._orient === $ && this._size.reverse();
    }, h.prototype._positionGroup = function() {
      var t = this.group, e = this._location, o = this._orient, s = this.dataZoomModel.getFirstTargetAxisModel(), a = s && s.get("inverse"), r = this._displayables.sliderGroup, i = (this._dataShadowInfo || {}).otherAxisInverse;
      r.attr(o === z && !a ? {
        scaleY: i ? 1 : -1,
        scaleX: 1
      } : o === z && a ? {
        scaleY: i ? 1 : -1,
        scaleX: -1
      } : o === $ && !a ? {
        scaleY: i ? -1 : 1,
        scaleX: 1,
        rotation: Math.PI / 2
      } : {
        scaleY: i ? -1 : 1,
        scaleX: -1,
        rotation: Math.PI / 2
      });
      var n = t.getBoundingRect([r]);
      t.x = e.x - n.x, t.y = e.y - n.y, t.markRedraw();
    }, h.prototype._getViewExtent = function() {
      return [0, this._size[0]];
    }, h.prototype._renderBackground = function() {
      var t = this.dataZoomModel, e = this._size, o = this._displayables.sliderGroup, s = t.get("brushSelect");
      o.add(new I({
        silent: !0,
        shape: {
          x: 0,
          y: 0,
          width: e[0],
          height: e[1]
        },
        style: {
          fill: t.get("backgroundColor")
        },
        z2: -40
      }));
      var a = new I({
        shape: {
          x: 0,
          y: 0,
          width: e[0],
          height: e[1]
        },
        style: {
          fill: "transparent"
        },
        z2: 0,
        onclick: y(this._onClickPanel, this)
      }), r = this.api.getZr();
      s ? (a.on("mousedown", this._onBrushStart, this), a.cursor = "crosshair", r.on("mousemove", this._onBrush), r.on("mouseup", this._onBrushEnd)) : (r.off("mousemove", this._onBrush), r.off("mouseup", this._onBrushEnd)), o.add(a);
    }, h.prototype._renderDataShadow = function() {
      var t = this._dataShadowInfo = this._prepareDataShadowInfo();
      if (this._displayables.dataShadowSegs = [], !t)
        return;
      var e = this._size, o = this._shadowSize || [], s = t.series, a = s.getRawData(), r = s.getShadowDim && s.getShadowDim(), i = r && a.getDimensionInfo(r) ? s.getShadowDim() : t.otherDim;
      if (i == null)
        return;
      var n = this._shadowPolygonPts, l = this._shadowPolylinePts;
      if (a !== this._shadowData || i !== this._shadowDim || e[0] !== o[0] || e[1] !== o[1]) {
        var d = a.getDataExtent(t.thisDim), u = a.getDataExtent(i), v = (u[1] - u[0]) * 0.3;
        u = [u[0] - v, u[1] + v];
        var p = [0, e[1]], f = [0, e[0]], m = [[e[0], 0], [0, 0]], c = [], S = f[1] / Math.max(1, a.count() - 1), E = e[0] / (d[1] - d[0]), b = t.thisAxis.type === "time", g = -S, Z = Math.round(a.count() / e[0]), A;
        a.each([t.thisDim, i], function(k, M, D) {
          if (Z > 0 && D % Z) {
            b || (g += S);
            return;
          }
          g = b ? (+k - d[0]) * E : g + S;
          var x = M == null || isNaN(M) || M === "", G = x ? 0 : _(M, u, p, !0);
          x && !A && D ? (m.push([m[m.length - 1][0], 0]), c.push([c[c.length - 1][0], 0])) : !x && A && (m.push([g, 0]), c.push([g, 0])), x || (m.push([g, G]), c.push([g, G])), A = x;
        }), n = this._shadowPolygonPts = m, l = this._shadowPolylinePts = c;
      }
      this._shadowData = a, this._shadowDim = i, this._shadowSize = [e[0], e[1]];
      var T = this.dataZoomModel;
      function L(k) {
        var M = T.getModel(k ? "selectedDataBackground" : "dataBackground"), D = new U(), x = new mt({
          shape: {
            points: n
          },
          segmentIgnoreThreshold: 1,
          style: M.getModel("areaStyle").getAreaStyle(),
          silent: !0,
          z2: -20
        }), G = new ft({
          shape: {
            points: l
          },
          segmentIgnoreThreshold: 1,
          style: M.getModel("lineStyle").getLineStyle(),
          silent: !0,
          z2: -19
        });
        return D.add(x), D.add(G), D;
      }
      for (var R = 0; R < 3; R++) {
        var B = L(R === 1);
        this._displayables.sliderGroup.add(B), this._displayables.dataShadowSegs.push(B);
      }
    }, h.prototype._prepareDataShadowInfo = function() {
      var t = this.dataZoomModel, e = t.get("showDataShadow");
      if (e !== !1) {
        var o, s = this.ecModel;
        return t.eachTargetAxis(function(a, r) {
          var i = t.getAxisProxy(a, r).getTargetSeriesModels();
          P(i, function(n) {
            if (!o && !(e !== !0 && q(wt, n.get("type")) < 0)) {
              var l = s.getComponent(nt(a), r).axis, d = bt(a), u, v = n.coordinateSystem;
              d != null && v.getOtherAxis && (u = v.getOtherAxis(l).inverse), d = n.getData().mapDimension(d);
              var p = n.getData().mapDimension(a);
              o = {
                thisAxis: l,
                series: n,
                thisDim: p,
                otherDim: d,
                otherAxisInverse: u
              };
            }
          }, this);
        }, this), o;
      }
    }, h.prototype._renderHandle = function() {
      var t = this.group, e = this._displayables, o = e.handles = [null, null], s = e.handleLabels = [null, null], a = this._displayables.sliderGroup, r = this._size, i = this.dataZoomModel, n = this.api, l = i.get("borderRadius") || 0, d = i.get("brushSelect"), u = e.filler = new I({
        silent: d,
        style: {
          fill: i.get("fillerColor")
        },
        textConfig: {
          position: "inside"
        }
      });
      a.add(u), a.add(new I({
        silent: !0,
        subPixelOptimize: !0,
        shape: {
          x: 0,
          y: 0,
          width: r[0],
          height: r[1],
          r: l
        },
        style: {
          // deprecated option
          stroke: i.get("dataBackgroundColor") || i.get("borderColor"),
          lineWidth: ct,
          fill: X.color.transparent
        }
      })), P([0, 1], function(E) {
        var b = i.get("handleIcon");
        !dt[b] && b.indexOf("path://") < 0 && b.indexOf("image://") < 0 && (b = "path://" + b, process.env.NODE_ENV !== "production" && ut("handleIcon now needs 'path://' prefix when using a path string"));
        var g = Y(b, -1, 0, 2, 2, null, !0);
        g.attr({
          cursor: Mt(this._orient),
          draggable: !0,
          drift: y(this._onDragMove, this, E),
          ondragend: y(this._onDragEnd, this),
          onmouseover: y(this._showDataInfo, this, !0),
          onmouseout: y(this._showDataInfo, this, !1),
          z2: 5
        });
        var Z = g.getBoundingRect(), A = i.get("handleSize");
        this._handleHeight = N(A, this._size[1]), this._handleWidth = Z.width / Z.height * this._handleHeight, g.setStyle(i.getModel("handleStyle").getItemStyle()), g.style.strokeNoScale = !0, g.rectHover = !0, g.ensureState("emphasis").style = i.getModel(["emphasis", "handleStyle"]).getItemStyle(), lt(g);
        var T = i.get("handleColor");
        T != null && (g.style.fill = T), a.add(o[E] = g);
        var L = i.getModel("textStyle"), R = i.get("handleLabel") || {}, B = R.show || !1;
        t.add(s[E] = new vt({
          silent: !0,
          invisible: !B,
          style: pt(L, {
            x: 0,
            y: 0,
            text: "",
            verticalAlign: "middle",
            align: "center",
            fill: L.getTextColor(),
            font: L.getFont()
          }),
          z2: 10
        }));
      }, this);
      var v = u;
      if (d) {
        var p = N(i.get("moveHandleSize"), r[1]), f = e.moveHandle = new O({
          style: i.getModel("moveHandleStyle").getItemStyle(),
          silent: !0,
          shape: {
            r: [0, 0, 2, 2],
            y: r[1] - 0.5,
            height: p
          }
        }), m = p * 0.8, c = e.moveHandleIcon = Y(i.get("moveHandleIcon"), -m / 2, -m / 2, m, m, X.color.neutral00, !0);
        c.silent = !0, c.y = r[1] + p / 2 - 0.5, f.ensureState("emphasis").style = i.getModel(["emphasis", "moveHandleStyle"]).getItemStyle();
        var S = Math.min(r[1] / 2, Math.max(p, 10));
        v = e.moveZone = new O({
          invisible: !0,
          shape: {
            y: r[1] - S,
            height: p + S
          }
        }), v.on("mouseover", function() {
          n.enterEmphasis(f);
        }).on("mouseout", function() {
          n.leaveEmphasis(f);
        }), a.add(f), a.add(c), a.add(v);
      }
      v.attr({
        draggable: !0,
        cursor: "default",
        drift: y(this._onDragMove, this, "all"),
        ondragstart: y(this._showDataInfo, this, !0),
        ondragend: y(this._onDragEnd, this),
        onmouseover: y(this._showDataInfo, this, !0),
        onmouseout: y(this._showDataInfo, this, !1)
      });
    }, h.prototype._resetInterval = function() {
      var t = this._range = this.dataZoomModel.getPercentRange(), e = this._getViewExtent();
      this._handleEnds = [_(t[0], [0, 100], e, !0), _(t[1], [0, 100], e, !0)];
    }, h.prototype._updateInterval = function(t, e) {
      var o = this.dataZoomModel, s = this._handleEnds, a = this._getViewExtent(), r = o.findRepresentativeAxisProxy().getMinMaxSpan(), i = [0, 100];
      W(e, s, a, o.get("zoomLock") ? "all" : t, r.minSpan != null ? _(r.minSpan, i, a, !0) : null, r.maxSpan != null ? _(r.maxSpan, i, a, !0) : null);
      var n = this._range, l = this._range = C([_(s[0], a, i, !0), _(s[1], a, i, !0)]);
      return !n || n[0] !== l[0] || n[1] !== l[1];
    }, h.prototype._updateView = function(t) {
      var e = this._displayables, o = this._handleEnds, s = C(o.slice()), a = this._size;
      P([0, 1], function(v) {
        var p = e.handles[v], f = this._handleHeight;
        p.attr({
          scaleX: f / 2,
          scaleY: f / 2,
          // This is a trick, by adding an extra tiny offset to let the default handle's end point align to the drag window.
          // NOTE: It may affect some custom shapes a bit. But we prefer to have better result by default.
          x: o[v] + (v ? -1 : 1),
          y: a[1] / 2 - f / 2
        });
      }, this), e.filler.setShape({
        x: s[0],
        y: 0,
        width: s[1] - s[0],
        height: a[1]
      });
      var r = {
        x: s[0],
        width: s[1] - s[0]
      };
      e.moveHandle && (e.moveHandle.setShape(r), e.moveZone.setShape(r), e.moveZone.getBoundingRect(), e.moveHandleIcon && e.moveHandleIcon.attr("x", r.x + r.width / 2));
      for (var i = e.dataShadowSegs, n = [0, s[0], s[1], a[0]], l = 0; l < i.length; l++) {
        var d = i[l], u = d.getClipPath();
        u || (u = new O(), d.setClipPath(u)), u.setShape({
          x: n[l],
          y: 0,
          width: n[l + 1] - n[l],
          height: a[1]
        });
      }
      this._updateDataInfo(t);
    }, h.prototype._updateDataInfo = function(t) {
      var e = this.dataZoomModel, o = this._displayables, s = o.handleLabels, a = this._orient, r = ["", ""];
      if (e.get("showDetail")) {
        var i = e.findRepresentativeAxisProxy();
        if (i) {
          var n = i.getAxisModel().axis, l = this._range, d = t ? i.calculateDataWindow({
            start: l[0],
            end: l[1]
          }).valueWindow : i.getDataValueWindow();
          r = [this._formatLabel(d[0], n), this._formatLabel(d[1], n)];
        }
      }
      var u = C(this._handleEnds.slice());
      v.call(this, 0), v.call(this, 1);
      function v(p) {
        var f = Q(o.handles[p].parent, this.group), m = tt(p === 0 ? "right" : "left", f), c = this._handleWidth / 2 + _t, S = F([u[p] + (p === 0 ? -c : c), this._size[1] / 2], f);
        s[p].setStyle({
          x: S[0],
          y: S[1],
          verticalAlign: a === z ? "middle" : m,
          align: a === z ? m : "center",
          text: r[p]
        });
      }
    }, h.prototype._formatLabel = function(t, e) {
      var o = this.dataZoomModel, s = o.get("labelFormatter"), a = o.get("labelPrecision");
      (a == null || a === "auto") && (a = e.getPixelPrecision());
      var r = t == null || isNaN(t) ? "" : e.type === "category" || e.type === "time" ? e.scale.getLabel({
        value: Math.round(t)
      }) : t.toFixed(Math.min(a, 20));
      return J(s) ? s(t, r) : K(s) ? s.replace("{value}", r) : r;
    }, h.prototype._showDataInfo = function(t) {
      var e = this.dataZoomModel.get("handleLabel") || {}, o = e.show || !1, s = this.dataZoomModel.getModel(["emphasis", "handleLabel"]), a = s.get("show") || !1, r = t || this._dragging ? a : o, i = this._displayables, n = i.handleLabels;
      n[0].attr("invisible", !r), n[1].attr("invisible", !r), i.moveHandle && this.api[r ? "enterEmphasis" : "leaveEmphasis"](i.moveHandle, 1);
    }, h.prototype._onDragMove = function(t, e, o, s) {
      this._dragging = !0, V(s.event);
      var a = this._displayables.sliderGroup.getLocalTransform(), r = F([e, o], a, !0), i = this._updateInterval(t, r[0]), n = this.dataZoomModel.get("realtime");
      this._updateView(!n), i && n && this._dispatchZoomAction(!0);
    }, h.prototype._onDragEnd = function() {
      this._dragging = !1, this._showDataInfo(!1);
      var t = this.dataZoomModel.get("realtime");
      !t && this._dispatchZoomAction(!1);
    }, h.prototype._onClickPanel = function(t) {
      var e = this._size, o = this._displayables.sliderGroup.transformCoordToLocal(t.offsetX, t.offsetY);
      if (!(o[0] < 0 || o[0] > e[0] || o[1] < 0 || o[1] > e[1])) {
        var s = this._handleEnds, a = (s[0] + s[1]) / 2, r = this._updateInterval("all", o[0] - a);
        this._updateView(), r && this._dispatchZoomAction(!1);
      }
    }, h.prototype._onBrushStart = function(t) {
      var e = t.offsetX, o = t.offsetY;
      this._brushStart = new gt(e, o), this._brushing = !0, this._brushStartTime = +/* @__PURE__ */ new Date();
    }, h.prototype._onBrushEnd = function(t) {
      if (this._brushing) {
        var e = this._displayables.brushRect;
        if (this._brushing = !1, !!e) {
          e.attr("ignore", !0);
          var o = e.shape, s = +/* @__PURE__ */ new Date();
          if (!(s - this._brushStartTime < 200 && Math.abs(o.width) < 5)) {
            var a = this._getViewExtent(), r = [0, 100], i = this._handleEnds = [o.x, o.x + o.width], n = this.dataZoomModel.findRepresentativeAxisProxy().getMinMaxSpan();
            W(0, i, a, 0, n.minSpan != null ? _(n.minSpan, r, a, !0) : null, n.maxSpan != null ? _(n.maxSpan, r, a, !0) : null), this._range = C([_(i[0], a, r, !0), _(i[1], a, r, !0)]), this._updateView(), this._dispatchZoomAction(!1);
          }
        }
      }
    }, h.prototype._onBrush = function(t) {
      this._brushing && (V(t.event), this._updateBrushRect(t.offsetX, t.offsetY));
    }, h.prototype._updateBrushRect = function(t, e) {
      var o = this._displayables, s = this.dataZoomModel, a = o.brushRect;
      a || (a = o.brushRect = new I({
        silent: !0,
        style: s.getModel("brushStyle").getItemStyle()
      }), o.sliderGroup.add(a)), a.attr("ignore", !1);
      var r = this._brushStart, i = this._displayables.sliderGroup, n = i.transformCoordToLocal(t, e), l = i.transformCoordToLocal(r.x, r.y), d = this._size;
      n[0] = Math.max(Math.min(d[0], n[0]), 0), a.setShape({
        x: l[0],
        y: 0,
        width: n[0] - l[0],
        height: d[1]
      });
    }, h.prototype._dispatchZoomAction = function(t) {
      var e = this._range;
      this.api.dispatchAction({
        type: "dataZoom",
        from: this.uid,
        dataZoomId: this.dataZoomModel.id,
        animation: t ? St : null,
        start: e[0],
        end: e[1]
      });
    }, h.prototype._findCoordRect = function() {
      var t, e = ht(this.dataZoomModel).infoList;
      if (!t && e.length) {
        var o = e[0].model.coordinateSystem;
        t = o.getRect && o.getRect();
      }
      if (!t) {
        var s = this.api.getWidth(), a = this.api.getHeight();
        t = {
          x: s * 0.2,
          y: a * 0.2,
          width: s * 0.6,
          height: a * 0.6
        };
      }
      return t;
    }, h.type = "dataZoom.slider", h;
  })(rt)
);
function bt(w) {
  var h = {
    x: "y",
    y: "x",
    radius: "angle",
    angle: "radius"
  };
  return h[w];
}
function Mt(w) {
  return w === "vertical" ? "ns-resize" : "ew-resize";
}
export {
  Yt as default
};
