// echarts@6.0.0/lib/component/dataZoom/SliderZoomModel.js
import { __extends as a } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import r from "./DataZoomModel.js";
import { inheritDefaultOption as c } from "../../util/component.js";
import o from "../../visual/tokens.js";
var p = (
  /** @class */
  (function(t) {
    a(e, t);
    function e() {
      var l = t !== null && t.apply(this, arguments) || this;
      return l.type = e.type, l;
    }
    return e.type = "dataZoom.slider", e.layoutMode = "box", e.defaultOption = c(r.defaultOption, {
      show: !0,
      // deault value can only be drived in view stage.
      right: "ph",
      top: "ph",
      width: "ph",
      height: "ph",
      left: null,
      bottom: null,
      borderColor: o.color.accent10,
      borderRadius: 0,
      backgroundColor: o.color.transparent,
      // dataBackgroundColor: '#ddd',
      dataBackground: {
        lineStyle: {
          color: o.color.accent30,
          width: 0.5
        },
        areaStyle: {
          color: o.color.accent20,
          opacity: 0.2
        }
      },
      selectedDataBackground: {
        lineStyle: {
          color: o.color.accent40,
          width: 0.5
        },
        areaStyle: {
          color: o.color.accent20,
          opacity: 0.3
        }
      },
      // Color of selected window.
      fillerColor: "rgba(135,175,274,0.2)",
      handleIcon: "path://M-9.35,34.56V42m0-40V9.5m-2,0h4a2,2,0,0,1,2,2v21a2,2,0,0,1-2,2h-4a2,2,0,0,1-2-2v-21A2,2,0,0,1-11.35,9.5Z",
      // Percent of the slider height
      handleSize: "100%",
      handleStyle: {
        color: o.color.neutral00,
        borderColor: o.color.accent20
      },
      moveHandleSize: 7,
      moveHandleIcon: "path://M-320.9-50L-320.9-50c18.1,0,27.1,9,27.1,27.1V85.7c0,18.1-9,27.1-27.1,27.1l0,0c-18.1,0-27.1-9-27.1-27.1V-22.9C-348-41-339-50-320.9-50z M-212.3-50L-212.3-50c18.1,0,27.1,9,27.1,27.1V85.7c0,18.1-9,27.1-27.1,27.1l0,0c-18.1,0-27.1-9-27.1-27.1V-22.9C-239.4-41-230.4-50-212.3-50z M-103.7-50L-103.7-50c18.1,0,27.1,9,27.1,27.1V85.7c0,18.1-9,27.1-27.1,27.1l0,0c-18.1,0-27.1-9-27.1-27.1V-22.9C-130.9-41-121.8-50-103.7-50z",
      moveHandleStyle: {
        color: o.color.accent40,
        opacity: 0.5
      },
      showDetail: !0,
      showDataShadow: "auto",
      realtime: !0,
      zoomLock: !1,
      textStyle: {
        color: o.color.tertiary
      },
      brushSelect: !0,
      brushStyle: {
        color: o.color.accent30,
        opacity: 0.3
      },
      emphasis: {
        handleLabel: {
          show: !0
        },
        handleStyle: {
          borderColor: o.color.accent40
        },
        moveHandleStyle: {
          opacity: 0.8
        }
      },
      defaultLocationEdgeGap: 15
    }), e;
  })(r)
);
export {
  p as default
};
