// prosemirror-tables@1.6.4/dist/index.js
import { Plugin as fe, PluginKey as Z, Selection as z, SelectionRange as ye, TextSelection as E, NodeSelection as Se } from "../../../../prosemirror-state@1.4.3/node_modules/prosemirror-state/dist/index.js";
import { Fragment as R, Slice as v } from "../../../../prosemirror-model@1.25.0/node_modules/prosemirror-model/dist/index.js";
import { DecorationSet as X, Decoration as q } from "../../../../prosemirror-view@1.38.1/node_modules/prosemirror-view/dist/index.js";
import { keydownHandler as Ae } from "../../../../prosemirror-keymap@1.2.2/node_modules/prosemirror-keymap/dist/index.js";
import { Transform as Re } from "../../../../prosemirror-transform@1.10.3/node_modules/prosemirror-transform/dist/index.js";
var J, U;
if (typeof WeakMap < "u") {
  let t = /* @__PURE__ */ new WeakMap();
  J = (e) => t.get(e), U = (e, o) => (t.set(e, o), o);
} else {
  const t = [];
  let o = 0;
  J = (n) => {
    for (let l = 0; l < t.length; l += 2)
      if (t[l] == n) return t[l + 1];
  }, U = (n, l) => (o == 10 && (o = 0), t[o++] = n, t[o++] = l);
}
var C = class {
  constructor(t, e, o, n) {
    this.width = t, this.height = e, this.map = o, this.problems = n;
  }
  // Find the dimensions of the cell at the given position.
  findCell(t) {
    for (let e = 0; e < this.map.length; e++) {
      const o = this.map[e];
      if (o != t) continue;
      const n = e % this.width, l = e / this.width | 0;
      let r = n + 1, s = l + 1;
      for (let c = 1; r < this.width && this.map[e + c] == o; c++)
        r++;
      for (let c = 1; s < this.height && this.map[e + this.width * c] == o; c++)
        s++;
      return { left: n, top: l, right: r, bottom: s };
    }
    throw new RangeError(`No cell with offset ${t} found`);
  }
  // Find the left side of the cell at the given position.
  colCount(t) {
    for (let e = 0; e < this.map.length; e++)
      if (this.map[e] == t)
        return e % this.width;
    throw new RangeError(`No cell with offset ${t} found`);
  }
  // Find the next cell in the given direction, starting from the cell
  // at `pos`, if any.
  nextCell(t, e, o) {
    const { left: n, right: l, top: r, bottom: s } = this.findCell(t);
    return e == "horiz" ? (o < 0 ? n == 0 : l == this.width) ? null : this.map[r * this.width + (o < 0 ? n - 1 : l)] : (o < 0 ? r == 0 : s == this.height) ? null : this.map[n + this.width * (o < 0 ? r - 1 : s)];
  }
  // Get the rectangle spanning the two given cells.
  rectBetween(t, e) {
    const {
      left: o,
      right: n,
      top: l,
      bottom: r
    } = this.findCell(t), {
      left: s,
      right: c,
      top: i,
      bottom: a
    } = this.findCell(e);
    return {
      left: Math.min(o, s),
      top: Math.min(l, i),
      right: Math.max(n, c),
      bottom: Math.max(r, a)
    };
  }
  // Return the position of all cells that have the top left corner in
  // the given rectangle.
  cellsInRect(t) {
    const e = [], o = {};
    for (let n = t.top; n < t.bottom; n++)
      for (let l = t.left; l < t.right; l++) {
        const r = n * this.width + l, s = this.map[r];
        o[s] || (o[s] = !0, !(l == t.left && l && this.map[r - 1] == s || n == t.top && n && this.map[r - this.width] == s) && e.push(s));
      }
    return e;
  }
  // Return the position at which the cell at the given row and column
  // starts, or would start, if a cell started there.
  positionAt(t, e, o) {
    for (let n = 0, l = 0; ; n++) {
      const r = l + o.child(n).nodeSize;
      if (n == t) {
        let s = e + t * this.width;
        const c = (t + 1) * this.width;
        for (; s < c && this.map[s] < l; ) s++;
        return s == c ? r - 1 : this.map[s];
      }
      l = r;
    }
  }
  // Find the table map for the given table node.
  static get(t) {
    return J(t) || U(t, xe(t));
  }
};
function xe(t) {
  if (t.type.spec.tableRole != "table")
    throw new RangeError("Not a table node: " + t.type.name);
  const e = Me(t), o = t.childCount, n = [];
  let l = 0, r = null;
  const s = [];
  for (let a = 0, d = e * o; a < d; a++) n[a] = 0;
  for (let a = 0, d = 0; a < o; a++) {
    const f = t.child(a);
    d++;
    for (let p = 0; ; p++) {
      for (; l < n.length && n[l] != 0; ) l++;
      if (p == f.childCount) break;
      const m = f.child(p), { colspan: g, rowspan: b, colwidth: _ } = m.attrs;
      for (let H = 0; H < b; H++) {
        if (H + a >= o) {
          (r || (r = [])).push({
            type: "overlong_rowspan",
            pos: d,
            n: b - H
          });
          break;
        }
        const j = l + H * e;
        for (let k = 0; k < g; k++) {
          n[j + k] == 0 ? n[j + k] = d : (r || (r = [])).push({
            type: "collision",
            row: a,
            pos: d,
            n: g - k
          });
          const L = _ && _[k];
          if (L) {
            const B = (j + k) % e * 2, K = s[B];
            K == null || K != L && s[B + 1] == 1 ? (s[B] = L, s[B + 1] = 1) : K == L && s[B + 1]++;
          }
        }
      }
      l += g, d += m.nodeSize;
    }
    const u = (a + 1) * e;
    let h = 0;
    for (; l < u; ) n[l++] == 0 && h++;
    h && (r || (r = [])).push({ type: "missing", row: a, n: h }), d++;
  }
  (e === 0 || o === 0) && (r || (r = [])).push({ type: "zero_sized" });
  const c = new C(e, o, n, r);
  let i = !1;
  for (let a = 0; !i && a < s.length; a += 2)
    s[a] != null && s[a + 1] < o && (i = !0);
  return i && ze(c, s, t), c;
}
function Me(t) {
  let e = -1, o = !1;
  for (let n = 0; n < t.childCount; n++) {
    const l = t.child(n);
    let r = 0;
    if (o)
      for (let s = 0; s < n; s++) {
        const c = t.child(s);
        for (let i = 0; i < c.childCount; i++) {
          const a = c.child(i);
          s + a.attrs.rowspan > n && (r += a.attrs.colspan);
        }
      }
    for (let s = 0; s < l.childCount; s++) {
      const c = l.child(s);
      r += c.attrs.colspan, c.attrs.rowspan > 1 && (o = !0);
    }
    e == -1 ? e = r : e != r && (e = Math.max(e, r));
  }
  return e;
}
function ze(t, e, o) {
  t.problems || (t.problems = []);
  const n = {};
  for (let l = 0; l < t.map.length; l++) {
    const r = t.map[l];
    if (n[r]) continue;
    n[r] = !0;
    const s = o.nodeAt(r);
    if (!s)
      throw new RangeError(`No cell with offset ${r} found`);
    let c = null;
    const i = s.attrs;
    for (let a = 0; a < i.colspan; a++) {
      const d = (l + a) % t.width, f = e[d * 2];
      f != null && (!i.colwidth || i.colwidth[a] != f) && ((c || (c = Ne(i)))[a] = f);
    }
    c && t.problems.unshift({
      type: "colwidth mismatch",
      pos: r,
      colwidth: c
    });
  }
}
function Ne(t) {
  if (t.colwidth) return t.colwidth.slice();
  const e = [];
  for (let o = 0; o < t.colspan; o++) e.push(0);
  return e;
}
function y(t) {
  let e = t.cached.tableNodeTypes;
  if (!e) {
    e = t.cached.tableNodeTypes = {};
    for (const o in t.nodes) {
      const n = t.nodes[o], l = n.spec.tableRole;
      l && (e[l] = n);
    }
  }
  return e;
}
var N = new Z("selectingCells");
function D(t) {
  for (let e = t.depth - 1; e > 0; e--)
    if (t.node(e).type.spec.tableRole == "row")
      return t.node(0).resolve(t.before(e + 1));
  return null;
}
function ve(t) {
  for (let e = t.depth; e > 0; e--) {
    const o = t.node(e).type.spec.tableRole;
    if (o === "cell" || o === "header_cell") return t.node(e);
  }
  return null;
}
function A(t) {
  const e = t.selection.$head;
  for (let o = e.depth; o > 0; o--)
    if (e.node(o).type.spec.tableRole == "row") return !0;
  return !1;
}
function O(t) {
  const e = t.selection;
  if ("$anchorCell" in e && e.$anchorCell)
    return e.$anchorCell.pos > e.$headCell.pos ? e.$anchorCell : e.$headCell;
  if ("node" in e && e.node && e.node.type.spec.tableRole == "cell")
    return e.$anchor;
  const o = D(e.$head) || ke(e.$head);
  if (o)
    return o;
  throw new RangeError(`No cell found around position ${e.head}`);
}
function ke(t) {
  for (let e = t.nodeAfter, o = t.pos; e; e = e.firstChild, o++) {
    const n = e.type.spec.tableRole;
    if (n == "cell" || n == "header_cell") return t.doc.resolve(o);
  }
  for (let e = t.nodeBefore, o = t.pos; e; e = e.lastChild, o--) {
    const n = e.type.spec.tableRole;
    if (n == "cell" || n == "header_cell")
      return t.doc.resolve(o - e.nodeSize);
  }
}
function Y(t) {
  return t.parent.type.spec.tableRole == "row" && !!t.nodeAfter;
}
function Ee(t) {
  return t.node(0).resolve(t.pos + t.nodeAfter.nodeSize);
}
function ee(t, e) {
  return t.depth == e.depth && t.pos >= e.start(-1) && t.pos <= e.end(-1);
}
function ue(t, e, o) {
  const n = t.node(-1), l = C.get(n), r = t.start(-1), s = l.nextCell(t.pos - r, e, o);
  return s == null ? null : t.node(0).resolve(r + s);
}
function T(t, e, o = 1) {
  const n = { ...t, colspan: t.colspan - o };
  return n.colwidth && (n.colwidth = n.colwidth.slice(), n.colwidth.splice(e, o), n.colwidth.some((l) => l > 0) || (n.colwidth = null)), n;
}
function he(t, e, o = 1) {
  const n = { ...t, colspan: t.colspan + o };
  if (n.colwidth) {
    n.colwidth = n.colwidth.slice();
    for (let l = 0; l < o; l++) n.colwidth.splice(e, 0, 0);
  }
  return n;
}
function Te(t, e, o) {
  const n = y(e.type.schema).header_cell;
  for (let l = 0; l < t.height; l++)
    if (e.nodeAt(t.map[o + l * t.width]).type != n)
      return !1;
  return !0;
}
var w = class M extends z {
  // A table selection is identified by its anchor and head cells. The
  // positions given to this constructor should point _before_ two
  // cells in the same table. They may be the same, to select a single
  // cell.
  constructor(e, o = e) {
    const n = e.node(-1), l = C.get(n), r = e.start(-1), s = l.rectBetween(
      e.pos - r,
      o.pos - r
    ), c = e.node(0), i = l.cellsInRect(s).filter((d) => d != o.pos - r);
    i.unshift(o.pos - r);
    const a = i.map((d) => {
      const f = n.nodeAt(d);
      if (!f)
        throw RangeError(`No cell with offset ${d} found`);
      const u = r + d + 1;
      return new ye(
        c.resolve(u),
        c.resolve(u + f.content.size)
      );
    });
    super(a[0].$from, a[0].$to, a), this.$anchorCell = e, this.$headCell = o;
  }
  map(e, o) {
    const n = e.resolve(o.map(this.$anchorCell.pos)), l = e.resolve(o.map(this.$headCell.pos));
    if (Y(n) && Y(l) && ee(n, l)) {
      const r = this.$anchorCell.node(-1) != n.node(-1);
      return r && this.isRowSelection() ? M.rowSelection(n, l) : r && this.isColSelection() ? M.colSelection(n, l) : new M(n, l);
    }
    return E.between(n, l);
  }
  // Returns a rectangular slice of table rows containing the selected
  // cells.
  content() {
    const e = this.$anchorCell.node(-1), o = C.get(e), n = this.$anchorCell.start(-1), l = o.rectBetween(
      this.$anchorCell.pos - n,
      this.$headCell.pos - n
    ), r = {}, s = [];
    for (let i = l.top; i < l.bottom; i++) {
      const a = [];
      for (let d = i * o.width + l.left, f = l.left; f < l.right; f++, d++) {
        const u = o.map[d];
        if (r[u]) continue;
        r[u] = !0;
        const h = o.findCell(u);
        let p = e.nodeAt(u);
        if (!p)
          throw RangeError(`No cell with offset ${u} found`);
        const m = l.left - h.left, g = h.right - l.right;
        if (m > 0 || g > 0) {
          let b = p.attrs;
          if (m > 0 && (b = T(b, 0, m)), g > 0 && (b = T(
            b,
            b.colspan - g,
            g
          )), h.left < l.left) {
            if (p = p.type.createAndFill(b), !p)
              throw RangeError(
                `Could not create cell with attrs ${JSON.stringify(b)}`
              );
          } else
            p = p.type.create(b, p.content);
        }
        if (h.top < l.top || h.bottom > l.bottom) {
          const b = {
            ...p.attrs,
            rowspan: Math.min(h.bottom, l.bottom) - Math.max(h.top, l.top)
          };
          h.top < l.top ? p = p.type.createAndFill(b) : p = p.type.create(b, p.content);
        }
        a.push(p);
      }
      s.push(e.child(i).copy(R.from(a)));
    }
    const c = this.isColSelection() && this.isRowSelection() ? e : s;
    return new v(R.from(c), 1, 1);
  }
  replace(e, o = v.empty) {
    const n = e.steps.length, l = this.ranges;
    for (let s = 0; s < l.length; s++) {
      const { $from: c, $to: i } = l[s], a = e.mapping.slice(n);
      e.replace(
        a.map(c.pos),
        a.map(i.pos),
        s ? v.empty : o
      );
    }
    const r = z.findFrom(
      e.doc.resolve(e.mapping.slice(n).map(this.to)),
      -1
    );
    r && e.setSelection(r);
  }
  replaceWith(e, o) {
    this.replace(e, new v(R.from(o), 0, 0));
  }
  forEachCell(e) {
    const o = this.$anchorCell.node(-1), n = C.get(o), l = this.$anchorCell.start(-1), r = n.cellsInRect(
      n.rectBetween(
        this.$anchorCell.pos - l,
        this.$headCell.pos - l
      )
    );
    for (let s = 0; s < r.length; s++)
      e(o.nodeAt(r[s]), l + r[s]);
  }
  // True if this selection goes all the way from the top to the
  // bottom of the table.
  isColSelection() {
    const e = this.$anchorCell.index(-1), o = this.$headCell.index(-1);
    if (Math.min(e, o) > 0) return !1;
    const n = e + this.$anchorCell.nodeAfter.attrs.rowspan, l = o + this.$headCell.nodeAfter.attrs.rowspan;
    return Math.max(n, l) == this.$headCell.node(-1).childCount;
  }
  // Returns the smallest column selection that covers the given anchor
  // and head cell.
  static colSelection(e, o = e) {
    const n = e.node(-1), l = C.get(n), r = e.start(-1), s = l.findCell(e.pos - r), c = l.findCell(o.pos - r), i = e.node(0);
    return s.top <= c.top ? (s.top > 0 && (e = i.resolve(r + l.map[s.left])), c.bottom < l.height && (o = i.resolve(
      r + l.map[l.width * (l.height - 1) + c.right - 1]
    ))) : (c.top > 0 && (o = i.resolve(r + l.map[c.left])), s.bottom < l.height && (e = i.resolve(
      r + l.map[l.width * (l.height - 1) + s.right - 1]
    ))), new M(e, o);
  }
  // True if this selection goes all the way from the left to the
  // right of the table.
  isRowSelection() {
    const e = this.$anchorCell.node(-1), o = C.get(e), n = this.$anchorCell.start(-1), l = o.colCount(this.$anchorCell.pos - n), r = o.colCount(this.$headCell.pos - n);
    if (Math.min(l, r) > 0) return !1;
    const s = l + this.$anchorCell.nodeAfter.attrs.colspan, c = r + this.$headCell.nodeAfter.attrs.colspan;
    return Math.max(s, c) == o.width;
  }
  eq(e) {
    return e instanceof M && e.$anchorCell.pos == this.$anchorCell.pos && e.$headCell.pos == this.$headCell.pos;
  }
  // Returns the smallest row selection that covers the given anchor
  // and head cell.
  static rowSelection(e, o = e) {
    const n = e.node(-1), l = C.get(n), r = e.start(-1), s = l.findCell(e.pos - r), c = l.findCell(o.pos - r), i = e.node(0);
    return s.left <= c.left ? (s.left > 0 && (e = i.resolve(
      r + l.map[s.top * l.width]
    )), c.right < l.width && (o = i.resolve(
      r + l.map[l.width * (c.top + 1) - 1]
    ))) : (c.left > 0 && (o = i.resolve(r + l.map[c.top * l.width])), s.right < l.width && (e = i.resolve(
      r + l.map[l.width * (s.top + 1) - 1]
    ))), new M(e, o);
  }
  toJSON() {
    return {
      type: "cell",
      anchor: this.$anchorCell.pos,
      head: this.$headCell.pos
    };
  }
  static fromJSON(e, o) {
    return new M(e.resolve(o.anchor), e.resolve(o.head));
  }
  static create(e, o, n = o) {
    return new M(e.resolve(o), e.resolve(n));
  }
  getBookmark() {
    return new De(this.$anchorCell.pos, this.$headCell.pos);
  }
};
w.prototype.visible = !1;
z.jsonID("cell", w);
var De = class pe {
  constructor(e, o) {
    this.anchor = e, this.head = o;
  }
  map(e) {
    return new pe(e.map(this.anchor), e.map(this.head));
  }
  resolve(e) {
    const o = e.resolve(this.anchor), n = e.resolve(this.head);
    return o.parent.type.spec.tableRole == "row" && n.parent.type.spec.tableRole == "row" && o.index() < o.parent.childCount && n.index() < n.parent.childCount && ee(o, n) ? new w(o, n) : z.near(n, 1);
  }
};
function _e(t) {
  if (!(t.selection instanceof w)) return null;
  const e = [];
  return t.selection.forEachCell((o, n) => {
    e.push(
      q.node(n, n + o.nodeSize, { class: "selectedCell" })
    );
  }), X.create(t.doc, e);
}
function He({ $from: t, $to: e }) {
  if (t.pos == e.pos || t.pos < e.pos - 6) return !1;
  let o = t.pos, n = e.pos, l = t.depth;
  for (; l >= 0 && !(t.after(l + 1) < t.end(l)); l--, o++)
    ;
  for (let r = e.depth; r >= 0 && !(e.before(r + 1) > e.start(r)); r--, n--)
    ;
  return o == n && /row|table/.test(t.node(l).type.spec.tableRole);
}
function Be({ $from: t, $to: e }) {
  let o, n;
  for (let l = t.depth; l > 0; l--) {
    const r = t.node(l);
    if (r.type.spec.tableRole === "cell" || r.type.spec.tableRole === "header_cell") {
      o = r;
      break;
    }
  }
  for (let l = e.depth; l > 0; l--) {
    const r = e.node(l);
    if (r.type.spec.tableRole === "cell" || r.type.spec.tableRole === "header_cell") {
      n = r;
      break;
    }
  }
  return o !== n && e.parentOffset === 0;
}
function Le(t, e, o) {
  const n = (e || t).selection, l = (e || t).doc;
  let r, s;
  if (n instanceof Se && (s = n.node.type.spec.tableRole)) {
    if (s == "cell" || s == "header_cell")
      r = w.create(l, n.from);
    else if (s == "row") {
      const c = l.resolve(n.from + 1);
      r = w.rowSelection(c, c);
    } else if (!o) {
      const c = C.get(n.node), i = n.from + 1, a = i + c.map[c.width * c.height - 1];
      r = w.create(l, i + 1, a);
    }
  } else n instanceof E && He(n) ? r = E.create(l, n.from) : n instanceof E && Be(n) && (r = E.create(l, n.$from.start(), n.$from.end()));
  return r && (e || (e = t.tr)).setSelection(r), e;
}
var $e = new Z("fix-tables");
function me(t, e, o, n) {
  const l = t.childCount, r = e.childCount;
  e: for (let s = 0, c = 0; s < r; s++) {
    const i = e.child(s);
    for (let a = c, d = Math.min(l, s + 3); a < d; a++)
      if (t.child(a) == i) {
        c = a + 1, o += i.nodeSize;
        continue e;
      }
    n(i, o), c < l && t.child(c).sameMarkup(i) ? me(t.child(c), i, o + 1, n) : i.nodesBetween(0, i.content.size, n, o + 1), o += i.nodeSize;
  }
}
function Fe(t, e) {
  let o;
  const n = (l, r) => {
    l.type.spec.tableRole == "table" && (o = Pe(t, l, r, o));
  };
  return e ? e.doc != t.doc && me(e.doc, t.doc, 0, n) : t.doc.descendants(n), o;
}
function Pe(t, e, o, n) {
  const l = C.get(e);
  if (!l.problems) return n;
  n || (n = t.tr);
  const r = [];
  for (let i = 0; i < l.height; i++) r.push(0);
  for (let i = 0; i < l.problems.length; i++) {
    const a = l.problems[i];
    if (a.type == "collision") {
      const d = e.nodeAt(a.pos);
      if (!d) continue;
      const f = d.attrs;
      for (let u = 0; u < f.rowspan; u++) r[a.row + u] += a.n;
      n.setNodeMarkup(
        n.mapping.map(o + 1 + a.pos),
        null,
        T(f, f.colspan - a.n, a.n)
      );
    } else if (a.type == "missing")
      r[a.row] += a.n;
    else if (a.type == "overlong_rowspan") {
      const d = e.nodeAt(a.pos);
      if (!d) continue;
      n.setNodeMarkup(n.mapping.map(o + 1 + a.pos), null, {
        ...d.attrs,
        rowspan: d.attrs.rowspan - a.n
      });
    } else if (a.type == "colwidth mismatch") {
      const d = e.nodeAt(a.pos);
      if (!d) continue;
      n.setNodeMarkup(n.mapping.map(o + 1 + a.pos), null, {
        ...d.attrs,
        colwidth: a.colwidth
      });
    } else if (a.type == "zero_sized") {
      const d = n.mapping.map(o);
      n.delete(d, d + e.nodeSize);
    }
  }
  let s, c;
  for (let i = 0; i < r.length; i++)
    r[i] && (s == null && (s = i), c = i);
  for (let i = 0, a = o + 1; i < l.height; i++) {
    const d = e.child(i), f = a + d.nodeSize, u = r[i];
    if (u > 0) {
      let h = "cell";
      d.firstChild && (h = d.firstChild.type.spec.tableRole);
      const p = [];
      for (let g = 0; g < u; g++) {
        const b = y(t.schema)[h].createAndFill();
        b && p.push(b);
      }
      const m = (i == 0 || s == i - 1) && c == i ? a + 1 : f - 1;
      n.insert(n.mapping.map(m), p);
    }
    a = f;
  }
  return n.setMeta($e, { fixTables: !0 });
}
function x(t) {
  const e = t.selection, o = O(t), n = o.node(-1), l = o.start(-1), r = C.get(n);
  return { ...e instanceof w ? r.rectBetween(
    e.$anchorCell.pos - l,
    e.$headCell.pos - l
  ) : r.findCell(o.pos - l), tableStart: l, map: r, table: n };
}
function ge(t, { map: e, tableStart: o, table: n }, l) {
  let r = l > 0 ? -1 : 0;
  Te(e, n, l + r) && (r = l == 0 || l == e.width ? null : 0);
  for (let s = 0; s < e.height; s++) {
    const c = s * e.width + l;
    if (l > 0 && l < e.width && e.map[c - 1] == e.map[c]) {
      const i = e.map[c], a = n.nodeAt(i);
      t.setNodeMarkup(
        t.mapping.map(o + i),
        null,
        he(a.attrs, l - e.colCount(i))
      ), s += a.attrs.rowspan - 1;
    } else {
      const i = r == null ? y(n.type.schema).cell : n.nodeAt(e.map[c + r]).type, a = e.positionAt(s, l, n);
      t.insert(t.mapping.map(o + a), i.createAndFill());
    }
  }
  return t;
}
function gt(t, e) {
  if (!A(t)) return !1;
  if (e) {
    const o = x(t);
    e(ge(t.tr, o, o.left));
  }
  return !0;
}
function wt(t, e) {
  if (!A(t)) return !1;
  if (e) {
    const o = x(t);
    e(ge(t.tr, o, o.right));
  }
  return !0;
}
function We(t, { map: e, table: o, tableStart: n }, l) {
  const r = t.mapping.maps.length;
  for (let s = 0; s < e.height; ) {
    const c = s * e.width + l, i = e.map[c], a = o.nodeAt(i), d = a.attrs;
    if (l > 0 && e.map[c - 1] == i || l < e.width - 1 && e.map[c + 1] == i)
      t.setNodeMarkup(
        t.mapping.slice(r).map(n + i),
        null,
        T(d, l - e.colCount(i))
      );
    else {
      const f = t.mapping.slice(r).map(n + i);
      t.delete(f, f + a.nodeSize);
    }
    s += d.rowspan;
  }
}
function Ct(t, e) {
  if (!A(t)) return !1;
  if (e) {
    const o = x(t), n = t.tr;
    if (o.left == 0 && o.right == o.map.width) return !1;
    for (let l = o.right - 1; We(n, o, l), l != o.left; l--) {
      const r = o.tableStart ? n.doc.nodeAt(o.tableStart - 1) : n.doc;
      if (!r)
        throw RangeError("No table found");
      o.table = r, o.map = C.get(r);
    }
    e(n);
  }
  return !0;
}
function Ie(t, e, o) {
  var n;
  const l = y(e.type.schema).header_cell;
  for (let r = 0; r < t.width; r++)
    if (((n = e.nodeAt(t.map[r + o * t.width])) == null ? void 0 : n.type) != l)
      return !1;
  return !0;
}
function we(t, { map: e, tableStart: o, table: n }, l) {
  var r;
  let s = o;
  for (let a = 0; a < l; a++) s += n.child(a).nodeSize;
  const c = [];
  let i = l > 0 ? -1 : 0;
  Ie(e, n, l + i) && (i = l == 0 || l == e.height ? null : 0);
  for (let a = 0, d = e.width * l; a < e.width; a++, d++)
    if (l > 0 && l < e.height && e.map[d] == e.map[d - e.width]) {
      const f = e.map[d], u = n.nodeAt(f).attrs;
      t.setNodeMarkup(o + f, null, {
        ...u,
        rowspan: u.rowspan + 1
      }), a += u.colspan - 1;
    } else {
      const f = i == null ? y(n.type.schema).cell : (r = n.nodeAt(e.map[d + i * e.width])) == null ? void 0 : r.type, u = f?.createAndFill();
      u && c.push(u);
    }
  return t.insert(s, y(n.type.schema).row.create(null, c)), t;
}
function bt(t, e) {
  if (!A(t)) return !1;
  if (e) {
    const o = x(t);
    e(we(t.tr, o, o.top));
  }
  return !0;
}
function yt(t, e) {
  if (!A(t)) return !1;
  if (e) {
    const o = x(t);
    e(we(t.tr, o, o.bottom));
  }
  return !0;
}
function Oe(t, { map: e, table: o, tableStart: n }, l) {
  let r = 0;
  for (let a = 0; a < l; a++) r += o.child(a).nodeSize;
  const s = r + o.child(l).nodeSize, c = t.mapping.maps.length;
  t.delete(r + n, s + n);
  const i = /* @__PURE__ */ new Set();
  for (let a = 0, d = l * e.width; a < e.width; a++, d++) {
    const f = e.map[d];
    if (!i.has(f)) {
      if (i.add(f), l > 0 && f == e.map[d - e.width]) {
        const u = o.nodeAt(f).attrs;
        t.setNodeMarkup(t.mapping.slice(c).map(f + n), null, {
          ...u,
          rowspan: u.rowspan - 1
        }), a += u.colspan - 1;
      } else if (l < e.height && f == e.map[d + e.width]) {
        const u = o.nodeAt(f), h = u.attrs, p = u.type.create(
          { ...h, rowspan: u.attrs.rowspan - 1 },
          u.content
        ), m = e.positionAt(l + 1, a, o);
        t.insert(t.mapping.slice(c).map(n + m), p), a += h.colspan - 1;
      }
    }
  }
}
function St(t, e) {
  if (!A(t)) return !1;
  if (e) {
    const o = x(t), n = t.tr;
    if (o.top == 0 && o.bottom == o.map.height) return !1;
    for (let l = o.bottom - 1; Oe(n, o, l), l != o.top; l--) {
      const r = o.tableStart ? n.doc.nodeAt(o.tableStart - 1) : n.doc;
      if (!r)
        throw RangeError("No table found");
      o.table = r, o.map = C.get(o.table);
    }
    e(n);
  }
  return !0;
}
function oe(t) {
  const e = t.content;
  return e.childCount == 1 && e.child(0).isTextblock && e.child(0).childCount == 0;
}
function je({ width: t, height: e, map: o }, n) {
  let l = n.top * t + n.left, r = l, s = (n.bottom - 1) * t + n.left, c = l + (n.right - n.left - 1);
  for (let i = n.top; i < n.bottom; i++) {
    if (n.left > 0 && o[r] == o[r - 1] || n.right < t && o[c] == o[c + 1])
      return !0;
    r += t, c += t;
  }
  for (let i = n.left; i < n.right; i++) {
    if (n.top > 0 && o[l] == o[l - t] || n.bottom < e && o[s] == o[s + t])
      return !0;
    l++, s++;
  }
  return !1;
}
function At(t, e) {
  const o = t.selection;
  if (!(o instanceof w) || o.$anchorCell.pos == o.$headCell.pos)
    return !1;
  const n = x(t), { map: l } = n;
  if (je(l, n)) return !1;
  if (e) {
    const r = t.tr, s = {};
    let c = R.empty, i, a;
    for (let d = n.top; d < n.bottom; d++)
      for (let f = n.left; f < n.right; f++) {
        const u = l.map[d * l.width + f], h = n.table.nodeAt(u);
        if (!(s[u] || !h))
          if (s[u] = !0, i == null)
            i = u, a = h;
          else {
            oe(h) || (c = c.append(h.content));
            const p = r.mapping.map(u + n.tableStart);
            r.delete(p, p + h.nodeSize);
          }
      }
    if (i == null || a == null)
      return !0;
    if (r.setNodeMarkup(i + n.tableStart, null, {
      ...he(
        a.attrs,
        a.attrs.colspan,
        n.right - n.left - a.attrs.colspan
      ),
      rowspan: n.bottom - n.top
    }), c.size) {
      const d = i + 1 + a.content.size, f = oe(a) ? i + 1 : d;
      r.replaceWith(f + n.tableStart, d + n.tableStart, c);
    }
    r.setSelection(
      new w(r.doc.resolve(i + n.tableStart))
    ), e(r);
  }
  return !0;
}
function Rt(t, e) {
  const o = y(t.schema);
  return Ke(({ node: n }) => o[n.type.spec.tableRole])(t, e);
}
function Ke(t) {
  return (e, o) => {
    var n;
    const l = e.selection;
    let r, s;
    if (l instanceof w) {
      if (l.$anchorCell.pos != l.$headCell.pos) return !1;
      r = l.$anchorCell.nodeAfter, s = l.$anchorCell.pos;
    } else {
      if (r = ve(l.$from), !r) return !1;
      s = (n = D(l.$from)) == null ? void 0 : n.pos;
    }
    if (r == null || s == null || r.attrs.colspan == 1 && r.attrs.rowspan == 1)
      return !1;
    if (o) {
      let c = r.attrs;
      const i = [], a = c.colwidth;
      c.rowspan > 1 && (c = { ...c, rowspan: 1 }), c.colspan > 1 && (c = { ...c, colspan: 1 });
      const d = x(e), f = e.tr;
      for (let h = 0; h < d.right - d.left; h++)
        i.push(
          a ? {
            ...c,
            colwidth: a && a[h] ? [a[h]] : null
          } : c
        );
      let u;
      for (let h = d.top; h < d.bottom; h++) {
        let p = d.map.positionAt(h, d.left, d.table);
        h == d.top && (p += r.nodeSize);
        for (let m = d.left, g = 0; m < d.right; m++, g++)
          m == d.left && h == d.top || f.insert(
            u = f.mapping.map(p + d.tableStart, 1),
            t({ node: r, row: h, col: m }).createAndFill(i[g])
          );
      }
      f.setNodeMarkup(
        s,
        t({ node: r, row: d.top, col: d.left }),
        i[0]
      ), l instanceof w && f.setSelection(
        new w(
          f.doc.resolve(l.$anchorCell.pos),
          u ? f.doc.resolve(u) : void 0
        )
      ), o(f);
    }
    return !0;
  };
}
function xt(t, e) {
  return function(o, n) {
    if (!A(o)) return !1;
    const l = O(o);
    if (l.nodeAfter.attrs[t] === e) return !1;
    if (n) {
      const r = o.tr;
      o.selection instanceof w ? o.selection.forEachCell((s, c) => {
        s.attrs[t] !== e && r.setNodeMarkup(c, null, {
          ...s.attrs,
          [t]: e
        });
      }) : r.setNodeMarkup(l.pos, null, {
        ...l.nodeAfter.attrs,
        [t]: e
      }), n(r);
    }
    return !0;
  };
}
function Ve(t) {
  return function(e, o) {
    if (!A(e)) return !1;
    if (o) {
      const n = y(e.schema), l = x(e), r = e.tr, s = l.map.cellsInRect(
        t == "column" ? {
          left: l.left,
          top: 0,
          right: l.right,
          bottom: l.map.height
        } : t == "row" ? {
          left: 0,
          top: l.top,
          right: l.map.width,
          bottom: l.bottom
        } : l
      ), c = s.map((i) => l.table.nodeAt(i));
      for (let i = 0; i < s.length; i++)
        c[i].type == n.header_cell && r.setNodeMarkup(
          l.tableStart + s[i],
          n.cell,
          c[i].attrs
        );
      if (r.steps.length == 0)
        for (let i = 0; i < s.length; i++)
          r.setNodeMarkup(
            l.tableStart + s[i],
            n.header_cell,
            c[i].attrs
          );
      o(r);
    }
    return !0;
  };
}
function ne(t, e, o) {
  const n = e.map.cellsInRect({
    left: 0,
    top: 0,
    right: t == "row" ? e.map.width : 1,
    bottom: t == "column" ? e.map.height : 1
  });
  for (let l = 0; l < n.length; l++) {
    const r = e.table.nodeAt(n[l]);
    if (r && r.type !== o.header_cell)
      return !1;
  }
  return !0;
}
function te(t, e) {
  return e = e || { useDeprecatedLogic: !1 }, e.useDeprecatedLogic ? Ve(t) : function(o, n) {
    if (!A(o)) return !1;
    if (n) {
      const l = y(o.schema), r = x(o), s = o.tr, c = ne("row", r, l), i = ne(
        "column",
        r,
        l
      ), d = (t === "column" ? c : t === "row" ? i : !1) ? 1 : 0, f = t == "column" ? {
        left: 0,
        top: d,
        right: 1,
        bottom: r.map.height
      } : t == "row" ? {
        left: d,
        top: 0,
        right: r.map.width,
        bottom: 1
      } : r, u = t == "column" ? i ? l.cell : l.header_cell : t == "row" ? c ? l.cell : l.header_cell : l.cell;
      r.map.cellsInRect(f).forEach((h) => {
        const p = h + r.tableStart, m = s.doc.nodeAt(p);
        m && s.setNodeMarkup(p, u, m.attrs);
      }), n(s);
    }
    return !0;
  };
}
te("row", {
  useDeprecatedLogic: !0
});
te("column", {
  useDeprecatedLogic: !0
});
var Mt = te("cell", {
  useDeprecatedLogic: !0
});
function Xe(t, e) {
  if (e < 0) {
    const o = t.nodeBefore;
    if (o) return t.pos - o.nodeSize;
    for (let n = t.index(-1) - 1, l = t.before(); n >= 0; n--) {
      const r = t.node(-1).child(n), s = r.lastChild;
      if (s)
        return l - 1 - s.nodeSize;
      l -= r.nodeSize;
    }
  } else {
    if (t.index() < t.parent.childCount - 1)
      return t.pos + t.nodeAfter.nodeSize;
    const o = t.node(-1);
    for (let n = t.indexAfter(-1), l = t.after(); n < o.childCount; n++) {
      const r = o.child(n);
      if (r.childCount) return l + 1;
      l += r.nodeSize;
    }
  }
  return null;
}
function zt(t) {
  return function(e, o) {
    if (!A(e)) return !1;
    const n = Xe(O(e), t);
    if (n == null) return !1;
    if (o) {
      const l = e.doc.resolve(n);
      o(
        e.tr.setSelection(E.between(l, Ee(l))).scrollIntoView()
      );
    }
    return !0;
  };
}
function Nt(t, e) {
  const o = t.selection.$anchor;
  for (let n = o.depth; n > 0; n--)
    if (o.node(n).type.spec.tableRole == "table")
      return e && e(
        t.tr.delete(o.before(n), o.after(n)).scrollIntoView()
      ), !0;
  return !1;
}
function $(t, e) {
  const o = t.selection;
  if (!(o instanceof w)) return !1;
  if (e) {
    const n = t.tr, l = y(t.schema).cell.createAndFill().content;
    o.forEachCell((r, s) => {
      r.content.eq(l) || n.replace(
        n.mapping.map(s + 1),
        n.mapping.map(s + r.nodeSize - 1),
        new v(l, 0, 0)
      );
    }), n.docChanged && e(n);
  }
  return !0;
}
function qe(t) {
  if (!t.size) return null;
  let { content: e, openStart: o, openEnd: n } = t;
  for (; e.childCount == 1 && (o > 0 && n > 0 || e.child(0).type.spec.tableRole == "table"); )
    o--, n--, e = e.child(0).content;
  const l = e.child(0), r = l.type.spec.tableRole, s = l.type.schema, c = [];
  if (r == "row")
    for (let i = 0; i < e.childCount; i++) {
      let a = e.child(i).content;
      const d = i ? 0 : Math.max(0, o - 1), f = i < e.childCount - 1 ? 0 : Math.max(0, n - 1);
      (d || f) && (a = G(
        y(s).row,
        new v(a, d, f)
      ).content), c.push(a);
    }
  else if (r == "cell" || r == "header_cell")
    c.push(
      o || n ? G(
        y(s).row,
        new v(e, o, n)
      ).content : e
    );
  else
    return null;
  return Je(s, c);
}
function Je(t, e) {
  const o = [];
  for (let l = 0; l < e.length; l++) {
    const r = e[l];
    for (let s = r.childCount - 1; s >= 0; s--) {
      const { rowspan: c, colspan: i } = r.child(s).attrs;
      for (let a = l; a < l + c; a++)
        o[a] = (o[a] || 0) + i;
    }
  }
  let n = 0;
  for (let l = 0; l < o.length; l++) n = Math.max(n, o[l]);
  for (let l = 0; l < o.length; l++)
    if (l >= e.length && e.push(R.empty), o[l] < n) {
      const r = y(t).cell.createAndFill(), s = [];
      for (let c = o[l]; c < n; c++)
        s.push(r);
      e[l] = e[l].append(R.from(s));
    }
  return { height: e.length, width: n, rows: e };
}
function G(t, e) {
  const o = t.createAndFill();
  return new Re(o).replace(0, o.content.size, e).doc;
}
function Ue({ width: t, height: e, rows: o }, n, l) {
  if (t != n) {
    const r = [], s = [];
    for (let c = 0; c < o.length; c++) {
      const i = o[c], a = [];
      for (let d = r[c] || 0, f = 0; d < n; f++) {
        let u = i.child(f % i.childCount);
        d + u.attrs.colspan > n && (u = u.type.createChecked(
          T(
            u.attrs,
            u.attrs.colspan,
            d + u.attrs.colspan - n
          ),
          u.content
        )), a.push(u), d += u.attrs.colspan;
        for (let h = 1; h < u.attrs.rowspan; h++)
          r[c + h] = (r[c + h] || 0) + u.attrs.colspan;
      }
      s.push(R.from(a));
    }
    o = s, t = n;
  }
  if (e != l) {
    const r = [];
    for (let s = 0, c = 0; s < l; s++, c++) {
      const i = [], a = o[c % e];
      for (let d = 0; d < a.childCount; d++) {
        let f = a.child(d);
        s + f.attrs.rowspan > l && (f = f.type.create(
          {
            ...f.attrs,
            rowspan: Math.max(1, l - f.attrs.rowspan)
          },
          f.content
        )), i.push(f);
      }
      r.push(R.from(i));
    }
    o = r, e = l;
  }
  return { width: t, height: e, rows: o };
}
function Ye(t, e, o, n, l, r, s) {
  const c = t.doc.type.schema, i = y(c);
  let a, d;
  if (l > e.width)
    for (let f = 0, u = 0; f < e.height; f++) {
      const h = o.child(f);
      u += h.nodeSize;
      const p = [];
      let m;
      h.lastChild == null || h.lastChild.type == i.cell ? m = a || (a = i.cell.createAndFill()) : m = d || (d = i.header_cell.createAndFill());
      for (let g = e.width; g < l; g++) p.push(m);
      t.insert(t.mapping.slice(s).map(u - 1 + n), p);
    }
  if (r > e.height) {
    const f = [];
    for (let p = 0, m = (e.height - 1) * e.width; p < Math.max(e.width, l); p++) {
      const g = p >= e.width ? !1 : o.nodeAt(e.map[m + p]).type == i.header_cell;
      f.push(
        g ? d || (d = i.header_cell.createAndFill()) : a || (a = i.cell.createAndFill())
      );
    }
    const u = i.row.create(null, R.from(f)), h = [];
    for (let p = e.height; p < r; p++) h.push(u);
    t.insert(t.mapping.slice(s).map(n + o.nodeSize - 2), h);
  }
  return !!(a || d);
}
function le(t, e, o, n, l, r, s, c) {
  if (s == 0 || s == e.height) return !1;
  let i = !1;
  for (let a = l; a < r; a++) {
    const d = s * e.width + a, f = e.map[d];
    if (e.map[d - e.width] == f) {
      i = !0;
      const u = o.nodeAt(f), { top: h, left: p } = e.findCell(f);
      t.setNodeMarkup(t.mapping.slice(c).map(f + n), null, {
        ...u.attrs,
        rowspan: s - h
      }), t.insert(
        t.mapping.slice(c).map(e.positionAt(s, p, o)),
        u.type.createAndFill({
          ...u.attrs,
          rowspan: h + u.attrs.rowspan - s
        })
      ), a += u.attrs.colspan - 1;
    }
  }
  return i;
}
function re(t, e, o, n, l, r, s, c) {
  if (s == 0 || s == e.width) return !1;
  let i = !1;
  for (let a = l; a < r; a++) {
    const d = a * e.width + s, f = e.map[d];
    if (e.map[d - 1] == f) {
      i = !0;
      const u = o.nodeAt(f), h = e.colCount(f), p = t.mapping.slice(c).map(f + n);
      t.setNodeMarkup(
        p,
        null,
        T(
          u.attrs,
          s - h,
          u.attrs.colspan - (s - h)
        )
      ), t.insert(
        p + u.nodeSize,
        u.type.createAndFill(
          T(u.attrs, 0, s - h)
        )
      ), a += u.attrs.rowspan - 1;
    }
  }
  return i;
}
function se(t, e, o, n, l) {
  let r = o ? t.doc.nodeAt(o - 1) : t.doc;
  if (!r)
    throw new Error("No table found");
  let s = C.get(r);
  const { top: c, left: i } = n, a = i + l.width, d = c + l.height, f = t.tr;
  let u = 0;
  function h() {
    if (r = o ? f.doc.nodeAt(o - 1) : f.doc, !r)
      throw new Error("No table found");
    s = C.get(r), u = f.mapping.maps.length;
  }
  Ye(f, s, r, o, a, d, u) && h(), le(f, s, r, o, i, a, c, u) && h(), le(f, s, r, o, i, a, d, u) && h(), re(f, s, r, o, c, d, i, u) && h(), re(f, s, r, o, c, d, a, u) && h();
  for (let p = c; p < d; p++) {
    const m = s.positionAt(p, i, r), g = s.positionAt(p, a, r);
    f.replace(
      f.mapping.slice(u).map(m + o),
      f.mapping.slice(u).map(g + o),
      new v(l.rows[p - c], 0, 0)
    );
  }
  h(), f.setSelection(
    new w(
      f.doc.resolve(o + s.positionAt(c, i, r)),
      f.doc.resolve(o + s.positionAt(d - 1, a - 1, r))
    )
  ), e(f);
}
var Ge = Ae({
  ArrowLeft: F("horiz", -1),
  ArrowRight: F("horiz", 1),
  ArrowUp: F("vert", -1),
  ArrowDown: F("vert", 1),
  "Shift-ArrowLeft": P("horiz", -1),
  "Shift-ArrowRight": P("horiz", 1),
  "Shift-ArrowUp": P("vert", -1),
  "Shift-ArrowDown": P("vert", 1),
  Backspace: $,
  "Mod-Backspace": $,
  Delete: $,
  "Mod-Delete": $
});
function W(t, e, o) {
  return o.eq(t.selection) ? !1 : (e && e(t.tr.setSelection(o).scrollIntoView()), !0);
}
function F(t, e) {
  return (o, n, l) => {
    if (!l) return !1;
    const r = o.selection;
    if (r instanceof w)
      return W(
        o,
        n,
        z.near(r.$headCell, e)
      );
    if (t != "horiz" && !r.empty) return !1;
    const s = Ce(l, t, e);
    if (s == null) return !1;
    if (t == "horiz")
      return W(
        o,
        n,
        z.near(o.doc.resolve(r.head + e), e)
      );
    {
      const c = o.doc.resolve(s), i = ue(c, t, e);
      let a;
      return i ? a = z.near(i, 1) : e < 0 ? a = z.near(o.doc.resolve(c.before(-1)), -1) : a = z.near(o.doc.resolve(c.after(-1)), 1), W(o, n, a);
    }
  };
}
function P(t, e) {
  return (o, n, l) => {
    if (!l) return !1;
    const r = o.selection;
    let s;
    if (r instanceof w)
      s = r;
    else {
      const i = Ce(l, t, e);
      if (i == null) return !1;
      s = new w(o.doc.resolve(i));
    }
    const c = ue(s.$headCell, t, e);
    return c ? W(
      o,
      n,
      new w(s.$anchorCell, c)
    ) : !1;
  };
}
function Qe(t, e) {
  const o = t.state.doc, n = D(o.resolve(e));
  return n ? (t.dispatch(t.state.tr.setSelection(new w(n))), !0) : !1;
}
function Ze(t, e, o) {
  if (!A(t.state)) return !1;
  let n = qe(o);
  const l = t.state.selection;
  if (l instanceof w) {
    n || (n = {
      width: 1,
      height: 1,
      rows: [
        R.from(
          G(y(t.state.schema).cell, o)
        )
      ]
    });
    const r = l.$anchorCell.node(-1), s = l.$anchorCell.start(-1), c = C.get(r).rectBetween(
      l.$anchorCell.pos - s,
      l.$headCell.pos - s
    );
    return n = Ue(n, c.right - c.left, c.bottom - c.top), se(t.state, t.dispatch, s, c, n), !0;
  } else if (n) {
    const r = O(t.state), s = r.start(-1);
    return se(
      t.state,
      t.dispatch,
      s,
      C.get(r.node(-1)).findCell(r.pos - s),
      n
    ), !0;
  } else
    return !1;
}
function et(t, e) {
  var o;
  if (e.ctrlKey || e.metaKey) return;
  const n = ie(t, e.target);
  let l;
  if (e.shiftKey && t.state.selection instanceof w)
    r(t.state.selection.$anchorCell, e), e.preventDefault();
  else if (e.shiftKey && n && (l = D(t.state.selection.$anchor)) != null && ((o = V(t, e)) == null ? void 0 : o.pos) != l.pos)
    r(l, e), e.preventDefault();
  else if (!n)
    return;
  function r(i, a) {
    let d = V(t, a);
    const f = N.getState(t.state) == null;
    if (!d || !ee(i, d))
      if (f) d = i;
      else return;
    const u = new w(i, d);
    if (f || !t.state.selection.eq(u)) {
      const h = t.state.tr.setSelection(u);
      f && h.setMeta(N, i.pos), t.dispatch(h);
    }
  }
  function s() {
    t.root.removeEventListener("mouseup", s), t.root.removeEventListener("dragstart", s), t.root.removeEventListener("mousemove", c), N.getState(t.state) != null && t.dispatch(t.state.tr.setMeta(N, -1));
  }
  function c(i) {
    const a = i, d = N.getState(t.state);
    let f;
    if (d != null)
      f = t.state.doc.resolve(d);
    else if (ie(t, a.target) != n && (f = V(t, e), !f))
      return s();
    f && r(f, a);
  }
  t.root.addEventListener("mouseup", s), t.root.addEventListener("dragstart", s), t.root.addEventListener("mousemove", c);
}
function Ce(t, e, o) {
  if (!(t.state.selection instanceof E)) return null;
  const { $head: n } = t.state.selection;
  for (let l = n.depth - 1; l >= 0; l--) {
    const r = n.node(l);
    if ((o < 0 ? n.index(l) : n.indexAfter(l)) != (o < 0 ? 0 : r.childCount)) return null;
    if (r.type.spec.tableRole == "cell" || r.type.spec.tableRole == "header_cell") {
      const c = n.before(l), i = e == "vert" ? o > 0 ? "down" : "up" : o > 0 ? "right" : "left";
      return t.endOfTextblock(i) ? c : null;
    }
  }
  return null;
}
function ie(t, e) {
  for (; e && e != t.dom; e = e.parentNode)
    if (e.nodeName == "TD" || e.nodeName == "TH")
      return e;
  return null;
}
function V(t, e) {
  const o = t.posAtCoords({
    left: e.clientX,
    top: e.clientY
  });
  return o && o ? D(t.state.doc.resolve(o.pos)) : null;
}
var tt = class {
  constructor(t, e) {
    this.node = t, this.defaultCellMinWidth = e, this.dom = document.createElement("div"), this.dom.className = "tableWrapper", this.table = this.dom.appendChild(document.createElement("table")), this.table.style.setProperty(
      "--default-cell-min-width",
      `${e}px`
    ), this.colgroup = this.table.appendChild(document.createElement("colgroup")), Q(t, this.colgroup, this.table, e), this.contentDOM = this.table.appendChild(document.createElement("tbody"));
  }
  update(t) {
    return t.type != this.node.type ? !1 : (this.node = t, Q(
      t,
      this.colgroup,
      this.table,
      this.defaultCellMinWidth
    ), !0);
  }
  ignoreMutation(t) {
    return t.type == "attributes" && (t.target == this.table || this.colgroup.contains(t.target));
  }
};
function Q(t, e, o, n, l, r) {
  var s;
  let c = 0, i = !0, a = e.firstChild;
  const d = t.firstChild;
  if (d) {
    for (let f = 0, u = 0; f < d.childCount; f++) {
      const { colspan: h, colwidth: p } = d.child(f).attrs;
      for (let m = 0; m < h; m++, u++) {
        const g = l == u ? r : p && p[m], b = g ? g + "px" : "";
        if (c += g || n, g || (i = !1), a)
          a.style.width != b && (a.style.width = b), a = a.nextSibling;
        else {
          const _ = document.createElement("col");
          _.style.width = b, e.appendChild(_);
        }
      }
    }
    for (; a; ) {
      const f = a.nextSibling;
      (s = a.parentNode) == null || s.removeChild(a), a = f;
    }
    i ? (o.style.width = c + "px", o.style.minWidth = "") : (o.style.width = "", o.style.minWidth = c + "px");
  }
}
var S = new Z(
  "tableColumnResizing"
);
function vt({
  handleWidth: t = 5,
  cellMinWidth: e = 25,
  defaultCellMinWidth: o = 100,
  View: n = tt,
  lastColumnResizable: l = !0
} = {}) {
  const r = new fe({
    key: S,
    state: {
      init(s, c) {
        var i, a;
        const d = (a = (i = r.spec) == null ? void 0 : i.props) == null ? void 0 : a.nodeViews, f = y(c.schema).table.name;
        return n && d && (d[f] = (u, h) => new n(u, o, h)), new ot(-1, !1);
      },
      apply(s, c) {
        return c.apply(s);
      }
    },
    props: {
      attributes: (s) => {
        const c = S.getState(s);
        return c && c.activeHandle > -1 ? { class: "resize-cursor" } : {};
      },
      handleDOMEvents: {
        mousemove: (s, c) => {
          nt(s, c, t, l);
        },
        mouseleave: (s) => {
          lt(s);
        },
        mousedown: (s, c) => {
          rt(s, c, e, o);
        }
      },
      decorations: (s) => {
        const c = S.getState(s);
        if (c && c.activeHandle > -1)
          return dt(s, c.activeHandle);
      },
      nodeViews: {}
    }
  });
  return r;
}
var ot = class I {
  constructor(e, o) {
    this.activeHandle = e, this.dragging = o;
  }
  apply(e) {
    const o = this, n = e.getMeta(S);
    if (n && n.setHandle != null)
      return new I(n.setHandle, !1);
    if (n && n.setDragging !== void 0)
      return new I(o.activeHandle, n.setDragging);
    if (o.activeHandle > -1 && e.docChanged) {
      let l = e.mapping.map(o.activeHandle, -1);
      return Y(e.doc.resolve(l)) || (l = -1), new I(l, o.dragging);
    }
    return o;
  }
};
function nt(t, e, o, n) {
  if (!t.editable) return;
  const l = S.getState(t.state);
  if (l && !l.dragging) {
    const r = it(e.target);
    let s = -1;
    if (r) {
      const { left: c, right: i } = r.getBoundingClientRect();
      e.clientX - c <= o ? s = ce(t, e, "left", o) : i - e.clientX <= o && (s = ce(t, e, "right", o));
    }
    if (s != l.activeHandle) {
      if (!n && s !== -1) {
        const c = t.state.doc.resolve(s), i = c.node(-1), a = C.get(i), d = c.start(-1);
        if (a.colCount(c.pos - d) + c.nodeAfter.attrs.colspan - 1 == a.width - 1)
          return;
      }
      be(t, s);
    }
  }
}
function lt(t) {
  if (!t.editable) return;
  const e = S.getState(t.state);
  e && e.activeHandle > -1 && !e.dragging && be(t, -1);
}
function rt(t, e, o, n) {
  var l;
  if (!t.editable) return !1;
  const r = (l = t.dom.ownerDocument.defaultView) != null ? l : window, s = S.getState(t.state);
  if (!s || s.activeHandle == -1 || s.dragging)
    return !1;
  const c = t.state.doc.nodeAt(s.activeHandle), i = st(t, s.activeHandle, c.attrs);
  t.dispatch(
    t.state.tr.setMeta(S, {
      setDragging: { startX: e.clientX, startWidth: i }
    })
  );
  function a(f) {
    r.removeEventListener("mouseup", a), r.removeEventListener("mousemove", d);
    const u = S.getState(t.state);
    u?.dragging && (ct(
      t,
      u.activeHandle,
      ae(u.dragging, f, o)
    ), t.dispatch(
      t.state.tr.setMeta(S, { setDragging: null })
    ));
  }
  function d(f) {
    if (!f.which) return a(f);
    const u = S.getState(t.state);
    if (u && u.dragging) {
      const h = ae(u.dragging, f, o);
      de(
        t,
        u.activeHandle,
        h,
        n
      );
    }
  }
  return de(
    t,
    s.activeHandle,
    i,
    n
  ), r.addEventListener("mouseup", a), r.addEventListener("mousemove", d), e.preventDefault(), !0;
}
function st(t, e, { colspan: o, colwidth: n }) {
  const l = n && n[n.length - 1];
  if (l) return l;
  const r = t.domAtPos(e);
  let c = r.node.childNodes[r.offset].offsetWidth, i = o;
  if (n)
    for (let a = 0; a < o; a++)
      n[a] && (c -= n[a], i--);
  return c / i;
}
function it(t) {
  for (; t && t.nodeName != "TD" && t.nodeName != "TH"; )
    t = t.classList && t.classList.contains("ProseMirror") ? null : t.parentNode;
  return t;
}
function ce(t, e, o, n) {
  const l = o == "right" ? -n : n, r = t.posAtCoords({
    left: e.clientX + l,
    top: e.clientY
  });
  if (!r) return -1;
  const { pos: s } = r, c = D(t.state.doc.resolve(s));
  if (!c) return -1;
  if (o == "right") return c.pos;
  const i = C.get(c.node(-1)), a = c.start(-1), d = i.map.indexOf(c.pos - a);
  return d % i.width == 0 ? -1 : a + i.map[d - 1];
}
function ae(t, e, o) {
  const n = e.clientX - t.startX;
  return Math.max(o, t.startWidth + n);
}
function be(t, e) {
  t.dispatch(
    t.state.tr.setMeta(S, { setHandle: e })
  );
}
function ct(t, e, o) {
  const n = t.state.doc.resolve(e), l = n.node(-1), r = C.get(l), s = n.start(-1), c = r.colCount(n.pos - s) + n.nodeAfter.attrs.colspan - 1, i = t.state.tr;
  for (let a = 0; a < r.height; a++) {
    const d = a * r.width + c;
    if (a && r.map[d] == r.map[d - r.width]) continue;
    const f = r.map[d], u = l.nodeAt(f).attrs, h = u.colspan == 1 ? 0 : c - r.colCount(f);
    if (u.colwidth && u.colwidth[h] == o) continue;
    const p = u.colwidth ? u.colwidth.slice() : at(u.colspan);
    p[h] = o, i.setNodeMarkup(s + f, null, { ...u, colwidth: p });
  }
  i.docChanged && t.dispatch(i);
}
function de(t, e, o, n) {
  const l = t.state.doc.resolve(e), r = l.node(-1), s = l.start(-1), c = C.get(r).colCount(l.pos - s) + l.nodeAfter.attrs.colspan - 1;
  let i = t.domAtPos(l.start(-1)).node;
  for (; i && i.nodeName != "TABLE"; )
    i = i.parentNode;
  i && Q(
    r,
    i.firstChild,
    i,
    n,
    c,
    o
  );
}
function at(t) {
  return Array(t).fill(0);
}
function dt(t, e) {
  var o;
  const n = [], l = t.doc.resolve(e), r = l.node(-1);
  if (!r)
    return X.empty;
  const s = C.get(r), c = l.start(-1), i = s.colCount(l.pos - c) + l.nodeAfter.attrs.colspan - 1;
  for (let a = 0; a < s.height; a++) {
    const d = i + a * s.width;
    if ((i == s.width - 1 || s.map[d] != s.map[d + 1]) && (a == 0 || s.map[d] != s.map[d - s.width])) {
      const f = s.map[d], u = c + f + r.nodeAt(f).nodeSize - 1, h = document.createElement("div");
      h.className = "column-resize-handle", (o = S.getState(t)) != null && o.dragging && n.push(
        q.node(
          c + f,
          c + f + r.nodeAt(f).nodeSize,
          {
            class: "column-resize-dragging"
          }
        )
      ), n.push(q.widget(u, h));
    }
  }
  return X.create(t.doc, n);
}
function kt({
  allowTableNodeSelection: t = !1
} = {}) {
  return new fe({
    key: N,
    // This piece of state is used to remember when a mouse-drag
    // cell-selection is happening, so that it can continue even as
    // transactions (which might move its anchor cell) come in.
    state: {
      init() {
        return null;
      },
      apply(e, o) {
        const n = e.getMeta(N);
        if (n != null) return n == -1 ? null : n;
        if (o == null || !e.docChanged) return o;
        const { deleted: l, pos: r } = e.mapping.mapResult(o);
        return l ? null : r;
      }
    },
    props: {
      decorations: _e,
      handleDOMEvents: {
        mousedown: et
      },
      createSelectionBetween(e) {
        return N.getState(e.state) != null ? e.state.selection : null;
      },
      handleTripleClick: Qe,
      handleKeyDown: Ge,
      handlePaste: Ze
    },
    appendTransaction(e, o, n) {
      return Le(
        n,
        Fe(n, o),
        t
      );
    }
  });
}
export {
  De as CellBookmark,
  w as CellSelection,
  ot as ResizeState,
  C as TableMap,
  tt as TableView,
  Ue as __clipCells,
  se as __insertCells,
  qe as __pastedCells,
  he as addColSpan,
  ge as addColumn,
  wt as addColumnAfter,
  gt as addColumnBefore,
  we as addRow,
  yt as addRowAfter,
  bt as addRowBefore,
  D as cellAround,
  ke as cellNear,
  Te as columnIsHeader,
  vt as columnResizing,
  S as columnResizingPluginKey,
  $ as deleteCellSelection,
  Ct as deleteColumn,
  St as deleteRow,
  Nt as deleteTable,
  Fe as fixTables,
  $e as fixTablesKey,
  zt as goToNextCell,
  Ze as handlePaste,
  ee as inSameTable,
  A as isInTable,
  At as mergeCells,
  Ee as moveCellForward,
  ue as nextCell,
  Y as pointsAtCell,
  T as removeColSpan,
  We as removeColumn,
  Oe as removeRow,
  Ie as rowIsHeader,
  x as selectedRect,
  O as selectionCell,
  xt as setCellAttr,
  Rt as splitCell,
  Ke as splitCellWithType,
  kt as tableEditing,
  N as tableEditingKey,
  y as tableNodeTypes,
  te as toggleHeader,
  Mt as toggleHeaderCell,
  Q as updateColumnsOnResize
};
