// @radix-ui/react-checkbox@1.1.3/dist/index.js
import * as n from "react";
import { useComposedRefs as M } from "../../../../../@radix-ui_react-compose-refs@1.1.1_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-compose-refs/dist/index.js";
import { createContextScope as O } from "../../../../../@radix-ui_react-context@1.1.1_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-context/dist/index.js";
import { composeEventHandlers as I } from "../../../../../@radix-ui_primitive@1.1.1/node_modules/@radix-ui/primitive/dist/index.js";
import { useControllableState as j } from "../../../../../@radix-ui_react-use-controllable-state@1.1.0_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-use-controllable-state/dist/index.js";
import { usePrevious as A } from "../../../../../@radix-ui_react-use-previous@1.1.0_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-use-previous/dist/index.js";
import { useSize as B } from "../../../../../@radix-ui_react-use-size@1.1.0_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-use-size/dist/index.js";
import { Presence as H } from "../../../../../@radix-ui_react-presence@1.1.2_@types_react-dom@18.3.1_@types_react@18.3.18_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/@radix-ui/react-presence/dist/index.js";
import { Primitive as S } from "../../../../../@radix-ui_react-primitive@2.0.1_@types_react-dom@18.3.1_@types_react@18.3.18_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/@radix-ui/react-primitive/dist/index.js";
import { jsxs as K, jsx as h } from "react/jsx-runtime";
var P = "Checkbox", [L] = O(P), [q, z] = L(P), g = n.forwardRef(
  (e, f) => {
    const {
      __scopeCheckbox: t,
      name: a,
      checked: l,
      defaultChecked: r,
      required: p,
      disabled: s,
      value: k = "on",
      onCheckedChange: b,
      form: i,
      ...v
    } = e, [d, C] = n.useState(null), x = M(f, (o) => C(o)), y = n.useRef(!1), R = d ? i || !!d.closest("form") : !0, [u = !1, E] = j({
      prop: l,
      defaultProp: r,
      onChange: b
    }), N = n.useRef(u);
    return n.useEffect(() => {
      const o = d?.form;
      if (o) {
        const m = () => E(N.current);
        return o.addEventListener("reset", m), () => o.removeEventListener("reset", m);
      }
    }, [d, E]), /* @__PURE__ */ K(q, { scope: t, state: u, disabled: s, children: [
      /* @__PURE__ */ h(
        S.button,
        {
          type: "button",
          role: "checkbox",
          "aria-checked": c(u) ? "mixed" : u,
          "aria-required": p,
          "data-state": D(u),
          "data-disabled": s ? "" : void 0,
          disabled: s,
          value: k,
          ...v,
          ref: x,
          onKeyDown: I(e.onKeyDown, (o) => {
            o.key === "Enter" && o.preventDefault();
          }),
          onClick: I(e.onClick, (o) => {
            E((m) => c(m) ? !0 : !m), R && (y.current = o.isPropagationStopped(), y.current || o.stopPropagation());
          })
        }
      ),
      R && /* @__PURE__ */ h(
        T,
        {
          control: d,
          bubbles: !y.current,
          name: a,
          value: k,
          checked: u,
          required: p,
          disabled: s,
          form: i,
          style: { transform: "translateX(-100%)" },
          defaultChecked: c(r) ? !1 : r
        }
      )
    ] });
  }
);
g.displayName = P;
var w = "CheckboxIndicator", _ = n.forwardRef(
  (e, f) => {
    const { __scopeCheckbox: t, forceMount: a, ...l } = e, r = z(w, t);
    return /* @__PURE__ */ h(H, { present: a || c(r.state) || r.state === !0, children: /* @__PURE__ */ h(
      S.span,
      {
        "data-state": D(r.state),
        "data-disabled": r.disabled ? "" : void 0,
        ...l,
        ref: f,
        style: { pointerEvents: "none", ...e.style }
      }
    ) });
  }
);
_.displayName = w;
var T = (e) => {
  const { control: f, checked: t, bubbles: a = !0, defaultChecked: l, ...r } = e, p = n.useRef(null), s = A(t), k = B(f);
  n.useEffect(() => {
    const i = p.current, v = window.HTMLInputElement.prototype, C = Object.getOwnPropertyDescriptor(v, "checked").set;
    if (s !== t && C) {
      const x = new Event("click", { bubbles: a });
      i.indeterminate = c(t), C.call(i, c(t) ? !1 : t), i.dispatchEvent(x);
    }
  }, [s, t, a]);
  const b = n.useRef(c(t) ? !1 : t);
  return /* @__PURE__ */ h(
    "input",
    {
      type: "checkbox",
      "aria-hidden": !0,
      defaultChecked: l ?? b.current,
      ...r,
      tabIndex: -1,
      ref: p,
      style: {
        ...e.style,
        ...k,
        position: "absolute",
        pointerEvents: "none",
        opacity: 0,
        margin: 0
      }
    }
  );
};
function c(e) {
  return e === "indeterminate";
}
function D(e) {
  return c(e) ? "indeterminate" : e ? "checked" : "unchecked";
}
var Z = g, $ = _;
export {
  g as Checkbox,
  _ as CheckboxIndicator,
  $ as Indicator,
  Z as Root
};
