// echarts@6.0.0/lib/coord/radar/Radar.js
import l from "./IndicatorAxis.js";
import f from "../../scale/Interval.js";
import { parsePercent as u } from "../../util/number.js";
import { each as d, map as v, isString as x, isNumber as y } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { alignScaleTicks as g } from "../axisAlignTicks.js";
import { createBoxLayoutReference as A } from "../../util/layout.js";
var D = (
  /** @class */
  (function() {
    function o(n, a, t) {
      this.dimensions = [], this._model = n, this._indicatorAxes = v(n.getIndicatorModels(), function(e, s) {
        var r = "indicator_" + s, i = new l(
          r,
          new f()
          // (indicatorModel.get('axisType') === 'log') ? new LogScale() : new IntervalScale()
        );
        return i.name = e.get("name"), i.model = e, e.axis = i, this.dimensions.push(r), i;
      }, this), this.resize(n, t);
    }
    return o.prototype.getIndicatorAxes = function() {
      return this._indicatorAxes;
    }, o.prototype.dataToPoint = function(n, a) {
      var t = this._indicatorAxes[a];
      return this.coordToPoint(t.dataToCoord(n), a);
    }, o.prototype.coordToPoint = function(n, a) {
      var t = this._indicatorAxes[a], e = t.angle, s = this.cx + n * Math.cos(e), r = this.cy - n * Math.sin(e);
      return [s, r];
    }, o.prototype.pointToData = function(n) {
      var a = n[0] - this.cx, t = n[1] - this.cy, e = Math.sqrt(a * a + t * t);
      a /= e, t /= e;
      for (var s = Math.atan2(-t, a), r = 1 / 0, i, h = -1, c = 0; c < this._indicatorAxes.length; c++) {
        var m = this._indicatorAxes[c], p = Math.abs(s - m.angle);
        p < r && (i = m, h = c, r = p);
      }
      return [h, +(i && i.coordToData(e))];
    }, o.prototype.resize = function(n, a) {
      var t = A(n, a).refContainer, e = n.get("center"), s = Math.min(t.width, t.height) / 2;
      this.cx = u(e[0], t.width) + t.x, this.cy = u(e[1], t.height) + t.y, this.startAngle = n.get("startAngle") * Math.PI / 180;
      var r = n.get("radius");
      (x(r) || y(r)) && (r = [0, r]), this.r0 = u(r[0], s), this.r = u(r[1], s), d(this._indicatorAxes, function(i, h) {
        i.setExtent(this.r0, this.r);
        var c = this.startAngle + h * Math.PI * 2 / this._indicatorAxes.length;
        c = Math.atan2(Math.sin(c), Math.cos(c)), i.angle = c;
      }, this);
    }, o.prototype.update = function(n, a) {
      var t = this._indicatorAxes, e = this._model;
      d(t, function(i) {
        i.scale.setExtent(1 / 0, -1 / 0);
      }), n.eachSeriesByType("radar", function(i, h) {
        if (!(i.get("coordinateSystem") !== "radar" || n.getComponent("radar", i.get("radarIndex")) !== e)) {
          var c = i.getData();
          d(t, function(m) {
            m.scale.unionExtentFromData(c, c.mapDimension(m.dim));
          });
        }
      }, this);
      var s = e.get("splitNumber"), r = new f();
      r.setExtent(0, s), r.setInterval(1), d(t, function(i, h) {
        g(i.scale, i.model, r);
      });
    }, o.prototype.convertToPixel = function(n, a, t) {
      return console.warn("Not implemented."), null;
    }, o.prototype.convertFromPixel = function(n, a, t) {
      return console.warn("Not implemented."), null;
    }, o.prototype.containPoint = function(n) {
      return console.warn("Not implemented."), !1;
    }, o.create = function(n, a) {
      var t = [];
      return n.eachComponent("radar", function(e) {
        var s = new o(e, n, a);
        t.push(s), e.coordinateSystem = s;
      }), n.eachSeriesByType("radar", function(e) {
        e.get("coordinateSystem") === "radar" && (e.coordinateSystem = t[e.get("radarIndex") || 0]);
      }), t;
    }, o.dimensions = [], o;
  })()
);
export {
  D as default
};
