// echarts@6.0.0/lib/coord/radar/RadarModel.js
import { __extends as T } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { map as G, defaults as h, merge as p, clone as w, isString as k, isFunction as S, mixin as A } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import I from "../axisDefault.js";
import E from "../../model/Model.js";
import { AxisModelCommonMixin as _ } from "../axisModelCommonMixin.js";
import C from "../../model/Component.js";
import f from "../../visual/tokens.js";
var r = I.value;
function l(i, a) {
  return h({
    show: a
  }, i);
}
var B = (
  /** @class */
  (function(i) {
    T(a, i);
    function a() {
      var s = i !== null && i.apply(this, arguments) || this;
      return s.type = a.type, s;
    }
    return a.prototype.optionUpdated = function() {
      var s = this.get("boundaryGap"), c = this.get("splitNumber"), d = this.get("scale"), v = this.get("axisLine"), g = this.get("axisTick"), y = this.get("axisLabel"), m = this.get("axisName"), b = this.get(["axisName", "show"]), n = this.get(["axisName", "formatter"]), L = this.get("axisNameGap"), N = this.get("triggerEvent"), M = G(this.get("indicator") || [], function(e) {
        e.max != null && e.max > 0 && !e.min ? e.min = 0 : e.min != null && e.min < 0 && !e.max && (e.max = 0);
        var u = m;
        e.color != null && (u = h({
          color: e.color
        }, m));
        var t = p(w(e), {
          boundaryGap: s,
          splitNumber: c,
          scale: d,
          axisLine: v,
          axisTick: g,
          // axisType: axisType,
          axisLabel: y,
          // Compatible with 2 and use text
          name: e.text,
          showName: b,
          nameLocation: "end",
          nameGap: L,
          // min: 0,
          nameTextStyle: u,
          triggerEvent: N
        }, !1);
        if (k(n)) {
          var x = t.name;
          t.name = n.replace("{value}", x ?? "");
        } else S(n) && (t.name = n(t.name, t));
        var o = new E(t, null, this.ecModel);
        return A(o, _.prototype), o.mainType = "radar", o.componentIndex = this.componentIndex, o;
      }, this);
      this._indicatorModels = M;
    }, a.prototype.getIndicatorModels = function() {
      return this._indicatorModels;
    }, a.type = "radar", a.defaultOption = {
      // zlevel: 0,
      z: 0,
      center: ["50%", "50%"],
      radius: "50%",
      startAngle: 90,
      axisName: {
        show: !0,
        color: f.color.axisLabel
        // formatter: null
        // textStyle: {}
      },
      boundaryGap: [0, 0],
      splitNumber: 5,
      axisNameGap: 15,
      scale: !1,
      // Polygon or circle
      shape: "polygon",
      axisLine: p({
        lineStyle: {
          color: f.color.neutral20
        }
      }, r.axisLine),
      axisLabel: l(r.axisLabel, !1),
      axisTick: l(r.axisTick, !1),
      // axisType: 'value',
      splitLine: l(r.splitLine, !0),
      splitArea: l(r.splitArea, !0),
      // {text, min, max}
      indicator: []
    }, a;
  })(C)
);
export {
  B as default
};
