// @tiptap/extension-bubble-menu@2.24.0/dist/index.js
import { isTextSelection as C, isNodeSelection as E, posToDOMRect as T, Extension as D } from "../../../../../@tiptap_core@2.24.0_@tiptap_pm@2.24.0/node_modules/@tiptap/core/dist/index.js";
import { Plugin as O, PluginKey as M } from "../../../../../prosemirror-state@1.4.3/node_modules/prosemirror-state/dist/index.js";
import B from "../../../../../tippy.js@6.3.7/node_modules/tippy.js/dist/tippy.esm.js";
class R {
  constructor({ editor: e, element: i, view: a, tippyOptions: m = {}, updateDelay: f = 250, shouldShow: c }) {
    this.preventHide = !1, this.shouldShow = ({ view: t, state: s, from: o, to: r }) => {
      const { doc: l, selection: h } = s, { empty: u } = h, p = !l.textBetween(o, r).length && C(s.selection), v = this.element.contains(document.activeElement);
      return !(!(t.hasFocus() || v) || u || p || !this.editor.isEditable);
    }, this.mousedownHandler = () => {
      this.preventHide = !0;
    }, this.dragstartHandler = () => {
      this.hide();
    }, this.focusHandler = () => {
      setTimeout(() => this.update(this.editor.view));
    }, this.blurHandler = ({ event: t }) => {
      var s;
      if (this.preventHide) {
        this.preventHide = !1;
        return;
      }
      t?.relatedTarget && (!((s = this.element.parentNode) === null || s === void 0) && s.contains(t.relatedTarget)) || t?.relatedTarget !== this.editor.view.dom && this.hide();
    }, this.tippyBlurHandler = (t) => {
      this.blurHandler({ event: t });
    }, this.handleDebouncedUpdate = (t, s) => {
      const o = !s?.selection.eq(t.state.selection), r = !s?.doc.eq(t.state.doc);
      !o && !r || (this.updateDebounceTimer && clearTimeout(this.updateDebounceTimer), this.updateDebounceTimer = window.setTimeout(() => {
        this.updateHandler(t, o, r, s);
      }, this.updateDelay));
    }, this.updateHandler = (t, s, o, r) => {
      var l, h, u;
      const { state: p, composing: v } = t, { selection: g } = p;
      if (v || !s && !o)
        return;
      this.createTooltip();
      const { ranges: b } = g, y = Math.min(...b.map((n) => n.$from.pos)), w = Math.max(...b.map((n) => n.$to.pos));
      if (!((l = this.shouldShow) === null || l === void 0 ? void 0 : l.call(this, {
        editor: this.editor,
        element: this.element,
        view: t,
        state: p,
        oldState: r,
        from: y,
        to: w
      }))) {
        this.hide();
        return;
      }
      (h = this.tippy) === null || h === void 0 || h.setProps({
        getReferenceClientRect: ((u = this.tippyOptions) === null || u === void 0 ? void 0 : u.getReferenceClientRect) || (() => {
          if (E(p.selection)) {
            let n = t.nodeDOM(y);
            if (n) {
              const H = n.dataset.nodeViewWrapper ? n : n.querySelector("[data-node-view-wrapper]");
              if (H && (n = H.firstChild), n)
                return n.getBoundingClientRect();
            }
          }
          return T(t, y, w);
        })
      }), this.show();
    }, this.editor = e, this.element = i, this.view = a, this.updateDelay = f, c && (this.shouldShow = c), this.element.addEventListener("mousedown", this.mousedownHandler, { capture: !0 }), this.view.dom.addEventListener("dragstart", this.dragstartHandler), this.editor.on("focus", this.focusHandler), this.editor.on("blur", this.blurHandler), this.tippyOptions = m, this.element.remove(), this.element.style.visibility = "visible";
  }
  createTooltip() {
    const { element: e } = this.editor.options, i = !!e.parentElement;
    this.tippy || !i || (this.tippy = B(e, {
      duration: 0,
      getReferenceClientRect: null,
      content: this.element,
      interactive: !0,
      trigger: "manual",
      placement: "top",
      hideOnClick: "toggle",
      ...this.tippyOptions
    }), this.tippy.popper.firstChild && this.tippy.popper.firstChild.addEventListener("blur", this.tippyBlurHandler));
  }
  update(e, i) {
    const { state: a } = e, m = a.selection.from !== a.selection.to;
    if (this.updateDelay > 0 && m) {
      this.handleDebouncedUpdate(e, i);
      return;
    }
    const f = !i?.selection.eq(e.state.selection), c = !i?.doc.eq(e.state.doc);
    this.updateHandler(e, f, c, i);
  }
  show() {
    var e;
    (e = this.tippy) === null || e === void 0 || e.show();
  }
  hide() {
    var e;
    (e = this.tippy) === null || e === void 0 || e.hide();
  }
  destroy() {
    var e, i;
    !((e = this.tippy) === null || e === void 0) && e.popper.firstChild && this.tippy.popper.firstChild.removeEventListener("blur", this.tippyBlurHandler), (i = this.tippy) === null || i === void 0 || i.destroy(), this.element.removeEventListener("mousedown", this.mousedownHandler, { capture: !0 }), this.view.dom.removeEventListener("dragstart", this.dragstartHandler), this.editor.off("focus", this.focusHandler), this.editor.off("blur", this.blurHandler);
  }
}
const _ = (d) => new O({
  key: typeof d.pluginKey == "string" ? new M(d.pluginKey) : d.pluginKey,
  view: (e) => new R({ view: e, ...d })
});
D.create({
  name: "bubbleMenu",
  addOptions() {
    return {
      element: null,
      tippyOptions: {},
      pluginKey: "bubbleMenu",
      updateDelay: void 0,
      shouldShow: null
    };
  },
  addProseMirrorPlugins() {
    return this.options.element ? [
      _({
        pluginKey: this.options.pluginKey,
        editor: this.editor,
        element: this.options.element,
        tippyOptions: this.options.tippyOptions,
        updateDelay: this.options.updateDelay,
        shouldShow: this.options.shouldShow
      })
    ] : [];
  }
});
export {
  _ as BubbleMenuPlugin,
  R as BubbleMenuView
};
