// zrender@6.0.0/lib/graphic/shape/BezierCurve.js
import { __extends as S } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import q from "../Path.js";
import { normalize as B } from "../../core/vector.js";
import { quadraticSubdivide as a, cubicSubdivide as d, cubicDerivativeAt as m, cubicAt as b, quadraticDerivativeAt as z, quadraticAt as A } from "../../core/curve.js";
var e = [], D = /* @__PURE__ */ (function() {
  function r() {
    this.x1 = 0, this.y1 = 0, this.x2 = 0, this.y2 = 0, this.cpx1 = 0, this.cpy1 = 0, this.percent = 1;
  }
  return r;
})();
function C(r, u, t) {
  var i = r.cpx2, f = r.cpy2;
  return i != null || f != null ? [
    (t ? m : b)(r.x1, r.cpx1, r.cpx2, r.x2, u),
    (t ? m : b)(r.y1, r.cpy1, r.cpy2, r.y2, u)
  ] : [
    (t ? z : A)(r.x1, r.cpx1, r.x2, u),
    (t ? z : A)(r.y1, r.cpy1, r.y2, u)
  ];
}
var P = (function(r) {
  S(u, r);
  function u(t) {
    return r.call(this, t) || this;
  }
  return u.prototype.getDefaultStyle = function() {
    return {
      stroke: "#000",
      fill: null
    };
  }, u.prototype.getDefaultShape = function() {
    return new D();
  }, u.prototype.buildPath = function(t, i) {
    var f = i.x1, x = i.y1, o = i.x2, n = i.y2, v = i.cpx1, y = i.cpy1, p = i.cpx2, l = i.cpy2, c = i.percent;
    c !== 0 && (t.moveTo(f, x), p == null || l == null ? (c < 1 && (a(f, v, o, c, e), v = e[1], o = e[2], a(x, y, n, c, e), y = e[1], n = e[2]), t.quadraticCurveTo(v, y, o, n)) : (c < 1 && (d(f, v, p, o, c, e), v = e[1], p = e[2], o = e[3], d(x, y, l, n, c, e), y = e[1], l = e[2], n = e[3]), t.bezierCurveTo(v, y, p, l, o, n)));
  }, u.prototype.pointAt = function(t) {
    return C(this.shape, t, !1);
  }, u.prototype.tangentAt = function(t) {
    var i = C(this.shape, t, !0);
    return B(i, i);
  }, u;
})(q);
P.prototype.type = "bezier-curve";
export {
  D as BezierCurveShape,
  P as default
};
