// zrender@6.0.0/lib/graphic/shape/Line.js
import { __extends as a } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import p from "../Path.js";
import { subPixelOptimizeLine as s } from "../helper/subPixelOptimize.js";
var x = {}, c = /* @__PURE__ */ (function() {
  function u() {
    this.x1 = 0, this.y1 = 0, this.x2 = 0, this.y2 = 0, this.percent = 1;
  }
  return u;
})(), v = (function(u) {
  a(r, u);
  function r(e) {
    return u.call(this, e) || this;
  }
  return r.prototype.getDefaultStyle = function() {
    return {
      stroke: "#000",
      fill: null
    };
  }, r.prototype.getDefaultShape = function() {
    return new c();
  }, r.prototype.buildPath = function(e, t) {
    var f, l, n, o;
    if (this.subPixelOptimize) {
      var y = s(x, t, this.style);
      f = y.x1, l = y.y1, n = y.x2, o = y.y2;
    } else
      f = t.x1, l = t.y1, n = t.x2, o = t.y2;
    var i = t.percent;
    i !== 0 && (e.moveTo(f, l), i < 1 && (n = f * (1 - i) + n * i, o = l * (1 - i) + o * i), e.lineTo(n, o));
  }, r.prototype.pointAt = function(e) {
    var t = this.shape;
    return [
      t.x1 * (1 - e) + t.x2 * e,
      t.y1 * (1 - e) + t.y2 * e
    ];
  }, r;
})(p);
v.prototype.type = "line";
export {
  c as LineShape,
  v as default
};
