// zrender@6.0.0/lib/graphic/shape/Rect.js
import { __extends as c } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import l from "../Path.js";
import { buildPath as s } from "../helper/roundRect.js";
import { subPixelOptimizeRect as y } from "../helper/subPixelOptimize.js";
var a = /* @__PURE__ */ (function() {
  function e() {
    this.x = 0, this.y = 0, this.width = 0, this.height = 0;
  }
  return e;
})(), v = {}, x = (function(e) {
  c(i, e);
  function i(h) {
    return e.call(this, h) || this;
  }
  return i.prototype.getDefaultShape = function() {
    return new a();
  }, i.prototype.buildPath = function(h, t) {
    var n, o, u, f;
    if (this.subPixelOptimize) {
      var r = y(v, t, this.style);
      n = r.x, o = r.y, u = r.width, f = r.height, r.r = t.r, t = r;
    } else
      n = t.x, o = t.y, u = t.width, f = t.height;
    t.r ? s(h, t) : h.rect(n, o, u, f);
  }, i.prototype.isZeroArea = function() {
    return !this.shape.width || !this.shape.height;
  }, i;
})(l);
x.prototype.type = "rect";
export {
  a as RectShape,
  x as default
};
