// echarts@6.0.0/lib/util/animation.js
var o = (
  /** @class */
  (function() {
    function r() {
      this._storage = [], this._elExistsMap = {};
    }
    return r.prototype.add = function(t, i, n, a, s) {
      return this._elExistsMap[t.id] ? !1 : (this._elExistsMap[t.id] = !0, this._storage.push({
        el: t,
        target: i,
        duration: n,
        delay: a,
        easing: s
      }), !0);
    }, r.prototype.finished = function(t) {
      return this._finishedCallback = t, this;
    }, r.prototype.start = function() {
      for (var t = this, i = this._storage.length, n = function() {
        i--, i <= 0 && (t._storage.length = 0, t._elExistsMap = {}, t._finishedCallback && t._finishedCallback());
      }, a = 0, s = this._storage.length; a < s; a++) {
        var e = this._storage[a];
        e.el.animateTo(e.target, {
          duration: e.duration,
          delay: e.delay,
          easing: e.easing,
          setToFinal: !0,
          done: n,
          aborted: n
        });
      }
      return this;
    }, r;
  })()
);
function h() {
  return new o();
}
export {
  h as createWrap
};
