// echarts@6.0.0/lib/util/clazz.js
import { __extends as d } from "../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { each as l, inherits as v, extend as C, assert as c, isFunction as E } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
var h = ".", u = "___EC__COMPONENT__CONTAINER___", p = "___EC__EXTENDED_CLASS___";
function o(s) {
  var t = {
    main: "",
    sub: ""
  };
  if (s) {
    var a = s.split(h);
    t.main = a[0] || "", t.sub = a[1] || "";
  }
  return t;
}
function m(s) {
  c(/^[a-zA-Z0-9_]+([.][a-zA-Z0-9_]+)?$/.test(s), 'componentType "' + s + '" illegal');
}
function D(s) {
  return !!(s && s[p]);
}
function M(s, t) {
  s.$constructor = s, s.extend = function(a) {
    process.env.NODE_ENV !== "production" && l(t, function(r) {
      a[r] || console.warn("Method `" + r + "` should be implemented" + (a.type ? " in " + a.type : "") + ".");
    });
    var e = this, n;
    return y(e) ? n = /** @class */
    (function(r) {
      d(i, r);
      function i() {
        return r.apply(this, arguments) || this;
      }
      return i;
    })(e) : (n = function() {
      (a.$constructor || e).apply(this, arguments);
    }, v(n, this)), C(n.prototype, a), n[p] = !0, n.extend = this.extend, n.superCall = x, n.superApply = b, n.superClass = e, n;
  };
}
function y(s) {
  return E(s) && /^class\s/.test(Function.prototype.toString.call(s));
}
function g(s, t) {
  s.extend = t.extend;
}
var N = Math.round(Math.random() * 10);
function O(s) {
  var t = ["__\0is_clz", N++].join("_");
  s.prototype[t] = !0, process.env.NODE_ENV !== "production" && c(!s.isInstance, 'The method "is" can not be defined.'), s.isInstance = function(a) {
    return !!(a && a[t]);
  };
}
function x(s, t) {
  for (var a = [], e = 2; e < arguments.length; e++)
    a[e - 2] = arguments[e];
  return this.superClass.prototype[t].apply(s, a);
}
function b(s, t, a) {
  return this.superClass.prototype[t].apply(s, a);
}
function w(s) {
  var t = {};
  s.registerClass = function(e) {
    var n = e.type || e.prototype.type;
    if (n) {
      m(n), e.prototype.type = n;
      var r = o(n);
      if (!r.sub)
        process.env.NODE_ENV !== "production" && t[r.main] && console.warn(r.main + " exists."), t[r.main] = e;
      else if (r.sub !== u) {
        var i = a(r);
        i[r.sub] = e;
      }
    }
    return e;
  }, s.getClass = function(e, n, r) {
    var i = t[e];
    if (i && i[u] && (i = n ? i[n] : null), r && !i)
      throw new Error(n ? "Component " + e + "." + (n || "") + " is used but not imported." : e + ".type should be specified.");
    return i;
  }, s.getClassesByMainType = function(e) {
    var n = o(e), r = [], i = t[n.main];
    return i && i[u] ? l(i, function(f, _) {
      _ !== u && r.push(f);
    }) : r.push(i), r;
  }, s.hasClass = function(e) {
    var n = o(e);
    return !!t[n.main];
  }, s.getAllClassMainTypes = function() {
    var e = [];
    return l(t, function(n, r) {
      e.push(r);
    }), e;
  }, s.hasSubTypes = function(e) {
    var n = o(e), r = t[n.main];
    return r && r[u];
  };
  function a(e) {
    var n = t[e.main];
    return (!n || !n[u]) && (n = t[e.main] = {}, n[u] = !0), n;
  }
}
export {
  O as enableClassCheck,
  M as enableClassExtend,
  w as enableClassManagement,
  D as isExtendedClass,
  g as mountExtend,
  o as parseClassType
};
