// echarts@6.0.0/lib/util/component.js
import { each as h, indexOf as g, merge as E } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { parseClassType as I } from "./clazz.js";
import { makePrintable as w } from "./log.js";
var A = Math.round(Math.random() * 10);
function G(i) {
  return [i || "", A++].join("_");
}
function M(i) {
  var u = {};
  i.registerSubTypeDefaulter = function(v, s) {
    var f = I(v);
    u[f.main] = s;
  }, i.determineSubType = function(v, s) {
    var f = s.type;
    if (!f) {
      var r = I(v).main;
      i.hasSubTypes(v) && u[r] && (f = u[r](s));
    }
    return f;
  };
}
function _(i, u) {
  i.topologicalTravel = function(r, e, c, a) {
    if (!r.length)
      return;
    var o = v(e), D = o.graph, p = o.noEntryList, n = {};
    for (h(r, function(t) {
      n[t] = !0;
    }); p.length; ) {
      var l = p.pop(), y = D[l], b = !!n[l];
      b && (c.call(a, l, y.originalDeps.slice()), delete n[l]), h(y.successor, b ? S : d);
    }
    h(n, function() {
      var t = "";
      throw process.env.NODE_ENV !== "production" && (t = w("Circular dependency may exists: ", n, r, e)), new Error(t);
    });
    function d(t) {
      D[t].entryCount--, D[t].entryCount === 0 && p.push(t);
    }
    function S(t) {
      n[t] = !0, d(t);
    }
  };
  function v(r) {
    var e = {}, c = [];
    return h(r, function(a) {
      var o = s(e, a), D = o.originalDeps = u(a), p = f(D, r);
      o.entryCount = p.length, o.entryCount === 0 && c.push(a), h(p, function(n) {
        g(o.predecessor, n) < 0 && o.predecessor.push(n);
        var l = s(e, n);
        g(l.successor, n) < 0 && l.successor.push(a);
      });
    }), {
      graph: e,
      noEntryList: c
    };
  }
  function s(r, e) {
    return r[e] || (r[e] = {
      predecessor: [],
      successor: []
    }), r[e];
  }
  function f(r, e) {
    var c = [];
    return h(r, function(a) {
      g(e, a) >= 0 && c.push(a);
    }), c;
  }
}
function j(i, u) {
  return E(E({}, i, !0), u, !0);
}
export {
  M as enableSubTypeDefaulter,
  _ as enableTopologicalTravel,
  G as getUID,
  j as inheritDefaultOption
};
