// echarts@6.0.0/lib/util/conditionalExpression.js
import { isObject as b, isArrayLike as P, isArray as _, map as y, keys as A, hasOwn as V, isString as C, isRegExp as L, isNumber as D } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { makePrintable as l, throwError as i } from "./log.js";
import { getRawValueParser as w, createFilterComparator as G } from "../data/helper/dataValueHelper.js";
var N = {
  value: "eq",
  // PENDING: not good for literal semantic?
  "<": "lt",
  "<=": "lte",
  ">": "gt",
  ">=": "gte",
  "=": "eq",
  "!=": "ne",
  "<>": "ne"
  // Might be misleading for sake of the difference between '==' and '===',
  // so don't support them.
  // '==': 'eq',
  // '===': 'seq',
  // '!==': 'sne'
  // PENDING: Whether support some common alias "ge", "le", "neq"?
  // ge: 'gte',
  // le: 'lte',
  // neq: 'ne',
}, M = (
  /** @class */
  (function() {
    function n(r) {
      var t = this._condVal = C(r) ? new RegExp(r) : L(r) ? r : null;
      if (t == null) {
        var e = "";
        process.env.NODE_ENV !== "production" && (e = l("Illegal regexp", r, "in")), i(e);
      }
    }
    return n.prototype.evaluate = function(r) {
      var t = typeof r;
      return C(t) ? this._condVal.test(r) : D(t) ? this._condVal.test(r + "") : !1;
    }, n;
  })()
), R = (
  /** @class */
  (function() {
    function n() {
    }
    return n.prototype.evaluate = function() {
      return this.value;
    }, n;
  })()
), S = (
  /** @class */
  (function() {
    function n() {
    }
    return n.prototype.evaluate = function() {
      for (var r = this.children, t = 0; t < r.length; t++)
        if (!r[t].evaluate())
          return !1;
      return !0;
    }, n;
  })()
), j = (
  /** @class */
  (function() {
    function n() {
    }
    return n.prototype.evaluate = function() {
      for (var r = this.children, t = 0; t < r.length; t++)
        if (r[t].evaluate())
          return !0;
      return !1;
    }, n;
  })()
), O = (
  /** @class */
  (function() {
    function n() {
    }
    return n.prototype.evaluate = function() {
      return !this.child.evaluate();
    }, n;
  })()
), q = (
  /** @class */
  (function() {
    function n() {
    }
    return n.prototype.evaluate = function() {
      for (var r = !!this.valueParser, t = this.getValue, e = t(this.valueGetterParam), a = r ? this.valueParser(e) : null, o = 0; o < this.subCondList.length; o++)
        if (!this.subCondList[o].evaluate(r ? a : e))
          return !1;
      return !0;
    }, n;
  })()
);
function d(n, r) {
  if (n === !0 || n === !1) {
    var t = new R();
    return t.value = n, t;
  }
  var e = "";
  return m(n) || (process.env.NODE_ENV !== "production" && (e = l("Illegal config. Expect a plain object but actually", n)), i(e)), n.and ? I("and", n, r) : n.or ? I("or", n, r) : n.not ? F(n, r) : K(n, r);
}
function I(n, r, t) {
  var e = r[n], a = "";
  process.env.NODE_ENV !== "production" && (a = l('"and"/"or" condition should only be `' + n + ": [...]` and must not be empty array.", "Illegal condition:", r)), _(e) || i(a), e.length || i(a);
  var o = n === "and" ? new S() : new j();
  return o.children = y(e, function(v) {
    return d(v, t);
  }), o.children.length || i(a), o;
}
function F(n, r) {
  var t = n.not, e = "";
  process.env.NODE_ENV !== "production" && (e = l('"not" condition should only be `not: {}`.', "Illegal condition:", n)), m(t) || i(e);
  var a = new O();
  return a.child = d(t, r), a.child || i(e), a;
}
function K(n, r) {
  for (var t = "", e = r.prepareGetValue(n), a = [], o = A(n), v = n.parser, c = v ? w(v) : null, f = 0; f < o.length; f++) {
    var u = o[f];
    if (!(u === "parser" || r.valueGetterAttrMap.get(u))) {
      var h = V(N, u) ? N[u] : u, g = n[u], p = c ? c(g) : g, E = G(h, p) || h === "reg" && new M(p);
      E || (process.env.NODE_ENV !== "production" && (t = l('Illegal relational operation: "' + u + '" in condition:', n)), i(t)), a.push(E);
    }
  }
  a.length || (process.env.NODE_ENV !== "production" && (t = l("Relational condition must have at least one operator.", "Illegal condition:", n)), i(t));
  var s = new q();
  return s.valueGetterParam = e, s.valueParser = c, s.getValue = r.getValue, s.subCondList = a, s;
}
function m(n) {
  return b(n) && !P(n);
}
var T = (
  /** @class */
  (function() {
    function n(r, t) {
      this._cond = d(r, t);
    }
    return n.prototype.evaluate = function() {
      return this._cond.evaluate();
    }, n;
  })()
);
function H(n, r) {
  return new T(n, r);
}
export {
  H as parseConditionalExpression
};
