// echarts@6.0.0/lib/util/decal.js
import fr from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/WeakMap.js";
import gr from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/LRU.js";
import { defaults as vr, isArray as ur, isString as M, isNumber as w, map as H } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { getLeastCommonMultiple as F } from "./number.js";
import { createSymbol as cr } from "./symbol.js";
import { brushSingle as sr } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/canvas/graphic.js";
import { platformApi as mr } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/platform.js";
var D = new fr(), G = new gr(100), Z = ["symbol", "symbolSize", "symbolKeepAspect", "color", "backgroundColor", "dashArrayX", "dashArrayY", "maxTileWidth", "maxTileHeight"];
function Mr(r, t) {
  if (r === "none")
    return null;
  var e = t.getDevicePixelRatio(), a = t.getZr(), o = a.painter.type === "svg";
  r.dirty && D.delete(r);
  var K = D.get(r);
  if (K)
    return K;
  var i = vr(r, {
    symbol: "rect",
    symbolSize: 1,
    symbolKeepAspect: !0,
    color: "rgba(0, 0, 0, 0.2)",
    backgroundColor: null,
    dashArrayX: 5,
    dashArrayY: 5,
    rotation: 0,
    maxTileWidth: 512,
    maxTileHeight: 512
  });
  i.backgroundColor === "none" && (i.backgroundColor = null);
  var g = {
    repeat: "repeat"
  };
  return _(g), g.rotation = i.rotation, g.scaleX = g.scaleY = o ? 1 : 1 / e, D.set(r, g), r.dirty = !1, g;
  function _(p) {
    for (var P = [e], C = !0, T = 0; T < Z.length; ++T) {
      var y = i[Z[T]];
      if (y != null && !ur(y) && !M(y) && !w(y) && typeof y != "boolean") {
        C = !1;
        break;
      }
      P.push(y);
    }
    var L;
    if (C) {
      L = P.join(",") + (o ? "-svg" : "");
      var x = G.get(L);
      x && (o ? p.svgElement = x : p.image = x);
    }
    var v = J(i.dashArrayX), c = pr(i.dashArrayY), d = q(i.symbol), X = yr(v), R = Q(c), f = !o && mr.createCanvas(), W = o && {
      tag: "g",
      attrs: {},
      key: "dcl",
      children: []
    }, b = $(), k;
    f && (f.width = b.width * e, f.height = b.height * e, k = f.getContext("2d")), O(), C && G.put(L, f || W), p.image = f, p.svgElement = W, p.svgWidth = b.width, p.svgHeight = b.height;
    function $() {
      for (var u = 1, n = 0, s = X.length; n < s; ++n)
        u = F(u, X[n]);
      for (var h = 1, n = 0, s = d.length; n < s; ++n)
        h = F(h, d[n].length);
      u *= h;
      var S = R * X.length * d.length;
      if (process.env.NODE_ENV !== "production") {
        var l = function(A) {
          console.warn("Calculated decal size is greater than " + A + " due to decal option settings so " + A + " is used for the decal size. Please consider changing the decal option to make a smaller decal or set " + A + " to be larger to avoid incontinuity.");
        };
        u > i.maxTileWidth && l("maxTileWidth"), S > i.maxTileHeight && l("maxTileHeight");
      }
      return {
        width: Math.max(1, Math.min(u, i.maxTileWidth)),
        height: Math.max(1, Math.min(S, i.maxTileHeight))
      };
    }
    function O() {
      k && (k.clearRect(0, 0, f.width, f.height), i.backgroundColor && (k.fillStyle = i.backgroundColor, k.fillRect(0, 0, f.width, f.height)));
      for (var u = 0, n = 0; n < c.length; ++n)
        u += c[n];
      if (u <= 0)
        return;
      for (var s = -R, h = 0, S = 0, l = 0; s < b.height; ) {
        if (h % 2 === 0) {
          for (var A = S / 2 % d.length, Y = 0, m = 0, B = 0; Y < b.width * 2; ) {
            for (var E = 0, n = 0; n < v[l].length; ++n)
              E += v[l][n];
            if (E <= 0)
              break;
            if (m % 2 === 0) {
              var V = (1 - i.symbolSize) * 0.5, j = Y + v[l][m] * V, I = s + c[h] * V, rr = v[l][m] * i.symbolSize, er = c[h] * i.symbolSize, tr = B / 2 % d[A].length;
              ar(j, I, rr, er, d[A][tr]);
            }
            Y += v[l][m], ++B, ++m, m === v[l].length && (m = 0);
          }
          ++l, l === v.length && (l = 0);
        }
        s += c[h], ++S, ++h, h === c.length && (h = 0);
      }
      function ar(ir, nr, or, lr, hr) {
        var z = o ? 1 : e, N = cr(hr, ir * z, nr * z, or * z, lr * z, i.color, i.symbolKeepAspect);
        if (o) {
          var U = a.painter.renderOneToVNode(N);
          U && W.children.push(U);
        } else
          sr(k, N);
      }
    }
  }
}
function q(r) {
  if (!r || r.length === 0)
    return [["rect"]];
  if (M(r))
    return [[r]];
  for (var t = !0, e = 0; e < r.length; ++e)
    if (!M(r[e])) {
      t = !1;
      break;
    }
  if (t)
    return q([r]);
  for (var a = [], e = 0; e < r.length; ++e)
    M(r[e]) ? a.push([r[e]]) : a.push(r[e]);
  return a;
}
function J(r) {
  if (!r || r.length === 0)
    return [[0, 0]];
  if (w(r)) {
    var t = Math.ceil(r);
    return [[t, t]];
  }
  for (var e = !0, a = 0; a < r.length; ++a)
    if (!w(r[a])) {
      e = !1;
      break;
    }
  if (e)
    return J([r]);
  for (var o = [], a = 0; a < r.length; ++a)
    if (w(r[a])) {
      var t = Math.ceil(r[a]);
      o.push([t, t]);
    } else {
      var t = H(r[a], function(g) {
        return Math.ceil(g);
      });
      t.length % 2 === 1 ? o.push(t.concat(t)) : o.push(t);
    }
  return o;
}
function pr(r) {
  if (!r || typeof r == "object" && r.length === 0)
    return [0, 0];
  if (w(r)) {
    var t = Math.ceil(r);
    return [t, t];
  }
  var e = H(r, function(a) {
    return Math.ceil(a);
  });
  return r.length % 2 ? e.concat(e) : e;
}
function yr(r) {
  return H(r, function(t) {
    return Q(t);
  });
}
function Q(r) {
  for (var t = 0, e = 0; e < r.length; ++e)
    t += r[e];
  return r.length % 2 === 1 ? t * 2 : t;
}
export {
  Mr as createOrUpdatePatternFromDecal
};
