// echarts@6.0.0/lib/util/ECEventProcessor.js
import { isString as m, each as y } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { parseClassType as E } from "./clazz.js";
var I = (
  /** @class */
  (function() {
    function p() {
    }
    return p.prototype.normalizeQuery = function(d) {
      var t = {}, o = {}, f = {};
      if (m(d)) {
        var v = E(d);
        t.mainType = v.main || null, t.subType = v.sub || null;
      } else {
        var a = ["Index", "Name", "Id"], l = {
          name: 1,
          dataIndex: 1,
          dataType: 1
        };
        y(d, function(n, r) {
          for (var e = !1, u = 0; u < a.length; u++) {
            var c = a[u], i = r.lastIndexOf(c);
            if (i > 0 && i === r.length - c.length) {
              var s = r.slice(0, i);
              s !== "data" && (t.mainType = s, t[c.toLowerCase()] = n, e = !0);
            }
          }
          l.hasOwnProperty(r) && (o[r] = n, e = !0), e || (f[r] = n);
        });
      }
      return {
        cptQuery: t,
        dataQuery: o,
        otherQuery: f
      };
    }, p.prototype.filter = function(d, t) {
      var o = this.eventInfo;
      if (!o)
        return !0;
      var f = o.targetEl, v = o.packedEvent, a = o.model, l = o.view;
      if (!a || !l)
        return !0;
      var n = t.cptQuery, r = t.dataQuery;
      return e(n, a, "mainType") && e(n, a, "subType") && e(n, a, "index", "componentIndex") && e(n, a, "name") && e(n, a, "id") && e(r, v, "name") && e(r, v, "dataIndex") && e(r, v, "dataType") && (!l.filterForExposedEvent || l.filterForExposedEvent(d, t.otherQuery, f, v));
      function e(u, c, i, s) {
        return u[i] == null || c[s || i] === u[i];
      }
    }, p.prototype.afterTrigger = function() {
      this.eventInfo = null;
    }, p;
  })()
);
export {
  I as ECEventProcessor
};
