// echarts@6.0.0/lib/util/format.js
import { isString as l, isObject as h, isArray as b, normalizeCssArray as y, each as v, isStringSafe as s, isNumber as w, trim as A } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { encodeHTML as m } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/dom.js";
import { isNumeric as T, parseDate as R, numericToNumber as N } from "./number.js";
import { format as S } from "./time.js";
function C(r) {
  if (!T(r))
    return l(r) ? r : "-";
  var e = (r + "").split(".");
  return e[0].replace(/(\d{1,3})(?=(?:\d{3})+(?!\d))/g, "$1,") + (e.length > 1 ? "." + e[1] : "");
}
function U(r, e) {
  return r = (r || "").toLowerCase().replace(/-(.)/g, function(n, o) {
    return o.toUpperCase();
  }), e && r && (r = r.charAt(0).toUpperCase() + r.slice(1)), r;
}
var D = y;
function V(r, e, n) {
  var o = "{yyyy}-{MM}-{dd} {HH}:{mm}:{ss}";
  function i(d) {
    return d && A(d) ? d : "-";
  }
  function a(d) {
    return !!(d != null && !isNaN(d) && isFinite(d));
  }
  var t = e === "time", c = r instanceof Date;
  if (t || c) {
    var f = t ? R(r) : r;
    if (isNaN(+f)) {
      if (c)
        return "-";
    } else return S(f, o, n);
  }
  if (e === "ordinal")
    return s(r) ? i(r) : w(r) && a(r) ? r + "" : "-";
  var p = N(r);
  return a(p) ? C(p) : s(r) ? i(r) : typeof r == "boolean" ? r + "" : "-";
}
var g = ["a", "b", "c", "d", "e", "f", "g"], u = function(r, e) {
  return "{" + r + (e ?? "") + "}";
};
function $(r, e, n) {
  b(e) || (e = [e]);
  var o = e.length;
  if (!o)
    return "";
  for (var i = e[0].$vars || [], a = 0; a < i.length; a++) {
    var t = g[a];
    r = r.replace(u(t), u(t, 0));
  }
  for (var c = 0; c < o; c++)
    for (var f = 0; f < i.length; f++) {
      var p = e[c][i[f]];
      r = r.replace(u(g[f], c), n ? m(p) : p);
    }
  return r;
}
function z(r, e, n) {
  return v(e, function(o, i) {
    r = r.replace("{" + i + "}", o);
  }), r;
}
function H(r, e) {
  var n = l(r) ? {
    color: r,
    extraCssText: e
  } : r || {}, o = n.color, i = n.type;
  e = n.extraCssText;
  var a = n.renderMode || "html";
  if (!o)
    return "";
  if (a === "html")
    return i === "subItem" ? '<span style="display:inline-block;vertical-align:middle;margin-right:8px;margin-left:3px;border-radius:4px;width:4px;height:4px;background-color:' + m(o) + ";" + (e || "") + '"></span>' : '<span style="display:inline-block;margin-right:4px;border-radius:10px;width:10px;height:10px;background-color:' + m(o) + ";" + (e || "") + '"></span>';
  var t = n.markerId || "markerX";
  return {
    renderMode: a,
    content: "{" + t + "|}  ",
    style: i === "subItem" ? {
      width: 4,
      height: 4,
      borderRadius: 2,
      backgroundColor: o
    } : {
      width: 10,
      height: 10,
      borderRadius: 5,
      backgroundColor: o
    }
  };
}
function F(r, e) {
  return e = e || "transparent", l(r) ? r : h(r) && r.colorStops && (r.colorStops[0] || {}).color || e;
}
function L(r, e) {
  if (e === "_blank" || e === "blank") {
    var n = window.open();
    n.opener = null, n.location.href = r;
  } else
    window.open(r, e);
}
export {
  C as addCommas,
  F as convertToColorString,
  m as encodeHTML,
  $ as formatTpl,
  z as formatTplSimple,
  H as getTooltipMarker,
  V as makeValueReadable,
  D as normalizeCssArray,
  U as toCamelCase,
  L as windowOpen
};
