// echarts@6.0.0/lib/util/graphic.js
import { createFromString as A, mergePath as B, extendFromString as N } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/tool/path.js";
import { invert as S, identity as D, mul as $, copy as F, create as _ } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/matrix.js";
import { applyTransform as k } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/vector.js";
import H from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Path.js";
import W from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/Transformable.js";
import E from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Image.js";
import { default as xr } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
import { default as Tr } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Text.js";
import j from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Circle.js";
import U from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Ellipse.js";
import X from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Sector.js";
import q from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Ring.js";
import V from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Polygon.js";
import Y from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Polyline.js";
import J from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Rect.js";
import K from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Line.js";
import Q from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/BezierCurve.js";
import Z from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Arc.js";
import { default as wr } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/CompoundPath.js";
import { default as Rr } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/LinearGradient.js";
import { default as Cr } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/RadialGradient.js";
import ee from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/BoundingRect.js";
import { default as Ir } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/OrientedBoundingRect.js";
import { default as zr } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/Point.js";
import { default as Ar } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/IncrementalDisplayable.js";
import { subPixelOptimize as re, subPixelOptimizeLine as te, subPixelOptimizeRect as ne } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/helper/subPixelOptimize.js";
import { isArray as ie, isNumber as oe, assert as ae, each as fe, hasOwn as ue, keys as pe, defaults as I, extend as me, isArrayLike as se, map as ve, clone as he, isString as ce } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { getECData as G } from "./innerStore.js";
import { updateProps as le } from "../animation/basicTransition.js";
import { initProps as Nr, isElementRemoved as Sr, removeElement as Dr, removeElementWithFadeOut as $r } from "../animation/basicTransition.js";
import { mathMax as m, mathAbs as h, mathMin as l } from "./number.js";
var x = {}, Ue = ["x", "y"], Xe = ["width", "height"];
function qe(e) {
  return H.extend(e);
}
var ge = N;
function Ve(e, r) {
  return ge(e, r);
}
function s(e, r) {
  x[e] = r;
}
function Ye(e) {
  if (x.hasOwnProperty(e))
    return x[e];
}
function de(e, r, t, n) {
  var i = A(e, r);
  return t && (n === "center" && (t = z(t, i.getBoundingRect())), ye(i, t)), i;
}
function Je(e, r, t) {
  var n = new E({
    style: {
      image: e,
      x: r.x,
      y: r.y,
      width: r.width,
      height: r.height
    },
    onload: function(i) {
      if (t === "center") {
        var o = {
          width: i.width,
          height: i.height
        };
        n.setStyle(z(r, o));
      }
    }
  });
  return n;
}
function z(e, r) {
  var t = r.width / r.height, n = e.height * t, i;
  n <= e.width ? i = e.height : (n = e.width, i = n / t);
  var o = e.x + e.width / 2, a = e.y + e.height / 2;
  return {
    x: o - n / 2,
    y: a - i / 2,
    width: n,
    height: i
  };
}
var Ke = B;
function ye(e, r) {
  if (e.applyTransform) {
    var t = e.getBoundingRect(), n = t.calculateTransform(r);
    e.applyTransform(n);
  }
}
function Qe(e, r) {
  return te(e, e, {
    lineWidth: r
  }), e;
}
function Ze(e, r) {
  return ne(e, e, r), e;
}
var er = re;
function rr(e, r) {
  for (var t = D([]); e && e !== r; )
    $(t, e.getLocalTransform(), t), e = e.parent;
  return t;
}
function xe(e, r, t) {
  return r && !se(r) && (r = W.getLocalTransform(r)), t && (r = S([], r)), k([], e, r);
}
function tr(e, r, t) {
  var n = r[4] === 0 || r[5] === 0 || r[0] === 0 ? 1 : h(2 * r[4] / r[0]), i = r[4] === 0 || r[5] === 0 || r[2] === 0 ? 1 : h(2 * r[4] / r[2]), o = [e === "left" ? -n : e === "right" ? n : 0, e === "top" ? -i : e === "bottom" ? i : 0];
  return o = xe(o, r, t), h(o[0]) > h(o[1]) ? o[0] > 0 ? "right" : "left" : o[1] > 0 ? "bottom" : "top";
}
function L(e) {
  return !e.isGroup;
}
function Pe(e) {
  return e.shape != null;
}
function nr(e, r, t) {
  if (!e || !r)
    return;
  function n(a) {
    var f = {};
    return a.traverse(function(p) {
      L(p) && p.anid && (f[p.anid] = p);
    }), f;
  }
  function i(a) {
    var f = {
      x: a.x,
      y: a.y,
      rotation: a.rotation
    };
    return Pe(a) && (f.shape = he(a.shape)), f;
  }
  var o = n(e);
  r.traverse(function(a) {
    if (L(a) && a.anid) {
      var f = o[a.anid];
      if (f) {
        var p = i(a);
        a.attr(i(f)), le(a, p, t, G(a).dataIndex);
      }
    }
  });
}
function ir(e, r) {
  return ve(e, function(t) {
    var n = t[0];
    n = m(n, r.x), n = l(n, r.x + r.width);
    var i = t[1];
    return i = m(i, r.y), i = l(i, r.y + r.height), [n, i];
  });
}
function or(e, r) {
  var t = m(e.x, r.x), n = l(e.x + e.width, r.x + r.width), i = m(e.y, r.y), o = l(e.y + e.height, r.y + r.height);
  if (n >= t && o >= i)
    return {
      x: t,
      y: i,
      width: n - t,
      height: o - i
    };
}
function ar(e, r, t) {
  var n = me({
    rectHover: !0
  }, r), i = n.style = {
    strokeNoScale: !0
  };
  if (t = t || {
    x: -1,
    y: -1,
    width: 2,
    height: 2
  }, e)
    return e.indexOf("image://") === 0 ? (i.image = e.slice(8), I(i, t), new E(n)) : de(e.replace("path://", ""), n, t, "center");
}
function fr(e, r, t, n, i) {
  for (var o = 0, a = i[i.length - 1]; o < i.length; o++) {
    var f = i[o];
    if (Te(e, r, t, n, f[0], f[1], a[0], a[1]))
      return !0;
    a = f;
  }
}
function Te(e, r, t, n, i, o, a, f) {
  var p = t - e, v = n - r, c = a - i, T = f - o, d = y(c, T, p, v);
  if (Oe(d))
    return !1;
  var O = e - i, w = r - o, b = y(O, w, p, v) / d;
  if (b < 0 || b > 1)
    return !1;
  var R = y(O, w, c, T) / d;
  return !(R < 0 || R > 1);
}
function y(e, r, t, n) {
  return e * n - t * r;
}
function Oe(e) {
  return e <= 1e-6 && e >= -1e-6;
}
function ur(e, r, t, n, i) {
  return r == null || (oe(r) ? u[0] = u[1] = u[2] = u[3] = r : (process.env.NODE_ENV !== "production" && ae(r.length === 4), u[0] = r[0], u[1] = r[1], u[2] = r[2], u[3] = r[3]), n && (u[0] = m(0, u[0]), u[1] = m(0, u[1]), u[2] = m(0, u[2]), u[3] = m(0, u[3])), t && (u[0] = -u[0], u[1] = -u[1], u[2] = -u[2], u[3] = -u[3]), C(e, u, "x", "width", 3, 1, i && i[0] || 0), C(e, u, "y", "height", 0, 2, i && i[1] || 0)), e;
}
var u = [0, 0, 0, 0];
function C(e, r, t, n, i, o, a) {
  var f = r[o] + r[i], p = e[n];
  e[n] += f, a = m(0, l(a, p)), e[n] < a ? (e[n] = a, e[t] += r[i] >= 0 ? -r[i] : r[o] >= 0 ? p + r[o] : h(f) > 1e-8 ? (p - a) * r[i] / f : 0) : e[t] -= r[i];
}
function pr(e) {
  var r = e.itemTooltipOption, t = e.componentModel, n = e.itemName, i = ce(r) ? {
    formatter: r
  } : r, o = t.mainType, a = t.componentIndex, f = {
    componentType: o,
    name: n,
    $vars: ["name"]
  };
  f[o + "Index"] = a;
  var p = e.formatterParamsExtra;
  p && fe(pe(p), function(c) {
    ue(f, c) || (f[c] = p[c], f.$vars.push(c));
  });
  var v = G(e.el);
  v.componentMainType = o, v.componentIndex = a, v.tooltipConfig = {
    name: n,
    option: I({
      content: n,
      encodeHTMLContent: !0,
      formatterParams: f
    }, i)
  };
}
function P(e, r) {
  var t;
  e.isGroup && (t = r(e)), t || e.traverse(r);
}
function mr(e, r) {
  if (e)
    if (ie(e))
      for (var t = 0; t < e.length; t++)
        P(e[t], r);
    else
      P(e, r);
}
function sr(e) {
  return !e || h(e[1]) < g && h(e[2]) < g || h(e[0]) < g && h(e[3]) < g;
}
var g = 1e-5;
function vr(e, r) {
  return e ? ee.copy(e, r) : r.clone();
}
function hr(e, r) {
  return r ? F(e || _(), r) : void 0;
}
function cr(e) {
  return {
    z: e.get("z") || 0,
    zlevel: e.get("zlevel") || 0
  };
}
function lr(e) {
  var r = -1 / 0, t = 1 / 0;
  P(e, function(o) {
    n(o), n(o.getTextContent()), n(o.getTextGuideLine());
  });
  function n(o) {
    if (!(!o || o.isGroup)) {
      var a = o.currentStates;
      if (a.length)
        for (var f = 0; f < a.length; f++)
          i(o.states[a[f]]);
      i(o);
    }
  }
  function i(o) {
    if (o) {
      var a = o.z2;
      a > r && (r = a), a < t && (t = a);
    }
  }
  return t > r && (t = r = 0), {
    min: t,
    max: r
  };
}
function gr(e, r, t) {
  M(e, r, t, -1 / 0);
}
function M(e, r, t, n) {
  if (e.ignoreModelZ)
    return n;
  var i = e.getTextContent(), o = e.getTextGuideLine(), a = e.isGroup;
  if (a)
    for (var f = e.childrenRef(), p = 0; p < f.length; p++)
      n = m(M(f[p], r, t, n), n);
  else
    e.z = r, e.zlevel = t, n = m(e.z2 || 0, n);
  if (i && (i.z = r, i.zlevel = t, isFinite(n) && (i.z2 = n + 2)), o) {
    var v = e.textGuideLineConfig;
    o.z = r, o.zlevel = t, isFinite(n) && (o.z2 = n + (v && v.showAbove ? 1 : -1));
  }
  return n;
}
s("circle", j);
s("ellipse", U);
s("sector", X);
s("ring", q);
s("polygon", V);
s("polyline", Y);
s("rect", J);
s("line", K);
s("bezierCurve", Q);
s("arc", Z);
export {
  Z as Arc,
  Q as BezierCurve,
  ee as BoundingRect,
  j as Circle,
  wr as CompoundPath,
  U as Ellipse,
  xr as Group,
  E as Image,
  Ar as IncrementalDisplayable,
  K as Line,
  Rr as LinearGradient,
  Ir as OrientedBoundingRect,
  H as Path,
  zr as Point,
  V as Polygon,
  Y as Polyline,
  Cr as RadialGradient,
  J as Rect,
  q as Ring,
  X as Sector,
  Tr as Text,
  Xe as WH,
  Ue as XY,
  xe as applyTransform,
  lr as calcZ2Range,
  ir as clipPointsByRect,
  or as clipRectByRect,
  ar as createIcon,
  vr as ensureCopyRect,
  hr as ensureCopyTransform,
  ur as expandOrShrinkRect,
  Ve as extendPath,
  qe as extendShape,
  Ye as getShapeClass,
  rr as getTransform,
  nr as groupTransition,
  Nr as initProps,
  sr as isBoundingRectAxisAligned,
  Sr as isElementRemoved,
  Te as lineLineIntersect,
  fr as linePolygonIntersect,
  Je as makeImage,
  de as makePath,
  Ke as mergePath,
  s as registerShape,
  Dr as removeElement,
  $r as removeElementWithFadeOut,
  ye as resizePath,
  cr as retrieveZInfo,
  pr as setTooltipConfig,
  er as subPixelOptimize,
  Qe as subPixelOptimizeLine,
  Ze as subPixelOptimizeRect,
  tr as transformDirection,
  mr as traverseElements,
  gr as traverseUpdateZ,
  le as updateProps
};
