// echarts@6.0.0/lib/util/jitter.js
import y from "../coord/cartesian/Axis2D.js";
import { makeInner as f } from "./model.js";
function j(r, i) {
  var a = r.coordinateSystem, t = a && a.type, e = a && a.getBaseAxis && a.getBaseAxis(), l = e && e.scale && e.scale.type, n = t === "cartesian2d" && l === "ordinal" || t === "single", v = i.model.get("jitter") > 0;
  return n && v;
}
var b = f();
function V(r, i, a, t) {
  if (r instanceof y) {
    var e = r.scale.type;
    if (e !== "category" && e !== "ordinal")
      return a;
  }
  var l = r.model, n = l.get("jitter"), v = l.get("jitterOverlap"), s = l.get("jitterMargin") || 0, c = r.scale.type === "ordinal" ? r.getBandWidth() : null;
  return n > 0 ? v ? M(a, n, c, t) : J(r, i, a, t, n, s) : a;
}
function M(r, i, a, t) {
  if (a === null)
    return r + (Math.random() - 0.5) * i;
  var e = a - t * 2, l = Math.min(Math.max(0, i), e);
  return r + (Math.random() - 0.5) * l;
}
function J(r, i, a, t, e, l) {
  var n = b(r);
  n.items || (n.items = []);
  var v = n.items, s = o(v, i, a, t, e, l, 1), c = o(v, i, a, t, e, l, -1), p = Math.abs(s - a) < Math.abs(c - a) ? s : c, m = r.scale.type === "ordinal" ? r.getBandWidth() : null, u = Math.abs(p - a);
  return u > e / 2 || m && u > m / 2 - t ? M(a, e, m, t) : (v.push({
    fixedCoord: i,
    floatCoord: p,
    r: t
  }), p);
}
function o(r, i, a, t, e, l, n) {
  for (var v = a, s = 0; s < r.length; s++) {
    var c = r[s], p = i - c.fixedCoord, m = v - c.floatCoord, u = p * p + m * m, h = t + c.r + l;
    if (u < h * h) {
      var g = c.floatCoord + Math.sqrt(h * h - p * p) * n;
      if (Math.abs(g - a) > e / 2)
        return Number.MAX_VALUE;
      if (n === 1 && g > v || n === -1 && g < v) {
        v = g, s = -1;
        continue;
      }
    }
  }
  return v;
}
export {
  V as fixJitter,
  j as needFixJitter
};
