// echarts@6.0.0/lib/util/layout.js
import { isObject as p, isArray as m, each as z, hasOwn as A, curry as R, defaults as T } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import P from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/BoundingRect.js";
import { parsePercent as d } from "./number.js";
import { normalizeCssArray as M } from "./format.js";
import { error as V } from "./log.js";
import { getCoordForBoxCoordSys as k, BoxCoordinateSystemCoordFrom as F } from "../core/CoordinateSystem.js";
var S = z, E = ["left", "right", "top", "bottom", "width", "height"], w = [["width", "left", "right"], ["height", "top", "bottom"]];
function B(r, e, h, o, a) {
  var t = 0, v = 0;
  o == null && (o = 1 / 0), a == null && (a = 1 / 0);
  var n = 0;
  e.eachChild(function(u, i) {
    var f = u.getBoundingRect(), l = e.childAt(i + 1), g = l && l.getBoundingRect(), c, y;
    if (r === "horizontal") {
      var b = f.width + (g ? -g.x + f.x : 0);
      c = t + b, c > o || u.newline ? (t = 0, c = b, v += n + h, n = f.height) : n = Math.max(n, f.height);
    } else {
      var x = f.height + (g ? -g.y + f.y : 0);
      y = v + x, y > a || u.newline ? (t += n + h, v = 0, y = x, n = f.width) : n = Math.max(n, f.width);
    }
    u.newline || (u.x = t, u.y = v, u.markRedraw(), r === "horizontal" ? t = c + h : v = y + h);
  });
}
var Q = B;
R(B, "vertical");
R(B, "horizontal");
function U(r, e) {
  return {
    left: r.getShallow("left", e),
    top: r.getShallow("top", e),
    right: r.getShallow("right", e),
    bottom: r.getShallow("bottom", e),
    width: r.getShallow("width", e),
    height: r.getShallow("height", e)
  };
}
function H(r, e) {
  var h = W(r, e, {
    enableLayoutOnlyByCenter: !0
  }), o = r.getBoxLayoutParams(), a, t;
  if (h.type === C.point)
    t = h.refPoint, a = L(o, {
      width: e.getWidth(),
      height: e.getHeight()
    });
  else {
    var v = r.get("center"), n = m(v) ? v : [v, v];
    a = L(o, h.refContainer), t = h.boxCoordFrom === F.coord2 ? h.refPoint : [d(n[0], a.width) + a.x, d(n[1], a.height) + a.y];
  }
  return {
    viewRect: a,
    center: t
  };
}
function X(r, e) {
  var h = H(r, e), o = h.viewRect, a = h.center, t = r.get("radius");
  m(t) || (t = [0, t]);
  var v = d(o.width, e.getWidth()), n = d(o.height, e.getHeight()), u = Math.min(v, n), i = d(t[0], u / 2), f = d(t[1], u / 2);
  return {
    cx: a[0],
    cy: a[1],
    r0: i,
    r: f,
    viewRect: o
  };
}
function L(r, e, h) {
  h = M(h || 0);
  var o = e.width, a = e.height, t = d(r.left, o), v = d(r.top, a), n = d(r.right, o), u = d(r.bottom, a), i = d(r.width, o), f = d(r.height, a), l = h[2] + h[0], g = h[1] + h[3], c = r.aspect;
  switch (isNaN(i) && (i = o - n - g - t), isNaN(f) && (f = a - u - l - v), c != null && (isNaN(i) && isNaN(f) && (c > o / a ? i = o * 0.8 : f = a * 0.8), isNaN(i) && (i = c * f), isNaN(f) && (f = i / c)), isNaN(t) && (t = o - n - i - g), isNaN(v) && (v = a - u - f - l), r.left || r.right) {
    case "center":
      t = o / 2 - i / 2 - h[3];
      break;
    case "right":
      t = o - i - g;
      break;
  }
  switch (r.top || r.bottom) {
    case "middle":
    case "center":
      v = a / 2 - f / 2 - h[0];
      break;
    case "bottom":
      v = a - f - l;
      break;
  }
  t = t || 0, v = v || 0, isNaN(i) && (i = o - g - t - (n || 0)), isNaN(f) && (f = a - l - v - (u || 0));
  var y = new P((e.x || 0) + t + h[3], (e.y || 0) + v + h[0], i, f);
  return y.margin = h, y;
}
function Y(r, e, h) {
  var o = r.getShallow("preserveAspect", !0);
  if (!o)
    return e;
  var a = e.width / e.height;
  if (Math.abs(Math.atan(h) - Math.atan(a)) < 1e-9)
    return e;
  var t = r.getShallow("preserveAspectAlign", !0), v = r.getShallow("preserveAspectVerticalAlign", !0), n = {
    width: e.width,
    height: e.height
  }, u = o === "cover";
  return a > h && !u || a < h && u ? (n.width = e.height * h, t === "left" ? n.left = 0 : t === "right" ? n.right = 0 : n.left = "center") : (n.height = e.width / h, v === "top" ? n.top = 0 : v === "bottom" ? n.bottom = 0 : n.top = "middle"), L(n, e);
}
var C = {
  rect: 1,
  point: 2
};
function W(r, e, h) {
  var o, a, t, v = r.boxCoordinateSystem, n;
  if (v) {
    var u = k(r), i = u.coord, f = u.from;
    if (v.dataToLayout) {
      t = C.rect, n = f;
      var l = v.dataToLayout(i);
      o = l.contentRect || l.rect;
    } else h && h.enableLayoutOnlyByCenter && v.dataToPoint ? (t = C.point, n = f, a = v.dataToPoint(i)) : process.env.NODE_ENV !== "production" && V(r.type + "[" + r.componentIndex + "]" + (" layout based on " + v.type + " is not supported."));
  }
  return t == null && (t = C.rect), t === C.rect && (o || (o = {
    x: 0,
    y: 0,
    width: e.getWidth(),
    height: e.getHeight()
  }), a = [o.x + o.width / 2, o.y + o.height / 2]), {
    type: t,
    refContainer: o,
    refPoint: a,
    boxCoordFrom: n
  };
}
function Z(r, e, h, o, a, t) {
  var v = !a || !a.hv || a.hv[0], n = !a || !a.hv || a.hv[1], u = a && a.boundingMode || "all";
  if (t = t || r, t.x = r.x, t.y = r.y, !v && !n)
    return !1;
  var i;
  if (u === "raw")
    i = r.type === "group" ? new P(0, 0, +e.width || 0, +e.height || 0) : r.getBoundingRect();
  else if (i = r.getBoundingRect(), r.needLocalTransform()) {
    var f = r.getLocalTransform();
    i = i.clone(), i.applyTransform(f);
  }
  var l = L(T({
    width: i.width,
    height: i.height
  }, e), h, o), g = v ? l.x - i.x : 0, c = n ? l.y - i.y : 0;
  return u === "raw" ? (t.x = g, t.y = c) : (t.x += g, t.y += c), t === r && r.markRedraw(), !0;
}
function _(r, e) {
  return r[w[e][0]] != null || r[w[e][1]] != null && r[w[e][2]] != null;
}
function O(r) {
  var e = r.layoutMode || r.constructor.layoutMode;
  return p(e) ? e : e ? {
    type: e
  } : null;
}
function I(r, e, h) {
  var o = h && h.ignoreSize;
  !m(o) && (o = [o, o]);
  var a = v(w[0], 0), t = v(w[1], 1);
  u(w[0], r, a), u(w[1], r, t);
  function v(i, f) {
    var l = {}, g = 0, c = {}, y = 0, b = 2;
    if (S(i, function(s) {
      c[s] = r[s];
    }), S(i, function(s) {
      A(e, s) && (l[s] = c[s] = e[s]), n(l, s) && g++, n(c, s) && y++;
    }), o[f])
      return n(e, i[1]) ? c[i[2]] = null : n(e, i[2]) && (c[i[1]] = null), c;
    if (y === b || !g)
      return c;
    if (g >= b)
      return l;
    for (var x = 0; x < i.length; x++) {
      var N = i[x];
      if (!A(l, N) && A(r, N)) {
        l[N] = r[N];
        break;
      }
    }
    return l;
  }
  function n(i, f) {
    return i[f] != null && i[f] !== "auto";
  }
  function u(i, f, l) {
    S(i, function(g) {
      f[g] = l[g];
    });
  }
}
function rr(r) {
  return D({}, r);
}
function D(r, e) {
  return e && r && S(E, function(h) {
    A(e, h) && (r[h] = e[h]);
  }), r;
}
export {
  C as BoxLayoutReferenceType,
  w as HV_NAMES,
  E as LOCATION_PARAMS,
  Y as applyPreserveAspect,
  Q as box,
  D as copyLayoutParams,
  W as createBoxLayoutReference,
  O as fetchLayoutMode,
  U as getBoxLayoutParams,
  X as getCircleLayout,
  rr as getLayoutParams,
  L as getLayoutRect,
  I as mergeLayoutParam,
  Z as positionElement,
  _ as sizeCalculable
};
