// echarts@6.0.0/lib/util/log.js
import { map as d, isString as p, eqNaN as E, isFunction as N, isRegExp as g } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
var l = "[ECharts] ", s = {}, y = typeof console < "u" && console.warn && console.log;
function t(n, r, o) {
  if (y) {
    if (o) {
      if (s[r])
        return;
      s[r] = !0;
    }
    console[n](l + r);
  }
}
function D(n, r) {
  t("log", n, r);
}
function O(n, r) {
  t("warn", n, r);
}
function m(n, r) {
  t("error", n, r);
}
function h(n) {
  process.env.NODE_ENV !== "production" && t("warn", "DEPRECATED: " + n, !0);
}
function I(n, r, o) {
  process.env.NODE_ENV !== "production" && h((o ? "[" + o + "]" : "") + (n + " is deprecated; use " + r + " instead."));
}
function S() {
  for (var n = [], r = 0; r < arguments.length; r++)
    n[r] = arguments[r];
  var o = "";
  if (process.env.NODE_ENV !== "production") {
    var i = function(e) {
      return e === void 0 ? "undefined" : e === 1 / 0 ? "Infinity" : e === -1 / 0 ? "-Infinity" : E(e) ? "NaN" : e instanceof Date ? "Date(" + e.toISOString() + ")" : N(e) ? "function () { ... }" : g(e) ? e + "" : null;
    };
    o = d(n, function(e) {
      if (p(e))
        return e;
      var u = i(e);
      if (u != null)
        return u;
      if (typeof JSON < "u" && JSON.stringify)
        try {
          return JSON.stringify(e, function(a, f) {
            var c = i(f);
            return c ?? f;
          });
        } catch {
          return "?";
        }
      else
        return "?";
    }).join(" ");
  }
  return o;
}
function R(n) {
  throw new Error(n);
}
export {
  h as deprecateLog,
  I as deprecateReplaceLog,
  m as error,
  D as log,
  S as makePrintable,
  R as throwError,
  O as warn
};
