// echarts@6.0.0/lib/util/model.js
import { createHashMap as M, each as v, isObject as g, isString as w, isNumber as D, isStringSafe as k, indexOf as O, assert as q, map as T, isArray as E } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import X from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/env.js";
import { isNumeric as Y, getRandomIdBase as G, round as m, getPrecision as y } from "./number.js";
import { error as p, warn as K } from "./log.js";
function A(n, e, r) {
  return (e - n) * r + n;
}
var P = "series\0", B = "\0_ec_\0";
function U(n) {
  return n instanceof Array ? n : n == null ? [] : [n];
}
function ln(n, e, r) {
  if (n) {
    n[e] = n[e] || {}, n.emphasis = n.emphasis || {}, n.emphasis[e] = n.emphasis[e] || {};
    for (var i = 0, u = r.length; i < u; i++) {
      var a = r[i];
      !n.emphasis[e].hasOwnProperty(a) && n[e].hasOwnProperty(a) && (n.emphasis[e][a] = n[e][a]);
    }
  }
}
var tn = ["fontStyle", "fontWeight", "fontSize", "fontFamily", "rich", "tag", "color", "textBorderColor", "textBorderWidth", "width", "height", "lineHeight", "align", "verticalAlign", "baseline", "shadowColor", "shadowBlur", "shadowOffsetX", "shadowOffsetY", "textShadowColor", "textShadowBlur", "textShadowOffsetX", "textShadowOffsetY", "backgroundColor", "borderColor", "borderWidth", "borderRadius", "padding"];
function fn(n) {
  return g(n) && !E(n) && !(n instanceof Date) ? n.value : n;
}
function on(n) {
  return g(n) && !(n instanceof Array);
}
function sn(n, e, r) {
  var i = r === "normalMerge", u = r === "replaceMerge", a = r === "replaceAll";
  n = n || [], e = (e || []).slice();
  var t = M();
  v(e, function(l, o) {
    if (!g(l)) {
      e[o] = null;
      return;
    }
    process.env.NODE_ENV !== "production" && (l.id != null && !R(l.id) && S(l.id), l.name != null && !R(l.name) && S(l.name));
  });
  var f = W(n, t, r);
  return (i || u) && z(f, n, t, e), i && C(f, e), i || u ? H(f, e, u) : a && $(f, e), J(f), f;
}
function W(n, e, r) {
  var i = [];
  if (r === "replaceAll")
    return i;
  for (var u = 0; u < n.length; u++) {
    var a = n[u];
    a && a.id != null && e.set(a.id, u), i.push({
      existing: r === "replaceMerge" || _(a) ? null : a,
      newOption: null,
      keyInfo: null,
      brandNew: null
    });
  }
  return i;
}
function z(n, e, r, i) {
  v(i, function(u, a) {
    if (!(!u || u.id == null)) {
      var t = I(u.id), f = r.get(t);
      if (f != null) {
        var l = n[f];
        q(!l.newOption, 'Duplicated option on id "' + t + '".'), l.newOption = u, l.existing = e[f], i[a] = null;
      }
    }
  });
}
function C(n, e) {
  v(e, function(r, i) {
    if (!(!r || r.name == null))
      for (var u = 0; u < n.length; u++) {
        var a = n[u].existing;
        if (!n[u].newOption && a && (a.id == null || r.id == null) && !_(r) && !_(a) && L("name", a, r)) {
          n[u].newOption = r, e[i] = null;
          return;
        }
      }
  });
}
function H(n, e, r) {
  v(e, function(i) {
    if (i) {
      for (
        var u, a = 0;
        // Be `!resultItem` only when `nextIdx >= result.length`.
        (u = n[a]) && (u.newOption || _(u.existing) || // In mode "replaceMerge", here no not-mapped-non-internal-existing.
        u.existing && i.id != null && !L("id", i, u.existing));
      )
        a++;
      u ? (u.newOption = i, u.brandNew = r) : n.push({
        newOption: i,
        brandNew: r,
        existing: null,
        keyInfo: null
      }), a++;
    }
  });
}
function $(n, e) {
  v(e, function(r) {
    n.push({
      newOption: r,
      brandNew: !0,
      existing: null,
      keyInfo: null
    });
  });
}
function J(n) {
  var e = M();
  v(n, function(r) {
    var i = r.existing;
    i && e.set(i.id, r);
  }), v(n, function(r) {
    var i = r.newOption;
    q(!i || i.id == null || !e.get(i.id) || e.get(i.id) === r, "id duplicates: " + (i && i.id)), i && i.id != null && e.set(i.id, r), !r.keyInfo && (r.keyInfo = {});
  }), v(n, function(r, i) {
    var u = r.existing, a = r.newOption, t = r.keyInfo;
    if (g(a)) {
      if (t.name = a.name != null ? I(a.name) : u ? u.name : P + i, u)
        t.id = I(u.id);
      else if (a.id != null)
        t.id = I(a.id);
      else {
        var f = 0;
        do
          t.id = "\0" + t.name + "\0" + f++;
        while (e.get(t.id));
      }
      e.set(t.id, r);
    }
  });
}
function L(n, e, r) {
  var i = N(e[n], null), u = N(r[n], null);
  return i != null && u != null && i === u;
}
function I(n) {
  if (process.env.NODE_ENV !== "production" && n == null)
    throw new Error();
  return N(n, "");
}
function N(n, e) {
  return n == null ? e : w(n) ? n : D(n) || k(n) ? n + "" : e;
}
function S(n) {
  process.env.NODE_ENV !== "production" && K("`" + n + "` is invalid id or name. Must be a string or number.");
}
function R(n) {
  return k(n) || Y(n);
}
function dn(n) {
  var e = n.name;
  return !!(e && e.indexOf(P));
}
function _(n) {
  return n && n.id != null && I(n.id).indexOf(B) === 0;
}
function vn(n) {
  return B + n;
}
function cn(n, e, r) {
  v(n, function(i) {
    var u = i.newOption;
    g(u) && (i.keyInfo.mainType = e, i.keyInfo.subType = Q(e, u, i.existing, r));
  });
}
function Q(n, e, r, i) {
  var u = e.type ? e.type : r ? r.subType : i.determineSubType(n, e);
  return u;
}
function hn(n, e) {
  var r = {}, i = {};
  return u(n || [], r), u(e || [], i, r), [a(r), a(i)];
  function u(t, f, l) {
    for (var o = 0, d = t.length; o < d; o++) {
      var s = N(t[o].seriesId, null);
      if (s == null)
        return;
      for (var c = U(t[o].dataIndex), h = l && l[s], x = 0, F = c.length; x < F; x++) {
        var b = c[x];
        h && h[b] ? h[b] = null : (f[s] || (f[s] = {}))[b] = 1;
      }
    }
  }
  function a(t, f) {
    var l = [];
    for (var o in t)
      if (t.hasOwnProperty(o) && t[o] != null)
        if (f)
          l.push(+o);
        else {
          var d = a(t[o], !0);
          d.length && l.push({
            seriesId: o,
            dataIndex: d
          });
        }
    return l;
  }
}
function xn(n, e) {
  if (e.dataIndexInside != null)
    return e.dataIndexInside;
  if (e.dataIndex != null)
    return E(e.dataIndex) ? T(e.dataIndex, function(r) {
      return n.indexOfRawIndex(r);
    }) : n.indexOfRawIndex(e.dataIndex);
  if (e.name != null)
    return E(e.name) ? T(e.name, function(r) {
      return n.indexOfName(r);
    }) : n.indexOfName(e.name);
}
function In() {
  var n = "__ec_inner_" + Z++;
  return function(e) {
    return e[n] || (e[n] = {});
  };
}
var Z = G();
function gn(n, e, r) {
  var i = j(e, r), u = i.mainTypeSpecified, a = i.queryOptionMap, t = i.others, f = t, l = r ? r.defaultMainType : null;
  return !u && l && a.set(l, {}), a.each(function(o, d) {
    var s = nn(n, d, o, {
      useDefault: l === d,
      enableAll: r && r.enableAll != null ? r.enableAll : !0,
      enableNone: r && r.enableNone != null ? r.enableNone : !0
    });
    f[d + "Models"] = s.models, f[d + "Model"] = s.models[0];
  }), f;
}
function j(n, e) {
  var r;
  if (w(n)) {
    var i = {};
    i[n + "Index"] = 0, r = i;
  } else
    r = n;
  var u = M(), a = {}, t = !1;
  return v(r, function(f, l) {
    if (l === "dataIndex" || l === "dataIndexInside") {
      a[l] = f;
      return;
    }
    var o = l.match(/^(\w+)(Index|Id|Name)$/) || [], d = o[1], s = (o[2] || "").toLowerCase();
    if (!(!d || !s || e && e.includeMainTypes && O(e.includeMainTypes, d) < 0)) {
      t = t || !!d;
      var c = u.get(d) || u.set(d, {});
      c[s] = f;
    }
  }), {
    mainTypeSpecified: t,
    queryOptionMap: u,
    others: a
  };
}
var V = {
  useDefault: !0,
  enableAll: !1,
  enableNone: !1
}, yn = {
  useDefault: !1,
  enableAll: !0,
  enableNone: !0
};
function nn(n, e, r, i) {
  i = i || V;
  var u = r.index, a = r.id, t = r.name, f = {
    models: null,
    specified: u != null || a != null || t != null
  };
  if (!f.specified) {
    var l = void 0;
    return f.models = i.useDefault && (l = n.getComponent(e)) ? [l] : [], f;
  }
  if (u === "none" || u === !1) {
    if (i.enableNone)
      return f.models = [], f;
    process.env.NODE_ENV !== "production" && p('`"none"` or `false` is not a valid value on index option.'), u = -1;
  }
  return u === "all" && (i.enableAll ? u = a = t = null : (process.env.NODE_ENV !== "production" && p('`"all"` is not a valid value on index option.'), u = -1)), f.models = n.queryComponents({
    mainType: e,
    index: u,
    id: a,
    name: t
  }), f;
}
function Nn(n, e, r) {
  n.setAttribute ? n.setAttribute(e, r) : n[e] = r;
}
function _n(n, e) {
  return n.getAttribute ? n.getAttribute(e) : n[e];
}
function Mn(n) {
  return n === "auto" ? X.domSupported ? "html" : "richText" : n || "html";
}
function bn(n, e) {
  var r = M(), i = [];
  return v(n, function(u) {
    var a = e(u);
    (r.get(a) || (i.push(a), r.set(a, []))).push(u);
  }), {
    keys: i,
    buckets: r
  };
}
function En(n, e, r, i, u) {
  var a = e == null || e === "auto";
  if (i == null)
    return i;
  if (D(i)) {
    var t = A(r || 0, i, u);
    return m(t, a ? Math.max(y(r || 0), y(i)) : e);
  } else {
    if (w(i))
      return u < 1 ? r : i;
    for (var f = [], l = r, o = i, d = Math.max(l ? l.length : 0, o.length), s = 0; s < d; ++s) {
      var c = n.getDimensionInfo(s);
      if (c && c.type === "ordinal")
        f[s] = (u < 1 && l ? l : o)[s];
      else {
        var h = l && l[s] ? l[s] : 0, x = o[s], t = A(h, x, u);
        f[s] = m(t, a ? Math.max(y(h), y(x)) : e);
      }
    }
    return f;
  }
}
var wn = (
  /** @class */
  (function() {
    function n() {
    }
    return n.prototype.reset = function(e, r, i, u) {
      return this._list = e, this._step = u = u || 1, this._idx = r, this._end = i ?? (u > 0 ? e.length : 0), this.item = null, this.key = NaN, this;
    }, n.prototype.next = function() {
      return (this._step > 0 ? this._idx < this._end : this._idx >= this._end) ? (this.item = this._list[this._idx], this.key = this._idx = this._idx + this._step, !0) : !1;
    }, n;
  })()
);
function Tn(n) {
  n.option = n.parentModel = n.ecModel = null;
}
export {
  wn as ListIterator,
  yn as MULTIPLE_REFERRING,
  V as SINGLE_REFERRING,
  tn as TEXT_STYLE_OPTIONS,
  Tn as clearTmpModel,
  hn as compressBatches,
  N as convertOptionIdName,
  ln as defaultEmphasis,
  _n as getAttribute,
  fn as getDataItemValue,
  Mn as getTooltipRenderMode,
  bn as groupData,
  En as interpolateRawValues,
  _ as isComponentIdInternal,
  on as isDataItemOption,
  dn as isNameSpecified,
  In as makeInner,
  vn as makeInternalComponentId,
  sn as mappingToExists,
  U as normalizeToArray,
  gn as parseFinder,
  j as preParseFinder,
  xn as queryDataIndex,
  nn as queryReferringComponents,
  Nn as setAttribute,
  cn as setComponentTypeToKeyInfo
};
