// echarts@6.0.0/lib/util/number.js
import { isString as h, reduce as P, map as N } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
var p = 1e-4, x = 20;
function S(r) {
  return r.replace(/^\s+|\s+$/g, "");
}
var l = Math.min, A = Math.max, R = Math.abs;
function _(r, e, n, u) {
  var t = e[0], i = e[1], a = n[0], d = n[1], f = i - t, o = d - a;
  if (f === 0)
    return o === 0 ? a : (a + d) / 2;
  if (u)
    if (f > 0) {
      if (r <= t)
        return a;
      if (r >= i)
        return d;
    } else {
      if (r >= t)
        return a;
      if (r <= i)
        return d;
    }
  else {
    if (r === t)
      return a;
    if (r === i)
      return d;
  }
  return (r - t) / f * o + a;
}
var y = g;
function g(r, e, n) {
  switch (r) {
    case "center":
    case "middle":
      r = "50%";
      break;
    case "left":
    case "top":
      r = "0%";
      break;
    case "right":
    case "bottom":
      r = "100%";
      break;
  }
  return E(r, e, n);
}
function E(r, e, n) {
  return h(r) ? S(r).match(/%$/) ? parseFloat(r) / 100 * e + (n || 0) : parseFloat(r) : r == null ? NaN : +r;
}
function F(r, e, n) {
  return e == null && (e = 10), e = Math.min(Math.max(0, e), x), r = (+r).toFixed(e), n ? r : +r;
}
function b(r) {
  return r.sort(function(e, n) {
    return e - n;
  }), r;
}
function m(r) {
  if (r = +r, isNaN(r))
    return 0;
  if (r > 1e-14) {
    for (var e = 1, n = 0; n < 15; n++, e *= 10)
      if (Math.round(r * e) / e === r)
        return n;
  }
  return O(r);
}
function O(r) {
  var e = r.toString().toLowerCase(), n = e.indexOf("e"), u = n > 0 ? +e.slice(n + 1) : 0, t = n > 0 ? n : e.length, i = e.indexOf("."), a = i < 0 ? 0 : t - 1 - i;
  return Math.max(0, a - u);
}
function G(r, e) {
  var n = Math.log, u = Math.LN10, t = Math.floor(n(r[1] - r[0]) / u), i = Math.round(n(R(e[1] - e[0])) / u), a = Math.min(Math.max(-t + i, 0), 20);
  return isFinite(a) ? a : 20;
}
function U(r, e) {
  var n = P(r, function(c, M) {
    return c + (isNaN(M) ? 0 : M);
  }, 0);
  if (n === 0)
    return [];
  for (var u = Math.pow(10, e), t = N(r, function(c) {
    return (isNaN(c) ? 0 : c) / n * u * 100;
  }), i = u * 100, a = N(t, function(c) {
    return Math.floor(c);
  }), d = P(a, function(c, M) {
    return c + M;
  }, 0), f = N(t, function(c, M) {
    return c - a[M];
  }); d < i; ) {
    for (var o = Number.NEGATIVE_INFINITY, s = null, v = 0, w = f.length; v < w; ++v)
      f[v] > o && (o = f[v], s = v);
    ++a[s], f[s] = 0, ++d;
  }
  return N(a, function(c) {
    return c / u;
  });
}
function k(r, e) {
  var n = Math.max(m(r), m(e)), u = r + e;
  return n > x ? u : F(u, n);
}
var q = 9007199254740991;
function Q(r) {
  var e = Math.PI * 2;
  return (r % e + e) % e;
}
function Z(r) {
  return r > -p && r < p;
}
var T = /^(?:(\d{4})(?:[-\/](\d{1,2})(?:[-\/](\d{1,2})(?:[T ](\d{1,2})(?::(\d{1,2})(?::(\d{1,2})(?:[.,](\d+))?)?)?(Z|[\+\-]\d\d:?\d\d)?)?)?)?)?$/;
function $(r) {
  if (r instanceof Date)
    return r;
  if (h(r)) {
    var e = T.exec(r);
    if (!e)
      return /* @__PURE__ */ new Date(NaN);
    if (e[8]) {
      var n = +e[4] || 0;
      return e[8].toUpperCase() !== "Z" && (n -= +e[8].slice(0, 3)), new Date(Date.UTC(+e[1], +(e[2] || 1) - 1, +e[3] || 1, n, +(e[5] || 0), +e[6] || 0, e[7] ? +e[7].substring(0, 3) : 0));
    } else
      return new Date(+e[1], +(e[2] || 1) - 1, +e[3] || 1, +e[4] || 0, +(e[5] || 0), +e[6] || 0, e[7] ? +e[7].substring(0, 3) : 0);
  } else if (r == null)
    return /* @__PURE__ */ new Date(NaN);
  return new Date(Math.round(r));
}
function z(r) {
  return Math.pow(10, I(r));
}
function I(r) {
  if (r === 0)
    return 0;
  var e = Math.floor(Math.log(r) / Math.LN10);
  return r / Math.pow(10, e) >= 10 && e++, e;
}
function X(r, e) {
  var n = I(r), u = Math.pow(10, n), t = r / u, i;
  return t < 1.5 ? i = 1 : t < 2.5 ? i = 2 : t < 4 ? i = 3 : t < 7 ? i = 5 : i = 10, r = i * u, n >= -20 ? +r.toFixed(n < 0 ? -n : 0) : r;
}
function B(r, e) {
  var n = (r.length - 1) * e + 1, u = Math.floor(n), t = +r[u - 1], i = n - u;
  return i ? t + i * (r[u] - t) : t;
}
function H(r) {
  r.sort(function(f, o) {
    return d(f, o, 0) ? -1 : 1;
  });
  for (var e = -1 / 0, n = 1, u = 0; u < r.length; ) {
    for (var t = r[u].interval, i = r[u].close, a = 0; a < 2; a++)
      t[a] <= e && (t[a] = e, i[a] = a ? 1 : 1 - n), e = t[a], n = i[a];
    t[0] === t[1] && i[0] * i[1] !== 1 ? r.splice(u, 1) : u++;
  }
  return r;
  function d(f, o, s) {
    return f.interval[s] < o.interval[s] || f.interval[s] === o.interval[s] && (f.close[s] - o.close[s] === (s ? -1 : 1) || !s && d(f, o, 1));
  }
}
function C(r) {
  var e = parseFloat(r);
  return e == r && (e !== 0 || !h(r) || r.indexOf("x") <= 0) ? e : NaN;
}
function V(r) {
  return !isNaN(C(r));
}
function Y() {
  return Math.round(Math.random() * 9);
}
function D(r, e) {
  return e === 0 ? r : D(e, r % e);
}
function j(r, e) {
  return r == null ? e : e == null ? r : r * e / D(r, e);
}
export {
  q as MAX_SAFE_INTEGER,
  k as addSafe,
  b as asc,
  D as getGreatestCommonDividor,
  j as getLeastCommonMultiple,
  U as getPercentSeats,
  G as getPixelPrecision,
  m as getPrecision,
  O as getPrecisionSafe,
  Y as getRandomIdBase,
  V as isNumeric,
  Z as isRadianAroundZero,
  _ as linearMap,
  R as mathAbs,
  A as mathMax,
  l as mathMin,
  X as nice,
  C as numericToNumber,
  $ as parseDate,
  y as parsePercent,
  g as parsePositionOption,
  E as parsePositionSizeOption,
  B as quantile,
  z as quantity,
  I as quantityExponent,
  H as reformIntervals,
  Q as remRadian,
  F as round
};
