// echarts@6.0.0/lib/util/states.js
import { each as p, isArrayLike as $, isObject as ee, keys as te, indexOf as V, extend as v, isArray as M } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { getECData as S } from "./innerStore.js";
import { liftColor as B } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/tool/color.js";
import { makeInner as Y, queryDataIndex as z } from "./model.js";
import re from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Path.js";
import { error as D } from "./log.js";
var A = 1, L = {}, R = Y(), O = Y(), w = 0, C = 1, I = 2, xe = ["emphasis", "blur", "select"], Be = ["normal", "emphasis", "blur", "select"], ne = 10, ie = 9, ae = "highlight", oe = "downplay", U = "select", le = "unselect", W = "toggleSelect", Ae = "selectchanged";
function g(e) {
  return e != null && e !== "none";
}
function E(e, t, r) {
  e.onHoverStateChange && (e.hoverState || 0) !== r && e.onHoverStateChange(t), e.hoverState = r;
}
function Z(e) {
  E(e, "emphasis", I);
}
function j(e) {
  e.hoverState === I && E(e, "normal", w);
}
function H(e) {
  E(e, "blur", C);
}
function q(e) {
  e.hoverState === C && E(e, "normal", w);
}
function ue(e) {
  e.selected = !0;
}
function ce(e) {
  e.selected = !1;
}
function P(e, t, r) {
  t(e, r);
}
function h(e, t, r) {
  P(e, t, r), e.isGroup && e.traverse(function(n) {
    P(n, t, r);
  });
}
function Le(e, t) {
  switch (t) {
    case "emphasis":
      e.hoverState = I;
      break;
    case "normal":
      e.hoverState = w;
      break;
    case "blur":
      e.hoverState = C;
      break;
    case "select":
      e.selected = !0;
  }
}
function fe(e, t, r, n) {
  for (var i = e.style, a = {}, o = 0; o < t.length; o++) {
    var u = t[o], c = i[u];
    a[u] = c ?? (n && n[u]);
  }
  for (var o = 0; o < e.animators.length; o++) {
    var l = e.animators[o];
    l.__fromStateTransition && l.__fromStateTransition.indexOf(r) < 0 && l.targetName === "style" && l.saveTo(a, t);
  }
  return a;
}
function se(e, t, r, n) {
  var i = r && V(r, "select") >= 0, a = !1;
  if (e instanceof re) {
    var o = R(e), u = i && o.selectFill || o.normalFill, c = i && o.selectStroke || o.normalStroke;
    if (g(u) || g(c)) {
      n = n || {};
      var l = n.style || {};
      l.fill === "inherit" ? (a = !0, n = v({}, n), l = v({}, l), l.fill = u) : !g(l.fill) && g(u) ? (a = !0, n = v({}, n), l = v({}, l), l.fill = B(u)) : !g(l.stroke) && g(c) && (a || (n = v({}, n), l = v({}, l)), l.stroke = B(c)), n.style = l;
    }
  }
  if (n && n.z2 == null) {
    a || (n = v({}, n));
    var s = e.z2EmphasisLift;
    n.z2 = e.z2 + (s ?? ne);
  }
  return n;
}
function ve(e, t, r) {
  if (r && r.z2 == null) {
    r = v({}, r);
    var n = e.z2SelectLift;
    r.z2 = e.z2 + (n ?? ie);
  }
  return r;
}
function he(e, t, r) {
  var n = V(e.currentStates, t) >= 0, i = e.style.opacity, a = n ? null : fe(e, ["opacity"], t, {
    opacity: 1
  });
  r = r || {};
  var o = r.style || {};
  return o.opacity == null && (r = v({}, r), o = v({
    // Already being applied 'emphasis'. DON'T mul opacity multiple times.
    opacity: n ? i : a.opacity * 0.1
  }, o), r.style = o), r;
}
function _(e, t) {
  var r = this.states[e];
  if (this.style) {
    if (e === "emphasis")
      return se(this, e, t, r);
    if (e === "blur")
      return he(this, e, r);
    if (e === "select")
      return ve(this, e, r);
  }
  return r;
}
function ge(e) {
  e.stateProxy = _;
  var t = e.getTextContent(), r = e.getTextGuideLine();
  t && (t.stateProxy = _), r && (r.stateProxy = _);
}
function k(e, t) {
  !J(e, t) && !e.__highByOuter && h(e, Z);
}
function G(e, t) {
  !J(e, t) && !e.__highByOuter && h(e, j);
}
function Pe(e, t) {
  e.__highByOuter |= 1 << (t || 0), h(e, Z);
}
function ke(e, t) {
  !(e.__highByOuter &= ~(1 << (t || 0))) && h(e, j);
}
function Ge(e) {
  h(e, H);
}
function Se(e) {
  h(e, q);
}
function pe(e) {
  h(e, ue);
}
function de(e) {
  h(e, ce);
}
function J(e, t) {
  return e.__highDownSilentOnTouch && t.zrByTouch;
}
function me(e) {
  var t = e.getModel(), r = [], n = [];
  t.eachComponent(function(i, a) {
    var o = O(a), u = i === "series", c = u ? e.getViewOfSeriesModel(a) : e.getViewOfComponentModel(a);
    !u && n.push(c), o.isBlured && (c.group.traverse(function(l) {
      q(l);
    }), u && r.push(a)), o.isBlured = !1;
  }), p(n, function(i) {
    i && i.toggleBlurSeries && i.toggleBlurSeries(r, !1, t);
  });
}
function T(e, t, r, n) {
  var i = n.getModel();
  r = r || "coordinateSystem";
  function a(l, s) {
    for (var f = 0; f < s.length; f++) {
      var d = l.getItemGraphicEl(s[f]);
      d && Se(d);
    }
  }
  if (e != null && !(!t || t === "none")) {
    var o = i.getSeriesByIndex(e), u = o.coordinateSystem;
    u && u.master && (u = u.master);
    var c = [];
    i.eachSeries(function(l) {
      var s = o === l, f = l.coordinateSystem;
      f && f.master && (f = f.master);
      var d = f && u ? f === u : s;
      if (!// Not blur other series if blurScope series
      (r === "series" && !s || r === "coordinateSystem" && !d || t === "series" && s)) {
        var X = n.getViewOfSeriesModel(l);
        if (X.group.traverse(function(x) {
          x.__highByOuter && s && t === "self" || H(x);
        }), $(t))
          a(l.getData(), t);
        else if (ee(t))
          for (var y = te(t), m = 0; m < y.length; m++)
            a(l.getData(y[m]), t[y[m]]);
        c.push(l), O(l).isBlured = !0;
      }
    }), i.eachComponent(function(l, s) {
      if (l !== "series") {
        var f = n.getViewOfComponentModel(s);
        f && f.toggleBlurSeries && f.toggleBlurSeries(c, !0, i);
      }
    });
  }
}
function N(e, t, r) {
  if (!(e == null || t == null)) {
    var n = r.getModel().getComponent(e, t);
    if (n) {
      O(n).isBlured = !0;
      var i = r.getViewOfComponentModel(n);
      !i || !i.focusBlurEnabled || i.group.traverse(function(a) {
        H(a);
      });
    }
  }
}
function Ne(e, t, r) {
  var n = e.seriesIndex, i = e.getData(t.dataType);
  if (!i) {
    process.env.NODE_ENV !== "production" && D("Unknown dataType " + t.dataType);
    return;
  }
  var a = z(i, t);
  a = (M(a) ? a[0] : a) || 0;
  var o = i.getItemGraphicEl(a);
  if (!o)
    for (var u = i.count(), c = 0; !o && c < u; )
      o = i.getItemGraphicEl(c++);
  if (o) {
    var l = S(o);
    T(n, l.focus, l.blurScope, r);
  } else {
    var s = e.get(["emphasis", "focus"]), f = e.get(["emphasis", "blurScope"]);
    s != null && T(n, s, f, r);
  }
}
function K(e, t, r, n) {
  var i = {
    focusSelf: !1,
    dispatchers: null
  };
  if (e == null || e === "series" || t == null || r == null)
    return i;
  var a = n.getModel().getComponent(e, t);
  if (!a)
    return i;
  var o = n.getViewOfComponentModel(a);
  if (!o || !o.findHighDownDispatchers)
    return i;
  for (var u = o.findHighDownDispatchers(r), c, l = 0; l < u.length; l++)
    if (process.env.NODE_ENV !== "production" && !b(u[l]) && D("param should be highDownDispatcher"), S(u[l]).focus === "self") {
      c = !0;
      break;
    }
  return {
    focusSelf: c,
    dispatchers: u
  };
}
function Fe(e, t, r) {
  process.env.NODE_ENV !== "production" && !b(e) && D("param should be highDownDispatcher");
  var n = S(e), i = K(n.componentMainType, n.componentIndex, n.componentHighDownName, r), a = i.dispatchers, o = i.focusSelf;
  a ? (o && N(n.componentMainType, n.componentIndex, r), p(a, function(u) {
    return k(u, t);
  })) : (T(n.seriesIndex, n.focus, n.blurScope, r), n.focus === "self" && N(n.componentMainType, n.componentIndex, r), k(e, t));
}
function Ve(e, t, r) {
  process.env.NODE_ENV !== "production" && !b(e) && D("param should be highDownDispatcher"), me(r);
  var n = S(e), i = K(n.componentMainType, n.componentIndex, n.componentHighDownName, r).dispatchers;
  i ? p(i, function(a) {
    return G(a, t);
  }) : G(e, t);
}
function Me(e, t, r) {
  if (Te(t)) {
    var n = t.dataType, i = e.getData(n), a = z(i, t);
    M(a) || (a = [a]), e[t.type === W ? "toggleSelect" : t.type === U ? "select" : "unselect"](a, n);
  }
}
function Ye(e) {
  var t = e.getAllData();
  p(t, function(r) {
    var n = r.data, i = r.type;
    n.eachItemGraphicEl(function(a, o) {
      e.isSelected(o, i) ? pe(a) : de(a);
    });
  });
}
function ze(e) {
  var t = [];
  return e.eachSeries(function(r) {
    var n = r.getAllData();
    p(n, function(i) {
      i.data;
      var a = i.type, o = r.getSelectedDataIndices();
      if (o.length > 0) {
        var u = {
          dataIndex: o,
          seriesIndex: r.seriesIndex
        };
        a != null && (u.dataType = a), t.push(u);
      }
    });
  }), t;
}
function De(e, t, r) {
  Q(e, !0), h(e, ge), ye(e, t, r);
}
function Ee(e) {
  Q(e, !1);
}
function Re(e, t, r, n) {
  n ? Ee(e) : De(e, t, r);
}
function ye(e, t, r) {
  var n = S(e);
  t != null ? (n.focus = t, n.blurScope = r) : n.focus && (n.focus = null);
}
var F = ["emphasis", "blur", "select"], _e = {
  itemStyle: "getItemStyle",
  lineStyle: "getLineStyle",
  areaStyle: "getAreaStyle"
};
function Ue(e, t, r, n) {
  r = r || "itemStyle";
  for (var i = 0; i < F.length; i++) {
    var a = F[i], o = t.getModel([a, r]), u = e.ensureState(a);
    u.style = n ? n(o) : o[_e[r]]();
  }
}
function Q(e, t) {
  var r = t === !1, n = e;
  e.highDownSilentOnTouch && (n.__highDownSilentOnTouch = e.highDownSilentOnTouch), (!r || n.__highDownDispatcher) && (n.__highByOuter = n.__highByOuter || 0, n.__highDownDispatcher = !r);
}
function b(e) {
  return !!(e && e.__highDownDispatcher);
}
function We(e, t, r) {
  var n = S(e);
  n.componentMainType = t.mainType, n.componentIndex = t.componentIndex, n.componentHighDownName = r;
}
function Ze(e) {
  var t = L[e];
  return t == null && A <= 32 && (t = L[e] = A++), t;
}
function Te(e) {
  var t = e.type;
  return t === U || t === le || t === W;
}
function je(e) {
  var t = e.type;
  return t === ae || t === oe;
}
function qe(e) {
  var t = R(e);
  t.normalFill = e.style.fill, t.normalStroke = e.style.stroke;
  var r = e.states.select || {};
  t.selectFill = r.style && r.style.fill || null, t.selectStroke = r.style && r.style.stroke || null;
}
export {
  Be as DISPLAY_STATES,
  oe as DOWNPLAY_ACTION_TYPE,
  ae as HIGHLIGHT_ACTION_TYPE,
  C as HOVER_STATE_BLUR,
  I as HOVER_STATE_EMPHASIS,
  w as HOVER_STATE_NORMAL,
  U as SELECT_ACTION_TYPE,
  Ae as SELECT_CHANGED_EVENT_TYPE,
  xe as SPECIAL_STATES,
  W as TOGGLE_SELECT_ACTION_TYPE,
  le as UNSELECT_ACTION_TYPE,
  ne as Z2_EMPHASIS_LIFT,
  ie as Z2_SELECT_LIFT,
  me as allLeaveBlur,
  N as blurComponent,
  T as blurSeries,
  Ne as blurSeriesFromHighlightPayload,
  Ee as disableHoverEmphasis,
  We as enableComponentHighDownFeatures,
  De as enableHoverEmphasis,
  ye as enableHoverFocus,
  Ge as enterBlur,
  Pe as enterEmphasis,
  k as enterEmphasisWhenMouseOver,
  pe as enterSelect,
  K as findComponentHighDownDispatchers,
  ze as getAllSelectedIndices,
  Ze as getHighlightDigit,
  Ve as handleGlobalMouseOutForHighDown,
  Fe as handleGlobalMouseOverForHighDown,
  b as isHighDownDispatcher,
  je as isHighDownPayload,
  Te as isSelectChangePayload,
  Se as leaveBlur,
  ke as leaveEmphasis,
  G as leaveEmphasisWhenMouseOut,
  de as leaveSelect,
  qe as savePathStates,
  Q as setAsHighDownDispatcher,
  ge as setDefaultStateProxy,
  Le as setStatesFlag,
  Ue as setStatesStylesFromModel,
  Re as toggleHoverEmphasis,
  Me as toggleSelectionFromPayload,
  Ye as updateSeriesElementSelection
};
