// echarts@6.0.0/lib/util/styleCompat.js
import { hasOwn as e, each as g } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import f from "../visual/tokens.js";
var r = {};
function C(o, t, l, n) {
  return o && (o.legacy || o.legacy !== !1 && !l && !n && t !== "tspan" && (t === "text" || e(o, "text")));
}
function O(o, t, l) {
  var n = o, a, x, d;
  if (t === "text")
    d = n;
  else {
    d = {}, e(n, "text") && (d.text = n.text), e(n, "rich") && (d.rich = n.rich), e(n, "textFill") && (d.fill = n.textFill), e(n, "textStroke") && (d.stroke = n.textStroke), e(n, "fontFamily") && (d.fontFamily = n.fontFamily), e(n, "fontSize") && (d.fontSize = n.fontSize), e(n, "fontStyle") && (d.fontStyle = n.fontStyle), e(n, "fontWeight") && (d.fontWeight = n.fontWeight), x = {
      type: "text",
      style: d,
      // ec4 does not support rectText trigger.
      // And when text position is different in normal and emphasis
      // => hover text trigger emphasis;
      // => text position changed, leave mouse pointer immediately;
      // That might cause incorrect state.
      silent: !0
    }, a = {};
    var i = e(n, "textPosition");
    l ? a.position = i ? n.textPosition : "inside" : i && (a.position = n.textPosition), e(n, "textPosition") && (a.position = n.textPosition), e(n, "textOffset") && (a.offset = n.textOffset), e(n, "textRotation") && (a.rotation = n.textRotation), e(n, "textDistance") && (a.distance = n.textDistance);
  }
  return S(d, o), g(d.rich, function(s) {
    S(s, s);
  }), {
    textConfig: a,
    textContent: x
  };
}
function S(o, t) {
  t && (t.font = t.textFont || t.font, e(t, "textStrokeWidth") && (o.lineWidth = t.textStrokeWidth), e(t, "textAlign") && (o.align = t.textAlign), e(t, "textVerticalAlign") && (o.verticalAlign = t.textVerticalAlign), e(t, "textLineHeight") && (o.lineHeight = t.textLineHeight), e(t, "textWidth") && (o.width = t.textWidth), e(t, "textHeight") && (o.height = t.textHeight), e(t, "textBackgroundColor") && (o.backgroundColor = t.textBackgroundColor), e(t, "textPadding") && (o.padding = t.textPadding), e(t, "textBorderColor") && (o.borderColor = t.textBorderColor), e(t, "textBorderWidth") && (o.borderWidth = t.textBorderWidth), e(t, "textBorderRadius") && (o.borderRadius = t.textBorderRadius), e(t, "textBoxShadowColor") && (o.shadowColor = t.textBoxShadowColor), e(t, "textBoxShadowBlur") && (o.shadowBlur = t.textBoxShadowBlur), e(t, "textBoxShadowOffsetX") && (o.shadowOffsetX = t.textBoxShadowOffsetX), e(t, "textBoxShadowOffsetY") && (o.shadowOffsetY = t.textBoxShadowOffsetY));
}
function W(o, t, l) {
  var n = o;
  n.textPosition = n.textPosition || l.position || "inside", l.offset != null && (n.textOffset = l.offset), l.rotation != null && (n.textRotation = l.rotation), l.distance != null && (n.textDistance = l.distance);
  var a = n.textPosition.indexOf("inside") >= 0, x = o.fill || f.color.neutral99;
  w(n, t);
  var d = n.textFill == null;
  return a ? d && (n.textFill = l.insideFill || f.color.neutral00, !n.textStroke && l.insideStroke && (n.textStroke = l.insideStroke), !n.textStroke && (n.textStroke = x), n.textStrokeWidth == null && (n.textStrokeWidth = 2)) : (d && (n.textFill = o.fill || l.outsideFill || f.color.neutral00), !n.textStroke && l.outsideStroke && (n.textStroke = l.outsideStroke)), n.text = t.text, n.rich = t.rich, g(t.rich, function(i) {
    w(i, i);
  }), n;
}
function w(o, t) {
  t && (e(t, "fill") && (o.textFill = t.fill), e(t, "stroke") && (o.textStroke = t.fill), e(t, "lineWidth") && (o.textStrokeWidth = t.lineWidth), e(t, "font") && (o.font = t.font), e(t, "fontStyle") && (o.fontStyle = t.fontStyle), e(t, "fontWeight") && (o.fontWeight = t.fontWeight), e(t, "fontSize") && (o.fontSize = t.fontSize), e(t, "fontFamily") && (o.fontFamily = t.fontFamily), e(t, "align") && (o.textAlign = t.align), e(t, "verticalAlign") && (o.textVerticalAlign = t.verticalAlign), e(t, "lineHeight") && (o.textLineHeight = t.lineHeight), e(t, "width") && (o.textWidth = t.width), e(t, "height") && (o.textHeight = t.height), e(t, "backgroundColor") && (o.textBackgroundColor = t.backgroundColor), e(t, "padding") && (o.textPadding = t.padding), e(t, "borderColor") && (o.textBorderColor = t.borderColor), e(t, "borderWidth") && (o.textBorderWidth = t.borderWidth), e(t, "borderRadius") && (o.textBorderRadius = t.borderRadius), e(t, "shadowColor") && (o.textBoxShadowColor = t.shadowColor), e(t, "shadowBlur") && (o.textBoxShadowBlur = t.shadowBlur), e(t, "shadowOffsetX") && (o.textBoxShadowOffsetX = t.shadowOffsetX), e(t, "shadowOffsetY") && (o.textBoxShadowOffsetY = t.shadowOffsetY), e(t, "textShadowColor") && (o.textShadowColor = t.textShadowColor), e(t, "textShadowBlur") && (o.textShadowBlur = t.textShadowBlur), e(t, "textShadowOffsetX") && (o.textShadowOffsetX = t.textShadowOffsetX), e(t, "textShadowOffsetY") && (o.textShadowOffsetY = t.textShadowOffsetY));
}
function v(o, t) {
  if (process.env.NODE_ENV !== "production") {
    var l = o + "^_^" + t;
    r[l] || (console.warn('[ECharts] DEPRECATED: "' + o + '" has been deprecated. ' + t), r[l] = !0);
  }
}
export {
  O as convertFromEC4CompatibleStyle,
  W as convertToEC4StyleForCustomSerise,
  C as isEC4CompatibleStyle,
  v as warnDeprecated
};
