// echarts@6.0.0/lib/util/symbol.js
import { isArray as y, retrieve2 as M, each as C } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { makeImage as k, makePath as R } from "./graphic.js";
import P from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/BoundingRect.js";
import { calculateTextPosition as B } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/contain/text.js";
import { parsePercent as T } from "./number.js";
import L from "../visual/tokens.js";
import h from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Path.js";
import _ from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Circle.js";
import f from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Rect.js";
import E from "../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Line.js";
var I = h.extend({
  type: "triangle",
  shape: {
    cx: 0,
    cy: 0,
    width: 0,
    height: 0
  },
  buildPath: function(i, n) {
    var r = n.cx, e = n.cy, t = n.width / 2, o = n.height / 2;
    i.moveTo(r, e - o), i.lineTo(r + t, e + o), i.lineTo(r - t, e + o), i.closePath();
  }
}), q = h.extend({
  type: "diamond",
  shape: {
    cx: 0,
    cy: 0,
    width: 0,
    height: 0
  },
  buildPath: function(i, n) {
    var r = n.cx, e = n.cy, t = n.width / 2, o = n.height / 2;
    i.moveTo(r, e - o), i.lineTo(r + t, e), i.lineTo(r, e + o), i.lineTo(r - t, e), i.closePath();
  }
}), z = h.extend({
  type: "pin",
  shape: {
    // x, y on the cusp
    x: 0,
    y: 0,
    width: 0,
    height: 0
  },
  buildPath: function(i, n) {
    var r = n.x, e = n.y, t = n.width / 5 * 3, o = Math.max(t, n.height), a = t / 2, d = a * a / (o - a), c = e - o + a + d, v = Math.asin(d / a), u = Math.cos(v) * a, g = Math.sin(v), x = Math.cos(v), l = a * 0.6, w = a * 0.7;
    i.moveTo(r - u, c + d), i.arc(r, c, a, Math.PI - v, Math.PI * 2 + v), i.bezierCurveTo(r + u - g * l, c + d + x * l, r, e - w, r, e), i.bezierCurveTo(r, e - w, r - u + g * l, c + d + x * l, r - u, c + d), i.closePath();
  }
}), A = h.extend({
  type: "arrow",
  shape: {
    x: 0,
    y: 0,
    width: 0,
    height: 0
  },
  buildPath: function(i, n) {
    var r = n.height, e = n.width, t = n.x, o = n.y, a = e / 3 * 2;
    i.moveTo(t, o), i.lineTo(t + a, o + r), i.lineTo(t, o + r / 4 * 3), i.lineTo(t - a, o + r), i.lineTo(t, o), i.closePath();
  }
}), D = {
  line: E,
  rect: f,
  roundRect: f,
  square: f,
  circle: _,
  diamond: q,
  pin: z,
  arrow: A,
  triangle: I
}, W = {
  line: function(i, n, r, e, t) {
    t.x1 = i, t.y1 = n + e / 2, t.x2 = i + r, t.y2 = n + e / 2;
  },
  rect: function(i, n, r, e, t) {
    t.x = i, t.y = n, t.width = r, t.height = e;
  },
  roundRect: function(i, n, r, e, t) {
    t.x = i, t.y = n, t.width = r, t.height = e, t.r = Math.min(r, e) / 4;
  },
  square: function(i, n, r, e, t) {
    var o = Math.min(r, e);
    t.x = i, t.y = n, t.width = o, t.height = o;
  },
  circle: function(i, n, r, e, t) {
    t.cx = i + r / 2, t.cy = n + e / 2, t.r = Math.min(r, e) / 2;
  },
  diamond: function(i, n, r, e, t) {
    t.cx = i + r / 2, t.cy = n + e / 2, t.width = r, t.height = e;
  },
  pin: function(i, n, r, e, t) {
    t.x = i + r / 2, t.y = n + e / 2, t.width = r, t.height = e;
  },
  arrow: function(i, n, r, e, t) {
    t.x = i + r / 2, t.y = n + e / 2, t.width = r, t.height = e;
  },
  triangle: function(i, n, r, e, t) {
    t.cx = i + r / 2, t.cy = n + e / 2, t.width = r, t.height = e;
  }
}, m = {};
C(D, function(i, n) {
  m[n] = new i();
});
var X = h.extend({
  type: "symbol",
  shape: {
    symbolType: "",
    x: 0,
    y: 0,
    width: 0,
    height: 0
  },
  calculateTextPosition: function(i, n, r) {
    var e = B(i, n, r), t = this.shape;
    return t && t.symbolType === "pin" && n.position === "inside" && (e.y = r.y + r.height * 0.4), e;
  },
  buildPath: function(i, n, r) {
    var e = n.symbolType;
    if (e !== "none") {
      var t = m[e];
      t || (e = "rect", t = m[e]), W[e](n.x, n.y, n.width, n.height, t.shape), t.buildPath(i, t.shape, r);
    }
  }
});
function Y(i, n) {
  if (this.type !== "image") {
    var r = this.style;
    this.__isEmptyBrush ? (r.stroke = i, r.fill = n || L.color.neutral00, r.lineWidth = 2) : this.shape.symbolType === "line" ? r.stroke = i : r.fill = i, this.markRedraw();
  }
}
function Z(i, n, r, e, t, o, a) {
  var d = i.indexOf("empty") === 0;
  d && (i = i.substr(5, 1).toLowerCase() + i.substr(6));
  var c;
  return i.indexOf("image://") === 0 ? c = k(i.slice(8), new P(n, r, e, t), a ? "center" : "cover") : i.indexOf("path://") === 0 ? c = R(i.slice(7), {}, new P(n, r, e, t), a ? "center" : "cover") : c = new X({
    shape: {
      symbolType: i,
      x: n,
      y: r,
      width: e,
      height: t
    }
  }), c.__isEmptyBrush = d, c.setColor = Y, o && c.setColor(o), c;
}
function $(i) {
  return y(i) || (i = [+i, +i]), [i[0] || 0, i[1] || 0];
}
function b(i, n) {
  if (i != null)
    return y(i) || (i = [i, i]), [T(i[0], n[0]) || 0, T(M(i[1], i[0]), n[1]) || 0];
}
export {
  Z as createSymbol,
  b as normalizeSymbolOffset,
  $ as normalizeSymbolSize,
  m as symbolBuildProxies
};
