// echarts@6.0.0/lib/util/throttle.js
var f = "\0__throttleOriginMethod", D = "\0__throttleRate", E = "\0__throttleType";
function O(n, a, r) {
  var l, e = 0, u = 0, t = null, i, g, m, c;
  a = a || 0;
  function s() {
    u = (/* @__PURE__ */ new Date()).getTime(), t = null, n.apply(g, m || []);
  }
  var T = function() {
    for (var v = [], o = 0; o < arguments.length; o++)
      v[o] = arguments[o];
    l = (/* @__PURE__ */ new Date()).getTime(), g = this, m = v;
    var h = c || a, x = c || r;
    c = null, i = l - (x ? e : u) - h, clearTimeout(t), x ? t = setTimeout(s, h) : i >= 0 ? s() : t = setTimeout(s, -i), e = l;
  };
  return T.clear = function() {
    t && (clearTimeout(t), t = null);
  }, T.debounceNextCall = function(v) {
    c = v;
  }, T;
}
function R(n, a, r, l) {
  var e = n[a];
  if (e) {
    var u = e[f] || e, t = e[E], i = e[D];
    if (i !== r || t !== l) {
      if (r == null || !l)
        return n[a] = u;
      e = n[a] = O(u, r, l === "debounce"), e[f] = u, e[E] = l, e[D] = r;
    }
    return e;
  }
}
function _(n, a) {
  var r = n[a];
  r && r[f] && (r.clear && r.clear(), n[a] = r[f]);
}
export {
  _ as clear,
  R as createOrUpdate,
  O as throttle
};
