// echarts@6.0.0/lib/util/time.js
import { isString as H, isFunction as b, each as p, isObject as k, isArray as F } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { parseDate as D } from "./number.js";
import { getLocaleModel as R, getDefaultLocaleModel as j, SYSTEM_LANG as x } from "../core/locale.js";
import B from "../model/Model.js";
import { getScaleBreakHelper as Q } from "../scale/break.js";
var V = 1e3, z = V * 60, J = z * 60, K = J * 24, de = K * 365, X = {
  year: /({yyyy}|{yy})/,
  month: /({MMMM}|{MMM}|{MM}|{M})/,
  day: /({dd}|{d})/,
  hour: /({HH}|{H}|{hh}|{h})/,
  minute: /({mm}|{m})/,
  second: /({ss}|{s})/,
  millisecond: /({SSS}|{S})/
}, g = {
  year: "{yyyy}",
  month: "{MMM}",
  day: "{d}",
  hour: "{HH}:{mm}",
  minute: "{HH}:{mm}",
  second: "{HH}:{mm}:{ss}",
  millisecond: "{HH}:{mm}:{ss} {SSS}"
}, Z = "{yyyy}-{MM}-{dd} {HH}:{mm}:{ss} {SSS}", M = "{yyyy}-{MM}-{dd}", ye = {
  year: "{yyyy}",
  month: "{yyyy}-{MM}",
  day: M,
  hour: M + " " + g.hour,
  minute: M + " " + g.minute,
  second: M + " " + g.second,
  millisecond: Z
}, S = ["year", "month", "day", "hour", "minute", "second", "millisecond"], ve = ["year", "half-year", "quarter", "month", "week", "half-week", "day", "half-day", "quarter-day", "hour", "minute", "second", "millisecond"];
function fe(e) {
  return !H(e) && !b(e) ? $(e) : e;
}
function $(e) {
  e = e || {};
  var t = {}, r = !0;
  return p(S, function(s) {
    r && (r = e[s] == null);
  }), p(S, function(s, l) {
    var o = e[s];
    t[s] = {};
    for (var a = null, c = l; c >= 0; c--) {
      var u = S[c], m = k(o) && !F(o) ? o[u] : o, n = void 0;
      F(m) ? (n = m.slice(), a = n[0] || "") : H(m) ? (a = m, n = [a]) : (a == null ? a = g[s] : X[u].test(a) || (a = t[u][u][0] + " " + a), n = [a], r && (n[1] = "{primary|" + a + "}")), t[s][u] = n;
    }
  }), t;
}
function i(e, t) {
  return e += "", "0000".substr(0, t - e.length) + e;
}
function C(e) {
  switch (e) {
    case "half-year":
    case "quarter":
      return "month";
    case "week":
    case "half-week":
      return "day";
    case "half-day":
    case "quarter-day":
      return "hour";
    default:
      return e;
  }
}
function he(e) {
  return e === C(e);
}
function Me(e) {
  switch (e) {
    case "year":
    case "month":
      return "day";
    case "millisecond":
      return "millisecond";
    default:
      return "second";
  }
}
function I(e, t, r, s) {
  var l = D(e), o = l[ee(r)](), a = l[Y(r)]() + 1, c = Math.floor((a - 1) / 3) + 1, u = l[E(r)](), m = l["get" + (r ? "UTC" : "") + "Day"](), n = l[A(r)](), d = (n - 1) % 12 + 1, y = l[G(r)](), v = l[O(r)](), f = l[w(r)](), N = n >= 12 ? "pm" : "am", L = N.toUpperCase(), _ = s instanceof B ? s : R(s || x) || j(), h = _.getModel("time"), q = h.get("month"), T = h.get("monthAbbr"), P = h.get("dayOfWeek"), W = h.get("dayOfWeekAbbr");
  return (t || "").replace(/{a}/g, N + "").replace(/{A}/g, L + "").replace(/{yyyy}/g, o + "").replace(/{yy}/g, i(o % 100 + "", 2)).replace(/{Q}/g, c + "").replace(/{MMMM}/g, q[a - 1]).replace(/{MMM}/g, T[a - 1]).replace(/{MM}/g, i(a, 2)).replace(/{M}/g, a + "").replace(/{dd}/g, i(u, 2)).replace(/{d}/g, u + "").replace(/{eeee}/g, P[m]).replace(/{ee}/g, W[m]).replace(/{e}/g, m + "").replace(/{HH}/g, i(n, 2)).replace(/{H}/g, n + "").replace(/{hh}/g, i(d + "", 2)).replace(/{h}/g, d + "").replace(/{mm}/g, i(y, 2)).replace(/{m}/g, y + "").replace(/{ss}/g, i(v, 2)).replace(/{s}/g, v + "").replace(/{SSS}/g, i(f, 3)).replace(/{S}/g, f + "");
}
function ge(e, t, r, s, l) {
  var o = null;
  if (H(r))
    o = r;
  else if (b(r)) {
    var a = {
      time: e.time,
      level: e.time.level
    }, c = Q();
    c && c.makeAxisLabelFormatterParamBreak(a, e.break), o = r(e.value, t, a);
  } else {
    var u = e.time;
    if (u) {
      var m = r[u.lowerTimeUnit][u.upperTimeUnit];
      o = m[Math.min(u.level, m.length - 1)] || "";
    } else {
      var n = U(e.value, l);
      o = r[n][n][0];
    }
  }
  return I(new Date(e.value), o, l, s);
}
function U(e, t) {
  var r = D(e), s = r[Y(t)]() + 1, l = r[E(t)](), o = r[A(t)](), a = r[G(t)](), c = r[O(t)](), u = r[w(t)](), m = u === 0, n = m && c === 0, d = n && a === 0, y = d && o === 0, v = y && l === 1, f = v && s === 1;
  return f ? "year" : v ? "month" : y ? "day" : d ? "hour" : n ? "minute" : m ? "second" : "millisecond";
}
function Se(e, t, r) {
  switch (t) {
    case "year":
      e[re(r)](0);
    case "month":
      e[ae(r)](1);
    case "day":
      e[te(r)](0);
    case "hour":
      e[ne(r)](0);
    case "minute":
      e[le(r)](0);
    case "second":
      e[oe(r)](0);
  }
  return e;
}
function ee(e) {
  return e ? "getUTCFullYear" : "getFullYear";
}
function Y(e) {
  return e ? "getUTCMonth" : "getMonth";
}
function E(e) {
  return e ? "getUTCDate" : "getDate";
}
function A(e) {
  return e ? "getUTCHours" : "getHours";
}
function G(e) {
  return e ? "getUTCMinutes" : "getMinutes";
}
function O(e) {
  return e ? "getUTCSeconds" : "getSeconds";
}
function w(e) {
  return e ? "getUTCMilliseconds" : "getMilliseconds";
}
function He(e) {
  return e ? "setUTCFullYear" : "setFullYear";
}
function re(e) {
  return e ? "setUTCMonth" : "setMonth";
}
function ae(e) {
  return e ? "setUTCDate" : "setDate";
}
function te(e) {
  return e ? "setUTCHours" : "setHours";
}
function ne(e) {
  return e ? "setUTCMinutes" : "setMinutes";
}
function le(e) {
  return e ? "setUTCSeconds" : "setSeconds";
}
function oe(e) {
  return e ? "setUTCMilliseconds" : "setMilliseconds";
}
export {
  K as ONE_DAY,
  J as ONE_HOUR,
  z as ONE_MINUTE,
  V as ONE_SECOND,
  de as ONE_YEAR,
  E as dateGetterName,
  ae as dateSetterName,
  I as format,
  ye as fullLeveledFormatter,
  ee as fullYearGetterName,
  He as fullYearSetterName,
  Me as getDefaultFormatPrecisionOfInterval,
  C as getPrimaryTimeUnit,
  U as getUnitFromValue,
  A as hoursGetterName,
  te as hoursSetterName,
  he as isPrimaryTimeUnit,
  ge as leveledFormat,
  w as millisecondsGetterName,
  oe as millisecondsSetterName,
  G as minutesGetterName,
  ne as minutesSetterName,
  Y as monthGetterName,
  re as monthSetterName,
  i as pad,
  fe as parseTimeAxisLabelFormatter,
  S as primaryTimeUnits,
  Se as roundTime,
  O as secondsGetterName,
  le as secondsSetterName,
  ve as timeUnits
};
