// @tiptap/extension-collaboration@2.24.0/dist/index.js
import { Extension as g } from "../../../../../@tiptap_core@2.24.0_@tiptap_pm@2.24.0/node_modules/@tiptap/core/dist/index.js";
import { Plugin as f, PluginKey as h } from "../../../../../prosemirror-state@1.4.3/node_modules/prosemirror-state/dist/index.js";
import { ySyncPluginKey as m, yUndoPluginKey as a } from "../../../../../y-prosemirror@1.3.6_prosemirror-model@1.25.0_prosemirror-state@1.4.3_prosemirror-view@1.38.1__5io44ux6u7ky2cicdaagpya6xa/node_modules/y-prosemirror/src/plugins/keys.js";
import { yUndoPlugin as p, redo as y, undo as v } from "../../../../../y-prosemirror@1.3.6_prosemirror-model@1.25.0_prosemirror-state@1.4.3_prosemirror-view@1.38.1__5io44ux6u7ky2cicdaagpya6xa/node_modules/y-prosemirror/src/plugins/undo-plugin.js";
import { ySyncPlugin as M } from "../../../../../y-prosemirror@1.3.6_prosemirror-model@1.25.0_prosemirror-state@1.4.3_prosemirror-view@1.38.1__5io44ux6u7ky2cicdaagpya6xa/node_modules/y-prosemirror/src/plugins/sync-plugin.js";
import { yXmlFragmentToProsemirrorJSON as S } from "../../../../../y-prosemirror@1.3.6_prosemirror-model@1.25.0_prosemirror-state@1.4.3_prosemirror-view@1.38.1__5io44ux6u7ky2cicdaagpya6xa/node_modules/y-prosemirror/src/lib.js";
g.create({
  name: "collaboration",
  priority: 1e3,
  addOptions() {
    return {
      document: null,
      field: "default",
      fragment: null
    };
  },
  addStorage() {
    return {
      isDisabled: !1
    };
  },
  onCreate() {
    this.editor.extensionManager.extensions.find((e) => e.name === "history") && console.warn('[tiptap warn]: "@tiptap/extension-collaboration" comes with its own history support and is not compatible with "@tiptap/extension-history".');
  },
  addCommands() {
    return {
      undo: () => ({ tr: e, state: n, dispatch: r }) => (e.setMeta("preventDispatch", !0), a.getState(n).undoManager.undoStack.length === 0 ? !1 : r ? v(n) : !0),
      redo: () => ({ tr: e, state: n, dispatch: r }) => (e.setMeta("preventDispatch", !0), a.getState(n).undoManager.redoStack.length === 0 ? !1 : r ? y(n) : !0)
    };
  },
  addKeyboardShortcuts() {
    return {
      "Mod-z": () => this.editor.commands.undo(),
      "Mod-y": () => this.editor.commands.redo(),
      "Shift-Mod-z": () => this.editor.commands.redo()
    };
  },
  addProseMirrorPlugins() {
    var e;
    const n = this.options.fragment ? this.options.fragment : this.options.document.getXmlFragment(this.options.field), r = p(this.options.yUndoOptions), i = r.spec.view;
    r.spec.view = (o) => {
      const { undoManager: t } = a.getState(o.state);
      t.restore && (t.restore(), t.restore = () => {
      });
      const s = i ? i(o) : void 0;
      return {
        destroy: () => {
          const c = t.trackedOrigins.has(t), l = t._observers;
          t.restore = () => {
            c && t.trackedOrigins.add(t), t.doc.on("afterTransaction", t.afterTransactionHandler), t._observers = l;
          }, s?.destroy && s.destroy();
        }
      };
    };
    const d = {
      ...this.options.ySyncOptions,
      onFirstRender: this.options.onFirstRender
    }, u = M(n, d);
    return this.editor.options.enableContentCheck && ((e = n.doc) === null || e === void 0 || e.on("beforeTransaction", () => {
      try {
        const o = S(n);
        if (o.content.length === 0)
          return;
        this.editor.schema.nodeFromJSON(o).check();
      } catch (o) {
        return this.editor.emit("contentError", {
          error: o,
          editor: this.editor,
          disableCollaboration: () => {
            var t;
            (t = n.doc) === null || t === void 0 || t.destroy(), this.storage.isDisabled = !0;
          }
        }), !1;
      }
    })), [
      u,
      r,
      // Only add the filterInvalidContent plugin if content checking is enabled
      this.editor.options.enableContentCheck && new f({
        key: new h("filterInvalidContent"),
        filterTransaction: () => {
          var o;
          return this.storage.isDisabled && ((o = n.doc) === null || o === void 0 || o.destroy()), !0;
        }
      })
    ].filter(Boolean);
  }
});
function x(e) {
  return !!e.getMeta(m);
}
export {
  x as isChangeOrigin
};
