// y-prosemirror@1.3.6/src/lib.js
import { yattr2markname as g } from "./plugins/sync-plugin.js";
import { createRelativePositionFromTypeIndex as c, XmlText as a, RelativePosition as _, findRootTypeKey as f, createAbsolutePositionFromRelativePosition as x, isParentOf as P, createID as b, XmlElement as w } from "../../../../yjs@13.6.27/node_modules/yjs/dist/yjs.js";
import { unexpectedCase as h } from "../../../../lib0@0.2.109/node_modules/lib0/error.js";
const R = (n, t, m) => {
  if (n === 0)
    return c(t, 0, t.length === 0 ? -1 : 0);
  let e = t._first === null ? null : (
    /** @type {Y.ContentType} */
    t._first.content.type
  );
  for (; e !== null && t !== e; ) {
    if (e instanceof a) {
      if (e._length >= n)
        return c(e, n, t.length === 0 ? -1 : 0);
      if (n -= e._length, e._item !== null && e._item.next !== null)
        e = /** @type {Y.ContentType} */
        e._item.next.content.type;
      else {
        do
          e = e._item === null ? null : e._item.parent, n--;
        while (e !== t && e !== null && e._item !== null && e._item.next === null);
        e !== null && e !== t && (e = e._item === null ? null : (
          /** @type {Y.ContentType} */
          /** @type Y.Item */
          e._item.next.content.type
        ));
      }
    } else {
      const l = (
        /** @type {any} */
        (m.get(e) || { nodeSize: 0 }).nodeSize
      );
      if (e._first !== null && n < l)
        e = /** @type {Y.ContentType} */
        e._first.content.type, n--;
      else {
        if (n === 1 && e._length === 0 && l > 1)
          return new _(e._item === null ? null : e._item.id, e._item === null ? f(e) : null, null);
        if (n -= l, e._item !== null && e._item.next !== null)
          e = /** @type {Y.ContentType} */
          e._item.next.content.type;
        else {
          if (n === 0)
            return e = e._item === null ? e : e._item.parent, new _(e._item === null ? null : e._item.id, e._item === null ? f(e) : null, null);
          do
            e = /** @type {Y.Item} */
            e._item.parent, n--;
          while (e !== t && /** @type {Y.Item} */
          e._item.next === null);
          e !== t && (e = /** @type {Y.ContentType} */
          /** @type {Y.Item} */
          /** @type {Y.Item} */
          e._item.next.content.type);
        }
      }
    }
    if (e === null)
      throw h();
    if (n === 0 && e.constructor !== a && e !== t)
      return y(e._item.parent, e._item);
  }
  return c(t, t._length, t.length === 0 ? -1 : 0);
}, y = (n, t) => {
  let m = null, e = null;
  return n._item === null ? e = f(n) : m = b(n._item.id.client, n._item.id.clock), new _(m, e, t.id);
}, S = (n, t, m, e) => {
  const l = x(m, n);
  if (l === null || l.type !== t && !P(t, l.type._item))
    return null;
  let r = l.type, i = 0;
  if (r.constructor === a)
    i = l.index;
  else if (r._item === null || !r._item.deleted) {
    let o = r._first, s = 0;
    for (; s < r._length && s < l.index && o !== null; ) {
      if (!o.deleted) {
        const u = (
          /** @type {Y.ContentType} */
          o.content.type
        );
        s++, u instanceof a ? i += u._length : i += /** @type {any} */
        e.get(u).nodeSize;
      }
      o = /** @type {Y.Item} */
      o.right;
    }
    i += 1;
  }
  for (; r !== t && r._item !== null; ) {
    const o = r._item.parent;
    if (o._item === null || !o._item.deleted) {
      i += 1;
      let s = (
        /** @type {Y.AbstractType} */
        o._first
      );
      for (; s !== null; ) {
        const u = (
          /** @type {Y.ContentType} */
          s.content.type
        );
        if (u === r)
          break;
        s.deleted || (u instanceof a ? i += u._length : i += /** @type {any} */
        e.get(u).nodeSize), s = s.right;
      }
    }
    r = /** @type {Y.AbstractType} */
    o;
  }
  return i - 1;
};
function T(n) {
  const t = n.toArray(), m = (e) => {
    let l;
    if (e instanceof a)
      l = e.toDelta().map(
        /** @param {any} d */
        (i) => {
          const o = {
            type: "text",
            text: i.insert
          };
          return i.attributes && (o.marks = Object.keys(i.attributes).map((s) => {
            const u = i.attributes[s], d = {
              type: g(s)
            };
            return Object.keys(u) && (d.attrs = u), d;
          })), o;
        }
      );
    else if (e instanceof w) {
      l = {
        type: e.nodeName
      };
      const r = e.getAttributes();
      Object.keys(r).length && (l.attrs = r);
      const i = e.toArray();
      i.length && (l.content = i.map(m).flat());
    } else
      h();
    return l;
  };
  return {
    type: "doc",
    content: t.map(m)
  };
}
export {
  R as absolutePositionToRelativePosition,
  S as relativePositionToAbsolutePosition,
  T as yXmlFragmentToProsemirrorJSON
};
