// echarts@6.0.0/lib/util/shape/sausage.js
import { __extends as S } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import "../graphic.js";
import y from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Path.js";
var A = (
  /** @class */
  /* @__PURE__ */ (function() {
    function h() {
      this.cx = 0, this.cy = 0, this.r0 = 0, this.r = 0, this.startAngle = 0, this.endAngle = Math.PI * 2, this.clockwise = !0;
    }
    return h;
  })()
), k = (
  /** @class */
  (function(h) {
    S(u, h);
    function u(n) {
      var t = h.call(this, n) || this;
      return t.type = "sausage", t;
    }
    return u.prototype.getDefaultShape = function() {
      return new A();
    }, u.prototype.buildPath = function(n, t) {
      var e = t.cx, i = t.cy, o = Math.max(t.r0 || 0, 0), c = Math.max(t.r, 0), f = (c - o) * 0.5, v = o + f, a = t.startAngle, r = t.endAngle, s = t.clockwise, l = Math.PI * 2, P = s ? r - a < l : a - r < l;
      P || (a = r - (s ? l : -l));
      var g = Math.cos(a), M = Math.sin(a), d = Math.cos(r), m = Math.sin(r);
      P ? (n.moveTo(g * o + e, M * o + i), n.arc(g * v + e, M * v + i, f, -Math.PI + a, a, !s)) : n.moveTo(g * c + e, M * c + i), n.arc(e, i, c, a, r, !s), n.arc(d * v + e, m * v + i, f, r - Math.PI * 2, r - Math.PI, !s), o !== 0 && n.arc(e, i, o, r, a, s);
    }, u;
  })(y)
);
export {
  k as default
};
