// @radix-ui/react-toggle@1.1.1/dist/index.js
import * as n from "react";
import { composeEventHandlers as f } from "../../../../../@radix-ui_primitive@1.1.1/node_modules/@radix-ui/primitive/dist/index.js";
import { useControllableState as m } from "../../../../../@radix-ui_react-use-controllable-state@1.1.0_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-use-controllable-state/dist/index.js";
import { Primitive as p } from "../../../../../@radix-ui_react-primitive@2.0.1_@types_react-dom@18.3.1_@types_react@18.3.18_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/@radix-ui/react-primitive/dist/index.js";
import { jsx as b } from "react/jsx-runtime";
var u = "Toggle", r = n.forwardRef((e, t) => {
  const { pressed: a, defaultPressed: s = !1, onPressedChange: d, ...i } = e, [o = !1, l] = m({
    prop: a,
    onChange: d,
    defaultProp: s
  });
  return /* @__PURE__ */ b(
    p.button,
    {
      type: "button",
      "aria-pressed": o,
      "data-state": o ? "on" : "off",
      "data-disabled": e.disabled ? "" : void 0,
      ...i,
      ref: t,
      onClick: f(e.onClick, () => {
        e.disabled || l(!o);
      })
    }
  );
});
r.displayName = u;
var C = r;
export {
  C as Root,
  r as Toggle
};
