// @radix-ui/react-avatar@1.1.2/dist/index.js
import * as i from "react";
import { createContextScope as C } from "../../../../../@radix-ui_react-context@1.1.1_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-context/dist/index.js";
import { useCallbackRef as I } from "../../../../../@radix-ui_react-use-callback-ref@1.1.0_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-use-callback-ref/dist/index.js";
import { useLayoutEffect as g } from "../../../../../@radix-ui_react-use-layout-effect@1.1.0_@types_react@18.3.18_react@18.3.1/node_modules/@radix-ui/react-use-layout-effect/dist/index.js";
import { Primitive as m } from "../../../../../@radix-ui_react-primitive@2.0.1_@types_react-dom@18.3.1_@types_react@18.3.18_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/@radix-ui/react-primitive/dist/index.js";
import { jsx as c } from "react/jsx-runtime";
var l = "Avatar", [R] = C(l), [_, v] = R(l), p = i.forwardRef(
  (e, r) => {
    const { __scopeAvatar: s, ...a } = e, [o, t] = i.useState("idle");
    return /* @__PURE__ */ c(
      _,
      {
        scope: s,
        imageLoadingStatus: o,
        onImageLoadingStatusChange: t,
        children: /* @__PURE__ */ c(m.span, { ...a, ref: r })
      }
    );
  }
);
p.displayName = l;
var A = "AvatarImage", S = i.forwardRef(
  (e, r) => {
    const { __scopeAvatar: s, src: a, onLoadingStatusChange: o = () => {
    }, ...t } = e, d = v(A, s), n = x(a, t.referrerPolicy), u = I((f) => {
      o(f), d.onImageLoadingStatusChange(f);
    });
    return g(() => {
      n !== "idle" && u(n);
    }, [n, u]), n === "loaded" ? /* @__PURE__ */ c(m.img, { ...t, ref: r, src: a }) : null;
  }
);
S.displayName = A;
var L = "AvatarFallback", w = i.forwardRef(
  (e, r) => {
    const { __scopeAvatar: s, delayMs: a, ...o } = e, t = v(L, s), [d, n] = i.useState(a === void 0);
    return i.useEffect(() => {
      if (a !== void 0) {
        const u = window.setTimeout(() => n(!0), a);
        return () => window.clearTimeout(u);
      }
    }, [a]), d && t.imageLoadingStatus !== "loaded" ? /* @__PURE__ */ c(m.span, { ...o, ref: r }) : null;
  }
);
w.displayName = L;
function x(e, r) {
  const [s, a] = i.useState("idle");
  return g(() => {
    if (!e) {
      a("error");
      return;
    }
    let o = !0;
    const t = new window.Image(), d = (n) => () => {
      o && a(n);
    };
    return a("loading"), t.onload = d("loaded"), t.onerror = d("error"), t.src = e, r && (t.referrerPolicy = r), () => {
      o = !1;
    };
  }, [e, r]), s;
}
var b = p, k = S, P = w;
export {
  p as Avatar,
  w as AvatarFallback,
  S as AvatarImage,
  P as Fallback,
  k as Image,
  b as Root
};
