// echarts@6.0.0/lib/legacy/dataSelectAction.js
import { extend as u, isString as v, isArray as a, each as h } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { deprecateLog as S, deprecateReplaceLog as y } from "../util/log.js";
import { queryDataIndex as E } from "../util/model.js";
function D(i, c) {
  function r(e, t) {
    var n = [];
    return e.eachComponent({
      mainType: "series",
      subType: i,
      query: t
    }, function(o) {
      n.push(o.seriesIndex);
    }), n;
  }
  h([[i + "ToggleSelect", "toggleSelect"], [i + "Select", "select"], [i + "UnSelect", "unselect"]], function(e) {
    c(e[0], function(t, n, o) {
      t = u({}, t), process.env.NODE_ENV !== "production" && y(t.type, e[1]), o.dispatchAction(u(t, {
        type: e[1],
        seriesIndex: r(n, t)
      }));
    });
  });
}
function l(i, c, r, e, t) {
  var n = i + c;
  r.isSilent(n) || (process.env.NODE_ENV !== "production" && S("event " + n + " is deprecated."), e.eachComponent({
    mainType: "series",
    subType: "pie"
  }, function(o) {
    for (var m = o.seriesIndex, s = o.option.selectedMap, p = t.selected, f = 0; f < p.length; f++)
      if (p[f].seriesIndex === m) {
        var g = o.getData(), d = E(g, t.fromActionPayload);
        r.trigger(n, {
          type: n,
          seriesId: o.id,
          name: a(d) ? g.getName(d[0]) : g.getName(d),
          selected: v(s) ? s : u({}, s)
        });
      }
  }));
}
function L(i, c, r) {
  i.on("selectchanged", function(e) {
    var t = r.getModel();
    e.isFromClick ? (l("map", "selectchanged", c, t, e), l("pie", "selectchanged", c, t, e)) : e.fromAction === "select" ? (l("map", "selected", c, t, e), l("pie", "selected", c, t, e)) : e.fromAction === "unselect" && (l("map", "unselected", c, t, e), l("pie", "unselected", c, t, e));
  });
}
export {
  D as createLegacyDataSelectAction,
  L as handleLegacySelectEvents
};
