// echarts@6.0.0/lib/coord/Axis.js
import { map as h, each as M } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { getPixelPrecision as b, linearMap as f, round as v } from "../util/number.js";
import { createAxisTicks as E, createAxisLabels as k, calculateCategoryInterval as L, createAxisLabelsComputingContext as x, AxisTickLabelComputingKind as g } from "./axisTickLabelBuilder.js";
var m = [0, 1], z = (
  /** @class */
  (function() {
    function r(t, n, e) {
      this.onBand = !1, this.inverse = !1, this.dim = t, this.scale = n, this._extent = e || [0, 0];
    }
    return r.prototype.contain = function(t) {
      var n = this._extent, e = Math.min(n[0], n[1]), i = Math.max(n[0], n[1]);
      return t >= e && t <= i;
    }, r.prototype.containData = function(t) {
      return this.scale.contain(this.scale.parse(t));
    }, r.prototype.getExtent = function() {
      return this._extent.slice();
    }, r.prototype.getPixelPrecision = function(t) {
      return b(t || this.scale.getExtent(), this._extent);
    }, r.prototype.setExtent = function(t, n) {
      var e = this._extent;
      e[0] = t, e[1] = n;
    }, r.prototype.dataToCoord = function(t, n) {
      var e = this._extent, i = this.scale;
      return t = i.normalize(i.parse(t)), this.onBand && i.type === "ordinal" && (e = e.slice(), y(e, i.count())), f(t, m, e, n);
    }, r.prototype.coordToData = function(t, n) {
      var e = this._extent, i = this.scale;
      this.onBand && i.type === "ordinal" && (e = e.slice(), y(e, i.count()));
      var a = f(t, e, m, n);
      return this.scale.scale(a);
    }, r.prototype.pointToData = function(t, n) {
    }, r.prototype.getTicksCoords = function(t) {
      t = t || {};
      var n = t.tickModel || this.getTickModel(), e = E(this, n, {
        breakTicks: t.breakTicks,
        pruneByBreak: t.pruneByBreak
      }), i = e.ticks, a = h(i, function(l) {
        return {
          coord: this.dataToCoord(this.scale.type === "ordinal" ? this.scale.getRawOrdinalNumber(l) : l),
          tickValue: l
        };
      }, this), o = n.get("alignWithLabel");
      return _(this, a, o, t.clamp), a;
    }, r.prototype.getMinorTicksCoords = function() {
      if (this.scale.type === "ordinal")
        return [];
      var t = this.model.getModel("minorTick"), n = t.get("splitNumber");
      n > 0 && n < 100 || (n = 5);
      var e = this.scale.getMinorTicks(n), i = h(e, function(a) {
        return h(a, function(o) {
          return {
            coord: this.dataToCoord(o),
            tickValue: o
          };
        }, this);
      }, this);
      return i;
    }, r.prototype.getViewLabels = function(t) {
      return t = t || x(g.determine), k(this, t).labels;
    }, r.prototype.getLabelModel = function() {
      return this.model.getModel("axisLabel");
    }, r.prototype.getTickModel = function() {
      return this.model.getModel("axisTick");
    }, r.prototype.getBandWidth = function() {
      var t = this._extent, n = this.scale.getExtent(), e = n[1] - n[0] + (this.onBand ? 1 : 0);
      e === 0 && (e = 1);
      var i = Math.abs(t[1] - t[0]);
      return Math.abs(i) / e;
    }, r.prototype.calculateCategoryInterval = function(t) {
      return t = t || x(g.determine), L(this, t);
    }, r;
  })()
);
function y(r, t) {
  var n = r[1] - r[0], e = t, i = n / e / 2;
  r[0] += i, r[1] -= i;
}
function _(r, t, n, e) {
  var i = t.length;
  if (!r.onBand || n || !i)
    return;
  var a = r.getExtent(), o, l;
  if (i === 1)
    t[0].coord = a[0], t[0].onBand = !0, o = t[1] = {
      coord: a[1],
      tickValue: t[0].tickValue,
      onBand: !0
    };
  else {
    var T = t[i - 1].tickValue - t[0].tickValue, p = (t[i - 1].coord - t[0].coord) / T;
    M(t, function(c) {
      c.coord -= p / 2, c.onBand = !0;
    });
    var d = r.scale.getExtent();
    l = 1 + d[1] - t[i - 1].tickValue, o = {
      coord: t[i - 1].coord + p * l,
      tickValue: d[1] + 1,
      onBand: !0
    }, t.push(o);
  }
  var B = a[0] > a[1];
  u(t[0].coord, a[0]) && (e ? t[0].coord = a[0] : t.shift()), e && u(a[0], t[0].coord) && t.unshift({
    coord: a[0],
    onBand: !0
  }), u(a[1], o.coord) && (e ? o.coord = a[1] : t.pop()), e && u(o.coord, a[1]) && t.push({
    coord: a[1],
    onBand: !0
  });
  function u(c, s) {
    return c = v(c), s = v(s), B ? c > s : c < s;
  }
}
export {
  z as default
};
