// echarts@6.0.0/lib/coord/axisModelCreator.js
import { __extends as h } from "../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import c from "./axisDefault.js";
import { fetchLayoutMode as M, getLayoutParams as A, mergeLayoutParam as v } from "../util/layout.js";
import x from "../data/OrdinalMeta.js";
import { AXIS_TYPES as _ } from "./axisCommonTypes.js";
import { each as O, merge as i } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { getAxisBreakHelper as k } from "../component/axis/axisBreakHelper.js";
function T(o, n, f, l) {
  O(_, function(B, a) {
    var d = i(i({}, c[a], !0), l, !0), m = (
      /** @class */
      (function(u) {
        h(e, u);
        function e() {
          var t = u !== null && u.apply(this, arguments) || this;
          return t.type = n + "Axis." + a, t;
        }
        return e.prototype.mergeDefaultAndTheme = function(t, r) {
          var p = M(this), y = p ? A(t) : {}, g = r.getTheme();
          i(t, g.get(a + "Axis")), i(t, this.getDefaultOption()), t.type = s(t), p && v(t, y, p);
        }, e.prototype.optionUpdated = function() {
          var t = this.option;
          t.type === "category" && (this.__ordinalMeta = x.createByAxisModel(this));
        }, e.prototype.getCategories = function(t) {
          var r = this.option;
          if (r.type === "category")
            return t ? r.data : this.__ordinalMeta.categories;
        }, e.prototype.getOrdinalMeta = function() {
          return this.__ordinalMeta;
        }, e.prototype.updateAxisBreaks = function(t) {
          var r = k();
          return r ? r.updateModelAxisBreak(this, t) : {
            breaks: []
          };
        }, e.type = n + "Axis." + a, e.defaultOption = d, e;
      })(f)
    );
    o.registerComponentModel(m);
  }), o.registerSubTypeDefaulter(n + "Axis", s);
}
function s(o) {
  return o.type || (o.data ? "category" : "value");
}
export {
  T as default
};
