// echarts@6.0.0/lib/coord/axisHelper.js
import { each as m, keys as y, isString as M, isFunction as d } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import w from "../scale/Ordinal.js";
import h from "../scale/Interval.js";
import E from "../scale/Scale.js";
import { prepareLayoutBarSeries as L, makeColumnLayout as O, retrieveColumnLayout as I } from "../layout/barGrid.js";
import k from "../scale/Time.js";
import S from "../scale/Log.js";
import { getStackedDimension as A } from "../data/helper/dataStackHelper.js";
import { ensureScaleRawExtentInfo as F } from "./scaleRawExtentInfo.js";
import { parseTimeAxisLabelFormatter as C } from "../util/time.js";
import { getScaleBreakHelper as b } from "../scale/break.js";
import { error as p } from "../util/log.js";
function N(e, t) {
  var n = e.type, a = F(e, t, e.getExtent()).calculate();
  e.setBlank(a.isBlank);
  var r = a.min, i = a.max, o = t.ecModel;
  if (o && n === "time") {
    var l = L("bar", o), u = !1;
    if (m(l, function(s) {
      u = u || s.getBaseAxis() === t.axis;
    }), u) {
      var v = O(l), f = B(r, i, t, v);
      r = f.min, i = f.max;
    }
  }
  return {
    extent: [r, i],
    // "fix" means "fixed", the value should not be
    // changed in the subsequent steps.
    fixMin: a.minFixed,
    fixMax: a.maxFixed
  };
}
function B(e, t, n, a) {
  var r = n.axis.getExtent(), i = Math.abs(r[1] - r[0]), o = I(a, n.axis);
  if (o === void 0)
    return {
      min: e,
      max: t
    };
  var l = 1 / 0;
  m(o, function(c) {
    l = Math.min(c.offset, l);
  });
  var u = -1 / 0;
  m(o, function(c) {
    u = Math.max(c.offset + c.width, u);
  }), l = Math.abs(l), u = Math.abs(u);
  var v = l + u, f = t - e, s = 1 - (l + u) / i, g = f / s - f;
  return t += g * (u / v), e -= g * (l / v), {
    min: e,
    max: t
  };
}
function Y(e, t) {
  var n = t, a = N(e, n), r = a.extent, i = n.get("splitNumber");
  e instanceof S && (e.base = n.get("logBase"));
  var o = e.type, l = n.get("interval"), u = o === "interval" || o === "time";
  e.setBreaksFromOption(T(n)), e.setExtent(r[0], r[1]), e.calcNiceExtent({
    splitNumber: i,
    fixMin: a.fixMin,
    fixMax: a.fixMax,
    minInterval: u ? n.get("minInterval") : null,
    maxInterval: u ? n.get("maxInterval") : null
  }), l != null && e.setInterval && e.setInterval(l);
}
function $(e, t) {
  if (t = t || e.get("type"), t)
    switch (t) {
      // Buildin scale
      case "category":
        return new w({
          ordinalMeta: e.getOrdinalMeta ? e.getOrdinalMeta() : e.getCategories(),
          extent: [1 / 0, -1 / 0]
        });
      case "time":
        return new k({
          locale: e.ecModel.getLocaleModel(),
          useUTC: e.ecModel.get("useUTC")
        });
      default:
        return new (E.getClass(t) || h)();
    }
}
function _(e) {
  var t = e.scale.getExtent(), n = t[0], a = t[1];
  return !(n > 0 && a > 0 || n < 0 && a < 0);
}
function H(e) {
  var t = e.getLabelModel().get("formatter");
  if (e.type === "time") {
    var n = C(t);
    return function(r, i) {
      return e.scale.getFormattedLabel(r, i, n);
    };
  } else {
    if (M(t))
      return function(r) {
        var i = e.scale.getLabel(r), o = t.replace("{value}", i ?? "");
        return o;
      };
    if (d(t)) {
      if (e.type === "category")
        return function(r, i) {
          return t(
            x(e, r),
            r.value - e.scale.getExtent()[0],
            null
            // Using `null` just for backward compat.
          );
        };
      var a = b();
      return function(r, i) {
        var o = null;
        return a && (o = a.makeAxisLabelFormatterParamBreak(o, r.break)), t(x(e, r), i, o);
      };
    } else
      return function(r) {
        return e.scale.getLabel(r);
      };
  }
}
function x(e, t) {
  return e.type === "category" ? e.scale.getLabel(t) : t.value;
}
function D(e) {
  var t = e.get("interval");
  return t ?? "auto";
}
function ee(e) {
  return e.type === "category" && D(e.getLabelModel()) === 0;
}
function R(e, t) {
  var n = {};
  return m(e.mapDimensionsAll(t), function(a) {
    n[A(e, a)] = !0;
  }), y(n);
}
function te(e, t, n) {
  t && m(R(t, n), function(a) {
    var r = t.getApproximateExtent(a);
    r[0] < e[0] && (e[0] = r[0]), r[1] > e[1] && (e[1] = r[1]);
  });
}
function re(e) {
  return e === "middle" || e === "center";
}
function ne(e) {
  return e.getShallow("show");
}
function T(e) {
  var t = e.get("breaks", !0);
  if (t != null) {
    if (!b()) {
      process.env.NODE_ENV !== "production" && p('Must `import {AxisBreak} from "echarts/features.js"; use(AxisBreak);` first if using breaks option.');
      return;
    }
    if (!V(e.axis)) {
      process.env.NODE_ENV !== "production" && p("Axis '" + e.axis.dim + "'-'" + e.axis.type + "' does not support break.");
      return;
    }
    return t;
  }
}
function V(e) {
  return (e.dim === "x" || e.dim === "y" || e.dim === "z" || e.dim === "single") && e.type !== "category";
}
export {
  $ as createScaleByModel,
  x as getAxisRawValue,
  R as getDataDimensionsOnAxis,
  D as getOptionCategoryInterval,
  N as getScaleExtent,
  _ as ifAxisCrossZero,
  re as isNameLocationCenter,
  H as makeLabelFormatter,
  Y as niceScaleExtent,
  T as retrieveAxisBreaksOption,
  ne as shouldAxisShow,
  ee as shouldShowAllLabels,
  te as unionAxisExtentFromData
};
