// echarts@6.0.0/lib/coord/axisTickLabelBuilder.js
import { filter as V, map as d, bind as W, isFunction as F, each as j } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { getBoundingRect as G } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/contain/text.js";
import { makeInner as N } from "../util/model.js";
import { makeLabelFormatter as h, getOptionCategoryInterval as R, shouldShowAllLabels as K } from "./axisHelper.js";
var q = N(), g = N(), y = {
  estimate: 1,
  determine: 2
};
function J(e) {
  return {
    out: {
      noPxChangeTryDetermine: []
    },
    kind: e
  };
}
function D(e, r) {
  var t = d(r, function(a) {
    return e.scale.parse(a);
  });
  return e.type === "time" && t.length > 0 && (t.sort(), t.unshift(t[0]), t.push(t[t.length - 1])), t;
}
function ne(e, r) {
  var t = e.getLabelModel().get("customValues");
  if (t) {
    var a = h(e), n = e.scale.getExtent(), l = D(e, t), o = V(l, function(u) {
      return u >= n[0] && u <= n[1];
    });
    return {
      labels: d(o, function(u) {
        var i = {
          value: u
        };
        return {
          formattedLabel: a(i),
          rawLabel: e.scale.getLabel(i),
          tickValue: u,
          time: void 0,
          break: void 0
        };
      })
    };
  }
  return e.type === "category" ? Q(e, r) : X(e);
}
function le(e, r, t) {
  var a = e.getTickModel().get("customValues");
  if (a) {
    var n = e.scale.getExtent(), l = D(e, a);
    return {
      ticks: V(l, function(o) {
        return o >= n[0] && o <= n[1];
      })
    };
  }
  return e.type === "category" ? U(e, r) : {
    ticks: d(e.scale.getTicks(t), function(o) {
      return o.value;
    })
  };
}
function Q(e, r) {
  var t = e.getLabelModel(), a = P(e, t, r);
  return !t.get("show") || e.scale.isBlank() ? {
    labels: []
  } : a;
}
function P(e, r, t) {
  var a = Z(e), n = R(r), l = t.kind === y.estimate;
  if (!l) {
    var o = B(a, n);
    if (o)
      return o;
  }
  var u, i;
  F(n) ? u = O(e, n) : (i = n === "auto" ? $(e, t) : n, u = z(e, i));
  var v = {
    labels: u,
    labelCategoryInterval: i
  };
  return l ? t.out.noPxChangeTryDetermine.push(function() {
    return T(a, n, v), !0;
  }) : T(a, n, v), v;
}
function U(e, r) {
  var t = Y(e), a = R(r), n = B(t, a);
  if (n)
    return n;
  var l, o;
  if ((!r.get("show") || e.scale.isBlank()) && (l = []), F(a))
    l = O(e, a, !0);
  else if (a === "auto") {
    var u = P(e, e.getLabelModel(), J(y.determine));
    o = u.labelCategoryInterval, l = d(u.labels, function(i) {
      return i.tickValue;
    });
  } else
    o = a, l = z(e, o, !0);
  return T(t, a, {
    ticks: l,
    tickCategoryInterval: o
  });
}
function X(e) {
  var r = e.scale.getTicks(), t = h(e);
  return {
    labels: d(r, function(a, n) {
      return {
        formattedLabel: t(a, n),
        rawLabel: e.scale.getLabel(a),
        tickValue: a.value,
        time: a.time,
        break: a.break
      };
    })
  };
}
var Y = S("axisTick"), Z = S("axisLabel");
function S(e) {
  return function(t) {
    return g(t)[e] || (g(t)[e] = {
      list: []
    });
  };
}
function B(e, r) {
  for (var t = 0; t < e.list.length; t++)
    if (e.list[t].key === r)
      return e.list[t].value;
}
function T(e, r, t) {
  return e.list.push({
    key: r,
    value: t
  }), t;
}
function $(e, r) {
  if (r.kind === y.estimate) {
    var t = e.calculateCategoryInterval(r);
    return r.out.noPxChangeTryDetermine.push(function() {
      return g(e).autoInterval = t, !0;
    }), t;
  }
  var a = g(e).autoInterval;
  return a ?? (g(e).autoInterval = e.calculateCategoryInterval(r));
}
function oe(e, r) {
  var t = r.kind, a = x(e), n = h(e), l = (a.axisRotate - a.labelRotate) / 180 * Math.PI, o = e.scale, u = o.getExtent(), i = o.count();
  if (u[1] - u[0] < 1)
    return 0;
  var v = 1, k = 40;
  i > k && (v = Math.max(1, Math.floor(i / k)));
  for (var c = u[0], C = e.dataToCoord(c + 1) - e.dataToCoord(c), I = Math.abs(C * Math.cos(l)), m = Math.abs(C * Math.sin(l)), s = 0, f = 0; c <= u[1]; c += v) {
    var b = 0, A = 0, E = G(n({
      value: c
    }), a.font, "center", "top");
    b = E.width * 1.3, A = E.height * 1.3, s = Math.max(s, b, 7), f = Math.max(f, A, 7);
  }
  var M = s / I, p = f / m;
  isNaN(M) && (M = 1 / 0), isNaN(p) && (p = 1 / 0);
  var L = Math.max(0, Math.floor(Math.min(M, p)));
  if (t === y.estimate)
    return r.out.noPxChangeTryDetermine.push(W(_, null, e, L, i)), L;
  var w = H(e, L, i);
  return w ?? L;
}
function _(e, r, t) {
  return H(e, r, t) == null;
}
function H(e, r, t) {
  var a = q(e.model), n = e.getExtent(), l = a.lastAutoInterval, o = a.lastTickCount;
  if (l != null && o != null && Math.abs(l - r) <= 1 && Math.abs(o - t) <= 1 && l > r && a.axisExtent0 === n[0] && a.axisExtent1 === n[1])
    return l;
  a.lastTickCount = t, a.lastAutoInterval = r, a.axisExtent0 = n[0], a.axisExtent1 = n[1];
}
function x(e) {
  var r = e.getLabelModel();
  return {
    axisRotate: e.getRotate ? e.getRotate() : e.isHorizontal && !e.isHorizontal() ? 90 : 0,
    labelRotate: r.get("rotate") || 0,
    font: r.getFont()
  };
}
function z(e, r, t) {
  var a = h(e), n = e.scale, l = n.getExtent(), o = e.getLabelModel(), u = [], i = Math.max((r || 0) + 1, 1), v = l[0], k = n.count();
  v !== 0 && i > 1 && k / i > 2 && (v = Math.round(Math.ceil(v / i) * i));
  var c = K(e), C = o.get("showMinLabel") || c, I = o.get("showMaxLabel") || c;
  C && v !== l[0] && s(l[0]);
  for (var m = v; m <= l[1]; m += i)
    s(m);
  I && m - i !== l[1] && s(l[1]);
  function s(f) {
    var b = {
      value: f
    };
    u.push(t ? f : {
      formattedLabel: a(b),
      rawLabel: n.getLabel(b),
      tickValue: f,
      time: void 0,
      break: void 0
    });
  }
  return u;
}
function O(e, r, t) {
  var a = e.scale, n = h(e), l = [];
  return j(a.getTicks(), function(o) {
    var u = a.getLabel(o), i = o.value;
    r(o.value, u) && l.push(t ? i : {
      formattedLabel: n(o),
      rawLabel: u,
      tickValue: i,
      time: void 0,
      break: void 0
    });
  }), l;
}
export {
  y as AxisTickLabelComputingKind,
  oe as calculateCategoryInterval,
  ne as createAxisLabels,
  J as createAxisLabelsComputingContext,
  le as createAxisTicks
};
