// echarts@6.0.0/lib/coord/axisAlignTicks.js
import { round as p, getPrecisionSafe as h } from "../util/number.js";
import b from "../scale/Interval.js";
import { getScaleExtent as w, retrieveAxisBreaksOption as S } from "./axisHelper.js";
import { warn as y } from "../util/log.js";
import { logTransform as O, increaseInterval as u, isValueNice as A } from "../scale/helper.js";
function D(i, v, f) {
  var m, l = b.prototype, c = l.getTicks.call(f), T = l.getTicks.call(f, {
    expandToNicedExtent: !0
  }), n = c.length - 1, E = l.getInterval.call(f), s = w(i, v), t = s.extent, o = s.fixMin, x = s.fixMax;
  i.type === "log" && (t = O(i.base, t, !0)), i.setBreaksFromOption(S(v)), i.setExtent(t[0], t[1]), i.calcNiceExtent({
    splitNumber: n,
    fixMin: o,
    fixMax: x
  });
  var k = l.getExtent.call(i);
  o && (t[0] = k[0]), x && (t[1] = k[1]);
  var e = l.getInterval.call(i), a = t[0], r = t[1];
  if (o && x)
    e = (r - a) / n;
  else if (o)
    for (r = t[0] + e * n; r < t[1] && isFinite(r) && isFinite(t[1]); )
      e = u(e), r = t[0] + e * n;
  else if (x)
    for (a = t[1] - e * n; a > t[0] && isFinite(a) && isFinite(t[0]); )
      e = u(e), a = t[1] - e * n;
  else {
    var F = i.getTicks().length - 1;
    F > n && (e = u(e));
    var g = e * n;
    r = Math.ceil(t[1] / e) * e, a = p(r - g), a < 0 && t[0] >= 0 ? (a = 0, r = p(g)) : r > 0 && t[1] <= 0 && (r = 0, a = -p(g));
  }
  var d = (c[0].value - T[0].value) / E, N = (c[n].value - T[n].value) / E;
  if (l.setExtent.call(i, a + e * d, r + e * N), l.setInterval.call(i, e), (d || N) && l.setNiceExtent.call(i, a + e, r - e), process.env.NODE_ENV !== "production") {
    var I = l.getTicks.call(i);
    I[1] && (!A(e) || h(I[1].value) > h(e)) && y("The ticks may be not readable when set min: " + v.get("min") + ", max: " + v.get("max") + (" and alignTicks: true. (" + ((m = v.axis) === null || m === void 0 ? void 0 : m.dim) + "AxisIndex: " + v.componentIndex + ")"), !0);
  }
}
export {
  D as alignScaleTicks
};
