// echarts@6.0.0/lib/coord/scaleRawExtentInfo.js
import { isFunction as h, isArray as v, eqNaN as m, assert as p } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { parsePercent as x } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/contain/text.js";
var c = (
  /** @class */
  (function() {
    function r(e, a, i) {
      this._prepareParams(e, a, i);
    }
    return r.prototype._prepareParams = function(e, a, i) {
      i[1] < i[0] && (i = [NaN, NaN]), this._dataMin = i[0], this._dataMax = i[1];
      var l = this._isOrdinal = e.type === "ordinal";
      this._needCrossZero = e.type === "interval" && a.getNeedCrossZero && a.getNeedCrossZero();
      var u = a.get("min", !0);
      u == null && (u = a.get("startValue", !0));
      var o = this._modelMinRaw = u;
      h(o) ? this._modelMinNum = d(e, o({
        min: i[0],
        max: i[1]
      })) : o !== "dataMin" && (this._modelMinNum = d(e, o));
      var n = this._modelMaxRaw = a.get("max", !0);
      if (h(n) ? this._modelMaxNum = d(e, n({
        min: i[0],
        max: i[1]
      })) : n !== "dataMax" && (this._modelMaxNum = d(e, n)), l)
        this._axisDataLen = a.getCategories().length;
      else {
        var t = a.get("boundaryGap"), s = v(t) ? t : [t || 0, t || 0];
        typeof s[0] == "boolean" || typeof s[1] == "boolean" ? (process.env.NODE_ENV !== "production" && console.warn('Boolean type for boundaryGap is only allowed for ordinal axis. Please use string in percentage instead, e.g., "20%". Currently, boundaryGap is set to be 0.'), this._boundaryGapInner = [0, 0]) : this._boundaryGapInner = [x(s[0], 1), x(s[1], 1)];
      }
    }, r.prototype.calculate = function() {
      var e = this._isOrdinal, a = this._dataMin, i = this._dataMax, l = this._axisDataLen, u = this._boundaryGapInner, o = e ? null : i - a || Math.abs(a), n = this._modelMinRaw === "dataMin" ? a : this._modelMinNum, t = this._modelMaxRaw === "dataMax" ? i : this._modelMaxNum, s = n != null, f = t != null;
      n == null && (n = e ? l ? 0 : NaN : a - u[0] * o), t == null && (t = e ? l ? l - 1 : NaN : i + u[1] * o), (n == null || !isFinite(n)) && (n = NaN), (t == null || !isFinite(t)) && (t = NaN);
      var N = m(n) || m(t) || e && !l;
      this._needCrossZero && (n > 0 && t > 0 && !s && (n = 0), n < 0 && t < 0 && !f && (t = 0));
      var M = this._determinedMin, _ = this._determinedMax;
      return M != null && (n = M, s = !0), _ != null && (t = _, f = !0), {
        min: n,
        max: t,
        minFixed: s,
        maxFixed: f,
        isBlank: N
      };
    }, r.prototype.modifyDataMinMax = function(e, a) {
      process.env.NODE_ENV !== "production" && p(!this.frozen), this[y[e]] = a;
    }, r.prototype.setDeterminedMinMax = function(e, a) {
      var i = w[e];
      process.env.NODE_ENV !== "production" && p(!this.frozen && this[i] == null), this[i] = a;
    }, r.prototype.freeze = function() {
      this.frozen = !0;
    }, r;
  })()
), w = {
  min: "_determinedMin",
  max: "_determinedMax"
}, y = {
  min: "_dataMin",
  max: "_dataMax"
};
function R(r, e, a) {
  var i = r.rawExtentInfo;
  return i || (i = new c(r, e, a), r.rawExtentInfo = i, i);
}
function d(r, e) {
  return e == null ? null : m(e) ? NaN : r.parse(e);
}
export {
  c as ScaleRawExtentInfo,
  R as ensureScaleRawExtentInfo,
  d as parseAxisModelMinMax
};
