// echarts@6.0.0/lib/coord/View.js
import { __extends as T } from "../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { applyTransform as l, copy as y } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/vector.js";
import { copy as v, create as d, invert as _ } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/matrix.js";
import p from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/BoundingRect.js";
import m from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/Transformable.js";
import { parsePercent as f } from "../util/number.js";
import { clone as g } from "../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { clampByZoomLimit as w } from "../component/helper/roamHelper.js";
var h = l, P = (
  /** @class */
  (function(i) {
    T(o, i);
    function o(r, e) {
      var t = i.call(this) || this;
      return t.type = "view", t.dimensions = ["x", "y"], t._roamTransformable = new m(), t._rawTransformable = new m(), t.name = r, t._opt = e, t;
    }
    return o.prototype.setBoundingRect = function(r, e, t, n) {
      return this._rect = new p(r, e, t, n), this._updateCenterAndZoom(), this._rect;
    }, o.prototype.getBoundingRect = function() {
      return this._rect;
    }, o.prototype.setViewRect = function(r, e, t, n) {
      this._transformTo(r, e, t, n), this._viewRect = new p(r, e, t, n);
    }, o.prototype._transformTo = function(r, e, t, n) {
      var s = this.getBoundingRect(), a = this._rawTransformable;
      a.transform = s.calculateTransform(new p(r, e, t, n));
      var c = a.parent;
      a.parent = null, a.decomposeTransform(), a.parent = c, this._updateTransform();
    }, o.prototype.setCenter = function(r) {
      var e = this._opt;
      e && e.api && e.ecModel && e.ecModel.getShallow("legacyViewCoordSysCenterBase") && r && (r = [f(r[0], e.api.getWidth()), f(r[1], e.api.getWidth())]), this._centerOption = g(r), this._updateCenterAndZoom();
    }, o.prototype.setZoom = function(r) {
      this._zoom = w(r || 1, this.zoomLimit), this._updateCenterAndZoom();
    }, o.prototype.getDefaultCenter = function() {
      var r = this.getBoundingRect(), e = r.x + r.width / 2, t = r.y + r.height / 2;
      return [e, t];
    }, o.prototype.getCenter = function() {
      return this._center || this.getDefaultCenter();
    }, o.prototype.getZoom = function() {
      return this._zoom || 1;
    }, o.prototype.getRoamTransform = function() {
      return this._roamTransformable.getLocalTransform();
    }, o.prototype._updateCenterAndZoom = function() {
      var r = this._centerOption, e = this._rect;
      r && e && (this._center = [f(r[0], e.width, e.x), f(r[1], e.height, e.y)]);
      var t = this._rawTransformable.getLocalTransform(), n = this._roamTransformable, s = this.getDefaultCenter(), a = this.getCenter(), c = this.getZoom();
      a = l([], a, t), s = l([], s, t), n.originX = a[0], n.originY = a[1], n.x = s[0] - a[0], n.y = s[1] - a[1], n.scaleX = n.scaleY = c, this._updateTransform();
    }, o.prototype._updateTransform = function() {
      var r = this._roamTransformable, e = this._rawTransformable;
      e.parent = r, r.updateTransform(), e.updateTransform(), v(this.transform || (this.transform = []), e.transform || d()), this._rawTransform = e.getLocalTransform(), this.invTransform = this.invTransform || [], _(this.invTransform, this.transform), this.decomposeTransform();
    }, o.prototype.getTransformInfo = function() {
      var r = this._rawTransformable, e = this._roamTransformable, t = new m();
      return t.transform = e.transform, t.decomposeTransform(), {
        roam: {
          x: t.x,
          y: t.y,
          scaleX: t.scaleX,
          scaleY: t.scaleY
        },
        raw: {
          x: r.x,
          y: r.y,
          scaleX: r.scaleX,
          scaleY: r.scaleY
        }
      };
    }, o.prototype.getViewRect = function() {
      return this._viewRect;
    }, o.prototype.getViewRectAfterRoam = function() {
      var r = this.getBoundingRect().clone();
      return r.applyTransform(this.transform), r;
    }, o.prototype.dataToPoint = function(r, e, t) {
      var n = e ? this._rawTransform : this.transform;
      return t = t || [], n ? h(t, r, n) : y(t, r);
    }, o.prototype.pointToData = function(r, e, t) {
      t = t || [];
      var n = this.invTransform;
      return n ? h(t, r, n) : (t[0] = r[0], t[1] = r[1], t);
    }, o.prototype.convertToPixel = function(r, e, t) {
      var n = u(e);
      return n === this ? n.dataToPoint(t) : null;
    }, o.prototype.convertFromPixel = function(r, e, t) {
      var n = u(e);
      return n === this ? n.pointToData(t) : null;
    }, o.prototype.containPoint = function(r) {
      return this.getViewRectAfterRoam().contain(r[0], r[1]);
    }, o.dimensions = ["x", "y"], o;
  })(m)
);
function u(i) {
  var o = i.seriesModel;
  return o ? o.coordinateSystem : null;
}
export {
  P as default
};
