// @tiptap/extension-color@2.11.7/dist/index.js
import "../../../../../@tiptap_extension-text-style@2.11.7_@tiptap_core@2.24.0_@tiptap_pm@2.24.0_/node_modules/@tiptap/extension-text-style/dist/index.js";
import { Extension as e } from "../../../../../@tiptap_core@2.24.0_@tiptap_pm@2.24.0/node_modules/@tiptap/core/dist/index.js";
const n = e.create({
  name: "color",
  addOptions() {
    return {
      types: ["textStyle"]
    };
  },
  addGlobalAttributes() {
    return [
      {
        types: this.options.types,
        attributes: {
          color: {
            default: null,
            parseHTML: (t) => {
              var r;
              return (r = t.style.color) === null || r === void 0 ? void 0 : r.replace(/['"]+/g, "");
            },
            renderHTML: (t) => t.color ? {
              style: `color: ${t.color}`
            } : {}
          }
        }
      }
    ];
  },
  addCommands() {
    return {
      setColor: (t) => ({ chain: r }) => r().setMark("textStyle", { color: t }).run(),
      unsetColor: () => ({ chain: t }) => t().setMark("textStyle", { color: null }).removeEmptyTextStyle().run()
    };
  }
});
export {
  n as Color,
  n as default
};
