// @atlaskit/pragmatic-drag-and-drop@1.7.4/dist/esm/make-adapter/make-drop-target.js
import U from "../../../../../../../@babel_runtime@7.28.6/node_modules/@babel/runtime/helpers/esm/defineProperty.js";
import V from "../../../../../../../@babel_runtime@7.28.6/node_modules/@babel/runtime/helpers/esm/toConsumableArray.js";
import { combine as Y } from "../public-utils/combine.js";
import { once as q } from "../public-utils/once.js";
import { addAttribute as z } from "../util/add-attribute.js";
function M(e, o) {
  var t = Object.keys(e);
  if (Object.getOwnPropertySymbols) {
    var a = Object.getOwnPropertySymbols(e);
    o && (a = a.filter(function(T) {
      return Object.getOwnPropertyDescriptor(e, T).enumerable;
    })), t.push.apply(t, a);
  }
  return t;
}
function E(e) {
  for (var o = 1; o < arguments.length; o++) {
    var t = arguments[o] != null ? arguments[o] : {};
    o % 2 ? M(Object(t), !0).forEach(function(a) {
      U(e, a, t[a]);
    }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : M(Object(t)).forEach(function(a) {
      Object.defineProperty(e, a, Object.getOwnPropertyDescriptor(t, a));
    });
  }
  return e;
}
function I(e, o) {
  var t = typeof Symbol < "u" && e[Symbol.iterator] || e["@@iterator"];
  if (!t) {
    if (Array.isArray(e) || (t = B(e)) || o) {
      t && (e = t);
      var a = 0, T = function() {
      };
      return { s: T, n: function() {
        return a >= e.length ? { done: !0 } : { done: !1, value: e[a++] };
      }, e: function(b) {
        throw b;
      }, f: T };
    }
    throw new TypeError(`Invalid attempt to iterate non-iterable instance.
In order to be iterable, non-array objects must have a [Symbol.iterator]() method.`);
  }
  var S, k = !0, C = !1;
  return { s: function() {
    t = t.call(e);
  }, n: function() {
    var b = t.next();
    return k = b.done, b;
  }, e: function(b) {
    C = !0, S = b;
  }, f: function() {
    try {
      k || t.return == null || t.return();
    } finally {
      if (C) throw S;
    }
  } };
}
function B(e, o) {
  if (e) {
    if (typeof e == "string") return H(e, o);
    var t = {}.toString.call(e).slice(8, -1);
    return t === "Object" && e.constructor && (t = e.constructor.name), t === "Map" || t === "Set" ? Array.from(e) : t === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? H(e, o) : void 0;
  }
}
function H(e, o) {
  (o == null || o > e.length) && (o = e.length);
  for (var t = 0, a = Array(o); t < o; t++) a[t] = e[t];
  return a;
}
function N(e) {
  return e.slice(0).reverse();
}
function F(e) {
  var o = e.typeKey, t = e.defaultDropEffect, a = /* @__PURE__ */ new WeakMap(), T = "data-drop-target-for-".concat(o), S = "[".concat(T, "]");
  function k(r) {
    return a.set(r.element, r), function() {
      return a.delete(r.element);
    };
  }
  function C(r) {
    if (process.env.NODE_ENV !== "production") {
      var p = a.get(r.element);
      p && console.warn("You have already registered a [".concat(o, "] dropTarget on the same element"), {
        existing: p,
        proposed: r
      }), r.element instanceof HTMLIFrameElement && console.warn(`
            We recommend not registering <iframe> elements as drop targets
            as it can result in some strange browser event ordering.
          `.replace(/\s{2,}/g, " ").trim());
    }
    var i = Y(z(r.element, {
      attribute: T,
      value: "true"
    }), k(r));
    return q(i);
  }
  function g(r) {
    var p, i, s, d, l = r.source, c = r.target, v = r.input, f = r.result, n = f === void 0 ? [] : f;
    if (c == null)
      return n;
    if (!(c instanceof Element))
      return c instanceof Node ? g({
        source: l,
        target: c.parentElement,
        input: v,
        result: n
      }) : n;
    var y = c.closest(S);
    if (y == null)
      return n;
    var u = a.get(y);
    if (u == null)
      return n;
    var h = {
      input: v,
      source: l,
      element: u.element
    };
    if (u.canDrop && !u.canDrop(h))
      return g({
        source: l,
        target: u.element.parentElement,
        input: v,
        result: n
      });
    var O = (p = (i = u.getData) === null || i === void 0 ? void 0 : i.call(u, h)) !== null && p !== void 0 ? p : {}, D = (s = (d = u.getDropEffect) === null || d === void 0 ? void 0 : d.call(u, h)) !== null && s !== void 0 ? s : t, m = {
      data: O,
      element: u.element,
      dropEffect: D,
      // we are collecting _actual_ drop targets, so these are
      // being applied _not_ due to stickiness
      isActiveDueToStickiness: !1
    };
    return g({
      source: l,
      target: u.element.parentElement,
      input: v,
      // Using bubble ordering. Same ordering as `event.getPath()`
      result: [].concat(V(n), [m])
    });
  }
  function b(r) {
    var p = r.eventName, i = r.payload, s = I(i.location.current.dropTargets), d;
    try {
      for (s.s(); !(d = s.n()).done; ) {
        var l, c = d.value, v = a.get(c.element), f = E(E({}, i), {}, {
          self: c
        });
        v == null || (l = v[p]) === null || l === void 0 || l.call(
          v,
          // I cannot seem to get the types right here.
          // TS doesn't seem to like that one event can need `nativeSetDragImage`
          // @ts-expect-error
          f
        );
      }
    } catch (n) {
      s.e(n);
    } finally {
      s.f();
    }
  }
  var R = {
    onGenerateDragPreview: b,
    onDrag: b,
    onDragStart: b,
    onDrop: b,
    onDropTargetChange: function(p) {
      var i = p.payload, s = new Set(i.location.current.dropTargets.map(function(A) {
        return A.element;
      })), d = /* @__PURE__ */ new Set(), l = I(i.location.previous.dropTargets), c;
      try {
        for (l.s(); !(c = l.n()).done; ) {
          var v, f = c.value;
          d.add(f.element);
          var n = a.get(f.element), y = s.has(f.element), u = E(E({}, i), {}, {
            self: f
          });
          if (n == null || (v = n.onDropTargetChange) === null || v === void 0 || v.call(n, u), !y) {
            var h;
            n == null || (h = n.onDragLeave) === null || h === void 0 || h.call(n, u);
          }
        }
      } catch (A) {
        l.e(A);
      } finally {
        l.f();
      }
      var O = I(i.location.current.dropTargets), D;
      try {
        for (O.s(); !(D = O.n()).done; ) {
          var m, j, P = D.value;
          if (!d.has(P.element)) {
            var K = E(E({}, i), {}, {
              self: P
            }), w = a.get(P.element);
            w == null || (m = w.onDropTargetChange) === null || m === void 0 || m.call(w, K), w == null || (j = w.onDragEnter) === null || j === void 0 || j.call(w, K);
          }
        }
      } catch (A) {
        O.e(A);
      } finally {
        O.f();
      }
    }
  };
  function W(r) {
    R[r.eventName](r);
  }
  function G(r) {
    var p = r.source, i = r.target, s = r.input, d = r.current, l = g({
      source: p,
      target: i,
      input: s
    });
    if (l.length >= d.length)
      return l;
    for (var c = N(d), v = N(l), f = [], n = 0; n < c.length; n++) {
      var y, u = c[n], h = v[n];
      if (h != null) {
        f.push(h);
        continue;
      }
      var O = f[n - 1], D = c[n - 1];
      if (O?.element !== D?.element)
        break;
      var m = a.get(u.element);
      if (!m)
        break;
      var j = {
        input: s,
        source: p,
        element: m.element
      };
      if (m.canDrop && !m.canDrop(j) || !((y = m.getIsSticky) !== null && y !== void 0 && y.call(m, j)))
        break;
      f.push(E(E({}, u), {}, {
        // making it clear to consumers this drop target is active due to stickiness
        isActiveDueToStickiness: !0
      }));
    }
    return N(f);
  }
  return {
    dropTargetForConsumers: C,
    getIsOver: G,
    dispatchEvent: W
  };
}
export {
  F as makeDropTarget
};
