// @atlaskit/pragmatic-drag-and-drop@1.7.4/dist/esm/make-adapter/make-adapter.js
import { lifecycle as c } from "../ledger/lifecycle-manager.js";
import { register as g } from "../ledger/usage-ledger.js";
import { makeDropTarget as y } from "./make-drop-target.js";
import { makeMonitor as T } from "./make-monitor.js";
function A(e) {
  var p = e.typeKey, s = e.mount, r = e.dispatchEventToSource, o = e.onPostDispatch, u = e.defaultDropEffect, i = T(), n = y({
    typeKey: p,
    defaultDropEffect: u
  });
  function m(t) {
    r?.(t), n.dispatchEvent(t), i.dispatchEvent(t), o?.(t);
  }
  function d(t) {
    var a = t.event, f = t.dragType;
    c.start({
      event: a,
      dragType: f,
      getDropTargetsOver: n.getIsOver,
      dispatchEvent: m
    });
  }
  function v() {
    function t() {
      var a = {
        canStart: c.canStart,
        start: d
      };
      return s(a);
    }
    return g({
      typeKey: p,
      mount: t
    });
  }
  return {
    registerUsage: v,
    dropTarget: n.dropTargetForConsumers,
    monitor: i.monitorForConsumers
  };
}
export {
  A as makeAdapter
};
