// @atlaskit/pragmatic-drag-and-drop@1.7.4/dist/esm/make-adapter/make-monitor.js
import O from "../../../../../../../@babel_runtime@7.28.6/node_modules/@babel/runtime/helpers/esm/defineProperty.js";
import { once as h } from "../public-utils/once.js";
function j(e, t) {
  var r = typeof Symbol < "u" && e[Symbol.iterator] || e["@@iterator"];
  if (!r) {
    if (Array.isArray(e) || (r = w(e)) || t) {
      r && (e = r);
      var n = 0, f = function() {
      };
      return { s: f, n: function() {
        return n >= e.length ? { done: !0 } : { done: !1, value: e[n++] };
      }, e: function(c) {
        throw c;
      }, f };
    }
    throw new TypeError(`Invalid attempt to iterate non-iterable instance.
In order to be iterable, non-array objects must have a [Symbol.iterator]() method.`);
  }
  var o, a = !0, u = !1;
  return { s: function() {
    r = r.call(e);
  }, n: function() {
    var c = r.next();
    return a = c.done, c;
  }, e: function(c) {
    u = !0, o = c;
  }, f: function() {
    try {
      a || r.return == null || r.return();
    } finally {
      if (u) throw o;
    }
  } };
}
function w(e, t) {
  if (e) {
    if (typeof e == "string") return p(e, t);
    var r = {}.toString.call(e).slice(8, -1);
    return r === "Object" && e.constructor && (r = e.constructor.name), r === "Map" || r === "Set" ? Array.from(e) : r === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(r) ? p(e, t) : void 0;
  }
}
function p(e, t) {
  (t == null || t > e.length) && (t = e.length);
  for (var r = 0, n = Array(t); r < t; r++) n[r] = e[r];
  return n;
}
function d(e, t) {
  var r = Object.keys(e);
  if (Object.getOwnPropertySymbols) {
    var n = Object.getOwnPropertySymbols(e);
    t && (n = n.filter(function(f) {
      return Object.getOwnPropertyDescriptor(e, f).enumerable;
    })), r.push.apply(r, n);
  }
  return r;
}
function A(e) {
  for (var t = 1; t < arguments.length; t++) {
    var r = arguments[t] != null ? arguments[t] : {};
    t % 2 ? d(Object(r), !0).forEach(function(n) {
      O(e, n, r[n]);
    }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(r)) : d(Object(r)).forEach(function(n) {
      Object.defineProperty(e, n, Object.getOwnPropertyDescriptor(r, n));
    });
  }
  return e;
}
function D() {
  var e = /* @__PURE__ */ new Set(), t = null;
  function r(o) {
    t && (!o.canMonitor || o.canMonitor(t.canMonitorArgs)) && t.active.add(o);
  }
  function n(o) {
    var a = A({}, o);
    e.add(a), r(a);
    function u() {
      e.delete(a), t && t.active.delete(a);
    }
    return h(u);
  }
  function f(o) {
    var a = o.eventName, u = o.payload;
    if (a === "onGenerateDragPreview") {
      t = {
        canMonitorArgs: {
          initial: u.location.initial,
          source: u.source
        },
        active: /* @__PURE__ */ new Set()
      };
      var i = j(e), c;
      try {
        for (i.s(); !(c = i.n()).done; ) {
          var b = c.value;
          r(b);
        }
      } catch (m) {
        i.e(m);
      } finally {
        i.f();
      }
    }
    if (t) {
      for (var g = Array.from(t.active), l = 0, v = g; l < v.length; l++) {
        var s = v[l];
        if (t.active.has(s)) {
          var y;
          (y = s[a]) === null || y === void 0 || y.call(s, u);
        }
      }
      a === "onDrop" && (t.active.clear(), t = null);
    }
  }
  return {
    dispatchEvent: f,
    monitorForConsumers: n
  };
}
export {
  D as makeMonitor
};
