// echarts@6.0.0/lib/chart/sunburst/sunburstLayout.js
import { parsePercent as f } from "../../util/number.js";
import { isArray as q, each as R, isFunction as M, map as G } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
var J = Math.PI / 180;
function at(n, i, u) {
  i.eachSeriesByType(n, function(r) {
    var t = r.get("center"), a = r.get("radius");
    q(a) || (a = [0, a]), q(t) || (t = [t, t]);
    var v = u.getWidth(), L = u.getHeight(), l = Math.min(v, L), N = f(t[0], v), P = f(t[1], L), h = f(a[0], l / 2), Q = f(a[1], l / 2), p = -r.get("startAngle") * J, k = r.get("minAngle") * J, s = r.getData().tree.root, c = r.getViewRoot(), T = c.depth, D = r.get("sort");
    D != null && K(c, D);
    var S = 0;
    R(c.children, function(e) {
      !isNaN(e.getValue()) && S++;
    });
    var C = c.getValue(), Z = Math.PI / (C || S) * 2, x = c.depth > 0, U = c.height - (x ? -1 : 1), y = (Q - h) / (U || 1), d = r.get("clockwise"), X = r.get("stillShowZeroSum"), Y = d ? 1 : -1, z = function(e, g) {
      if (e) {
        var w = g;
        if (e !== s) {
          var E = e.getValue(), o = C === 0 && X ? Z : E * Z;
          o < k && (o = k), w = g + Y * o;
          var F = e.depth - T - (x ? -1 : 1), H = h + y * F, W = h + y * (F + 1), m = r.getLevelModel(e);
          if (m) {
            var I = m.get("r0", !0), V = m.get("r", !0), A = m.get("radius", !0);
            A != null && (I = A[0], V = A[1]), I != null && (H = f(I, l / 2)), V != null && (W = f(V, l / 2));
          }
          e.setLayout({
            angle: o,
            startAngle: g,
            endAngle: w,
            clockwise: d,
            cx: N,
            cy: P,
            r0: H,
            r: W
          });
        }
        if (e.children && e.children.length) {
          var j = 0;
          R(e.children, function(O) {
            j += z(O, g + j);
          });
        }
        return w - g;
      }
    };
    if (x) {
      var $ = h, b = h + y, B = Math.PI * 2;
      s.setLayout({
        angle: B,
        startAngle: p,
        endAngle: p + B,
        clockwise: d,
        cx: N,
        cy: P,
        r0: $,
        r: b
      });
    }
    z(c, p);
  });
}
function K(n, i) {
  var u = n.children || [];
  n.children = _(u, i), u.length && R(n.children, function(r) {
    K(r, i);
  });
}
function _(n, i) {
  if (M(i)) {
    var u = G(n, function(t, a) {
      var v = t.getValue();
      return {
        params: {
          depth: t.depth,
          height: t.height,
          dataIndex: t.dataIndex,
          getValue: function() {
            return v;
          }
        },
        index: a
      };
    });
    return u.sort(function(t, a) {
      return i(t.params, a.params);
    }), G(u, function(t) {
      return n[t.index];
    });
  } else {
    var r = i === "asc";
    return n.sort(function(t, a) {
      var v = (t.getValue() - a.getValue()) * (r ? 1 : -1);
      return v === 0 ? (t.dataIndex - a.dataIndex) * (r ? -1 : 1) : v;
    });
  }
}
export {
  at as default
};
