// echarts@6.0.0/lib/chart/sunburst/SunburstPiece.js
import { __extends as H } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { extend as F, each as O, concatArray as J, isNumber as X } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { SPECIAL_STATES as Y, toggleHoverEmphasis as j, DISPLAY_STATES as k } from "../../util/states.js";
import { createTextStyle as q } from "../../label/labelStyle.js";
import { getECData as Z } from "../../util/innerStore.js";
import { getSectorCornerRadius as z } from "../helper/sectorHelper.js";
import { createOrUpdatePatternFromDecal as B } from "../../util/decal.js";
import { initProps as G, updateProps as K, saveOldStyle as Q } from "../../animation/basicTransition.js";
import { normalizeRadian as _ } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/contain/util.js";
import { isRadianAroundZero as W } from "../../util/number.js";
import $ from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Text.js";
import N from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Sector.js";
var ee = 2, te = 4, he = (
  /** @class */
  (function(E) {
    H(A, E);
    function A(m, t, l, c) {
      var e = E.call(this) || this;
      e.z2 = ee, e.textConfig = {
        inside: !0
      }, Z(e).seriesIndex = t.seriesIndex;
      var r = new $({
        z2: te,
        silent: m.getModel().get(["label", "silent"])
      });
      return e.setTextContent(r), e.updateData(!0, m, t, l, c), e;
    }
    return A.prototype.updateData = function(m, t, l, c, e) {
      this.node = t, t.piece = this, l = l || this._seriesModel, c = c || this._ecModel;
      var r = this;
      Z(r).dataIndex = t.dataIndex;
      var i = t.getModel(), p = i.getModel("emphasis"), S = t.getLayout(), s = F({}, S);
      s.label = null;
      var u = t.getVisual("style");
      u.lineJoin = "bevel";
      var I = t.getVisual("decal");
      I && (u.decal = B(I, e));
      var b = z(i.getModel("itemStyle"), s, !0);
      F(s, b), O(Y, function(a) {
        var g = r.ensureState(a), n = i.getModel([a, "itemStyle"]);
        g.style = n.getItemStyle();
        var h = z(n, s);
        h && (g.shape = h);
      }), m ? (r.setShape(s), r.shape.r = S.r0, G(r, {
        shape: {
          r: S.r
        }
      }, l, t.dataIndex)) : (K(r, {
        shape: s
      }, l), Q(r)), r.useStyle(u), this._updateLabel(l);
      var x = i.getShallow("cursor");
      x && r.attr("cursor", x), this._seriesModel = l || this._seriesModel, this._ecModel = c || this._ecModel;
      var d = p.get("focus"), o = d === "relative" ? J(t.getAncestorsIndices(), t.getDescendantIndices()) : d === "ancestor" ? t.getAncestorsIndices() : d === "descendant" ? t.getDescendantIndices() : d;
      j(this, o, p.get("blurScope"), p.get("disabled"));
    }, A.prototype._updateLabel = function(m) {
      var t = this, l = this.node.getModel(), c = l.getModel("label"), e = this.node.getLayout(), r = e.endAngle - e.startAngle, i = (e.startAngle + e.endAngle) / 2, p = Math.cos(i), S = Math.sin(i), s = this, u = s.getTextContent(), I = this.node.dataIndex, b = c.get("minAngle") / 180 * Math.PI, x = c.get("show") && !(b != null && Math.abs(r) < b);
      u.ignore = !x, O(k, function(o) {
        var a = o === "normal" ? l.getModel("label") : l.getModel([o, "label"]), g = o === "normal", n = g ? u : u.ensureState(o), h = m.getFormattedLabel(I, o);
        g && (h = h || t.node.name), n.style = q(a, {}, null, o !== "normal", !0), h && (n.style.text = h);
        var C = a.get("show");
        C != null && !g && (n.ignore = !C);
        var D = d(a, "position"), R = g ? s : s.states[o], U = R.style.fill;
        R.textConfig = {
          outsideFill: a.get("color") === "inherit" ? U : null,
          inside: D !== "outside"
        };
        var f, T = d(a, "distance") || 0, v = d(a, "align"), y = d(a, "rotate"), w = Math.PI * 0.5, V = Math.PI * 1.5, L = _(y === "tangential" ? Math.PI / 2 - i : i), M = L > w && !W(L - w) && L < V;
        D === "outside" ? (f = e.r + T, v = M ? "right" : "left") : !v || v === "center" ? (r === 2 * Math.PI && e.r0 === 0 ? f = 0 : f = (e.r + e.r0) / 2, v = "center") : v === "left" ? (f = e.r0 + T, v = M ? "right" : "left") : v === "right" && (f = e.r - T, v = M ? "left" : "right"), n.style.align = v, n.style.verticalAlign = d(a, "verticalAlign") || "middle", n.x = f * p + e.cx, n.y = f * S + e.cy;
        var P = 0;
        y === "radial" ? P = _(-i) + (M ? Math.PI : 0) : y === "tangential" ? P = _(Math.PI / 2 - i) + (M ? Math.PI : 0) : X(y) && (P = y * Math.PI / 180), n.rotation = _(P);
      });
      function d(o, a) {
        var g = o.get(a);
        return g ?? c.get(a);
      }
      u.dirtyStyle();
    }, A;
  })(N)
);
export {
  he as default
};
