// echarts@6.0.0/lib/chart/sunburst/SunburstSeries.js
import { __extends as m } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { map as y, each as v, isArray as n } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import g from "../../model/Series.js";
import b from "../../data/Tree.js";
import { wrapTreePathInfo as w } from "../helper/treeHelper.js";
import D from "../../model/Model.js";
import M from "../helper/enableAriaDecalForTree.js";
var O = (
  /** @class */
  (function(o) {
    m(t, o);
    function t() {
      var e = o !== null && o.apply(this, arguments) || this;
      return e.type = t.type, e.ignoreStyleOnData = !0, e;
    }
    return t.prototype.getInitialData = function(e, r) {
      var a = {
        name: e.name,
        children: e.data
      };
      p(a);
      var d = this._levelModels = y(e.levels || [], function(i) {
        return new D(i, this, r);
      }, this), l = b.createTree(a, this, h);
      function h(i) {
        i.wrapMethod("getItemModel", function(s, f) {
          var c = l.getNodeByDataIndex(f), u = d[c.depth];
          return u && (s.parentModel = u), s;
        });
      }
      return l.data;
    }, t.prototype.optionUpdated = function() {
      this.resetViewRoot();
    }, t.prototype.getDataParams = function(e) {
      var r = o.prototype.getDataParams.apply(this, arguments), a = this.getData().tree.getNodeByDataIndex(e);
      return r.treePathInfo = w(a, this), r;
    }, t.prototype.getLevelModel = function(e) {
      return this._levelModels && this._levelModels[e.depth];
    }, t.prototype.getViewRoot = function() {
      return this._viewRoot;
    }, t.prototype.resetViewRoot = function(e) {
      e ? this._viewRoot = e : e = this._viewRoot;
      var r = this.getRawData().tree.root;
      (!e || e !== r && !r.contains(e)) && (this._viewRoot = r);
    }, t.prototype.enableAriaDecal = function() {
      M(this);
    }, t.type = "series.sunburst", t.defaultOption = {
      // zlevel: 0,
      z: 2,
      // 默认全局居中
      center: ["50%", "50%"],
      radius: [0, "75%"],
      // 默认顺时针
      clockwise: !0,
      startAngle: 90,
      // 最小角度改为0
      minAngle: 0,
      // If still show when all data zero.
      stillShowZeroSum: !0,
      // 'rootToNode', 'link', or false
      nodeClick: "rootToNode",
      renderLabelForZeroData: !1,
      label: {
        // could be: 'radial', 'tangential', or 'none'
        rotate: "radial",
        show: !0,
        opacity: 1,
        // 'left' is for inner side of inside, and 'right' is for outer
        // side for inside
        align: "center",
        position: "inside",
        distance: 5,
        silent: !0
      },
      itemStyle: {
        borderWidth: 1,
        borderColor: "white",
        borderType: "solid",
        shadowBlur: 0,
        shadowColor: "rgba(0, 0, 0, 0.2)",
        shadowOffsetX: 0,
        shadowOffsetY: 0,
        opacity: 1
      },
      emphasis: {
        focus: "descendant"
      },
      blur: {
        itemStyle: {
          opacity: 0.2
        },
        label: {
          opacity: 0.1
        }
      },
      // Animation type can be expansion, scale.
      animationType: "expansion",
      animationDuration: 1e3,
      animationDurationUpdate: 500,
      data: [],
      /**
       * Sort order.
       *
       * Valid values: 'desc', 'asc', null, or callback function.
       * 'desc' and 'asc' for descend and ascendant order;
       * null for not sorting;
       * example of callback function:
       * function(nodeA, nodeB) {
       *     return nodeA.getValue() - nodeB.getValue();
       * }
       */
      sort: "desc"
    }, t;
  })(g)
);
function p(o) {
  var t = 0;
  v(o.children, function(r) {
    p(r);
    var a = r.value;
    n(a) && (a = a[0]), t += a;
  });
  var e = o.value;
  n(e) && (e = e[0]), (e == null || isNaN(e)) && (e = t), e < 0 && (e = 0), n(o.value) ? o.value[0] = e : o.value = e;
}
export {
  O as default
};
