// echarts@6.0.0/lib/chart/sunburst/sunburstAction.js
import { extend as r } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { deprecateReplaceLog as h } from "../../util/log.js";
import { retrieveTargetInfo as p, aboveViewRoot as f } from "../helper/treeHelper.js";
var g = "sunburstRootToNode", d = "sunburstHighlight", a = "sunburstUnhighlight";
function N(u) {
  u.registerAction({
    type: g,
    update: "updateView"
  }, function(t, o) {
    o.eachComponent({
      mainType: "series",
      subType: "sunburst",
      query: t
    }, n);
    function n(i, s) {
      var e = p(t, [g], i);
      if (e) {
        var c = i.getViewRoot();
        c && (t.direction = f(c, e.node) ? "rollUp" : "drillDown"), i.resetViewRoot(e.node);
      }
    }
  }), u.registerAction({
    type: d,
    update: "none"
  }, function(t, o, n) {
    t = r({}, t), o.eachComponent({
      mainType: "series",
      subType: "sunburst",
      query: t
    }, i);
    function i(s) {
      var e = p(t, [d], s);
      e && (t.dataIndex = e.node.dataIndex);
    }
    process.env.NODE_ENV !== "production" && h("sunburstHighlight", "highlight"), n.dispatchAction(r(t, {
      type: "highlight"
    }));
  }), u.registerAction({
    type: a,
    update: "updateView"
  }, function(t, o, n) {
    t = r({}, t), process.env.NODE_ENV !== "production" && h("sunburstUnhighlight", "downplay"), n.dispatchAction(r(t, {
      type: "downplay"
    }));
  });
}
export {
  g as ROOT_TO_NODE_ACTION,
  N as installSunburstAction
};
