// echarts@6.0.0/lib/chart/sunburst/SunburstView.js
import { __extends as I } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import { curry as O } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import E from "../../view/Chart.js";
import _ from "./SunburstPiece.js";
import L from "../../data/DataDiffer.js";
import { ROOT_TO_NODE_ACTION as x } from "./sunburstAction.js";
import { windowOpen as F } from "../../util/format.js";
var K = (
  /** @class */
  (function(h) {
    I(f, h);
    function f() {
      var e = h !== null && h.apply(this, arguments) || this;
      return e.type = f.type, e;
    }
    return f.prototype.render = function(e, o, a, c) {
      var r = this;
      this.seriesModel = e, this.api = a, this.ecModel = o;
      var u = e.getData(), n = u.tree.root, v = e.getViewRoot(), l = this.group, P = e.get("renderLabelForZeroData"), g = [];
      v.eachNode(function(t) {
        g.push(t);
      });
      var d = this._oldChildren || [];
      D(g, d), R(n, v), this._initEvents(), this._oldChildren = g;
      function D(t, i) {
        if (t.length === 0 && i.length === 0)
          return;
        new L(i, t, p, p).add(m).update(m).remove(O(m, null)).execute();
        function p(s) {
          return s.getId();
        }
        function m(s, y) {
          var V = s == null ? null : t[s], k = y == null ? null : i[y];
          T(V, k);
        }
      }
      function T(t, i) {
        if (!P && t && !t.getValue() && (t = null), t !== n && i !== n) {
          if (i && i.piece)
            t ? (i.piece.updateData(!1, t, e, o, a), u.setItemGraphicEl(t.dataIndex, i.piece)) : b(i);
          else if (t) {
            var p = new _(t, e, o, a);
            l.add(p), u.setItemGraphicEl(t.dataIndex, p);
          }
        }
      }
      function b(t) {
        t && t.piece && (l.remove(t.piece), t.piece = null);
      }
      function R(t, i) {
        i.depth > 0 ? (r.virtualPiece ? r.virtualPiece.updateData(!1, t, e, o, a) : (r.virtualPiece = new _(t, e, o, a), l.add(r.virtualPiece)), i.piece.off("click"), r.virtualPiece.on("click", function(p) {
          r._rootToNode(i.parentNode);
        })) : r.virtualPiece && (l.remove(r.virtualPiece), r.virtualPiece = null);
      }
    }, f.prototype._initEvents = function() {
      var e = this;
      this.group.off("click"), this.group.on("click", function(o) {
        var a = !1, c = e.seriesModel.getViewRoot();
        c.eachNode(function(r) {
          if (!a && r.piece && r.piece === o.target) {
            var u = r.getModel().get("nodeClick");
            if (u === "rootToNode")
              e._rootToNode(r);
            else if (u === "link") {
              var n = r.getModel(), v = n.get("link");
              if (v) {
                var l = n.get("target", !0) || "_blank";
                F(v, l);
              }
            }
            a = !0;
          }
        });
      });
    }, f.prototype._rootToNode = function(e) {
      e !== this.seriesModel.getViewRoot() && this.api.dispatchAction({
        type: x,
        from: this.uid,
        seriesId: this.seriesModel.id,
        targetNode: e
      });
    }, f.prototype.containPoint = function(e, o) {
      var a = o.getData(), c = a.getItemLayout(0);
      if (c) {
        var r = e[0] - c.cx, u = e[1] - c.cy, n = Math.sqrt(r * r + u * u);
        return n <= c.r && n >= c.r0;
      }
    }, f.type = "sunburst", f;
  })(E)
);
export {
  K as default
};
