// @tiptap/extension-table@3.15.3/dist/index.js
import { Extension as $, Node as x, callOrReturn as _, getExtensionField as O, mergeAttributes as M, findParentNodeClosestToPos as k } from "../../../../../@tiptap_core@2.24.0_@tiptap_pm@2.24.0/node_modules/@tiptap/core/dist/index.js";
import { TextSelection as I } from "../../../../../prosemirror-state@1.4.3/node_modules/prosemirror-state/dist/index.js";
import { columnResizing as P, tableEditing as z, CellSelection as N, fixTables as D, goToNextCell as H, setCellAttr as j, mergeCells as W, splitCell as v, toggleHeaderCell as V, toggleHeader as S, deleteTable as B, deleteRow as K, addRowAfter as F, addRowBefore as U, deleteColumn as q, addColumnAfter as G, addColumnBefore as J } from "../../../../../prosemirror-tables@1.6.4/node_modules/prosemirror-tables/dist/index.js";
var Q = x.create({
  name: "tableCell",
  addOptions() {
    return {
      HTMLAttributes: {}
    };
  },
  content: "block+",
  addAttributes() {
    return {
      colspan: {
        default: 1
      },
      rowspan: {
        default: 1
      },
      colwidth: {
        default: null,
        parseHTML: (e) => {
          var t, n;
          const l = e.getAttribute("colwidth"), u = l ? l.split(",").map((r) => parseInt(r, 10)) : null;
          if (!u) {
            const r = (t = e.closest("table")) == null ? void 0 : t.querySelectorAll("colgroup > col"), i = Array.from(((n = e.parentElement) == null ? void 0 : n.children) || []).indexOf(e);
            if (i && i > -1 && r && r[i]) {
              const a = r[i].getAttribute("width");
              return a ? [parseInt(a, 10)] : null;
            }
          }
          return u;
        }
      }
    };
  },
  tableRole: "cell",
  isolating: !0,
  parseHTML() {
    return [{ tag: "td" }];
  },
  renderHTML({ HTMLAttributes: e }) {
    return ["td", M(this.options.HTMLAttributes, e), 0];
  }
}), X = x.create({
  name: "tableHeader",
  addOptions() {
    return {
      HTMLAttributes: {}
    };
  },
  content: "block+",
  addAttributes() {
    return {
      colspan: {
        default: 1
      },
      rowspan: {
        default: 1
      },
      colwidth: {
        default: null,
        parseHTML: (e) => {
          const t = e.getAttribute("colwidth");
          return t ? t.split(",").map((l) => parseInt(l, 10)) : null;
        }
      }
    };
  },
  tableRole: "header_cell",
  isolating: !0,
  parseHTML() {
    return [{ tag: "th" }];
  },
  renderHTML({ HTMLAttributes: e }) {
    return ["th", M(this.options.HTMLAttributes, e), 0];
  }
}), Y = x.create({
  name: "tableRow",
  addOptions() {
    return {
      HTMLAttributes: {}
    };
  },
  content: "(tableCell | tableHeader)*",
  tableRole: "row",
  parseHTML() {
    return [{ tag: "tr" }];
  },
  renderHTML({ HTMLAttributes: e }) {
    return ["tr", M(this.options.HTMLAttributes, e), 0];
  }
});
function R(e, t) {
  return t ? ["width", `${Math.max(t, e)}px`] : ["min-width", `${e}px`];
}
function E(e, t, n, l, u, r) {
  var i;
  let a = 0, f = !0, c = t.firstChild;
  const b = e.firstChild;
  if (b !== null)
    for (let m = 0, y = 0; m < b.childCount; m += 1) {
      const { colspan: s, colwidth: d } = b.child(m).attrs;
      for (let o = 0; o < s; o += 1, y += 1) {
        const h = u === y ? r : d && d[o], C = h ? `${h}px` : "";
        if (a += h || l, h || (f = !1), c) {
          if (c.style.width !== C) {
            const [g, w] = R(l, h);
            c.style.setProperty(g, w);
          }
          c = c.nextSibling;
        } else {
          const g = document.createElement("col"), [w, A] = R(l, h);
          g.style.setProperty(w, A), t.appendChild(g);
        }
      }
    }
  for (; c; ) {
    const m = c.nextSibling;
    (i = c.parentNode) == null || i.removeChild(c), c = m;
  }
  const p = e.attrs.style && typeof e.attrs.style == "string" && /\bwidth\s*:/i.test(e.attrs.style);
  f && !p ? (n.style.width = `${a}px`, n.style.minWidth = "") : (n.style.width = "", n.style.minWidth = `${a}px`);
}
var Z = class {
  constructor(e, t) {
    this.node = e, this.cellMinWidth = t, this.dom = document.createElement("div"), this.dom.className = "tableWrapper", this.table = this.dom.appendChild(document.createElement("table")), e.attrs.style && (this.table.style.cssText = e.attrs.style), this.colgroup = this.table.appendChild(document.createElement("colgroup")), E(e, this.colgroup, this.table, t), this.contentDOM = this.table.appendChild(document.createElement("tbody"));
  }
  update(e) {
    return e.type !== this.node.type ? !1 : (this.node = e, E(e, this.colgroup, this.table, this.cellMinWidth), !0);
  }
  ignoreMutation(e) {
    const t = e.target, n = this.dom.contains(t), l = this.contentDOM.contains(t);
    return !!(n && !l && (e.type === "attributes" || e.type === "childList" || e.type === "characterData"));
  }
};
function ee(e, t, n, l) {
  let u = 0, r = !0;
  const i = [], a = e.firstChild;
  if (!a)
    return {};
  for (let p = 0, m = 0; p < a.childCount; p += 1) {
    const { colspan: y, colwidth: s } = a.child(p).attrs;
    for (let d = 0; d < y; d += 1, m += 1) {
      const o = n === m ? l : s && s[d];
      u += o || t, o || (r = !1);
      const [h, C] = R(t, o);
      i.push(["col", { style: `${h}: ${C}` }]);
    }
  }
  const f = r ? `${u}px` : "", c = r ? "" : `${u}px`;
  return { colgroup: ["colgroup", {}, ...i], tableWidth: f, tableMinWidth: c };
}
function L(e, t) {
  return e.createAndFill();
}
function te(e) {
  if (e.cached.tableNodeTypes)
    return e.cached.tableNodeTypes;
  const t = {};
  return Object.keys(e.nodes).forEach((n) => {
    const l = e.nodes[n];
    l.spec.tableRole && (t[l.spec.tableRole] = l);
  }), e.cached.tableNodeTypes = t, t;
}
function le(e, t, n, l, u) {
  const r = te(e), i = [], a = [];
  for (let c = 0; c < n; c += 1) {
    const b = L(r.cell);
    if (b && a.push(b), l) {
      const p = L(r.header_cell);
      p && i.push(p);
    }
  }
  const f = [];
  for (let c = 0; c < t; c += 1)
    f.push(r.row.createChecked(null, l && c === 0 ? i : a));
  return r.table.createChecked(null, f);
}
function re(e) {
  return e instanceof N;
}
var T = ({ editor: e }) => {
  const { selection: t } = e.state;
  if (!re(t))
    return !1;
  let n = 0;
  const l = k(t.ranges[0].$from, (r) => r.type.name === "table");
  return l?.node.descendants((r) => {
    if (r.type.name === "table")
      return !1;
    ["tableCell", "tableHeader"].includes(r.type.name) && (n += 1);
  }), n === t.ranges.length ? (e.commands.deleteTable(), !0) : !1;
}, ne = "";
function oe(e) {
  return (e || "").replace(/\s+/g, " ").trim();
}
function ae(e, t, n = {}) {
  var l;
  const u = (l = n.cellLineSeparator) != null ? l : ne;
  if (!e || !e.content || e.content.length === 0)
    return "";
  const r = [];
  e.content.forEach((s) => {
    const d = [];
    s.content && s.content.forEach((o) => {
      let h = "";
      o.content && Array.isArray(o.content) && o.content.length > 1 ? h = o.content.map((A) => t.renderChildren(A)).join(u) : h = o.content ? t.renderChildren(o.content) : "";
      const C = oe(h), g = o.type === "tableHeader";
      d.push({ text: C, isHeader: g });
    }), r.push(d);
  });
  const i = r.reduce((s, d) => Math.max(s, d.length), 0);
  if (i === 0)
    return "";
  const a = new Array(i).fill(0);
  r.forEach((s) => {
    var d;
    for (let o = 0; o < i; o += 1) {
      const C = (((d = s[o]) == null ? void 0 : d.text) || "").length;
      C > a[o] && (a[o] = C), a[o] < 3 && (a[o] = 3);
    }
  });
  const f = (s, d) => s + " ".repeat(Math.max(0, d - s.length)), c = r[0], b = c.some((s) => s.isHeader);
  let p = `
`;
  const m = new Array(i).fill(0).map((s, d) => b && c[d] && c[d].text || "");
  return p += `| ${m.map((s, d) => f(s, a[d])).join(" | ")} |
`, p += `| ${a.map((s) => "-".repeat(Math.max(3, s))).join(" | ")} |
`, (b ? r.slice(1) : r).forEach((s) => {
    p += `| ${new Array(i).fill(0).map((d, o) => f(s[o] && s[o].text || "", a[o])).join(" | ")} |
`;
  }), p;
}
var se = ae, ie = x.create({
  name: "table",
  // @ts-ignore
  addOptions() {
    return {
      HTMLAttributes: {},
      resizable: !1,
      renderWrapper: !1,
      handleWidth: 5,
      cellMinWidth: 25,
      // TODO: fix
      View: Z,
      lastColumnResizable: !0,
      allowTableNodeSelection: !1
    };
  },
  content: "tableRow+",
  tableRole: "table",
  isolating: !0,
  group: "block",
  parseHTML() {
    return [{ tag: "table" }];
  },
  renderHTML({ node: e, HTMLAttributes: t }) {
    const { colgroup: n, tableWidth: l, tableMinWidth: u } = ee(e, this.options.cellMinWidth), r = t.style;
    function i() {
      return r || (l ? `width: ${l}` : `min-width: ${u}`);
    }
    const a = [
      "table",
      M(this.options.HTMLAttributes, t, {
        style: i()
      }),
      n,
      ["tbody", 0]
    ];
    return this.options.renderWrapper ? ["div", { class: "tableWrapper" }, a] : a;
  },
  parseMarkdown: (e, t) => {
    const n = [];
    if (e.header) {
      const l = [];
      e.header.forEach((u) => {
        l.push(t.createNode("tableHeader", {}, [{ type: "paragraph", content: t.parseInline(u.tokens) }]));
      }), n.push(t.createNode("tableRow", {}, l));
    }
    return e.rows && e.rows.forEach((l) => {
      const u = [];
      l.forEach((r) => {
        u.push(t.createNode("tableCell", {}, [{ type: "paragraph", content: t.parseInline(r.tokens) }]));
      }), n.push(t.createNode("tableRow", {}, u));
    }), t.createNode("table", void 0, n);
  },
  renderMarkdown: (e, t) => se(e, t),
  addCommands() {
    return {
      insertTable: ({ rows: e = 3, cols: t = 3, withHeaderRow: n = !0 } = {}) => ({ tr: l, dispatch: u, editor: r }) => {
        const i = le(r.schema, e, t, n);
        if (u) {
          const a = l.selection.from + 1;
          l.replaceSelectionWith(i).scrollIntoView().setSelection(I.near(l.doc.resolve(a)));
        }
        return !0;
      },
      addColumnBefore: () => ({ state: e, dispatch: t }) => J(e, t),
      addColumnAfter: () => ({ state: e, dispatch: t }) => G(e, t),
      deleteColumn: () => ({ state: e, dispatch: t }) => q(e, t),
      addRowBefore: () => ({ state: e, dispatch: t }) => U(e, t),
      addRowAfter: () => ({ state: e, dispatch: t }) => F(e, t),
      deleteRow: () => ({ state: e, dispatch: t }) => K(e, t),
      deleteTable: () => ({ state: e, dispatch: t }) => B(e, t),
      mergeCells: () => ({ state: e, dispatch: t }) => W(e, t),
      splitCell: () => ({ state: e, dispatch: t }) => v(e, t),
      toggleHeaderColumn: () => ({ state: e, dispatch: t }) => S("column")(e, t),
      toggleHeaderRow: () => ({ state: e, dispatch: t }) => S("row")(e, t),
      toggleHeaderCell: () => ({ state: e, dispatch: t }) => V(e, t),
      mergeOrSplit: () => ({ state: e, dispatch: t }) => W(e, t) ? !0 : v(e, t),
      setCellAttribute: (e, t) => ({ state: n, dispatch: l }) => j(e, t)(n, l),
      goToNextCell: () => ({ state: e, dispatch: t }) => H(1)(e, t),
      goToPreviousCell: () => ({ state: e, dispatch: t }) => H(-1)(e, t),
      fixTables: () => ({ state: e, dispatch: t }) => (t && D(e), !0),
      setCellSelection: (e) => ({ tr: t, dispatch: n }) => {
        if (n) {
          const l = N.create(t.doc, e.anchorCell, e.headCell);
          t.setSelection(l);
        }
        return !0;
      }
    };
  },
  addKeyboardShortcuts() {
    return {
      Tab: () => this.editor.commands.goToNextCell() ? !0 : this.editor.can().addRowAfter() ? this.editor.chain().addRowAfter().goToNextCell().run() : !1,
      "Shift-Tab": () => this.editor.commands.goToPreviousCell(),
      Backspace: T,
      "Mod-Backspace": T,
      Delete: T,
      "Mod-Delete": T
    };
  },
  addProseMirrorPlugins() {
    return [
      ...this.options.resizable && this.editor.isEditable ? [
        P({
          handleWidth: this.options.handleWidth,
          cellMinWidth: this.options.cellMinWidth,
          defaultCellMinWidth: this.options.cellMinWidth,
          View: this.options.View,
          lastColumnResizable: this.options.lastColumnResizable
        })
      ] : [],
      z({
        allowTableNodeSelection: this.options.allowTableNodeSelection
      })
    ];
  },
  extendNodeSchema(e) {
    const t = {
      name: e.name,
      options: e.options,
      storage: e.storage
    };
    return {
      tableRole: _(O(e, "tableRole", t))
    };
  }
}), pe = $.create({
  name: "tableKit",
  addExtensions() {
    const e = [];
    return this.options.table !== !1 && e.push(ie.configure(this.options.table)), this.options.tableCell !== !1 && e.push(Q.configure(this.options.tableCell)), this.options.tableHeader !== !1 && e.push(X.configure(this.options.tableHeader)), this.options.tableRow !== !1 && e.push(Y.configure(this.options.tableRow)), e;
  }
});
export {
  ne as DEFAULT_CELL_LINE_SEPARATOR,
  ie as Table,
  Q as TableCell,
  X as TableHeader,
  pe as TableKit,
  Y as TableRow,
  Z as TableView,
  ee as createColGroup,
  le as createTable,
  ae as renderTableToMarkdown,
  E as updateColumns
};
