// echarts@6.0.0/lib/theme/dark.js
import t from "../visual/tokens.js";
var e = t.darkColor, a = e.background, r = function() {
  return {
    axisLine: {
      lineStyle: {
        color: e.axisLine
      }
    },
    splitLine: {
      lineStyle: {
        color: e.axisSplitLine
      }
    },
    splitArea: {
      areaStyle: {
        color: [e.backgroundTint, e.backgroundTransparent]
      }
    },
    minorSplitLine: {
      lineStyle: {
        color: e.axisMinorSplitLine
      }
    },
    axisLabel: {
      color: e.axisLabel
    },
    axisName: {}
  };
}, l = {
  label: {
    color: e.secondary
  },
  itemStyle: {
    borderColor: e.borderTint
  },
  dividerLineStyle: {
    color: e.border
  }
}, i = {
  darkMode: !0,
  color: e.theme,
  backgroundColor: a,
  axisPointer: {
    lineStyle: {
      color: e.border
    },
    crossStyle: {
      color: e.borderShade
    },
    label: {
      color: e.tertiary
    }
  },
  legend: {
    textStyle: {
      color: e.secondary
    },
    pageTextStyle: {
      color: e.tertiary
    }
  },
  textStyle: {
    color: e.secondary
  },
  title: {
    textStyle: {
      color: e.primary
    },
    subtextStyle: {
      color: e.quaternary
    }
  },
  toolbox: {
    iconStyle: {
      borderColor: e.accent50
    }
  },
  tooltip: {
    backgroundColor: e.neutral20,
    defaultBorderColor: e.border,
    textStyle: {
      color: e.tertiary
    }
  },
  dataZoom: {
    borderColor: e.accent10,
    textStyle: {
      color: e.tertiary
    },
    brushStyle: {
      color: e.backgroundTint
    },
    handleStyle: {
      color: e.neutral00,
      borderColor: e.accent20
    },
    moveHandleStyle: {
      color: e.accent40
    },
    emphasis: {
      handleStyle: {
        borderColor: e.accent50
      }
    },
    dataBackground: {
      lineStyle: {
        color: e.accent30
      },
      areaStyle: {
        color: e.accent20
      }
    },
    selectedDataBackground: {
      lineStyle: {
        color: e.accent50
      },
      areaStyle: {
        color: e.accent30
      }
    }
  },
  visualMap: {
    textStyle: {
      color: e.secondary
    },
    handleStyle: {
      borderColor: e.neutral30
    }
  },
  timeline: {
    lineStyle: {
      color: e.accent10
    },
    label: {
      color: e.tertiary
    },
    controlStyle: {
      color: e.accent30,
      borderColor: e.accent30
    }
  },
  calendar: {
    itemStyle: {
      color: e.neutral00,
      borderColor: e.neutral20
    },
    dayLabel: {
      color: e.tertiary
    },
    monthLabel: {
      color: e.secondary
    },
    yearLabel: {
      color: e.secondary
    }
  },
  matrix: {
    x: l,
    y: l,
    backgroundColor: {
      borderColor: e.axisLine
    },
    body: {
      itemStyle: {
        borderColor: e.borderTint
      }
    }
  },
  timeAxis: r(),
  logAxis: r(),
  valueAxis: r(),
  categoryAxis: r(),
  line: {
    symbol: "circle"
  },
  graph: {
    color: e.theme
  },
  gauge: {
    title: {
      color: e.secondary
    },
    axisLine: {
      lineStyle: {
        color: [[1, e.neutral05]]
      }
    },
    axisLabel: {
      color: e.axisLabel
    },
    detail: {
      color: e.primary
    }
  },
  candlestick: {
    itemStyle: {
      color: "#f64e56",
      color0: "#54ea92",
      borderColor: "#f64e56",
      borderColor0: "#54ea92"
      // borderColor: '#ca2824',
      // borderColor0: '#09a443'
    }
  },
  funnel: {
    itemStyle: {
      borderColor: e.background
    }
  },
  radar: (function() {
    var o = r();
    return o.axisName = {
      color: e.axisLabel
    }, o.axisLine.lineStyle.color = e.neutral20, o;
  })(),
  treemap: {
    breadcrumb: {
      itemStyle: {
        color: e.neutral20,
        textStyle: {
          color: e.secondary
        }
      },
      emphasis: {
        itemStyle: {
          color: e.neutral30
        }
      }
    }
  },
  sunburst: {
    itemStyle: {
      borderColor: e.background
    }
  },
  map: {
    itemStyle: {
      borderColor: e.border,
      areaColor: e.neutral10
    },
    label: {
      color: e.tertiary
    },
    emphasis: {
      label: {
        color: e.primary
      },
      itemStyle: {
        areaColor: e.highlight
      }
    },
    select: {
      label: {
        color: e.primary
      },
      itemStyle: {
        areaColor: e.highlight
      }
    }
  },
  geo: {
    itemStyle: {
      borderColor: e.border,
      areaColor: e.neutral10
    },
    emphasis: {
      label: {
        color: e.primary
      },
      itemStyle: {
        areaColor: e.highlight
      }
    },
    select: {
      label: {
        color: e.primary
      },
      itemStyle: {
        color: e.highlight
      }
    }
  }
};
i.categoryAxis.splitLine.show = !1;
export {
  i as default
};
