// echarts@6.0.0/lib/chart/radar/RadarView.js
import { __extends as x } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import * as k from "../../util/graphic.js";
import { setStatesStylesFromModel as E, toggleHoverEmphasis as T } from "../../util/states.js";
import { defaults as _, each as F, clone as O, extend as H, map as X } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import { normalizeSymbolSize as Y, createSymbol as Z } from "../../util/symbol.js";
import $ from "../../view/Chart.js";
import { setLabelStyle as j, getLabelStatesModels as q } from "../../label/labelStyle.js";
import B from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Image.js";
import { initProps as b, saveOldStyle as V, updateProps as G } from "../../animation/basicTransition.js";
import J from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Polygon.js";
import K from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/shape/Polyline.js";
import L from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/graphic/Group.js";
var pe = (
  /** @class */
  (function(d) {
    x(y, d);
    function y() {
      var m = d !== null && d.apply(this, arguments) || this;
      return m.type = y.type, m;
    }
    return y.prototype.render = function(m, Q, U) {
      var v = m.coordinateSystem, S = this.group, n = m.getData(), g = this._data;
      function z(e, r) {
        var t = e.getItemVisual(r, "symbol") || "circle";
        if (t !== "none") {
          var o = Y(e.getItemVisual(r, "symbolSize")), l = Z(t, -1, -1, 2, 2), p = e.getItemVisual(r, "symbolRotate") || 0;
          return l.attr({
            style: {
              strokeNoScale: !0
            },
            z2: 100,
            scaleX: o[0] / 2,
            scaleY: o[1] / 2,
            rotation: p * Math.PI / 180 || 0
          }), l;
        }
      }
      function h(e, r, t, o, l, p) {
        t.removeAll();
        for (var a = 0; a < r.length - 1; a++) {
          var s = z(o, l);
          s && (s.__dimIdx = a, e[a] ? (s.setPosition(e[a]), k[p ? "initProps" : "updateProps"](s, {
            x: r[a][0],
            y: r[a][1]
          }, m, l)) : s.setPosition(r[a]), t.add(s));
        }
      }
      function I(e) {
        return X(e, function(r) {
          return [v.cx, v.cy];
        });
      }
      n.diff(g).add(function(e) {
        var r = n.getItemLayout(e);
        if (r) {
          var t = new J(), o = new K(), l = {
            shape: {
              points: r
            }
          };
          t.shape.points = I(r), o.shape.points = I(r), b(t, l, m, e), b(o, l, m, e);
          var p = new L(), a = new L();
          p.add(o), p.add(t), p.add(a), h(o.shape.points, r, a, n, e, !0), n.setItemGraphicEl(e, p);
        }
      }).update(function(e, r) {
        var t = g.getItemGraphicEl(r), o = t.childAt(0), l = t.childAt(1), p = t.childAt(2), a = {
          shape: {
            points: n.getItemLayout(e)
          }
        };
        a.shape.points && (h(o.shape.points, a.shape.points, p, n, e, !1), V(l), V(o), G(o, a, m), G(l, a, m), n.setItemGraphicEl(e, t));
      }).remove(function(e) {
        S.remove(g.getItemGraphicEl(e));
      }).execute(), n.eachItemGraphicEl(function(e, r) {
        var t = n.getItemModel(r), o = e.childAt(0), l = e.childAt(1), p = e.childAt(2), a = n.getItemVisual(r, "style"), s = a.fill;
        S.add(e), o.useStyle(_(t.getModel("lineStyle").getLineStyle(), {
          fill: "none",
          stroke: s
        })), E(o, t, "lineStyle"), E(l, t, "areaStyle");
        var M = t.getModel("areaStyle"), A = M.isEmpty() && M.parentModel.isEmpty();
        l.ignore = A, F(["emphasis", "select", "blur"], function(i) {
          var u = t.getModel([i, "areaStyle"]), c = u.isEmpty() && u.parentModel.isEmpty();
          l.ensureState(i).ignore = c && A;
          var C = t.getModel([i, "lineStyle"]).getLineStyle();
          o.ensureState(i).style = C;
          var D = u.getAreaStyle();
          l.ensureState(i).style = D;
          var R = t.getModel([i, "itemStyle"]).getItemStyle();
          p.eachChild(function(w) {
            w.ensureState(i).style = O(R);
          });
        }), l.useStyle(_(t.getModel("areaStyle").getAreaStyle(), {
          fill: s,
          opacity: 0.7,
          decal: a.decal
        }));
        var f = t.getModel("emphasis");
        p.eachChild(function(i) {
          if (i instanceof B) {
            var u = i.style;
            i.useStyle(H({
              // TODO other properties like x, y ?
              image: u.image,
              x: u.x,
              y: u.y,
              width: u.width,
              height: u.height
            }, a));
          } else
            i.useStyle(a), i.setColor(s), i.style.strokeNoScale = !0;
          var c = n.getStore().get(n.getDimensionIndex(i.__dimIdx), r);
          (c == null || isNaN(c)) && (c = ""), j(i, q(t), {
            labelFetcher: n.hostModel,
            labelDataIndex: r,
            labelDimIndex: i.__dimIdx,
            defaultText: c,
            inheritColor: s,
            defaultOpacity: a.opacity
          });
        }), T(e, f.get("focus"), f.get("blurScope"), f.get("disabled"));
      }), this._data = n;
    }, y.prototype.remove = function() {
      this.group.removeAll(), this._data = null;
    }, y.type = "radar", y;
  })($)
);
export {
  pe as default
};
