// echarts@6.0.0/lib/chart/radar/RadarSeries.js
import { __extends as h } from "../../../../../../tslib@2.3.0/node_modules/tslib/tslib.es6.js";
import g from "../../model/Series.js";
import v from "../helper/createSeriesDataSimply.js";
import { bind as u, map as c } from "../../../../../../zrender@6.0.0/node_modules/zrender/lib/core/util.js";
import S from "../../visual/LegendVisualProvider.js";
import { retrieveVisualColorForTooltipMarker as D, createTooltipMarkup as y } from "../../component/tooltip/tooltipMarkup.js";
var P = (
  /** @class */
  (function(i) {
    h(e, i);
    function e() {
      var t = i !== null && i.apply(this, arguments) || this;
      return t.type = e.type, t.hasSymbolVisual = !0, t;
    }
    return e.prototype.init = function(t) {
      i.prototype.init.apply(this, arguments), this.legendVisualProvider = new S(u(this.getData, this), u(this.getRawData, this));
    }, e.prototype.getInitialData = function(t, a) {
      return v(this, {
        generateCoord: "indicator_",
        generateCoordCount: 1 / 0
      });
    }, e.prototype.formatTooltip = function(t, a, n) {
      var o = this.getData(), r = this.coordinateSystem, l = r.getIndicatorAxes(), s = this.getData().getName(t), m = s === "" ? this.name : s, f = D(this, t);
      return y("section", {
        header: m,
        sortBlocks: !0,
        blocks: c(l, function(p) {
          var d = o.get(o.mapDimension(p.dim), t);
          return y("nameValue", {
            markerType: "subItem",
            markerColor: f,
            name: p.name,
            value: d,
            sortParam: d
          });
        })
      });
    }, e.prototype.getTooltipPosition = function(t) {
      if (t != null) {
        for (var a = this.getData(), n = this.coordinateSystem, o = a.getValues(c(n.dimensions, function(m) {
          return a.mapDimension(m);
        }), t), r = 0, l = o.length; r < l; r++)
          if (!isNaN(o[r])) {
            var s = n.getIndicatorAxes();
            return n.coordToPoint(s[r].dataToCoord(o[r]), r);
          }
      }
    }, e.type = "series.radar", e.dependencies = ["radar"], e.defaultOption = {
      // zlevel: 0,
      z: 2,
      colorBy: "data",
      coordinateSystem: "radar",
      legendHoverLink: !0,
      radarIndex: 0,
      lineStyle: {
        width: 2,
        type: "solid",
        join: "round"
      },
      label: {
        position: "top"
      },
      // areaStyle: {
      // },
      // itemStyle: {}
      symbolSize: 8
      // symbolRotate: null
    }, e;
  })(g)
);
export {
  P as default
};
