// gridstack@12.3.3/dist/dd-droppable.js
import { DDManager as i } from "./dd-manager.js";
import { DDBaseImplement as h } from "./dd-base-impl.js";
import { Utils as o } from "./utils.js";
import { isTouch as p, pointerenter as a, pointerleave as l } from "./dd-touch.js";
class E extends h {
  constructor(e, t = {}) {
    super(), this.el = e, this.option = t, this._mouseEnter = this._mouseEnter.bind(this), this._mouseLeave = this._mouseLeave.bind(this), this.enable(), this._setupAccept();
  }
  on(e, t) {
    super.on(e, t);
  }
  off(e) {
    super.off(e);
  }
  enable() {
    this.disabled !== !1 && (super.enable(), this.el.classList.add("ui-droppable"), this.el.classList.remove("ui-droppable-disabled"), this.el.addEventListener("mouseenter", this._mouseEnter), this.el.addEventListener("mouseleave", this._mouseLeave), p && (this.el.addEventListener("pointerenter", a), this.el.addEventListener("pointerleave", l)));
  }
  disable(e = !1) {
    this.disabled !== !0 && (super.disable(), this.el.classList.remove("ui-droppable"), e || this.el.classList.add("ui-droppable-disabled"), this.el.removeEventListener("mouseenter", this._mouseEnter), this.el.removeEventListener("mouseleave", this._mouseLeave), p && (this.el.removeEventListener("pointerenter", a), this.el.removeEventListener("pointerleave", l)));
  }
  destroy() {
    this.disable(!0), this.el.classList.remove("ui-droppable"), this.el.classList.remove("ui-droppable-disabled"), super.destroy();
  }
  updateOption(e) {
    return Object.keys(e).forEach((t) => this.option[t] = e[t]), this._setupAccept(), this;
  }
  /** @internal called when the cursor enters our area - prepare for a possible drop and track leaving */
  _mouseEnter(e) {
    if (!i.dragElement || !this._canDrop(i.dragElement.el))
      return;
    e.preventDefault(), e.stopPropagation(), i.dropElement && i.dropElement !== this && i.dropElement._mouseLeave(e, !0), i.dropElement = this;
    const t = o.initEvent(e, { target: this.el, type: "dropover" });
    this.option.over && this.option.over(t, this._ui(i.dragElement)), this.triggerEvent("dropover", t), this.el.classList.add("ui-droppable-over");
  }
  /** @internal called when the item is leaving our area, stop tracking if we had moving item */
  _mouseLeave(e, t = !1) {
    if (!i.dragElement || i.dropElement !== this)
      return;
    e.preventDefault(), e.stopPropagation();
    const n = o.initEvent(e, { target: this.el, type: "dropout" });
    if (this.option.out && this.option.out(n, this._ui(i.dragElement)), this.triggerEvent("dropout", n), i.dropElement === this && (delete i.dropElement, !t)) {
      let s, r = this.el.parentElement;
      for (; !s && r; )
        s = r.ddElement?.ddDroppable, r = r.parentElement;
      s && s._mouseEnter(e);
    }
  }
  /** item is being dropped on us - called by the drag mouseup handler - this calls the client drop event */
  drop(e) {
    e.preventDefault();
    const t = o.initEvent(e, { target: this.el, type: "drop" });
    this.option.drop && this.option.drop(t, this._ui(i.dragElement)), this.triggerEvent("drop", t);
  }
  /** @internal true if element matches the string/method accept option */
  _canDrop(e) {
    return e && (!this.accept || this.accept(e));
  }
  /** @internal */
  _setupAccept() {
    return this.option.accept ? (typeof this.option.accept == "string" ? this.accept = (e) => e.classList.contains(this.option.accept) || e.matches(this.option.accept) : this.accept = this.option.accept, this) : this;
  }
  /** @internal */
  _ui(e) {
    return {
      draggable: e.el,
      ...e.ui()
    };
  }
}
export {
  E as DDDroppable
};
