// gridstack@12.3.3/dist/dd-gridstack.js
import { Utils as b } from "./utils.js";
import { DDManager as i } from "./dd-manager.js";
import { DDElement as D } from "./dd-element.js";
class h {
  /**
   * Enable/disable/configure resizing for grid elements.
   *
   * @param el - Grid item element(s) to configure
   * @param opts - Resize options or command ('enable', 'disable', 'destroy', 'option', or config object)
   * @param key - Option key when using 'option' command
   * @param value - Option value when using 'option' command
   * @returns this instance for chaining
   *
   * @example
   * dd.resizable(element, 'enable');  // Enable resizing
   * dd.resizable(element, 'option', 'minWidth', 100);  // Set minimum width
   */
  resizable(r, e, t, l) {
    return this._getDDElements(r, e).forEach((a) => {
      if (e === "disable" || e === "enable")
        a.ddResizable && a.ddResizable[e]();
      else if (e === "destroy")
        a.ddResizable && a.cleanResizable();
      else if (e === "option")
        a.setupResizable({ [t]: l });
      else {
        const d = a.el.gridstackNode.grid;
        let n = a.el.getAttribute("gs-resize-handles") || d.opts.resizable.handles || "e,s,se";
        n === "all" && (n = "n,e,s,w,se,sw,ne,nw");
        const g = !d.opts.alwaysShowResizeHandle;
        a.setupResizable({
          ...d.opts.resizable,
          handles: n,
          autoHide: g,
          start: e.start,
          stop: e.stop,
          resize: e.resize
        });
      }
    }), this;
  }
  /**
   * Enable/disable/configure dragging for grid elements.
   *
   * @param el - Grid item element(s) to configure
   * @param opts - Drag options or command ('enable', 'disable', 'destroy', 'option', or config object)
   * @param key - Option key when using 'option' command
   * @param value - Option value when using 'option' command
   * @returns this instance for chaining
   *
   * @example
   * dd.draggable(element, 'enable');  // Enable dragging
   * dd.draggable(element, {handle: '.drag-handle'});  // Configure drag handle
   */
  draggable(r, e, t, l) {
    return this._getDDElements(r, e).forEach((a) => {
      if (e === "disable" || e === "enable")
        a.ddDraggable && a.ddDraggable[e]();
      else if (e === "destroy")
        a.ddDraggable && a.cleanDraggable();
      else if (e === "option")
        a.setupDraggable({ [t]: l });
      else {
        const s = a.el.gridstackNode.grid;
        a.setupDraggable({
          ...s.opts.draggable,
          // containment: (grid.parentGridNode && grid.opts.dragOut === false) ? grid.el.parentElement : (grid.opts.draggable.containment || null),
          start: e.start,
          stop: e.stop,
          drag: e.drag
        });
      }
    }), this;
  }
  dragIn(r, e) {
    return this._getDDElements(r).forEach((t) => t.setupDraggable(e)), this;
  }
  droppable(r, e, t, l) {
    return typeof e.accept == "function" && !e._accept && (e._accept = e.accept, e.accept = (a) => e._accept(a)), this._getDDElements(r, e).forEach((a) => {
      e === "disable" || e === "enable" ? a.ddDroppable && a.ddDroppable[e]() : e === "destroy" ? a.ddDroppable && a.cleanDroppable() : e === "option" ? a.setupDroppable({ [t]: l }) : a.setupDroppable(e);
    }), this;
  }
  /** true if element is droppable */
  isDroppable(r) {
    return !!(r?.ddElement?.ddDroppable && !r.ddElement.ddDroppable.disabled);
  }
  /** true if element is draggable */
  isDraggable(r) {
    return !!(r?.ddElement?.ddDraggable && !r.ddElement.ddDraggable.disabled);
  }
  /** true if element is draggable */
  isResizable(r) {
    return !!(r?.ddElement?.ddResizable && !r.ddElement.ddResizable.disabled);
  }
  on(r, e, t) {
    return this._getDDElements(r).forEach((l) => l.on(e, (a) => {
      t(a, i.dragElement ? i.dragElement.el : a.target, i.dragElement ? i.dragElement.helper : null);
    })), this;
  }
  off(r, e) {
    return this._getDDElements(r).forEach((t) => t.off(e)), this;
  }
  /** @internal returns a list of DD elements, creating them on the fly by default unless option is to destroy or disable */
  _getDDElements(r, e) {
    const t = r.gridstack || e !== "destroy" && e !== "disable", l = b.getElements(r);
    return l.length ? l.map((s) => s.ddElement || (t ? D.init(s) : null)).filter((s) => s) : [];
  }
}
export {
  h as DDGridStack
};
