// gridstack@12.3.3/dist/dd-resizable.js
import { DDResizableHandle as m } from "./dd-resizable-handle.js";
import { DDBaseImplement as g } from "./dd-base-impl.js";
import { Utils as r } from "./utils.js";
import { DDManager as n } from "./dd-manager.js";
class d extends g {
  // have to be public else complains for HTMLElementExtendOpt ?
  constructor(t, e = {}) {
    super(), this.el = t, this.option = e, this.rectScale = { x: 1, y: 1 }, this._ui = () => {
      const i = this.el.parentElement.getBoundingClientRect(), h = {
        width: this.originalRect.width,
        height: this.originalRect.height + this.scrolled,
        left: this.originalRect.left,
        top: this.originalRect.top - this.scrolled
      }, o = this.temporalRect || h;
      return {
        position: {
          left: (o.left - i.left) * this.rectScale.x,
          top: (o.top - i.top) * this.rectScale.y
        },
        size: {
          width: o.width * this.rectScale.x,
          height: o.height * this.rectScale.y
        }
        /* Gridstack ONLY needs position set above... keep around in case.
        element: [this.el], // The object representing the element to be resized
        helper: [], // TODO: not support yet - The object representing the helper that's being resized
        originalElement: [this.el],// we don't wrap here, so simplify as this.el //The object representing the original element before it is wrapped
        originalPosition: { // The position represented as { left, top } before the resizable is resized
          left: this.originalRect.left - containmentRect.left,
          top: this.originalRect.top - containmentRect.top
        },
        originalSize: { // The size represented as { width, height } before the resizable is resized
          width: this.originalRect.width,
          height: this.originalRect.height
        }
        */
      };
    }, this._mouseOver = this._mouseOver.bind(this), this._mouseOut = this._mouseOut.bind(this), this.enable(), this._setupAutoHide(this.option.autoHide), this._setupHandlers();
  }
  on(t, e) {
    super.on(t, e);
  }
  off(t) {
    super.off(t);
  }
  enable() {
    super.enable(), this.el.classList.remove("ui-resizable-disabled"), this._setupAutoHide(this.option.autoHide);
  }
  disable() {
    super.disable(), this.el.classList.add("ui-resizable-disabled"), this._setupAutoHide(!1);
  }
  destroy() {
    this._removeHandlers(), this._setupAutoHide(!1), delete this.el, super.destroy();
  }
  updateOption(t) {
    const e = t.handles && t.handles !== this.option.handles, s = t.autoHide && t.autoHide !== this.option.autoHide;
    return Object.keys(t).forEach((i) => this.option[i] = t[i]), e && (this._removeHandlers(), this._setupHandlers()), s && this._setupAutoHide(this.option.autoHide), this;
  }
  /** @internal turns auto hide on/off */
  _setupAutoHide(t) {
    return t ? (this.el.classList.add("ui-resizable-autohide"), this.el.addEventListener("mouseover", this._mouseOver), this.el.addEventListener("mouseout", this._mouseOut)) : (this.el.classList.remove("ui-resizable-autohide"), this.el.removeEventListener("mouseover", this._mouseOver), this.el.removeEventListener("mouseout", this._mouseOut), n.overResizeElement === this && delete n.overResizeElement), this;
  }
  /** @internal */
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  _mouseOver(t) {
    n.overResizeElement || n.dragElement || (n.overResizeElement = this, this.el.classList.remove("ui-resizable-autohide"));
  }
  /** @internal */
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  _mouseOut(t) {
    n.overResizeElement === this && (delete n.overResizeElement, this.el.classList.add("ui-resizable-autohide"));
  }
  /** @internal */
  _setupHandlers() {
    return this.handlers = this.option.handles.split(",").map((t) => t.trim()).map((t) => new m(this.el, t, {
      start: (e) => {
        this._resizeStart(e);
      },
      stop: (e) => {
        this._resizeStop(e);
      },
      move: (e) => {
        this._resizing(e, t);
      }
    })), this;
  }
  /** @internal */
  _resizeStart(t) {
    this.sizeToContent = r.shouldSizeToContent(this.el.gridstackNode, !0), this.originalRect = this.el.getBoundingClientRect(), this.scrollEl = r.getScrollElement(this.el), this.scrollY = this.scrollEl.scrollTop, this.scrolled = 0, this.startEvent = t, this._setupHelper(), this._applyChange();
    const e = r.initEvent(t, { type: "resizestart", target: this.el });
    return this.option.start && this.option.start(e, this._ui()), this.el.classList.add("ui-resizable-resizing"), this.triggerEvent("resizestart", e), this;
  }
  /** @internal */
  _resizing(t, e) {
    this.scrolled = this.scrollEl.scrollTop - this.scrollY, this.temporalRect = this._getChange(t, e), this._applyChange();
    const s = r.initEvent(t, { type: "resize", target: this.el });
    return this.option.resize && this.option.resize(s, this._ui()), this.triggerEvent("resize", s), this;
  }
  /** @internal */
  _resizeStop(t) {
    const e = r.initEvent(t, { type: "resizestop", target: this.el });
    return this._cleanHelper(), this.option.stop && this.option.stop(e), this.el.classList.remove("ui-resizable-resizing"), this.triggerEvent("resizestop", e), delete this.startEvent, delete this.originalRect, delete this.temporalRect, delete this.scrollY, delete this.scrolled, this;
  }
  /** @internal */
  _setupHelper() {
    this.elOriginStyleVal = d._originStyleProp.map((s) => this.el.style[s]), this.parentOriginStylePosition = this.el.parentElement.style.position;
    const t = this.el.parentElement, e = r.getValuesFromTransformedElement(t);
    return this.rectScale = {
      x: e.xScale,
      y: e.yScale
    }, getComputedStyle(this.el.parentElement).position.match(/static/) && (this.el.parentElement.style.position = "relative"), this.el.style.position = "absolute", this.el.style.opacity = "0.8", this;
  }
  /** @internal */
  _cleanHelper() {
    return d._originStyleProp.forEach((t, e) => {
      this.el.style[t] = this.elOriginStyleVal[e] || null;
    }), this.el.parentElement.style.position = this.parentOriginStylePosition || null, this;
  }
  /** @internal */
  _getChange(t, e) {
    const s = this.startEvent, i = {
      width: this.originalRect.width,
      height: this.originalRect.height + this.scrolled,
      left: this.originalRect.left,
      top: this.originalRect.top - this.scrolled
    }, h = t.clientX - s.clientX, o = this.sizeToContent ? 0 : t.clientY - s.clientY;
    let a, c;
    e.indexOf("e") > -1 ? i.width += h : e.indexOf("w") > -1 && (i.width -= h, i.left += h, a = !0), e.indexOf("s") > -1 ? i.height += o : e.indexOf("n") > -1 && (i.height -= o, i.top += o, c = !0);
    const l = this._constrainSize(i.width, i.height, a, c);
    return Math.round(i.width) !== Math.round(l.width) && (e.indexOf("w") > -1 && (i.left += i.width - l.width), i.width = l.width), Math.round(i.height) !== Math.round(l.height) && (e.indexOf("n") > -1 && (i.top += i.height - l.height), i.height = l.height), i;
  }
  /** @internal constrain the size to the set min/max values */
  _constrainSize(t, e, s, i) {
    const h = this.option, o = (s ? h.maxWidthMoveLeft : h.maxWidth) || Number.MAX_SAFE_INTEGER, a = h.minWidth / this.rectScale.x || t, c = (i ? h.maxHeightMoveUp : h.maxHeight) || Number.MAX_SAFE_INTEGER, l = h.minHeight / this.rectScale.y || e, p = Math.min(o, Math.max(a, t)), u = Math.min(c, Math.max(l, e));
    return { width: p, height: u };
  }
  /** @internal */
  _applyChange() {
    let t = { left: 0, top: 0, width: 0, height: 0 };
    if (this.el.style.position === "absolute") {
      const e = this.el.parentElement, { left: s, top: i } = e.getBoundingClientRect();
      t = { left: s, top: i, width: 0, height: 0 };
    }
    return this.temporalRect ? (Object.keys(this.temporalRect).forEach((e) => {
      const s = this.temporalRect[e], i = e === "width" || e === "left" ? this.rectScale.x : e === "height" || e === "top" ? this.rectScale.y : 1;
      this.el.style[e] = (s - t[e]) * i + "px";
    }), this) : this;
  }
  /** @internal */
  _removeHandlers() {
    return this.handlers.forEach((t) => t.destroy()), delete this.handlers, this;
  }
}
d._originStyleProp = ["width", "height", "position", "left", "top", "opacity", "zIndex"];
export {
  d as DDResizable
};
