// gridstack@12.3.3/dist/dd-touch.js
import { DDManager as u } from "./dd-manager.js";
import { Utils as i } from "./utils.js";
const m = typeof window < "u" && typeof document < "u" && ("ontouchstart" in document || "ontouchstart" in window || window.DocumentTouch && document instanceof window.DocumentTouch || navigator.maxTouchPoints > 0 || navigator.msMaxTouchPoints > 0);
class t {
}
function o(e, n) {
  e.touches.length > 1 || (e.cancelable && e.preventDefault(), i.simulateMouseEvent(e.changedTouches[0], n));
}
function r(e, n) {
  e.cancelable && e.preventDefault(), i.simulateMouseEvent(e, n);
}
function s(e) {
  t.touchHandled || (t.touchHandled = !0, o(e, "mousedown"));
}
function d(e) {
  t.touchHandled && o(e, "mousemove");
}
function f(e) {
  if (!t.touchHandled)
    return;
  t.pointerLeaveTimeout && (window.clearTimeout(t.pointerLeaveTimeout), delete t.pointerLeaveTimeout);
  const n = !!u.dragElement;
  o(e, "mouseup"), n || o(e, "click"), t.touchHandled = !1;
}
function l(e) {
  e.pointerType !== "mouse" && e.target.releasePointerCapture(e.pointerId);
}
function p(e) {
  u.dragElement && e.pointerType !== "mouse" && r(e, "mouseenter");
}
function h(e) {
  u.dragElement && e.pointerType !== "mouse" && (t.pointerLeaveTimeout = window.setTimeout(() => {
    delete t.pointerLeaveTimeout, r(e, "mouseleave");
  }, 10));
}
export {
  m as isTouch,
  l as pointerdown,
  p as pointerenter,
  h as pointerleave,
  f as touchend,
  d as touchmove,
  s as touchstart
};
