// gridstack@12.3.3/dist/gridstack.js
import { GridStackEngine as N } from "./gridstack-engine.js";
import { Utils as h, obsolete as H } from "./utils.js";
import { gridDefaults as p } from "./types.js";
import { DDGridStack as A } from "./dd-gridstack.js";
import { isTouch as D } from "./dd-touch.js";
import { DDManager as _ } from "./dd-manager.js";
import { DDResizable as U } from "./dd-resizable.js";
import { DDDraggable as P } from "./dd-draggable.js";
import { DDResizableHandle as F } from "./dd-resizable-handle.js";
const f = new A();
class d {
  /**
   * initializing the HTML element, or selector string, into a grid will return the grid. Calling it again will
   * simply return the existing instance (ignore any passed options). There is also an initAll() version that support
   * multiple grids initialization at once. Or you can use addGrid() to create the entire grid from JSON.
   * @param options grid options (optional)
   * @param elOrString element or CSS selector (first one used) to convert to a grid (default to '.grid-stack' class selector)
   *
   * @example
   * const grid = GridStack.init();
   *
   * Note: the HTMLElement (of type GridHTMLElement) will store a `gridstack: GridStack` value that can be retrieve later
   * const grid = document.querySelector('.grid-stack').gridstack;
   */
  static init(t = {}, e = ".grid-stack") {
    if (typeof document > "u")
      return null;
    const i = d.getGridElement(e);
    return i ? (i.gridstack || (i.gridstack = new d(i, h.cloneDeep(t))), i.gridstack) : (console.error(typeof e == "string" ? 'GridStack.initAll() no grid was found with selector "' + e + `" - element missing or wrong selector ?
Note: ".grid-stack" is required for proper CSS styling and drag/drop, and is the default selector.` : "GridStack.init() no grid element was passed."), null);
  }
  /**
   * Will initialize a list of elements (given a selector) and return an array of grids.
   * @param options grid options (optional)
   * @param selector elements selector to convert to grids (default to '.grid-stack' class selector)
   *
   * @example
   * const grids = GridStack.initAll();
   * grids.forEach(...)
   */
  static initAll(t = {}, e = ".grid-stack") {
    const i = [];
    return typeof document > "u" || (d.getGridElements(e).forEach((s) => {
      s.gridstack || (s.gridstack = new d(s, h.cloneDeep(t))), i.push(s.gridstack);
    }), i.length === 0 && console.error('GridStack.initAll() no grid was found with selector "' + e + `" - element missing or wrong selector ?
Note: ".grid-stack" is required for proper CSS styling and drag/drop, and is the default selector.`)), i;
  }
  /**
   * call to create a grid with the given options, including loading any children from JSON structure. This will call GridStack.init(), then
   * grid.load() on any passed children (recursively). Great alternative to calling init() if you want entire grid to come from
   * JSON serialized data, including options.
   * @param parent HTML element parent to the grid
   * @param opt grids options used to initialize the grid, and list of children
   */
  static addGrid(t, e = {}) {
    if (!t)
      return null;
    let i = t;
    if (i.gridstack) {
      const o = i.gridstack;
      return e && (o.opts = { ...o.opts, ...e }), e.children !== void 0 && o.load(e.children), o;
    }
    return (!t.classList.contains("grid-stack") || d.addRemoveCB) && (d.addRemoveCB ? i = d.addRemoveCB(t, e, !0, !0) : i = h.createDiv(["grid-stack", e.class], t)), d.init(e, i);
  }
  /** call this method to register your engine instead of the default one.
   * See instead `GridStackOptions.engineClass` if you only need to
   * replace just one instance.
   */
  static registerEngine(t) {
    d.engineClass = t;
  }
  /**
   * @internal create placeholder DIV as needed
   * @returns the placeholder element for indicating drop zones during drag operations
   */
  get placeholder() {
    if (!this._placeholder) {
      this._placeholder = h.createDiv([this.opts.placeholderClass, p.itemClass, this.opts.itemClass]);
      const t = h.createDiv(["placeholder-content"], this._placeholder);
      this.opts.placeholderText && (t.textContent = this.opts.placeholderText);
    }
    return this._placeholder;
  }
  /**
   * Construct a grid item from the given element and options
   * @param el the HTML element tied to this grid after it's been initialized
   * @param opts grid options - public for classes to access, but use methods to modify!
   */
  constructor(t, e = {}) {
    this.el = t, this.opts = e, this.animationDelay = 310, this._gsEventHandler = {}, this._extraDragRow = 0, this.dragTransform = { xScale: 1, yScale: 1, xOffset: 0, yOffset: 0 }, t.gridstack = this, this.opts = e = e || {}, t.classList.contains("grid-stack") || this.el.classList.add("grid-stack"), e.row && (e.minRow = e.maxRow = e.row, delete e.row);
    const i = h.toNumber(t.getAttribute("gs-row"));
    e.column === "auto" && delete e.column, e.alwaysShowResizeHandle !== void 0 && (e._alwaysShowResizeHandle = e.alwaysShowResizeHandle);
    const s = e.columnOpts;
    if (s) {
      const g = s.breakpoints;
      !s.columnWidth && !g?.length ? delete e.columnOpts : (s.columnMax = s.columnMax || 12, g?.length > 1 && g.sort((a, l) => (l.w || 0) - (a.w || 0)));
    }
    const n = {
      ...h.cloneDeep(p),
      column: h.toNumber(t.getAttribute("gs-column")) || p.column,
      minRow: i || h.toNumber(t.getAttribute("gs-min-row")) || p.minRow,
      maxRow: i || h.toNumber(t.getAttribute("gs-max-row")) || p.maxRow,
      staticGrid: h.toBool(t.getAttribute("gs-static")) || p.staticGrid,
      sizeToContent: h.toBool(t.getAttribute("gs-size-to-content")) || void 0,
      draggable: {
        handle: (e.handleClass ? "." + e.handleClass : e.handle ? e.handle : "") || p.draggable.handle
      },
      removableOptions: {
        accept: e.itemClass || p.removableOptions.accept,
        decline: p.removableOptions.decline
      }
    };
    t.getAttribute("gs-animate") && (n.animate = h.toBool(t.getAttribute("gs-animate"))), e = h.defaults(e, n), this._initMargin(), this.checkDynamicColumn(), this._updateColumnVar(e), e.rtl === "auto" && (e.rtl = t.style.direction === "rtl"), e.rtl && this.el.classList.add("grid-stack-rtl");
    const r = this.el.closest("." + p.itemClass)?.gridstackNode;
    if (r && (r.subGrid = this, this.parentGridNode = r, this.el.classList.add("grid-stack-nested"), r.el.classList.add("grid-stack-sub-grid")), this._isAutoCellHeight = e.cellHeight === "auto", this._isAutoCellHeight || e.cellHeight === "initial")
      this.cellHeight(void 0);
    else {
      typeof e.cellHeight == "number" && e.cellHeightUnit && e.cellHeightUnit !== p.cellHeightUnit && (e.cellHeight = e.cellHeight + e.cellHeightUnit, delete e.cellHeightUnit);
      const g = e.cellHeight;
      delete e.cellHeight, this.cellHeight(g);
    }
    e.alwaysShowResizeHandle === "mobile" && (e.alwaysShowResizeHandle = D), this._setStaticClass();
    const u = e.engineClass || d.engineClass || N;
    if (this.engine = new u({
      column: this.getColumn(),
      float: e.float,
      maxRow: e.maxRow,
      onChange: (g) => {
        g.forEach((a) => {
          const l = a.el;
          l && (a._removeDOM ? (l && l.remove(), delete a._removeDOM) : this._writePosAttr(l, a));
        }), this._updateContainerHeight();
      }
    }), e.auto && (this.batchUpdate(), this.engine._loading = !0, this.getGridItems().forEach((g) => this._prepareElement(g)), delete this.engine._loading, this.batchUpdate(!1)), e.children) {
      const g = e.children;
      delete e.children, g.length && this.load(g);
    }
    this.setAnimation(), e.subGridDynamic && !_.pauseDrag && (_.pauseDrag = !0), e.draggable?.pause !== void 0 && (_.pauseDrag = e.draggable.pause), this._setupRemoveDrop(), this._setupAcceptWidget(), this._updateResizeEvent();
  }
  _updateColumnVar(t = this.opts) {
    this.el.classList.add("gs-" + t.column), typeof t.column == "number" && this.el.style.setProperty("--gs-column-width", `${100 / t.column}%`);
  }
  /**
   * add a new widget and returns it.
   *
   * Widget will be always placed even if result height is more than actual grid height.
   * You need to use `willItFit()` before calling addWidget for additional check.
   * See also `makeWidget(el)` for DOM element.
   *
   * @example
   * const grid = GridStack.init();
   * grid.addWidget({w: 3, content: 'hello'});
   *
   * @param w GridStackWidget definition. used MakeWidget(el) if you have dom element instead.
   */
  addWidget(t) {
    if (!t)
      return;
    if (typeof t == "string") {
      console.error("V11: GridStack.addWidget() does not support string anymore. see #2736");
      return;
    }
    if (t.ELEMENT_NODE)
      return console.error("V11: GridStack.addWidget() does not support HTMLElement anymore. use makeWidget()"), this.makeWidget(t);
    let e, i = t;
    if (i.grid = this, i.el ? e = i.el : d.addRemoveCB ? e = d.addRemoveCB(this.el, t, !0, !1) : e = this.createWidgetDivs(i), !e)
      return;
    if (i = e.gridstackNode, i && e.parentElement === this.el && this.engine.nodes.find((n) => n._id === i._id))
      return e;
    const s = this._readAttr(e);
    return h.defaults(t, s), this.engine.prepareNode(t), this.el.appendChild(e), this.makeWidget(e, t), e;
  }
  /**
   * Create the default grid item divs and content (possibly lazy loaded) by using GridStack.renderCB().
   *
   * @param n GridStackNode definition containing widget configuration
   * @returns the created HTML element with proper grid item structure
   *
   * @example
   * const element = grid.createWidgetDivs({ w: 2, h: 1, content: 'Hello World' });
   */
  createWidgetDivs(t) {
    const e = h.createDiv(["grid-stack-item", this.opts.itemClass]), i = h.createDiv(["grid-stack-item-content"], e);
    return h.lazyLoad(t) ? t.visibleObservable || (t.visibleObservable = new IntersectionObserver(([s]) => {
      s.isIntersecting && (t.visibleObservable?.disconnect(), delete t.visibleObservable, d.renderCB(i, t), t.grid?.prepareDragDrop(t.el));
    }), window.setTimeout(() => t.visibleObservable?.observe(e))) : d.renderCB(i, t), e;
  }
  /**
   * Convert an existing gridItem element into a sub-grid with the given (optional) options, else inherit them
   * from the parent's subGrid options.
   * @param el gridItem element to convert
   * @param ops (optional) sub-grid options, else default to node, then parent settings, else defaults
   * @param nodeToAdd (optional) node to add to the newly created sub grid (used when dragging over existing regular item)
   * @param saveContent if true (default) the html inside .grid-stack-content will be saved to child widget
   * @returns newly created grid
   */
  makeSubGrid(t, e, i, s = !0) {
    let n = t.gridstackNode;
    if (n || (n = this.makeWidget(t).gridstackNode), n.subGrid?.el)
      return n.subGrid;
    let o, r = this;
    for (; r && !o; )
      o = r.opts?.subGridOpts, r = r.parentGridNode?.grid;
    e = h.cloneDeep({
      // by default sub-grid inherit from us | parent, other than id, children, etc...
      ...this.opts,
      id: void 0,
      children: void 0,
      column: "auto",
      columnOpts: void 0,
      layout: "list",
      subGridOpts: void 0,
      ...o || {},
      ...e || n.subGridOpts || {}
    }), n.subGridOpts = e;
    let u;
    e.column === "auto" && (u = !0, e.column = Math.max(n.w || 1, i?.w || 1), delete e.columnOpts);
    let g = n.el.querySelector(".grid-stack-item-content"), a, l;
    if (s && (this._removeDD(n.el), l = { ...n, x: 0, y: 0 }, h.removeInternalForSave(l), delete l.subGridOpts, n.content && (l.content = n.content, delete n.content), d.addRemoveCB ? a = d.addRemoveCB(this.el, l, !0, !1) : (a = h.createDiv(["grid-stack-item"]), a.appendChild(g), g = h.createDiv(["grid-stack-item-content"], n.el)), this.prepareDragDrop(n.el)), i) {
      const b = u ? e.column : n.w, m = n.h + i.h, v = n.el.style;
      v.transition = "none", this.update(n.el, { w: b, h: m }), setTimeout(() => v.transition = null);
    }
    const c = n.subGrid = d.addGrid(g, e);
    return i?._moving && (c._isTemp = !0), u && (c._autoColumn = !0), s && c.makeWidget(a, l), i && (i._moving ? window.setTimeout(() => h.simulateMouseEvent(i._event, "mouseenter", c.el), 0) : c.makeWidget(n.el, n)), this.resizeToContentCheck(!1, n), c;
  }
  /**
   * called when an item was converted into a nested grid to accommodate a dragged over item, but then item leaves - return back
   * to the original grid-item. Also called to remove empty sub-grids when last item is dragged out (since re-creating is simple)
   */
  removeAsSubGrid(t) {
    const e = this.parentGridNode?.grid;
    e && (e.batchUpdate(), e.removeWidget(this.parentGridNode.el, !0, !0), this.engine.nodes.forEach((i) => {
      i.x += this.parentGridNode.x, i.y += this.parentGridNode.y, e.makeWidget(i.el, i);
    }), e.batchUpdate(!1), this.parentGridNode && delete this.parentGridNode.subGrid, delete this.parentGridNode, t && window.setTimeout(() => h.simulateMouseEvent(t._event, "mouseenter", e.el), 0));
  }
  /**
   * saves the current layout returning a list of widgets for serialization which might include any nested grids.
   * @param saveContent if true (default) the latest html inside .grid-stack-content will be saved to GridStackWidget.content field, else it will
   * be removed.
   * @param saveGridOpt if true (default false), save the grid options itself, so you can call the new GridStack.addGrid()
   * to recreate everything from scratch. GridStackOptions.children would then contain the widget list instead.
   * @param saveCB callback for each node -> widget, so application can insert additional data to be saved into the widget data structure.
   * @param column if provided, the grid will be saved for the given column size (IFF we have matching internal saved layout, or current layout).
   * Otherwise it will use the largest possible layout (say 12 even if rendering at 1 column) so we can restore to all layouts.
   * NOTE: if you want to save to currently display layout, pass this.getColumn() as column.
   * NOTE2: nested grids will ALWAYS save to the container size to be in sync with parent.
   * @returns list of widgets or full grid option, including .children list of widgets
   */
  save(t = !0, e = !1, i = d.saveCB, s) {
    const n = this.engine.save(t, i, s);
    if (n.forEach((o) => {
      if (t && o.el && !o.subGrid && !i) {
        const r = o.el.querySelector(".grid-stack-item-content");
        o.content = r?.innerHTML, o.content || delete o.content;
      } else if (!t && !i && delete o.content, o.subGrid?.el) {
        const r = o.w || o.subGrid.getColumn(), u = o.subGrid.save(t, e, i, r);
        o.subGridOpts = e ? u : { children: u }, delete o.subGrid;
      }
      delete o.el;
    }), e) {
      const o = h.cloneDeep(this.opts);
      o.marginBottom === o.marginTop && o.marginRight === o.marginLeft && o.marginTop === o.marginRight && (o.margin = o.marginTop, delete o.marginTop, delete o.marginRight, delete o.marginBottom, delete o.marginLeft), o.rtl === (this.el.style.direction === "rtl") && (o.rtl = "auto"), this._isAutoCellHeight && (o.cellHeight = "auto"), this._autoColumn && (o.column = "auto");
      const r = o._alwaysShowResizeHandle;
      return delete o._alwaysShowResizeHandle, r !== void 0 ? o.alwaysShowResizeHandle = r : delete o.alwaysShowResizeHandle, h.removeInternalAndSame(o, p), o.children = n, o;
    }
    return n;
  }
  /**
   * Load widgets from a list. This will call update() on each (matching by id) or add/remove widgets that are not there.
   * Used to restore a grid layout for a saved layout list (see `save()`).
   *
   * @param items list of widgets definition to update/create
   * @param addRemove boolean (default true) or callback method can be passed to control if and how missing widgets can be added/removed, giving
   * the user control of insertion.
   * @returns the grid instance for chaining
   *
   * @example
   * // Basic usage with saved layout
   * const savedLayout = grid.save(); // Save current layout
   * // ... later restore it
   * grid.load(savedLayout);
   *
   * // Load with custom add/remove callback
   * grid.load(layout, (items, grid, add) => {
   *   if (add) {
   *     // Custom logic for adding new widgets
   *     items.forEach(item => {
   *       const el = document.createElement('div');
   *       el.innerHTML = item.content || '';
   *       grid.addWidget(el, item);
   *     });
   *   } else {
   *     // Custom logic for removing widgets
   *     items.forEach(item => grid.removeWidget(item.el));
   *   }
   * });
   *
   * // Load without adding/removing missing widgets
   * grid.load(layout, false);
   *
   * @see {@link http://gridstackjs.com/demo/serialization.html} for complete example
   */
  load(t, e = d.addRemoveCB || !0) {
    t = h.cloneDeep(t);
    const i = this.getColumn();
    t.forEach((a) => {
      a.w = a.w || a.minW || 1, a.h = a.h || a.minH || 1;
    }), t = h.sort(t), this.engine.skipCacheUpdate = this._ignoreLayoutsNodeChange = !0;
    let s = 0;
    t.forEach((a) => {
      s = Math.max(s, (a.x || 0) + a.w);
    }), s > this.engine.defaultColumn && (this.engine.defaultColumn = s), s > i && (this.engine.nodes.length === 0 && this.responseLayout ? (this.engine.nodes = t, this.engine.columnChanged(s, i, this.responseLayout), t = this.engine.nodes, this.engine.nodes = [], delete this.responseLayout) : this.engine.cacheLayout(t, s, !0));
    const n = d.addRemoveCB;
    typeof e == "function" && (d.addRemoveCB = e);
    const o = [];
    this.batchUpdate();
    const r = !this.engine.nodes.length, u = r && this.opts.animate;
    u && this.setAnimation(!1), !r && e && [...this.engine.nodes].forEach((l) => {
      if (!l.id)
        return;
      h.find(t, l.id) || (d.addRemoveCB && d.addRemoveCB(this.el, l, !1, !1), o.push(l), this.removeWidget(l.el, !0, !1));
    }), this.engine._loading = !0;
    const g = [];
    return this.engine.nodes = this.engine.nodes.filter((a) => h.find(t, a.id) ? (g.push(a), !1) : !0), t.forEach((a) => {
      const l = h.find(g, a.id);
      if (l) {
        if (h.shouldSizeToContent(l) && (a.h = l.h), this.engine.nodeBoundFix(a), (a.autoPosition || a.x === void 0 || a.y === void 0) && (a.w = a.w || l.w, a.h = a.h || l.h, this.engine.findEmptyPosition(a)), this.engine.nodes.push(l), h.samePos(l, a) && this.engine.nodes.length > 1 && (this.moveNode(l, { ...a, forceCollide: !0 }), h.copyPos(a, l)), this.update(l.el, a), a.subGridOpts?.children) {
          const c = l.el.querySelector(".grid-stack");
          c && c.gridstack && c.gridstack.load(a.subGridOpts.children);
        }
      } else e && this.addWidget(a);
    }), delete this.engine._loading, this.engine.removedNodes = o, this.batchUpdate(!1), delete this._ignoreLayoutsNodeChange, delete this.engine.skipCacheUpdate, n ? d.addRemoveCB = n : delete d.addRemoveCB, u && this.setAnimation(!0, !0), this;
  }
  /**
   * use before calling a bunch of `addWidget()` to prevent un-necessary relayouts in between (more efficient)
   * and get a single event callback. You will see no changes until `batchUpdate(false)` is called.
   */
  batchUpdate(t = !0) {
    return this.engine.batchUpdate(t), t || (this._updateContainerHeight(), this._triggerRemoveEvent(), this._triggerAddEvent(), this._triggerChangeEvent()), this;
  }
  /**
   * Gets the current cell height in pixels. This takes into account the unit type and converts to pixels if necessary.
   *
   * @param forcePixel if true, forces conversion to pixels even when cellHeight is specified in other units
   * @returns the cell height in pixels
   *
   * @example
   * const height = grid.getCellHeight();
   * console.log('Cell height:', height, 'px');
   *
   * // Force pixel conversion
   * const pixelHeight = grid.getCellHeight(true);
   */
  getCellHeight(t = !1) {
    if (this.opts.cellHeight && this.opts.cellHeight !== "auto" && (!t || !this.opts.cellHeightUnit || this.opts.cellHeightUnit === "px"))
      return this.opts.cellHeight;
    if (this.opts.cellHeightUnit === "rem")
      return this.opts.cellHeight * parseFloat(getComputedStyle(document.documentElement).fontSize);
    if (this.opts.cellHeightUnit === "em")
      return this.opts.cellHeight * parseFloat(getComputedStyle(this.el).fontSize);
    if (this.opts.cellHeightUnit === "cm")
      return this.opts.cellHeight * (96 / 2.54);
    if (this.opts.cellHeightUnit === "mm")
      return this.opts.cellHeight * (96 / 2.54) / 10;
    const e = this.el.querySelector("." + this.opts.itemClass);
    if (e) {
      const s = h.toNumber(e.getAttribute("gs-h")) || 1;
      return Math.round(e.offsetHeight / s);
    }
    const i = parseInt(this.el.getAttribute("gs-current-row"));
    return i ? Math.round(this.el.getBoundingClientRect().height / i) : this.opts.cellHeight;
  }
  /**
   * Update current cell height - see `GridStackOptions.cellHeight` for format by updating eh Browser CSS variable.
   *
   * @param val the cell height. Options:
   *   - `undefined`: cells content will be made square (match width minus margin)
   *   - `0`: the CSS will be generated by the application instead
   *   - number: height in pixels
   *   - string: height with units (e.g., '70px', '5rem', '2em')
   * @returns the grid instance for chaining
   *
   * @example
   * grid.cellHeight(100);     // 100px height
   * grid.cellHeight('70px');  // explicit pixel height
   * grid.cellHeight('5rem');  // relative to root font size
   * grid.cellHeight(grid.cellWidth() * 1.2); // aspect ratio
   * grid.cellHeight('auto');  // auto-size based on content
   */
  cellHeight(t) {
    if (t !== void 0 && this._isAutoCellHeight !== (t === "auto") && (this._isAutoCellHeight = t === "auto", this._updateResizeEvent()), (t === "initial" || t === "auto") && (t = void 0), t === void 0) {
      const i = -this.opts.marginRight - this.opts.marginLeft + this.opts.marginTop + this.opts.marginBottom;
      t = this.cellWidth() + i;
    }
    const e = h.parseHeight(t);
    return this.opts.cellHeightUnit === e.unit && this.opts.cellHeight === e.h ? this : (this.opts.cellHeightUnit = e.unit, this.opts.cellHeight = e.h, this.el.style.setProperty("--gs-cell-height", `${this.opts.cellHeight}${this.opts.cellHeightUnit}`), this._updateContainerHeight(), this.resizeToContentCheck(), this);
  }
  /** Gets current cell width. */
  /**
   * Gets the current cell width in pixels. This is calculated based on the grid container width divided by the number of columns.
   *
   * @returns the cell width in pixels
   *
   * @example
   * const width = grid.cellWidth();
   * console.log('Cell width:', width, 'px');
   *
   * // Use cell width to calculate widget dimensions
   * const widgetWidth = width * 3; // For a 3-column wide widget
   */
  cellWidth() {
    return this._widthOrContainer() / this.getColumn();
  }
  /** return our expected width (or parent) , and optionally of window for dynamic column check */
  _widthOrContainer(t = !1) {
    return t && this.opts.columnOpts?.breakpointForWindow ? window.innerWidth : this.el.clientWidth || this.el.parentElement.clientWidth || window.innerWidth;
  }
  /** checks for dynamic column count for our current size, returning true if changed */
  checkDynamicColumn() {
    const t = this.opts.columnOpts;
    if (!t || !t.columnWidth && !t.breakpoints?.length)
      return !1;
    const e = this.getColumn();
    let i = e;
    const s = this._widthOrContainer(!0);
    if (t.columnWidth)
      i = Math.min(Math.round(s / t.columnWidth) || 1, t.columnMax);
    else {
      i = t.columnMax;
      let n = 0;
      for (; n < t.breakpoints.length && s <= t.breakpoints[n].w; )
        i = t.breakpoints[n++].c || e;
    }
    if (i !== e) {
      const n = t.breakpoints?.find((o) => o.c === i);
      return this.column(i, n?.layout || t.layout), !0;
    }
    return !1;
  }
  /**
   * Re-layout grid items to reclaim any empty space. This is useful after removing widgets
   * or when you want to optimize the layout.
   *
   * @param layout layout type. Options:
   *   - 'compact' (default): might re-order items to fill any empty space
   *   - 'list': keep the widget left->right order the same, even if that means leaving an empty slot if things don't fit
   * @param doSort re-sort items first based on x,y position. Set to false to do your own sorting ahead (default: true)
   * @returns the grid instance for chaining
   *
   * @example
   * // Compact layout after removing widgets
   * grid.removeWidget('.widget-to-remove');
   * grid.compact();
   *
   * // Use list layout (preserve order)
   * grid.compact('list');
   *
   * // Compact without sorting first
   * grid.compact('compact', false);
   */
  compact(t = "compact", e = !0) {
    return this.engine.compact(t, e), this._triggerChangeEvent(), this;
  }
  /**
   * Set the number of columns in the grid. Will update existing widgets to conform to new number of columns,
   * as well as cache the original layout so you can revert back to previous positions without loss.
   *
   * Requires `gridstack-extra.css` or `gridstack-extra.min.css` for [2-11] columns,
   * else you will need to generate correct CSS.
   * See: https://github.com/gridstack/gridstack.js#change-grid-columns
   *
   * @param column Integer > 0 (default 12)
   * @param layout specify the type of re-layout that will happen. Options:
   *   - 'moveScale' (default): scale widget positions and sizes
   *   - 'move': keep widget sizes, only move positions
   *   - 'scale': keep widget positions, only scale sizes
   *   - 'none': don't change widget positions or sizes
   *   Note: items will never be outside of the current column boundaries.
   *   Ignored for `column=1` as we always want to vertically stack.
   * @returns the grid instance for chaining
   *
   * @example
   * // Change to 6 columns with default scaling
   * grid.column(6);
   *
   * // Change to 4 columns, only move positions
   * grid.column(4, 'move');
   *
   * // Single column layout (vertical stack)
   * grid.column(1);
   */
  column(t, e = "moveScale") {
    if (!t || t < 1 || this.opts.column === t)
      return this;
    const i = this.getColumn();
    return this.opts.column = t, this.engine ? (this.engine.column = t, this.el.classList.remove("gs-" + i), this._updateColumnVar(), this.engine.columnChanged(i, t, e), this._isAutoCellHeight && this.cellHeight(), this.resizeToContentCheck(!0), this._ignoreLayoutsNodeChange = !0, this._triggerChangeEvent(), delete this._ignoreLayoutsNodeChange, this) : (this.responseLayout = e, this);
  }
  /**
   * Get the number of columns in the grid (default 12).
   *
   * @returns the current number of columns in the grid
   *
   * @example
   * const columnCount = grid.getColumn(); // returns 12 by default
   */
  getColumn() {
    return this.opts.column;
  }
  /**
   * Returns an array of grid HTML elements (no placeholder) - used to iterate through our children in DOM order.
   * This method excludes placeholder elements and returns only actual grid items.
   *
   * @returns array of GridItemHTMLElement instances representing all grid items
   *
   * @example
   * const items = grid.getGridItems();
   * items.forEach(item => {
   *   console.log('Item ID:', item.gridstackNode.id);
   * });
   */
  getGridItems() {
    return Array.from(this.el.children).filter((t) => t.matches("." + this.opts.itemClass) && !t.matches("." + this.opts.placeholderClass));
  }
  /**
   * Returns true if change callbacks should be ignored due to column change, sizeToContent, loading, etc.
   * This is useful for callers who want to implement dirty flag functionality.
   *
   * @returns true if change callbacks are currently being ignored
   *
   * @example
   * if (!grid.isIgnoreChangeCB()) {
   *   // Process the change event
   *   console.log('Grid layout changed');
   * }
   */
  isIgnoreChangeCB() {
    return this._ignoreLayoutsNodeChange;
  }
  /**
   * Destroys a grid instance. DO NOT CALL any methods or access any vars after this as it will free up members.
   * @param removeDOM if `false` grid and items HTML elements will not be removed from the DOM (Optional. Default `true`).
   */
  destroy(t = !0) {
    if (this.el)
      return this.offAll(), this._updateResizeEvent(!0), this.setStatic(!0, !1), this.setAnimation(!1), t ? this.el.parentNode.removeChild(this.el) : (this.removeAll(t), this.el.removeAttribute("gs-current-row")), this.parentGridNode && delete this.parentGridNode.subGrid, delete this.parentGridNode, delete this.opts, delete this._placeholder?.gridstackNode, delete this._placeholder, delete this.engine, delete this.el.gridstack, delete this.el, this;
  }
  /**
   * Enable/disable floating widgets (default: `false`). When enabled, widgets can float up to fill empty spaces.
   * See [example](http://gridstackjs.com/demo/float.html)
   *
   * @param val true to enable floating, false to disable
   * @returns the grid instance for chaining
   *
   * @example
   * grid.float(true);  // Enable floating
   * grid.float(false); // Disable floating (default)
   */
  float(t) {
    return this.opts.float !== t && (this.opts.float = this.engine.float = t, this._triggerChangeEvent()), this;
  }
  /**
   * Get the current float mode setting.
   *
   * @returns true if floating is enabled, false otherwise
   *
   * @example
   * const isFloating = grid.getFloat();
   * console.log('Floating enabled:', isFloating);
   */
  getFloat() {
    return this.engine.float;
  }
  /**
   * Get the position of the cell under a pixel on screen.
   * @param position the position of the pixel to resolve in
   * absolute coordinates, as an object with top and left properties
   * @param useDocRelative if true, value will be based on document position vs parent position (Optional. Default false).
   * Useful when grid is within `position: relative` element
   *
   * Returns an object with properties `x` and `y` i.e. the column and row in the grid.
   */
  getCellFromPixel(t, e = !1) {
    const i = this.el.getBoundingClientRect();
    let s;
    e ? s = { top: i.top + document.documentElement.scrollTop, left: i.left } : s = { top: this.el.offsetTop, left: this.el.offsetLeft };
    const n = t.left - s.left, o = t.top - s.top, r = i.width / this.getColumn(), u = i.height / parseInt(this.el.getAttribute("gs-current-row"));
    return { x: Math.floor(n / r), y: Math.floor(o / u) };
  }
  /**
   * Returns the current number of rows, which will be at least `minRow` if set.
   * The row count is based on the highest positioned widget in the grid.
   *
   * @returns the current number of rows in the grid
   *
   * @example
   * const rowCount = grid.getRow();
   * console.log('Grid has', rowCount, 'rows');
   */
  getRow() {
    return Math.max(this.engine.getRow(), this.opts.minRow || 0);
  }
  /**
   * Checks if the specified rectangular area is empty (no widgets occupy any part of it).
   *
   * @param x the x coordinate (column) of the area to check
   * @param y the y coordinate (row) of the area to check
   * @param w the width in columns of the area to check
   * @param h the height in rows of the area to check
   * @returns true if the area is completely empty, false if any widget overlaps
   *
   * @example
   * // Check if a 2x2 area at position (1,1) is empty
   * if (grid.isAreaEmpty(1, 1, 2, 2)) {
   *   console.log('Area is available for placement');
   * }
   */
  isAreaEmpty(t, e, i, s) {
    return this.engine.isAreaEmpty(t, e, i, s);
  }
  /**
   * If you add elements to your grid by hand (or have some framework creating DOM), you have to tell gridstack afterwards to make them widgets.
   * If you want gridstack to add the elements for you, use `addWidget()` instead.
   * Makes the given element a widget and returns it.
   *
   * @param els widget or single selector to convert.
   * @param options widget definition to use instead of reading attributes or using default sizing values
   * @returns the converted GridItemHTMLElement
   *
   * @example
   * const grid = GridStack.init();
   *
   * // Create HTML content manually, possibly looking like:
   * // <div id="item-1" gs-x="0" gs-y="0" gs-w="3" gs-h="2"></div>
   * grid.el.innerHTML = '<div id="item-1" gs-w="3"></div><div id="item-2"></div>';
   *
   * // Convert existing elements to widgets
   * grid.makeWidget('#item-1'); // Uses gs-* attributes from DOM
   * grid.makeWidget('#item-2', {w: 2, h: 1, content: 'Hello World'});
   *
   * // Or pass DOM element directly
   * const element = document.getElementById('item-3');
   * grid.makeWidget(element, {x: 0, y: 1, w: 4, h: 2});
   */
  makeWidget(t, e) {
    const i = d.getElement(t);
    if (!i || i.gridstackNode)
      return i;
    i.parentElement || this.el.appendChild(i), this._prepareElement(i, !0, e);
    const s = i.gridstackNode;
    this._updateContainerHeight(), s.subGridOpts && this.makeSubGrid(i, s.subGridOpts, void 0, !1);
    let n;
    return this.opts.column === 1 && !this._ignoreLayoutsNodeChange && (n = this._ignoreLayoutsNodeChange = !0), this._triggerAddEvent(), this._triggerChangeEvent(), n && delete this._ignoreLayoutsNodeChange, i;
  }
  on(t, e) {
    return t.indexOf(" ") !== -1 ? (t.split(" ").forEach((s) => this.on(s, e)), this) : (t === "change" || t === "added" || t === "removed" || t === "enable" || t === "disable" ? (t === "enable" || t === "disable" ? this._gsEventHandler[t] = (s) => e(s) : this._gsEventHandler[t] = (s) => {
      s.detail && e(s, s.detail);
    }, this.el.addEventListener(t, this._gsEventHandler[t])) : t === "drag" || t === "dragstart" || t === "dragstop" || t === "resizestart" || t === "resize" || t === "resizestop" || t === "dropped" || t === "resizecontent" ? this._gsEventHandler[t] = e : console.error("GridStack.on(" + t + ") event not supported"), this);
  }
  /**
   * unsubscribe from the 'on' event GridStackEvent
   * @param name of the event (see possible values) or list of names space separated
   */
  off(t) {
    return t.indexOf(" ") !== -1 ? (t.split(" ").forEach((i) => this.off(i)), this) : ((t === "change" || t === "added" || t === "removed" || t === "enable" || t === "disable") && this._gsEventHandler[t] && this.el.removeEventListener(t, this._gsEventHandler[t]), delete this._gsEventHandler[t], this);
  }
  /**
   * Remove all event handlers from the grid. This is useful for cleanup when destroying a grid.
   *
   * @returns the grid instance for chaining
   *
   * @example
   * grid.offAll(); // Remove all event listeners
   */
  offAll() {
    return Object.keys(this._gsEventHandler).forEach((t) => this.off(t)), this;
  }
  /**
   * Removes widget from the grid.
   * @param el  widget or selector to modify
   * @param removeDOM if `false` DOM element won't be removed from the tree (Default? true).
   * @param triggerEvent if `false` (quiet mode) element will not be added to removed list and no 'removed' callbacks will be called (Default? true).
   */
  removeWidget(t, e = !0, i = !0) {
    return t ? (d.getElements(t).forEach((s) => {
      if (s.parentElement && s.parentElement !== this.el)
        return;
      let n = s.gridstackNode;
      n || (n = this.engine.nodes.find((o) => s === o.el)), n && (e && d.addRemoveCB && d.addRemoveCB(this.el, n, !1, !1), delete s.gridstackNode, this._removeDD(s), this.engine.removeNode(n, e, i), e && s.parentElement && s.remove());
    }), i && (this._triggerRemoveEvent(), this._triggerChangeEvent()), this) : (console.error("Error: GridStack.removeWidget(undefined) called"), this);
  }
  /**
   * Removes all widgets from the grid.
   * @param removeDOM if `false` DOM elements won't be removed from the tree (Default? `true`).
   * @param triggerEvent if `false` (quiet mode) element will not be added to removed list and no 'removed' callbacks will be called (Default? true).
   */
  removeAll(t = !0, e = !0) {
    return this.engine.nodes.forEach((i) => {
      t && d.addRemoveCB && d.addRemoveCB(this.el, i, !1, !1), delete i.el.gridstackNode, this.opts.staticGrid || this._removeDD(i.el);
    }), this.engine.removeAll(t, e), e && this._triggerRemoveEvent(), this;
  }
  /**
   * Toggle the grid animation state.  Toggles the `grid-stack-animate` class.
   * @param doAnimate if true the grid will animate.
   * @param delay if true setting will be set on next event loop.
   */
  setAnimation(t = this.opts.animate, e) {
    return e ? setTimeout(() => {
      this.opts && this.setAnimation(t);
    }) : t ? this.el.classList.add("grid-stack-animate") : this.el.classList.remove("grid-stack-animate"), this.opts.animate = t, this;
  }
  /** @internal */
  hasAnimationCSS() {
    return this.el.classList.contains("grid-stack-animate");
  }
  /**
   * Toggle the grid static state, which permanently removes/add Drag&Drop support, unlike disable()/enable() that just turns it off/on.
   * Also toggle the grid-stack-static class.
   * @param val if true the grid become static.
   * @param updateClass true (default) if css class gets updated
   * @param recurse true (default) if sub-grids also get updated
   */
  setStatic(t, e = !0, i = !0) {
    return !!this.opts.staticGrid === t ? this : (t ? this.opts.staticGrid = !0 : delete this.opts.staticGrid, this._setupRemoveDrop(), this._setupAcceptWidget(), this.engine.nodes.forEach((s) => {
      this.prepareDragDrop(s.el), s.subGrid && i && s.subGrid.setStatic(t, e, i);
    }), e && this._setStaticClass(), this);
  }
  /**
   * Updates the passed in options on the grid (similar to update(widget) for for the grid options).
   * @param options PARTIAL grid options to update - only items specified will be updated.
   * NOTE: not all options updating are currently supported (lot of code, unlikely to change)
   */
  updateOptions(t) {
    const e = this.opts;
    return t === e ? this : (t.acceptWidgets !== void 0 && (e.acceptWidgets = t.acceptWidgets, this._setupAcceptWidget()), t.animate !== void 0 && this.setAnimation(t.animate), t.cellHeight && this.cellHeight(t.cellHeight), t.class !== void 0 && t.class !== e.class && (e.class && this.el.classList.remove(e.class), t.class && this.el.classList.add(t.class)), t.columnOpts ? (this.opts.columnOpts = t.columnOpts, this.checkDynamicColumn()) : t.columnOpts === null && this.opts.columnOpts ? (delete this.opts.columnOpts, this._updateResizeEvent()) : typeof t.column == "number" && this.column(t.column), t.margin !== void 0 && this.margin(t.margin), t.staticGrid !== void 0 && this.setStatic(t.staticGrid), t.disableDrag !== void 0 && !t.staticGrid && this.enableMove(!t.disableDrag), t.disableResize !== void 0 && !t.staticGrid && this.enableResize(!t.disableResize), t.float !== void 0 && this.float(t.float), t.row !== void 0 ? (e.minRow = e.maxRow = e.row = t.row, this._updateContainerHeight()) : (t.minRow !== void 0 && (e.minRow = t.minRow, this._updateContainerHeight()), t.maxRow !== void 0 && (e.maxRow = t.maxRow)), t.children?.length && this.load(t.children), this);
  }
  /**
   * Updates widget position/size and other info. This is used to change widget properties after creation.
   * Can update position, size, content, and other widget properties.
   *
   * Note: If you need to call this on all nodes, use load() instead which will update what changed.
   * Setting the same x,y for multiple items will be indeterministic and likely unwanted.
   *
   * @param els widget element(s) or selector to modify
   * @param opt new widget options (x,y,w,h, etc.). Only those set will be updated.
   * @returns the grid instance for chaining
   *
   * @example
   * // Update widget size and position
   * grid.update('.my-widget', { x: 2, y: 1, w: 3, h: 2 });
   *
   * // Update widget content
   * grid.update(widget, { content: '<p>New content</p>' });
   *
   * // Update multiple properties
   * grid.update('#my-widget', {
   *   w: 4,
   *   h: 3,
   *   noResize: true,
   *   locked: true
   * });
   */
  update(t, e) {
    return d.getElements(t).forEach((i) => {
      const s = i?.gridstackNode;
      if (!s)
        return;
      const n = { ...h.copyPos({}, s), ...h.cloneDeep(e) };
      this.engine.nodeBoundFix(n), delete n.autoPosition;
      const o = ["x", "y", "w", "h"];
      let r;
      if (o.some((a) => n[a] !== void 0 && n[a] !== s[a]) && (r = {}, o.forEach((a) => {
        r[a] = n[a] !== void 0 ? n[a] : s[a], delete n[a];
      })), !r && (n.minW || n.minH || n.maxW || n.maxH) && (r = {}), n.content !== void 0) {
        const a = i.querySelector(".grid-stack-item-content");
        a && a.textContent !== n.content && (s.content = n.content, d.renderCB(a, n), s.subGrid?.el && (a.appendChild(s.subGrid.el), s.subGrid._updateContainerHeight())), delete n.content;
      }
      let u = !1, g = !1;
      for (const a in n)
        a[0] !== "_" && s[a] !== n[a] && (s[a] = n[a], u = !0, g = g || !this.opts.staticGrid && (a === "noResize" || a === "noMove" || a === "locked"));
      if (h.sanitizeMinMax(s), r) {
        const a = r.w !== void 0 && r.w !== s.w;
        this.moveNode(s, r), a && s.subGrid ? s.subGrid.onResize(this.hasAnimationCSS() ? s.w : void 0) : this.resizeToContentCheck(a, s), delete s._orig;
      }
      (r || u) && this._writeAttr(i, s), g && this.prepareDragDrop(s.el), d.updateCB && d.updateCB(s);
    }), this;
  }
  moveNode(t, e) {
    const i = t._updating;
    i || this.engine.cleanNodes().beginUpdate(t), this.engine.moveNode(t, e), this._updateContainerHeight(), i || (this._triggerChangeEvent(), this.engine.endUpdate());
  }
  /**
   * Updates widget height to match the content height to avoid vertical scrollbars or dead space.
   * This automatically adjusts the widget height based on its content size.
   *
   * Note: This assumes only 1 child under resizeToContentParent='.grid-stack-item-content'
   * (sized to gridItem minus padding) that represents the entire content size.
   *
   * @param el the grid item element to resize
   *
   * @example
   * // Resize a widget to fit its content
   * const widget = document.querySelector('.grid-stack-item');
   * grid.resizeToContent(widget);
   *
   * // This is commonly used with dynamic content:
   * widget.querySelector('.content').innerHTML = 'New longer content...';
   * grid.resizeToContent(widget);
   */
  resizeToContent(t) {
    if (!t || (t.classList.remove("size-to-content-max"), !t.clientHeight))
      return;
    const e = t.gridstackNode;
    if (!e)
      return;
    const i = e.grid;
    if (!i || t.parentElement !== i.el)
      return;
    const s = i.getCellHeight(!0);
    if (!s)
      return;
    let n = e.h ? e.h * s : t.clientHeight, o;
    if (e.resizeToContentParent && (o = t.querySelector(e.resizeToContentParent)), o || (o = t.querySelector(d.resizeToContentParent)), !o)
      return;
    const r = t.clientHeight - o.clientHeight, u = e.h ? e.h * s - r : o.clientHeight;
    let g;
    if (e.subGrid) {
      g = e.subGrid.getRow() * e.subGrid.getCellHeight(!0);
      const c = e.subGrid.el.getBoundingClientRect(), b = t.getBoundingClientRect();
      g += c.top - b.top;
    } else {
      if (e.subGridOpts?.children?.length)
        return;
      {
        const c = o.firstElementChild;
        if (!c) {
          console.error(`Error: GridStack.resizeToContent() widget id:${e.id} '${d.resizeToContentParent}'.firstElementChild is null, make sure to have a div like container. Skipping sizing.`);
          return;
        }
        g = c.getBoundingClientRect().height || u;
      }
    }
    if (u === g)
      return;
    n += g - u;
    let a = Math.ceil(n / s);
    const l = Number.isInteger(e.sizeToContent) ? e.sizeToContent : 0;
    l && a > l && (a = l, t.classList.add("size-to-content-max")), e.minH && a < e.minH ? a = e.minH : e.maxH && a > e.maxH && (a = e.maxH), a !== e.h && (i._ignoreLayoutsNodeChange = !0, i.moveNode(e, { h: a }), delete i._ignoreLayoutsNodeChange);
  }
  /** call the user resize (so they can do extra work) else our build in version */
  resizeToContentCBCheck(t) {
    d.resizeToContentCB ? d.resizeToContentCB(t) : this.resizeToContent(t);
  }
  /**
   * Rotate widgets by swapping their width and height. This is typically called when the user presses 'r' during dragging.
   * The rotation swaps the w/h dimensions and adjusts min/max constraints accordingly.
   *
   * @param els widget element(s) or selector to rotate
   * @param relative optional pixel coordinate relative to upper/left corner to rotate around (keeps that cell under cursor)
   * @returns the grid instance for chaining
   *
   * @example
   * // Rotate a specific widget
   * grid.rotate('.my-widget');
   *
   * // Rotate with relative positioning during drag
   * grid.rotate(widget, { left: 50, top: 30 });
   */
  rotate(t, e) {
    return d.getElements(t).forEach((i) => {
      const s = i.gridstackNode;
      if (!h.canBeRotated(s))
        return;
      const n = { w: s.h, h: s.w, minH: s.minW, minW: s.minH, maxH: s.maxW, maxW: s.maxH };
      if (e) {
        const r = e.left > 0 ? Math.floor(e.left / this.cellWidth()) : 0, u = e.top > 0 ? Math.floor(e.top / this.opts.cellHeight) : 0;
        n.x = s.x + r - (s.h - (u + 1)), n.y = s.y + u - r;
      }
      Object.keys(n).forEach((r) => {
        n[r] === void 0 && delete n[r];
      });
      const o = s._orig;
      this.update(i, n), s._orig = o;
    }), this;
  }
  /**
   * Updates the margins which will set all 4 sides at once - see `GridStackOptions.margin` for format options.
   * Supports CSS string format of 1, 2, or 4 values or a single number.
   *
   * @param value margin value - can be:
   *   - Single number: `10` (applies to all sides)
   *   - Two values: `'10px 20px'` (top/bottom, left/right)
   *   - Four values: `'10px 20px 5px 15px'` (top, right, bottom, left)
   * @returns the grid instance for chaining
   *
   * @example
   * grid.margin(10);           // 10px all sides
   * grid.margin('10px 20px');  // 10px top/bottom, 20px left/right
   * grid.margin('5px 10px 15px 20px'); // Different for each side
   */
  margin(t) {
    if (!(typeof t == "string" && t.split(" ").length > 1)) {
      const i = h.parseHeight(t);
      if (this.opts.marginUnit === i.unit && this.opts.margin === i.h)
        return;
    }
    return this.opts.margin = t, this.opts.marginTop = this.opts.marginBottom = this.opts.marginLeft = this.opts.marginRight = void 0, this._initMargin(), this;
  }
  /**
   * Returns the current margin value as a number (undefined if the 4 sides don't match).
   * This only returns a number if all sides have the same margin value.
   *
   * @returns the margin value in pixels, or undefined if sides have different values
   *
   * @example
   * const margin = grid.getMargin();
   * if (margin !== undefined) {
   *   console.log('Uniform margin:', margin, 'px');
   * } else {
   *   console.log('Margins are different on different sides');
   * }
   */
  getMargin() {
    return this.opts.margin;
  }
  /**
   * Returns true if the height of the grid will be less than the vertical
   * constraint. Always returns true if grid doesn't have height constraint.
   * @param node contains x,y,w,h,auto-position options
   *
   * @example
   * if (grid.willItFit(newWidget)) {
   *   grid.addWidget(newWidget);
   * } else {
   *   alert('Not enough free space to place the widget');
   * }
   */
  willItFit(t) {
    if (arguments.length > 1) {
      console.warn("gridstack.ts: `willItFit(x,y,w,h,autoPosition)` is deprecated. Use `willItFit({x, y,...})`. It will be removed soon");
      const e = arguments;
      let i = 0, s = { x: e[i++], y: e[i++], w: e[i++], h: e[i++], autoPosition: e[i++] };
      return this.willItFit(s);
    }
    return this.engine.willItFit(t);
  }
  /** @internal */
  _triggerChangeEvent() {
    if (this.engine.batchMode)
      return this;
    const t = this.engine.getDirtyNodes(!0);
    return t && t.length && (this._ignoreLayoutsNodeChange || this.engine.layoutsNodesChange(t), this._triggerEvent("change", t)), this.engine.saveInitial(), this;
  }
  /** @internal */
  _triggerAddEvent() {
    if (this.engine.batchMode)
      return this;
    if (this.engine.addedNodes?.length) {
      this._ignoreLayoutsNodeChange || this.engine.layoutsNodesChange(this.engine.addedNodes), this.engine.addedNodes.forEach((e) => {
        delete e._dirty;
      });
      const t = [...this.engine.addedNodes];
      this.engine.addedNodes = [], this._triggerEvent("added", t);
    }
    return this;
  }
  /** @internal */
  _triggerRemoveEvent() {
    if (this.engine.batchMode)
      return this;
    if (this.engine.removedNodes?.length) {
      const t = [...this.engine.removedNodes];
      this.engine.removedNodes = [], this._triggerEvent("removed", t);
    }
    return this;
  }
  /** @internal */
  _triggerEvent(t, e) {
    const i = e ? new CustomEvent(t, { bubbles: !1, detail: e }) : new Event(t);
    let s = this;
    for (; s.parentGridNode; )
      s = s.parentGridNode.grid;
    return s.el.dispatchEvent(i), this;
  }
  /** @internal */
  _updateContainerHeight() {
    if (!this.engine || this.engine.batchMode)
      return this;
    const t = this.parentGridNode;
    let e = this.getRow() + this._extraDragRow;
    const i = this.opts.cellHeight, s = this.opts.cellHeightUnit;
    if (!i)
      return this;
    if (!t && !this.opts.minRow) {
      const n = h.parseHeight(getComputedStyle(this.el).minHeight);
      if (n.h > 0 && n.unit === s) {
        const o = Math.floor(n.h / i);
        e < o && (e = o);
      }
    }
    return this.el.setAttribute("gs-current-row", String(e)), this.el.style.removeProperty("min-height"), this.el.style.removeProperty("height"), e && (this.el.style[t ? "minHeight" : "height"] = e * i + s), t && h.shouldSizeToContent(t) && t.grid.resizeToContentCBCheck(t.el), this;
  }
  /** @internal */
  _prepareElement(t, e = !1, i) {
    i = i || this._readAttr(t), t.gridstackNode = i, i.el = t, i.grid = this, i = this.engine.addNode(i, e), this._writeAttr(t, i), t.classList.add(p.itemClass, this.opts.itemClass);
    const s = h.shouldSizeToContent(i);
    return s ? t.classList.add("size-to-content") : t.classList.remove("size-to-content"), s && this.resizeToContentCheck(!1, i), h.lazyLoad(i) || this.prepareDragDrop(i.el), this;
  }
  /** @internal write position CSS vars and x,y,w,h attributes (not used for CSS but by users) back to element */
  _writePosAttr(t, e) {
    return (!e._moving && !e._resizing || this._placeholder === t) && (t.style.top = e.y ? e.y === 1 ? "var(--gs-cell-height)" : `calc(${e.y} * var(--gs-cell-height))` : null, t.style.left = e.x ? e.x === 1 ? "var(--gs-column-width)" : `calc(${e.x} * var(--gs-column-width))` : null, t.style.width = e.w > 1 ? `calc(${e.w} * var(--gs-column-width))` : null, t.style.height = e.h > 1 ? `calc(${e.h} * var(--gs-cell-height))` : null), e.x > 0 ? t.setAttribute("gs-x", String(e.x)) : t.removeAttribute("gs-x"), e.y > 0 ? t.setAttribute("gs-y", String(e.y)) : t.removeAttribute("gs-y"), e.w > 1 ? t.setAttribute("gs-w", String(e.w)) : t.removeAttribute("gs-w"), e.h > 1 ? t.setAttribute("gs-h", String(e.h)) : t.removeAttribute("gs-h"), this;
  }
  /** @internal call to write any default attributes back to element */
  _writeAttr(t, e) {
    if (!e)
      return this;
    this._writePosAttr(t, e);
    const i = {
      // autoPosition: 'gs-auto-position', // no need to write out as already in node and doesn't affect CSS
      noResize: "gs-no-resize",
      noMove: "gs-no-move",
      locked: "gs-locked",
      id: "gs-id",
      sizeToContent: "gs-size-to-content"
    };
    for (const s in i)
      e[s] ? t.setAttribute(i[s], String(e[s])) : t.removeAttribute(i[s]);
    return this;
  }
  /** @internal call to read any default attributes from element */
  _readAttr(t, e = !0) {
    const i = {};
    i.x = h.toNumber(t.getAttribute("gs-x")), i.y = h.toNumber(t.getAttribute("gs-y")), i.w = h.toNumber(t.getAttribute("gs-w")), i.h = h.toNumber(t.getAttribute("gs-h")), i.autoPosition = h.toBool(t.getAttribute("gs-auto-position")), i.noResize = h.toBool(t.getAttribute("gs-no-resize")), i.noMove = h.toBool(t.getAttribute("gs-no-move")), i.locked = h.toBool(t.getAttribute("gs-locked"));
    const s = t.getAttribute("gs-size-to-content");
    s && (s === "true" || s === "false" ? i.sizeToContent = h.toBool(s) : i.sizeToContent = parseInt(s, 10)), i.id = t.getAttribute("gs-id"), i.maxW = h.toNumber(t.getAttribute("gs-max-w")), i.minW = h.toNumber(t.getAttribute("gs-min-w")), i.maxH = h.toNumber(t.getAttribute("gs-max-h")), i.minH = h.toNumber(t.getAttribute("gs-min-h")), e && (i.w === 1 && t.removeAttribute("gs-w"), i.h === 1 && t.removeAttribute("gs-h"), i.maxW && t.removeAttribute("gs-max-w"), i.minW && t.removeAttribute("gs-min-w"), i.maxH && t.removeAttribute("gs-max-h"), i.minH && t.removeAttribute("gs-min-h"));
    for (const n in i) {
      if (!i.hasOwnProperty(n))
        return;
      !i[n] && i[n] !== 0 && n !== "sizeToContent" && delete i[n];
    }
    return i;
  }
  /** @internal */
  _setStaticClass() {
    const t = ["grid-stack-static"];
    return this.opts.staticGrid ? (this.el.classList.add(...t), this.el.setAttribute("gs-static", "true")) : (this.el.classList.remove(...t), this.el.removeAttribute("gs-static")), this;
  }
  /**
   * called when we are being resized - check if the one Column Mode needs to be turned on/off
   * and remember the prev columns we used, or get our count from parent, as well as check for cellHeight==='auto' (square)
   * or `sizeToContent` gridItem options.
   */
  onResize(t = this.el?.clientWidth) {
    if (!t || this.prevWidth === t)
      return;
    this.prevWidth = t, this.batchUpdate();
    let e = !1;
    return this._autoColumn && this.parentGridNode ? this.opts.column !== this.parentGridNode.w && (this.column(this.parentGridNode.w, this.opts.layout || "list"), e = !0) : e = this.checkDynamicColumn(), this._isAutoCellHeight && this.cellHeight(), this.engine.nodes.forEach((i) => {
      i.subGrid && i.subGrid.onResize();
    }), this._skipInitialResize || this.resizeToContentCheck(e), delete this._skipInitialResize, this.batchUpdate(!1), this;
  }
  /** resizes content for given node (or all) if shouldSizeToContent() is true */
  resizeToContentCheck(t = !1, e = void 0) {
    if (this.engine) {
      if (t && this.hasAnimationCSS())
        return setTimeout(() => this.resizeToContentCheck(!1, e), this.animationDelay);
      if (e)
        h.shouldSizeToContent(e) && this.resizeToContentCBCheck(e.el);
      else if (this.engine.nodes.some((i) => h.shouldSizeToContent(i))) {
        const i = [...this.engine.nodes];
        this.batchUpdate(), i.forEach((s) => {
          h.shouldSizeToContent(s) && this.resizeToContentCBCheck(s.el);
        }), this._ignoreLayoutsNodeChange = !0, this.batchUpdate(!1), this._ignoreLayoutsNodeChange = !1;
      }
      this._gsEventHandler.resizecontent && this._gsEventHandler.resizecontent(null, e ? [e] : this.engine.nodes);
    }
  }
  /** add or remove the grid element size event handler */
  _updateResizeEvent(t = !1) {
    const e = !this.parentGridNode && (this._isAutoCellHeight || this.opts.sizeToContent || this.opts.columnOpts || this.engine.nodes.find((i) => i.sizeToContent));
    return !t && e && !this.resizeObserver ? (this._sizeThrottle = h.throttle(() => this.onResize(), this.opts.cellHeightThrottle), this.resizeObserver = new ResizeObserver(() => this._sizeThrottle()), this.resizeObserver.observe(this.el), this._skipInitialResize = !0) : (t || !e) && this.resizeObserver && (this.resizeObserver.disconnect(), delete this.resizeObserver, delete this._sizeThrottle), this;
  }
  /** @internal convert a potential selector into actual element */
  static getElement(t = ".grid-stack-item") {
    return h.getElement(t);
  }
  /** @internal */
  static getElements(t = ".grid-stack-item") {
    return h.getElements(t);
  }
  /** @internal */
  static getGridElement(t) {
    return d.getElement(t);
  }
  /** @internal */
  static getGridElements(t) {
    return h.getElements(t);
  }
  /** @internal initialize margin top/bottom/left/right and units */
  _initMargin() {
    let t, e = 0, i = [];
    typeof this.opts.margin == "string" && (i = this.opts.margin.split(" ")), i.length === 2 ? (this.opts.marginTop = this.opts.marginBottom = i[0], this.opts.marginLeft = this.opts.marginRight = i[1]) : i.length === 4 ? (this.opts.marginTop = i[0], this.opts.marginRight = i[1], this.opts.marginBottom = i[2], this.opts.marginLeft = i[3]) : (t = h.parseHeight(this.opts.margin), this.opts.marginUnit = t.unit, e = this.opts.margin = t.h), ["marginTop", "marginRight", "marginBottom", "marginLeft"].forEach((o) => {
      this.opts[o] === void 0 ? this.opts[o] = e : (t = h.parseHeight(this.opts[o]), this.opts[o] = t.h, delete this.opts.margin);
    }), this.opts.marginUnit = t.unit, this.opts.marginTop === this.opts.marginBottom && this.opts.marginLeft === this.opts.marginRight && this.opts.marginTop === this.opts.marginRight && (this.opts.margin = this.opts.marginTop);
    const n = this.el.style;
    return n.setProperty("--gs-item-margin-top", `${this.opts.marginTop}${this.opts.marginUnit}`), n.setProperty("--gs-item-margin-bottom", `${this.opts.marginBottom}${this.opts.marginUnit}`), n.setProperty("--gs-item-margin-right", `${this.opts.marginRight}${this.opts.marginUnit}`), n.setProperty("--gs-item-margin-left", `${this.opts.marginLeft}${this.opts.marginUnit}`), this;
  }
  /* ===========================================================================================
   * drag&drop methods that used to be stubbed out and implemented in dd-gridstack.ts
   * but caused loading issues in prod - see https://github.com/gridstack/gridstack.js/issues/2039
   * ===========================================================================================
   */
  /**
   * Get the global drag & drop implementation instance.
   * This provides access to the underlying drag & drop functionality.
   *
   * @returns the DDGridStack instance used for drag & drop operations
   *
   * @example
   * const dd = GridStack.getDD();
   * // Access drag & drop functionality
   */
  static getDD() {
    return f;
  }
  /**
   * call to setup dragging in from the outside (say toolbar), by specifying the class selection and options.
   * Called during GridStack.init() as options, but can also be called directly (last param are used) in case the toolbar
   * is dynamically create and needs to be set later.
   * @param dragIn string selector (ex: '.sidebar-item') or list of dom elements
   * @param dragInOptions options - see DDDragOpt. (default: {handle: '.grid-stack-item-content', appendTo: 'body'}
   * @param widgets GridStackWidget def to assign to each element which defines what to create on drop
   * @param root optional root which defaults to document (for shadow dom pass the parent HTMLDocument)
   */
  static setupDragIn(t, e, i, s = document) {
    e?.pause !== void 0 && (_.pauseDrag = e.pause), e = { appendTo: "body", helper: "clone", ...e || {} }, (typeof t == "string" ? h.getElements(t, s) : t).forEach((o, r) => {
      f.isDraggable(o) || f.dragIn(o, e), i?.[r] && (o.gridstackNode = i[r]);
    });
  }
  /**
   * Enables/Disables dragging by the user for specific grid elements.
   * For all items and future items, use enableMove() instead. No-op for static grids.
   *
   * Note: If you want to prevent an item from moving due to being pushed around by another
   * during collision, use the 'locked' property instead.
   *
   * @param els widget element(s) or selector to modify
   * @param val if true widget will be draggable, assuming the parent grid isn't noMove or static
   * @returns the grid instance for chaining
   *
   * @example
   * // Make specific widgets draggable
   * grid.movable('.my-widget', true);
   *
   * // Disable dragging for specific widgets
   * grid.movable('#fixed-widget', false);
   */
  movable(t, e) {
    return this.opts.staticGrid ? this : (d.getElements(t).forEach((i) => {
      const s = i.gridstackNode;
      s && (e ? delete s.noMove : s.noMove = !0, this.prepareDragDrop(s.el));
    }), this);
  }
  /**
   * Enables/Disables user resizing for specific grid elements.
   * For all items and future items, use enableResize() instead. No-op for static grids.
   *
   * @param els widget element(s) or selector to modify
   * @param val if true widget will be resizable, assuming the parent grid isn't noResize or static
   * @returns the grid instance for chaining
   *
   * @example
   * // Make specific widgets resizable
   * grid.resizable('.my-widget', true);
   *
   * // Disable resizing for specific widgets
   * grid.resizable('#fixed-size-widget', false);
   */
  resizable(t, e) {
    return this.opts.staticGrid ? this : (d.getElements(t).forEach((i) => {
      const s = i.gridstackNode;
      s && (e ? delete s.noResize : s.noResize = !0, this.prepareDragDrop(s.el));
    }), this);
  }
  /**
   * Temporarily disables widgets moving/resizing.
   * If you want a more permanent way (which freezes up resources) use `setStatic(true)` instead.
   *
   * Note: This is a no-op for static grids.
   *
   * This is a shortcut for:
   * ```typescript
   * grid.enableMove(false);
   * grid.enableResize(false);
   * ```
   *
   * @param recurse if true (default), sub-grids also get updated
   * @returns the grid instance for chaining
   *
   * @example
   * // Disable all interactions
   * grid.disable();
   *
   * // Disable only this grid, not sub-grids
   * grid.disable(false);
   */
  disable(t = !0) {
    if (!this.opts.staticGrid)
      return this.enableMove(!1, t), this.enableResize(!1, t), this._triggerEvent("disable"), this;
  }
  /**
   * Re-enables widgets moving/resizing - see disable().
   * Note: This is a no-op for static grids.
   *
   * This is a shortcut for:
   * ```typescript
   * grid.enableMove(true);
   * grid.enableResize(true);
   * ```
   *
   * @param recurse if true (default), sub-grids also get updated
   * @returns the grid instance for chaining
   *
   * @example
   * // Re-enable all interactions
   * grid.enable();
   *
   * // Enable only this grid, not sub-grids
   * grid.enable(false);
   */
  enable(t = !0) {
    if (!this.opts.staticGrid)
      return this.enableMove(!0, t), this.enableResize(!0, t), this._triggerEvent("enable"), this;
  }
  /**
   * Enables/disables widget moving for all widgets. No-op for static grids.
   * Note: locally defined items (with noMove property) still override this setting.
   *
   * @param doEnable if true widgets will be movable, if false moving is disabled
   * @param recurse if true (default), sub-grids also get updated
   * @returns the grid instance for chaining
   *
   * @example
   * // Enable moving for all widgets
   * grid.enableMove(true);
   *
   * // Disable moving for all widgets
   * grid.enableMove(false);
   *
   * // Enable only this grid, not sub-grids
   * grid.enableMove(true, false);
   */
  enableMove(t, e = !0) {
    return this.opts.staticGrid ? this : (t ? delete this.opts.disableDrag : this.opts.disableDrag = !0, this.engine.nodes.forEach((i) => {
      this.prepareDragDrop(i.el), i.subGrid && e && i.subGrid.enableMove(t, e);
    }), this);
  }
  /**
   * Enables/disables widget resizing for all widgets. No-op for static grids.
   * Note: locally defined items (with noResize property) still override this setting.
   *
   * @param doEnable if true widgets will be resizable, if false resizing is disabled
   * @param recurse if true (default), sub-grids also get updated
   * @returns the grid instance for chaining
   *
   * @example
   * // Enable resizing for all widgets
   * grid.enableResize(true);
   *
   * // Disable resizing for all widgets
   * grid.enableResize(false);
   *
   * // Enable only this grid, not sub-grids
   * grid.enableResize(true, false);
   */
  enableResize(t, e = !0) {
    return this.opts.staticGrid ? this : (t ? delete this.opts.disableResize : this.opts.disableResize = !0, this.engine.nodes.forEach((i) => {
      this.prepareDragDrop(i.el), i.subGrid && e && i.subGrid.enableResize(t, e);
    }), this);
  }
  /** @internal call when drag (and drop) needs to be cancelled (Esc key) */
  cancelDrag() {
    const t = this._placeholder?.gridstackNode;
    t && (t._isExternal ? (t._isAboutToRemove = !0, this.engine.removeNode(t)) : t._isAboutToRemove && d._itemRemoving(t.el, !1), this.engine.restoreInitial());
  }
  /** @internal removes any drag&drop present (called during destroy) */
  _removeDD(t) {
    return f.draggable(t, "destroy").resizable(t, "destroy"), t.gridstackNode && delete t.gridstackNode._initDD, delete t.ddElement, this;
  }
  /** @internal called to add drag over to support widgets being added externally */
  _setupAcceptWidget() {
    if (this.opts.staticGrid || !this.opts.acceptWidgets && !this.opts.removable)
      return f.droppable(this.el, "destroy"), this;
    let t, e;
    const i = (s, n, o) => {
      o = o || n;
      const r = o.gridstackNode;
      if (!r)
        return;
      if (!r.grid?.el) {
        o.style.transform = `scale(${1 / this.dragTransform.xScale},${1 / this.dragTransform.yScale})`;
        const c = o.getBoundingClientRect();
        o.style.left = c.x + (this.dragTransform.xScale - 1) * (s.clientX - c.x) / this.dragTransform.xScale + "px", o.style.top = c.y + (this.dragTransform.yScale - 1) * (s.clientY - c.y) / this.dragTransform.yScale + "px", o.style.transformOrigin = "0px 0px";
      }
      let { top: u, left: g } = o.getBoundingClientRect();
      const a = this.el.getBoundingClientRect();
      g -= a.left, u -= a.top;
      const l = {
        position: {
          top: u * this.dragTransform.xScale,
          left: g * this.dragTransform.yScale
        }
      };
      if (r._temporaryRemoved) {
        if (r.x = Math.max(0, Math.round(g / e)), r.y = Math.max(0, Math.round(u / t)), delete r.autoPosition, this.engine.nodeBoundFix(r), !this.engine.willItFit(r)) {
          if (r.autoPosition = !0, !this.engine.willItFit(r)) {
            f.off(n, "drag");
            return;
          }
          r._willFitPos && (h.copyPos(r, r._willFitPos), delete r._willFitPos);
        }
        this._onStartMoving(o, s, l, r, e, t);
      } else
        this._dragOrResize(o, s, l, r, e, t);
    };
    return f.droppable(this.el, {
      accept: (s) => {
        const n = s.gridstackNode || this._readAttr(s, !1);
        if (n?.grid === this)
          return !0;
        if (!this.opts.acceptWidgets)
          return !1;
        let o = !0;
        if (typeof this.opts.acceptWidgets == "function")
          o = this.opts.acceptWidgets(s);
        else {
          const r = this.opts.acceptWidgets === !0 ? ".grid-stack-item" : this.opts.acceptWidgets;
          o = s.matches(r);
        }
        if (o && n && this.opts.maxRow) {
          const r = { w: n.w, h: n.h, minW: n.minW, minH: n.minH };
          o = this.engine.willItFit(r);
        }
        return o;
      }
    }).on(this.el, "dropover", (s, n, o) => {
      let r = o?.gridstackNode || n.gridstackNode;
      if (r?.grid === this && !r._temporaryRemoved)
        return !1;
      if (r?._sidebarOrig && (r.w = r._sidebarOrig.w, r.h = r._sidebarOrig.h), r?.grid && r.grid !== this && !r._temporaryRemoved && r.grid._leave(n, o), o = o || n, e = this.cellWidth(), t = this.getCellHeight(!0), !r) {
        const a = o.getAttribute("data-gs-widget") || o.getAttribute("gridstacknode");
        if (a) {
          try {
            r = JSON.parse(a);
          } catch {
            console.error("Gridstack dropover: Bad JSON format: ", a);
          }
          o.removeAttribute("data-gs-widget"), o.removeAttribute("gridstacknode");
        }
        r || (r = this._readAttr(o)), r._sidebarOrig = { w: r.w, h: r.h };
      }
      r.grid || (r.el || (r = { ...r }), r._isExternal = !0, o.gridstackNode = r);
      const u = r.w || Math.round(o.offsetWidth / e) || 1, g = r.h || Math.round(o.offsetHeight / t) || 1;
      return r.grid && r.grid !== this ? (n._gridstackNodeOrig || (n._gridstackNodeOrig = r), n.gridstackNode = r = { ...r, w: u, h: g, grid: this }, delete r.x, delete r.y, this.engine.cleanupNode(r).nodeBoundFix(r), r._initDD = r._isExternal = // DOM needs to be re-parented on a drop
      r._temporaryRemoved = !0) : (r.w = u, r.h = g, r._temporaryRemoved = !0), d._itemRemoving(r.el, !1), f.on(n, "drag", i), i(s, n, o), !1;
    }).on(this.el, "dropout", (s, n, o) => {
      const r = o?.gridstackNode || n.gridstackNode;
      return r && (!r.grid || r.grid === this) && (this._leave(n, o), this._isTemp && this.removeAsSubGrid(r)), !1;
    }).on(this.el, "drop", (s, n, o) => {
      const r = o?.gridstackNode || n.gridstackNode;
      if (r?.grid === this && !r._isExternal)
        return !1;
      const u = !!this.placeholder.parentElement, g = n !== o;
      this.placeholder.remove(), delete this.placeholder.gridstackNode, u && this.opts.animate && (this.setAnimation(!1), this.setAnimation(!0, !0));
      const a = n._gridstackNodeOrig;
      if (delete n._gridstackNodeOrig, u && a?.grid && a.grid !== this) {
        const c = a.grid;
        c.engine.removeNodeFromLayoutCache(a), c.engine.removedNodes.push(a), c._triggerRemoveEvent()._triggerChangeEvent(), c.parentGridNode && !c.engine.nodes.length && c.opts.subGridDynamic && c.removeAsSubGrid();
      }
      if (!r || (u && (this.engine.cleanupNode(r), r.grid = this), delete r.grid?._isTemp, f.off(n, "drag"), o !== n ? (o.remove(), n = o) : n.remove(), this._removeDD(n), !u))
        return !1;
      const l = r.subGrid?.el?.gridstack;
      return h.copyPos(r, this._readAttr(this.placeholder)), h.removePositioningStyles(n), g && (r.content || r.subGridOpts || d.addRemoveCB) ? (delete r.el, n = this.addWidget(r)) : (this._prepareElement(n, !0, r), this.el.appendChild(n), this.resizeToContentCheck(!1, r), l && (l.parentGridNode = r), this._updateContainerHeight()), this.engine.addedNodes.push(r), this._triggerAddEvent(), this._triggerChangeEvent(), this.engine.endUpdate(), this._gsEventHandler.dropped && this._gsEventHandler.dropped({ ...s, type: "dropped" }, a && a.grid ? a : void 0, r), !1;
    }), this;
  }
  /** @internal mark item for removal */
  static _itemRemoving(t, e) {
    if (!t)
      return;
    const i = t ? t.gridstackNode : void 0;
    !i?.grid || t.classList.contains(i.grid.opts.removableOptions.decline) || (e ? i._isAboutToRemove = !0 : delete i._isAboutToRemove, e ? t.classList.add("grid-stack-item-removing") : t.classList.remove("grid-stack-item-removing"));
  }
  /** @internal called to setup a trash drop zone if the user specifies it */
  _setupRemoveDrop() {
    if (typeof this.opts.removable != "string")
      return this;
    const t = document.querySelector(this.opts.removable);
    return t ? (!this.opts.staticGrid && !f.isDroppable(t) && f.droppable(t, this.opts.removableOptions).on(t, "dropover", (e, i) => d._itemRemoving(i, !0)).on(t, "dropout", (e, i) => d._itemRemoving(i, !1)), this) : this;
  }
  /**
   * prepares the element for drag&drop - this is normally called by makeWidget() unless are are delay loading
   * @param el GridItemHTMLElement of the widget
   * @param [force=false]
   * */
  prepareDragDrop(t, e = !1) {
    const i = t?.gridstackNode;
    if (!i)
      return;
    const s = i.noMove || this.opts.disableDrag, n = i.noResize || this.opts.disableResize, o = this.opts.staticGrid || s && n;
    if ((e || o) && (i._initDD && (this._removeDD(t), delete i._initDD), o && t.classList.add("ui-draggable-disabled", "ui-resizable-disabled"), !e))
      return this;
    if (!i._initDD) {
      let r, u;
      const g = (c, b) => {
        this.triggerEvent(c, c.target), r = this.cellWidth(), u = this.getCellHeight(!0), this._onStartMoving(t, c, b, i, r, u);
      }, a = (c, b) => {
        this._dragOrResize(t, c, b, i, r, u);
      }, l = (c) => {
        this.placeholder.remove(), delete this.placeholder.gridstackNode, delete i._moving, delete i._resizing, delete i._event, delete i._lastTried;
        const b = i.w !== i._orig.w, m = c.target;
        if (!(!m.gridstackNode || m.gridstackNode.grid !== this)) {
          if (i.el = m, i._isAboutToRemove) {
            const v = t.gridstackNode.grid;
            v._gsEventHandler[c.type] && v._gsEventHandler[c.type](c, m), v.engine.nodes.push(i), v.removeWidget(t, !0, !0);
          } else
            h.removePositioningStyles(m), i._temporaryRemoved ? (this._writePosAttr(m, i), this.engine.addNode(i)) : this._writePosAttr(m, i), this.triggerEvent(c, m);
          this._extraDragRow = 0, this._updateContainerHeight(), this._triggerChangeEvent(), this.engine.endUpdate(), c.type === "resizestop" && (Number.isInteger(i.sizeToContent) && (i.sizeToContent = i.h), this.resizeToContentCheck(b, i));
        }
      };
      f.draggable(t, {
        start: g,
        stop: l,
        drag: a
      }).resizable(t, {
        start: g,
        stop: l,
        resize: a
      }), i._initDD = !0;
    }
    return f.draggable(t, s ? "disable" : "enable").resizable(t, n ? "disable" : "enable"), this;
  }
  /** @internal handles actual drag/resize start */
  _onStartMoving(t, e, i, s, n, o) {
    if (this.engine.cleanNodes().beginUpdate(s), this._writePosAttr(this.placeholder, s), this.el.appendChild(this.placeholder), this.placeholder.gridstackNode = s, s.grid?.el)
      this.dragTransform = h.getValuesFromTransformedElement(t);
    else if (this.placeholder && this.placeholder.closest(".grid-stack")) {
      const r = this.placeholder.closest(".grid-stack");
      this.dragTransform = h.getValuesFromTransformedElement(r);
    } else
      this.dragTransform = {
        xScale: 1,
        xOffset: 0,
        yScale: 1,
        yOffset: 0
      };
    if (s.el = this.placeholder, s._lastUiPosition = i.position, s._prevYPix = i.position.top, s._moving = e.type === "dragstart", s._resizing = e.type === "resizestart", delete s._lastTried, e.type === "dropover" && s._temporaryRemoved && (this.engine.addNode(s), s._moving = !0), this.engine.cacheRects(n, o, this.opts.marginTop, this.opts.marginRight, this.opts.marginBottom, this.opts.marginLeft), e.type === "resizestart") {
      const r = this.getColumn() - s.x, u = (this.opts.maxRow || Number.MAX_SAFE_INTEGER) - s.y;
      f.resizable(t, "option", "minWidth", n * Math.min(s.minW || 1, r)).resizable(t, "option", "minHeight", o * Math.min(s.minH || 1, u)).resizable(t, "option", "maxWidth", n * Math.min(s.maxW || Number.MAX_SAFE_INTEGER, r)).resizable(t, "option", "maxWidthMoveLeft", n * Math.min(s.maxW || Number.MAX_SAFE_INTEGER, s.x + s.w)).resizable(t, "option", "maxHeight", o * Math.min(s.maxH || Number.MAX_SAFE_INTEGER, u)).resizable(t, "option", "maxHeightMoveUp", o * Math.min(s.maxH || Number.MAX_SAFE_INTEGER, s.y + s.h));
    }
  }
  /** @internal handles actual drag/resize */
  _dragOrResize(t, e, i, s, n, o) {
    const r = { ...s._orig };
    let u, g = this.opts.marginLeft, a = this.opts.marginRight, l = this.opts.marginTop, c = this.opts.marginBottom;
    const b = Math.round(o * 0.1), m = Math.round(n * 0.1);
    if (g = Math.min(g, m), a = Math.min(a, m), l = Math.min(l, b), c = Math.min(c, b), e.type === "drag") {
      if (s._temporaryRemoved)
        return;
      const C = i.position.top - s._prevYPix;
      s._prevYPix = i.position.top, this.opts.draggable.scroll !== !1 && h.updateScrollPosition(t, i.position, C);
      const y = i.position.left + (i.position.left > s._lastUiPosition.left ? -a : g), z = i.position.top + (i.position.top > s._lastUiPosition.top ? -c : l);
      r.x = Math.round(y / n), r.y = Math.round(z / o);
      const x = this._extraDragRow;
      if (this.engine.collide(s, r)) {
        const w = this.getRow();
        let R = Math.max(0, r.y + s.h - w);
        this.opts.maxRow && w + R > this.opts.maxRow && (R = Math.max(0, this.opts.maxRow - w)), this._extraDragRow = R;
      } else
        this._extraDragRow = 0;
      if (this._extraDragRow !== x && this._updateContainerHeight(), s.x === r.x && s.y === r.y)
        return;
    } else if (e.type === "resize") {
      if (r.x < 0 || (h.updateScrollResize(e, t, o), r.w = Math.round((i.size.width - g) / n), r.h = Math.round((i.size.height - l) / o), s.w === r.w && s.h === r.h) || s._lastTried && s._lastTried.w === r.w && s._lastTried.h === r.h)
        return;
      const C = i.position.left + g, y = i.position.top + l;
      r.x = Math.round(C / n), r.y = Math.round(y / o), u = !0;
    }
    s._event = e, s._lastTried = r;
    const v = {
      x: i.position.left + g,
      y: i.position.top + l,
      w: (i.size ? i.size.width : s.w * n) - g - a,
      h: (i.size ? i.size.height : s.h * o) - l - c
    };
    if (this.engine.moveNodeCheck(s, { ...r, cellWidth: n, cellHeight: o, rect: v, resizing: u })) {
      s._lastUiPosition = i.position, this.engine.cacheRects(n, o, l, a, c, g), delete s._skipDown, u && s.subGrid && s.subGrid.onResize(), this._extraDragRow = 0, this._updateContainerHeight();
      const C = e.target;
      s._sidebarOrig || this._writePosAttr(C, s), this.triggerEvent(e, C);
    }
  }
  /** call given event callback on our main top-most grid (if we're nested) */
  triggerEvent(t, e) {
    let i = this;
    for (; i.parentGridNode; )
      i = i.parentGridNode.grid;
    i._gsEventHandler[t.type] && i._gsEventHandler[t.type](t, e);
  }
  /** @internal called when item leaving our area by either cursor dropout event
   * or shape is outside our boundaries. remove it from us, and mark temporary if this was
   * our item to start with else restore prev node values from prev grid it came from.
   */
  _leave(t, e) {
    e = e || t;
    const i = e.gridstackNode;
    if (!i || (e.style.transform = e.style.transformOrigin = null, f.off(t, "drag"), i._temporaryRemoved))
      return;
    i._temporaryRemoved = !0, this.engine.removeNode(i), i.el = i._isExternal && e ? e : t;
    const s = i._sidebarOrig;
    i._isExternal && this.engine.cleanupNode(i), i._sidebarOrig = s, this.opts.removable === !0 && d._itemRemoving(t, !0), t._gridstackNodeOrig ? (t.gridstackNode = t._gridstackNodeOrig, delete t._gridstackNodeOrig) : i._isExternal && this.engine.restoreInitial();
  }
  // legacy method removed
  commit() {
    return H(this, this.batchUpdate(!1), "commit", "batchUpdate", "5.2"), this;
  }
}
d.renderCB = (E, t) => {
  E && t?.content && (E.textContent = t.content);
};
d.resizeToContentParent = ".grid-stack-item-content";
d.Utils = h;
d.Engine = N;
d.GDRev = "12.3.2";
export {
  P as DDDraggable,
  A as DDGridStack,
  _ as DDManager,
  U as DDResizable,
  F as DDResizableHandle,
  d as GridStack,
  N as GridStackEngine,
  h as Utils,
  p as gridDefaults,
  H as obsolete
};
