// gridstack@12.3.3/dist/gridstack-engine.js
import { Utils as u } from "./utils.js";
class y {
  constructor(t = {}) {
    this.addedNodes = [], this.removedNodes = [], this.defaultColumn = 12, this.column = t.column || this.defaultColumn, this.column > this.defaultColumn && (this.defaultColumn = this.column), this.maxRow = t.maxRow, this._float = t.float, this.nodes = t.nodes || [], this.onChange = t.onChange;
  }
  /**
   * Enable/disable batch mode for multiple operations to optimize performance.
   * When enabled, layout updates are deferred until batch mode is disabled.
   *
   * @param flag true to enable batch mode, false to disable and apply changes
   * @param doPack if true (default), pack/compact nodes when disabling batch mode
   * @returns the engine instance for chaining
   *
   * @example
   * // Start batch mode for multiple operations
   * engine.batchUpdate(true);
   * engine.addNode(node1);
   * engine.addNode(node2);
   * engine.batchUpdate(false); // Apply all changes at once
   */
  batchUpdate(t = !0, i = !0) {
    return !!this.batchMode === t ? this : (this.batchMode = t, t ? (this._prevFloat = this._float, this._float = !0, this.cleanNodes(), this.saveInitial()) : (this._float = this._prevFloat, delete this._prevFloat, i && this._packNodes(), this._notify()), this);
  }
  // use entire row for hitting area (will use bottom reverse sorted first) if we not actively moving DOWN and didn't already skip
  _useEntireRowArea(t, i) {
    return (!this.float || this.batchMode && !this._prevFloat) && !this._hasLocked && (!t._moving || t._skipDown || i.y <= t.y);
  }
  /** @internal fix collision on given 'node', going to given new location 'nn', with optional 'collide' node already found.
   * return true if we moved. */
  _fixCollisions(t, i = t, s, h = {}) {
    if (this.sortNodes(-1), s = s || this.collide(t, i), !s)
      return !1;
    if (t._moving && !h.nested && !this.float && this.swap(t, s))
      return !0;
    let e = i;
    !this._loading && this._useEntireRowArea(t, i) && (e = { x: 0, w: this.column, y: i.y, h: i.h }, s = this.collide(t, e, h.skip));
    let r = !1;
    const f = { nested: !0, pack: !1 };
    let l = 0;
    for (; s = s || this.collide(t, e, h.skip); ) {
      if (l++ > this.nodes.length * 2)
        throw new Error("Infinite collide check");
      let o;
      if (s.locked || this._loading || t._moving && !t._skipDown && i.y > t.y && !this.float && // can take space we had, or before where we're going
      (!this.collide(s, { ...s, y: t.y }, t) || !this.collide(s, { ...s, y: i.y - s.h }, t))) {
        t._skipDown = t._skipDown || i.y > t.y;
        const a = { ...i, y: s.y + s.h, ...f };
        o = this._loading && u.samePos(t, a) ? !0 : this.moveNode(t, a), (s.locked || this._loading) && o ? u.copyPos(i, t) : !s.locked && o && h.pack && (this._packNodes(), i.y = s.y + s.h, u.copyPos(t, i)), r = r || o;
      } else
        o = this.moveNode(s, { ...s, y: i.y + i.h, skip: t, ...f });
      if (!o)
        return r;
      s = void 0;
    }
    return r;
  }
  /**
   * Return the first node that intercepts/collides with the given node or area.
   * Used for collision detection during drag and drop operations.
   *
   * @param skip the node to skip in collision detection (usually the node being moved)
   * @param area the area to check for collisions (defaults to skip node's area)
   * @param skip2 optional second node to skip in collision detection
   * @returns the first colliding node, or undefined if no collision
   *
   * @example
   * const colliding = engine.collide(draggedNode, {x: 2, y: 1, w: 2, h: 1});
   * if (colliding) {
   *   console.log('Would collide with:', colliding.id);
   * }
   */
  collide(t, i = t, s) {
    const h = t._id, e = s?._id;
    return this.nodes.find((r) => r._id !== h && r._id !== e && u.isIntercepted(r, i));
  }
  /**
   * Return all nodes that intercept/collide with the given node or area.
   * Similar to collide() but returns all colliding nodes instead of just the first.
   *
   * @param skip the node to skip in collision detection
   * @param area the area to check for collisions (defaults to skip node's area)
   * @param skip2 optional second node to skip in collision detection
   * @returns array of all colliding nodes
   *
   * @example
   * const allCollisions = engine.collideAll(draggedNode);
   * console.log('Colliding with', allCollisions.length, 'nodes');
   */
  collideAll(t, i = t, s) {
    const h = t._id, e = s?._id;
    return this.nodes.filter((r) => r._id !== h && r._id !== e && u.isIntercepted(r, i));
  }
  /** does a pixel coverage collision based on where we started, returning the node that has the most coverage that is >50% mid line */
  directionCollideCoverage(t, i, s) {
    if (!i.rect || !t._rect)
      return;
    const h = t._rect, e = { ...i.rect };
    e.y > h.y ? (e.h += e.y - h.y, e.y = h.y) : e.h += h.y - e.y, e.x > h.x ? (e.w += e.x - h.x, e.x = h.x) : e.w += h.x - e.x;
    let r, f = 0.5;
    for (let l of s) {
      if (l.locked || !l._rect)
        break;
      const o = l._rect;
      let a = Number.MAX_VALUE, c = Number.MAX_VALUE;
      h.y < o.y ? a = (e.y + e.h - o.y) / o.h : h.y + h.h > o.y + o.h && (a = (o.y + o.h - e.y) / o.h), h.x < o.x ? c = (e.x + e.w - o.x) / o.w : h.x + h.w > o.x + o.w && (c = (o.x + o.w - e.x) / o.w);
      const n = Math.min(c, a);
      n > f && (f = n, r = l);
    }
    return i.collide = r, r;
  }
  /** does a pixel coverage returning the node that has the most coverage by area */
  /*
  protected collideCoverage(r: GridStackPosition, collides: GridStackNode[]): {collide: GridStackNode, over: number} {
    const collide: GridStackNode;
    const overMax = 0;
    collides.forEach(n => {
      if (n.locked || !n._rect) return;
      const over = Utils.areaIntercept(r, n._rect);
      if (over > overMax) {
        overMax = over;
        collide = n;
      }
    });
    return {collide, over: overMax};
  }
  */
  /**
   * Cache the pixel rectangles for all nodes used for collision detection during drag operations.
   * This optimization converts grid coordinates to pixel coordinates for faster collision detection.
   *
   * @param w width of a single grid cell in pixels
   * @param h height of a single grid cell in pixels
   * @param top top margin/padding in pixels
   * @param right right margin/padding in pixels
   * @param bottom bottom margin/padding in pixels
   * @param left left margin/padding in pixels
   * @returns the engine instance for chaining
   *
   * @internal This is typically called by GridStack during resize events
   */
  cacheRects(t, i, s, h, e, r) {
    return this.nodes.forEach((f) => f._rect = {
      y: f.y * i + s,
      x: f.x * t + r,
      w: f.w * t - r - h,
      h: f.h * i - s - e
    }), this;
  }
  /**
   * Attempt to swap the positions of two nodes if they meet swapping criteria.
   * Nodes can swap if they are the same size or in the same column/row, not locked, and touching.
   *
   * @param a first node to swap
   * @param b second node to swap
   * @returns true if swap was successful, false if not possible, undefined if not applicable
   *
   * @example
   * const swapped = engine.swap(nodeA, nodeB);
   * if (swapped) {
   *   console.log('Nodes swapped successfully');
   * }
   */
  swap(t, i) {
    if (!i || i.locked || !t || t.locked)
      return !1;
    function s() {
      const e = i.x, r = i.y;
      return i.x = t.x, i.y = t.y, t.h != i.h ? (t.x = e, t.y = i.y + i.h) : t.w != i.w ? (t.x = i.x + i.w, t.y = r) : (t.x = e, t.y = r), t._dirty = i._dirty = !0, !0;
    }
    let h;
    if (t.w === i.w && t.h === i.h && (t.x === i.x || t.y === i.y) && (h = u.isTouching(t, i)))
      return s();
    if (h !== !1) {
      if (t.w === i.w && t.x === i.x && (h || (h = u.isTouching(t, i)))) {
        if (i.y < t.y) {
          const e = t;
          t = i, i = e;
        }
        return s();
      }
      if (h !== !1) {
        if (t.h === i.h && t.y === i.y && (h || (h = u.isTouching(t, i)))) {
          if (i.x < t.x) {
            const e = t;
            t = i, i = e;
          }
          return s();
        }
        return !1;
      }
    }
  }
  /**
   * Check if the specified rectangular area is empty (no nodes occupy any part of it).
   *
   * @param x the x coordinate (column) of the area to check
   * @param y the y coordinate (row) of the area to check
   * @param w the width in columns of the area to check
   * @param h the height in rows of the area to check
   * @returns true if the area is completely empty, false if any node overlaps
   *
   * @example
   * if (engine.isAreaEmpty(2, 1, 3, 2)) {
   *   console.log('Area is available for placement');
   * }
   */
  isAreaEmpty(t, i, s, h) {
    const e = { x: t || 0, y: i || 0, w: s || 1, h: h || 1 };
    return !this.collide(e);
  }
  /**
   * Re-layout grid items to reclaim any empty space.
   * This optimizes the grid layout by moving items to fill gaps.
   *
   * @param layout layout algorithm to use:
   *   - 'compact' (default): find truly empty spaces, may reorder items
   *   - 'list': keep the sort order exactly the same, move items up sequentially
   * @param doSort if true (default), sort nodes by position before compacting
   * @returns the engine instance for chaining
   *
   * @example
   * // Compact to fill empty spaces
   * engine.compact();
   *
   * // Compact preserving item order
   * engine.compact('list');
   */
  compact(t = "compact", i = !0) {
    if (this.nodes.length === 0)
      return this;
    i && this.sortNodes();
    const s = this.batchMode;
    s || this.batchUpdate();
    const h = this._inColumnResize;
    h || (this._inColumnResize = !0);
    const e = this.nodes;
    return this.nodes = [], e.forEach((r, f, l) => {
      let o;
      r.locked || (r.autoPosition = !0, t === "list" && f && (o = l[f - 1])), this.addNode(r, !1, o);
    }), h || delete this._inColumnResize, s || this.batchUpdate(!1), this;
  }
  /**
   * Enable/disable floating widgets (default: `false`).
   * When floating is enabled, widgets can move up to fill empty spaces.
   * See [example](http://gridstackjs.com/demo/float.html)
   *
   * @param val true to enable floating, false to disable
   *
   * @example
   * engine.float = true;  // Enable floating
   * engine.float = false; // Disable floating (default)
   */
  set float(t) {
    this._float !== t && (this._float = t || !1, t || this._packNodes()._notify());
  }
  /**
   * Get the current floating mode setting.
   *
   * @returns true if floating is enabled, false otherwise
   *
   * @example
   * const isFloating = engine.float;
   * console.log('Floating enabled:', isFloating);
   */
  get float() {
    return this._float || !1;
  }
  /**
   * Sort the nodes array from first to last, or reverse.
   * This is called during collision/placement operations to enforce a specific order.
   *
   * @param dir sort direction: 1 for ascending (first to last), -1 for descending (last to first)
   * @returns the engine instance for chaining
   *
   * @example
   * engine.sortNodes();    // Sort ascending (default)
   * engine.sortNodes(-1);  // Sort descending
   */
  sortNodes(t = 1) {
    return this.nodes = u.sort(this.nodes, t), this;
  }
  /** @internal called to top gravity pack the items back OR revert back to original Y positions when floating */
  _packNodes() {
    return this.batchMode ? this : (this.sortNodes(), this.float ? this.nodes.forEach((t) => {
      if (t._updating || t._orig === void 0 || t.y === t._orig.y)
        return;
      let i = t.y;
      for (; i > t._orig.y; )
        --i, this.collide(t, { x: t.x, y: i, w: t.w, h: t.h }) || (t._dirty = !0, t.y = i);
    }) : this.nodes.forEach((t, i) => {
      if (!t.locked)
        for (; t.y > 0; ) {
          const s = i === 0 ? 0 : t.y - 1;
          if (!(i === 0 || !this.collide(t, { x: t.x, y: s, w: t.w, h: t.h })))
            break;
          t._dirty = t.y !== s, t.y = s;
        }
    }), this);
  }
  /**
   * Prepare and validate a node's coordinates and values for the current grid.
   * This ensures the node has valid position, size, and properties before being added to the grid.
   *
   * @param node the node to prepare and validate
   * @param resizing if true, resize the node down if it's out of bounds; if false, move it to fit
   * @returns the prepared node with valid coordinates
   *
   * @example
   * const node = { w: 3, h: 2, content: 'Hello' };
   * const prepared = engine.prepareNode(node);
   * console.log('Node prepared at:', prepared.x, prepared.y);
   */
  prepareNode(t, i) {
    t._id = t._id ?? y._idSeq++;
    const s = t.id;
    if (s) {
      let e = 1;
      for (; this.nodes.find((r) => r.id === t.id && r !== t); )
        t.id = s + "_" + e++;
    }
    (t.x === void 0 || t.y === void 0 || t.x === null || t.y === null) && (t.autoPosition = !0);
    const h = { x: 0, y: 0, w: 1, h: 1 };
    return u.defaults(t, h), t.autoPosition || delete t.autoPosition, t.noResize || delete t.noResize, t.noMove || delete t.noMove, u.sanitizeMinMax(t), typeof t.x == "string" && (t.x = Number(t.x)), typeof t.y == "string" && (t.y = Number(t.y)), typeof t.w == "string" && (t.w = Number(t.w)), typeof t.h == "string" && (t.h = Number(t.h)), isNaN(t.x) && (t.x = h.x, t.autoPosition = !0), isNaN(t.y) && (t.y = h.y, t.autoPosition = !0), isNaN(t.w) && (t.w = h.w), isNaN(t.h) && (t.h = h.h), this.nodeBoundFix(t, i), t;
  }
  /**
   * Part 2 of preparing a node to fit inside the grid - validates and fixes coordinates and dimensions.
   * This ensures the node fits within grid boundaries and respects min/max constraints.
   *
   * @param node the node to validate and fix
   * @param resizing if true, resize the node to fit; if false, move the node to fit
   * @returns the engine instance for chaining
   *
   * @example
   * // Fix a node that might be out of bounds
   * engine.nodeBoundFix(node, true); // Resize to fit
   * engine.nodeBoundFix(node, false); // Move to fit
   */
  nodeBoundFix(t, i) {
    const s = t._orig || u.copyPos({}, t);
    if (t.maxW && (t.w = Math.min(t.w || 1, t.maxW)), t.maxH && (t.h = Math.min(t.h || 1, t.maxH)), t.minW && (t.w = Math.max(t.w || 1, t.minW)), t.minH && (t.h = Math.max(t.h || 1, t.minH)), (t.x || 0) + (t.w || 1) > this.column && this.column < this.defaultColumn && !this._inColumnResize && !this.skipCacheUpdate && t._id != null && this.findCacheLayout(t, this.defaultColumn) === -1) {
      const e = { ...t };
      e.autoPosition || e.x === void 0 ? (delete e.x, delete e.y) : e.x = Math.min(this.defaultColumn - 1, e.x), e.w = Math.min(this.defaultColumn, e.w || 1), this.cacheOneLayout(e, this.defaultColumn);
    }
    return t.w > this.column ? t.w = this.column : t.w < 1 && (t.w = 1), this.maxRow && t.h > this.maxRow ? t.h = this.maxRow : t.h < 1 && (t.h = 1), t.x < 0 && (t.x = 0), t.y < 0 && (t.y = 0), t.x + t.w > this.column && (i ? t.w = this.column - t.x : t.x = this.column - t.w), this.maxRow && t.y + t.h > this.maxRow && (i ? t.h = this.maxRow - t.y : t.y = this.maxRow - t.h), u.samePos(t, s) || (t._dirty = !0), this;
  }
  /**
   * Returns a list of nodes that have been modified from their original values.
   * This is used to track which nodes need DOM updates.
   *
   * @param verify if true, performs additional verification by comparing current vs original positions
   * @returns array of nodes that have been modified
   *
   * @example
   * const changed = engine.getDirtyNodes();
   * console.log('Modified nodes:', changed.length);
   *
   * // Get verified dirty nodes
   * const verified = engine.getDirtyNodes(true);
   */
  getDirtyNodes(t) {
    return t ? this.nodes.filter((i) => i._dirty && !u.samePos(i, i._orig)) : this.nodes.filter((i) => i._dirty);
  }
  /** @internal call this to call onChange callback with dirty nodes so DOM can be updated */
  _notify(t) {
    if (this.batchMode || !this.onChange)
      return this;
    const i = (t || []).concat(this.getDirtyNodes());
    return this.onChange(i), this;
  }
  /**
   * Clean all dirty and last tried information from nodes.
   * This resets the dirty state tracking for all nodes.
   *
   * @returns the engine instance for chaining
   *
   * @internal
   */
  cleanNodes() {
    return this.batchMode ? this : (this.nodes.forEach((t) => {
      delete t._dirty, delete t._lastTried;
    }), this);
  }
  /**
   * Save the initial position/size of all nodes to track real dirty state.
   * This creates a snapshot of current positions that can be restored later.
   *
   * Note: Should be called right after change events and before move/resize operations.
   *
   * @returns the engine instance for chaining
   *
   * @internal
   */
  saveInitial() {
    return this.nodes.forEach((t) => {
      t._orig = u.copyPos({}, t), delete t._dirty;
    }), this._hasLocked = this.nodes.some((t) => t.locked), this;
  }
  /**
   * Restore all nodes back to their initial values.
   * This is typically called when canceling an operation (e.g., Esc key during drag).
   *
   * @returns the engine instance for chaining
   *
   * @internal
   */
  restoreInitial() {
    return this.nodes.forEach((t) => {
      !t._orig || u.samePos(t, t._orig) || (u.copyPos(t, t._orig), t._dirty = !0);
    }), this._notify(), this;
  }
  /**
   * Find the first available empty spot for the given node dimensions.
   * Updates the node's x,y attributes with the found position.
   *
   * @param node the node to find a position for (w,h must be set)
   * @param nodeList optional list of nodes to check against (defaults to engine nodes)
   * @param column optional column count (defaults to engine column count)
   * @param after optional node to start search after (maintains order)
   * @returns true if an empty position was found and node was updated
   *
   * @example
   * const node = { w: 2, h: 1 };
   * if (engine.findEmptyPosition(node)) {
   *   console.log('Found position at:', node.x, node.y);
   * }
   */
  findEmptyPosition(t, i = this.nodes, s = this.column, h) {
    const e = h ? h.y * s + (h.x + h.w) : 0;
    let r = !1;
    for (let f = e; !r; ++f) {
      const l = f % s, o = Math.floor(f / s);
      if (l + t.w > s)
        continue;
      const a = { x: l, y: o, w: t.w, h: t.h };
      i.find((c) => u.isIntercepted(a, c)) || ((t.x !== l || t.y !== o) && (t._dirty = !0), t.x = l, t.y = o, delete t.autoPosition, r = !0);
    }
    return r;
  }
  /**
   * Add the given node to the grid, handling collision detection and re-packing.
   * This is the main method for adding new widgets to the engine.
   *
   * @param node the node to add to the grid
   * @param triggerAddEvent if true, adds node to addedNodes list for event triggering
   * @param after optional node to place this node after (for ordering)
   * @returns the added node (or existing node if duplicate)
   *
   * @example
   * const node = { x: 0, y: 0, w: 2, h: 1, content: 'Hello' };
   * const added = engine.addNode(node, true);
   */
  addNode(t, i = !1, s) {
    const h = this.nodes.find((r) => r._id === t._id);
    if (h)
      return h;
    this._inColumnResize ? this.nodeBoundFix(t) : this.prepareNode(t), delete t._temporaryRemoved, delete t._removeDOM;
    let e;
    return t.autoPosition && this.findEmptyPosition(t, this.nodes, this.column, s) && (delete t.autoPosition, e = !0), this.nodes.push(t), i && this.addedNodes.push(t), e || this._fixCollisions(t), this.batchMode || this._packNodes()._notify(), t;
  }
  /**
   * Remove the given node from the grid.
   *
   * @param node the node to remove
   * @param removeDOM if true (default), marks node for DOM removal
   * @param triggerEvent if true, adds node to removedNodes list for event triggering
   * @returns the engine instance for chaining
   *
   * @example
   * engine.removeNode(node, true, true);
   */
  removeNode(t, i = !0, s = !1) {
    return this.nodes.find((h) => h._id === t._id) ? (s && this.removedNodes.push(t), i && (t._removeDOM = !0), this.nodes = this.nodes.filter((h) => h._id !== t._id), t._isAboutToRemove || this._packNodes(), this._notify([t]), this) : this;
  }
  /**
   * Remove all nodes from the grid.
   *
   * @param removeDOM if true (default), marks all nodes for DOM removal
   * @param triggerEvent if true (default), triggers removal events
   * @returns the engine instance for chaining
   *
   * @example
   * engine.removeAll(); // Remove all nodes
   */
  removeAll(t = !0, i = !0) {
    if (delete this._layouts, !this.nodes.length)
      return this;
    t && this.nodes.forEach((h) => h._removeDOM = !0);
    const s = this.nodes;
    return this.removedNodes = i ? s : [], this.nodes = [], this._notify(s);
  }
  /**
   * Check if a node can be moved to a new position, considering layout constraints.
   * This is a safer version of moveNode() that validates the move first.
   *
   * For complex cases (like maxRow constraints), it simulates the move in a clone first,
   * then applies the changes only if they meet all specifications.
   *
   * @param node the node to move
   * @param o move options including target position
   * @returns true if the node was successfully moved
   *
   * @example
   * const canMove = engine.moveNodeCheck(node, { x: 2, y: 1 });
   * if (canMove) {
   *   console.log('Node moved successfully');
   * }
   */
  moveNodeCheck(t, i) {
    if (!this.changedPosConstrain(t, i))
      return !1;
    if (i.pack = !0, !this.maxRow)
      return this.moveNode(t, i);
    let s;
    const h = new y({
      column: this.column,
      float: this.float,
      nodes: this.nodes.map((r) => r._id === t._id ? (s = { ...r }, s) : { ...r })
    });
    if (!s)
      return !1;
    const e = h.moveNode(s, i) && h.getRow() <= Math.max(this.getRow(), this.maxRow);
    if (!e && !i.resizing && i.collide) {
      const r = i.collide.el.gridstackNode;
      if (this.swap(t, r))
        return this._notify(), !0;
    }
    return e ? (h.nodes.filter((r) => r._dirty).forEach((r) => {
      const f = this.nodes.find((l) => l._id === r._id);
      f && (u.copyPos(f, r), f._dirty = !0);
    }), this._notify(), !0) : !1;
  }
  /** return true if can fit in grid height constrain only (always true if no maxRow) */
  willItFit(t) {
    if (delete t._willFitPos, !this.maxRow)
      return !0;
    const i = new y({
      column: this.column,
      float: this.float,
      nodes: this.nodes.map((h) => ({ ...h }))
    }), s = { ...t };
    return this.cleanupNode(s), delete s.el, delete s._id, delete s.content, delete s.grid, i.addNode(s), i.getRow() <= this.maxRow ? (t._willFitPos = u.copyPos({}, s), !0) : !1;
  }
  /** true if x,y or w,h are different after clamping to min/max */
  changedPosConstrain(t, i) {
    return i.w = i.w || t.w, i.h = i.h || t.h, t.x !== i.x || t.y !== i.y ? !0 : (t.maxW && (i.w = Math.min(i.w, t.maxW)), t.maxH && (i.h = Math.min(i.h, t.maxH)), t.minW && (i.w = Math.max(i.w, t.minW)), t.minH && (i.h = Math.max(i.h, t.minH)), t.w !== i.w || t.h !== i.h);
  }
  /** return true if the passed in node was actually moved (checks for no-op and locked) */
  moveNode(t, i) {
    if (!t || /*node.locked ||*/
    !i)
      return !1;
    let s;
    i.pack === void 0 && !this.batchMode && (s = i.pack = !0), typeof i.x != "number" && (i.x = t.x), typeof i.y != "number" && (i.y = t.y), typeof i.w != "number" && (i.w = t.w), typeof i.h != "number" && (i.h = t.h);
    const h = t.w !== i.w || t.h !== i.h, e = u.copyPos({}, t, !0);
    if (u.copyPos(e, i), this.nodeBoundFix(e, h), u.copyPos(i, e), !i.forceCollide && u.samePos(t, i))
      return !1;
    const r = u.copyPos({}, t), f = this.collideAll(t, e, i.skip);
    let l = !0;
    if (f.length) {
      const o = t._moving && !i.nested;
      let a = o ? this.directionCollideCoverage(t, i, f) : f[0];
      if (o && a && t.grid?.opts?.subGridDynamic && !t.grid._isTemp) {
        const c = u.areaIntercept(i.rect, a._rect), n = u.area(i.rect), _ = u.area(a._rect);
        c / (n < _ ? n : _) > 0.8 && (a.grid.makeSubGrid(a.el, void 0, t), a = void 0);
      }
      a ? l = !this._fixCollisions(t, e, a, i) : (l = !1, s && delete i.pack);
    }
    return l && !u.samePos(t, e) && (t._dirty = !0, u.copyPos(t, e)), i.pack && this._packNodes()._notify(), !u.samePos(t, r);
  }
  getRow() {
    return this.nodes.reduce((t, i) => Math.max(t, i.y + i.h), 0);
  }
  beginUpdate(t) {
    return t._updating || (t._updating = !0, delete t._skipDown, this.batchMode || this.saveInitial()), this;
  }
  endUpdate() {
    const t = this.nodes.find((i) => i._updating);
    return t && (delete t._updating, delete t._skipDown), this;
  }
  /** saves a copy of the largest column layout (eg 12 even when rendering 1 column) so we don't loose orig layout, unless explicity column
   * count to use is given. returning a list of widgets for serialization
   * @param saveElement if true (default), the element will be saved to GridStackWidget.el field, else it will be removed.
   * @param saveCB callback for each node -> widget, so application can insert additional data to be saved into the widget data structure.
   * @param column if provided, the grid will be saved for the given column count (IFF we have matching internal saved layout, or current layout).
   * Note: nested grids will ALWAYS save the container w to match overall layouts (parent + child) to be consistent.
  */
  save(t = !0, i, s) {
    const h = this._layouts?.length || 0;
    let e;
    h && (s ? s !== this.column && (e = this._layouts[s]) : this.column !== h - 1 && (e = this._layouts[h - 1]));
    const r = [];
    return this.sortNodes(), this.nodes.forEach((f) => {
      const l = e?.find((a) => a._id === f._id), o = { ...f, ...l || {} };
      u.removeInternalForSave(o, !t), i && i(f, o), r.push(o);
    }), r;
  }
  /** @internal called whenever a node is added or moved - updates the cached layouts */
  layoutsNodesChange(t) {
    return !this._layouts || this._inColumnResize ? this : (this._layouts.forEach((i, s) => {
      if (!i || s === this.column)
        return this;
      if (s < this.column)
        this._layouts[s] = void 0;
      else {
        const h = s / this.column;
        t.forEach((e) => {
          if (!e._orig)
            return;
          const r = i.find((f) => f._id === e._id);
          r && (r.y >= 0 && e.y !== e._orig.y && (r.y += e.y - e._orig.y), e.x !== e._orig.x && (r.x = Math.round(e.x * h)), e.w !== e._orig.w && (r.w = Math.round(e.w * h)));
        });
      }
    }), this);
  }
  /**
   * @internal Called to scale the widget width & position up/down based on the column change.
   * Note we store previous layouts (especially original ones) to make it possible to go
   * from say 12 -> 1 -> 12 and get back to where we were.
   *
   * @param prevColumn previous number of columns
   * @param column  new column number
   * @param layout specify the type of re-layout that will happen (position, size, etc...).
   * Note: items will never be outside of the current column boundaries. default (moveScale). Ignored for 1 column
   */
  columnChanged(t, i, s = "moveScale") {
    if (!this.nodes.length || !i || t === i)
      return this;
    const h = s === "compact" || s === "list";
    h && this.sortNodes(1), i < t && this.cacheLayout(this.nodes, t), this.batchUpdate();
    let e = [], r = h ? this.nodes : u.sort(this.nodes, -1);
    if (i > t && this._layouts) {
      const f = this._layouts[i] || [], l = this._layouts.length - 1;
      !f.length && t !== l && this._layouts[l]?.length && (t = l, this._layouts[l].forEach((o) => {
        const a = r.find((c) => c._id === o._id);
        a && (!h && !o.autoPosition && (a.x = o.x ?? a.x, a.y = o.y ?? a.y), a.w = o.w ?? a.w, (o.x == null || o.y === void 0) && (a.autoPosition = !0));
      })), f.forEach((o) => {
        const a = r.findIndex((c) => c._id === o._id);
        if (a !== -1) {
          const c = r[a];
          if (h) {
            c.w = o.w;
            return;
          }
          (o.autoPosition || isNaN(o.x) || isNaN(o.y)) && this.findEmptyPosition(o, e), o.autoPosition || (c.x = o.x ?? c.x, c.y = o.y ?? c.y, c.w = o.w ?? c.w, e.push(c)), r.splice(a, 1);
        }
      });
    }
    if (h)
      this.compact(s, !1);
    else {
      if (r.length)
        if (typeof s == "function")
          s(i, t, e, r);
        else {
          const f = h || s === "none" ? 1 : i / t, l = s === "move" || s === "moveScale", o = s === "scale" || s === "moveScale";
          r.forEach((a) => {
            a.x = i === 1 ? 0 : l ? Math.round(a.x * f) : Math.min(a.x, i - 1), a.w = i === 1 || t === 1 ? 1 : o ? Math.round(a.w * f) || 1 : Math.min(a.w, i), e.push(a);
          }), r = [];
        }
      e = u.sort(e, -1), this._inColumnResize = !0, this.nodes = [], e.forEach((f) => {
        this.addNode(f, !1), delete f._orig;
      });
    }
    return this.nodes.forEach((f) => delete f._orig), this.batchUpdate(!1, !h), delete this._inColumnResize, this;
  }
  /**
   * call to cache the given layout internally to the given location so we can restore back when column changes size
   * @param nodes list of nodes
   * @param column corresponding column index to save it under
   * @param clear if true, will force other caches to be removed (default false)
   */
  cacheLayout(t, i, s = !1) {
    const h = [];
    return t.forEach((e, r) => {
      if (e._id === void 0) {
        const f = e.id ? this.nodes.find((l) => l.id === e.id) : void 0;
        e._id = f?._id ?? y._idSeq++;
      }
      h[r] = { x: e.x, y: e.y, w: e.w, _id: e._id };
    }), this._layouts = s ? [] : this._layouts || [], this._layouts[i] = h, this;
  }
  /**
   * call to cache the given node layout internally to the given location so we can restore back when column changes size
   * @param node single node to cache
   * @param column corresponding column index to save it under
   */
  cacheOneLayout(t, i) {
    t._id = t._id ?? y._idSeq++;
    const s = { x: t.x, y: t.y, w: t.w, _id: t._id };
    (t.autoPosition || t.x === void 0) && (delete s.x, delete s.y, t.autoPosition && (s.autoPosition = !0)), this._layouts = this._layouts || [], this._layouts[i] = this._layouts[i] || [];
    const h = this.findCacheLayout(t, i);
    return h === -1 ? this._layouts[i].push(s) : this._layouts[i][h] = s, this;
  }
  findCacheLayout(t, i) {
    return this._layouts?.[i]?.findIndex((s) => s._id === t._id) ?? -1;
  }
  removeNodeFromLayoutCache(t) {
    if (this._layouts)
      for (let i = 0; i < this._layouts.length; i++) {
        const s = this.findCacheLayout(t, i);
        s !== -1 && this._layouts[i].splice(s, 1);
      }
  }
  /** called to remove all internal values but the _id */
  cleanupNode(t) {
    for (const i in t)
      i[0] === "_" && i !== "_id" && delete t[i];
    return this;
  }
}
y._idSeq = 0;
export {
  y as GridStackEngine
};
