// bind-event-listener@3.0.0/dist/bind-all.js
import { __exports as t } from "../../../../../../_virtual/bind-all.js";
import { __require as b } from "./bind.js";
var a;
function g() {
  if (a) return t;
  a = 1;
  var i = t && t.__assign || function() {
    return i = Object.assign || function(n) {
      for (var r, e = 1, o = arguments.length; e < o; e++) {
        r = arguments[e];
        for (var u in r) Object.prototype.hasOwnProperty.call(r, u) && (n[u] = r[u]);
      }
      return n;
    }, i.apply(this, arguments);
  };
  Object.defineProperty(t, "__esModule", { value: !0 }), t.bindAll = void 0;
  var l = /* @__PURE__ */ b();
  function d(n) {
    if (!(typeof n > "u"))
      return typeof n == "boolean" ? {
        capture: n
      } : n;
  }
  function s(n, r) {
    if (r == null)
      return n;
    var e = i(i({}, n), { options: i(i({}, d(r)), d(n.options)) });
    return e;
  }
  function c(n, r, e) {
    var o = r.map(function(u) {
      var f = s(u, e);
      return (0, l.bind)(n, f);
    });
    return function() {
      o.forEach(function(f) {
        return f();
      });
    };
  }
  return t.bindAll = c, t;
}
export {
  g as __require
};
